From 53e1a57c72953de1bfcc94c3257c724e7fa1f88f Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/block/entity/ShulkerBoxBlockEntity.java b/net/minecraft/world/level/block/entity/ShulkerBoxBlockEntity.java
index 5f6078e4cc06c059c01c7a20421f2f3a364a058b..286335cb3d8a5ce48a48357b014db1781004c2bd 100644
--- a/net/minecraft/world/level/block/entity/ShulkerBoxBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/ShulkerBoxBlockEntity.java
@@ -32,7 +32,7 @@ import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 import org.jspecify.annotations.Nullable;
 
-public class ShulkerBoxBlockEntity extends RandomizableContainerBlockEntity implements WorldlyContainer {
+public class ShulkerBoxBlockEntity extends RandomizableContainerBlockEntity implements WorldlyContainer, io.canvasmc.canvas.world.block.SleepingBlockEntity { // Canvas - block entity sleeping
     public static final int COLUMNS = 9;
     public static final int ROWS = 3;
     public static final int CONTAINER_SIZE = 27;
@@ -48,6 +48,30 @@ public class ShulkerBoxBlockEntity extends RandomizableContainerBlockEntity impl
     private float progress;
     private float progressOld;
     private final @Nullable DyeColor color;
+    // Canvas start - block entity sleeping
+    private net.minecraft.world.level.chunk.LevelChunk.@Nullable RebindableTickingBlockEntityWrapper tickWrapper = null;
+    private @Nullable TickingBlockEntity sleepingTicker = null;
+
+    @Override
+    public net.minecraft.world.level.chunk.LevelChunk.@Nullable RebindableTickingBlockEntityWrapper lithium$getTickWrapper() {
+        return this.tickWrapper;
+    }
+
+    @Override
+    public void lithium$setTickWrapper(final net.minecraft.world.level.chunk.LevelChunk.@Nullable RebindableTickingBlockEntityWrapper tickWrapper) {
+        this.tickWrapper = tickWrapper;
+    }
+
+    @Override
+    public @Nullable TickingBlockEntity lithium$getSleepingTicker() {
+        return this.sleepingTicker;
+    }
+
+    @Override
+    public void lithium$setSleepingTicker(final @Nullable TickingBlockEntity sleepingTicker) {
+        this.sleepingTicker = sleepingTicker;
+    }
+    // Canvas end - block entity sleeping
 
     // CraftBukkit start - add fields and methods
     public List<org.bukkit.entity.HumanEntity> transaction = new java.util.ArrayList<>();
@@ -134,6 +158,11 @@ public class ShulkerBoxBlockEntity extends RandomizableContainerBlockEntity impl
                     doNeighborUpdates(level, pos, blockState);
                 }
         }
+        // Canvas start - block entity sleeping
+        if (this.animationStatus == ShulkerBoxBlockEntity.AnimationStatus.CLOSED && this.progressOld == 0.0f && this.progress == 0.0f) {
+            this.lithium$startSleeping();
+        }
+        // Canvas end - block entity sleeping
     }
 
     public ShulkerBoxBlockEntity.AnimationStatus getAnimationStatus() {
@@ -174,6 +203,11 @@ public class ShulkerBoxBlockEntity extends RandomizableContainerBlockEntity impl
 
     @Override
     public boolean triggerEvent(final int b0, final int b1) {
+        // Canvas start - block entity sleeping
+        if (this.sleepingTicker != null) {
+            this.wakeUpNow();
+        }
+        // Canvas end - block entity sleeping
         if (b0 == 1) {
             this.openCount = b1;
             if (b1 == 0) {
