--- a/net/minecraft/world/entity/animal/AbstractFish.java
+++ b/net/minecraft/world/entity/animal/AbstractFish.java
@@ -173,7 +_,7 @@
 
         @Override
         public void tick() {
-            if (this.fish.isEyeInFluid(FluidTags.WATER)) {
+            if (this.fish.canvas$isEyeInWater()) { // Canvas - optimize fluidOnEyes
                 this.fish.setDeltaMovement(this.fish.getDeltaMovement().add(0.0, 0.005, 0.0));
             }
 
