--- a/net/minecraft/world/entity/animal/HappyGhast.java
+++ b/net/minecraft/world/entity/animal/HappyGhast.java
@@ -543,7 +_,7 @@
             boundingBox.maxZ + 1.0
         );
 
-        for (Player player : this.level().players()) {
+        for (Player player : this.level().getLocalPlayers()) { // Canvas - region threading
             if (!player.isSpectator()) {
                 Entity rootVehicle = player.getRootVehicle();
                 if (!(rootVehicle instanceof HappyGhast) && aabb.contains(rootVehicle.position())) {
