From 53e1a57c72953de1bfcc94c3257c724e7fa1f88f Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/BaseSpawner.java b/net/minecraft/world/level/BaseSpawner.java
index 4657083b6c32bd72604aab122b1f44deeb4ac0a4..68dbc0a791b73688f99047e2e4ffa526e0aa2d5b 100644
--- a/net/minecraft/world/level/BaseSpawner.java
+++ b/net/minecraft/world/level/BaseSpawner.java
@@ -45,14 +45,77 @@ public abstract class BaseSpawner {
     public @Nullable SpawnData nextSpawnData;
     private double spin;
     private double oSpin;
-    public int minSpawnDelay = 200;
-    public int maxSpawnDelay = 800;
-    public int spawnCount = 4;
+    // Canvas start - spawner configs
+    public int minSpawnDelay = getMinSpawnDelayDefault();
+    public int maxSpawnDelay = getMaxSpawnDelayDefault();
+    public int spawnCount = getSpawnCountDefault();
+    // Canvas end - spawner configs
     private @Nullable Entity displayEntity;
-    public int maxNearbyEntities = 6;
-    public int requiredPlayerRange = 16;
-    public int spawnRange = 4;
+    // Canvas start - spawner configs
+    public int maxNearbyEntities = getMaxNearbyEntitiesDefault();
+    public int requiredPlayerRange = getRequiredPlayerRangeDefault();
+    public int spawnRange = getSpawnRangeDefault();
+    // Canvas end - spawner configs
     private int tickDelay = 0; // Paper - Configurable mob spawner tick rate
+    // Canvas start - spawner configs
+
+    // this is instantiated on the region that we are being placed at, so we are technically
+    // able to fetch the current region and pull the Canvas config from there
+
+    private int getMinSpawnDelayDefault() {
+        final io.papermc.paper.threadedregions.RegionizedWorldData worldData = io.papermc.paper.threadedregions.TickRegionScheduler.getCurrentRegionizedWorldData();
+        if (worldData == null) {
+            // return vanilla default
+            return 200;
+        }
+        return worldData.world.canvasConfig().blocks.spawner.minSpawnDelay;
+    }
+
+    private int getMaxSpawnDelayDefault() {
+        final io.papermc.paper.threadedregions.RegionizedWorldData worldData = io.papermc.paper.threadedregions.TickRegionScheduler.getCurrentRegionizedWorldData();
+        if (worldData == null) {
+            // return vanilla default
+            return 800;
+        }
+        return worldData.world.canvasConfig().blocks.spawner.maxSpawnDelay;
+    }
+
+    private int getSpawnCountDefault() {
+        final io.papermc.paper.threadedregions.RegionizedWorldData worldData = io.papermc.paper.threadedregions.TickRegionScheduler.getCurrentRegionizedWorldData();
+        if (worldData == null) {
+            // return vanilla default
+            return 4;
+        }
+        return worldData.world.canvasConfig().blocks.spawner.spawnCount;
+    }
+
+    private int getMaxNearbyEntitiesDefault() {
+        final io.papermc.paper.threadedregions.RegionizedWorldData worldData = io.papermc.paper.threadedregions.TickRegionScheduler.getCurrentRegionizedWorldData();
+        if (worldData == null) {
+            // return vanilla default
+            return 6;
+        }
+        return worldData.world.canvasConfig().blocks.spawner.maxNearbyEntities;
+    }
+
+    private int getRequiredPlayerRangeDefault() {
+        final io.papermc.paper.threadedregions.RegionizedWorldData worldData = io.papermc.paper.threadedregions.TickRegionScheduler.getCurrentRegionizedWorldData();
+        if (worldData == null) {
+            // return vanilla default
+            return 16;
+        }
+        return worldData.world.canvasConfig().blocks.spawner.requiredPlayerRange;
+    }
+
+    private int getSpawnRangeDefault() {
+        final io.papermc.paper.threadedregions.RegionizedWorldData worldData = io.papermc.paper.threadedregions.TickRegionScheduler.getCurrentRegionizedWorldData();
+        if (worldData == null) {
+            // return vanilla default
+            return 4;
+        }
+        return worldData.world.canvasConfig().blocks.spawner.spawnRange;
+    }
+    // Canvas end - spawner configs
 
     public void setEntityId(final EntityType<?> type, final @Nullable Level level, final RandomSource random, final BlockPos pos) {
         this.getOrCreateNextSpawnData(level, random, pos).getEntityToSpawn().putString("id", BuiltInRegistries.ENTITY_TYPE.getKey(type).toString());
@@ -60,7 +123,7 @@ public abstract class BaseSpawner {
     }
 
     public boolean isNearPlayer(final Level level, final BlockPos pos) {
-        return level.hasNearbyAlivePlayerThatAffectsSpawning(pos.getX() + 0.5, pos.getY() + 0.5, pos.getZ() + 0.5, this.requiredPlayerRange); // Paper - Affects Spawning API
+        return level.canvas$hasNearbyAlivePlayerThatAffectsSpawningForSpawner(pos.getX() + 0.5, pos.getY() + 0.5, pos.getZ() + 0.5, this.requiredPlayerRange); // Paper - Affects Spawning API // Canvas - optimize spawners
     }
 
     public void clientTick(final Level level, final BlockPos pos) {
@@ -157,7 +220,7 @@ public abstract class BaseSpawner {
                                 return;
                             }
 
-                            int nearBy = level.getEntities(
+                            int nearBy = level.canvasConfig().blocks.spawner.disableMaxNearbyEntitiesCheck ? Integer.MIN_VALUE : level.getEntities( // Canvas - spawner configs
                                     EntityTypeTest.forExactClass(entity.getClass()),
                                     new AABB(pos.getX(), pos.getY(), pos.getZ(), pos.getX() + 1, pos.getY() + 1, pos.getZ() + 1).inflate(this.spawnRange),
                                     EntitySelector.NO_SPECTATORS
@@ -187,6 +250,7 @@ public abstract class BaseSpawner {
                                     mob.aware = false;
                                 }
                                 // Spigot end
+                                mob.canvas$fromSpawner = true; // Canvas - spawner configs
                             }
 
                             // Paper start
@@ -240,13 +304,15 @@ public abstract class BaseSpawner {
         this.spawnPotentials = input.read("SpawnPotentials", SpawnData.LIST_CODEC)
             .orElseGet(() -> WeightedList.of(this.nextSpawnData != null ? this.nextSpawnData : new SpawnData()));
         // Paper start - use int if set
-        this.minSpawnDelay = input.getIntOr("Paper.MinSpawnDelay", input.getIntOr("MinSpawnDelay", 200));
-        this.maxSpawnDelay = input.getIntOr("Paper.MaxSpawnDelay", input.getIntOr("MaxSpawnDelay", 800));
+        this.minSpawnDelay = input.getIntOr("Paper.MinSpawnDelay", input.getIntOr("MinSpawnDelay", getMinSpawnDelayDefault())); // Canvas - spawner configs
+        this.maxSpawnDelay = input.getIntOr("Paper.MaxSpawnDelay", input.getIntOr("MaxSpawnDelay", getMaxSpawnDelayDefault())); // Canvas - spawner configs
         // Paper end - use int if set
-        this.spawnCount = input.getIntOr("SpawnCount", 4);
-        this.maxNearbyEntities = input.getIntOr("MaxNearbyEntities", 6);
-        this.requiredPlayerRange = input.getIntOr("RequiredPlayerRange", 16);
-        this.spawnRange = input.getIntOr("SpawnRange", 4);
+        // Canvas start - spawner configs
+        this.spawnCount = input.getIntOr("SpawnCount", getSpawnCountDefault());
+        this.maxNearbyEntities = input.getIntOr("MaxNearbyEntities", getMaxNearbyEntitiesDefault());
+        this.requiredPlayerRange = input.getIntOr("RequiredPlayerRange", getRequiredPlayerRangeDefault());
+        this.spawnRange = input.getIntOr("SpawnRange", getSpawnRangeDefault());
+        // Canvas end - spawner configs
         this.displayEntity = null;
     }
 
