From 53e1a57c72953de1bfcc94c3257c724e7fa1f88f Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/NaturalSpawner.java b/net/minecraft/world/level/NaturalSpawner.java
index 2e76d0870aebcb605ca7f990560472c33951cdf1..9f26a61a5dd1f0ebeb5627745fa92f01314a9b0b 100644
--- a/net/minecraft/world/level/NaturalSpawner.java
+++ b/net/minecraft/world/level/NaturalSpawner.java
@@ -69,24 +69,49 @@ public final class NaturalSpawner {
         final LocalMobCapCalculator localMobCapCalculator
     ) {
         // Paper start - Optional per player mob spawns
-        return createState(spawnableChunkCount, entities, chunkGetter, localMobCapCalculator, false);
+        return createState(spawnableChunkCount, localMobCapCalculator, false, io.papermc.paper.threadedregions.TickRegionScheduler.getCurrentRegionizedWorldData()); // Canvas - optimize natural spawning
     }
 
+    // Canvas start - optimize natural spawning
+    private static int[] calculateLimits(final io.papermc.paper.threadedregions.RegionizedWorldData worldData, final int spawnableChunkCount) {
+        int[] retVal = new int[SPAWNING_CATEGORIES.length];
+        ServerLevel world = worldData.world;
+        for (MobCategory category : MobCategory.values()) {
+            if (category == MobCategory.MISC) continue;
+            // normally we would check if the category is valid for spawn limits, however the SPAWNING_CATEGORIES
+            // array filters this for us already, so we don't need to do an extra call
+            int rawLimit = world.getWorld().getSpawnLimit(org.bukkit.craftbukkit.util.CraftSpawnCategory.toBukkit(category));
+            // we need to apply the same ratio calculation as canSpawnForCategoryGlobal
+            retVal[category.ordinal()] = rawLimit * spawnableChunkCount / NaturalSpawner.MAGIC_NUMBER;
+        }
+        return retVal;
+    }
+
+    // Canvas end - optimize natural spawning
     public static NaturalSpawner.SpawnState createState(
         final int spawnableChunkCount,
-        final Iterable<Entity> entities,
-        final NaturalSpawner.ChunkGetter chunkGetter,
         final LocalMobCapCalculator localMobCapCalculator,
-        final boolean countMobs
+        final boolean countMobs, final io.papermc.paper.threadedregions.RegionizedWorldData worldData // Canvas - optimize natural spawning
     ) {
         // Paper end - Optional per player mob spawns
         PotentialCalculator spawnPotential = new PotentialCalculator();
+        int[] limits = calculateLimits(worldData, spawnableChunkCount); // Canvas - optimize natural spawning
         Object2IntOpenHashMap<MobCategory> mobCounts = new Object2IntOpenHashMap<>();
 
-        for (Entity entity : entities) {
-            if (!(entity instanceof Mob mob && (mob.isPersistenceRequired() || mob.requiresCustomPersistence()))) {
+        // Canvas start - optimize natural spawning
+        final boolean perPlayerMobSpawning = worldData.world.paperConfig().entities.spawning.perPlayerMobSpawns;
+        // optimized iteration path, go by chunks
+        for (LevelChunk chunk : worldData.getChunks().getRawDataUnchecked()) {
+            if (chunk == null) continue; // unlikely, but check to be sure
+            for (Entity entity : chunk.moonrise$getChunkHolder().getEntityChunk().getEntityArr()) {
+                if (entity == null) continue; // very likely, must check
                 MobCategory category = entity.getType().getCategory();
-                if (category != MobCategory.MISC) {
+                // if we skip because we hit the limit, it won't change the outcome, since if it skips because
+                // of the limit hit, the category will be filtered anyway later
+                if (category == MobCategory.MISC) continue;
+                final boolean shouldSkipBecauseOfLimit = limits[category.ordinal()] <= mobCounts.getInt(category);
+                if (!perPlayerMobSpawning && shouldSkipBecauseOfLimit) continue;
+                if (!(entity instanceof Mob mob && (mob.isPersistenceRequired() || mob.requiresCustomPersistence()))) {
                     // Paper start - Only count natural spawns
                     if (!entity.level().paperConfig().entities.spawning.countAllMobsForSpawning &&
                         !(entity.spawnReason == org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL ||
@@ -95,28 +120,44 @@ public final class NaturalSpawner {
                     }
                     // Paper end - Only count natural spawns
                     BlockPos pos = entity.blockPosition();
-                    chunkGetter.query(ChunkPos.pack(pos), chunk -> {
+
+                    // getting the rough biome is expensive, if we are skipping because of
+                    // the limit, why tf are we gonna do this, since there is no impact by skipping this anyway
+
+                    // the only real reason we don't skip the full thing with per-player mob spawns is because
+                    // per-player mob spawns require 'updatePlayerMobTypeMap' to be called for accurate spawning
+
+                    // as a result, if we are doing per-player spawning, we try and skip as much as possible to
+                    // reduce expensive CPU time without costing vanilla mob spawning functionality
+
+                    // Note: this should be safe, this only effects the rate of spawning, and specifically for a
+                    // few mobs for a few nether biomes, so farms shouldn't be impacted
+                    if (!perPlayerMobSpawning || !shouldSkipBecauseOfLimit) {
+                        // only do this lookup if we aren't using per-player spawns(since we skip above), and
+                        // if we are doing per-player mob spawning and not over the limit yet
                         MobSpawnSettings.MobSpawnCost mobSpawnCost = getRoughBiome(pos, chunk).getMobSettings().getMobSpawnCost(entity.getType());
                         if (mobSpawnCost != null) {
                             spawnPotential.addCharge(entity.blockPosition(), mobSpawnCost.charge());
                         }
+                    }
 
-                        if (localMobCapCalculator != null && entity instanceof Mob) { // Paper - Optional per player mob spawns
-                            localMobCapCalculator.addMob(chunk.getPos(), category);
-                        }
+                    // only runs if we are not doing per-player
+                    if (localMobCapCalculator != null && entity instanceof Mob) {
+                        localMobCapCalculator.addMob(chunk.getPos(), category);
+                    }
 
-                        mobCounts.addTo(category, 1);
-                        // Paper start - Optional per player mob spawns
-                        if (countMobs) {
-                            chunk.level.getChunkSource().chunkMap.updatePlayerMobTypeMap(entity);
-                        }
-                        // Paper end - Optional per player mob spawns
-                    });
+                    mobCounts.addTo(category, 1);
+                    // Paper start - Optional per player mob spawns
+                    if (countMobs) {
+                        chunk.level.getChunkSource().chunkMap.updatePlayerMobTypeMap(entity);
+                    }
+                    // Paper end - Optional per player mob spawns
                 }
             }
         }
+        // Canvas end - optimize natural spawning
 
-        return new NaturalSpawner.SpawnState(spawnableChunkCount, mobCounts, spawnPotential, localMobCapCalculator);
+        return new NaturalSpawner.SpawnState(spawnableChunkCount, mobCounts, spawnPotential, localMobCapCalculator, worldData, limits); // Canvas - optimize natural spawning
     }
 
     private static Biome getRoughBiome(final BlockPos pos, final ChunkAccess chunk) {
@@ -160,6 +201,11 @@ public final class NaturalSpawner {
     public static void spawnForChunk(
         final ServerLevel level, final LevelChunk chunk, final NaturalSpawner.SpawnState state, final List<MobCategory> spawningCategories
     ) {
+        // Canvas start - optimize natural spawning
+        final ca.spottedleaf.moonrise.common.list.ReferenceList<net.minecraft.server.level.ServerPlayer> inRange =
+            level.moonrise$getNearbyPlayers().getPlayers(chunk.getPos(), ca.spottedleaf.moonrise.common.misc.NearbyPlayers.NearbyMapType.TICK_VIEW_DISTANCE);
+        net.minecraft.server.level.ServerPlayer closestPlayer = null;
+        // Canvas end - optimize natural spawning
         for (MobCategory mobCategory : spawningCategories) {
             // Paper start - Optional per player mob spawns
             final boolean canSpawn;
@@ -174,12 +220,17 @@ public final class NaturalSpawner {
 
                 // Apply per-player limit
                 int minDiff = Integer.MAX_VALUE;
-                final ca.spottedleaf.moonrise.common.list.ReferenceList<net.minecraft.server.level.ServerPlayer> inRange =
-                    level.moonrise$getNearbyPlayers().getPlayers(chunk.getPos(), ca.spottedleaf.moonrise.common.misc.NearbyPlayers.NearbyMapType.TICK_VIEW_DISTANCE);
+                // Canvas - move up
                 if (inRange != null) {
                     final net.minecraft.server.level.ServerPlayer[] backingSet = inRange.getRawDataUnchecked();
                     for (int k = 0, len = inRange.size(); k < len; k++) {
-                        minDiff = Math.min(limit - level.getChunkSource().chunkMap.getMobCountNear(backingSet[k], mobCategory), minDiff);
+                        // Canvas start - optimize natural spawning
+                        int diff = limit - level.getChunkSource().chunkMap.getMobCountNear(backingSet[k], mobCategory);
+                        if (diff < minDiff) {
+                            minDiff = diff;
+                            closestPlayer = backingSet[k];
+                        }
+                        // Canvas end - optimize natural spawning
                     }
                 }
 
@@ -189,8 +240,10 @@ public final class NaturalSpawner {
                 canSpawn = state.canSpawnForCategoryLocal(mobCategory, chunk.getPos());
             }
             if (canSpawn) {
-                spawnCategoryForChunk(mobCategory, level, chunk, state::canSpawn, state::afterSpawn,
-                    maxSpawns, level.paperConfig().entities.spawning.perPlayerMobSpawns ? level.getChunkSource().chunkMap::updatePlayerMobTypeMap : null);
+                // Canvas start - optimize natural spawning
+                spawnCategoryForChunk(mobCategory, state, chunk, state::canSpawn, state::afterSpawn,
+                    maxSpawns, level.paperConfig().entities.spawning.perPlayerMobSpawns ? level.getChunkSource().chunkMap::updatePlayerMobTypeMap : null, closestPlayer);
+                // Canvas end - optimize natural spawning
                 // Paper end - Optional per player mob spawns
             }
         }
@@ -206,31 +259,34 @@ public final class NaturalSpawner {
     }
     // Paper end - Add mobcaps commands
 
+    // Canvas start - optimize natural spawning
     public static void spawnCategoryForChunk(
         final MobCategory mobCategory,
-        final ServerLevel level,
-        final LevelChunk chunk,
-        final NaturalSpawner.SpawnPredicate extraTest,
-        final NaturalSpawner.AfterSpawnCallback spawnCallback
-    ) {
-        // Paper start - Optional per player mob spawns
-        spawnCategoryForChunk(mobCategory, level, chunk, extraTest, spawnCallback, Integer.MAX_VALUE, null);
-    }
-    public static void spawnCategoryForChunk(
-        final MobCategory mobCategory,
-        final ServerLevel level,
+        final SpawnState spawnState,
         final LevelChunk chunk,
         final NaturalSpawner.SpawnPredicate extraTest,
         final NaturalSpawner.AfterSpawnCallback spawnCallback,
         final int maxSpawns,
-        final Consumer<Entity> trackEntity
+        final Consumer<Entity> trackEntity, final Player closestPlayer
     ) {
-        // Paper end - Optional per player mob spawns
-        BlockPos start = getRandomPosWithin(level, chunk);
-        if (start.getY() >= level.getMinY() + 1) {
-            spawnCategoryForPosition(mobCategory, level, chunk, start, extraTest, spawnCallback, maxSpawns, trackEntity); // Paper - Optional per player mob spawns
+        final io.papermc.paper.threadedregions.RegionizedWorldData regionizedWorldData = spawnState.regionizedWorldData;
+        final ServerLevel world = regionizedWorldData.world;
+
+        final ChunkPos pos = chunk.getPos();
+        final RandomSource random = regionizedWorldData.simpleUnsafeLocalRandom;
+
+        // using the min bounds and then adding an RNG with bound of 16 means it won't escape this chunk
+        int randomX = pos.getMinBlockX() + random.nextInt(16);
+        int randomZ = pos.getMinBlockZ() + random.nextInt(16);
+        int surfaceY = chunk.getHeight(Heightmap.Types.WORLD_SURFACE, randomX, randomZ) + 1;
+        int randomY = Mth.randomBetweenInclusive(random, world.getMinY(), surfaceY);
+
+        BlockPos start = new BlockPos(randomX, randomY, randomZ);
+        if (start.getY() >= world.getMinY() + 1) {
+            spawnCategoryForPosition(mobCategory, spawnState, chunk, start, extraTest, spawnCallback, maxSpawns, trackEntity, closestPlayer); // Paper - Optional per player mob spawns
         }
     }
+    // Canvas end - optimize natural spawning
 
     @VisibleForDebug
     public static void spawnCategoryForPosition(final MobCategory mobCategory, final ServerLevel level, final BlockPos start) {
@@ -245,61 +301,94 @@ public final class NaturalSpawner {
         final NaturalSpawner.SpawnPredicate extraTest,
         final NaturalSpawner.AfterSpawnCallback spawnCallback
     ) {
-        // Paper start - Optional per player mob spawns
-        spawnCategoryForPosition(mobCategory, level, chunk, start, extraTest, spawnCallback, Integer.MAX_VALUE, null);
+        throw new UnsupportedOperationException(); // Canvas - optimize natural spawning
+    }
+    // Canvas start - optimize natural spawning
+
+    private static int clamp(final int val, final int min, final int max) {
+        return Math.max(min, Math.min(val, max));
     }
+
     public static void spawnCategoryForPosition(
         final MobCategory mobCategory,
-        final ServerLevel level,
-        final ChunkAccess chunk,
+        final SpawnState spawnState, // Canvas - use state
+        final LevelChunk chunk, // Canvas - pass as LevelChunk
         final BlockPos start,
         final NaturalSpawner.SpawnPredicate extraTest,
         final NaturalSpawner.AfterSpawnCallback spawnCallback,
         final int maxSpawns,
-        final @Nullable Consumer<Entity> trackEntity
+        final @Nullable Consumer<Entity> trackEntity, final Player closestPlayer // Canvas - pass player
     ) {
         // Paper end - Optional per player mob spawns
+        final io.papermc.paper.threadedregions.RegionizedWorldData regionizedWorldData = spawnState.regionizedWorldData;
+        final ServerLevel level = regionizedWorldData.world;
+        final RandomSource fastRandom = regionizedWorldData.simpleUnsafeLocalRandom;
+        final net.minecraft.world.level.border.WorldBorder border = level.getWorldBorder();
+
+        final int minX = chunk.getPos().getMinBlockX();
+        final int minZ = chunk.getPos().getMinBlockZ();
+        final int maxX = chunk.getPos().getMaxBlockX();
+        final int maxZ = chunk.getPos().getMaxBlockZ();
+
         StructureManager structureManager = level.structureManager();
         ChunkGenerator generator = level.getChunkSource().getGenerator();
         int yStart = start.getY();
-        BlockState state = level.getBlockStateIfLoadedAndInBounds(start); // Paper - don't load chunks for mob spawn
-        if (state != null && !state.isRedstoneConductor(chunk, start)) { // Paper - don't load chunks for mob spawn
+        // the pos is literally in bounds of this chunk, don't fetch the chunk again
+        BlockState state = chunk.getBlockStateFinal(start.getX(), start.getY(), start.getZ());
+        if (!state.isRedstoneConductor(chunk, start)) { // Paper - don't load chunks for mob spawn
+    // Canvas end - optimize natural spawning
             BlockPos.MutableBlockPos pos = new BlockPos.MutableBlockPos();
             int clusterSize = 0;
 
             for (int groupCount = 0; groupCount < 3; groupCount++) {
-                int x = start.getX();
-                int z = start.getZ();
+                int centerX = start.getX(); // Canvas - rename
+                int centerZ = start.getZ(); // Canvas - rename
                 int ss = 6;
                 MobSpawnSettings.SpawnerData currentSpawnData = null;
                 SpawnGroupData groupData = null;
-                int max = Mth.ceil(level.random.nextFloat() * 4.0F);
+                int max = Mth.ceil(fastRandom.nextFloat() * 4.0F); // Canvas - use faster random
                 int groupSize = 0;
 
                 for (int ll = 0; ll < max; ll++) {
-                    x += level.random.nextInt(6) - level.random.nextInt(6);
-                    z += level.random.nextInt(6) - level.random.nextInt(6);
+                    // Canvas start - optimize natural spawning
+                    // this cuts random calls in half, which is def a micro-opt, but
+                    // reduces possible random calls from 48 -> 24, while almost preserving
+                    // the vanilla spread logic. the vanilla logic uses triangular
+                    // distribution, and this one uses uniform distribution, meaning
+                    // mob clusters will spawn slightly more spread out, rather than clustered
+                    // more towards the center. barely noticeable, but worth mentioning
+                    long randomPair = fastRandom.nextLong();
+                    int x = centerX + (int)((randomPair & 0xFFFF) * 11L >> 16) - 5;
+                    int z = centerZ + (int)(((randomPair >>> 32) & 0xFFFF) * 11L >> 16) - 5;
+
+                    // clamp to the chunk bounds, so we can reduce chunk lookup calls since
+                    // all blocks will be within this chunk
+                    x = clamp(x, minX, maxX);
+                    z = clamp(z, minZ, maxZ);
+                    // Canvas end - optimize natural spawning
                     pos.set(x, yStart, z);
                     double xx = x + 0.5;
                     double zz = z + 0.5;
-                    Player nearestPlayer = level.getNearestPlayer(xx, yStart, zz, -1.0, false);
+                    // we run through an inRange set already, so why not try and use those results to help us out?
+                    // the biggest difference we would face here is like... 16 blocks max?
+                    Player nearestPlayer = closestPlayer == null ? level.getNearestPlayer(xx, yStart, zz, -1.0, false) : closestPlayer; // Canvas - optimize natural spawning
                     if (nearestPlayer != null) {
                         double nearestPlayerDistanceSqr = nearestPlayer.distanceToSqr(xx, yStart, zz);
-                        if (level.isLoadedAndInBounds(pos) && isRightDistanceToPlayerAndSpawnPoint(level, chunk, pos, nearestPlayerDistanceSqr)) { // Paper - don't load chunks for mob spawn
+                        if (border.isWithinBounds(pos) && isRightDistanceToPlayerAndSpawnPoint(level, chunk, pos, nearestPlayerDistanceSqr)) { // Paper - don't load chunks for mob spawn // Canvas - reduce chunk lookups
                             if (currentSpawnData == null) {
                                 Optional<MobSpawnSettings.SpawnerData> nextSpawnData = getRandomSpawnMobAt(
-                                    level, structureManager, generator, mobCategory, level.random, pos
+                                    chunk, structureManager, generator, mobCategory, fastRandom, pos // Canvas - pass chunk and fast random
                                 );
                                 if (nextSpawnData.isEmpty()) {
                                     break;
                                 }
 
                                 currentSpawnData = nextSpawnData.get();
-                                max = currentSpawnData.minCount() + level.random.nextInt(1 + currentSpawnData.maxCount() - currentSpawnData.minCount());
+                                max = currentSpawnData.minCount() + fastRandom.nextInt(1 + currentSpawnData.maxCount() - currentSpawnData.minCount()); // Canvas - optimize natural spawning
                             }
 
                             // Paper start - PreCreatureSpawnEvent
-                            PreSpawnStatus doSpawning = isValidSpawnPostitionForType(level, mobCategory, structureManager, generator, currentSpawnData, pos, nearestPlayerDistanceSqr);
+                            PreSpawnStatus doSpawning = isValidSpawnPostitionForType(level, mobCategory, structureManager, generator, currentSpawnData, pos, chunk, fastRandom, nearestPlayerDistanceSqr); // Canvas - pass chunk and random
                             // Paper start - per player mob count backoff
                             if (doSpawning == PreSpawnStatus.ABORT || doSpawning == PreSpawnStatus.CANCELLED) {
                                 level.getChunkSource().chunkMap.updateFailurePlayerMobTypeMap(pos.getX() >> 4, pos.getZ() >> 4, mobCategory);
@@ -316,7 +405,7 @@ public final class NaturalSpawner {
                                     return;
                                 }
 
-                                mob.snapTo(xx, yStart, zz, level.random.nextFloat() * 360.0F, 0.0F);
+                                mob.snapTo(xx, yStart, zz, fastRandom.nextFloat() * 360.0F, 0.0F); // Canvas - optimize natural spawning
                                 if (isValidPositionForMob(level, mob, nearestPlayerDistanceSqr)) {
                                     groupData = mob.finalizeSpawn(
                                         level, level.getCurrentDifficultyAt(mob.blockPosition()), EntitySpawnReason.NATURAL, groupData
@@ -362,8 +451,12 @@ public final class NaturalSpawner {
                 && respawnData.pos().closerToCenterThan(new Vec3(pos.getX() + 0.5, pos.getY(), pos.getZ() + 0.5), 24.0)) {
                 return false;
             } else {
-                ChunkPos chunkPos = ChunkPos.containing(pos);
-                return Objects.equals(chunkPos, chunk.getPos()) || level.canSpawnEntitiesInChunk(chunkPos);
+                // Canvas start - optimize natural spawning
+                // 'pos' is a mutable block pos clamped to the passed ChunkAccess, so this check is stupid
+                // ChunkPos chunkPos = ChunkPos.containing(pos);
+                // return Objects.equals(chunkPos, chunk.getPos()) || level.canSpawnEntitiesInChunk(chunkPos);
+                return true;
+                // Canvas end - optimize natural spawning
             }
         }
     }
@@ -383,6 +476,8 @@ public final class NaturalSpawner {
         final ChunkGenerator generator,
         final MobSpawnSettings.SpawnerData currentSpawnData,
         final BlockPos.MutableBlockPos pos,
+        final LevelChunk chunk, // Canvas - pass chunk
+        final RandomSource random, // Canvas - pass random
         final double nearestPlayerDistanceSqr
     ) {
         // Paper start - PreCreatureSpawnEvent
@@ -401,9 +496,9 @@ public final class NaturalSpawner {
         // Paper end - PreCreatureSpawnEvent
             && (type.canSpawnFarFromPlayer() || !(nearestPlayerDistanceSqr > type.getCategory().getDespawnDistance() * type.getCategory().getDespawnDistance()))
             && type.canSummon()
-            && canSpawnMobAt(level, structureManager, generator, mobCategory, currentSpawnData, pos)
+            && canSpawnMobAt(chunk, structureManager, generator, mobCategory, currentSpawnData, pos) // Canvas - pass chunk
             && SpawnPlacements.isSpawnPositionOk(type, level, pos)
-            && SpawnPlacements.checkSpawnRules(type, level, EntitySpawnReason.NATURAL, pos, level.random)
+            && SpawnPlacements.checkSpawnRules(type, level, EntitySpawnReason.NATURAL, pos, random) // Canvas - pass faster random
             && level.noCollision(type.getSpawnAABB(pos.getX() + 0.5, pos.getY(), pos.getZ() + 0.5));
         return success ? PreSpawnStatus.SUCCESS : PreSpawnStatus.FAIL; // Paper - PreCreatureSpawnEvent
     }
@@ -433,47 +528,47 @@ public final class NaturalSpawner {
     }
 
     private static Optional<MobSpawnSettings.SpawnerData> getRandomSpawnMobAt(
-        final ServerLevel level,
+        final LevelChunk chunk, // Canvas - pass chunk
         final StructureManager structureManager,
         final ChunkGenerator generator,
         final MobCategory mobCategory,
         final RandomSource random,
         final BlockPos pos
     ) {
-        Holder<Biome> biome = level.getBiome(pos);
+        Holder<Biome> biome = chunk.level.getBiome(pos); // Canvas - pass chunk
         return mobCategory == MobCategory.WATER_AMBIENT && biome.is(BiomeTags.REDUCED_WATER_AMBIENT_SPAWNS) && random.nextFloat() < 0.98F
             ? Optional.empty()
-            : mobsAt(level, structureManager, generator, mobCategory, pos, biome).getRandom(random);
+            : mobsAt(chunk, structureManager, generator, mobCategory, pos, biome).getRandom(random); // Canvas - pass chunk
     }
 
     private static boolean canSpawnMobAt(
-        final ServerLevel level,
+        final LevelChunk chunk, // Canvas - pass chunk
         final StructureManager structureManager,
         final ChunkGenerator generator,
         final MobCategory mobCategory,
         final MobSpawnSettings.SpawnerData spawnerData,
         final BlockPos pos
     ) {
-        return mobsAt(level, structureManager, generator, mobCategory, pos, null).contains(spawnerData);
+        return mobsAt(chunk, structureManager, generator, mobCategory, pos, null).contains(spawnerData); // Canvas - pass chunk
     }
 
     private static WeightedList<MobSpawnSettings.SpawnerData> mobsAt(
-        final ServerLevel level,
+        final LevelChunk chunk, // Canvas - pass chunk
         final StructureManager structureManager,
         final ChunkGenerator generator,
         final MobCategory mobCategory,
         final BlockPos pos,
         final @Nullable Holder<Biome> biome
     ) {
-        return isInNetherFortressBounds(pos, level, mobCategory, structureManager)
+        return isInNetherFortressBounds(pos, chunk, mobCategory, structureManager) // Canvas - pass chunk
             ? NetherFortressStructure.FORTRESS_ENEMIES
-            : generator.getMobsAt(biome != null ? biome : level.getBiome(pos), structureManager, mobCategory, pos);
+            : generator.getMobsAt(biome != null ? biome : chunk.level.getBiome(pos), structureManager, mobCategory, pos, chunk.getAllReferences()); // Canvas - reduce chunk lookup
     }
 
     public static boolean isInNetherFortressBounds(
-        final BlockPos pos, final ServerLevel level, final MobCategory category, final StructureManager structureManager
+        final BlockPos pos, final LevelChunk chunk, final MobCategory category, final StructureManager structureManager // Canvas - pass chunk
     ) {
-        if (category == MobCategory.MONSTER && level.getBlockState(pos.below()).is(Blocks.NETHER_BRICKS)) {
+        if (category == MobCategory.MONSTER && chunk.getBlockState(pos.below()).is(Blocks.NETHER_BRICKS)) { // Canvas - reduce chunk lookups
             Structure fortress = structureManager.registryAccess().lookupOrThrow(Registries.STRUCTURE).getValue(BuiltinStructures.FORTRESS);
             return fortress != null && structureManager.getStructureAt(pos, fortress).isValid();
         } else {
@@ -620,18 +715,22 @@ public final class NaturalSpawner {
         private @Nullable BlockPos lastCheckedPos;
         private @Nullable EntityType<?> lastCheckedType;
         private double lastCharge;
+        public final io.papermc.paper.threadedregions.RegionizedWorldData regionizedWorldData; // Canvas - optimize natural spawning
+        public final int[] limits; // Canvas - optimize natural spawning
 
         private SpawnState(
             final int spawnableChunkCount,
             final Object2IntOpenHashMap<MobCategory> mobCategoryCounts,
             final PotentialCalculator spawnPotential,
-            final LocalMobCapCalculator localMobCapCalculator
+            final LocalMobCapCalculator localMobCapCalculator, io.papermc.paper.threadedregions.RegionizedWorldData regionizedWorldData, int[] limits // Canvas - optimize natural spawning
         ) {
             this.spawnableChunkCount = spawnableChunkCount;
             this.mobCategoryCounts = mobCategoryCounts;
             this.spawnPotential = spawnPotential;
             this.localMobCapCalculator = localMobCapCalculator;
             this.unmodifiableMobCategoryCounts = Object2IntMaps.unmodifiable(mobCategoryCounts);
+            this.regionizedWorldData = regionizedWorldData; // Canvas - optimize natural spawning
+            this.limits = limits; // Canvas - optimize natural spawning
         }
 
         private boolean canSpawn(final EntityType<?> type, final BlockPos testPos, final ChunkAccess chunk) {
