--- a/net/minecraft/server/gui/MinecraftServerGui.java
+++ b/net/minecraft/server/gui/MinecraftServerGui.java
@@ -51,7 +_,7 @@
         } catch (Exception var3) {
         }
 
-        final JFrame jFrame = new JFrame("Purpur Minecraft server"); // Purpur - Improve GUI
+        final JFrame jFrame = new JFrame("Canvas Minecraft server"); // Purpur - Improve GUI // Canvas - Rebrand
         final MinecraftServerGui minecraftServerGui = new MinecraftServerGui(server);
         jFrame.setDefaultCloseOperation(2);
         jFrame.add(minecraftServerGui);
@@ -59,7 +_,7 @@
         jFrame.setLocationRelativeTo(null);
         jFrame.setVisible(true);
         // Paper start - Improve ServerGUI
-        jFrame.setName("Purpur Minecraft server"); // Purpur - Improve GUI
+        jFrame.setName("Canvas Minecraft server"); // Purpur - Improve GUI // Canvas - Rebrand
         try {
             jFrame.setIconImage(javax.imageio.ImageIO.read(java.util.Objects.requireNonNull(MinecraftServerGui.class.getClassLoader().getResourceAsStream("logo.png"))));
         } catch (java.io.IOException ignore) {
@@ -69,7 +_,7 @@
             @Override
             public void windowClosing(WindowEvent event) {
                 if (!minecraftServerGui.isClosing.getAndSet(true)) {
-                    jFrame.setTitle("Purpur Minecraft server - shutting down!"); // Purpur - Improve GUI
+                    jFrame.setTitle("Canvas Minecraft server - shutting down!"); // Purpur - Improve GUI // Canvas - Rebrand
                     server.halt(true);
                     minecraftServerGui.runFinalizers();
                 }
@@ -192,7 +_,7 @@
         }
     }
 
-    void runFinalizers() {
+    public void runFinalizers() { // Canvas - public
         this.finalizers.forEach(Runnable::run);
     }
 
