--- a/net/minecraft/gametest/framework/GameTestServer.java
+++ b/net/minecraft/gametest/framework/GameTestServer.java
@@ -337,6 +_,18 @@
         return false;
     }
 
+    // Canvas start - bump netty
+    @Override
+    public boolean isIoUringEnabled() {
+        return false;
+    }
+
+    @Override
+    public boolean isKqueueEnabled() {
+        return false;
+    }
+
+    // Canvas end - bump netty
     @Override
     public boolean isCommandBlockEnabled() {
         return true;
