--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftChestBoat.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftChestBoat.java
@@ -6,11 +_,11 @@
 import org.bukkit.inventory.Inventory;
 
 public abstract class CraftChestBoat extends CraftBoat implements org.bukkit.entity.ChestBoat, com.destroystokyo.paper.loottable.PaperLootableEntityInventory { // Paper
-    private final Inventory inventory;
+    // private final Inventory inventory; // Canvas - fix chest boat inventories on region switch/removal
 
     public CraftChestBoat(CraftServer server, AbstractChestBoat entity) {
         super(server, entity);
-        this.inventory = new CraftInventory(entity);
+        // this.inventory = new CraftInventory(entity); // Canvas - fix chest boat inventories on region switch/removal
     }
 
     @Override
@@ -21,6 +_,6 @@
 
     @Override
     public Inventory getInventory() {
-        return this.inventory;
+        return new CraftInventory(getHandle()); // Canvas - fix chest boat inventories on region switch/removal
     }
 }
