From 37c0802d8eb83234e7cd8fab3253d198d2d53b39 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas paperServer File Patches


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntityType.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntityType.java
index bf59ea6025abad4e3688e04316dbbc171c5343d2..eeaf036b579ffe2c646c60c397696cee609289d7 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntityType.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntityType.java
@@ -15,13 +15,22 @@ import org.bukkit.entity.EntityType;
 
 public class CraftEntityType {
 
+    private static final java.util.Map<net.minecraft.world.entity.EntityType<?>, EntityType> MINECRAFT_TO_BUKKIT_KEY_CACHE = new java.util.concurrent.ConcurrentHashMap<>(); // Canvas - cache minecraftToBukkit convert
     public static EntityType minecraftToBukkit(net.minecraft.world.entity.EntityType<?> minecraft) {
         Preconditions.checkArgument(minecraft != null);
 
+        // Canvas start - cache minecraftToBukkit convert
+        EntityType asBukkit = MINECRAFT_TO_BUKKIT_KEY_CACHE.get(minecraft);
+
+        if (io.canvasmc.canvas.Config.INSTANCE.enableCachedMTBEntityTypeConvert && asBukkit != null) {
+            return asBukkit;
+        }
+        // Canvas end - cache minecraftToBukkit convert
         net.minecraft.core.Registry<net.minecraft.world.entity.EntityType<?>> registry = CraftRegistry.getMinecraftRegistry(Registries.ENTITY_TYPE);
         EntityType bukkit = Registry.ENTITY_TYPE.get(CraftNamespacedKey.fromMinecraft(registry.getResourceKey(minecraft).orElseThrow().identifier()));
 
         Preconditions.checkArgument(bukkit != null);
+        MINECRAFT_TO_BUKKIT_KEY_CACHE.put(minecraft, bukkit); // Canvas - cache minecraftToBukkit convert
 
         return bukkit;
     }
