--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntityType.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntityType.java
@@ -15,11 +_,19 @@
 
 public class CraftEntityType {
 
+    private static final java.util.Map<net.minecraft.world.entity.EntityType<?>, EntityType> MINECRAFT_TO_BUKKIT_KEY_CACHE = new java.util.concurrent.ConcurrentHashMap<>(); // Canvas - cache minecraftToBukkit convert
     public static EntityType minecraftToBukkit(net.minecraft.world.entity.EntityType<?> minecraft) {
         Preconditions.checkArgument(minecraft != null);
 
+        // Canvas start - cache minecraftToBukkit convert
+        EntityType asBukkit = MINECRAFT_TO_BUKKIT_KEY_CACHE.get(minecraft);
+        if (io.canvasmc.canvas.GlobalConfiguration.getInstance().cacheMinecraft2BukkitEntityTypeConversion && asBukkit != null) {
+            return asBukkit;
+        }
+        // Canvas end - cache minecraftToBukkit conver
         EntityType bukkit = Registry.ENTITY_TYPE.get(CraftNamespacedKey.fromMinecraft(net.minecraft.world.entity.EntityType.getKey(minecraft)));
         Preconditions.checkArgument(bukkit != null);
+        MINECRAFT_TO_BUKKIT_KEY_CACHE.put(minecraft, bukkit); // Canvas - cache minecraftToBukkit convert
         return bukkit;
     }
 
