--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntityType.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntityType.java
@@ -15,13 +_,22 @@
 
 public class CraftEntityType {
 
+    private static final java.util.Map<net.minecraft.world.entity.EntityType<?>, EntityType> MINECRAFT_TO_BUKKIT_KEY_CACHE = new java.util.concurrent.ConcurrentHashMap<>(); // Canvas - cache minecraftToBukkit convert
     public static EntityType minecraftToBukkit(net.minecraft.world.entity.EntityType<?> minecraft) {
         Preconditions.checkArgument(minecraft != null);
 
+        // Canvas start - cache minecraftToBukkit convert
+        EntityType asBukkit = MINECRAFT_TO_BUKKIT_KEY_CACHE.get(minecraft);
+
+        if (io.canvasmc.canvas.Config.INSTANCE.enableCachedMTBEntityTypeConvert && asBukkit != null) {
+            return asBukkit;
+        }
+        // Canvas end - cache minecraftToBukkit convert
         net.minecraft.core.Registry<net.minecraft.world.entity.EntityType<?>> registry = CraftRegistry.getMinecraftRegistry(Registries.ENTITY_TYPE);
         EntityType bukkit = Registry.ENTITY_TYPE.get(CraftNamespacedKey.fromMinecraft(registry.getResourceKey(minecraft).orElseThrow().location()));
 
         Preconditions.checkArgument(bukkit != null);
+        MINECRAFT_TO_BUKKIT_KEY_CACHE.put(minecraft, bukkit); // Canvas - cache minecraftToBukkit convert
 
         return bukkit;
     }
