--- a/net/minecraft/core/component/PatchedDataComponentMap.java
+++ b/net/minecraft/core/component/PatchedDataComponentMap.java
@@ -1,5 +_,6 @@
 package net.minecraft.core.component;
 
+import io.canvasmc.canvas.util.fastutil.Reference2ObjectConcurrentHashMap;
 import it.unimi.dsi.fastutil.objects.Reference2ObjectArrayMap;
 import it.unimi.dsi.fastutil.objects.Reference2ObjectMap;
 import it.unimi.dsi.fastutil.objects.Reference2ObjectMaps;
@@ -25,7 +_,7 @@
 
     private PatchedDataComponentMap(DataComponentMap prototype, Reference2ObjectMap<DataComponentType<?>, Optional<?>> patch, boolean copyOnWtite) {
         this.prototype = prototype;
-        this.patch = patch;
+        this.patch = patch instanceof Reference2ObjectConcurrentHashMap<DataComponentType<?>, Optional<?>> ? patch : new Reference2ObjectConcurrentHashMap<>(patch); // Canvas - concurrent
         this.copyOnWrite = copyOnWtite;
     }
 
@@ -136,7 +_,7 @@
 
     private void ensureMapOwnership() {
         if (this.copyOnWrite) {
-            this.patch = new Reference2ObjectArrayMap<>(this.patch);
+            this.patch = new Reference2ObjectConcurrentHashMap<>(this.patch); // Canvas - concurrent
             this.copyOnWrite = false;
         }
     }
