From 56a92b2c479a7161c340134ba931fb3d6a8e6f79 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/boss/wither/WitherBoss.java b/net/minecraft/world/entity/boss/wither/WitherBoss.java
index 1ab44f34b5d18d12259717444370f886b3a56144..0faada8a881273d7987ec7d8ea425ab0f8454069 100644
--- a/net/minecraft/world/entity/boss/wither/WitherBoss.java
+++ b/net/minecraft/world/entity/boss/wither/WitherBoss.java
@@ -137,6 +137,13 @@ public class WitherBoss extends Monster implements RangedAttackMob {
         super.setCustomName(name);
         this.bossEvent.setName(this.getDisplayName());
     }
+    // Pufferfish start - optimize suffocation
+
+    @Override
+    public boolean shouldCheckForSuffocation() {
+        return true;
+    }
+    // Pufferfish end
 
     @Override
     public SoundEvent getAmbientSound() {
