--- a/net/minecraft/world/level/block/BambooStalkBlock.java
+++ b/net/minecraft/world/level/block/BambooStalkBlock.java
@@ -81,12 +_,17 @@
 
     @Override
     public @Nullable BlockState getStateForPlacement(final BlockPlaceContext context) {
-        FluidState fluidState = context.getLevel().getFluidState(context.getClickedPos());
+        // Canvas start - optimize random tick
+        final BlockPos clickedPos = context.getClickedPos();
+        net.minecraft.world.level.chunk.LevelChunk chunk = context.getLevel().getChunkIfLoaded(clickedPos);
+        if (chunk == null) return null;
+        FluidState fluidState = chunk.getFluidState(clickedPos);
+        // Canvas end - optimize random tick
         if (!fluidState.isEmpty()) {
             return null;
         }
 
-        BlockState belowState = context.getLevel().getBlockState(context.getClickedPos().below());
+        BlockState belowState = chunk.getBlockState(context.getClickedPos().below()); // Canvas - optimize random tick
         if (belowState.is(BlockTags.SUPPORTS_BAMBOO)) {
             if (belowState.is(Blocks.BAMBOO_SAPLING)) {
                 return this.defaultBlockState().setValue(AGE, 0);
@@ -94,7 +_,7 @@
                 int age = belowState.getValue(AGE) > 0 ? 1 : 0;
                 return this.defaultBlockState().setValue(AGE, age);
             } else {
-                BlockState aboveState = context.getLevel().getBlockState(context.getClickedPos().above());
+                BlockState aboveState = chunk.getBlockState(context.getClickedPos().above()); // Canvas - optimize random tick
                 return aboveState.is(Blocks.BAMBOO)
                     ? this.defaultBlockState().setValue(AGE, aboveState.getValue(AGE))
                     : Blocks.BAMBOO_SAPLING.defaultBlockState();
@@ -191,9 +_,13 @@
     }
 
     protected void growBamboo(final BlockState state, final Level level, final BlockPos pos, final RandomSource random, final int height) {
-        BlockState belowState = level.getBlockState(pos.below());
+        // Canvas start - optimize random tick
+        net.minecraft.world.level.chunk.LevelChunk chunk = level.getChunkIfLoaded(pos);
+        if (chunk == null) return;
+        BlockState belowState = chunk.getBlockState(pos.below());
         BlockPos twoBelowPos = pos.below(2);
-        BlockState twoBelowState = level.getBlockState(twoBelowPos);
+        BlockState twoBelowState = chunk.getBlockState(twoBelowPos);
+        // Canvas end - optimize random tick
         BambooLeaves leaves = BambooLeaves.NONE;
         boolean shouldUpdateOthers = false; // CraftBukkit
         if (height >= 1) {
