--- a/net/minecraft/world/level/block/BambooStalkBlock.java
+++ b/net/minecraft/world/level/block/BambooStalkBlock.java
@@ -81,11 +_,16 @@
 
     @Override
     public @Nullable BlockState getStateForPlacement(BlockPlaceContext context) {
-        FluidState fluidState = context.getLevel().getFluidState(context.getClickedPos());
+        // Canvas start - optimize random tick
+        final BlockPos clickedPos = context.getClickedPos();
+        net.minecraft.world.level.chunk.LevelChunk chunk = context.getLevel().getChunkIfLoaded(clickedPos);
+        if (chunk == null) return null;
+        FluidState fluidState = chunk.getFluidState(clickedPos);
+        // Canvas end - optimize random tick
         if (!fluidState.isEmpty()) {
             return null;
         } else {
-            BlockState blockState = context.getLevel().getBlockState(context.getClickedPos().below());
+            BlockState blockState = chunk.getBlockState(context.getClickedPos().below()); // Canvas - optimize random tick
             if (blockState.is(BlockTags.BAMBOO_PLANTABLE_ON)) {
                 if (blockState.is(Blocks.BAMBOO_SAPLING)) {
                     return this.defaultBlockState().setValue(AGE, 0);
@@ -93,7 +_,7 @@
                     int i = blockState.getValue(AGE) > 0 ? 1 : 0;
                     return this.defaultBlockState().setValue(AGE, i);
                 } else {
-                    BlockState blockState1 = context.getLevel().getBlockState(context.getClickedPos().above());
+                    BlockState blockState1 = chunk.getBlockState(context.getClickedPos().above()); // Canvas - optimize random tick
                     return blockState1.is(Blocks.BAMBOO)
                         ? this.defaultBlockState().setValue(AGE, blockState1.getValue(AGE))
                         : Blocks.BAMBOO_SAPLING.defaultBlockState();
@@ -186,9 +_,13 @@
     }
 
     protected void growBamboo(BlockState state, Level level, BlockPos pos, RandomSource random, int age) {
-        BlockState blockState = level.getBlockState(pos.below());
+        // Canvas start - optimize random tick
+        net.minecraft.world.level.chunk.LevelChunk chunk = level.getChunkIfLoaded(pos);
+        if (chunk == null) return;
+        BlockState blockState = chunk.getBlockState(pos.below());
         BlockPos blockPos = pos.below(2);
-        BlockState blockState1 = level.getBlockState(blockPos);
+        BlockState blockState1 = chunk.getBlockState(blockPos);
+        // Canvas end - optimize random tick
         BambooLeaves bambooLeaves = BambooLeaves.NONE;
         boolean shouldUpdateOthers = false; // CraftBukkit
         if (age >= 1) {
