--- a/net/minecraft/world/level/block/Blocks.java
+++ b/net/minecraft/world/level/block/Blocks.java
@@ -5983,6 +_,12 @@
     private static Block register(
         final ResourceKey<Block> id, final Function<BlockBehaviour.Properties, Block> factory, final BlockBehaviour.Properties properties
     ) {
+        // Canvas start - MC-223153
+        if (io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaFixes.mc223153 && id.equals(BlockItemIds.RAW_COPPER_BLOCK.block())) {
+            Block block = factory.apply(properties.sound(SoundType.COPPER).setId(id));
+            return Registry.register(BuiltInRegistries.BLOCK, id, block);
+        }
+        // Canvas end - MC-223153
         Block block = factory.apply(properties.setId(id));
         return Registry.register(BuiltInRegistries.BLOCK, id, block);
     }
