--- a/net/minecraft/world/level/block/Blocks.java
+++ b/net/minecraft/world/level/block/Blocks.java
@@ -7218,6 +_,11 @@
     }
 
     private static Block register(final String id, final BlockBehaviour.Properties properties) {
+        // Canvas start - MC-223153
+        if (io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaFixes.mc223153 && id.equalsIgnoreCase("raw_copper_block")) {
+            return register(id, Block::new, properties.sound(SoundType.COPPER));
+        }
+        // Canvas end - MC-223153
         return register(id, Block::new, properties);
     }
 
