--- a/net/minecraft/world/level/block/Block.java
+++ b/net/minecraft/world/level/block/Block.java
@@ -385,8 +_,8 @@
             ItemEntity itemEntity = itemEntitySupplier.get();
             itemEntity.setDefaultPickUpDelay();
             // CraftBukkit start
-            if (level.captureDrops != null) {
-                level.captureDrops.add(itemEntity);
+            if (level.captureDrops.get() != null) { // Canvas - threadlocal
+                level.captureDrops.get().add(itemEntity); // Canvas - threadlocal
             } else {
                 level.addFreshEntity(itemEntity);
             }
