--- a/net/minecraft/world/level/block/Block.java
+++ b/net/minecraft/world/level/block/Block.java
@@ -385,11 +_,13 @@
             ItemEntity itemEntity = itemEntitySupplier.get();
             itemEntity.setDefaultPickUpDelay();
             // CraftBukkit start
-            if (level.captureDrops != null) {
-                level.captureDrops.add(itemEntity);
+            synchronized (level.captureDrops) { // Canvas
+            if (level.captureDrops.get() != null) { // Canvas
+                level.captureDrops.get().add(itemEntity); // Canvas
             } else {
                 level.addFreshEntity(itemEntity);
             }
+            } // Canvas
             // CraftBukkit end
         }
     }
