--- a/net/minecraft/world/level/block/CropBlock.java
+++ b/net/minecraft/world/level/block/CropBlock.java
@@ -163,6 +_,7 @@
     }
 
     protected static boolean hasSufficientLight(LevelReader level, BlockPos pos) {
+        if (io.canvasmc.canvas.Config.INSTANCE.cropsIgnoreLightCheck) return true; // Canvas - crops ignore light check config
         return level.getRawBrightness(pos, 0) >= 8;
     }
 
