--- a/net/minecraft/world/level/block/DaylightDetectorBlock.java
+++ b/net/minecraft/world/level/block/DaylightDetectorBlock.java
@@ -110,7 +_,7 @@
     }
 
     private static void tickEntity(Level level, BlockPos pos, BlockState state, DaylightDetectorBlockEntity blockEntity) {
-        if (level.getGameTime() % 20L == 0L) {
+        if (level.getRedstoneGameTime() % 20L == 0L) { // Canvas - Threaded Regions
             updateSignalStrength(state, level, pos);
         }
     }
