--- a/net/minecraft/world/level/block/EnchantingTableBlock.java
+++ b/net/minecraft/world/level/block/EnchantingTableBlock.java
@@ -45,8 +_,32 @@
     }
 
     public static boolean isValidBookShelf(final Level level, final BlockPos pos, final BlockPos offset) {
-        return level.getBlockState(pos.offset(offset)).is(BlockTags.ENCHANTMENT_POWER_PROVIDER)
-            && level.getBlockState(pos.offset(offset.getX() / 2, offset.getY(), offset.getZ() / 2)).is(BlockTags.ENCHANTMENT_POWER_TRANSMITTER);
+        // Canvas start - block configs
+        final BlockState offsetState = level.getBlockState(pos.offset(offset));
+
+        // must be a power provider
+        if (!offsetState.is(BlockTags.ENCHANTMENT_POWER_PROVIDER)) {
+            if (level.canvasConfig().blocks.fullChiseledBookShelvesCountAsValidEnchantPowerSources
+                && offsetState.is(Blocks.CHISELED_BOOKSHELF)) {
+
+                // all slots have to be filled for it to be a valid bookshelf
+                for (final net.minecraft.world.level.block.state.properties.BooleanProperty slotOccupiedProperty : ChiseledBookShelfBlock.SLOT_OCCUPIED_PROPERTIES) {
+                    if (!offsetState.getValue(slotOccupiedProperty)) {
+                        return false;
+                    }
+                }
+
+            }
+            else return false;
+        }
+
+        // check if something is blocking the shelf
+        if (!level.getBlockState(pos.offset(offset.getX() / 2, offset.getY(), offset.getZ() / 2)).is(BlockTags.ENCHANTMENT_POWER_TRANSMITTER)) {
+            return false;
+        }
+
+        return true;
+        // Canvas end - block configs
     }
 
     @Override
