--- a/net/minecraft/world/level/block/EndGatewayBlock.java
+++ b/net/minecraft/world/level/block/EndGatewayBlock.java
@@ -92,6 +_,7 @@
     protected void entityInside(BlockState state, Level level, BlockPos pos, Entity entity, InsideBlockEffectApplier effectApplier) {
         if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(level, pos)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
         if (entity.canUsePortal(false)
+            && (!io.canvasmc.canvas.Config.INSTANCE.blacklistNonPlayerEntitiesFromEnteringGatewayPortals || entity instanceof net.minecraft.server.level.ServerPlayer) // Canvas - blacklist non player entities from portals configs
             && !level.isClientSide
             && level.getBlockEntity(pos) instanceof TheEndGatewayBlockEntity theEndGatewayBlockEntity
             && !theEndGatewayBlockEntity.isCoolingDown()) {
