--- a/net/minecraft/world/level/block/EndPortalBlock.java
+++ b/net/minecraft/world/level/block/EndPortalBlock.java
@@ -67,7 +_,10 @@
         final boolean isPrecise
     ) {
         if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(level, pos)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
-        if (entity.canUsePortal(false)) {
+        // Canvas start - blacklist non player entities from portals configs
+        if (entity.canUsePortal(false) &&
+            (!io.canvasmc.canvas.GlobalConfiguration.getInstance().blacklistNonPlayerEntitiesFromEnteringEndPortals || entity instanceof ServerPlayer)) {
+        // Canvas end - blacklist non player entities from portals configs
             // CraftBukkit start - Entity in portal
             org.bukkit.event.entity.EntityPortalEnterEvent event = new org.bukkit.event.entity.EntityPortalEnterEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level), org.bukkit.PortalType.ENDER); // Paper - add portal type
             if (!event.callEvent()) return; // Paper - make cancellable
@@ -76,6 +_,12 @@
                 if (level.paperConfig().misc.disableEndCredits) {player.seenCredits = true; return;} // Paper - Option to disable end credits
                 player.showEndCredits();
             } else {
+                // Canvas start - unsafe end portal teleportation
+                if (io.canvasmc.canvas.GlobalConfiguration.getInstance().allowUnsafeEndPortalTeleportation && !(entity instanceof net.minecraft.world.entity.player.Player)) {
+                    entity.endPortalLogicAsync(pos);
+                    return;
+                }
+                // Canvas end - unsafe end portal teleportation
                 entity.setAsInsidePortal(this, pos);
             }
         }
@@ -134,13 +_,15 @@
 
     // Folia start - region threading
+    @Override
+    public boolean requiresPortalEntryBlockOwnership() {
+        return false;
+    }
+
     @Override
     public boolean portalAsync(ServerLevel sourceWorld, Entity portalTarget, BlockPos portalPos) {
         if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(portalTarget)) {
             return false;
         }
-        if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(sourceWorld, portalPos)) {
-            return false;
-        }
 
         return portalTarget.endPortalLogicAsync(portalPos);
     }
