--- a/net/minecraft/world/level/block/FarmBlock.java
+++ b/net/minecraft/world/level/block/FarmBlock.java
@@ -189,7 +_,10 @@
                 int x = xOff + dx;
                 for (int dy = 0; dy <= 1; ++dy) {
                     int y = dy + yOff;
-                    net.minecraft.world.level.chunk.LevelChunk chunk = (net.minecraft.world.level.chunk.LevelChunk)level.getChunk(x >> 4, z >> 4);
+                    // Canvas start - don't call syncload
+                    @Nullable net.minecraft.world.level.chunk.LevelChunk chunk = (net.minecraft.world.level.chunk.LevelChunk)level.getChunkIfLoadedImmediately(x >> 4, z >> 4);
+                    if (chunk == null) continue;
+                    // Canvas end
                     net.minecraft.world.level.material.FluidState fluid = chunk.getBlockStateFinal(x, y, z).getFluidState();
                     if (fluid.is(FluidTags.WATER)) {
                         return true;
