--- a/net/minecraft/world/level/block/FireBlock.java
+++ b/net/minecraft/world/level/block/FireBlock.java
@@ -341,7 +_,11 @@
     }
 
     private int getIgniteOdds(LevelReader level, BlockPos pos) {
-        if (!level.isEmptyBlock(pos)) {
+        // Canvas start - prevent syncload
+        BlockState state = level.getBlockStateIfLoaded(pos);
+        if (state == null) return 0;
+        if (!state.isAir()) {
+        // Canvas end
             return 0;
         } else {
             int i = 0;
