--- a/net/minecraft/world/level/block/FlowerPotBlock.java
+++ b/net/minecraft/world/level/block/FlowerPotBlock.java
@@ -81,6 +_,7 @@
             level.setBlock(pos, blockState, 3);
             level.gameEvent(player, GameEvent.BLOCK_CHANGE, pos);
             player.awardStat(Stats.POT_FLOWER);
+            if (io.canvasmc.canvas.Config.INSTANCE.fixes.mc231743) player.awardStat(Stats.ITEM_USED.get(player.getItemInHand(hand).getItem())); // Canvas - MC-231743
             stack.consume(1, player);
             return InteractionResult.SUCCESS;
         }
