--- a/net/minecraft/world/level/block/LeavesBlock.java
+++ b/net/minecraft/world/level/block/LeavesBlock.java
@@ -50,11 +_,13 @@
 
     @Override
     protected boolean isRandomlyTicking(BlockState state) {
+        if (io.canvasmc.canvas.Config.INSTANCE.disableLeafDecay) return false; // Canvas - disable leaf decay config
         return state.getValue(DISTANCE) == 7 && !state.getValue(PERSISTENT);
     }
 
     @Override
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
+        if (io.canvasmc.canvas.Config.INSTANCE.disableLeafDecay) return; // Canvas - disable leaf decay config
         if (this.decaying(state)) {
             // CraftBukkit start
             org.bukkit.event.block.LeavesDecayEvent event = new org.bukkit.event.block.LeavesDecayEvent(org.bukkit.craftbukkit.block.CraftBlock.at(level, pos));
@@ -75,6 +_,7 @@
 
     @Override
     protected void tick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
+        if (io.canvasmc.canvas.Config.INSTANCE.disableLeafDecay) return; // Canvas - disable leaf decay config
         level.setBlock(pos, updateDistance(state, level, pos), 3);
     }
 
