--- a/net/minecraft/world/level/block/LeavesBlock.java
+++ b/net/minecraft/world/level/block/LeavesBlock.java
@@ -82,7 +_,23 @@
 
     @Override
     protected void tick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
-        level.setBlock(pos, updateDistance(state, level, pos), 3);
+        // Canvas start - dont set the block if it doesnt need to
+        int newValue = 7;
+        int oldValue = state.getValue(DISTANCE);
+        BlockPos.MutableBlockPos mutable = new BlockPos.MutableBlockPos();
+
+        for (Direction direction : Direction.values()) {
+            mutable.setWithOffset(pos, direction);
+            newValue = Math.min(newValue, getDistanceAt(level.getBlockState(mutable)) + 1);
+            if (newValue == 1) {
+                break;
+            }
+        }
+
+        if (newValue != oldValue) {
+            level.setBlock(pos, state.setValue(DISTANCE, newValue), 3);
+        }
+        // Canvas end
     }
 
     @Override
