--- a/net/minecraft/world/level/block/LiquidBlock.java
+++ b/net/minecraft/world/level/block/LiquidBlock.java
@@ -29,6 +_,7 @@
 import net.minecraft.world.level.block.state.StateDefinition;
 import net.minecraft.world.level.block.state.properties.BlockStateProperties;
 import net.minecraft.world.level.block.state.properties.IntegerProperty;
+import net.minecraft.world.level.chunk.LevelChunk;
 import net.minecraft.world.level.material.FlowingFluid;
 import net.minecraft.world.level.material.Fluid;
 import net.minecraft.world.level.material.FluidState;
@@ -189,17 +_,24 @@
 
             for (Direction direction : POSSIBLE_FLOW_DIRECTIONS) {
                 BlockPos blockPos = pos.relative(direction.getOpposite());
-                if (level.getFluidState(blockPos).is(FluidTags.WATER)) {
+                // Canvas start - dont call syncload
+                @Nullable LevelChunk possibleChunk = level.getChunkIfLoaded(blockPos);
+                if (possibleChunk == null) continue;
+                if (possibleChunk.getFluidState(blockPos).is(FluidTags.WATER)) {
+                    // calling 'level.getFluidState(pos)' is fine now, given the position provided to the method
+                    // cannot be in an unloaded chunk, since it's either placed or a neighbor(which needs to be loaded)
+                // Canvas end
                     Block block = level.getFluidState(pos).isSource() ? Blocks.OBSIDIAN : Blocks.COBBLESTONE;
                     // CraftBukkit start
                     if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level, pos, block.defaultBlockState())) {
+                        level.setBlock(pos, block.defaultBlockState(), 3); // Canvas
                         this.fizz(level, pos);
                     }
                     // CraftBukkit end
                     return false;
                 }
 
-                if (isSoulSoil && level.getBlockState(blockPos).is(Blocks.BLUE_ICE)) {
+                if (isSoulSoil && possibleChunk.getBlockState(blockPos).is(Blocks.BLUE_ICE)) { // Canvas - dont call syncload
                     // CraftBukkit start
                     if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockFormEvent(level, pos, Blocks.BASALT.defaultBlockState())) {
                         this.fizz(level, pos);
