--- a/net/minecraft/world/level/block/NetherWartBlock.java
+++ b/net/minecraft/world/level/block/NetherWartBlock.java
@@ -17,7 +_,7 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public class NetherWartBlock extends VegetationBlock {
+public class NetherWartBlock extends VegetationBlock implements BonemealableBlock { // Canvas - nether wart bonemealable
     public static final MapCodec<NetherWartBlock> CODEC = simpleCodec(NetherWartBlock::new);
     public static final int MAX_AGE = 3;
     public static final IntegerProperty AGE = BlockStateProperties.AGE_3;
@@ -66,4 +_,48 @@
     protected void createBlockStateDefinition(final StateDefinition.Builder<Block, BlockState> builder) {
         builder.add(AGE);
     }
+    // Canvas start - nether wart bonemealable
+
+    @Override
+    public boolean isValidBonemealTarget(
+        final LevelReader levelReader,
+        final BlockPos pos,
+        final BlockState state
+    ) {
+        return state.getValue(AGE) < MAX_AGE && levelReader instanceof net.minecraft.server.level.ServerLevel level && level.canvasConfig().farming.netherWartBonemealable;
+    }
+
+    @Override
+    public boolean isBonemealSuccess(
+        final net.minecraft.world.level.Level level,
+        final RandomSource random,
+        final BlockPos pos,
+        final BlockState state
+    ) {
+        return level.canvasConfig().farming.netherWartBonemealable;
+    }
+
+    @Override
+    public void performBonemeal(
+        final ServerLevel level,
+        final RandomSource random,
+        final BlockPos pos,
+        final BlockState state
+    ) {
+        int age = Math.min(MAX_AGE, state.getValue(AGE) + this.getBonemealAgeIncrease(level));
+        org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, pos, this.getStateForAge(age), Block.UPDATE_CLIENTS);
+        level.sendParticles(
+            net.minecraft.core.particles.ParticleTypes.HAPPY_VILLAGER,
+            pos.getX() + 0.5, pos.getY() + 0.5, pos.getZ() + 0.5, 10, 0.25, 0.25, 0.25, 0
+        );
+    }
+
+    private int getBonemealAgeIncrease(final net.minecraft.world.level.Level level) {
+        return net.minecraft.util.Mth.nextInt(level.getRandom(), 1, 3);
+    }
+
+    private BlockState getStateForAge(final int age) {
+        return this.defaultBlockState().setValue(AGE, age);
+    }
+    // Canvas end - nether wart bonemealable
 }
