--- a/net/minecraft/world/level/block/NetherPortalBlock.java
+++ b/net/minecraft/world/level/block/NetherPortalBlock.java
@@ -120,7 +_,10 @@
         final boolean isPrecise
     ) {
         if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(level, pos)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
-        if (entity.canUsePortal(false)) {
+            // Canvas start - blacklist non player entities from portals configs
+            if (entity.canUsePortal(false) &&
+                (!io.canvasmc.canvas.Config.INSTANCE.blacklistNonPlayerEntitiesFromEnteringNetherPortals || entity instanceof net.minecraft.server.level.ServerPlayer)) {
+            // Canvas end - blacklist non player entities from portals configs
             // CraftBukkit start - Entity in portal
             org.bukkit.event.entity.EntityPortalEnterEvent event = new org.bukkit.event.entity.EntityPortalEnterEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level), org.bukkit.PortalType.NETHER); // Paper - add portal type
             level.getCraftServer().getPluginManager().callEvent(event);
@@ -143,7 +_,7 @@
 
     @Override
     public @Nullable TeleportTransition getPortalDestination(final ServerLevel currentLevel, final Entity entity, final BlockPos portalEntryPos) {
-        ResourceKey<Level> newDimension = currentLevel.getTypeKey() == net.minecraft.world.level.dimension.LevelStem.NETHER ? Level.OVERWORLD : Level.NETHER; // CraftBukkit
+        ResourceKey<Level> newDimension = currentLevel.getTypeKey() == net.minecraft.world.level.dimension.LevelStem.NETHER ? io.canvasmc.canvas.Config.INSTANCE.fetchRespawnDimensionKey() : Level.NETHER; // CraftBukkit // Canvas - default respawn dimension config
         ServerLevel newLevel = currentLevel.getServer().getLevel(newDimension);
         // Paper start - Add EntityPortalReadyEvent
         io.papermc.paper.event.entity.EntityPortalReadyEvent portalReadyEvent = new io.papermc.paper.event.entity.EntityPortalReadyEvent(entity.getBukkitEntity(), newLevel == null ? null : newLevel.getWorld(), org.bukkit.PortalType.NETHER);
