--- a/net/minecraft/world/level/block/Portal.java
+++ b/net/minecraft/world/level/block/Portal.java
@@ -16,6 +_,10 @@
     // Folia start - region threading
     public boolean portalAsync(ServerLevel sourceWorld, Entity portalTarget, BlockPos portalPos);
 
+    default boolean requiresPortalEntryBlockOwnership() {
+        return true;
+    }
+
     // Folia end - region threading
     default Portal.Transition getLocalTransition() {
         return Portal.Transition.NONE;
