--- a/net/minecraft/world/level/block/RedstoneTorchBlock.java
+++ b/net/minecraft/world/level/block/RedstoneTorchBlock.java
@@ -6,6 +_,7 @@
 import java.util.Map;
 import java.util.WeakHashMap;
 import javax.annotation.Nullable;
+import io.canvasmc.canvas.region.ServerRegions;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.core.particles.DustParticleOptions;
@@ -73,10 +_,10 @@
     protected void tick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
         boolean hasNeighborSignal = this.hasNeighborSignal(level, pos, state);
         // Paper start - Faster redstone torch rapid clock removal
-        java.util.ArrayDeque<RedstoneTorchBlock.Toggle> redstoneUpdateInfos = level.redstoneUpdateInfos;
+        java.util.ArrayDeque<RedstoneTorchBlock.Toggle> redstoneUpdateInfos = ServerRegions.getTickData(level).redstoneUpdateInfos; // Canvas - Threaded Regions
         if (redstoneUpdateInfos != null) {
             RedstoneTorchBlock.Toggle curr;
-            while ((curr = redstoneUpdateInfos.peek()) != null && level.getGameTime() - curr.when > 60L) {
+            while ((curr = redstoneUpdateInfos.peek()) != null && level.getRedstoneGameTime() - curr.when > 60L) { // Canvas - Threaded Regions
                 redstoneUpdateInfos.poll();
             }
         }
@@ -154,13 +_,13 @@
 
     private static boolean isToggledTooFrequently(Level level, BlockPos pos, boolean logToggle) {
         // Paper start - Faster redstone torch rapid clock removal
-        java.util.ArrayDeque<RedstoneTorchBlock.Toggle> list = level.redstoneUpdateInfos;
+        java.util.ArrayDeque<RedstoneTorchBlock.Toggle> list = ServerRegions.getTickData((ServerLevel) level).redstoneUpdateInfos; // Canvas - Threaded Regions
         if (list == null) {
-            list = level.redstoneUpdateInfos = new java.util.ArrayDeque<>();
+            list = ServerRegions.getTickData((ServerLevel) level).redstoneUpdateInfos = new java.util.ArrayDeque<>(); // Canvas - Threaded Regions
         }
         // Paper end - Faster redstone torch rapid clock removal
         if (logToggle) {
-            list.add(new RedstoneTorchBlock.Toggle(pos.immutable(), level.getGameTime()));
+            list.add(new RedstoneTorchBlock.Toggle(pos.immutable(), level.getRedstoneGameTime())); // Canvas - Threaded Regions
         }
 
         int i = 0;
@@ -182,12 +_,18 @@
     }
 
     public static class Toggle {
-        final BlockPos pos;
-        final long when;
+        public final BlockPos pos; // Canvas - public
+        long when; // Canvas - mutable
 
         public Toggle(BlockPos pos, long when) {
             this.pos = pos;
             this.when = when;
         }
+        // Canvas start - Threaded Regions
+
+        public void offsetTime(long offset) {
+            this.when += offset;
+        }
+        // Canvas end
     }
 }
