--- a/net/minecraft/world/level/block/SaplingBlock.java
+++ b/net/minecraft/world/level/block/SaplingBlock.java
@@ -56,18 +_,18 @@
             level.setBlock(pos, state.cycle(STAGE), 4);
         } else {
             // CraftBukkit start
-            if (level.captureTreeGeneration) {
+            if (level.captureTreeGeneration.get()) { // Canvas - threadlocal
                 this.treeGrower.growTree(level, level.getChunkSource().getGenerator(), pos, state, random);
             } else {
-                level.captureTreeGeneration = true;
+                level.captureTreeGeneration.set(true); // Canvas - threadlocal
                 this.treeGrower.growTree(level, level.getChunkSource().getGenerator(), pos, state, random);
-                level.captureTreeGeneration = false;
-                if (!level.capturedBlockStates.isEmpty()) {
+                level.captureTreeGeneration.set(false); // Canvas - threadlocal
+                if (!level.capturedBlockStates.get().isEmpty()) { // Canvas - threadlocal
                     org.bukkit.TreeType treeType = SaplingBlock.treeType;
                     SaplingBlock.treeType = null;
                     org.bukkit.Location location = org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level.getWorld());
-                    java.util.List<org.bukkit.block.BlockState> blocks = new java.util.ArrayList<>(level.capturedBlockStates.values());
-                    level.capturedBlockStates.clear();
+                    java.util.List<org.bukkit.block.BlockState> blocks = new java.util.ArrayList<>(level.capturedBlockStates.get().values()); // Canvas - threadlocal
+                    level.capturedBlockStates.get().clear(); // Canvas - threadlocal
                     org.bukkit.event.world.StructureGrowEvent event = null;
                     if (treeType != null) {
                         event = new org.bukkit.event.world.StructureGrowEvent(location, treeType, false, null, blocks);
