--- a/net/minecraft/world/level/block/SaplingBlock.java
+++ b/net/minecraft/world/level/block/SaplingBlock.java
@@ -56,12 +_,12 @@
             level.setBlock(pos, state.cycle(STAGE), 4);
         } else {
             // CraftBukkit start
-            if (level.captureTreeGeneration) {
+            if (level.captureTreeGeneration.get()) { // Canvas - atomic
                 this.treeGrower.growTree(level, level.getChunkSource().getGenerator(), pos, state, random);
             } else {
-                level.captureTreeGeneration = true;
+                level.captureTreeGeneration.set(true); // Canvas - atomic
                 this.treeGrower.growTree(level, level.getChunkSource().getGenerator(), pos, state, random);
-                level.captureTreeGeneration = false;
+                level.captureTreeGeneration.set(false); // Canvas - atomic
                 if (!level.capturedBlockStates.isEmpty()) {
                     org.bukkit.TreeType treeType = SaplingBlock.treeType;
                     SaplingBlock.treeType = null;
