--- a/net/minecraft/world/level/block/SlimeBlock.java
+++ b/net/minecraft/world/level/block/SlimeBlock.java
@@ -1,6 +_,7 @@
 package net.minecraft.world.level.block;
 
 import com.mojang.serialization.MapCodec;
+import io.canvasmc.canvas.Config;
 import net.minecraft.core.BlockPos;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.LivingEntity;
@@ -41,7 +_,7 @@
     private void bounceUp(Entity entity) {
         Vec3 deltaMovement = entity.getDeltaMovement();
         if (deltaMovement.y < 0.0) {
-            double d = entity instanceof LivingEntity ? 1.0 : 0.8;
+            double d = entity instanceof LivingEntity ? Config.INSTANCE.blocks.livingEntityVelocityReboundFactor : Config.INSTANCE.blocks.nonLivingEntityVelocityReboundFactor; // Canvas - slime rebound velocity config
             entity.setDeltaMovement(deltaMovement.x, -deltaMovement.y * d, deltaMovement.z);
         }
     }
