--- a/net/minecraft/world/level/block/SpreadingSnowyDirtBlock.java
+++ b/net/minecraft/world/level/block/SpreadingSnowyDirtBlock.java
@@ -52,7 +_,7 @@
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
         if (this instanceof GrassBlock && level.paperConfig().tickRates.grassSpread != 1 && (level.paperConfig().tickRates.grassSpread < 1 || (io.papermc.paper.threadedregions.RegionizedServer.getCurrentTick() + pos.hashCode()) % level.paperConfig().tickRates.grassSpread != 0)) { return; } // Paper - Configurable random tick rates for blocks // Folia - regionised ticking
         // Paper start - Perf: optimize dirt and snow spreading
-        final net.minecraft.world.level.chunk.ChunkAccess cachedBlockChunk = level.getChunkIfLoaded(pos);
+        final net.minecraft.world.level.chunk.LevelChunk cachedBlockChunk = level.getChunkIfLoaded(pos); // Canvas - optimize random tick
         if (cachedBlockChunk == null) { // Is this needed?
             return;
         }
@@ -66,24 +_,27 @@
             // CraftBukkit end
             level.setBlockAndUpdate(pos, Blocks.DIRT.defaultBlockState());
         } else {
-            if (level.getMaxLocalRawBrightness(pos.above()) >= 9) {
+            if (io.canvasmc.canvas.Config.INSTANCE.disableGrassLightChecks || level.getMaxLocalRawBrightness(pos.above()) >= 9) { // Canvas - disable grass light checks config
                 BlockState blockState = this.defaultBlockState();
 
                 for (int i = 0; i < 4; i++) {
                     BlockPos blockPos = pos.offset(random.nextInt(3) - 1, random.nextInt(5) - 3, random.nextInt(3) - 1);
                     // Paper start - Perf: optimize dirt and snow spreading
-                    if (pos.getX() == blockPos.getX() && pos.getY() == blockPos.getY() && pos.getZ() == blockPos.getZ()) {
+                    if (pos == blockPos) { // Canvas - optimize random tick
                         continue;
                     }
 
-                    final net.minecraft.world.level.chunk.ChunkAccess access;
+                    final net.minecraft.world.level.chunk.LevelChunk access; // Canvas - optimize random tick
                     if (cachedBlockChunk.locX == blockPos.getX() >> 4 && cachedBlockChunk.locZ == blockPos.getZ() >> 4) {
                         access = cachedBlockChunk;
                     } else {
-                        access = level.getChunkAt(blockPos);
+                        // Canvas start - optimize random tick
+                        access = level.getChunkIfLoaded(blockPos);
+                        if (access == null) continue;
                     }
-                    if (access.getBlockState(blockPos).is(Blocks.DIRT) && SpreadingSnowyDirtBlock.canPropagate(access, blockState, level, blockPos)) {
-                        org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(level, pos, blockPos, blockState.setValue(SNOWY, isSnowySetting(access.getBlockState(blockPos.above()))), Block.UPDATE_ALL); // CraftBukkit
+                    if (access.getBlockStateFinal(blockPos.getX(), blockPos.getY(), blockPos.getZ()).is(Blocks.DIRT) && SpreadingSnowyDirtBlock.canPropagate(access, blockState, level, blockPos)) {
+                        org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(level, pos, blockPos, blockState.setValue(SNOWY, isSnowySetting(access.getBlockStateFinal(blockPos.getX(), blockPos.getY() + 1, blockPos.getZ()))), Block.UPDATE_ALL); // CraftBukkit
+                        // Canvas end - optimize random tick
                         // Paper end - Perf: optimize dirt and snow spreading
                     }
                 }
