--- a/net/minecraft/world/level/block/SugarCaneBlock.java
+++ b/net/minecraft/world/level/block/SugarCaneBlock.java
@@ -19,7 +_,7 @@
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.VoxelShape;
 
-public class SugarCaneBlock extends Block {
+public class SugarCaneBlock extends Block implements BonemealableBlock { // Canvas - sugar cane bonemealable
     public static final MapCodec<SugarCaneBlock> CODEC = simpleCodec(SugarCaneBlock::new);
     public static final IntegerProperty AGE = BlockStateProperties.AGE_15;
     private static final VoxelShape SHAPE = Block.column(12.0, 0.0, 16.0);
@@ -112,4 +_,88 @@
     protected void createBlockStateDefinition(final StateDefinition.Builder<Block, BlockState> builder) {
         builder.add(AGE);
     }
+    // Canvas start - bonemealable sugarcane
+
+    @Override
+    public boolean isValidBonemealTarget(
+        final LevelReader levelReader,
+        final BlockPos pos,
+        final BlockState state
+    ) {
+        if (!(levelReader instanceof ServerLevel level) || !level.canvasConfig().farming.sugarCaneBonemealable) return false;
+
+        int heightAbove = this.getHeightAboveUpToMax(level, pos);
+        int heightBelow = this.getHeightBelowUpToMax(level, pos);
+        BlockPos growthPos = pos.above(heightAbove + 1);
+        return heightAbove + heightBelow + 1 < level.paperConfig().maxGrowthHeight.reeds
+            && level.isInsideBuildHeight(growthPos)
+            && level.isEmptyBlock(growthPos);
+    }
+
+    @Override
+    public boolean isBonemealSuccess(
+        final net.minecraft.world.level.Level level,
+        final RandomSource random,
+        final BlockPos pos,
+        final BlockState state
+    ) {
+        return level.canvasConfig().farming.sugarCaneBonemealable;
+    }
+
+    @Override
+    public void performBonemeal(
+        final ServerLevel level,
+        final RandomSource random,
+        final BlockPos pos,
+        final BlockState state
+    ) {
+        if (!level.canvasConfig().farming.sugarCaneBonemealable) return;
+
+        int heightAbove = this.getHeightAboveUpToMax(level, pos);
+        int heightBelow = this.getHeightBelowUpToMax(level, pos);
+        int totalHeight = heightAbove + heightBelow + 1;
+        int newCane = 1 + random.nextInt(2);
+
+        for (int i = 0; i < newCane; i++) {
+            BlockPos topPos = pos.above(heightAbove);
+            BlockState topState = level.getBlockState(topPos);
+            BlockPos growthPos = topPos.above();
+            if (totalHeight >= level.paperConfig().maxGrowthHeight.reeds || !topState.is(Blocks.SUGAR_CANE) || !level.isEmptyBlock(growthPos) || level.isOutsideBuildHeight(growthPos)) {
+                return;
+            }
+
+            growCane(level, topPos);
+            heightAbove++;
+            totalHeight++;
+        }
+    }
+
+    protected void growCane(final ServerLevel level, final BlockPos pos) {
+        org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(level, pos, pos.above(), Blocks.SUGAR_CANE.defaultBlockState(), Block.UPDATE_ALL);
+        level.sendParticles(
+            net.minecraft.core.particles.ParticleTypes.HAPPY_VILLAGER,
+            pos.getX() + 0.5, pos.getY() + 0.5, pos.getZ() + 0.5, 10, 0.25, 0.25, 0.25, 0
+        );
+    }
+
+    protected int getHeightAboveUpToMax(final BlockGetter level, final BlockPos pos) {
+        int height = 0;
+
+        while (height < ((ServerLevel) level).paperConfig().maxGrowthHeight.reeds && level.getBlockState(pos.above(height + 1)).is(Blocks.SUGAR_CANE)) {
+            height++;
+        }
+
+        return height;
+    }
+
+    protected int getHeightBelowUpToMax(final BlockGetter level, final BlockPos pos) {
+        int height = 0;
+
+        while (height < ((net.minecraft.world.level.Level) level).paperConfig().maxGrowthHeight.reeds && level.getBlockState(pos.below(height + 1)).is(Blocks.SUGAR_CANE)) {
+            height++;
+        }
+
+        return height;
+    }
+    // Canvas end - bonemealable sugarcane
 }
