--- a/net/minecraft/world/level/block/VineBlock.java
+++ b/net/minecraft/world/level/block/VineBlock.java
@@ -20,6 +_,7 @@
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.block.state.StateDefinition;
 import net.minecraft.world.level.block.state.properties.BooleanProperty;
+import net.minecraft.world.level.chunk.LevelChunk;
 import net.minecraft.world.phys.shapes.CollisionContext;
 import net.minecraft.world.phys.shapes.Shapes;
 import net.minecraft.world.phys.shapes.VoxelShape;
@@ -293,7 +_,18 @@
         int i1 = 5;
 
         for (BlockPos blockPos : iterable) {
-            if (blockReader.getBlockState(blockPos).is(this)) {
+            // Canvas start - don't call syncload
+            @org.jetbrains.annotations.Nullable LevelChunk chunk = null;
+            if (blockReader instanceof ServerLevel serverLevel) {
+                chunk = serverLevel.getChunkIfLoaded(blockPos);
+                // chunk not loaded, skip
+                if (chunk == null) {
+                    continue;
+                }
+            }
+            // by the time we get here, it's not null
+            if (chunk.getBlockState(blockPos).is(this)) {
+            // Canvas end
                 if (--i1 <= 0) {
                     return false;
                 }
