--- a/net/minecraft/world/level/timers/TimerQueue.java
+++ b/net/minecraft/world/level/timers/TimerQueue.java
@@ -11,6 +_,7 @@
 import java.util.PriorityQueue;
 import java.util.Queue;
 import java.util.Set;
+import java.util.concurrent.PriorityBlockingQueue;
 import java.util.stream.Stream;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.nbt.ListTag;
@@ -24,7 +_,7 @@
     private static final String TIMER_NAME_TAG = "Name";
     private static final String TIMER_TRIGGER_TIME_TAG = "TriggerTime";
     private final TimerCallbacks<T> callbacksRegistry;
-    private final Queue<TimerQueue.Event<T>> queue = new PriorityQueue<>(createComparator());
+    private final Queue<TimerQueue.Event<T>> queue = new PriorityBlockingQueue<>(11, createComparator()); // Canvas - threadsafe
     private UnsignedLong sequentialId = UnsignedLong.ZERO;
     private final Table<String, Long, TimerQueue.Event<T>> events = HashBasedTable.create();
 
