--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -5,6 +_,7 @@
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
 import com.mojang.datafixers.util.Pair;
+import io.canvasmc.canvas.LevelAccess;
 import io.papermc.paper.FeatureHooks;
 import io.papermc.paper.raytracing.RayTraceTarget;
 import io.papermc.paper.registry.RegistryAccess;
@@ -2431,6 +_,13 @@
 
         return this.getStructures(x, z, struct -> registry.getKey(struct).equals(key));
     }
+    // Canvas start
+
+    @Override
+    public @NotNull LevelAccess getLevelAccess() {
+        return this.getHandle();
+    }
+    // Canvas end
 
     private List<GeneratedStructure> getStructures(int x, int z, Predicate<net.minecraft.world.level.levelgen.structure.Structure> predicate) {
         List<GeneratedStructure> structures = new ArrayList<>();
