--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -232,7 +_,7 @@
             if (chunk == null) {
                 continue;
             }
-            size += chunk.blockEntities.size();
+            size += chunk.canvas$getAllBlockEntities().length; // Canvas - optimize block entity fetching
         }
         return size;
     }
@@ -1305,12 +_,7 @@
     public void save(boolean flush) {
         org.spigotmc.AsyncCatcher.catchOp("world save"); // Spigot
         this.server.checkSaveState();
-        boolean oldSave = this.world.noSave;
-
-        this.world.noSave = false;
-        this.world.save(null, flush, false);
-
-        this.world.noSave = oldSave;
+        this.world.canvas$worldUnloadTicket.propagate(new io.canvasmc.canvas.util.ticket.UnloadTicket((b) -> {}, true)); // Canvas - region threading
     }
 
     @Override
@@ -2536,4 +_,67 @@
         return POINTERS_SUPPLIER.view(this);
     }
     // Paper end
+    // Canvas start - region threading
+
+    @Override
+    @NotNull
+    public io.canvasmc.canvas.region.WorldRegionizer getRegionizer() {
+        return this.world.regioniser;
+    }
+
+    @Override
+    public boolean isTickThreadFor(@NotNull Block block) {
+        return ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.world, ((CraftBlock) block).getPosition());
+    }
+
+    @Override
+    public boolean isTickThreadFor(@NotNull Block block, int blockRadius) {
+        return ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.world, ((CraftBlock) block).getPosition(), blockRadius);
+    }
+
+    @Override
+    public boolean isTickThreadFor(@NotNull Chunk chunk) {
+        return isTickThreadFor(chunk.getX(), chunk.getZ());
+    }
+
+    @Override
+    public boolean isTickThreadFor(@NotNull Location position) {
+        return ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.world, CraftLocation.toVec3(position));
+    }
+
+    @Override
+    public boolean isTickThreadFor(int chunkX, int chunkZ) {
+        return ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.world, chunkX, chunkZ);
+    }
+
+    @Override
+    public boolean isTickThreadFor(@NotNull BoundingBox aabb) {
+        return ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.world, new AABB(aabb.getMinX(), aabb.getMinY(), aabb.getMinZ(), aabb.getMaxX(), aabb.getMaxY(), aabb.getMaxZ()));
+    }
+
+    @Override
+    public boolean isTickThreadFor(double blockX, double blockZ) {
+        return ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.world, blockX, blockZ);
+    }
+
+    @Override
+    public boolean isTickThreadFor(Location position, Vector velocity, int buffer) {
+        return ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.world, CraftLocation.toVec3(position), org.bukkit.craftbukkit.util.CraftVector.toVec3(velocity), buffer);
+    }
+
+    @Override
+    public boolean isTickThreadFor(int fromChunkX, int fromChunkZ, int toChunkX, int toChunkZ) {
+        return ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.world, fromChunkX, fromChunkZ, toChunkX, toChunkZ);
+    }
+
+    @Override
+    public boolean isTickThreadFor(int chunkX, int chunkZ, int radius) {
+        return ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.world, chunkX, chunkZ, radius);
+    }
+
+    @Override
+    public boolean isTickThreadFor(Entity entity) {
+        return ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(((CraftEntity) entity).getHandleRaw());
+    }
+    // Canvas end - region threading
 }
