--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -3160,4 +_,19 @@
         }
     }
     // Folia end - region TPS API
+    // Canvas start - region threading
+    // note: in impl the world is provided, same as 'getWorld()'
+    @Override
+    public <T> io.canvasmc.canvas.region.RegionTickData.@NotNull IRegionizedData<T> createRegionizedData(java.util.function.@NotNull BiFunction<io.canvasmc.canvas.region.RegionTickData, org.bukkit.World, T> supplier, io.canvasmc.canvas.region.RegionTickData.IRegionizedData.IRegionizedCallback<T> callback) {
+        return io.papermc.paper.threadedregions.TickRegions.TickRegionData.createRegionizedDataApiHandle(supplier, callback);
+    }
+
+    @Override
+    public <T> T getLocalRegionizedData(final io.canvasmc.canvas.region.RegionTickData.IRegionizedData<T> regionizedData) {
+        io.papermc.paper.threadedregions.ThreadedRegionizer.ThreadedRegion<io.papermc.paper.threadedregions.TickRegions.TickRegionData, io.papermc.paper.threadedregions.TickRegions.TickRegionSectionData>
+            region = io.papermc.paper.threadedregions.TickRegionScheduler.getCurrentRegion();
+        if (region == null) throw new IllegalStateException("Must be on a tick thread runner that currently owns a region");
+        return region.getData().getOrCreateFromIRegionizedData(regionizedData);
+    }
+    // Canvas end - region threading
 }
