From c9d43594f1dc856ec561804d690b933859e1d7f9 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas paperServer File Patches


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 0bfb8b5c936d863c618323d72143c9ca06080fa8..524ef901fcac85dafeb7f24f53fe3274c2f7c528 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -380,6 +380,18 @@ public final class CraftServer implements Server {
         return ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(((org.bukkit.craftbukkit.entity.CraftEntity) entity).getHandleRaw());
     }
 
+    // Canvas start - expand region threading api
+    @Override
+    public boolean isOwnedByCurrentRegion(@org.jspecify.annotations.NonNull final World world, final org.bukkit.util.@org.jspecify.annotations.NonNull BoundingBox aabb) {
+        return ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(((CraftWorld) world).getHandle(), new net.minecraft.world.phys.AABB(aabb.getMinX(), aabb.getMinY(), aabb.getMinZ(), aabb.getMaxX(), aabb.getMaxY(), aabb.getMaxZ()));
+    }
+
+    @Override
+    public boolean isOwnedByCurrentRegion(@org.jspecify.annotations.NonNull final Location position, final org.bukkit.util.@org.jspecify.annotations.NonNull Vector velocity, final int buffer) {
+        return ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(((CraftWorld) position.getWorld()).getHandle(), CraftLocation.toVec3(position), org.bukkit.craftbukkit.util.CraftVector.toVec3(velocity), buffer);
+    }
+
+    // Canvas end - expand region threading api
     @Override
     public final boolean isGlobalTickThread() {
         return io.papermc.paper.threadedregions.RegionizedServer.isGlobalTickThread(); // Folia - region threading API
@@ -3156,4 +3168,24 @@ public final class CraftServer implements Server {
         }
     }
     // Folia end - region TPS API
+    // Canvas start - region threading
+    // note: in impl the world is provided, same as 'getWorld()'
+    @Override
+    public <T> io.canvasmc.canvas.region.RegionTickData.@NotNull IRegionizedData<T> createRegionizedData(java.util.function.@NotNull BiFunction<io.canvasmc.canvas.region.RegionTickData, org.bukkit.World, T> supplier, io.canvasmc.canvas.region.RegionTickData.IRegionizedData.IRegionizedCallback<T> callback) {
+        return io.papermc.paper.threadedregions.TickRegions.TickRegionData.createRegionizedDataApiHandle(supplier, callback);
+    }
+
+    @Override
+    public <T> T getLocalRegionizedData(final io.canvasmc.canvas.region.RegionTickData.IRegionizedData<T> regionizedData) {
+        io.papermc.paper.threadedregions.ThreadedRegionizer.ThreadedRegion<io.papermc.paper.threadedregions.TickRegions.TickRegionData, io.papermc.paper.threadedregions.TickRegions.TickRegionSectionData>
+            region = io.papermc.paper.threadedregions.TickRegionScheduler.getCurrentRegion();
+        if (region == null) throw new IllegalStateException("Must be on a tick thread runner that currently owns a region");
+        return region.getData().getOrCreateFromIRegionizedData(regionizedData);
+    }
+
+    @Override
+    public io.canvasmc.canvas.region.RegionThreadingTickManager getRegionThreadingTickRateManager() {
+        return io.canvasmc.canvas.tick.ServerRegionTickManager.API_INSTANCE;
+    }
+    // Canvas end - region threading
 }
