--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -237,7 +_,7 @@
             if (chunk == null) {
                 continue;
             }
-            size += chunk.blockEntities.size();
+            size += chunk.canvas$getBlockEntitiesCount(); // Canvas - optimize block entity fetching
         }
         return size;
     }
