--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -2431,6 +_,13 @@
 
         return this.getStructures(x, z, struct -> registry.getKey(struct).equals(key));
     }
+    // Canvas start
+
+    @Override
+    public @NotNull io.canvasmc.canvas.LevelAccess getLevelAccess() {
+        return this.getHandle();
+    }
+    // Canvas end
 
     private List<GeneratedStructure> getStructures(int x, int z, Predicate<net.minecraft.world.level.levelgen.structure.Structure> predicate) {
         List<GeneratedStructure> structures = new ArrayList<>();
