From 37c0802d8eb83234e7cd8fab3253d198d2d53b39 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas paperServer File Patches


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
index c8d179d936d57e8222d87b7fd943dbddfa8446a0..39f38267614c74003e4a99bc308a3c6e254a0a76 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
@@ -136,10 +136,13 @@ public class CraftChunk implements Chunk {
         int index = 0;
         ChunkAccess chunk = this.getHandle(ChunkStatus.FULL);
 
-        BlockState[] states = new BlockState[chunk.blockEntities.size()];
-        for (BlockPos pos : chunk.blockEntities.keySet()) {
-            states[index++] = CraftBlock.at(this.level, pos).getState(useSnapshot);
+        // Canvas start - optimize block entity fetching
+        net.minecraft.world.level.block.entity.BlockEntity[] blockEntities = chunk.canvas$getAllBlockEntities();
+        BlockState[] states = new BlockState[blockEntities.length];
+        for (net.minecraft.world.level.block.entity.BlockEntity blockEntity : blockEntities) {
+            states[index++] = CraftBlock.at(this.level, blockEntity.getBlockPos()).getState(useSnapshot);
         }
+        // Canvas end - optimize block entity fetching
 
         return states;
     }
@@ -154,8 +157,11 @@ public class CraftChunk implements Chunk {
         ChunkAccess chunk = this.getHandle(ChunkStatus.FULL);
 
         List<BlockState> states = new ArrayList<>();
-        for (BlockPos pos : chunk.blockEntities.keySet()) {
-            Block block = CraftBlock.at(this.level, pos);
+        // Canvas start - optimize block entity fetching
+        net.minecraft.world.level.block.entity.BlockEntity[] blockEntities = chunk.canvas$getAllBlockEntities();
+        for (net.minecraft.world.level.block.entity.BlockEntity blockEntity : blockEntities) {
+            Block block = CraftBlock.at(this.level, blockEntity.getBlockPos());
+        // Canvas end - optimize block entity fetching
             if (blockPredicate.test(block)) {
                 states.add(block.getState(useSnapshot));
             }
