--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -406,7 +_,7 @@
         }
         final java.util.concurrent.CompletableFuture<ChunkAccess> future = new java.util.concurrent.CompletableFuture<>();
         ca.spottedleaf.moonrise.common.PlatformHooks.get().scheduleChunkLoad(
-            this.world, x, z, false, ChunkStatus.EMPTY, true, ca.spottedleaf.concurrentutil.util.Priority.NORMAL, future::complete
+            this.world, x, z, false, ChunkStatus.EMPTY, true, ca.spottedleaf.concurrentutil.util.Priority.BLOCKING, future::complete // Canvas - use blocking priority
         );
         world.getChunkSource().mainThreadProcessor.managedBlock(future::isDone);
         return future.thenApply(c -> {
@@ -750,11 +_,11 @@
 
     @Override
     public boolean generateTree(Location loc, TreeType type, BlockChangeDelegate delegate) {
-        this.world.captureTreeGeneration = true;
-        this.world.captureBlockStates = true;
+        this.world.captureTreeGeneration.set(true); // Canvas - atomic
+        this.world.captureBlockStates.set(true); // Canvas - atomic
         boolean grownTree = this.generateTree(loc, type);
-        this.world.captureBlockStates = false;
-        this.world.captureTreeGeneration = false;
+        this.world.captureBlockStates.set(false); // Canvas - atomic
+        this.world.captureTreeGeneration.set(false); // Canvas - atomic
         if (grownTree) { // Copy block data to delegate
             for (BlockState blockstate : this.world.capturedBlockStates.values()) {
                 BlockPos position = ((CraftBlockState) blockstate).getPosition();
@@ -2404,6 +_,13 @@
 
         return this.getStructures(x, z, struct -> registry.getKey(struct).equals(key));
     }
+    // Canvas start
+
+    @Override
+    public @NotNull io.canvasmc.canvas.LevelAccess getLevelAccess() {
+        return this.getHandle();
+    }
+    // Canvas end
 
     private List<GeneratedStructure> getStructures(int x, int z, Predicate<net.minecraft.world.level.levelgen.structure.Structure> predicate) {
         List<GeneratedStructure> structures = new ArrayList<>();
