--- a/net/minecraft/world/item/equipment/ArmorMaterials.java
+++ b/net/minecraft/world/item/equipment/ArmorMaterials.java
@@ -25,7 +_,7 @@
         25, makeDefense(2, 5, 6, 2, 5), 9, SoundEvents.ARMOR_EQUIP_TURTLE, 0.0F, 0.0F, ItemTags.REPAIRS_TURTLE_HELMET, EquipmentAssets.TURTLE_SCUTE
     );
     ArmorMaterial NETHERITE = new ArmorMaterial(
-        37, makeDefense(3, 6, 8, 3, 11), 15, SoundEvents.ARMOR_EQUIP_NETHERITE, 3.0F, 0.1F, ItemTags.REPAIRS_NETHERITE_ARMOR, EquipmentAssets.NETHERITE
+        37, makeDefense(3, 6, 8, 3, 11), 15, SoundEvents.ARMOR_EQUIP_NETHERITE, 3.0F, io.canvasmc.canvas.Config.INSTANCE.combat.ignoreNetheriteKnockbackResistance ? 0.0F : 0.1F, ItemTags.REPAIRS_NETHERITE_ARMOR, EquipmentAssets.NETHERITE // Canvas - combat configs
     );
     ArmorMaterial ARMADILLO_SCUTE = new ArmorMaterial(
         4, makeDefense(3, 6, 8, 3, 11), 10, SoundEvents.ARMOR_EQUIP_WOLF, 0.0F, 0.0F, ItemTags.REPAIRS_WOLF_ARMOR, EquipmentAssets.ARMADILLO_SCUTE
