From a1940d59a7eeb8b7dee60550bc75804e3611cb1d Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/item/equipment/ArmorMaterials.java b/net/minecraft/world/item/equipment/ArmorMaterials.java
index 421ad7772c9a6d562182016eacc34b1633c02301..c6d913079b7aa67842d08b07c02c5806fccc9d64 100644
--- a/net/minecraft/world/item/equipment/ArmorMaterials.java
+++ b/net/minecraft/world/item/equipment/ArmorMaterials.java
@@ -28,7 +28,7 @@ public interface ArmorMaterials {
         25, makeDefense(2, 5, 6, 2, 5), 9, SoundEvents.ARMOR_EQUIP_TURTLE, 0.0F, 0.0F, ItemTags.REPAIRS_TURTLE_HELMET, EquipmentAssets.TURTLE_SCUTE
     );
     ArmorMaterial NETHERITE = new ArmorMaterial(
-        37, makeDefense(3, 6, 8, 3, 19), 15, SoundEvents.ARMOR_EQUIP_NETHERITE, 3.0F, 0.1F, ItemTags.REPAIRS_NETHERITE_ARMOR, EquipmentAssets.NETHERITE
+        37, makeDefense(3, 6, 8, 3, 19), 15, SoundEvents.ARMOR_EQUIP_NETHERITE, 3.0F, io.canvasmc.canvas.Config.INSTANCE.combat.ignoreNetheriteKnockbackResistance ? 0.0F : 0.1F/*TODO-make this configurable more..?*/, ItemTags.REPAIRS_NETHERITE_ARMOR, EquipmentAssets.NETHERITE // Canvas - combat configs
     );
     ArmorMaterial ARMADILLO_SCUTE = new ArmorMaterial(
         4, makeDefense(3, 6, 8, 3, 11), 10, SoundEvents.ARMOR_EQUIP_WOLF, 0.0F, 0.0F, ItemTags.REPAIRS_WOLF_ARMOR, EquipmentAssets.ARMADILLO_SCUTE
