--- a/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
+++ b/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
@@ -126,6 +_,23 @@
         super.inactiveTick();
     }
     // Paper end - EAR 2
+    // Canvas start - MC-261810
+
+    @Override
+    protected void onBelowWorld() {
+        if (
+            io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaFixes.mc261810 &&
+            this.attachedToEntity != null &&
+            ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.attachedToEntity)
+        ) {
+            // we shouldn't discard the entity if the attached entity is still
+            // in the same tick region and we are below the void, otherwise the
+            // firework gets discarded and the boost stops like 1 tick in
+            return;
+        }
+        super.onBelowWorld();
+    }
+    // Canvas end - MC-261810
 
     @Override
     public void tick() {
@@ -147,7 +_,7 @@
 
             if (this.attachedToEntity != null) {
                 Vec3 handAngle;
-                if (this.attachedToEntity.isFallFlying()) {
+                if (this.attachedToEntity.isFallFlying() && (!io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaFixes.mc119754 || !this.attachedToEntity.isSpectator())) { // Canvas - MC-119754
                     Vec3 lookAngle = this.attachedToEntity.getLookAngle();
                     double power = 1.5;
                     double powerAdd = 0.1;
