From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/network/chat/SignedMessageChain.java b/net/minecraft/network/chat/SignedMessageChain.java
index 9395c6a58ba37fb00c492065b8045f54de0a2246..c7bef2318a68caae0be5d57dfcdc53927849a805 100644
--- a/net/minecraft/network/chat/SignedMessageChain.java
+++ b/net/minecraft/network/chat/SignedMessageChain.java
@@ -49,7 +49,7 @@ public class SignedMessageChain {
                     throw new SignedMessageChain.DecodeException(SignedMessageChain.DecodeException.CHAIN_BROKEN);
                 }
 
-                if (body.timeStamp().isBefore(SignedMessageChain.this.lastTimeStamp)) {
+                if (!io.canvasmc.canvas.GlobalConfiguration.getInstance().chat.disableChatVerificationOrder && body.timeStamp().isBefore(SignedMessageChain.this.lastTimeStamp)) { // Canvas - allow disabling chat verification order
                     this.setChainBroken();
                     throw new SignedMessageChain.DecodeException(SignedMessageChain.DecodeException.OUT_OF_ORDER_CHAT, org.bukkit.event.player.PlayerKickEvent.Cause.OUT_OF_ORDER_CHAT); // Paper - kick event causes
                 }
