From 53e1a57c72953de1bfcc94c3257c724e7fa1f88f Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/item/crafting/ShapelessRecipe.java b/net/minecraft/world/item/crafting/ShapelessRecipe.java
index 53a818f14709d1978dbbc4aa7367be733a8ee47f..786d512ce9820f1a552b750a4d1c1b9bf3a68941 100644
--- a/net/minecraft/world/item/crafting/ShapelessRecipe.java
+++ b/net/minecraft/world/item/crafting/ShapelessRecipe.java
@@ -38,10 +38,18 @@ public class ShapelessRecipe extends NormalCraftingRecipe {
     public static final RecipeSerializer<ShapelessRecipe> SERIALIZER = new RecipeSerializer<>(MAP_CODEC, STREAM_CODEC);
     private final ItemStackTemplate result;
     private final List<Ingredient> ingredients;
+    private final boolean isBukkit; // Pufferfish
 
+    // Pufferfish start
     public ShapelessRecipe(
         final Recipe.CommonInfo commonInfo, final CraftingRecipe.CraftingBookInfo bookInfo, final ItemStackTemplate result, final List<Ingredient> ingredients
     ) {
+        this(commonInfo, bookInfo, result, ingredients, false);
+    }
+
+    public ShapelessRecipe(
+        final Recipe.CommonInfo commonInfo, final CraftingRecipe.CraftingBookInfo bookInfo, final ItemStackTemplate result, final List<Ingredient> ingredients, final boolean isBukkit
+    ) { this.isBukkit = isBukkit; // Pufferfish end
         super(commonInfo, bookInfo);
         this.result = result;
         this.ingredients = ingredients;
@@ -74,6 +82,27 @@ public class ShapelessRecipe extends NormalCraftingRecipe {
 
     @Override
     public boolean matches(final CraftingInput input, final Level level) {
+        // Pufferfish start
+        if (!this.isBukkit) {
+            java.util.List<Ingredient> ingredients = com.google.common.collect.Lists.newArrayList(this.ingredients.toArray(new Ingredient[0]));
+
+            inventory: for (int index = 0; index < input.size(); index++) {
+                ItemStack item = input.getItem(index);
+
+                if (!item.isEmpty()) {
+                    for (int count = 0; count < ingredients.size(); count++) {
+                        if (ingredients.get(count).test(item)) {
+                            ingredients.remove(count);
+                            continue inventory;
+                        }
+                    }
+                    return false;
+                }
+            }
+
+            return ingredients.isEmpty();
+        }
+        // Pufferfish end
         // Paper start - Improve exact choice recipe ingredients & unwrap ternary
         if (input.ingredientCount() != this.ingredients.size()) {
             return false;
