From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/item/ItemStack.java b/net/minecraft/world/item/ItemStack.java
index 4e483fcf80d8583e00ad75c18557de8036989cc2..834d52db7a371c02c99d099ada83b3127b787fa6 100644
--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -865,11 +865,11 @@ public final class ItemStack implements DataComponentHolder, ItemInstance {
     }
 
     public static boolean isSameItem(final ItemStack a, final ItemStack b) {
-        return a.is(b.getItem());
+        return a == b || a.is(b.getItem()); // Canvas - optimize matching checks
     }
 
     public static boolean isSameItemSameComponents(final ItemStack a, final ItemStack b) {
-        return a.is(b.getItem()) && (a.isEmpty() && b.isEmpty() || Objects.equals(a.components, b.components));
+        return a == b || a.is(b.getItem()) && (a.isEmpty() && b.isEmpty() || Objects.equals(a.components, b.components)); // Canvas - optimize matching checks
     }
 
     public static boolean matchesIgnoringComponents(final ItemStack a, final ItemStack b, final Predicate<DataComponentType<?>> ignoredPredicate) {
