From 53e1a57c72953de1bfcc94c3257c724e7fa1f88f Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/item/ItemStack.java b/net/minecraft/world/item/ItemStack.java
index 0a79d1f25f07459847f2eda2f5da8eb9dfa46d60..585f6caa9627236c4a1f83949abe1bb38aba1b7c 100644
--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -865,11 +865,11 @@ public final class ItemStack implements DataComponentHolder, ItemInstance {
     }
 
     public static boolean isSameItem(final ItemStack a, final ItemStack b) {
-        return a.is(b.getItem());
+        return a == b || a.is(b.getItem()); // Canvas - optimize matching checks
     }
 
     public static boolean isSameItemSameComponents(final ItemStack a, final ItemStack b) {
-        return a.is(b.getItem()) && (a.isEmpty() && b.isEmpty() || Objects.equals(a.components, b.components));
+        return a == b || a.is(b.getItem()) && (a.isEmpty() && b.isEmpty() || Objects.equals(a.components, b.components)); // Canvas - optimize matching checks
     }
 
     public static boolean matchesIgnoringComponents(final ItemStack a, final ItemStack b, final Predicate<DataComponentType<?>> ignoredPredicate) {
