From 23441ae46065042fd5e624f07307c4cb48391dc3 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/item/enchantment/Enchantment.java b/net/minecraft/world/item/enchantment/Enchantment.java
index 0d981b86fa7ee8713e2a419fc2f377defbc2ed9c..80657e0f55c8c608e0ad45f7a785af668c0abd38 100644
--- a/net/minecraft/world/item/enchantment/Enchantment.java
+++ b/net/minecraft/world/item/enchantment/Enchantment.java
@@ -124,7 +124,15 @@ public record Enchantment(Component description, Enchantment.EnchantmentDefiniti
     }
 
     public boolean matchingSlot(final EquipmentSlot slot) {
-        return this.definition.slots().stream().anyMatch(group -> group.test(slot));
+        // Canvas start - remove streams
+        for (EquipmentSlotGroup equipmentSlotGroup : this.definition.slots()) {
+            if (equipmentSlotGroup.test(slot)) {
+                return true;
+            }
+        }
+
+        return false;
+        // Canvas end - remove streams
     }
 
     public boolean isPrimaryItem(final ItemStack item) {
