--- a/net/minecraft/world/item/enchantment/Enchantment.java
+++ b/net/minecraft/world/item/enchantment/Enchantment.java
@@ -131,7 +_,15 @@
     }
 
     public boolean matchingSlot(EquipmentSlot slot) {
-        return this.definition.slots().stream().anyMatch(equipmentSlotGroup -> equipmentSlotGroup.test(slot));
+        // Canvas start - remove streams
+        for (EquipmentSlotGroup equipmentSlotGroup : this.definition.slots()) {
+            if (equipmentSlotGroup.test(slot)) {
+                return true;
+            }
+        }
+
+        return false;
+        // Canvas end - remove streams
     }
 
     public boolean isPrimaryItem(ItemStack stack) {
