--- a/ca/spottedleaf/moonrise/patches/chunk_system/io/MoonriseRegionFileIO.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/io/MoonriseRegionFileIO.java
@@ -167,6 +_,7 @@
 
     public static void flush(final MinecraftServer server) {
         for (final ServerLevel world : server.getAllLevels()) {
+            LOGGER.info("Flushing I/O for level {}", world.name()); // Canvas - debug flushing
             flush(world);
         }
     }
