--- a/net/minecraft/world/item/crafting/ShapelessRecipe.java
+++ b/net/minecraft/world/item/crafting/ShapelessRecipe.java
@@ -38,10 +_,18 @@
     public static final RecipeSerializer<ShapelessRecipe> SERIALIZER = new RecipeSerializer<>(MAP_CODEC, STREAM_CODEC);
     private final ItemStackTemplate result;
     private final List<Ingredient> ingredients;
+    private final boolean isBukkit; // Pufferfish
 
+    // Pufferfish start
     public ShapelessRecipe(
         final Recipe.CommonInfo commonInfo, final CraftingRecipe.CraftingBookInfo bookInfo, final ItemStackTemplate result, final List<Ingredient> ingredients
     ) {
+        this(commonInfo, bookInfo, result, ingredients, false);
+    }
+
+    public ShapelessRecipe(
+        final Recipe.CommonInfo commonInfo, final CraftingRecipe.CraftingBookInfo bookInfo, final ItemStackTemplate result, final List<Ingredient> ingredients, final boolean isBukkit
+    ) { this.isBukkit = isBukkit; // Pufferfish end
         super(commonInfo, bookInfo);
         this.result = result;
         this.ingredients = ingredients;
@@ -74,6 +_,27 @@
 
     @Override
     public boolean matches(final CraftingInput input, final Level level) {
+        // Pufferfish start
+        if (!this.isBukkit) {
+            java.util.List<Ingredient> ingredients = com.google.common.collect.Lists.newArrayList(this.ingredients.toArray(new Ingredient[0]));
+
+            inventory: for (int index = 0; index < input.size(); index++) {
+                ItemStack item = input.getItem(index);
+
+                if (!item.isEmpty()) {
+                    for (int count = 0; count < ingredients.size(); count++) {
+                        if (ingredients.get(count).test(item)) {
+                            ingredients.remove(count);
+                            continue inventory;
+                        }
+                    }
+                    return false;
+                }
+            }
+
+            return ingredients.isEmpty();
+        }
+        // Pufferfish end
         if (input.ingredientCount() != this.ingredients.size()) {
             return false;
         } else {
