--- a/net/minecraft/world/level/block/piston/PistonBaseBlock.java
+++ b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
@@ -139,7 +_,7 @@
                 && pistonMovingBlockEntity.isExtending()
                 && (
                     pistonMovingBlockEntity.getProgress(0.0F) < 0.5F
-                        || level.getGameTime() == pistonMovingBlockEntity.getLastTicked()
+                        || level.getRedstoneGameTime() == pistonMovingBlockEntity.getLastTicked() // Canvas - Threaded Regions
                         || ((ServerLevel)level).isHandlingTick()
                 )) {
                 i = 2;
