From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/data/advancements/packs/VanillaNetherAdvancements.java b/net/minecraft/data/advancements/packs/VanillaNetherAdvancements.java
index 262de32bfbb0c766f353839cf97cd9bdf9e8123d..b1effb554b0c8d7973dd2cf2cfcdd44102bd7d37 100644
--- a/net/minecraft/data/advancements/packs/VanillaNetherAdvancements.java
+++ b/net/minecraft/data/advancements/packs/VanillaNetherAdvancements.java
@@ -151,7 +151,7 @@ public class VanillaNetherAdvancements implements AdvancementSubProvider {
             .addCriterion(
                 "killed_ghast",
                 KilledTrigger.TriggerInstance.playerKilledEntity(
-                    EntityPredicate.Builder.entity().of(entityTypes, EntityType.GHAST).located(LocationPredicate.Builder.inDimension(Level.OVERWORLD))
+                    EntityPredicate.Builder.entity().of(entityTypes, EntityType.GHAST).located(LocationPredicate.Builder.inDimension(io.canvasmc.canvas.GlobalConfiguration.fetchRespawnDimensionKey())) // Canvas - default respawn dimension config - this is where the nether portal ties to, so we should change this to match that
                 )
             )
             .save(output, "nether/uneasy_alliance");
@@ -442,7 +442,7 @@ public class VanillaNetherAdvancements implements AdvancementSubProvider {
                 "ride_entity_distance",
                 DistanceTrigger.TriggerInstance.rideEntityInLava(
                     EntityPredicate.Builder.entity()
-                        .located(LocationPredicate.Builder.inDimension(Level.OVERWORLD))
+                        .located(LocationPredicate.Builder.inDimension(io.canvasmc.canvas.GlobalConfiguration.fetchRespawnDimensionKey())) // Canvas - default respawn dimension config - this is where the nether portal ties to, so we should change this to match that
                         .vehicle(EntityPredicate.Builder.entity().of(entityTypes, EntityType.STRIDER)),
                     DistancePredicate.horizontal(MinMaxBounds.Doubles.atLeast(50.0))
                 )
