From 0790c032210252a6406b62762400f31b9c07fa8d Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas paperServer File Patches


diff --git a/src/main/java/io/papermc/paper/antixray/ChunkPacketBlockControllerAntiXray.java b/src/main/java/io/papermc/paper/antixray/ChunkPacketBlockControllerAntiXray.java
index 582937bbb8b5eb7372205d62be6522206fa789d5..f30b6a7a3d3588bb0554e9e0fa9d9a7bcd046dff 100644
--- a/src/main/java/io/papermc/paper/antixray/ChunkPacketBlockControllerAntiXray.java
+++ b/src/main/java/io/papermc/paper/antixray/ChunkPacketBlockControllerAntiXray.java
@@ -186,15 +186,20 @@ public final class ChunkPacketBlockControllerAntiXray extends ChunkPacketBlockCo
             return;
         }
 
+        // Canvas start - region threading
+        // from below
+        LevelChunk chunk = chunkPacketInfo.getChunk();
+        int x = chunk.getPos().x();
+        int z = chunk.getPos().z();
         if (!Bukkit.isPrimaryThread()) {
             // Plugins?
-            MinecraftServer.getServer().scheduleOnMain(() -> modifyBlocks(chunkPacket, chunkPacketInfo));
+            io.papermc.paper.threadedregions.RegionizedServer.getInstance().taskQueue.queueChunkTask(
+                chunk.level, x, z, () -> modifyBlocks(chunkPacket, chunkPacketInfo));
+            // Canvas end - region threading
             return;
         }
 
-        LevelChunk chunk = chunkPacketInfo.getChunk();
-        int x = chunk.getPos().x();
-        int z = chunk.getPos().z();
+        // Canvas - region threading - move up
         Level level = chunk.getLevel();
         ((ChunkPacketInfoAntiXray) chunkPacketInfo).setNearbyChunks(level.getChunkIfLoaded(x - 1, z), level.getChunkIfLoaded(x + 1, z), level.getChunkIfLoaded(x, z - 1), level.getChunkIfLoaded(x, z + 1));
         executor.execute((Runnable) chunkPacketInfo);
