--- a/net/minecraft/world/level/block/CropBlock.java
+++ b/net/minecraft/world/level/block/CropBlock.java
@@ -164,6 +_,7 @@
     }
 
     protected static boolean hasSufficientLight(final LevelReader level, final BlockPos pos) {
+        if (level instanceof ServerLevel serverLevel && serverLevel.canvasConfig().farming.cropsIgnoreLightCheck) return true; // Canvas - crops ignore light check config
         return level.getRawBrightness(pos, 0) >= 8;
     }
 
