--- a/net/minecraft/world/level/block/SpreadingSnowyBlock.java
+++ b/net/minecraft/world/level/block/SpreadingSnowyBlock.java
@@ -64,7 +_,7 @@
         Optional<Block> baseBlock = blocks.getOptional(this.baseBlock);
         if (!baseBlock.isEmpty()) {
             // Paper start - Perf: optimize dirt and snow spreading
-            final net.minecraft.world.level.chunk.ChunkAccess cachedChunk = level.getChunkIfLoaded(pos);
+            final net.minecraft.world.level.chunk.LevelChunk cachedChunk = level.getChunkIfLoaded(pos); // Canvas - optimize random tick
             if (cachedChunk == null) { // Is this needed?
                 return;
             }
@@ -78,24 +_,27 @@
                 // CraftBukkit end
                 level.setBlockAndUpdate(pos, baseBlock.get().defaultBlockState());
             } else {
-                if (level.getMaxLocalRawBrightness(pos.above()) >= 9) {
+                if (level.canvasConfig().disableGrassLightChecks || level.getMaxLocalRawBrightness(pos.above()) >= 9) { // Canvas - disable grass light checks config
                     BlockState defaultBlockState = this.defaultBlockState();
 
                     for (int i = 0; i < 4; i++) {
                         BlockPos testPos = pos.offset(random.nextInt(3) - 1, random.nextInt(5) - 3, random.nextInt(3) - 1);
                         // Paper start - Perf: optimize dirt and snow spreading
-                        if (pos.getX() == testPos.getX() && pos.getY() == testPos.getY() && pos.getZ() == testPos.getZ()) {
+                        if (pos == testPos) { // Canvas - optimize random tick
                             continue;
                         }
 
-                        final net.minecraft.world.level.chunk.ChunkAccess access;
+                        final net.minecraft.world.level.chunk.LevelChunk access; // Canvas - optimize random tick
                         if (cachedChunk.locX == testPos.getX() >> 4 && cachedChunk.locZ == testPos.getZ() >> 4) {
                             access = cachedChunk;
                         } else {
-                            access = level.getChunkAt(testPos);
+                            // Canvas start - optimize random tick
+                            access = level.getChunkIfLoaded(testPos);
+                            if (access == null) continue;
                         }
-                        if (access.getBlockState(testPos).is(baseBlock.get()) && SpreadingSnowyBlock.canPropagate(access, defaultBlockState, testPos)) {
-                            org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(level, pos, testPos, defaultBlockState.setValue(SNOWY, isSnowySetting(access.getBlockState(testPos.above()))), Block.UPDATE_ALL); // CraftBukkit
+                        if (access.getBlockStateFinal(testPos.getX(), testPos.getY(), testPos.getZ()).is(baseBlock.get()) && SpreadingSnowyBlock.canPropagate(access, defaultBlockState, testPos)) {
+                            org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(level, pos, testPos, defaultBlockState.setValue(SNOWY, isSnowySetting(access.getBlockStateFinal(testPos.getX(), testPos.getY() + 1, testPos.getZ()))), Block.UPDATE_ALL); // CraftBukkit
+                            // Canvas end - optimize random tick
                             // Paper end - Perf: optimize dirt and snow spreading
                         }
                     }
