--- a/net/minecraft/world/level/redstone/CollectingNeighborUpdater.java
+++ b/net/minecraft/world/level/redstone/CollectingNeighborUpdater.java
@@ -1,10 +_,14 @@
 package net.minecraft.world.level.redstone;
 
 import com.mojang.logging.LogUtils;
-import java.util.ArrayDeque;
 import java.util.ArrayList;
+import java.util.Deque;
 import java.util.List;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.atomic.AtomicInteger;
 import javax.annotation.Nullable;
+import io.canvasmc.canvas.Config;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.world.flag.FeatureFlags;
@@ -17,13 +_,13 @@
     private static final Logger LOGGER = LogUtils.getLogger();
     private final Level level;
     private final int maxChainedNeighborUpdates;
-    private final ArrayDeque<CollectingNeighborUpdater.NeighborUpdates> stack = new ArrayDeque<>();
-    private final List<CollectingNeighborUpdater.NeighborUpdates> addedThisLayer = new ArrayList<>();
+    private final Deque<NeighborUpdates> stack = new ConcurrentLinkedDeque<>(); // Canvas - fix CME/NPE
+    private final List<CollectingNeighborUpdater.NeighborUpdates> addedThisLayer = new CopyOnWriteArrayList<>(); // Canvas - use concurrent
     private int count = 0;
 
     public CollectingNeighborUpdater(Level level, int maxChainedNeighborUpdates) {
         this.level = level;
-        this.maxChainedNeighborUpdates = maxChainedNeighborUpdates;
+        this.maxChainedNeighborUpdates = Config.INSTANCE.maxChainedNeighborUpdates; // Canvas - configurable max chained neighbor updates
     }
 
     @Override
@@ -57,7 +_,7 @@
                 this.stack.push(updates);
             }
         } else if (this.count - 1 == this.maxChainedNeighborUpdates) {
-            LOGGER.error("Too many chained neighbor updates. Skipping the rest. First skipped position: " + pos.toShortString());
+            LOGGER.error("Too many chained neighbor updates. Skipping the rest. First skipped position: {}", pos.toShortString(), new Throwable()); // Canvas - debug chained neighbor updates
         }
 
         if (!flag) {
@@ -75,6 +_,7 @@
                 this.addedThisLayer.clear();
                 CollectingNeighborUpdater.NeighborUpdates neighborUpdates = this.stack.peek();
 
+                if (neighborUpdates == null) continue; // Canvas - ensure not null
                 while (this.addedThisLayer.isEmpty()) {
                     if (!neighborUpdates.runNext(this.level)) {
                         this.stack.pop();
@@ -105,21 +_,23 @@
         private Orientation orientation;
         @Nullable
         private final Direction skipDirection;
-        private int idx = 0;
+        private final AtomicInteger idx = new AtomicInteger(); // Canvas - atomic
 
         MultiNeighborUpdate(BlockPos sourcePos, Block sourceBlock, @Nullable Orientation orientation, @Nullable Direction skipDirection) {
             this.sourcePos = sourcePos;
             this.sourceBlock = sourceBlock;
             this.orientation = orientation;
+            // Canvas start - atomic
             this.skipDirection = skipDirection;
-            if (NeighborUpdater.UPDATE_ORDER[this.idx] == skipDirection) {
-                this.idx++;
+            if (NeighborUpdater.UPDATE_ORDER[this.idx.get()] == skipDirection) {
+            // Canvas end
+                this.idx.getAndIncrement();
             }
         }
 
         @Override
         public boolean runNext(Level level) {
-            Direction direction = NeighborUpdater.UPDATE_ORDER[this.idx++];
+            Direction direction = NeighborUpdater.UPDATE_ORDER[this.idx.getAndIncrement()]; // Canvas - atomic
             BlockPos blockPos = this.sourcePos.relative(direction);
             BlockState blockState = level.getBlockState(blockPos);
             Orientation orientation = null;
@@ -134,11 +_,13 @@
             }
 
             NeighborUpdater.executeUpdate(level, blockState, blockPos, this.sourceBlock, orientation, false, this.sourcePos); // Paper - Add source block to BlockPhysicsEvent
-            if (this.idx < NeighborUpdater.UPDATE_ORDER.length && NeighborUpdater.UPDATE_ORDER[this.idx] == this.skipDirection) {
-                this.idx++;
+            // Canvas start - atomic
+            if (this.idx.get() < NeighborUpdater.UPDATE_ORDER.length && NeighborUpdater.UPDATE_ORDER[this.idx.get()] == this.skipDirection) {
+                this.idx.getAndIncrement();
+            // Canvas end
             }
 
-            return this.idx < NeighborUpdater.UPDATE_ORDER.length;
+            return this.idx.get() < NeighborUpdater.UPDATE_ORDER.length; // Canvas - atomic
         }
     }
 
