--- a/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
@@ -37,6 +_,7 @@
 import net.minecraft.world.inventory.BeaconMenu;
 import net.minecraft.world.inventory.ContainerData;
 import net.minecraft.world.inventory.ContainerLevelAccess;
+import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.Level;
 import net.minecraft.world.level.block.BeaconBeamBlock;
 import net.minecraft.world.level.block.Blocks;
@@ -223,7 +_,7 @@
         }
 
         int i = blockEntity.levels; final int originalLevels = i; // Paper - OBFHELPER
-        if (level.getGameTime() % 80L == 0L) {
+        if (level.getRedstoneGameTime() % 80L == 0L) { // Canvas - Threaded Regions
             if (!blockEntity.beamSections.isEmpty()) {
                 blockEntity.levels = updateBase(level, x, y, z);
             }
@@ -357,7 +_,7 @@
             list = level.getEntitiesOfClass(Player.class, aabb); // Diff from applyEffect
         } else {
             list = new java.util.ArrayList<>();
-            for (final Player player : level.players()) {
+            for (final Player player : level.getLocalPlayers(new ChunkPos(pos))) { // Canvas - Threaded Regions
                 if (!net.minecraft.world.entity.EntitySelector.NO_SPECTATORS.test(player)) continue;
                 if (player.getBoundingBox().intersects(aabb)) {
                     list.add(player);
