--- a/net/minecraft/world/level/block/entity/BeehiveBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BeehiveBlockEntity.java
@@ -80,7 +_,7 @@
     private static final int MIN_TICKS_BEFORE_REENTERING_HIVE = 400;
     private static final int MIN_OCCUPATION_TICKS_NECTAR = 2400;
     public static final int MIN_OCCUPATION_TICKS_NECTARLESS = 600;
-    private List<BeehiveBlockEntity.BeeData> stored = Lists.newArrayList();
+    private List<BeehiveBlockEntity.BeeData> stored = Lists.newCopyOnWriteArrayList(); // Canvas - concurrent
     @Nullable
     public BlockPos savedFlowerPos;
     public int maxBees = MAX_OCCUPANTS; // CraftBukkit - allow setting max amount of bees a hive can hold
