--- a/net/minecraft/world/level/block/entity/BlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BlockEntity.java
@@ -39,6 +_,11 @@
     protected boolean remove;
     private BlockState blockState;
     private DataComponentMap components = DataComponentMap.EMPTY;
+    // Canvas start
+
+    public void lithium$handleSetChanged() {
+    }
+    // Canvas end
 
     public BlockEntity(BlockEntityType<?> type, BlockPos pos, BlockState blockState) {
         this.type = type;
@@ -217,6 +_,7 @@
             if (ignoreBlockEntityUpdates) return; // Paper - Perf: Optimize Hoppers
             setChanged(this.level, this.worldPosition, this.blockState);
         }
+        this.lithium$handleSetChanged(); // Canvas
     }
 
     protected static void setChanged(Level level, BlockPos pos, BlockState state) {
