--- a/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
@@ -39,7 +_,7 @@
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
-public class CampfireBlockEntity extends BlockEntity implements Clearable {
+public class CampfireBlockEntity extends BlockEntity implements Clearable, io.canvasmc.canvas.world.block.SleepingBlockEntity { // Canvas - block entity sleeping
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final int BURN_COOL_SPEED = 2;
     private static final int NUM_SLOTS = 4;
@@ -47,6 +_,31 @@
     public final int[] cookingProgress = new int[4];
     public final int[] cookingTime = new int[4];
     public final boolean[] stopCooking = new boolean[4]; // Paper - Add more Campfire API
+    // Canvas start - block entity sleeping
+    private net.minecraft.world.level.chunk.LevelChunk.@Nullable RebindableTickingBlockEntityWrapper tickWrapper = null;
+    private @Nullable TickingBlockEntity sleepingTicker = null;
+
+    @Override
+    public net.minecraft.world.level.chunk.LevelChunk.@Nullable RebindableTickingBlockEntityWrapper lithium$getTickWrapper() {
+        return this.tickWrapper;
+    }
+
+    @Override
+    public void lithium$setTickWrapper(net.minecraft.world.level.chunk.LevelChunk.@Nullable RebindableTickingBlockEntityWrapper tickWrapper) {
+        this.tickWrapper = tickWrapper;
+        this.lithium$setSleepingTicker(null);
+    }
+
+    @Override
+    public @Nullable TickingBlockEntity lithium$getSleepingTicker() {
+        return this.sleepingTicker;
+    }
+
+    @Override
+    public void lithium$setSleepingTicker(@Nullable TickingBlockEntity sleepingTicker) {
+        this.sleepingTicker = sleepingTicker;
+    }
+    // Canvas end - block entity sleeping
 
     public CampfireBlockEntity(final BlockPos worldPosition, final BlockState blockState) {
         super(BlockEntityType.CAMPFIRE, worldPosition, blockState);
@@ -111,6 +_,10 @@
 
         if (changed) {
             setChanged(level, pos, state);
+        // Canvas start - block entity sleeping
+        } else {
+            entity.lithium$startSleeping();
+        // Canvas end - block entity sleeping
         }
     }
 
@@ -126,6 +_,10 @@
 
         if (changed) {
             setChanged(level, pos, state);
+        // Canvas start - block entity sleeping
+        } else {
+            entity.lithium$startSleeping();
+        // Canvas end - block entity sleeping
         }
     }
 
@@ -183,6 +_,7 @@
             System.arraycopy(cookingState, 0, this.stopCooking, 0, Math.min(this.stopCooking.length, bytes.capacity()));
         });
         // Paper end - Add more Campfire API
+        this.wakeUpNow(); // Canvas - block entity sleeping
     }
 
     @Override
@@ -238,6 +_,7 @@
                 // CraftBukkit end
                 this.cookingProgress[slot] = 0;
                 this.items.set(slot, placeItem.consumeAndReturn(1, sourceEntity));
+                this.wakeUpNow(); // Canvas - block entity sleeping
                 serverLevel.gameEvent(GameEvent.BLOCK_CHANGE, this.getBlockPos(), GameEvent.Context.of(sourceEntity, this.getBlockState()));
                 this.markUpdated();
                 return true;
