--- a/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
@@ -217,7 +_,7 @@
                 teleport, net.minecraft.world.entity.Entity.TELEPORT_FLAG_LOAD_CHUNK | net.minecraft.world.entity.Entity.TELEPORT_FLAG_TELEPORT_PASSENGERS,
                 (net.minecraft.world.entity.Entity teleportedEntity) -> {
                     // adjust to the final exit position
-                    Vec3 adjusted = Vec3.atCenterOf(TheEndGatewayBlockEntity.findExitPosition(portalWorld, teleportPos));
+                    Vec3 adjusted = Vec3.atCenterOf(TheEndGatewayBlockEntity.findExitPosition(portalWorld, teleportedEntity.blockPosition())); // Canvas - region threading
                     // teleportTo will adjust rider positions
                     teleportedEntity.teleportTo(adjusted.x, adjusted.y, adjusted.z);
 
@@ -232,6 +_,12 @@
 
     @Nullable
     public Vec3 getPortalPosition(ServerLevel level, BlockPos pos) {
+        // Canvas start - ensure tick thread
+        if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(level, pos)) {
+            this.trySearchForExit(level, pos);
+            return null;
+        }
+        // Canvas end - ensure tick thread
         if (this.exitPortal == null && level.getTypeKey() == net.minecraft.world.level.dimension.LevelStem.END) { // CraftBukkit - work in alternate worlds
             BlockPos blockPos = findOrCreateValidTeleportPos(level, pos);
             blockPos = blockPos.above(10);
