From 53e1a57c72953de1bfcc94c3257c724e7fa1f88f Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
index 43bfa179a8363a1dedd8de083479b7a56965c918..010e34d5339723efdf8605d61d15942036ccae89 100644
--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
@@ -58,6 +58,12 @@ public class Snowball extends ThrowableItemProjectile {
         Entity entity = hitResult.getEntity();
         int damage = entity instanceof Blaze ? 3 : 0;
         entity.hurt(this.damageSources().thrown(this, this.getOwner()), damage);
+        // Canvas start - snowball can knockback player config
+        if (level().canvasConfig().combat.snowballCanKnockbackPlayers && entity instanceof net.minecraft.server.level.ServerPlayer serverPlayer) {
+            serverPlayer.knockback(0.4F, this.getX() - entity.getX(), this.getZ() - entity.getZ());
+            serverPlayer.hurtMarked = true;
+        }
+        // Canvas end - snowball can knockback player config
     }
 
     @Override
