From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
index 6e828818909df2b31349398639257b7a6f89fcd5..2a20da29d26499a8bde56eabc11e52457bfb8100 100644
--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
@@ -56,6 +56,12 @@ public class Snowball extends ThrowableItemProjectile {
         Entity entity = hitResult.getEntity();
         int damage = entity instanceof Blaze ? 3 : 0;
         entity.hurt(this.damageSources().thrown(this, this.getOwner()), damage);
+        // Canvas start - snowball can knockback player config
+        if (level().canvasConfig().combat.snowballCanKnockbackPlayers && entity instanceof net.minecraft.server.level.ServerPlayer serverPlayer) {
+            serverPlayer.knockback(0.4F, this.getX() - entity.getX(), this.getZ() - entity.getZ());
+            serverPlayer.hurtMarked = true;
+        }
+        // Canvas end - snowball can knockback player config
     }
 
     @Override
