From 53e1a57c72953de1bfcc94c3257c724e7fa1f88f Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEgg.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEgg.java
index 79c9818999759302138f195c93b3245916243014..25935bdb66157aeb2135c4265968511f5cf26e2f 100644
--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEgg.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEgg.java
@@ -60,6 +60,12 @@ public class ThrownEgg extends ThrowableItemProjectile {
     protected void onHitEntity(final EntityHitResult hitResult) {
         super.onHitEntity(hitResult);
         hitResult.getEntity().hurt(this.damageSources().thrown(this, this.getOwner()), 0.0F);
+        // Canvas start - egg can knockback player config
+        if (level().canvasConfig().combat.eggCanKnockbackPlayers && hitResult.getEntity() instanceof net.minecraft.server.level.ServerPlayer serverPlayer) {
+            serverPlayer.knockback(0.4F, this.getX() - hitResult.getEntity().getX(), this.getZ() - hitResult.getEntity().getZ());
+            serverPlayer.hurtMarked = true;
+        }
+        // Canvas end - egg can knockback player config
     }
 
     @Override
