From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/projectile/FishingHook.java b/net/minecraft/world/entity/projectile/FishingHook.java
index e39cbe227e1308d34d84f2097049d869e2ce7e93..0cd700a4e9d5a5b514a7f8c2f02b8fd868818452 100644
--- a/net/minecraft/world/entity/projectile/FishingHook.java
+++ b/net/minecraft/world/entity/projectile/FishingHook.java
@@ -516,6 +516,11 @@ public class FishingHook extends Projectile {
         if (!this.level().isClientSide() && playerOwner != null && !this.shouldStopFishing(playerOwner)) {
             int i = 0;
             if (this.hookedIn != null) {
+                // Canvas start - combat configs
+                if (!io.canvasmc.canvas.Config.INSTANCE.combat.fishingRodPulls) {
+                    return 0; // act as if we canceled the pull
+                }
+                // Canvas end - combat configs
                 // CraftBukkit start
                 org.bukkit.event.player.PlayerFishEvent playerFishEvent = new org.bukkit.event.player.PlayerFishEvent((org.bukkit.entity.Player) playerOwner.getBukkitEntity(), this.hookedIn.getBukkitEntity(), (org.bukkit.entity.FishHook) this.getBukkitEntity(), org.bukkit.craftbukkit.CraftEquipmentSlot.getHand(hand), org.bukkit.event.player.PlayerFishEvent.State.CAUGHT_ENTITY); // Paper - Add hand parameter to PlayerFishEvent
                 if (!playerFishEvent.callEvent()) {
@@ -609,6 +614,14 @@ public class FishingHook extends Projectile {
         if (owner != null) {
             Vec3 vec3 = new Vec3(owner.getX() - this.getX(), owner.getY() - this.getY(), owner.getZ() - this.getZ()).scale(0.1);
             entity.setDeltaMovement(entity.getDeltaMovement().add(vec3));
+            // Canvas start - MC-100991
+            if (io.canvasmc.canvas.Config.INSTANCE.fixes.mc100991 && entity instanceof net.minecraft.world.entity.LivingEntity livingEntity) {
+                livingEntity.getCombatTracker().recordDamage(
+                    level().damageSources().thrown(this, owner),
+                    livingEntity.getHealth()
+                );
+            }
+            // Canvas end - MC-100991
         }
     }
 
