From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/projectile/Projectile.java b/net/minecraft/world/entity/projectile/Projectile.java
index da218b6f84f8501eec8086abf2b61d8552741fc9..7b8a2489d5ee0a95c88df600d20241b83dbd4495 100644
--- a/net/minecraft/world/entity/projectile/Projectile.java
+++ b/net/minecraft/world/entity/projectile/Projectile.java
@@ -45,10 +45,46 @@ public abstract class Projectile extends Entity implements TraceableEntity {
     public boolean hasBeenShot = false;
     private @Nullable Entity lastDeflectedBy;
     protected boolean hitCancelled = false; // CraftBukkit
+    private long canvas$ticketTimer = 0L; // Canvas - configurable projectile types can load chunks
 
     protected Projectile(EntityType<? extends Projectile> type, Level level) {
         super(type, level);
     }
+    // Pufferfish start
+
+    private int loadedLifetime = 0;
+
+    @Override
+    public void setPos(double x, double y, double z) {
+        // Canvas start - region threading
+        if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this)) {
+            super.setPos(x, y, z);
+            return;
+        }
+        io.papermc.paper.threadedregions.RegionizedWorldData currWorldData = this.level().getCurrentWorldData();
+        long currentTick = currWorldData.regionData.getCurrentTick();
+        if (currWorldData.canvas$loadedTick != currentTick) {
+            currWorldData.canvas$loadedTick = currentTick;
+            currWorldData.canvas$loadedThisTick = 0;
+        }
+        // Canvas end - region threading
+        int previousX = Mth.floor(this.getX()) >> 4, previousZ = Mth.floor(this.getZ()) >> 4;
+        int newX = Mth.floor(x) >> 4, newZ = Mth.floor(z) >> 4;
+        if (previousX != newX || previousZ != newZ) {
+            boolean isLoaded = ((net.minecraft.server.level.ServerChunkCache) this.level().getChunkSource()).getChunkAtIfLoadedImmediately(newX, newZ) != null;
+            if (!isLoaded) {
+                if (currWorldData.canvas$loadedThisTick > io.canvasmc.canvas.Config.INSTANCE.projectiles.maxProjectileLoadsPerTick) { // Canvas - region threading
+                    if (++this.loadedLifetime > io.canvasmc.canvas.Config.INSTANCE.projectiles.maxProjectileLoadsPerProjectile) {
+                        this.discard();
+                    }
+                    return;
+                }
+                currWorldData.canvas$loadedThisTick++; // Canvas - region threading
+            }
+        }
+        super.setPos(x, y, z);
+    }
+    // Pufferfish end
 
     protected void setOwner(@Nullable EntityReference<Entity> owner) {
         this.owner = owner;
@@ -132,6 +168,25 @@ public abstract class Projectile extends Entity implements TraceableEntity {
         this.checkLeftOwner();
         super.tick();
         this.leftOwnerChecked = false;
+        // Canvas start - configurable projectile types can load chunks
+        if (
+            !(this instanceof net.minecraft.world.entity.projectile.throwableitemprojectile.ThrownEnderpearl) // ender pearls have their own logic
+            && this.isAlive()
+            && this.getType().canvas$isProjectileAndLoadsChunksOverride
+        ) {
+            BlockPos blockPos = BlockPos.containing(this.position());
+            int sectionPosX = net.minecraft.core.SectionPos.blockToSectionCoord(this.position().x());
+            int sectionPosZ = net.minecraft.core.SectionPos.blockToSectionCoord(this.position().x());
+            if ((
+                --this.canvas$ticketTimer <= 0L
+                    || sectionPosX != net.minecraft.core.SectionPos.blockToSectionCoord(blockPos.getX())
+                    || sectionPosZ != net.minecraft.core.SectionPos.blockToSectionCoord(blockPos.getZ())
+            )
+                && this.getOwner() instanceof net.minecraft.server.level.ServerPlayer) {
+                this.canvas$ticketTimer = net.minecraft.server.level.ServerPlayer.placeEnderPearlTicket((ServerLevel) this.level(), this.chunkPosition()) - 1L;
+            }
+        }
+        // Canvas end - configurable projectile types can load chunks
     }
 
     protected void checkLeftOwner() {
@@ -365,7 +420,10 @@ public abstract class Projectile extends Entity implements TraceableEntity {
             EntityHitResult entityHitResult = (EntityHitResult)result;
             Entity entity = entityHitResult.getEntity();
             if (entity.getType().is(EntityTypeTags.REDIRECTABLE_PROJECTILE) && entity instanceof Projectile projectile) {
-                projectile.deflect(ProjectileDeflection.AIM_DEFLECT, this.getOwner(), this.owner, true);
+                // Canvas start - configurable cross-region redirect deflection
+                Entity owner = io.canvasmc.canvas.Config.INSTANCE.projectiles.crossRegionRedirectableProjectileDeflection ? this.getOwnerRaw() : this.getOwner();
+                projectile.deflect(ProjectileDeflection.AIM_DEFLECT, owner, this.owner, true);
+                // Canvas end - configurable cross-region redirect deflection
             }
 
             this.onHitEntity(entityHitResult);
