From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/behavior/Behavior.java b/net/minecraft/world/entity/ai/behavior/Behavior.java
index 5b0cadd2544fb2a627822e645ff32fec2e9cfda9..c1dbe273c6c06028567d8b894937563971ce5b71 100644
--- a/net/minecraft/world/entity/ai/behavior/Behavior.java
+++ b/net/minecraft/world/entity/ai/behavior/Behavior.java
@@ -15,6 +15,8 @@ public abstract class Behavior<E extends LivingEntity> implements BehaviorContro
     private final int minDuration;
     private final int maxDuration;
     private final String configKey; // Paper - configurable behavior tick rate and timings
+    private long canvas$cachedMemoryModCount = -1; // Canvas - optimize brain
+    private boolean canvas$cachedHasRequiredMemoryState; // Canvas - optimize brain
 
     public Behavior(Map<MemoryModuleType<?>, MemoryStatus> entryCondition) {
         this(entryCondition, 60);
@@ -27,7 +29,7 @@ public abstract class Behavior<E extends LivingEntity> implements BehaviorContro
     public Behavior(Map<MemoryModuleType<?>, MemoryStatus> entryCondition, int minDuration, int maxDuration) {
         this.minDuration = minDuration;
         this.maxDuration = maxDuration;
-        this.entryCondition = entryCondition;
+        this.entryCondition = new it.unimi.dsi.fastutil.objects.Reference2ObjectOpenHashMap<>(entryCondition); // Canvas - optimize brain
         // Paper start - configurable behavior tick rate and timings
         String key = io.papermc.paper.util.MappingEnvironment.reobf() ? io.papermc.paper.util.ObfHelper.INSTANCE.deobfClassName(this.getClass().getName()) : this.getClass().getName();
         int lastSeparator = key.lastIndexOf('.');
@@ -103,17 +105,26 @@ public abstract class Behavior<E extends LivingEntity> implements BehaviorContro
         return this.getClass().getSimpleName();
     }
 
-    protected boolean hasRequiredMemories(E owner) {
-        for (Entry<MemoryModuleType<?>, MemoryStatus> entry : this.entryCondition.entrySet()) {
-            MemoryModuleType<?> memoryModuleType = entry.getKey();
-            MemoryStatus memoryStatus = entry.getValue();
-            if (!owner.getBrain().checkMemory(memoryModuleType, memoryStatus)) {
-                return false;
+    // Canvas start - optimize brain
+    protected boolean hasRequiredMemories(E entity) {
+        net.minecraft.world.entity.ai.Brain<?> brain = entity.getBrain();
+        long modCount = brain.getModCount();
+        if (this.canvas$cachedMemoryModCount == modCount) {
+            return this.canvas$cachedHasRequiredMemoryState;
+        }
+        this.canvas$cachedMemoryModCount = modCount;
+
+        it.unimi.dsi.fastutil.objects.ObjectIterator<it.unimi.dsi.fastutil.objects.Reference2ObjectMap.Entry<MemoryModuleType<?>, MemoryStatus>> fastIterator = ((it.unimi.dsi.fastutil.objects.Reference2ObjectOpenHashMap<MemoryModuleType<?>, MemoryStatus>) this.entryCondition).reference2ObjectEntrySet().fastIterator();
+        while (fastIterator.hasNext()) {
+            it.unimi.dsi.fastutil.objects.Reference2ObjectMap.Entry<MemoryModuleType<?>, MemoryStatus> entry = fastIterator.next();
+            if (!brain.checkMemory(entry.getKey(), entry.getValue())) {
+                return this.canvas$cachedHasRequiredMemoryState = false;
             }
         }
 
-        return true;
+        return this.canvas$cachedHasRequiredMemoryState = true;
     }
+    // Canvas end - optimize brain
 
     public static enum Status {
         STOPPED,
