From 56a92b2c479a7161c340134ba931fb3d6a8e6f79 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/behavior/SetLookAndInteract.java b/net/minecraft/world/entity/ai/behavior/SetLookAndInteract.java
index 71b8f9a7d4fe97f56fe1fc03b48e4dcd455ccd1e..2998f1a06e786b171ac1f7d2df0e5f7f2022ba9b 100644
--- a/net/minecraft/world/entity/ai/behavior/SetLookAndInteract.java
+++ b/net/minecraft/world/entity/ai/behavior/SetLookAndInteract.java
@@ -20,7 +20,7 @@ public class SetLookAndInteract {
                     i,
                     (lookTarget, interactionTarget, nearestEntities) -> (level, body, timestamp) -> {
                         Optional<LivingEntity> closest = i.<NearestVisibleLivingEntities>get(nearestEntities)
-                            .findClosest(e -> e.distanceToSqr(body) <= interactionRangeSqr && e.is(type));
+                            .findClosest(e -> e.is(type) && e.distanceToSqr(body) <= interactionRangeSqr); // Canvas - optimize
                         if (closest.isEmpty()) {
                             return false;
                         }
