From 56a92b2c479a7161c340134ba931fb3d6a8e6f79 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/behavior/TradeWithVillager.java b/net/minecraft/world/entity/ai/behavior/TradeWithVillager.java
index 736ad21b73a19113c6f0f3bce6aa1cc0a41485a8..1eae6a24066ee58e4fa378f8892b4c186250a538 100644
--- a/net/minecraft/world/entity/ai/behavior/TradeWithVillager.java
+++ b/net/minecraft/world/entity/ai/behavior/TradeWithVillager.java
@@ -73,7 +73,17 @@ public class TradeWithVillager extends Behavior<Villager> {
     private static Set<Item> figureOutWhatIAmWillingToTrade(final Villager myBody, final Villager target) {
         ImmutableSet<Item> targetItems = target.getVillagerData().profession().value().requestedItems();
         ImmutableSet<Item> selfItems = myBody.getVillagerData().profession().value().requestedItems();
-        return targetItems.stream().filter(entry -> !selfItems.contains(entry)).collect(Collectors.toSet());
+        // Canvas start - remove streams
+        Set<Item> result = new java.util.HashSet<>();
+
+        for (Item item : targetItems) {
+            if (!selfItems.contains(item)) {
+                result.add(item);
+            }
+        }
+
+        return result;
+        // Canvas end - remove streams
     }
 
     private static void throwHalfStack(final Villager villager, final Set<Item> items, final LivingEntity target) {
