From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Fri, 8 Dec 2023 03:43:26 +0000
Subject: [PATCH] RAM Details Exposure


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index ee732ed80e84749fdb9c6938a984e74a58044ff1..224f3df8c99df10e72515e59df7184a1f0b36032 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -108,6 +108,41 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull
     public me.dueris.canvas.CraftServerGui getServerGui();
 
+    /**
+     * Retrieves the total memory size.
+     *
+     * @return The total memory size in bytes, returned as a long.
+     */
+    public long getResourcesTotal();
+
+    /**
+     * Retrieves the amount of free memory.
+     *
+     * @return The amount of free memory in bytes, returned as a long.
+     */
+    public long getResourcesFree();
+
+    /**
+     * Retrieves the maximum memory size.
+     *
+     * @return The maximum memory size in bytes, returned as a long.
+     */
+    public long getResourcesMax();
+
+    /**
+     * Retrieves the amount of used memory (total - free).
+     *
+     * @return The amount of used memory in bytes, returned as a long.
+     */
+    public long getResourcesUsedMem();
+
+    /**
+     * Retrieves the used memory as a percentage of the maximum memory.
+     *
+     * @return The used memory percentage, where 0% represents no memory usage, and 100% represents full memory usage.
+     */
+    public int getResourcesUsedPercent();
+
     // Canvas end
 
     /**
