From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Fri, 27 Oct 2023 07:18:33 +0000
Subject: [PATCH] Expose StartupArgs to Server


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 224f3df8c99df10e72515e59df7184a1f0b36032..719b59119cd54d94e02fffbbc789abf7c79c2047 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -143,6 +143,12 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     public int getResourcesUsedPercent();
 
+    /**
+     * Retrieves the array of startup arguments used to launch the server.
+     *
+     * @return The array of startup arguments passed to the server during its launch.
+     */
+    public String[] getStartupArgs();
     // Canvas end
 
     /**
