From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Fri, 24 Nov 2023 14:45:21 -0800
Subject: [PATCH] Create FakeBlock/GhostBlock api


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 07e75978b4fc0e446e8aa46a40be5e371dc1c11b..3e66acfec675f3a6dc4eed9facc75f49c33c9c4d 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -156,7 +156,38 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
         int z = Block.getBlockKeyZ(key);
         return new Location(this, x, y, z);
     }
-    // Paper end
+    // Paper end -/- Canvas start
+
+    /**
+     * Creates a "ghost block" that the player its sent to can walk on and interact with, like the block was actually there
+     * @param location
+     * @param material
+     * @param sender
+     */
+    public void createFakePhysicalBlock(Location location, Material material, Player sender);
+
+    /**
+     * Creates a "ghost block" that the player its sent to, that is non-interactable and acts like an AIR block
+     * @param location
+     * @param material
+     * @param sender
+     */
+    public void createGhostBlock(Location location, Material material, Player sender);
+
+    /**
+     * Creates a "ghost block" that all players online are sent, where they can walk on and interact with, like the block was actually there
+     * @param location
+     * @param material
+     */
+    public void createFakePhysicalBlock(Location location, Material material);
+
+    /**
+     * Creates a "ghost block" that all players online are sent, that is non-interactable and acts like an AIR block
+     * @param location
+     * @param material
+     */
+    public void createGhostBlock(Location location, Material material);
+    // Canvas end
 
     /**
      * Gets the highest non-empty (impassable) block at the given coordinates.
