From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 1229ed65 AH <1229ed65b2ccd3@crankymonkey.info>
Date: Fri, 26 Jan 2024 17:54:53 +0000
Subject: [PATCH] LevelResource api


diff --git a/src/main/java/me/dueris/canvas/util/LevelResource.java b/src/main/java/me/dueris/canvas/util/LevelResource.java
new file mode 100644
index 0000000000000000000000000000000000000000..7d92a9779ab304da79009758bc4b1381f3cd0db1
--- /dev/null
+++ b/src/main/java/me/dueris/canvas/util/LevelResource.java
@@ -0,0 +1,39 @@
+package me.dueris.canvas.util;
+
+import java.nio.file.*;
+import java.io.*;
+
+public interface LevelResource {
+    
+    public Path datapackDir();
+    
+    public Path advancementsDir();
+    
+    public Path playerStatsDir();
+    
+    public Path playerdataDir();
+    
+    public Path legacyPlayerdataDir();
+    
+    public File leveldataFile();
+    
+    public File leveldataOldFile();
+    
+    public File iconFile();
+    
+    public File lockFile();
+    
+    public Path generatedDir();
+    
+    public File mapResourceFile();
+    
+    public Path mainWorldRoot();
+    
+    public Path serverRoot();
+    
+    public File serverPropertiesFile();
+    
+    public Path logsDir();
+    
+    public Path cacheDir();
+}
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 0f44289872f17d6a80dc25c7f9237df00c10031d..70a1eb6b6bcd9261fddd8d98cd155c1ed5a12468 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -159,6 +159,11 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     public ServerProperties getServerProperties();
 
+    /**
+     * Gets the LevelResource Constants from Canvas
+     */
+    public me.dueris.canvas.util.LevelResource getLevelResources();
+
     public TimeManager getTimeManager();
     // Canvas end
 
