From 56a92b2c479a7161c340134ba931fb3d6a8e6f79 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/server/commands/EnchantCommand.java b/net/minecraft/server/commands/EnchantCommand.java
index 129db1b3368b08b2dea943f5417bb984fb075f1e..372996032ecca12cbfa758594ede3bfc5f7f6d5f 100644
--- a/net/minecraft/server/commands/EnchantCommand.java
+++ b/net/minecraft/server/commands/EnchantCommand.java
@@ -67,8 +67,8 @@ public class EnchantCommand {
         final CommandSourceStack source, final Collection<? extends Entity> targets, final Holder<Enchantment> enchantmentHolder, final int level
     ) throws CommandSyntaxException {
         Enchantment enchantment = enchantmentHolder.value();
-        if (level > enchantment.getMaxLevel()) {
-            throw ERROR_LEVEL_TOO_HIGH.create(level, enchantment.getMaxLevel());
+        if (level > (io.canvasmc.canvas.GlobalConfiguration.getInstance().enchantCommand.uncapMaxLevel ? 255 : enchantment.getMaxLevel())) { // Canvas - enchant command configs
+            throw ERROR_LEVEL_TOO_HIGH.create(level, (io.canvasmc.canvas.GlobalConfiguration.getInstance().enchantCommand.uncapMaxLevel ? 255 : enchantment.getMaxLevel())); // Canvas - enchant command configs
         }
 
         // Canvas start - region threading
@@ -88,8 +88,10 @@ public class EnchantCommand {
             if (entity instanceof LivingEntity target) {
                 ItemStack item = target.getMainHandItem();
                 if (!item.isEmpty()) {
-                    if (enchantment.canEnchant(item)
-                        && EnchantmentHelper.isEnchantmentCompatible(EnchantmentHelper.getEnchantmentsForCrafting(item).keySet(), enchantmentHolder)) {
+                    // Canvas start - enchant command configs
+                    if ((io.canvasmc.canvas.GlobalConfiguration.getInstance().enchantCommand.allowEnchantsOnUnsupportedItems || enchantment.canEnchant(item))
+                        && (io.canvasmc.canvas.GlobalConfiguration.getInstance().enchantCommand.allowEnchantingWithIncompatibleEnchants || EnchantmentHelper.isEnchantmentCompatible(EnchantmentHelper.getEnchantmentsForCrafting(item).keySet(), enchantmentHolder))) {
+                    // Canvas end - enchant command configs
                         item.enchant(enchantmentHolder, level);
                         return (list) -> list.add(entity); // Canvas - region threading
                     } else if (targets.size() == 1) {
