--- a/net/minecraft/world/level/dimension/end/EndDragonFight.java
+++ b/net/minecraft/world/level/dimension/end/EndDragonFight.java
@@ -268,7 +_,7 @@
             for (int i1 = -8; i1 <= 8; i1++) {
                 LevelChunk chunk = this.level.getChunk(i, i1);
 
-                for (BlockEntity blockEntity : chunk.getBlockEntities().values()) {
+                for (BlockEntity blockEntity : chunk.canvas$getAllBlockEntities()) { // Canvas - optimize block entity fetching
                     if (blockEntity instanceof TheEndPortalBlockEntity) {
                         return true;
                     }
@@ -287,7 +_,7 @@
             for (int i1 = -8 + chunkPos.z; i1 <= 8 + chunkPos.z; i1++) {
                 LevelChunk chunk = this.level.getChunk(i, i1);
 
-                for (BlockEntity blockEntity : chunk.getBlockEntities().values()) {
+                for (BlockEntity blockEntity : chunk.canvas$getAllBlockEntities()) { // Canvas - optimize block entity fetching
                     if (blockEntity instanceof TheEndPortalBlockEntity) {
                         BlockPattern.BlockPatternMatch blockPatternMatch = this.exitPortalPattern.find(this.level, blockEntity.getBlockPos());
                         if (blockPatternMatch != null) {
