--- a/net/minecraft/world/entity/vehicle/OldMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/OldMinecartBehavior.java
@@ -399,7 +_,7 @@
                 }
             }
         } else {
-            for (Entity entity1 : this.level().getEntities(this.minecart, aabb)) {
+            for (Entity entity1 : this.level().getEntitiesOfClass(AbstractMinecart.class, aabb, entity -> entity != this.minecart)) { // Canvas - optimize entity type predicates
                 if (!this.minecart.hasPassenger(entity1) && entity1.isPushable() && entity1 instanceof AbstractMinecart) {
                     // CraftBukkit start
                     org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
