--- a/net/minecraft/network/protocol/PacketUtils.java
+++ b/net/minecraft/network/protocol/PacketUtils.java
@@ -49,7 +_,12 @@
             };
             // ignore retired state, if removed then we don't want the packet to be handled
             if (processor instanceof net.minecraft.server.network.ServerGamePacketListenerImpl gamePacketListener) {
-                gamePacketListener.player.queuePacketTask(run);
+                // Canvas start - fix end credits
+                if (gamePacketListener.player.isDisplayingEndCredits) io.papermc.paper.threadedregions.RegionizedServer.getInstance().addTask(run);
+                else {
+                    gamePacketListener.player.queuePacketTask(run);
+                }
+                // Canvas end - fix end credits
             } else if (processor instanceof net.minecraft.server.network.ServerConfigurationPacketListenerImpl configurationPacketListener) {
                 io.papermc.paper.threadedregions.RegionizedServer.getInstance().addTask(run);
             } else if (processor instanceof net.minecraft.server.network.ServerLoginPacketListenerImpl loginPacketListener) {
