From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/server/network/ServerLoginPacketListenerImpl.java b/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
index 69f694995848173c2f3137aa8e9706be6fd3a9ae..c34e75e4a81262102fb10b1bb237e9d85480372f 100644
--- a/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
@@ -423,11 +423,32 @@ public class ServerLoginPacketListenerImpl implements ServerLoginPacketListener,
         this.disconnect(ServerCommonPacketListenerImpl.DISCONNECT_UNEXPECTED_QUERY);
     }
 
+    private volatile boolean canvas$switchedAlready = false; // Canvas - async protocol switch
     @Override
     public void handleLoginAcknowledgement(ServerboundLoginAcknowledgedPacket packet) {
+        // Canvas start - async protocol switch
+        if (io.canvasmc.canvas.Config.INSTANCE.networking.asyncProtocolSwitch && !canvas$switchedAlready) {
+            canvas$switchedAlready = true;
+            this.connection.canvas$setupOutboundProtocolAsync(ConfigurationProtocols.CLIENTBOUND)
+                .addListener(unused -> {
+                    try {
+                        // just run PacketUtils, this will schedule this to the global tick for us
+                        net.minecraft.network.protocol.PacketUtils.ensureRunningOnSameThread(packet, this, this.server.packetProcessor());
+                    } catch (net.minecraft.server.RunningOnDifferentThreadException | io.papermc.paper.util.ServerStopRejectedExecutionException ignored) {
+                    } catch (java.util.concurrent.RejectedExecutionException rejExc) {
+                        this.connection.disconnect(Component.translatable("multiplayer.disconnect.server_shutdown"));
+                    } catch (ClassCastException castClass) {
+                        LOGGER.error("Received {} that couldn't be processed", packet.getClass(), castClass);
+                        this.connection.disconnect(Component.translatable("multiplayer.disconnect.invalid_packet"));
+                    }
+                }
+            );
+            return;
+        }
         net.minecraft.network.protocol.PacketUtils.ensureRunningOnSameThread(packet, this, this.server.packetProcessor()); // CraftBukkit
         Validate.validState(this.state == ServerLoginPacketListenerImpl.State.PROTOCOL_SWITCHING, "Unexpected login acknowledgement packet");
-        this.connection.setupOutboundProtocol(ConfigurationProtocols.CLIENTBOUND);
+        if (!io.canvasmc.canvas.Config.INSTANCE.networking.asyncProtocolSwitch) this.connection.setupOutboundProtocol(ConfigurationProtocols.CLIENTBOUND);
+        // Canvas end - async protocol switch
         CommonListenerCookie commonListenerCookie = CommonListenerCookie.createInitial(Objects.requireNonNull(this.authenticatedProfile), this.transferred);
         ServerConfigurationPacketListenerImpl serverConfigurationPacketListenerImpl = new ServerConfigurationPacketListenerImpl(
             this.server, this.connection, commonListenerCookie
