From 53e1a57c72953de1bfcc94c3257c724e7fa1f88f Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/util/ParallelSearchRadiusIteration.java b/ca/spottedleaf/moonrise/patches/chunk_system/util/ParallelSearchRadiusIteration.java
index 28d796a3fddebf0bb25ba759969aadecc5a6a2de..497e77cf502b14c9560deafc397927ae3ed53150 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/util/ParallelSearchRadiusIteration.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/util/ParallelSearchRadiusIteration.java
@@ -58,7 +58,7 @@ public final class ParallelSearchRadiusIteration {
     }
 
     public static long[] getSearchIteration(final int radius) {
-        return SEARCH_RADIUS_ITERATION_LIST[radius];
+        return SEARCH_RADIUS_ITERATION_LIST[Math.min(SEARCH_RADIUS_ITERATION_LIST.length - 1, radius)]; // Canvas - temporary fix for Moonrise#130
     }
 
     public static long[] getEuclideanIteration(final int radius) {
