From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/server/Main.java b/net/minecraft/server/Main.java
index 38b50aaf6473c7692d3372ad7b956355be989bd4..cbd185eabb9ffd94cc9a4f3a6dbe1470ffdb70ce 100644
--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -66,6 +66,19 @@ public class Main {
     public static void main(final OptionSet options) { // CraftBukkit - replaces main(String[] args)
         io.papermc.paper.log.LogManagerShutdownThread.hook(); // Paper - Improved watchdog support
         SharedConstants.tryDetectVersion();
+        // Canvas start - validate EULA early
+        Path eulaFile = Paths.get("eula.txt");
+        Eula eula = new Eula(eulaFile);
+        boolean eulaAgreed = Boolean.getBoolean("com.mojang.eula.agree");
+        if (eulaAgreed) {
+            LOGGER.warn("You have used the Paper command line EULA agreement flag.");
+            LOGGER.warn("By using this setting you are indicating your agreement to Mojang's EULA (https://aka.ms/MinecraftEULA).");
+            LOGGER.warn("If you do not agree to the above EULA please stop your server and remove this flag immediately.");
+        } else if (!eula.hasAgreedToEULA()) {
+            LOGGER.info("You need to agree to the EULA in order to run the server. Go to eula.txt for more info.");
+            return;
+        }
+        // Canvas end - validate EULA early
         /* CraftBukkit start - Replace everything
         OptionParser parser = new OptionParser();
         OptionSpec<Void> nogui = parser.accepts("nogui");
@@ -118,8 +131,7 @@ public class Main {
             }
             // Paper end
             RegionFileVersion.configure(settings.getProperties().regionFileComression);
-            Path eulaFile = Paths.get("eula.txt");
-            Eula eula = new Eula(eulaFile);
+            // Canvas - validate EULA early - move up
             // Paper start - load config files early for access below if needed
             org.bukkit.configuration.file.YamlConfiguration bukkitConfiguration = io.papermc.paper.configuration.PaperConfigurations.loadLegacyConfigFile((File) options.valueOf("bukkit-settings"));
             org.bukkit.configuration.file.YamlConfiguration spigotConfiguration = io.papermc.paper.configuration.PaperConfigurations.loadLegacyConfigFile((File) options.valueOf("spigot-settings"));
@@ -142,17 +154,7 @@ public class Main {
                 return;
             }
 
-            // Paper start - eula system property
-            boolean eulaAgreed = Boolean.getBoolean("com.mojang.eula.agree");
-            if (eulaAgreed) {
-                LOGGER.warn("You have used the Paper command line EULA agreement flag.");
-                LOGGER.warn("By using this setting you are indicating your agreement to Mojang's EULA (https://aka.ms/MinecraftEULA).");
-                LOGGER.warn("If you do not agree to the above EULA please stop your server and remove this flag immediately.");
-            } else if (!eula.hasAgreedToEULA()) {
-                // Paper end - eula system property
-                LOGGER.info("You need to agree to the EULA in order to run the server. Go to eula.txt for more info.");
-                return;
-            }
+            // Canvas - validate EULA early - move up
 
             // Paper start - Detect headless JRE
             String awtException = io.papermc.paper.util.ServerEnvironment.awtDependencyCheck();
