From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 9c13c02cf60ed472b91e269bf57015a6dbaac4e4..460f3f8bd1ee491b974accc543ef5a3dc5923761 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1346,11 +1346,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
                 LOGGER.info("*************************************************************************************");
             }
             // Paper end - Add onboarding message for initial server start
-            // Paper start - Improve outdated version checking
-            if (System.getProperty("paper.disableStartupVersionCheck") == null && io.papermc.paper.configuration.GlobalConfiguration.get().updateChecker.enabled) {
-                CompletableFuture.runAsync(com.destroystokyo.paper.PaperVersionFetcher::getUpdateStatusStartupMessage, io.papermc.paper.util.MCUtil.ASYNC_EXECUTOR);
-            }
-            // Paper end - Improve outdated version checking
+            // Canvas - Rebrand
 
             while (this.running) {
                 final long tickStart = System.nanoTime(); // Paper - improve tick loop
@@ -1858,6 +1854,8 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         if (false) while (!this.processQueue.isEmpty()) { // Folia - region threading
             this.processQueue.remove().run();
         }
+        region.world.getCurrentWorldData().tpsbar.tick(); // Canvas - tpsbar
+        region.world.getCurrentWorldData().rambar.tick(); // Canvas - rambar
 
         this.updateEffectiveRespawnData();
 
@@ -1999,7 +1997,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     }
 
     public String getServerModName() {
-        return io.papermc.paper.ServerBuildInfo.buildInfo().brandName(); // Paper
+        return io.canvasmc.canvas.GlobalConfiguration.getInstance().serverModName; // Canvas - Use configurable server branding
     }
 
     public ServerClockManager clockManager() {
@@ -2308,6 +2306,13 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         return 256;
     }
 
+    // Canvas start - no chat reports
+    public boolean canvas$ncrEnforceSecureProfile() {
+        if (io.canvasmc.canvas.GlobalConfiguration.getInstance().chat.disableChatReporting) return false; // NCR doesn't require secure profile
+        return this.enforceSecureProfile();
+    }
+
+    // Canvas end - no chat reports
     public boolean enforceSecureProfile() {
         return false;
     }
