From 53e1a57c72953de1bfcc94c3257c724e7fa1f88f Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index c0d915d58b7d1a7b90af4df8b1ac5f30b18f756a..1440608c41de5cc12eb39e2695b995dd5ba55c6a 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1356,11 +1356,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
                 LOGGER.info("*************************************************************************************");
             }
             // Paper end - Add onboarding message for initial server start
-            // Paper start - Improve outdated version checking
-            if (System.getProperty("paper.disableStartupVersionCheck") == null && io.papermc.paper.configuration.GlobalConfiguration.get().updateChecker.enabled) {
-                CompletableFuture.runAsync(com.destroystokyo.paper.PaperVersionFetcher::getUpdateStatusStartupMessage, io.papermc.paper.util.MCUtil.ASYNC_EXECUTOR);
-            }
-            // Paper end - Improve outdated version checking
+            // Canvas - Rebrand
 
             while (this.running) {
                 final long tickStart = System.nanoTime(); // Paper - improve tick loop
@@ -1925,6 +1921,8 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         if (false) while (!this.processQueue.isEmpty()) { // Folia - region threading
             this.processQueue.remove().run();
         }
+        region.world.getCurrentWorldData().tpsbar.tick(); // Canvas - tpsbar
+        region.world.getCurrentWorldData().rambar.tick(); // Canvas - rambar
 
         this.updateEffectiveRespawnData();
 
@@ -2066,7 +2064,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     }
 
     public String getServerModName() {
-        return io.papermc.paper.ServerBuildInfo.buildInfo().brandName(); // Paper
+        return io.canvasmc.canvas.GlobalConfiguration.getInstance().serverModName; // Canvas - Use configurable server branding
     }
 
     public ServerClockManager clockManager() {
@@ -2374,6 +2372,13 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         return 256;
     }
 
+    // Canvas start - no chat reports
+    public boolean canvas$ncrEnforceSecureProfile() {
+        if (io.canvasmc.canvas.GlobalConfiguration.getInstance().chat.disableChatReporting) return false; // NCR doesn't require secure profile
+        return this.enforceSecureProfile();
+    }
+
+    // Canvas end - no chat reports
     public boolean enforceSecureProfile() {
         return false;
     }
