From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/commands/Commands.java b/net/minecraft/commands/Commands.java
index 1a22759c4dba723920b5d116b6d55592a3bcce22..3d7b52a0b75b0bbda86322931bdc370b9d46c173 100644
--- a/net/minecraft/commands/Commands.java
+++ b/net/minecraft/commands/Commands.java
@@ -194,7 +194,7 @@ public class Commands {
         AdvancementCommands.register(this.dispatcher);
         AttributeCommand.register(this.dispatcher, context);
         ExecuteCommand.register(this.dispatcher, context);
-        //BossBarCommands.register(this.dispatcher, context); // Folia - region threading - TODO
+        BossBarCommands.register(this.dispatcher, context);
         ClearInventoryCommands.register(this.dispatcher, context);
         //CloneCommands.register(this.dispatcher, context); // Folia - region threading - TODO
         DamageCommand.register(this.dispatcher, context);
@@ -221,7 +221,7 @@ public class Commands {
         KillCommand.register(this.dispatcher);
         ListPlayersCommand.register(this.dispatcher);
         LocateCommand.register(this.dispatcher, context);
-        //LootCommand.register(this.dispatcher, context); // Folia - region threading - TODO later
+        LootCommand.register(this.dispatcher, context);
         MsgCommand.register(this.dispatcher);
         ParticleCommand.register(this.dispatcher, context);
         PlaceCommand.register(this.dispatcher);
@@ -242,11 +242,11 @@ public class Commands {
         SetSpawnCommand.register(this.dispatcher);
         SetWorldSpawnCommand.register(this.dispatcher);
         SpectateCommand.register(this.dispatcher);
-        //SpreadPlayersCommand.register(this.dispatcher); // Folia - region threading - TODO later
+        SpreadPlayersCommand.register(this.dispatcher);
         StopSoundCommand.register(this.dispatcher);
         StopwatchCommand.register(this.dispatcher);
         SummonCommand.register(this.dispatcher, context);
-        //TagCommand.register(this.dispatcher); // Folia - region threading - TODO later
+        TagCommand.register(this.dispatcher); // Canvas - TODO - do better
         //TeamCommand.register(this.dispatcher, context); // Folia - region threading - TODO later
         //TeamMsgCommand.register(this.dispatcher); // Folia - region threading - TODO later
         TeleportCommand.register(this.dispatcher);
@@ -259,6 +259,7 @@ public class Commands {
         WaypointCommand.register(this.dispatcher, context);
         WeatherCommand.register(this.dispatcher);
         WorldBorderCommand.register(this.dispatcher);
+        io.canvasmc.canvas.command.RootCommandTree.INSTANCE.build(this.dispatcher); // Canvas - Rebrand - register commands
         if (JvmProfiler.INSTANCE.isAvailable()) {
             JfrCommand.register(this.dispatcher);
         }
@@ -288,7 +289,7 @@ public class Commands {
             PardonCommand.register(this.dispatcher);
             PardonIpCommand.register(this.dispatcher);
             //PerfCommand.register(this.dispatcher); // Folia - region threading - TODO later
-            //SaveAllCommand.register(this.dispatcher); // Folia - region threading - TODO later
+            SaveAllCommand.register(this.dispatcher);
             SaveOffCommand.register(this.dispatcher);
             SaveOnCommand.register(this.dispatcher);
             SetPlayerIdleTimeoutCommand.register(this.dispatcher);
