--- a/net/minecraft/world/level/chunk/ChunkGenerator.java
+++ b/net/minecraft/world/level/chunk/ChunkGenerator.java
@@ -196,7 +_,7 @@
             int chunkOriginX = SectionPos.blockToSectionCoord(pos.getX());
             int chunkOriginZ = SectionPos.blockToSectionCoord(pos.getZ());
 
-            for (int radius = 0; radius <= maxSearchRadius; radius++) {
+            int radius = maxSearchRadius; { // Canvas - optimize treasure map locating
                 boolean foundSomething = false;
 
                 for (Entry<StructurePlacement, Set<Holder<Structure>>> entry : randomSpreadEntries) {
@@ -507,6 +_,12 @@
     ) {
         Map<Structure, LongSet> structures = structureManager.getAllStructuresAt(pos);
 
+    // Canvas start - optimize natural spawning
+        return getMobsAt(biome, structureManager, mobCategory, pos, structures);
+    }
+
+    public WeightedList<MobSpawnSettings.SpawnerData> getMobsAt(final Holder<Biome> biome, final StructureManager structureManager, final MobCategory mobCategory, final BlockPos pos, final Map<Structure, LongSet> structures) {
+    // Canvas end - optimize natural spawning
         for (Entry<Structure, LongSet> entry : structures.entrySet()) {
             Structure structure = entry.getKey();
             StructureSpawnOverride override = structure.spawnOverrides().get(mobCategory);
