--- a/net/minecraft/world/level/chunk/ChunkGenerator.java
+++ b/net/minecraft/world/level/chunk/ChunkGenerator.java
@@ -229,7 +_,7 @@
             int chunkOriginX = SectionPos.blockToSectionCoord(pos.getX());
             int chunkOriginZ = SectionPos.blockToSectionCoord(pos.getZ());
 
-            for (int radius = 0; radius <= maxSearchRadius; radius++) {
+            for (int radius = (io.canvasmc.canvas.GlobalConfiguration.getInstance().chunkSystem.optimizeTreasureMapLocating ? maxSearchRadius : 0); radius <= maxSearchRadius; radius++) { // Canvas - optimize treasure map locating - just search the largest radius to avoid spamming iterations
                 boolean foundSomething = false;
 
                 for (Entry<StructurePlacement, Set<Holder<Structure>>> entry : randomSpreadEntries) {
