From 2be0fbf099df19422462cdb5b961fc00062d256b Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/chunk/ChunkGenerator.java b/net/minecraft/world/level/chunk/ChunkGenerator.java
index 7e7b096b6e6be0481313105d68d8457915a1baba..6fba71fbd64dff2f83a5b888fd947731314a110e 100644
--- a/net/minecraft/world/level/chunk/ChunkGenerator.java
+++ b/net/minecraft/world/level/chunk/ChunkGenerator.java
@@ -505,6 +505,12 @@ public abstract class ChunkGenerator {
     ) {
         Map<Structure, LongSet> structures = structureManager.getAllStructuresAt(pos);
 
+    // Canvas start - optimize natural spawning
+        return getMobsAt(biome, structureManager, mobCategory, pos, structures);
+    }
+
+    public WeightedList<MobSpawnSettings.SpawnerData> getMobsAt(Holder<Biome> biome, StructureManager structureManager, MobCategory mobCategory, BlockPos pos, Map<Structure, LongSet> structures) {
+    // Canvas end - optimize natural spawning
         for (Entry<Structure, LongSet> entry : structures.entrySet()) {
             Structure structure = entry.getKey();
             StructureSpawnOverride override = structure.spawnOverrides().get(mobCategory);
