--- a/net/minecraft/world/level/chunk/ChunkGenerator.java
+++ b/net/minecraft/world/level/chunk/ChunkGenerator.java
@@ -505,6 +_,12 @@
     ) {
         Map<Structure, LongSet> structures = structureManager.getAllStructuresAt(pos);
 
+    // Canvas start - optimize natural spawning
+        return getMobsAt(biome, structureManager, mobCategory, pos, structures);
+    }
+
+    public WeightedList<MobSpawnSettings.SpawnerData> getMobsAt(Holder<Biome> biome, StructureManager structureManager, MobCategory mobCategory, BlockPos pos, Map<Structure, LongSet> structures) {
+    // Canvas end - optimize natural spawning
         for (Entry<Structure, LongSet> entry : structures.entrySet()) {
             Structure structure = entry.getKey();
             StructureSpawnOverride override = structure.spawnOverrides().get(mobCategory);
