--- a/net/minecraft/world/level/chunk/LevelChunkSection.java
+++ b/net/minecraft/world/level/chunk/LevelChunkSection.java
@@ -18,11 +_,14 @@
     public static final int SECTION_HEIGHT = 16;
     public static final int SECTION_SIZE = 4096;
     public static final int BIOME_CONTAINER_BITS = 2;
-    public short nonEmptyBlockCount; // Paper - package private
+    public volatile short nonEmptyBlockCount; // Paper - package private // Leaf - async chunk sending - volatile
     public short tickingBlockCount;
     public short tickingFluidCount;
     public final PalettedContainer<BlockState> states;
     private PalettedContainer<Holder<Biome>> biomes; // CraftBukkit - read/write
+    public short waterCount; // Canvas - optimize fluid pushing
+    public short bubbleCount; // Canvas - optimize fluid pushing
+    public short lavaCount; // Canvas - optimize fluid pushing
 
     // Paper start - block counting
     private static final it.unimi.dsi.fastutil.shorts.ShortArrayList FULL_LIST = new it.unimi.dsi.fastutil.shorts.ShortArrayList(16*16*16);
@@ -52,6 +_,9 @@
         this.nonEmptyBlockCount = section.nonEmptyBlockCount;
         this.tickingBlockCount = section.tickingBlockCount;
         this.tickingFluidCount = section.tickingFluidCount;
+        this.waterCount = section.waterCount; // Canvas - optimize fluid pushing
+        this.bubbleCount = section.bubbleCount; // Canvas - optimize fluid pushing
+        this.lavaCount = section.lavaCount; // Canvas - optimize fluid pushing
         this.states = section.states.copy();
         this.biomes = section.biomes.copy();
     }
@@ -159,6 +_,26 @@
                 this.tickingBlockCount++;
             }
         }
+        // Canvas start - optimize fluid pushing
+        if (blockState.isWater) {
+            this.waterCount--;
+        }
+        if (blockState.isBubble) {
+            this.bubbleCount--;
+        }
+        if (blockState.isLava) {
+            this.lavaCount--;
+        }
+        if (state.isWater) {
+            this.waterCount++;
+        }
+        if (state.isBubble) {
+            this.bubbleCount++;
+        }
+        if (state.isLava) {
+            this.lavaCount++;
+        }
+        // Canvas end - optimize fluid pushing
 
         if (!!fluidState1.isRandomlyTicking()) { // Paper - block counting
             this.tickingFluidCount++;
@@ -189,6 +_,8 @@
         // Paper start - block counting
         // reset, then recalculate
         this.nonEmptyBlockCount = (short)0;
+        this.waterCount = (short)0; // Canvas - optimize fluid pushing
+        this.lavaCount = (short)0; // Canvas - optimize fluid pushing
         this.tickingBlockCount = (short)0;
         this.tickingFluidCount = (short)0;
         this.specialCollidingBlocks = (short)0;
@@ -246,6 +_,17 @@
                     if (fluid.isRandomlyTicking()) {
                         this.tickingFluidCount += (short)paletteCount;
                     }
+                    // Canvas start - optimize fluid pushing
+                    if (state.isWater) {
+                        this.waterCount += (short)paletteCount;
+                    }
+                    if (state.isBubble) {
+                        this.bubbleCount += (short)paletteCount;
+                    }
+                    if (state.isLava) {
+                        this.lavaCount += (short)paletteCount;
+                    }
+                    // Canvas end - optimize fluid pushing
                 }
             }
         }
