--- a/net/minecraft/world/level/chunk/LevelChunkSection.java
+++ b/net/minecraft/world/level/chunk/LevelChunkSection.java
@@ -23,6 +_,9 @@
     public short tickingFluidCount;
     public final PalettedContainer<BlockState> states;
     private PalettedContainer<Holder<Biome>> biomes; // CraftBukkit - read/write
+    public short canvas$waterCount; // Canvas - optimize fluid pushing
+    public short canvas$bubbleCount; // Canvas - optimize fluid pushing
+    public short canvas$lavaCount; // Canvas - optimize fluid pushing
 
     // Paper start - block counting
     private static final it.unimi.dsi.fastutil.shorts.ShortArrayList FULL_LIST = new it.unimi.dsi.fastutil.shorts.ShortArrayList(16*16*16);
@@ -52,6 +_,9 @@
         this.nonEmptyBlockCount = section.nonEmptyBlockCount;
         this.tickingBlockCount = section.tickingBlockCount;
         this.tickingFluidCount = section.tickingFluidCount;
+        this.canvas$waterCount = section.canvas$waterCount; // Canvas - optimize fluid pushing
+        this.canvas$bubbleCount = section.canvas$bubbleCount; // Canvas - optimize fluid pushing
+        this.canvas$lavaCount = section.canvas$lavaCount; // Canvas - optimize fluid pushing
         this.states = section.states.copy();
         this.biomes = section.biomes.copy();
     }
@@ -159,6 +_,26 @@
                 this.tickingBlockCount++;
             }
         }
+        // Canvas start - optimize fluid pushing
+        if (blockState.canvas$isWater) {
+            this.canvas$waterCount--;
+        }
+        if (blockState.canvas$isBubble) {
+            this.canvas$bubbleCount--;
+        }
+        if (blockState.canvas$isLava) {
+            this.canvas$lavaCount--;
+        }
+        if (state.canvas$isWater) {
+            this.canvas$waterCount++;
+        }
+        if (state.canvas$isBubble) {
+            this.canvas$bubbleCount++;
+        }
+        if (state.canvas$isLava) {
+            this.canvas$lavaCount++;
+        }
+        // Canvas end - optimize fluid pushing
 
         if (!!fluidState1.isRandomlyTicking()) { // Paper - block counting
             this.tickingFluidCount++;
@@ -189,6 +_,8 @@
         // Paper start - block counting
         // reset, then recalculate
         this.nonEmptyBlockCount = (short)0;
+        this.canvas$waterCount = (short)0; // Canvas - optimize fluid pushing
+        this.canvas$lavaCount = (short)0; // Canvas - optimize fluid pushing
         this.tickingBlockCount = (short)0;
         this.tickingFluidCount = (short)0;
         this.specialCollidingBlocks = (short)0;
@@ -246,6 +_,17 @@
                     if (fluid.isRandomlyTicking()) {
                         this.tickingFluidCount += (short)paletteCount;
                     }
+                    // Canvas start - optimize fluid pushing
+                    if (state.canvas$isWater) {
+                        this.canvas$waterCount += (short)paletteCount;
+                    }
+                    if (state.canvas$isBubble) {
+                        this.canvas$bubbleCount += (short)paletteCount;
+                    }
+                    if (state.canvas$isLava) {
+                        this.canvas$lavaCount += (short)paletteCount;
+                    }
+                    // Canvas end - optimize fluid pushing
                 }
             }
         }
