--- a/net/minecraft/world/level/chunk/Palette.java
+++ b/net/minecraft/world/level/chunk/Palette.java
@@ -25,4 +_,9 @@
     public interface Factory {
         <A> Palette<A> create(int bits, IdMap<A> registry, PaletteResize<A> resizeHandler, List<A> values);
     }
+    // Canvas start - maybeHasOrCatch
+    public default boolean maybeHasOrCatch(Predicate<T> filter, T defaultValue) {
+        return this.maybeHas(filter);
+    }
+    // Canvas end
 }
