From 3cf7451517453a52df7cb21a31824aadeaf7c6d9 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/chunk/ProtoChunk.java b/net/minecraft/world/level/chunk/ProtoChunk.java
index d3bde6eb0edd24d5e5352388788d30938712a9c7..56fa0ba6a447ca0969dff268514289ebd54f4f0a 100644
--- a/net/minecraft/world/level/chunk/ProtoChunk.java
+++ b/net/minecraft/world/level/chunk/ProtoChunk.java
@@ -189,16 +189,19 @@ public class ProtoChunk extends ChunkAccess {
     @Override
     public void setBlockEntity(final BlockEntity blockEntity) {
         this.pendingBlockEntities.remove(blockEntity.getBlockPos());
-        this.blockEntities.put(blockEntity.getBlockPos(), blockEntity);
+        // Canvas start - optimize block entity fetching
+        BlockPos pos = blockEntity.getBlockPos();
+        this.canvas$setBlockEntityInBuckets(pos.getX(), pos.getY(), pos.getZ(), blockEntity);
+        // Canvas end - optimize block entity fetching
     }
 
     @Override
     public @Nullable BlockEntity getBlockEntity(final BlockPos pos) {
-        return this.blockEntities.get(pos);
+        return this.canvas$getFromBuckets(pos.getX(), pos.getY(), pos.getZ()); // Canvas - optimize block entity fetching
     }
 
     public Map<BlockPos, BlockEntity> getBlockEntities() {
-        return this.blockEntities;
+        return this.canvas$buildBlockEntityMap(); // Canvas - optimize block entity fetching
     }
 
     public void addEntity(final CompoundTag tag) {
@@ -298,7 +301,7 @@ public class ProtoChunk extends ChunkAccess {
 
     @Override
     public void removeBlockEntity(final BlockPos pos) {
-        this.blockEntities.remove(pos);
+        this.canvas$removeFromBuckets(pos.getX(), pos.getY(), pos.getZ()); // Canvas - optimize block entity fetching
         this.pendingBlockEntities.remove(pos);
     }
 
