--- a/net/minecraft/world/level/chunk/ProtoChunk.java
+++ b/net/minecraft/world/level/chunk/ProtoChunk.java
@@ -192,17 +_,20 @@
     @Override
     public void setBlockEntity(BlockEntity blockEntity) {
         this.pendingBlockEntities.remove(blockEntity.getBlockPos());
-        this.blockEntities.put(blockEntity.getBlockPos(), blockEntity);
+        // Canvas start - optimize block entity fetching
+        BlockPos pos = blockEntity.getBlockPos();
+        this.canvas$setBlockEntityInBuckets(pos.getX(), pos.getY(), pos.getZ(), blockEntity);
+        // Canvas end - optimize block entity fetching
     }
 
     @Nullable
     @Override
     public BlockEntity getBlockEntity(BlockPos pos) {
-        return this.blockEntities.get(pos);
+        return this.canvas$getFromBuckets(pos.getX(), pos.getY(), pos.getZ()); // Canvas - optimize block entity fetching
     }
 
     public Map<BlockPos, BlockEntity> getBlockEntities() {
-        return this.blockEntities;
+        return this.canvas$buildBlockEntityMap(); // Canvas - optimize block entity fetching
     }
 
     public void addEntity(CompoundTag tag) {
@@ -303,7 +_,7 @@
 
     @Override
     public void removeBlockEntity(BlockPos pos) {
-        this.blockEntities.remove(pos);
+        this.canvas$removeFromBuckets(pos.getX(), pos.getY(), pos.getZ()); // Canvas - optimize block entity fetching
         this.pendingBlockEntities.remove(pos);
     }
 
