--- a/src/main/java/org/purpurmc/purpur/command/TPSBarCommand.java
+++ b/src/main/java/org/purpurmc/purpur/command/TPSBarCommand.java
@@ -1,6 +_,7 @@
 package org.purpurmc.purpur.command;
 
 import com.mojang.brigadier.CommandDispatcher;
+import io.canvasmc.canvas.Config;
 import net.kyori.adventure.text.Component;
 import net.kyori.adventure.text.format.NamedTextColor;
 import net.kyori.adventure.text.minimessage.MiniMessage;
@@ -29,7 +_,7 @@
 
     private static int execute(CommandSourceStack sender, Collection<ServerPlayer> targets) {
         for (ServerPlayer player : targets) {
-            boolean result = TPSBarTask.instance().togglePlayer(player.getBukkitEntity());
+            boolean result = TPSBarTask.instance().togglePlayer(player.getBukkitEntity()) && (!Config.INSTANCE.ticking.enableThreadedRegionizing || TPSBarTask.RegionBarTask.instance().togglePlayer(player.getBukkitEntity())); // Canvas - Threaded Regions
             player.tpsBar(result);
 
             Component output = MiniMessage.miniMessage().deserialize(PurpurConfig.tpsbarCommandOutput,
