From 9e9970d708edecf8556f1b2db11dec8fc04b5815 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/animal/fish/AbstractFish.java b/net/minecraft/world/entity/animal/fish/AbstractFish.java
index 62303a59d3e0ec28bc3d6bf6fc459527ad80b1f1..6375c150d4a0a0be0f1aa6b5b9f26a8d25f93416 100644
--- a/net/minecraft/world/entity/animal/fish/AbstractFish.java
+++ b/net/minecraft/world/entity/animal/fish/AbstractFish.java
@@ -170,7 +170,7 @@ public abstract class AbstractFish extends WaterAnimal implements Bucketable {
 
         @Override
         public void tick() {
-            if (this.fish.isEyeInFluid(FluidTags.WATER)) {
+            if (this.fish.canvas$isEyeInWater()) { // Canvas - optimize fluidOnEyes
                 this.fish.setDeltaMovement(this.fish.getDeltaMovement().add(0.0, 0.005, 0.0));
             }
 
