From 5462f8f54a2af66b338154bdab1b4b47bd772d5d Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/animal/fish/Pufferfish.java b/net/minecraft/world/entity/animal/fish/Pufferfish.java
index 9981452e2dda0675defd279c4460031dae5fd25f..9916ed69c962d0baf981de5a57ccdd53e1cb5421 100644
--- a/net/minecraft/world/entity/animal/fish/Pufferfish.java
+++ b/net/minecraft/world/entity/animal/fish/Pufferfish.java
@@ -160,6 +160,7 @@ public class Pufferfish extends AbstractFish {
         int puffState = this.getPuffState();
         if (player instanceof ServerPlayer serverPlayer
             && puffState > 0
+            && (!io.canvasmc.canvas.Config.INSTANCE.fixes.mc155509 || this.isAlive()) // Canvas - MC-155509
             && player.hurtServer(serverPlayer.level(), this.damageSources().mobAttack(this), 1 + puffState)) {
             if (!this.isSilent()) {
                 serverPlayer.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.PUFFER_FISH_STING, 0.0F));
