--- a/net/minecraft/world/entity/animal/fish/Pufferfish.java
+++ b/net/minecraft/world/entity/animal/fish/Pufferfish.java
@@ -160,6 +_,7 @@
         int puffState = this.getPuffState();
         if (player instanceof ServerPlayer serverPlayer
             && puffState > 0
+            && (!io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaFixes.mc155509 || this.isAlive()) // Canvas - MC-155509
             && player.hurtServer(serverPlayer.level(), this.damageSources().mobAttack(this), 1 + puffState)) {
             if (!this.isSilent()) {
                 serverPlayer.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.PUFFER_FISH_STING, 0.0F));
