--- a/net/minecraft/world/level/levelgen/Aquifer.java
+++ b/net/minecraft/world/level/levelgen/Aquifer.java
@@ -99,6 +_,259 @@
         private static final int[][] SURFACE_SAMPLING_OFFSETS_IN_CHUNKS = new int[][]{
             {0, 0}, {-2, -1}, {-1, -1}, {0, -1}, {1, -1}, {-3, 0}, {-2, 0}, {-1, 0}, {1, 0}, {-2, 1}, {-1, 1}, {0, 1}, {1, 1}
         };
+        // Canvas start - optimize aquifer
+        private int c2me$packed1;
+        private int c2me$packed2;
+        private int c2me$packed3;
+        private int c2me$packed4;
+
+        private double c2me$mutableDoubleThingy;
+        private short[] c2me$packedBlockPositions;
+
+        private static int c2me$unpackPackedX(int packed) {
+            return packed >> 8;
+        }
+
+        private static int c2me$unpackPackedY(int packed) {
+            return (packed >> 4) & 0b1111;
+        }
+
+        private static int c2me$unpackPackedZ(int packed) {
+            return packed & 0b1111;
+        }
+
+        private static int c2me$unpackPackedDist(int packed) {
+            return packed >> 20;
+        }
+
+        private static int c2me$unpackPackedPosIdx(int packed) {
+            return packed & 0xffff;
+        }
+
+        private @Nullable BlockState aquiferExtracted$applyPost(DensityFunction.FunctionContext pos, double density, int j, int i, int k) {
+            Aquifer.FluidStatus fluidLevel2 = this.c2me$getWaterLevelIndexed(c2me$unpackPackedPosIdx(this.c2me$packed1));
+            double d = similarity(c2me$unpackPackedDist(this.c2me$packed1), c2me$unpackPackedDist(this.c2me$packed2));
+            BlockState blockState = fluidLevel2.at(j);
+            if (d <= 0.0) {
+                if (d >= FLOWING_UPDATE_SIMULARITY) {
+                    Aquifer.FluidStatus fluidLevel3 = this.c2me$getWaterLevelIndexed(c2me$unpackPackedPosIdx(this.c2me$packed2));
+                    this.shouldScheduleFluidUpdate = !fluidLevel2.equals(fluidLevel3);
+                } else {
+                    this.shouldScheduleFluidUpdate = false;
+                }
+                return blockState;
+            } else if (blockState.is(Blocks.WATER) && this.globalFluidPicker.computeFluid(i, j - 1, k).at(j - 1).is(Blocks.LAVA)) {
+                this.shouldScheduleFluidUpdate = true;
+                return blockState;
+            } else {
+//            MutableDouble mutableDouble = new MutableDouble(Double.NaN); // 234MB/s alloc rate at 480 cps
+                this.c2me$mutableDoubleThingy = Double.NaN;
+                Aquifer.FluidStatus fluidLevel3 = this.c2me$getWaterLevelIndexed(c2me$unpackPackedPosIdx(this.c2me$packed2));
+                double e = d * this.c2me$calculateDensityModified(pos, fluidLevel2, fluidLevel3);
+                if (density + e > 0.0) {
+                    this.shouldScheduleFluidUpdate = false;
+                    return null;
+                } else {
+                    return aquiferExtracted$getFinalBlockState(pos, density, d, fluidLevel2, fluidLevel3, blockState);
+                }
+            }
+        }
+
+        private BlockState aquiferExtracted$getFinalBlockState(DensityFunction.FunctionContext pos, double density, double d, Aquifer.FluidStatus fluidLevel2, Aquifer.FluidStatus fluidLevel3, BlockState blockState) {
+            Aquifer.FluidStatus fluidLevel4 = this.c2me$getWaterLevelIndexed(c2me$unpackPackedPosIdx(this.c2me$packed3));
+            int dist1 = c2me$unpackPackedDist(this.c2me$packed1);
+            int dist2 = c2me$unpackPackedDist(this.c2me$packed2);
+            int dist3 = c2me$unpackPackedDist(this.c2me$packed3);
+            int dist4 = c2me$unpackPackedDist(this.c2me$packed4);
+            double f = similarity(dist1, dist3);
+            if (aquiferExtracted$extractedCheckFG(pos, density, d, fluidLevel2, f, fluidLevel4)) return null;
+
+            double h = similarity(dist2, dist3);
+            if (aquiferExtracted$extractedCheckFG(pos, density, d, fluidLevel3, h, fluidLevel4)) return null;
+
+//        this.shouldScheduleFluidUpdate = true;
+            boolean bl = !fluidLevel2.equals(fluidLevel3);
+            boolean bl2 = h >= FLOWING_UPDATE_SIMULARITY && !fluidLevel3.equals(fluidLevel4);
+            boolean bl3 = f >= FLOWING_UPDATE_SIMULARITY && !fluidLevel2.equals(fluidLevel4);
+            if (!bl && !bl2 && !bl3) {
+                this.shouldScheduleFluidUpdate = f >= FLOWING_UPDATE_SIMULARITY
+                    && similarity(dist1, dist4) >= FLOWING_UPDATE_SIMULARITY
+                    && !fluidLevel2.equals(this.c2me$getWaterLevelIndexed(c2me$unpackPackedPosIdx(this.c2me$packed4)));
+            } else {
+                this.shouldScheduleFluidUpdate = true;
+            }
+            return blockState;
+        }
+
+        private boolean aquiferExtracted$extractedCheckFG(DensityFunction.FunctionContext pos, double density, double d, Aquifer.FluidStatus fluidLevel2, double f, Aquifer.FluidStatus fluidLevel4) {
+            if (f > 0.0) {
+                double g = d * f * this.c2me$calculateDensityModified(pos, fluidLevel2, fluidLevel4);
+                if (density + g > 0.0) {
+                    this.shouldScheduleFluidUpdate = false;
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        private void aquiferExtracted$refreshDistPosIdx(int x, int y, int z) {
+            int gx = (x - 5) >> 4;
+            int gy = Math.floorDiv(y + 1, 12);
+            int gz = (z - 5) >> 4;
+            int A = Integer.MAX_VALUE;
+            int B = Integer.MAX_VALUE;
+            int C = Integer.MAX_VALUE;
+            int D = Integer.MAX_VALUE;
+
+            int index = 12; // 12 max
+            for (int offY = -1; offY <= 1; ++offY) {
+                int gymul = (gy + offY) * 12;
+                for (int offZ = 0; offZ <= 1; ++offZ) {
+                    int gzmul = (gz + offZ) << 4;
+
+                    int index0 = index - 1;
+                    int posIdx0 = this.getIndex(gx, gy + offY, gz + offZ);
+                    int position0 = this.c2me$packedBlockPositions[posIdx0];
+                    int dx0 = (gx << 4) + c2me$unpackPackedX(position0) - x;
+                    int dy0 = gymul + c2me$unpackPackedY(position0) - y;
+                    int dz0 = gzmul + c2me$unpackPackedZ(position0) - z;
+                    int dist_0 = dx0 * dx0 + dy0 * dy0 + dz0 * dz0;
+
+                    int index1 = index - 2;
+                    int posIdx1 = posIdx0 + 1;
+                    int position1 = this.c2me$packedBlockPositions[posIdx1];
+                    int dx1 = ((gx + 1) << 4) + c2me$unpackPackedX(position1) - x;
+                    int dy1 = gymul + c2me$unpackPackedY(position1) - y;
+                    int dz1 = gzmul + c2me$unpackPackedZ(position1) - z;
+                    int dist_1 = dx1 * dx1 + dy1 * dy1 + dz1 * dz1;
+
+                    int p0 = (dist_0 << 20) | (index0 << 16) | posIdx0;
+                    if (p0 <= C) {
+                        int n01 = Math.max(A, p0);
+                        A = Math.min(A, p0);
+
+                        int n02 = Math.max(B, n01);
+                        B = Math.min(B, n01);
+
+                        int n03 = Math.max(C, n02);
+                        C = Math.min(C, n02);
+
+                        D = Math.min(D, n03);
+                    }
+
+                    int p1 = (dist_1 << 20) | (index1 << 16) | posIdx1;
+                    if (p1 <= C) {
+                        int n11 = Math.max(A, p1);
+                        A = Math.min(A, p1);
+
+                        int n12 = Math.max(B, n11);
+                        B = Math.min(B, n11);
+
+                        int n13 = Math.max(C, n12);
+                        C = Math.min(C, n12);
+
+                        D = Math.min(D, n13);
+                    }
+
+                    index -= 2;
+                }
+            }
+
+//        this.c2me$dist1 = A >> 20;
+//        this.c2me$dist2 = B >> 20;
+//        this.c2me$dist3 = C >> 20;
+//        this.c2me$pos1 = this.blockPositions[A & 0xffff];
+//        this.c2me$pos2 = this.blockPositions[B & 0xffff];
+//        this.c2me$pos3 = this.blockPositions[C & 0xffff];
+            this.c2me$packed1 = A;
+            this.c2me$packed2 = B;
+            this.c2me$packed3 = C;
+            this.c2me$packed4 = D;
+        }
+
+        private Aquifer.FluidStatus c2me$getWaterLevelIndexed(int index) {
+            return this.getAquiferStatus(index);
+        }
+
+        private double c2me$calculateDensityModified(
+            DensityFunction.FunctionContext pos, Aquifer.FluidStatus fluidLevel, Aquifer.FluidStatus fluidLevel2
+        ) {
+            int i = pos.blockY();
+            BlockState blockState = fluidLevel.at(i);
+            BlockState blockState2 = fluidLevel2.at(i);
+            if ((!blockState.is(Blocks.LAVA) || !blockState2.is(Blocks.WATER)) && (!blockState.is(Blocks.WATER) || !blockState2.is(Blocks.LAVA))) {
+                int j = Math.abs(fluidLevel.fluidLevel - fluidLevel2.fluidLevel);
+                if (j == 0) {
+                    return 0.0;
+                } else {
+                    double d = 0.5 * (double)(fluidLevel.fluidLevel + fluidLevel2.fluidLevel);
+                    final double q = aquiferExtracted$getQ(i, d, j);
+
+                    return aquiferExtracted$postCalculateDensityModified(pos, q);
+                }
+            } else {
+                return 2.0;
+            }
+        }
+
+        private double aquiferExtracted$postCalculateDensity(DensityFunction.FunctionContext pos, MutableDouble mutableDouble, double q) {
+            double r;
+            if (!(q < -2.0) && !(q > 2.0)) {
+                double s = mutableDouble.getValue();
+                if (Double.isNaN(s)) {
+                    double t = this.barrierNoise.compute(pos);
+                    mutableDouble.setValue(t);
+                    r = t;
+                } else {
+                    r = s;
+                }
+            } else {
+                r = 0.0;
+            }
+
+            return 2.0 * (r + q);
+        }
+
+        private double aquiferExtracted$postCalculateDensityModified(DensityFunction.FunctionContext pos, double q) {
+            double r;
+            if (!(q < -2.0) && !(q > 2.0)) {
+                double s = this.c2me$mutableDoubleThingy;
+                if (Double.isNaN(s)) {
+                    double t = this.barrierNoise.compute(pos);
+                    this.c2me$mutableDoubleThingy = t;
+                    r = t;
+                } else {
+                    r = s;
+                }
+            } else {
+                r = 0.0;
+            }
+
+            return 2.0 * (r + q);
+        }
+
+        private static double aquiferExtracted$getQ(double i, double d, double j) {
+            double e = i + 0.5 - d;
+            double f = j / 2.0;
+            double o = f - Math.abs(e);
+            double q;
+            if (e > 0.0) {
+                if (o > 0.0) {
+                    q = o / 1.5;
+                } else {
+                    q = o / 2.5;
+                }
+            } else {
+                double p = 3.0 + o;
+                if (p > 0.0) {
+                    q = p / 3.0;
+                } else {
+                    q = p / 10.0;
+                }
+            }
+            return q;
+        }
+        // Canvas end - optimize aquifer
 
         NoiseBasedAquifer(
             NoiseChunk noiseChunk,
@@ -136,6 +_,44 @@
             );
             int i6 = gridY(i5 + 12) - -1;
             this.skipSamplingAboveY = fromGridY(i6, 11) - 1;
+            // Canvas start - optimize aquifer
+            if (io.canvasmc.canvas.Config.INSTANCE.chunks.optimizeAquifer) {
+                if (this.aquiferLocationCache.length % (this.gridSizeX * this.gridSizeZ) != 0) {
+                    throw new AssertionError("Array length");
+                }
+
+                int sizeY = this.aquiferLocationCache.length / (this.gridSizeX * this.gridSizeZ);
+
+                this.c2me$packedBlockPositions = new short[this.aquiferLocationCache.length];
+
+                final RandomSource random = io.canvasmc.canvas.util.Util.getRandom(this.positionalRandomFactory);
+                // index: y, z, x
+                for (int y = 0; y < sizeY; y++) {
+                    for (int z = 0; z < this.gridSizeZ; z++) {
+                        for (int x = 0; x < this.gridSizeX; x++) {
+                            final int x1 = x + this.minGridX;
+                            final int y1 = y + this.minGridY;
+                            final int z1 = z + this.minGridZ;
+                            io.canvasmc.canvas.util.Util.derive(this.positionalRandomFactory, random, x1, y1, z1);
+                            int r0 = random.nextInt(10);
+                            int r1 = random.nextInt(9);
+                            int r2 = random.nextInt(10);
+                            int x2 = x1 * 16 + r0;
+                            int y2 = y1 * 12 + r1;
+                            int z2 = z1 * 16 + r2;
+                            int index = this.getIndex(x1, y1, z1);
+                            this.aquiferLocationCache[index] = BlockPos.asLong(x2, y2, z2);
+                            this.c2me$packedBlockPositions[index] = (short) ((r0 << 8) | (r1 << 4) | r2);
+                        }
+                    }
+                }
+                for (long blockPosition : this.aquiferLocationCache) {
+                    if (blockPosition == Long.MAX_VALUE) {
+                        throw new AssertionError("Array initialization");
+                    }
+                }
+            }
+            // Canvas end - optimize aquifer
         }
 
         private int getIndex(int gridX, int gridY, int gridZ) {
@@ -147,6 +_,29 @@
 
         @Override
         public @Nullable BlockState computeSubstance(DensityFunction.FunctionContext context, double substance) {
+            // Canvas start - optimize aquifer
+            if (io.canvasmc.canvas.Config.INSTANCE.chunks.optimizeAquifer) {
+                int i = context.blockX();
+                int j = context.blockY();
+                int k = context.blockZ();
+                if (substance > 0.0) {
+                    this.shouldScheduleFluidUpdate = false;
+                    return null;
+                } else {
+                    Aquifer.FluidStatus fluidLevel = this.globalFluidPicker.computeFluid(i, j, k);
+                    if (j > this.skipSamplingAboveY) {
+                        this.shouldScheduleFluidUpdate = false;
+                        return fluidLevel.at(j);
+                    } if (fluidLevel.at(j).is(Blocks.LAVA)) {
+                        this.shouldScheduleFluidUpdate = false;
+                        return Blocks.LAVA.defaultBlockState();
+                    } else {
+                        aquiferExtracted$refreshDistPosIdx(i, j, k);
+                        return aquiferExtracted$applyPost(context, substance, j, i, k);
+                    }
+                }
+            }
+            // Canvas end - optimize aquifer
             if (substance > 0.0) {
                 this.shouldScheduleFluidUpdate = false;
                 return null;
@@ -302,6 +_,27 @@
         private double calculatePressure(
             DensityFunction.FunctionContext context, MutableDouble substance, Aquifer.FluidStatus firstFluid, Aquifer.FluidStatus secondFluid
         ) {
+            // Canvas start - optimize aquifer
+            // DensityFunction.NoisePos pos, MutableDouble mutableDouble, AquiferSampler.FluidLevel fluidLevel, AquiferSampler.FluidLevel fluidLevel2
+            if (io.canvasmc.canvas.Config.INSTANCE.chunks.optimizeAquifer) {
+                int i = context.blockY();
+                BlockState blockState = firstFluid.at(i);
+                BlockState blockState2 = secondFluid.at(i);
+                if ((!blockState.is(Blocks.LAVA) || !blockState2.is(Blocks.WATER)) && (!blockState.is(Blocks.WATER) || !blockState2.is(Blocks.LAVA))) {
+                    int j = Math.abs(firstFluid.fluidLevel - secondFluid.fluidLevel);
+                    if (j == 0) {
+                        return 0.0;
+                    } else {
+                        double d = 0.5 * (double)(firstFluid.fluidLevel + secondFluid.fluidLevel);
+                        final double q = aquiferExtracted$getQ(i, d, j);
+
+                        return aquiferExtracted$postCalculateDensity(context, substance, q);
+                    }
+                } else {
+                    return 2.0;
+                }
+            }
+            // Canvas end - optimize aquifer
             int i = context.blockY();
             BlockState blockState = firstFluid.at(i);
             BlockState blockState1 = secondFluid.at(i);
@@ -438,6 +_,34 @@
         }
 
         private int computeSurfaceLevel(int x, int y, int z, Aquifer.FluidStatus fluidStatus, int maxSurfaceLevel, boolean fluidPresent) {
+            // Canvas start - optimize aquifer
+            if (io.canvasmc.canvas.Config.INSTANCE.chunks.optimizeAquifer) {
+                DensityFunction.SinglePointContext unblendedNoisePos = new DensityFunction.SinglePointContext(x, y, z);
+                double d;
+                double e;
+                if (OverworldBiomeBuilder.isDeepDarkRegion(this.erosion, this.depth, unblendedNoisePos)) {
+                    d = -1.0;
+                    e = -1.0;
+                } else {
+                    int i = maxSurfaceLevel + 8 - y;
+                    double f = fluidPresent ? Mth.clampedLerp((double) i / 64.0, 1.0, 0.0) : 0.0; // inline
+                    double g = Mth.clamp(this.fluidLevelFloodednessNoise.compute(unblendedNoisePos), -1.0, 1.0);
+                    d = g + 0.8 + (f - 1.0) * 1.2; // inline
+                    e = g + 0.3 + (f - 1.0) * 1.1; // inline
+                }
+
+                int i;
+                if (e > 0.0) {
+                    i = fluidStatus.fluidLevel;
+                } else if (d > 0.0) {
+                    i = this.computeRandomizedFluidSurfaceLevel(x, y, z, maxSurfaceLevel);
+                } else {
+                    i = DimensionType.WAY_BELOW_MIN_Y;
+                }
+
+                return i;
+            }
+            // Canvas end - optimize aquifer
             DensityFunction.SinglePointContext singlePointContext = new DensityFunction.SinglePointContext(x, y, z);
             double d;
             double d1;
@@ -482,6 +_,22 @@
         }
 
         private BlockState computeFluidType(int x, int y, int z, Aquifer.FluidStatus fluidStatus, int surfaceLevel) {
+            // Canvas start - optimize aquifer
+            if (io.canvasmc.canvas.Config.INSTANCE.chunks.optimizeAquifer) {
+                BlockState blockState = fluidStatus.fluidType;
+                if (surfaceLevel <= -10 && surfaceLevel != DimensionType.WAY_BELOW_MIN_Y && fluidStatus.fluidType != Blocks.LAVA.defaultBlockState()) {
+                    int k = x >> 6; // floorDiv(blockX, 64)
+                    int l = Math.floorDiv(y, 40);
+                    int m = z >> 6; // floorDiv(blockZ, 64)
+                    double d = this.lavaNoise.compute(new DensityFunction.SinglePointContext(k, l, m));
+                    if (Math.abs(d) > 0.3) {
+                        blockState = Blocks.LAVA.defaultBlockState();
+                    }
+                }
+
+                return blockState;
+            }
+            // Canvas end - optimize aquifer
             BlockState blockState = fluidStatus.fluidType;
             if (surfaceLevel <= -10 && surfaceLevel != DimensionType.WAY_BELOW_MIN_Y && fluidStatus.fluidType != Blocks.LAVA.defaultBlockState()) {
                 int i = 64;
