--- a/net/minecraft/world/level/levelgen/BelowZeroRetrogen.java
+++ b/net/minecraft/world/level/levelgen/BelowZeroRetrogen.java
@@ -10,6 +_,7 @@
 import java.util.function.Predicate;
 import java.util.stream.LongStream;
 import javax.annotation.Nullable;
+import io.canvasmc.canvas.Config;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
 import net.minecraft.core.registries.BuiltInRegistries;
@@ -82,6 +_,11 @@
     }
 
     public void applyBedrockMask(ProtoChunk chunk) {
+        // Canvas start
+        if (Config.INSTANCE.chunks.smoothBedrockLayer) {
+            return;
+        }
+        // Canvas end
         LevelHeightAccessor heightAccessorForGeneration = chunk.getHeightAccessorForGeneration();
         int minY = heightAccessorForGeneration.getMinY();
         int maxY = heightAccessorForGeneration.getMaxY();
