--- a/net/minecraft/world/level/levelgen/DensityFunctions.java
+++ b/net/minecraft/world/level/levelgen/DensityFunctions.java
@@ -278,13 +_,13 @@
         return new DensityFunctions.FindTopSurface(density, upperBound, lowerBound, stepSize);
     }
 
-    private record Ap2(
+    public record Ap2( // Leaf - C2ME - DensityFunction compiler - private -> public
         @Override DensityFunctions.TwoArgumentSimpleFunction.Type type,
         @Override DensityFunction argument1,
         @Override DensityFunction argument2,
         @Override double minValue,
         @Override double maxValue
-    ) implements DensityFunctions.TwoArgumentSimpleFunction {
+    ) implements DensityFunctions.TwoArgumentSimpleFunction { // Leaf - C2ME - DensityFunction compiler
         @Override
         public double compute(final DensityFunction.FunctionContext context) {
             double v1 = this.argument1.compute(context);
@@ -301,14 +_,25 @@
         public void fillArray(final double[] output, final DensityFunction.ContextProvider contextProvider) {
             this.argument1.fillArray(output, contextProvider);
             switch (this.type) {
-                case ADD:
-                    double[] v2 = new double[output.length];
+                // Leaf start - C2ME - DensityFunction compiler
+                case ADD: {
+                    final com.ishland.c2me.opts.dfc.common.gen.jvm.util.DfcObjectCache objectCache =
+                        contextProvider instanceof com.ishland.c2me.opts.dfc.common.ducks.IDfcObjectCacheCapable cacheCapable
+                            ? cacheCapable.c2me$getDfcObjectCache()
+                            : null;
+                    final double[] v2 = objectCache == null ? new double[output.length] : objectCache.getDoubleArray(output.length, false);
                     this.argument2.fillArray(v2, contextProvider);
 
                     for (int i = 0; i < output.length; i++) {
                         output[i] += v2[i];
                     }
+
+                    if (objectCache != null) {
+                        objectCache.recycle(v2);
+                    }
                     break;
+                }
+                // Leaf end - C2ME - DensityFunction compiler
                 case MUL:
                     for (int i = 0; i < output.length; i++) {
                         double v = output[i];
@@ -339,7 +_,7 @@
         }
     }
 
-    enum BeardifierMarker implements DensityFunctions.BeardifierOrMarker {
+    public enum BeardifierMarker implements DensityFunctions.BeardifierOrMarker { // Leaf - C2ME - DensityFunction compiler - package-private -> public
         INSTANCE;
 
         @Override
@@ -372,7 +_,7 @@
         }
     }
 
-    enum BlendAlpha implements DensityFunction.SimpleFunction {
+    public enum BlendAlpha implements DensityFunction.SimpleFunction { // Leaf - C2ME - DensityFunction compiler - package-private -> public
         INSTANCE;
 
         public static final KeyDispatchDataCodec<DensityFunction> CODEC = KeyDispatchDataCodec.of(MapCodec.unit(INSTANCE));
@@ -403,7 +_,7 @@
         }
     }
 
-    enum BlendOffset implements DensityFunction.SimpleFunction {
+    public enum BlendOffset implements DensityFunction.SimpleFunction { // Leaf - C2ME - DensityFunction compiler - package-private -> public
         INSTANCE;
 
         public static final KeyDispatchDataCodec<DensityFunction> CODEC = KeyDispatchDataCodec.of(MapCodec.unit(INSTANCE));
@@ -434,7 +_,7 @@
         }
     }
 
-    protected record Clamp(@Override DensityFunction input, @Override double minValue, @Override double maxValue) implements DensityFunctions.PureTransformer {
+    public record Clamp(@Override DensityFunction input, @Override double minValue, @Override double maxValue) implements DensityFunctions.PureTransformer { // Leaf - C2ME - DensityFunction compiler - protected -> public
         private static final MapCodec<DensityFunctions.Clamp> DATA_CODEC = RecordCodecBuilder.mapCodec(
             i -> i.group(
                     DensityFunction.CODEC.fieldOf("input").forGetter(DensityFunctions.Clamp::input),
@@ -461,7 +_,7 @@
         }
     }
 
-    private record Constant(double value) implements DensityFunction.SimpleFunction {
+    public record Constant(double value) implements DensityFunction.SimpleFunction { // Leaf - C2ME - DensityFunction compiler - private -> public
         private static final KeyDispatchDataCodec<DensityFunctions.Constant> CODEC = DensityFunctions.singleArgumentCodec(
             DensityFunctions.NOISE_VALUE_CODEC, DensityFunctions.Constant::new, DensityFunctions.Constant::value
         );
@@ -493,7 +_,7 @@
         }
     }
 
-    protected static final class EndIslandDensityFunction implements DensityFunction.SimpleFunction {
+    public static final class EndIslandDensityFunction implements DensityFunction.SimpleFunction { // Leaf - C2ME - DensityFunction compiler - protected -> public
         public static final KeyDispatchDataCodec<DensityFunctions.EndIslandDensityFunction> CODEC = KeyDispatchDataCodec.of(
             MapCodec.unit(new DensityFunctions.EndIslandDensityFunction(0L))
         );
@@ -577,9 +_,27 @@
         public KeyDispatchDataCodec<? extends DensityFunction> codec() {
             return CODEC;
         }
+
+        // Leaf start - C2ME - DensityFunction compiler
+        @Override
+        public boolean equals(final Object object) {
+            if (this == object) {
+                return true;
+            }
+            if (object == null || this.getClass() != object.getClass()) {
+                return false;
+            }
+            return java.util.Objects.equals(this.islandNoise, ((DensityFunctions.EndIslandDensityFunction)object).islandNoise);
+        }
+
+        @Override
+        public int hashCode() {
+            return java.util.Objects.hashCode(this.islandNoise);
+        }
+        // Leaf end - C2ME - DensityFunction compiler
     }
 
-    private record FindTopSurface(DensityFunction density, DensityFunction upperBound, int lowerBound, int cellHeight) implements DensityFunction {
+    public record FindTopSurface(DensityFunction density, DensityFunction upperBound, int lowerBound, int cellHeight) implements DensityFunction { // Leaf - C2ME - DensityFunction compiler - private -> public
         private static final MapCodec<DensityFunctions.FindTopSurface> DATA_CODEC = RecordCodecBuilder.mapCodec(
             i -> i.group(
                     DensityFunction.CODEC.fieldOf("density").forGetter(DensityFunctions.FindTopSurface::density),
@@ -668,7 +_,7 @@
         }
     }
 
-    private record IntervalSelect(DensityFunction input, DoubleList thresholds, List<DensityFunction> functions) implements DensityFunction {
+    public record IntervalSelect(DensityFunction input, DoubleList thresholds, List<DensityFunction> functions) implements DensityFunction { // Leaf - C2ME - DensityFunction compiler - private -> public
         private static final Codec<DoubleList> THRESHOLDS_CODEC = DensityFunctions.NOISE_VALUE_CODEC.listOf().xmap(DoubleArrayList::new, Function.identity());
         public static final MapCodec<DensityFunctions.IntervalSelect> DATA_CODEC = RecordCodecBuilder.<DensityFunctions.IntervalSelect>mapCodec(
                 i -> i.group(
@@ -755,8 +_,8 @@
         }
     }
 
-    protected record Mapped(DensityFunctions.Mapped.Type type, @Override DensityFunction input, @Override double minValue, @Override double maxValue)
-        implements DensityFunctions.PureTransformer {
+    public record Mapped(DensityFunctions.Mapped.Type type, @Override DensityFunction input, @Override double minValue, @Override double maxValue)
+        implements DensityFunctions.PureTransformer { // Leaf - C2ME - DensityFunction compiler - protected -> public
         public static DensityFunctions.Mapped create(final DensityFunctions.Mapped.Type type, final DensityFunction input) {
             double minValue = input.minValue();
             double maxValue = input.maxValue();
@@ -828,7 +_,8 @@
         }
     }
 
-    record Marker(@Override DensityFunctions.Marker.Type type, @Override DensityFunction wrapped) implements DensityFunctions.MarkerOrMarked {
+    public record Marker(@Override DensityFunctions.Marker.Type type, @Override DensityFunction wrapped)
+        implements DensityFunctions.MarkerOrMarked, com.ishland.c2me.opts.dfc.common.ducks.IFastCacheLike { // Leaf - C2ME - DensityFunction compiler - package-private -> public
         @Override
         public double compute(final DensityFunction.FunctionContext context) {
             return this.wrapped.compute(context);
@@ -849,6 +_,77 @@
             return this.type == DensityFunctions.Marker.Type.BlendDensity ? Double.POSITIVE_INFINITY : this.wrapped.maxValue();
         }
 
+        // Leaf start - C2ME - DensityFunction compiler
+        @Override
+        public double c2me$getCached(
+            final int x,
+            final int y,
+            final int z,
+            final com.ishland.c2me.opts.dfc.common.ast.EvalType evalType
+        ) {
+            return Double.longBitsToDouble(com.ishland.c2me.opts.dfc.common.ducks.IFastCacheLike.CACHE_MISS_NAN_BITS);
+        }
+
+        @Override
+        public boolean c2me$getCached(
+            final double[] result,
+            final int[] x,
+            final int[] y,
+            final int[] z,
+            final com.ishland.c2me.opts.dfc.common.ast.EvalType evalType
+        ) {
+            return false;
+        }
+
+        @Override
+        public void c2me$cache(
+            final int x,
+            final int y,
+            final int z,
+            final com.ishland.c2me.opts.dfc.common.ast.EvalType evalType,
+            final double cached
+        ) {
+        }
+
+        @Override
+        public void c2me$cache(
+            final double[] result,
+            final int[] x,
+            final int[] y,
+            final int[] z,
+            final com.ishland.c2me.opts.dfc.common.ast.EvalType evalType
+        ) {
+        }
+
+        @Override
+        public boolean c2me$isActualCache() {
+            return switch (this.type) {
+                case FlatCache, CacheAllInCell, Cache2D, CacheOnce -> true;
+                default -> false;
+            };
+        }
+
+        @Override
+        public String c2me$describeCacheLike() {
+            return "wrapping, " + this.type.getSerializedName();
+        }
+
+        @Override
+        public DensityFunction c2me$getDelegate() {
+            return this.wrapped;
+        }
+
+        @Override
+        public DensityFunction c2me$withDelegate(final DensityFunction delegate) {
+            return new DensityFunctions.Marker(this.type, delegate);
+        }
+
+        @Override
+        public DensityFunction mapAll(final DensityFunction.Visitor visitor) {
+            return visitor.apply(this.c2me$withDelegate(this.wrapped.mapAll(visitor)));
+        }
+        // Leaf end - C2ME - DensityFunction compiler
+
         public enum Type implements StringRepresentable {
             Interpolated("interpolated"),
             FlatCache("flat_cache"),
@@ -889,9 +_,9 @@
         }
     }
 
-    private record MulOrAdd(
+    public record MulOrAdd(
         DensityFunctions.MulOrAdd.Type specificType, @Override DensityFunction input, @Override double minValue, @Override double maxValue, double argument
-    ) implements DensityFunctions.TwoArgumentSimpleFunction, DensityFunctions.PureTransformer {
+    ) implements DensityFunctions.TwoArgumentSimpleFunction, DensityFunctions.PureTransformer { // Leaf - C2ME - DensityFunction compiler - private -> public
         @Override
         public DensityFunctions.TwoArgumentSimpleFunction.Type type() {
             return this.specificType == DensityFunctions.MulOrAdd.Type.MUL
@@ -944,7 +_,7 @@
         }
     }
 
-    protected record Noise(DensityFunction.NoiseHolder noise, @Deprecated double xzScale, double yScale) implements DensityFunction {
+    public record Noise(DensityFunction.NoiseHolder noise, @Deprecated double xzScale, double yScale) implements DensityFunction { // Leaf - C2ME - DensityFunction compiler - protected -> public
         public static final MapCodec<DensityFunctions.Noise> DATA_CODEC = RecordCodecBuilder.mapCodec(
             i -> i.group(
                     DensityFunction.NoiseHolder.CODEC.fieldOf("noise").forGetter(DensityFunctions.Noise::noise),
@@ -1006,8 +_,8 @@
         double transform(final double input);
     }
 
-    private record RangeChoice(DensityFunction input, double minInclusive, double maxExclusive, DensityFunction whenInRange, DensityFunction whenOutOfRange)
-        implements DensityFunction {
+    public record RangeChoice(DensityFunction input, double minInclusive, double maxExclusive, DensityFunction whenInRange, DensityFunction whenOutOfRange)
+        implements DensityFunction { // Leaf - C2ME - DensityFunction compiler - private -> public
         public static final MapCodec<DensityFunctions.RangeChoice> DATA_CODEC = RecordCodecBuilder.mapCodec(
             i -> i.group(
                     DensityFunction.CODEC.fieldOf("input").forGetter(DensityFunctions.RangeChoice::input),
@@ -1063,7 +_,7 @@
         }
     }
 
-    protected record Shift(@Override DensityFunction.NoiseHolder offsetNoise) implements DensityFunctions.ShiftNoise {
+    public record Shift(@Override DensityFunction.NoiseHolder offsetNoise) implements DensityFunctions.ShiftNoise { // Leaf - C2ME - DensityFunction compiler - protected -> public
         private static final KeyDispatchDataCodec<DensityFunctions.Shift> CODEC = DensityFunctions.singleArgumentCodec(
             DensityFunction.NoiseHolder.CODEC, DensityFunctions.Shift::new, DensityFunctions.Shift::offsetNoise
         );
@@ -1084,7 +_,7 @@
         }
     }
 
-    protected record ShiftA(@Override DensityFunction.NoiseHolder offsetNoise) implements DensityFunctions.ShiftNoise {
+    public record ShiftA(@Override DensityFunction.NoiseHolder offsetNoise) implements DensityFunctions.ShiftNoise { // Leaf - C2ME - DensityFunction compiler - protected -> public
         private static final KeyDispatchDataCodec<DensityFunctions.ShiftA> CODEC = DensityFunctions.singleArgumentCodec(
             DensityFunction.NoiseHolder.CODEC, DensityFunctions.ShiftA::new, DensityFunctions.ShiftA::offsetNoise
         );
@@ -1105,7 +_,7 @@
         }
     }
 
-    protected record ShiftB(@Override DensityFunction.NoiseHolder offsetNoise) implements DensityFunctions.ShiftNoise {
+    public record ShiftB(@Override DensityFunction.NoiseHolder offsetNoise) implements DensityFunctions.ShiftNoise { // Leaf - C2ME - DensityFunction compiler - protected -> public
         private static final KeyDispatchDataCodec<DensityFunctions.ShiftB> CODEC = DensityFunctions.singleArgumentCodec(
             DensityFunction.NoiseHolder.CODEC, DensityFunctions.ShiftB::new, DensityFunctions.ShiftB::offsetNoise
         );
@@ -1149,9 +_,9 @@
         }
     }
 
-    protected record ShiftedNoise(
+    public record ShiftedNoise(
         DensityFunction shiftX, DensityFunction shiftY, DensityFunction shiftZ, double xzScale, double yScale, DensityFunction.NoiseHolder noise
-    ) implements DensityFunction {
+    ) implements DensityFunction { // Leaf - C2ME - DensityFunction compiler - protected -> public
         private static final MapCodec<DensityFunctions.ShiftedNoise> DATA_CODEC = RecordCodecBuilder.mapCodec(
             i -> i.group(
                     DensityFunction.CODEC.fieldOf("shift_x").forGetter(DensityFunctions.ShiftedNoise::shiftX),
@@ -1401,7 +_,7 @@
         }
     }
 
-    private record YClampedGradient(int fromY, int toY, double fromValue, double toValue) implements DensityFunction.SimpleFunction {
+    public record YClampedGradient(int fromY, int toY, double fromValue, double toValue) implements DensityFunction.SimpleFunction { // Leaf - C2ME - DensityFunction compiler - private -> public
         private static final MapCodec<DensityFunctions.YClampedGradient> DATA_CODEC = RecordCodecBuilder.mapCodec(
             i -> i.group(
                     Codec.intRange(DimensionType.MIN_Y * 2, DimensionType.MAX_Y * 2).fieldOf("from_y").forGetter(DensityFunctions.YClampedGradient::fromY),
