--- a/net/minecraft/world/level/levelgen/NoiseRouterData.java
+++ b/net/minecraft/world/level/levelgen/NoiseRouterData.java
@@ -501,8 +_,8 @@
         return DensityFunctions.lerp(densityFunction2, bottomDelta, densityFunction);
     }
 
-    protected static final class QuantizedSpaghettiRarity {
-        protected static double getSphaghettiRarity2D(double value) {
+    public static final class QuantizedSpaghettiRarity { // Canvas - protected -> public
+        public static double getSphaghettiRarity2D(double value) { // Canvas - protected -> public
             if (value < -0.75) {
                 return 0.5;
             } else if (value < -0.5) {
@@ -514,7 +_,7 @@
             }
         }
 
-        protected static double getSpaghettiRarity3D(double value) {
+        public static double getSpaghettiRarity3D(double value) { // Canvas - protected -> public
             if (value < -0.5) {
                 return 0.75;
             } else if (value < 0.0) {
