--- a/net/minecraft/world/level/levelgen/NoiseChunk.java
+++ b/net/minecraft/world/level/levelgen/NoiseChunk.java
@@ -21,7 +_,7 @@
 import net.minecraft.world.level.levelgen.material.MaterialRuleList;
 import org.jspecify.annotations.Nullable;
 
-public class NoiseChunk implements DensityFunction.FunctionContext, DensityFunction.ContextProvider {
+public class NoiseChunk implements DensityFunction.FunctionContext, DensityFunction.ContextProvider, com.ishland.c2me.base.mixin.access.IChunkNoiseSampler, com.ishland.c2me.opts.dfc.common.ducks.IDfcObjectCacheCapable, com.ishland.c2me.opts.dfc.common.ducks.ICoordinatesFilling, com.ishland.c2me.opts.dfc.common.ducks.IPreloadedCoordinates { // Leaf - C2ME - DensityFunction compiler
     private final int cellCountXZ;
     private final int cellCountY;
     private final int cellNoiseMinY;
@@ -57,7 +_,25 @@
     private long interpolationCounter;
     private long arrayInterpolationCounter;
     private int arrayIndex;
-    private final DensityFunction.ContextProvider sliceFillingContextProvider = new DensityFunction.ContextProvider() {
+
+    // Leaf start - C2ME - DensityFunction compiler
+    private int c2me$oldStartBlockX = Integer.MAX_VALUE;
+    private int c2me$oldStartBlockY = Integer.MAX_VALUE;
+    private int c2me$oldStartBlockZ = Integer.MAX_VALUE;
+    private int[] c2me$cachedXArray;
+    private int[] c2me$cachedYArray;
+    private int[] c2me$cachedZArray;
+    NoiseChunk.NoiseInterpolator[] c2me$interpolatorsArray;
+    private final com.ishland.c2me.opts.dfc.common.gen.jvm.util.DfcObjectCache c2Me$dfcObjectCache =
+        io.canvasmc.canvas.GlobalConfiguration.getInstance().chunkSystem.densityFunctionCompiler
+            ? new com.ishland.c2me.opts.dfc.common.gen.jvm.util.DfcObjectCache.Impl()
+            : com.ishland.c2me.opts.dfc.common.gen.jvm.util.DfcObjectCache.Noop.INSTANCE;
+    private final Map<com.ishland.c2me.opts.dfc.common.gen.jvm.CompiledEntry, com.ishland.c2me.opts.dfc.common.gen.jvm.CompiledEntry>
+        c2me$compiledEntryCache = new it.unimi.dsi.fastutil.objects.Reference2ReferenceOpenHashMap<>();
+    // Leaf end - C2ME - DensityFunction compiler
+
+    private final DensityFunction.ContextProvider sliceFillingContextProvider = new NoiseChunk.SliceFillingContextProvider(); // Leaf - C2ME - DensityFunction compiler
+    private final class SliceFillingContextProvider implements DensityFunction.ContextProvider, com.ishland.c2me.opts.dfc.common.ducks.IDfcObjectCacheCapable, com.ishland.c2me.opts.dfc.common.ducks.ICoordinatesFilling { // Leaf - C2ME - DensityFunction compiler
         @Override
         public DensityFunction.FunctionContext forIndex(final int cellYIndex) {
             NoiseChunk.this.cellStartBlockY = (cellYIndex + NoiseChunk.this.cellNoiseMinY) * NoiseChunk.this.cellHeight;
@@ -77,7 +_,23 @@
                 output[cellYIndex] = function.compute(NoiseChunk.this);
             }
         }
-    };
+
+        // Leaf start - C2ME - DensityFunction compiler
+        @Override
+        public com.ishland.c2me.opts.dfc.common.gen.jvm.util.DfcObjectCache c2me$getDfcObjectCache() {
+            return NoiseChunk.this.c2me$getDfcObjectCache();
+        }
+
+        @Override
+        public void c2me$fillCoordinates(final int[] x, final int[] y, final int[] z) {
+            for (int i = 0; i < NoiseChunk.this.cellCountY + 1; i++) {
+                x[i] = NoiseChunk.this.cellStartBlockX + NoiseChunk.this.inCellX;
+                y[i] = (i + NoiseChunk.this.cellNoiseMinY) * NoiseChunk.this.cellHeight;
+                z[i] = NoiseChunk.this.cellStartBlockZ + NoiseChunk.this.inCellZ;
+            }
+        }
+    }
+    // Leaf end - C2ME - DensityFunction compiler
 
     public static NoiseChunk forChunk(
         final ChunkAccess chunk,
@@ -140,7 +_,7 @@
         }
 
         NoiseRouter router = randomState.router();
-        NoiseRouter wrappedRouter = router.mapAll(this::wrap);
+        NoiseRouter wrappedRouter = router.mapAll(this.c2me$wrapVisitor(this::wrap)); // Leaf - C2ME - DensityFunction compiler
         this.preliminarySurfaceLevel = wrappedRouter.preliminarySurfaceLevel();
         if (!settings.isAquifersEnabled()) {
             this.aquifer = Aquifer.createDisabled(globalFluidPicker);
@@ -153,10 +_,16 @@
         }
 
         List<NoiseChunk.BlockStateFiller> builder = new ArrayList<>();
-        DensityFunction fullNoiseValue = DensityFunctions.cacheAllInCell(
-                DensityFunctions.add(wrappedRouter.finalDensity(), DensityFunctions.BeardifierMarker.INSTANCE)
-            )
-            .mapAll(this::wrap);
+        // Leaf start - C2ME - DensityFunction compiler
+        DensityFunction fullNoiseValueTemp = wrappedRouter.c2me$getFinalFinalDensity();
+        if (fullNoiseValueTemp == null) {
+            fullNoiseValueTemp = DensityFunctions.cacheAllInCell(
+                    DensityFunctions.add(wrappedRouter.finalDensity(), DensityFunctions.BeardifierMarker.INSTANCE)
+                )
+                .mapAll(this.c2me$wrapVisitor(this::wrap));
+        }
+        final DensityFunction fullNoiseValue = fullNoiseValueTemp;
+        // Leaf end - C2ME - DensityFunction compiler
         this.fullNoiseDensity = fullNoiseValue;
         builder.add(context -> this.aquifer.computeSubstance(context, fullNoiseValue.compute(context)));
         if (settings.oreVeinsEnabled()) {
@@ -168,12 +_,14 @@
 
     protected Climate.Sampler cachedClimateSampler(final NoiseRouter noises, final List<Climate.ParameterPoint> spawnTarget) {
         return new Climate.Sampler(
-            noises.temperature().mapAll(this::wrap),
-            noises.vegetation().mapAll(this::wrap),
-            noises.continents().mapAll(this::wrap),
-            noises.erosion().mapAll(this::wrap),
-            noises.depth().mapAll(this::wrap),
-            noises.ridges().mapAll(this::wrap),
+            // Leaf start - C2ME - DensityFunction compiler
+            noises.temperature().mapAll(this.c2me$wrapVisitor(this::wrap)),
+            noises.vegetation().mapAll(this.c2me$wrapVisitor(this::wrap)),
+            noises.continents().mapAll(this.c2me$wrapVisitor(this::wrap)),
+            noises.erosion().mapAll(this.c2me$wrapVisitor(this::wrap)),
+            noises.depth().mapAll(this.c2me$wrapVisitor(this::wrap)),
+            noises.ridges().mapAll(this.c2me$wrapVisitor(this::wrap)),
+            // Leaf end - C2ME - DensityFunction compiler
             spawnTarget
         );
     }
@@ -201,6 +_,179 @@
         return this.cellStartBlockZ + this.inCellZ;
     }
 
+    // Leaf start - C2ME - DensityFunction compiler
+    @Override
+    public com.ishland.c2me.opts.dfc.common.gen.jvm.util.DfcObjectCache c2me$getDfcObjectCache() {
+        return this.c2Me$dfcObjectCache;
+    }
+
+    @Override
+    public void c2me$fillCoordinates(final int[] x, final int[] y, final int[] z) {
+        int index = 0;
+        for (int yInCell = this.cellHeight - 1; yInCell >= 0; yInCell--) {
+            final int blockY = this.cellStartBlockY + yInCell;
+            for (int xInCell = 0; xInCell < this.cellWidth; xInCell++) {
+                final int blockX = this.cellStartBlockX + xInCell;
+                for (int zInCell = 0; zInCell < this.cellWidth; zInCell++) {
+                    x[index] = blockX;
+                    y[index] = blockY;
+                    z[index] = this.cellStartBlockZ + zInCell;
+                    index++;
+                }
+            }
+        }
+    }
+
+    private void c2me$reloadCachedArrays() {
+        if (this.c2me$cachedXArray == null || this.c2me$cachedYArray == null || this.c2me$cachedZArray == null) {
+            final int length = this.cellWidth * this.cellWidth * this.cellHeight;
+            this.c2me$cachedXArray = this.c2Me$dfcObjectCache.getIntArray(length, false);
+            this.c2me$cachedYArray = this.c2Me$dfcObjectCache.getIntArray(length, false);
+            this.c2me$cachedZArray = this.c2Me$dfcObjectCache.getIntArray(length, false);
+        }
+        if (this.c2me$oldStartBlockX != this.cellStartBlockX
+            || this.c2me$oldStartBlockY != this.cellStartBlockY
+            || this.c2me$oldStartBlockZ != this.cellStartBlockZ) {
+            this.c2me$fillCoordinates(this.c2me$cachedXArray, this.c2me$cachedYArray, this.c2me$cachedZArray);
+            this.c2me$oldStartBlockX = this.cellStartBlockX;
+            this.c2me$oldStartBlockY = this.cellStartBlockY;
+            this.c2me$oldStartBlockZ = this.cellStartBlockZ;
+        }
+    }
+
+    @Override
+    public int[] c2me$getXArray() {
+        this.c2me$reloadCachedArrays();
+        return this.c2me$cachedXArray;
+    }
+
+    @Override
+    public int[] c2me$getYArray() {
+        this.c2me$reloadCachedArrays();
+        return this.c2me$cachedYArray;
+    }
+
+    @Override
+    public int[] c2me$getZArray() {
+        this.c2me$reloadCachedArrays();
+        return this.c2me$cachedZArray;
+    }
+
+    private DensityFunction.Visitor c2me$wrapVisitor(final DensityFunction.Visitor visitor) {
+        if (!io.canvasmc.canvas.GlobalConfiguration.getInstance().chunkSystem.densityFunctionCompiler) {
+            return visitor;
+        }
+        return new com.ishland.c2me.opts.dfc.common.gen.DelegatingBlendingCachingAwareVisitor(
+            visitor, this.c2me$compiledEntryCache, !this.blender.isEmpty()
+        );
+    }
+
+    @Override
+    public int getStartBlockX() {
+        return this.cellStartBlockX;
+    }
+
+    @Override
+    public int getStartBlockY() {
+        return this.cellStartBlockY;
+    }
+
+    @Override
+    public int getStartBlockZ() {
+        return this.cellStartBlockZ;
+    }
+
+    @Override
+    public int getHorizontalCellBlockCount() {
+        return this.cellWidth;
+    }
+
+    @Override
+    public int getVerticalCellBlockCount() {
+        return this.cellHeight;
+    }
+
+    @Override
+    public boolean getIsInInterpolationLoop() {
+        return this.interpolating;
+    }
+
+    @Override
+    public boolean getIsSamplingForCaches() {
+        return this.fillingCell;
+    }
+
+    @Override
+    public int getStartBiomeX() {
+        return this.firstNoiseX;
+    }
+
+    @Override
+    public int getStartBiomeZ() {
+        return this.firstNoiseZ;
+    }
+
+    @Override
+    public int getHorizontalCellCount() {
+        return this.cellCountXZ;
+    }
+
+    @Override
+    public int getVerticalCellCount() {
+        return this.cellCountY;
+    }
+
+    @Override
+    public int getMinimumCellY() {
+        return this.cellNoiseMinY;
+    }
+
+    @Override
+    public int getCellBlockX() {
+        return this.inCellX;
+    }
+
+    @Override
+    public int getCellBlockY() {
+        return this.inCellY;
+    }
+
+    @Override
+    public int getCellBlockZ() {
+        return this.inCellZ;
+    }
+
+    @Override
+    public BlockState invokeSampleBlockState() {
+        return this.getInterpolatedState();
+    }
+
+    @Override
+    public int getHorizontalBiomeEnd() {
+        return this.noiseSizeXZ;
+    }
+
+    @Override
+    public DensityFunctions.BeardifierOrMarker getBeardifying() {
+        return this.beardifier;
+    }
+
+    @Override
+    public int getStartCellX() {
+        return this.firstCellX;
+    }
+
+    @Override
+    public int getStartCellZ() {
+        return this.firstCellZ;
+    }
+
+    @Override
+    public Blender getBlender() {
+        return this.blender;
+    }
+    // Leaf end - C2ME - DensityFunction compiler
+
     public int maxPreliminarySurfaceLevel(final int minBlockX, final int minBlockZ, final int maxBlockX, final int maxBlockZ) {
         int maxY = Integer.MIN_VALUE;
 
@@ -311,10 +_,20 @@
         this.fillingCell = false;
     }
 
+    // Leaf start - C2ME - DensityFunction compiler
+    private NoiseChunk.NoiseInterpolator[] c2me$initInterpolatorsArray() {
+        NoiseChunk.NoiseInterpolator[] interpolatorsArray = this.c2me$interpolatorsArray;
+        if (interpolatorsArray == null) {
+            interpolatorsArray = this.c2me$interpolatorsArray = this.interpolators.toArray(NoiseChunk.NoiseInterpolator[]::new);
+        }
+        return interpolatorsArray;
+    }
+    // Leaf end - C2ME - DensityFunction compiler
+
     public void updateForY(final int posY, final double factorY) {
         this.inCellY = posY - this.cellStartBlockY;
 
-        for (NoiseChunk.NoiseInterpolator i : this.interpolators) {
+        for (NoiseChunk.NoiseInterpolator i : this.c2me$initInterpolatorsArray()) { // Leaf - C2ME - DensityFunction compiler - array iteration
             i.updateForY(factorY);
         }
     }
@@ -322,7 +_,7 @@
     public void updateForX(final int posX, final double factorX) {
         this.inCellX = posX - this.cellStartBlockX;
 
-        for (NoiseChunk.NoiseInterpolator i : this.interpolators) {
+        for (NoiseChunk.NoiseInterpolator i : this.c2me$initInterpolatorsArray()) { // Leaf - C2ME - DensityFunction compiler - array iteration
             i.updateForX(factorX);
         }
     }
@@ -331,7 +_,7 @@
         this.inCellZ = posZ - this.cellStartBlockZ;
         this.interpolationCounter++;
 
-        for (NoiseChunk.NoiseInterpolator i : this.interpolators) {
+        for (NoiseChunk.NoiseInterpolator i : this.c2me$initInterpolatorsArray()) { // Leaf - C2ME - DensityFunction compiler - array iteration
             i.updateForZ(factorZ);
         }
     }
@@ -345,7 +_,11 @@
     }
 
     public void swapSlices() {
-        this.interpolators.forEach(NoiseChunk.NoiseInterpolator::swapSlices);
+        // Leaf start - C2ME - DensityFunction compiler
+        for (NoiseChunk.NoiseInterpolator noiseInterpolator : this.c2me$initInterpolatorsArray()) {
+            noiseInterpolator.swapSlices();
+        }
+        // Leaf end - C2ME - DensityFunction compiler
     }
 
     public Aquifer aquifer() {
@@ -373,13 +_,31 @@
     }
 
     protected DensityFunction wrap(final DensityFunction function) {
-        return this.wrapped.computeIfAbsent(function, this::wrapNew);
+        // Canvas start - reduce worldgen allocations
+        // avoid lambda allocation
+        DensityFunction func = this.wrapped.get(function);
+
+        if (func == null) {
+            func = this.wrapNew(function);
+            this.wrapped.put(function, func);
+        }
+
+        return func;
+        // Canvas end - reduce worldgen allocations
     }
 
     private DensityFunction wrapNew(final DensityFunction function) {
+        // Leaf start - C2ME - DensityFunction compiler
+        if (io.canvasmc.canvas.GlobalConfiguration.getInstance().chunkSystem.densityFunctionCompiler
+            && this.interpolating
+            && function instanceof DensityFunctions.Marker) {
+            throw new IllegalStateException("Cannot create more wrapping during interpolation loop");
+        }
+        // Leaf end - C2ME - DensityFunction compiler
         return switch (function) {
-            case DensityFunctions.Marker(DensityFunctions.Marker.Type type, DensityFunction wrapped) -> {
-                switch (type) {
+            case DensityFunctions.Marker marker -> {
+                final DensityFunction wrapped = marker.wrapped();
+                switch (marker.type()) {
                     case Interpolated:
                         yield new NoiseChunk.NoiseInterpolator(wrapped);
                     case FlatCache:
@@ -407,7 +_,7 @@
         };
     }
 
-    private class BlendAlpha implements NoiseChunk.NoiseChunkDensityFunction {
+    public class BlendAlpha implements NoiseChunk.NoiseChunkDensityFunction { // Leaf - C2ME - DensityFunction compiler - private -> public
         @Override
         public DensityFunction wrapped() {
             return DensityFunctions.BlendAlpha.INSTANCE;
@@ -487,7 +_,7 @@
         }
     }
 
-    private class BlendOffset implements NoiseChunk.NoiseChunkDensityFunction {
+    public class BlendOffset implements NoiseChunk.NoiseChunkDensityFunction { // Leaf - C2ME - DensityFunction compiler - private -> public
         @Override
         public DensityFunction wrapped() {
             return DensityFunctions.BlendOffset.INSTANCE;
@@ -529,8 +_,8 @@
         @Nullable BlockState calculate(final DensityFunction.FunctionContext context);
     }
 
-    private static class Cache2D implements NoiseChunk.NoiseChunkDensityFunction, DensityFunctions.MarkerOrMarked {
-        private final DensityFunction function;
+    public static class Cache2D implements NoiseChunk.NoiseChunkDensityFunction, DensityFunctions.MarkerOrMarked, com.ishland.c2me.opts.dfc.common.ducks.IFastCacheLike { // Leaf - C2ME - DensityFunction compiler - private -> public
+        private DensityFunction function; // Leaf - C2ME - DensityFunction compiler - private final -> private
         private long lastPos2D = ChunkPos.INVALID_CHUNK_POS;
         private double lastValue;
 
@@ -567,10 +_,78 @@
         public DensityFunctions.Marker.Type type() {
             return DensityFunctions.Marker.Type.Cache2D;
         }
+
+        // Leaf start - C2ME - DensityFunction compiler
+        @Override
+        public double c2me$getCached(
+            final int x,
+            final int y,
+            final int z,
+            final com.ishland.c2me.opts.dfc.common.ast.EvalType evalType
+        ) {
+            return this.lastPos2D == ChunkPos.pack(x, z)
+                ? this.lastValue
+                : Double.longBitsToDouble(com.ishland.c2me.opts.dfc.common.ducks.IFastCacheLike.CACHE_MISS_NAN_BITS);
+        }
+
+        @Override
+        public boolean c2me$getCached(
+            final double[] result,
+            final int[] x,
+            final int[] y,
+            final int[] z,
+            final com.ishland.c2me.opts.dfc.common.ast.EvalType evalType
+        ) {
+            return false;
+        }
+
+        @Override
+        public void c2me$cache(
+            final int x,
+            final int y,
+            final int z,
+            final com.ishland.c2me.opts.dfc.common.ast.EvalType evalType,
+            final double cached
+        ) {
+            this.lastPos2D = ChunkPos.pack(x, z);
+            this.lastValue = cached;
+        }
+
+        @Override
+        public void c2me$cache(
+            final double[] result,
+            final int[] x,
+            final int[] y,
+            final int[] z,
+            final com.ishland.c2me.opts.dfc.common.ast.EvalType evalType
+        ) {
+        }
+
+        @Override
+        public boolean c2me$isActualCache() {
+            return true;
+        }
+
+        @Override
+        public String c2me$describeCacheLike() {
+            return "cache_2d";
+        }
+
+        @Override
+        public DensityFunction c2me$getDelegate() {
+            return this.function;
+        }
+
+        @Override
+        public DensityFunction c2me$withDelegate(final DensityFunction delegate) {
+            this.function = delegate;
+            return this;
+        }
+        // Leaf end - C2ME - DensityFunction compiler
     }
 
-    private class CacheAllInCell implements NoiseChunk.NoiseChunkDensityFunction, DensityFunctions.MarkerOrMarked {
-        private final DensityFunction noiseFiller;
+    private class CacheAllInCell implements NoiseChunk.NoiseChunkDensityFunction, DensityFunctions.MarkerOrMarked, com.ishland.c2me.opts.dfc.common.ducks.IFastCacheLike { // Leaf - C2ME - DensityFunction compiler
+        private DensityFunction noiseFiller; // Leaf - C2ME - DensityFunction compiler - private final -> private
         private final double[] values;
 
         private CacheAllInCell(final DensityFunction noiseFiller) {
@@ -581,6 +_,19 @@
 
         @Override
         public double compute(final DensityFunction.FunctionContext context) {
+            // Leaf start - C2ME - DensityFunction compiler
+            if (!(context instanceof NoiseChunk)
+                && context instanceof com.ishland.c2me.opts.dfc.common.gen.jvm.vif.NoisePosVanillaInterface vanillaInterface
+                && vanillaInterface.getType() == com.ishland.c2me.opts.dfc.common.ast.EvalType.INTERPOLATION
+                && NoiseChunk.this.interpolating) {
+                final int x = context.blockX() - NoiseChunk.this.cellStartBlockX;
+                final int y = context.blockY() - NoiseChunk.this.cellStartBlockY;
+                final int z = context.blockZ() - NoiseChunk.this.cellStartBlockZ;
+                return x >= 0 && y >= 0 && z >= 0 && x < NoiseChunk.this.cellWidth && y < NoiseChunk.this.cellHeight && z < NoiseChunk.this.cellWidth
+                    ? this.values[((NoiseChunk.this.cellHeight - 1 - y) * NoiseChunk.this.cellWidth + x) * NoiseChunk.this.cellWidth + z]
+                    : this.noiseFiller.compute(context);
+            }
+            // Leaf end - C2ME - DensityFunction compiler
             if (context != NoiseChunk.this) {
                 return this.noiseFiller.compute(context);
             }
@@ -611,14 +_,119 @@
         public DensityFunctions.Marker.Type type() {
             return DensityFunctions.Marker.Type.CacheAllInCell;
         }
+
+        // Leaf start - C2ME - DensityFunction compiler
+        @Override
+        public double c2me$getCached(
+            final int x,
+            final int y,
+            final int z,
+            final com.ishland.c2me.opts.dfc.common.ast.EvalType evalType
+        ) {
+            if (evalType == com.ishland.c2me.opts.dfc.common.ast.EvalType.INTERPOLATION && NoiseChunk.this.interpolating) {
+                final int cellX = x - NoiseChunk.this.cellStartBlockX;
+                final int cellY = y - NoiseChunk.this.cellStartBlockY;
+                final int cellZ = z - NoiseChunk.this.cellStartBlockZ;
+                if (cellX >= 0
+                    && cellY >= 0
+                    && cellZ >= 0
+                    && cellX < NoiseChunk.this.cellWidth
+                    && cellY < NoiseChunk.this.cellHeight
+                    && cellZ < NoiseChunk.this.cellWidth) {
+                    return this.values[((NoiseChunk.this.cellHeight - 1 - cellY) * NoiseChunk.this.cellWidth + cellX) * NoiseChunk.this.cellWidth + cellZ];
+                }
+            }
+            return Double.longBitsToDouble(com.ishland.c2me.opts.dfc.common.ducks.IFastCacheLike.CACHE_MISS_NAN_BITS);
+        }
+
+        @Override
+        public boolean c2me$getCached(
+            final double[] result,
+            final int[] x,
+            final int[] y,
+            final int[] z,
+            final com.ishland.c2me.opts.dfc.common.ast.EvalType evalType
+        ) {
+            if (evalType == com.ishland.c2me.opts.dfc.common.ast.EvalType.INTERPOLATION && NoiseChunk.this.interpolating) {
+                for (int i = 0; i < result.length; i++) {
+                    final int cellX = x[i] - NoiseChunk.this.cellStartBlockX;
+                    final int cellY = y[i] - NoiseChunk.this.cellStartBlockY;
+                    final int cellZ = z[i] - NoiseChunk.this.cellStartBlockZ;
+                    if (cellX >= 0
+                        && cellY >= 0
+                        && cellZ >= 0
+                        && cellX < NoiseChunk.this.cellWidth
+                        && cellY < NoiseChunk.this.cellHeight
+                        && cellZ < NoiseChunk.this.cellWidth) {
+                        result[i] = this.values[((NoiseChunk.this.cellHeight - 1 - cellY) * NoiseChunk.this.cellWidth + cellX) * NoiseChunk.this.cellWidth + cellZ];
+                    } else {
+                        System.out.println("partial cell cache hit");
+                        return false;
+                    }
+                }
+            }
+            return false;
+        }
+
+        @Override
+        public void c2me$cache(
+            final int x,
+            final int y,
+            final int z,
+            final com.ishland.c2me.opts.dfc.common.ast.EvalType evalType,
+            final double cached
+        ) {
+        }
+
+        @Override
+        public void c2me$cache(
+            final double[] result,
+            final int[] x,
+            final int[] y,
+            final int[] z,
+            final com.ishland.c2me.opts.dfc.common.ast.EvalType evalType
+        ) {
+        }
+
+        @Override
+        public boolean c2me$isActualCache() {
+            return true;
+        }
+
+        @Override
+        public String c2me$describeCacheLike() {
+            return "cache_all_in_cell";
+        }
+
+        @Override
+        public DensityFunction c2me$getDelegate() {
+            return this.noiseFiller;
+        }
+
+        @Override
+        public DensityFunction c2me$withDelegate(final DensityFunction delegate) {
+            this.noiseFiller = delegate;
+            return this;
+        }
+        // Leaf end - C2ME - DensityFunction compiler
     }
 
-    private class CacheOnce implements NoiseChunk.NoiseChunkDensityFunction, DensityFunctions.MarkerOrMarked {
-        private final DensityFunction function;
+    public class CacheOnce implements NoiseChunk.NoiseChunkDensityFunction, DensityFunctions.MarkerOrMarked, com.ishland.c2me.opts.dfc.common.ducks.IFastCacheLike { // Leaf - C2ME - DensityFunction compiler - private -> public
+        private DensityFunction function; // Leaf - C2ME - DensityFunction compiler - private final -> private
         private long lastCounter;
         private long lastArrayCounter;
         private double lastValue;
         private double @Nullable [] lastArray;
+        // Leaf start - C2ME - DensityFunction compiler
+        private double c2me$lastValue = Double.NaN;
+        private int c2me$lastX = Integer.MIN_VALUE;
+        private int c2me$lastY = Integer.MIN_VALUE;
+        private int c2me$lastZ = Integer.MIN_VALUE;
+        private int[] c2me$lastXa;
+        private int[] c2me$lastYa;
+        private int[] c2me$lastZa;
+        private double[] c2me$lastValuea;
+        // Leaf end - C2ME - DensityFunction compiler
 
         private CacheOnce(final DensityFunction function) {
             this.function = function;
@@ -626,6 +_,32 @@
 
         @Override
         public double compute(final DensityFunction.FunctionContext context) {
+            // Leaf start - C2ME - DensityFunction compiler
+            if (!(context instanceof NoiseChunk)) {
+                final int blockX = context.blockX();
+                final int blockY = context.blockY();
+                final int blockZ = context.blockZ();
+                if (this.c2me$lastValuea != null) {
+                    for (int i = 0; i < this.c2me$lastValuea.length; i++) {
+                        if (this.c2me$lastXa[i] == blockX && this.c2me$lastYa[i] == blockY && this.c2me$lastZa[i] == blockZ) {
+                            return this.c2me$lastValuea[i];
+                        }
+                    }
+                }
+                if (!Double.isNaN(this.c2me$lastValue)
+                    && this.c2me$lastX == blockX
+                    && this.c2me$lastY == blockY
+                    && this.c2me$lastZ == blockZ) {
+                    return this.c2me$lastValue;
+                }
+                final double value = this.function.compute(context);
+                this.c2me$lastValue = value;
+                this.c2me$lastX = blockX;
+                this.c2me$lastY = blockY;
+                this.c2me$lastZ = blockZ;
+                return value;
+            }
+            // Leaf end - C2ME - DensityFunction compiler
             if (context != NoiseChunk.this) {
                 return this.function.compute(context);
             }
@@ -646,6 +_,27 @@
 
         @Override
         public void fillArray(final double[] output, final DensityFunction.ContextProvider contextProvider) {
+            // Leaf start - C2ME - DensityFunction compiler
+            if (!(contextProvider instanceof NoiseChunk)) {
+                if (contextProvider instanceof com.ishland.c2me.opts.dfc.common.gen.jvm.vif.EachApplierVanillaInterface vanillaInterface) {
+                    if (this.c2me$lastValuea != null
+                        && java.util.Arrays.equals(vanillaInterface.getY(), this.c2me$lastYa)
+                        && java.util.Arrays.equals(vanillaInterface.getX(), this.c2me$lastXa)
+                        && java.util.Arrays.equals(vanillaInterface.getZ(), this.c2me$lastZa)) {
+                        System.arraycopy(this.c2me$lastValuea, 0, output, 0, this.c2me$lastValuea.length);
+                    } else {
+                        this.function.fillArray(output, contextProvider);
+                        this.c2me$lastValuea = java.util.Arrays.copyOf(output, output.length);
+                        this.c2me$lastXa = vanillaInterface.getX();
+                        this.c2me$lastYa = vanillaInterface.getY();
+                        this.c2me$lastZa = vanillaInterface.getZ();
+                    }
+                } else {
+                    this.function.fillArray(output, contextProvider);
+                }
+                return;
+            }
+            // Leaf end - C2ME - DensityFunction compiler
             if (this.lastArray != null && this.lastArrayCounter == NoiseChunk.this.arrayInterpolationCounter) {
                 System.arraycopy(this.lastArray, 0, output, 0, output.length);
             } else {
@@ -669,10 +_,109 @@
         public DensityFunctions.Marker.Type type() {
             return DensityFunctions.Marker.Type.CacheOnce;
         }
+
+        // Leaf start - C2ME - DensityFunction compiler
+        @Override
+        public double c2me$getCached(
+            final int x,
+            final int y,
+            final int z,
+            final com.ishland.c2me.opts.dfc.common.ast.EvalType evalType
+        ) {
+            if (this.c2me$lastValuea != null) {
+                for (int i = 0; i < this.c2me$lastValuea.length; i++) {
+                    if (this.c2me$lastXa[i] == x && this.c2me$lastYa[i] == y && this.c2me$lastZa[i] == z) {
+                        return this.c2me$lastValuea[i];
+                    }
+                }
+            }
+            if (!Double.isNaN(this.c2me$lastValue)
+                && this.c2me$lastX == x
+                && this.c2me$lastY == y
+                && this.c2me$lastZ == z) {
+                return this.c2me$lastValue;
+            }
+            return Double.longBitsToDouble(com.ishland.c2me.opts.dfc.common.ducks.IFastCacheLike.CACHE_MISS_NAN_BITS);
+        }
+
+        @Override
+        public boolean c2me$getCached(
+            final double[] result,
+            final int[] x,
+            final int[] y,
+            final int[] z,
+            final com.ishland.c2me.opts.dfc.common.ast.EvalType evalType
+        ) {
+            if (this.c2me$lastValuea != null
+                && java.util.Arrays.equals(y, this.c2me$lastYa)
+                && java.util.Arrays.equals(x, this.c2me$lastXa)
+                && java.util.Arrays.equals(z, this.c2me$lastZa)) {
+                System.arraycopy(this.c2me$lastValuea, 0, result, 0, this.c2me$lastValuea.length);
+                return true;
+            }
+            return false;
+        }
+
+        @Override
+        public void c2me$cache(
+            final int x,
+            final int y,
+            final int z,
+            final com.ishland.c2me.opts.dfc.common.ast.EvalType evalType,
+            final double cached
+        ) {
+            this.c2me$lastValue = cached;
+            this.c2me$lastX = x;
+            this.c2me$lastY = y;
+            this.c2me$lastZ = z;
+        }
+
+        @Override
+        public void c2me$cache(
+            final double[] result,
+            final int[] x,
+            final int[] y,
+            final int[] z,
+            final com.ishland.c2me.opts.dfc.common.ast.EvalType evalType
+        ) {
+            if (this.c2me$lastValuea != null && this.c2me$lastValuea.length == result.length) {
+                System.arraycopy(result, 0, this.c2me$lastValuea, 0, result.length);
+                System.arraycopy(x, 0, this.c2me$lastXa, 0, result.length);
+                System.arraycopy(y, 0, this.c2me$lastYa, 0, result.length);
+                System.arraycopy(z, 0, this.c2me$lastZa, 0, result.length);
+            } else {
+                this.c2me$lastValuea = java.util.Arrays.copyOf(result, result.length);
+                this.c2me$lastXa = java.util.Arrays.copyOf(x, x.length);
+                this.c2me$lastYa = java.util.Arrays.copyOf(y, y.length);
+                this.c2me$lastZa = java.util.Arrays.copyOf(z, z.length);
+            }
+        }
+
+        @Override
+        public boolean c2me$isActualCache() {
+            return true;
+        }
+
+        @Override
+        public String c2me$describeCacheLike() {
+            return "cache_once";
+        }
+
+        @Override
+        public DensityFunction c2me$getDelegate() {
+            return this.function;
+        }
+
+        @Override
+        public DensityFunction c2me$withDelegate(final DensityFunction delegate) {
+            this.function = delegate;
+            return this;
+        }
+        // Leaf end - C2ME - DensityFunction compiler
     }
 
-    private class FlatCache implements NoiseChunk.NoiseChunkDensityFunction, DensityFunctions.MarkerOrMarked {
-        private final DensityFunction noiseFiller;
+    public class FlatCache implements NoiseChunk.NoiseChunkDensityFunction, DensityFunctions.MarkerOrMarked, com.ishland.c2me.opts.dfc.common.ducks.IFastCacheLike { // Leaf - C2ME - DensityFunction compiler - private -> public
+        private DensityFunction noiseFiller; // Leaf - C2ME - DensityFunction compiler - mutable delegate
         private final double[] values;
         private final int sizeXZ;
 
@@ -717,6 +_,84 @@
         public DensityFunctions.Marker.Type type() {
             return DensityFunctions.Marker.Type.FlatCache;
         }
+
+        // Leaf start - C2ME - DensityFunction compiler
+        @Override
+        public double c2me$getCached(
+            final int x,
+            final int y,
+            final int z,
+            final com.ishland.c2me.opts.dfc.common.ast.EvalType evalType
+        ) {
+            final int localX = QuartPos.fromBlock(x) - NoiseChunk.this.firstNoiseX;
+            final int localZ = QuartPos.fromBlock(z) - NoiseChunk.this.firstNoiseZ;
+            return localX >= 0 && localZ >= 0 && localX < this.sizeXZ && localZ < this.sizeXZ
+                ? this.values[localX + localZ * this.sizeXZ]
+                : Double.longBitsToDouble(com.ishland.c2me.opts.dfc.common.ducks.IFastCacheLike.CACHE_MISS_NAN_BITS);
+        }
+
+        @Override
+        public boolean c2me$getCached(
+            final double[] result,
+            final int[] x,
+            final int[] y,
+            final int[] z,
+            final com.ishland.c2me.opts.dfc.common.ast.EvalType evalType
+        ) {
+            for (int i = 0; i < result.length; i++) {
+                final int localX = QuartPos.fromBlock(x[i]) - NoiseChunk.this.firstNoiseX;
+                final int localZ = QuartPos.fromBlock(z[i]) - NoiseChunk.this.firstNoiseZ;
+                if (localX >= 0 && localZ >= 0 && localX < this.sizeXZ && localZ < this.sizeXZ) {
+                    result[i] = this.values[localX + localZ * this.sizeXZ];
+                } else {
+                    System.out.println("partial flat cache hit");
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        @Override
+        public void c2me$cache(
+            final int x,
+            final int y,
+            final int z,
+            final com.ishland.c2me.opts.dfc.common.ast.EvalType evalType,
+            final double cached
+        ) {
+        }
+
+        @Override
+        public void c2me$cache(
+            final double[] result,
+            final int[] x,
+            final int[] y,
+            final int[] z,
+            final com.ishland.c2me.opts.dfc.common.ast.EvalType evalType
+        ) {
+        }
+
+        @Override
+        public boolean c2me$isActualCache() {
+            return true;
+        }
+
+        @Override
+        public String c2me$describeCacheLike() {
+            return "flat_cache";
+        }
+
+        @Override
+        public DensityFunction c2me$getDelegate() {
+            return this.noiseFiller;
+        }
+
+        @Override
+        public DensityFunction c2me$withDelegate(final DensityFunction delegate) {
+            this.noiseFiller = delegate;
+            return this;
+        }
+        // Leaf end - C2ME - DensityFunction compiler
     }
 
     private interface NoiseChunkDensityFunction extends DensityFunction {
@@ -733,10 +_,10 @@
         }
     }
 
-    public class NoiseInterpolator implements NoiseChunk.NoiseChunkDensityFunction, DensityFunctions.MarkerOrMarked {
+    public class NoiseInterpolator implements NoiseChunk.NoiseChunkDensityFunction, DensityFunctions.MarkerOrMarked, com.ishland.c2me.opts.dfc.common.ducks.IFastCacheLike, com.ishland.c2me.base.mixin.access.IChunkNoiseSamplerDensityInterpolator { // Leaf - C2ME - DensityFunction compiler
         private double[][] slice0;
         private double[][] slice1;
-        private final DensityFunction noiseFiller;
+        private DensityFunction noiseFiller; // Leaf - C2ME - DensityFunction compiler - private final -> private
         private double noise000;
         private double noise001;
         private double noise100;
@@ -801,6 +_,33 @@
 
         @Override
         public double compute(final DensityFunction.FunctionContext context) {
+            // Leaf start - C2ME - DensityFunction compiler
+            if (!(context instanceof NoiseChunk)
+                && context instanceof com.ishland.c2me.opts.dfc.common.gen.jvm.vif.NoisePosVanillaInterface vanillaInterface
+                && vanillaInterface.getType() == com.ishland.c2me.opts.dfc.common.ast.EvalType.INTERPOLATION) {
+                if (!NoiseChunk.this.interpolating) {
+                    return this.noiseFiller.compute(context);
+                }
+                final int cellX = context.blockX() - NoiseChunk.this.cellStartBlockX;
+                final int cellY = context.blockY() - NoiseChunk.this.cellStartBlockY;
+                final int cellZ = context.blockZ() - NoiseChunk.this.cellStartBlockZ;
+                return NoiseChunk.this.fillingCell
+                    ? Mth.lerp3(
+                        (double)cellX / NoiseChunk.this.cellWidth,
+                        (double)cellY / NoiseChunk.this.cellHeight,
+                        (double)cellZ / NoiseChunk.this.cellWidth,
+                        this.noise000,
+                        this.noise100,
+                        this.noise010,
+                        this.noise110,
+                        this.noise001,
+                        this.noise101,
+                        this.noise011,
+                        this.noise111
+                    )
+                    : this.value;
+            }
+            // Leaf end - C2ME - DensityFunction compiler
             if (context != NoiseChunk.this) {
                 return this.noiseFiller.compute(context);
             } else if (!NoiseChunk.this.interpolating) {
@@ -848,5 +_,136 @@
         public DensityFunctions.Marker.Type type() {
             return DensityFunctions.Marker.Type.Interpolated;
         }
+
+        // Leaf start - C2ME - DensityFunction compiler
+        @Override
+        public double c2me$getCached(
+            final int x,
+            final int y,
+            final int z,
+            final com.ishland.c2me.opts.dfc.common.ast.EvalType evalType
+        ) {
+            if (evalType == com.ishland.c2me.opts.dfc.common.ast.EvalType.INTERPOLATION) {
+                if (!NoiseChunk.this.interpolating) {
+                    throw new IllegalStateException("Trying to sample interpolator outside the interpolation loop");
+                }
+                if (NoiseChunk.this.fillingCell) {
+                    final int cellX = x - NoiseChunk.this.cellStartBlockX;
+                    final int cellY = y - NoiseChunk.this.cellStartBlockY;
+                    final int cellZ = z - NoiseChunk.this.cellStartBlockZ;
+                    return Mth.lerp3(
+                        (double)cellX / NoiseChunk.this.cellWidth,
+                        (double)cellY / NoiseChunk.this.cellHeight,
+                        (double)cellZ / NoiseChunk.this.cellWidth,
+                        this.noise000,
+                        this.noise100,
+                        this.noise010,
+                        this.noise110,
+                        this.noise001,
+                        this.noise101,
+                        this.noise011,
+                        this.noise111
+                    );
+                }
+                return this.value;
+            }
+            return Double.longBitsToDouble(com.ishland.c2me.opts.dfc.common.ducks.IFastCacheLike.CACHE_MISS_NAN_BITS);
+        }
+
+        @Override
+        public boolean c2me$getCached(
+            final double[] result,
+            final int[] x,
+            final int[] y,
+            final int[] z,
+            final com.ishland.c2me.opts.dfc.common.ast.EvalType evalType
+        ) {
+            if (evalType == com.ishland.c2me.opts.dfc.common.ast.EvalType.INTERPOLATION
+                && NoiseChunk.this.interpolating
+                && NoiseChunk.this.fillingCell) {
+                for (int i = 0; i < result.length; i++) {
+                    final int cellX = x[i] - NoiseChunk.this.cellStartBlockX;
+                    final int cellY = y[i] - NoiseChunk.this.cellStartBlockY;
+                    final int cellZ = z[i] - NoiseChunk.this.cellStartBlockZ;
+                    result[i] = Mth.lerp3(
+                        (double)cellX / NoiseChunk.this.cellWidth,
+                        (double)cellY / NoiseChunk.this.cellHeight,
+                        (double)cellZ / NoiseChunk.this.cellWidth,
+                        this.noise000,
+                        this.noise100,
+                        this.noise010,
+                        this.noise110,
+                        this.noise001,
+                        this.noise101,
+                        this.noise011,
+                        this.noise111
+                    );
+                }
+                return true;
+            }
+            return false;
+        }
+
+        @Override
+        public void c2me$cache(
+            final int x,
+            final int y,
+            final int z,
+            final com.ishland.c2me.opts.dfc.common.ast.EvalType evalType,
+            final double cached
+        ) {
+        }
+
+        @Override
+        public void c2me$cache(
+            final double[] result,
+            final int[] x,
+            final int[] y,
+            final int[] z,
+            final com.ishland.c2me.opts.dfc.common.ast.EvalType evalType
+        ) {
+        }
+
+        @Override
+        public boolean c2me$isActualCache() {
+            return false;
+        }
+
+        @Override
+        public String c2me$describeCacheLike() {
+            return "interpolated";
+        }
+
+        @Override
+        public DensityFunction c2me$getDelegate() {
+            return this.noiseFiller;
+        }
+
+        @Override
+        public DensityFunction c2me$withDelegate(final DensityFunction delegate) {
+            this.noiseFiller = delegate;
+            return this;
+        }
+
+        @Override
+        public void invokeInterpolateX(final double deltaX) {
+            this.updateForX(deltaX);
+        }
+
+        @Override
+        public void invokeInterpolateY(final double deltaY) {
+            this.updateForY(deltaY);
+        }
+
+        @Override
+        public void invokeInterpolateZ(final double deltaZ) {
+            this.updateForZ(deltaZ);
+        }
+
+        @Override
+        public void invokeSwapBuffers() {
+            this.swapSlices();
+        }
+        // Leaf end - C2ME - DensityFunction compiler
     }
 }
