--- a/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
+++ b/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
@@ -307,6 +_,15 @@
         int i = Mth.floorDiv(minY, noiseSettings.getCellHeight());
         int i1 = Mth.floorDiv(noiseSettings.height(), noiseSettings.getCellHeight());
         return i1 <= 0 ? CompletableFuture.completedFuture(chunk) : CompletableFuture.supplyAsync(() -> {
+            // Canvas start - optimize noise gen
+            if (io.canvasmc.canvas.Config.INSTANCE.chunks.optimizeNoiseGeneration) {
+                // remove the rest of this bs, we don't need it
+                @Nullable ChunkAccess chunkWithNoise = this.doFill(
+                    blender, structureManager, randomState, chunk, i, i1);
+
+                return chunkWithNoise;
+            }
+            // Canvas end - optimize noise gen
             int sectionIndex = chunk.getSectionIndex(i1 * noiseSettings.getCellHeight() - 1 + minY);
             int sectionIndex1 = chunk.getSectionIndex(minY);
             Set<LevelChunkSection> set = Sets.newHashSet();
@@ -385,7 +_,11 @@
 
                                 interpolatedState = this.debugPreliminarySurfaceLevel(noiseChunk, i10, i7, i13, interpolatedState);
                                 if (interpolatedState != AIR && !SharedConstants.debugVoidTerrain(chunk.getPos())) {
-                                    section.setBlockState(i11, i8, i14, interpolatedState, false);
+                                    // Canvas start - optimize noise gen
+                                    if (io.canvasmc.canvas.Config.INSTANCE.chunks.optimizeNoiseGeneration && section.states.strategy != null) {
+                                        canvas$populateNoiseWrapSetBlockStateOperation(section, i11, i8, i14, interpolatedState, false);
+                                    } else section.setBlockState(i11, i8, i14, interpolatedState, false);
+                                    // Canvas end - optimize noise gen
                                     heightmapUnprimed.update(i11, i7, i14, interpolatedState);
                                     heightmapUnprimed1.update(i11, i7, i14, interpolatedState);
                                     if (aquifer.shouldScheduleFluidUpdate() && !interpolatedState.getFluidState().isEmpty()) {
@@ -405,6 +_,38 @@
         noiseChunk.stopInterpolation();
         return chunk;
     }
+    // Canvas start - optimize noise gen
+
+    private BlockState canvas$populateNoiseWrapSetBlockStateOperation(LevelChunkSection chunkSection, int chunkSectionBlockPosX, int chunkSectionBlockPosY, int chunkSectionBlockPosZ, BlockState blockState, boolean lock) {
+        chunkSection.nonEmptyBlockCount += 1;
+
+        if (!blockState.getFluidState().isEmpty()) {
+            chunkSection.tickingFluidCount += 1;
+        }
+
+        if (blockState.isRandomlyTicking()) {
+            chunkSection.tickingBlockCount += 1;
+        }
+
+        if (blockState.canvas$isWater) {
+            chunkSection.canvas$waterCount++;
+        }
+
+        if (blockState.canvas$isBubble) {
+            chunkSection.canvas$bubbleCount++;
+        }
+
+        if (blockState.canvas$isLava) {
+            chunkSection.canvas$lavaCount++;
+        }
+
+        chunkSection.states.getAndSetUnchecked(
+            chunkSectionBlockPosX, chunkSectionBlockPosY, chunkSectionBlockPosZ, blockState
+        );
+
+        return blockState;
+    }
+    // Canvas end - optimize noise gen
 
     private BlockState debugPreliminarySurfaceLevel(NoiseChunk chunk, int x, int y, int z, BlockState state) {
         if (SharedConstants.DEBUG_AQUIFERS && z >= 0 && z % 4 == 0) {
