--- a/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
+++ b/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
@@ -356,6 +_,15 @@
         int cellYMin = Mth.floorDiv(minY, noiseSettings.getCellHeight());
         int cellCountY = Mth.floorDiv(noiseSettings.height(), noiseSettings.getCellHeight());
         return cellCountY <= 0 ? CompletableFuture.completedFuture(centerChunk) : CompletableFuture.supplyAsync(() -> {
+            // Canvas start - optimize noise gen
+            if (io.canvasmc.canvas.Config.INSTANCE.chunks.optimizeNoiseGeneration) {
+                // remove the rest of this bs, we don't need it
+                @Nullable ChunkAccess chunkWithNoise = this.doFill(
+                    blender, structureManager, randomState, centerChunk, cellYMin, cellCountY);
+
+                return chunkWithNoise;
+            }
+            // Canvas end - optimize noise gen
             int topSectionIndex = centerChunk.getSectionIndex(cellCountY * noiseSettings.getCellHeight() - 1 + minY);
             int bottomSectionIndex = centerChunk.getSectionIndex(minY);
             Set<LevelChunkSection> sections = Sets.newHashSet();
@@ -441,7 +_,11 @@
 
                                 state = this.debugPreliminarySurfaceLevel(noiseChunk, posX, posY, posZ, state);
                                 if (state != AIR && !SharedConstants.debugVoidTerrain(centerChunk.getPos())) {
-                                    section.setBlockState(xInSection, yInSection, zInSection, state, false);
+                                    // Canvas start - optimize noise gen
+                                    if (io.canvasmc.canvas.Config.INSTANCE.chunks.optimizeNoiseGeneration && section.states.strategy != null) {
+                                        canvas$populateNoiseWrapSetBlockStateOperation(section, xInSection, yInSection, zInSection, state, false);
+                                    } else section.setBlockState(xInSection, yInSection, zInSection, state, false);
+                                    // Canvas end - optimize noise gen
                                     oceanFloor.update(xInSection, posY, zInSection, state);
                                     worldSurface.update(xInSection, posY, zInSection, state);
                                     if (aquifer.shouldScheduleFluidUpdate() && !state.getFluidState().isEmpty()) {
@@ -461,6 +_,38 @@
         noiseChunk.stopInterpolation();
         return centerChunk;
     }
+    // Canvas start - optimize noise gen
+
+    private BlockState canvas$populateNoiseWrapSetBlockStateOperation(LevelChunkSection chunkSection, int chunkSectionBlockPosX, int chunkSectionBlockPosY, int chunkSectionBlockPosZ, BlockState blockState, boolean lock) {
+        chunkSection.nonEmptyBlockCount += 1;
+
+        if (!blockState.getFluidState().isEmpty()) {
+            chunkSection.tickingFluidCount += 1;
+        }
+
+        if (blockState.isRandomlyTicking()) {
+            chunkSection.tickingBlockCount += 1;
+        }
+
+        if (blockState.canvas$isWater) {
+            chunkSection.canvas$waterCount++;
+        }
+
+        if (blockState.canvas$isBubble) {
+            chunkSection.canvas$bubbleCount++;
+        }
+
+        if (blockState.canvas$isLava) {
+            chunkSection.canvas$lavaCount++;
+        }
+
+        chunkSection.states.getAndSetUnchecked(
+            chunkSectionBlockPosX, chunkSectionBlockPosY, chunkSectionBlockPosZ, blockState
+        );
+
+        return blockState;
+    }
+    // Canvas end - optimize noise gen
 
     private BlockState debugPreliminarySurfaceLevel(final NoiseChunk noiseChunk, final int posX, final int posY, final int posZ, BlockState state) {
         if (SharedConstants.DEBUG_AQUIFERS && posZ >= 0 && posZ % 4 == 0) {
