--- a/net/minecraft/world/level/levelgen/NoiseChunk.java
+++ b/net/minecraft/world/level/levelgen/NoiseChunk.java
@@ -362,7 +_,17 @@
     }
 
     protected DensityFunction wrap(DensityFunction densityFunction) {
-        return this.wrapped.computeIfAbsent(densityFunction, this::wrapNew);
+        // Canvas start - reduce worldgen allocations
+        // avoid lambda allocation
+        DensityFunction func = this.wrapped.get(densityFunction);
+
+        if (func == null) {
+            func = this.wrapNew(densityFunction);
+            this.wrapped.put(densityFunction, func);
+        }
+
+        return func;
+        // Canvas end - reduce worldgen allocations
     }
 
     private DensityFunction wrapNew(DensityFunction densityFunction) {
