--- a/net/minecraft/world/level/levelgen/RandomState.java
+++ b/net/minecraft/world/level/levelgen/RandomState.java
@@ -16,8 +_,8 @@
 public final class RandomState {
     final PositionalRandomFactory random;
     private final HolderGetter<NormalNoise.NoiseParameters> noises;
-    private final NoiseRouter router;
-    private final Climate.Sampler sampler;
+    private NoiseRouter router; // Canvas - mutable
+    private Climate.Sampler sampler; // Canvas - mutable
     private final SurfaceSystem surfaceSystem;
     private final PositionalRandomFactory aquiferRandom;
     private final PositionalRandomFactory oreRandom;
