--- a/net/minecraft/world/level/levelgen/RandomState.java
+++ b/net/minecraft/world/level/levelgen/RandomState.java
@@ -81,7 +_,7 @@
             }
         }
 
-        this.router = settings.noiseRouter().mapAll(new NoiseWiringHelper());
+        NoiseRouter wiredRouter = settings.noiseRouter().mapAll(new NoiseWiringHelper()); // Leaf - C2ME - DensityFunction compiler
         DensityFunction.Visitor noiseFlattener = new DensityFunction.Visitor() {
             private final Map<DensityFunction, DensityFunction> wrapped = new HashMap<>();
 
@@ -98,15 +_,63 @@
                 return this.wrapped.computeIfAbsent(input, this::wrapNew);
             }
         };
-        this.sampler = new Climate.Sampler(
-            this.router.temperature().mapAll(noiseFlattener),
-            this.router.vegetation().mapAll(noiseFlattener),
-            this.router.continents().mapAll(noiseFlattener),
-            this.router.erosion().mapAll(noiseFlattener),
-            this.router.depth().mapAll(noiseFlattener),
-            this.router.ridges().mapAll(noiseFlattener),
+        // Leaf start - C2ME - DensityFunction compiler
+        Climate.Sampler wiredSampler = new Climate.Sampler(
+            wiredRouter.temperature().mapAll(noiseFlattener),
+            wiredRouter.vegetation().mapAll(noiseFlattener),
+            wiredRouter.continents().mapAll(noiseFlattener),
+            wiredRouter.erosion().mapAll(noiseFlattener),
+            wiredRouter.depth().mapAll(noiseFlattener),
+            wiredRouter.ridges().mapAll(noiseFlattener),
             settings.spawnTarget()
         );
+        // Leaf end - C2ME - DensityFunction compiler
+
+        // Leaf start - C2ME - DensityFunction compiler
+        if (io.canvasmc.canvas.GlobalConfiguration.getInstance().chunkSystem.densityFunctionCompiler) {
+            final com.google.common.base.Stopwatch stopwatch = com.google.common.base.Stopwatch.createStarted();
+            final DensityFunction finalFinalDensity = DensityFunctions.add(wiredRouter.finalDensity(), DensityFunctions.BeardifierMarker.INSTANCE);
+            final NoiseRouter originalRouter = wiredRouter;
+            originalRouter.c2me$setFinalFinalDensity(finalFinalDensity);
+
+            final com.ishland.c2me.opts.dfc.common.gen.jvm.BytecodeGen.Context generationContext =
+                com.ishland.c2me.opts.dfc.common.gen.jvm.BytecodeGen.initContext();
+            wiredRouter = new NoiseRouter(
+                generationContext.compileDelayed("barrier", wiredRouter.barrierNoise()),
+                generationContext.compileDelayed("fluid_level_floodedness", wiredRouter.fluidLevelFloodednessNoise()),
+                generationContext.compileDelayed("fluid_level_spread", wiredRouter.fluidLevelSpreadNoise()),
+                generationContext.compileDelayed("lava", wiredRouter.lavaNoise()),
+                generationContext.compileDelayed("temperature", wiredRouter.temperature()),
+                generationContext.compileDelayed("vegetation", wiredRouter.vegetation()),
+                generationContext.compileDelayed("continents", wiredRouter.continents()),
+                generationContext.compileDelayed("erosion", wiredRouter.erosion()),
+                generationContext.compileDelayed("depth", wiredRouter.depth()),
+                generationContext.compileDelayed("ridges", wiredRouter.ridges()),
+                generationContext.compileDelayed("preliminary_surface_level", wiredRouter.preliminarySurfaceLevel()),
+                generationContext.compileDelayed("final_density", wiredRouter.finalDensity()),
+                generationContext.compileDelayed("vein_toggle", wiredRouter.veinToggle()),
+                generationContext.compileDelayed("vein_ridged", wiredRouter.veinRidged()),
+                generationContext.compileDelayed("vein_gap", wiredRouter.veinGap())
+            );
+            wiredRouter.c2me$setFinalFinalDensity(generationContext.compileDelayed("final_final_density", finalFinalDensity));
+            wiredRouter.c2me$setOriginalNoiseRouter(originalRouter);
+
+            wiredSampler = new Climate.Sampler(
+                generationContext.compileDelayed("multiNoiseSampler_temperature", wiredSampler.temperature()),
+                generationContext.compileDelayed("multiNoiseSampler_vegetation", wiredSampler.humidity()),
+                generationContext.compileDelayed("multiNoiseSampler_continents", wiredSampler.continentalness()),
+                generationContext.compileDelayed("multiNoiseSampler_erosion", wiredSampler.erosion()),
+                generationContext.compileDelayed("multiNoiseSampler_depth", wiredSampler.depth()),
+                generationContext.compileDelayed("multiNoiseSampler_ridges", wiredSampler.weirdness()),
+                wiredSampler.spawnTarget()
+            );
+            com.ishland.c2me.opts.dfc.common.gen.jvm.BytecodeGen.finalizeCompilation(generationContext);
+            stopwatch.stop();
+            io.canvasmc.canvas.GlobalConfiguration.LOGGER.info("Density function compilation finished in {}", stopwatch);
+        }
+        this.router = wiredRouter;
+        this.sampler = wiredSampler;
+        // Leaf end - C2ME - DensityFunction compiler
     }
 
     public NormalNoise getOrCreateNoise(final ResourceKey<NormalNoise.NoiseParameters> noise) {
