--- a/net/minecraft/world/level/levelgen/SurfaceRules.java
+++ b/net/minecraft/world/level/levelgen/SurfaceRules.java
@@ -592,8 +_,13 @@
     private record SequenceRule(List<SurfaceRules.SurfaceRule> rules) implements SurfaceRules.SurfaceRule {
         @Override
         public @Nullable BlockState tryApply(final int blockX, final int blockY, final int blockZ) {
-            for (SurfaceRules.SurfaceRule rule : this.rules) {
-                BlockState state = rule.tryApply(blockX, blockY, blockZ);
+            // Canvas start - reduce worldgen allocations
+            // avoid iterator allocation
+            int size = this.rules.size();
+            //noinspection ForLoopReplaceableByForEach
+            for (int i = 0; i < size; i++) {
+                BlockState state = this.rules.get(i).tryApply(blockX, blockY, blockZ);
+            // Canvas end - reduce worldgen allocations
                 if (state != null) {
                     return state;
                 }
