--- a/net/minecraft/world/level/levelgen/SurfaceRules.java
+++ b/net/minecraft/world/level/levelgen/SurfaceRules.java
@@ -13,6 +_,7 @@
 import java.util.function.Predicate;
 import java.util.function.Supplier;
 import javax.annotation.Nullable;
+import io.canvasmc.canvas.Config;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
 import net.minecraft.core.Registry;
@@ -390,13 +_,29 @@
             }
         }
 
-        static class SteepMaterialCondition extends SurfaceRules.LazyXZCondition {
+        public static class SteepMaterialCondition extends SurfaceRules.LazyXZCondition { // Canvas - package-private -> public
             SteepMaterialCondition(SurfaceRules.Context context) {
                 super(context);
             }
 
             @Override
             protected boolean compute() {
+                // Canvas start - MC-258859
+                if (Config.INSTANCE.mc258859) {
+                    int x = this.context.blockX & 15;
+                    int z = this.context.blockZ & 15;
+                    ChunkAccess chunk = this.context.chunk;
+                    int south = chunk.getHeight(Heightmap.Types.WORLD_SURFACE_WG, x, Math.max(z - 1, 0));
+                    int north = chunk.getHeight(Heightmap.Types.WORLD_SURFACE_WG, x, Math.min(z + 1, 15));
+                    if (Math.abs(north - south) >= 4) {
+                        return true;
+                    } else {
+                        int west = chunk.getHeight(Heightmap.Types.WORLD_SURFACE_WG, Math.max(x - 1, 0), z);
+                        int east = chunk.getHeight(Heightmap.Types.WORLD_SURFACE_WG, Math.min(x + 1, 15), z);
+                        return Math.abs(west - east) >= 4;
+                    }
+                }
+                // Canvas end
                 int i = this.context.blockX & 15;
                 int i1 = this.context.blockZ & 15;
                 int max = Math.max(i1 - 1, 0);
