--- a/net/minecraft/world/level/levelgen/WorldgenRandom.java
+++ b/net/minecraft/world/level/levelgen/WorldgenRandom.java
@@ -73,7 +_,7 @@
     }
 
     public static enum Algorithm {
-        LEGACY(LegacyRandomSource::new),
+        LEGACY(ThreadSafeLegacyRandomSource::new), // Canvas - threadsafe random
         XOROSHIRO(XoroshiroRandomSource::new);
 
         private final LongFunction<RandomSource> constructor;
