--- a/net/minecraft/world/level/levelgen/WorldOptions.java
+++ b/net/minecraft/world/level/levelgen/WorldOptions.java
@@ -21,6 +_,7 @@
     );
     public static final WorldOptions DEMO_OPTIONS = new WorldOptions("North Carolina".hashCode(), true, true);
     private final long seed;
+    private final long biomeZoomSeed; // Canvas - cache biome zoom seed
     private final boolean generateStructures;
     private final boolean generateBonusChest;
     private final Optional<String> legacyCustomOptions;
@@ -42,7 +_,14 @@
         this.generateStructures = generateStructures;
         this.generateBonusChest = generateBonusChest;
         this.legacyCustomOptions = legacyCustomOptions;
-    }
+        this.biomeZoomSeed = net.minecraft.world.level.biome.BiomeManager.obfuscateSeed(seed); // Canvas - cache biome zoom seed
+    }
+
+    // Canvas start - cache biome zoom seed
+    public long biomeZoomSeed() {
+        return this.biomeZoomSeed;
+    }
+    // Canvas end - cache biome zoom seed
 
     public long seed() {
         return this.seed;
