--- a/net/minecraft/world/level/levelgen/Xoroshiro128PlusPlus.java
+++ b/net/minecraft/world/level/levelgen/Xoroshiro128PlusPlus.java
@@ -5,8 +_,8 @@
 import net.minecraft.Util;
 
 public class Xoroshiro128PlusPlus {
-    private long seedLo;
-    private long seedHi;
+    public long seedLo; // Canvas - private -> public
+    public long seedHi; // Canvas - private -> public
     public static final Codec<Xoroshiro128PlusPlus> CODEC = Codec.LONG_STREAM
         .comapFlatMap(
             longStream -> Util.fixedSize(longStream, 2).map(longs -> new Xoroshiro128PlusPlus(longs[0], longs[1])),
