--- a/net/minecraft/world/level/levelgen/XoroshiroRandomSource.java
+++ b/net/minecraft/world/level/levelgen/XoroshiroRandomSource.java
@@ -10,7 +_,7 @@
     private static final double DOUBLE_UNIT = 1.110223E-16F;
     public static final Codec<XoroshiroRandomSource> CODEC = Xoroshiro128PlusPlus.CODEC
         .xmap(xoroshiro128PlusPlus -> new XoroshiroRandomSource(xoroshiro128PlusPlus), xoroshiroRandomSource -> xoroshiroRandomSource.randomNumberGenerator);
-    private Xoroshiro128PlusPlus randomNumberGenerator;
+    public Xoroshiro128PlusPlus randomNumberGenerator; // Canvas - private -> public
     private final MarsagliaPolarGaussian gaussianSource = new MarsagliaPolarGaussian(this);
 
     public XoroshiroRandomSource(long seed) {
@@ -106,14 +_,13 @@
         return this.randomNumberGenerator.nextLong() >>> 64 - bits;
     }
 
-    public static class XoroshiroPositionalRandomFactory implements PositionalRandomFactory {
-        private final long seedLo;
-        private final long seedHi;
-
-        public XoroshiroPositionalRandomFactory(long seedLo, long seedHi) {
-            this.seedLo = seedLo;
-            this.seedHi = seedHi;
-        }
+    // Canvas start
+    /**
+     * @param seedLo Canvas - private -> public
+     * @param seedHi Canvas - private -> public
+     */
+    public record XoroshiroPositionalRandomFactory(long seedLo, long seedHi) implements PositionalRandomFactory {
+    // Canvas end
 
         @Override
         public RandomSource at(int x, int y, int z) {
