--- a/net/minecraft/util/debugchart/DebugSampleSubscriptionTracker.java
+++ b/net/minecraft/util/debugchart/DebugSampleSubscriptionTracker.java
@@ -1,8 +_,10 @@
 package net.minecraft.util.debugchart;
 
 import com.google.common.collect.Maps;
+import java.util.Collections;
 import java.util.EnumMap;
 import java.util.LinkedList;
+import java.util.List;
 import java.util.Map;
 import java.util.Queue;
 import net.minecraft.Util;
@@ -15,7 +_,7 @@
     public static final int STOP_SENDING_AFTER_MS = 10000;
     private final PlayerList playerList;
     private final EnumMap<RemoteDebugSampleType, Map<ServerPlayer, DebugSampleSubscriptionTracker.SubscriptionStartedAt>> subscriptions;
-    private final Queue<DebugSampleSubscriptionTracker.SubscriptionRequest> subscriptionRequestQueue = new LinkedList<>();
+    private final List<SubscriptionRequest> subscriptionRequestQueue = Collections.synchronizedList(new LinkedList<>()); // Canvas
 
     public DebugSampleSubscriptionTracker(PlayerList playerList) {
         this.playerList = playerList;
