--- a/net/minecraft/world/entity/monster/Zombie.java
+++ b/net/minecraft/world/entity/monster/Zombie.java
@@ -223,7 +_,7 @@
                     this.doUnderWaterConversion();
                 }
             } else if (this.convertsInWater()) {
-                if (this.isEyeInFluid(FluidTags.WATER)) {
+                if (this.canvas$isEyeInWater()) { // Canvas - optimize fluidOnEyes
                     this.inWaterTime++;
                     if (this.inWaterTime >= 600) {
                         this.startUnderWaterConversion(300);
@@ -358,10 +_,15 @@
                     int i2 = floor1 + Mth.nextInt(this.random, 7, 40) * Mth.nextInt(this.random, -1, 1);
                     int i3 = floor2 + Mth.nextInt(this.random, 7, 40) * Mth.nextInt(this.random, -1, 1);
                     BlockPos blockPos = new BlockPos(i1, i2, i3);
+                    // Canvas start - prevent reinforcement checks from loading chunks
+                    if (this.level().getChunkIfLoadedImmediately(blockPos.getX() >> 4, blockPos.getZ() >> 4) == null) {
+                        continue;
+                    }
+                    // Canvas end - prevent reinforcement checks from loading chunks
                     if (SpawnPlacements.isSpawnPositionOk(type, level, blockPos)
                         && SpawnPlacements.checkSpawnRules(type, level, EntitySpawnReason.REINFORCEMENT, blockPos, level.random)) {
                         zombie.setPos(i1, i2, i3);
-                        if (!level.hasNearbyAlivePlayerThatAffectsSpawning(i1, i2, i3, 7.0) // Paper - affects spawning api
+                        if (!level.canvas$hasNearbyAlivePlayerThatAffectsSpawningForZombie(i1, i2, i3, 7.0) // Paper - affects spawning api // Canvas - optimize nearby alive player checks for zombies
                             && level.isUnobstructed(zombie)
                             && level.noCollision(zombie)
                             && (zombie.canSpawnInLiquids() || !level.containsAnyLiquid(zombie.getBoundingBox()))) {
