--- a/net/minecraft/world/entity/ai/behavior/BabyFollowAdult.java
+++ b/net/minecraft/world/entity/ai/behavior/BabyFollowAdult.java
@@ -29,6 +_,12 @@
                             return false;
                         } else {
                             LivingEntity livingEntity = instance.get(memoryAccessor);
+                            // Canvas start - region threading
+                            if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(livingEntity)) {
+                                memoryAccessor.erase();
+                                return true;
+                            }
+                            // Canvas end - region threading
                             if (entity.closerThan(livingEntity, followRange.getMaxValue() + 1) && !entity.closerThan(livingEntity, followRange.getMinValue())) {
                                 // CraftBukkit start
                                 org.bukkit.event.entity.EntityTargetLivingEntityEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityTargetLivingEvent(entity, livingEntity, org.bukkit.event.entity.EntityTargetEvent.TargetReason.FOLLOW_LEADER);
@@ -39,6 +_,12 @@
                                     memoryAccessor.erase();
                                     return true;
                                 }
+                                // Canvas start - region threading
+                                if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(((org.bukkit.craftbukkit.entity.CraftEntity)event.getTarget()).getHandleRaw())) {
+                                    memoryAccessor.erase();
+                                    return true;
+                                }
+                                // Canvas end - region threading
                                 livingEntity = ((org.bukkit.craftbukkit.entity.CraftLivingEntity) event.getTarget()).getHandle();
                                 // CraftBukkit end
                                 WalkTarget walkTarget = new WalkTarget(
