From 30322d32a063dc3b6c41ba7f9d75c2e508ff7cf1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java b/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
index 7399e0e47e125f95790eaa5b5e863ce96688e5b7..9efe85659cb663ee688946d90a558742770e476b 100644
--- a/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
+++ b/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
@@ -115,10 +115,33 @@ public class BehaviorUtils {
 
     public static SectionPos findSectionClosestToVillage(final ServerLevel level, final SectionPos center, final int radius) {
         int distToVillage = level.sectionsToVillage(center);
-        return SectionPos.cube(center, radius)
-            .filter(s -> level.sectionsToVillage(s) < distToVillage)
-            .min(Comparator.comparingInt(level::sectionsToVillage))
-            .orElse(center);
+        // Canvas start - remove streams
+        SectionPos closestSection = center;
+        int closestDistance = distToVillage;
+
+        final int lowerX = center.getX() - radius;
+        final int lowerY = center.getY() - radius;
+        final int lowerZ = center.getZ() - radius;
+        final int upperX = center.getX() + radius;
+        final int upperY = center.getY() + radius;
+        final int upperZ = center.getZ() + radius;
+
+        for (int x = lowerX; x <= upperX; x++) {
+            for (int z = lowerZ; z <= upperZ; z++) {
+                for (int y = lowerY; y <= upperY; y++) {
+                    SectionPos pos = SectionPos.of(x, y, z);
+                    int distance = level.sectionsToVillage(pos);
+
+                    if (distance < closestDistance) {
+                        closestDistance = distance;
+                        closestSection = pos;
+                    }
+                }
+            }
+        }
+
+        return closestSection;
+        // Canvas end - remove streams
     }
 
     public static boolean isWithinAttackRange(final Mob body, final LivingEntity target, final int projectileAttackRangeMargin) {
