--- a/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
+++ b/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
@@ -101,10 +_,33 @@
 
     public static SectionPos findSectionClosestToVillage(ServerLevel level, SectionPos sectionPos, int radius) {
         int i = level.sectionsToVillage(sectionPos);
-        return SectionPos.cube(sectionPos, radius)
-            .filter(pos -> level.sectionsToVillage(pos) < i)
-            .min(Comparator.comparingInt(level::sectionsToVillage))
-            .orElse(sectionPos);
+        // Canvas start - remove streams
+        SectionPos closestSection = sectionPos;
+        int closestDistance = i;
+
+        final int lowerX = sectionPos.getX() - radius;
+        final int lowerY = sectionPos.getY() - radius;
+        final int lowerZ = sectionPos.getZ() - radius;
+        final int upperX = sectionPos.getX() + radius;
+        final int upperY = sectionPos.getY() + radius;
+        final int upperZ = sectionPos.getZ() + radius;
+
+        for (int x = lowerX; x <= upperX; x++) {
+            for (int z = lowerZ; z <= upperZ; z++) {
+                for (int y = lowerY; y <= upperY; y++) {
+                    SectionPos pos = SectionPos.of(x, y, z);
+                    int distance = level.sectionsToVillage(pos);
+
+                    if (distance < closestDistance) {
+                        closestDistance = distance;
+                        closestSection = pos;
+                    }
+                }
+            }
+        }
+
+        return closestSection;
+        // Canvas end - remove streams
     }
 
     public static boolean isWithinAttackRange(Mob mob, LivingEntity target, int cooldown) {
