From 4b22e4fb667ff3bf3c489ff7a94018cc27157519 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/behavior/Behavior.java b/net/minecraft/world/entity/ai/behavior/Behavior.java
index b620d38e61ae560074b551a92afefe4022d5078e..fe92da446c130301e485071b37023b672c13fffc 100644
--- a/net/minecraft/world/entity/ai/behavior/Behavior.java
+++ b/net/minecraft/world/entity/ai/behavior/Behavior.java
@@ -16,6 +16,8 @@ public abstract class Behavior<E extends LivingEntity> implements BehaviorContro
     private final int minDuration;
     private final int maxDuration;
     private final String configKey; // Paper - configurable behavior tick rate and timings
+    private long lithium$cachedMemoryModCount = -1; // Canvas - optimize brain
+    private boolean lithium$cachedHasRequiredMemoryState; // Canvas - optimize brain
 
     public Behavior(final Map<MemoryModuleType<?>, MemoryStatus> entryCondition) {
         this(entryCondition, DEFAULT_DURATION);
@@ -28,7 +30,7 @@ public abstract class Behavior<E extends LivingEntity> implements BehaviorContro
     public Behavior(final Map<MemoryModuleType<?>, MemoryStatus> entryCondition, final int minDuration, final int maxDuration) {
         this.minDuration = minDuration;
         this.maxDuration = maxDuration;
-        this.entryCondition = entryCondition;
+        this.entryCondition = new it.unimi.dsi.fastutil.objects.Reference2ObjectOpenHashMap<>(entryCondition); // Canvas - optimize brain
         // Paper start - configurable behavior tick rate and timings
         String key = this.getClass().getName();
         int lastSeparator = key.lastIndexOf('.');
@@ -110,15 +112,24 @@ public abstract class Behavior<E extends LivingEntity> implements BehaviorContro
     }
 
     protected boolean hasRequiredMemories(final E body) {
-        for (Entry<MemoryModuleType<?>, MemoryStatus> entry : this.entryCondition.entrySet()) {
-            MemoryModuleType<?> memoryType = entry.getKey();
-            MemoryStatus requiredStatus = entry.getValue();
-            if (!body.getBrain().checkMemory(memoryType, requiredStatus)) {
-                return false;
+        // Canvas start - optimize brain
+        net.minecraft.world.entity.ai.Brain<?> brain = body.getBrain();
+        long modCount = brain.lithium$getMemoryValueModCount();
+        if (this.lithium$cachedMemoryModCount == modCount) {
+            return this.lithium$cachedHasRequiredMemoryState;
+        }
+        this.lithium$cachedMemoryModCount = modCount;
+
+        it.unimi.dsi.fastutil.objects.ObjectIterator<it.unimi.dsi.fastutil.objects.Reference2ObjectMap.Entry<net.minecraft.world.entity.ai.memory.MemoryModuleType<?>, net.minecraft.world.entity.ai.memory.MemoryStatus>> fastIterator = ((it.unimi.dsi.fastutil.objects.Reference2ObjectOpenHashMap<net.minecraft.world.entity.ai.memory.MemoryModuleType<?>, net.minecraft.world.entity.ai.memory.MemoryStatus>) this.entryCondition).reference2ObjectEntrySet().fastIterator();
+        while (fastIterator.hasNext()) {
+            it.unimi.dsi.fastutil.objects.Reference2ObjectMap.Entry<MemoryModuleType<?>, MemoryStatus> entry = fastIterator.next();
+            if (!brain.checkMemory(entry.getKey(), entry.getValue())) {
+                return this.lithium$cachedHasRequiredMemoryState = false;
             }
         }
 
-        return true;
+        return this.lithium$cachedHasRequiredMemoryState = true;
+        // Canvas end - optimize brain
     }
 
     public static enum Status {
