--- a/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
+++ b/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
@@ -69,6 +_,7 @@
     }
 
     public static void setWalkAndLookTargetMemories(LivingEntity entity, PositionTracker positionTracker, float speedModifier, int closeEnoughDist) {
+        if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(entity.level(), positionTracker.currentBlockPosition())) return; // Canvas - region threading
         WalkTarget walkTarget = new WalkTarget(positionTracker, speedModifier, closeEnoughDist);
         entity.getBrain().setMemory(MemoryModuleType.LOOK_TARGET, positionTracker);
         entity.getBrain().setMemory(MemoryModuleType.WALK_TARGET, walkTarget);
@@ -100,10 +_,33 @@
 
     public static SectionPos findSectionClosestToVillage(ServerLevel serverLevel, SectionPos sectionPos, int radius) {
         int i = serverLevel.sectionsToVillage(sectionPos);
-        return SectionPos.cube(sectionPos, radius)
-            .filter(pos -> serverLevel.sectionsToVillage(pos) < i)
-            .min(Comparator.comparingInt(serverLevel::sectionsToVillage))
-            .orElse(sectionPos);
+        // Canvas start - remove streams
+        SectionPos closestSection = sectionPos;
+        int closestDistance = i;
+
+        final int lowerX = sectionPos.getX() - radius;
+        final int lowerY = sectionPos.getY() - radius;
+        final int lowerZ = sectionPos.getZ() - radius;
+        final int upperX = sectionPos.getX() + radius;
+        final int upperY = sectionPos.getY() + radius;
+        final int upperZ = sectionPos.getZ() + radius;
+
+        for (int x = lowerX; x <= upperX; x++) {
+            for (int z = lowerZ; z <= upperZ; z++) {
+                for (int y = lowerY; y <= upperY; y++) {
+                    SectionPos pos = SectionPos.of(x, y, z);
+                    int distance = serverLevel.sectionsToVillage(pos);
+
+                    if (distance < closestDistance) {
+                        closestDistance = distance;
+                        closestSection = pos;
+                    }
+                }
+            }
+        }
+
+        return closestSection;
+        // Canvas end - remove streams
     }
 
     public static boolean isWithinAttackRange(Mob mob, LivingEntity target, int cooldown) {
