--- a/net/minecraft/world/entity/ai/behavior/CelebrateVillagersSurvivedRaid.java
+++ b/net/minecraft/world/entity/ai/behavior/CelebrateVillagersSurvivedRaid.java
@@ -1,6 +_,7 @@
 package net.minecraft.world.entity.ai.behavior;
 
 import com.google.common.collect.ImmutableMap;
+import io.canvasmc.canvas.Config;
 import it.unimi.dsi.fastutil.ints.IntList;
 import java.util.List;
 import javax.annotation.Nullable;
@@ -52,7 +_,7 @@
             owner.playCelebrateSound();
         }
 
-        if (random.nextInt(200) == 0 && MoveToSkySeeingSpot.hasNoBlocksAbove(level, owner, owner.blockPosition())) {
+        if ((Config.INSTANCE.entities.villagerCelebrationFireworksChance != -1 && random.nextInt(100) < Config.INSTANCE.entities.villagerCelebrationFireworksChance) && MoveToSkySeeingSpot.hasNoBlocksAbove(level, owner, owner.blockPosition())) { // Canvas - villager celebration chance
             DyeColor dyeColor = Util.getRandom(DyeColor.values(), random);
             int randomInt = random.nextInt(3);
             ItemStack firework = this.getFirework(dyeColor, randomInt);
