From e64d3ec2ef48556e0a10e7686e1c057ce53f22aa Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/behavior/GateBehavior.java b/net/minecraft/world/entity/ai/behavior/GateBehavior.java
index 9fb025b05cacc901d72d93c16cf2b7e48d4dd9a6..c5bd9d5da977f8b714736437f49374e20fba574b 100644
--- a/net/minecraft/world/entity/ai/behavior/GateBehavior.java
+++ b/net/minecraft/world/entity/ai/behavior/GateBehavior.java
@@ -85,7 +85,17 @@ public class GateBehavior<E extends LivingEntity> implements BehaviorControl<E>
             }
         }
         // Paper end - Perf: Remove streams from hot code
-        if (this.behaviors.stream().noneMatch(behavior -> behavior.getStatus() == Behavior.Status.RUNNING)) {
+        // Canvas start - remove streams
+        boolean hasRunningTask = false;
+        for (final BehaviorControl<? super E> behavior : this.behaviors) {
+            if (behavior.getStatus() == Behavior.Status.RUNNING) {
+                hasRunningTask = true;
+                break;
+            }
+        }
+
+        if (!hasRunningTask) {
+        // Canvas end - remove streams
             this.doStop(level, body, timestamp);
         }
     }
