--- a/net/minecraft/world/entity/ai/behavior/InsideBrownianWalk.java
+++ b/net/minecraft/world/entity/ai/behavior/InsideBrownianWalk.java
@@ -24,12 +_,18 @@
                                 .map(BlockPos::immutable)
                                 .collect(Util.toMutableList());
                             Collections.shuffle(list);
-                            list.stream()
-                                .filter(pos -> !level.canSeeSky(pos))
-                                .filter(pos -> level.loadedAndEntityCanStandOn(pos, mob))
-                                .filter(pos -> level.noCollision(mob))
-                                .findFirst()
-                                .ifPresent(pos -> walkTarget.set(new WalkTarget(pos, speedModifier, 0)));
+                            // Canvas start - remove streams
+                            for (BlockPos pos1 : list) {
+                                if (!level.canSeeSky(pos1)) {
+                                    if (level.loadedAndEntityCanStandOn(pos1, mob)) {
+                                        if (level.noCollision(mob)) {
+                                            walkTarget.set(new WalkTarget(pos1, speedModifier, 0));
+                                            break;
+                                        }
+                                    }
+                                }
+                            }
+                            // Canvas end
                             return true;
                         }
                     }
