From 73e3ddbec8cd225e42fdf1e4683ca01a6c53e643 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/behavior/InsideBrownianWalk.java b/net/minecraft/world/entity/ai/behavior/InsideBrownianWalk.java
index 37d902e81435f3a29f9a2706bbc2e86791c8f4b6..8351f73f8116afb9dd6399e704ddbdb8e21262f4 100644
--- a/net/minecraft/world/entity/ai/behavior/InsideBrownianWalk.java
+++ b/net/minecraft/world/entity/ai/behavior/InsideBrownianWalk.java
@@ -20,16 +20,50 @@ public class InsideBrownianWalk {
                             return false;
                         } else {
                             BlockPos bodyPos = body.blockPosition();
-                            List<BlockPos> poses = BlockPos.betweenClosedStream(bodyPos.offset(-1, -1, -1), bodyPos.offset(1, 1, 1))
-                                .map(BlockPos::immutable)
-                                .collect(Util.toMutableList());
-                            Collections.shuffle(poses);
-                            poses.stream()
-                                .filter(pos -> !level.canSeeSky(pos))
-                                .filter(pos -> level.loadedAndEntityCanStandOn(pos, body))
-                                .filter(pos -> level.noCollision(body))
-                                .findFirst()
-                                .ifPresent(target -> walkTarget.set(new WalkTarget(target, speedModifier, 0)));
+                            // Canvas start - remove streams
+                            BlockPos minPos = bodyPos.offset(-1, -1, -1);
+                            BlockPos maxPos = bodyPos.offset(1, 1, 1);
+                            // remove list, use BlockPos[], precompute total size
+                            int minX = minPos.getX();
+                            int minY = minPos.getY();
+                            int minZ = minPos.getZ();
+                            int maxX = maxPos.getX();
+                            int maxY = maxPos.getY();
+                            int maxZ = maxPos.getZ();
+
+                            int sizeX = maxX - minX + 1;
+                            int sizeY = maxY - minY + 1;
+                            int sizeZ = maxZ - minZ + 1;
+                            int arrSize = sizeX * sizeY * sizeZ;
+
+                            BlockPos[] positions = new BlockPos[arrSize];
+                            int index = 0;
+
+                            for (int x = minX; x <= maxX; x++) {
+                                for (int y = minY; y <= maxY; y++) {
+                                    for (int z = minZ; z <= maxZ; z++) {
+                                        positions[index++] = new BlockPos(x, y, z).immutable();
+                                    }
+                                }
+                            }
+
+                            // copied from it.unimi.dsi.fastutil.objects.ObjectArrays.shuffle() and modified to use 'mob.random'
+                            for (int idx = positions.length; idx-- != 0;) {
+                                final int p = body.random.nextInt(idx + 1);
+                                final BlockPos t = positions[idx];
+                                positions[idx] = positions[p];
+                                positions[p] = t;
+                            }
+
+                            for (BlockPos pos : positions) {
+                                if (!level.canSeeSky(pos) &&
+                                    level.loadedAndEntityCanStandOn(pos, body) &&
+                                    level.noCollision(body)) {
+                                    walkTarget.set(new WalkTarget(pos, speedModifier, 0));
+                                    break;
+                                }
+                            }
+                            // Canvas end - remove streams
                             return true;
                         }
                     }
