From 0ce24b7bccfa2d01cdd888671365a9ba0973fab1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java b/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java
index 843d6897a43ed9e9d28b29608401b78266f57f61..8f596d7fdc7374a0027476c515f79a1a2259e966 100644
--- a/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java
+++ b/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java
@@ -55,11 +55,15 @@ public class SetClosestHomeAsWalkTarget {
                                         return true;
                                     }
                                 };
-                                Set<Pair<Holder<PoiType>, BlockPos>> pois = poiManager.findAllWithType(
-                                        p -> p.is(PoiTypes.HOME), cacheTest, body.blockPosition(), AcquirePoi.SCAN_RANGE, PoiManager.Occupancy.ANY
-                                    )
-                                    .collect(Collectors.toSet());
-                                Path path = AcquirePoi.findPathToPois(body, pois);
+                                // Canvas start - reroute to paper poi logic
+                                java.util.List<Pair<Holder<PoiType>, BlockPos>> poips = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
+                                ca.spottedleaf.moonrise.patches.poi_lookup.PoiAccess.findAnyPoiPositions(
+                                    poiManager, p -> p.is(PoiTypes.HOME), cacheTest,
+                                    body.blockPosition(), net.minecraft.world.entity.ai.behavior.AcquirePoi.SCAN_RANGE, // 48
+                                    PoiManager.Occupancy.ANY, false, Integer.MAX_VALUE, poips
+                                );
+                                Path path = AcquirePoi.findPathToPois(body, new java.util.HashSet<>(poips));
+                                // Canvas end - reroute to paper poi logic
                                 if (path != null && path.canReach()) {
                                     BlockPos targetPos = path.getTarget();
                                     Optional<Holder<PoiType>> type = poiManager.getType(targetPos);
