--- a/net/minecraft/world/entity/ai/behavior/SetLookAndInteract.java
+++ b/net/minecraft/world/entity/ai/behavior/SetLookAndInteract.java
@@ -20,7 +_,7 @@
                     i,
                     (lookTarget, interactionTarget, nearestEntities) -> (level, body, timestamp) -> {
                         Optional<LivingEntity> closest = i.<NearestVisibleLivingEntities>get(nearestEntities)
-                            .findClosest(e -> e.distanceToSqr(body) <= interactionRangeSqr && e.is(type));
+                            .findClosest(e -> e.is(type) && e.distanceToSqr(body) <= interactionRangeSqr); // Canvas - optimize
                         if (closest.isEmpty()) {
                             return false;
                         }
