--- a/net/minecraft/world/entity/ai/behavior/Swim.java
+++ b/net/minecraft/world/entity/ai/behavior/Swim.java
@@ -14,7 +_,7 @@
     }
 
     public static <T extends Mob> boolean shouldSwim(T mob) {
-        return mob.isInWater() && mob.getFluidHeight(FluidTags.WATER) > mob.getFluidJumpThreshold() || mob.isInLava();
+        return mob.isInWater() && mob.canvas$getWaterFluidHeight() > mob.getFluidJumpThreshold() || mob.isInLava(); // Canvas - optimize fluid height
     }
 
     @Override
