From 12dcdabd241e6617b95f15084fa04e5aaff0bb2b Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/behavior/Swim.java b/net/minecraft/world/entity/ai/behavior/Swim.java
index 06771975710ebc5867f843b2007144c998bddd5f..82a0edb9f3fe186f7720105f997bcbebd6c433d0 100644
--- a/net/minecraft/world/entity/ai/behavior/Swim.java
+++ b/net/minecraft/world/entity/ai/behavior/Swim.java
@@ -14,7 +14,7 @@ public class Swim<T extends Mob> extends Behavior<T> {
     }
 
     public static <T extends Mob> boolean shouldSwim(final T mob) {
-        return mob.isInWater() && mob.getFluidHeight(FluidTags.WATER) > mob.getFluidJumpThreshold() || mob.isInLava();
+        return mob.isInWater() && mob.canvas$getWaterFluidHeight() > mob.getFluidJumpThreshold() || mob.isInLava(); // Canvas - optimize fluid height
     }
 
     @Override
