From 5462f8f54a2af66b338154bdab1b4b47bd772d5d Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/behavior/TradeWithVillager.java b/net/minecraft/world/entity/ai/behavior/TradeWithVillager.java
index 1d405a77c4cf70bd818a4f59163e5b319808d5f3..8c51d213e3bcee15d78679b79894a0a2ec4cb2c6 100644
--- a/net/minecraft/world/entity/ai/behavior/TradeWithVillager.java
+++ b/net/minecraft/world/entity/ai/behavior/TradeWithVillager.java
@@ -73,7 +73,17 @@ public class TradeWithVillager extends Behavior<Villager> {
     private static Set<Item> figureOutWhatIAmWillingToTrade(final Villager myBody, final Villager target) {
         ImmutableSet<Item> targetItems = target.getVillagerData().profession().value().requestedItems();
         ImmutableSet<Item> selfItems = myBody.getVillagerData().profession().value().requestedItems();
-        return targetItems.stream().filter(entry -> !selfItems.contains(entry)).collect(Collectors.toSet());
+        // Canvas start - remove streams
+        Set<Item> result = new java.util.HashSet<>();
+
+        for (Item item : targetItems) {
+            if (!selfItems.contains(item)) {
+                result.add(item);
+            }
+        }
+
+        return result;
+        // Canvas end - remove streams
     }
 
     private static void throwHalfStack(final Villager villager, final Set<Item> items, final LivingEntity target) {
