--- a/net/minecraft/world/entity/ai/behavior/VillagerGoalPackages.java
+++ b/net/minecraft/world/entity/ai/behavior/VillagerGoalPackages.java
@@ -54,13 +_,15 @@
                     MemoryModuleType.POTENTIAL_JOB_SITE,
                     true,
                     Optional.empty(),
-                    (l, p) -> true
+                    (l, p) -> true, (body) -> body.level().canvasConfig().entities.villagers.reduceJobSitePoiSearchRange ? 16 : 48, (body) -> body.level().canvasConfig().entities.villagers.villagerAcquirePoiTasksLoadChunks // Canvas - configurable scan range and if poi should load
                 )
             ),
             Pair.of(7, new GoToPotentialJobSite(speedModifier)),
             Pair.of(8, YieldJobSite.create(speedModifier)),
-            Pair.of(10, AcquirePoi.create(p -> p.is(PoiTypes.HOME), MemoryModuleType.HOME, false, Optional.of((byte)14), VillagerGoalPackages::validateBedPoi)),
-            Pair.of(10, AcquirePoi.create(p -> p.is(PoiTypes.MEETING), MemoryModuleType.MEETING_POINT, true, Optional.of((byte)14))),
+            // Canvas start - configurable scan range and if poi should load
+            Pair.of(10, AcquirePoi.create(p -> p.is(PoiTypes.HOME), MemoryModuleType.HOME, MemoryModuleType.HOME, false, Optional.of((byte)14), VillagerGoalPackages::validateBedPoi, (body) -> body.level().canvasConfig().entities.villagers.reduceHomePoiSearchRange ? 16 : 48, (body) -> body.level().canvasConfig().entities.villagers.villagerAcquirePoiTasksLoadChunks)),
+            Pair.of(10, AcquirePoi.create(p -> p.is(PoiTypes.MEETING), MemoryModuleType.MEETING_POINT, MemoryModuleType.MEETING_POINT, true, Optional.of((byte)14), (_, _) -> true, (body) -> body.level().canvasConfig().entities.villagers.reduceMeetingPointPoiSearchRange ? 16 : 48, (body) -> body.level().canvasConfig().entities.villagers.villagerAcquirePoiTasksLoadChunks)),
+            // Canvas end - configurable scan range and if poi should load
             Pair.of(10, AssignProfessionFromJobSite.create()),
             Pair.of(10, ResetProfession.create())
         );
