--- a/net/minecraft/world/entity/ai/behavior/VillagerGoalPackages.java
+++ b/net/minecraft/world/entity/ai/behavior/VillagerGoalPackages.java
@@ -21,6 +_,8 @@
 import net.minecraft.world.entity.raid.Raid;
 import net.minecraft.world.level.block.BedBlock;
 import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.chunk.LevelChunk;
+import org.jetbrains.annotations.Nullable;
 
 public class VillagerGoalPackages {
     private static final float STROLL_SPEED_MODIFIER = 0.4F;
@@ -69,7 +_,14 @@
     }
 
     private static boolean validateBedPoi(ServerLevel level, BlockPos pos) {
-        BlockState blockState = level.getBlockState(pos);
+        // Canvas start - dont call syncload
+        @Nullable LevelChunk chunk = level.getChunkIfLoaded(pos);
+        if (chunk == null) {
+            // chunk not loaded, so there is no bed, return false.
+            return false;
+        }
+        BlockState blockState = chunk.getBlockState(pos);
+        // Canvas end
         return blockState.is(BlockTags.BEDS) && !blockState.getValue(BedBlock.OCCUPIED);
     }
 
