--- a/net/minecraft/world/entity/ai/behavior/YieldJobSite.java
+++ b/net/minecraft/world/entity/ai/behavior/YieldJobSite.java
@@ -48,23 +_,24 @@
                             return true;
                         }
 
-                        i.<List<LivingEntity>>get(nearestEntities)
-                            .stream()
-                            .filter(v -> v instanceof Villager && v != body)
-                            .map(v -> (Villager)v)
-                            .filter(LivingEntity::isAlive)
-                            .filter(v -> nearbyWantsJobsite(poiType.get(), v, poiPos))
-                            .findFirst()
-                            .ifPresent(nearbyVillager -> {
+                        // Canvas start - remove streams
+                        for (LivingEntity living : i.get(nearestEntities)) {
+                            if (living instanceof Villager villager &&
+                                living != body && villager.isAlive() &&
+                                nearbyWantsJobsite(poiType.get(), villager, poiPos)
+                            ) {
                                 walkTarget.erase();
                                 lookTarget.erase();
                                 potentialJob.erase();
-                                if (nearbyVillager.getBrain().getMemory(MemoryModuleType.JOB_SITE).isEmpty()) {
-                                    BehaviorUtils.setWalkAndLookTargetMemories(nearbyVillager, poiPos, speedModifier, 1);
-                                    nearbyVillager.getBrain().setMemory(MemoryModuleType.POTENTIAL_JOB_SITE, GlobalPos.of(level.dimension(), poiPos));
+                                if (villager.getBrain().getMemory(MemoryModuleType.JOB_SITE).isEmpty()) {
+                                    BehaviorUtils.setWalkAndLookTargetMemories(villager, poiPos, speedModifier, 1);
+                                    villager.getBrain().setMemory(MemoryModuleType.POTENTIAL_JOB_SITE, GlobalPos.of(level.dimension(), poiPos));
                                     level.debugSynchronizers().updatePoi(poiPos);
                                 }
-                            });
+                                break;
+                            }
+                        }
+                        // Canvas end - remove streams
                         return true;
                     }
                 )
