--- a/net/minecraft/world/entity/ai/behavior/YieldJobSite.java
+++ b/net/minecraft/world/entity/ai/behavior/YieldJobSite.java
@@ -44,23 +_,26 @@
                             if (type.isEmpty()) {
                                 return true;
                             } else {
-                                instance.<List<LivingEntity>>get(nearestLivingEntities)
-                                    .stream()
-                                    .filter(nearEntity -> nearEntity instanceof Villager && nearEntity != villager)
-                                    .map(nearEntity -> (Villager)nearEntity)
-                                    .filter(LivingEntity::isAlive)
-                                    .filter(nearVillager -> nearbyWantsJobsite(type.get(), nearVillager, blockPos))
-                                    .findFirst()
-                                    .ifPresent(nearVillager -> {
-                                        walkTarget.erase();
-                                        lookTarget.erase();
-                                        potentialJobSite.erase();
-                                        if (nearVillager.getBrain().getMemory(MemoryModuleType.JOB_SITE).isEmpty()) {
-                                            BehaviorUtils.setWalkAndLookTargetMemories(nearVillager, blockPos, speedModifier, 1);
-                                            nearVillager.getBrain().setMemory(MemoryModuleType.POTENTIAL_JOB_SITE, GlobalPos.of(level.dimension(), blockPos));
-                                            DebugPackets.sendPoiTicketCountPacket(level, blockPos);
+                                // Canvas start - remove streams
+                                List<LivingEntity> mobsList = instance.get(nearestLivingEntities);
+                                for (LivingEntity nearEntity : mobsList) {
+                                    if (nearEntity instanceof Villager nearVillager && nearEntity != villager && nearEntity.isAlive()) {
+                                        if (nearbyWantsJobsite(type.get(), nearVillager, blockPos)) {
+                                            walkTarget.erase();
+                                            lookTarget.erase();
+                                            potentialJobSite.erase();
+
+                                            if (nearVillager.getBrain().getMemory(MemoryModuleType.JOB_SITE).isEmpty()) {
+                                                BehaviorUtils.setWalkAndLookTargetMemories(nearVillager, blockPos, speedModifier, 1);
+                                                nearVillager.getBrain().setMemory(MemoryModuleType.POTENTIAL_JOB_SITE, GlobalPos.of(level.dimension(), blockPos));
+                                                DebugPackets.sendPoiTicketCountPacket(level, blockPos);
+                                            }
+
+                                            break;
                                         }
-                                    });
+                                    }
+                                }
+                                // Canvas end - remove streams
                                 return true;
                             }
                         }
