--- a/src/main/java/org/bukkit/map/MapPalette.java
+++ b/src/main/java/org/bukkit/map/MapPalette.java
@@ -395,9 +_,15 @@
         temp.getRGB(0, 0, temp.getWidth(), temp.getHeight(), pixels, 0, temp.getWidth());
 
         byte[] result = new byte[temp.getWidth() * temp.getHeight()];
+        // Pufferfish start - SIMD
+        if (!io.canvasmc.canvas.simd.SIMDDetection.isEnabled) {
         for (int i = 0; i < pixels.length; i++) {
             result[i] = matchColor(new Color(pixels[i], true));
         }
+        } else {
+            io.canvasmc.canvas.simd.VectorMapPalette.matchColorVectorized(pixels, result);
+        }
+        // Pufferfish end - SIMD
         return result;
     }
 
