--- a/net/minecraft/world/level/storage/LevelData.java
+++ b/net/minecraft/world/level/storage/LevelData.java
@@ -41,7 +_,7 @@
     }
 
     public record RespawnData(GlobalPos globalPos, float yaw, float pitch) {
-        public static final LevelData.RespawnData DEFAULT = new LevelData.RespawnData(GlobalPos.of(Level.OVERWORLD, BlockPos.ZERO), 0.0F, 0.0F);
+        public static final LevelData.RespawnData DEFAULT = new LevelData.RespawnData(GlobalPos.of(io.canvasmc.canvas.Config.INSTANCE.fetchRespawnDimensionKey(), BlockPos.ZERO), 0.0F, 0.0F); // Canvas - default respawn dimension config
         public static final MapCodec<LevelData.RespawnData> MAP_CODEC = RecordCodecBuilder.mapCodec(
             instance -> instance.group(
                     GlobalPos.MAP_CODEC.forGetter(LevelData.RespawnData::globalPos),
