From cd913ee721ba4988f5421e570c30420c95c78467 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/storage/LevelData.java b/net/minecraft/world/level/storage/LevelData.java
index 96d5ea24515d4fdc0edfc62ce33b0a434be69894..748e5b69fd0697b5aeec82df089ccc2730f5667b 100644
--- a/net/minecraft/world/level/storage/LevelData.java
+++ b/net/minecraft/world/level/storage/LevelData.java
@@ -31,7 +31,7 @@ public interface LevelData {
     }
 
     public record RespawnData(GlobalPos globalPos, float yaw, float pitch) {
-        public static final LevelData.RespawnData DEFAULT = new LevelData.RespawnData(GlobalPos.of(Level.OVERWORLD, BlockPos.ZERO), 0.0F, 0.0F);
+        public static final LevelData.RespawnData DEFAULT = new LevelData.RespawnData(GlobalPos.of(io.canvasmc.canvas.Config.INSTANCE.fetchRespawnDimensionKey(), BlockPos.ZERO), 0.0F, 0.0F); // Canvas - default respawn dimension config
         public static final MapCodec<LevelData.RespawnData> MAP_CODEC = RecordCodecBuilder.mapCodec(
             i -> i.group(
                     GlobalPos.MAP_CODEC.forGetter(LevelData.RespawnData::globalPos),
