--- a/net/minecraft/world/level/storage/PrimaryLevelData.java
+++ b/net/minecraft/world/level/storage/PrimaryLevelData.java
@@ -11,6 +_,8 @@
 import java.util.UUID;
 import java.util.stream.Collectors;
 import javax.annotation.Nullable;
+import io.canvasmc.canvas.server.level.distance.component.GlobalDistanceComponent;
+import io.canvasmc.canvas.server.level.distance.component.WorldSpecificViewDistanceComponents;
 import net.minecraft.CrashReportCategory;
 import net.minecraft.SharedConstants;
 import net.minecraft.Util;
@@ -189,7 +_,7 @@
         Lifecycle worldGenSettingsLifecycle
     ) {
         long _long = tag.get("Time").asLong(0L);
-        return new PrimaryLevelData(
+        PrimaryLevelData data = new PrimaryLevelData( // Canvas - world specific view distance
             tag.get("Player").flatMap(CompoundTag.CODEC::parse).result().orElse(null),
             tag.get("WasModded").asBoolean(false),
             new BlockPos(tag.get("SpawnX").asInt(0), tag.get("SpawnY").asInt(0), tag.get("SpawnZ").asInt(0)),
@@ -224,6 +_,13 @@
             specialWorldProperty,
             worldGenSettingsLifecycle
         );
+        // Canvas start - world specific view distance
+        GlobalDistanceComponent component = new GlobalDistanceComponent();
+        component.globalSimulationDistance = tag.get("GlobalSimulationDistance").asInt(0);
+        component.globalViewDistance = tag.get("GlobalViewDistance").asInt(0);
+        WorldSpecificViewDistanceComponents.GLOBAL_DISTANCES.put(data, component);
+        return data;
+        // Canvas end
     }
 
     @Override
@@ -298,6 +_,12 @@
         }
         nbt.putString("Bukkit.Version", org.bukkit.Bukkit.getName() + "/" + org.bukkit.Bukkit.getVersion() + "/" + org.bukkit.Bukkit.getBukkitVersion()); // CraftBukkit
         this.world.getWorld().storeBukkitValues(nbt); // CraftBukkit - add pdc
+        // Canvas start - world specific view distance
+        if (!WorldSpecificViewDistanceComponents.GLOBAL_DISTANCES.containsKey(this)) {
+            WorldSpecificViewDistanceComponents.GLOBAL_DISTANCES.put(this, new GlobalDistanceComponent());
+        }
+        WorldSpecificViewDistanceComponents.GLOBAL_DISTANCES.get(this).writeToNbt(nbt, registry);
+        // Canvas end
     }
 
     private static ListTag stringCollectionToTag(Set<String> stringCollection) {
