--- a/net/minecraft/world/level/storage/PrimaryLevelData.java
+++ b/net/minecraft/world/level/storage/PrimaryLevelData.java
@@ -72,6 +_,7 @@
     private boolean wasModded;
     private final Set<String> removedFeatureFlags;
     private final TimerQueue<MinecraftServer> scheduledEvents;
+    public final io.canvasmc.canvas.chunk.PerWorldDistanceConfig canvas$distanceConfig; // Canvas - per world distance
 
     // CraftBukkit start - Add world and pdc
     public net.minecraft.core.Registry<net.minecraft.world.level.dimension.LevelStem> customDimensions;
@@ -115,6 +_,7 @@
         WorldOptions worldOptions,
         PrimaryLevelData.SpecialWorldProperty specialWorldProperty,
         Lifecycle worldGenSettingsLifecycle
+        , io.canvasmc.canvas.chunk.PerWorldDistanceConfig canvas$distanceConfig // Canvas - per world distance
     ) {
         this.wasModded = wasModded;
         this.respawnData = respawnData;
@@ -142,6 +_,7 @@
         this.worldOptions = worldOptions;
         this.specialWorldProperty = specialWorldProperty;
         this.worldGenSettingsLifecycle = worldGenSettingsLifecycle;
+        this.canvas$distanceConfig = canvas$distanceConfig; // Canvas - per world distance
     }
 
     public PrimaryLevelData(
@@ -174,6 +_,7 @@
             worldOptions,
             specialWorldProperty,
             worldGenSettingsLifecycle
+            , io.canvasmc.canvas.chunk.PerWorldDistanceConfig.DEFAULT // Canvas - per world distance
         );
     }
 
@@ -218,6 +_,7 @@
             worldOptions,
             specialWorldProperty,
             worldGenSettingsLifecycle
+            , tag.get("Canvas.Distance").read(io.canvasmc.canvas.chunk.PerWorldDistanceConfig.CODEC).resultOrPartial(LOGGER::error).orElse(io.canvasmc.canvas.chunk.PerWorldDistanceConfig.DEFAULT) // Canvas - per world distance
         );
         // Paper start
         data.respawnDimension = tag.get(PAPER_RESPAWN_DIMENSION)
@@ -293,6 +_,7 @@
         tag.storeNullable("WanderingTraderId", UUIDUtil.CODEC, this.wanderingTraderId);
         tag.putString("Bukkit.Version", org.bukkit.Bukkit.getName() + "/" + org.bukkit.Bukkit.getVersion() + "/" + org.bukkit.Bukkit.getBukkitVersion()); // CraftBukkit
         this.world.getWorld().storeBukkitValues(tag); // CraftBukkit - add pdc
+        tag.store("Canvas.Distance", io.canvasmc.canvas.chunk.PerWorldDistanceConfig.CODEC, this.canvas$distanceConfig); // Canvas - per world distance
     }
 
     private static ListTag stringCollectionToTag(Set<String> stringCollection) {
