--- a/net/minecraft/world/level/storage/PrimaryLevelData.java
+++ b/net/minecraft/world/level/storage/PrimaryLevelData.java
@@ -44,6 +_,7 @@
     private final Set<String> knownServerBrands;
     private boolean wasModded;
     private final Set<String> removedFeatureFlags;
+    public final io.canvasmc.canvas.world.PerWorldDistanceConfig canvas$distanceConfig; // Canvas - per world distance
 
     private PrimaryLevelData(
         final @Nullable UUID singlePlayerUUID,
@@ -57,6 +_,7 @@
         final LevelSettings settings,
         final PrimaryLevelData.SpecialWorldProperty specialWorldProperty,
         final Lifecycle worldGenSettingsLifecycle
+        , io.canvasmc.canvas.world.PerWorldDistanceConfig canvas$distanceConfig // Canvas - per world distance
     ) {
         this.wasModded = wasModded;
         this.respawnData = respawnData;
@@ -69,6 +_,7 @@
         this.settings = settings;
         this.specialWorldProperty = specialWorldProperty;
         this.worldGenSettingsLifecycle = worldGenSettingsLifecycle;
+        this.canvas$distanceConfig = canvas$distanceConfig; // Canvas - per world distance
     }
 
     public PrimaryLevelData(final LevelSettings levelSettings, final PrimaryLevelData.SpecialWorldProperty specialWorldProperty, final Lifecycle lifecycle) {
@@ -84,6 +_,7 @@
             levelSettings.copy(),
             specialWorldProperty,
             lifecycle
+            , io.canvasmc.canvas.world.PerWorldDistanceConfig.DEFAULT // Canvas - per world distance
         );
     }
 
@@ -107,6 +_,7 @@
             settings,
             specialWorldProperty,
             worldGenSettingsLifecycle
+            , input.get("Canvas.Distance").read(io.canvasmc.canvas.world.PerWorldDistanceConfig.CODEC).resultOrPartial(LOGGER::error).orElse(io.canvasmc.canvas.world.PerWorldDistanceConfig.DEFAULT) // Canvas - per world distance
         );
         // Paper start
         data.respawnDimension = input.get(PAPER_RESPAWN_DIMENSION)
@@ -153,6 +_,7 @@
 
         tag.store(WorldDataConfiguration.MAP_CODEC, this.settings.dataConfiguration());
         tag.putString("Bukkit.Version", org.bukkit.Bukkit.getName() + "/" + org.bukkit.Bukkit.getVersion() + "/" + org.bukkit.Bukkit.getBukkitVersion()); // CraftBukkit
+        tag.store("Canvas.Distance", io.canvasmc.canvas.world.PerWorldDistanceConfig.CODEC, this.canvas$distanceConfig); // Canvas - per world distance
     }
 
     public static void writeLastPlayed(final CompoundTag tag) {
