From 9cc14a34f9eda110d805d11d2cf52e6b4b265a19 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/storage/PrimaryLevelData.java b/net/minecraft/world/level/storage/PrimaryLevelData.java
index b62cbb3a20f7bea0a36e744e2978a9f3a6c9d442..1930e14c548918df44138beba12a455d3cd30754 100644
--- a/net/minecraft/world/level/storage/PrimaryLevelData.java
+++ b/net/minecraft/world/level/storage/PrimaryLevelData.java
@@ -44,6 +44,7 @@ public class PrimaryLevelData implements ServerLevelData, WorldData {
     private final Set<String> knownServerBrands;
     private boolean wasModded;
     private final Set<String> removedFeatureFlags;
+    public final io.canvasmc.canvas.world.PerWorldDistanceConfig canvas$distanceConfig; // Canvas - per world distance
 
     private PrimaryLevelData(
         final @Nullable UUID singlePlayerUUID,
@@ -57,6 +58,7 @@ public class PrimaryLevelData implements ServerLevelData, WorldData {
         final LevelSettings settings,
         final PrimaryLevelData.SpecialWorldProperty specialWorldProperty,
         final Lifecycle worldGenSettingsLifecycle
+        , io.canvasmc.canvas.world.PerWorldDistanceConfig canvas$distanceConfig // Canvas - per world distance
     ) {
         this.wasModded = wasModded;
         this.respawnData = respawnData;
@@ -69,6 +71,7 @@ public class PrimaryLevelData implements ServerLevelData, WorldData {
         this.settings = settings;
         this.specialWorldProperty = specialWorldProperty;
         this.worldGenSettingsLifecycle = worldGenSettingsLifecycle;
+        this.canvas$distanceConfig = canvas$distanceConfig; // Canvas - per world distance
     }
 
     public PrimaryLevelData(final LevelSettings levelSettings, final PrimaryLevelData.SpecialWorldProperty specialWorldProperty, final Lifecycle lifecycle) {
@@ -84,6 +87,7 @@ public class PrimaryLevelData implements ServerLevelData, WorldData {
             levelSettings.copy(),
             specialWorldProperty,
             lifecycle
+            , io.canvasmc.canvas.world.PerWorldDistanceConfig.DEFAULT // Canvas - per world distance
         );
     }
 
@@ -107,6 +111,7 @@ public class PrimaryLevelData implements ServerLevelData, WorldData {
             settings,
             specialWorldProperty,
             worldGenSettingsLifecycle
+            , input.get("Canvas.Distance").read(io.canvasmc.canvas.world.PerWorldDistanceConfig.CODEC).resultOrPartial(LOGGER::error).orElse(io.canvasmc.canvas.world.PerWorldDistanceConfig.DEFAULT) // Canvas - per world distance
         );
         // Paper start
         data.respawnDimension = input.get(PAPER_RESPAWN_DIMENSION)
@@ -153,6 +158,7 @@ public class PrimaryLevelData implements ServerLevelData, WorldData {
 
         tag.store(WorldDataConfiguration.MAP_CODEC, this.settings.dataConfiguration());
         tag.putString("Bukkit.Version", org.bukkit.Bukkit.getName() + "/" + org.bukkit.Bukkit.getVersion() + "/" + org.bukkit.Bukkit.getBukkitVersion()); // CraftBukkit
+        tag.store("Canvas.Distance", io.canvasmc.canvas.world.PerWorldDistanceConfig.CODEC, this.canvas$distanceConfig); // Canvas - per world distance
     }
 
     public static void writeLastPlayed(final CompoundTag tag) {
