--- a/net/minecraft/world/level/storage/PrimaryLevelData.java
+++ b/net/minecraft/world/level/storage/PrimaryLevelData.java
@@ -73,6 +_,7 @@
     private boolean wasModded;
     private final Set<String> removedFeatureFlags;
     private final TimerQueue<MinecraftServer> scheduledEvents;
+    public final io.canvasmc.canvas.chunk.PerWorldDistanceConfig canvas$distanceConfig; // Canvas - per world distance
 
     // CraftBukkit start - Add world and pdc
     public net.minecraft.core.Registry<net.minecraft.world.level.dimension.LevelStem> customDimensions;
@@ -117,6 +_,7 @@
         WorldOptions worldOptions,
         PrimaryLevelData.SpecialWorldProperty specialWorldProperty,
         Lifecycle worldGenSettingsLifecycle
+        , io.canvasmc.canvas.chunk.PerWorldDistanceConfig canvas$distanceConfig // Canvas - per world distance
     ) {
         this.wasModded = wasModded;
         this.spawnPos = spawnPos;
@@ -145,6 +_,7 @@
         this.worldOptions = worldOptions;
         this.specialWorldProperty = specialWorldProperty;
         this.worldGenSettingsLifecycle = worldGenSettingsLifecycle;
+        this.canvas$distanceConfig = canvas$distanceConfig; // Canvas - per world distance
     }
 
     public PrimaryLevelData(
@@ -178,6 +_,7 @@
             worldOptions,
             specialWorldProperty,
             worldGenSettingsLifecycle
+            , io.canvasmc.canvas.chunk.PerWorldDistanceConfig.DEFAULT // Canvas - per world distance
         );
     }
 
@@ -223,6 +_,7 @@
             worldOptions,
             specialWorldProperty,
             worldGenSettingsLifecycle
+            , tag.get("Canvas.Distance").read(io.canvasmc.canvas.chunk.PerWorldDistanceConfig.CODEC).resultOrPartial(LOGGER::error).orElse(io.canvasmc.canvas.chunk.PerWorldDistanceConfig.DEFAULT) // Canvas - per world distance
         );
     }
 
@@ -293,6 +_,7 @@
         nbt.storeNullable("WanderingTraderId", UUIDUtil.CODEC, this.wanderingTraderId);
         nbt.putString("Bukkit.Version", org.bukkit.Bukkit.getName() + "/" + org.bukkit.Bukkit.getVersion() + "/" + org.bukkit.Bukkit.getBukkitVersion()); // CraftBukkit
         this.world.getWorld().storeBukkitValues(nbt); // CraftBukkit - add pdc
+        nbt.store("Canvas.Distance", io.canvasmc.canvas.chunk.PerWorldDistanceConfig.CODEC, this.canvas$distanceConfig); // Canvas - per world distance
     }
 
     private static ListTag stringCollectionToTag(Set<String> stringCollection) {
