--- a/paper-server/build.gradle.kts
+++ b/paper-server/build.gradle.kts
@@ -2,12 +_,13 @@
 import io.papermc.paperweight.attribute.DevBundleOutput
 import io.papermc.paperweight.util.*
 import java.time.Instant
+import io.papermc.paperweight.tasks.MergeAccessTransforms
 
 plugins {
     `java-library`
     `maven-publish`
     idea
-    id("io.papermc.paperweight.core")
+    id("io.canvasmc.weaver.core")
     id("io.papermc.fill.gradle") version "1.0.12"
 }
 
@@ -18,12 +_,31 @@
     paperclip("io.papermc:paperclip:3.0.4")
 }
 
+val mergeMinecraftATs by tasks.registering<MergeAccessTransforms> {
+    firstFile.set(paperweight.activeFork.flatMap { it.buildDataDir.file("${it.name}.at") })
+    secondFile.set(paperweight.activeFork.flatMap { it.buildDataDir.file("folia.at") })
+}
+
 paperweight {
     minecraftVersion = providers.gradleProperty("mcVersion")
+    filterPatches = false
     gitFilePatches = false
 
+    val canvas = forks.register("canvas") {
+        additionalAts.set(mergeMinecraftATs.flatMap { it.outputFile })
+        upstream.patchDir("paperServer") {
+            upstreamPath = "paper-server"
+            excludes = setOf("src/minecraft", "patches", "build.gradle.kts")
+            patchesDir = rootDirectory.dir("canvas-server/paper-patches")
+            outputDir = rootDirectory.dir("paper-server")
+        }
+    }
+
+    activeFork = canvas
+
     updatingMinecraft {
         // oldPaperCommit = "d4fe85375af18bfa88f44d7c1e6a61904ae550cc"
+        // oldForkCommit = "58d6715718521ea774b786fa9ca1e6fc2db14702"
     }
 }
 
@@ -86,7 +_,19 @@
     }
 }
 
-val log4jPlugins = sourceSets.create("log4jPlugins")
+sourceSets {
+    main {
+        java { srcDir("../paper-server/src/main/java"); srcDir("../paper-server/src/generated/java") }
+        resources { srcDir("../paper-server/src/main/resources") }
+    }
+    test {
+        java { srcDir("../paper-server/src/test/java") }
+        resources { srcDir("../paper-server/src/test/resources") }
+    }
+}
+val log4jPlugins = sourceSets.create("log4jPlugins") {
+    java { srcDir("../paper-server/src/log4jPlugins/java") }
+}
 configurations.named(log4jPlugins.compileClasspathConfigurationName) {
     extendsFrom(configurations.compileClasspath.get())
 }
@@ -109,12 +_,14 @@
 }
 
 dependencies {
-    implementation(project(":paper-api"))
+    implementation(projects.canvasApi)
     implementation("ca.spottedleaf:leafpile:1.0.0")
-    implementation("org.jline:jline-terminal-ffm:3.27.1") // use ffm on java 22+
-    implementation("org.jline:jline-terminal-jni:3.27.1") // fall back to jni on java 21
+    implementation("org.jline:jline-terminal-ffm:3.29.0") // use ffm on java 22+
+    implementation("org.jline:jline-terminal-jni:3.29.0") // fall back to jni on java 21
     implementation("net.minecrell:terminalconsoleappender:1.3.0")
     implementation("net.kyori:adventure-text-serializer-ansi")
+    implementation("net.openhft:affinity:2026.2")
+    implementation("io.canvasmc.httpclient:httpclient:1.0.1")
 
     /*
       Required to add the missing Log4j2Plugins.dat file from log4j-core
@@ -138,7 +_,10 @@
     // Deps that were previously in the API but have now been moved here for backwards compat, eventually to be removed
     runtimeOnly("commons-lang:commons-lang:2.6")
     runtimeOnly("org.xerial:sqlite-jdbc:3.49.1.0")
-    runtimeOnly("com.mysql:mysql-connector-j:9.2.0")
+    runtimeOnly("com.mysql:mysql-connector-j:9.2.0") {
+        exclude("com.google.protobuf", "protobuf-java")
+    }
+    runtimeOnly("com.google.protobuf:protobuf-java:4.30.1")
     runtimeOnly("com.lmax:disruptor:3.4.4")
     implementation("com.googlecode.json-simple:json-simple:1.1.1") { // change to runtimeOnly once Timings is removed
         isTransitive = false // includes junit
@@ -148,7 +_,7 @@
     testRuntimeOnly("org.junit.platform:junit-platform-launcher:6.0.3")
     testImplementation("org.junit.jupiter:junit-jupiter:6.0.3")
     testImplementation("org.junit.platform:junit-platform-suite-engine:6.0.3")
-    testImplementation("org.hamcrest:hamcrest:2.2")
+    testImplementation("org.hamcrest:hamcrest:3.0")
     testImplementation("org.mockito:mockito-core:5.22.0")
     mockitoAgent("org.mockito:mockito-core:5.22.0") { isTransitive = false } // Configure mockito agent that is needed in newer java versions
     testImplementation("org.ow2.asm:asm-tree:9.9.1")
@@ -169,20 +_,30 @@
         val implementationVersion = "$mcVersion-${build ?: "DEV"}-$gitHash"
         val date = git.exec(providers, "show", "-s", "--format=%ci", gitHash).get().trim()
         val gitBranch = git.exec(providers, "rev-parse", "--abbrev-ref", "HEAD").get().trim()
+        val canvasContributors = listOf(
+            "VeguiDev",
+            "BaconCat1",
+            "Euphillya",
+            "Jsinco",
+            "Biquaternions"
+        )
         attributes(
             "Main-Class" to "org.bukkit.craftbukkit.Main",
-            "Implementation-Title" to "Paper",
+            "Implementation-Title" to "Canvas",
             "Implementation-Version" to implementationVersion,
             "Implementation-Vendor" to date,
-            "Specification-Title" to "Paper",
+            "Specification-Title" to "Canvas",
             "Specification-Version" to project.version,
-            "Specification-Vendor" to "Paper Team",
-            "Brand-Id" to "papermc:paper",
-            "Brand-Name" to "Paper",
+            "Specification-Vendor" to "Canvas Team",
+            "Brand-Id" to "canvasmc:canvas",
+            "Brand-Name" to "Canvas",
+            "Brand-Vendor" to "Canvas Team",
+            "Brand-Website" to "https://canvasmc.io",
             "Build-Number" to (build ?: ""),
             "Build-Time" to buildTime.toString(),
             "Git-Branch" to gitBranch,
             "Git-Commit" to gitHash,
+            "Contributors" to canvasContributors,
         )
         for (tld in setOf("net", "com", "org")) {
             attributes("$tld/bukkit", "Sealed" to true)
@@ -240,6 +_,10 @@
         excludeTags("Slow")
     }
 
+    if (paperweight.updatingMinecraft.oldForkCommit.isPresent) {
+        environment("CANVAS_TEST_STRICT", "true")
+    }
+
     // Configure mockito agent that is needed in newer java versions
     val provider = objects.newInstance<MockitoAgentProvider>()
     provider.fileCollection.from(mockitoAgent)
@@ -273,17 +_,18 @@
     javaLauncher.set(project.javaToolchains.launcherFor {
         languageVersion.set(JavaLanguageVersion.of(25))
         // TODO - JB runtime 25 has issues with spark rn
-        // vendor.set(JvmVendorSpec.JETBRAINS)
+        vendor.set(JvmVendorSpec.GRAAL_VM)
     })
-    //jvmArgs("-XX:+AllowEnhancedClassRedefinition")
+    jvmArgs("--add-modules=jdk.incubator.vector")
 
-    if (rootProject.childProjects["test-plugin"] != null) {
-        val testPluginJar = rootProject.project(":test-plugin").tasks.jar.flatMap { it.archiveFile }
-        inputs.file(testPluginJar)
-        args("-add-plugin=${testPluginJar.get().asFile.absolutePath}")
+    for (entry in rootProject.childProjects) {
+        if (entry.key == "test-plugin" || entry.key.endsWith("-debug", true) || entry.key.endsWith("-plugin", true)) {
+            val testPluginJar = rootProject.project(":${entry.key}").tasks.jar.flatMap { it.archiveFile }
+            inputs.file(testPluginJar)
+            args("-add-plugin=${testPluginJar.get().asFile.absolutePath}")
+        }
     }
 
-    args("--nogui")
     systemProperty("net.kyori.adventure.text.warnWhenLegacyFormattingDetected", true)
     if (providers.gradleProperty("paper.runDisableWatchdog").getOrElse("false") == "true") {
         systemProperty("disable.watchdog", true)
@@ -293,8 +_,8 @@
     systemProperty("paper.disableMigrationDelay", true)
     systemProperty("paper.updatingMinecraft", providers.gradleProperty("updatingMinecraft").getOrElse("false").toBoolean())
 
-    val memoryGb = providers.gradleProperty("paper.runMemoryGb").getOrElse("2")
-    minHeapSize = "${memoryGb}G"
+    val memoryGb = providers.gradleProperty("paper.runMemoryGb").getOrElse("7")
+    // minHeapSize = "${memoryGb}G"
     maxHeapSize = "${memoryGb}G"
 
     doFirst {
@@ -326,7 +_,7 @@
     mainClass.set(null as String?)
 }
 
-fill {
+/* fill {
     project("paper")
     versionFamily(paperweight.minecraftVersion.map { it.split(".", "-").takeWhile { part -> part.toIntOrNull() != null }.take(2).joinToString(".") })
     version(paperweight.minecraftVersion)
@@ -340,5 +_,38 @@
                 nameResolver.set { project, _, version, build -> "$project-$version-$build.jar" }
             }
         }
+    }
+} */
+
+val sortFoliaATs = tasks.register<SortATs>("sortFoliaATs") {
+    atFile.set(mergeMinecraftATs.flatMap { it.secondFile })
+    atFileOut.set(mergeMinecraftATs.flatMap { it.secondFile })
+}
+
+afterEvaluate {
+    tasks.named<io.papermc.paperweight.core.tasks.patching.RebuildFilePatches>("rebuildMinecraftSourcePatches") {
+        atFile.set(mergeMinecraftATs.flatMap { it.firstFile })
+        atFileOut.set(mergeMinecraftATs.flatMap { it.firstFile })
+        dependsOn(sortFoliaATs)
+    }
+}
+
+abstract class SortATs : DefaultTask() {
+
+    @get:InputFile
+    abstract val atFile: RegularFileProperty
+
+    @get:OutputFile
+    abstract val atFileOut: RegularFileProperty
+
+    @TaskAction
+    fun run() {
+        val ats = paper.libs.org.cadixdev.at.io.AccessTransformFormats.FML.read(atFile.path)
+
+        paper.libs.org.cadixdev.at.io.AccessTransformFormats.FML.writeLF(
+            atFileOut.path,
+            ats,
+            "# This file is auto generated, any changes may be overridden!\n# See CONTRIBUTING.md on how to add access transformers."
+        )
     }
 }
