--- a/folia-server/build.gradle.kts
+++ b/folia-server/build.gradle.kts
@@ -1,6 +_,7 @@
 import io.papermc.fill.model.BuildChannel
 import io.papermc.paperweight.attribute.DevBundleOutput
 import io.papermc.paperweight.util.*
+import io.papermc.paperweight.util.constants.MACHE_PARAM_MAPPINGS_CONFIG
 import io.papermc.paperweight.util.data.FileEntry
 import paper.libs.com.google.gson.annotations.SerializedName
 import java.time.Instant
@@ -10,7 +_,7 @@
     `java-library`
     `maven-publish`
     idea
-    id("io.papermc.paperweight.core")
+    id("io.canvasmc.weaver.core")
     id("io.papermc.fill.gradle") version "1.0.7"
 }
 
@@ -35,7 +_,26 @@
         }
     }
 
-    activeFork = fork
+    val canvas = forks.register("canvas") {
+        forks = fork
+        javadocMappings(configurations.named(MACHE_PARAM_MAPPINGS_CONFIG))
+        upstream.patchRepo("paperServer") {
+            upstreamRepo = fork.patchedRepo("paperServer")
+            patchesDir = rootDirectory.dir("canvas-server/paper-patches")
+            additionalAts = rootDirectory.file("build-data/canvas.at")
+            outputDir = rootDirectory.dir("paper-server")
+        }
+
+        upstream.patchDir("foliaServer") {
+            upstreamPath = "folia-server"
+            excludes = setOf("src/minecraft", "paper-patches", "minecraft-patches", "build.gradle.kts", "build.gradle.kts.patch")
+            patchesDir = rootDirectory.dir("canvas-server/folia-patches")
+            additionalAts = rootDirectory.file("build-data/canvas.at")
+            outputDir = rootDirectory.dir("folia-server")
+        }
+    }
+
+    activeFork = canvas
 
 
     spigot {
@@ -73,7 +_,7 @@
 }
 val services = objects.newInstance<Services>()
 
-if (project.providers.gradleProperty("publishDevBundle").isPresent) {
+if (true) {
     val devBundleComponent = services.softwareComponentFactory.adhoc("devBundle")
     components.add(devBundleComponent)
 
@@ -123,10 +_,14 @@
     main {
         java { srcDir("../paper-server/src/main/java"); srcDir("../paper-server/src/generated/java") }
         resources { srcDir("../paper-server/src/main/resources") }
+        java { srcDir("../folia-server/src/main/java") }
+        resources { srcDir("../folia-server/src/main/resources") }
     }
     test {
         java { srcDir("../paper-server/src/test/java") }
         resources { srcDir("../paper-server/src/test/resources") }
+        java { srcDir("../folia-server/src/test/java") }
+        resources { srcDir("../folia-server/src/test/resources") }
     }
 }
 val log4jPlugins = sourceSets.create("log4jPlugins") {
@@ -153,10 +_,10 @@
 }
 
 dependencies {
-    implementation(project(":folia-api"))
+    implementation(projects.canvasApi)
     implementation("ca.spottedleaf:concurrentutil:0.0.3")
-    implementation("org.jline:jline-terminal-ffm:3.27.1") // use ffm on java 22+
-    implementation("org.jline:jline-terminal-jni:3.27.1") // fall back to jni on java 21
+    implementation("org.jline:jline-terminal-ffm:3.29.0") // use ffm on java 22+
+    implementation("org.jline:jline-terminal-jni:3.29.0") // fall back to jni on java 21
     implementation("net.minecrell:terminalconsoleappender:1.3.0")
     implementation("net.kyori:adventure-text-serializer-ansi")
     runtimeConfiguration(sourceSets.main.map { it.runtimeClasspath })
@@ -167,23 +_,27 @@
       all its classes to check if they are plugins.
       Scanning takes about 1-2 seconds so adding this speeds up the server start.
      */
-    implementation("org.apache.logging.log4j:log4j-core:2.24.1")
-    log4jPlugins.annotationProcessorConfigurationName("org.apache.logging.log4j:log4j-core:2.24.1") // Needed to generate meta for our Log4j plugins
+    implementation("org.apache.logging.log4j:log4j-core:2.24.3")
+    log4jPlugins.annotationProcessorConfigurationName("org.apache.logging.log4j:log4j-core:2.24.3") // Needed to generate meta for our Log4j plugins
     runtimeOnly(log4jPlugins.output)
     alsoShade(log4jPlugins.output)
 
     implementation("com.velocitypowered:velocity-native:3.4.0-SNAPSHOT") {
         isTransitive = false
     }
-    implementation("io.netty:netty-codec-haproxy:4.1.118.Final") // Add support for proxy protocol
-    implementation("org.apache.logging.log4j:log4j-iostreams:2.24.1")
+    implementation("io.netty:netty-codec-haproxy:4.2.7.Final") // Add support for proxy protocol
+    implementation("io.netty:netty-all:4.2.7.Final")
+    implementation("org.apache.logging.log4j:log4j-iostreams:2.24.3")
     implementation("org.ow2.asm:asm-commons:9.8")
     implementation("org.spongepowered:configurate-yaml:4.2.0")
 
     // Deps that were previously in the API but have now been moved here for backwards compat, eventually to be removed
     runtimeOnly("commons-lang:commons-lang:2.6")
     runtimeOnly("org.xerial:sqlite-jdbc:3.49.1.0")
-    runtimeOnly("com.mysql:mysql-connector-j:9.2.0")
+    runtimeOnly("com.mysql:mysql-connector-j:9.2.0") {
+        exclude("com.google.protobuf", "protobuf-java")
+    }
+    runtimeOnly("com.google.protobuf:protobuf-java:4.30.1")
     runtimeOnly("com.lmax:disruptor:3.4.4")
     implementation("com.googlecode.json-simple:json-simple:1.1.1") { // change to runtimeOnly once Timings is removed
         isTransitive = false // includes junit
@@ -192,11 +_,11 @@
     testImplementation("io.github.classgraph:classgraph:4.8.179") // For mob goal test
     testImplementation("org.junit.jupiter:junit-jupiter:5.12.2")
     testImplementation("org.junit.platform:junit-platform-suite-engine:1.12.2")
-    testImplementation("org.hamcrest:hamcrest:2.2")
-    testImplementation("org.mockito:mockito-core:5.14.1")
-    mockitoAgent("org.mockito:mockito-core:5.14.1") { isTransitive = false } // Configure mockito agent that is needed in newer java versions
+    testImplementation("org.hamcrest:hamcrest:3.0")
+    testImplementation("org.mockito:mockito-core:5.16.1")
+    mockitoAgent("org.mockito:mockito-core:5.16.1") { isTransitive = false } // Configure mockito agent that is needed in newer java versions
     testImplementation("org.ow2.asm:asm-tree:9.8")
-    testImplementation("org.junit-pioneer:junit-pioneer:2.2.0") // CartesianTest
+    testImplementation("org.junit-pioneer:junit-pioneer:2.3.0") // CartesianTest
 
     implementation("net.neoforged:srgutils:1.0.9") // Mappings handling
     implementation("net.neoforged:AutoRenamingTool:2.0.3") // Remap plugins
@@ -224,14 +_,14 @@
         val gitBranch = git.exec(providers, "rev-parse", "--abbrev-ref", "HEAD").get().trim()
         attributes(
             "Main-Class" to "org.bukkit.craftbukkit.Main",
-            "Implementation-Title" to "Folia",
+            "Implementation-Title" to "Canvas",
             "Implementation-Version" to implementationVersion,
             "Implementation-Vendor" to date,
-            "Specification-Title" to "Folia",
+            "Specification-Title" to "Canvas",
             "Specification-Version" to project.version,
-            "Specification-Vendor" to "Paper Team",
-            "Brand-Id" to "papermc:folia",
-            "Brand-Name" to "Folia",
+            "Specification-Vendor" to "Canvas Team",
+            "Brand-Id" to "canvasmc:canvas",
+            "Brand-Name" to "Canvas",
             "Build-Number" to (build ?: ""),
             "Build-Time" to buildTime.toString(),
             "Git-Branch" to gitBranch,
@@ -328,25 +_,24 @@
         .asFile
     javaLauncher.set(project.javaToolchains.launcherFor {
         languageVersion.set(JavaLanguageVersion.of(21))
-        vendor.set(JvmVendorSpec.JETBRAINS)
+        vendor.set(JvmVendorSpec.GRAAL_VM)
     })
-    jvmArgs("-XX:+AllowEnhancedClassRedefinition")
+    jvmArgs("--add-modules=jdk.incubator.vector")
 
-    if (rootProject.childProjects["test-plugin"] != null) {
-        val testPluginJar = rootProject.project(":test-plugin").tasks.jar.flatMap { it.archiveFile }
+    if (rootProject.childProjects["canvas-test-plugin"] != null) {
+        val testPluginJar = rootProject.project(":canvas-test-plugin").tasks.jar.flatMap { it.archiveFile }
         inputs.file(testPluginJar)
         args("-add-plugin=${testPluginJar.get().asFile.absolutePath}")
     }
 
-    args("--nogui")
     systemProperty("net.kyori.adventure.text.warnWhenLegacyFormattingDetected", true)
     if (providers.gradleProperty("paper.runDisableWatchdog").getOrElse("false") == "true") {
         systemProperty("disable.watchdog", true)
     }
     systemProperty("io.papermc.paper.suppress.sout.nags", true)
 
-    val memoryGb = providers.gradleProperty("paper.runMemoryGb").getOrElse("2")
-    minHeapSize = "${memoryGb}G"
+    val memoryGb = providers.gradleProperty("paper.runMemoryGb").getOrElse("9") // increased memory
+    // minHeapSize = "${memoryGb}G"
     maxHeapSize = "${memoryGb}G"
 
     doFirst {
