From 9cc14a34f9eda110d805d11d2cf52e6b4b265a19 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/schedule/Activity.java b/net/minecraft/world/entity/schedule/Activity.java
index 23a614ea83abf7abdbf4f5b34751d08962c46cee..1a03d13be4f08ad4737a6e7f35c1bcda729b3363 100644
--- a/net/minecraft/world/entity/schedule/Activity.java
+++ b/net/minecraft/world/entity/schedule/Activity.java
@@ -32,10 +32,12 @@ public class Activity {
     public static final Activity DIG = register("dig");
     private final String name;
     private final int hashCode;
+    public final int id; // Canvas - optimize brain
 
-    private Activity(final String name) {
+    private Activity(final String name, int id) { // Canvas - optimize brain
         this.name = name;
         this.hashCode = name.hashCode();
+        this.id = id; // Canvas - optimize brain
     }
 
     public String getName() {
@@ -43,7 +45,7 @@ public class Activity {
     }
 
     private static Activity register(final String name) {
-        return Registry.register(BuiltInRegistries.ACTIVITY, name, new Activity(name));
+        return Registry.register(BuiltInRegistries.ACTIVITY, name, new Activity(name, BuiltInRegistries.ACTIVITY.size())); // Canvas - optimize brain
     }
 
     @Override
