From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/levelgen/NoiseChunk.java b/net/minecraft/world/level/levelgen/NoiseChunk.java
index b7f4943b6c862f4b99aa7920af3fc91fc9fd0b76..e8ec64b49241b1f9c5c3f1d535f1556f933aba7c 100644
--- a/net/minecraft/world/level/levelgen/NoiseChunk.java
+++ b/net/minecraft/world/level/levelgen/NoiseChunk.java
@@ -371,7 +371,17 @@ public class NoiseChunk implements DensityFunction.ContextProvider, DensityFunct
     }
 
     protected DensityFunction wrap(DensityFunction densityFunction) {
-        return this.wrapped.computeIfAbsent(densityFunction, this::wrapNew);
+        // Canvas start - reduce worldgen allocations
+        // avoid lambda allocation
+        DensityFunction func = this.wrapped.get(densityFunction);
+
+        if (func == null) {
+            func = this.wrapNew(densityFunction);
+            this.wrapped.put(densityFunction, func);
+        }
+
+        return func;
+        // Canvas end - reduce worldgen allocations
     }
 
     private DensityFunction wrapNew(DensityFunction densityFunction) {
