From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/levelgen/SurfaceRules.java b/net/minecraft/world/level/levelgen/SurfaceRules.java
index c19a6843ba53d5c1850b45b82a967b9dd397e6b8..8d90f15505997db0aece8b546e8d4f8aeaa3d6d6 100644
--- a/net/minecraft/world/level/levelgen/SurfaceRules.java
+++ b/net/minecraft/world/level/levelgen/SurfaceRules.java
@@ -397,6 +397,22 @@ public class SurfaceRules {
 
             @Override
             protected boolean compute() {
+                // Canvas start - MC-258859
+                if (io.canvasmc.canvas.Config.INSTANCE.fixes.mc258859) {
+                    int x = this.context.blockX & 15;
+                    int z = this.context.blockZ & 15;
+                    ChunkAccess chunk = this.context.chunk;
+                    int south = chunk.getHeight(Heightmap.Types.WORLD_SURFACE_WG, x, Math.max(z - 1, 0));
+                    int north = chunk.getHeight(Heightmap.Types.WORLD_SURFACE_WG, x, Math.min(z + 1, 15));
+                    if (Math.abs(north - south) >= 4) {
+                        return true;
+                    } else {
+                        int west = chunk.getHeight(Heightmap.Types.WORLD_SURFACE_WG, Math.max(x - 1, 0), z);
+                        int east = chunk.getHeight(Heightmap.Types.WORLD_SURFACE_WG, Math.min(x + 1, 15), z);
+                        return Math.abs(west - east) >= 4;
+                    }
+                }
+                // Canvas end - MC-258859
                 int i = this.context.blockX & 15;
                 int i1 = this.context.blockZ & 15;
                 int max = Math.max(i1 - 1, 0);
@@ -580,8 +596,13 @@ public class SurfaceRules {
     record SequenceRule(List<SurfaceRules.SurfaceRule> rules) implements SurfaceRules.SurfaceRule {
         @Override
         public @Nullable BlockState tryApply(int x, int y, int z) {
-            for (SurfaceRules.SurfaceRule surfaceRule : this.rules) {
-                BlockState blockState = surfaceRule.tryApply(x, y, z);
+            // Canvas start - reduce worldgen allocations
+            // avoid iterator allocation
+            int size = this.rules.size();
+            //noinspection ForLoopReplaceableByForEach
+            for (int i = 0; i < size; i++) {
+                BlockState blockState = this.rules.get(i).tryApply(x, y, z);
+            // Canvas end - reduce worldgen allocations
                 if (blockState != null) {
                     return blockState;
                 }
