--- a/src/main/java/ca/spottedleaf/moonrise/common/util/MoonriseConstants.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/util/MoonriseConstants.java
@@ -1,10 +_,11 @@
 package ca.spottedleaf.moonrise.common.util;
 
 import ca.spottedleaf.moonrise.common.PlatformHooks;
+import io.canvasmc.canvas.Config;
 
 public final class MoonriseConstants {
 
-    public static final int MAX_VIEW_DISTANCE = Integer.getInteger(PlatformHooks.get().getBrand() + ".MaxViewDistance", 32);
+    public static final int MAX_VIEW_DISTANCE = Integer.getInteger(PlatformHooks.get().getBrand() + ".MaxViewDistance", Config.INSTANCE.chunkGeneration.maxViewDistance);
 
     private MoonriseConstants() {}
 
