--- a/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java
@@ -97,6 +_,13 @@
 
     private static final AtomicInteger ID_GENERATOR = new AtomicInteger();
 
+    // Canvas start
+    public TickThread(final Runnable runnable) {
+        super(runnable);
+        id = ID_GENERATOR.incrementAndGet();
+    }
+
+    // Canvas end
     public TickThread(final String name) {
         this(null, name);
     }
