From c9d43594f1dc856ec561804d690b933859e1d7f9 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas paperServer File Patches


diff --git a/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java b/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java
index 6c74787321748702022787b65ff5465f8bcec4ad..43f28f75c877fed9d83c098804cbc54a267f9e24 100644
--- a/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java
@@ -243,10 +243,21 @@ public class TickThread extends Thread {
         if (entity == null) {
             return true;
         }
+        // Canvas start - optimize thread check
+        final Thread thread = Thread.currentThread();
+        // not on tick thread, can't possibly own the entity
+        if (!(thread instanceof TickThread tickThread)) return false;
+        // not on tick runner, meaning not part of the region scheduler, nope
+        if (!(tickThread instanceof TickRegionScheduler.TickThreadRunner threadRunner)) {
+            // if is shutdown thread, it owns the entity
+            // note, shutdown thread is instanceof TickThread, so this works
+            return tickThread instanceof RegionShutdownThread;
+        }
+        // Canvas end - optimize thread check
         final ThreadedRegionizer.ThreadedRegion<TickRegions.TickRegionData, TickRegions.TickRegionSectionData> region =
-            TickRegionScheduler.getCurrentRegion();
+            threadRunner.currentTickingRegion;// Canvas - optimize thread check
         if (region == null) {
-            if (RegionizedServer.isGlobalTickThread()) {
+            if (RegionizedServer.getGlobalTickData() == threadRunner.currentTickingTask) {// Canvas - optimize thread check
                 if (entity instanceof net.minecraft.server.level.ServerPlayer serverPlayer) {
                     final net.minecraft.server.network.ServerGamePacketListenerImpl possibleBad = serverPlayer.connection;
                     if (possibleBad == null) {
@@ -265,9 +276,7 @@ public class TickThread extends Thread {
                     return false;
                 }
             }
-            if (isShutdownThread()) {
-                return true;
-            }
+            // Canvas - optimize thread check
             if (entity instanceof net.minecraft.server.level.ServerPlayer serverPlayer) {
                 // off-main access to server player is never ok, server player is owned by one of global context or region context always
                 return false;
@@ -286,7 +295,7 @@ public class TickThread extends Thread {
             return false;
         }
 
-        final RegionizedWorldData worldData = io.papermc.paper.threadedregions.TickRegionScheduler.getCurrentRegionizedWorldData();
+        final RegionizedWorldData worldData = threadRunner.currentTickingWorldRegionizedData; // Canvas - optimize thread check
 
         // pass through the check if the entity is removed and we own its chunk
         if (worldData.hasEntity(entity)) {
