--- a/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java
@@ -133,7 +_,7 @@
         this(group, run, name, ID_GENERATOR.incrementAndGet());
     }
 
-    private TickThread(final ThreadGroup group, final Runnable run, final String name, final int id) {
+    public TickThread(final ThreadGroup group, final Runnable run, final String name, final int id) { // Canvas - TODO - at
         super(group, run, name);
         this.id = id;
     }
