--- a/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java
@@ -171,10 +_,19 @@
     }
 
     public static boolean isTickThreadFor(final Level world, final int chunkX, final int chunkZ) {
-        final ThreadedRegionizer.ThreadedRegion<TickRegions.TickRegionData, TickRegions.TickRegionSectionData> region =
-            TickRegionScheduler.getCurrentRegion();
+        // Canvas start - optimize thread check
+        final Thread thread = Thread.currentThread();
+        final ThreadedRegionizer.ThreadedRegion<TickRegions.TickRegionData, TickRegions.TickRegionSectionData> region;
+        if (!(thread instanceof io.papermc.paper.threadedregions.TickRegionScheduler.TickThreadRunner threadRunner)) {
+            if (thread instanceof RegionShutdownThread shutdownThread) {
+                region = shutdownThread.shuttingDown;
+            }
+            else region = null;
+        }
+        else region = threadRunner.currentTickingRegion;
+        // Canvas end - optimize thread check
         if (region == null) {
-            return isShutdownThread();
+            return thread.getClass() == RegionShutdownThread.class; // Canvas - optimize thread check
         }
         return ((net.minecraft.server.level.ServerLevel)world).regioniser.getRegionAtUnsynchronised(chunkX, chunkZ) == region;
     }
@@ -209,10 +_,19 @@
     }
 
     public static boolean isTickThreadFor(final Level world, final int fromChunkX, final int fromChunkZ, final int toChunkX, final int toChunkZ) {
-        final ThreadedRegionizer.ThreadedRegion<TickRegions.TickRegionData, TickRegions.TickRegionSectionData> region =
-            TickRegionScheduler.getCurrentRegion();
+        // Canvas start - optimize thread check
+        final Thread thread = Thread.currentThread();
+        final ThreadedRegionizer.ThreadedRegion<TickRegions.TickRegionData, TickRegions.TickRegionSectionData> region;
+        if (!(thread instanceof io.papermc.paper.threadedregions.TickRegionScheduler.TickThreadRunner threadRunner)) {
+            if (thread instanceof RegionShutdownThread shutdownThread) {
+                region = shutdownThread.shuttingDown;
+            }
+            else region = null;
+        }
+        else region = threadRunner.currentTickingRegion;
+        // Canvas end - optimize thread check
         if (region == null) {
-            return isShutdownThread();
+            return thread.getClass() == RegionShutdownThread.class; // Canvas - optimize thread check
         }
 
         final int shift = ((net.minecraft.server.level.ServerLevel)world).regioniser.sectionChunkShift;
@@ -243,10 +_,19 @@
         if (entity == null) {
             return true;
         }
-        final ThreadedRegionizer.ThreadedRegion<TickRegions.TickRegionData, TickRegions.TickRegionSectionData> region =
-            TickRegionScheduler.getCurrentRegion();
+        // Canvas start - optimize thread check
+        final Thread thread = Thread.currentThread();
+        final ThreadedRegionizer.ThreadedRegion<TickRegions.TickRegionData, TickRegions.TickRegionSectionData> region;
+        if (!(thread instanceof io.papermc.paper.threadedregions.TickRegionScheduler.TickThreadRunner threadRunner)) {
+            if (thread instanceof RegionShutdownThread shutdownThread) {
+                region = shutdownThread.shuttingDown;
+            }
+            else region = null;
+        }
+        else region = threadRunner.currentTickingRegion;
+        // Canvas end - optimize thread check
         if (region == null) {
-            if (RegionizedServer.isGlobalTickThread()) {
+            if (thread instanceof io.papermc.paper.threadedregions.TickRegionScheduler.TickThreadRunner threadRunner && RegionizedServer.getGlobalTickData() == threadRunner.currentTickingTask) { // Canvas - optimize thread check
                 if (entity instanceof net.minecraft.server.level.ServerPlayer serverPlayer) {
                     final net.minecraft.server.network.ServerGamePacketListenerImpl possibleBad = serverPlayer.connection;
                     if (possibleBad == null) {
@@ -265,7 +_,7 @@
                     return false;
                 }
             }
-            if (isShutdownThread()) {
+            if (thread.getClass() == RegionShutdownThread.class) { // Canvas - optimize thread check
                 return true;
             }
             if (entity instanceof net.minecraft.server.level.ServerPlayer serverPlayer) {
@@ -286,7 +_,7 @@
             return false;
         }
 
-        final RegionizedWorldData worldData = io.papermc.paper.threadedregions.TickRegionScheduler.getCurrentRegionizedWorldData();
+        final RegionizedWorldData worldData = region.getData().world.worldRegionData.get(); // Canvas - optimize thread check
 
         // pass through the check if the entity is removed and we own its chunk
         if (worldData.hasEntity(entity)) {
