From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/PotentialCalculator.java b/net/minecraft/world/level/PotentialCalculator.java
index c9ff08679ba711bb86477ad39526cf9a41f5f315..74728de5ba077bb7df22b766f55b371b6e60fff9 100644
--- a/net/minecraft/world/level/PotentialCalculator.java
+++ b/net/minecraft/world/level/PotentialCalculator.java
@@ -5,7 +5,7 @@ import java.util.List;
 import net.minecraft.core.BlockPos;
 
 public class PotentialCalculator {
-    private final List<PotentialCalculator.PointCharge> charges = Lists.newArrayList();
+    private final List<PotentialCalculator.PointCharge> charges = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(); // Canvas - optimized collection
 
     public void addCharge(BlockPos pos, double charge) {
         if (charge != 0.0) {
