--- a/net/minecraft/world/level/dimension/end/EnderDragonFight.java
+++ b/net/minecraft/world/level/dimension/end/EnderDragonFight.java
@@ -333,7 +_,7 @@
             for (int z = -8; z <= 8; z++) {
                 LevelChunk chunk = this.level.getChunk(x, z);
 
-                for (BlockEntity blockEntity : chunk.getBlockEntities().values()) {
+                for (BlockEntity blockEntity : chunk.canvas$getAllBlockEntities()) { // Canvas - optimize block entity fetching
                     if (blockEntity instanceof TheEndPortalBlockEntity) {
                         return true;
                     }
@@ -351,7 +_,7 @@
             for (int z = -8 + chunkOrigin.z(); z <= 8 + chunkOrigin.z(); z++) {
                 LevelChunk chunk = this.level.getChunk(x, z);
 
-                for (BlockEntity blockEntity : chunk.getBlockEntities().values()) {
+                for (BlockEntity blockEntity : chunk.canvas$getAllBlockEntities()) { // Canvas - optimize block entity fetching
                     if (blockEntity instanceof TheEndPortalBlockEntity) {
                         BlockPattern.BlockPatternMatch match = this.exitPortalPattern.find(this.level, blockEntity.getBlockPos());
                         if (match != null) {
