From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java b/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
index ca5448b2531d2f7126132723c4fa8d49835a4850..04854c3fa0b001f5edca5cecb1531034df981db5 100644
--- a/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
+++ b/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
@@ -476,13 +476,20 @@ public class WalkNodeEvaluator extends NodeEvaluator {
         return pathType;
     }
 
-    protected static PathType getPathTypeFromState(BlockGetter level, BlockPos pos) {
+    // Canvas start - optimize path type cache
+    private static final PathType[] PATH_TYPES = PathType.values();
+    public static PathType getPathTypeFromState(BlockGetter level, BlockPos pos) {
         // Paper start - Do not load chunks during pathfinding
         BlockState blockState = level.getBlockStateIfLoaded(pos);
         if (blockState == null) {
             return PathType.BLOCKED;
         }
         // Paper end
+        return PATH_TYPES[blockState.pathType];
+    }
+
+    public static PathType canvas$getPathTypeFromBlockState(BlockState blockState) {
+    // Canvas end - optimize path type cache
         Block block = blockState.getBlock();
         if (blockState.isAir()) {
             return PathType.OPEN;
