From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/behavior/TradeWithVillager.java b/net/minecraft/world/entity/ai/behavior/TradeWithVillager.java
index 256fb863683d66235db4fb2b65fb1495672a67bf..660785d3b0066444875091b8c7422466e0b2de38 100644
--- a/net/minecraft/world/entity/ai/behavior/TradeWithVillager.java
+++ b/net/minecraft/world/entity/ai/behavior/TradeWithVillager.java
@@ -71,9 +71,19 @@ public class TradeWithVillager extends Behavior<Villager> {
     }
 
     private static Set<Item> figureOutWhatIAmWillingToTrade(Villager villager, Villager other) {
-        ImmutableSet<Item> set = other.getVillagerData().profession().value().requestedItems();
-        ImmutableSet<Item> set1 = villager.getVillagerData().profession().value().requestedItems();
-        return set.stream().filter(item -> !set1.contains(item)).collect(Collectors.toSet());
+        // Canvas start - remove streams
+        ImmutableSet<Item> otherItems = other.getVillagerData().profession().value().requestedItems();
+        ImmutableSet<Item> villagerItems = villager.getVillagerData().profession().value().requestedItems();
+        Set<Item> result = new java.util.HashSet<>();
+
+        for (Item item : otherItems) {
+            if (!villagerItems.contains(item)) {
+                result.add(item);
+            }
+        }
+
+        return result;
+        // Canvas end - remove streams
     }
 
     private static void throwHalfStack(Villager villager, Set<Item> stack, LivingEntity entity) {
