From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java b/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
index 3b39887a234da7a8aea2c84eb8385f6b54fefca4..94c1a078c4200dce524adab33ea40d76ba4f7bfd 100644
--- a/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
+++ b/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
@@ -513,6 +513,8 @@ public class WalkNodeEvaluator extends NodeEvaluator {
         return blockPathType;
     }
 
+    // Canvas start - optimize path type cache
+    private static final PathType[] PATH_TYPES = PathType.values();
     public static PathType getPathTypeFromState(final BlockGetter level, final BlockPos pos) {
         // Paper start - Do not load chunks during pathfinding
         BlockState blockState = level.getBlockStateIfLoaded(pos);
@@ -520,6 +522,11 @@ public class WalkNodeEvaluator extends NodeEvaluator {
             return PathType.BLOCKED;
         }
         // Paper end
+        return PATH_TYPES[blockState.pathType];
+    }
+
+    public static PathType canvas$getPathTypeFromBlockState(final BlockState blockState) {
+    // Canvas end - optimize path type cache
         Block block = blockState.getBlock();
         if (blockState.isAir()) {
             return PathType.OPEN;
