--- a/net/minecraft/world/attribute/EnvironmentAttributeSystem.java
+++ b/net/minecraft/world/attribute/EnvironmentAttributeSystem.java
@@ -62,7 +_,19 @@
     static void addDefaultLayers(EnvironmentAttributeSystem.Builder builder, Level level) {
         RegistryAccess registryAccess = level.registryAccess();
         BiomeManager biomeManager = level.getBiomeManager();
-        LongSupplier longSupplier = level::getDayTime;
+        // Canvas start - region threading
+        LongSupplier longSupplier = () -> {
+            io.papermc.paper.threadedregions.RegionizedWorldData worldData =
+                io.papermc.paper.threadedregions.TickRegionScheduler.getCurrentRegionizedWorldData();
+            // tbh worldData shouldn't be null, but just to be sure we check
+            if (worldData == null || worldData.world != level) {
+                // world not ours, get day time safely
+                return level.getLevelData().getDayTime();
+            }
+            // on correct world, get daytime from region data
+            return worldData.getTickData().dayTime();
+        };
+        // Canvas end - region threading
         addDimensionLayer(builder, level.dimensionType());
         addBiomeLayer(builder, registryAccess.lookupOrThrow(Registries.BIOME), biomeManager);
         level.dimensionType().timelines().forEach(holder -> builder.addTimelineLayer((Holder<Timeline>)holder, longSupplier));
