From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/util/context/ContextKeySet.java b/net/minecraft/util/context/ContextKeySet.java
index 3ae040c4910d64839e7b1c4da266abb35b84bb0a..f8209453bec027c7545cd283e7b8b602717f9e65 100644
--- a/net/minecraft/util/context/ContextKeySet.java
+++ b/net/minecraft/util/context/ContextKeySet.java
@@ -9,8 +9,8 @@ public class ContextKeySet {
     private final Set<ContextKey<?>> allowed;
 
     ContextKeySet(Set<ContextKey<?>> required, Set<ContextKey<?>> allowed) {
-        this.required = Set.copyOf(required);
-        this.allowed = Set.copyOf(Sets.union(required, allowed));
+        this.required = new it.unimi.dsi.fastutil.objects.ObjectOpenHashSet<>(required); // Canvas - optimize ContextMap#create
+        this.allowed = new it.unimi.dsi.fastutil.objects.ObjectOpenHashSet<>(Sets.union(required, allowed)); // Canvas - optimize ContextMap#create
     }
 
     public Set<ContextKey<?>> required() {
