From 9e9970d708edecf8556f1b2db11dec8fc04b5815 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/navigation/WallClimberNavigation.java b/net/minecraft/world/entity/ai/navigation/WallClimberNavigation.java
index da73c38857960aa2977164dfbe421cbf02673d25..a3c75532fe221022b7f49bba6bb10033dffe27ad 100644
--- a/net/minecraft/world/entity/ai/navigation/WallClimberNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/WallClimberNavigation.java
@@ -44,11 +44,11 @@ public class WallClimberNavigation extends GroundPathNavigation {
             super.tick();
         } else {
             if (this.pathToPosition != null) {
-                if (!this.pathToPosition.closerToCenterThan(this.mob.position(), this.mob.getBbWidth())
+                if (!this.pathToPosition.closerToCenterThan(this.mob.position(), io.canvasmc.canvas.Config.INSTANCE.fixes.mc94054 ? Math.max(this.mob.getBbWidth(), 1.0F) : this.mob.getBbWidth()) // Canvas - MC-94054
                     && (
                         !(this.mob.getY() > this.pathToPosition.getY())
                             || !BlockPos.containing(this.pathToPosition.getX(), this.mob.getY(), this.pathToPosition.getZ())
-                                .closerToCenterThan(this.mob.position(), this.mob.getBbWidth())
+                                .closerToCenterThan(this.mob.position(), io.canvasmc.canvas.Config.INSTANCE.fixes.mc94054 ? Math.max(this.mob.getBbWidth(), 1.0F) : this.mob.getBbWidth()) // Canvas - MC-94054
                     )) {
                     this.mob
                         .getMoveControl()
