--- a/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
@@ -396,7 +_,7 @@
                 }
             }
         } else {
-            for (Entity entity : this.level().getEntities(this.minecart, hitbox)) {
+            for (Entity entity : this.level().getEntitiesOfClass(AbstractMinecart.class, hitbox, entity -> entity != this.minecart)) { // Canvas - optimize entity type predicates
                 if (!this.minecart.hasPassenger(entity) && entity.isPushable() && entity instanceof AbstractMinecart) {
                     // CraftBukkit start
                     org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
