--- a/net/minecraft/world/level/levelgen/material/MaterialRuleList.java
+++ b/net/minecraft/world/level/levelgen/material/MaterialRuleList.java
@@ -9,13 +_,17 @@
     @Nullable
     @Override
     public BlockState calculate(DensityFunction.FunctionContext context) {
-        for (NoiseChunk.BlockStateFiller blockStateFiller : this.materialRuleList) {
-            BlockState blockState = blockStateFiller.calculate(context);
-            if (blockState != null) {
-                return blockState;
-            }
+        // Canvas start - reduce worldgen allocations
+        // avoid iterator allocation
+        BlockState blockState = null;
+        int length = this.materialRuleList.length;
+
+        for (int i = 0; blockState == null && i < length; i++) {
+            NoiseChunk.BlockStateFiller blockStateFiller = this.materialRuleList[i];
+            blockState = blockStateFiller.calculate(context);
         }
 
-        return null;
+        return blockState;
+        // Canvas end - reduce worldgen allocations
     }
 }
