--- a/src/main/java/io/papermc/paper/world/migration/WorldFolderMigration.java
+++ b/src/main/java/io/papermc/paper/world/migration/WorldFolderMigration.java
@@ -49,6 +_,7 @@
             case VANILLA_MIGRATION -> VanillaWorldMigration.migrate(context);
             case NO_OP -> {}
         }
+        io.canvasmc.canvas.util.migration.CanvasFileMigration.initMigration(context); // Canvas - 26.1+ file migration
     }
 
     public static void migrateApiWorld(
