--- a/ca/spottedleaf/moonrise/patches/chunk_tick_iteration/ChunkTickConstants.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_tick_iteration/ChunkTickConstants.java
@@ -1,7 +_,9 @@
 package ca.spottedleaf.moonrise.patches.chunk_tick_iteration;
 
+import io.canvasmc.canvas.Config;
+
 public final class ChunkTickConstants {
 
-    public static final int PLAYER_SPAWN_TRACK_RANGE = 8;
+    public static final int PLAYER_SPAWN_TRACK_RANGE = (int) Math.round(Config.INSTANCE.tickDistanceMaps.playerNearChunkDetectionRange / 16.0); // Canvas - optimize isChunkNearPlayer checks
 
 }
