--- a/net/minecraft/world/entity/boss/wither/WitherBoss.java
+++ b/net/minecraft/world/entity/boss/wither/WitherBoss.java
@@ -135,6 +_,13 @@
         super.setCustomName(name);
         this.bossEvent.setName(this.getDisplayName());
     }
+    // Pufferfish start - optimize suffocation
+
+    @Override
+    public boolean shouldCheckForSuffocation() {
+        return true;
+    }
+    // Pufferfish end
 
     @Override
     public SoundEvent getAmbientSound() {
