From 6321679ed132e14ba66d57ae4ff74f98abc436b1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/boss/wither/WitherBoss.java b/net/minecraft/world/entity/boss/wither/WitherBoss.java
index 419c29c003b3e690a89a14c0a9b3eb2ac3500175..209fc7b390d4fd134fd2ead82989adc62f2df64f 100644
--- a/net/minecraft/world/entity/boss/wither/WitherBoss.java
+++ b/net/minecraft/world/entity/boss/wither/WitherBoss.java
@@ -138,6 +138,13 @@ public class WitherBoss extends Monster implements RangedAttackMob {
         super.setCustomName(name);
         this.bossEvent.setName(this.getDisplayName());
     }
+    // Pufferfish start - optimize suffocation
+
+    @Override
+    public boolean shouldCheckForSuffocation() {
+        return true;
+    }
+    // Pufferfish end
 
     @Override
     public SoundEvent getAmbientSound() {
