--- a/net/minecraft/world/entity/npc/Villager.java
+++ b/net/minecraft/world/entity/npc/Villager.java
@@ -867,7 +_,17 @@
 
     private int countFoodPointsInInventory() {
         SimpleContainer inventory = this.getInventory();
-        return FOOD_POINTS.entrySet().stream().mapToInt(entry -> inventory.countItem(entry.getKey()) * entry.getValue()).sum();
+        // Canvas start - remove streams
+        int sum = 0;
+
+        for (Map.Entry<Item, Integer> entry : FOOD_POINTS.entrySet()) {
+            Item item = entry.getKey();
+            int points = entry.getValue();
+            sum += inventory.countItem(item) * points;
+        }
+
+        return sum;
+        // Canvas end - remove streams
     }
 
     public boolean hasFarmSeeds() {
