--- a/net/minecraft/world/level/levelgen/blending/Blender.java
+++ b/net/minecraft/world/level/levelgen/blending/Blender.java
@@ -149,7 +_,7 @@
     private static double heightToOffset(final double height) {
         double dimensionFactor = 1.0;
         double targetY = height + 0.5;
-        double targetYMod = Mth.positiveModulo(targetY, 8.0);
+        double targetYMod = Mth.positiveModuloForPositiveIntegerDenominator(targetY, 8.0); // Canvas - faster floating-point positive modulo
         return 1.0 * (32.0 * (targetY - 128.0) - 3.0 * (targetY - 120.0) * targetYMod + 3.0 * targetYMod * targetYMod) / (128.0 * (32.0 - 3.0 * targetYMod));
     }
 
