--- a/net/minecraft/world/entity/animal/sniffer/Sniffer.java
+++ b/net/minecraft/world/entity/animal/sniffer/Sniffer.java
@@ -266,7 +_,16 @@
 
     private boolean canDig(BlockPos pos) {
         return this.level().getBlockState(pos).is(BlockTags.SNIFFER_DIGGABLE_BLOCK)
-            && this.getExploredPositions().noneMatch(globalPos -> GlobalPos.of(this.level().dimension(), pos).equals(globalPos))
+            // Canvas start - region threading
+            && this.getExploredPositions().noneMatch(globalPos -> {
+                if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(
+                    net.minecraft.server.MinecraftServer.getServer().getLevel(globalPos.dimension()), globalPos.pos()
+                )) {
+                    return false;
+                }
+                return GlobalPos.of(this.level().dimension(), pos).equals(globalPos);
+            })
+            // Canvas end - region threading
             && Optional.ofNullable(this.getNavigation().createPath(pos, 1)).map(Path::canReach).orElse(false);
     }
 
