--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -166,9 +_,9 @@
             this.yo = this.getY();
             this.zo = this.getZ();
             Vec3 deltaMovement = this.getDeltaMovement();
-            if (this.isInWater() && this.getFluidHeight(FluidTags.WATER) > 0.1F) {
+            if (this.isInWater() && canvas$getWaterFluidHeight() > 0.1F) { // Canvas - optimize fluid height
                 this.setUnderwaterMovement();
-            } else if (this.isInLava() && this.getFluidHeight(FluidTags.LAVA) > 0.1F) {
+            } else if (this.isInLava() && canvas$getLavaFluidHeight() > 0.1F) { // Canvas - optimize fluid height
                 this.setUnderLavaMovement();
             } else {
                 this.applyGravity();
@@ -340,6 +_,15 @@
         return !this.isInvulnerableToBase(damageSource) && this.getItem().canBeHurtBy(damageSource);
     }
 
+    // Canvas start - invulnerable to lightning and explosions config
+    @Override
+    public boolean isInvulnerableToBase(final DamageSource damageSource) {
+        if (io.canvasmc.canvas.Config.INSTANCE.itemEntitiesImmuneToExplosions && damageSource.is(net.minecraft.tags.DamageTypeTags.IS_EXPLOSION)) return true;
+        if (io.canvasmc.canvas.Config.INSTANCE.itemEntitiesImmuneToLightning && damageSource.is(net.minecraft.tags.DamageTypeTags.IS_LIGHTNING)) return true;
+        return super.isInvulnerableToBase(damageSource);
+    }
+
+    // Canvas end - invulnerable to lightning and explosions config
     @Override
     public final boolean hurtServer(ServerLevel level, DamageSource damageSource, float amount) {
         if (this.isInvulnerableToBase(damageSource)) {
@@ -612,4 +_,13 @@
     public SlotAccess getSlot(int slot) {
         return slot == 0 ? SlotAccess.of(this::getItem, this::setItem) : super.getSlot(slot);
     }
+    // Canvas start - MC-206922
+
+    @Override
+    public void thunderHit(final ServerLevel level, final net.minecraft.world.entity.LightningBolt lightning) {
+        if (!io.canvasmc.canvas.Config.INSTANCE.fixes.mc206922 || this.tickCount > 8) {
+            super.thunderHit(level, lightning);
+        }
+    }
+    // Canvas end - MC-206922
 }
