--- a/net/minecraft/world/level/levelgen/structure/ScatteredFeaturePiece.java
+++ b/net/minecraft/world/level/levelgen/structure/ScatteredFeaturePiece.java
@@ -12,7 +_,7 @@
     protected final int width;
     protected final int height;
     protected final int depth;
-    protected int heightPosition = -1;
+    protected volatile int heightPosition = -1; // Canvas - volatile
 
     protected ScatteredFeaturePiece(StructurePieceType type, int x, int y, int z, int width, int height, int depth, Direction orientation) {
         super(type, 0, StructurePiece.makeBoundingBox(x, y, z, orientation, width, height, depth));
