From 53e1a57c72953de1bfcc94c3257c724e7fa1f88f Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/animal/fish/Pufferfish.java b/net/minecraft/world/entity/animal/fish/Pufferfish.java
index 9981452e2dda0675defd279c4460031dae5fd25f..7719b3f233f03aeddc0b118a6fd11ad9c5209aa4 100644
--- a/net/minecraft/world/entity/animal/fish/Pufferfish.java
+++ b/net/minecraft/world/entity/animal/fish/Pufferfish.java
@@ -160,6 +160,7 @@ public class Pufferfish extends AbstractFish {
         int puffState = this.getPuffState();
         if (player instanceof ServerPlayer serverPlayer
             && puffState > 0
+            && (!io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaFixes.mc155509 || this.isAlive()) // Canvas - MC-155509
             && player.hurtServer(serverPlayer.level(), this.damageSources().mobAttack(this), 1 + puffState)) {
             if (!this.isSilent()) {
                 serverPlayer.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.PUFFER_FISH_STING, 0.0F));
