From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/server/network/config/PrepareSpawnTask.java b/net/minecraft/server/network/config/PrepareSpawnTask.java
index 439fc764794b3a67dffa898b91f50a1878737bc3..9d0057bd3a54b2f11850e808d3d64b517435c080 100644
--- a/net/minecraft/server/network/config/PrepareSpawnTask.java
+++ b/net/minecraft/server/network/config/PrepareSpawnTask.java
@@ -79,7 +79,7 @@ public class PrepareSpawnTask implements ConfigurationTask {
                 if (bWorld != null) {
                     resourceKey = ((org.bukkit.craftbukkit.CraftWorld) bWorld).getHandle().dimension();
                 } else {
-                    resourceKey = net.minecraft.world.level.Level.OVERWORLD;
+                    resourceKey = io.canvasmc.canvas.GlobalConfiguration.fetchRespawnDimensionKey(); // Canvas - default respawn dimension config
                     invalidPlayerWorld[0] = true;
                 }
             }
