From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris Bauer <jedimastertoothless@hotmail.com>
Date: Thu, 26 Oct 2023 20:54:10 +0000
Subject: [PATCH] Use Sonatype


diff --git a/.gitignore b/.gitignore
index 3811c0d849a3eb028ed1a6b7a2d4747f7f570448..09d00408604e75c6ff418ce2d2a7ded744d3487f 100644
--- a/.gitignore
+++ b/.gitignore
@@ -46,3 +46,4 @@ dependency-reduced-pom.xml
 # vs code
 /.vscode
 /.factorypath
+bin/
diff --git a/build.gradle.kts b/build.gradle.kts
index ae887244ca5bd1907415195a60af1d6c642d7abd..5a742f0ab5727cdad3cbc1714cf1c7a3863c8ed9 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -117,9 +117,9 @@ tasks.compileTestJava {
 // Paper end
 
 publishing {
-    publications.create<MavenPublication>("maven") {
-        artifact(tasks.shadowJar)
-    }
+    // publications.create<MavenPublication>("maven") {
+    //     artifact(tasks.shadowJar)
+    // }
 }
 
 relocation {
@@ -228,3 +228,40 @@ tasks.registerRunTask("runDev") {
     classpath(sourceSets.main.map { it.runtimeClasspath })
     jvmArgs("-DPaper.pushPaperAssetsRoot=true")
 }
+
+// Canvas start
+publishing {
+    publications {
+        create<MavenPublication>("mavenJava") {
+            pom {
+                name.set("Canvas-Server")
+                description.set("Empowering the impossible in Minecraft")
+                url.set("https://github.com/Dueris/Canvas")
+                licenses {
+                    license {
+                        name.set("GNU General Public License 3.0")
+                        url.set("https://www.gnu.org/licenses/gpl-3.0.en.html")
+                    }
+                }
+            }
+            groupId = "io.github.dueris"
+            artifactId = "canvas-server"
+            from(components["java"])
+        }
+    }
+    repositories {
+        maven {
+            name = "sonatype"
+            if (version.toString().endsWith("SNAPSHOT")) {
+                url = uri("https://s01.oss.sonatype.org/content/repositories/snapshots/")
+            } else {
+                url = uri("https://s01.oss.sonatype.org/service/local/staging/deploy/maven2/")
+            }
+            credentials {
+                username=System.getenv("OSSRH_USERNAME")
+                password=System.getenv("OSSRH_PASSWORD")
+            }
+        }
+    }
+}
+// Canvas end
\ No newline at end of file
diff --git a/gradle.properties b/gradle.properties
new file mode 100644
index 0000000000000000000000000000000000000000..bdae19149cb5e173a8538e3682a2efa6b19d6b08
--- /dev/null
+++ b/gradle.properties
@@ -0,0 +1,2 @@
+ossrhUsername=System.getenv("OSSRH_USERNAME")
+ossrhPassword=System.getenv("OSSRH_PASSWORD")
\ No newline at end of file
