From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <159172167+PurpleWolfMC@users.noreply.github.com>
Date: Thu, 25 Apr 2024 09:27:06 +0000
Subject: [PATCH] Expose Launch Args


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 56b6a8aedff5e83ca79aebd8857bd2392d8d87a8..3fb3114fe690d4becedddf1f9682f1727258fc5e 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -273,6 +273,9 @@ import javax.annotation.Nonnull; // Paper
 public final class CraftServer implements Server {
     private final String serverName = "Purpur"; // Paper // Pufferfish // Purpur
     private final String serverVersion;
+    // Canvas start
+    public static String[] startupArgs = null;
+    // Canvas end
     private final String bukkitVersion = Versioning.getBukkitVersion();
     private final Logger logger = Logger.getLogger("Minecraft");
     private final ServicesManager servicesManager = new SimpleServicesManager();
@@ -678,26 +681,35 @@ public final class CraftServer implements Server {
         return this.console.getGui();
     }
 
+    @Override
     public long getResourcesTotal(){
         return this.getGuiHandle().getGuiStatsComponent().getRamDetails().getGraphData().getTotal();
     }
 
+    @Override
     public long getResourcesFree(){
         return this.getGuiHandle().getGuiStatsComponent().getRamDetails().getGraphData().getFree();
     }
 
+    @Override
     public long getResourcesMax(){
         return this.getGuiHandle().getGuiStatsComponent().getRamDetails().getGraphData().getMax();
     }
 
+    @Override
     public long getResourcesUsedMem(){
         return this.getGuiHandle().getGuiStatsComponent().getRamDetails().getGraphData().getUsedMem();
     }
 
+    @Override
     public int getResourcesUsedPercent(){
         return this.getGuiHandle().getGuiStatsComponent().getRamDetails().getGraphData().getUsedPercent();
     }
 
+    @Override
+    public String[] getStartupArgs(){
+        return CraftServer.startupArgs != null ? CraftServer.startupArgs : null;
+    }
     // Canvas end
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index fcc8527f5583af7792ff8b18a8c3d416de7a9f37..e3fc025fe836a42bddb8b6ec15861104cfc66178 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -394,6 +394,7 @@ public class Main {
             tryPreloadClass(io.papermc.paper.chunk.system.RegionizedPlayerChunkLoader.PlayerChunkLoaderData.class.getName());
             // Paper end
         }
+        CraftServer.startupArgs = args; // Canvas - Expose Launch Args
     }
 
     // Paper start
