From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolf1909 <160778222+PurpleWolf1909@users.noreply.github.com>
Date: Tue, 30 Apr 2024 21:37:05 +0000
Subject: [PATCH] More ModelPart API


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index dd1d4f17656576f60c47f16b28fa09ccb10b73fc..5021cf06ebb20a2abc8300cfc2f1d9e6ad9bdf25 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -443,6 +443,16 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
         this.bukkitPickUpLoot = true;
         this.maxHealthCache = this.getMaxHealth();
     }
+    // Canvas start - ModelPart api
+
+    public io.github.dueris.canvas.entity.player.PlayerModelPart minecraftToCanvasModelPart(net.minecraft.world.entity.player.PlayerModelPart modelPart){
+        return io.github.dueris.canvas.entity.player.PlayerModelPart.valueOf(modelPart.toString());
+    }
+
+    public net.minecraft.world.entity.player.PlayerModelPart canvasToMinecraftModelPart(io.github.dueris.canvas.entity.player.PlayerModelPart modelPart){
+        return net.minecraft.world.entity.player.PlayerModelPart.valueOf(modelPart.toString());
+    }
+    // Canvas end
 
     // Use method to resend items in hands in case of client desync, because the item use got cancelled.
     // For example, when cancelling the leash event
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 72eb043e99ccd35832b707bc21dc4369696a1177..07a9bc7347de0ee798c6d2f2918733c9b4bced7e 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -3291,6 +3291,13 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         }
         return (int) (duration.toMillis() / 50L);
     }
+    // Canvas start - PlayerModelPart api
+
+    @Override
+    public boolean hasModelPartShown(io.github.dueris.canvas.entity.player.PlayerModelPart modelPart){
+        return this.getHandle().isModelPartShown(this.getHandle().canvasToMinecraftModelPart(modelPart));
+    }
+    // Canvas end
 
     @Override
     public void clearTitle() {
