From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <159172167+PurpleWolfMC@users.noreply.github.com>
Date: Thu, 25 Apr 2024 08:28:26 +0000
Subject: [PATCH] More Block API


diff --git a/src/main/java/me/dueris/canvas/CanvasConfig.java b/src/main/java/me/dueris/canvas/CanvasConfig.java
index 2b84679ab281474d30fcd98ce51a24af19c1a068..bbf155dc497e740db472356fe5dd743af8f7d158 100644
--- a/src/main/java/me/dueris/canvas/CanvasConfig.java
+++ b/src/main/java/me/dueris/canvas/CanvasConfig.java
@@ -167,5 +167,14 @@ public class CanvasConfig {
         }
         return builder.build();
     }
-
+    
+    public static boolean useCanvasBedSetDye = true;
+    public static boolean makeCampfireAlwaysLots = false;
+    public static int farmlandSearchRadius = 4;
+
+    private static void blocks(){
+        useCanvasBedSetDye = getBoolean("blocks.useCanvasBedSetDyeMethod", useCanvasBedSetDye);
+        makeCampfireAlwaysLots = getBoolean("block.forceCampfireEmitLotsOfParticles", makeCampfireAlwaysLots);
+        farmlandSearchRadius = getInt("block.farmlandSearchRadius", farmlandSearchRadius);
+    }
 }
diff --git a/src/main/java/net/minecraft/world/level/block/FarmBlock.java b/src/main/java/net/minecraft/world/level/block/FarmBlock.java
index 0c39126ce51439cce05747161aba43bce33a12d8..11b40a3e3cc2bf08a71cff5772beb41a3e7c55f6 100644
--- a/src/main/java/net/minecraft/world/level/block/FarmBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/FarmBlock.java
@@ -166,15 +166,18 @@ public class FarmBlock extends Block {
         return world.getBlockState(pos.above()).is(BlockTags.MAINTAINS_FARMLAND);
     }
 
+    // Canvas start
+    private static int searchRadius = me.dueris.canvas.CanvasConfig.farmlandSearchRadius;
+    
     private static boolean isNearWater(LevelReader world, BlockPos pos) {
         // Paper start - Perf: remove abstract block iteration
         int xOff = pos.getX();
         int yOff = pos.getY();
         int zOff = pos.getZ();
 
-        for (int dz = -4; dz <= 4; ++dz) {
+        for (int dz = -searchRadius; dz <= searchRadius; ++dz) {
             int z = dz + zOff;
-            for (int dx = -4; dx <= 4; ++dx) {
+            for (int dx = -searchRadius; dx <= searchRadius; ++dx) {
                 int x = xOff + dx;
                 for (int dy = 0; dy <= 1; ++dy) {
                     int y = dy + yOff;
@@ -190,6 +193,7 @@ public class FarmBlock extends Block {
         return ((ServerLevel) world).purpurConfig.farmlandGetsMoistFromBelow && world.getFluidState(pos.relative(Direction.DOWN)).is(FluidTags.WATER); // Purpur;
         // Paper end - Perf: remove abstract block iteration
     }
+    // Canvas end
 
     @Override
     protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
diff --git a/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
index 9f7b49a6a50cb53d914c8f4b5132c035c5ab71da..1a14d5fce5f2cfafd9433a8decfef11bde87d560 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
@@ -133,7 +133,7 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
 
         if (randomsource.nextFloat() < 0.11F) {
             for (i = 0; i < randomsource.nextInt(2) + 2; ++i) {
-                CampfireBlock.makeParticles(world, pos, (Boolean) state.getValue(CampfireBlock.SIGNAL_FIRE), false);
+                CampfireBlock.makeParticles(world, pos, (Boolean) state.getValue(CampfireBlock.SIGNAL_FIRE), getShouldBeLots(false)); // Canvas
             }
         }
 
@@ -154,6 +154,13 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
         }
 
     }
+    // Canvas start
+    private static boolean forceLots = me.dueris.canvas.CanvasConfig.makeCampfireAlwaysLots;
+
+    public static boolean getShouldBeLots(boolean original){
+        return forceLots ? true : original;
+    }
+    // Canvas end
 
     public NonNullList<ItemStack> getItems() {
         return this.items;
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBed.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBed.java
index 6595ca064fa410d6249c928c0009ea167d9e6d6e..dccd7ad325c98a81770dd573895cbc7a7b1e5d09 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBed.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBed.java
@@ -1,7 +1,11 @@
 package org.bukkit.craftbukkit.block;
 
 import net.minecraft.world.level.block.entity.BedBlockEntity;
+
+import java.nio.channels.UnsupportedAddressTypeException;
+
 import org.bukkit.DyeColor;
+import org.bukkit.Material;
 import org.bukkit.World;
 import org.bukkit.block.Bed;
 
@@ -57,7 +61,15 @@ public class CraftBed extends CraftBlockEntityState<BedBlockEntity> implements B
 
     @Override
     public void setColor(DyeColor color) {
-        throw new UnsupportedOperationException("Must set block type to appropriate bed colour");
+        // throw new UnsupportedOperationException("Must set block type to appropriate bed colour"); - Canvas - why tho lol
+        if(me.dueris.canvas.CanvasConfig.useCanvasBedSetDye){
+            this.getBlock().setType(Material.valueOf(
+                color.toDyeString().toUpperCase() + "_BED"
+            ));
+        }else{
+            // Then we can use Paper/Bukkits throw exception
+            throw new UnsupportedOperationException("Must set block type to appropriate bed colour");
+        }
     }
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftCampfire.java b/src/main/java/org/bukkit/craftbukkit/block/CraftCampfire.java
index cc641dc70dc12b902bd9cad93fffded9032f1e82..5e3df69406a391563761de4cd76e7091222567d6 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftCampfire.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftCampfire.java
@@ -20,6 +20,13 @@ public class CraftCampfire extends CraftBlockEntityState<CampfireBlockEntity> im
     public int getSize() {
         return this.getSnapshot().getItems().size();
     }
+    // Canvas start
+
+    @Override
+    public void extinguish(){
+        this.getSnapshot().dowse();
+    }
+    // Canvas end
 
     @Override
     public ItemStack getItem(int index) {
