From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Sat, 6 Jan 2024 03:59:20 +0000
Subject: [PATCH] Allow setting the respawn dimension of the player


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 7417ccb27524e4bd87100ec9f9c943d41b8a99c9..d68504e799f3c2ec684e30bbf60de784675751f3 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -234,7 +234,7 @@ public class ServerPlayer extends Player {
     private Vec3 enteredLavaOnVehiclePosition;
     private SectionPos lastSectionPos;
     private ChunkTrackingView chunkTrackingView;
-    private ResourceKey<Level> respawnDimension;
+    public ResourceKey<Level> respawnDimension = Level.OVERWORLD; // Canvas - private -> public
     @Nullable
     private BlockPos respawnPosition;
     private boolean respawnForced;
@@ -336,6 +336,12 @@ public class ServerPlayer extends Player {
             return input.setSendViewDistance(distance);
         });
     }
+    // Canvas start
+
+    public void setRespawnDimension(ServerLevel level){
+        this.respawnDimension = level.dimension();
+    }
+    // Canvas end
     // Paper end - replace player chunk loader
     // Paper start - optimise chunk tick iteration
     public double lastEntitySpawnRadiusSquared = -1.0;
@@ -351,7 +357,7 @@ public class ServerPlayer extends Player {
         this.language =  null; // Paper - default to null
         this.lastSectionPos = SectionPos.of(0, 0, 0);
         this.chunkTrackingView = ChunkTrackingView.EMPTY;
-        this.respawnDimension = Level.OVERWORLD;
+        this.respawnDimension = getRespawnDimension();
         this.wardenSpawnTracker = new WardenSpawnTracker(0, 0, 0);
         this.containerSynchronizer = new ContainerSynchronizer() {
             @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 1708c81987af9bce8c1b4499ee4e092cce582ce2..7f3ed7fc3f8ac4ae7854879834ebb480d17564a7 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1582,6 +1582,18 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         }
         return null;
     }
+    // Canvas start
+
+    @Override
+    public void setRespawnDimension(org.bukkit.World world){
+        this.getHandle().setRespawnDimension(((CraftWorld)world).getHandle());
+    }
+
+    @Override
+    public org.bukkit.World getRespawnDimension(){
+        return this.getHandle().level().getServer().getLevel(this.getHandle().respawnDimension).getWorld();
+    }
+    // Canvas end
 
     @Override
     public void setBedSpawnLocation(Location location) {
