From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <159172167+PurpleWolfMC@users.noreply.github.com>
Date: Wed, 24 Apr 2024 08:22:47 +0000
Subject: [PATCH] Make EntityList use LinkedQueue as Handle


diff --git a/src/main/java/com/destroystokyo/paper/util/maplist/EntityList.java b/src/main/java/com/destroystokyo/paper/util/maplist/EntityList.java
index 6913bfd95ce69e2e9fff3158151d6d1eb7871c28..c4213691e91bc05e74fe897285ab4d097a68d153 100644
--- a/src/main/java/com/destroystokyo/paper/util/maplist/EntityList.java
+++ b/src/main/java/com/destroystokyo/paper/util/maplist/EntityList.java
@@ -5,6 +5,7 @@ import net.minecraft.world.entity.Entity;
 import java.util.Arrays;
 import java.util.Iterator;
 import java.util.NoSuchElementException;
+import java.util.concurrent.ConcurrentLinkedQueue; // Canvas
 
 // list with O(1) remove & contains
 /**
@@ -12,26 +13,27 @@ import java.util.NoSuchElementException;
  */
 public final class EntityList implements Iterable<Entity> {
 
-    protected final Int2IntOpenHashMap entityToIndex = new Int2IntOpenHashMap(2, 0.8f);
+    /* protected final Int2IntOpenHashMap entityToIndex = new Int2IntOpenHashMap(2, 0.8f); // Canvas
     {
         this.entityToIndex.defaultReturnValue(Integer.MIN_VALUE);
-    }
+    } */ // Canvas
+    protected final ConcurrentLinkedQueue<Entity> handle = new ConcurrentLinkedQueue<>(); // Canvas
 
     //protected static final Entity[] EMPTY_LIST = new Entity[0]; // Canvas
 
-    protected Entity[] entities = me.dueris.canvas.Constants.EMPTY_entity_arr; // Canvas
-    protected int count;
+    // protected Entity[] entities = me.dueris.canvas.Constants.EMPTY_entity_arr; // Canvas // Make EntityList Concurrent
+    // protected int count; // Canvas
 
     public int size() {
-        return this.count;
+        return this.handle.size(); // Canvas
     }
 
     public boolean contains(final Entity entity) {
-        return this.entityToIndex.containsKey(entity.getId());
+        return this.handle.contains(entity); // Canvas
     }
 
     public boolean remove(final Entity entity) {
-        final int index = this.entityToIndex.remove(entity.getId());
+        /* final int index = this.entityToIndex.remove(entity.getId()); // Canvas
         if (index == Integer.MIN_VALUE) {
             return false;
         }
@@ -46,11 +48,12 @@ public final class EntityList implements Iterable<Entity> {
         this.entities[index] = end;
         this.entities[endIndex] = null;
 
-        return true;
+        return true; */ // Canvas
+        return this.handle.remove(entity); // Canvas
     }
 
     public boolean add(final Entity entity) {
-        final int count = this.count;
+        /* final int count = this.count; // Canvas
         final int currIndex = this.entityToIndex.putIfAbsent(entity.getId(), count);
 
         if (currIndex != Integer.MIN_VALUE) {
@@ -67,33 +70,53 @@ public final class EntityList implements Iterable<Entity> {
         list[count] = entity;
         this.count = count + 1;
 
-        return true;
+        return true; */ // Canvas
+        return this.handle.add(entity); // Canvas
     }
 
     public Entity getChecked(final int index) {
-        if (index < 0 || index >= this.count) {
-            throw new IndexOutOfBoundsException("Index: " + index + " is out of bounds, size: " + this.count);
+        if (index < 0 || index >= this.size()) { // Canvas
+            throw new IndexOutOfBoundsException("Index: " + index + " is out of bounds, size: " + this.size());
         }
-        return this.entities[index];
+        return this.getUnchecked(index); // Canvas
     }
 
     public Entity getUnchecked(final int index) {
-        return this.entities[index];
+        // Canvas start
+        // return this.entities[index];
+        int i = 0;
+        for (Entity element : this.handle) {
+            if (i == index) {
+                return element;
+            }
+            i++;
+        }
+        return null;
+        // Canvas end
     }
 
     public Entity[] getRawData() {
-        return this.entities;
+        // Canvas start
+        // return this.entities;
+        Entity[] entityArray = new Entity[this.size()];
+        int i = 0;
+        for (Entity entity : this.handle) {
+            entityArray[i++] = entity;
+        }
+        return entityArray;
+        // Canvas end
     }
 
     public void clear() {
-        this.entityToIndex.clear();
+        /* this.entityToIndex.clear(); // Canvas
         Arrays.fill(this.entities, 0, this.count, null);
-        this.count = 0;
+        this.count = 0; */ // Canvas
+        this.handle.clear(); // Canvas
     }
 
     @Override
     public Iterator<Entity> iterator() {
-        return new Iterator<Entity>() {
+        /* return new Iterator<Entity>() { // Canvas
 
             Entity lastRet;
             int current;
@@ -123,6 +146,7 @@ public final class EntityList implements Iterable<Entity> {
                 EntityList.this.remove(lastRet);
                 --this.current;
             }
-        };
+        }; */ // Canvas
+        return this.handle.iterator(); // Canvas
     }
 }
