--- a/net/minecraft/world/level/BaseSpawner.java
+++ b/net/minecraft/world/level/BaseSpawner.java
@@ -60,7 +_,7 @@
     }
 
     public boolean isNearPlayer(final Level level, final BlockPos pos) {
-        return level.hasNearbyAlivePlayerThatAffectsSpawning(pos.getX() + 0.5, pos.getY() + 0.5, pos.getZ() + 0.5, this.requiredPlayerRange); // Paper - Affects Spawning API
+        return level.canvas$hasNearbyAlivePlayerThatAffectsSpawningForSpawner(pos.getX() + 0.5, pos.getY() + 0.5, pos.getZ() + 0.5, this.requiredPlayerRange); // Paper - Affects Spawning API // Canvas - optimize spawners
     }
 
     public void clientTick(final Level level, final BlockPos pos) {
@@ -157,7 +_,7 @@
                                 return;
                             }
 
-                            int nearBy = level.getEntities(
+                            int nearBy = level.canvasConfig().spawner.disableMaxNearbyEntitiesCheck ? Integer.MIN_VALUE : level.getEntities( // Canvas - spawner configs
                                     EntityTypeTest.forExactClass(entity.getClass()),
                                     new AABB(pos.getX(), pos.getY(), pos.getZ(), pos.getX() + 1, pos.getY() + 1, pos.getZ() + 1).inflate(this.spawnRange),
                                     EntitySelector.NO_SPECTATORS
@@ -187,6 +_,7 @@
                                     mob.aware = false;
                                 }
                                 // Spigot end
+                                mob.canvas$fromSpawner = true; // Canvas - spawner configs
                             }
 
                             // Paper start
