From 31ebb67cda7f3f00de773448b77d99e750a63ace Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/BaseSpawner.java b/net/minecraft/world/level/BaseSpawner.java
index 4657083b6c32bd72604aab122b1f44deeb4ac0a4..2d768963b658a41cbb192f3f7af43492e525695e 100644
--- a/net/minecraft/world/level/BaseSpawner.java
+++ b/net/minecraft/world/level/BaseSpawner.java
@@ -45,13 +45,13 @@ public abstract class BaseSpawner {
     public @Nullable SpawnData nextSpawnData;
     private double spin;
     private double oSpin;
-    public int minSpawnDelay = 200;
-    public int maxSpawnDelay = 800;
-    public int spawnCount = 4;
+    public int minSpawnDelay = io.canvasmc.canvas.Config.INSTANCE.spawner.minSpawnDelay; // Canvas - spawner configs
+    public int maxSpawnDelay = io.canvasmc.canvas.Config.INSTANCE.spawner.maxSpawnDelay; // Canvas - spawner configs
+    public int spawnCount = io.canvasmc.canvas.Config.INSTANCE.spawner.spawnCount; // Canvas - spawner configs
     private @Nullable Entity displayEntity;
-    public int maxNearbyEntities = 6;
-    public int requiredPlayerRange = 16;
-    public int spawnRange = 4;
+    public int maxNearbyEntities = io.canvasmc.canvas.Config.INSTANCE.spawner.maxNearbyEntities; // Canvas - spawner configs
+    public int requiredPlayerRange = io.canvasmc.canvas.Config.INSTANCE.spawner.requiredPlayerRange; // Canvas - spawner configs
+    public int spawnRange = io.canvasmc.canvas.Config.INSTANCE.spawner.spawnRange; // Canvas - spawner configs
     private int tickDelay = 0; // Paper - Configurable mob spawner tick rate
 
     public void setEntityId(final EntityType<?> type, final @Nullable Level level, final RandomSource random, final BlockPos pos) {
@@ -60,7 +60,7 @@ public abstract class BaseSpawner {
     }
 
     public boolean isNearPlayer(final Level level, final BlockPos pos) {
-        return level.hasNearbyAlivePlayerThatAffectsSpawning(pos.getX() + 0.5, pos.getY() + 0.5, pos.getZ() + 0.5, this.requiredPlayerRange); // Paper - Affects Spawning API
+        return level.canvas$hasNearbyAlivePlayerThatAffectsSpawningForSpawner(pos.getX() + 0.5, pos.getY() + 0.5, pos.getZ() + 0.5, this.requiredPlayerRange); // Paper - Affects Spawning API // Canvas - optimize spawners
     }
 
     public void clientTick(final Level level, final BlockPos pos) {
@@ -157,7 +157,7 @@ public abstract class BaseSpawner {
                                 return;
                             }
 
-                            int nearBy = level.getEntities(
+                            int nearBy = io.canvasmc.canvas.Config.INSTANCE.spawner.disableMaxNearbyEntitiesCheck ? Integer.MIN_VALUE : level.getEntities( // Canvas - spawner configs
                                     EntityTypeTest.forExactClass(entity.getClass()),
                                     new AABB(pos.getX(), pos.getY(), pos.getZ(), pos.getX() + 1, pos.getY() + 1, pos.getZ() + 1).inflate(this.spawnRange),
                                     EntitySelector.NO_SPECTATORS
@@ -187,6 +187,7 @@ public abstract class BaseSpawner {
                                     mob.aware = false;
                                 }
                                 // Spigot end
+                                mob.canvas$fromSpawner = true; // Canvas - spawner configs
                             }
 
                             // Paper start
