--- a/net/minecraft/world/level/BaseSpawner.java
+++ b/net/minecraft/world/level/BaseSpawner.java
@@ -4,6 +_,7 @@
 import java.util.Optional;
 import java.util.function.Function;
 import javax.annotation.Nullable;
+import io.canvasmc.canvas.Config;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.particles.ParticleTypes;
 import net.minecraft.core.registries.BuiltInRegistries;
@@ -36,14 +_,14 @@
     public SpawnData nextSpawnData;
     private double spin;
     private double oSpin;
-    public int minSpawnDelay = 200;
-    public int maxSpawnDelay = 800;
-    public int spawnCount = 4;
+    public int minSpawnDelay = Config.INSTANCE.spawner.minSpawnDelay; // Canvas - spawner config
+    public int maxSpawnDelay = Config.INSTANCE.spawner.maxSpawnDelay; // Canvas - spawner config
+    public int spawnCount = Config.INSTANCE.spawner.spawnCount; // Canvas - spawner config
     @Nullable
     private Entity displayEntity;
-    public int maxNearbyEntities = 6;
-    public int requiredPlayerRange = 16;
-    public int spawnRange = 4;
+    public int maxNearbyEntities = Config.INSTANCE.spawner.maxNearbyEntities; // Canvas - spawner config
+    public int requiredPlayerRange = Config.INSTANCE.spawner.requiredPlayerRange; // Canvas - spawner config
+    public int spawnRange = Config.INSTANCE.spawner.spawnRange; // Canvas - spawner config
     private int tickDelay = 0; // Paper - Configurable mob spawner tick rate
 
     public void setEntityId(EntityType<?> type, @Nullable Level level, RandomSource random, BlockPos pos) {
