From 3cf7451517453a52df7cb21a31824aadeaf7c6d9 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/PotentialCalculator.java b/net/minecraft/world/level/PotentialCalculator.java
index c6fb54abfe6c950c3693ef0c53ed6264ab87f0cb..85227d7b3d26484abcee444916a299909422bb66 100644
--- a/net/minecraft/world/level/PotentialCalculator.java
+++ b/net/minecraft/world/level/PotentialCalculator.java
@@ -5,7 +5,7 @@ import java.util.List;
 import net.minecraft.core.BlockPos;
 
 public class PotentialCalculator {
-    private final List<PotentialCalculator.PointCharge> charges = Lists.newArrayList();
+    private final List<PotentialCalculator.PointCharge> charges = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(); // Canvas - optimized collection
 
     public void addCharge(final BlockPos pos, final double charge) {
         if (charge != 0.0) {
