--- a/net/minecraft/world/level/PotentialCalculator.java
+++ b/net/minecraft/world/level/PotentialCalculator.java
@@ -5,7 +_,7 @@
 import net.minecraft.core.BlockPos;
 
 public class PotentialCalculator {
-    private final List<PotentialCalculator.PointCharge> charges = Lists.newArrayList();
+    private final List<PotentialCalculator.PointCharge> charges = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(); // Canvas - optimized collection
 
     public void addCharge(final BlockPos pos, final double charge) {
         if (charge != 0.0) {
