From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index b4eec72beed20fdb9426129c2e5cc447b72676c1..3098789f589ee71f42f09499e095e14f895d31e3 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -174,6 +174,7 @@ public abstract class Player extends Avatar implements ContainerUser {
     public boolean affectsSpawning = true; // Paper - Affects Spawning API
     public net.kyori.adventure.util.TriState flyingFallDamage = net.kyori.adventure.util.TriState.NOT_SET; // Paper - flying fall damage
     public int sixRowEnderchestSlotCount = -1; // Purpur - Barrels and enderchests 6 rows
+    public boolean canvas$isTemporarilyBlocking = false; // Canvas - implement sword blocking
 
     // CraftBukkit start
     public boolean fauxSleeping;
@@ -285,7 +286,7 @@ public abstract class Player extends Avatar implements ContainerUser {
             this.setPos(nx, this.getY(), nz);
         }
 
-        this.attackStrengthTicker++;
+        if (!level().canvasConfig().combat.restoreOldAttackDelayMechanics) this.attackStrengthTicker++; // Canvas - combat configs
         this.itemSwapTicker++;
         ItemStack mainHandItemStack = this.getMainHandItem();
         if (!ItemStack.matches(this.lastItemInMainHand, mainHandItemStack)) {
@@ -1051,9 +1052,10 @@ public abstract class Player extends Avatar implements ContainerUser {
                     baseDamage += attackingItemStack.getItem().getAttackDamageBonus(entity, baseDamage, damageSource);
                     boolean criticalAttack = fullStrengthAttack && this.canCriticalAttack(entity);
                     criticalAttack = criticalAttack && !this.level().paperConfig().entities.behavior.disablePlayerCrits; // Paper - Toggleable player crits
+                    criticalAttack = criticalAttack && (!this.level().canvasConfig().combat.disableCritsWhileSprinting || !this.isSprinting()); // Canvas - combat configs
                     if (criticalAttack) {
                         damageSource = damageSource.critical(); // Paper - critical damage API
-                        baseDamage *= 1.5F;
+                        baseDamage *= this.level().canvasConfig().combat.criticalHitMultiplier; // Canvas - combat configs
                     }
 
                     float totalDamage = baseDamage + magicBoost;
@@ -1067,7 +1069,7 @@ public abstract class Player extends Avatar implements ContainerUser {
                     boolean wasHurt = entity.hurtOrSimulate(damageSource, totalDamage);
                     if (wasHurt) {
                         this.causeExtraKnockback(entity, this.getKnockback(entity, damageSource) + (knockbackAttack ? 0.5F : 0.0F), oldMovement);
-                        if (sweepAttack) {
+                        if (sweepAttack && !this.level().canvasConfig().combat.disableSweepingEdge) { // Canvas - combat configs
                             this.doSweepAttack(entity, baseDamage, damageSource, attackStrengthScale);
                         }
 
@@ -1159,7 +1161,7 @@ public abstract class Player extends Avatar implements ContainerUser {
         if (entity instanceof LivingEntity) {
             float actualDamage = oldLivingEntityHealth - ((LivingEntity)entity).getHealth();
             this.awardStat(Stats.DAMAGE_DEALT, Math.round(actualDamage * 10.0F));
-            if (this.level() instanceof ServerLevel && actualDamage > 2.0F) {
+            if (!level().canvasConfig().visuals.particles.disableNewCombatParticles && this.level() instanceof ServerLevel && actualDamage > 2.0F) { // Canvas - particles config
                 int count = (int)(actualDamage * 0.5);
                 ((ServerLevel)this.level())
                     .sendParticles(ParticleTypes.DAMAGE_INDICATOR, entity.getX(), entity.getY(0.5), entity.getZ(), count, 0.1, 0.0, 0.1, 0.2);
@@ -1946,6 +1948,7 @@ public abstract class Player extends Avatar implements ContainerUser {
     }
 
     public float getCurrentItemAttackStrengthDelay() {
+        if (level().canvasConfig().combat.restoreOldAttackDelayMechanics) return 0.0F; // Canvas - combat configs
         return (float)(1.0 / this.getAttributeValue(Attributes.ATTACK_SPEED) * 20.0);
     }
 
@@ -1956,6 +1959,7 @@ public abstract class Player extends Avatar implements ContainerUser {
     }
 
     public float getAttackStrengthScale(final float a) {
+        if (level().canvasConfig().combat.restoreOldAttackDelayMechanics) return 1.0F; // Canvas - combat configs
         return Mth.clamp((this.attackStrengthTicker + a) / this.getCurrentItemAttackStrengthDelay(), 0.0F, 1.0F);
     }
 
