From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/server/players/SleepStatus.java b/net/minecraft/server/players/SleepStatus.java
index 2a7ae521654ad5c9f392baa5562e64bb71b13097..51a128e7b8034e76823b2ccc3835b14cdda6bda3 100644
--- a/net/minecraft/server/players/SleepStatus.java
+++ b/net/minecraft/server/players/SleepStatus.java
@@ -15,9 +15,24 @@ public class SleepStatus {
 
     public boolean areEnoughDeepSleeping(int requiredSleepPercentage, List<ServerPlayer> sleepingPlayers) {
         // CraftBukkit start
-        int i = (int) sleepingPlayers.stream().filter(player -> player.isSleepingLongEnough() || player.fauxSleeping).count();
-        boolean anyDeepSleep = sleepingPlayers.stream().anyMatch(Player::isSleepingLongEnough);
-        return anyDeepSleep && i >= this.sleepersNeeded(requiredSleepPercentage);
+        // Canvas start - remove streams
+        int count = 0;
+        boolean anyPlayerSleeping = false;
+
+        for (ServerPlayer player : sleepingPlayers) {
+            final boolean isSleepingLongEnough = player.isSleepingLongEnough();
+
+            if (isSleepingLongEnough) {
+                anyPlayerSleeping = true;
+            }
+
+            if (isSleepingLongEnough || player.fauxSleeping) {
+                count++;
+            }
+        }
+
+        return anyPlayerSleeping && count >= this.sleepersNeeded(requiredSleepPercentage);
+        // Canvas end - remove streams
         // CraftBukkit end
     }
 
