--- a/net/minecraft/network/protocol/game/ClientboundPlayerInfoUpdatePacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundPlayerInfoUpdatePacket.java
@@ -2,6 +_,7 @@
 
 import com.google.common.base.MoreObjects;
 import com.mojang.authlib.GameProfile;
+import java.util.ArrayList;
 import java.util.Collection;
 import java.util.EnumSet;
 import java.util.List;
@@ -31,7 +_,14 @@
 
     public ClientboundPlayerInfoUpdatePacket(EnumSet<ClientboundPlayerInfoUpdatePacket.Action> actions, Collection<ServerPlayer> players) {
         this.actions = actions;
-        this.entries = players.stream().map(ClientboundPlayerInfoUpdatePacket.Entry::new).toList();
+        // Canvas start - replace stream
+        List<Entry> list = new ArrayList<>();
+        for (ServerPlayer player : players) {
+            Entry entry = new Entry(player);
+            list.add(entry);
+        }
+        this.entries = list;
+        // Canvas end
     }
 
     public ClientboundPlayerInfoUpdatePacket(ClientboundPlayerInfoUpdatePacket.Action action, ServerPlayer player) {
