--- a/net/minecraft/network/protocol/game/ClientboundContainerSetContentPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundContainerSetContentPacket.java
@@ -30,6 +_,7 @@
 
     @Override
     public boolean packetTooLarge(net.minecraft.network.Connection manager) {
+        if (io.canvasmc.canvas.GlobalConfiguration.getInstance().networking.disablePaperPacketOverflowContainerFix) return false; // Canvas - container overflow config
         for (int i = 0 ; i < this.items.size() ; i++) {
             manager.send(new ClientboundContainerSetSlotPacket(this.containerId, this.stateId, i, this.items.get(i)));
         }
