From 4b22e4fb667ff3bf3c489ff7a94018cc27157519 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/network/protocol/game/ClientboundContainerSetContentPacket.java b/net/minecraft/network/protocol/game/ClientboundContainerSetContentPacket.java
index 4c8ad28b42a7fc21dc59e6fe40263462cce74496..2d4c5a3972cc7135bb17c80b554905cea7b9787b 100644
--- a/net/minecraft/network/protocol/game/ClientboundContainerSetContentPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundContainerSetContentPacket.java
@@ -30,6 +30,7 @@ public record ClientboundContainerSetContentPacket(int containerId, int stateId,
 
     @Override
     public boolean packetTooLarge(net.minecraft.network.Connection manager) {
+        if (io.canvasmc.canvas.Config.INSTANCE.networking.disablePaperPacketOverflowContainerFix) return false; // Canvas - container overflow config
         for (int i = 0 ; i < this.items.size() ; i++) {
             manager.send(new ClientboundContainerSetSlotPacket(this.containerId, this.stateId, i, this.items.get(i)));
         }
