--- a/net/minecraft/network/protocol/game/ClientboundUpdateAttributesPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundUpdateAttributesPacket.java
@@ -32,6 +_,16 @@
         this.attributes = Lists.newArrayList();
 
         for (AttributeInstance value : values) {
+            // Canvas start - combat configs
+            if (io.canvasmc.canvas.Config.INSTANCE.combat.disableAttackHitDelay
+                && value.getAttribute().is(net.minecraft.world.entity.ai.attributes.Attributes.ATTACK_SPEED)) {
+                this.attributes.add(
+                    new ClientboundUpdateAttributesPacket.AttributeSnapshot(
+                        value.getAttribute(), 20, new java.util.HashSet<>()
+                    )
+                );
+            }
+            // Canvas end - combat configs
             this.attributes.add(new ClientboundUpdateAttributesPacket.AttributeSnapshot(value.getAttribute(), value.getBaseValue(), value.getModifiers()));
         }
     }
