From 652f21c70faf38fd698db5fb44e853a1e1a939d7 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java b/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
index 42bfd4b3b626ef66b0f2ef8438add5342b367800..dc9169494a5ef25f07f84ab513e26d5d94d3c5dc 100644
--- a/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
+++ b/net/minecraft/network/protocol/game/ClientboundLevelChunkPacketData.java
@@ -61,17 +61,17 @@ public class ClientboundLevelChunkPacketData {
         this.blockEntitiesData = Lists.newArrayList();
         int totalTileEntities = 0; // Paper - Handle oversized block entities in chunks
 
-        for (Entry<BlockPos, BlockEntity> entry : levelChunk.getBlockEntities().entrySet()) {
+        for (BlockEntity blockEntity : levelChunk.canvas$getAllBlockEntities()) { // Canvas - optimize block entity fetching
             // Paper start - Handle oversized block entities in chunks
             if (++totalTileEntities > BLOCK_ENTITY_LIMIT) {
-                net.minecraft.network.protocol.Packet<ClientGamePacketListener> packet = entry.getValue().getUpdatePacket();
+                net.minecraft.network.protocol.Packet<ClientGamePacketListener> packet = blockEntity.getUpdatePacket(); // Canvas - optimize block entity fetching
                 if (packet != null) {
                     this.extraPackets.add(packet);
                     continue;
                 }
             }
             // Paper end - Handle oversized block entities in chunks
-            this.blockEntitiesData.add(ClientboundLevelChunkPacketData.BlockEntityInfo.create(entry.getValue()));
+            this.blockEntitiesData.add(ClientboundLevelChunkPacketData.BlockEntityInfo.create(blockEntity)); // Canvas - optimize block entity fetching
         }
     }
 
