--- a/net/minecraft/network/protocol/game/ClientboundUpdateAttributesPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundUpdateAttributesPacket.java
@@ -31,11 +_,13 @@
         this.entityId = entityId;
         this.attributes = Lists.newArrayList();
 
+        java.util.Set<AttributeModifier> emptySet = new java.util.HashSet<>(); // Canvas - combat configs
         for (AttributeInstance attributeInstance : attributes) {
+            boolean apply = io.canvasmc.canvas.Config.INSTANCE.combat.disableAttackHitDelay && attributeInstance.getAttribute().is(net.minecraft.world.entity.ai.attributes.Attributes.ATTACK_SPEED); // Canvas - combat configs
             this.attributes
                 .add(
                     new ClientboundUpdateAttributesPacket.AttributeSnapshot(
-                        attributeInstance.getAttribute(), attributeInstance.getBaseValue(), attributeInstance.getModifiers()
+                        attributeInstance.getAttribute(), apply ? 20 : attributeInstance.getBaseValue(), apply ? emptySet : attributeInstance.getModifiers() // Canvas - combat configs
                     )
                 );
         }
