From 0ce24b7bccfa2d01cdd888671365a9ba0973fab1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/network/protocol/game/ClientboundUpdateAttributesPacket.java b/net/minecraft/network/protocol/game/ClientboundUpdateAttributesPacket.java
index 344260d191e4c80671062cd29ce0146a16ea7f1d..58f4920b0e6d89c8b262459879bb7f6560d121c5 100644
--- a/net/minecraft/network/protocol/game/ClientboundUpdateAttributesPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundUpdateAttributesPacket.java
@@ -32,6 +32,16 @@ public class ClientboundUpdateAttributesPacket implements Packet<ClientGamePacke
         this.attributes = Lists.newArrayList();
 
         for (AttributeInstance value : values) {
+            // Canvas start - combat configs
+            if (io.canvasmc.canvas.Config.INSTANCE.combat.disableAttackHitDelay
+                && value.getAttribute().is(net.minecraft.world.entity.ai.attributes.Attributes.ATTACK_SPEED)) {
+                this.attributes.add(
+                    new ClientboundUpdateAttributesPacket.AttributeSnapshot(
+                        value.getAttribute(), 20, new java.util.HashSet<>()
+                    )
+                );
+            }
+            // Canvas end - combat configs
             this.attributes.add(new ClientboundUpdateAttributesPacket.AttributeSnapshot(value.getAttribute(), value.getBaseValue(), value.getModifiers()));
         }
     }
