--- a/net/minecraft/network/protocol/game/ClientboundUpdateAttributesPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundUpdateAttributesPacket.java
@@ -32,6 +_,20 @@
         this.attributes = Lists.newArrayList();
 
         for (AttributeInstance value : values) {
+            // Canvas start - combat configs
+            final io.papermc.paper.threadedregions.RegionizedWorldData regionizedWorldData =
+                io.papermc.paper.threadedregions.TickRegionScheduler.getCurrentRegionizedWorldData();
+            if (value.getAttribute().is(net.minecraft.world.entity.ai.attributes.Attributes.ATTACK_SPEED) &&
+                regionizedWorldData.world.canvasConfig().combat.restoreOldAttackDelayMechanics
+            ) {
+                this.attributes.add(
+                    new ClientboundUpdateAttributesPacket.AttributeSnapshot(
+                        value.getAttribute(), 20, new java.util.HashSet<>()
+                    )
+                );
+                continue;
+            }
+            // Canvas end - combat configs
             this.attributes.add(new ClientboundUpdateAttributesPacket.AttributeSnapshot(value.getAttribute(), value.getBaseValue(), value.getModifiers()));
         }
     }
