From 593a33a13babbc895838d407163563cd3688a7c9 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/network/protocol/game/ServerboundChatPacket.java b/net/minecraft/network/protocol/game/ServerboundChatPacket.java
index c9623b551609c77e8f143e492fe07f80b5a07dc0..e4204f8043c68d6e635cbe9a34104ecbba924ea0 100644
--- a/net/minecraft/network/protocol/game/ServerboundChatPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundChatPacket.java
@@ -11,6 +11,15 @@ import org.jspecify.annotations.Nullable;
 
 public record ServerboundChatPacket(String message, Instant timeStamp, long salt, @Nullable MessageSignature signature, LastSeenMessages.Update lastSeenMessages)
     implements Packet<ServerGamePacketListener> {
+    // Canvas start - no chat reports
+
+    @Override
+    @Nullable
+    public MessageSignature signature() {
+        return io.canvasmc.canvas.Config.INSTANCE.enableNoChatReports ? null : signature;
+    }
+
+    // Canvas end - no chat reports
     public static final StreamCodec<FriendlyByteBuf, ServerboundChatPacket> STREAM_CODEC = Packet.codec(
         ServerboundChatPacket::write, ServerboundChatPacket::new
     );
