--- a/net/minecraft/network/protocol/game/ServerboundChatCommandSignedPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundChatCommandSignedPacket.java
@@ -1,6 +_,7 @@
 package net.minecraft.network.protocol.game;
 
 import java.time.Instant;
+import io.canvasmc.canvas.Config;
 import net.minecraft.commands.arguments.ArgumentSignatures;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.chat.LastSeenMessages;
@@ -36,4 +_,14 @@
     public void handle(ServerGamePacketListener handler) {
         handler.handleSignedChatCommand(this);
     }
+    // Canvas start - no chat reports
+
+    @Override
+    public ArgumentSignatures argumentSignatures() {
+        if (Config.INSTANCE.networking.noChatReports.enable) {
+            return ArgumentSignatures.EMPTY;
+        }
+        return argumentSignatures;
+    }
+    // Canvas end
 }
