--- a/net/minecraft/network/protocol/game/ServerboundChatPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundChatPacket.java
@@ -2,6 +_,7 @@
 
 import java.time.Instant;
 import javax.annotation.Nullable;
+import io.canvasmc.canvas.Config;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.chat.LastSeenMessages;
 import net.minecraft.network.chat.MessageSignature;
@@ -36,4 +_,15 @@
     public void handle(ServerGamePacketListener handler) {
         handler.handleChat(this);
     }
+    // Canvas start - no chat reports
+
+    @Override
+    @Nullable
+    public MessageSignature signature() {
+        if (Config.INSTANCE.noChatReports.enable) {
+            return null;
+        }
+        return signature;
+    }
+    // Canvas end
 }
