--- a/net/minecraft/network/protocol/game/ServerboundChatPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundChatPacket.java
@@ -11,6 +_,15 @@
 
 public record ServerboundChatPacket(String message, Instant timeStamp, long salt, @Nullable MessageSignature signature, LastSeenMessages.Update lastSeenMessages)
     implements Packet<ServerGamePacketListener> {
+    // Canvas start - no chat reports
+
+    @Override
+    @Nullable
+    public MessageSignature signature() {
+        return io.canvasmc.canvas.Config.INSTANCE.enableNoChatReports ? null : signature;
+    }
+
+    // Canvas end - no chat reports
     public static final StreamCodec<FriendlyByteBuf, ServerboundChatPacket> STREAM_CODEC = Packet.codec(
         ServerboundChatPacket::write, ServerboundChatPacket::new
     );
