From 3cf7451517453a52df7cb21a31824aadeaf7c6d9 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/network/protocol/game/ServerboundChatCommandSignedPacket.java b/net/minecraft/network/protocol/game/ServerboundChatCommandSignedPacket.java
index 646e86b6a278a5cd3363e24ed9668359d3b250fd..8dd4ca2f5ec4b4872ccb4a7316e2fc13bcba323d 100644
--- a/net/minecraft/network/protocol/game/ServerboundChatCommandSignedPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundChatCommandSignedPacket.java
@@ -11,6 +11,14 @@ import net.minecraft.network.protocol.PacketType;
 public record ServerboundChatCommandSignedPacket(
     String command, Instant timeStamp, long salt, ArgumentSignatures argumentSignatures, LastSeenMessages.Update lastSeenMessages
 ) implements Packet<ServerGamePacketListener> {
+    // Canvas start - no chat reports
+
+    @Override
+    public ArgumentSignatures argumentSignatures() {
+        return io.canvasmc.canvas.Config.INSTANCE.enableNoChatReports ? ArgumentSignatures.EMPTY : argumentSignatures;
+    }
+
+    // Canvas end - no chat reports
     public static final StreamCodec<FriendlyByteBuf, ServerboundChatCommandSignedPacket> STREAM_CODEC = Packet.codec(
         ServerboundChatCommandSignedPacket::write, ServerboundChatCommandSignedPacket::new
     );
