--- a/net/minecraft/network/protocol/game/ServerboundChatCommandSignedPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundChatCommandSignedPacket.java
@@ -11,6 +_,14 @@
 public record ServerboundChatCommandSignedPacket(
     String command, Instant timeStamp, long salt, ArgumentSignatures argumentSignatures, LastSeenMessages.Update lastSeenMessages
 ) implements Packet<ServerGamePacketListener> {
+    // Canvas start - no chat reports
+
+    @Override
+    public ArgumentSignatures argumentSignatures() {
+        return io.canvasmc.canvas.Config.INSTANCE.enableNoChatReports ? ArgumentSignatures.EMPTY : argumentSignatures;
+    }
+
+    // Canvas end - no chat reports
     public static final StreamCodec<FriendlyByteBuf, ServerboundChatCommandSignedPacket> STREAM_CODEC = Packet.codec(
         ServerboundChatCommandSignedPacket::write, ServerboundChatCommandSignedPacket::new
     );
