--- a/net/minecraft/network/protocol/game/ServerboundChatCommandSignedPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundChatCommandSignedPacket.java
@@ -35,4 +_,11 @@
     public void handle(final ServerGamePacketListener listener) {
         listener.handleSignedChatCommand(this);
     }
+    // Canvas start - no chat reports
+
+    @Override
+    public ArgumentSignatures argumentSignatures() {
+        return io.canvasmc.canvas.GlobalConfiguration.getInstance().chat.disableChatReporting ? ArgumentSignatures.EMPTY : argumentSignatures;
+    }
+    // Canvas end - no chat reports
 }
