--- a/net/minecraft/network/protocol/game/ServerboundChatSessionUpdatePacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundChatSessionUpdatePacket.java
@@ -1,10 +_,13 @@
 package net.minecraft.network.protocol.game;
 
+import io.canvasmc.canvas.Config;
 import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.RemoteChatSession;
 import net.minecraft.network.codec.StreamCodec;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.PacketType;
+import net.minecraft.server.network.ServerGamePacketListenerImpl;
 
 public record ServerboundChatSessionUpdatePacket(RemoteChatSession.Data chatSession) implements Packet<ServerGamePacketListener> {
     public static final StreamCodec<FriendlyByteBuf, ServerboundChatSessionUpdatePacket> STREAM_CODEC = Packet.codec(
@@ -26,6 +_,19 @@
 
     @Override
     public void handle(ServerGamePacketListener handler) {
+        // Canvas start - no chat reports
+        if (Config.INSTANCE.noChatReports.enable) {
+            var impl = (ServerGamePacketListenerImpl) handler;
+
+            if (!impl.getPlayer().getServer().isSingleplayerOwner(impl.getPlayer().getGameProfile())) {
+                if (Config.INSTANCE.noChatReports.demandOnClient) {
+                    impl.disconnect(Component.literal(Config.INSTANCE.noChatReports.disconnectDemandOnClientMessage));
+                }
+            }
+
+            return;
+        }
+        // Canvas end
         handler.handleChatSessionUpdate(this);
     }
 }
