From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java b/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
index 722f378bb3726200b6fe88948bf1e1d3514a7b5c..1639bf7cf9a70930e8fa593ddd3922c8918e5230 100644
--- a/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
+++ b/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
@@ -116,6 +116,7 @@ public class RangedBowAttackGoal<T extends Monster & RangedAttackMob> extends Go
                 }
 
                 this.mob.lookAt(target, 30.0F, 30.0F);
+                if (io.canvasmc.canvas.Config.INSTANCE.fixes.mc121706) this.mob.getLookControl().setLookAt(target, 30f, 30f); // Canvas - MC-121706
             } else {
                 this.mob.getLookControl().setLookAt(target, 30.0F, 30.0F);
             }
