From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/block/BambooStalkBlock.java b/net/minecraft/world/level/block/BambooStalkBlock.java
index 81e2a279d4c29f5fe52387875489239515a8c82b..83ad254f2cb4ac8471ab7cf14851bb93a4bc150a 100644
--- a/net/minecraft/world/level/block/BambooStalkBlock.java
+++ b/net/minecraft/world/level/block/BambooStalkBlock.java
@@ -81,11 +81,16 @@ public class BambooStalkBlock extends Block implements BonemealableBlock {
 
     @Override
     public @Nullable BlockState getStateForPlacement(BlockPlaceContext context) {
-        FluidState fluidState = context.getLevel().getFluidState(context.getClickedPos());
+        // Canvas start - optimize random tick
+        final BlockPos clickedPos = context.getClickedPos();
+        net.minecraft.world.level.chunk.LevelChunk chunk = context.getLevel().getChunkIfLoaded(clickedPos);
+        if (chunk == null) return null;
+        FluidState fluidState = chunk.getFluidState(clickedPos);
+        // Canvas end - optimize random tick
         if (!fluidState.isEmpty()) {
             return null;
         } else {
-            BlockState blockState = context.getLevel().getBlockState(context.getClickedPos().below());
+            BlockState blockState = chunk.getBlockState(context.getClickedPos().below()); // Canvas - optimize random tick
             if (blockState.is(BlockTags.BAMBOO_PLANTABLE_ON)) {
                 if (blockState.is(Blocks.BAMBOO_SAPLING)) {
                     return this.defaultBlockState().setValue(AGE, 0);
@@ -93,7 +98,7 @@ public class BambooStalkBlock extends Block implements BonemealableBlock {
                     int i = blockState.getValue(AGE) > 0 ? 1 : 0;
                     return this.defaultBlockState().setValue(AGE, i);
                 } else {
-                    BlockState blockState1 = context.getLevel().getBlockState(context.getClickedPos().above());
+                    BlockState blockState1 = chunk.getBlockState(context.getClickedPos().above()); // Canvas - optimize random tick
                     return blockState1.is(Blocks.BAMBOO)
                         ? this.defaultBlockState().setValue(AGE, blockState1.getValue(AGE))
                         : Blocks.BAMBOO_SAPLING.defaultBlockState();
@@ -186,9 +191,13 @@ public class BambooStalkBlock extends Block implements BonemealableBlock {
     }
 
     protected void growBamboo(BlockState state, Level level, BlockPos pos, RandomSource random, int age) {
-        BlockState blockState = level.getBlockState(pos.below());
+        // Canvas start - optimize random tick
+        net.minecraft.world.level.chunk.LevelChunk chunk = level.getChunkIfLoaded(pos);
+        if (chunk == null) return;
+        BlockState blockState = chunk.getBlockState(pos.below());
         BlockPos blockPos = pos.below(2);
-        BlockState blockState1 = level.getBlockState(blockPos);
+        BlockState blockState1 = chunk.getBlockState(blockPos);
+        // Canvas end - optimize random tick
         BambooLeaves bambooLeaves = BambooLeaves.NONE;
         boolean shouldUpdateOthers = false; // CraftBukkit
         if (age >= 1) {
