From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/block/Blocks.java b/net/minecraft/world/level/block/Blocks.java
index 2eb6fe73adeda3e525e147b7f7034bdec651e04a..bb3ee476b9d4d113ac471e1c006cc7b5f496149c 100644
--- a/net/minecraft/world/level/block/Blocks.java
+++ b/net/minecraft/world/level/block/Blocks.java
@@ -7287,6 +7287,11 @@ public class Blocks {
     }
 
     private static Block register(String name, BlockBehaviour.Properties properties) {
+        // Canvas start - MC-223153
+        if (io.canvasmc.canvas.Config.INSTANCE.fixes.mc223153 && name.equalsIgnoreCase("raw_copper_block")) {
+            properties = properties.sound(SoundType.COPPER);
+        }
+        // Canvas end - MC-223153
         return register(name, Block::new, properties);
     }
 
