From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/block/ChestBlock.java b/net/minecraft/world/level/block/ChestBlock.java
index fa2dce759419f62825000718361f47c6f208cdb2..b23e58b5cb27ce2c861f42d9ab38632be47bb2ed 100644
--- a/net/minecraft/world/level/block/ChestBlock.java
+++ b/net/minecraft/world/level/block/ChestBlock.java
@@ -92,6 +92,28 @@ public class ChestBlock extends AbstractChestBlock<ChestBlockEntity> implements
             return Optional.empty();
         }
     };
+    // Canvas start - optimize hoppers
+    private static final DoubleBlockCombiner.Combiner<ChestBlockEntity, Container> NULLABLE_DIRECT_CONTAINER = new DoubleBlockCombiner.Combiner<ChestBlockEntity, Container>() {
+        @Override
+        public Container acceptDouble(ChestBlockEntity first, ChestBlockEntity second) {
+            return new CompoundContainer(first, second);
+        }
+
+        @Override
+        public Container acceptSingle(ChestBlockEntity single) {
+            return single;
+        }
+
+        @Override
+        public Container acceptNone() {
+            // the data flow issue is that all methods return non-null by default
+            // however when we use this we have nullability handling since we
+            // do expect the return value to be nullable
+            //noinspection DataFlowIssue
+            return null;
+        }
+    };
+    // Canvas end - optimize hoppers
     public static DoubleBlockCombiner.Combiner<ChestBlockEntity, Optional<MenuProvider>> MENU_PROVIDER_COMBINER = new DoubleBlockCombiner.Combiner<ChestBlockEntity, Optional<MenuProvider>>() {
         @Override
         public Optional<MenuProvider> acceptDouble(final ChestBlockEntity first, final ChestBlockEntity second) {
@@ -306,6 +328,14 @@ public class ChestBlock extends AbstractChestBlock<ChestBlockEntity> implements
         return this.blockEntityType.get();
     }
 
+    // Canvas start - optimize hoppers
+    public static @Nullable Container canvas$getContainerForHoppers(ChestBlock chest, BlockState state, Level level, BlockPos pos, boolean override, Container def) {
+        // when conducting tests, the code bellow the injected code took ~17.1ms out of the ~40.45ms total
+        // to tick the hoppers in our testing environment. ~20.14ms was from this method
+        return chest.combine(state, level, pos, override).apply(NULLABLE_DIRECT_CONTAINER);
+    }
+
+    // Canvas end - optimize hoppers
     public static @Nullable Container getContainer(ChestBlock chest, BlockState state, Level level, BlockPos pos, boolean override) {
         return chest.combine(state, level, pos, override).apply(CHEST_COMBINER).orElse(null);
     }
