From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/block/CropBlock.java b/net/minecraft/world/level/block/CropBlock.java
index 1cf40fafd822d976ef4822335c60d8017659916f..b95b0f86b14b71861a263e0db2327ac083dc1cb9 100644
--- a/net/minecraft/world/level/block/CropBlock.java
+++ b/net/minecraft/world/level/block/CropBlock.java
@@ -163,6 +163,7 @@ public class CropBlock extends VegetationBlock implements BonemealableBlock {
     }
 
     protected static boolean hasSufficientLight(LevelReader level, BlockPos pos) {
+        if (io.canvasmc.canvas.Config.INSTANCE.cropsIgnoreLightCheck) return true; // Canvas - crops ignore light check config
         return level.getRawBrightness(pos, 0) >= 8;
     }
 
