From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/block/FarmBlock.java b/net/minecraft/world/level/block/FarmBlock.java
index eb21b20a72062fc53b8f94aa98dcd380c8ad4db3..c7a0c898f4b464ecb3a95d01c0685a7ea79a338b 100644
--- a/net/minecraft/world/level/block/FarmBlock.java
+++ b/net/minecraft/world/level/block/FarmBlock.java
@@ -41,7 +41,7 @@ public class FarmBlock extends Block {
 
     protected FarmBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(MOISTURE, 0));
+        this.registerDefaultState(this.stateDefinition.any().setValue(MOISTURE, io.canvasmc.canvas.Config.INSTANCE.farmlandAlwaysMoist ? MAX_MOISTURE : 0)); // Canvas - farmland always moist config
     }
 
     @Override
@@ -97,7 +97,7 @@ public class FarmBlock extends Block {
         int moistureValue = state.getValue(MOISTURE);
         if (moistureValue > 0 && level.paperConfig().tickRates.wetFarmland != 1 && (level.paperConfig().tickRates.wetFarmland < 1 || (level.getRedstoneGameTime() + pos.hashCode()) % level.paperConfig().tickRates.wetFarmland != 0)) { return; } // Paper - Configurable random tick rates for blocks // Folia - region threading
         if (moistureValue == 0 && level.paperConfig().tickRates.dryFarmland != 1 && (level.paperConfig().tickRates.dryFarmland < 1 || (level.getRedstoneGameTime() + pos.hashCode()) % level.paperConfig().tickRates.dryFarmland != 0)) { return; } // Paper - Configurable random tick rates for blocks // Folia - region threading
-        if (!isNearWater(level, pos) && !level.isRainingAt(pos.above())) {
+        if (!io.canvasmc.canvas.Config.INSTANCE.farmlandAlwaysMoist && !isNearWater(level, pos) && !level.isRainingAt(pos.above())) { // Canvas - farmland always moist config - if farmland can be moist then we allow moisture changes, otherwise we assume near water
             if (moistureValue > 0) {
                 org.bukkit.craftbukkit.event.CraftEventFactory.handleMoistureChangeEvent(level, pos, state.setValue(MOISTURE, moistureValue - 1), Block.UPDATE_CLIENTS); // CraftBukkit
             } else if (!shouldMaintainFarmland(level, pos)) {
@@ -112,6 +112,7 @@ public class FarmBlock extends Block {
     public void fallOn(Level level, BlockState state, BlockPos pos, Entity entity, double fallDistance) {
         super.fallOn(level, state, pos, entity, fallDistance); // CraftBukkit - moved here as game rules / events shouldn't affect fall damage.
         if (level instanceof ServerLevel serverLevel
+            && !io.canvasmc.canvas.Config.INSTANCE.disableFarmlandTrampling // Canvas - disable farmland trampling config
             && level.random.nextFloat() < fallDistance - 0.5
             && entity instanceof LivingEntity
             && (entity instanceof Player || serverLevel.getGameRules().get(GameRules.MOB_GRIEFING))
