--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
@@ -549,15 +_,15 @@
         UseOnContext context = new UseOnContext(world, null, InteractionHand.MAIN_HAND, Items.BONE_MEAL.getDefaultInstance(), new BlockHitResult(Vec3.ZERO, direction, this.getPosition(), false));
 
         // SPIGOT-6895: Call StructureGrowEvent and BlockFertilizeEvent
-        world.captureTreeGeneration = true;
+        world.captureTreeGeneration.set(true); // Canvas - threadlocal
         InteractionResult result = BoneMealItem.applyBonemeal(context);
-        world.captureTreeGeneration = false;
+        world.captureTreeGeneration.set(false); // Canvas - threadlocal
 
-        if (world.capturedBlockStates.size() > 0) {
+        if (world.capturedBlockStates.get().size() > 0) { // Canvas - threadlocal
             TreeType treeType = SaplingBlock.treeType;
             SaplingBlock.treeType = null;
-            List<BlockState> blocks = new ArrayList<>(world.capturedBlockStates.values());
-            world.capturedBlockStates.clear();
+            List<BlockState> blocks = new ArrayList<>(world.capturedBlockStates.get().values()); // Canvas - threadlocal
+            world.capturedBlockStates.get().clear(); // Canvas - threadlocal
             StructureGrowEvent structureEvent = null;
 
             if (treeType != null) {
