--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
@@ -314,7 +_,7 @@
 
     @Override
     public org.bukkit.block.BlockState getState() {
-        return CraftBlockStates.getBlockState(this);
+        return CraftBlockStates.getBlockState(this, io.canvasmc.canvas.Config.INSTANCE.tileEntitySnapshotCreation); // Canvas - don't use snapshots for acquiring blockstate
     }
 
     @Override
