--- a/net/minecraft/core/NonNullList.java
+++ b/net/minecraft/core/NonNullList.java
@@ -12,23 +_,23 @@
     private final @Nullable E defaultValue;
 
     public static <E> NonNullList<E> create() {
-        return new NonNullList<>(Lists.newArrayList(), null);
+        return new NonNullList<>(new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(), null); // Canvas - optimized collection
     }
 
     public static <E> NonNullList<E> createWithCapacity(final int capacity) {
-        return new NonNullList<>(Lists.newArrayListWithCapacity(capacity), null);
+        return new NonNullList<>(new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(capacity), null); // Canvas - optimized collection
     }
 
     public static <E> NonNullList<E> withSize(final int size, final E defaultValue) {
         Objects.requireNonNull(defaultValue);
         Object[] objects = new Object[size];
         Arrays.fill(objects, defaultValue);
-        return new NonNullList<>(Arrays.asList((E[])objects), defaultValue);
+        return new NonNullList<>(new it.unimi.dsi.fastutil.objects.ObjectArrayList<>((E[])objects), defaultValue); // Canvas - optimized collection
     }
 
     @SafeVarargs
     public static <E> NonNullList<E> of(final E defaultValue, final E... values) {
-        return new NonNullList<>(Arrays.asList(values), defaultValue);
+        return new NonNullList<>(new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(values), defaultValue); // Canvas - optimized collection
     }
 
     protected NonNullList(final List<E> list, final @Nullable E defaultValue) {
