From 5fef61e193d31b59e870ef514121ea8263f383b9 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas paperServer File Patches


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 10afaa655979baa4577c6f67c630c661e3f5681a..bc4256828d3ed392f87038c002da882aed7b236d 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -230,7 +230,7 @@ public class CraftWorld extends CraftRegionAccessor implements World {
             if (chunk == null) {
                 continue;
             }
-            size += chunk.blockEntities.size();
+            size += chunk.canvas$getBlockEntitiesCount(); // Canvas - optimize block entity fetching
         }
         return size;
     }
