From 5fef61e193d31b59e870ef514121ea8263f383b9 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas paperServer File Patches


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
index f020f310c4bca49a23de91bea72ccd9941c0b03c..ed7996c8eb8132f2cadcb6b07f09accf139a8680 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
@@ -134,10 +134,13 @@ public class CraftChunk implements Chunk {
         int index = 0;
         ChunkAccess chunk = this.getHandle(ChunkStatus.FULL);
 
-        BlockState[] states = new BlockState[chunk.blockEntities.size()];
-        for (BlockPos pos : chunk.blockEntities.keySet()) {
-            states[index++] = CraftBlock.at(this.level, pos).getState(useSnapshot);
+        // Canvas start - optimize block entity fetching
+        net.minecraft.world.level.block.entity.BlockEntity[] blockEntities = chunk.canvas$getAllBlockEntities();
+        BlockState[] states = new BlockState[blockEntities.length];
+        for (net.minecraft.world.level.block.entity.BlockEntity blockEntity : blockEntities) {
+            states[index++] = CraftBlock.at(this.level, blockEntity.getBlockPos()).getState(useSnapshot);
         }
+        // Canvas end - optimize block entity fetching
 
         return states;
     }
@@ -152,8 +155,11 @@ public class CraftChunk implements Chunk {
         ChunkAccess chunk = this.getHandle(ChunkStatus.FULL);
 
         List<BlockState> states = new ArrayList<>();
-        for (BlockPos pos : chunk.blockEntities.keySet()) {
-            Block block = CraftBlock.at(this.level, pos);
+        // Canvas start - optimize block entity fetching
+        net.minecraft.world.level.block.entity.BlockEntity[] blockEntities = chunk.canvas$getAllBlockEntities();
+        for (net.minecraft.world.level.block.entity.BlockEntity blockEntity : blockEntities) {
+            Block block = CraftBlock.at(this.level, blockEntity.getBlockPos());
+        // Canvas end - optimize block entity fetching
             if (blockPredicate.test(block)) {
                 states.add(block.getState(useSnapshot));
             }
