From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Toffikk <slodkitofik@gmail.com>
Date: Thu, 9 Apr 2026 22:12:23 +0200
Subject: [PATCH] Paper Changes


diff --git a/src/main/java/com/destroystokyo/paper/entity/PaperPathfinder.java b/src/main/java/com/destroystokyo/paper/entity/PaperPathfinder.java
index 25779c87106a35929efa8dc8ffd94dc7e490de6c..0a9da3d25d7f5eb992c45e370c379746692369b4 100644
--- a/src/main/java/com/destroystokyo/paper/entity/PaperPathfinder.java
+++ b/src/main/java/com/destroystokyo/paper/entity/PaperPathfinder.java
@@ -8,9 +8,9 @@ import javax.annotation.Nullable;
 import net.minecraft.world.level.pathfinder.Node;
 import net.minecraft.world.level.pathfinder.Path;
 import org.bukkit.Location;
-import org.bukkit.craftbukkit.entity.CraftLivingEntity;
+import org.bukkit.craftbukkit.entity.CraftEntity;
 import org.bukkit.craftbukkit.util.CraftLocation;
-import org.bukkit.entity.LivingEntity;
+import org.bukkit.entity.Entity;
 import org.bukkit.entity.Mob;
 
 public class PaperPathfinder implements com.destroystokyo.paper.entity.Pathfinder {
@@ -57,9 +57,9 @@ public class PaperPathfinder implements com.destroystokyo.paper.entity.Pathfinde
 
     @Nullable
     @Override
-    public PathResult findPath(LivingEntity target) {
+    public PathResult findPath(Entity target) {
         Preconditions.checkArgument(target != null, "Target can not be null");
-        Path path = this.entity.getNavigation().createPath(((CraftLivingEntity) target).getHandle(), 0);
+        Path path = this.entity.getNavigation().createPath(((CraftEntity) target).getHandle(), 0);
         return path != null ? new PaperPathResult(path) : null;
     }
 
diff --git a/src/main/java/io/papermc/paper/registry/set/NamedRegistryKeySetImpl.java b/src/main/java/io/papermc/paper/registry/set/NamedRegistryKeySetImpl.java
index 8a9d9b44ec3ceb6cc6a7fc97814294a25cedfd9b..e056d9407a1ce95c99df7fdb29cb59f331348b43 100644
--- a/src/main/java/io/papermc/paper/registry/set/NamedRegistryKeySetImpl.java
+++ b/src/main/java/io/papermc/paper/registry/set/NamedRegistryKeySetImpl.java
@@ -2,6 +2,7 @@ package io.papermc.paper.registry.set;
 
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Iterables;
+import io.papermc.paper.adventure.PaperAdventure;
 import io.papermc.paper.registry.PaperRegistries;
 import io.papermc.paper.registry.RegistryAccess;
 import io.papermc.paper.registry.RegistryKey;
@@ -51,7 +52,7 @@ public record NamedRegistryKeySetImpl<T extends Keyed, M>( // TODO remove Keyed
     public @Unmodifiable Collection<T> resolve(final Registry<T> registry) {
         final ImmutableList.Builder<T> builder = ImmutableList.builder();
         for (final Holder<M> holder : this.namedSet) {
-            builder.add(registry.getOrThrow(CraftNamespacedKey.fromMinecraft(((Holder.Reference<?>) holder).key().identifier())));
+            builder.add(registry.getOrThrow(PaperAdventure.asAdventure(((Holder.Reference<?>) holder).key().identifier())));
         }
         return builder.build();
     }
diff --git a/src/main/java/io/papermc/paper/util/CountingOps.java b/src/main/java/io/papermc/paper/util/CountingOps.java
new file mode 100644
index 0000000000000000000000000000000000000000..20026e54b9906367e66e6bf61d88b98c5f4c4412
--- /dev/null
+++ b/src/main/java/io/papermc/paper/util/CountingOps.java
@@ -0,0 +1,308 @@
+package io.papermc.paper.util;
+
+import com.mojang.datafixers.util.Pair;
+import com.mojang.serialization.DataResult;
+import com.mojang.serialization.DynamicOps;
+import com.mojang.serialization.ListBuilder;
+import com.mojang.serialization.MapLike;
+import com.mojang.serialization.RecordBuilder;
+import com.mojang.serialization.RecordBuilder.AbstractUniversalBuilder;
+import io.netty.handler.codec.DecoderException;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.IntStream;
+import java.util.stream.LongStream;
+import java.util.stream.Stream;
+import net.minecraft.nbt.Tag;
+import net.minecraft.util.AbstractListBuilder;
+
+public final class CountingOps implements DynamicOps<CountingOps.Value> {
+    public static final CountingOps INSTANCE = new CountingOps(Tag.MAX_DEPTH);
+
+    private final int maxDepth;
+
+    public CountingOps(final int maxDepth) {
+        this.maxDepth = maxDepth;
+    }
+
+    @Override
+    public <U> U convertTo(final DynamicOps<U> outOps, final Value input) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Value empty() {
+        return Value.EMPTY;
+    }
+
+    @Override
+    public Value emptyMap() {
+        return this.container(Value.Kind.MAP, 0);
+    }
+
+    @Override
+    public Value emptyList() {
+        return this.container(Value.Kind.LIST, 0);
+    }
+
+    @Override
+    public Value createNumeric(final Number value) {
+        return Value.PRIMITIVE;
+    }
+
+    @Override
+    public Value createString(final String value) {
+        return Value.PRIMITIVE;
+    }
+
+    @Override
+    public DataResult<Number> getNumberValue(final Value input) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public DataResult<String> getStringValue(final Value input) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public DataResult<Value> mergeToList(final Value input, final Value value) {
+        final Accumulator accumulator = new Accumulator();
+        if (input.kind == Value.Kind.LIST) {
+            accumulator.acceptDepth(input.depth - 1);
+        } else if (input.kind != Value.Kind.EMPTY) {
+            accumulator.accept(input);
+        }
+        accumulator.accept(value);
+        return DataResult.success(this.container(Value.Kind.LIST, accumulator));
+    }
+
+    @Override
+    public DataResult<Value> mergeToList(final Value input, final List<Value> values) {
+        final Accumulator accumulator = new Accumulator();
+        if (input.kind == Value.Kind.LIST) {
+            accumulator.acceptDepth(input.depth - 1);
+        } else if (input.kind != Value.Kind.EMPTY) {
+            accumulator.accept(input);
+        }
+        values.forEach(accumulator::accept);
+        return DataResult.success(this.container(Value.Kind.LIST, accumulator));
+    }
+
+    @Override
+    public DataResult<Value> mergeToMap(final Value input, final Value key, final Value value) {
+        final Accumulator accumulator = new Accumulator();
+        if (input.kind == Value.Kind.MAP) {
+            accumulator.acceptDepth(input.depth - 1);
+        } else if (input.kind != Value.Kind.EMPTY) {
+            accumulator.accept(input);
+        }
+        accumulator.accept(value);
+        return DataResult.success(this.container(Value.Kind.MAP, accumulator));
+    }
+
+    @Override
+    public DataResult<Value> mergeToMap(final Value input, final Map<Value, Value> values) {
+        final Accumulator accumulator = new Accumulator();
+        if (input.kind == Value.Kind.MAP) {
+            accumulator.acceptDepth(input.depth - 1);
+        } else if (input.kind != Value.Kind.EMPTY) {
+            accumulator.accept(input);
+        }
+        values.forEach((key, value) -> {
+            accumulator.accept(value);
+        });
+        return DataResult.success(this.container(Value.Kind.MAP, accumulator));
+    }
+
+    @Override
+    public DataResult<Value> mergeToMap(final Value input, final MapLike<Value> values) {
+        final Accumulator accumulator = new Accumulator();
+        if (input.kind == Value.Kind.MAP) {
+            accumulator.acceptDepth(input.depth - 1);
+        } else if (input.kind != Value.Kind.EMPTY) {
+            accumulator.accept(input);
+        }
+        values.entries().forEach(entry -> {
+            accumulator.accept(entry.getSecond());
+        });
+        return DataResult.success(this.container(Value.Kind.MAP, accumulator));
+    }
+
+    @Override
+    public DataResult<Stream<Pair<Value, Value>>> getMapValues(final Value input) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public DataResult<Stream<Value>> getStream(final Value input) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Value createMap(final Stream<Pair<Value, Value>> map) {
+        final Accumulator accumulator = new Accumulator();
+        map.forEach(entry -> accumulator.accept(entry.getSecond()));
+        return this.container(Value.Kind.MAP, accumulator);
+    }
+
+    @Override
+    public Value createMap(final Map<Value, Value> map) {
+        final Accumulator accumulator = new Accumulator();
+        map.forEach((key, value) -> {
+            accumulator.accept(value);
+        });
+        return this.container(Value.Kind.MAP, accumulator);
+    }
+
+    @Override
+    public Value createList(final Stream<Value> input) {
+        final Accumulator accumulator = new Accumulator();
+        input.forEach(accumulator::accept);
+        return this.container(Value.Kind.LIST, accumulator);
+    }
+
+    @Override
+    public Value createByteList(final ByteBuffer input) {
+        return Value.PRIMITIVE;
+    }
+
+    @Override
+    public Value createIntList(final IntStream input) {
+        return Value.PRIMITIVE;
+    }
+
+    @Override
+    public Value createLongList(final LongStream input) {
+        return Value.PRIMITIVE;
+    }
+
+    @Override
+    public Value remove(final Value input, final String key) {
+        return input;
+    }
+
+    @Override
+    public RecordBuilder<Value> mapBuilder() {
+        return new MapBuilder(this);
+    }
+
+    @Override
+    public ListBuilder<Value> listBuilder() {
+        return new CountingListBuilder(this);
+    }
+
+    @Override
+    public String toString() {
+        return "Counting";
+    }
+
+    private Value container(final Value.Kind kind, final Accumulator accumulator) {
+        return this.container(kind, accumulator.maxDepth);
+    }
+
+    private Value container(final Value.Kind kind, final int childDepth) {
+        final int depth = childDepth + 1;
+        if (depth > this.maxDepth) {
+            throw new DecoderException("NBT depth exceeded: " + depth + " > " + this.maxDepth);
+        }
+        return new Value(kind, depth);
+    }
+
+    private static final class Accumulator {
+        private int maxDepth;
+
+        private void accept(final Value value) {
+            this.acceptDepth(value.depth);
+        }
+
+        private void acceptDepth(final int depth) {
+            this.maxDepth = Math.max(this.maxDepth, depth);
+        }
+    }
+
+    public static final class Value {
+        private static final Value EMPTY = new Value(Kind.EMPTY, 0);
+        private static final Value PRIMITIVE = new Value(Kind.PRIMITIVE, 0);
+
+        private final Kind kind;
+        private final int depth;
+
+        private Value(final Kind kind, final int depth) {
+            this.kind = kind;
+            this.depth = depth;
+        }
+
+        public int depth() {
+            return this.depth;
+        }
+
+        private enum Kind {
+            EMPTY,
+            PRIMITIVE,
+            LIST,
+            MAP
+        }
+    }
+
+    private static final class CountingListBuilder extends AbstractListBuilder<Value, Accumulator> {
+        private final CountingOps ops;
+
+        private CountingListBuilder(final CountingOps ops) {
+            super(ops);
+            this.ops = ops;
+        }
+
+        @Override
+        protected Accumulator initBuilder() {
+            return new Accumulator();
+        }
+
+        @Override
+        protected Accumulator append(final Accumulator builder, final Value value) {
+            builder.accept(value);
+            return builder;
+        }
+
+        @Override
+        protected DataResult<Value> build(final Accumulator builder, final Value prefix) {
+            if (prefix.kind == Value.Kind.LIST) {
+                builder.acceptDepth(prefix.depth - 1);
+            } else if (prefix.kind != Value.Kind.EMPTY) {
+                builder.accept(prefix);
+            }
+            return DataResult.success(this.ops.container(Value.Kind.LIST, builder));
+        }
+    }
+
+    private static final class MapBuilder extends AbstractUniversalBuilder<Value, Accumulator> {
+        private final CountingOps ops;
+
+        private MapBuilder(final CountingOps ops) {
+            super(ops);
+            this.ops = ops;
+        }
+
+        @Override
+        protected Accumulator initBuilder() {
+            return new Accumulator();
+        }
+
+        @Override
+        protected Accumulator append(final Value key, final Value value, final Accumulator builder) {
+            builder.accept(value);
+            return builder;
+        }
+
+        @Override
+        protected DataResult<Value> build(final Accumulator builder, final Value prefix) {
+            if (prefix.kind == Value.Kind.MAP) {
+                builder.acceptDepth(prefix.depth - 1);
+            } else if (prefix.kind != Value.Kind.EMPTY) {
+                builder.accept(prefix);
+            }
+            return DataResult.success(this.ops.container(Value.Kind.MAP, builder));
+        }
+    }
+}
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 9d4d97fcb0f3cc94e9a263c61cc0e7e58fd322c3..8df0966d4345606c3c8adbe0ef1f3ac8841c24fb 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -80,37 +80,6 @@ public class Main {
                         .ofType(Integer.class)
                         .describedAs("Server size");
 
-                this.acceptsAll(asList("d", "date-format"), "Format of the date to display in the console (for log entries)")
-                        .withRequiredArg()
-                        .ofType(SimpleDateFormat.class)
-                        .describedAs("Log date format");
-
-                this.accepts("log-pattern", "Specifies the log filename pattern")
-                        .withRequiredArg()
-                        .ofType(String.class)
-                        .defaultsTo("server.log")
-                        .describedAs("Log filename");
-
-                this.accepts("log-limit", "Limits the maximum size of the log file (0 = unlimited)")
-                        .withRequiredArg()
-                        .ofType(Integer.class)
-                        .defaultsTo(0)
-                        .describedAs("Max log size");
-
-                this.accepts("log-count", "Specified how many log files to cycle through")
-                        .withRequiredArg()
-                        .ofType(Integer.class)
-                        .defaultsTo(1)
-                        .describedAs("Log count");
-
-                this.accepts("log-append", "Whether to append to the log file")
-                        .withRequiredArg()
-                        .ofType(Boolean.class)
-                        .defaultsTo(true)
-                        .describedAs("Log append");
-
-                this.accepts("log-strip-color", "Strips color codes from log file");
-
                 this.acceptsAll(asList("b", "bukkit-settings"), "File for bukkit settings")
                         .withRequiredArg()
                         .ofType(File.class)
@@ -137,6 +106,8 @@ public class Main {
 
                 this.accepts("demo", "Demo mode");
 
+                this.accepts("bonusChest", "Enable the bonus chest");
+
                 this.accepts("initSettings", "Only create configuration files and then exit"); // SPIGOT-5761: Add initSettings option
 
                 this.acceptsAll(asList("S", "spigot-settings"), "File for spigot settings")
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftCopperGolem.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftCopperGolem.java
index f0aec398dbfdaf074b725439f11a1bc733299e83..958b749ff030626e321a0f13e9b422be23320a7b 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftCopperGolem.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftCopperGolem.java
@@ -3,6 +3,7 @@ package org.bukkit.craftbukkit.entity;
 import com.google.common.base.Preconditions;
 import io.papermc.paper.entity.PaperShearable;
 import io.papermc.paper.world.WeatheringCopperState;
+import net.minecraft.world.entity.animal.golem.CopperGolemState;
 import net.minecraft.world.level.block.WeatheringCopper;
 import org.bukkit.craftbukkit.CraftServer;
 import org.bukkit.entity.CopperGolem;
@@ -31,6 +32,17 @@ public class CraftCopperGolem extends CraftGolem implements CopperGolem, PaperSh
         this.getHandle().setWeatherState(WeatheringCopper.WeatherState.valueOf(state.name()));
     }
 
+    @Override
+    public CopperGolem.State getGolemState() {
+        return State.valueOf(this.getHandle().getState().name());
+    }
+
+    @Override
+    public void setGolemState(final CopperGolem.State state) {
+        Preconditions.checkArgument(state != null, "state cannot be null");
+        this.getHandle().setState(CopperGolemState.valueOf(state.name()));
+    }
+
     @Override
     public Oxidizing getOxidizing() {
         long value = this.getHandle().nextWeatheringTick;
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftAbstractInventoryView.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftAbstractInventoryView.java
index 9dea16f74523b7b4633d7615bf3dfad4d086ab09..307f3703b6dd4732fb88050d98a2a72dcc78d713 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftAbstractInventoryView.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftAbstractInventoryView.java
@@ -53,6 +53,18 @@ public abstract class CraftAbstractInventoryView implements InventoryView {
         Preconditions.checkArgument(rawSlot >= 0, "Negative, non outside slot %s", rawSlot);
         Preconditions.checkArgument(rawSlot < this.countSlots(), "Slot %s greater than inventory slot count", rawSlot);
 
+        // Paper start - Fix crafter slot ID conversions
+        if (this.getType() == InventoryType.CRAFTER) {
+            // Raw slot ID for crafter inventory views is 45 for result slot and 0-8 for crafting grid slots.
+            // Crafter inventory size is 10. Only check 0-8 and 45.
+            if (rawSlot < (this.getTopInventory().getSize() - 1) || rawSlot == 45) {
+                return this.getTopInventory();
+            } else {
+                return this.getBottomInventory();
+            }
+        }
+        // Paper end - Fix crafter slot ID conversions
+
         if (rawSlot < this.getTopInventory().getSize()) {
             return this.getTopInventory();
         } else {
@@ -62,6 +74,43 @@ public abstract class CraftAbstractInventoryView implements InventoryView {
 
     @Override
     public int convertSlot(final int rawSlot) {
+        // Paper start - Fix crafter slot ID conversions
+        // Crafter inventory size is 10, but the view uses non-contiguous raw slot IDs 0-8 (grid) and 45 (result).
+        // The numInTop check and slot number shift lower in this method assume contiguous slot IDs.
+        if (this.getType() == InventoryType.CRAFTER) {
+            /*
+             * Raw Slots:
+             *
+             *       0  1  2
+             *       3  4  5     45
+             *       6  7  8
+             * 9  10 11 12 13 14 15 16 17
+             * 18 19 20 21 22 23 24 25 26
+             * 27 28 29 30 31 32 33 34 35
+             * 36 37 38 39 40 41 42 43 44
+             */
+
+            /*
+             * Converted Slots:
+             *
+             *       0  1  2
+             *       3  4  5     9
+             *       6  7  8
+             * 9  10 11 12 13 14 15 16 17
+             * 18 19 20 21 22 23 24 25 26
+             * 27 28 29 30 31 32 33 34 35
+             * 0  1  2  3  4  5  6  7  8
+             */
+            if (rawSlot == 45) {
+                return 9; // Result
+            } else if (rawSlot >= 36) {
+                return rawSlot - 36; // Quickbar
+            } else {
+                return rawSlot; // Crafting grid or player inventory
+            }
+        }
+        // Paper end - Fix crafter slot ID conversions
+
         int numInTop = this.getTopInventory().getSize();
         // Index from the top inventory as having slots from [0,size]
         if (rawSlot < numInTop) {
@@ -129,7 +178,18 @@ public abstract class CraftAbstractInventoryView implements InventoryView {
     @Override
     public InventoryType.SlotType getSlotType(final int slot) {
         InventoryType.SlotType type = InventoryType.SlotType.CONTAINER;
-        if (slot >= 0 && slot < this.getTopInventory().getSize()) {
+        // Paper start - Fix crafter slot ID conversions
+        if (this.getType() == InventoryType.CRAFTER) {
+            // Crafter inventory size is 10, but the view uses non-contiguous raw slot IDs 0-8 (grid) and 45 (result).
+            if (slot < 0) {
+                type = InventoryType.SlotType.OUTSIDE;
+            } else if (slot == 45) {
+                type = InventoryType.SlotType.RESULT;
+            } else if (slot > 35) {
+                type = InventoryType.SlotType.QUICKBAR;
+            }
+        } else if (slot >= 0 && slot < this.getTopInventory().getSize()) {
+            // Paper end - Fix crafter slot ID conversions
             switch (this.getType()) {
                 case BLAST_FURNACE:
                 case FURNACE:
diff --git a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
index 936b812843dfc1b58e4fb3c41a0f9683258a1a98..764e47b152b966f0e4fba458817473cffb98baf6 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
@@ -14,7 +14,7 @@ import com.mojang.serialization.Dynamic;
 import com.mojang.serialization.JavaOps;
 import com.mojang.serialization.JsonOps;
 import io.papermc.paper.adventure.AdventureCodecs;
-import io.papermc.paper.adventure.PaperAdventure;
+import io.papermc.paper.entity.EntitySerializationFlag;
 import io.papermc.paper.registry.RegistryKey;
 import java.io.File;
 import java.io.IOException;
@@ -28,7 +28,6 @@ import java.util.Map;
 import java.util.Set;
 import java.util.logging.Level;
 import java.util.stream.Stream;
-import io.papermc.paper.entity.EntitySerializationFlag;
 import net.kyori.adventure.text.event.HoverEvent;
 import net.minecraft.SharedConstants;
 import net.minecraft.advancements.AdvancementHolder;
@@ -61,7 +60,6 @@ import org.bukkit.Bukkit;
 import org.bukkit.Keyed;
 import org.bukkit.Material;
 import org.bukkit.NamespacedKey;
-import org.bukkit.Registry;
 import org.bukkit.UnsafeValues;
 import org.bukkit.World;
 import org.bukkit.advancement.Advancement;
@@ -830,7 +828,7 @@ public final class CraftMagicNumbers implements UnsafeValues {
 
     @Override
     public void setBiomeKey(org.bukkit.RegionAccessor accessor, int x, int y, int z, org.bukkit.NamespacedKey biomeKey) {
-        accessor.setBiome(x, y, z, org.bukkit.Registry.BIOME.getOrThrow(biomeKey));
+        accessor.setBiome(x, y, z, io.papermc.paper.registry.RegistryAccess.registryAccess().getRegistry(RegistryKey.BIOME).getOrThrow(biomeKey));
     }
 
     @Override
diff --git a/src/test/java/io/papermc/paper/item/ItemStackDataComponentTest.java b/src/test/java/io/papermc/paper/item/ItemStackDataComponentTest.java
index 8a83468ec2f4cbd2c4fab4f5e075bae90b7ed46e..0e4a75e94df306120486a613192f3b6c47530e44 100644
--- a/src/test/java/io/papermc/paper/item/ItemStackDataComponentTest.java
+++ b/src/test/java/io/papermc/paper/item/ItemStackDataComponentTest.java
@@ -17,10 +17,9 @@ import io.papermc.paper.datacomponent.item.MapItemColor;
 import io.papermc.paper.datacomponent.item.PotDecorations;
 import io.papermc.paper.datacomponent.item.Tool;
 import io.papermc.paper.datacomponent.item.TooltipDisplay;
-import io.papermc.paper.registry.RegistryAccess;
 import io.papermc.paper.registry.RegistryKey;
+import io.papermc.paper.registry.keys.tags.BlockTypeTagKeys;
 import io.papermc.paper.registry.set.RegistrySet;
-import io.papermc.paper.registry.tag.TagKey;
 import java.util.List;
 import java.util.Map;
 import java.util.function.BiConsumer;
@@ -226,7 +225,7 @@ class ItemStackDataComponentTest {
                     TriState.TRUE
                 ),
                 Tool.rule(
-                    RegistryAccess.registryAccess().getRegistry(RegistryKey.BLOCK).getTag(TagKey.create(RegistryKey.BLOCK, NamespacedKey.minecraft("bamboo_blocks"))),
+                    Registry.BLOCK.getTag(BlockTypeTagKeys.BAMBOO_BLOCKS),
                     2F,
                     TriState.TRUE
                 )
