From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Tue, 21 Apr 2026 19:45:56 -0700
Subject: [PATCH] Rebrand


diff --git a/src/main/java/com/destroystokyo/paper/Metrics.java b/src/main/java/com/destroystokyo/paper/Metrics.java
index 9f4d73fd321f6f0cd19fdd1b63068d2ea6c0816b..5421a90b3713e36dc536974d76fc39beb382008c 100644
--- a/src/main/java/com/destroystokyo/paper/Metrics.java
+++ b/src/main/java/com/destroystokyo/paper/Metrics.java
@@ -593,7 +593,7 @@ public class Metrics {
             boolean logFailedRequests = config.getBoolean("logFailedRequests", false);
             // Only start Metrics, if it's enabled in the config
             if (config.getBoolean("enabled", true)) {
-                Metrics metrics = new Metrics("Paper", serverUUID, logFailedRequests, Bukkit.getLogger());
+                Metrics metrics = new Metrics("Canvas", serverUUID, logFailedRequests, Bukkit.getLogger()); // Canvas - Rebrand
 
                 metrics.addCustomChart(new Metrics.SimplePie("minecraft_version", () -> {
                     String minecraftVersion = Bukkit.getVersion();
@@ -607,11 +607,17 @@ public class Metrics {
                 final String implVersion = org.bukkit.craftbukkit.Main.class.getPackage().getImplementationVersion();
                 if (implVersion != null) {
                     final String buildOrHash = implVersion.substring(implVersion.lastIndexOf('-') + 1);
-                    paperVersion = "git-Paper-%s-%s".formatted(Bukkit.getServer().getMinecraftVersion(), buildOrHash);
+                    // Canvas start - new versioning schem for bstats
+                    io.papermc.paper.ServerBuildInfo buildInfo = io.papermc.paper.ServerBuildInfo.buildInfo();
+                    final int num = buildInfo.buildNumber().orElse(-1);
+                    paperVersion = "Canvas-V3-" +
+                        buildInfo.minecraftVersionName() + "-" +
+                        (num != -1 ? num : "DEV");
+                    // Canvas end - new versioning schem for bstats
                 } else {
                     paperVersion = "unknown";
                 }
-                metrics.addCustomChart(new Metrics.SimplePie("paper_version", () -> paperVersion));
+                metrics.addCustomChart(new Metrics.SimplePie("canvas_version", () -> paperVersion)); // Canvas - Rebrand
 
                 metrics.addCustomChart(new Metrics.DrilldownPie("java_version", () -> {
                     Map<String, Map<String, Integer>> map = new HashMap<>();
diff --git a/src/main/java/com/destroystokyo/paper/console/PaperConsole.java b/src/main/java/com/destroystokyo/paper/console/PaperConsole.java
index 6ee39b534b8d992655bc0cef3c299d12cbae0034..0fafeb6a1dae6e9a707e752114889da1a3eeeb95 100644
--- a/src/main/java/com/destroystokyo/paper/console/PaperConsole.java
+++ b/src/main/java/com/destroystokyo/paper/console/PaperConsole.java
@@ -20,7 +20,7 @@ public final class PaperConsole extends SimpleTerminalConsole {
     @Override
     protected LineReader buildReader(LineReaderBuilder builder) {
         builder
-                .appName("Paper")
+                .appName(io.papermc.paper.ServerBuildInfo.buildInfo().brandName()) // Canvas - Rebrand
                 .variable(LineReader.HISTORY_FILE, java.nio.file.Paths.get(".console_history"))
                 .completer(new ConsoleCommandCompleter(this.server))
                 .option(LineReader.Option.COMPLETE_IN_WORD, true);
diff --git a/src/main/java/io/papermc/paper/PaperServerInternalAPIBridge.java b/src/main/java/io/papermc/paper/PaperServerInternalAPIBridge.java
index f4d5123d9488ff066032575de85fde2a1006a118..3fcdea18a3eed08fb7eec2bf81cd52321d120989 100644
--- a/src/main/java/io/papermc/paper/PaperServerInternalAPIBridge.java
+++ b/src/main/java/io/papermc/paper/PaperServerInternalAPIBridge.java
@@ -181,7 +181,7 @@ public class PaperServerInternalAPIBridge implements InternalAPIBridge {
 
     @Override
     public VersionFetcher getVersionFetcher() {
-        return new PaperVersionFetcher();
+        return new io.canvasmc.canvas.util.version.CanvasVersionFetcher(); // Canvas - Rebrand
     }
 
     @Override
diff --git a/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java b/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
index 74ffdc823e66fc5ec027c4b7c462382bcbfe2be2..5681dbb4c3dcf7ffd60db5074c8164740144460a 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
@@ -21,7 +21,11 @@ public record ServerBuildInfoImpl(
     OptionalInt buildNumber,
     Instant buildTime,
     Optional<String> gitBranch,
-    Optional<String> gitCommit
+    // Canvas start - improve version command
+    Optional<String> gitCommit,
+    Optional<String> brandVendor,
+    Optional<String> brandWebsite
+    // Canvas end - improve version command
 ) implements ServerBuildInfo {
     private static final String ATTRIBUTE_BRAND_ID = "Brand-Id";
     private static final String ATTRIBUTE_BRAND_NAME = "Brand-Name";
@@ -29,8 +33,16 @@ public record ServerBuildInfoImpl(
     private static final String ATTRIBUTE_BUILD_NUMBER = "Build-Number";
     private static final String ATTRIBUTE_GIT_BRANCH = "Git-Branch";
     private static final String ATTRIBUTE_GIT_COMMIT = "Git-Commit";
+    private static final String ATTRIBUTE_BRAND_VENDOR = "Brand-Vendor"; // Canvas - improve version command
+    private static final String ATTRIBUTE_BRAND_WEBSITE = "Brand-Website"; // Canvas - improve version command
 
     private static final String BRAND_PAPER_NAME = "Paper";
+    // Canvas start - Rebrand
+    private static final String BRAND_CANVAS_NAME = "Canvas";
+
+    // private static final Key BRAND_CANVAS_ID = Key.key("canvasmc", "canvas");
+    // private static final Key BRAND_FOLIA_ID = Key.key("papermc", "folia");
+    // Canvas end - Rebrand
 
     private static final String BUILD_DEV = "DEV";
 
@@ -42,9 +54,9 @@ public record ServerBuildInfoImpl(
         this(
             getManifestAttribute(manifest, ATTRIBUTE_BRAND_ID)
                 .map(Key::key)
-                .orElse(BRAND_PAPER_ID),
+                .orElse(BRAND_CANVAS_ID), // Canvas - Rebrand
             getManifestAttribute(manifest, ATTRIBUTE_BRAND_NAME)
-                .orElse(BRAND_PAPER_NAME),
+                .orElse(BRAND_CANVAS_NAME), // Canvas - Rebrand
             SharedConstants.getCurrentVersion().id(),
             SharedConstants.getCurrentVersion().name(),
             getManifestAttribute(manifest, ATTRIBUTE_BUILD_NUMBER)
@@ -55,13 +67,17 @@ public record ServerBuildInfoImpl(
                 .map(Instant::parse)
                 .orElse(Main.BOOT_TIME),
             getManifestAttribute(manifest, ATTRIBUTE_GIT_BRANCH),
-            getManifestAttribute(manifest, ATTRIBUTE_GIT_COMMIT)
+            // Canvas start - improve version command
+            getManifestAttribute(manifest, ATTRIBUTE_GIT_COMMIT),
+            getManifestAttribute(manifest, ATTRIBUTE_BRAND_VENDOR),
+            getManifestAttribute(manifest, ATTRIBUTE_BRAND_WEBSITE)
+            // Canvas end - improve version command
         );
     }
 
     @Override
     public boolean isBrandCompatible(final @NotNull Key brandId) {
-        return brandId.equals(this.brandId);
+        return brandId.equals(this.brandId) || brandId.equals(BRAND_PAPER_ID) || brandId.equals(BRAND_FOLIA_ID); // Canvas - Rebrand
     }
 
     @Override
diff --git a/src/main/java/io/papermc/paper/command/PaperVersionCommand.java b/src/main/java/io/papermc/paper/command/PaperVersionCommand.java
index 5f17c295e2158b93ed6f77012c0ac5507ac88284..1df07aa472e63fc045e04b1fa169a61b516a2cd3 100644
--- a/src/main/java/io/papermc/paper/command/PaperVersionCommand.java
+++ b/src/main/java/io/papermc/paper/command/PaperVersionCommand.java
@@ -170,9 +170,13 @@ public class PaperVersionCommand {
                this.versionFetcher.getVersionMessage()
            );
 
+           // Canvas start - improve version command
+           final Component bukkitComponent = message.hoverEvent(Component.translatable("chat.copy.click", NamedTextColor.WHITE))
+               .clickEvent(ClickEvent.copyToClipboard(PlainTextComponentSerializer.plainText().serialize(message)));
+           final Component versionFetcherComponent = this.versionFetcher.getFullOutMessage();
            return new ComputedVersion(
-               message.hoverEvent(Component.translatable("chat.copy.click", NamedTextColor.WHITE))
-                   .clickEvent(ClickEvent.copyToClipboard(PlainTextComponentSerializer.plainText().serialize(message))),
+               versionFetcherComponent == null ? bukkitComponent : versionFetcherComponent,
+           // Canvas end - improve version command
                System.currentTimeMillis()
            );
        });
