From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Tue, 23 Dec 2025 13:04:36 -0800
Subject: [PATCH] Support Spark


diff --git a/src/main/java/io/papermc/paper/SparksFly.java b/src/main/java/io/papermc/paper/SparksFly.java
index b332645ed65928100f580221d8a9948bc77e362e..8cfabe08364d14db1c7c28b814e69f38a8d490a9 100644
--- a/src/main/java/io/papermc/paper/SparksFly.java
+++ b/src/main/java/io/papermc/paper/SparksFly.java
@@ -42,7 +42,7 @@ public final class SparksFly {
         // Folia - region threading
         this.logger = Logger.getLogger(ID);
         this.logger.log(Level.INFO, "This server bundles the spark profiler. For more information please visit https://docs.papermc.io/paper/profiling");
-        this.spark = PaperSparkModule.create(Compatibility.VERSION_1_0, server, this.logger, new PaperScheduler() {
+        this.spark = io.canvasmc.canvas.spark.FoliaSparkPlugin.create(Compatibility.VERSION_1_0, server, this.logger, new PaperScheduler() { // Canvas - support spark
             @Override
             public void executeAsync(final Runnable runnable) {
                 MCUtil.scheduleAsyncTask(this.catching(runnable, "asynchronous"));
@@ -116,7 +116,7 @@ public final class SparksFly {
 
     private void enable() {
         if (!this.enabled) {
-            if (false) { // Folia - disable in-built spark profiler
+            if (GlobalConfiguration.get().spark.enabled) { // Folia - disable in-built spark profiler // Canvas - support spark
                 this.enabled = true;
                 this.spark.enable();
             } else {
@@ -168,7 +168,7 @@ public final class SparksFly {
     }
 
     public static boolean isPluginPreferred() {
-        return true; // Folia - disable in-built spark profiler
+        return Boolean.getBoolean(PREFER_SPARK_PLUGIN_PROPERTY); // Folia - disable in-built spark profiler // Canvas - support spark
     }
 
     private static boolean isPluginEnabled(final Server server) {
@@ -176,7 +176,7 @@ public final class SparksFly {
     }
 
     private static boolean shouldEnableImmediately() {
-        return GlobalConfiguration.get().spark.enableImmediately;
+        return false; // Canvas - support spark
     }
 
     public static final class CommandImpl extends Command {
