From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Tue, 21 Apr 2026 20:23:42 -0700
Subject: [PATCH] Replace Moonrise Executor


diff --git a/src/main/java/ca/spottedleaf/moonrise/common/util/MoonriseCommon.java b/src/main/java/ca/spottedleaf/moonrise/common/util/MoonriseCommon.java
index b9f788693b0c4be48e8a923a3dec50555d135e7e..34e098e0718c213be2517e3edf4ba53f680e2d05 100644
--- a/src/main/java/ca/spottedleaf/moonrise/common/util/MoonriseCommon.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/util/MoonriseCommon.java
@@ -14,30 +14,15 @@ public final class MoonriseCommon {
 
     private static final Logger LOGGER = LogUtils.getClassLogger();
 
-    public static final long WORKER_QUEUE_HOLD_TIME = (long)(20.0e6); // 20ms
-    public static final BalancedPrioritisedThreadPool WORKER_POOL = new BalancedPrioritisedThreadPool(
-        WORKER_QUEUE_HOLD_TIME,
-        new ThreadFactory() {
-            private final AtomicInteger idGenerator = new AtomicInteger();
-
-            @Override
-            public Thread newThread(final Runnable run) {
-                final Thread thread = new Thread(run, PlatformHooks.get().getBrand() + " Common Worker #" + this.idGenerator.getAndIncrement());
-
-                thread.setDaemon(true);
-                thread.setUncaughtExceptionHandler(new Thread.UncaughtExceptionHandler() {
-                    @Override
-                    public void uncaughtException(final Thread thread, final Throwable throwable) {
-                        LOGGER.error("Uncaught exception in thread " + thread.getName(), throwable);
-                    }
-                });
+    // Canvas start - replace moonrise executor
+    public static final long WORKER_QUEUE_HOLD_TIME = (long)(1.0e6); // 1ms
+    public static final long IO_WORKER_QUEUE_HOLD_TIME = (long)(1.0e6); // 1ms
+    public static io.canvasmc.canvas.world.chunk.BalancedChunkSystem WORKER_POOL;
+    public static io.canvasmc.canvas.world.chunk.BalancedChunkSystem IO_POOL;
 
-                return thread;
-            }
-        }
-    );
-    public static final BalancedPrioritisedThreadPool.OrderedStreamGroup CLIENT_GROUP = MoonriseCommon.WORKER_POOL.createOrderedStreamGroup();
-    public static final BalancedPrioritisedThreadPool.OrderedStreamGroup SERVER_GROUP = MoonriseCommon.WORKER_POOL.createOrderedStreamGroup();
+    public static io.canvasmc.canvas.world.chunk.BalancedChunkSystem.OrderedStreamGroup SERVER_GROUP;
+    public static io.canvasmc.canvas.world.chunk.BalancedChunkSystem.OrderedStreamGroup SERVER_IO_GROUP;
+    // Canvas end - replace moonrise executor
 
     public static void adjustWorkerThreads(final int configWorkerThreads, final int configIoThreads) {
         int defaultWorkerThreads = OSNuma.getNativeInstance().getTotalCores()  / 2;
@@ -56,36 +41,54 @@ public final class MoonriseCommon {
 
         final int ioThreads = Math.max(1, configIoThreads);
 
-        WORKER_POOL.adjustThreadCount(workerThreads);
-        IO_POOL.adjustThreadCount(ioThreads);
+    // Canvas start - replace moonrise executor
+        if (WORKER_POOL != null && IO_POOL != null) {
+            // we just need to adjust the thread count, cannot recreate instances
+            WORKER_POOL.adjustThreadCount(workerThreads);
+            IO_POOL.adjustThreadCount(ioThreads);
+        } else {
+            // setup chunk system
+            LOGGER.info("Setting up ls_wg chunk system");
+
+            // build instances
+            WORKER_POOL = new io.canvasmc.canvas.world.chunk.BalancedChunkSystem(
+                WORKER_QUEUE_HOLD_TIME, workerThreads,
+                new java.util.concurrent.ThreadFactory() {
+                    private final AtomicInteger idGenerator = new AtomicInteger();
+
+                    @Override
+                    public Thread newThread(final Runnable run) {
+                        Thread thread = new Thread(run, "ls_wg worker #" + this.idGenerator.getAndIncrement());
+                        thread.setDaemon(true);
+                        thread.setUncaughtExceptionHandler((thread1, throwable) -> LOGGER.error("Uncaught exception in thread {}", thread1.getName(), throwable));
+                        return thread;
+                    }
+                }, "ls_wg"
+            );
+            IO_POOL = new io.canvasmc.canvas.world.chunk.BalancedChunkSystem(
+                IO_WORKER_QUEUE_HOLD_TIME, ioThreads,
+                new java.util.concurrent.ThreadFactory() {
+                    private final AtomicInteger idGenerator = new AtomicInteger();
+
+                    @Override
+                    public Thread newThread(final Runnable run) {
+                        Thread thread = new Thread(run, "ls_wg/io worker #" + this.idGenerator.getAndIncrement());
+                        thread.setDaemon(true);
+                        thread.setUncaughtExceptionHandler((thread1, throwable) -> LOGGER.error("Uncaught exception in thread {}", thread1.getName(), throwable));
+                        return thread;
+                    }
+                }, "ls_wg/io"
+            );
+
+            // setup server groups
+            SERVER_GROUP = WORKER_POOL.createChunkOrderedStreamGroup();
+            SERVER_IO_GROUP = IO_POOL.createChunkOrderedStreamGroup();
+        }
 
-        LOGGER.info(PlatformHooks.get().getBrand() + " is using " + workerThreads + " worker threads, " + ioThreads + " I/O threads");
+        LOGGER.info("Running LS ChunkSystem with {} worker threads and {} io threads", workerThreads, ioThreads);
     }
 
-    public static final long IO_QUEUE_HOLD_TIME = (long)(25.0e6); // 25ms
-    public static final BalancedPrioritisedThreadPool IO_POOL = new BalancedPrioritisedThreadPool(
-        IO_QUEUE_HOLD_TIME,
-        new ThreadFactory() {
-                private final AtomicInteger idGenerator = new AtomicInteger();
-
-                @Override
-                public Thread newThread(final Runnable run) {
-                    final Thread thread = new Thread(run, PlatformHooks.get().getBrand() + " I/O Worker #" + this.idGenerator.getAndIncrement());
-
-                    thread.setDaemon(true);
-                    thread.setUncaughtExceptionHandler(new Thread.UncaughtExceptionHandler() {
-                        @Override
-                        public void uncaughtException(final Thread thread, final Throwable throwable) {
-                            LOGGER.error("Uncaught exception in thread " + thread.getName(), throwable);
-                        }
-                    });
-
-                    return thread;
-                }
-            }
-    );
-    public static final BalancedPrioritisedThreadPool.OrderedStreamGroup CLIENT_IO_GROUP = IO_POOL.createOrderedStreamGroup();
-    public static final BalancedPrioritisedThreadPool.OrderedStreamGroup SERVER_IO_GROUP = IO_POOL.createOrderedStreamGroup();
+    // Canvas end - replace moonrise executor
 
     public static void haltExecutors() {
         MoonriseCommon.WORKER_POOL.shutdown(false);
