From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Tue, 21 Apr 2026 20:26:25 -0700
Subject: [PATCH] Send state to player on region update


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 6b4fc8e96d9b8659c5c99eadbaf5c5fd76d3763d..1ffa28fb580204c13ba3d4e3a8329a32c40ea366 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -261,6 +261,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         this.lastRegion = region;
         if (region != null && this.packetProcessor.hasPackets()) {
             region.regionData.setHasPackets();
+            region.regionData.getRegionSchedulingHandle().getTickManager().sendStateToPlayer(getHandle()); // Canvas - region threading
         }
     }
 
