From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Thu, 9 Apr 2026 16:57:55 -0700
Subject: [PATCH] Send state to player on region update


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 10e717a7dfe5a4659f029aa90268a8a62a96f4b5..6cfcb96dc98e50c0f49809b0686db02ba34dd774 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -259,6 +259,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         this.lastRegion = region;
         if (region != null && this.packetProcessor.hasPackets()) {
             region.regionData.setHasPackets();
+            region.regionData.getRegionSchedulingHandle().getTickManager().sendStateToPlayer(getHandle()); // Canvas - region threading
         }
     }
 
