From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Tue, 21 Apr 2026 20:26:56 -0700
Subject: [PATCH] Fix End Credits


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 8571de3671f64582d4800bb33e2e9474c71488f6..8beec4e880622c68ffe30c147ea5b019b2c2578c 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -270,6 +270,13 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
     }
 
     public <T extends net.minecraft.network.PacketListener> void schedulePacket(T listener, net.minecraft.network.protocol.Packet<T> packet) {
+        // Canvas start - region threading
+        if (this.getHandle().canvas$isDisplayingEndCredits) {
+            // if displaying end credits, we want to run this on the global tick
+            io.papermc.paper.threadedregions.RegionizedServer.getInstance().schedulePacket(listener, packet);
+            return;
+        }
+        // Canvas end - region threading
         if (!this.packetProcessor.scheduleIfPossible(listener, packet)) {
             return;
         }
