From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Thu, 9 Apr 2026 16:36:21 -0700
Subject: [PATCH] Fix End Credits


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index b7d227e405336febc5695409b6a4cc7029a2528f..6864b9e3b4f569e0e2b5eef1f0d0cb21dd9164bb 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -264,6 +264,13 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
     }
 
     public <T extends net.minecraft.network.PacketListener> void schedulePacket(T listener, net.minecraft.network.protocol.Packet<T> packet) {
+        // Canvas start - region threading
+        if (this.getHandle().canvas$isDisplayingEndCredits) {
+            // if displaying end credits, we want to run this on the global tick
+            io.papermc.paper.threadedregions.RegionizedServer.getInstance().schedulePacket(listener, packet);
+            return;
+        }
+        // Canvas end - region threading
         if (!this.packetProcessor.scheduleIfPossible(listener, packet)) {
             return;
         }
