From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Sat, 16 May 2026 18:14:42 -0700
Subject: [PATCH] Restore Scoreboards


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index f7922eea2431a2726ed539251254d92b46ec5ae9..ec37cc9d6c082df340a6e15db0f110e1414c8524 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -2585,6 +2585,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
     }
 
     public void setScoreboardOverride(@Nullable CraftScoreboard scoreboardOverride) {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(getHandle(), "Cannot set player board override async"); // Canvas - region threading
         this.scoreboardOverride = scoreboardOverride;
     }
 
@@ -2595,6 +2596,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player, PluginMessa
         if (!(scoreboard instanceof CraftScoreboard craftScoreboard)) {
             throw new IllegalArgumentException("Cannot set player scoreboard to an unregistered Scoreboard");
         }
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(getHandle(), "Cannot set player board async"); // Canvas - region threading
 
         this.server.getScoreboardManager().setPlayerBoard(this, craftScoreboard);
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftObjective.java b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftObjective.java
index f1a8529bac83de33d4a27bde60b83877119443d7..1ffc76e5da3d982d4750b825fc1de6f2402f62c4 100644
--- a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftObjective.java
+++ b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftObjective.java
@@ -86,6 +86,7 @@ final class CraftObjective extends CraftScoreboardComponent implements Objective
     @Override
     public void setDisplaySlot(DisplaySlot slot) {
         this.checkState();
+        this.checkMainThread("Cannot set display slot async"); // Canvas - region threading
 
         Scoreboard board = this.getScoreboard().getHandle();
 
@@ -201,6 +202,12 @@ final class CraftObjective extends CraftScoreboardComponent implements Objective
     void checkState() {
         Preconditions.checkState(this.getScoreboard().getHandle().getObjective(this.objective.getName()) != null, "Unregistered scoreboard component");
     }
+    // Canvas start - region threading
+
+    void checkMainThread(final String reason) {
+        this.objective.getScoreboard().ensureMainThread(reason);
+    }
+    // Canvas end - region threading
 
     @Override
     public int hashCode() {
diff --git a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScoreboard.java b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScoreboard.java
index b959fe4311fc6cbae3dcc052ef4710f8c8aefec6..674a2f56d6f72be753a55df153eb1c1b1516ee94 100644
--- a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScoreboard.java
+++ b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScoreboard.java
@@ -32,7 +32,6 @@ public final class CraftScoreboard implements org.bukkit.scoreboard.Scoreboard {
 
     @Override
     public Objective registerNewObjective(String name, Criteria criteria, net.kyori.adventure.text.Component displayName, RenderType renderType) throws IllegalArgumentException {
-        if (true) throw new UnsupportedOperationException(); // Folia - not supported yet
         if (displayName == null) {
             displayName = net.kyori.adventure.text.Component.empty();
         }
@@ -41,6 +40,7 @@ public final class CraftScoreboard implements org.bukkit.scoreboard.Scoreboard {
         Preconditions.checkArgument(renderType != null, "RenderType cannot be null");
         Preconditions.checkArgument(name.length() <= Short.MAX_VALUE, "The name '%s' is longer than the limit of 32767 characters (%s)", name, name.length());
         Preconditions.checkArgument(this.getHandle().getObjective(name) == null, "An objective of name '%s' already exists", name);
+        this.getHandle().ensureMainThread("Cannot register new objective async"); // Canvas - region threading
         // Paper start - lazily track plugin scoreboards
         if (((CraftCriteria) criteria).criteria != net.minecraft.world.scores.criteria.ObjectiveCriteria.DUMMY && !this.registeredGlobally) {
             net.minecraft.server.MinecraftServer.getServer().server.getScoreboardManager().registerScoreboardForVanilla(this);
@@ -178,10 +178,10 @@ public final class CraftScoreboard implements org.bukkit.scoreboard.Scoreboard {
 
     @Override
     public Team registerNewTeam(String name) {
-        if (true) throw new UnsupportedOperationException(); // Folia - not supported yet
         Preconditions.checkArgument(name != null, "Team name cannot be null");
         Preconditions.checkArgument(name.length() <= Short.MAX_VALUE, "Team name '%s' is longer than the limit of 32767 characters (%s)", name, name.length());
         Preconditions.checkArgument(this.getHandle().getPlayerTeam(name) == null, "Team name '%s' is already in use", name);
+        this.getHandle().ensureMainThread("Cannot register new team async"); // Canvas - region threading
 
         return new CraftTeam(this, this.getHandle().addPlayerTeam(name));
     }
@@ -206,8 +206,8 @@ public final class CraftScoreboard implements org.bukkit.scoreboard.Scoreboard {
 
     @Override
     public void clearSlot(DisplaySlot slot) {
-        if (true) throw new UnsupportedOperationException(); // Folia - not supported yet
         Preconditions.checkArgument(slot != null, "Slot cannot be null");
+        this.getHandle().ensureMainThread("Cannot clear display slot async"); // Canvas - region threading
         this.getHandle().setDisplayObjective(CraftScoreboardTranslations.fromBukkitSlot(slot), null);
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScoreboardManager.java b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScoreboardManager.java
index ab76c7f9bb0a4980a07a4d08c0d3ecadbca9668b..20993546c00c3e75fe327557e5b42319b8e7f453 100644
--- a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScoreboardManager.java
+++ b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScoreboardManager.java
@@ -22,7 +22,7 @@ import org.bukkit.scoreboard.ScoreboardManager;
 public final class CraftScoreboardManager implements ScoreboardManager {
     private final CraftScoreboard mainScoreboard;
     private final MinecraftServer server;
-    private final Collection<CraftScoreboard> scoreboards = new WeakCollection<>();
+    private final Collection<CraftScoreboard> scoreboards = new io.canvasmc.canvas.util.WeakConcurrentCollection<>(); // Canvas - region threading
 
     public CraftScoreboardManager(MinecraftServer server, net.minecraft.world.scores.Scoreboard scoreboard) {
         this.mainScoreboard = new CraftScoreboard(scoreboard);
@@ -38,8 +38,7 @@ public final class CraftScoreboardManager implements ScoreboardManager {
 
     @Override
     public CraftScoreboard getNewScoreboard() {
-        if (true) throw new UnsupportedOperationException(); // Folia - not supported yet
-        org.spigotmc.AsyncCatcher.catchOp("scoreboard creation"); // Spigot
+        this.server.getScoreboard().ensureMainThread("Cannot create new scoreboard async"); // Canvas - region threading
         CraftScoreboard scoreboard = new CraftScoreboard(new ServerScoreboard(this.server));
         if (io.papermc.paper.configuration.GlobalConfiguration.get().scoreboards.trackPluginScoreboards) {
             scoreboard.registeredGlobally = true;
@@ -49,7 +48,7 @@ public final class CraftScoreboardManager implements ScoreboardManager {
     }
 
     public void registerScoreboardForVanilla(CraftScoreboard scoreboard) {
-        org.spigotmc.AsyncCatcher.catchOp("scoreboard registration");
+        server.getScoreboard().ensureMainThread("Cannot register CraftScoreboard async"); // Canvas - region threading
         this.scoreboards.add(scoreboard);
     }
 
@@ -59,7 +58,7 @@ public final class CraftScoreboardManager implements ScoreboardManager {
     }
 
     public void setPlayerBoard(CraftPlayer player, CraftScoreboard scoreboard) {
-        if (true) throw new UnsupportedOperationException(); // Folia - not supported yet
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(player.getHandle(), "Cannot set player board async"); // Canvas - region threading
         net.minecraft.world.scores.Scoreboard oldBoard = this.getPlayerBoard(player).getHandle();
         net.minecraft.world.scores.Scoreboard newBoard = scoreboard.getHandle();
         if (oldBoard == newBoard) {
