From 9cc14a34f9eda110d805d11d2cf52e6b4b265a19 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/server/level/ChunkMap.java b/net/minecraft/server/level/ChunkMap.java
index 6fcf0ee02caf94164c1d9a6e0fe2f813a8cc73a7..72ea750a880c4f8707c0ab82d081780cf8719efa 100644
--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -244,7 +244,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
 
         final int index = entity.getType().getCategory().ordinal();
         final ca.spottedleaf.moonrise.common.list.ReferenceList<ServerPlayer> inRange =
-            this.level.moonrise$getNearbyPlayers().getPlayers(entity.chunkPosition(), ca.spottedleaf.moonrise.common.misc.NearbyPlayers.NearbyMapType.TICK_VIEW_DISTANCE);
+            this.level.moonrise$getNearbyPlayers().getPlayers(entity.chunkPosition(), ca.spottedleaf.moonrise.common.misc.NearbyPlayers.NearbyMapType.SPAWN_RANGE); // Canvas - fix mobs walking outside simulation distance with higher view distance causing mob caps to be ignored for per-player entities
         if (inRange == null) {
             return;
         }
@@ -650,7 +650,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
                 printFuture(holder.getEntityTickingChunkFuture()),
                 this.ticketStorage.getTicketDebugString(posKey, false),
                 this.anyPlayerCloseEnoughForSpawning(pos),
-                fullChunk.<Integer>map(c -> c.getBlockEntities().size()).orElse(0),
+                fullChunk.<Integer>map(c -> c.canvas$getAllBlockEntities().length).orElse(0), // Canvas - optimize block entity fetching
                 this.ticketStorage.getTicketDebugString(posKey, true),
                 this.distanceManager.getChunkLevel(posKey, true),
                 fullChunk.<Integer>map(levelChunk -> levelChunk.getBlockTicks().count()).orElse(0),
@@ -1349,7 +1349,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
                 }
                 // Folia end - region threading
                 // CraftBukkit start - respect vanish API
-                if (visibleToPlayer && (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(player) || !player.getBukkitEntity().canSee(this.entity.getBukkitEntity()))) { // Paper - only consider hits // Folia - region threading
+                if (visibleToPlayer && (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(player) || !player.getBukkitEntity().canSeeChunkMapUpdatePlayer(this.entity.getBukkitEntity()))) { // Paper - only consider hits // Folia - region threading // Leaf - optimize canSee checks
                     visibleToPlayer = false;
                 }
                 // CraftBukkit end
