--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -244,7 +_,7 @@
 
         final int index = entity.getType().getCategory().ordinal();
         final ca.spottedleaf.moonrise.common.list.ReferenceList<ServerPlayer> inRange =
-            this.level.moonrise$getNearbyPlayers().getPlayers(entity.chunkPosition(), ca.spottedleaf.moonrise.common.misc.NearbyPlayers.NearbyMapType.TICK_VIEW_DISTANCE);
+            this.level.moonrise$getNearbyPlayers().getPlayers(entity.chunkPosition(), ca.spottedleaf.moonrise.common.misc.NearbyPlayers.NearbyMapType.SPAWN_RANGE); // Canvas - fix mobs walking outside simulation distance with higher view distance causing mob caps to be ignored for per-player entities
         if (inRange == null) {
             return;
         }
@@ -650,7 +_,7 @@
                 printFuture(holder.getEntityTickingChunkFuture()),
                 this.ticketStorage.getTicketDebugString(posKey, false),
                 this.anyPlayerCloseEnoughForSpawning(pos),
-                fullChunk.<Integer>map(c -> c.getBlockEntities().size()).orElse(0),
+                fullChunk.<Integer>map(LevelChunk::canvas$getBlockEntitiesCount).orElse(0), // Canvas - optimize block entity fetching
                 this.ticketStorage.getTicketDebugString(posKey, true),
                 this.distanceManager.getChunkLevel(posKey, true),
                 fullChunk.<Integer>map(levelChunk -> levelChunk.getBlockTicks().count()).orElse(0),
@@ -1327,7 +_,7 @@
                 double distanceSquared = deltaToPlayerX * deltaToPlayerX + deltaToPlayerZ * deltaToPlayerZ; // Paper
                 double rangeSquared = visibleRange * visibleRange;
                 // Paper start - Configurable entity tracking range by Y
-                boolean visibleToPlayer = distanceSquared <= rangeSquared;
+                boolean visibleToPlayer = distanceSquared <= rangeSquared && (!(this.entity instanceof net.minecraft.server.level.ServerPlayer) || !level.canvasConfig().visuals.dontTrackPlayersInEntityTracking); // Canvas - allow disabling player viewing eachother in worlds
                 if (visibleToPlayer && level.paperConfig().entities.trackingRangeY.enabled) {
                     double rangeY = level.paperConfig().entities.trackingRangeY.get(this.entity, -1);
                     if (rangeY != -1) {
@@ -1345,7 +_,7 @@
                 }
                 // Folia end - region threading
                 // CraftBukkit start - respect vanish API
-                if (visibleToPlayer && (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(player) || !player.getBukkitEntity().canSee(this.entity.getBukkitEntity()))) { // Paper - only consider hits // Folia - region threading
+                if (visibleToPlayer && (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(player) || !player.getBukkitEntity().canSeeChunkMapUpdatePlayer(this.entity.getBukkitEntity()))) { // Paper - only consider hits // Folia - region threading // Leaf - optimize canSee checks
                     visibleToPlayer = false;
                 }
                 // CraftBukkit end
