--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -267,13 +_,12 @@
     }
 
     // Paper start - per player mob count backoff
-    public void updateFailurePlayerMobTypeMap(int chunkX, int chunkZ, net.minecraft.world.entity.MobCategory mobCategory) {
+    public void updateFailurePlayerMobTypeMap(net.minecraft.world.entity.MobCategory mobCategory, ca.spottedleaf.moonrise.common.list.ReferenceList<net.minecraft.server.level.ServerPlayer> inRange) { // Canvas - optimize natural spawning
         if (!this.level.paperConfig().entities.spawning.perPlayerMobSpawns) {
             return;
         }
         int idx = mobCategory.ordinal();
-        final ca.spottedleaf.moonrise.common.list.ReferenceList<ServerPlayer> inRange =
-            this.level.moonrise$getNearbyPlayers().getPlayersByChunk(chunkX, chunkZ, ca.spottedleaf.moonrise.common.misc.NearbyPlayers.NearbyMapType.TICK_VIEW_DISTANCE);
+        // Canvas - optimize natural spawning
         if (inRange == null) {
             return;
         }
@@ -664,7 +_,7 @@
                 printFuture(chunkHolder.getEntityTickingChunkFuture()),
                 this.ticketStorage.getTicketDebugString(longKey, false),
                 this.anyPlayerCloseEnoughForSpawning(chunkPos),
-                optional1.<Integer>map(chunk -> chunk.getBlockEntities().size()).orElse(0),
+                optional1.<Integer>map(chunk -> chunk.canvas$getAllBlockEntities().length).orElse(0), // Canvas - optimize block entity fetching
                 this.ticketStorage.getTicketDebugString(longKey, true),
                 this.distanceManager.getChunkLevel(longKey, true),
                 optional1.<Integer>map(chunk -> chunk.getBlockTicks().count()).orElse(0),
@@ -1277,7 +_,7 @@
                 double d1 = vec3_dx * vec3_dx + vec3_dz * vec3_dz; // Paper
                 double d2 = d * d;
                 // Paper start - Configurable entity tracking range by Y
-                boolean flag = d1 <= d2;
+                boolean flag = d1 <= d2 && !entity.isCulled(player); // Canvas - entity culling
                 if (flag && level.paperConfig().entities.trackingRangeY.enabled) {
                     double rangeY = level.paperConfig().entities.trackingRangeY.get(this.entity, -1);
                     if (rangeY != -1) {
@@ -1293,7 +_,7 @@
                 }
                 // Folia end - region threading
                 // CraftBukkit start - respect vanish API
-                if (flag && (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(player) || !player.getBukkitEntity().canSee(this.entity.getBukkitEntity()))) { // Paper - only consider hits // Folia - region threading
+                if (flag && (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(player) || !player.getBukkitEntity().canSeeChunkMapUpdatePlayer(this.entity.getBukkitEntity()))) { // Paper - only consider hits // Folia - region threading // Leaf - optimize canSee checks
                     flag = false;
                 }
                 // CraftBukkit end
