--- a/net/minecraft/server/level/DistanceManager.java
+++ b/net/minecraft/server/level/DistanceManager.java
@@ -81,7 +_,7 @@
     }
     // Paper end - chunk tick iteration optimisation
 
-    protected void purgeStaleTickets() {
+    public void purgeStaleTickets() { // Canvas - protected -> public
         this.moonrise$getChunkHolderManager().tick(); // Paper - rewrite chunk system
     }
 
@@ -178,15 +_,13 @@
 
     public boolean inEntityTickingRange(long chunkPos) {
         // Paper start - rewrite chunk system
-        final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder chunkHolder = this.moonrise$getChunkHolderManager().getChunkHolder(chunkPos);
-        return chunkHolder != null && chunkHolder.isEntityTickingReady();
+        return this.moonrise$getChunkHolderManager().entityTickingChunkHolders.containsLong(chunkPos); // Canvas - faster impl
         // Paper end - rewrite chunk system
     }
 
     public boolean inBlockTickingRange(long chunkPos) {
         // Paper start - rewrite chunk system
-        final ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder chunkHolder = this.moonrise$getChunkHolderManager().getChunkHolder(chunkPos);
-        return chunkHolder != null && chunkHolder.isTickingReady();
+        return this.moonrise$getChunkHolderManager().blockTickingChunkHolders.containsLong(chunkPos); // Canvas - faster impl
         // Paper end - rewrite chunk system
     }
 
