--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -455,6 +_,8 @@
     public boolean sentListPacket = false;
     public boolean suppressTrackerForLogin = false; // Paper - Fire PlayerJoinEvent when Player is actually ready
     // CraftBukkit end
+    public final io.canvasmc.canvas.world.RegionizedTpsBar.DisplayManager canvas$tpsBarDisplay = io.canvasmc.canvas.world.RegionizedTpsBar.DisplayManager.createNew(this); // Canvas - tpsbar
+    public final int canvas$infoBucketIndex; // Canvas - optimize playerlist tick
     public boolean isRealPlayer; // Paper
     public com.destroystokyo.paper.event.entity.@Nullable PlayerNaturallySpawnCreaturesEvent playerNaturallySpawnedEvent; // Paper - PlayerNaturallySpawnCreaturesEvent
     public org.bukkit.event.player.PlayerQuitEvent.@Nullable QuitReason quitReason = null; // Paper - Add API for quit reason; there are a lot of changes to do if we change all methods leading to the event
@@ -506,6 +_,7 @@
         this.bukkitPickUpLoot = true;
         this.maxHealthCache = this.getMaxHealth();
         // CraftBukkit end
+        this.canvas$infoBucketIndex = Math.floorMod(this.uuid.hashCode(), net.minecraft.server.players.PlayerList.SEND_PLAYER_INFO_INTERVAL); // Canvas - optimize playerlist tick
     }
 
     // Folia start - region threading
@@ -528,8 +_,8 @@
         double amountX = selectMaxX - selectMinX;
         double amountZ = selectMaxZ - selectMinZ;
 
-        int selectX = amountX < 1.0 ? Mth.floor(worldBorder.getCenterX()) : (int)Mth.floor((amountX + 1.0) * random.nextDouble() + selectMinX);
-        int selectZ = amountZ < 1.0 ? Mth.floor(worldBorder.getCenterZ()) : (int)Mth.floor((amountZ + 1.0) * random.nextDouble() + selectMinZ);
+        int selectX = amountX < 0.0 ? Mth.floor(worldBorder.getCenterX()) : Mth.floor(amountX * random.nextDouble() + selectMinX); // Luminol - correct respawn place
+        int selectZ = amountZ < 0.0 ? Mth.floor(worldBorder.getCenterZ()) : Mth.floor(amountZ * random.nextDouble() + selectMinZ); // Luminol - correct respawn place
 
         return new BlockPos(selectX, 0, selectZ);
     }
@@ -540,10 +_,21 @@
     }
 
     private static BlockPos findSpawnAround(ServerLevel world, BlockPos selected) {
+        // Luminol start - Correct player respawn place
+        BlockPos inChunk;
+        inChunk = PlayerSpawnFinder.getOverworldRespawnPos(world, selected.getX(), selected.getZ());
+        if (inChunk != null) {
+            AABB checkVolume = PlayerSpawnFinder.PLAYER_DIMENSIONS.makeBoundingBox((double)inChunk.getX() + 0.5, (double)inChunk.getY(), (double)inChunk.getZ() + 0.5);
+
+            if (world.noCollision(null, checkVolume, true)) {
+                return inChunk;
+            }
+        }
+        // Luminol end - Correct player respawn place
         // try hard to find, so that we don't attempt another chunk load
         for (int dz = -SPAWN_RADIUS_SELECTION_SEARCH; dz <= SPAWN_RADIUS_SELECTION_SEARCH; ++dz) {
             for (int dx = -SPAWN_RADIUS_SELECTION_SEARCH; dx <= SPAWN_RADIUS_SELECTION_SEARCH; ++dx) {
-                BlockPos inChunk = PlayerSpawnFinder.getOverworldRespawnPos(world, selected.getX() + dx, selected.getZ() + dz);
+                inChunk = PlayerSpawnFinder.getOverworldRespawnPos(world, selected.getX() + dx, selected.getZ() + dz); // Luminol - Correct player respawn place
                 if (inChunk == null) {
                     continue;
                 }
@@ -675,6 +_,7 @@
         this.respawnConfig = input.read("respawn", ServerPlayer.RespawnConfig.CODEC).orElse(null);
         this.spawnExtraParticlesOnFall = input.getBooleanOr("spawn_extra_particles_on_fall", false);
         this.raidOmenPosition = input.read("raid_omen_position", BlockPos.CODEC).orElse(null);
+        this.canvas$tpsBarDisplay.updateFromEntry(input.read("TpsBar", io.canvasmc.canvas.world.RegionizedTpsBar.Entry.CODEC).orElse(io.canvasmc.canvas.world.RegionizedTpsBar.Entry.FALLBACK)); // Canvas - tpsbar
         // Paper start - Expand PlayerGameModeChangeEvent
         this.loadGameTypes(input);
     }
@@ -716,6 +_,7 @@
             output.store("ShoulderEntityRight", CompoundTag.CODEC, this.getShoulderEntityRight());
         }
         this.getBukkitEntity().setExtraData(output); // CraftBukkit
+        output.store("TpsBar", io.canvasmc.canvas.world.RegionizedTpsBar.Entry.CODEC, this.canvas$tpsBarDisplay.serializeDisplay()); // Canvas - tpsbar
     }
 
     private void saveParentVehicle(final ValueOutput playerOutput) {
@@ -1448,6 +_,7 @@
         if (killer != null) {
             this.awardStat(Stats.ENTITY_KILLED_BY.get(killer.getType()));
             killer.awardKillScore(this, source);
+            if (source.is(DamageTypes.WITHER_SKULL)) this.canvas$isDyingFromWitherRose = true; // Canvas - fix Folia#388
             this.createWitherRose(killer);
         }
 
@@ -1720,7 +_,7 @@
         }
         // Canvas end - region threading
         this.respawn((player) -> {
-            CriteriaTriggers.CHANGED_DIMENSION.trigger(player, Level.END, Level.OVERWORLD);
+            CriteriaTriggers.CHANGED_DIMENSION.trigger(player, Level.END, io.canvasmc.canvas.GlobalConfiguration.fetchRespawnDimensionKey()); // Canvas - default respawn dimension config
         }, org.bukkit.event.player.PlayerRespawnEvent.RespawnReason.END_PORTAL, true);
     }
 
@@ -1853,7 +_,7 @@
         // find and modify respawn block state
         if (respawnWorld == null || respawnConfig == null) {
             // default to regular spawn
-            fudgeSpawnLocation(this.server.getLevel(Level.OVERWORLD), spawnPosComplete);
+            fudgeSpawnLocation(this.server.getLevel(io.canvasmc.canvas.GlobalConfiguration.fetchRespawnDimensionKey()), spawnPosComplete); // Canvas - default respawn dimension config
         } else {
             // load chunk for block
             // give at least 1 radius of loaded chunks so that we do not sync load anything
@@ -1874,7 +_,7 @@
                             null, false, com.destroystokyo.paper.event.player.PlayerSetSpawnEvent.Cause.PLAYER_RESPAWN
                         );
                         // default to regular spawn
-                        fudgeSpawnLocation(this.server.getLevel(Level.OVERWORLD), spawnPosComplete);
+                        fudgeSpawnLocation(this.server.getLevel(io.canvasmc.canvas.GlobalConfiguration.fetchRespawnDimensionKey()), spawnPosComplete); // Canvas - default respawn dimension config
                         return;
                     }
 
@@ -2226,7 +_,7 @@
             CriteriaTriggers.CHANGED_DIMENSION.trigger(this, oldKey, newKey);
         }
 
-        if (maindimensionkey == Level.NETHER && maindimensionkey1 == Level.OVERWORLD && this.enteredNetherPosition != null) {
+        if (maindimensionkey == Level.NETHER && maindimensionkey1 == io.canvasmc.canvas.GlobalConfiguration.fetchRespawnDimensionKey() && this.enteredNetherPosition != null) { // Canvas - default respawn dimension config
             // CraftBukkit end
             if (!respawning) CriteriaTriggers.NETHER_TRAVEL.trigger(this, this.enteredNetherPosition); // Canvas - region threading
         }
@@ -2406,7 +_,7 @@
         if (this.spawnExtraParticlesOnFall && onGround && this.fallDistance > 0.0) {
             Vec3 centered = pos.getCenter().add(0.0, 0.5, 0.0);
             int particles = (int)Mth.clamp(50.0 * this.fallDistance, 0.0, 200.0);
-            this.level()
+            if (!level().canvasConfig().visuals.particles.disableFallParticles) this.level() // Canvas - particles config
                 .sendParticles(new BlockParticleOption(ParticleTypes.BLOCK, onState), centered.x, centered.y, centered.z, particles, 0.3F, 0.3F, 0.3F, 0.15F);
             this.spawnExtraParticlesOnFall = false;
         }
@@ -2753,6 +_,7 @@
 
     public void disconnect() {
         this.disconnected = true;
+        if (this.canvas$isTemporarilyBlocking) this.releaseUsingItem(); // Canvas - implement sword blocking
         this.ejectPassengers();
 
         // Paper start - Workaround vehicle not tracking the passenger disconnection dismount
@@ -3642,7 +_,7 @@
     }
 
     public Set<DebugSubscription<?>> debugSubscriptions() {
-        return !this.server.debugSubscribers().hasRequiredPermissions(this) ? Set.of() : this.requestedDebugSubscriptions;
+        return Set.of(); // Canvas - disable debug subscriptions
     }
 
     public record RespawnConfig(LevelData.RespawnData respawnData, boolean forced) {
@@ -3655,7 +_,7 @@
         );
 
         private static ResourceKey<Level> getDimensionOrDefault(final ServerPlayer.@Nullable RespawnConfig respawnConfig) {
-            return respawnConfig != null ? respawnConfig.respawnData().dimension() : Level.OVERWORLD;
+            return respawnConfig != null ? respawnConfig.respawnData().dimension() : io.canvasmc.canvas.GlobalConfiguration.fetchRespawnDimensionKey(); // Canvas - default respawn dimension config
         }
 
         public boolean isSamePosition(final ServerPlayer.@Nullable RespawnConfig other) {
