From a1940d59a7eeb8b7dee60550bc75804e3611cb1d Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/server/level/ServerEntity.java b/net/minecraft/server/level/ServerEntity.java
index ff49ffb122bb46b261b8d16f3f253746a2e4e8be..ea15814b5114dab9651823a26b07e36677bde89e 100644
--- a/net/minecraft/server/level/ServerEntity.java
+++ b/net/minecraft/server/level/ServerEntity.java
@@ -164,7 +164,7 @@ public class ServerEntity {
                 boolean positionChanged = (vec3_dx * vec3_dx + vec3_dy * vec3_dy + vec3_dz * vec3_dz) >= 7.62939453125E-6D;
                 // Paper end - reduce allocation of Vec3D here
                 Packet<ClientGamePacketListener> packet = null;
-                boolean pos = positionChanged || this.tickCount % 60 == 0;
+                // boolean pos = positionChanged || this.tickCount % 60 == 0; // Canvas - reduce useless move packets
                 boolean sentPosition = false;
                 boolean sentRotation = false;
                 long xa = this.positionCodec.encodeX(currentPosition);
@@ -181,21 +181,36 @@ public class ServerEntity {
                     packet = ClientboundEntityPositionSyncPacket.of(this.entity);
                     sentPosition = true;
                     sentRotation = true;
-                } else if ((!pos || !shouldSendRotation) && !(this.entity instanceof AbstractArrow)) {
-                    if (pos) {
-                        packet = new ClientboundMoveEntityPacket.Pos(this.entity.getId(), (short)xa, (short)ya, (short)za, this.entity.onGround());
-                        sentPosition = true;
-                    } else if (shouldSendRotation) {
-                        packet = new ClientboundMoveEntityPacket.Rot(this.entity.getId(), yRotn, xRotn, this.entity.onGround());
-                        sentRotation = true;
-                    }
+                // Canvas start - reduce useless move packets
                 } else {
-                    packet = new ClientboundMoveEntityPacket.PosRot(this.entity.getId(), (short)xa, (short)ya, (short)za, yRotn, xRotn, this.entity.onGround());
-                    sentPosition = true;
-                    sentRotation = true;
+                    if (positionChanged || shouldSendRotation || this.entity instanceof AbstractArrow) {
+                        if (this.entity instanceof AbstractArrow || (positionChanged && shouldSendRotation)) {
+                            packet = new ClientboundMoveEntityPacket.PosRot(
+                                this.entity.getId(), (short)xa, (short)ya, (short)za, yRotn, xRotn, this.entity.onGround()
+                            );
+                            sentPosition = true;
+                            sentRotation = true;
+                        } else if (positionChanged) {
+                            packet = new ClientboundMoveEntityPacket.Pos(
+                                this.entity.getId(), (short)xa, (short)ya, (short)za, this.entity.onGround()
+                            );
+                            sentPosition = true;
+                        } else {
+                            packet = new ClientboundMoveEntityPacket.Rot(
+                                this.entity.getId(), yRotn, xRotn, this.entity.onGround()
+                            );
+                            sentRotation = true;
+                        }
+                    }
+
+                    if (io.canvasmc.canvas.Config.INSTANCE.networking.reduceUselessMovePackets
+                        && canvas$isUselessMoveEntityPacket(packet)) {
+                        packet = null;
+                    }
+                // Canvas end - reduce useless move packets
                 }
 
-                if (this.entity.needsSync || this.trackDelta || this.entity instanceof LivingEntity && ((LivingEntity)this.entity).isFallFlying()) {
+                if (this.entity.needsSync || this.trackDelta || this.entity instanceof LivingEntity && ((LivingEntity)this.entity).isFallFlying() && this.entity.getDeltaMovement() != this.lastSentMovement) { // Canvas - reduce useless move packets
                     Vec3 movement = this.entity.getDeltaMovement();
                     double diff = movement.distanceToSqr(this.lastSentMovement);
                     if (diff > 1.0E-7 || diff > 0.0 && movement.lengthSqr() == 0.0) {
@@ -210,9 +225,41 @@ public class ServerEntity {
                                         )
                                     )
                                 );
-                        } else {
+                        // Canvas start - filter ClientboundSetEntityMotionPacket
+                        } else if (!io.canvasmc.canvas.Config.INSTANCE.networking.filterClientboundSetEntityMotionPacket) {
+                            // sends all of them if not filtering
                             this.synchronizer.sendToTrackingPlayers(new ClientboundSetEntityMotionPacket(this.entity.getId(), this.lastSentMovement));
+                        } else {
+                            /*
+                               Filtering is enabled.
+
+                               Entities with a 360-degree rotation and movement, and item and ender eye entities
+                               get royally screwed over with this change, as it removes the synchronization
+                               of the entity delta movement.
+
+                               This packet, when disabled, can reduce bandwidth by 60%, which is insane, but we do
+                               need to add a rough filter that ensures there is no visual impact on the client
+
+                               The only entities needing this are:
+                               1. Squid, only mobs with 360° range of rotation
+                               2. Item Entities, disabling this causes glitches and often slightly desyncs their position within 0.5 blocks on the client side. But we filter it if
+                                  it hasn't moved, since for some reason it sends those packets anyway???
+                               3. Ender Eyes, they depend on their velocity being set bc the packet sets the velocity on the client, but the client doesn't handle the
+                                    ender eye velocity on its own bc the ender eye velocity is in relation to the stronghold structure which the client doesn't have access to
+                               4. Shulker bullets use random to determine where they are going, so we need to send velocity packets to ensure it remains smooth on the client
+
+                               This filter helps keep unneeded packets reduced, but also ensures that we keep smooth visual gameplay
+                            */
+                            if (
+                                (this.entity instanceof net.minecraft.world.entity.item.ItemEntity && positionChanged) ||
+                                this.entity instanceof net.minecraft.world.entity.projectile.EyeOfEnder ||
+                                this.entity instanceof net.minecraft.world.entity.animal.squid.Squid ||
+                                this.entity instanceof net.minecraft.world.entity.projectile.ShulkerBullet
+                            ) {
+                                this.synchronizer.sendToTrackingPlayers(new ClientboundSetEntityMotionPacket(this.entity.getId(), this.lastSentMovement));
+                            }
                         }
+                        // Canvas end - filter ClientboundSetEntityMotionPacket
                     }
                 }
 
@@ -268,6 +315,15 @@ public class ServerEntity {
             this.synchronizer.sendToTrackingPlayersAndSelf(new ClientboundSetEntityMotionPacket(this.entity));
         }
     }
+    // Canvas start - reduce useless move packets
+
+    private boolean canvas$isUselessMoveEntityPacket(@Nullable Packet<?> packet) {
+        if (packet instanceof ClientboundMoveEntityPacket.Pos posPacket) {
+            return posPacket.getXa() == 0 && posPacket.getYa() == 0 && posPacket.getZa() == 0;
+        }
+        return false;
+    }
+    // Canvas end - reduce useless move packets
 
     private void handleMinecartPosRot(final NewMinecartBehavior newMinecartBehavior, final byte yRotn, final byte xRotn, final boolean shouldSendRotation) {
         this.sendDirtyEntityData();
