From c8924af9ea84aa98154140092c21d52c00c82d9c Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/server/level/ServerChunkCache.java b/net/minecraft/server/level/ServerChunkCache.java
index ea909ea1143c5fb62afcd302d4ffbd156d457f76..53dc9b9f57d6364779ee26adc06f0b193478c00b 100644
--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -161,7 +161,8 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
 
         // TODO check on update: impl of forEachBlockTickingChunk will only iterate ENTITY ticking chunks!
         // TODO check on update: consumer just runs tickChunk
-        final ca.spottedleaf.moonrise.common.list.ReferenceList<net.minecraft.world.level.chunk.LevelChunk> entityTickingChunks = world.getCurrentWorldData().getEntityTickingChunks(); // Folia - region threading
+        io.papermc.paper.threadedregions.RegionizedWorldData worldData = world.getCurrentWorldData(); // Canvas - optimize random tick
+        final ca.spottedleaf.moonrise.common.list.ReferenceList<net.minecraft.world.level.chunk.LevelChunk> entityTickingChunks = worldData.getEntityTickingChunks(); // Folia - region threading // Canvas - optimize random tick
 
         // note: we can use the backing array here because:
         // 1. we do not care about new additions
@@ -169,9 +170,25 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
         final LevelChunk[] raw = entityTickingChunks.getRawDataUnchecked();
         final int size = entityTickingChunks.size();
 
+        // Canvas start - optimize random tick
+        final net.minecraft.world.level.levelgen.BitRandomSource random = worldData.simpleUnsafeLocalRandom;
+        // don't tick this if we are not raining or thundering
+        if (!world.paperConfig().environment.disableIceAndSnow && (world.isRaining() || world.isThundering())) {
+            int currentIceAndSnowTick = random.nextInt(48 * 16);
+            for (int i = 0; i < size; i++) {
+                currentIceAndSnowTick -= randomTickSpeed;
+                if (currentIceAndSnowTick <= 0) {
+                    currentIceAndSnowTick = random.nextInt(48 * 16);
+                    LevelChunk chunk = raw[i];
+                    ChunkPos pos = chunk.getPos();
+                    world.tickPrecipitation(world.getBlockRandomPos(pos.getMinBlockX(), 0, pos.getMinBlockZ(), 15));
+                }
+            }
+        }
+        // Canvas end - optimize random tick
         java.util.Objects.checkFromToIndex(0, size, raw.length);
         for (int i = 0; i < size; ++i) {
-            world.tickChunk(raw[i], randomTickSpeed);
+            world.tickChunk(raw[i], randomTickSpeed, worldData); // Canvas - optimize random tick
 
             // call mid-tick tasks for chunk system
             if ((i & 7) == 0) {
@@ -539,9 +556,9 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
                 }
                 // Paper end - per player mob spawning backoff
             }
-            spawnCookie = NaturalSpawner.createState(chunkCount, regionizedWorldData.getLoadedEntities(), this::getFullChunk, null, true); // Folia - region threading - note: function only cares about loaded entities, doesn't need all
+            spawnCookie = NaturalSpawner.createState(chunkCount, null, true, regionizedWorldData); // Folia - region threading - note: function only cares about loaded entities, doesn't need all // Canvas - optimize natural spawning
         } else {
-            spawnCookie = NaturalSpawner.createState(chunkCount, regionizedWorldData.getLoadedEntities(), this::getFullChunk, !this.level.paperConfig().entities.spawning.perPlayerMobSpawns ? new LocalMobCapCalculator(this.chunkMap) : null, false); // Folia - region threading - note: function only cares about loaded entities, doesn't need all
+            spawnCookie = NaturalSpawner.createState(chunkCount, !this.level.paperConfig().entities.spawning.perPlayerMobSpawns ? new LocalMobCapCalculator(this.chunkMap) : null, false, regionizedWorldData); // Folia - region threading - note: function only cares about loaded entities, doesn't need all // Canvas - optimize natural spawning
         }
         // Paper end - Optional per player mob spawns
         regionizedWorldData.lastSpawnState = spawnCookie; // Folia - region threading
@@ -555,6 +572,7 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
                 chunkRange = Math.min(chunkRange, 8);
                 player.playerNaturallySpawnedEvent = new com.destroystokyo.paper.event.entity.PlayerNaturallySpawnCreaturesEvent(player.getBukkitEntity(), (byte) chunkRange);
                 player.playerNaturallySpawnedEvent.callEvent();
+                regionizedWorldData.natureSpawnChunkMap.addPlayer(entityPlayer); // Canvas - optimize collect spawning chunks
             }
             // Paper end - PlayerNaturallySpawnCreaturesEvent
             boolean spawnPersistent = this.level.ticksPerSpawnCategory.getLong(org.bukkit.entity.SpawnCategory.ANIMAL) != 0L && this.level.getRedstoneGameTime() % this.level.ticksPerSpawnCategory.getLong(org.bukkit.entity.SpawnCategory.ANIMAL) == 0L; // CraftBukkit // Folia - region threading
@@ -566,16 +584,21 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
         List<LevelChunk> spawningChunks = regionizedWorldData.temporaryChunkTickList; // Folia - region threading
 
         try {
-            this.chunkMap.collectSpawningChunks(spawningChunks);
+            // Canvas start - optimize collect spawning chunks
+            regionizedWorldData.natureSpawnChunkMap.build();
+            regionizedWorldData.natureSpawnChunkMap.collectSpawningChunks(regionizedWorldData.getEntityTickingChunks(), list);
+            // this.chunkMap.collectSpawningChunks(spawningChunks);
+            // Canvas end - optimize collect spawning chunks
             // Paper start - chunk tick iteration optimisation
-            this.shuffleRandom.setSeed(this.level.getRandom().nextLong());
-            if (!this.level.paperConfig().entities.spawning.perPlayerMobSpawns) Util.shuffle(spawningChunks, this.level.getRandom()); // Paper - Optional per player mob spawns; do not need this when per-player is enabled
+            // this.shuffleRandom.setSeed(this.level.getRandom().nextLong()); // Canvas - optimize collect spawning chunks
+            // if (!this.level.paperConfig().entities.spawning.perPlayerMobSpawns) Util.shuffle(spawningChunks, this.level.getRandom()); // Paper - Optional per player mob spawns; do not need this when per-player is enabled // Canvas - optimize collect spawning chunks
             // Paper end - chunk tick iteration optimisation
 
             for (LevelChunk chunk : spawningChunks) {
                 this.tickSpawningChunk(chunk, timeDiff, spawningCategories, spawnCookie);
             }
         } finally {
+            regionizedWorldData.natureSpawnChunkMap.clear(); // Canvas - optimize collect spawning chunks
             spawningChunks.clear();
         }
 
