--- a/net/minecraft/server/level/TicketType.java
+++ b/net/minecraft/server/level/TicketType.java
@@ -64,6 +_,7 @@
     public static final TicketType<Long> TELEPORT_HOLD_TICKET = ca.spottedleaf.moonrise.patches.chunk_system.ticket.ChunkSystemTicketType.create("folia:teleport_hold_ticket", Long::compareTo);
     public static final TicketType REGION_SCHEDULER_API_HOLD = register("folia:region_scheduler_api_hold", 0L, false, TicketType.TicketUse.LOADING_AND_SIMULATION);
     // Folia end - region threading
+    public static final TicketType REGION_PROFILING_HOLD = register("canvas:region_profiler_hold", 0L, false, TicketType.TicketUse.LOADING_AND_SIMULATION); // Canvas - rewrite scheduler
 
     public static TicketType register(String name, long timeout, boolean persist, TicketType.TicketUse use) {
         return Registry.register(BuiltInRegistries.TICKET_TYPE, name, new TicketType(timeout, persist, use));
