--- a/net/minecraft/server/level/WorldGenRegion.java
+++ b/net/minecraft/server/level/WorldGenRegion.java
@@ -125,7 +_,7 @@
         this.levelData = level.getLevelData();
         this.random = level.getChunkSource().randomState().getOrCreateRandomFactory(WORLDGEN_REGION_RANDOM).at(this.center.getPos().getWorldPosition());
         this.dimensionType = level.dimensionType();
-        this.biomeManager = new BiomeManager(this, BiomeManager.obfuscateSeed(this.seed));
+        this.biomeManager = new BiomeManager(this, level.worldGenSettings.options().biomeZoomSeed()); // Canvas - cache biome zoom seed
         ChunkPos centerPos = center.getPos();
         this.centerChunkX = centerPos.x();
         this.centerChunkZ = centerPos.z();
