--- a/net/minecraft/world/entity/monster/zombie/Zombie.java
+++ b/net/minecraft/world/entity/monster/zombie/Zombie.java
@@ -342,10 +_,15 @@
                 int yt = y + Mth.nextInt(this.random, 7, 40) * Mth.nextInt(this.random, -1, 1);
                 int zt = z + Mth.nextInt(this.random, 7, 40) * Mth.nextInt(this.random, -1, 1);
                 BlockPos spawnPos = new BlockPos(xt, yt, zt);
+                // Canvas start - prevent reinforcement checks from loading chunks
+                if (this.level().getChunkIfLoadedImmediately(spawnPos.getX() >> 4, spawnPos.getZ() >> 4) == null) {
+                    continue;
+                }
+                // Canvas end - prevent reinforcement checks from loading chunks
                 if (SpawnPlacements.isSpawnPositionOk(type, level, spawnPos)
                     && SpawnPlacements.checkSpawnRules(type, level, EntitySpawnReason.REINFORCEMENT, spawnPos, level.getRandom())) {
                     reinforcement.setPos(xt, yt, zt);
-                    if (!level.hasNearbyAlivePlayerThatAffectsSpawning(xt, yt, zt, 7.0) // Paper - affects spawning api
+                    if (!level.canvas$hasNearbyAlivePlayerThatAffectsSpawningForZombie(xt, yt, zt, 7.0) // Paper - affects spawning api // Canvas - optimize nearby alive player checks for zombies
                         && level.isUnobstructed(reinforcement)
                         && level.noCollision(reinforcement)
                         && (reinforcement.canSpawnInLiquids() || !level.containsAnyLiquid(reinforcement.getBoundingBox()))) {
