From fa0385f66589f695ee335f5beee4acd758b26a48 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/monster/zombie/ZombieVillager.java b/net/minecraft/world/entity/monster/zombie/ZombieVillager.java
index 9bae1d1b35c4b9aa343cedd5a53e05d09db6a786..eb8313f179762200a7029ca5f11b91b35a2ac657 100644
--- a/net/minecraft/world/entity/monster/zombie/ZombieVillager.java
+++ b/net/minecraft/world/entity/monster/zombie/ZombieVillager.java
@@ -280,6 +280,16 @@ public class ZombieVillager extends Zombie implements VillagerDataHolder {
                 if (!this.isSilent()) {
                     level.levelEvent(null, LevelEvent.SOUND_ZOMBIE_CONVERTED, this.blockPosition(), 0);
                 }
+                // Canvas start - MC-200418
+                if (
+                    io.canvasmc.canvas.Config.INSTANCE.fixes.mc200418
+                        && mob.isPassenger()
+                        && mob.getVehicle() instanceof net.minecraft.world.entity.animal.chicken.Chicken
+                        && mob.isBaby())
+                {
+                    mob.removeVehicle();
+                }
+                // Canvas end - MC-200418
                 // CraftBukkit start
             }, org.bukkit.event.entity.EntityTransformEvent.TransformReason.CURED, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.CURED // CraftBukkit
         );
