From 73e3ddbec8cd225e42fdf1e4683ca01a6c53e643 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/monster/zombie/Zombie.java b/net/minecraft/world/entity/monster/zombie/Zombie.java
index 0008aa0ca02fc4ea67b795b3f0a0e9656233ce89..f0bc297169ee6f70e510e877729cc3289b9487ce 100644
--- a/net/minecraft/world/entity/monster/zombie/Zombie.java
+++ b/net/minecraft/world/entity/monster/zombie/Zombie.java
@@ -228,7 +228,7 @@ public class Zombie extends Monster {
                     this.doUnderWaterConversion(serverLevel);
                 }
             } else if (this.convertsInWater()) {
-                if (this.isEyeInFluid(FluidTags.WATER)) {
+                if (this.canvas$isEyeInWater()) { // Canvas - optimize fluidOnEyes
                     this.inWaterTime++;
                     if (this.inWaterTime >= 600) {
                         this.startUnderWaterConversion(300);
@@ -342,10 +342,15 @@ public class Zombie extends Monster {
                     int yt = y + Mth.nextInt(this.random, 7, 40) * Mth.nextInt(this.random, -1, 1);
                     int zt = z + Mth.nextInt(this.random, 7, 40) * Mth.nextInt(this.random, -1, 1);
                     BlockPos spawnPos = new BlockPos(xt, yt, zt);
+                    // Canvas start - prevent reinforcement checks from loading chunks
+                    if (this.level().getChunkIfLoadedImmediately(spawnPos.getX() >> 4, spawnPos.getZ() >> 4) == null) {
+                        continue;
+                    }
+                    // Canvas end - prevent reinforcement checks from loading chunks
                     if (SpawnPlacements.isSpawnPositionOk(type, level, spawnPos)
                         && SpawnPlacements.checkSpawnRules(type, level, EntitySpawnReason.REINFORCEMENT, spawnPos, level.getRandom())) {
                         reinforcement.setPos(xt, yt, zt);
-                        if (!level.hasNearbyAlivePlayerThatAffectsSpawning(xt, yt, zt, 7.0) // Paper - affects spawning api
+                        if (!level.canvas$hasNearbyAlivePlayerThatAffectsSpawningForZombie(xt, yt, zt, 7.0) // Paper - affects spawning api // Canvas - optimize nearby alive player checks for zombies
                             && level.isUnobstructed(reinforcement)
                             && level.noCollision(reinforcement)
                             && (reinforcement.canSpawnInLiquids() || !level.containsAnyLiquid(reinforcement.getBoundingBox()))) {
