--- a/net/minecraft/world/level/block/Blocks.java
+++ b/net/minecraft/world/level/block/Blocks.java
@@ -7052,6 +_,11 @@
     }
 
     private static Block register(String name, BlockBehaviour.Properties properties) {
+        // Canvas start - MC-223153
+        if (io.canvasmc.canvas.Config.INSTANCE.fixes.mc223153 && name.equalsIgnoreCase("raw_copper_block")) {
+            properties = properties.sound(SoundType.COPPER);
+        }
+        // Canvas end - MC-223153
         return register(name, Block::new, properties);
     }
 
