--- a/net/minecraft/world/level/block/CactusBlock.java
+++ b/net/minecraft/world/level/block/CactusBlock.java
@@ -52,25 +_,46 @@
 
     @Override
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
-        BlockPos blockPos = pos.above();
-        if (level.isEmptyBlock(blockPos)) {
+        // Leaf start - decrease chunk/block lookups
+        final int x = pos.getX();
+        final int z = pos.getZ();
+        net.minecraft.world.level.chunk.LevelChunk chunk = level.getChunkIfLoaded(x >> 4, z >> 4);
+        if (chunk == null) return;
+
+        final int y = pos.getY();
+        if (chunk.getBlockState(x, y + 1, z).isAir()) {
+        // Leaf end - decrease chunk/block lookups
             int i = 1;
             int ageValue = state.getValue(AGE);
 
-            while (level.getBlockState(pos.below(i)).is(this)) {
+            while (chunk.getBlockState(pos.below(i)).is(this)) { // Leaf - decrease chunk/block lookups
                 if (++i == level.paperConfig().maxGrowthHeight.cactus && ageValue == 15) { // Paper - Configurable cactus/bamboo/reed growth height
                     return;
                 }
             }
 
-            if (ageValue == 8 && this.canSurvive(this.defaultBlockState(), level, pos.above())) {
+            // Leaf start - decrease chunk/block lookups
+            BlockPos blockPos = null;
+            if (ageValue == 8 && this.canSurvive(this.defaultBlockState(), level, blockPos = pos.above())) {
+            // Leaf end - decrease chunk/block lookups
                 double d = i >= level.paperConfig().maxGrowthHeight.cactus ? 0.25 : 0.1; // Paper - Configurable cactus/bamboo/reed growth height
                 if (random.nextDouble() <= d) {
                     org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, blockPos, Blocks.CACTUS_FLOWER.defaultBlockState(), 3);
                 }
             } else if (ageValue == 15 && i < level.paperConfig().maxGrowthHeight.cactus) { // Paper - Configurable cactus/bamboo/reed growth height
+                // Leaf start - check if the cactus can even survive being placed
+                if (io.canvasmc.canvas.Config.INSTANCE.cactusCheckSurvivalBeforeGrowth && !canSurvive(level, blockPos = pos.above())) { // Leaf - Decrease chunk/block lookups
+                    level.levelEvent(LevelEvent.PARTICLES_DESTROY_BLOCK, blockPos, Block.getId(state));
+                    // We're going to fake the block breaking to match vanilla standards.
+                    for (net.minecraft.world.item.ItemStack drop : Block.getDrops(state, level, pos, null)) { // Use base cactus since we don't place a block
+                        Block.popResource(level, blockPos, drop);
+                    }
+                    level.setBlock(pos, state.setValue(CactusBlock.AGE, 0), Block.UPDATE_NONE);
+                    return;
+                }
+                // Leaf end - check if the cactus can even survive being placed
                 // Paper start
-                if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, blockPos, this.defaultBlockState(), 3)) {
+                if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, blockPos == null ? blockPos = pos.above() : blockPos, this.defaultBlockState(), 3)) { // Leaf - decrease chunk/block lookups
                     return;
                 }
                 // Paper end
@@ -115,6 +_,11 @@
 
     @Override
     protected boolean canSurvive(BlockState state, LevelReader level, BlockPos pos) {
+    // Leaf start - check if the cactus can even survive being placed
+        return canSurvive(level, pos);
+    }
+    protected boolean canSurvive(LevelReader level, BlockPos pos) {
+    // Leaf end - check if the cactus can even survive being placed
         for (Direction direction : Direction.Plane.HORIZONTAL) {
             BlockState blockState = level.getBlockState(pos.relative(direction));
             if (blockState.isSolid() || level.getFluidState(pos.relative(direction)).is(FluidTags.LAVA)) {
