--- a/net/minecraft/world/level/block/FarmBlock.java
+++ b/net/minecraft/world/level/block/FarmBlock.java
@@ -41,7 +_,7 @@
 
     protected FarmBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(MOISTURE, 0));
+        this.registerDefaultState(this.stateDefinition.any().setValue(MOISTURE, io.canvasmc.canvas.Config.INSTANCE.farmlandAlwaysMoist ? MAX_MOISTURE : 0)); // Canvas - farmland always moist config
     }
 
     @Override
@@ -97,7 +_,7 @@
         int moistureValue = state.getValue(MOISTURE);
         if (moistureValue > 0 && level.paperConfig().tickRates.wetFarmland != 1 && (level.paperConfig().tickRates.wetFarmland < 1 || (level.getRedstoneGameTime() + pos.hashCode()) % level.paperConfig().tickRates.wetFarmland != 0)) { return; } // Paper - Configurable random tick rates for blocks // Folia - region threading
         if (moistureValue == 0 && level.paperConfig().tickRates.dryFarmland != 1 && (level.paperConfig().tickRates.dryFarmland < 1 || (level.getRedstoneGameTime() + pos.hashCode()) % level.paperConfig().tickRates.dryFarmland != 0)) { return; } // Paper - Configurable random tick rates for blocks // Folia - region threading
-        if (!isNearWater(level, pos) && !level.isRainingAt(pos.above())) {
+        if (!io.canvasmc.canvas.Config.INSTANCE.farmlandAlwaysMoist && !isNearWater(level, pos) && !level.isRainingAt(pos.above())) { // Canvas - farmland always moist config - if farmland can be moist then we allow moisture changes, otherwise we assume near water
             if (moistureValue > 0) {
                 org.bukkit.craftbukkit.event.CraftEventFactory.handleMoistureChangeEvent(level, pos, state.setValue(MOISTURE, moistureValue - 1), 2); // CraftBukkit
             } else if (!shouldMaintainFarmland(level, pos)) {
@@ -112,6 +_,7 @@
     public void fallOn(Level level, BlockState state, BlockPos pos, Entity entity, double fallDistance) {
         super.fallOn(level, state, pos, entity, fallDistance); // CraftBukkit - moved here as game rules / events shouldn't affect fall damage.
         if (level instanceof ServerLevel serverLevel
+            && !io.canvasmc.canvas.Config.INSTANCE.disableFarmlandTrampling // Canvas - disable farmland trampling config
             && level.random.nextFloat() < fallDistance - 0.5
             && entity instanceof LivingEntity
             && (entity instanceof Player || serverLevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING))
