From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Sat, 18 Jan 2025 20:46:44 -0800
Subject: [PATCH] Allow ignoring moved too quickly when lagging


diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index f3cfb655ebeb2499c812450832ac14e280cf6a41..f15699c1283c1839f6dc0ec83f39b71f129cd85b 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -8,6 +8,7 @@ import com.mojang.brigadier.ParseResults;
 import com.mojang.brigadier.StringReader;
 import com.mojang.brigadier.suggestion.Suggestions;
 import com.mojang.logging.LogUtils;
+import io.canvasmc.canvas.Config;
 import it.unimi.dsi.fastutil.ints.Int2ObjectMaps;
 import it.unimi.dsi.fastutil.ints.Int2ObjectMap.Entry;
 import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
@@ -1438,6 +1439,7 @@ public class ServerGamePacketListenerImpl
                                 if (this.shouldCheckPlayerMovement(isFallFlying)) {
                                     float f2 = isFallFlying ? 300.0F : 100.0F;
                                     if (d7 - d6 > Math.max(f2, Mth.square(org.spigotmc.SpigotConfig.movedTooQuicklyMultiplier * (float) i * speed))) {
+                                        if (!(Config.INSTANCE.ignoreMovedTooQuicklyWhenLagging && player.serverLevel().lagging)) { // Canvas
                                         // CraftBukkit end
                                         // Paper start - Add fail move event
                                         io.papermc.paper.event.player.PlayerFailMoveEvent event = fireFailMove(io.papermc.paper.event.player.PlayerFailMoveEvent.FailReason.MOVED_TOO_QUICKLY,
@@ -1449,6 +1451,7 @@ public class ServerGamePacketListenerImpl
                                             this.teleport(this.player.getX(), this.player.getY(), this.player.getZ(), this.player.getYRot(), this.player.getXRot());
                                             return;
                                         }
+                                        } // Canvas
                                         // Paper end - Add fail move event
                                     }
                                 }
