From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Sun, 19 Jan 2025 00:32:26 -0800
Subject: [PATCH] Mob AI Goal Disable Config


diff --git a/net/minecraft/world/entity/ai/goal/GoalSelector.java b/net/minecraft/world/entity/ai/goal/GoalSelector.java
index 0ed26e4866b36c67ff8628ce2fe421ae69467231..67ebb310230bdcbf4890bd7b884f9601af6df549 100644
--- a/net/minecraft/world/entity/ai/goal/GoalSelector.java
+++ b/net/minecraft/world/entity/ai/goal/GoalSelector.java
@@ -1,6 +1,7 @@
 package net.minecraft.world.entity.ai.goal;
 
 import com.google.common.annotations.VisibleForTesting;
+import io.canvasmc.canvas.Config;
 import it.unimi.dsi.fastutil.objects.ObjectLinkedOpenHashSet;
 import java.util.EnumMap;
 import java.util.EnumSet;
@@ -29,6 +30,11 @@ public class GoalSelector {
     private int curRate; // Paper - EAR 2
 
     public void addGoal(int priority, Goal goal) {
+        // Canvas start
+        if (Config.COMPILED_DISABLED_GOAL_CLASSES.contains(goal.getClass())) {
+            return;
+        }
+        // Canvas end
         this.availableGoals.add(new WrappedGoal(priority, goal));
     }
 
