From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java b/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
index 780b53eab14a087436e5f27c8d646042f4b357c9..dd46216e483a0577fb7d7e3a2182fd25ab2fbc1f 100644
--- a/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
@@ -397,7 +397,7 @@ public class OldMinecartBehavior extends MinecartBehavior {
                 }
             }
         } else {
-            for (Entity entity1 : this.level().getEntities(this.minecart, aabb)) {
+            for (Entity entity1 : this.level().getEntitiesOfClass(AbstractMinecart.class, aabb, entity -> entity != this.minecart)) { // Canvas - optimize entity type predicates
                 if (!this.minecart.hasPassenger(entity1) && entity1.isPushable() && entity1 instanceof AbstractMinecart) {
                     // CraftBukkit start
                     org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
