--- a/net/minecraft/world/ticks/LevelTicks.java
+++ b/net/minecraft/world/ticks/LevelTicks.java
@@ -1,21 +_,23 @@
 package net.minecraft.world.ticks;
 
+import io.netty.util.internal.ConcurrentSet;
 import it.unimi.dsi.fastutil.longs.Long2LongMap;
 import it.unimi.dsi.fastutil.longs.Long2LongMaps;
 import it.unimi.dsi.fastutil.longs.Long2LongOpenHashMap;
 import it.unimi.dsi.fastutil.longs.Long2ObjectMap;
+import it.unimi.dsi.fastutil.longs.Long2ObjectMaps;
 import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap;
 import it.unimi.dsi.fastutil.longs.Long2LongMap.Entry;
 import it.unimi.dsi.fastutil.objects.ObjectIterator;
-import it.unimi.dsi.fastutil.objects.ObjectOpenCustomHashSet;
-import java.util.ArrayDeque;
 import java.util.ArrayList;
 import java.util.Comparator;
 import java.util.List;
 import java.util.LongSummaryStatistics;
-import java.util.PriorityQueue;
 import java.util.Queue;
 import java.util.Set;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.PriorityBlockingQueue;
+import java.util.concurrent.atomic.AtomicInteger;
 import java.util.function.BiConsumer;
 import java.util.function.LongPredicate;
 import java.util.function.Predicate;
@@ -32,17 +_,22 @@
     private static final Comparator<LevelChunkTicks<?>> CONTAINER_DRAIN_ORDER = (levelChunkTicks, levelChunkTicks1) -> ScheduledTick.INTRA_TICK_DRAIN_ORDER
         .compare(levelChunkTicks.peek(), levelChunkTicks1.peek());
     private final LongPredicate tickCheck;
-    private final Long2ObjectMap<LevelChunkTicks<T>> allContainers = new Long2ObjectOpenHashMap<>();
-    private final Long2LongMap nextTickForContainer = Util.make(new Long2LongOpenHashMap(), map -> map.defaultReturnValue(Long.MAX_VALUE));
-    private final Queue<LevelChunkTicks<T>> containersToTick = new PriorityQueue<>(CONTAINER_DRAIN_ORDER);
-    private final Queue<ScheduledTick<T>> toRunThisTick = new ArrayDeque<>();
-    private final List<ScheduledTick<T>> alreadyRunThisTick = new ArrayList<>();
-    private final Set<ScheduledTick<?>> toRunThisTickSet = new ObjectOpenCustomHashSet<>(ScheduledTick.UNIQUE_TICK_HASH);
+    private final Long2ObjectMap<LevelChunkTicks<T>> allContainers = Long2ObjectMaps.synchronize(new Long2ObjectOpenHashMap<>()); // Canvas
+    private final Long2LongMap nextTickForContainer = Util.make(Long2LongMaps.synchronize(new Long2LongOpenHashMap()), map -> map.defaultReturnValue(Long.MAX_VALUE)); // Canvas - synchronize
+    private final Queue<LevelChunkTicks<T>> containersToTick = new PriorityBlockingQueue<>(11, CONTAINER_DRAIN_ORDER); // Canvas
+    private final Queue<ScheduledTick<T>> toRunThisTick = new ConcurrentLinkedQueue<>(); // Canvas
+    private final AtomicInteger toRunThisTickCount = new AtomicInteger(0); // Canvas
+    private final List<ScheduledTick<T>> alreadyRunThisTick = new ArrayList<>(); // Canvas
+    private final Set<ScheduledTick<?>> toRunThisTickSet = new ConcurrentSet<>(); // Canvas
     private final BiConsumer<LevelChunkTicks<T>, ScheduledTick<T>> chunkScheduleUpdater = (levelChunkTicks, scheduledTick) -> {
         if (scheduledTick.equals(levelChunkTicks.peek())) {
             this.updateContainerScheduling(scheduledTick);
         }
     };
+    // Canvas start
+    public int ticksSinceLastFluidTickCall = 0;
+    public int ticksSinceLastBlockTickCall = 0;
+    // Canvas end
 
     public LevelTicks(LongPredicate tickCheck) {
         this.tickCheck = tickCheck;
@@ -53,7 +_,9 @@
         this.allContainers.put(packedChunkPos, chunkTicks);
         ScheduledTick<T> scheduledTick = chunkTicks.peek();
         if (scheduledTick != null) {
+            synchronized (this.nextTickForContainer) { // Canvas
             this.nextTickForContainer.put(packedChunkPos, scheduledTick.triggerTick());
+            } // Canvas
         }
 
         chunkTicks.setOnTickAdded(this.chunkScheduleUpdater);
@@ -62,7 +_,9 @@
     public void removeContainer(ChunkPos chunkPos) {
         long packedChunkPos = chunkPos.toLong();
         LevelChunkTicks<T> levelChunkTicks = this.allContainers.remove(packedChunkPos);
+        synchronized (this.nextTickForContainer) { // Canvas
         this.nextTickForContainer.remove(packedChunkPos);
+        } // Canvas
         if (levelChunkTicks != null) {
             levelChunkTicks.setOnTickAdded(null);
         }
@@ -84,7 +_,7 @@
         profilerFiller.push("collect");
         this.collectTicks(gameTime, maxAllowedTicks, profilerFiller);
         profilerFiller.popPush("run");
-        profilerFiller.incrementCounter("ticksToRun", this.toRunThisTick.size());
+        profilerFiller.incrementCounter("ticksToRun", this.toRunThisTickCount.get()); // Canvas
         this.runCollectedTicks(ticker);
         profilerFiller.popPush("cleanup");
         this.cleanupAfterTick();
@@ -99,6 +_,7 @@
     }
 
     private void sortContainersToTick(long gameTime) {
+        synchronized (this.nextTickForContainer) { // Canvas
         ObjectIterator<Entry> objectIterator = Long2LongMaps.fastIterator(this.nextTickForContainer);
 
         while (objectIterator.hasNext()) {
@@ -122,6 +_,7 @@
                 }
             }
         }
+        } // Canvas
     }
 
     private void drainContainers(long gameTime, int maxAllowedTicks) {
@@ -148,7 +_,9 @@
     }
 
     private void updateContainerScheduling(ScheduledTick<T> tick) {
+        synchronized (this.nextTickForContainer) { // Canvas
         this.nextTickForContainer.put(ChunkPos.asLong(tick.pos()), tick.triggerTick());
+        } // Canvas
     }
 
     private void drainFromCurrentContainer(Queue<LevelChunkTicks<T>> containersToTick, LevelChunkTicks<T> levelChunkTicks, long gameTime, int maxAllowedTicks) {
@@ -171,16 +_,19 @@
     }
 
     private void scheduleForThisTick(ScheduledTick<T> tick) {
+        if (tick == null) return; // Canvas - technically this is nullable.
         this.toRunThisTick.add(tick);
+        this.toRunThisTickCount.incrementAndGet(); // Canvas
     }
 
     private boolean canScheduleMoreTicks(int maxAllowedTicks) {
-        return this.toRunThisTick.size() < maxAllowedTicks;
+        return this.toRunThisTickCount.get() < maxAllowedTicks; // Canvas
     }
 
     private void runCollectedTicks(BiConsumer<BlockPos, T> ticker) {
         while (!this.toRunThisTick.isEmpty()) {
             ScheduledTick<T> scheduledTick = this.toRunThisTick.poll();
+            this.toRunThisTickCount.decrementAndGet(); // Canvas
             if (!this.toRunThisTickSet.isEmpty()) {
                 this.toRunThisTickSet.remove(scheduledTick);
             }
@@ -192,6 +_,7 @@
 
     private void cleanupAfterTick() {
         this.toRunThisTick.clear();
+        this.toRunThisTickCount.set(0); // Canvas
         this.containersToTick.clear();
         this.alreadyRunThisTick.clear();
         this.toRunThisTickSet.clear();
@@ -233,6 +_,7 @@
     }
 
     public void clearArea(BoundingBox area) {
+        synchronized (this.nextTickForContainer) { // Canvas
         Predicate<ScheduledTick<T>> predicate = scheduledTick -> area.isInside(scheduledTick.pos());
         this.forContainersInArea(area, (pos, container) -> {
             ScheduledTick<T> scheduledTick = container.peek();
@@ -248,6 +_,7 @@
         });
         this.alreadyRunThisTick.removeIf(predicate);
         this.toRunThisTick.removeIf(predicate);
+        } // Canvas
     }
 
     public void copyArea(BoundingBox area, Vec3i offset) {
