--- a/net/minecraft/world/ticks/LevelTicks.java
+++ b/net/minecraft/world/ticks/LevelTicks.java
@@ -1,9 +_,15 @@
 package net.minecraft.world.ticks;
 
+import io.canvasmc.canvas.Config;
+import io.canvasmc.canvas.region.ServerRegions;
+import io.canvasmc.canvas.util.fastutil.Long2LongConcurrentHashMap;
+import io.canvasmc.canvas.util.fastutil.Long2ObjectConcurrentHashMap;
+import io.papermc.paper.threadedregions.ThreadedRegionizer;
 import it.unimi.dsi.fastutil.longs.Long2LongMap;
 import it.unimi.dsi.fastutil.longs.Long2LongMaps;
 import it.unimi.dsi.fastutil.longs.Long2LongOpenHashMap;
 import it.unimi.dsi.fastutil.longs.Long2ObjectMap;
+import it.unimi.dsi.fastutil.longs.Long2ObjectMaps;
 import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap;
 import it.unimi.dsi.fastutil.longs.Long2LongMap.Entry;
 import it.unimi.dsi.fastutil.objects.ObjectIterator;
@@ -16,9 +_,11 @@
 import java.util.PriorityQueue;
 import java.util.Queue;
 import java.util.Set;
+import java.util.concurrent.locks.ReentrantLock;
 import java.util.function.BiConsumer;
 import java.util.function.LongPredicate;
 import java.util.function.Predicate;
+import it.unimi.dsi.fastutil.objects.ObjectSets;
 import net.minecraft.Util;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.SectionPos;
@@ -32,23 +_,105 @@
     private static final Comparator<LevelChunkTicks<?>> CONTAINER_DRAIN_ORDER = (levelChunkTicks, levelChunkTicks1) -> ScheduledTick.INTRA_TICK_DRAIN_ORDER
         .compare(levelChunkTicks.peek(), levelChunkTicks1.peek());
     private final LongPredicate tickCheck;
-    private final Long2ObjectMap<LevelChunkTicks<T>> allContainers = new Long2ObjectOpenHashMap<>();
-    private final Long2LongMap nextTickForContainer = Util.make(new Long2LongOpenHashMap(), map -> map.defaultReturnValue(Long.MAX_VALUE));
-    private final Queue<LevelChunkTicks<T>> containersToTick = new PriorityQueue<>(CONTAINER_DRAIN_ORDER);
-    private final Queue<ScheduledTick<T>> toRunThisTick = new ArrayDeque<>();
+    public final Long2ObjectMap<LevelChunkTicks<T>> allContainers = new Long2ObjectConcurrentHashMap<>(16, 0.75F); // Canvas - concurrent - private -> public
+    private final Long2LongMap nextTickForContainer = Util.make(new Long2LongConcurrentHashMap(Long.MAX_VALUE), map -> map.defaultReturnValue(Long.MAX_VALUE)); // Canvas - concurrent
+    private final Queue<LevelChunkTicks<T>> containersToTick = com.google.common.collect.Queues.synchronizedQueue(new PriorityQueue<>(CONTAINER_DRAIN_ORDER)); // Canvas - synchronized
+    private final Queue<ScheduledTick<T>> toRunThisTick = new java.util.concurrent.ConcurrentLinkedDeque<>(); // Canvas - concurrent
     private final List<ScheduledTick<T>> alreadyRunThisTick = new ArrayList<>();
-    private final Set<ScheduledTick<?>> toRunThisTickSet = new ObjectOpenCustomHashSet<>(ScheduledTick.UNIQUE_TICK_HASH);
+    private final Set<ScheduledTick<?>> toRunThisTickSet = ObjectSets.synchronize(new ObjectOpenCustomHashSet<>(ScheduledTick.UNIQUE_TICK_HASH)); // Canvas - synchronized
     private final BiConsumer<LevelChunkTicks<T>, ScheduledTick<T>> chunkScheduleUpdater = (levelChunkTicks, scheduledTick) -> {
         if (scheduledTick.equals(levelChunkTicks.peek())) {
             this.updateContainerScheduling(scheduledTick);
         }
     };
-
-    public LevelTicks(LongPredicate tickCheck) {
+    public int ticksSinceLastBlockTickCall; // Canvas - ticks between config
+    public int ticksSinceLastFluidTickCall; // Canvas - ticks between config
+
+    // Canvas start - Threaded Regions
+    public final net.minecraft.server.level.ServerLevel world;
+    public final boolean isBlock;
+    public final boolean isWorldRegion;
+
+    public void merge(final LevelTicks<T> into, final long tickOffset) {
+        // note: containersToTick, toRunThisTick, alreadyRunThisTick, toRunThisTickSet
+        // are all transient state, only ever non-empty during tick. But merging regions occurs while there
+        // is no tick happening, so we assume they are empty.
+        for (final java.util.Iterator<Long2ObjectMap.Entry<LevelChunkTicks<T>>> iterator =
+             (this.allContainers).long2ObjectEntrySet().iterator();
+             iterator.hasNext();) {
+            final Long2ObjectMap.Entry<LevelChunkTicks<T>> entry = iterator.next();
+            final LevelChunkTicks<T> tickContainer = entry.getValue();
+            tickContainer.offsetTicks(tickOffset);
+            into.allContainers.put(entry.getLongKey(), tickContainer);
+        }
+        for (final java.util.Iterator<Long2LongMap.Entry> iterator = (this.nextTickForContainer).long2LongEntrySet().iterator();
+             iterator.hasNext();) {
+            final Long2LongMap.Entry entry = iterator.next();
+            into.nextTickForContainer.put(entry.getLongKey(), entry.getLongValue() + tickOffset);
+        }
+    }
+
+    public void split(final int chunkToRegionShift,
+                      final it.unimi.dsi.fastutil.longs.Long2ReferenceOpenHashMap<LevelTicks<T>> regionToData) {
+        for (final java.util.Iterator<Long2ObjectMap.Entry<LevelChunkTicks<T>>> iterator =
+             (this.allContainers).long2ObjectEntrySet().iterator();
+             iterator.hasNext();) {
+            final Long2ObjectMap.Entry<LevelChunkTicks<T>> entry = iterator.next();
+
+            final long chunkKey = entry.getLongKey();
+            final int chunkX = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkX(chunkKey);
+            final int chunkZ = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkZ(chunkKey);
+
+            final long regionSectionKey = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(
+                chunkX >> chunkToRegionShift, chunkZ >> chunkToRegionShift
+            );
+            // Should always be non-null, since containers are removed on unload.
+            LevelTicks<T> poss = regionToData.get(regionSectionKey);
+            if (poss == null) continue;
+            poss.allContainers.put(chunkKey, entry.getValue());
+        }
+        for (final java.util.Iterator<Long2LongMap.Entry> iterator = (this.nextTickForContainer).long2LongEntrySet().iterator();
+             iterator.hasNext();) {
+            final Long2LongMap.Entry entry = iterator.next();
+            final long chunkKey = entry.getLongKey();
+            final int chunkX = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkX(chunkKey);
+            final int chunkZ = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkZ(chunkKey);
+
+            final long regionSectionKey = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkKey(
+                chunkX >> chunkToRegionShift, chunkZ >> chunkToRegionShift
+            );
+
+            // Should always be non-null, since containers are removed on unload.
+            LevelTicks<T> poss = regionToData.get(regionSectionKey);
+            if (poss == null) continue;
+            poss.nextTickForContainer.put(chunkKey, entry.getLongValue());
+        }
+    }
+
+    public LevelTicks(LongPredicate tickCheck, net.minecraft.server.level.ServerLevel world, boolean isBlock, boolean isWorldRegion) {
+        this.world = world;
+        this.isBlock = isBlock;
+        this.isWorldRegion = isWorldRegion;
+        // Canvas end
         this.tickCheck = tickCheck;
     }
 
+    private volatile ThreadLocal<Boolean> checkRegionized = ThreadLocal.withInitial(() -> true);
     public void addContainer(ChunkPos chunkPos, LevelChunkTicks<T> chunkTicks) {
+        // Canvas start - Threaded Regions
+        if (Config.INSTANCE.ticking.enableThreadedRegionizing && checkRegionized.get()) {
+            // we must ensure we are regionized properly
+            int chunkX = chunkPos.x;
+            int chunkZ = chunkPos.z;
+            ThreadedRegionizer.ThreadedRegion<ServerRegions.TickRegionData, ServerRegions.TickRegionSectionData> correctRegion = this.world.regioniser.getRegionAtUnsynchronised(chunkX, chunkZ);
+            LevelTicks<T> regionized = (LevelTicks<T>) (this.isBlock ? correctRegion.getData().tickData.getBlockLevelTicks() : correctRegion.getData().tickData.getFluidLevelTicks());
+            regionized.checkRegionized.set(false);
+            regionized.addContainer(chunkPos, chunkTicks);
+            regionized.checkRegionized.set(true);
+            return;
+        }
+        if (this.isWorldRegion && Config.INSTANCE.ticking.enableThreadedRegionizing) throw new RuntimeException("Cannot add container to world region while regionized!");
+        // Canvas end
         long packedChunkPos = chunkPos.toLong();
         this.allContainers.put(packedChunkPos, chunkTicks);
         ScheduledTick<T> scheduledTick = chunkTicks.peek();
@@ -56,10 +_,35 @@
             this.nextTickForContainer.put(packedChunkPos, scheduledTick.triggerTick());
         }
 
-        chunkTicks.setOnTickAdded(this.chunkScheduleUpdater);
+        // Canvas start - Threaded Regions
+        // at this point, we should be properly regionized (if applicable)
+        // so we poll the tick data HERE, as we cannot trust when 'onTickAdded' is called and where
+        final boolean isBlock = this.isBlock;
+        final net.minecraft.server.level.ServerLevel world = this.world;
+        ServerRegions.WorldTickData worldData = ServerRegions.getTickData(world);
+        // make sure the lambda contains no reference to this LevelTicks
+        chunkTicks.setOnTickAdded((LevelChunkTicks<T> levelChunkTicks, ScheduledTick<T> tick) -> {
+            if (tick.equals(levelChunkTicks.peek())) {
+                ((LevelTicks<T>)(isBlock ? worldData.getBlockLevelTicks() : worldData.getFluidLevelTicks())).updateContainerScheduling(tick);
+            }
+        });
+        // Canvas end
     }
 
     public void removeContainer(ChunkPos chunkPos) {
+        // Canvas start - Threaded Regions
+        if (Config.INSTANCE.ticking.enableThreadedRegionizing && checkRegionized.get()) {
+            // we must ensure we are regionized properly
+            int chunkX = chunkPos.x;
+            int chunkZ = chunkPos.z;
+            ThreadedRegionizer.ThreadedRegion<ServerRegions.TickRegionData, ServerRegions.TickRegionSectionData> correctRegion = this.world.regioniser.getRegionAtUnsynchronised(chunkX, chunkZ);
+            LevelTicks<T> regionized = (LevelTicks<T>) (this.isBlock ? correctRegion.getData().tickData.getBlockLevelTicks() : correctRegion.getData().tickData.getFluidLevelTicks());
+            regionized.checkRegionized.set(false);
+            regionized.removeContainer(chunkPos);
+            regionized.checkRegionized.set(true);
+            return;
+        }
+        // Canvas end
         long packedChunkPos = chunkPos.toLong();
         LevelChunkTicks<T> levelChunkTicks = this.allContainers.remove(packedChunkPos);
         this.nextTickForContainer.remove(packedChunkPos);
@@ -71,6 +_,20 @@
     @Override
     public void schedule(ScheduledTick<T> tick) {
         long packedChunkPos = ChunkPos.asLong(tick.pos());
+        // Canvas start - Threaded Regions
+        if (Config.INSTANCE.ticking.enableThreadedRegionizing && checkRegionized.get()) {
+            // we must ensure we are regionized properly
+            ChunkPos pos = new ChunkPos(tick.pos());
+            int chunkX = pos.x;
+            int chunkZ = pos.z;
+            ThreadedRegionizer.ThreadedRegion<ServerRegions.TickRegionData, ServerRegions.TickRegionSectionData> correctRegion = this.world.regioniser.getRegionAtUnsynchronised(chunkX, chunkZ);
+            LevelTicks<T> regionized = (LevelTicks<T>) (this.isBlock ? correctRegion.getData().tickData.getBlockLevelTicks() : correctRegion.getData().tickData.getFluidLevelTicks());
+            regionized.checkRegionized.set(false);
+            regionized.schedule(tick);
+            regionized.checkRegionized.set(true);
+            return;
+        }
+        // Canvas end
         LevelChunkTicks<T> levelChunkTicks = this.allContainers.get(packedChunkPos);
         if (levelChunkTicks == null) {
             Util.logAndPauseIfInIde("Trying to schedule tick in not loaded position " + tick.pos());
