--- a/net/minecraft/world/ticks/ProtoChunkTicks.java
+++ b/net/minecraft/world/ticks/ProtoChunkTicks.java
@@ -7,8 +_,8 @@
 import net.minecraft.core.BlockPos;
 
 public class ProtoChunkTicks<T> implements SerializableTickContainer<T>, TickContainerAccess<T> {
-    private final List<SavedTick<T>> ticks = Lists.newArrayList();
-    private final Set<SavedTick<?>> ticksPerPosition = new ObjectOpenCustomHashSet<>(SavedTick.UNIQUE_TICK_HASH);
+    private final List<SavedTick<T>> ticks = java.util.Collections.synchronizedList(Lists.newArrayList()); // Canvas - synchronized
+    private final Set<SavedTick<?>> ticksPerPosition = it.unimi.dsi.fastutil.objects.ObjectSets.synchronize(new ObjectOpenCustomHashSet<>(SavedTick.UNIQUE_TICK_HASH)); // Canvas - synchronized
 
     @Override
     public void schedule(ScheduledTick<T> tick) {
