--- a/net/minecraft/world/item/EnderEyeItem.java
+++ b/net/minecraft/world/item/EnderEyeItem.java
@@ -109,14 +_,37 @@
         } else {
             player.startUsingItem(hand);
             if (level instanceof ServerLevel serverLevel) {
-                BlockPos blockPos = serverLevel.findNearestMapStructure(StructureTags.EYE_OF_ENDER_LOCATED, player.blockPosition(), 100, false);
-                if (blockPos == null) {
-                    return InteractionResult.CONSUME;
-                }
-
+                // Canvas start - async locator
+                BlockPos blockPos = BlockPos.ZERO;
                 EyeOfEnder eyeOfEnder = new EyeOfEnder(level, player.getX(), player.getY(0.5), player.getZ());
+
+                eyeOfEnder.asyncLocator$locateTaskOngoing = true;
+                io.canvasmc.canvas.chunk.structure.AsyncLocator.locate(
+                    serverLevel,
+                    StructureTags.EYE_OF_ENDER_LOCATED,
+                    player.blockPosition(),
+                    100,
+                    false
+                ).thenOnRegionOrNull(representable -> {
+                    BlockPos pos = representable.get();
+                    eyeOfEnder.asyncLocator$locateTaskOngoing = false;
+                    if (eyeOfEnder.isRemoved()) {
+                        // don't signal to if the eye is removed
+                        return;
+                    }
+                    if (pos != null) {
+                        eyeOfEnder.signalTo(Vec3.atLowerCornerOf(pos));
+                        CriteriaTriggers.USED_ENDER_EYE.trigger((ServerPlayer) player, pos);
+                        player.awardStat(Stats.ITEM_USED.get(this));
+                    }
+                }, () -> {
+                    eyeOfEnder.asyncLocator$locateTaskOngoing = false;
+                    // Set the entity's life to long enough that it dies
+                    eyeOfEnder.life = Integer.MAX_VALUE - 100;
+                });
+                // Canvas end - async locator
                 eyeOfEnder.setItem(itemInHand);
-                eyeOfEnder.signalTo(Vec3.atLowerCornerOf(blockPos));
+                // Canvas - async locator
                 level.gameEvent(GameEvent.PROJECTILE_SHOOT, eyeOfEnder.position(), GameEvent.Context.of(player));
                 // CraftBukkit start
                 if (!level.addFreshEntity(eyeOfEnder)) {
@@ -130,7 +_,7 @@
                 float f = Mth.lerp(level.random.nextFloat(), 0.33F, 0.5F);
                 level.playSound(null, player.getX(), player.getY(), player.getZ(), SoundEvents.ENDER_EYE_LAUNCH, SoundSource.NEUTRAL, 1.0F, f);
                 itemInHand.consume(1, player);
-                player.awardStat(Stats.ITEM_USED.get(this));
+                // Canvas - async locator
             }
 
             return InteractionResult.SUCCESS_SERVER;
