--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -883,11 +_,11 @@
     }
 
     public static boolean isSameItem(ItemStack stack, ItemStack other) {
-        return stack.is(other.getItem());
+        return stack == other || stack.is(other.getItem()); // Canvas - optimize matching checks
     }
 
     public static boolean isSameItemSameComponents(ItemStack stack, ItemStack other) {
-        return stack.is(other.getItem()) && (stack.isEmpty() && other.isEmpty() || Objects.equals(stack.components, other.components));
+        return stack == other || stack.is(other.getItem()) && (stack.isEmpty() && other.isEmpty() || Objects.equals(stack.components, other.components)); // Canvas - optimize matching checks
     }
 
     public static MapCodec<ItemStack> lenientOptionalFieldOf(String fieldName) {
