--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -388,24 +_,24 @@
             ServerLevel serverLevel = (ServerLevel) context.getLevel();
 
             if (!(item instanceof BucketItem/* || item instanceof SolidBucketItem*/)) { // if not bucket // Paper - Fix cancelled powdered snow bucket placement
-                serverLevel.captureBlockStates = true;
+                serverLevel.captureBlockStates.set(true); // Canvas - atomic
                 // special case bonemeal
                 if (item == Items.BONE_MEAL) {
-                    serverLevel.captureTreeGeneration = true;
+                    serverLevel.captureTreeGeneration.set(true); // Canvas - atomic
                 }
             }
             InteractionResult interactionResult;
             try {
                 interactionResult = item.useOn(context);
             } finally {
-                serverLevel.captureBlockStates = false;
+                serverLevel.captureBlockStates.set(false); // Canvas - atomic
             }
             DataComponentPatch newPatch = this.components.asPatch();
             int newCount = this.getCount();
             this.setCount(oldCount);
             this.restorePatch(previousPatch);
-            if (interactionResult.consumesAction() && serverLevel.captureTreeGeneration && !serverLevel.capturedBlockStates.isEmpty()) {
-                serverLevel.captureTreeGeneration = false;
+            if (interactionResult.consumesAction() && serverLevel.captureTreeGeneration.get() && !serverLevel.capturedBlockStates.isEmpty()) { // Canvas - atomic
+                serverLevel.captureTreeGeneration.set(false); // Canvas - atomic
                 org.bukkit.Location location = org.bukkit.craftbukkit.util.CraftLocation.toBukkit(clickedPos, serverLevel.getWorld());
                 org.bukkit.TreeType treeType = net.minecraft.world.level.block.SaplingBlock.treeType;
                 net.minecraft.world.level.block.SaplingBlock.treeType = null;
@@ -439,7 +_,7 @@
                 SignItem.openSign = null; // SPIGOT-6758 - Reset on early return
                 return interactionResult;
             }
-            serverLevel.captureTreeGeneration = false;
+            serverLevel.captureTreeGeneration.set(false); // Canvas - atomic
             if (player != null && interactionResult instanceof InteractionResult.Success success && success.wasItemInteraction()) {
                 InteractionHand hand = context.getHand();
                 org.bukkit.event.block.BlockPlaceEvent placeEvent = null;
@@ -457,14 +_,14 @@
                     placeEvent.getPlayer().updateInventory();
                     serverLevel.capturedTileEntities.clear(); // Paper - Allow chests to be placed with NBT data; clear out block entities as chests and such will pop loot
                     // revert back all captured blocks
-                    serverLevel.preventPoiUpdated = true; // CraftBukkit - SPIGOT-5710
+                    serverLevel.preventPoiUpdated.set(true); // CraftBukkit - SPIGOT-5710 // Canvas - atomic
                     serverLevel.isBlockPlaceCancelled = true; // Paper - prevent calling cleanup logic when undoing a block place upon a cancelled BlockPlaceEvent
                     for (org.bukkit.block.BlockState blockstate : blocks) {
                         blockstate.update(true, false);
                         ((org.bukkit.craftbukkit.block.CraftBlock) blockstate.getBlock()).getNMS().getBlock().forgetPlacer(); // Purpur - Store placer on Block when placed
                     }
                     serverLevel.isBlockPlaceCancelled = false; // Paper - prevent calling cleanup logic when undoing a block place upon a cancelled BlockPlaceEvent
-                    serverLevel.preventPoiUpdated = false;
+                    serverLevel.preventPoiUpdated.set(false); // Canvas - atomic
 
                     SignItem.openSign = null; // SPIGOT-6758 - Reset on early return
                 } else {
