--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -388,7 +_,7 @@
             ServerLevel serverLevel = (ServerLevel) context.getLevel();
 
             if (!(item instanceof BucketItem/* || item instanceof SolidBucketItem*/)) { // if not bucket // Paper - Fix cancelled powdered snow bucket placement
-                serverLevel.captureBlockStates = true;
+                serverLevel.captureBlockStates.set(true); // Canvas - atomic
                 // special case bonemeal
                 if (item == Items.BONE_MEAL) {
                     serverLevel.captureTreeGeneration = true;
@@ -398,7 +_,7 @@
             try {
                 interactionResult = item.useOn(context);
             } finally {
-                serverLevel.captureBlockStates = false;
+                serverLevel.captureBlockStates.set(false); // Canvas - atomic
             }
             DataComponentPatch newPatch = this.components.asPatch();
             int newCount = this.getCount();
