From bd14df94165f1c34a7f632a7de807b770022d5c5 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/item/MaceItem.java b/net/minecraft/world/item/MaceItem.java
index 529e9aae6e645d45f4b223990e325aaabff3de0c..368bad409c9f8b5e1eb472529352841262645b56 100644
--- a/net/minecraft/world/item/MaceItem.java
+++ b/net/minecraft/world/item/MaceItem.java
@@ -104,7 +104,7 @@ public class MaceItem extends Item {
             } else {
                 double fallHeightThreshold1 = 3.0;
                 double fallHeightThreshold2 = 8.0;
-                double fallDistance = attacker.fallDistance;
+                double fallDistance = io.canvasmc.canvas.Config.INSTANCE.combat.mace.ignoreFallDistance ? 1.0D : Math.min(attacker.fallDistance, io.canvasmc.canvas.Config.INSTANCE.combat.mace.fallDistanceLimit == -1 ? attacker.fallDistance : io.canvasmc.canvas.Config.INSTANCE.combat.mace.fallDistanceLimit); // Canvas - mace config
                 double damage;
                 if (fallDistance <= 3.0) {
                     damage = 4.0 * fallDistance;
