--- a/net/minecraft/world/item/MaceItem.java
+++ b/net/minecraft/world/item/MaceItem.java
@@ -104,7 +_,7 @@
             } else {
                 double fallHeightThreshold1 = 3.0;
                 double fallHeightThreshold2 = 8.0;
-                double fallDistance = attacker.fallDistance;
+                double fallDistance = attacker.level().canvasConfig().combat.mace.ignoreFallDistance ? 1.0D : Math.min(attacker.fallDistance, attacker.level().canvasConfig().combat.mace.fallDistanceLimit == -1 ? attacker.fallDistance : attacker.level().canvasConfig().combat.mace.fallDistanceLimit); // Canvas - mace config
                 double damage;
                 if (fallDistance <= 3.0) {
                     damage = 4.0 * fallDistance;
