--- a/net/minecraft/world/item/MapItem.java
+++ b/net/minecraft/world/item/MapItem.java
@@ -274,6 +_,7 @@
         if (!level.isClientSide) {
             MapItemSavedData savedData = getSavedData(stack, level);
             if (savedData != null) {
+                synchronized (savedData) { // Canvas
                 if (entity instanceof Player player) {
                     savedData.tickCarriedBy(player, stack);
                 }
@@ -281,6 +_,7 @@
                 if (!savedData.locked && (isSelected || entity instanceof Player && ((Player)entity).getOffhandItem() == stack)) {
                     this.update(level, entity, savedData);
                 }
+                } // Canvas
             }
         }
     }
