From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/item/enchantment/Enchantment.java b/net/minecraft/world/item/enchantment/Enchantment.java
index 0d981b86fa7ee8713e2a419fc2f377defbc2ed9c..d5e7cc8d6c784a0b34d73a031c8f6a0d20ab4f4b 100644
--- a/net/minecraft/world/item/enchantment/Enchantment.java
+++ b/net/minecraft/world/item/enchantment/Enchantment.java
@@ -124,7 +124,15 @@ public record Enchantment(Component description, Enchantment.EnchantmentDefiniti
     }
 
     public boolean matchingSlot(final EquipmentSlot slot) {
-        return this.definition.slots().stream().anyMatch(group -> group.test(slot));
+        // Canvas start - remove streams
+        for (EquipmentSlotGroup slotGroup : this.definition.slots()) {
+            if (slotGroup.test(slot)) {
+                return true;
+            }
+        }
+
+        return false;
+        // Canvas end - remove streams
     }
 
     public boolean isPrimaryItem(final ItemStack item) {
