From 56a92b2c479a7161c340134ba931fb3d6a8e6f79 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java b/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
index 4f91849366ea4c6a28a42d4b0916e9b340c2cb31..a11325f35ea3d8631c94f653de35a2b624186083 100644
--- a/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
+++ b/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
@@ -513,6 +513,8 @@ public class WalkNodeEvaluator extends NodeEvaluator {
         return blockPathType;
     }
 
+    // Canvas start - optimize path type cache
+    private static final PathType[] PATH_TYPES = PathType.values();
     protected static PathType getPathTypeFromState(final BlockGetter level, final BlockPos pos) {
         // Paper start - Do not load chunks during pathfinding
         BlockState blockState = level.getBlockStateIfLoaded(pos);
@@ -520,6 +522,11 @@ public class WalkNodeEvaluator extends NodeEvaluator {
             return PathType.BLOCKED;
         }
         // Paper end
+        return PATH_TYPES[blockState.pathType];
+    }
+
+    public static PathType canvas$getPathTypeFromBlockState(final BlockState blockState) {
+    // Canvas end - optimize path type cache
         Block block = blockState.getBlock();
         if (blockState.isAir()) {
             return PathType.OPEN;
