From 56a92b2c479a7161c340134ba931fb3d6a8e6f79 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/levelgen/NoiseChunk.java b/net/minecraft/world/level/levelgen/NoiseChunk.java
index 117c3e33d49b92ca0807c576237fe135e40c524d..ef1305981fbd44693ad28d722f60dd3901a7b45d 100644
--- a/net/minecraft/world/level/levelgen/NoiseChunk.java
+++ b/net/minecraft/world/level/levelgen/NoiseChunk.java
@@ -373,7 +373,17 @@ public class NoiseChunk implements DensityFunction.FunctionContext, DensityFunct
     }
 
     protected DensityFunction wrap(final DensityFunction function) {
-        return this.wrapped.computeIfAbsent(function, this::wrapNew);
+        // Canvas start - reduce worldgen allocations
+        // avoid lambda allocation
+        DensityFunction func = this.wrapped.get(function);
+
+        if (func == null) {
+            func = this.wrapNew(function);
+            this.wrapped.put(function, func);
+        }
+
+        return func;
+        // Canvas end - reduce worldgen allocations
     }
 
     private DensityFunction wrapNew(final DensityFunction function) {
