From 56a92b2c479a7161c340134ba931fb3d6a8e6f79 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/levelgen/SurfaceRules.java b/net/minecraft/world/level/levelgen/SurfaceRules.java
index 7a3b0da821352b228ca2cd1d8a288238715a377d..e90df8cbddb0458be13a046090133ece55e0f29a 100644
--- a/net/minecraft/world/level/levelgen/SurfaceRules.java
+++ b/net/minecraft/world/level/levelgen/SurfaceRules.java
@@ -666,8 +666,13 @@ public class SurfaceRules {
     private record SequenceRule(List<SurfaceRules.SurfaceRule> rules) implements SurfaceRules.SurfaceRule {
         @Override
         public @Nullable BlockState tryApply(final int blockX, final int blockY, final int blockZ) {
-            for (SurfaceRules.SurfaceRule rule : this.rules) {
-                BlockState state = rule.tryApply(blockX, blockY, blockZ);
+            // Canvas start - reduce worldgen allocations
+            // avoid iterator allocation
+            int size = this.rules.size();
+            //noinspection ForLoopReplaceableByForEach
+            for (int i = 0; i < size; i++) {
+                BlockState state = this.rules.get(i).tryApply(blockX, blockY, blockZ);
+            // Canvas end - reduce worldgen allocations
                 if (state != null) {
                     return state;
                 }
