--- a/net/minecraft/world/entity/monster/AbstractSkeleton.java
+++ b/net/minecraft/world/entity/monster/AbstractSkeleton.java
@@ -3,6 +_,7 @@
 import java.time.LocalDate;
 import java.time.temporal.ChronoField;
 import javax.annotation.Nullable;
+import io.canvasmc.canvas.Config;
 import net.minecraft.core.BlockPos;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.server.level.ServerLevel;
@@ -186,6 +_,17 @@
         double d1 = target.getY(0.3333333333333333) - arrow.getY();
         double d2 = target.getZ() - this.getZ();
         double squareRoot = Math.sqrt(d * d + d2 * d2);
+        // Canvas start - skeleton aim config
+        int inaccuracyModifier = Config.INSTANCE.entities.skeletonAimAccuracy;
+        if (inaccuracyModifier > 0) {
+            RandomSource random = this.getRandom();
+            double inaccuracyFactor = inaccuracyModifier * 0.05;
+            d += random.nextGaussian() * inaccuracyFactor;
+            d1 += random.nextGaussian() * inaccuracyFactor;
+            d2 += random.nextGaussian() * inaccuracyFactor;
+        }
+
+        // Canvas end
         if (this.level() instanceof ServerLevel serverLevel) {
             // CraftBukkit start
             org.bukkit.event.entity.EntityShootBowEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityShootBowEvent(this, this.getMainHandItem(), arrow.getPickupItem(), arrow, net.minecraft.world.InteractionHand.MAIN_HAND, 0.8F, true); // Paper - improve entity shhot bow event - add arrow stack to event
