--- a/net/minecraft/world/entity/monster/Drowned.java
+++ b/net/minecraft/world/entity/monster/Drowned.java
@@ -298,9 +_,11 @@
         if (!this.level().isClientSide) {
             if (this.isEffectiveAi() && this.isUnderWater() && this.wantsToSwim()) {
                 this.navigation = this.waterNavigation;
+                (this).lithium$updateNavigationRegistration(); // Canvas
                 this.setSwimming(true);
             } else {
                 this.navigation = this.groundNavigation;
+                (this).lithium$updateNavigationRegistration(); // Canvas
                 this.setSwimming(false);
             }
         }
@@ -313,7 +_,7 @@
 
     protected boolean closeToNextPos() {
         Path path = this.getNavigation().getPath();
-        if (path != null) {
+        if (path != null && path.isProcessed()) { // Canvas - ensure path is processed
             BlockPos target = path.getTarget();
             if (target != null) {
                 double d = this.distanceToSqr(target.getX(), target.getY(), target.getZ());
@@ -401,6 +_,7 @@
             this.drowned.setSearchingForLand(false);
             this.drowned.navigation = this.drowned.groundNavigation;
             super.start();
+            (this.drowned).lithium$updateNavigationRegistration(); // Canvas
         }
 
         @Override
