From 562a609f7063b6400aca83fa84572c4902aa25af Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/monster/EnderMan.java b/net/minecraft/world/entity/monster/EnderMan.java
index 55a7cac408614d0efdece87a5380bd1fb7406200..f5af7a65a0627ceeedb9869d8e5da291a668b87b 100644
--- a/net/minecraft/world/entity/monster/EnderMan.java
+++ b/net/minecraft/world/entity/monster/EnderMan.java
@@ -298,11 +298,17 @@ public class EnderMan extends Monster implements NeutralMob {
     private boolean teleport(final double x, final double y, final double z) {
         BlockPos.MutableBlockPos pos = new BlockPos.MutableBlockPos(x, y, z);
 
-        while (pos.getY() > this.level().getMinY() && !this.level().getBlockState(pos).blocksMotion()) {
+        // Pufferfish start - single chunk lookup
+        net.minecraft.world.level.chunk.LevelChunk chunk = this.level().getChunkIfLoaded(pos);
+        if (chunk == null) {
+            return false;
+        }
+        // Pufferfish end - single chunk lookup
+        while (pos.getY() > this.level().getMinY() && !chunk.getBlockState(pos).blocksMotion()) { // Pufferfish - single chunk lookup
             pos.move(Direction.DOWN);
         }
 
-        BlockState blockState = this.level().getBlockState(pos);
+        BlockState blockState = chunk.getBlockState(pos); // Pufferfish - single chunk lookup
         boolean couldStandOn = blockState.blocksMotion();
         boolean isWet = blockState.getFluidState().is(FluidTags.WATER);
         if (couldStandOn && !isWet) {
