--- a/net/minecraft/world/entity/monster/RangedAttackMob.java
+++ b/net/minecraft/world/entity/monster/RangedAttackMob.java
@@ -7,6 +_,6 @@
     void performRangedAttack(LivingEntity target, float power);
 
     default float rangedAttackUncertainty(final Level level) {
-        return 14 - level.getDifficulty().getId() * 4;
+        return Math.max(level.canvasConfig().entities.rangedAttackMobUncertaintyBaseValue - level.getDifficulty().getId() * 4, 0); // Canvas - ranged attack uncertainty config
     }
 }
