From 60fbf10de1ffff26663053a967c13ae1e63a5a7a Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/monster/Strider.java b/net/minecraft/world/entity/monster/Strider.java
index 8f58f2f9dae3809c72cf59c0da3523c8923177c2..132e955706e262c4549c5c2dc836b06845627904 100644
--- a/net/minecraft/world/entity/monster/Strider.java
+++ b/net/minecraft/world/entity/monster/Strider.java
@@ -297,7 +297,7 @@ public class Strider extends Animal implements ItemSteerable {
             BlockState stateOn = this.getBlockStateOnLegacy();
             boolean inWarmBlocks = stateInside.is(BlockTags.STRIDER_WARM_BLOCKS)
                 || stateOn.is(BlockTags.STRIDER_WARM_BLOCKS)
-                || this.getFluidHeight(FluidTags.LAVA) > 0.0;
+                || canvas$getLavaFluidHeight() > 0.0; // Canvas - optimize fluid height
             boolean onWarmStrider = this.getVehicle() instanceof Strider strider && !strider.isSuffocating();
             // CraftBukkit start
             boolean suffocating = !inWarmBlocks && !onWarmStrider;
@@ -360,7 +360,7 @@ public class Strider extends Animal implements ItemSteerable {
 
     @Override
     protected boolean canAddPassenger(final Entity passenger) {
-        return !this.isVehicle() && !this.isEyeInFluid(FluidTags.LAVA);
+        return !this.isVehicle() && !canvas$isEyeInLava(); // Canvas - optimize fluidOnEyes
     }
 
     @Override
