From 73e3ddbec8cd225e42fdf1e4683ca01a6c53e643 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/monster/Witch.java b/net/minecraft/world/entity/monster/Witch.java
index 205b55db821ab81883c23c194cdb12cd57061fb9..08dccdf3c2b4ebb16ea6681bd28bd44cee6d0ce3 100644
--- a/net/minecraft/world/entity/monster/Witch.java
+++ b/net/minecraft/world/entity/monster/Witch.java
@@ -138,7 +138,7 @@ public class Witch extends Raider implements RangedAttackMob {
                 }
             } else {
                 Holder<Potion> potion = null;
-                if (this.random.nextFloat() < 0.15F && this.isEyeInFluid(FluidTags.WATER) && !this.hasEffect(MobEffects.WATER_BREATHING)) {
+                if (this.random.nextFloat() < 0.15F && this.canvas$isEyeInWater() && !this.hasEffect(MobEffects.WATER_BREATHING)) { // Canvas - optimize fluidOnEyes
                     potion = Potions.WATER_BREATHING;
                 } else if (this.random.nextFloat() < 0.15F
                     && (this.isOnFire() || this.getLastDamageSource() != null && this.getLastDamageSource().is(DamageTypeTags.IS_FIRE))
