--- a/net/minecraft/world/entity/monster/ZombieVillager.java
+++ b/net/minecraft/world/entity/monster/ZombieVillager.java
@@ -238,6 +_,16 @@
                 if (!this.isSilent()) {
                     level.levelEvent(null, 1027, this.blockPosition(), 0);
                 }
+                // Canvas start - MC-200418
+                if (
+                    io.canvasmc.canvas.Config.INSTANCE.fixes.mc200418
+                        && mob.isPassenger()
+                        && mob.getVehicle() instanceof net.minecraft.world.entity.animal.Chicken
+                        && mob.isBaby())
+                {
+                    mob.removeVehicle();
+                }
+                // Canvas end - MC-200418
                 // CraftBukkit start
             }, org.bukkit.event.entity.EntityTransformEvent.TransformReason.CURED, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.CURED // CraftBukkit
         );
