--- a/paper-api/build.gradle.kts
+++ b/paper-api/build.gradle.kts
@@ -4,6 +_,7 @@
     `java-library`
     `maven-publish`
     idea
+    id("io.canvasmc.weaver.dependency-bridge")
 }
 
 java {
@@ -45,7 +_,7 @@
     // api dependencies are listed transitively to API consumers
     api("com.google.guava:guava:33.6.0-jre")
     api("com.google.code.gson:gson:2.14.0")
-    api("org.yaml:snakeyaml:2.2")
+    api("org.yaml:snakeyaml:2.3")
     api("org.joml:joml:1.10.8") {
         isTransitive = false // https://github.com/JOML-CI/JOML/issues/352
     }
@@ -68,9 +_,10 @@
     apiAndDocs("net.kyori:adventure-text-serializer-plain")
     apiAndDocs("net.kyori:adventure-text-logger-slf4j")
 
-    api("org.apache.maven:maven-resolver-provider:3.9.6") // make API dependency for Paper Plugins
-    implementation("org.apache.maven.resolver:maven-resolver-connector-basic:1.9.18")
-    implementation("org.apache.maven.resolver:maven-resolver-transport-http:1.9.18")
+    api("org.apache.maven:maven-resolver-provider:3.9.9") // make API dependency for Paper Plugins
+    implementation("org.apache.maven.resolver:maven-resolver-connector-basic:1.9.22")
+    implementation("org.apache.maven.resolver:maven-resolver-transport-http:1.9.22")
+    compileOnly("org.apache.commons:commons-lang3:3.17.0")
 
     // Annotations - Slowly migrate to jspecify
     val annotations = "org.jetbrains:annotations:$annotationsVersion"
@@ -87,14 +_,14 @@
     // Test dependencies
     testImplementation("org.apache.commons:commons-lang3:3.20.0")
     testImplementation("org.junit.jupiter:junit-jupiter:6.0.3")
-    testImplementation("org.hamcrest:hamcrest:2.2")
+    testImplementation("org.hamcrest:hamcrest:3.0")
     testImplementation("org.mockito:mockito-core:5.22.0")
     testImplementation("org.ow2.asm:asm-tree:9.9.1")
     mockitoAgent("org.mockito:mockito-core:5.22.0") { isTransitive = false } // configure mockito agent that is needed in newer java versions
     testRuntimeOnly("org.junit.platform:junit-platform-launcher:6.0.3")
 }
 
-val generatedDir: java.nio.file.Path = layout.projectDirectory.dir("src/generated/java").asFile.toPath()
+val generatedDir: java.nio.file.Path = rootProject.layout.projectDirectory.dir("paper-api/src/generated/java").asFile.toPath()
 idea {
     module {
         generatedSourceDirs.add(generatedDir.toFile())
@@ -104,6 +_,18 @@
     main {
         java {
             srcDir(generatedDir)
+            srcDir(file("../paper-api/src/main/java"))
+        }
+        resources {
+            srcDir(file("../paper-api/src/main/resources"))
+        }
+    }
+    test {
+        java {
+            srcDir(file("../paper-api/src/test/java"))
+        }
+        resources {
+            srcDir(file("../paper-api/src/test/resources"))
         }
     }
 }
@@ -138,6 +_,32 @@
             suppressPomMetadataWarningsFor(it)
         }
         from(components["java"])
+        // Canvas start - add more info to the publication
+        pom {
+            name.set("canvas-api")
+            description.set("API bundle for the Canvas Minecraft server software")
+            url.set("https://github.com/CraftCanvasMC/Canvas")
+            licenses {
+                license {
+                    name.set("GPL-3.0")
+                    url.set("https://github.com/CraftCanvasMC/Canvas/blob/HEAD/LICENSE")
+                    distribution.set("repo")
+                }
+            }
+            developers {
+                developer {
+                    id.set("canvas-team")
+                    name.set("Canvas Team")
+                    organization.set("CanvasMC")
+                    organizationUrl.set("https://canvasmc.io")
+                    roles.add("developer")
+                }
+            }
+            scm {
+                url.set("https://github.com/CraftCanvasMC/Canvas")
+            }
+        }
+        // Canvas end - add more info to the publication
     }
 }
 
@@ -178,6 +_,10 @@
     }
 }
 
+tasks.withType<JavaCompile>().configureEach {
+    options.compilerArgs.add("--add-modules=jdk.incubator.vector")
+}
+
 abstract class Services {
     @get:Inject
     abstract val fileSystemOperations: FileSystemOperations
@@ -186,7 +_,9 @@
 
 tasks.withType<Javadoc>().configureEach {
     val options = options as StandardJavadocDocletOptions
-    options.overview = "src/main/javadoc/overview.html"
+    options.addStringOption("-add-modules", "jdk.incubator.vector")
+    options.addStringOption("Xdoclint:none", "-quiet")
+    options.overview = "../paper-api/src/main/javadoc/overview.html"
     options.use()
     options.isDocFilesSubDirs = true
     options.links(
@@ -213,11 +_,11 @@
     }
 
     // workaround for https://github.com/gradle/gradle/issues/4046
-    inputs.dir("src/main/javadoc").withPropertyName("javadoc-sourceset")
+    inputs.dir("../paper-api/src/main/javadoc").withPropertyName("javadoc-sourceset")
     val fsOps = services.fileSystemOperations
     doLast {
         fsOps.copy {
-            from("src/main/javadoc") {
+            from("../paper-api/src/main/javadoc") {
                 include("**/doc-files/**")
             }
             into("build/docs/javadoc")
