From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/projectile/hurtingprojectile/DragonFireball.java b/net/minecraft/world/entity/projectile/hurtingprojectile/DragonFireball.java
index 5bdec8986c88f33b6e6fe806e8e3ed0f69cb617f..66ed996fd33da1fa881e4968994e2bc1d9a6e602 100644
--- a/net/minecraft/world/entity/projectile/hurtingprojectile/DragonFireball.java
+++ b/net/minecraft/world/entity/projectile/hurtingprojectile/DragonFireball.java
@@ -32,7 +32,7 @@ public class DragonFireball extends AbstractHurtingProjectile {
         super.onHit(result);
         if (result.getType() != HitResult.Type.ENTITY || !this.ownedBy(((EntityHitResult)result).getEntity())) {
             if (!this.level().isClientSide()) {
-                List<LivingEntity> entitiesOfClass = this.level().getEntitiesOfClass(LivingEntity.class, this.getBoundingBox().inflate(4.0, 2.0, 4.0));
+                List<LivingEntity> entitiesOfClass = this.level().getEntitiesOfClass(LivingEntity.class, this.getBoundingBox().inflate(4.0, 2.0, 4.0), ca.spottedleaf.moonrise.common.util.TickThread::isTickThreadFor); // Canvas - region threading
                 AreaEffectCloud areaEffectCloud = new AreaEffectCloud(this.level(), this.getX(), this.getY(), this.getZ());
                 Entity owner = this.getOwner();
                 if (owner instanceof LivingEntity) {
