--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -111,7 +_,7 @@
     public static final Component CHAT_FILTERED_FULL = Component.translatable("chat.filtered_full");
     public static final Component DUPLICATE_LOGIN_DISCONNECT_MESSAGE = Component.translatable("multiplayer.disconnect.duplicate_login");
     private static final Logger LOGGER = LogUtils.getLogger();
-    public static final int SEND_PLAYER_INFO_INTERVAL = 600;
+    public static final int SEND_PLAYER_INFO_INTERVAL = io.canvasmc.canvas.GlobalConfiguration.getInstance().networking.playerInfoSendInterval; // Canvas - optimize playerlist tick
     private static final ThreadLocal<SimpleDateFormat> BAN_DATE_FORMAT = ThreadLocal.withInitial(() -> new SimpleDateFormat("yyyy-MM-dd 'at' HH:mm:ss z", Locale.ROOT)); // Folia - region threading - SDF is not thread-safe
     private final MinecraftServer server;
     public final List<ServerPlayer> players = new java.util.concurrent.CopyOnWriteArrayList(); // CraftBukkit - ArrayList -> CopyOnWriteArrayList: Iterator safety
@@ -130,6 +_,7 @@
     private int simulationDistance;
     private boolean allowCommandsForAllPlayers;
     private int sendAllPlayerInfoIn;
+    private final PlayerListBucket[] canvas$infoBuckets = new PlayerListBucket[SEND_PLAYER_INFO_INTERVAL]; // Canvas - optimize playerlist tick
 
     // CraftBukkit start
     private org.bukkit.craftbukkit.CraftServer cserver;
@@ -192,7 +_,7 @@
     private boolean countConnection(Connection conn, int limit) {
         synchronized (this.connectionsStateLock) {
             int count = this.usersCountedAgainstLimit.size();
-            if (count >= limit) {
+            if (count > limit) { // Canvas - fix Folia max player count breaking Vanilla
                 return false;
             }
             this.usersCountedAgainstLimit.add(conn);
@@ -258,7 +_,7 @@
                 !immediateRespawn,
                 doLimitedCrafting,
                 player.createCommonSpawnInfo(level),
-                this.server.enforceSecureProfile()
+                io.canvasmc.canvas.Config.INSTANCE.enableNoChatReports || this.server.enforceSecureProfile() // Canvas - no chat reports
             )
         );
         player.getBukkitEntity().sendSupportedChannels(); // CraftBukkit
@@ -292,6 +_,8 @@
 
         // player.connection.send(ClientboundPlayerInfoUpdatePacket.createPlayerInitializing(this.players)); // CraftBukkit - replaced with loop below
         this.players.add(player);
+        player.lastSave = System.nanoTime(); // Canvas - init lastSave on join
+        this.canvas$pushToBuckets(player); // Canvas - optimize playerlist tick
         this.playersByName.put(player.getScoreboardName().toLowerCase(java.util.Locale.ROOT), player); // Spigot
         this.playersByUUID.put(player.getUUID(), player);
         // this.broadcastAll(ClientboundPlayerInfoUpdatePacket.createPlayerInitializing(List.of(player))); // CraftBukkit - replaced with loop below
@@ -498,6 +_,7 @@
         if (advancements != null) {
             advancements.save();
         }
+        new io.canvasmc.canvas.event.PlayerSaveEvent(player.getBukkitEntity(), false).callEvent(); // Canvas - player save event
     }
 
     public net.kyori.adventure.text.@Nullable Component remove(final ServerPlayer player) { // CraftBukkit - return string // Paper - return Component
@@ -566,6 +_,7 @@
         player.getBukkitEntity().packetProcessor.close(); // Folia - region threading
         player.getAdvancements().stopListening();
         this.players.remove(player);
+        this.canvas$removeFromBuckets(player); // Canvas - optimize playerlist tick
         this.playersByName.remove(player.getScoreboardName().toLowerCase(java.util.Locale.ROOT)); // Spigot
         this.server.getCustomBossEvents().onPlayerDisconnect(player);
         UUID uuid = player.getUUID();
@@ -593,6 +_,7 @@
         // This removes the scoreboard (and player reference) for the specific player in the manager
         this.cserver.getScoreboardManager().removePlayer(player.getBukkitEntity());
         // CraftBukkit end
+        new io.canvasmc.canvas.event.PlayerSaveEvent(player.getBukkitEntity(), true).callEvent(); // Canvas - player save event
         return playerQuitEvent.quitMessage(); // Paper - Adventure
     }
 
@@ -801,6 +_,20 @@
     }
 
     public void tick() {
+        // Canvas start - optimize playerlist tick
+        if (io.canvasmc.canvas.GlobalConfiguration.getInstance().networking.alternativePlayerListTick) {
+            PlayerListBucket bucket = this.canvas$infoBuckets[this.sendAllPlayerInfoIn];
+
+            if (bucket != null) {
+                bucket.tick(this.players.toArray(new ServerPlayer[0]));
+            }
+
+            if (++this.sendAllPlayerInfoIn >= SEND_PLAYER_INFO_INTERVAL) {
+                this.sendAllPlayerInfoIn = 0;
+            }
+            return;
+        }
+        // Canvas end - optimize playerlist tick
         if (++this.sendAllPlayerInfoIn > 600) {
             // CraftBukkit start
             ServerPlayer[] players = this.players.toArray(new ServerPlayer[0]); // Folia - region threading
@@ -812,6 +_,61 @@
             this.sendAllPlayerInfoIn = 0;
         }
     }
+    // Canvas start - optimize playerlist tick
+
+    public void canvas$pushToBuckets(ServerPlayer player) {
+        if (!io.papermc.paper.threadedregions.RegionizedServer.isGlobalTickThread()) {
+            io.papermc.paper.threadedregions.RegionizedServer.getInstance().addTask(() -> canvas$pushToBuckets(player));
+            return;
+        }
+        PlayerListBucket bucket = this.canvas$infoBuckets[player.canvas$infoBucketIndex];
+        if (bucket == null) {
+            this.canvas$infoBuckets[player.canvas$infoBucketIndex] = new PlayerListBucket(new ServerPlayer[]{player});
+            return;
+        }
+        this.canvas$infoBuckets[player.canvas$infoBucketIndex] = bucket = new PlayerListBucket(java.util.Arrays.copyOf(bucket.bucket(), bucket.bucket().length + 1));
+        bucket.bucket()[bucket.bucket().length - 1] = player;
+    }
+
+    public void canvas$removeFromBuckets(ServerPlayer player) {
+        if (!io.papermc.paper.threadedregions.RegionizedServer.isGlobalTickThread()) {
+            io.papermc.paper.threadedregions.RegionizedServer.getInstance().addTask(() -> canvas$removeFromBuckets(player));
+            return;
+        }
+        PlayerListBucket bucket = this.canvas$infoBuckets[player.canvas$infoBucketIndex];
+        if (bucket != null) {
+            if (bucket.bucket().length == 1) {
+                if (bucket.bucket()[0] == player) {
+                    this.canvas$infoBuckets[player.canvas$infoBucketIndex] = null;
+                }
+                return;
+            }
+
+            for (int i = 0; i < bucket.bucket().length; i++) {
+                if (bucket.bucket()[i] == player) {
+                    bucket.bucket()[i] = bucket.bucket()[bucket.bucket().length - 1];
+                    this.canvas$infoBuckets[player.canvas$infoBucketIndex] = new PlayerListBucket(java.util.Arrays.copyOf(bucket.bucket(), bucket.bucket().length - 1));
+                }
+            }
+        }
+    }
+
+    private record PlayerListBucket(ServerPlayer[] bucket) {
+        public void tick(ServerPlayer[] allPlayers) {
+            List<ServerPlayer> players = java.util.Arrays.asList(allPlayers);
+
+            for (int i = bucket.length - 1; i >= 0; i--) {
+                ServerPlayer target = bucket[i];
+                target.connection.send(
+                    new ClientboundPlayerInfoUpdatePacket(
+                        EnumSet.of(ClientboundPlayerInfoUpdatePacket.Action.UPDATE_LATENCY),
+                        com.google.common.collect.Collections2.filter(players, t -> target.getBukkitEntity().canSee(t.getBukkitEntity()))
+                    )
+                );
+            }
+        }
+    }
+    // Canvas end - optimize playerlist tick
 
     // CraftBukkit start - add a world/entity limited version
     public void broadcastAll(Packet packet, net.minecraft.world.entity.player.Player entityhuman) {
@@ -1026,16 +_,19 @@
         final Packet<?> packet
     ) {
         for (ServerPlayer player : this.players) { // Folia - region threading
-            // CraftBukkit start - Test if player receiving packet can see the source of the packet
-            if (except != null && !player.getBukkitEntity().canSee(except.getBukkitEntity())) {
-               continue;
-            }
-            // CraftBukkit end
+            // Canvas - optimize - move down
             if (player != except && player.level().dimension() == dimension) {
                 double xd = x - player.getX();
                 double yd = y - player.getY();
                 double zd = z - player.getZ();
                 if (xd * xd + yd * yd + zd * zd < range * range) {
+                    // Canvas start - from above
+                    // CraftBukkit start - Test if player receiving packet can see the source of the packet
+                    if (except != null && !player.getBukkitEntity().canSee(except.getBukkitEntity())) {
+                        continue;
+                    }
+                    // CraftBukkit end
+                    // Canvas end - from above
                     player.connection.send(packet);
                 }
             }
@@ -1053,13 +_,13 @@
             int numSaved = 0;
             final long now = System.nanoTime(); // Folia - region threading
             final long timeInterval = (long)interval * io.papermc.paper.threadedregions.TickRegionScheduler.getTimeBetweenTicks(); // Folia - region threading // Canvas - region threading
-            for (final ServerPlayer player : this.players) {
+            for (final ServerPlayer player : interval == -1 ? this.players : io.papermc.paper.threadedregions.TickRegionScheduler.getCurrentRegionizedWorldData().getLocalPlayers()) { // Canvas - optimize player save
             // Folia start - region threading
-            if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(player)) {
+            if (interval == -1 && !ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(player)) { // Canvas - optimize player save
                 continue;
             }
             // Folia end - region threading
-            if (interval == -1 || now - player.lastSave >= timeInterval) { // Folia - region threading
+            if (interval == -1 || player.lastSave == ServerPlayer.LAST_SAVE_ABSENT || now - player.lastSave >= timeInterval) { // Folia - region threading // Canvas - fix overflow on LAST_SAVE_ABSENT
                 this.save(player);
                 if (interval != -1 && ++numSaved >= io.papermc.paper.configuration.GlobalConfiguration.get().playerAutoSave.maxPerTick()) {
                     break;
@@ -1270,6 +_,7 @@
     }
 
     public boolean verifyChatTrusted(final PlayerChatMessage message) {
+        if (io.canvasmc.canvas.Config.INSTANCE.enableNoChatReports) return true; // Canvas - no chat reports
         return message.hasSignature() && !message.hasExpiredServer(Instant.now());
     }
 
@@ -1330,7 +_,12 @@
         //this.broadcastAll(new ClientboundSetChunkCacheRadiusPacket(viewDistance)); // Paper - rewrite chunk system
 
         for (ServerLevel level : this.server.getAllLevels()) {
-            level.getChunkSource().setViewDistance(viewDistance);
+            // Canvas start - per world distance
+            io.canvasmc.canvas.world.PerWorldDistanceConfig distanceConfig = level.serverLevelData.canvas$distanceConfig;
+            if (!distanceConfig.isViewDistanceOverridden()) {
+                level.getChunkSource().setViewDistance(viewDistance);
+            }
+            // Canvas end - per world distance
         }
     }
 
@@ -1339,7 +_,12 @@
         //this.broadcastAll(new ClientboundSetSimulationDistancePacket(simulationDistance)); // Paper - rewrite chunk system
 
         for (ServerLevel level : this.server.getAllLevels()) {
-            level.getChunkSource().setSimulationDistance(simulationDistance);
+            // Canvas start - per world distance
+            io.canvasmc.canvas.world.PerWorldDistanceConfig distanceConfig = level.serverLevelData.canvas$distanceConfig;
+            if (!distanceConfig.isSimulationDistanceOverridden()) {
+                level.getChunkSource().setSimulationDistance(simulationDistance);
+            }
+            // Canvas end - per world distance
         }
     }
 
