From 23441ae46065042fd5e624f07307c4cb48391dc3 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/server/players/SleepStatus.java b/net/minecraft/server/players/SleepStatus.java
index 61870e387c93a5f0e62180372438b7011a294cd6..e53cf30607fae750b74cd49a2f731bffdc39bb85 100644
--- a/net/minecraft/server/players/SleepStatus.java
+++ b/net/minecraft/server/players/SleepStatus.java
@@ -15,9 +15,23 @@ public class SleepStatus {
 
     public boolean areEnoughDeepSleeping(final int sleepPercentageNeeded, final List<ServerPlayer> players) {
         // CraftBukkit start
-        int deepSleepers = (int)players.stream().filter(player -> player.isSleepingLongEnough() || player.fauxSleeping).count();
-        boolean anyDeepSleep = players.stream().anyMatch(Player::isSleepingLongEnough);
-        return anyDeepSleep && deepSleepers >= this.sleepersNeeded(sleepPercentageNeeded);
+        // Canvas start - remove streams
+        long count = 0L;
+        boolean anyDeepSleep = false;
+
+        for (net.minecraft.server.level.ServerPlayer player : players) {
+            final boolean isSleepingLongEnough = player.isSleepingLongEnough();
+
+            if (isSleepingLongEnough || player.fauxSleeping) {
+                count++;
+                if (!anyDeepSleep && isSleepingLongEnough) {
+                    anyDeepSleep = true;
+                }
+            }
+        }
+
+        return anyDeepSleep && count >= this.sleepersNeeded(sleepPercentageNeeded);
+        // Canvas end - remove streams
         // CraftBukkit end
     }
 
