--- a/net/minecraft/server/players/SleepStatus.java
+++ b/net/minecraft/server/players/SleepStatus.java
@@ -15,9 +_,23 @@
 
     public boolean areEnoughDeepSleeping(final int sleepPercentageNeeded, final List<ServerPlayer> players) {
         // CraftBukkit start
-        int deepSleepers = (int)players.stream().filter(player -> player.isSleepingLongEnough() || player.fauxSleeping).count();
-        boolean anyDeepSleep = players.stream().anyMatch(Player::isSleepingLongEnough);
-        return anyDeepSleep && deepSleepers >= this.sleepersNeeded(sleepPercentageNeeded);
+        // Canvas start - remove streams
+        long count = 0L;
+        boolean anyDeepSleep = false;
+
+        for (net.minecraft.server.level.ServerPlayer player : players) {
+            final boolean isSleepingLongEnough = player.isSleepingLongEnough();
+
+            if (isSleepingLongEnough || player.fauxSleeping) {
+                count++;
+                if (!anyDeepSleep && isSleepingLongEnough) {
+                    anyDeepSleep = true;
+                }
+            }
+        }
+
+        return anyDeepSleep && count >= this.sleepersNeeded(sleepPercentageNeeded);
+        // Canvas end - remove streams
         // CraftBukkit end
     }
 
