--- a/net/minecraft/world/Clearable.java
+++ b/net/minecraft/world/Clearable.java
@@ -3,11 +_,14 @@
 import javax.annotation.Nullable;
 
 public interface Clearable {
+    final Object clearingLock = new Object(); // Canvas
     void clearContent();
 
     static void tryClear(@Nullable Object object) {
+        synchronized (clearingLock) { // Canvas
         if (object instanceof Clearable) {
             ((Clearable)object).clearContent();
         }
+        } // Canvas
     }
 }
