--- a/ca/spottedleaf/moonrise/patches/chunk_system/util/ParallelSearchRadiusIteration.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/util/ParallelSearchRadiusIteration.java
@@ -23,6 +_,11 @@
     }
 
     public static long[] getSearchIteration(final int radius) {
+        // Canvas start - ensure bounds are kept
+        if (radius >= SEARCH_RADIUS_ITERATION_LIST.length) {
+            return SEARCH_RADIUS_ITERATION_LIST[SEARCH_RADIUS_ITERATION_LIST.length - 1];
+        }
+        // Canvas end
         return SEARCH_RADIUS_ITERATION_LIST[radius];
     }
 
