From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <159172167+PurpleWolfMC@users.noreply.github.com>
Date: Mon, 29 Apr 2024 18:48:00 +0000
Subject: [PATCH] Dont Spam Warnings During Compile


diff --git a/build.gradle.kts b/build.gradle.kts
index e51d37397e3f13f53f7a9c6d09f0b8965c0333d6..136d17798a5add70035da8239a592139bebac2ef 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -71,6 +71,12 @@ dependencies {
 paperweight {
     craftBukkitPackageVersion.set("v1_20_R4") // also needs to be updated in MappingEnvironment
 }
+// Canvas start - lets not spam warnings
+tasks.compileJava {
+    options.compilerArgs.add("-Xlint:-deprecation")
+    options.isWarnings = false
+}
+// Canvas end
 
 tasks.jar {
     archiveClassifier.set("dev")
