From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sun, 19 Nov 2023 13:13:01 -0800
Subject: [PATCH] Add "isIllegalItem()" to CraftItemStack


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
index f1e1953f2dc65dc615b7b7b648c37b195d3b4c25..9ca8d319045c6576ab8bb3af605b509020b35c23 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
@@ -116,6 +116,12 @@ public final class CraftItemStack extends ItemStack {
     }
 
     public net.minecraft.world.item.ItemStack handle;
+    // Canvas start - why no getHandle method?
+
+    public net.minecraft.world.item.ItemStack getHandle(){
+        return this.handle;
+    }
+    // Canvas end
 
     /**
      * Mirror
@@ -668,6 +674,31 @@ public final class CraftItemStack extends ItemStack {
     static Material getType(net.minecraft.world.item.ItemStack item) {
         return item == null ? Material.AIR : CraftItemType.minecraftToBukkit(item.getItem());
     }
+    // Canvas start
+
+    public boolean isIllegalItem(){
+        return CraftItemStack.isIllegalItem(this);
+    }
+
+    public static boolean isIllegalItem(ItemStack itemStack){
+        for(Enchantment enchantment : itemStack.getEnchantments().keySet()){
+            if(enchantment.getMaxLevel() < itemStack.getEnchantments().get(enchantment)){
+                return true;
+            }
+        }
+
+        if(
+            itemStack.getDamage() < 0 ||
+            itemStack.getMaxItemUseDuration() < itemStack.getDamage() ||
+            itemStack.getItemMeta() == null ||
+            itemStack.getAmount() > 64
+        ){
+            return true;
+        } else {
+            return false;
+        }
+    }
+    // Canvas end
 
     @Override
     public boolean setItemMeta(ItemMeta itemMeta) {
