From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Fri, 24 Nov 2023 14:45:11 -0800
Subject: [PATCH] Create FakeBlock/GhostBlock api


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 523a6afd7f1dcb0b4ddf5ede0415487c104296bc..3e26419322ff7d0f2471f0ac84af69aaeb03a93e 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -1028,6 +1028,44 @@ public class CraftWorld extends CraftRegionAccessor implements World {
     public Block getBlockAt(Location location) {
         return this.getBlockAt(location.getBlockX(), location.getBlockY(), location.getBlockZ());
     }
+    // Canvas start
+
+    @Override
+    public void createFakePhysicalBlock(final Location location, final org.bukkit.Material material, final Player sender) {
+        sender.sendBlockChange(location, material.createBlockData());
+    }
+
+    @Override
+    public void createGhostBlock(final Location location, final org.bukkit.Material material, final Player sender) {
+        org.bukkit.craftbukkit.entity.CraftBlockDisplay entity = (org.bukkit.craftbukkit.entity.CraftBlockDisplay) this.spawnEntity(location, org.bukkit.entity.EntityType.BLOCK_DISPLAY);
+        entity.setBlock(material.createBlockData());
+        net.minecraft.network.protocol.game.ClientboundAddEntityPacket entityPacket = new net.minecraft.network.protocol.game.ClientboundAddEntityPacket(
+            entity.getHandle()
+        );
+        CraftPlayer toSend = (CraftPlayer) sender;
+        toSend.getHandle().connection.send(entityPacket);
+    }
+
+    @Override
+    public void createFakePhysicalBlock(final Location location, final org.bukkit.Material material) {
+        for(Player player : Bukkit.getOnlinePlayers()){
+            player.sendBlockChange(location, material.createBlockData());
+        }
+    }
+
+    @Override
+    public void createGhostBlock(final Location location, final org.bukkit.Material material) {
+        org.bukkit.craftbukkit.entity.CraftBlockDisplay entity = (org.bukkit.craftbukkit.entity.CraftBlockDisplay) this.spawnEntity(location, org.bukkit.entity.EntityType.BLOCK_DISPLAY);
+        entity.setBlock(material.createBlockData());
+        net.minecraft.network.protocol.game.ClientboundAddEntityPacket entityPacket = new net.minecraft.network.protocol.game.ClientboundAddEntityPacket(
+            entity.getHandle()
+        );
+        for(Player sender : Bukkit.getOnlinePlayers()){
+            CraftPlayer toSend = (CraftPlayer) sender;
+            toSend.getHandle().connection.send(entityPacket);
+        }
+    }
+    // Canvas end
 
     @Override
     public Chunk getChunkAt(Location location) {
