From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Mon, 3 Jun 2024 15:54:52 -0700
Subject: [PATCH] Add configurable chunk builder priority


diff --git a/src/main/java/io/papermc/paper/chunk/system/scheduling/ChunkTaskScheduler.java b/src/main/java/io/papermc/paper/chunk/system/scheduling/ChunkTaskScheduler.java
index 71e79ac5e87277de3793d5ddd55bad8318e32338..cc9d7e4adae625e76e73fadca2c8a2d8920b92ff 100644
--- a/src/main/java/io/papermc/paper/chunk/system/scheduling/ChunkTaskScheduler.java
+++ b/src/main/java/io/papermc/paper/chunk/system/scheduling/ChunkTaskScheduler.java
@@ -101,7 +101,7 @@ public final class ChunkTaskScheduler {
         workerThreads = new ca.spottedleaf.concurrentutil.executor.standard.PrioritisedThreadPool(
             "Paper Chunk System Worker Pool", newChunkSystemWorkerThreads,
             (final Thread thread, final Integer id) -> {
-                thread.setPriority(Thread.NORM_PRIORITY - 2);
+                thread.setPriority(Thread.NORM_PRIORITY + me.dueris.canvas.CanvasConfig.threadPriorityOffset);
                 thread.setName("Tuinity Chunk System Worker #" + id.intValue());
                 thread.setUncaughtExceptionHandler(io.papermc.paper.chunk.system.scheduling.NewChunkHolder.CHUNKSYSTEM_UNCAUGHT_EXCEPTION_HANDLER);
             }, (long)(20.0e6)); // 20ms
diff --git a/src/main/java/me/dueris/canvas/CanvasConfig.java b/src/main/java/me/dueris/canvas/CanvasConfig.java
index 9825b47fab89c05958ebfb93d830634fdd723ff2..8df9d7a33b1750621666ebe9223895eee36c407d 100644
--- a/src/main/java/me/dueris/canvas/CanvasConfig.java
+++ b/src/main/java/me/dueris/canvas/CanvasConfig.java
@@ -212,9 +212,11 @@ public class CanvasConfig {
     public static int maxEntitiesToReducePathfindingRates = 2;
 
     public static boolean disableFeatures = false;
+    public static int threadPriorityOffset = -2;
 
     private static void chunkGen() {
         disableFeatures = getBoolean("chunkgen.disablefeatures", disableFeatures);
+        threadPriorityOffset = getInt("chunkgen.threadPriorityOffset", threadPriorityOffset);
     }
 
     private static void general(){
