--- a/net/minecraft/world/entity/ai/attributes/AttributeInstance.java
+++ b/net/minecraft/world/entity/ai/attributes/AttributeInstance.java
@@ -8,6 +_,7 @@
 import java.util.Collection;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.function.Consumer;
 import javax.annotation.Nullable;
 import net.minecraft.core.Holder;
@@ -24,8 +_,10 @@
     private final Map<AttributeModifier.Operation, Map<ResourceLocation, AttributeModifier>> modifiersByOperation = Maps.newEnumMap(
         AttributeModifier.Operation.class
     );
-    private final Map<ResourceLocation, AttributeModifier> modifierById = new Object2ObjectArrayMap<>();
-    private final Map<ResourceLocation, AttributeModifier> permanentModifiers = new Object2ObjectArrayMap<>();
+    // Canvas start
+    private final Map<ResourceLocation, AttributeModifier> modifierById = new ConcurrentHashMap<>();
+    private final Map<ResourceLocation, AttributeModifier> permanentModifiers = new ConcurrentHashMap<>();
+    // Canvas end
     private double baseValue;
     private boolean dirty = true;
     private double cachedValue;
