--- a/net/minecraft/world/entity/ai/attributes/AttributeInstance.java
+++ b/net/minecraft/world/entity/ai/attributes/AttributeInstance.java
@@ -171,7 +_,7 @@
     }
 
     private Collection<AttributeModifier> getModifiersOrEmpty(AttributeModifier.Operation operation) {
-        return this.modifiersByOperation.getOrDefault(operation, Map.of()).values();
+        return this.modifiersByOperation.computeIfAbsent(operation, unnamed -> new Object2ObjectOpenHashMap<>()).values(); // Canvas - use optimized collection
     }
 
     public void replaceFrom(AttributeInstance instance) {
