--- a/io/papermc/paper/threadedregions/RegionizedServer.java
+++ b/io/papermc/paper/threadedregions/RegionizedServer.java
@@ -62,6 +_,14 @@
     }
 
     public void init() {
+        // Canvas start - default respawn dimension config
+        // validate respawn dimension config entry first
+        net.minecraft.resources.ResourceKey<net.minecraft.world.level.Level> respawnDimension = io.canvasmc.canvas.GlobalConfiguration.fetchRespawnDimensionKey();
+        ServerLevel level = MinecraftServer.getServer().getLevel(respawnDimension);
+        if (level == null) {
+            throw new IllegalStateException("Respawn dimension invalid, could not fetch dimension from value '" + respawnDimension + "'");
+        }
+        // Canvas end - default respawn dimension config
         // call init event _before_ scheduling anything
         new RegionizedServerInitEvent().callEvent();
 
