From 9e9970d708edecf8556f1b2db11dec8fc04b5815 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/io/papermc/paper/threadedregions/RegionShutdownThread.java b/io/papermc/paper/threadedregions/RegionShutdownThread.java
index aeb0f79fe57a1546dad74bb4dfa7dfa52c08702f..7a122dc53f5bde8ebee09b9dca18bab2780ac2ec 100644
--- a/io/papermc/paper/threadedregions/RegionShutdownThread.java
+++ b/io/papermc/paper/threadedregions/RegionShutdownThread.java
@@ -25,6 +25,7 @@ public final class RegionShutdownThread extends ca.spottedleaf.moonrise.common.u
         this.setUncaughtExceptionHandler((thread, thr) -> {
             LOGGER.error("Error shutting down server", thr);
         });
+        this.setPriority(10); // Canvas - bump priority
     }
 
     static ThreadedRegionizer.ThreadedRegion<TickRegions.TickRegionData, TickRegions.TickRegionSectionData> getRegion() {
@@ -138,6 +139,7 @@ public final class RegionShutdownThread extends ca.spottedleaf.moonrise.common.u
                         player.containerMenu.setCarried(ItemStack.EMPTY);
                         player.drop(carried, false);
                     }
+                    if (player.canvas$isTemporarilyBlocking) player.releaseUsingItem(); // Canvas - implement sword blocking
                 } catch (final Throwable thr) {
                     LOGGER.error("Failed to close player inventory for player: " + player, thr);
                 }
