--- a/io/papermc/paper/threadedregions/RegionizedServer.java
+++ b/io/papermc/paper/threadedregions/RegionizedServer.java
@@ -79,6 +_,12 @@
         // that essentially crashes the server on the first tick of a freshly generated world.
         // This fixes it, because the race condition is caused by the global tick not being run first
         // which is required to process ticket updates so we don't fail any thread ownership checks
+        // validate respawn dimension config entry first
+        net.minecraft.resources.ResourceKey<net.minecraft.world.level.Level> respawnDimension = io.canvasmc.canvas.Config.INSTANCE.fetchRespawnDimensionKey();
+        ServerLevel level = MinecraftServer.getServer().getLevel(respawnDimension);
+        if (level == null) {
+            throw new IllegalStateException("Respawn dimension invalid, could not fetch dimension from value '" + respawnDimension + "'");
+        }
         this.globalTick((int) this.tickHandle.getCurrentTick());
         this.tickHandle.setInitialStart(System.nanoTime());
         // Canvas end - rewrite scheduler
@@ -452,6 +_,42 @@
     // A global tick only updates things like weather / worldborder, basically anything in the world that is
     // NOT tied to a specific region, but rather shared amongst all of them.
     private void globalTick(final ServerLevel world, final int tickCount) {
+        // Canvas start - fix world loading/unloading
+        if (world.isUnloading) {
+            // we are unloading, return.
+            AtomicBoolean isSingularRegionAlive = new AtomicBoolean(false);
+            world.regioniser.computeForAllRegionsUnsynchronised((region) -> {
+                if (!region.getData().isClosed) {
+                    // region still active, possibly still saving or hasn't gotten to that yet
+                    isSingularRegionAlive.set(true);
+                }
+            });
+            if (!isSingularRegionAlive.get()) {
+                LOGGER.info("Finalizing unload of world '" + world.getWorld().getName() + "'");
+                // double check that *all* players are removed from this world, since they may
+                // be added in some weird spot during shutdown of the world that we didn't catch previously
+                for (final net.minecraft.server.level.ServerPlayer player : world.players()) {
+                    player.getBukkitEntity().kick(net.kyori.adventure.text.Component.text("World unloading"), org.bukkit.event.player.PlayerKickEvent.Cause.WORLD_UNLOAD);
+                }
+                // all regions saved and players are removed, we can now close this world
+                world.moonrise$getChunkTaskScheduler().halt(false, 0L);
+                world.chunkSource.getDataStorage().close();
+                world.moonrise$getChunkTaskScheduler().chunkHolderManager.close(false, true, true, false, false);
+                io.papermc.paper.FeatureHooks.closeEntityManager(world, false); // SPIGOT-6722: close entityManager // Paper - chunk system
+                try {
+                    world.levelStorageAccess.close();
+                } catch (java.io.IOException e) {
+                    throw new RuntimeException("Unable to close level storage access for world unload", e);
+                }
+                this.worlds.remove(world);
+                MinecraftServer.getServer().removeLevel(world);
+                MinecraftServer.getServer().server.worlds.remove(world.getWorld().getName().toLowerCase(java.util.Locale.ROOT));
+                LOGGER.info("Completed unload of world '{}'", world.getWorld().getName());
+                world.unloadFuture.complete(true);
+            }
+            return;
+        }
+        // Canvas end - fix world loading/unloading
         // needs
         // worldborder tick
         // advancing the weather cycle
@@ -480,6 +_,18 @@
 
         // time ticking (TODO API synchronisation?)
         this.tickTime(world, tickCount);
+        // Canvas start - fix end credits
+        List<net.minecraft.server.level.ServerPlayer> players = world.getServer().getPlayerList().seeingEndCredits;
+        for (java.util.Iterator<net.minecraft.server.level.ServerPlayer> iterator = players.iterator(); iterator.hasNext(); ) {
+            final net.minecraft.server.level.ServerPlayer serverPlayer = iterator.next();
+            if (serverPlayer.hasDisconnected()) {
+                iterator.remove();
+                LOGGER.info("Player disconnected while viewing end credits, removing from fallback keepalive");
+                continue;
+            }
+            serverPlayer.connection.keepConnectionAlive();
+        }
+        // Canvas end - fix end credits
 
         world.updateTickData();
 
