From bd14df94165f1c34a7f632a7de807b770022d5c5 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/io/papermc/paper/threadedregions/RegionizedServer.java b/io/papermc/paper/threadedregions/RegionizedServer.java
index ee70bdd32927fbbce5cb6975df31df41f3fa8edb..fb796be55d71a8819199a9a67d28e24f1acd555a 100644
--- a/io/papermc/paper/threadedregions/RegionizedServer.java
+++ b/io/papermc/paper/threadedregions/RegionizedServer.java
@@ -63,6 +63,14 @@ public final class RegionizedServer {
     }
 
     public void init() {
+        // Canvas start - default respawn dimension config
+        // validate respawn dimension config entry first
+        net.minecraft.resources.ResourceKey<net.minecraft.world.level.Level> respawnDimension = io.canvasmc.canvas.Config.INSTANCE.fetchRespawnDimensionKey();
+        ServerLevel level = MinecraftServer.getServer().getLevel(respawnDimension);
+        if (level == null) {
+            throw new IllegalStateException("Respawn dimension invalid, could not fetch dimension from value '" + respawnDimension + "'");
+        }
+        // Canvas end - default respawn dimension config
         // call init event _before_ scheduling anything
         new RegionizedServerInitEvent().callEvent();
 
