--- a/io/papermc/paper/threadedregions/RegionShutdownThread.java
+++ b/io/papermc/paper/threadedregions/RegionShutdownThread.java
@@ -25,6 +_,7 @@
         this.setUncaughtExceptionHandler((thread, thr) -> {
             LOGGER.error("Error shutting down server", thr);
         });
+        this.setPriority(10); // Canvas - bump priority
     }
 
     static ThreadedRegionizer.ThreadedRegion<TickRegions.TickRegionData, TickRegions.TickRegionSectionData> getRegion() {
@@ -138,6 +_,7 @@
                         player.containerMenu.setCarried(ItemStack.EMPTY);
                         player.drop(carried, false);
                     }
+                    if (player.canvas$isTemporarilyBlocking) player.releaseUsingItem(); // Canvas - implement sword blocking
                 } catch (final Throwable thr) {
                     LOGGER.error("Failed to close player inventory for player: " + player, thr);
                 }
@@ -148,9 +_,16 @@
             this.shuttingDown = null;
         }
     }
+    // Canvas start - region threading
+    private static volatile boolean canvas$isShutdown = false;
+    public static boolean canvas$isShutdown() {
+        return canvas$isShutdown;
+    }
+    // Canvas end - region threading
 
     @Override
     public final void run() {
+        canvas$isShutdown = true; // Canvas - region threading
         // await scheduler termination
         LOGGER.info("Awaiting scheduler termination for 60s...");
         if (TickRegions.getScheduler().halt(true, TimeUnit.SECONDS.toNanos(60L))) {
@@ -161,6 +_,32 @@
         }
 
         MinecraftServer.getServer().stopServer(); // stop part 1: most logic, kicking players, plugins, etc
+        // Canvas start - improve player removal logic
+        // this.playerList.removeAll(this.isRestarting); // Paper
+        // try { Thread.sleep(100); } catch (InterruptedException ex) {}
+        net.minecraft.server.players.PlayerList playerList = MinecraftServer.getServer().getPlayerList();
+        LOGGER.info("Removing players");
+        // waiting for 100ms is a stupid fix
+        List<java.util.concurrent.CompletableFuture<Void>> futures = new ArrayList<>();
+        for (final ServerPlayer serverPlayer : playerList.players) {
+            final net.minecraft.network.Connection connection = serverPlayer.connection.connection;
+            if (connection.channel == null || !connection.channel.isOpen()) continue;
+            java.util.concurrent.CompletableFuture<Void> future = new java.util.concurrent.CompletableFuture<>();
+            connection.channel.closeFuture().addListener(f -> future.complete(null));
+            futures.add(future);
+        }
+        // now, remove players and wait
+        playerList.removeAll(MinecraftServer.getServer().isRestarting);
+        if (!futures.isEmpty()) {
+            if (io.canvasmc.canvas.util.CFUtil.waitFor(java.util.concurrent.CompletableFuture.allOf(futures.toArray(new java.util.concurrent.CompletableFuture[0])),
+                java.util.concurrent.TimeUnit.SECONDS, 20)) {
+                LOGGER.info("Players removed");
+            } else {
+                LOGGER.warn("Players did not all get removed within 20 seconds, proceeding with shutdown anyways");
+            }
+        }
+        MinecraftServer.getServer().getConnection().stop();
+        // Canvas end - improve player removal logic
         // halt all chunk systems first so that any in-progress chunk generation stops
         LOGGER.info("Halting chunk systems...");
         for (final ServerLevel world : MinecraftServer.getServer().getAllLevels()) {
