--- a/io/papermc/paper/threadedregions/RegionizedServer.java
+++ b/io/papermc/paper/threadedregions/RegionizedServer.java
@@ -63,6 +_,14 @@
     }
 
     public void init() {
+        // Canvas start - default respawn dimension config
+        // validate respawn dimension config entry first
+        net.minecraft.resources.ResourceKey<net.minecraft.world.level.Level> respawnDimension = io.canvasmc.canvas.Config.INSTANCE.fetchRespawnDimensionKey();
+        ServerLevel level = MinecraftServer.getServer().getLevel(respawnDimension);
+        if (level == null) {
+            throw new IllegalStateException("Respawn dimension invalid, could not fetch dimension from value '" + respawnDimension + "'");
+        }
+        // Canvas end - default respawn dimension config
         // call init event _before_ scheduling anything
         new RegionizedServerInitEvent().callEvent();
 
@@ -361,6 +_,7 @@
 
         // player list
         MinecraftServer.getServer().getPlayerList().tick();
+        MinecraftServer.getServer().pearls.autosave(); // Canvas - region threading
     }
 
     private void tickPlayerSample() {
@@ -401,9 +_,14 @@
             }
 
             if (isNotOwnedByGlobalRegion(conn)) {
-                // we actually require that the owning regions remove the connection for us, as it is possible
-                // that ownership is transferred back to us
-                continue;
+                // Canvas start - region threading
+                if (!(conn.getPacketListener() instanceof ServerGamePacketListenerImpl gamePacketListener
+                    && gamePacketListener.player.canvas$isDisplayingEndCredits)) {
+                    // we actually require that the owning regions remove the connection for us,
+                    // as it is possible that ownership is transferred back to us
+                    continue;
+                }
+                // Canvas end - region threading
             }
 
             if (!conn.isConnected()) {
