--- a/io/papermc/paper/threadedregions/TeleportUtils.java
+++ b/io/papermc/paper/threadedregions/TeleportUtils.java
@@ -41,7 +_,7 @@
                             return;
                         }
                         (useFromRootVehicle ? realFrom.getRootVehicle() : realFrom).teleportAsync(
-                            ((CraftWorld)loc.getWorld()).getHandle(), pos, null, null, null,
+                            ((CraftWorld)loc.getWorld()).getHandle(), pos, yaw, pitch, null, // Canvas - region threading
                             cause, teleportFlags, onComplete
                         );
                     },
