--- a/io/papermc/paper/threadedregions/TickRegions.java
+++ b/io/papermc/paper/threadedregions/TickRegions.java
@@ -4,7 +_,7 @@
 public class TickRegions {
 
     public static int getRegionChunkShift() {
-        return ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ThreadedTicketLevelPropagator.SECTION_SHIFT;
+        return io.canvasmc.canvas.Config.INSTANCE.ticking.regionGridExponent; // Canvas - Threaded Regions
     }
 
 }
