--- a/io/papermc/paper/threadedregions/TickRegions.java
+++ b/io/papermc/paper/threadedregions/TickRegions.java
@@ -196,6 +_,15 @@
         private final RegionStats regionStats;
         private final java.util.concurrent.atomic.AtomicBoolean hasPackets = new java.util.concurrent.atomic.AtomicBoolean(false); // Canvas - rewrite scheduler
 
+        // Canvas start - fix world load and unload
+        public volatile boolean isClosed = false;
+        public final io.canvasmc.canvas.util.ticket.TicketHolder<io.canvasmc.canvas.util.ticket.SaveAllTicket> saveAllTicket = new io.canvasmc.canvas.util.ticket.TicketHolder<>(); // Canvas - fix commands with region threading
+
+        public void markClosed() {
+            this.isClosed = true;
+        }
+
+        // Canvas end - fix world load and unload
         private TickRegionData(final ThreadedRegionizer.ThreadedRegion<TickRegionData, TickRegionSectionData> region) {
             this.region = region;
             this.world = region.regioniser.world;
