--- a/io/papermc/paper/threadedregions/TickRegions.java
+++ b/io/papermc/paper/threadedregions/TickRegions.java
@@ -196,6 +_,15 @@
         private final RegionStats regionStats;
         private final java.util.concurrent.atomic.AtomicBoolean hasPackets = new java.util.concurrent.atomic.AtomicBoolean(false); // Canvas - rewrite scheduler
 
+        // Canvas start - fix world loading/unloading
+        public volatile boolean isClosed = false;
+        public final java.util.concurrent.atomic.AtomicReference<io.canvasmc.canvas.command.SaveAllTicket> saveAllTicket = new java.util.concurrent.atomic.AtomicReference<>(null); // Canvas - async save-all command
+
+        public void markClosed() {
+            this.isClosed = true;
+        }
+
+        // Canvas end - fix world loading/unloading
         private TickRegionData(final ThreadedRegionizer.ThreadedRegion<TickRegionData, TickRegionSectionData> region) {
             this.region = region;
             this.world = region.regioniser.world;
