--- a/io/papermc/paper/threadedregions/TickRegionScheduler.java
+++ b/io/papermc/paper/threadedregions/TickRegionScheduler.java
@@ -69,6 +_,7 @@
             public Thread newThread(final Runnable run) {
                 final Thread ret = new TickThreadRunner(this.threadGroup, run, "Folia Region Scheduler Thread #" + this.idGenerator.getAndIncrement());
                 ret.setUncaughtExceptionHandler(TickRegionScheduler.this::uncaughtException);
+                ret.setPriority(8); // Canvas - bump priority to match main thread in Paper
                 return ret;
             }
         };
