From bd14df94165f1c34a7f632a7de807b770022d5c5 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/io/papermc/paper/threadedregions/TickRegions.java b/io/papermc/paper/threadedregions/TickRegions.java
index 7336bd81d01823320c3d6377b35079f73fb17506..c0a4269ea775523782cc0fc7e89b9f349e589bc7 100644
--- a/io/papermc/paper/threadedregions/TickRegions.java
+++ b/io/papermc/paper/threadedregions/TickRegions.java
@@ -250,6 +250,7 @@ public final class TickRegions implements ThreadedRegionizer.RegionCallbacks<Tic
         private final RegionStats regionStats;
 
         private final AtomicBoolean hasPackets = new AtomicBoolean(false);
+        public final io.canvasmc.canvas.util.ticket.TicketHolder<io.canvasmc.canvas.util.ticket.SaveAllTicket> canvas$saveAllTicket = new io.canvasmc.canvas.util.ticket.TicketHolder<>(); // Canvas - fix commands with region threading
 
         private TickRegionData(final ThreadedRegionizer.ThreadedRegion<TickRegionData, TickRegionSectionData> region) {
             this.region = region;
