diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index e0bf9a3..f4e4c0c 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -180,6 +180,7 @@ public abstract class Player extends Avatar implements ContainerUser {
     public boolean affectsSpawning = true; // Paper - Affects Spawning API
     public net.kyori.adventure.util.TriState flyingFallDamage = net.kyori.adventure.util.TriState.NOT_SET; // Paper - flying fall damage
 	 public int sixRowEnderchestSlotCount = -1; // Purpur - Barrels and enderchests 6 rows
+    public boolean canvas$isTemporarilyBlocking = false; // Canvas - implement sword blocking
 
     // CraftBukkit start
     public boolean fauxSleeping;
@@ -288,7 +289,7 @@ public abstract class Player extends Avatar implements ContainerUser {
             this.setPos(d, this.getY(), d1);
         }
 
-        this.attackStrengthTicker++;
+        if (!io.canvasmc.canvas.Config.INSTANCE.combat.disableAttackHitDelay) this.attackStrengthTicker++; // Canvas - combat configs
         this.itemSwapTicker++;
         ItemStack mainHandItem = this.getMainHandItem();
         if (!ItemStack.matches(this.lastItemInMainHand, mainHandItem)) {
@@ -299,7 +300,7 @@ public abstract class Player extends Avatar implements ContainerUser {
             this.lastItemInMainHand = mainHandItem.copy();
         }
 
-        if (!this.isEyeInFluid(FluidTags.WATER) && this.isEquipped(Items.TURTLE_HELMET)) {
+        if (!canvas$isEyeInWater() && this.isEquipped(Items.TURTLE_HELMET)) { // Canvas - optimize fluidOnEyes
             this.turtleHelmetTick();
         }
 
@@ -328,7 +329,7 @@ public abstract class Player extends Avatar implements ContainerUser {
     }
 
     protected boolean updateIsUnderwater() {
-        this.wasUnderwater = this.isEyeInFluid(FluidTags.WATER);
+        this.wasUnderwater = this.canvas$isEyeInWater(); // Canvas - optimize fluidOnEyes
         return this.wasUnderwater;
     }
 
@@ -649,7 +650,7 @@ public abstract class Player extends Avatar implements ContainerUser {
         }
 
         destroySpeed *= (float)this.getAttributeValue(Attributes.BLOCK_BREAK_SPEED);
-        if (this.isEyeInFluid(FluidTags.WATER)) {
+        if (this.canvas$isEyeInWater()) { // Canvas - optimize fluidOnEyes
             destroySpeed *= (float)this.getAttribute(Attributes.SUBMERGED_MINING_SPEED).getValue();
         }
 
@@ -1051,9 +1052,10 @@ public abstract class Player extends Avatar implements ContainerUser {
                     f += weaponItem.getItem().getAttackDamageBonus(target, f, damageSource);
                     boolean flag2 = flag && this.canCriticalAttack(target);
                     flag2 = flag2 && !this.level().paperConfig().entities.behavior.disablePlayerCrits; // Paper - Toggleable player crits
+                    flag2 = flag2 && (!io.canvasmc.canvas.Config.INSTANCE.combat.disableCritsWhileSprinting || !this.isSprinting()); // Canvas - combat configs
                     if (flag2) {
                         damageSource = damageSource.critical(); // Paper - critical damage API
-                        f *= 1.5F;
+                        f *= io.canvasmc.canvas.Config.INSTANCE.combat.criticalHitMultiplier; // Canvas - combat configs
                     }
 
                     float f2 = f + f1;
@@ -1067,7 +1069,7 @@ public abstract class Player extends Avatar implements ContainerUser {
                     boolean flag3 = target.hurtOrSimulate(damageSource, f2);
                     if (flag3) {
                         this.causeExtraKnockback(target, this.getKnockback(target, damageSource) + (flag1 ? 0.5F : 0.0F), deltaMovement);
-                        if (isSweepAttack) {
+                        if (isSweepAttack && !io.canvasmc.canvas.Config.INSTANCE.combat.disableSweepingEdge) { // Canvas - combat configs
                             this.doSweepAttack(target, f, damageSource, attackStrengthScale);
                         }
 
@@ -1152,7 +1154,7 @@ public abstract class Player extends Avatar implements ContainerUser {
         if (target instanceof LivingEntity) {
             float f = previousHealth - ((LivingEntity)target).getHealth();
             this.awardStat(Stats.DAMAGE_DEALT, Math.round(f * 10.0F));
-            if (this.level() instanceof ServerLevel && f > 2.0F) {
+            if (!io.canvasmc.canvas.Config.INSTANCE.particles.disableNewCombatParticles && this.level() instanceof ServerLevel && f > 2.0F) { // Canvas - particles config
                 int i = (int)(f * 0.5);
                 ((ServerLevel)this.level())
                     .sendParticles(ParticleTypes.DAMAGE_INDICATOR, target.getX(), target.getY(0.5), target.getZ(), i, 0.1, 0.0, 0.1, 0.2);
@@ -1982,6 +1984,7 @@ public abstract class Player extends Avatar implements ContainerUser {
     }
 
     public float getCurrentItemAttackStrengthDelay() {
+        if (io.canvasmc.canvas.Config.INSTANCE.combat.disableAttackHitDelay) return 0.0F; // Canvas - combat configs
         return (float)(1.0 / this.getAttributeValue(Attributes.ATTACK_SPEED) * 20.0);
     }
 
@@ -1992,6 +1995,7 @@ public abstract class Player extends Avatar implements ContainerUser {
     }
 
     public float getAttackStrengthScale(float adjustTicks) {
+        if (io.canvasmc.canvas.Config.INSTANCE.combat.disableAttackHitDelay) return 1.0F; // Canvas - combat configs
         return Mth.clamp((this.attackStrengthTicker + adjustTicks) / this.getCurrentItemAttackStrengthDelay(), 0.0F, 1.0F);
     }
 
