--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -175,6 +_,7 @@
     public boolean affectsSpawning = true; // Paper - Affects Spawning API
     public net.kyori.adventure.util.TriState flyingFallDamage = net.kyori.adventure.util.TriState.NOT_SET; // Paper - flying fall damage
     public int sixRowEnderchestSlotCount = -1; // Purpur - Barrels and enderchests 6 rows
+    public boolean canvas$isTemporarilyBlocking = false; // Canvas - implement sword blocking
 
     // CraftBukkit start
     public boolean fauxSleeping;
@@ -282,7 +_,7 @@
             this.setPos(nx, this.getY(), nz);
         }
 
-        this.attackStrengthTicker++;
+        if (!io.canvasmc.canvas.Config.INSTANCE.combat.disableAttackHitDelay) this.attackStrengthTicker++; // Canvas - combat configs
         this.itemSwapTicker++;
         ItemStack mainHandItemStack = this.getMainHandItem();
         if (!ItemStack.matches(this.lastItemInMainHand, mainHandItemStack)) {
@@ -1043,9 +_,10 @@
                     baseDamage += attackingItemStack.getItem().getAttackDamageBonus(entity, baseDamage, damageSource);
                     boolean criticalAttack = fullStrengthAttack && this.canCriticalAttack(entity);
                     criticalAttack = criticalAttack && !this.level().paperConfig().entities.behavior.disablePlayerCrits; // Paper - Toggleable player crits
+                    criticalAttack = criticalAttack && (!io.canvasmc.canvas.Config.INSTANCE.combat.disableCritsWhileSprinting || !this.isSprinting()); // Canvas - combat configs
                     if (criticalAttack) {
                         damageSource = damageSource.critical(); // Paper - critical damage API
-                        baseDamage *= 1.5F;
+                        baseDamage *= io.canvasmc.canvas.Config.INSTANCE.combat.criticalHitMultiplier; // Canvas - combat configs
                     }
 
                     float totalDamage = baseDamage + magicBoost;
@@ -1059,7 +_,7 @@
                     boolean wasHurt = entity.hurtOrSimulate(damageSource, totalDamage);
                     if (wasHurt) {
                         this.causeExtraKnockback(entity, this.getKnockback(entity, damageSource) + (knockbackAttack ? 0.5F : 0.0F), oldMovement);
-                        if (sweepAttack) {
+                        if (sweepAttack && !io.canvasmc.canvas.Config.INSTANCE.combat.disableSweepingEdge) { // Canvas - combat configs
                             this.doSweepAttack(entity, baseDamage, damageSource, attackStrengthScale);
                         }
 
@@ -1151,7 +_,7 @@
         if (entity instanceof LivingEntity) {
             float actualDamage = oldLivingEntityHealth - ((LivingEntity)entity).getHealth();
             this.awardStat(Stats.DAMAGE_DEALT, Math.round(actualDamage * 10.0F));
-            if (this.level() instanceof ServerLevel && actualDamage > 2.0F) {
+            if (!io.canvasmc.canvas.Config.INSTANCE.particles.disableNewCombatParticles && this.level() instanceof ServerLevel && actualDamage > 2.0F) { // Canvas - particles config
                 int count = (int)(actualDamage * 0.5);
                 ((ServerLevel)this.level())
                     .sendParticles(ParticleTypes.DAMAGE_INDICATOR, entity.getX(), entity.getY(0.5), entity.getZ(), count, 0.1, 0.0, 0.1, 0.2);
@@ -1946,6 +_,7 @@
     }
 
     public float getCurrentItemAttackStrengthDelay() {
+        if (io.canvasmc.canvas.Config.INSTANCE.combat.disableAttackHitDelay) return 0.0F; // Canvas - combat configs
         return (float)(1.0 / this.getAttributeValue(Attributes.ATTACK_SPEED) * 20.0);
     }
 
@@ -1956,6 +_,7 @@
     }
 
     public float getAttackStrengthScale(final float a) {
+        if (io.canvasmc.canvas.Config.INSTANCE.combat.disableAttackHitDelay) return 1.0F; // Canvas - combat configs
         return Mth.clamp((this.attackStrengthTicker + a) / this.getCurrentItemAttackStrengthDelay(), 0.0F, 1.0F);
     }
 
