From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Sat, 17 Jan 2026 20:37:55 -0800
Subject: [PATCH] PR Party 2


diff --git a/net/minecraft/commands/Commands.java b/net/minecraft/commands/Commands.java
index 368ed1b17857a53c7ac307d84e72781d3489463c..668ce850c9197b0f52859d74a8d3a8a2d3a390e3 100644
--- a/net/minecraft/commands/Commands.java
+++ b/net/minecraft/commands/Commands.java
@@ -202,7 +202,7 @@ public class Commands {
         //DataPackCommand.register(this.dispatcher, context); // Folia - region threading - TODO
         //DebugCommand.register(this.dispatcher); // Folia - region threading - TODO
         DefaultGameModeCommands.register(this.dispatcher);
-        //DialogCommand.register(this.dispatcher, context); // Folia - region threading - TODO
+        DialogCommand.register(this.dispatcher, context);
         DifficultyCommand.register(this.dispatcher);
         EffectCommands.register(this.dispatcher, context);
         EmoteCommands.register(this.dispatcher);
diff --git a/net/minecraft/server/commands/DialogCommand.java b/net/minecraft/server/commands/DialogCommand.java
index 15ff470c19c3bad50133bca3ef3dc0fa5c5b421d..7b14299049754ac786f212c8a746b79e49822869 100644
--- a/net/minecraft/server/commands/DialogCommand.java
+++ b/net/minecraft/server/commands/DialogCommand.java
@@ -46,7 +46,11 @@ public class DialogCommand {
 
     private static int showDialog(CommandSourceStack source, Collection<ServerPlayer> targets, Holder<Dialog> dialog) {
         for (ServerPlayer serverPlayer : targets) {
-            serverPlayer.openDialog(dialog);
+            // Folia start - region threading
+            serverPlayer.getBukkitEntity().taskScheduler.schedule((ServerPlayer nmsPlayer) -> {
+                serverPlayer.openDialog(dialog);
+            }, null, 1L);
+            // Folia end - region threading
         }
 
         if (targets.size() == 1) {
@@ -60,7 +64,11 @@ public class DialogCommand {
 
     private static int clearDialog(CommandSourceStack source, Collection<ServerPlayer> targets) {
         for (ServerPlayer serverPlayer : targets) {
-            serverPlayer.connection.send(ClientboundClearDialogPacket.INSTANCE);
+            // Folia start - region threading
+            serverPlayer.getBukkitEntity().taskScheduler.schedule((ServerPlayer nmsPlayer) -> {
+                serverPlayer.connection.send(ClientboundClearDialogPacket.INSTANCE);
+            }, null, 1L);
+            // Folia end - region threading
         }
 
         if (targets.size() == 1) {
