--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -25,6 +_,12 @@
      */
     Key BRAND_PURPUR_ID = Key.key("purpurmc", "purpur");
     // Purpur end
+    // Canvas start
+    /**
+     * The brand id for Canvas.
+     */
+    Key BRAND_CANVAS_ID = Key.key("canvasmc", "canvas");
+    // Canvas end
     /**
      * Gets the {@code ServerBuildInfo}.
      *
