--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -206,7 +_,7 @@
             boolean moved = Mth.floor(this.xo) != Mth.floor(this.getX())
                 || Mth.floor(this.yo) != Mth.floor(this.getY())
                 || Mth.floor(this.zo) != Mth.floor(this.getZ());
-            int rate = moved ? 2 : 40;
+            int rate = moved && !level().canvasConfig().entities.itemEntities.itemEntitiesWaitTwoSecondsForMergeCheckAlways ? 2 : 40; // Canvas - item entity wait 2s for merge check config
             if (this.tickCount % rate == 0 && !this.level().isClientSide() && this.isMergable()) {
                 this.mergeWithNeighbours();
             }
@@ -268,7 +_,7 @@
                     }
                     // Paper end - Fix items merging through walls
                     this.tryToMerge(entity);
-                    if (this.isRemoved()) {
+                    if (!this.isMergable()) { // Canvas - if this isn't mergeable, don't merge - we used to only check for if we are removed, ignoring if we aren't even mergeable ourselves anymore
                         break;
                     }
                 }
@@ -329,6 +_,15 @@
         return !this.getItem().canBeHurtBy(this.damageSources().inFire()) || super.fireImmune();
     }
 
+    // Canvas start - invulnerable to lightning and explosions config
+    @Override
+    public boolean isInvulnerableToBase(final DamageSource damageSource) {
+        if (level().canvasConfig().entities.itemEntities.itemEntitiesImmuneToExplosions && damageSource.is(net.minecraft.tags.DamageTypeTags.IS_EXPLOSION)) return true;
+        if (level().canvasConfig().entities.itemEntities.itemEntitiesImmuneToLightning && damageSource.is(net.minecraft.tags.DamageTypeTags.IS_LIGHTNING)) return true;
+        return super.isInvulnerableToBase(damageSource);
+    }
+
+    // Canvas end - invulnerable to lightning and explosions config
     @Override
     protected boolean shouldPlayLavaHurtSound() {
         return this.health <= 0 || this.tickCount % 10 == 0;
@@ -599,4 +_,13 @@
     public @Nullable SlotAccess getSlot(final int slot) {
         return slot == 0 ? SlotAccess.of(this::getItem, this::setItem) : super.getSlot(slot);
     }
+    // Canvas start - MC-206922
+
+    @Override
+    public void thunderHit(final ServerLevel level, final net.minecraft.world.entity.LightningBolt lightning) {
+        if (!io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaFixes.mc206922 || this.tickCount > 8) {
+            super.thunderHit(level, lightning);
+        }
+    }
+    // Canvas end - MC-206922
 }
