--- a/net/minecraft/world/entity/item/ItemEntity.java
+++ b/net/minecraft/world/entity/item/ItemEntity.java
@@ -327,6 +_,15 @@
         return !this.getItem().canBeHurtBy(this.damageSources().inFire()) || super.fireImmune();
     }
 
+    // Canvas start - invulnerable to lightning and explosions config
+    @Override
+    public boolean isInvulnerableToBase(final DamageSource damageSource) {
+        if (io.canvasmc.canvas.Config.INSTANCE.itemEntitiesImmuneToExplosions && damageSource.is(net.minecraft.tags.DamageTypeTags.IS_EXPLOSION)) return true;
+        if (io.canvasmc.canvas.Config.INSTANCE.itemEntitiesImmuneToLightning && damageSource.is(net.minecraft.tags.DamageTypeTags.IS_LIGHTNING)) return true;
+        return super.isInvulnerableToBase(damageSource);
+    }
+
+    // Canvas end - invulnerable to lightning and explosions config
     @Override
     protected boolean shouldPlayLavaHurtSound() {
         return this.health <= 0 || this.tickCount % 10 == 0;
@@ -593,4 +_,13 @@
     public @Nullable SlotAccess getSlot(final int slot) {
         return slot == 0 ? SlotAccess.of(this::getItem, this::setItem) : super.getSlot(slot);
     }
+    // Canvas start - MC-206922
+
+    @Override
+    public void thunderHit(final ServerLevel level, final net.minecraft.world.entity.LightningBolt lightning) {
+        if (!io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaFixes.mc206922 || this.tickCount > 8) {
+            super.thunderHit(level, lightning);
+        }
+    }
+    // Canvas end - MC-206922
 }
