--- a/net/minecraft/world/entity/item/PrimedTnt.java
+++ b/net/minecraft/world/entity/item/PrimedTnt.java
@@ -2,6 +_,7 @@
 
 import java.util.Optional;
 import javax.annotation.Nullable;
+import io.canvasmc.canvas.Config;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.particles.ParticleTypes;
 import net.minecraft.core.registries.Registries;
@@ -57,6 +_,15 @@
     private boolean usedPortal;
     public float explosionPower = 4.0F;
     public boolean isIncendiary = false; // CraftBukkit - add field
+    // Canvas start - merge tnt logic
+    float setPower = 4.0F;
+
+    public void addPower() {
+        if (setPower < 16.0F) {
+            setPower += 0.1F;
+        }
+    }
+    // Canvas end
 
     public PrimedTnt(EntityType<? extends PrimedTnt> entityType, Level level) {
         super(entityType, level);
@@ -160,6 +_,13 @@
             return;
         }
         // CraftBukkit end
+        // Canvas start - merge tnt logic
+        // The reason we dont add to HEAD is so we respect the plugin event.
+        if (Config.INSTANCE.blocks.tnt.mergeTntLogic) {
+            this.level().explode(this, Explosion.getDefaultDamageSource(this.level(), this), this.usedPortal ? USED_PORTAL_DAMAGE_CALCULATOR : null, this.getX(), this.getY(0.0625), this.getZ(), this.setPower, false, Level.ExplosionInteraction.TNT);
+            return;
+        }
+        // Canvas end
         this.level()
             .explode(
                 this,
