From 56a92b2c479a7161c340134ba931fb3d6a8e6f79 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/block/Blocks.java b/net/minecraft/world/level/block/Blocks.java
index f94f8049572c860652c97c10ecd888ef2c338262..76484a0e6dfabac532134f941bfe7fdfb870a72f 100644
--- a/net/minecraft/world/level/block/Blocks.java
+++ b/net/minecraft/world/level/block/Blocks.java
@@ -5983,6 +5983,12 @@ public class Blocks {
     private static Block register(
         final ResourceKey<Block> id, final Function<BlockBehaviour.Properties, Block> factory, final BlockBehaviour.Properties properties
     ) {
+        // Canvas start - MC-223153
+        if (io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaFixes.mc223153 && id.equals(BlockItemIds.RAW_COPPER_BLOCK.block())) {
+            Block block = factory.apply(properties.sound(SoundType.COPPER).setId(id));
+            return Registry.register(BuiltInRegistries.BLOCK, id, block);
+        }
+        // Canvas end - MC-223153
         Block block = factory.apply(properties.setId(id));
         return Registry.register(BuiltInRegistries.BLOCK, id, block);
     }
