From 56a92b2c479a7161c340134ba931fb3d6a8e6f79 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/block/EnchantingTableBlock.java b/net/minecraft/world/level/block/EnchantingTableBlock.java
index 8d3a6345588384e755e1866c70afcb905070dbd8..5f818537ccf6d13cf62904342b8b388efcb8b25c 100644
--- a/net/minecraft/world/level/block/EnchantingTableBlock.java
+++ b/net/minecraft/world/level/block/EnchantingTableBlock.java
@@ -46,8 +46,32 @@ public class EnchantingTableBlock extends BaseEntityBlock {
     }
 
     public static boolean isValidBookShelf(final Level level, final BlockPos pos, final BlockPos offset) {
-        return level.getBlockState(pos.offset(offset)).is(BlockTags.ENCHANTMENT_POWER_PROVIDER)
-            && level.getBlockState(pos.offset(offset.getX() / 2, offset.getY(), offset.getZ() / 2)).is(BlockTags.ENCHANTMENT_POWER_TRANSMITTER);
+        // Canvas start - block configs
+        final BlockState offsetState = level.getBlockState(pos.offset(offset));
+
+        // must be a power provider
+        if (!offsetState.is(BlockTags.ENCHANTMENT_POWER_PROVIDER)) {
+            if (level.canvasConfig().blocks.fullChiseledBookShelvesCountAsValidEnchantPowerSources
+                && offsetState.is(Blocks.CHISELED_BOOKSHELF)) {
+
+                // all slots have to be filled for it to be a valid bookshelf
+                for (final net.minecraft.world.level.block.state.properties.BooleanProperty slotOccupiedProperty : ChiseledBookShelfBlock.SLOT_OCCUPIED_PROPERTIES) {
+                    if (!offsetState.getValue(slotOccupiedProperty)) {
+                        return false;
+                    }
+                }
+
+            }
+            else return false;
+        }
+
+        // check if something is blocking the shelf
+        if (!level.getBlockState(pos.offset(offset.getX() / 2, offset.getY(), offset.getZ() / 2)).is(BlockTags.ENCHANTMENT_POWER_TRANSMITTER)) {
+            return false;
+        }
+
+        return true;
+        // Canvas end - block configs
     }
 
     @Override
