From 56a92b2c479a7161c340134ba931fb3d6a8e6f79 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/block/LeavesBlock.java b/net/minecraft/world/level/block/LeavesBlock.java
index 23ff958d35aa9199f361aa88cca7fd676f035709..a6b5a539a2d069eca8e806b5c5fdd945338faec1 100644
--- a/net/minecraft/world/level/block/LeavesBlock.java
+++ b/net/minecraft/world/level/block/LeavesBlock.java
@@ -65,6 +65,7 @@ public abstract class LeavesBlock extends Block implements SimpleWaterloggedBloc
 
     @Override
     protected void randomTick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
+        if (level.canvasConfig().farming.disableLeafDecay) return; // Canvas - disable leaf decay config
         if (this.decaying(state)) {
             // CraftBukkit start
             org.bukkit.event.block.LeavesDecayEvent event = new org.bukkit.event.block.LeavesDecayEvent(org.bukkit.craftbukkit.block.CraftBlock.at(level, pos));
@@ -85,6 +86,7 @@ public abstract class LeavesBlock extends Block implements SimpleWaterloggedBloc
 
     @Override
     protected void tick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
+        if (level.canvasConfig().farming.disableLeafDecay) return; // Canvas - disable leaf decay config
         level.setBlock(pos, updateDistance(state, level, pos), Block.UPDATE_ALL);
     }
 
