From 56a92b2c479a7161c340134ba931fb3d6a8e6f79 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/block/NetherPortalBlock.java b/net/minecraft/world/level/block/NetherPortalBlock.java
index a91580d397fc27f072f383876819843d08352942..68c1e2e311626c8470178f7a153bb780af158344 100644
--- a/net/minecraft/world/level/block/NetherPortalBlock.java
+++ b/net/minecraft/world/level/block/NetherPortalBlock.java
@@ -122,7 +122,10 @@ public class NetherPortalBlock extends Block implements Portal {
         final boolean isPrecise
     ) {
         if (!new io.papermc.paper.event.entity.EntityInsideBlockEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.block.CraftBlock.at(level, pos)).callEvent()) { return; } // Paper - Add EntityInsideBlockEvent
-        if (entity.canUsePortal(false)) {
+        // Canvas start - blacklist non player entities from portals configs
+        if (entity.canUsePortal(false) &&
+            (!io.canvasmc.canvas.GlobalConfiguration.getInstance().blacklistNonPlayerEntitiesFromEnteringNetherPortals || entity instanceof net.minecraft.server.level.ServerPlayer)) {
+        // Canvas end - blacklist non player entities from portals configs
             // CraftBukkit start - Entity in portal
             org.bukkit.event.entity.EntityPortalEnterEvent event = new org.bukkit.event.entity.EntityPortalEnterEvent(entity.getBukkitEntity(), org.bukkit.craftbukkit.util.CraftLocation.toBukkit(pos, level), org.bukkit.PortalType.NETHER); // Paper - add portal type
             level.getCraftServer().getPluginManager().callEvent(event);
