From fa0385f66589f695ee335f5beee4acd758b26a48 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/block/state/BlockBehaviour.java b/net/minecraft/world/level/block/state/BlockBehaviour.java
index 42918460ae0fffdde1e7fab5baec6a3662b10be7..85181249ac6cab266c51a4ce319343fcd921ef0b 100644
--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -488,6 +488,24 @@ public abstract class BlockBehaviour implements FeatureElement {
         private boolean emptyCollisionShape;
         private boolean emptyConstantCollisionShape;
         private VoxelShape constantCollisionShape;
+        // Canvas start - optimize path type cache
+        // inspired by Cache-block-state-tags patch from Leaf
+        public byte pathType; // 8bit, better this enum, as it only has 25 values
+        public int tagFlag; // caches blockstate flags
+
+        public void canvas$initCaches() {
+            this.tagFlag = io.canvasmc.canvas.world.block.BlockTags.init(asState());
+            this.pathType = (byte) net.minecraft.world.level.pathfinder.WalkNodeEvaluator.canvas$getPathTypeFromBlockState(this.asState()).ordinal();
+        }
+
+        public static void canvas$initPathTypeCache() {
+            for (Block block : BuiltInRegistries.BLOCK) {
+                for (BlockState blockState : block.getStateDefinition().getPossibleStates()) {
+                    blockState.canvas$initCaches();
+                }
+            }
+        }
+        // Canvas end - optimize path type cache
 
         private static void initCaches(final VoxelShape shape, final boolean neighbours) {
             ((ca.spottedleaf.moonrise.patches.collisions.shape.CollisionVoxelShape)shape).moonrise$isFullBlock();
