--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -490,6 +_,24 @@
         private boolean emptyCollisionShape;
         private boolean emptyConstantCollisionShape;
         private VoxelShape constantCollisionShape;
+        // Canvas start - optimize path type cache
+        // inspired by Cache-block-state-tags patch from Leaf
+        public byte pathType; // 8bit, better this enum, as it only has 25 values
+        public int tagFlag; // caches blockstate flags
+
+        public void canvas$initCaches() {
+            this.tagFlag = io.canvasmc.canvas.world.block.BlockTags.init(asState());
+            this.pathType = (byte) net.minecraft.world.level.pathfinder.WalkNodeEvaluator.canvas$getPathTypeFromBlockState(this.asState()).ordinal();
+        }
+
+        public static void canvas$initPathTypeCache() {
+            for (Block block : BuiltInRegistries.BLOCK) {
+                for (BlockState state : block.getStateDefinition().getPossibleStates()) {
+                    state.canvas$initCaches();
+                }
+            }
+        }
+        // Canvas end - optimize path type cache
 
         private static void initCaches(final VoxelShape shape, final boolean neighbours) {
             ((ca.spottedleaf.moonrise.patches.collisions.shape.CollisionVoxelShape)shape).moonrise$isFullBlock();
