--- a/net/minecraft/world/entity/ai/sensing/GolemSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/GolemSensor.java
@@ -34,7 +_,15 @@
     public static void checkForNearbyGolem(LivingEntity livingEntity) {
         Optional<List<LivingEntity>> memory = livingEntity.getBrain().getMemory(MemoryModuleType.NEAREST_LIVING_ENTITIES);
         if (!memory.isEmpty()) {
-            boolean flag = memory.get().stream().anyMatch(entity -> entity.getType().equals(EntityType.IRON_GOLEM));
+            // Canvas start - remove streams
+            boolean flag = false;
+            for (LivingEntity entity : memory.get()) {
+                if (entity.getType().equals(EntityType.IRON_GOLEM)) {
+                    flag = true;
+                    break;
+                }
+            }
+            // Canvas end - remove streams
             if (flag) {
                 golemDetected(livingEntity);
             }
