From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/network/PacketEncoder.java b/net/minecraft/network/PacketEncoder.java
index 27316db9eb78660c6f54e1d1ea4dc7d584c7bb55..167b7d554d7f4f78faedef01439f56212809eb02 100644
--- a/net/minecraft/network/PacketEncoder.java
+++ b/net/minecraft/network/PacketEncoder.java
@@ -60,7 +60,7 @@ public class PacketEncoder<T extends PacketListener> extends MessageToByteEncode
     // packet size is encoded into 3-byte varint
     private static final int MAX_FINAL_PACKET_SIZE = (1 << 21) - 1;
     // Vanilla Max size for the encoder (before compression)
-    private static final int MAX_PACKET_SIZE = 8388608;
+    private static int MAX_PACKET_SIZE = io.canvasmc.canvas.GlobalConfiguration.getInstance().networking.maximumPacketBytes; // Canvas - configurable max packet bytes
 
     public static class PacketTooLargeException extends RuntimeException {
         private final Packet<?> packet;
