From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/network/codec/ByteBufCodecs.java b/net/minecraft/network/codec/ByteBufCodecs.java
index 43270171bf22ed87c015593289380ea05eebfb57..487fd97b96863a0f2de593a8600b02f2cc01277f 100644
--- a/net/minecraft/network/codec/ByteBufCodecs.java
+++ b/net/minecraft/network/codec/ByteBufCodecs.java
@@ -774,9 +774,35 @@ public interface ByteBufCodecs {
 
             @Override
             public void encode(ByteBuf buffer, JsonElement value) {
+                // Canvas start - no chat reports
+                if (io.canvasmc.canvas.Config.INSTANCE.enableNoChatReports && this.canvas$isServerStatusElement(value)) {
+                    com.google.gson.JsonObject object = value.getAsJsonObject();
+                    object.addProperty("preventsChatReports", true);
+
+                    String string = GSON.toJson(object);
+                    Utf8String.write(buffer, string, 32767);
+                    return;
+                }
+                // Canvas end - no chat reports
                 String string = GSON.toJson(value);
                 Utf8String.write(buffer, string, maxLength);
             }
+            // Canvas start - no chat reports
+
+            private boolean canvas$isServerStatusElement(JsonElement element) {
+                if (!element.isJsonObject())
+                    return false;
+
+                com.google.gson.JsonObject object = element.getAsJsonObject();
+
+                if (!object.has("version") || !object.get("version").isJsonObject())
+                    return false;
+
+                com.google.gson.JsonObject version = object.get("version").getAsJsonObject();
+
+                return version.has("name") && version.has("protocol");
+            }
+            // Canvas end - no chat reports
         };
     }
 }
