From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/block/entity/BlockEntity.java b/net/minecraft/world/level/block/entity/BlockEntity.java
index c8699e6461f9def2f9fbd96d466b8a8fb75f2bc5..34a76b0000f46847004fffc9d84bbbc10f958784 100644
--- a/net/minecraft/world/level/block/entity/BlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BlockEntity.java
@@ -50,6 +50,7 @@ public abstract class BlockEntity implements DebugValueSource {
     protected boolean remove;
     private BlockState blockState;
     private DataComponentMap components = DataComponentMap.EMPTY;
+    public boolean canvas$isValid; // Canvas - optimize block entities
 
     // Folia start - region ticking
     public void updateTicks(final long fromTickOffset, final long fromRedstoneTimeOffset) {
@@ -63,6 +64,7 @@ public abstract class BlockEntity implements DebugValueSource {
         this.validateBlockState(blockState);
         this.blockState = blockState;
         this.persistentDataContainer = new org.bukkit.craftbukkit.persistence.CraftPersistentDataContainer(DATA_TYPE_REGISTRY); // Paper - always init
+        this.canvas$isValid = type.isValid(blockState); // Canvas - optimize block entities
     }
 
     private void validateBlockState(BlockState state) {
@@ -236,6 +238,11 @@ public abstract class BlockEntity implements DebugValueSource {
             if (IGNORE_TILE_UPDATES.get().booleanValue()) return; // Paper - Perf: Optimize Hoppers // Folia - region threading
             setChanged(this.level, this.worldPosition, this.blockState);
         }
+        // Canvas start - block entity sleeping
+        if (this instanceof io.canvasmc.canvas.world.block.SleepingBlockEntity sleepingBlockEntity) {
+            sleepingBlockEntity.onChange();
+        }
+        // Canvas end - block entity sleeping
     }
 
     protected static void setChanged(Level level, BlockPos pos, BlockState state) {
@@ -306,6 +313,7 @@ public abstract class BlockEntity implements DebugValueSource {
     public void setBlockState(BlockState blockState) {
         this.validateBlockState(blockState);
         this.blockState = blockState;
+        this.canvas$isValid = this.type.isValid(blockState); // Canvas - optimize block entities
     }
 
     protected void applyImplicitComponents(DataComponentGetter componentGetter) {
@@ -383,7 +391,7 @@ public abstract class BlockEntity implements DebugValueSource {
 
     // CraftBukkit start - add method
     public org.bukkit.inventory.@Nullable InventoryHolder getOwner() {
-        return getOwner(true);
+        return getOwner(io.canvasmc.canvas.Config.INSTANCE.tileEntitySnapshotCreation); // Canvas - don't use snapshots for acquiring blockstate
     }
 
     public org.bukkit.inventory.@Nullable InventoryHolder getOwner(boolean useSnapshot) {
