From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/block/entity/CampfireBlockEntity.java b/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
index a27f578eb4d2dbcdb5cd72cb0aa3ec98cdb543c2..fa97e56fe971ea2f519aee233c50aed4198bd6f9 100644
--- a/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
@@ -39,7 +39,7 @@ import net.minecraft.world.level.storage.ValueOutput;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
-public class CampfireBlockEntity extends BlockEntity implements Clearable {
+public class CampfireBlockEntity extends BlockEntity implements Clearable, io.canvasmc.canvas.world.block.SleepingBlockEntity { // Canvas - block entity sleeping
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final int BURN_COOL_SPEED = 2;
     private static final int NUM_SLOTS = 4;
@@ -47,6 +47,31 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
     public final int[] cookingProgress = new int[4];
     public final int[] cookingTime = new int[4];
     public final boolean[] stopCooking = new boolean[4]; // Paper - Add more Campfire API
+    // Canvas start - block entity sleeping
+    private net.minecraft.world.level.chunk.LevelChunk.RebindableTickingBlockEntityWrapper tickWrapper = null;
+    private TickingBlockEntity sleepingTicker = null;
+
+    @Override
+    public net.minecraft.world.level.chunk.LevelChunk.RebindableTickingBlockEntityWrapper lithium$getTickWrapper() {
+        return this.tickWrapper;
+    }
+
+    @Override
+    public void lithium$setTickWrapper(net.minecraft.world.level.chunk.LevelChunk.RebindableTickingBlockEntityWrapper tickWrapper) {
+        this.tickWrapper = tickWrapper;
+        this.lithium$setSleepingTicker(null);
+    }
+
+    @Override
+    public TickingBlockEntity lithium$getSleepingTicker() {
+        return this.sleepingTicker;
+    }
+
+    @Override
+    public void lithium$setSleepingTicker(TickingBlockEntity sleepingTicker) {
+        this.sleepingTicker = sleepingTicker;
+    }
+    // Canvas end - block entity sleeping
 
     public CampfireBlockEntity(BlockPos pos, BlockState blockState) {
         super(BlockEntityType.CAMPFIRE, pos, blockState);
@@ -113,6 +138,10 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
 
         if (flag) {
             setChanged(level, pos, state);
+        // Canvas start - block entity sleeping
+        } else {
+            campfire.lithium$startSleeping();
+        // Canvas end - block entity sleeping
         }
     }
 
@@ -128,6 +157,10 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
 
         if (flag) {
             setChanged(level, pos, state);
+        // Canvas start - block entity sleeping
+        } else {
+            blockEntity.lithium$startSleeping();
+        // Canvas end - block entity sleeping
         }
     }
 
@@ -184,6 +217,7 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
             System.arraycopy(cookingState, 0, this.stopCooking, 0, Math.min(this.stopCooking.length, bytes.capacity()));
         });
         // Paper end - Add more Campfire API
+        this.wakeUpNow(); // Canvas - block entity sleeping
     }
 
     @Override
@@ -239,6 +273,7 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
                 // CraftBukkit end
                 this.cookingProgress[i] = 0;
                 this.items.set(i, stack.consumeAndReturn(1, entity));
+                this.wakeUpNow(); // Canvas - block entity sleeping
                 level.gameEvent(GameEvent.BLOCK_CHANGE, this.getBlockPos(), GameEvent.Context.of(entity, this.getBlockState()));
                 this.markUpdated();
                 return true;
