From 56a92b2c479a7161c340134ba931fb3d6a8e6f79 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/block/entity/CampfireBlockEntity.java b/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
index 64d0d60d9025de2b5eb16b31c6437372d4c20c5e..e2838c451f1feca29aec6a4692f40599c75192fd 100644
--- a/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
@@ -39,7 +39,7 @@ import net.minecraft.world.level.storage.ValueOutput;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
-public class CampfireBlockEntity extends BlockEntity implements Clearable {
+public class CampfireBlockEntity extends BlockEntity implements Clearable, io.canvasmc.canvas.world.block.SleepingBlockEntity { // Canvas - block entity sleeping
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final int BURN_COOL_SPEED = 2;
     private static final int NUM_SLOTS = 4;
@@ -47,6 +47,31 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
     public final int[] cookingProgress = new int[4];
     public final int[] cookingTime = new int[4];
     public final boolean[] stopCooking = new boolean[4]; // Paper - Add more Campfire API
+    // Canvas start - block entity sleeping
+    private net.minecraft.world.level.chunk.LevelChunk.@Nullable RebindableTickingBlockEntityWrapper tickWrapper = null;
+    private @Nullable TickingBlockEntity sleepingTicker = null;
+
+    @Override
+    public net.minecraft.world.level.chunk.LevelChunk.@Nullable RebindableTickingBlockEntityWrapper lithium$getTickWrapper() {
+        return this.tickWrapper;
+    }
+
+    @Override
+    public void lithium$setTickWrapper(net.minecraft.world.level.chunk.LevelChunk.@Nullable RebindableTickingBlockEntityWrapper tickWrapper) {
+        this.tickWrapper = tickWrapper;
+        this.lithium$setSleepingTicker(null);
+    }
+
+    @Override
+    public @Nullable TickingBlockEntity lithium$getSleepingTicker() {
+        return this.sleepingTicker;
+    }
+
+    @Override
+    public void lithium$setSleepingTicker(@Nullable TickingBlockEntity sleepingTicker) {
+        this.sleepingTicker = sleepingTicker;
+    }
+    // Canvas end - block entity sleeping
 
     public CampfireBlockEntity(final BlockPos worldPosition, final BlockState blockState) {
         super(BlockEntityTypes.CAMPFIRE, worldPosition, blockState);
@@ -111,6 +136,10 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
 
         if (changed) {
             setChanged(level, pos, state);
+        // Canvas start - block entity sleeping
+        } else {
+            entity.lithium$startSleeping();
+        // Canvas end - block entity sleeping
         }
     }
 
@@ -126,6 +155,10 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
 
         if (changed) {
             setChanged(level, pos, state);
+        // Canvas start - block entity sleeping
+        } else {
+            entity.lithium$startSleeping();
+        // Canvas end - block entity sleeping
         }
     }
 
@@ -183,6 +216,7 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
             System.arraycopy(cookingState, 0, this.stopCooking, 0, Math.min(this.stopCooking.length, bytes.capacity()));
         });
         // Paper end - Add more Campfire API
+        this.wakeUpNow(); // Canvas - block entity sleeping
     }
 
     @Override
@@ -235,6 +269,7 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
                 // CraftBukkit end
                 this.cookingProgress[slot] = 0;
                 this.items.set(slot, placeItem.consumeAndReturn(1, sourceEntity));
+                this.wakeUpNow(); // Canvas - block entity sleeping
                 serverLevel.gameEvent(GameEvent.BLOCK_CHANGE, this.getBlockPos(), GameEvent.Context.of(sourceEntity, this.getBlockState()));
                 this.markUpdated();
                 return true;
