--- a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/server/ServerEntityLookup.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/server/ServerEntityLookup.java
@@ -70,7 +_,7 @@
     protected void entitySectionChangeCallback(final Entity entity,
                                                final int oldSectionX, final int oldSectionY, final int oldSectionZ,
                                                final int newSectionX, final int newSectionY, final int newSectionZ) {
-        if (entity instanceof ServerPlayer player) {
+        if (entity instanceof ServerPlayer player && (oldSectionX != newSectionX || oldSectionZ != newSectionZ)) { // Canvas - reduce unnecessary nearby player ticks
             ((ChunkSystemServerLevel)this.serverWorld).moonrise$getNearbyPlayers().tickPlayer(player);
         }
         if (entity instanceof ThrownEnderpearl enderpearl && (oldSectionX != newSectionX || oldSectionZ != newSectionZ)) {
