From 9cc14a34f9eda110d805d11d2cf52e6b4b265a19 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
index 43bfa179a8363a1dedd8de083479b7a56965c918..599edaedc38b106b4005ed13654e7ac8c967f6d6 100644
--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/Snowball.java
@@ -58,6 +58,12 @@ public class Snowball extends ThrowableItemProjectile {
         Entity entity = hitResult.getEntity();
         int damage = entity instanceof Blaze ? 3 : 0;
         entity.hurt(this.damageSources().thrown(this, this.getOwner()), damage);
+        // Canvas start - snowball can knockback player config
+        if (io.canvasmc.canvas.Config.INSTANCE.snowballCanKnockback && entity instanceof net.minecraft.server.level.ServerPlayer serverPlayer) {
+            serverPlayer.knockback(0.4F, this.getX() - entity.getX(), this.getZ() - entity.getZ());
+            serverPlayer.hurtMarked = true;
+        }
+        // Canvas end - snowball can knockback player config
     }
 
     @Override
