--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
@@ -52,11 +_,15 @@
     }
 
     private void deregisterFromCurrentOwner() {
-        // Folia - region threading - we remove the registration logic, we do not need to fetch the owner
+        if (this.getOwner() instanceof ServerPlayer serverPlayer) {
+            serverPlayer.deregisterEnderPearl(this);
+        }
     }
 
     private void registerToCurrentOwner() {
-        // Folia - region threading - we remove the registration logic, we do not need to fetch the owner
+        if (this.getOwner() instanceof ServerPlayer serverPlayer) {
+            serverPlayer.registerEnderPearl(this);
+        }
     }
 
     @Override
@@ -263,7 +_,7 @@
             int sectionPosZ = SectionPos.blockToSectionCoord(this.position().z());
             Entity entity = this.owner != null ? findOwnerIncludingDeadPlayer(serverLevel, this.owner.getUUID()) : null;
             if (entity instanceof ServerPlayer serverPlayer
-                && !entity.isAlive()
+                && entity.getBukkitEntity().taskScheduler.isRetired() // Canvas - Fix ender pearls despawning when switching dimensions
                 && !serverPlayer.wonGame
                 && serverPlayer.level().getGameRules().get(GameRules.ENDER_PEARLS_VANISH_ON_DEATH)) {
                 this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
