--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEgg.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEgg.java
@@ -61,6 +_,12 @@
     protected void onHitEntity(final EntityHitResult hitResult) {
         super.onHitEntity(hitResult);
         hitResult.getEntity().hurt(this.damageSources().thrown(this, this.getOwner()), 0.0F);
+        // Canvas start - egg can knockback player config
+        if (level().canvasConfig().combat.eggCanKnockbackPlayers && hitResult.getEntity() instanceof net.minecraft.server.level.ServerPlayer serverPlayer) {
+            serverPlayer.knockback(0.4F, this.getX() - hitResult.getEntity().getX(), this.getZ() - hitResult.getEntity().getZ());
+            serverPlayer.hurtMarked = true;
+        }
+        // Canvas end - egg can knockback player config
     }
 
     @Override
