--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
@@ -52,11 +_,17 @@
     }
 
     private void deregisterFromCurrentOwner() {
-        // Folia - region threading - we remove the registration logic, we do not need to fetch the owner
+        if (!io.canvasmc.canvas.Config.INSTANCE.restoreVanillaEnderPearlBehavior) return; // Canvas - restore vanilla ender pearl behavior
+        if (this.getOwner() instanceof ServerPlayer serverPlayer) {
+            serverPlayer.deregisterEnderPearl(this);
+        }
     }
 
     private void registerToCurrentOwner() {
-        // Folia - region threading - we remove the registration logic, we do not need to fetch the owner
+        if (!io.canvasmc.canvas.Config.INSTANCE.restoreVanillaEnderPearlBehavior) return; // Canvas - restore vanilla ender pearl behavior
+        if (this.getOwner() instanceof ServerPlayer serverPlayer) {
+            serverPlayer.registerEnderPearl(this);
+        }
     }
 
     @Override
