From 3cf7451517453a52df7cb21a31824aadeaf7c6d9 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEgg.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEgg.java
index 79c9818999759302138f195c93b3245916243014..f15d9b9fc882108431eedd33e70b884b7b9baf9c 100644
--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEgg.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEgg.java
@@ -60,6 +60,12 @@ public class ThrownEgg extends ThrowableItemProjectile {
     protected void onHitEntity(final EntityHitResult hitResult) {
         super.onHitEntity(hitResult);
         hitResult.getEntity().hurt(this.damageSources().thrown(this, this.getOwner()), 0.0F);
+        // Canvas start - egg can knockback player config
+        if (io.canvasmc.canvas.Config.INSTANCE.eggCanKnockback && hitResult.getEntity() instanceof net.minecraft.server.level.ServerPlayer serverPlayer) {
+            serverPlayer.knockback(0.4F, this.getX() - hitResult.getEntity().getX(), this.getZ() - hitResult.getEntity().getZ());
+            serverPlayer.hurtMarked = true;
+        }
+        // Canvas end - egg can knockback player config
     }
 
     @Override
