--- a/src/main/java/org/bukkit/event/server/ServerEvent.java
+++ b/src/main/java/org/bukkit/event/server/ServerEvent.java
@@ -9,7 +_,7 @@
 public abstract class ServerEvent extends Event {
 
     public ServerEvent() {
-        super(!Bukkit.isPrimaryThread()); // Paper
+        super(Bukkit.server != null && !Bukkit.isPrimaryThread()); // Paper // Canvas
     }
 
     public ServerEvent(boolean isAsync) {
