From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Pancakse <pancakseyt@gmail.com>
Date: Fri, 9 Feb 2024 11:01:37 -0700
Subject: [PATCH] Build Changes

I wrote this at school lol

diff --git a/build.gradle.kts b/build.gradle.kts
index 50cc3c1e5968d95a21133e4d554a93e181aba0f6..3f0c449c373c015c1cf10165ff402e29e82ac5de 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -20,7 +20,7 @@ minecraft {
 
 dependencies {
     implementation("com.squareup:javapoet:1.13.0")
-    implementation(project(":purpur-api"))
+    implementation(project(":canvas-api")) // Canvas
 }
 
 group = "io.papermc.paper"
