From 56a92b2c479a7161c340134ba931fb3d6a8e6f79 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/decoration/ArmorStand.java b/net/minecraft/world/entity/decoration/ArmorStand.java
index 9e4293c6df1851a852423b469577388d2254e169..e786fa71c4b4804aa4b032796953eedd6f937107 100644
--- a/net/minecraft/world/entity/decoration/ArmorStand.java
+++ b/net/minecraft/world/entity/decoration/ArmorStand.java
@@ -206,6 +206,7 @@ public class ArmorStand extends LivingEntity {
 
     @Override
     protected void pushEntities() {
+        if (!level().canvasConfig().entities.entityCollisionMode.allEntitiesCanBePushed()) return; // Canvas - only push players config
         if (!this.level().paperConfig().entities.armorStands.doCollisionEntityLookups) return; // Paper - Option to prevent armor stands from doing entity lookups
         for (Entity entity : this.level().getEntitiesOfClass(AbstractMinecart.class, this.getBoundingBox(), RIDABLE_MINECARTS)) { // Paper - optimise collisions
             if (this.distanceToSqr(entity) <= 0.2) {
@@ -355,7 +356,14 @@ public class ArmorStand extends LivingEntity {
             // CraftBukkit end
             // Paper start - avoid duplicate event call
             org.bukkit.event.entity.EntityDeathEvent event = this.brokenByAnything(level, source);
-            if (!event.isCancelled()) this.kill(level, source, false); // CraftBukkit
+            // Canvas start - MC-132878
+            if (!event.isCancelled()) {
+                if (io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaFixes.mc132878) {
+                    this.showBreakingParticles();
+                }
+                this.kill(level, source, false);
+            } // CraftBukkit
+            // Canvas end - MC-132878
             // Paper end
             return false;
         }
@@ -463,7 +471,14 @@ public class ArmorStand extends LivingEntity {
         if (health <= 0.5F) {
             // Paper start - avoid duplicate event call
             org.bukkit.event.entity.EntityDeathEvent event = this.brokenByAnything(level, source);
-            if (!event.isCancelled()) this.kill(level, source, false); // CraftBukkit
+            // Canvas start - MC-132878
+            if (!event.isCancelled()) {
+                if (io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaFixes.mc132878) {
+                    this.showBreakingParticles();
+                }
+                this.kill(level, source, false);
+            } // CraftBukkit
+            // Canvas end - MC-132878
             // Paper end
         } else {
             this.setHealth(health);
