--- a/net/minecraft/server/bossevents/CustomBossEvent.java
+++ b/net/minecraft/server/bossevents/CustomBossEvent.java
@@ -20,7 +_,7 @@
 public class CustomBossEvent extends ServerBossEvent {
     private static final int DEFAULT_MAX = 100;
     private final Identifier id;
-    private final Set<UUID> players = Sets.newHashSet();
+    private final Set<UUID> players = Sets.newConcurrentHashSet(); // Canvas - fix commands with region threading - TODO - do better
     private int value;
     private int max = 100;
     // CraftBukkit start
