--- a/net/minecraft/server/bossevents/CustomBossEvents.java
+++ b/net/minecraft/server/bossevents/CustomBossEvents.java
@@ -18,7 +_,7 @@
 public class CustomBossEvents {
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final Codec<Map<Identifier, CustomBossEvent.Packed>> EVENTS_CODEC = Codec.unboundedMap(Identifier.CODEC, CustomBossEvent.Packed.CODEC);
-    private final Map<Identifier, CustomBossEvent> events = Maps.newHashMap();
+    private final Map<Identifier, CustomBossEvent> events = Maps.newConcurrentMap(); // Canvas - fix commands with region threading - TODO - do better
 
     public @Nullable CustomBossEvent get(Identifier id) {
         return this.events.get(id);
