From 4b22e4fb667ff3bf3c489ff7a94018cc27157519 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/server/bossevents/CustomBossEvent.java b/net/minecraft/server/bossevents/CustomBossEvent.java
index 8e319737d2e3f97fe105ec9e258775729b6b64fb..9cd18d939f79aff5520523bb1cb6818576532929 100644
--- a/net/minecraft/server/bossevents/CustomBossEvent.java
+++ b/net/minecraft/server/bossevents/CustomBossEvent.java
@@ -20,7 +20,7 @@ import net.minecraft.world.BossEvent;
 public class CustomBossEvent extends ServerBossEvent {
     private static final int DEFAULT_MAX = 100;
     private final Identifier customId;
-    private final Set<UUID> players = Sets.newHashSet();
+    private final Set<UUID> players = Sets.newConcurrentHashSet(); // Canvas - fix commands with region threading - TODO - do better
     private int value;
     private int max = 100;
     private final Runnable dirtyCallback;
