From 0ce24b7bccfa2d01cdd888671365a9ba0973fab1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/server/bossevents/CustomBossEvents.java b/net/minecraft/server/bossevents/CustomBossEvents.java
index fb55ed68a1f8470b551030a27ca2656555c93d9e..5756f7aca1a80782a88b0e82ef82a626172be539 100644
--- a/net/minecraft/server/bossevents/CustomBossEvents.java
+++ b/net/minecraft/server/bossevents/CustomBossEvents.java
@@ -26,7 +26,7 @@ public class CustomBossEvents extends SavedData {
     public static final SavedDataType<CustomBossEvents> TYPE = new SavedDataType<>(
         Identifier.withDefaultNamespace("custom_boss_events"), CustomBossEvents::new, CODEC, DataFixTypes.SAVED_DATA_CUSTOM_BOSS_EVENTS
     );
-    private final Map<Identifier, CustomBossEvent> events = Maps.newHashMap();
+    private final Map<Identifier, CustomBossEvent> events = Maps.newConcurrentMap(); // Canvas - fix commands with region threading - TODO - do better
 
     public @Nullable CustomBossEvent get(final Identifier id) {
         return this.events.get(id);
