--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/executor/RadiusAwarePrioritisedExecutor.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/executor/RadiusAwarePrioritisedExecutor.java
@@ -473,6 +_,7 @@
             synchronized (this.scheduler) {
                 final DependencyNode node = this.dependencyNode;
                 this.dependencyNode = null;
+                if (node == null) return; // Canvas
                 toSchedule = node.tree.returnNode(node);
             }
 
