--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/executor/RadiusAwarePrioritisedExecutor.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/executor/RadiusAwarePrioritisedExecutor.java
@@ -5,6 +_,7 @@
 import ca.spottedleaf.moonrise.common.util.CoordinateUtils;
 import it.unimi.dsi.fastutil.longs.Long2ReferenceOpenHashMap;
 import it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet;
+import net.minecraft.server.level.ServerLevel;
 import java.util.ArrayList;
 import java.util.Comparator;
 import java.util.List;
@@ -111,14 +_,14 @@
     }
 
     public PrioritisedExecutor.PrioritisedTask createTask(final int chunkX, final int chunkZ, final int radius,
-                                                          final Runnable run, final Priority priority) {
+                                                          final Runnable run, final Priority priority, final ServerLevel world) { // Canvas
         if (radius < 0) {
             throw new IllegalArgumentException("Radius must be > 0: " + radius);
         }
-        return new Task(this, chunkX, chunkZ, radius, run, priority);
+        return new Task(this, chunkX, chunkZ, radius, run, priority, world); // Canvas
     }
 
-    public PrioritisedExecutor.PrioritisedTask createTask(final int chunkX, final int chunkZ, final int radius,
+    /* public PrioritisedExecutor.PrioritisedTask createTask(final int chunkX, final int chunkZ, final int radius, // Canvas
                                                           final Runnable run) {
         return this.createTask(chunkX, chunkZ, radius, run, Priority.NORMAL);
     }
@@ -139,13 +_,13 @@
         ret.queue();
 
         return ret;
-    }
+    } */ // Canvas
 
     public PrioritisedExecutor.PrioritisedTask createInfiniteRadiusTask(final Runnable run, final Priority priority) {
-        return new Task(this, 0, 0, -1, run, priority);
+        return new Task(this, 0, 0, -1, run, priority, null); // Canvas
     }
 
-    public PrioritisedExecutor.PrioritisedTask createInfiniteRadiusTask(final Runnable run) {
+    /* public PrioritisedExecutor.PrioritisedTask createInfiniteRadiusTask(final Runnable run) { // Canvas
         return this.createInfiniteRadiusTask(run, Priority.NORMAL);
     }
 
@@ -155,7 +_,7 @@
         ret.queue();
 
         return ret;
-    }
+    } */ // Canvas
 
     public PrioritisedExecutor.PrioritisedTask queueInfiniteRadiusTask(final Runnable run) {
         final PrioritisedExecutor.PrioritisedTask ret = this.createInfiniteRadiusTask(run, Priority.NORMAL);
@@ -431,27 +_,31 @@
         }
     }
 
-    private static final class Task implements PrioritisedExecutor.PrioritisedTask, Runnable {
+    public static final class Task implements PrioritisedExecutor.PrioritisedTask, Runnable { // Canvas - private -> public
 
         // task specific fields
         private final RadiusAwarePrioritisedExecutor scheduler;
-        private final int chunkX;
-        private final int chunkZ;
+        // Canvas start - private -> public
+        public final int chunkX;
+        public final int chunkZ;
+        // Canvas end
         private final int radius;
         private Runnable run;
         private Priority priority;
+        public final ServerLevel world; // Canvas
 
         private DependencyNode dependencyNode;
         private PrioritisedExecutor.PrioritisedTask queuedTask;
 
         private Task(final RadiusAwarePrioritisedExecutor scheduler, final int chunkX, final int chunkZ, final int radius,
-                     final Runnable run, final Priority priority) {
+                     final Runnable run, final Priority priority, final ServerLevel world) { // Canvas
             this.scheduler = scheduler;
             this.chunkX = chunkX;
             this.chunkZ = chunkZ;
             this.radius = radius;
             this.run = run;
             this.priority = priority;
+            this.world = world; // Canvas
         }
 
         private boolean isFiniteRadius() {
@@ -473,6 +_,7 @@
             synchronized (this.scheduler) {
                 final DependencyNode node = this.dependencyNode;
                 this.dependencyNode = null;
+                if (node == null) return; // Canvas - NPE
                 toSchedule = node.tree.returnNode(node);
             }
 
@@ -489,6 +_,7 @@
             final Runnable run = this.run;
             this.run = null;
             try {
+                if (run == null) return; // Canvas - NPE
                 run.run();
             } finally {
                 this.returnNode();
