--- a/net/minecraft/world/entity/animal/axolotl/AxolotlAi.java
+++ b/net/minecraft/world/entity/animal/axolotl/AxolotlAi.java
@@ -142,7 +_,13 @@
         Optional<PositionTracker> memory = entity.getBrain().getMemory(MemoryModuleType.LOOK_TARGET);
         if (memory.isPresent()) {
             BlockPos blockPos = memory.get().currentBlockPosition();
-            return level.isWaterAt(blockPos) == entity.isInWater();
+            // Canvas start - region threading
+            net.minecraft.world.level.material.FluidState fluidIfLoaded = level.getFluidIfLoaded(blockPos);
+            if (fluidIfLoaded == null) {
+                return false;
+            }
+            return fluidIfLoaded.is(net.minecraft.tags.FluidTags.WATER) == entity.isInWater();
+            // Canvas end - region threading
         } else {
             return false;
         }
