From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/dimension/end/EndDragonFight.java b/net/minecraft/world/level/dimension/end/EndDragonFight.java
index fb0cc5e5b461b3b887b948b92fdd9564d1ee4c20..d087da8b0a0026fa33cbcb3f603b1cc4f9b52003 100644
--- a/net/minecraft/world/level/dimension/end/EndDragonFight.java
+++ b/net/minecraft/world/level/dimension/end/EndDragonFight.java
@@ -265,7 +265,7 @@ public class EndDragonFight {
             for (int i1 = -8; i1 <= 8; i1++) {
                 LevelChunk chunk = this.level.getChunk(i, i1);
 
-                for (BlockEntity blockEntity : chunk.getBlockEntities().values()) {
+                for (BlockEntity blockEntity : chunk.canvas$getAllBlockEntities()) { // Canvas - optimize block entity fetching
                     if (blockEntity instanceof TheEndPortalBlockEntity) {
                         return true;
                     }
@@ -283,7 +283,7 @@ public class EndDragonFight {
             for (int i1 = -8 + chunkPos.z; i1 <= 8 + chunkPos.z; i1++) {
                 LevelChunk chunk = this.level.getChunk(i, i1);
 
-                for (BlockEntity blockEntity : chunk.getBlockEntities().values()) {
+                for (BlockEntity blockEntity : chunk.canvas$getAllBlockEntities()) { // Canvas - optimize block entity fetching
                     if (blockEntity instanceof TheEndPortalBlockEntity) {
                         BlockPattern.BlockPatternMatch blockPatternMatch = this.exitPortalPattern.find(this.level, blockEntity.getBlockPos());
                         if (blockPatternMatch != null) {
