From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Fri, 26 Sep 2025 20:01:56 -0700
Subject: [PATCH] Rewrite Moonrise Executor


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java
index 65716aa20fec95533249ba39bbf10dd570c7c64f..6f94f1277fc7b28a41228b4894504ccbc524fa99 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java
@@ -67,11 +67,7 @@ public final class ChunkTaskScheduler {
     private static final Logger LOGGER = LogUtils.getClassLogger();
 
     public static void init(final boolean useParallelGen) {
-        for (final PrioritisedThreadPool.ExecutorGroup.ThreadPoolExecutor executor : MoonriseCommon.RADIUS_AWARE_GROUP.getAllExecutors()) {
-            executor.setMaxParallelism(useParallelGen ? -1 : 1);
-        }
-
-        LOGGER.info("Chunk system is using population gen parallelism: " + useParallelGen);
+        // Canvas - rewrite moonrise executor
     }
 
     public static final TicketType CHUNK_LOAD = ChunkSystemTicketType.create("chunk_system:chunk_load", Long::compareTo);
@@ -116,12 +112,14 @@ public final class ChunkTaskScheduler {
 
     public final ServerLevel world;
     public final RadiusAwarePrioritisedExecutor radiusAwareScheduler;
-    public final PrioritisedThreadPool.ExecutorGroup.ThreadPoolExecutor parallelGenExecutor;
-    private final PrioritisedThreadPool.ExecutorGroup.ThreadPoolExecutor radiusAwareGenExecutor;
-    public final PrioritisedThreadPool.ExecutorGroup.ThreadPoolExecutor loadExecutor;
+    // Canvas start - rewrite moonrise executor
+    public final io.canvasmc.canvas.chunk.TheChunkSystem.ExecutorGroup.ThreadPoolExecutor parallelGenExecutor;
+    private final io.canvasmc.canvas.chunk.TheChunkSystem.ExecutorGroup.ThreadPoolExecutor radiusAwareGenExecutor;
+    public final io.canvasmc.canvas.chunk.TheChunkSystem.ExecutorGroup.ThreadPoolExecutor loadExecutor;
     public final PrioritisedThreadPool.ExecutorGroup.ThreadPoolExecutor ioExecutor;
-    public final PrioritisedThreadPool.ExecutorGroup.ThreadPoolExecutor compressionExecutor;
-    public final PrioritisedThreadPool.ExecutorGroup.ThreadPoolExecutor saveExecutor;
+    public final io.canvasmc.canvas.chunk.TheChunkSystem.ExecutorGroup.ThreadPoolExecutor compressionExecutor;
+    public final io.canvasmc.canvas.chunk.TheChunkSystem.ExecutorGroup.ThreadPoolExecutor saveExecutor;
+    // Canvas end - rewrite moonrise executor
 
     // Folia - regionised ticking
 
@@ -199,9 +197,7 @@ public final class ChunkTaskScheduler {
                 // FULL is executed on main.
         );
 
-        for (final ChunkStatus status : parallelCapableStatus) {
-            ((ChunkSystemChunkStatus)status).moonrise$setParallelCapable(true);
-        }
+        // Canvas - rewrite moonrise executor
     }
 
     private static final int[] ACCESS_RADIUS_TABLE_LOAD = new int[ChunkStatus.getStatusList().size()];
@@ -292,14 +288,16 @@ public final class ChunkTaskScheduler {
         this.lockShift = Math.max(((ChunkSystemServerLevel)world).moonrise$getRegionChunkShift(), ThreadedTicketLevelPropagator.SECTION_SHIFT);
         this.schedulingLockArea = new ReentrantAreaLock(this.getChunkSystemLockShift());
 
-        this.parallelGenExecutor = MoonriseCommon.PARALLEL_GEN_GROUP.createExecutor(-1, MoonriseCommon.WORKER_QUEUE_HOLD_TIME, 0);
-        this.radiusAwareGenExecutor = MoonriseCommon.RADIUS_AWARE_GROUP.createExecutor(1, MoonriseCommon.WORKER_QUEUE_HOLD_TIME, 0);
-        this.loadExecutor = MoonriseCommon.LOAD_GROUP.createExecutor(-1, MoonriseCommon.WORKER_QUEUE_HOLD_TIME, 0);
+        // Canvas start - rewrite moonrise executor
+        this.parallelGenExecutor = MoonriseCommon.PARALLEL_GEN_GROUP.createExecutor(MoonriseCommon.WORKER_QUEUE_HOLD_TIME, 0);
+        this.radiusAwareGenExecutor = MoonriseCommon.RADIUS_AWARE_GROUP.createExecutor(MoonriseCommon.WORKER_QUEUE_HOLD_TIME, 0);
+        this.loadExecutor = MoonriseCommon.LOAD_GROUP.createExecutor(MoonriseCommon.WORKER_QUEUE_HOLD_TIME, 0);
         this.radiusAwareScheduler = new RadiusAwarePrioritisedExecutor(this.radiusAwareGenExecutor, 16);
         this.ioExecutor = MoonriseCommon.SERVER_REGION_IO_GROUP.createExecutor(-1, MoonriseCommon.IO_QUEUE_HOLD_TIME, 0);
         // we need a separate executor here so that on shutdown we can continue to process I/O tasks
-        this.compressionExecutor = MoonriseCommon.LOAD_GROUP.createExecutor(-1, MoonriseCommon.WORKER_QUEUE_HOLD_TIME, 0);
-        this.saveExecutor = MoonriseCommon.LOAD_GROUP.createExecutor(-1, MoonriseCommon.WORKER_QUEUE_HOLD_TIME, 0);
+        this.compressionExecutor = MoonriseCommon.LOAD_GROUP.createExecutor(MoonriseCommon.WORKER_QUEUE_HOLD_TIME, 0);
+        this.saveExecutor = MoonriseCommon.LOAD_GROUP.createExecutor(MoonriseCommon.WORKER_QUEUE_HOLD_TIME, 0);
+        // Canvas end - rewrite moonrise executor
         this.chunkHolderManager = new ChunkHolderManager(world, this);
     }
 
diff --git a/com/mojang/datafixers/types/Type.java b/com/mojang/datafixers/types/Type.java
index c917e6cf076fa2e427d49a1732bde99a9c62d81a..544b51bda02ab9c030ed7f85ee396d394ce5fb42 100644
--- a/com/mojang/datafixers/types/Type.java
+++ b/com/mojang/datafixers/types/Type.java
@@ -192,7 +192,7 @@ public abstract class Type<A> implements App<Type.Mu, A> {
             PENDING_REWRITE_CACHE.remove(key);
             return result;
         }
-        return (Optional<RewriteResult<A, ?>>) pending.join();
+        return (Optional<RewriteResult<A, ?>>) io.canvasmc.canvas.util.CFUtil.join(pending); // Canvas - rewrite moonrise executor
     }
 
     public <FT, FR> Type<?> getSetType(final OpticFinder<FT> optic, final Type<FR> newType) {
diff --git a/com/mojang/math/Transformation.java b/com/mojang/math/Transformation.java
index 5fb382be4d86328690c49f2a5a0c3ec698a38e21..09bbad3da186545375011612b8140b4034f27952 100644
--- a/com/mojang/math/Transformation.java
+++ b/com/mojang/math/Transformation.java
@@ -52,6 +52,7 @@ public final class Transformation {
         } else {
             this.matrix = matrix;
         }
+        this.ensureDecomposed(); // Canvas - rewrite moonrise executor
     }
 
     public Transformation(@Nullable Vector3f translation, @Nullable Quaternionf leftRotation, @Nullable Vector3f scale, @Nullable Quaternionf rightRotation) {
@@ -61,6 +62,7 @@ public final class Transformation {
         this.scale = scale != null ? scale : new Vector3f(1.0F, 1.0F, 1.0F);
         this.rightRotation = rightRotation != null ? rightRotation : new Quaternionf();
         this.decomposed = true;
+        this.ensureDecomposed(); // Canvas - rewrite moonrise executor
     }
 
     public static Transformation identity() {
diff --git a/net/minecraft/world/level/levelgen/feature/stateproviders/RandomizedIntStateProvider.java b/net/minecraft/world/level/levelgen/feature/stateproviders/RandomizedIntStateProvider.java
index 21cbf6c1723feb1813d8cd5106e36594d140d987..94a12d0d05504da185c549fab9153a6390594aaf 100644
--- a/net/minecraft/world/level/levelgen/feature/stateproviders/RandomizedIntStateProvider.java
+++ b/net/minecraft/world/level/levelgen/feature/stateproviders/RandomizedIntStateProvider.java
@@ -55,17 +55,20 @@ public class RandomizedIntStateProvider extends BlockStateProvider {
 
     @Override
     public BlockState getState(RandomSource random, BlockPos pos) {
-        BlockState state = this.source.getState(random, pos);
-        if (this.property == null || !state.hasProperty(this.property)) {
-            IntegerProperty integerProperty = findProperty(state, this.propertyName);
-            if (integerProperty == null) {
-                return state;
+        // Canvas start - rewrite moonrise executor
+        BlockState blockState = this.source.getState(random, pos);
+        IntegerProperty propertyLocal = this.property; // used as cache only
+        if (propertyLocal == null || !blockState.hasProperty(propertyLocal)) {
+            IntegerProperty intProperty = findProperty(blockState, this.propertyName);
+            if (intProperty == null) {
+                return blockState;
             }
 
-            this.property = integerProperty;
+            this.property = propertyLocal = intProperty;
         }
 
-        return state.setValue(this.property, this.values.sample(random));
+        return blockState.trySetValue(propertyLocal, Integer.valueOf(this.values.sample(random)));
+        // Canvas end - rewrite moonrise executor
     }
 
     @Nullable
diff --git a/net/minecraft/world/level/levelgen/structure/ScatteredFeaturePiece.java b/net/minecraft/world/level/levelgen/structure/ScatteredFeaturePiece.java
index e2036a80eff3dc1a9ec625880d4aab6ef71d84fa..2046eab8e85d6f48ab9d9c7c9643fffc0c06b1ec 100644
--- a/net/minecraft/world/level/levelgen/structure/ScatteredFeaturePiece.java
+++ b/net/minecraft/world/level/levelgen/structure/ScatteredFeaturePiece.java
@@ -12,7 +12,7 @@ public abstract class ScatteredFeaturePiece extends StructurePiece {
     protected final int width;
     protected final int height;
     protected final int depth;
-    protected int heightPosition = -1;
+    protected volatile int heightPosition = -1; // Canvas - rewrite moonrise executor
 
     protected ScatteredFeaturePiece(StructurePieceType type, int x, int y, int z, int width, int height, int depth, Direction orientation) {
         super(type, 0, StructurePiece.makeBoundingBox(x, y, z, orientation, width, height, depth));
diff --git a/net/minecraft/world/level/levelgen/structure/structures/DesertPyramidPiece.java b/net/minecraft/world/level/levelgen/structure/structures/DesertPyramidPiece.java
index 0d2451a9ade43650dbbcbab69ce0f6e8f69b5aee..0a3baf3036ffc10eb964a35494a0b15d2eeb571f 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/DesertPyramidPiece.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/DesertPyramidPiece.java
@@ -23,29 +23,44 @@ import net.minecraft.world.level.storage.loot.BuiltInLootTables;
 public class DesertPyramidPiece extends ScatteredFeaturePiece {
     public static final int WIDTH = 21;
     public static final int DEPTH = 21;
-    private final boolean[] hasPlacedChest = new boolean[4];
-    private final List<BlockPos> potentialSuspiciousSandWorldPositions = new ArrayList<>();
+    private final java.util.concurrent.atomic.AtomicReferenceArray<Boolean> hasPlacedChestAtomic = new java.util.concurrent.atomic.AtomicReferenceArray<>(new Boolean[4]); // Canvas - rewrite moonrise executor
+    private final java.util.Set<BlockPos> potentialSuspiciousSandWorldPositions = com.google.common.collect.Sets.newConcurrentHashSet(); // Canvas - rewrite moonrise executor
     private BlockPos randomCollapsedRoofPos = BlockPos.ZERO;
+    // Canvas start - rewrite moonrise executor
+    private void onInit() {
+        for(int i = 0; i < this.hasPlacedChestAtomic.length(); ++i) {
+            if (this.hasPlacedChestAtomic.get(i) == null) {
+                this.hasPlacedChestAtomic.set(i, false);
+            }
+        }
+    }
+    // Canvas end - rewrite moonrise executor
 
     public DesertPyramidPiece(RandomSource random, int x, int z) {
         super(StructurePieceType.DESERT_PYRAMID_PIECE, x, 64, z, 21, 15, 21, getRandomHorizontalDirection(random));
+        onInit(); // Canvas - rewrite moonrise executor
     }
 
     public DesertPyramidPiece(CompoundTag tag) {
         super(StructurePieceType.DESERT_PYRAMID_PIECE, tag);
-        this.hasPlacedChest[0] = tag.getBooleanOr("hasPlacedChest0", false);
-        this.hasPlacedChest[1] = tag.getBooleanOr("hasPlacedChest1", false);
-        this.hasPlacedChest[2] = tag.getBooleanOr("hasPlacedChest2", false);
-        this.hasPlacedChest[3] = tag.getBooleanOr("hasPlacedChest3", false);
+        // Canvas start - rewrite moonrise executor
+        this.hasPlacedChestAtomic.set(0, tag.getBooleanOr("hasPlacedChest0", false));
+        this.hasPlacedChestAtomic.set(1, tag.getBooleanOr("hasPlacedChest1", false));
+        this.hasPlacedChestAtomic.set(2, tag.getBooleanOr("hasPlacedChest2", false));
+        this.hasPlacedChestAtomic.set(3, tag.getBooleanOr("hasPlacedChest3", false));
+        onInit();
+        // Canvas end - rewrite moonrise executor
     }
 
     @Override
     protected void addAdditionalSaveData(StructurePieceSerializationContext context, CompoundTag tag) {
         super.addAdditionalSaveData(context, tag);
-        tag.putBoolean("hasPlacedChest0", this.hasPlacedChest[0]);
-        tag.putBoolean("hasPlacedChest1", this.hasPlacedChest[1]);
-        tag.putBoolean("hasPlacedChest2", this.hasPlacedChest[2]);
-        tag.putBoolean("hasPlacedChest3", this.hasPlacedChest[3]);
+        // Canvas start - rewrite moonrise executor
+        tag.putBoolean("hasPlacedChest0", this.hasPlacedChestAtomic.get(0));
+        tag.putBoolean("hasPlacedChest1", this.hasPlacedChestAtomic.get(1));
+        tag.putBoolean("hasPlacedChest2", this.hasPlacedChestAtomic.get(2));
+        tag.putBoolean("hasPlacedChest3", this.hasPlacedChestAtomic.get(3));
+        // Canvas end - rewrite moonrise executor
     }
 
     @Override
@@ -287,12 +302,12 @@ public class DesertPyramidPiece extends ScatteredFeaturePiece {
             this.placeBlock(level, Blocks.CUT_SANDSTONE.defaultBlockState(), 10, -11, 13, box);
 
             for (Direction direction : Direction.Plane.HORIZONTAL) {
-                if (!this.hasPlacedChest[direction.get2DDataValue()]) {
+                if (!this.hasPlacedChestAtomic.get(direction.get2DDataValue())) { // Canvas - rewrite moonrise executor
                     int i4 = direction.getStepX() * 2;
                     int i5 = direction.getStepZ() * 2;
-                    this.hasPlacedChest[direction.get2DDataValue()] = this.createChest(
+                    this.hasPlacedChestAtomic.set(direction.get2DDataValue(), this.createChest( // Canvas - rewrite moonrise executor
                         level, box, random, 10 + i4, -11, 10 + i5, BuiltInLootTables.DESERT_PYRAMID
-                    );
+                    )); // Canvas - rewrite moonrise executor
                 }
             }
 
@@ -419,7 +434,7 @@ public class DesertPyramidPiece extends ScatteredFeaturePiece {
         this.randomCollapsedRoofPos = new BlockPos(this.getWorldX(i1, randomInt), this.getWorldY(y), this.getWorldZ(i1, randomInt));
     }
 
-    public List<BlockPos> getPotentialSuspiciousSandWorldPositions() {
+    public java.util.Set<BlockPos> getPotentialSuspiciousSandWorldPositions() { // Canvas - rewrite moonrise executor
         return this.potentialSuspiciousSandWorldPositions;
     }
 
diff --git a/net/minecraft/world/level/levelgen/structure/structures/JungleTemplePiece.java b/net/minecraft/world/level/levelgen/structure/structures/JungleTemplePiece.java
index d28f1a25e2137955402e41679d1a4220a0136579..c346e9026a093938d6959d3f024b4f3667b767ba 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/JungleTemplePiece.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/JungleTemplePiece.java
@@ -30,10 +30,12 @@ import net.minecraft.world.level.storage.loot.BuiltInLootTables;
 public class JungleTemplePiece extends ScatteredFeaturePiece {
     public static final int WIDTH = 12;
     public static final int DEPTH = 15;
-    private boolean placedMainChest;
-    private boolean placedHiddenChest;
-    private boolean placedTrap1;
-    private boolean placedTrap2;
+    // Canvas start - rewrite moonrise executor
+    private final java.util.concurrent.atomic.AtomicBoolean placedMainChest = new java.util.concurrent.atomic.AtomicBoolean(false);
+    private final java.util.concurrent.atomic.AtomicBoolean placedHiddenChest = new java.util.concurrent.atomic.AtomicBoolean(false);
+    private final java.util.concurrent.atomic.AtomicBoolean placedTrap1 = new java.util.concurrent.atomic.AtomicBoolean(false);
+    private final java.util.concurrent.atomic.AtomicBoolean placedTrap2 = new java.util.concurrent.atomic.AtomicBoolean(false);
+    // Canvas end - rewrite moonrise executor
     private static final JungleTemplePiece.MossStoneSelector STONE_SELECTOR = new JungleTemplePiece.MossStoneSelector();
 
     public JungleTemplePiece(RandomSource random, int x, int z) {
@@ -42,19 +44,23 @@ public class JungleTemplePiece extends ScatteredFeaturePiece {
 
     public JungleTemplePiece(CompoundTag tag) {
         super(StructurePieceType.JUNGLE_PYRAMID_PIECE, tag);
-        this.placedMainChest = tag.getBooleanOr("placedMainChest", false);
-        this.placedHiddenChest = tag.getBooleanOr("placedHiddenChest", false);
-        this.placedTrap1 = tag.getBooleanOr("placedTrap1", false);
-        this.placedTrap2 = tag.getBooleanOr("placedTrap2", false);
+        // Canvas start - rewrite moonrise executor
+        this.placedMainChest.set(tag.getBooleanOr("placedMainChest", false));
+        this.placedHiddenChest.set(tag.getBooleanOr("placedHiddenChest", false));
+        this.placedTrap1.set(tag.getBooleanOr("placedTrap1", false));
+        this.placedTrap2.set(tag.getBooleanOr("placedTrap2", false));
+        // Canvas end - rewrite moonrise executor
     }
 
     @Override
     protected void addAdditionalSaveData(StructurePieceSerializationContext context, CompoundTag tag) {
         super.addAdditionalSaveData(context, tag);
-        tag.putBoolean("placedMainChest", this.placedMainChest);
-        tag.putBoolean("placedHiddenChest", this.placedHiddenChest);
-        tag.putBoolean("placedTrap1", this.placedTrap1);
-        tag.putBoolean("placedTrap2", this.placedTrap2);
+        // Canvas start - rewrite moonrise executor
+        tag.putBoolean("placedMainChest", this.placedMainChest.get());
+        tag.putBoolean("placedHiddenChest", this.placedHiddenChest.get());
+        tag.putBoolean("placedTrap1", this.placedTrap1.get());
+        tag.putBoolean("placedTrap2", this.placedTrap2.get());
+        // Canvas end - rewrite moonrise executor
     }
 
     @Override
@@ -236,8 +242,10 @@ public class JungleTemplePiece extends ScatteredFeaturePiece {
                 box
             );
             this.placeBlock(level, Blocks.MOSSY_COBBLESTONE.defaultBlockState(), 3, -3, 1, box);
-            if (!this.placedTrap1) {
-                this.placedTrap1 = this.createDispenser(level, box, random, 3, -2, 1, Direction.NORTH, BuiltInLootTables.JUNGLE_TEMPLE_DISPENSER);
+            // Canvas start - rewrite moonrise executor
+            if (!this.placedTrap1.get()) {
+                this.placedTrap1.set(this.createDispenser(level, box, random, 3, -2, 1, Direction.NORTH, BuiltInLootTables.JUNGLE_TEMPLE_DISPENSER));
+            // Canvas end - rewrite moonrise executor
             }
 
             this.placeBlock(level, Blocks.VINE.defaultBlockState().setValue(VineBlock.SOUTH, true), 3, -2, 2, box);
@@ -328,14 +336,18 @@ public class JungleTemplePiece extends ScatteredFeaturePiece {
             );
             this.placeBlock(level, Blocks.MOSSY_COBBLESTONE.defaultBlockState(), 9, -3, 4, box);
             this.placeBlock(level, blockState4, 9, -2, 4, box);
-            if (!this.placedTrap2) {
-                this.placedTrap2 = this.createDispenser(level, box, random, 9, -2, 3, Direction.WEST, BuiltInLootTables.JUNGLE_TEMPLE_DISPENSER);
+            // Canvas start - rewrite moonrise executor
+            if (!this.placedTrap2.get()) {
+                this.placedTrap2.set(this.createDispenser(level, box, random, 9, -2, 3, Direction.WEST, BuiltInLootTables.JUNGLE_TEMPLE_DISPENSER));
+            // Canvas end - rewrite moonrise executor
             }
 
             this.placeBlock(level, Blocks.VINE.defaultBlockState().setValue(VineBlock.EAST, true), 8, -1, 3, box);
             this.placeBlock(level, Blocks.VINE.defaultBlockState().setValue(VineBlock.EAST, true), 8, -2, 3, box);
-            if (!this.placedMainChest) {
-                this.placedMainChest = this.createChest(level, box, random, 8, -3, 3, BuiltInLootTables.JUNGLE_TEMPLE);
+            // Canvas start - rewrite moonrise executor
+            if (!this.placedMainChest.get()) {
+                this.placedMainChest.set(this.createChest(level, box, random, 8, -3, 3, BuiltInLootTables.JUNGLE_TEMPLE));
+            // Canvas end - rewrite moonrise executor
             }
 
             this.placeBlock(level, Blocks.MOSSY_COBBLESTONE.defaultBlockState(), 9, -3, 2, box);
@@ -378,8 +390,10 @@ public class JungleTemplePiece extends ScatteredFeaturePiece {
             this.placeBlock(level, Blocks.STICKY_PISTON.defaultBlockState().setValue(PistonBaseBlock.FACING, Direction.WEST), 10, -2, 8, box);
             this.placeBlock(level, Blocks.STICKY_PISTON.defaultBlockState().setValue(PistonBaseBlock.FACING, Direction.WEST), 10, -1, 8, box);
             this.placeBlock(level, Blocks.REPEATER.defaultBlockState().setValue(RepeaterBlock.FACING, Direction.NORTH), 10, -2, 10, box);
-            if (!this.placedHiddenChest) {
-                this.placedHiddenChest = this.createChest(level, box, random, 9, -3, 10, BuiltInLootTables.JUNGLE_TEMPLE);
+            // Canvas start - rewrite moonrise executor
+            if (!this.placedHiddenChest.get()) {
+                this.placedHiddenChest.set(this.createChest(level, box, random, 9, -3, 10, BuiltInLootTables.JUNGLE_TEMPLE));
+            // Canvas end - rewrite moonrise executor
             }
         }
     }
diff --git a/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java b/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
index 53d3bf1d2a1debe46e276b1db25b420be4ad9958..f56688fab03f25045238410e0f884726db2c48b5 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
@@ -91,7 +91,7 @@ public class MineshaftPieces {
     public static class MineShaftCorridor extends MineshaftPieces.MineShaftPiece {
         private final boolean hasRails;
         private final boolean spiderCorridor;
-        private boolean hasPlacedSpider;
+        private volatile boolean hasPlacedSpider; // Canvas - rewrite moonrise executor
         private final int numSections;
 
         public MineShaftCorridor(CompoundTag tag) {
@@ -950,7 +950,7 @@ public class MineshaftPieces {
     }
 
     public static class MineShaftRoom extends MineshaftPieces.MineShaftPiece {
-        private final List<BoundingBox> childEntranceBoxes = Lists.newLinkedList();
+        private final List<BoundingBox> childEntranceBoxes = java.util.Collections.synchronizedList(Lists.newLinkedList()); // Canvas - rewrite moonrise executor
 
         public MineShaftRoom(int genDepth, RandomSource random, int x, int z, MineshaftStructure.Type type) {
             super(
diff --git a/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java b/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java
index ae4ffcf1859e8ff7f8fbc91246e66e20f5c33dd7..0c505366ce4e46e98c3bd65d761f67f72ae1254a 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java
@@ -643,7 +643,7 @@ public class NetherFortressPieces {
         private static final int WIDTH = 5;
         private static final int HEIGHT = 7;
         private static final int DEPTH = 5;
-        private boolean isNeedingChest;
+        private volatile boolean isNeedingChest; // Canvas - rewrite moonrise executor
 
         public CastleSmallCorridorLeftTurnPiece(int genDepth, RandomSource random, BoundingBox box, Direction orientation) {
             super(StructurePieceType.NETHER_FORTRESS_CASTLE_SMALL_CORRIDOR_LEFT_TURN, genDepth, box);
@@ -773,7 +773,7 @@ public class NetherFortressPieces {
         private static final int WIDTH = 5;
         private static final int HEIGHT = 7;
         private static final int DEPTH = 5;
-        private boolean isNeedingChest;
+        private volatile boolean isNeedingChest; // Canvas - rewrite moonrise executor
 
         public CastleSmallCorridorRightTurnPiece(int genDepth, RandomSource random, BoundingBox box, Direction orientation) {
             super(StructurePieceType.NETHER_FORTRESS_CASTLE_SMALL_CORRIDOR_RIGHT_TURN, genDepth, box);
@@ -1008,7 +1008,7 @@ public class NetherFortressPieces {
         private static final int WIDTH = 7;
         private static final int HEIGHT = 8;
         private static final int DEPTH = 9;
-        private boolean hasPlacedSpawner;
+        private volatile boolean hasPlacedSpawner; // Canvas - rewrite moonrise executor
 
         public MonsterThrone(int genDepth, BoundingBox box, Direction orientation) {
             super(StructurePieceType.NETHER_FORTRESS_MONSTER_THRONE, genDepth, box);
@@ -1094,7 +1094,7 @@ public class NetherFortressPieces {
         }
     }
 
-    abstract static class NetherBridgePiece extends StructurePiece {
+    public abstract static class NetherBridgePiece extends StructurePiece { // Canvas - TODO - at
         protected NetherBridgePiece(StructurePieceType type, int genDepth, BoundingBox boundingBox) {
             super(type, genDepth, boundingBox);
         }
@@ -1112,7 +1112,7 @@ public class NetherFortressPieces {
             int i = 0;
 
             for (NetherFortressPieces.PieceWeight pieceWeight : weights) {
-                if (pieceWeight.maxPlaceCount > 0 && pieceWeight.placeCount < pieceWeight.maxPlaceCount) {
+                if (pieceWeight.maxPlaceCount > 0 && pieceWeight.placeCount.get() < pieceWeight.maxPlaceCount) { // Canvas - rewrite moonrise executor
                     flag = true;
                 }
 
@@ -1152,7 +1152,7 @@ public class NetherFortressPieces {
                             pieceWeight, pieces, random, x, y, z, orientation, genDepth
                         );
                         if (netherBridgePiece != null) {
-                            pieceWeight.placeCount++;
+                            pieceWeight.placeCount.set(pieceWeight.placeCount.get() + 1); // Canvas - rewrite moonrise executor
                             startPiece.previousPiece = pieceWeight;
                             if (!pieceWeight.isValid()) {
                                 weights.remove(pieceWeight);
@@ -1387,7 +1387,7 @@ public class NetherFortressPieces {
     static class PieceWeight {
         public final Class<? extends NetherFortressPieces.NetherBridgePiece> pieceClass;
         public final int weight;
-        public int placeCount;
+        public final ThreadLocal<Integer> placeCount = ThreadLocal.withInitial(() -> 0); // Canvas - rewrite moonrise executor
         public final int maxPlaceCount;
         public final boolean allowInRow;
 
@@ -1403,11 +1403,11 @@ public class NetherFortressPieces {
         }
 
         public boolean doPlace(int genDepth) {
-            return this.maxPlaceCount == 0 || this.placeCount < this.maxPlaceCount;
+            return this.maxPlaceCount == 0 || this.placeCount.get() < this.maxPlaceCount; // Canvas - rewrite moonrise executor
         }
 
         public boolean isValid() {
-            return this.maxPlaceCount == 0 || this.placeCount < this.maxPlaceCount;
+            return this.maxPlaceCount == 0 || this.placeCount.get() < this.maxPlaceCount; // Canvas - rewrite moonrise executor
         }
     }
 
@@ -1545,24 +1545,24 @@ public class NetherFortressPieces {
     }
 
     public static class StartPiece extends NetherFortressPieces.BridgeCrossing {
-        public NetherFortressPieces.PieceWeight previousPiece;
+        public volatile NetherFortressPieces.PieceWeight previousPiece; // Canvas - rewrite moonrise executor
         public List<NetherFortressPieces.PieceWeight> availableBridgePieces;
         public List<NetherFortressPieces.PieceWeight> availableCastlePieces;
         public final List<StructurePiece> pendingChildren = Lists.newArrayList();
 
         public StartPiece(RandomSource random, int x, int z) {
             super(x, z, getRandomHorizontalDirection(random));
-            this.availableBridgePieces = Lists.newArrayList();
+            this.availableBridgePieces = java.util.Collections.synchronizedList(Lists.newArrayList()); // Canvas - rewrite moonrise executor
 
             for (NetherFortressPieces.PieceWeight pieceWeight : NetherFortressPieces.BRIDGE_PIECE_WEIGHTS) {
-                pieceWeight.placeCount = 0;
+                pieceWeight.placeCount.remove(); // Canvas - rewrite moonrise executor
                 this.availableBridgePieces.add(pieceWeight);
             }
 
-            this.availableCastlePieces = Lists.newArrayList();
+            this.availableCastlePieces = java.util.Collections.synchronizedList(Lists.newArrayList()); // Canvas - rewrite moonrise executor
 
             for (NetherFortressPieces.PieceWeight pieceWeight : NetherFortressPieces.CASTLE_PIECE_WEIGHTS) {
-                pieceWeight.placeCount = 0;
+                pieceWeight.placeCount.remove(); // Canvas - rewrite moonrise executor
                 this.availableCastlePieces.add(pieceWeight);
             }
         }
diff --git a/net/minecraft/world/level/levelgen/structure/structures/OceanMonumentPieces.java b/net/minecraft/world/level/levelgen/structure/structures/OceanMonumentPieces.java
index 999ac28a9e371f685554cdec2e645ebe11ddfc92..15454532b55c2acd4f37a4b869ae8b0e99a83217 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/OceanMonumentPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/OceanMonumentPieces.java
@@ -160,8 +160,8 @@ public class OceanMonumentPieces {
         private static final int DEPTH = 58;
         public static final int BIOME_RANGE_CHECK = 29;
         private static final int TOP_POSITION = 61;
-        private OceanMonumentPieces.RoomDefinition sourceRoom;
-        private OceanMonumentPieces.RoomDefinition coreRoom;
+        private volatile OceanMonumentPieces.RoomDefinition sourceRoom; // Canvas - rewrite moonrise executor
+        private volatile OceanMonumentPieces.RoomDefinition coreRoom; // Canvas - rewrite moonrise executor
         private final List<OceanMonumentPieces.OceanMonumentPiece> childPieces = Lists.newArrayList();
 
         public MonumentBuilding(RandomSource random, int x, int z, Direction orientation) {
@@ -1897,9 +1897,9 @@ public class OceanMonumentPieces {
         final int index;
         final OceanMonumentPieces.RoomDefinition[] connections = new OceanMonumentPieces.RoomDefinition[6];
         final boolean[] hasOpening = new boolean[6];
-        boolean claimed;
-        boolean isSource;
-        private int scanIndex;
+        volatile boolean claimed; // Canvas - rewrite moonrise executor
+        volatile boolean isSource; // Canvas - rewrite moonrise executor
+        private volatile int scanIndex; // Canvas - rewrite moonrise executor
 
         public RoomDefinition(int index) {
             this.index = index;
diff --git a/net/minecraft/world/level/levelgen/structure/structures/StrongholdPieces.java b/net/minecraft/world/level/levelgen/structure/structures/StrongholdPieces.java
index 1f1ee6e2d020cd06184313d19523ea928cf242c8..6f785ede9c9864dd705f0c24f6402db2b7e694dc 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/StrongholdPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/StrongholdPieces.java
@@ -65,32 +65,32 @@ public class StrongholdPieces {
             }
         }
     };
-    private static List<StrongholdPieces.PieceWeight> currentPieces;
-    static Class<? extends StrongholdPieces.StrongholdPiece> imposedPiece;
-    private static int totalWeight;
+    private static final ThreadLocal<List<PieceWeight>> currentPieces = new ThreadLocal<List<PieceWeight>>(); // Canvas - rewrite moonrise executor
+    static final ThreadLocal<Class<? extends StrongholdPiece>> imposedPiece = new ThreadLocal<Class<? extends StrongholdPiece>>(); // Canvas - rewrite moonrise executor
+    private static final ThreadLocal<Integer> totalWeight = ThreadLocal.withInitial(() -> 0); // Canvas - rewrite moonrise executor
     static final StrongholdPieces.SmoothStoneSelector SMOOTH_STONE_SELECTOR = new StrongholdPieces.SmoothStoneSelector();
 
     public static void resetPieces() {
-        currentPieces = Lists.newArrayList();
+        currentPieces.set(Lists.newArrayList()); // Canvas - rewrite moonrise executor
 
         for (StrongholdPieces.PieceWeight pieceWeight : STRONGHOLD_PIECE_WEIGHTS) {
-            pieceWeight.placeCount = 0;
-            currentPieces.add(pieceWeight);
+            pieceWeight.placeCount.set(0); // Canvas - rewrite moonrise executor
+            currentPieces.get().add(pieceWeight); // Canvas - rewrite moonrise executor
         }
 
-        imposedPiece = null;
+        imposedPiece.set(null); // Canvas - rewrite moonrise executor
     }
 
     private static boolean updatePieceWeight() {
         boolean flag = false;
-        totalWeight = 0;
+        totalWeight.set(0); // Canvas - rewrite moonrise executor
 
-        for (StrongholdPieces.PieceWeight pieceWeight : currentPieces) {
-            if (pieceWeight.maxPlaceCount > 0 && pieceWeight.placeCount < pieceWeight.maxPlaceCount) {
+        for (StrongholdPieces.PieceWeight pieceWeight : currentPieces.get()) { // Canvas - rewrite moonrise executor
+            if (pieceWeight.maxPlaceCount > 0 && pieceWeight.placeCount.get() < pieceWeight.maxPlaceCount) { // Canvas - rewrite moonrise executor
                 flag = true;
             }
 
-            totalWeight = totalWeight + pieceWeight.weight;
+            totalWeight.set(totalWeight.get() + pieceWeight.weight); // Canvas - rewrite moonrise executor
         }
 
         return flag;
@@ -140,9 +140,9 @@ public class StrongholdPieces {
         if (!updatePieceWeight()) {
             return null;
         } else {
-            if (imposedPiece != null) {
-                StrongholdPieces.StrongholdPiece strongholdPiece = findAndCreatePieceFactory(imposedPiece, pieces, random, x, y, z, direction, genDepth);
-                imposedPiece = null;
+            if (imposedPiece.get() != null) { // Canvas - rewrite moonrise executor
+                StrongholdPieces.StrongholdPiece strongholdPiece = findAndCreatePieceFactory(imposedPiece.get(), pieces, random, x, y, z, direction, genDepth); // Canvas - rewrite moonrise executor
+                imposedPiece.set(null); // Canvas - rewrite moonrise executor
                 if (strongholdPiece != null) {
                     return strongholdPiece;
                 }
@@ -152,9 +152,9 @@ public class StrongholdPieces {
 
             while (i < 5) {
                 i++;
-                int randomInt = random.nextInt(totalWeight);
+                int randomInt = random.nextInt(totalWeight.get()); // Canvas - rewrite moonrise executor
 
-                for (StrongholdPieces.PieceWeight pieceWeight : currentPieces) {
+                for (StrongholdPieces.PieceWeight pieceWeight : currentPieces.get()) { // Canvas - rewrite moonrise executor
                     randomInt -= pieceWeight.weight;
                     if (randomInt < 0) {
                         if (!pieceWeight.doPlace(genDepth) || pieceWeight == piece.previousPiece) {
@@ -165,10 +165,10 @@ public class StrongholdPieces {
                             pieceWeight.pieceClass, pieces, random, x, y, z, direction, genDepth
                         );
                         if (strongholdPiece1 != null) {
-                            pieceWeight.placeCount++;
+                            pieceWeight.placeCount.set(pieceWeight.placeCount.get() + 1); // Canvas - rewrite moonrise executor
                             piece.previousPiece = pieceWeight;
                             if (!pieceWeight.isValid()) {
-                                currentPieces.remove(pieceWeight);
+                                currentPieces.get().remove(pieceWeight); // Canvas - rewrite moonrise executor
                             }
 
                             return strongholdPiece1;
@@ -204,7 +204,7 @@ public class StrongholdPieces {
         private static final int WIDTH = 5;
         private static final int HEIGHT = 5;
         private static final int DEPTH = 7;
-        private boolean hasPlacedChest;
+        private volatile boolean hasPlacedChest; // Canvas - rewrite moonrise executor
 
         public ChestCorridor(int genDepth, RandomSource random, BoundingBox box, Direction orientation) {
             super(StructurePieceType.STRONGHOLD_CHEST_CORRIDOR, genDepth, box);
@@ -690,7 +690,7 @@ public class StrongholdPieces {
     static class PieceWeight {
         public final Class<? extends StrongholdPieces.StrongholdPiece> pieceClass;
         public final int weight;
-        public int placeCount;
+        public final ThreadLocal<Integer> placeCount = ThreadLocal.withInitial(() -> 0); // Canvas - rewrite moonrise executor
         public final int maxPlaceCount;
 
         public PieceWeight(Class<? extends StrongholdPieces.StrongholdPiece> pieceClass, int weight, int maxPlaceCount) {
@@ -700,11 +700,11 @@ public class StrongholdPieces {
         }
 
         public boolean doPlace(int genDepth) {
-            return this.maxPlaceCount == 0 || this.placeCount < this.maxPlaceCount;
+            return this.maxPlaceCount == 0 || this.placeCount.get() < this.maxPlaceCount; // Canvas - rewrite moonrise executor
         }
 
         public boolean isValid() {
-            return this.maxPlaceCount == 0 || this.placeCount < this.maxPlaceCount;
+            return this.maxPlaceCount == 0 || this.placeCount.get() < this.maxPlaceCount; // Canvas - rewrite moonrise executor
         }
     }
 
@@ -712,7 +712,7 @@ public class StrongholdPieces {
         protected static final int WIDTH = 11;
         protected static final int HEIGHT = 8;
         protected static final int DEPTH = 16;
-        private boolean hasPlacedSpawner;
+        private volatile boolean hasPlacedSpawner; // Canvas - rewrite moonrise executor
 
         public PortalRoom(int genDepth, BoundingBox box, Direction orientation) {
             super(StructurePieceType.STRONGHOLD_PORTAL_ROOM, genDepth, box);
@@ -1174,7 +1174,7 @@ public class StrongholdPieces {
         @Override
         public void addChildren(StructurePiece piece, StructurePieceAccessor pieces, RandomSource random) {
             if (this.isSource) {
-                StrongholdPieces.imposedPiece = StrongholdPieces.FiveCrossing.class;
+                StrongholdPieces.imposedPiece.set(StrongholdPieces.FiveCrossing.class); // Canvas - rewrite moonrise executor
             }
 
             this.generateSmallDoorChildForward((StrongholdPieces.StartPiece)piece, pieces, random, 1, 1);
@@ -1223,10 +1223,10 @@ public class StrongholdPieces {
     }
 
     public static class StartPiece extends StrongholdPieces.StairsDown {
-        public StrongholdPieces.PieceWeight previousPiece;
+        public volatile StrongholdPieces.PieceWeight previousPiece; // Canvas - rewrite moonrise executor
         @Nullable
-        public StrongholdPieces.PortalRoom portalRoomPiece;
-        public final List<StructurePiece> pendingChildren = Lists.newArrayList();
+        public volatile StrongholdPieces.PortalRoom portalRoomPiece; // Canvas - rewrite moonrise executor
+        public final List<StructurePiece> pendingChildren = java.util.Collections.synchronizedList(Lists.newArrayList()); // Canvas - rewrite moonrise executor
 
         public StartPiece(RandomSource random, int x, int z) {
             super(StructurePieceType.STRONGHOLD_START, 0, x, z, getRandomHorizontalDirection(random));
@@ -1377,7 +1377,7 @@ public class StrongholdPieces {
         }
     }
 
-    abstract static class StrongholdPiece extends StructurePiece {
+    public abstract static class StrongholdPiece extends StructurePiece { // Canvas - TODO - at
         protected StrongholdPieces.StrongholdPiece.SmallDoorType entryDoor = StrongholdPieces.StrongholdPiece.SmallDoorType.OPENING;
 
         protected StrongholdPiece(StructurePieceType type, int genDepth, BoundingBox boundingBox) {
diff --git a/net/minecraft/world/level/levelgen/structure/structures/SwampHutPiece.java b/net/minecraft/world/level/levelgen/structure/structures/SwampHutPiece.java
index 99999ce7f4918419cc487e6d8ce2591aac111f00..9c6c4e2f4d45c6161ed9fce8b1c2c57f48b86762 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/SwampHutPiece.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/SwampHutPiece.java
@@ -23,8 +23,8 @@ import net.minecraft.world.level.levelgen.structure.pieces.StructurePieceSeriali
 import net.minecraft.world.level.levelgen.structure.pieces.StructurePieceType;
 
 public class SwampHutPiece extends ScatteredFeaturePiece {
-    private boolean spawnedWitch;
-    private boolean spawnedCat;
+    private volatile boolean spawnedWitch; // Canvas - rewrite moonrise executor
+    private volatile boolean spawnedCat; // Canvas - rewrite moonrise executor
 
     public SwampHutPiece(RandomSource random, int x, int z) {
         super(StructurePieceType.SWAMPLAND_HUT, x, 64, z, 7, 7, 9, getRandomHorizontalDirection(random));
diff --git a/net/minecraft/world/level/levelgen/structure/structures/WoodlandMansionPieces.java b/net/minecraft/world/level/levelgen/structure/structures/WoodlandMansionPieces.java
index 750162571a2277966fcddd01053f38b81e9f636f..175bfd3906c847ae8328b07d07c723a7de40b85d 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/WoodlandMansionPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/WoodlandMansionPieces.java
@@ -126,7 +126,7 @@ public class WoodlandMansionPieces {
             int i = 11;
             this.entranceX = 7;
             this.entranceY = 4;
-            this.baseGrid = new WoodlandMansionPieces.SimpleGrid(11, 11, 5);
+            this.baseGrid = new io.canvasmc.canvas.util.chunk.ConcurrentFlagMatrix(11, 11, 5); // Canvas - rewrite moonrise executor
             this.baseGrid.set(this.entranceX, this.entranceY, this.entranceX + 1, this.entranceY + 1, 3);
             this.baseGrid.set(this.entranceX - 1, this.entranceY, this.entranceX - 1, this.entranceY + 1, 2);
             this.baseGrid.set(this.entranceX + 2, this.entranceY - 2, this.entranceX + 3, this.entranceY + 3, 5);
@@ -145,14 +145,14 @@ public class WoodlandMansionPieces {
             }
 
             this.floorRooms = new WoodlandMansionPieces.SimpleGrid[3];
-            this.floorRooms[0] = new WoodlandMansionPieces.SimpleGrid(11, 11, 5);
-            this.floorRooms[1] = new WoodlandMansionPieces.SimpleGrid(11, 11, 5);
-            this.floorRooms[2] = new WoodlandMansionPieces.SimpleGrid(11, 11, 5);
+            this.floorRooms[0] = new io.canvasmc.canvas.util.chunk.ConcurrentFlagMatrix(11, 11, 5); // Canvas - rewrite moonrise executor
+            this.floorRooms[1] = new io.canvasmc.canvas.util.chunk.ConcurrentFlagMatrix(11, 11, 5); // Canvas - rewrite moonrise executor
+            this.floorRooms[2] = new io.canvasmc.canvas.util.chunk.ConcurrentFlagMatrix(11, 11, 5); // Canvas - rewrite moonrise executor
             this.identifyRooms(this.baseGrid, this.floorRooms[0]);
             this.identifyRooms(this.baseGrid, this.floorRooms[1]);
             this.floorRooms[0].set(this.entranceX + 1, this.entranceY, this.entranceX + 1, this.entranceY + 1, 8388608);
             this.floorRooms[1].set(this.entranceX + 1, this.entranceY, this.entranceX + 1, this.entranceY + 1, 8388608);
-            this.thirdFloorGrid = new WoodlandMansionPieces.SimpleGrid(this.baseGrid.width, this.baseGrid.height, 5);
+            this.thirdFloorGrid = new io.canvasmc.canvas.util.chunk.ConcurrentFlagMatrix(this.baseGrid.width, this.baseGrid.height, 5); // Canvas - rewrite moonrise executor
             this.setupThirdFloor();
             this.identifyRooms(this.thirdFloorGrid, this.floorRooms[2]);
         }
@@ -1139,9 +1139,9 @@ public class WoodlandMansionPieces {
     }
 
     static class PlacementData {
-        public Rotation rotation;
-        public BlockPos position;
-        public String wallType;
+        public volatile Rotation rotation; // Canvas - rewrite moonrise executor
+        public volatile BlockPos position; // Canvas - rewrite moonrise executor
+        public volatile String wallType; // Canvas - rewrite moonrise executor
     }
 
     static class SecondFloorRoomCollection extends WoodlandMansionPieces.FloorRoomCollection {
@@ -1181,7 +1181,7 @@ public class WoodlandMansionPieces {
         }
     }
 
-    static class SimpleGrid {
+    public static class SimpleGrid { // Canvas - TODO - at
         private final int[][] grid;
         final int width;
         final int height;
diff --git a/net/minecraft/world/level/levelgen/structure/templatesystem/StructurePlaceSettings.java b/net/minecraft/world/level/levelgen/structure/templatesystem/StructurePlaceSettings.java
index 05027cc20d174d78bef118cd2ba545ac56e1559c..9706a81327cce6eea97970942e07659fc9203997 100644
--- a/net/minecraft/world/level/levelgen/structure/templatesystem/StructurePlaceSettings.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/StructurePlaceSettings.java
@@ -22,7 +22,7 @@ public class StructurePlaceSettings {
     @Nullable
     private RandomSource random;
     public int palette = -1; // CraftBukkit - Set initial value so we know if the palette has been set forcefully
-    private final List<StructureProcessor> processors = Lists.newArrayList();
+    private final List<StructureProcessor> processors = java.util.Collections.synchronizedList(Lists.newArrayList()); // Canvas - rewrite moonrise executor
     private boolean knownShape;
     private boolean finalizeEntities;
 
diff --git a/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java b/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
index f21e612a35d6ac4482dbf5d14e506959659e371a..ef99e0f52e06f3d986037e59d966be0ee4bf6b66 100644
--- a/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
@@ -75,8 +75,8 @@ public class StructureTemplate {
     public static final String ENTITY_TAG_BLOCKPOS = "blockPos";
     public static final String ENTITY_TAG_NBT = "nbt";
     public static final String SIZE_TAG = "size";
-    public final List<StructureTemplate.Palette> palettes = Lists.newArrayList();
-    public final List<StructureTemplate.StructureEntityInfo> entityInfoList = Lists.newArrayList();
+    public final List<StructureTemplate.Palette> palettes = java.util.Collections.synchronizedList(Lists.newArrayList()); // Canvas - rewrite moonrise executor
+    public final List<StructureTemplate.StructureEntityInfo> entityInfoList = java.util.Collections.synchronizedList(Lists.newArrayList()); // Canvas - rewrite moonrise executor
     private Vec3i size = Vec3i.ZERO;
     private String author = "?";
     // CraftBukkit start - data containers
