--- a/io/papermc/paper/threadedregions/commands/CommandServerHealth.java
+++ b/io/papermc/paper/threadedregions/commands/CommandServerHealth.java
@@ -274,6 +_,16 @@
                 .append(Component.text(" - ", LIST, TextDecoration.BOLD))
                 .append(Component.text("Online Players: ", PRIMARY))
                 .append(Component.text(Bukkit.getOnlinePlayers().size() + "\n", INFORMATION))
+                // Canvas start - add build channel and scheduler
+
+                .append(Component.text(" - ", LIST, TextDecoration.BOLD))
+                .append(Component.text("Build Channel: ", PRIMARY))
+                .append(Component.text(io.canvasmc.canvas.Config.ACTIVE_BUILD_CHANNEL.name() + "\n", io.canvasmc.canvas.Config.ACTIVE_BUILD_CHANNEL.color()))
+
+                .append(Component.text(" - ", LIST, TextDecoration.BOLD))
+                .append(Component.text("Scheduler: ", PRIMARY))
+                .append(Component.text(io.papermc.paper.configuration.GlobalConfiguration.get().threadedRegions.scheduler + "\n", INFORMATION))
+                // Canvas end - add build channel and scheduler
 
                 .append(Component.text(" - ", LIST, TextDecoration.BOLD))
                 .append(Component.text("Total regions: ", PRIMARY))
