--- a/io/papermc/paper/threadedregions/commands/CommandServerHealth.java
+++ b/io/papermc/paper/threadedregions/commands/CommandServerHealth.java
@@ -274,6 +_,69 @@
                 .append(Component.text(" - ", LIST, TextDecoration.BOLD))
                 .append(Component.text("Online Players: ", PRIMARY))
                 .append(Component.text(Bukkit.getOnlinePlayers().size() + "\n", INFORMATION))
+                // Canvas start - expand TPS cmd
+
+                .append(Component.text(" - ", LIST, TextDecoration.BOLD))
+                .append(Component.text("Scheduler Type: ", PRIMARY))
+                .append(Component.text(io.papermc.paper.configuration.GlobalConfiguration.get().threadedRegions.scheduler + "\n", INFORMATION))
+
+                .append(Component.text(" - ", LIST, TextDecoration.BOLD))
+                .append(Component.text("Worlds: ", PRIMARY))
+                .append(((java.util.function.Supplier<Component>) () -> {
+                    final java.util.Map<ServerLevel, List<ObjectObjectImmutablePair<ThreadedRegionizer.ThreadedRegion<TickRegions.TickRegionData, TickRegions.TickRegionSectionData>, TickData.TickReportData>>>
+                        world2Regions = new java.util.HashMap<>();
+
+                    for (int i = 0, len = regionsBelowThreshold.size(); i < len; ++i) {
+                        final ObjectObjectImmutablePair<ThreadedRegionizer.ThreadedRegion<TickRegions.TickRegionData, TickRegions.TickRegionSectionData>, TickData.TickReportData>
+                            pair = regionsBelowThreshold.get(i);
+                        final ServerLevel world = pair.left().regioniser.world;
+                        world2Regions.putIfAbsent(world, new ArrayList<>());
+                        world2Regions.get(world).add(pair);
+                    }
+
+                    final TextComponent.Builder builder = Component.text();
+
+                    builder.append(Component.text(world2Regions.size() + "\n", INFORMATION));
+
+                    for (final java.util.Map.Entry<ServerLevel, List<ObjectObjectImmutablePair<ThreadedRegionizer.ThreadedRegion<TickRegions.TickRegionData, TickRegions.TickRegionSectionData>, TickData.TickReportData>>> entry : world2Regions.entrySet()) {
+                        final List<ObjectObjectImmutablePair<ThreadedRegionizer.ThreadedRegion<TickRegions.TickRegionData, TickRegions.TickRegionSectionData>, TickData.TickReportData>>
+                            list = entry.getValue();
+                        final ServerLevel world = entry.getKey();
+
+                        builder.append(Component.text("   ", LIST));
+                        builder.append(Component.text(" - ", LIST));
+                        builder.append(Component.text(world.dimension().identifier().getPath(), INFORMATION));
+                        builder.append(Component.text(": ", PRIMARY));
+                        builder.append(Component.text(list.size() + " region" + (list.size() != 1 ? "s" : ""), SECONDARY));
+                        builder.append(Component.text(", ", PRIMARY));
+
+                        double worldUtil = 0.0;
+                        double combinedTpsVals = 0.0D;
+
+                        for (final ObjectObjectImmutablePair<ThreadedRegionizer.ThreadedRegion<TickRegions.TickRegionData, TickRegions.TickRegionSectionData>, TickData.TickReportData> pair : list) {
+
+                            final TickData.TickReportData report = pair.right();
+
+                            final double util = report == null ? 0.0D : report.utilisation();
+                            final double tps = report == null ? TickRegionScheduler.getTickRate() : report.tpsData().segmentAll().average();
+
+                            worldUtil += util;
+                            combinedTpsVals += tps;
+                        }
+
+                        double averagedTps = list.size() == 1 ? combinedTpsVals : combinedTpsVals / list.size();
+
+                        builder.append(Component.text(TWO_DECIMAL_PLACES.get().format(averagedTps), CommandUtil.getColourForTPS(averagedTps)));
+                        builder.append(Component.text(" average TPS", SECONDARY));
+                        builder.append(Component.text(", ", PRIMARY));
+                        builder.append(Component.text(ONE_DECIMAL_PLACES.get().format(worldUtil * 100.0) + "/" + ONE_DECIMAL_PLACES.get().format(maxThreadCount * 100.0) + "%", CommandUtil.getUtilisationColourRegion(worldUtil)));
+                        builder.append(Component.text(" util\n", SECONDARY));
+                    }
+
+                    return builder.build();
+                }).get())
+
+                // Canvas end - expand TPS cmd
 
                 .append(Component.text(" - ", LIST, TextDecoration.BOLD))
                 .append(Component.text("Total regions: ", PRIMARY))
@@ -293,16 +_,13 @@
                 .append(Component.text(TWO_DECIMAL_PLACES.get().format(genRate) + "\n", INFORMATION))
 
                 .append(Component.text(" - ", LIST, TextDecoration.BOLD))
-                .append(Component.text("Lowest Region TPS: ", PRIMARY))
-                .append(Component.text(TWO_DECIMAL_PLACES.get().format(minTps) + "\n", CommandUtil.getColourForTPS(minTps)))
-
-
-                .append(Component.text(" - ", LIST, TextDecoration.BOLD))
-                .append(Component.text("Median Region TPS: ", PRIMARY))
-                .append(Component.text(TWO_DECIMAL_PLACES.get().format(medianTps) + "\n", CommandUtil.getColourForTPS(medianTps)))
-
-                .append(Component.text(" - ", LIST, TextDecoration.BOLD))
-                .append(Component.text("Highest Region TPS: ", PRIMARY))
+                // Canvas start - compact min/med/max tps
+                .append(Component.text("Min/Median/Max TPS: ", PRIMARY))
+                .append(Component.text(TWO_DECIMAL_PLACES.get().format(minTps), CommandUtil.getColourForTPS(minTps)))
+                .append(Component.text("/", LIST))
+                .append(Component.text(TWO_DECIMAL_PLACES.get().format(medianTps), CommandUtil.getColourForTPS(medianTps)))
+                .append(Component.text("/", LIST))
+                // Canvas end - compact min/med/max tps
                 .append(Component.text(TWO_DECIMAL_PLACES.get().format(maxTps) + "\n", CommandUtil.getColourForTPS(maxTps)))
 
                 .append(Component.text("Highest ", HEADER, TextDecoration.BOLD))
