--- a/io/papermc/paper/threadedregions/commands/CommandServerHealth.java
+++ b/io/papermc/paper/threadedregions/commands/CommandServerHealth.java
@@ -267,6 +_,19 @@
             lowestRegionsBuilder.append(line);
         }
 
+        // Canvas start - add build channel
+        io.canvasmc.canvas.util.ApiClient.Channel channel = io.canvasmc.canvas.util.ApiClient.Channel.UNKNOWN;
+        io.papermc.paper.ServerBuildInfo buildInfo = io.papermc.paper.ServerBuildInfo.buildInfo();
+        int build = buildInfo.buildNumber().orElse(-1);
+        if (build == -1) {
+            channel = io.canvasmc.canvas.util.ApiClient.Channel.LOCAL;
+        } else {
+            try {
+                channel = io.canvasmc.canvas.util.CanvasVersionFetcher.CLIENT.getBuild(build).channel();
+            } catch (Throwable ignored) {
+            }
+        }
+        // Canvas end - add build channel
         sender.sendMessage(
             Component.text()
                 .append(Component.text("Server Health Report\n", HEADER, TextDecoration.BOLD))
@@ -274,6 +_,17 @@
                 .append(Component.text(" - ", LIST, TextDecoration.BOLD))
                 .append(Component.text("Online Players: ", PRIMARY))
                 .append(Component.text(Bukkit.getOnlinePlayers().size() + "\n", INFORMATION))
+                // Canvas start - add build channel and scheduler
+
+                .append(Component.text(" - ", LIST, TextDecoration.BOLD))
+                .append(Component.text("Build Channel: ", PRIMARY))
+                .append(Component.text(channel.name() + "\n", channel.color())
+                    .hoverEvent(HoverEvent.hoverEvent(HoverEvent.Action.SHOW_TEXT, Component.text(build == -1 ? "Development build, use caution" : "Build - " + build))))
+
+                .append(Component.text(" - ", LIST, TextDecoration.BOLD))
+                .append(Component.text("Scheduler: ", PRIMARY))
+                .append(Component.text(io.papermc.paper.configuration.GlobalConfiguration.get().threadedRegions.scheduler + "\n", INFORMATION))
+                // Canvas end - add build channel and scheduler
 
                 .append(Component.text(" - ", LIST, TextDecoration.BOLD))
                 .append(Component.text("Total regions: ", PRIMARY))
