From 3cf7451517453a52df7cb21a31824aadeaf7c6d9 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/io/papermc/paper/threadedregions/commands/CommandServerHealth.java b/io/papermc/paper/threadedregions/commands/CommandServerHealth.java
index 9100fa2575707d9fdacf7e731fd4d89f64bd12e3..de7c9fcd28250c21c6a572c313fca3868544b328 100644
--- a/io/papermc/paper/threadedregions/commands/CommandServerHealth.java
+++ b/io/papermc/paper/threadedregions/commands/CommandServerHealth.java
@@ -274,6 +274,26 @@ public final class CommandServerHealth extends Command {
                 .append(Component.text(" - ", LIST, TextDecoration.BOLD))
                 .append(Component.text("Online Players: ", PRIMARY))
                 .append(Component.text(Bukkit.getOnlinePlayers().size() + "\n", INFORMATION))
+                // Canvas start - add build channel and scheduler
+
+                .append(Component.text(" - ", LIST, TextDecoration.BOLD))
+                .append(Component.text("Build Channel: ", PRIMARY))
+                .append(
+                    Component.text(io.canvasmc.canvas.Config.ACTIVE_BUILD_CHANNEL.name() + "\n", TextColor.color(io.canvasmc.canvas.Config.ACTIVE_BUILD_CHANNEL.color()))
+                        .hoverEvent(HoverEvent.showText(
+                            Component.text("Running ", HEADER).append(Component.text(io.papermc.paper.ServerBuildInfo.buildInfo().brandName(), PRIMARY))
+                                .appendNewline()
+                                .append(Component.text(io.canvasmc.canvas.Config.ACTIVE_BUILD_CHANNEL == io.canvasmc.canvas.util.version.ApiClient.BuildStatus.LOCAL
+                                    ? "Local-Build" : String.valueOf(io.papermc.paper.ServerBuildInfo.buildInfo().buildNumber().orElse(-1)), SECONDARY))
+                                .append(Component.text("-", LIST))
+                                .append(Component.text(io.papermc.paper.ServerBuildInfo.buildInfo().gitBranch().orElse("Unknown Branch"), SECONDARY))
+                        ))
+                )
+
+                .append(Component.text(" - ", LIST, TextDecoration.BOLD))
+                .append(Component.text("Scheduler: ", PRIMARY))
+                .append(Component.text(io.papermc.paper.configuration.GlobalConfiguration.get().threadedRegions.scheduler + "\n", INFORMATION))
+                // Canvas end - add build channel and scheduler
 
                 .append(Component.text(" - ", LIST, TextDecoration.BOLD))
                 .append(Component.text("Total regions: ", PRIMARY))
