From 53e1a57c72953de1bfcc94c3257c724e7fa1f88f Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/raid/Raid.java b/net/minecraft/world/entity/raid/Raid.java
index 24cc4a15731edcb701596fe7051e188637457e37..4ca5cf3f8b90306194dc26ad160013af6ce1a6f6 100644
--- a/net/minecraft/world/entity/raid/Raid.java
+++ b/net/minecraft/world/entity/raid/Raid.java
@@ -554,7 +554,7 @@ public class Raid {
             double x3 = playerLoc.x + 13.0 / distBtwn * (raidLoc.x - playerLoc.x);
             double z3 = playerLoc.z + 13.0 / distBtwn * (raidLoc.z - playerLoc.z);
             if (distBtwn <= 64.0 || playersInRaid.contains(player)) {
-                player.connection.send(new ClientboundSoundPacket(SoundEvents.RAID_HORN, SoundSource.NEUTRAL, x3, player.getY(), z3, 64.0F, 1.0F, seed));
+                player.connection.send(new ClientboundSoundPacket(SoundEvents.RAID_HORN, io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaFixes.mc245394 ? SoundSource.HOSTILE : SoundSource.NEUTRAL, x3, player.getY(), z3, 64.0F, 1.0F, seed)); // Canvas - MC-245394
             }
         }
     }
