From 400a26b77b55ac5095fe4c99f70e24489a3947c7 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/animal/fox/Fox.java b/net/minecraft/world/entity/animal/fox/Fox.java
index cc60dbca2714f0da46c4b33bce6e631ae1094761..7925ccbdf5c2f3a96ff57d3b06a035685abc424b 100644
--- a/net/minecraft/world/entity/animal/fox/Fox.java
+++ b/net/minecraft/world/entity/animal/fox/Fox.java
@@ -1057,7 +1057,7 @@ public class Fox extends Animal {
 
         @Override
         public boolean canUse() {
-            return Fox.this.isInWater() && Fox.this.getFluidHeight(FluidTags.WATER) > 0.25 || Fox.this.isInLava();
+            return Fox.this.isInWater() && Fox.this.canvas$getWaterFluidHeight() > 0.25 || Fox.this.isInLava(); // Canvas - optimize fluid height
         }
     }
 
