--- a/net/minecraft/server/waypoints/ServerWaypointManager.java
+++ b/net/minecraft/server/waypoints/ServerWaypointManager.java
@@ -27,25 +_,12 @@
 
     @Override
     public void updateWaypoint(WaypointTransmitter waypoint) {
-        if (this.waypoints.contains(waypoint)) {
-            Map<ServerPlayer, WaypointTransmitter.Connection> map = Tables.transpose(this.connections).row(waypoint);
-            SetView<ServerPlayer> set = Sets.difference(this.players, map.keySet());
-
-            for (Entry<ServerPlayer, WaypointTransmitter.Connection> entry : ImmutableSet.copyOf(map.entrySet())) {
-                this.updateConnection(entry.getKey(), waypoint, entry.getValue());
-            }
-
-            for (ServerPlayer serverPlayer : set) {
-                this.createConnection(serverPlayer, waypoint);
-            }
-        }
+        // Canvas - region threading
     }
 
     @Override
     public void untrackWaypoint(WaypointTransmitter waypoint) {
-        this.connections.column(waypoint).forEach((serverPlayer, connection) -> connection.disconnect());
-        Tables.transpose(this.connections).row(waypoint).clear();
-        this.waypoints.remove(waypoint);
+        // Canvas - region threading
     }
 
     public void addPlayer(ServerPlayer player) {
@@ -61,18 +_,15 @@
     }
 
     public void breakAllConnections() {
-        this.connections.values().forEach(WaypointTransmitter.Connection::disconnect);
-        this.connections.clear();
+        // Canvas - region threading
     }
 
     public void remakeConnections(WaypointTransmitter waypoint) {
-        for (ServerPlayer serverPlayer : this.players) {
-            this.createConnection(serverPlayer, waypoint);
-        }
+        // Canvas - region threading
     }
 
     public Set<WaypointTransmitter> transmitters() {
-        return this.waypoints;
+        throw new UnsupportedOperationException("Unsupported in region threading");
     }
 
     private static boolean isLocatorBarEnabledFor(ServerPlayer player) {
@@ -80,36 +_,10 @@
     }
 
     private void createConnection(ServerPlayer player, WaypointTransmitter waypoint) {
-        if (player != waypoint) {
-            if (isLocatorBarEnabledFor(player)) {
-                waypoint.makeWaypointConnectionWith(player).ifPresentOrElse(connection -> {
-                    this.connections.put(player, waypoint, connection);
-                    connection.connect();
-                }, () -> {
-                    WaypointTransmitter.Connection connection = this.connections.remove(player, waypoint);
-                    if (connection != null) {
-                        connection.disconnect();
-                    }
-                });
-            }
-        }
+        // Canvas - region threading
     }
 
     private void updateConnection(ServerPlayer player, WaypointTransmitter waypoint, WaypointTransmitter.Connection connection) {
-        if (player != waypoint) {
-            if (isLocatorBarEnabledFor(player)) {
-                if (!connection.isBroken()) {
-                    connection.update();
-                } else {
-                    waypoint.makeWaypointConnectionWith(player).ifPresentOrElse(connection1 -> {
-                        connection1.connect();
-                        this.connections.put(player, waypoint, connection1);
-                    }, () -> {
-                        connection.disconnect();
-                        this.connections.remove(player, waypoint);
-                    });
-                }
-            }
-        }
+        // Canvas - region threading
     }
 }
