--- a/net/minecraft/server/waypoints/ServerWaypointManager.java
+++ b/net/minecraft/server/waypoints/ServerWaypointManager.java
@@ -22,11 +_,20 @@
 
     @Override
     public void trackWaypoint(WaypointTransmitter waypoint) {
-        // Folia - region threading
+        throw new UnsupportedOperationException("Unsupported in region threading");
+        /*
+        this.waypoints.add(waypoint);
+
+        for (ServerPlayer serverPlayer : this.players) {
+            this.createConnection(serverPlayer, waypoint);
+        }
+         */
     }
 
     @Override
     public void updateWaypoint(WaypointTransmitter waypoint) {
+        throw new UnsupportedOperationException("Unsupported in region threading");
+        /*
         if (this.waypoints.contains(waypoint)) {
             Map<ServerPlayer, WaypointTransmitter.Connection> map = Tables.transpose(this.connections).row(waypoint);
             SetView<ServerPlayer> set = Sets.difference(this.players, map.keySet());
@@ -39,47 +_,92 @@
                 this.createConnection(serverPlayer, waypoint);
             }
         }
+         */
     }
 
     @Override
     public void untrackWaypoint(WaypointTransmitter waypoint) {
+        throw new UnsupportedOperationException("Unsupported in region threading");
+        /*
         this.connections.column(waypoint).forEach((serverPlayer, connection) -> connection.disconnect());
         Tables.transpose(this.connections).row(waypoint).clear();
         this.waypoints.remove(waypoint);
+         */
     }
 
     public void addPlayer(ServerPlayer player) {
-        // Folia - region threading
+        throw new UnsupportedOperationException("Unsupported in region threading");
+        /*
+        this.players.add(player);
+
+        for (WaypointTransmitter waypointTransmitter : this.waypoints) {
+            this.createConnection(player, waypointTransmitter);
+        }
+
+        if (player.isTransmittingWaypoint()) {
+            this.trackWaypoint((WaypointTransmitter)player);
+        }
+         */
     }
 
     public void updatePlayer(ServerPlayer player) {
-        // Folia - region threading
+        throw new UnsupportedOperationException("Unsupported in region threading");
+        /*
+        Map<WaypointTransmitter, WaypointTransmitter.Connection> map = this.connections.row(player);
+        SetView<WaypointTransmitter> set = Sets.difference(this.waypoints, map.keySet());
+
+        for (Entry<WaypointTransmitter, WaypointTransmitter.Connection> entry : ImmutableSet.copyOf(map.entrySet())) {
+            this.updateConnection(player, entry.getKey(), entry.getValue());
+        }
+
+        for (WaypointTransmitter waypointTransmitter : set) {
+            this.createConnection(player, waypointTransmitter);
+        }
+         */
     }
 
     public void removePlayer(ServerPlayer player) {
-        // Folia - region threading
+        throw new UnsupportedOperationException("Unsupported in region threading");
+        /*
+        this.connections.row(player).values().removeIf(connection -> {
+            connection.disconnect();
+            return true;
+        });
+        this.untrackWaypoint((WaypointTransmitter)player);
+        this.players.remove(player);
+         */
     }
 
     public void breakAllConnections() {
+        throw new UnsupportedOperationException("Unsupported in region threading");
+        /*
         this.connections.values().forEach(WaypointTransmitter.Connection::disconnect);
         this.connections.clear();
+         */
     }
 
     public void remakeConnections(WaypointTransmitter waypoint) {
+        throw new UnsupportedOperationException("Unsupported in region threading");
+        /*
         for (ServerPlayer serverPlayer : this.players) {
             this.createConnection(serverPlayer, waypoint);
         }
+         */
     }
 
     public Set<WaypointTransmitter> transmitters() {
+        if (true) throw new UnsupportedOperationException("Unsupported in region threading");
         return this.waypoints;
     }
 
     private static boolean isLocatorBarEnabledFor(ServerPlayer player) {
+        if (true) throw new UnsupportedOperationException("Unsupported in region threading");
         return player.level().getGameRules().getBoolean(GameRules.RULE_LOCATOR_BAR); // Paper - per world gamerules
     }
 
     private void createConnection(ServerPlayer player, WaypointTransmitter waypoint) {
+        throw new UnsupportedOperationException("Unsupported in region threading");
+        /*
         if (player != waypoint) {
             if (isLocatorBarEnabledFor(player)) {
                 waypoint.makeWaypointConnectionWith(player).ifPresentOrElse(connection -> {
@@ -93,9 +_,12 @@
                 });
             }
         }
+         */
     }
 
     private void updateConnection(ServerPlayer player, WaypointTransmitter waypoint, WaypointTransmitter.Connection connection) {
+        throw new UnsupportedOperationException("Unsupported in region threading");
+        /*
         if (player != waypoint) {
             if (isLocatorBarEnabledFor(player)) {
                 if (!connection.isBroken()) {
@@ -111,5 +_,6 @@
                 }
             }
         }
+         */
     }
 }
