--- a/net/minecraft/world/entity/animal/Pufferfish.java
+++ b/net/minecraft/world/entity/animal/Pufferfish.java
@@ -163,6 +_,7 @@
         int puffState = this.getPuffState();
         if (entity instanceof ServerPlayer serverPlayer
             && puffState > 0
+            && (!io.canvasmc.canvas.Config.INSTANCE.fixes.mc155509 || this.isAlive()) // Canvas - MC-155509
             && entity.hurtServer(serverPlayer.level(), this.damageSources().mobAttack(this), 1 + puffState)) {
             if (!this.isSilent()) {
                 serverPlayer.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.PUFFER_FISH_STING, 0.0F));
