--- a/net/minecraft/world/entity/animal/Rabbit.java
+++ b/net/minecraft/world/entity/animal/Rabbit.java
@@ -216,7 +_,7 @@
     public void jumpFromGround() {
         super.jumpFromGround();
         double speedModifier = this.moveControl.getSpeedModifier();
-        if (speedModifier > 0.0) {
+        if (speedModifier > 0.0 || true) { // Canvas - if we say jump, jump?
             double d = this.getDeltaMovement().horizontalDistanceSqr();
             if (d < 0.01) {
                 this.moveRelative(0.1F, new Vec3(0.0, 0.0, 1.0));
