--- a/net/minecraft/world/inventory/LoomMenu.java
+++ b/net/minecraft/world/inventory/LoomMenu.java
@@ -2,6 +_,7 @@
 
 import com.google.common.collect.ImmutableList;
 import java.util.List;
+import io.canvasmc.canvas.Config;
 import net.minecraft.core.Holder;
 import net.minecraft.core.HolderGetter;
 import net.minecraft.core.component.DataComponents;
@@ -219,7 +_,7 @@
 
             if (holder != null) {
                 BannerPatternLayers bannerPatternLayers = item.getOrDefault(DataComponents.BANNER_PATTERNS, BannerPatternLayers.EMPTY);
-                boolean flag = bannerPatternLayers.layers().size() >= 6;
+                boolean flag = bannerPatternLayers.layers().size() >= Config.INSTANCE.blocks.loomMaxPatternCount; // Canvas - max pattern count
                 if (flag) {
                     this.selectedBannerPatternIndex.set(-1);
                     this.resultSlot.set(ItemStack.EMPTY);
