--- a/src/main/java/ca/spottedleaf/moonrise/common/list/ShortList.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/list/ShortList.java
@@ -1,11 +_,13 @@
 package ca.spottedleaf.moonrise.common.list;
 
+import it.unimi.dsi.fastutil.shorts.Short2ShortMap;
+import it.unimi.dsi.fastutil.shorts.Short2ShortMaps;
 import it.unimi.dsi.fastutil.shorts.Short2ShortOpenHashMap;
 import java.util.Arrays;
 
 public final class ShortList {
 
-    private final Short2ShortOpenHashMap map = new Short2ShortOpenHashMap();
+    private final Short2ShortMap map = Short2ShortMaps.synchronize(new Short2ShortOpenHashMap());
     {
         this.map.defaultReturnValue(Short.MIN_VALUE);
     }
@@ -15,11 +_,11 @@
     private short[] byIndex = EMPTY_LIST;
     private short count;
 
-    public int size() {
+    public synchronized int size() {
         return (int)this.count;
     }
 
-    public short getRaw(final int index) {
+    public synchronized short getRaw(final int index) {
         return this.byIndex[index];
     }
 
@@ -30,7 +_,7 @@
         }
     }
 
-    public boolean add(final short value) {
+    public synchronized boolean add(final short value) {
         final int count = (int)this.count;
         final short currIndex = this.map.putIfAbsent(value, (short)count);
 
@@ -51,7 +_,7 @@
         return true;
     }
 
-    public boolean remove(final short value) {
+    public synchronized boolean remove(final short value) {
         final short index = this.map.remove(value);
         if (index == Short.MIN_VALUE) {
             return false;
@@ -70,7 +_,7 @@
         return true;
     }
 
-    public void clear() {
+    public synchronized void clear() {
         this.count = (short)0;
         this.map.clear();
     }
