--- a/net/minecraft/server/commands/RideCommand.java
+++ b/net/minecraft/server/commands/RideCommand.java
@@ -66,12 +_,31 @@
             throw ERROR_MOUNTING_LOOP.create();
         } else if (target.level() != vehicle.level()) {
             throw ERROR_WRONG_DIMENSION.create();
-        } else if (!target.startRiding(vehicle, true)) {
-            throw ERROR_MOUNT_FAILED.create(target.getDisplayName(), vehicle.getDisplayName());
-        } else {
-            source.sendSuccess(() -> Component.translatable("commands.ride.mount.success", target.getDisplayName(), vehicle.getDisplayName()), true);
-            return 1;
+        // Canvas start - fix commands with region threading
         }
+        // by this point, all other checks are done, and we just need to attempt to ride
+        // we teleport to the vehicle location first, so they are both on the correct region
+        target.teleportAsync(
+            (net.minecraft.server.level.ServerLevel) vehicle.level(),
+            vehicle.position(),
+            vehicle.getYRot(),
+            vehicle.getXRot(),
+            net.minecraft.world.phys.Vec3.ZERO,
+            org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.COMMAND,
+            Entity.TELEPORT_FLAG_LOAD_CHUNK, (entity) -> {
+                try {
+                    if (!entity.startRiding(vehicle, true)) {
+                        throw ERROR_MOUNT_FAILED.create(target.getDisplayName(), vehicle.getDisplayName());
+                    } else {
+                        source.sendSuccess(() -> Component.translatable("commands.ride.mount.success", target.getDisplayName(), vehicle.getDisplayName()), true);
+                    }
+                } catch (CommandSyntaxException exception) {
+                    source.sendFailure(Component.literal(exception.getMessage()));
+                }
+            }
+        );
+        return 1;
+        // Canvas end - fix commands with region threading
     }
 
     private static int dismount(CommandSourceStack source, Entity target) throws CommandSyntaxException {
