--- a/net/minecraft/server/commands/RotateCommand.java
+++ b/net/minecraft/server/commands/RotateCommand.java
@@ -74,13 +_,17 @@
 
     private static int rotate(CommandSourceStack source, Entity entity, Coordinates coordinates) {
         Vec2 rotation = coordinates.getRotation(source);
+        entity.getBukkitEntity().taskScheduler.schedule((entity1) -> { // Canvas - fix commands with region threading
         entity.forceSetRotation(rotation.y, rotation.x);
+        }, null, 0L); // Canvas - fix commands with region threading
         source.sendSuccess(() -> Component.translatable("commands.rotate.success", entity.getDisplayName()), true);
         return 1;
     }
 
     private static int rotate(CommandSourceStack source, Entity entity, LookAt lookAt) {
+        entity.getBukkitEntity().taskScheduler.schedule((entity1) -> { // Canvas - fix commands with region threading
         lookAt.perform(source, entity);
+        }, null, 0L); // Canvas - fix commands with region threading
         source.sendSuccess(() -> Component.translatable("commands.rotate.success", entity.getDisplayName()), true);
         return 1;
     }
