--- a/net/minecraft/server/commands/TeleportCommand.java
+++ b/net/minecraft/server/commands/TeleportCommand.java
@@ -275,8 +_,8 @@
             double d = relatives.contains(Relative.X) ? x - target.getX() : x;
             double d1 = relatives.contains(Relative.Y) ? y - target.getY() : y;
             double d2 = relatives.contains(Relative.Z) ? z - target.getZ() : z;
-            float f = relatives.contains(Relative.Y_ROT) ? yRot - target.getYRot() : yRot;
-            float f1 = relatives.contains(Relative.X_ROT) ? xRot - target.getXRot() : xRot;
+            float f = target == source.getEntity() ? target.getYRot() : relatives.contains(Relative.Y_ROT) ? yRot - target.getYRot() : yRot; // Canvas - fix pitch/yaw with teleport command
+            float f1 = target == source.getEntity() ? target.getXRot() : relatives.contains(Relative.X_ROT) ? xRot - target.getXRot() : xRot; // Canvas - fix pitch/yaw with teleport command
             float f2 = Mth.wrapDegrees(f);
             float f3 = Mth.wrapDegrees(f1);
             // Folia start - region threading
