--- a/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
+++ b/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
@@ -146,7 +_,7 @@
 
             if (this.attachedToEntity != null) {
                 Vec3 handHoldingItemAngle;
-                if (this.attachedToEntity.isFallFlying()) {
+                if (this.attachedToEntity.isFallFlying() && (!io.canvasmc.canvas.Config.INSTANCE.fixes.mc119754 || !this.attachedToEntity.isSpectator())) { // Canvas - MC-119754
                     Vec3 lookAngle = this.attachedToEntity.getLookAngle();
                     double d = 1.5;
                     double d1 = 0.1;
