--- a/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
+++ b/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
@@ -1,5 +_,6 @@
 package net.minecraft.world.entity.projectile;
 
+import io.canvasmc.canvas.Config;
 import it.unimi.dsi.fastutil.doubles.DoubleDoubleImmutablePair;
 import java.util.List;
 import java.util.OptionalInt;
@@ -364,4 +_,14 @@
         double d1 = entity.position().z - this.position().z;
         return DoubleDoubleImmutablePair.of(d, d1);
     }
+    // Canvas start
+
+    @Override
+    public boolean shouldBeSaved() {
+        if (Config.INSTANCE.disableFireworkSaving) {
+            return false;
+        }
+        return super.shouldBeSaved();
+    }
+    // Canvas end
 }
