From 60fbf10de1ffff26663053a967c13ae1e63a5a7a Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/projectile/FireworkRocketEntity.java b/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
index ca9e604cb81cf3c3b0312dd02265388251872283..787f9d050f632ad8ac28144eaf1a1b61b69a0dff 100644
--- a/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
+++ b/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
@@ -147,7 +147,7 @@ public class FireworkRocketEntity extends Projectile implements ItemSupplier {
 
             if (this.attachedToEntity != null) {
                 Vec3 handAngle;
-                if (this.attachedToEntity.isFallFlying()) {
+                if (this.attachedToEntity.isFallFlying() && (!io.canvasmc.canvas.Config.INSTANCE.fixes.mc119754 || !this.attachedToEntity.isSpectator())) { // Canvas - MC-119754
                     Vec3 lookAngle = this.attachedToEntity.getLookAngle();
                     double power = 1.5;
                     double powerAdd = 0.1;
