--- a/net/minecraft/world/entity/projectile/FishingHook.java
+++ b/net/minecraft/world/entity/projectile/FishingHook.java
@@ -520,6 +_,11 @@
         if (!this.level().isClientSide && playerOwner != null && !this.shouldStopFishing(playerOwner)) {
             int i = 0;
             if (this.hookedIn != null) {
+                // Canvas start - combat configs
+                if (!io.canvasmc.canvas.Config.INSTANCE.combat.fishingRodPulls) {
+                    return 0; // act as if we canceled the pull
+                }
+                // Canvas end - combat configs
                 // CraftBukkit start
                 org.bukkit.event.player.PlayerFishEvent playerFishEvent = new org.bukkit.event.player.PlayerFishEvent((org.bukkit.entity.Player) playerOwner.getBukkitEntity(), this.hookedIn.getBukkitEntity(), (org.bukkit.entity.FishHook) this.getBukkitEntity(), org.bukkit.craftbukkit.CraftEquipmentSlot.getHand(hand), org.bukkit.event.player.PlayerFishEvent.State.CAUGHT_ENTITY); // Paper - Add hand parameter to PlayerFishEvent
                 if (!playerFishEvent.callEvent()) {
@@ -613,6 +_,14 @@
         if (owner != null) {
             Vec3 vec3 = new Vec3(owner.getX() - this.getX(), owner.getY() - this.getY(), owner.getZ() - this.getZ()).scale(0.1);
             entity.setDeltaMovement(entity.getDeltaMovement().add(vec3));
+            // Canvas start - MC-100991
+            if (io.canvasmc.canvas.Config.INSTANCE.fixes.mc100991 && entity instanceof net.minecraft.world.entity.LivingEntity livingEntity) {
+                livingEntity.getCombatTracker().recordDamage(
+                    level().damageSources().thrown(this, owner),
+                    livingEntity.getHealth()
+                );
+            }
+            // Canvas end - MC-100991
         }
     }
 
