From 30322d32a063dc3b6c41ba7f9d75c2e508ff7cf1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/projectile/ProjectileDeflection.java b/net/minecraft/world/entity/projectile/ProjectileDeflection.java
index 46258d08575033e98ac5ebfd66cf958cb088cc5f..7f8fedd4c7954c0f9f440c074cba9a4aa84398ab 100644
--- a/net/minecraft/world/entity/projectile/ProjectileDeflection.java
+++ b/net/minecraft/world/entity/projectile/ProjectileDeflection.java
@@ -17,7 +17,7 @@ public interface ProjectileDeflection {
     };
     ProjectileDeflection AIM_DEFLECT = (projectile, entity, random) -> {
         if (entity != null) {
-            Vec3 lookAngle = entity.getLookAngle();
+            Vec3 lookAngle = entity.canvas$getLookAngleThreadSafe(); // Canvas - region threading
             projectile.setDeltaMovement(lookAngle);
             projectile.needsSync = true;
         }
