--- a/net/minecraft/world/entity/projectile/ProjectileDeflection.java
+++ b/net/minecraft/world/entity/projectile/ProjectileDeflection.java
@@ -17,7 +_,7 @@
     };
     ProjectileDeflection AIM_DEFLECT = (projectile, entity, random) -> {
         if (entity != null) {
-            Vec3 lookAngle = entity.getLookAngle();
+            Vec3 lookAngle = entity.canvas$getLookAngleThreadSafe(); // Canvas - region threading
             projectile.setDeltaMovement(lookAngle);
             projectile.needsSync = true;
         }
