From a3c96b695f9e13d71875813ec2583d31d8569479 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/projectile/ShulkerBullet.java b/net/minecraft/world/entity/projectile/ShulkerBullet.java
index 27c143bcc0b64b951027be0a6cac0bba9f641cda..c7eeb33832459e22470db4826a354ea76034695d 100644
--- a/net/minecraft/world/entity/projectile/ShulkerBullet.java
+++ b/net/minecraft/world/entity/projectile/ShulkerBullet.java
@@ -226,6 +226,16 @@ public class ShulkerBullet extends Projectile {
 
         Vec3 movement = this.getDeltaMovement();
         this.setPos(this.position().add(movement));
+        // Canvas start - MC-221257
+        if (io.canvasmc.canvas.Config.INSTANCE.fixes.mc221257 && this.isInWater()) {
+            Vec3 vec3 = this.position();
+            Vec3 vec32 = this.getDeltaMovement();
+
+            for (int i = 0; i < 4; ++i) {
+                this.level().addParticle(ParticleTypes.BUBBLE, vec3.x - vec32.x * (double) 0.25F, vec3.y - vec32.y * (double) 0.25F, vec3.z - vec32.z * (double) 0.25F, vec32.x, vec32.y, vec32.z);
+            }
+        }
+        // Canvas end - MC-22157
         this.applyEffectsFromBlocks();
         if (this.portalProcess != null && this.portalProcess.isInsidePortalThisTick()) {
             this.handlePortal();
@@ -304,6 +314,12 @@ public class ShulkerBullet extends Projectile {
             if (target instanceof LivingEntity livingTarget) {
                 livingTarget.addEffect(new MobEffectInstance(MobEffects.LEVITATION, 200), MoreObjects.firstNonNull(owner, this), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); // CraftBukkit
             }
+            // Canvas start - MC-227337
+            if (io.canvasmc.canvas.Config.INSTANCE.fixes.mc227337) {
+                ((ServerLevel) this.level()).sendParticles(ParticleTypes.EXPLOSION, this.getX(), this.getY(), this.getZ(), 2, 0.2D, 0.2D, 0.2D, 0.0D);
+                this.playSound(SoundEvents.SHULKER_BULLET_HIT, 1.0F, 1.0F);
+            }
+            // Canvas end - MC-227337
         }
     }
 
