--- a/net/minecraft/world/entity/projectile/ShulkerBullet.java
+++ b/net/minecraft/world/entity/projectile/ShulkerBullet.java
@@ -231,6 +_,16 @@
 
         Vec3 deltaMovement = this.getDeltaMovement();
         this.setPos(this.position().add(deltaMovement));
+        // Canvas start - MC-221257
+        if (io.canvasmc.canvas.Config.INSTANCE.fixes.mc221257 && this.isInWater()) {
+            Vec3 vec3 = this.position();
+            Vec3 vec32 = this.getDeltaMovement();
+
+            for (int i = 0; i < 4; ++i) {
+                this.level().addParticle(ParticleTypes.BUBBLE, vec3.x - vec32.x * (double) 0.25F, vec3.y - vec32.y * (double) 0.25F, vec3.z - vec32.z * (double) 0.25F, vec32.x, vec32.y, vec32.z);
+            }
+        }
+        // Canvas end - MC-22157
         this.applyEffectsFromBlocks();
         if (this.portalProcess != null && this.portalProcess.isInsidePortalThisTick()) {
             this.handlePortal();
@@ -318,6 +_,12 @@
             if (entity instanceof LivingEntity livingEntity1) {
                 livingEntity1.addEffect(new MobEffectInstance(MobEffects.LEVITATION, 200), MoreObjects.firstNonNull(owner, this), org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ATTACK); // CraftBukkit
             }
+            // Canvas start - MC-227337
+            if (io.canvasmc.canvas.Config.INSTANCE.fixes.mc227337) {
+                ((ServerLevel) this.level()).sendParticles(ParticleTypes.EXPLOSION, this.getX(), this.getY(), this.getZ(), 2, 0.2D, 0.2D, 0.2D, 0.0D);
+                this.playSound(SoundEvents.SHULKER_BULLET_HIT, 1.0F, 1.0F);
+            }
+            // Canvas end - MC-227337
         }
     }
 
