--- a/net/minecraft/world/entity/projectile/Snowball.java
+++ b/net/minecraft/world/entity/projectile/Snowball.java
@@ -1,5 +_,6 @@
 package net.minecraft.world.entity.projectile;
 
+import io.canvasmc.canvas.Config;
 import net.minecraft.core.particles.ItemParticleOption;
 import net.minecraft.core.particles.ParticleOptions;
 import net.minecraft.core.particles.ParticleTypes;
@@ -94,4 +_,14 @@
             this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.HIT); // CraftBukkit - add Bukkit remove cause
         }
     }
+    // Canvas start
+
+    @Override
+    public boolean shouldBeSaved() {
+        if (Config.INSTANCE.entities.disableSnowballSaving) {
+            return false;
+        }
+        return super.shouldBeSaved();
+    }
+    // Canvas end
 }
