--- a/net/minecraft/world/entity/projectile/ThrownEgg.java
+++ b/net/minecraft/world/entity/projectile/ThrownEgg.java
@@ -58,6 +_,12 @@
     protected void onHitEntity(EntityHitResult result) {
         super.onHitEntity(result);
         result.getEntity().hurt(this.damageSources().thrown(this, this.getOwner()), 0.0F);
+        // Canvas start - egg can knockback player config
+        if (io.canvasmc.canvas.Config.INSTANCE.eggCanKnockback && result.getEntity() instanceof net.minecraft.server.level.ServerPlayer serverPlayer) {
+            serverPlayer.knockback(0.4F, this.getX() - result.getEntity().getX(), this.getZ() - result.getEntity().getZ());
+            serverPlayer.hurtMarked = true;
+        }
+        // Canvas end - egg can knockback player config
     }
 
     /**
