--- a/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
+++ b/net/minecraft/world/entity/projectile/ThrownEnderpearl.java
@@ -52,11 +_,23 @@
     }
 
     private void deregisterFromCurrentOwner() {
-        // Folia - region threading - we remove the registration logic, we do not need to fetch the owner
+        // Canvas start - region threading
+        if (io.canvasmc.canvas.Config.INSTANCE.restoreVanillaEnderPearlBehavior) {
+            if (this.getOwner() instanceof ServerPlayer serverPlayer) {
+                serverPlayer.deregisterEnderPearl(this);
+            }
+        }
+        // Canvas end - region threading
     }
 
     private void registerToCurrentOwner() {
-        // Folia - region threading - we remove the registration logic, we do not need to fetch the owner
+        // Canvas start - region threading
+        if (io.canvasmc.canvas.Config.INSTANCE.restoreVanillaEnderPearlBehavior) {
+            if (this.getOwner() instanceof ServerPlayer serverPlayer) {
+                serverPlayer.registerEnderPearl(this);
+            }
+        }
+        // Canvas end - region threading
     }
 
     @Nullable
