From 56a92b2c479a7161c340134ba931fb3d6a8e6f79 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/monster/zombie/ZombieVillager.java b/net/minecraft/world/entity/monster/zombie/ZombieVillager.java
index 816ab1327fa57d134fc20d23b6614a9999351c6c..4827d4111232d15cbd60f2efb24dbc0ba19e6161 100644
--- a/net/minecraft/world/entity/monster/zombie/ZombieVillager.java
+++ b/net/minecraft/world/entity/monster/zombie/ZombieVillager.java
@@ -286,6 +286,16 @@ public class ZombieVillager extends Zombie implements VillagerDataHolder {
                 if (!this.isSilent()) {
                     level.levelEvent(null, LevelEvent.SOUND_ZOMBIE_CONVERTED, this.blockPosition(), 0);
                 }
+                // Canvas start - MC-200418
+                if (
+                    io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaFixes.mc200418
+                        && villager.isPassenger()
+                        && villager.getVehicle() instanceof net.minecraft.world.entity.animal.chicken.Chicken
+                        && villager.isBaby())
+                {
+                    villager.removeVehicle();
+                }
+                // Canvas end - MC-200418
                 // CraftBukkit start
             }, org.bukkit.event.entity.EntityTransformEvent.TransformReason.CURED, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.CURED // CraftBukkit
         );
