--- a/net/minecraft/world/entity/EntityType.java
+++ b/net/minecraft/world/entity/EntityType.java
@@ -6,12 +_,14 @@
 import java.util.Optional;
 import java.util.Set;
 import java.util.Spliterator;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Consumer;
 import java.util.function.Function;
 import java.util.function.Supplier;
 import java.util.stream.Stream;
 import java.util.stream.StreamSupport;
 import javax.annotation.Nullable;
+import io.canvasmc.canvas.Config;
 import net.minecraft.Util;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
@@ -1365,6 +_,34 @@
     public float getHeight() {
         return this.dimensions.height();
     }
+    // Canvas start
+    private final AtomicReference<ResourceLocation> cached = new AtomicReference<>();
+    private final AtomicReference<Config.EntityMask> cachedMask = new AtomicReference<>();
+
+    public boolean isCachedResource() {
+        return cached.get() == null;
+    }
+
+    public void cacheResource(ResourceLocation location) {
+        cached.set(location);
+    }
+
+    public ResourceLocation cachedLocation() {
+        return cached.get();
+    }
+
+    public boolean isCachedMask() {
+        return cachedMask.get() == null;
+    }
+
+    public void cacheMask(Config.EntityMask mask) {
+        cachedMask.set(mask);
+    }
+
+    public Config.EntityMask getMask() {
+        return cachedMask.get();
+    }
+    // Canvas end
 
     @Override
     public FeatureFlagSet requiredFeatures() {
