--- a/net/minecraft/world/entity/EntityType.java
+++ b/net/minecraft/world/entity/EntityType.java
@@ -1225,9 +_,30 @@
     private final float spawnDimensionsScale;
     private final FeatureFlagSet requiredFeatures;
     private final boolean allowedInPeaceful;
+    public boolean canvas$isProjectileAndLoadsChunksOverride = false; // Canvas - entity type rules modification configurations
 
     private static <T extends Entity> EntityType<T> register(final ResourceKey<EntityType<?>> id, final EntityType.Builder<T> builder) {
-        return Registry.register(BuiltInRegistries.ENTITY_TYPE, id, builder.build(id));
+        // Canvas start - entity type rules modification configurations
+        // includes non tickable entities config
+        // includes projectile load chunks config
+        EntityType<T> type = builder.build(id);
+        final Identifier identifier = id.identifier();
+        if (type.getBaseClass().isAssignableFrom(net.minecraft.world.entity.projectile.Projectile.class)
+            && io.canvasmc.canvas.Config.INSTANCE.projectiles.loadChunks.stream().anyMatch(s -> {
+            String normalizedConfig = s.toLowerCase().replace(':', '_');
+            String normalizedLoc = identifier.toString().toLowerCase().replace(':', '_');
+            if (normalizedConfig.startsWith("minecraft_")) {
+                normalizedConfig = normalizedConfig.substring("minecraft_".length());
+            }
+            if (normalizedLoc.startsWith("minecraft_")) {
+                normalizedLoc = normalizedLoc.substring("minecraft_".length());
+            }
+            return normalizedConfig.equals(normalizedLoc);
+        })) {
+            type.canvas$isProjectileAndLoadsChunksOverride = true;
+        }
+        return Registry.register(BuiltInRegistries.ENTITY_TYPE, id, type);
+        // Canvas end - entity type rules modification configurations
     }
 
     private static ResourceKey<EntityType<?>> vanillaEntityId(final String vanillaId) {
