--- a/net/minecraft/world/entity/EntityType.java
+++ b/net/minecraft/world/entity/EntityType.java
@@ -1092,9 +_,30 @@
     public EntityDimensions dimensions;
     private final float spawnDimensionsScale;
     private final FeatureFlagSet requiredFeatures;
+    public boolean canvas$isProjectileAndLoadsChunksOverride = false; // Canvas - entity type rules modification configurations
 
     private static <T extends Entity> EntityType<T> register(ResourceKey<EntityType<?>> key, EntityType.Builder<T> builder) {
-        return Registry.register(BuiltInRegistries.ENTITY_TYPE, key, builder.build(key));
+        // Canvas start - entity type rules modification configurations
+        // includes non tickable entities config
+        // includes projectile load chunks config
+        EntityType<T> entityType = builder.build(key);
+        final ResourceLocation location = key.location();
+        if (entityType.getBaseClass().isAssignableFrom(net.minecraft.world.entity.projectile.Projectile.class)
+            && io.canvasmc.canvas.Config.INSTANCE.projectiles.loadChunks.stream().anyMatch(s -> {
+            String normalizedConfig = s.toLowerCase().replace(':', '_');
+            String normalizedLoc = location.toString().toLowerCase().replace(':', '_');
+            if (normalizedConfig.startsWith("minecraft_")) {
+                normalizedConfig = normalizedConfig.substring("minecraft_".length());
+            }
+            if (normalizedLoc.startsWith("minecraft_")) {
+                normalizedLoc = normalizedLoc.substring("minecraft_".length());
+            }
+            return normalizedConfig.equals(normalizedLoc);
+        })) {
+            entityType.canvas$isProjectileAndLoadsChunksOverride = true;
+        }
+        return Registry.register(BuiltInRegistries.ENTITY_TYPE, key, entityType);
+        // Canvas end - entity type rules modification configurations
     }
 
     private static ResourceKey<EntityType<?>> vanillaEntityId(String name) {
