--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -902,7 +_,7 @@
         }
 
         //if (this instanceof ServerPlayer) this.handlePortal(); // CraftBukkit - Moved up to postTick // Folia - region threading - ONLY allow in postTick()
-        if (this.canSpawnSprintParticle()) {
+        if (!level().canvasConfig().visuals.particles.disableSprintParticles && this.canSpawnSprintParticle()) { // Canvas - particles config
             this.spawnSprintParticle();
         }
 
@@ -937,7 +_,16 @@
 
         this.checkBelowWorld();
         if (!this.level().isClientSide()) {
+            // Canvas start - hide flames on entities with fire resistance or invis
+            if (this instanceof net.minecraft.world.entity.LivingEntity livingEntity) {
+                this.setSharedFlagOnFire(this.remainingFireTicks > 0 &&
+                    (!livingEntity.level().canvasConfig().visuals.hideFlamesOnEntitiesWithFireResistance || !livingEntity.hasEffect(net.minecraft.world.effect.MobEffects.FIRE_RESISTANCE)) &&
+                    (!livingEntity.level().canvasConfig().visuals.hideFlamesOnEntitiesWithInvisibility || !livingEntity.isInvisible())
+                );
+            } else {
+            // Canvas end - hide flames on entities with fire resistance or invis
             this.setSharedFlagOnFire(this.remainingFireTicks > 0);
+            } // Canvas - hide flames on entities with fire resistance or invis
         }
 
         this.firstTick = false;
@@ -1760,6 +_,11 @@
         boolean debugEntityBlockIntersections = this.level instanceof ServerLevel serverLevel
             && serverLevel.getServer().debugSubscribers().hasAnySubscriberFor(DebugSubscriptions.ENTITY_BLOCK_INTERSECTIONS);
         AtomicInteger iterations = new AtomicInteger();
+        // Canvas start - optimize checkInsideBlocks
+        // use element array for lambdas
+        final net.minecraft.world.level.chunk.ChunkAccess[] cachedChunk = new  net.minecraft.world.level.chunk.ChunkAccess[] { null };
+        final long[] lastChunkPos = { Long.MIN_VALUE };
+        // Canvas end - optimize checkInsideBlocks
         BlockGetter.forEachBlockIntersectedBetween(
             from,
             to,
@@ -1773,13 +_,25 @@
                     return false;
                 }
 
-                // Folia start - region threading
-                if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(Entity.this.level, blockIntersection, 16)) {
-                    return false;
-                }
-                // Folia end - region threading
+                // Canvas start - optimize checkInsideBlocks
+                // remove thread check here, as if the chunk is loaded, and we are NEXT TO IT, then it
+                // should always be inside the current region
                 iterations.set(iteration);
-                BlockState state = this.level().getBlockState(blockIntersection);
+                final int chunkX = blockIntersection.getX() >> 4;
+                final int chunkZ = blockIntersection.getZ() >> 4;
+                final long chunkLongPos = ((long) chunkZ << 32) | (chunkX & 0xFFFFFFFFL);
+                if (lastChunkPos[0] != chunkLongPos) {
+                    // update cache, this is a different chunk than previous
+                    lastChunkPos[0] = chunkLongPos;
+                    cachedChunk[0] = this.level.getChunkIfLoaded(chunkX, chunkZ);
+                }
+                net.minecraft.world.level.chunk.ChunkAccess chunk = cachedChunk[0];
+                if (chunk == null) {
+                    // return as if it were air
+                    return true;
+                }
+                BlockState state = chunk.getBlockState(blockIntersection);
+                // Canvas end - optimize checkInsideBlocks
                 if (state.isAir()) {
                     if (debugEntityBlockIntersections) {
                         this.debugBlockIntersection((ServerLevel)this.level(), blockIntersection.immutable(), false, false);
@@ -1790,7 +_,7 @@
                     VoxelShape intersectShape = state.getEntityInsideCollisionShape(this.level(), blockIntersection, this);
                     boolean insideBlock = intersectShape == Shapes.block()
                         || this.collidedWithShapeMovingFrom(from, to, intersectShape.move(new Vec3(blockIntersection)).toAabbs());
-                    boolean insideFluid = this.collidedWithFluid(state.getFluidState(), blockIntersection, from, to);
+                    boolean insideFluid = (state.tagFlag & io.canvasmc.canvas.world.block.BlockTags.FLUID) != 0 && this.collidedWithFluid(state.getFluidState(), blockIntersection, from, to); // Canvas - optimize checkInsideBlocks
                     if ((insideBlock || insideFluid) && visitedBlocks.add(blockIntersection.asLong())) {
                         if (insideBlock) {
                             try {
@@ -1841,6 +_,7 @@
     }
 
     public boolean collidedWithFluid(final FluidState fluidState, final BlockPos blockPos, final Vec3 from, final Vec3 to) {
+        if (fluidState.isEmpty()) return false; // Canvas - optimize checkInsideBlocks
         AABB fluidAABB = fluidState.getAABB(this.level(), blockPos);
         return fluidAABB != null && this.collidedWithShapeMovingFrom(from, to, List.of(fluidAABB));
     }
@@ -2127,6 +_,7 @@
 
         float yt = Mth.floor(this.getY());
 
+        if (!level().canvasConfig().visuals.particles.disableWaterSplashParticles) { // Canvas - particles config
         for (int i = 0; i < 1.0F + this.dimensions.width() * 20.0F; i++) {
             double xo = (this.random.nextDouble() * 2.0 - 1.0) * this.dimensions.width();
             double zo = (this.random.nextDouble() * 2.0 - 1.0) * this.dimensions.width();
@@ -2141,6 +_,7 @@
             double zo = (this.random.nextDouble() * 2.0 - 1.0) * this.dimensions.width();
             this.level().addParticle(ParticleTypes.SPLASH, this.getX() + xo, yt + 1.0F, this.getZ() + zo, movement.x, movement.y, movement.z);
         }
+        } // Canvas - particles config
 
         this.gameEvent(GameEvent.SPLASH);
     }
@@ -2321,8 +_,8 @@
     }
 
     public void push(final Entity entity) {
+        if (!entity.noPhysics && !this.noPhysics) { // Leaf - collision physics check before vehicle check
         if (!this.isPassengerOfSameVehicle(entity)) {
-            if (!entity.noPhysics && !this.noPhysics) {
                 if (this.level.paperConfig().collisions.onlyPlayersCollide && !(entity instanceof ServerPlayer || this instanceof ServerPlayer)) return; // Paper - Collision option for requiring a player participant
                 double xa = entity.getX() - this.getX();
                 double za = entity.getZ() - this.getZ();
@@ -3839,6 +_,7 @@
     }
 
     protected static void sendBubbleColumnParticles(final Level level, final BlockPos pos) {
+        if (level.canvasConfig().visuals.particles.disableBubbleColumnParticles) return; // Canvas - particles config
         if (level instanceof ServerLevel serverLevel) {
             RandomSource random = level.getRandom();
 
@@ -4617,7 +_,7 @@
     public boolean endPortalLogicAsync(BlockPos portalPos) {
         ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this, "Cannot portal entity async");
 
-        ServerLevel destination = this.level().getServer().getLevel(this.level().getTypeKey() == net.minecraft.world.level.dimension.LevelStem.END ? Level.OVERWORLD : Level.END);
+        ServerLevel destination = this.level().getServer().getLevel(this.level().getTypeKey() == net.minecraft.world.level.dimension.LevelStem.END ? io.canvasmc.canvas.GlobalConfiguration.fetchRespawnDimensionKey() : Level.END); // Canvas - default respawn dimension config
         if (destination == null) {
             // wat
             return false;
@@ -4629,7 +_,7 @@
     public boolean netherPortalLogicAsync(BlockPos portalPos) {
         ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this, "Cannot portal entity async");
 
-        ServerLevel destination = this.level().getServer().getLevel(this.level().getTypeKey() == net.minecraft.world.level.dimension.LevelStem.NETHER ? Level.OVERWORLD : Level.NETHER);
+        ServerLevel destination = this.level().getServer().getLevel(this.level().getTypeKey() == net.minecraft.world.level.dimension.LevelStem.NETHER ? io.canvasmc.canvas.GlobalConfiguration.fetchRespawnDimensionKey() : Level.NETHER); // Canvas - default respawn dimension config
         if (destination == null) {
             // wat
             return false;
@@ -5237,7 +_,7 @@
 
     public boolean canTeleport(final Level from, final Level to) {
         if (!this.isAlive() || !this.valid) return false; // Paper - Fix item duplication and teleport issues
-        if (from.dimension() == Level.END && to.dimension() == Level.OVERWORLD) {
+        if (from.dimension() == Level.END && to.dimension() == io.canvasmc.canvas.GlobalConfiguration.fetchRespawnDimensionKey()) { // Canvas - default respawn dimension config
             for (Entity passenger : this.getPassengers()) {
                 if (passenger instanceof ServerPlayer player && !player.seenCredits) {
                     return false;
