From 2be0fbf099df19422462cdb5b961fc00062d256b Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/EntityType.java b/net/minecraft/world/entity/EntityType.java
index 76143f9e9997e5d48348cb85d9ee6f57a966f5ea..761bbd35f7cbe39d8e77e7b774f7a78821d8a25a 100644
--- a/net/minecraft/world/entity/EntityType.java
+++ b/net/minecraft/world/entity/EntityType.java
@@ -1225,9 +1225,30 @@ public class EntityType<T extends Entity> implements EntityTypeTest<Entity, T>,
     private final float spawnDimensionsScale;
     private final FeatureFlagSet requiredFeatures;
     private final boolean allowedInPeaceful;
+    public boolean canvas$isProjectileAndLoadsChunksOverride = false; // Canvas - entity type rules modification configurations
 
     private static <T extends Entity> EntityType<T> register(final ResourceKey<EntityType<?>> id, final EntityType.Builder<T> builder) {
-        return Registry.register(BuiltInRegistries.ENTITY_TYPE, id, builder.build(id));
+        // Canvas start - entity type rules modification configurations
+        // includes non tickable entities config
+        // includes projectile load chunks config
+        EntityType<T> entityType = builder.build(id);
+        final Identifier location = id.identifier();
+        if (entityType.getBaseClass().isAssignableFrom(net.minecraft.world.entity.projectile.Projectile.class)
+            && io.canvasmc.canvas.Config.INSTANCE.projectiles.loadChunks.stream().anyMatch(s -> {
+            String normalizedConfig = s.toLowerCase().replace(':', '_');
+            String normalizedLoc = location.toString().toLowerCase().replace(':', '_');
+            if (normalizedConfig.startsWith("minecraft_")) {
+                normalizedConfig = normalizedConfig.substring("minecraft_".length());
+            }
+            if (normalizedLoc.startsWith("minecraft_")) {
+                normalizedLoc = normalizedLoc.substring("minecraft_".length());
+            }
+            return normalizedConfig.equals(normalizedLoc);
+        })) {
+            entityType.canvas$isProjectileAndLoadsChunksOverride = true;
+        }
+        return Registry.register(BuiltInRegistries.ENTITY_TYPE, key, entityType);
+        // Canvas end - entity type rules modification configurations
     }
 
     private static ResourceKey<EntityType<?>> vanillaEntityId(final String vanillaId) {
