--- a/net/minecraft/world/entity/ExperienceOrb.java
+++ b/net/minecraft/world/entity/ExperienceOrb.java
@@ -114,6 +_,13 @@
 
         this.setValue(value);
     }
+    // Canvas start - fire res orbs
+
+    @Override
+    public boolean fireImmune() {
+        return level().canvasConfig().entities.experienceOrbsAreFireResistant;
+    }
+    // Canvas end - fire res orbs
 
     public ExperienceOrb(final EntityType<? extends ExperienceOrb> type, final Level level) {
         super(type, level);
@@ -246,8 +_,8 @@
     }
 
     private void scanForMerges() {
-        if (this.level() instanceof ServerLevel) {
-            for (ExperienceOrb orb : this.level().getEntities(EntityTypeTest.forClass(ExperienceOrb.class), this.getBoundingBox().inflate(0.5), this::canMerge)) {
+        if (this.level() instanceof ServerLevel && (!this.level().canvasConfig().entities.fastOrbs || this.tickCount % 2 == 0)) { // Canvas - fast orbs
+            for (ExperienceOrb orb : this.level().getEntities(EntityTypeTest.forClass(ExperienceOrb.class), this.getBoundingBox().inflate(level().canvasConfig().entities.fastOrbs ? 0.9 : 0.5), this::canMerge)) { // Canvas - fast orbs
                 this.merge(orb);
             }
         }
@@ -263,6 +_,24 @@
     }
     public static void awardWithDirection(final ServerLevel level, final Vec3 pos, final Vec3 roughDirection, int amount, org.bukkit.entity.ExperienceOrb.@Nullable SpawnReason reason, final @Nullable Entity triggerId, final @Nullable Entity sourceId) {
     // Paper end - add reason to orbs
+        // Canvas start - fast orbs
+        if (level.canvasConfig().entities.fastOrbs) {
+            if (amount <= 0) return; // skip if count is 0
+            int experienceValue = getExperienceValue(amount);
+            AABB aabb = AABB.ofSize(pos, 1.0, 1.0, 1.0);
+            List<ExperienceOrb> entities = level.getEntities(
+                EntityTypeTest.forClass(ExperienceOrb.class), aabb, orb -> canMerge(orb, -1, -1) // we are running fast orbs so it just checks if it's alive, the 2 int vals don't matter
+            );
+            if (!entities.isEmpty()) {
+                ExperienceOrb orb = entities.getFirst();
+                orb.age = 0;
+                orb.setValue(orb.getValue() + experienceValue);
+            } else {
+                level.addFreshEntity(new ExperienceOrb(level, pos, roughDirection, experienceValue, reason, triggerId, sourceId));
+            }
+            return;
+        }
+        // Canvas end - fast orbs
         while (amount > 0) {
             int newCount = getExperienceValue(amount);
             amount -= newCount;
@@ -288,10 +_,12 @@
     }
 
     private boolean canMerge(final ExperienceOrb orb) {
+        if (level().canvasConfig().entities.fastOrbs) return orb.isAlive() && !this.is(orb); // Canvas - fast orbs
         return orb != this && canMerge(orb, this.getId(), this.getValue());
     }
 
     private static boolean canMerge(final ExperienceOrb orb, final int id, final int value) {
+        if (orb.level().canvasConfig().entities.fastOrbs) return orb.isAlive(); // Canvas - fast orbs
         return !orb.isRemoved() && (orb.getId() - id) % io.papermc.paper.configuration.GlobalConfiguration.get().misc.xpOrbGroupsPerArea.or(ORB_GROUPS_PER_AREA) == 0 && orb.getValue() == value; // Paper - Configure how many orbs will merge together
     }
 
@@ -301,6 +_,14 @@
             return;
         }
         // Paper end - call orb merge event
+        // Canvas start - fast orbs
+        if (level().canvasConfig().entities.fastOrbs) {
+            this.setValue(this.getValue() + orb.getValue());
+            this.age = Math.min(this.age, orb.age);
+            orb.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.MERGE); // CraftBukkit - add Bukkit remove cause
+            return;
+        }
+        // Canvas end - fast orbs
         this.count = this.count + orb.count;
         this.age = Math.min(this.age, orb.age);
         orb.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.MERGE); // CraftBukkit - add Bukkit remove cause
@@ -357,7 +_,7 @@
     public void playerTouch(final Player player) {
         if (player instanceof ServerPlayer serverPlayer) {
             if (player.takeXpDelay == 0 && new com.destroystokyo.paper.event.player.PlayerPickupExperienceEvent(serverPlayer.getBukkitEntity(), (org.bukkit.entity.ExperienceOrb) this.getBukkitEntity()).callEvent()) { // Paper - PlayerPickupExperienceEvent
-                player.takeXpDelay = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerXpCooldownEvent(player, 2, org.bukkit.event.player.PlayerExpCooldownChangeEvent.ChangeReason.PICKUP_ORB).getNewCooldown(); // CraftBukkit - entity.takeXpDelay = 2;
+                player.takeXpDelay = org.bukkit.craftbukkit.event.CraftEventFactory.callPlayerXpCooldownEvent(player, level().canvasConfig().entities.fastOrbs ? 0 : 2, org.bukkit.event.player.PlayerExpCooldownChangeEvent.ChangeReason.PICKUP_ORB).getNewCooldown(); // CraftBukkit - entity.takeXpDelay = 2; // Canvas - fast orbs
                 player.take(this, 1);
                 int remaining = this.repairPlayerItems(serverPlayer, this.getValue());
                 if (remaining > 0) {
