From 562a609f7063b6400aca83fa84572c4902aa25af Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 20b043cea20570fd32bc9bcfa09c09275a3df068..88a024b5f27df43020df808285b3ce2c28ca4aaa 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -180,7 +180,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     private static final int FREE_FALL_EVENTS_PER_ELYTRA_BREAK = 2;
     public static final float BASE_JUMP_POWER = 0.42F;
     protected static final float DEFAULT_KNOCKBACK = 0.4F;
-    protected static final int INVULNERABLE_DURATION = 20;
+    protected static final int INVULNERABLE_DURATION = io.canvasmc.canvas.Config.INSTANCE.combat.invulnerabilityTicks; // Canvas - combat configs
     private static final double MAX_LINE_OF_SIGHT_TEST_RANGE = 128.0;
     protected static final int LIVING_ENTITY_FLAG_IS_USING = 1;
     protected static final int LIVING_ENTITY_FLAG_OFF_HAND = 2;
@@ -283,6 +283,8 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     public List<DefaultDrop> drops = new java.util.ArrayList<>(); // Paper - Restore vanilla drops behavior
     public final org.bukkit.craftbukkit.attribute.CraftAttributeMap craftAttributes;
     public boolean collides = true;
+    public boolean canvas$fromSpawner = false; // Canvas - spawner configs
+    protected boolean canvas$isDyingFromWitherRose = false; // Canvas - fix Folia#388
     public Set<UUID> collidableExemptions = new java.util.HashSet<>();
     public boolean bukkitPickUpLoot;
     public org.bukkit.craftbukkit.entity.CraftLivingEntity getBukkitLivingEntity() { return (org.bukkit.craftbukkit.entity.CraftLivingEntity) super.getBukkitEntity(); } // Paper
@@ -393,7 +395,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         if (this.level() instanceof ServerLevel level && onGround && this.fallDistance > 0.0) {
             this.onChangedBlock(level, pos);
             double power = Math.max(0, Mth.floor(this.calculateFallPower(this.fallDistance)));
-            if (power > 0.0 && !onState.isAir()) {
+            if (power > 0.0 && !(onState.isAir() || (io.canvasmc.canvas.Config.INSTANCE.fixes.mc30391 && (this instanceof net.minecraft.world.entity.animal.chicken.Chicken || this instanceof net.minecraft.world.entity.monster.Blaze || this instanceof WitherBoss)))) { // Canvas - MC-30391
                 double x = this.getX();
                 double y = this.getY();
                 double z = this.getZ();
@@ -409,11 +411,13 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
                 double scale = Math.min(0.2F + power / 15.0, 2.5);
                 int particles = (int)(150.0 * scale);
                 // CraftBukkit start - visibility api
+                if (!io.canvasmc.canvas.Config.INSTANCE.particles.disableFallParticles) { // Canvas - particles config
                 if (this instanceof ServerPlayer) {
                     level.sendParticlesSource((ServerPlayer)this, new BlockParticleOption(ParticleTypes.BLOCK, onState), false, false, x, y, z, particles, 0.0, 0.0, 0.0, 0.15F);
                 } else {
                 level.sendParticles(new BlockParticleOption(ParticleTypes.BLOCK, onState), x, y, z, particles, 0.0, 0.0, 0.0, 0.15F);
                 }
+                } // Canvas - particles config
                 // CraftBukkit end
             }
         }
@@ -450,7 +454,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         super.baseTick();
         if (this.isAlive() && this.level() instanceof ServerLevel level) {
             boolean isPlayer = this instanceof Player;
-            if (this.isInWall()) {
+            if (shouldCheckForSuffocation() && this.isInWall()) { // Pufferfish - optimize suffocation
                 this.hurtServer(level, this.damageSources().inWall(), 1.0F);
             } else if (isPlayer && !level.getWorldBorder().isWithinBounds(this.getBoundingBox())) {
                 double dist = level.getWorldBorder().getDistanceToBorder(this) + level.getWorldBorder().getSafeZone();
@@ -529,6 +533,19 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         this.yRotO = this.getYRot();
         this.xRotO = this.getXRot();
     }
+    // Pufferfish start - optimize suffocation
+
+    public boolean couldPossiblyBeHurt(float amount) {
+        if ((float) this.invulnerableTime > (float) this.invulnerableDuration / 2.0F && amount <= this.lastHurt) {
+            return false;
+        }
+        return true;
+    }
+
+    public boolean shouldCheckForSuffocation() {
+        return !io.canvasmc.canvas.Config.INSTANCE.enableSuffocationOptimization || (tickCount % 10 == 0 && couldPossiblyBeHurt(1.0F));
+    }
+    // Pufferfish end
 
     protected boolean shouldTakeDrowningDamage() {
         return this.getAirSupply() <= -20;
@@ -553,9 +570,9 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     }
 
     protected void tryAddFrost() {
-        if (!this.getBlockStateOnLegacy().isAir()) {
+        if (true) { // Canvas - check frozen ticks before landing block
             int ticksFrozen = this.getTicksFrozen();
-            if (ticksFrozen > 0) {
+            if (ticksFrozen > 0 && !this.getBlockStateOnLegacy().isAir()) { // Canvas - check frozen ticks before landing block
                 AttributeInstance speed = this.getAttribute(Attributes.MOVEMENT_SPEED);
                 if (speed == null) {
                     return;
@@ -595,6 +612,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     public boolean broadcastedDeath = false; // Folia - region threading
     protected void tickDeath() {
         this.deathTime++;
+        if (io.canvasmc.canvas.Config.INSTANCE.combat.removeRedDeathAnimation) this.deathTime = 20; // Canvas - remove red death animation
         if (this.deathTime >= 20 && !this.level().isClientSide() && !this.isRemoved()) {
             this.level().broadcastEntityEvent(this, EntityEvent.POOF); this.broadcastedDeath = true; // Folia - region threading - death has been broadcasted
             if (!(this instanceof ServerPlayer)) this.remove(Entity.RemovalReason.KILLED, org.bukkit.event.entity.EntityRemoveEvent.Cause.DEATH); // CraftBukkit - add Bukkit remove cause // Folia - region threading - don't remove, we want the tick scheduler to be running
@@ -754,6 +772,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
 
         super.remove(reason, eventCause); // CraftBukkit
         this.brain.clearMemories();
+        if (io.canvasmc.canvas.Config.INSTANCE.fixes.mc298464) this.combatTracker.resetCombatState(); // Canvas - fix MC-298464
     }
 
     @Override
@@ -812,6 +831,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         if (this.locatorBarIcon.hasData()) {
             output.store("locator_bar_icon", Waypoint.Icon.CODEC, this.locatorBarIcon);
         }
+        output.store("from_spawner", com.mojang.serialization.Codec.BOOL, this.canvas$fromSpawner); // Canvas - spawner configs
     }
 
     // Paper start - Extend dropItem API
@@ -927,6 +947,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         this.locatorBarIcon = input.read("locator_bar_icon", Waypoint.Icon.CODEC).orElseGet(Waypoint.Icon::new);
         this.currentImpulseContextResetGraceTime = input.getIntOr("current_impulse_context_reset_grace_time", 0);
         this.currentImpulseImpactPos = input.read("current_explosion_impact_pos", Vec3.CODEC).orElse(null);
+        this.canvas$fromSpawner = input.read("from_spawner", com.mojang.serialization.Codec.BOOL).orElse(this.canvas$fromSpawner); // Canvas - spawner configs
     }
 
     @Override
@@ -999,7 +1020,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
             }
 
             List<ParticleOptions> particles = this.entityData.get(DATA_EFFECT_PARTICLES);
-            if (!particles.isEmpty()) {
+            if (!particles.isEmpty() && !io.canvasmc.canvas.Config.INSTANCE.particles.disableEffectParticles) { // Canvas - particles config
                 boolean isAmbient = this.entityData.get(DATA_EFFECT_AMBIENCE_ID);
                 int bound = this.isInvisible() ? 15 : 4;
                 int ambientFactor = isAmbient ? 5 : 1;
@@ -1030,14 +1051,18 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     }
 
     private void updateSynchronizedMobEffectParticles() {
-        List<ParticleOptions> visibleEffectParticles = this.activeEffects
-            .values()
-            .stream()
-            .filter(MobEffectInstance::isVisible)
-            .map(MobEffectInstance::getParticleOptions)
-            .toList();
+        // Canvas start - remove streams
+        List<ParticleOptions> visibleEffectParticles = new java.util.ArrayList<>();
+        final Collection<MobEffectInstance> effectsValues = this.activeEffects.values(); // Canvas - optimized collection
+
+        for (MobEffectInstance effect : effectsValues) { // Canvas - optimized collection
+            if (effect.isVisible()) {
+                visibleEffectParticles.add(effect.getParticleOptions());
+            }
+        }
+        // Canvas end - remove streams
         this.entityData.set(DATA_EFFECT_PARTICLES, visibleEffectParticles);
-        this.entityData.set(DATA_EFFECT_AMBIENCE_ID, areAllEffectsAmbient(this.activeEffects.values()));
+        this.entityData.set(DATA_EFFECT_AMBIENCE_ID, areAllEffectsAmbient(effectsValues)); // Canvas - optimized collection
     }
 
     private void updateGlowingStatus() {
@@ -1477,7 +1502,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
             }
 
             boolean tookFullDamage = true;
-            if (this.invulnerableTime > (float)this.invulnerableDuration / 2.0F && !source.is(DamageTypeTags.BYPASSES_COOLDOWN)) { // CraftBukkit - restore use of maxNoDamageTicks
+            if (this.invulnerableTime > 0.0F && !source.is(DamageTypeTags.BYPASSES_COOLDOWN)) { // CraftBukkit - restore use of maxNoDamageTicks // Canvas - combat configs
                 if (damage <= this.lastHurt) {
                     return false;
                 }
@@ -1506,10 +1531,10 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
                 }
                 if (this instanceof ServerPlayer && event.getDamage() == 0 && originalDamage == 0) return false; // Paper - revert to vanilla damage - players are not affected by damage that is 0 - skip damage if the vanilla damage is 0 and was not modified by plugins in the event.
                 this.lastHurt = damage;
-                this.invulnerableTime = this.invulnerableDuration; // CraftBukkit - restore use of maxNoDamageTicks
+                this.invulnerableTime = io.canvasmc.canvas.Config.INSTANCE.combat.invulnerabilityTicks; // CraftBukkit - restore use of maxNoDamageTicks // Canvas - combat configs
                 // this.actuallyHurt(level, source, damage);
                 // CraftBukkit end
-                this.hurtDuration = 10;
+                this.hurtDuration = io.canvasmc.canvas.Config.INSTANCE.combat.invulnerabilityTicks; // Canvas - combat configs
                 this.hurtTime = this.hurtDuration;
             }
 
@@ -1892,6 +1917,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
                     }
                 }
                 // Paper end
+                if (source.is(DamageTypes.WITHER_SKULL)) this.canvas$isDyingFromWitherRose = true; // Canvas - fix Folia#388
                 this.createWitherRose(killer);
                 }
 
@@ -1908,7 +1934,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     protected void createWitherRose(final @Nullable LivingEntity killer) {
         if (this.level() instanceof ServerLevel serverLevel) {
             boolean var6 = false;
-            if (this.dead && killer instanceof WitherBoss) { // Paper
+            if (this.dead && (killer instanceof WitherBoss || this.canvas$isDyingFromWitherRose)) { // Paper // Canvas - fix Folia#388
                 if (serverLevel.getGameRules().get(GameRules.MOB_GRIEFING)) {
                     BlockPos pos = this.blockPosition();
                     BlockState state = Blocks.WITHER_ROSE.defaultBlockState();
@@ -2219,7 +2245,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
 
     @Override
     public boolean isAlive() {
-        return !this.isRemoved() && this.getHealth() > 0.0F && !this.dead; // Paper - Check this.dead
+        return !this.dead && !this.isRemoved() && this.getHealth() > 0.0F; // Paper - Check this.dead // Canvas - cheapest check first
     }
 
     public boolean isLookingAtMe(
@@ -2350,7 +2376,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
 
     @Override
     public void animateHurt(final float yaw) {
-        this.hurtDuration = 10;
+        this.hurtDuration = io.canvasmc.canvas.Config.INSTANCE.combat.invulnerabilityTicks; // Canvas - combat configs
         this.hurtTime = this.hurtDuration;
     }
 
@@ -2707,7 +2733,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     public void handleDamageEvent(final DamageSource source) {
         this.walkAnimation.setSpeed(1.5F);
         this.invulnerableTime = this.invulnerableDuration; // Paper - configurable invulnerable duration
-        this.hurtDuration = 10;
+        this.hurtDuration = io.canvasmc.canvas.Config.INSTANCE.combat.invulnerabilityTicks; // Canvas - combat configs
         this.hurtTime = this.hurtDuration;
         SoundEvent hurtSound = this.getHurtSound(source);
         if (hurtSound != null) {
@@ -2774,7 +2800,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
                 this.swapHandItems();
                 break;
             case 60:
-                this.makePoofParticles();
+                if (!io.canvasmc.canvas.Config.INSTANCE.particles.disableDeathParticles) this.makePoofParticles(); // Canvas - particles config
                 break;
             case 65:
                 this.breakItem(this.getItemBySlot(EquipmentSlot.BODY));
@@ -2839,6 +2865,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     }
 
     protected void updateSwingTime() {
+        if (!this.swinging && this.swingTime == 0) return; // Canvas - skip unnecessary calculations if player is not flying or swing
         int currentSwingDuration = this.getCurrentSwingDuration();
         if (this.swinging) {
             this.swingTime++;
@@ -3178,7 +3205,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
             waterWalker *= 0.5F;
         }
 
-        if (waterWalker > 0.0F) {
+        if (waterWalker > 0.0F && (!io.canvasmc.canvas.Config.INSTANCE.fixes.mc136249 || autoSpinAttackTicks == 0)) { // Canvas - MC-136249
             slowDown += (0.54600006F - slowDown) * waterWalker;
             speed += (this.getSpeed() - speed) * waterWalker;
         }
@@ -3803,7 +3830,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
             this.checkAutoSpinAttack(beforeTravelBox, this.getBoundingBox());
         }
 
-        this.pushEntities();
+        if (io.canvasmc.canvas.Config.INSTANCE.entityCollisionMode.allEntitiesCanBePushed()) this.pushEntities(); // Canvas - entity collision mode config
         // Paper start - Add EntityMoveEvent
         if (((ServerLevel) this.level()).getCurrentWorldData().hasEntityMoveEvent && !(this instanceof Player)) { // Folia - region threading
             if (this.xo != this.getX() || this.yo != this.getY() || this.zo != this.getZ() || this.yRotO != this.getYRot() || this.xRotO != this.getXRot()) {
@@ -3839,6 +3866,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     protected void updateFallFlying() {
         this.checkFallDistanceAccumulation();
         if (!this.level().isClientSide()) {
+            if (!this.isFallFlying() && this.fallFlyTicks == 0) return; // Canvas - skip unnecessary calculations if player is not flying
             if (!this.canGlide()) {
                 if (this.getSharedFlag(Entity.FLAG_FALL_FLYING) && !CraftEventFactory.callToggleGlideEvent(this, false).isCancelled()) // CraftBukkit
                 this.setSharedFlag(Entity.FLAG_FALL_FLYING, false);
@@ -3877,6 +3905,12 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     }
 
     protected void pushEntities() {
+        // Canvas start - only players push config
+        io.canvasmc.canvas.world.entity.EntityCollisionMode collisionMode = io.canvasmc.canvas.Config.INSTANCE.entityCollisionMode;
+        if (collisionMode.noCollisions()) {
+            return;
+        }
+        // Canvas end - only players push config
         // Paper start - don't run getEntities if we're not going to use its result
         if (!this.isPushable()) {
             return;
@@ -3892,7 +3926,11 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
             return;
         }
         // Paper end - don't run getEntities if we're not going to use its result
-        List<Entity> pushableEntities = this.level().getPushableEntities(this, this.getBoundingBox());
+        // Canvas start - only players push config
+        List<? extends Entity> pushableEntities = collisionMode.onlyPlayersPushable() ?
+            this.level().canvas$getIntersectingPlayers(this, this.getBoundingBox(), EntitySelector.pushableBy(this))
+            : this.level().getPushableEntities(this, this.getBoundingBox());
+        // Canvas end - only players push config
         if (!pushableEntities.isEmpty()) {
             if (this.level() instanceof ServerLevel serverLevel) {
                 // Paper - don't run getEntities if we're not going to use its result; moved up
@@ -4068,6 +4106,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
 
     @Override
     public boolean isCollidable(boolean ignoreClimbing) {
+        if (this.canvas$fromSpawner && io.canvasmc.canvas.Config.INSTANCE.spawner.noCollisions) return false; // Canvas - spawner configs
         return this.isAlive() && !this.isSpectator() && (ignoreClimbing || !this.onClimbable()) && this.collides; // CraftBukkit
         // Paper end - Climbing should not bypass cramming gamerule
     }
@@ -4185,6 +4224,18 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
                 );
             }
 
+            // Canvas start - item entity spread config
+            if (this.isDeadOrDying()) {
+                float halfedRandom = this.random.nextFloat() * 0.5F;
+                float piRandom = this.random.nextFloat() * (float) (Math.PI * 2);
+                double factor = io.canvasmc.canvas.Config.INSTANCE.itemEntitySpreadFactor;
+                entity.setDeltaMovement(
+                    (-Mth.sin(piRandom) * halfedRandom) * factor,
+                    0.2F * factor,
+                    (Mth.cos(piRandom) * halfedRandom) * factor
+                );
+            }
+            // Canvas end - item entity spread config
             return entity;
         }
     }
