--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -273,7 +_,7 @@
     }
 
     public boolean setTarget(@Nullable LivingEntity target, @Nullable EntityTargetEvent.TargetReason reason) {
-        if (this.getTargetRaw() == target) { // Folia - region threading
+        if (this.getTargetRaw() == target || !ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(target)) { // Folia - region threading // Canvas - Folia#386 - don't set mob target to entity off owning region
             return false;
         }
         if (reason != null) {
@@ -337,6 +_,7 @@
             this.resetAmbientSoundTime();
             this.playAmbientSound();
         }
+        if (io.canvasmc.canvas.Config.INSTANCE.fixes.mc183990 && this.target != null && this.target.isDeadOrDying()) this.setTarget(null, org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_DIED); // Canvas - MC-183990
     }
 
     @Override
