--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -358,6 +_,7 @@
             this.resetAmbientSoundTime();
             this.playAmbientSound();
         }
+        if (io.canvasmc.canvas.Config.INSTANCE.fixes.mc183990 && this.target != null && this.target.isDeadOrDying()) this.setTarget(null, org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_DIED); // Canvas - MC-183990
     }
 
     @Override
