--- a/net/minecraft/world/entity/PathfinderMob.java
+++ b/net/minecraft/world/entity/PathfinderMob.java
@@ -1,5 +_,6 @@
 package net.minecraft.world.entity;
 
+import io.canvasmc.canvas.Config;
 import net.minecraft.core.BlockPos;
 import net.minecraft.world.entity.ai.goal.Goal;
 import net.minecraft.world.entity.ai.goal.PanicGoal;
@@ -9,6 +_,7 @@
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.LevelReader;
 import net.minecraft.world.phys.Vec3;
+import java.util.List;
 
 public abstract class PathfinderMob extends Mob {
     protected static final float DEFAULT_WALK_TARGET_VALUE = 0.0F;
@@ -35,6 +_,13 @@
     public boolean isPathFinding() {
         return !this.getNavigation().isDone();
     }
+    // Canvas start - reduce movement when crammed
+
+    public boolean canTickMoveOptimized() {
+        List<Entity> entityList = this.level().getEntities(this, this.getBoundingBox(), entity -> !entity.isRemoved() && entity != this);
+        return entityList.isEmpty() || entityList.size() < Config.INSTANCE.entities.cramming.crammedThreshold;
+    }
+    // Canvas end
 
     public boolean isPanicking() {
         if (this.brain.hasMemoryValue(MemoryModuleType.IS_PANICKING)) {
