--- a/net/minecraft/world/entity/SpawnPlacements.java
+++ b/net/minecraft/world/entity/SpawnPlacements.java
@@ -45,7 +_,7 @@
 import net.minecraft.world.level.levelgen.Heightmap;
 
 public class SpawnPlacements {
-    private static final Map<EntityType<?>, SpawnPlacements.Data> DATA_BY_TYPE = Maps.newHashMap();
+    private static final Map<EntityType<?>, SpawnPlacements.Data> DATA_BY_TYPE = Maps.newConcurrentMap(); // Canvas - concurrent
 
     private static <T extends Mob> void register(
         EntityType<T> entityType, SpawnPlacementType spawnPlacementType, Heightmap.Types heightmapType, SpawnPlacements.SpawnPredicate<T> predicate
