--- a/net/minecraft/world/entity/TamableAnimal.java
+++ b/net/minecraft/world/entity/TamableAnimal.java
@@ -310,11 +_,17 @@
     }
 
     private boolean canTeleportTo(BlockPos pos) {
+        // Canvas start - region threading
+        net.minecraft.world.level.chunk.LevelChunk chunk  = this.level().getChunkIfLoaded(pos);
+        if (chunk == null) {
+            return false;
+        }
+        // Canvas end - region threading
         PathType pathTypeStatic = WalkNodeEvaluator.getPathTypeStatic(this, pos);
         if (pathTypeStatic != PathType.WALKABLE) {
             return false;
         } else {
-            BlockState blockState = this.level().getBlockState(pos.below());
+            BlockState blockState = chunk.getBlockState(pos.below()); // Canvas - region threading
             if (!this.canFlyToOwner() && blockState.getBlock() instanceof LeavesBlock) {
                 return false;
             } else {
