--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
@@ -253,7 +_,12 @@
         // Folia start - region threading
         this.close(save, halt, true, true, true);
     }
+    // Canvas start - add 'flush' arg
     public void close(final boolean save, final boolean halt, final boolean first, final boolean last, final boolean checkRegions) {
+        this.close(save, halt, first, last, checkRegions, true);
+    }
+    public void close(final boolean save, final boolean halt, final boolean first, final boolean last, final boolean checkRegions, final boolean flush) {
+    // Canvas end - add 'flush' arg
         // Folia end - region threading
         TickThread.ensureTickThread("Closing world off-main");
         if (first && halt) { // Folia - region threading
@@ -266,7 +_,7 @@
         }
 
         if (save) {
-            this.saveAllChunksRegionised(true, true, true, first, last, checkRegions); // Folia - region threading
+            this.saveAllChunksRegionised(flush, true, true, first, last, checkRegions); // Folia - region threading // Canvas - add 'flush' arg
         }
 
         if (last) { // Folia - region threading
