From 6321679ed132e14ba66d57ae4ff74f98abc436b1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/util/context/ContextMap.java b/net/minecraft/util/context/ContextMap.java
index 439624ea6d329705a15be4c9ae2f90ef6d78bd70..c191d2db9771e8ff1011e2b5c324e6886f2fd78d 100644
--- a/net/minecraft/util/context/ContextMap.java
+++ b/net/minecraft/util/context/ContextMap.java
@@ -69,17 +69,35 @@ public class ContextMap {
         }
 
         public ContextMap create(final ContextKeySet paramSet) {
-            Set<ContextKey<?>> notAllowed = Sets.difference(this.params.keySet(), paramSet.allowed());
-            if (!notAllowed.isEmpty()) {
-                throw new IllegalArgumentException("Parameters not allowed in this parameter set: " + notAllowed);
-            } else {
-                Set<ContextKey<?>> missingRequired = Sets.difference(paramSet.required(), this.params.keySet());
-                if (!missingRequired.isEmpty()) {
-                    throw new IllegalArgumentException("Missing required parameters: " + missingRequired);
-                } else {
-                    return new ContextMap(this.params);
+            // Leaf start - optimize ContextMap#create
+            final Set<ContextKey<?>> allowed = paramSet.allowed();
+            final Set<ContextKey<?>> required = paramSet.required();
+            final Set<ContextKey<?>> params = this.params.keySet();
+
+            ContextKey<?> invalid = null;
+            for (ContextKey<?> key : params) {
+                if (!allowed.contains(key)) {
+                    invalid = key;
+                    break;
+                }
+            }
+            if (invalid != null) {
+                throw new IllegalArgumentException("Parameter not allowed in this parameter set: " + invalid);
+            }
+
+            ContextKey<?> missing = null;
+            for (ContextKey<?> key : required) {
+                if (!this.params.containsKey(key)) {
+                    missing = key;
+                    break;
                 }
             }
+            if (missing != null) {
+                throw new IllegalArgumentException("Missing required parameter: " + missing);
+            }
+
+            return new ContextMap(this.params);
+            // Leaf end - optimize ContextMap#create
         }
     }
 }
