--- a/net/minecraft/util/context/ContextKeySet.java
+++ b/net/minecraft/util/context/ContextKeySet.java
@@ -9,8 +_,8 @@
     private final Set<ContextKey<?>> allowed;
 
     ContextKeySet(Set<ContextKey<?>> required, Set<ContextKey<?>> allowed) {
-        this.required = Set.copyOf(required);
-        this.allowed = Set.copyOf(Sets.union(required, allowed));
+        this.required = new it.unimi.dsi.fastutil.objects.ObjectOpenHashSet<>(required); // Canvas - optimize ContextMap#create
+        this.allowed = new it.unimi.dsi.fastutil.objects.ObjectOpenHashSet<>(Sets.union(required, allowed)); // Canvas - optimize ContextMap#create
     }
 
     public Set<ContextKey<?>> required() {
