From cd913ee721ba4988f5421e570c30420c95c78467 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/util/context/ContextKeySet.java b/net/minecraft/util/context/ContextKeySet.java
index 7b0c6e7b1dfbf8eec3150cfeb5e80b904c784f79..99b02361ce91df35ab21d53795db171fd1bafd95 100644
--- a/net/minecraft/util/context/ContextKeySet.java
+++ b/net/minecraft/util/context/ContextKeySet.java
@@ -9,8 +9,8 @@ public class ContextKeySet {
     private final Set<ContextKey<?>> allowed;
 
     private ContextKeySet(final Set<ContextKey<?>> required, final Set<ContextKey<?>> optional) {
-        this.required = Set.copyOf(required);
-        this.allowed = Set.copyOf(Sets.union(required, optional));
+        this.required = new it.unimi.dsi.fastutil.objects.ObjectOpenHashSet<>(required); // Canvas - optimize ContextMap#create
+        this.allowed = new it.unimi.dsi.fastutil.objects.ObjectOpenHashSet<>(Sets.union(required, optional)); // Canvas - optimize ContextMap#create
     }
 
     public Set<ContextKey<?>> required() {
