--- a/src/main/java/org/spigotmc/RestartCommand.java
+++ b/src/main/java/org/spigotmc/RestartCommand.java
@@ -2,6 +_,7 @@
 
 import java.io.File;
 import java.util.Locale;
+import io.canvasmc.canvas.server.MultiWatchdogThread;
 import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.ServerPlayer;
 import org.bukkit.command.Command;
@@ -40,7 +_,7 @@
                 System.out.println("Startup script '" + SpigotConfig.restartScript + "' does not exist! Stopping server.");
             }
             // Stop the watchdog
-            WatchdogThread.doStop();
+            MultiWatchdogThread.doStop(); // Canvas - rewrite watchdog
 
             shutdownServer(isRestarting);
             // Paper end
