--- a/src/main/java/org/spigotmc/SpigotConfig.java
+++ b/src/main/java/org/spigotmc/SpigotConfig.java
@@ -110,7 +_,7 @@
         return SpigotConfig.config.getBoolean(path, SpigotConfig.config.getBoolean(path));
     }
 
-    private static int getInt(String path, int def) {
+    public static int getInt(String path, int def) { // Canvas - private -> public
         SpigotConfig.config.addDefault(path, def);
         return SpigotConfig.config.getInt(path, SpigotConfig.config.getInt(path));
     }
