--- a/src/main/java/org/spigotmc/TicksPerSecondCommand.java
+++ b/src/main/java/org/spigotmc/TicksPerSecondCommand.java
@@ -1,5 +_,6 @@
 package org.spigotmc;
 
+import io.canvasmc.canvas.command.ThreadedTickDiagnosis;
 import net.kyori.adventure.text.Component;
 import net.kyori.adventure.text.JoinConfiguration;
 import net.kyori.adventure.text.TextComponent;
@@ -33,6 +_,12 @@
         if (!this.testPermission(sender)) {
             return true;
         }
+        // Canvas start - we have our own new tick command
+        if (true) {
+            ThreadedTickDiagnosis.dump(sender);
+            return true;
+        }
+        // Canvas end
 
         // Paper start - Further improve tick handling
         double[] tps = org.bukkit.Bukkit.getTPS();
