From a1940d59a7eeb8b7dee60550bc75804e3611cb1d Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/goal/FloatGoal.java b/net/minecraft/world/entity/ai/goal/FloatGoal.java
index 1b609edf8ad5ebb14b8a27384178eac4c0f1f58f..35c8ff11a015ff4e7c5e8ed58579b305411afac2 100644
--- a/net/minecraft/world/entity/ai/goal/FloatGoal.java
+++ b/net/minecraft/world/entity/ai/goal/FloatGoal.java
@@ -16,7 +16,7 @@ public class FloatGoal extends Goal {
 
     @Override
     public boolean canUse() {
-        return this.mob.isInWater() && this.mob.getFluidHeight(FluidTags.WATER) > this.mob.getFluidJumpThreshold() || this.mob.isInLava();
+        return this.mob.isInWater() && this.mob.canvas$getWaterFluidHeight() > this.mob.getFluidJumpThreshold() || this.mob.isInLava(); // Canvas - optimize fluid height
     }
 
     @Override
