--- a/net/minecraft/world/entity/ai/goal/FloatGoal.java
+++ b/net/minecraft/world/entity/ai/goal/FloatGoal.java
@@ -16,7 +_,7 @@
 
     @Override
     public boolean canUse() {
-        return this.mob.isInWater() && this.mob.getFluidHeight(FluidTags.WATER) > this.mob.getFluidJumpThreshold() || this.mob.isInLava();
+        return this.mob.isInWater() && this.mob.canvas$getWaterFluidHeight() > this.mob.getFluidJumpThreshold() || this.mob.isInLava(); // Canvas - optimize fluid height
     }
 
     @Override
