--- a/net/minecraft/world/entity/ai/goal/GoalSelector.java
+++ b/net/minecraft/world/entity/ai/goal/GoalSelector.java
@@ -1,6 +_,7 @@
 package net.minecraft.world.entity.ai.goal;
 
 import com.google.common.annotations.VisibleForTesting;
+import io.canvasmc.canvas.Config;
 import it.unimi.dsi.fastutil.objects.ObjectLinkedOpenHashSet;
 import java.util.EnumMap;
 import java.util.EnumSet;
@@ -29,7 +_,19 @@
     private int curRate; // Paper - EAR 2
 
     public void addGoal(int priority, Goal goal) {
-        this.availableGoals.add(new WrappedGoal(priority, goal));
+        // Canvas start
+        WrappedGoal wrappedGoal = new WrappedGoal(priority, goal);
+        if (Config.COMPILED_GOAL_MASKS.containsKey(goal.getClass())) {
+            Config.Entities.GoalMask mask = Config.COMPILED_GOAL_MASKS.get(goal.getClass());
+            if (mask.disableGoal) {
+                // We don't even try adding it, server has disabled the goal.
+                return;
+            } else if (mask.goalTickDelay > 0) {
+                wrappedGoal.setTickDelay(mask.goalTickDelay);
+            }
+        }
+        this.availableGoals.add(wrappedGoal);
+        // Canvas end
     }
 
     @VisibleForTesting
@@ -83,7 +_,7 @@
         return true;
     }
 
-    public void tick() {
+    public void tick(int tickCount) { // Canvas
         ProfilerFiller profilerFiller = Profiler.get();
         profilerFiller.push("goalCleanup");
 
@@ -99,7 +_,8 @@
 
         for (WrappedGoal wrappedGoalx : this.availableGoals) {
             // Paper start
-            if (!wrappedGoalx.isRunning() && !goalContainsAnyFlags(wrappedGoalx, this.goalTypes) && goalCanBeReplacedForAllFlags(wrappedGoalx, this.lockedFlags) && wrappedGoalx.canUse()) {
+            if (wrappedGoalx == null) continue; // Canvas - NPE
+            if (!wrappedGoalx.isRunning() && !goalContainsAnyFlags(wrappedGoalx, this.goalTypes) && goalCanBeReplacedForAllFlags(wrappedGoalx, this.lockedFlags) && wrappedGoalx.canUse(tickCount)) { // Canvas
                 long flagIterator = wrappedGoalx.getFlags().getBackingSet();
                 int wrappedGoalSize = wrappedGoalx.getFlags().size();
                 for (int i = 0; i < wrappedGoalSize; ++i) {
