--- a/net/minecraft/world/entity/ai/goal/GoalSelector.java
+++ b/net/minecraft/world/entity/ai/goal/GoalSelector.java
@@ -1,6 +_,7 @@
 package net.minecraft.world.entity.ai.goal;
 
 import com.google.common.annotations.VisibleForTesting;
+import io.canvasmc.canvas.Config;
 import it.unimi.dsi.fastutil.objects.ObjectLinkedOpenHashSet;
 import java.util.EnumMap;
 import java.util.EnumSet;
@@ -29,6 +_,11 @@
     private int curRate; // Paper - EAR 2
 
     public void addGoal(int priority, Goal goal) {
+        // Canvas start
+        if (Config.COMPILED_DISABLED_GOAL_CLASSES.contains(goal.getClass())) {
+            return;
+        }
+        // Canvas end
         this.availableGoals.add(new WrappedGoal(priority, goal));
     }
 
@@ -99,6 +_,7 @@
 
         for (WrappedGoal wrappedGoalx : this.availableGoals) {
             // Paper start
+            if (wrappedGoalx == null) continue; // Canvas - NPE
             if (!wrappedGoalx.isRunning() && !goalContainsAnyFlags(wrappedGoalx, this.goalTypes) && goalCanBeReplacedForAllFlags(wrappedGoalx, this.lockedFlags) && wrappedGoalx.canUse()) {
                 long flagIterator = wrappedGoalx.getFlags().getBackingSet();
                 int wrappedGoalSize = wrappedGoalx.getFlags().size();
