--- a/net/minecraft/world/entity/ai/goal/MoveToBlockGoal.java
+++ b/net/minecraft/world/entity/ai/goal/MoveToBlockGoal.java
@@ -1,6 +_,8 @@
 package net.minecraft.world.entity.ai.goal;
 
 import java.util.EnumSet;
+import io.canvasmc.canvas.Config;
+import io.canvasmc.canvas.region.ServerRegions;
 import net.minecraft.core.BlockPos;
 import net.minecraft.world.entity.PathfinderMob;
 import net.minecraft.world.level.LevelReader;
@@ -43,6 +_,7 @@
 
     @Override
     public boolean canUse() {
+        if (Config.INSTANCE.entities.cramming.reduceEntityMoveWhenCrammed && (ServerRegions.getTickData(this.mob.level().level()).nonThreadsafeRandom.nextInt(100) >= 10 && !this.mob.canTickMoveOptimized())) return false; // Canvas - reduce movement when crammed
         if (this.nextStartTick > 0) {
             this.nextStartTick--;
             return false;
