--- a/net/minecraft/world/entity/ai/goal/MoveToBlockGoal.java
+++ b/net/minecraft/world/entity/ai/goal/MoveToBlockGoal.java
@@ -119,6 +_,7 @@
                 for (int x = 0; x <= r; x = x > 0 ? -x : 1 - x) {
                     for (int z = x < r && x > -r ? r : 0; z <= r; z = z > 0 ? -z : 1 - z) {
                         pos.setWithOffset(mobPos, x, y - 1, z);
+                        if (!this.mob.level().hasChunkAt(pos)) continue; // Pufferfish - if this block isn't loaded, continue
                         if (this.mob.isWithinHome(pos) && this.isValidTarget(this.mob.level(), pos)) {
                             this.blockPos = pos;
                             this.mob.movingTarget = pos == BlockPos.ZERO ? null : pos.immutable(); // Paper
