--- a/net/minecraft/world/entity/ai/goal/RandomStrollGoal.java
+++ b/net/minecraft/world/entity/ai/goal/RandomStrollGoal.java
@@ -2,6 +_,8 @@
 
 import java.util.EnumSet;
 import javax.annotation.Nullable;
+import io.canvasmc.canvas.Config;
+import io.canvasmc.canvas.region.ServerRegions;
 import net.minecraft.world.entity.PathfinderMob;
 import net.minecraft.world.entity.ai.util.DefaultRandomPos;
 import net.minecraft.world.phys.Vec3;
@@ -35,6 +_,7 @@
 
     @Override
     public boolean canUse() {
+        if (Config.INSTANCE.entities.cramming.reduceEntityMoveWhenCrammed && (ServerRegions.getTickData(this.mob.level().level()).nonThreadsafeRandom.nextInt(100) >= 10 && !this.mob.canTickMoveOptimized())) return false; // Canvas - reduce movement when crammed
         if (this.mob.hasControllingPassenger()) {
             return false;
         } else {
