From 566bd10dc5d91480e58a5345fdfc0ebbb9138662 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java b/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
index 50f12f2581267827c394134cd51a09016a5fa2b0..4c8cf03dbc09a5c02aec2ddaf5a4d9f233bd4afb 100644
--- a/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
+++ b/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
@@ -116,6 +116,7 @@ public class RangedBowAttackGoal<T extends Monster & RangedAttackMob> extends Go
                 }
 
                 this.mob.lookAt(target, 30.0F, 30.0F);
+                if (io.canvasmc.canvas.Config.INSTANCE.fixes.mc121706) this.mob.getLookControl().setLookAt(target, 30f, 30f); // Canvas - MC-121706
             } else {
                 this.mob.getLookControl().setLookAt(target, 30.0F, 30.0F);
             }
