--- a/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
+++ b/net/minecraft/world/entity/ai/goal/RangedBowAttackGoal.java
@@ -116,6 +_,7 @@
                 }
 
                 this.mob.lookAt(target, 30.0F, 30.0F);
+                if (io.canvasmc.canvas.Config.INSTANCE.fixes.mc121706) this.mob.getLookControl().setLookAt(target, 30f, 30f); // Canvas - MC-121706
             } else {
                 this.mob.getLookControl().setLookAt(target, 30.0F, 30.0F);
             }
