--- a/net/minecraft/world/entity/ai/goal/WrappedGoal.java
+++ b/net/minecraft/world/entity/ai/goal/WrappedGoal.java
@@ -7,6 +_,7 @@
     private final Goal goal;
     private final int priority;
     private boolean isRunning;
+    private int delay = 0; // Canvas
 
     public WrappedGoal(int priority, Goal goal) {
         this.priority = priority;
@@ -16,6 +_,23 @@
     public boolean canBeReplacedBy(WrappedGoal other) {
         return this.isInterruptable() && other.getPriority() < this.getPriority();
     }
+    // Canvas start
+
+    public void setTickDelay(int delay) {
+        this.delay = delay;
+    }
+
+    public int getDelay() {
+        return delay;
+    }
+
+    public boolean canUse(int tickCount) {
+        if (delay > 0) {
+            return tickCount % delay == 0;
+        }
+        return canUse();
+    }
+    // Canvas end
 
     @Override
     public boolean canUse() {
