--- a/src/main/java/io/papermc/paper/threadedregions/EntityScheduler.java
+++ b/src/main/java/io/papermc/paper/threadedregions/EntityScheduler.java
@@ -169,6 +_,7 @@
             }
             final ScheduledTask task = this.currentlyExecuting.pollFirst();
 
+            if (task == null) break; // Canvas - break if null
             if (this.tickCount != RETIRED_TICK_COUNT) {
                 ((Consumer<Entity>)task.run).accept(thisEntity);
             } else {
