--- a/src/main/java/io/papermc/paper/threadedregions/EntityScheduler.java
+++ b/src/main/java/io/papermc/paper/threadedregions/EntityScheduler.java
@@ -144,7 +_,7 @@
         final List<ScheduledTask> toRun;
         synchronized (this.stateLock) {
             if (this.tickCount == RETIRED_TICK_COUNT) {
-                throw new IllegalStateException("Ticking retired scheduler");
+                return; // Canvas - the entity can be removed concurrently, do not crash the server because of this.
             }
             ++this.tickCount;
             if (this.oneTimeDelayed.isEmpty()) {
@@ -169,6 +_,7 @@
             }
             final ScheduledTask task = this.currentlyExecuting.pollFirst();
 
+            if (task == null) break; // Canvas - break if null
             if (this.tickCount != RETIRED_TICK_COUNT) {
                 ((Consumer<Entity>)task.run).accept(thisEntity);
             } else {
