--- a/net/minecraft/world/entity/decoration/ItemFrame.java
+++ b/net/minecraft/world/entity/decoration/ItemFrame.java
@@ -130,7 +_,7 @@
         } else {
             BlockState blockState = this.level().getBlockState(this.pos.relative(this.getDirection().getOpposite()));
             return (blockState.isSolid() || this.getDirection().getAxis().isHorizontal() && DiodeBlock.isDiode(blockState))
-                && this.level().getEntities(this, this.getBoundingBox(), HANGING_ENTITY).isEmpty();
+                && this.level().getEntitiesOfClass(HangingEntity.class, this.getBoundingBox(), entity -> entity != this).isEmpty(); // Canvas - optimize entity type predicates
         }
     }
 
