--- a/net/minecraft/server/commands/EnchantCommand.java
+++ b/net/minecraft/server/commands/EnchantCommand.java
@@ -67,8 +_,8 @@
         final CommandSourceStack source, final Collection<? extends Entity> targets, final Holder<Enchantment> enchantmentHolder, final int level
     ) throws CommandSyntaxException {
         Enchantment enchantment = enchantmentHolder.value();
-        if (level > enchantment.getMaxLevel()) {
-            throw ERROR_LEVEL_TOO_HIGH.create(level, enchantment.getMaxLevel());
+        if (level > (io.canvasmc.canvas.GlobalConfiguration.getInstance().enchantCommand.uncapMaxLevel ? 255 : enchantment.getMaxLevel())) { // Canvas - enchant command configs
+            throw ERROR_LEVEL_TOO_HIGH.create(level, (io.canvasmc.canvas.GlobalConfiguration.getInstance().enchantCommand.uncapMaxLevel ? 255 : enchantment.getMaxLevel())); // Canvas - enchant command configs
         }
 
         // Canvas start - region threading
@@ -88,8 +_,10 @@
             if (entity instanceof LivingEntity target) {
                 ItemStack item = target.getMainHandItem();
                 if (!item.isEmpty()) {
-                    if (enchantment.canEnchant(item)
-                        && EnchantmentHelper.isEnchantmentCompatible(EnchantmentHelper.getEnchantmentsForCrafting(item).keySet(), enchantmentHolder)) {
+                    // Canvas start - enchant command configs
+                    if ((io.canvasmc.canvas.GlobalConfiguration.getInstance().enchantCommand.allowEnchantsOnUnsupportedItems || enchantment.canEnchant(item))
+                        && (io.canvasmc.canvas.GlobalConfiguration.getInstance().enchantCommand.allowEnchantingWithIncompatibleEnchants || EnchantmentHelper.isEnchantmentCompatible(EnchantmentHelper.getEnchantmentsForCrafting(item).keySet(), enchantmentHolder))) {
+                    // Canvas end - enchant command configs
                         item.enchant(enchantmentHolder, level);
                         return (list) -> list.add(entity); // Canvas - region threading
                     } else if (targets.size() == 1) {
