--- a/net/minecraft/server/commands/FillCommand.java
+++ b/net/minecraft/server/commands/FillCommand.java
@@ -161,6 +_,8 @@
         } else {
             List<BlockPos> list = Lists.newArrayList();
             ServerLevel level = source.getLevel();
+            // Canvas start - move to level-main
+            level.schedule(level.wrapRunnable(() -> {
             int i1 = 0;
 
             for (BlockPos blockPos : BlockPos.betweenClosed(area.minX(), area.minY(), area.minZ(), area.maxX(), area.maxY(), area.maxZ())) {
@@ -183,12 +_,19 @@
             }
 
             if (i1 == 0) {
-                throw ERROR_FAILED.create();
+                /* try {
+                    throw ERROR_FAILED.create();
+                } catch (CommandSyntaxException e) {
+                    source.sendFailure(Component.literal(ERROR_FAILED.create().getMessage()));
+                    throw new RuntimeException("Unable to execute command", e);
+                } */
             } else {
                 int i2 = i1;
                 source.sendSuccess(() -> Component.translatable("commands.fill.success", i2), true);
-                return i1;
             }
+            }));
+            return 0;
+            // Canvas end
         }
     }
 
