From bd14df94165f1c34a7f632a7de807b770022d5c5 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/server/commands/LootCommand.java b/net/minecraft/server/commands/LootCommand.java
index 1d5d3b5af5057cfa2c7b9b495932042e4f5a71ad..5c0085c2d62fd00b939a23b53a9582bfc209a458 100644
--- a/net/minecraft/server/commands/LootCommand.java
+++ b/net/minecraft/server/commands/LootCommand.java
@@ -267,8 +267,14 @@ public class LootCommand {
     }
 
     private static int blockDistribute(final CommandSourceStack source, final BlockPos pos, final List<ItemStack> drops, final LootCommand.Callback callback) throws CommandSyntaxException {
-        Container container = getContainer(source, pos);
+        // Canvas start - region threading
         List<ItemStack> usedItems = Lists.newArrayListWithCapacity(drops.size());
+        io.papermc.paper.threadedregions.RegionizedServer.getInstance().taskQueue.queueOrExecuteTickTask(
+            source.getLevel(), pos.getX() >> 4, pos.getZ() >> 4,
+            () -> {
+                try {
+        // Canvas end - region threading
+        Container container = getContainer(source, pos);
 
         for (ItemStack drop : drops) {
             if (distributeToContainer(container, drop.copy())) {
@@ -278,6 +284,13 @@ public class LootCommand {
         }
 
         callback.accept(usedItems);
+        // Canvas start - region threading
+                } catch (CommandSyntaxException syntaxException) {
+                    source.sendFailure(Component.literal(syntaxException.getMessage()));
+                }
+            }
+        );
+        // Canvas end - region threading
         return usedItems.size();
     }
 
@@ -314,10 +327,16 @@ public class LootCommand {
         final List<ItemStack> drops,
         final LootCommand.Callback callback
     ) throws CommandSyntaxException {
+        // Canvas start - region threading
+        List<ItemStack> usedItems = Lists.newArrayListWithCapacity(drops.size());
+        io.papermc.paper.threadedregions.RegionizedServer.getInstance().taskQueue.queueOrExecuteTickTask(
+            source.getLevel(), pos.getX() >> 4, pos.getZ() >> 4,
+            () -> {
+                try {
+        // Canvas end - region threading
         Container container = getContainer(source, pos);
         int maxSlot = container.getContainerSize();
         if (startSlot >= 0 && startSlot < maxSlot) {
-            List<ItemStack> usedItems = Lists.newArrayListWithCapacity(drops.size());
 
             for (int i = 0; i < slotCount; i++) {
                 int slot = startSlot + i;
@@ -329,10 +348,18 @@ public class LootCommand {
             }
 
             callback.accept(usedItems);
-            return usedItems.size();
+            return; // Canvas - region threading
         } else {
             throw ItemCommands.ERROR_TARGET_INAPPLICABLE_SLOT.create(startSlot);
         }
+        // Canvas start - region threading
+                } catch (CommandSyntaxException syntaxException) {
+                    source.sendFailure(Component.literal(syntaxException.getMessage()));
+                }
+            }
+        );
+        return usedItems.size();
+        // Canvas end - region threading
     }
 
     private static boolean canMergeItems(final ItemStack a, final ItemStack b) {
@@ -343,10 +370,12 @@ public class LootCommand {
         List<ItemStack> usedItems = Lists.newArrayListWithCapacity(drops.size());
 
         for (ItemStack drop : drops) {
-            for (ServerPlayer player : players) {
+            for (ServerPlayer serverPlayer : players) { // Canvas - region threading
+                serverPlayer.getBukkitEntity().taskScheduler.scheduleOrExecute((ServerPlayer player) -> { // Canvas - region threading
                 if (player.getInventory().add(drop.copy())) {
                     usedItems.add(drop);
                 }
+                }); // Canvas - region threading
             }
         }
 
@@ -369,13 +398,15 @@ public class LootCommand {
     ) throws CommandSyntaxException {
         List<ItemStack> usedItems = Lists.newArrayListWithCapacity(drops.size());
 
-        for (Entity entity : entities) {
+        for (Entity anEntity : entities) { // Canvas - region threading
+            anEntity.getBukkitEntity().taskScheduler.scheduleOrExecute((entity) -> { // Canvas - region threading
             if (entity instanceof ServerPlayer player) {
                 setSlots(entity, drops, startSlot, count, usedItems);
                 player.containerMenu.broadcastChanges();
             } else {
                 setSlots(entity, drops, startSlot, count, usedItems);
             }
+            }); // Canvas - region threading
         }
 
         callback.accept(usedItems);
@@ -385,9 +416,15 @@ public class LootCommand {
     private static int dropInWorld(final CommandSourceStack source, final Vec3 pos, final List<ItemStack> drops, final LootCommand.Callback callback) throws CommandSyntaxException {
         ServerLevel level = source.getLevel();
         drops.forEach(drop -> {
+            // Canvas start - region threading
+            io.papermc.paper.threadedregions.RegionizedServer.getInstance().taskQueue.queueOrExecuteTickTask(
+                source.getLevel(), net.minecraft.util.Mth.floor(pos.x) >> 4, net.minecraft.util.Mth.floor(pos.z) >> 4,
+                () -> {
+            // Canvas end - region threading
             ItemEntity entity = new ItemEntity(level, pos.x, pos.y, pos.z, drop.copy());
             entity.setDefaultPickUpDelay();
             level.addFreshEntity(entity);
+            }); // Canvas - region threading
         });
         callback.accept(drops);
         return drops.size();
@@ -432,6 +469,12 @@ public class LootCommand {
     ) throws CommandSyntaxException {
         CommandSourceStack source = context.getSource();
         ServerLevel level = source.getLevel();
+        // Canvas start - region threading
+        io.papermc.paper.threadedregions.RegionizedServer.getInstance().taskQueue.queueOrExecuteTickTask(
+            source.getLevel(), pos.getX() >> 4, pos.getZ() >> 4,
+            () -> {
+                try {
+        // Canvas end - region threading
         BlockState blockState = level.getBlockState(pos);
         BlockEntity blockEntity = level.getBlockEntity(pos);
         Optional<ResourceKey<LootTable>> lootTable = blockState.getBlock().getLootTable();
@@ -445,8 +488,16 @@ public class LootCommand {
                 .withOptionalParameter(LootContextParams.THIS_ENTITY, source.getEntity())
                 .withParameter(LootContextParams.TOOL, tool);
             List<ItemStack> drops = blockState.getDrops(lootParams);
-            return output.accept(context, drops, usedItems -> callback(source, usedItems, lootTable.get()));
+            output.accept(context, drops, usedItems -> callback(source, usedItems, lootTable.get())); // Canvas - region threading
         }
+        // Canvas start - region threading
+                } catch (CommandSyntaxException syntaxException) {
+                    context.getSource().sendFailure(Component.literal(syntaxException.getMessage()));
+                }
+            }
+        );
+        return 0;
+        // Canvas end - region threading
     }
 
     private static int dropKillLoot(final CommandContext<CommandSourceStack> context, final Entity target, final LootCommand.DropConsumer output) throws CommandSyntaxException {
