From 4d4ec3af021afd5dcd0fd24e659eea8297196156 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/server/commands/SaveAllCommand.java b/net/minecraft/server/commands/SaveAllCommand.java
index bc6d01996dde0870f6ebbda14176352777938a0c..1049925e6bc9dadc268129123b9f41b690d76e38 100644
--- a/net/minecraft/server/commands/SaveAllCommand.java
+++ b/net/minecraft/server/commands/SaveAllCommand.java
@@ -23,12 +23,40 @@ public class SaveAllCommand {
     private static int saveAll(final CommandSourceStack source, final boolean flush) throws CommandSyntaxException {
         source.sendSuccess(() -> Component.translatable("commands.save.saving"), false);
         MinecraftServer server = source.getServer();
-        boolean success = server.saveEverything(true, flush, true);
-        if (!success) {
-            throw ERROR_FAILED.create();
-        } else {
-            source.sendSuccess(() -> Component.translatable("commands.save.success"), true);
-            return 1;
+        // Canvas start - region threading
+        final java.util.concurrent.atomic.AtomicInteger regionCount = new java.util.concurrent.atomic.AtomicInteger();
+        final java.util.concurrent.atomic.AtomicInteger totalCount = new java.util.concurrent.atomic.AtomicInteger();
+        Runnable callback = () -> {
+            if (regionCount.decrementAndGet() == 0 || totalCount.get() == 0) {
+                source.sendSuccess(() -> Component.translatable("commands.save.success"), true);
+            }
+        };
+        for (final net.minecraft.server.level.ServerLevel world : server.getAllLevels()) {
+            world.regioniser.computeForAllRegions((region) -> {
+                regionCount.getAndIncrement();
+                totalCount.getAndIncrement();
+                region.getData().canvas$saveAllTicket.propagate(
+                    new io.canvasmc.canvas.util.ticket.SaveAllTicket(callback, (ex) -> {
+                        source.sendFailure(Component.literal(ERROR_FAILED.create().getMessage()));
+                        MinecraftServer.LOGGER.error(ERROR_FAILED.create().getMessage(), ex);
+                    }, flush)
+                );
+            });
         }
+
+        // pearl save rewrite
+        MinecraftServer.getServer().pearls.save((result) -> {
+            if (result) {
+                source.sendSuccess(() -> Component.literal("Saved pearl data"), true);
+            } else {
+                source.sendFailure(Component.literal("Failed to save pearl data, check console for details"));
+            }
+        });
+
+        // if no regions loaded, run callback
+        source.sendSuccess(() -> Component.literal("Saving " + totalCount.get() + " regions"), true);
+        if (totalCount.get() == 0) callback.run();
+        return 1;
+        // Canvas end - region threading
     }
 }
