--- a/net/minecraft/server/commands/SummonCommand.java
+++ b/net/minecraft/server/commands/SummonCommand.java
@@ -3,6 +_,7 @@
 import com.mojang.brigadier.CommandDispatcher;
 import com.mojang.brigadier.exceptions.CommandSyntaxException;
 import com.mojang.brigadier.exceptions.SimpleCommandExceptionType;
+import io.canvasmc.canvas.Config;
 import net.minecraft.commands.CommandBuildContext;
 import net.minecraft.commands.CommandSourceStack;
 import net.minecraft.commands.Commands;
@@ -15,6 +_,7 @@
 import net.minecraft.core.registries.Registries;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.chat.Component;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.entity.Entity;
 import net.minecraft.world.entity.EntitySpawnReason;
@@ -41,7 +_,7 @@
                                 ResourceArgument.getSummonableEntityType(context1, "entity"),
                                 context1.getSource().getPosition(),
                                 new CompoundTag(),
-                                true
+                                true, true // Canvas
                             )
                         )
                         .then(
@@ -52,7 +_,7 @@
                                         ResourceArgument.getSummonableEntityType(context1, "entity"),
                                         Vec3Argument.getVec3(context1, "pos"),
                                         new CompoundTag(),
-                                        true
+                                        true, true // Canvas
                                     )
                                 )
                                 .then(
@@ -63,7 +_,7 @@
                                                 ResourceArgument.getSummonableEntityType(context1, "entity"),
                                                 Vec3Argument.getVec3(context1, "pos"),
                                                 CompoundTagArgument.getCompoundTag(context1, "nbt"),
-                                                false
+                                                false, true // Canvas
                                             )
                                         )
                                 )
@@ -102,7 +_,15 @@
         }
     }
 
-    private static int spawnEntity(CommandSourceStack source, Holder.Reference<EntityType<?>> type, Vec3 pos, CompoundTag tag, boolean randomizeProperties) throws CommandSyntaxException {
+    // Canvas start - summon batching
+    public static int spawnEntity(CommandSourceStack source, Holder.Reference<EntityType<?>> type, Vec3 pos, CompoundTag tag, boolean randomizeProperties, boolean shouldBatch) throws CommandSyntaxException {
+        if (shouldBatch && Config.INSTANCE.batchSummonCommandTasks) {
+            MinecraftServer.getServer().summonCommandManager.batch(
+                source, type, pos, tag, randomizeProperties
+            );
+            return 1;
+        }
+    // Canvas end
         Entity entity = createEntity(source, type, pos, tag, randomizeProperties);
         source.sendSuccess(() -> Component.translatable("commands.summon.success", entity.getDisplayName()), true);
         return 1;
