--- a/net/minecraft/server/commands/TeleportCommand.java
+++ b/net/minecraft/server/commands/TeleportCommand.java
@@ -240,8 +_,8 @@
             double relativeOrAbsoluteX = relatives.contains(Relative.X) ? x - victim.getX() : x;
             double relativeOrAbsoluteY = relatives.contains(Relative.Y) ? y - victim.getY() : y;
             double relativeOrAbsoluteZ = relatives.contains(Relative.Z) ? z - victim.getZ() : z;
-            float relativeOrAbsoluteYRot = relatives.contains(Relative.Y_ROT) ? yRot - victim.getYRot() : yRot;
-            float relativeOrAbsoluteXRot = relatives.contains(Relative.X_ROT) ? xRot - victim.getXRot() : xRot;
+            float relativeOrAbsoluteYRot = relatives.contains(Relative.Y_ROT) && source.getEntity() != victim ? yRot - victim.getYRot() : yRot; // Canvas - region threading
+            float relativeOrAbsoluteXRot = relatives.contains(Relative.X_ROT) && source.getEntity() != victim ? xRot - victim.getXRot() : xRot; // Canvas - region threading
             float newYRot = Mth.wrapDegrees(relativeOrAbsoluteYRot);
             float newXRot = Mth.wrapDegrees(relativeOrAbsoluteXRot);
             // Folia start - region threading
