--- a/net/minecraft/server/commands/TickCommand.java
+++ b/net/minecraft/server/commands/TickCommand.java
@@ -18,6 +_,11 @@
     private static final String DEFAULT_TICKRATE = String.valueOf(20);
 
     public static void register(CommandDispatcher<CommandSourceStack> dispatcher) {
+        // we disable this because with canvas
+        // threading, this behavior becomes unpredictable
+        // and possibly unstable.
+        // TODO - make this work?
+        if (true) return; // Canvas - disable tick command
         dispatcher.register(
             Commands.literal("tick")
                 .requires(commandSourceStack -> commandSourceStack.hasPermission(3))
