--- a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
@@ -383,8 +_,12 @@
             final int centerZ = PlayerChunkLoaderData.this.lastChunkZ;
 
             return Integer.compare(
-                Math.abs(c1x - centerX) + Math.abs(c1z - centerZ),
-                Math.abs(c2x - centerX) + Math.abs(c2z - centerZ)
+                // Canvas start - use Euclidean distance squared
+                io.canvasmc.canvas.Config.INSTANCE.chunks.useEuclideanDistanceSquared ?
+                    (c1x - centerX) * (c1x - centerX) + (c1z - centerZ) * (c1z - centerZ) : Math.abs(c1x - centerX) + Math.abs(c1z - centerZ),
+                io.canvasmc.canvas.Config.INSTANCE.chunks.useEuclideanDistanceSquared ?
+                    (c2x - centerX) * (c2x - centerX) + (c2z - centerZ) * (c2z - centerZ) : Math.abs(c2x - centerX) + Math.abs(c2z - centerZ)
+                // Canvas end - use Euclidean distance squared
             );
         };
         private final LongHeapPriorityQueue sendQueue = new LongHeapPriorityQueue(CLOSEST_MANHATTAN_DIST);
