From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/server/commands/SaveAllCommand.java b/net/minecraft/server/commands/SaveAllCommand.java
index e37b71402179c73146ff67babc228ccfb996f8f4..f16e36cdf0ae4f7f2f61c7716770709bdf889058 100644
--- a/net/minecraft/server/commands/SaveAllCommand.java
+++ b/net/minecraft/server/commands/SaveAllCommand.java
@@ -21,14 +21,41 @@ public class SaveAllCommand {
     }
 
     private static int saveAll(CommandSourceStack source, boolean flush) throws CommandSyntaxException {
+        // Canvas start - fix commands with region threading
         source.sendSuccess(() -> Component.translatable("commands.save.saving"), false);
         MinecraftServer server = source.getServer();
-        boolean flag = server.saveEverything(true, flush, true);
-        if (!flag) {
-            throw ERROR_FAILED.create();
-        } else {
-            source.sendSuccess(() -> Component.translatable("commands.save.success"), true);
-            return 1;
+        final java.util.concurrent.atomic.AtomicInteger rc = new java.util.concurrent.atomic.AtomicInteger();
+        final java.util.concurrent.atomic.AtomicInteger count = new java.util.concurrent.atomic.AtomicInteger();
+        Runnable callback = () -> {
+            if (rc.decrementAndGet() == 0 || count.get() == 0) {
+                source.sendSuccess(() -> Component.translatable("commands.save.success"), true);
+            }
+        };
+        for (final net.minecraft.server.level.ServerLevel world : server.getAllLevels()) {
+            world.regioniser.computeForAllRegions((region) -> {
+                rc.getAndIncrement();
+                count.getAndIncrement();
+                region.getData().canvas$saveAllTicket.propagate(
+                    new io.canvasmc.canvas.util.ticket.SaveAllTicket(callback, (ex) -> {
+                        source.sendFailure(Component.literal(ERROR_FAILED.create().getMessage()));
+                        MinecraftServer.LOGGER.error(ERROR_FAILED.create().getMessage(), ex);
+                    }, flush)
+                );
+            });
         }
+        // Canvas start - region threading
+        MinecraftServer.getServer().pearls.save((result) -> {
+            if (result) {
+                source.sendSuccess(() -> Component.literal("Saved pearl data"), true);
+            } else {
+                source.sendFailure(Component.literal("Failed to save pearl data, check console for details"));
+            }
+        });
+        // Canvas end - region threading
+        // if no regions loaded, run callback
+        source.sendSuccess(() -> Component.literal("Saving " + count.get() + " regions"), true);
+        if (count.get() == 0) callback.run();
+        return 1;
+        // Canvas end - fix commands with region threading
     }
 }
