From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/util/context/ContextKeySet.java b/net/minecraft/util/context/ContextKeySet.java
index d7d95a4a72d62b2f57391b19c9873d482c64a61e..ec532f0ec8d9666a60023ad7f7d9f73bc9529268 100644
--- a/net/minecraft/util/context/ContextKeySet.java
+++ b/net/minecraft/util/context/ContextKeySet.java
@@ -9,8 +9,8 @@ public class ContextKeySet {
     private final Set<ContextKey<?>> allowed;
 
     private ContextKeySet(final Set<ContextKey<?>> required, final Set<ContextKey<?>> optional) {
-        this.required = Set.copyOf(required);
-        this.allowed = Set.copyOf(Sets.union(required, optional));
+        this.required = new it.unimi.dsi.fastutil.objects.ObjectOpenHashSet<>(required); // Canvas - optimize ContextMap#create
+        this.allowed = new it.unimi.dsi.fastutil.objects.ObjectOpenHashSet<>(Sets.union(required, optional)); // Canvas - optimize ContextMap#create
     }
 
     public Set<ContextKey<?>> required() {
