--- a/net/minecraft/data/advancements/packs/VanillaNetherAdvancements.java
+++ b/net/minecraft/data/advancements/packs/VanillaNetherAdvancements.java
@@ -151,7 +_,7 @@
             .addCriterion(
                 "killed_ghast",
                 KilledTrigger.TriggerInstance.playerKilledEntity(
-                    EntityPredicate.Builder.entity().of(entityTypes, EntityType.GHAST).located(LocationPredicate.Builder.inDimension(Level.OVERWORLD))
+                    EntityPredicate.Builder.entity().of(entityTypes, EntityType.GHAST).located(LocationPredicate.Builder.inDimension(io.canvasmc.canvas.Config.INSTANCE.fetchRespawnDimensionKey())) // Canvas - default respawn dimension config - this is where the nether portal ties to, so we should change this to match that
                 )
             )
             .save(output, "nether/uneasy_alliance");
@@ -442,7 +_,7 @@
                 "ride_entity_distance",
                 DistanceTrigger.TriggerInstance.rideEntityInLava(
                     EntityPredicate.Builder.entity()
-                        .located(LocationPredicate.Builder.inDimension(Level.OVERWORLD))
+                        .located(LocationPredicate.Builder.inDimension(io.canvasmc.canvas.Config.INSTANCE.fetchRespawnDimensionKey())) // Canvas - default respawn dimension config - this is where the nether portal ties to, so we should change this to match that
                         .vehicle(EntityPredicate.Builder.entity().of(entityTypes, EntityType.STRIDER)),
                     DistancePredicate.horizontal(MinMaxBounds.Doubles.atLeast(50.0))
                 )
