--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/task/ChunkFullTask.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/task/ChunkFullTask.java
@@ -10,6 +_,7 @@
 import ca.spottedleaf.moonrise.patches.chunk_system.level.poi.PoiChunk;
 import ca.spottedleaf.moonrise.patches.chunk_system.scheduling.ChunkTaskScheduler;
 import ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder;
+import io.canvasmc.canvas.server.chunk.ChunkPriorityManager;
 import net.minecraft.server.level.ServerChunkCache;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.world.level.chunk.ChunkAccess;
@@ -138,7 +_,7 @@
         if (!Priority.isValidPriority(priority)) {
             throw new IllegalArgumentException("Invalid priority " + priority);
         }
-        this.convertToFullTask.setPriority(priority);
+        this.convertToFullTask.setPriority(this.world.getPriorityManager().getPriority(chunkX, chunkZ, ChunkPriorityManager.TaskType.FULL)); // Canvas - optimize moonrise
     }
 
     @Override
