--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/task/ChunkProgressionTask.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/task/ChunkProgressionTask.java
@@ -81,7 +_,7 @@
 
     private void complete0(final ChunkAccess chunk, final Throwable throwable) {
         if ((boolean)COMPLETED_HANDLE.getAndSet((ChunkProgressionTask)this, (boolean)true)) {
-            throw new IllegalStateException("Already completed");
+            return; // Canvas - maybe dont crash the server when its already completed?
         }
         this.completedChunk = chunk;
         this.completedThrowable = throwable;
