--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/task/GenericDataLoadTask.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/task/GenericDataLoadTask.java
@@ -45,9 +_,11 @@
     protected final PrioritisedExecutor.PrioritisedTask processOnMain;
 
     protected final ChunkTaskScheduler scheduler;
-    protected final ServerLevel world;
-    protected final int chunkX;
-    protected final int chunkZ;
+    // Canvas start - protected -> public
+    public final ServerLevel world;
+    public final int chunkX;
+    public final int chunkZ;
+    // Canvas end
     protected final MoonriseRegionFileIO.RegionFileType type;
 
     public GenericDataLoadTask(final ChunkTaskScheduler scheduler, final ServerLevel world, final int chunkX,
@@ -292,7 +_,7 @@
         }
     }
 
-    private final class ProcessOffMainTask implements Runnable {
+    public final class ProcessOffMainTask implements Runnable { // Canvas - private -> public
 
         private CompoundTag data;
         private Throwable throwable;
@@ -333,9 +_,15 @@
                 GenericDataLoadTask.this.onComplete((TaskResult<FinalCompletion, Throwable>)newData);
             }
         }
+        // Canvas start
+
+        public GenericDataLoadTask<OnMain,FinalCompletion> loadTask() {
+            return GenericDataLoadTask.this;
+        }
+        // Canvas end
     }
 
-    private final class ProcessOnMainTask implements Runnable {
+    public final class ProcessOnMainTask implements Runnable { // Canvas - private -> public
 
         private OnMain data;
         private Throwable throwable;
@@ -350,6 +_,12 @@
 
             GenericDataLoadTask.this.onComplete(result);
         }
+        // Canvas start
+
+        public GenericDataLoadTask<OnMain,FinalCompletion> loadTask() {
+            return GenericDataLoadTask.this;
+        }
+        // Canvas end
     }
 
     protected static final class LoadDataFromDiskTask {
