--- a/net/minecraft/world/level/levelgen/material/MaterialRuleList.java
+++ b/net/minecraft/world/level/levelgen/material/MaterialRuleList.java
@@ -8,13 +_,16 @@
 public record MaterialRuleList(NoiseChunk.BlockStateFiller[] materialRuleList) implements NoiseChunk.BlockStateFiller {
     @Override
     public @Nullable BlockState calculate(final DensityFunction.FunctionContext context) {
-        for (NoiseChunk.BlockStateFiller rule : this.materialRuleList) {
-            BlockState state = rule.calculate(context);
-            if (state != null) {
-                return state;
-            }
-        }
+        // Canvas start - reduce worldgen allocations
+        // avoid iterator allocation
+        BlockState state = null;
+        int ruleCount = this.materialRuleList.length;
 
-        return null;
+        for (int i = 0; state == null && i < ruleCount; i++) {
+            NoiseChunk.BlockStateFiller rule = this.materialRuleList[i];
+            state = rule.calculate(context);
+        }
+        return state;
+        // Canvas end - reduce worldgen allocations
     }
 }
