--- a/net/minecraft/world/entity/schedule/Activity.java
+++ b/net/minecraft/world/entity/schedule/Activity.java
@@ -32,10 +_,12 @@
     public static final Activity DIG = register("dig");
     private final String name;
     private final int hashCode;
+    public final int id; // Canvas - optimize brain
 
-    private Activity(String name) {
+    private Activity(String name, int id) { // Canvas - optimize brain
         this.name = name;
         this.hashCode = name.hashCode();
+        this.id = id; // Canvas - optimize brain
     }
 
     public String getName() {
@@ -43,7 +_,7 @@
     }
 
     private static Activity register(String key) {
-        return Registry.register(BuiltInRegistries.ACTIVITY, key, new Activity(key));
+        return Registry.register(BuiltInRegistries.ACTIVITY, key, new Activity(key, BuiltInRegistries.ACTIVITY.size())); // Canvas - optimize brain
     }
 
     @Override
