--- a/net/minecraft/world/damagesource/CombatTracker.java
+++ b/net/minecraft/world/damagesource/CombatTracker.java
@@ -47,6 +_,11 @@
     public void recordDamageAndCheckCombatState(final CombatEntry combatEntry) {
         final DamageSource source = combatEntry.source();
     // Paper end - Combat tracker API
+        // Canvas start - MC-298464
+        if (io.canvasmc.canvas.Config.INSTANCE.fixes.mc298464) {
+            this.entries.clear();
+        }
+        // Canvas end - MC-298464
         this.entries.add(combatEntry);
         this.lastDamageTime = this.mob.tickCount;
         this.takingDamage = true;
