From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/levelgen/structure/pools/SinglePoolElement.java b/net/minecraft/world/level/levelgen/structure/pools/SinglePoolElement.java
index d6ee47406d65f66c2559242816471add8c3f9f8f..7595320aa93a5effcc820ad420bd1e2b504b993c 100644
--- a/net/minecraft/world/level/levelgen/structure/pools/SinglePoolElement.java
+++ b/net/minecraft/world/level/levelgen/structure/pools/SinglePoolElement.java
@@ -118,8 +118,15 @@ public class SinglePoolElement extends StructurePoolElement {
         StructureTemplateManager structureTemplateManager, BlockPos pos, Rotation rotation, RandomSource random
     ) {
         List<StructureTemplate.JigsawBlockInfo> jigsaws = this.getTemplate(structureTemplateManager).getJigsaws(pos, rotation);
+        // Canvas start - optimize structure gen
+        if (io.canvasmc.canvas.Config.INSTANCE.chunks.structures.optimizeStructureGen) {
+            structureLayoutOptimizer$fasterJigsawListShuffling1(jigsaws, random);
+            structureLayoutOptimizer$fasterJigsawListShuffling2(jigsaws);
+        } else {
         Util.shuffle(jigsaws, random);
         sortBySelectionPriority(jigsaws);
+        }
+        // Canvas end - optimize structure gen
         return jigsaws;
     }
 
@@ -190,6 +197,14 @@ public class SinglePoolElement extends StructurePoolElement {
     public String toString() {
         return "Single[" + this.template + "]";
     }
+    // Canvas start - optimize structure gen
+
+    private void structureLayoutOptimizer$fasterJigsawListShuffling1(List<StructureTemplate.JigsawBlockInfo> list, RandomSource randomSource) {
+        io.canvasmc.canvas.util.structure.GeneralUtils.shuffleAndPrioritize(list, randomSource);
+    }
+
+    private void structureLayoutOptimizer$fasterJigsawListShuffling2(List<StructureTemplate.JigsawBlockInfo> structureBlockInfos) {}
+    // Canvas end - optimize structure gen
 
     @VisibleForTesting
     public Identifier getTemplateLocation() {
