--- a/src/main/java/org/purpurmc/purpur/task/BossBarTask.java
+++ b/src/main/java/org/purpurmc/purpur/task/BossBarTask.java
@@ -1,5 +_,6 @@
 package org.purpurmc.purpur.task;
 
+import io.canvasmc.canvas.Config;
 import net.kyori.adventure.bossbar.BossBar;
 import net.minecraft.server.level.ServerPlayer;
 import org.bukkit.Bukkit;
@@ -91,12 +_,14 @@
     public static void startAll() {
         RamBarTask.instance().start();
         TPSBarTask.instance().start();
+        if (Config.INSTANCE.ticking.enableThreadedRegionizing) TPSBarTask.RegionBarTask.instance().start(); // Canvas - Threaded Regions
         CompassTask.instance().start();
     }
 
     public static void stopAll() {
         RamBarTask.instance().stop();
         TPSBarTask.instance().stop();
+        if (Config.INSTANCE.ticking.enableThreadedRegionizing) TPSBarTask.RegionBarTask.instance().stop(); // Canvas - Threaded Regions
         CompassTask.instance().stop();
     }
 
@@ -107,6 +_,7 @@
         }
         if (player.tpsBar()) {
             TPSBarTask.instance().addPlayer(bukkit);
+            if (Config.INSTANCE.ticking.enableThreadedRegionizing) TPSBarTask.RegionBarTask.instance().addPlayer(bukkit); // Canvas - Threaded Regions
         }
         if (player.compassBar()) {
             CompassTask.instance().addPlayer(bukkit);
@@ -116,6 +_,7 @@
     public static void removeFromAll(Player player) {
         RamBarTask.instance().removePlayer(player);
         TPSBarTask.instance().removePlayer(player);
+        if (Config.INSTANCE.ticking.enableThreadedRegionizing) TPSBarTask.RegionBarTask.instance().removePlayer(player); // Canvas - Threaded Regions
         CompassTask.instance().removePlayer(player);
     }
 }
