--- a/net/minecraft/server/Bootstrap.java
+++ b/net/minecraft/server/Bootstrap.java
@@ -9,6 +_,7 @@
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.function.Function;
 import java.util.function.Supplier;
+import io.canvasmc.canvas.CanvasBootstrap;
 import net.minecraft.SharedConstants;
 import net.minecraft.SuppressForbidden;
 import net.minecraft.commands.Commands;
@@ -187,7 +_,7 @@
 
     public static void validate() {
         checkBootstrapCalled(() -> "validate");
-        if (SharedConstants.IS_RUNNING_IN_IDE) {
+        if (SharedConstants.IS_RUNNING_IN_IDE || CanvasBootstrap.RUNNING_IN_IDE) { // Canvas - ide
             getMissingTranslations().forEach(string -> LOGGER.error("Missing translations: {}", string));
             Commands.validate();
         }
