--- a/net/minecraft/server/Eula.java
+++ b/net/minecraft/server/Eula.java
@@ -6,6 +_,7 @@
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.Properties;
+import io.canvasmc.canvas.CanvasBootstrap;
 import net.minecraft.SharedConstants;
 import net.minecraft.util.CommonLinks;
 import org.slf4j.Logger;
@@ -17,7 +_,7 @@
 
     public Eula(Path file) {
         this.file = file;
-        this.agreed = SharedConstants.IS_RUNNING_IN_IDE || this.readFile();
+        this.agreed = SharedConstants.IS_RUNNING_IN_IDE || CanvasBootstrap.RUNNING_IN_IDE || this.readFile(); // Canvas - ide
     }
 
     private boolean readFile() {
