From fa0385f66589f695ee335f5beee4acd758b26a48 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/server/Main.java b/net/minecraft/server/Main.java
index cf6152c5eaadf2586b00cad19b56401372f8bde0..7a76d61b789e6bd0cdd4dee798e4b205af5fa02e 100644
--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -66,6 +66,20 @@ public class Main {
     public static void main(final OptionSet options) { // CraftBukkit - replaces main(String[] args)
         io.papermc.paper.log.LogManagerShutdownThread.hook(); // Paper - Improved watchdog support
         SharedConstants.tryDetectVersion();
+        // Canvas start - validate EULA early
+        Path eulaFile = Paths.get("eula.txt");
+        Eula eula = new Eula(eulaFile);
+        boolean eulaAgreed = Boolean.getBoolean("com.mojang.eula.agree");
+        if (eulaAgreed) {
+            LOGGER.error("You have used the Spigot command line EULA agreement flag.");
+            LOGGER.error("By using this setting you are indicating your agreement to Mojang's EULA (https://aka.ms/MinecraftEULA).");
+            LOGGER.error("If you do not agree to the above EULA please stop your server and remove this flag immediately.");
+        }
+        if (!eula.hasAgreedToEULA() && !eulaAgreed) {
+            LOGGER.info("You need to agree to the EULA in order to run the server. Go to eula.txt for more info.");
+            return;
+        }
+        // Canvas end - validate EULA early
         /* CraftBukkit start - Replace everything
         OptionParser parser = new OptionParser();
         OptionSpec<Void> nogui = parser.accepts("nogui");
@@ -118,8 +132,7 @@ public class Main {
             }
             // Paper end
             RegionFileVersion.configure(settings.getProperties().regionFileComression);
-            Path eulaFile = Paths.get("eula.txt");
-            Eula eula = new Eula(eulaFile);
+            // Canvas - validate EULA early - move up
             // Paper start - load config files early for access below if needed
             org.bukkit.configuration.file.YamlConfiguration bukkitConfiguration = io.papermc.paper.configuration.PaperConfigurations.loadLegacyConfigFile((File) options.valueOf("bukkit-settings"));
             org.bukkit.configuration.file.YamlConfiguration spigotConfiguration = io.papermc.paper.configuration.PaperConfigurations.loadLegacyConfigFile((File) options.valueOf("spigot-settings"));
@@ -142,18 +155,7 @@ public class Main {
                 return;
             }
 
-            // Spigot start
-            boolean eulaAgreed = Boolean.getBoolean("com.mojang.eula.agree");
-            if (eulaAgreed) {
-                LOGGER.error("You have used the Spigot command line EULA agreement flag.");
-                LOGGER.error("By using this setting you are indicating your agreement to Mojang's EULA (https://aka.ms/MinecraftEULA).");
-                LOGGER.error("If you do not agree to the above EULA please stop your server and remove this flag immediately.");
-            }
-            if (!eula.hasAgreedToEULA() && !eulaAgreed) {
-                // Spigot end
-                LOGGER.info("You need to agree to the EULA in order to run the server. Go to eula.txt for more info.");
-                return;
-            }
+            // Canvas - validate EULA early - move up
 
             // Paper start - Detect headless JRE
             String awtException = io.papermc.paper.util.ServerEnvironment.awtDependencyCheck();
