--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -288,6 +_,7 @@
     public boolean isIteratingOverLevels = false; // Paper - Throw exception on world create while being ticked
     private final Set<String> pluginsBlockingSleep = new java.util.HashSet<>(); // Paper - API to allow/disallow tick sleeping
     public static final long SERVER_INIT = System.nanoTime(); // Paper - Lag compensation
+    public final io.canvasmc.canvas.world.entity.EnderPearls pearls = io.canvasmc.canvas.world.entity.EnderPearls.read(); // Canvas - region threading
     // Paper start - improve tick loop
     public final ca.spottedleaf.moonrise.common.time.TickData tickTimes1s  = new ca.spottedleaf.moonrise.common.time.TickData(java.util.concurrent.TimeUnit.SECONDS.toNanos(1L));
     public final ca.spottedleaf.moonrise.common.time.TickData tickTimes5s  = new ca.spottedleaf.moonrise.common.time.TickData(java.util.concurrent.TimeUnit.SECONDS.toNanos(5L));
@@ -735,6 +_,7 @@
             );
         }
         this.addLevel(serverLevel);
+        LOGGER.info("Using world distance config of {} for world '{}'", serverLevel.serverLevelData.canvas$distanceConfig, serverLevel); // Canvas - per world distance
         this.initWorld(serverLevel, serverLevelData, worldOptions);
     }
     public void initWorld(ServerLevel serverLevel, net.minecraft.world.level.storage.PrimaryLevelData serverLevelData, WorldOptions worldOptions) {
@@ -1065,15 +_,7 @@
         }
         // CraftBukkit end
         if (io.papermc.paper.plugin.PluginInitializerManager.instance().pluginRemapper != null) io.papermc.paper.plugin.PluginInitializerManager.instance().pluginRemapper.shutdown(); // Paper - Plugin remapping
-        this.getConnection().stop();
-        this.isSaving = true;
-        if (this.playerList != null) {
-            //LOGGER.info("Saving players"); // Folia - move to shutdown thread logic
-            //this.playerList.saveAll(); // Folia - move to shutdown thread logic
-            this.playerList.removeAll(this.isRestarting); // Paper
-            try { Thread.sleep(100); } catch (InterruptedException ex) {} // CraftBukkit - SPIGOT-625 - give server at least a chance to send packets
-        }
-
+        this.isSaving = true; // Canvas - improve player removal logic
         // Folia start - region threading
         if (true) {
             // the rest till part 2 is handled by the region shutdown thread
@@ -1327,11 +_,7 @@
                 LOGGER.info("*************************************************************************************");
             }
             // Paper end - Add onboarding message for initial server start
-            // Paper start - Improve outdated version checking
-            if (System.getProperty("paper.disableStartupVersionCheck") == null && io.papermc.paper.configuration.GlobalConfiguration.get().updateChecker.enabled) {
-                CompletableFuture.runAsync(com.destroystokyo.paper.PaperVersionFetcher::getUpdateStatusStartupMessage, io.papermc.paper.util.MCUtil.ASYNC_EXECUTOR);
-            }
-            // Paper end - Improve outdated version checking
+            // Canvas - Rebrand
 
             while (this.running) {
                 final long tickStart = System.nanoTime(); // Paper - improve tick loop
@@ -1732,6 +_,7 @@
         // Folia end - region threading
         //this.tickCount++; // Folia - region threading
         //this.tickRateManager.tick(); // Folia - region threading
+        region.world.getCurrentWorldData().tpsbar.tick(); // Canvas - tpsbar
         this.tickChildren(hasTimeLeft, region); // Folia - region threading
         if (false && nanos - this.lastServerStatus >= STATUS_EXPIRE_TIME_NANOS) { // Folia - region threading
             this.lastServerStatus = nanos;
@@ -1857,6 +_,21 @@
         //io.papermc.paper.adventure.providers.ClickCallbackProviderImpl.DIALOG_CLICK_MANAGER.handleQueue(this.tickCount); // Paper // Folia - region threading - moved to global tick
         //this.getFunctions().tick(); // Folia - region threading - TODO Purge functions
         this.updateEffectiveRespawnData();
+        // Canvas start - async save-all command
+        region.canvas$saveAllTicket.consumeIfPresent(ticket -> {
+            try {
+                ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)region.world).moonrise$getChunkTaskScheduler().chunkHolderManager.saveAllChunksRegionised(
+                    ticket.flush(), false, false, false, false, true, false
+                );
+                this.getPlayerList().saveAll();
+                LOGGER.info("Saved chunks in region around chunk {} in world '{}'", region.region.getCenterChunk(), region.region.regioniser.world.getWorld().getName());
+            } catch (Throwable thrown) {
+                ticket.exceptionPropagator().accept(thrown);
+            } finally {
+                ticket.callback().run();
+            }
+        });
+        // Canvas end - async save-all command
 
         // CraftBukkit start
         // Run tasks that are waiting on processing
@@ -2015,7 +_,7 @@
 
     @DontObfuscate
     public String getServerModName() {
-        return io.papermc.paper.ServerBuildInfo.buildInfo().brandName(); // Paper
+        return io.canvasmc.canvas.Config.INSTANCE.serverModName; // Canvas - Use configurable server branding
     }
 
     public SystemReport fillSystemReport(SystemReport systemReport) {
@@ -2315,6 +_,13 @@
         return 256;
     }
 
+    // Canvas start - no chat reports
+    public boolean canvas$ncrEnforceSecureProfile() {
+        if (io.canvasmc.canvas.Config.INSTANCE.enableNoChatReports) return false; // NCR doesn't require secure profile
+        return this.enforceSecureProfile();
+    }
+
+    // Canvas end - no chat reports
     public boolean enforceSecureProfile() {
         return false;
     }
