--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -736,6 +_,7 @@
             );
         }
         this.addLevel(serverLevel);
+        LOGGER.info("Using world distance config of {} for world '{}'", serverLevel.serverLevelData.canvas$distanceConfig, serverLevel); // Canvas - per world distance
         this.initWorld(serverLevel, serverLevelData, worldOptions);
     }
     public void initWorld(ServerLevel serverLevel, net.minecraft.world.level.storage.PrimaryLevelData serverLevelData, WorldOptions worldOptions) {
@@ -1319,11 +_,7 @@
                 LOGGER.info("*************************************************************************************");
             }
             // Paper end - Add onboarding message for initial server start
-            // Paper start - Improve outdated version checking
-            if (System.getProperty("paper.disableStartupVersionCheck") == null && io.papermc.paper.configuration.GlobalConfiguration.get().updateChecker.enabled) {
-                CompletableFuture.runAsync(com.destroystokyo.paper.PaperVersionFetcher::getUpdateStatusStartupMessage, io.papermc.paper.util.MCUtil.ASYNC_EXECUTOR);
-            }
-            // Paper end - Improve outdated version checking
+            // Canvas - Rebrand
 
             while (this.running) {
                 final long tickStart = System.nanoTime(); // Paper - improve tick loop
@@ -1873,12 +_,29 @@
         //io.papermc.paper.adventure.providers.ClickCallbackProviderImpl.DIALOG_CLICK_MANAGER.handleQueue(this.tickCount); // Paper // Folia - region threading - moved to global tick
         //this.getFunctions().tick(); // Folia - region threading - TODO Purge functions
         this.updateEffectiveRespawnData();
+        // Canvas start - async save-all command
+        region.canvas$saveAllTicket.consumeIfPresent(ticket -> {
+            try {
+                ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)region.world).moonrise$getChunkTaskScheduler().chunkHolderManager.saveAllChunksRegionised(
+                    ticket.flush(), false, false, false, false, true, false
+                );
+                this.getPlayerList().saveAll();
+                LOGGER.info("Saved chunks in region around chunk {} in world '{}'", region.region.getCenterChunk(), region.region.regioniser.world.getWorld().getName());
+            } catch (Throwable thrown) {
+                ticket.exceptionPropagator().accept(thrown);
+            } finally {
+                ticket.callback().run();
+            }
+        });
+        // Canvas end - async save-all command
 
         // CraftBukkit start
         // Run tasks that are waiting on processing
         if (false) while (!this.processQueue.isEmpty()) { // Folia - region threading
             this.processQueue.remove().run();
         }
+        region.world.getCurrentWorldData().tpsbar.tick(); // Canvas - tpsbar
+        region.world.getCurrentWorldData().rambar.tick(regionizedWorldData.getLocalPlayers()); // Canvas - rambar
 
         // Send time updates to everyone, it will get the right time from the world the player is in.
         // Paper start - Perf: Optimize time updates
@@ -2031,7 +_,7 @@
 
     @DontObfuscate
     public String getServerModName() {
-        return io.papermc.paper.ServerBuildInfo.buildInfo().brandName(); // Paper
+        return io.canvasmc.canvas.Config.INSTANCE.serverModName; // Canvas - Use configurable server branding
     }
 
     public SystemReport fillSystemReport(SystemReport systemReport) {
@@ -2331,6 +_,13 @@
         return 256;
     }
 
+    // Canvas start - no chat reports
+    public boolean canvas$ncrEnforceSecureProfile() {
+        if (io.canvasmc.canvas.Config.INSTANCE.enableNoChatReports) return false; // NCR doesn't require secure profile
+        return this.enforceSecureProfile();
+    }
+
+    // Canvas end - no chat reports
     public boolean enforceSecureProfile() {
         return false;
     }
