--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -787,6 +_,7 @@
             );
         }
         this.addLevel(serverLevel);
+        LOGGER.info("Using world distance config of {} for world '{}'", serverLevel.serverLevelData.canvas$distanceConfig, serverLevel); // Canvas - per world distance
         this.initWorld(serverLevel);
     }
     public void initWorld(ServerLevel overworld) {
@@ -1359,11 +_,7 @@
                 LOGGER.info("*************************************************************************************");
             }
             // Paper end - Add onboarding message for initial server start
-            // Paper start - Improve outdated version checking
-            if (System.getProperty("paper.disableStartupVersionCheck") == null && io.papermc.paper.configuration.GlobalConfiguration.get().updateChecker.enabled) {
-                CompletableFuture.runAsync(com.destroystokyo.paper.PaperVersionFetcher::getUpdateStatusStartupMessage, io.papermc.paper.util.MCUtil.ASYNC_EXECUTOR);
-            }
-            // Paper end - Improve outdated version checking
+            // Canvas - Rebrand
 
             while (this.running) {
                 final long tickStart = System.nanoTime(); // Paper - improve tick loop
@@ -1927,8 +_,24 @@
         if (false) while (!this.processQueue.isEmpty()) { // Folia - region threading
             this.processQueue.remove().run();
         }
+        region.world.getCurrentWorldData().tpsbar.tick(); // Canvas - tpsbar
 
         this.updateEffectiveRespawnData();
+        // Canvas start - region threading - save all command
+        region.canvas$saveAllTicket.consumeIfPresent(ticket -> {
+            try {
+                ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)region.world).moonrise$getChunkTaskScheduler().chunkHolderManager.saveAllChunksRegionised(
+                    ticket.flush(), false, false, false, false, true, false
+                );
+                this.getPlayerList().saveAll();
+                LOGGER.info("Saved chunks in region around chunk {} in world '{}'", region.region.getCenterChunk(), region.region.regioniser.world.getWorld().getName());
+            } catch (Throwable thrown) {
+                ticket.exceptionPropagator().accept(thrown);
+            } finally {
+                ticket.callback().run();
+            }
+        });
+        // Canvas end - region threading - save all command
 
         //this.isIteratingOverLevels = true; // Paper - Throw exception on world create while being ticked // Folia - region threading
         if (true) { ServerLevel level = region.world; // Folia - region threading
@@ -2048,7 +_,7 @@
     }
 
     public String getServerModName() {
-        return io.papermc.paper.ServerBuildInfo.buildInfo().brandName(); // Paper
+        return io.canvasmc.canvas.Config.INSTANCE.serverModName; // Canvas - Use configurable server branding
     }
 
     public ServerClockManager clockManager() {
@@ -2356,6 +_,13 @@
         return 256;
     }
 
+    // Canvas start - no chat reports
+    public boolean canvas$ncrEnforceSecureProfile() {
+        if (io.canvasmc.canvas.Config.INSTANCE.enableNoChatReports) return false; // NCR doesn't require secure profile
+        return this.enforceSecureProfile();
+    }
+
+    // Canvas end - no chat reports
     public boolean enforceSecureProfile() {
         return false;
     }
