From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/core/NonNullList.java b/net/minecraft/core/NonNullList.java
index ede0a5403739a2cf8bf101d7b12748c8acce87f5..1bcd9ea7e57e641193f49c39351e39b408aeb361 100644
--- a/net/minecraft/core/NonNullList.java
+++ b/net/minecraft/core/NonNullList.java
@@ -12,23 +12,23 @@ public class NonNullList<E> extends AbstractList<E> {
     private final @Nullable E defaultValue;
 
     public static <E> NonNullList<E> create() {
-        return new NonNullList<>(Lists.newArrayList(), null);
+        return new NonNullList<>(new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(), null); // Canvas - optimized collection
     }
 
     public static <E> NonNullList<E> createWithCapacity(int initialCapacity) {
-        return new NonNullList<>(Lists.newArrayListWithCapacity(initialCapacity), null);
+        return new NonNullList<>(new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(initialCapacity), null); // Canvas - optimized collection
     }
 
     public static <E> NonNullList<E> withSize(int size, E defaultValue) {
         Objects.requireNonNull(defaultValue);
         Object[] objects = new Object[size];
         Arrays.fill(objects, defaultValue);
-        return new NonNullList<>(Arrays.asList((E[])objects), defaultValue);
+        return new NonNullList<>(new it.unimi.dsi.fastutil.objects.ObjectArrayList<>((E[])objects), defaultValue); // Canvas - optimized collection
     }
 
     @SafeVarargs
     public static <E> NonNullList<E> of(E defaultValue, E... elements) {
-        return new NonNullList<>(Arrays.asList(elements), defaultValue);
+        return new NonNullList<>(new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(elements), defaultValue); // Canvas - optimized collection
     }
 
     protected NonNullList(List<E> list, @Nullable E defaultValue) {
