--- a/net/minecraft/core/dispenser/DispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/DispenseItemBehavior.java
@@ -500,7 +_,7 @@
                     }
                 }
 
-                level.captureTreeGeneration = true;
+                level.captureTreeGeneration.set(true); // Canvas - atomic
                 // CraftBukkit end
                 if (!BoneMealItem.growCrop(item, level, blockPos) && !BoneMealItem.growWaterPlant(item, level, blockPos, null)) {
                     this.setSuccess(false);
@@ -508,7 +_,7 @@
                     level.levelEvent(1505, blockPos, 15);
                 }
                 // CraftBukkit start
-                level.captureTreeGeneration = false;
+                level.captureTreeGeneration.set(false); // Canvas - atomic
                 if (level.capturedBlockStates.size() > 0) {
                     org.bukkit.TreeType treeType = net.minecraft.world.level.block.SaplingBlock.treeType;
                     net.minecraft.world.level.block.SaplingBlock.treeType = null;
