From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/behavior/AcquirePoi.java b/net/minecraft/world/entity/ai/behavior/AcquirePoi.java
index 2940dbc31a20eb49e24418ec84b73f2bfbdcfc32..74f9c5927fc06ee0d8f150c2d11f07a528e8035d 100644
--- a/net/minecraft/world/entity/ai/behavior/AcquirePoi.java
+++ b/net/minecraft/world/entity/ai/behavior/AcquirePoi.java
@@ -54,6 +54,21 @@ public class AcquirePoi {
         final Optional<Byte> onPoiAcquisitionEvent,
         final BiPredicate<ServerLevel, BlockPos> validPoi
     ) {
+    // Canvas start - configurable scan range and if poi should load
+        return create(poiType, memoryToValidate, memoryToAcquire, onlyIfAdult, onPoiAcquisitionEvent, validPoi, (_) -> SCAN_RANGE, (_) -> ca.spottedleaf.moonrise.patches.poi_lookup.PoiAccess.LOAD_FOR_SEARCHING);
+    }
+
+    public static BehaviorControl<PathfinderMob> create(
+        final Predicate<Holder<PoiType>> poiType,
+        final MemoryModuleType<GlobalPos> memoryToValidate,
+        final MemoryModuleType<GlobalPos> memoryToAcquire,
+        final boolean onlyIfAdult,
+        final Optional<Byte> onPoiAcquisitionEvent,
+        final BiPredicate<ServerLevel, BlockPos> validPoi,
+        final java.util.function.Function<Mob, Integer> scanRange,
+        final java.util.function.Function<Mob, Boolean> shouldLoad
+    ) {
+    // Canvas end - configurable scan range and if poi should load
         int batchSize = 5;
         int rate = 20;
         MutableLong nextScheduledStart = new MutableLong(0L);
@@ -96,7 +111,7 @@ public class AcquirePoi {
                         };
                         // Paper start - optimise POI searches
                         java.util.List<Pair<Holder<PoiType>, BlockPos>> poiPositionsRaw = new java.util.ArrayList<>();
-                        ca.spottedleaf.moonrise.patches.poi_lookup.PoiAccess.findNearestPoiPositions(poiManager, poiType, cacheTest, body.blockPosition(), SCAN_RANGE, Double.MAX_VALUE, PoiManager.Occupancy.HAS_SPACE, ca.spottedleaf.moonrise.patches.poi_lookup.PoiAccess.LOAD_FOR_SEARCHING, 5, poiPositionsRaw);
+                        ca.spottedleaf.moonrise.patches.poi_lookup.PoiAccess.findNearestPoiPositions(poiManager, poiType, cacheTest, body.blockPosition(), scanRange.apply(body), Double.MAX_VALUE, PoiManager.Occupancy.HAS_SPACE, shouldLoad.apply(body), 5, poiPositionsRaw); // Canvas - configurable scan range and if poi should load
                         Set<Pair<Holder<PoiType>, BlockPos>> poiPositions = new java.util.HashSet<>(poiPositionsRaw.size());
                         for (Pair<Holder<PoiType>, BlockPos> pair : poiPositionsRaw) {
                             if (validPoi.test(level, pair.getSecond())) {
