From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java b/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java
index 5bd7473b0db1a87f905c01c19ddf99e3a447bcfa..583942e37b4b12c274904a20f6c84dc7252e5268 100644
--- a/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java
+++ b/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java
@@ -58,11 +58,15 @@ public class SetClosestHomeAsWalkTarget {
                                 batchCache.put(key, lastUpdate.longValue() + CACHE_TIMEOUT);
                                 return true;
                             };
-                            Set<Pair<Holder<PoiType>, BlockPos>> pois = poiManager.findAllWithType(
-                                    p -> p.is(PoiTypes.HOME), cacheTest, body.blockPosition(), AcquirePoi.SCAN_RANGE, PoiManager.Occupancy.ANY
-                                )
-                                .collect(Collectors.toSet());
-                            Path path = AcquirePoi.findPathToPois(body, pois);
+                            // Canvas start - reroute to paper poi logic
+                            java.util.List<Pair<Holder<PoiType>, BlockPos>> poips = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
+                            ca.spottedleaf.moonrise.patches.poi_lookup.PoiAccess.findAnyPoiPositions(
+                                poiManager, p -> p.is(PoiTypes.HOME), cacheTest,
+                                body.blockPosition(), net.minecraft.world.entity.ai.behavior.AcquirePoi.SCAN_RANGE, // 48
+                                PoiManager.Occupancy.ANY, false, Integer.MAX_VALUE, poips
+                            );
+                            Path path = AcquirePoi.findPathToPois(body, new java.util.HashSet<>(poips));
+                            // Canvas end - reroute to paper poi logic
                             if (path != null && path.canReach()) {
                                 BlockPos targetPos = path.getTarget();
                                 Optional<Holder<PoiType>> type = poiManager.getType(targetPos);
