--- a/net/minecraft/util/CubicSpline.java
+++ b/net/minecraft/util/CubicSpline.java
@@ -270,8 +_,49 @@
         }
 
         private static int findIntervalStart(final float[] locations, final float input) {
-            return Mth.binarySearch(0, locations.length, i -> input < locations[i]) - 1;
+            // Leaf start - C2ME - DensityFunction compiler
+            int min = 0;
+            int length = locations.length;
+
+            while (length > 0) {
+                int half = length / 2;
+                int middle = min + half;
+                if (input < locations[middle]) {
+                    length = half;
+                } else {
+                    min = middle + 1;
+                    length -= half + 1;
+                }
+            }
+
+            return min - 1;
+            // Leaf end - C2ME - DensityFunction compiler
+        }
+
+        // Leaf start - C2ME - DensityFunction compiler
+        @Override
+        public boolean equals(final Object object) {
+            if (this == object) return true;
+            if (object == null || this.getClass() != object.getClass()) return false;
+            CubicSpline.Multipoint<?> that = (CubicSpline.Multipoint<?>)object;
+            return java.util.Objects.equals(this.coordinate, that.coordinate())
+                && java.util.Arrays.equals(this.locations, that.locations())
+                && java.util.Objects.equals(this.values, that.values())
+                && java.util.Arrays.equals(this.derivatives, that.derivatives());
+        }
+
+        @Override
+        public int hashCode() {
+            int result = 1;
+
+            result = 31 * result + java.util.Objects.hashCode(this.coordinate);
+            result = 31 * result + java.util.Arrays.hashCode(this.locations);
+            result = 31 * result + java.util.Objects.hashCode(this.values);
+            result = 31 * result + java.util.Arrays.hashCode(this.derivatives);
+
+            return result;
         }
+        // Leaf end - C2ME - DensityFunction compiler
 
         @VisibleForTesting
         @Override
