--- a/net/minecraft/util/Mth.java
+++ b/net/minecraft/util/Mth.java
@@ -175,6 +_,18 @@
         return (input % mod + mod) % mod;
     }
 
+    // Canvas start - faster floating-point positive modulo
+    public static float positiveModuloForPositiveIntegerDenominator(final float input, final float mod) {
+        float modulo = input % mod;
+        return modulo < 0 ? modulo + mod : modulo;
+    }
+
+    public static double positiveModuloForPositiveIntegerDenominator(final double input, final double mod) {
+        double modulo = input % mod;
+        return modulo < 0 ? modulo + mod : modulo;
+    }
+    // Canvas end - faster floating-point positive modulo
+
     public static boolean isMultipleOf(final int dividend, final int divisor) {
         return dividend % divisor == 0;
     }
