--- a/net/minecraft/util/Mth.java
+++ b/net/minecraft/util/Mth.java
@@ -4,6 +_,8 @@
 import java.util.UUID;
 import java.util.function.IntPredicate;
 import java.util.stream.IntStream;
+import io.canvasmc.canvas.Config;
+import io.canvasmc.canvas.util.CompactSineLUT;
 import net.minecraft.Util;
 import net.minecraft.core.Vec3i;
 import net.minecraft.world.phys.AABB;
@@ -29,7 +_,7 @@
     public static final Vector3f Y_AXIS = new Vector3f(0.0F, 1.0F, 0.0F);
     public static final Vector3f X_AXIS = new Vector3f(1.0F, 0.0F, 0.0F);
     public static final Vector3f Z_AXIS = new Vector3f(0.0F, 0.0F, 1.0F);
-    private static final float[] SIN = Util.make(new float[65536], floats -> {
+    public static final float[] SIN = Util.make(new float[65536], floats -> { // Canvas - private -> public
         for (int i1 = 0; i1 < floats.length; i1++) {
             floats[i1] = (float)Math.sin(i1 * Math.PI * 2.0 / 65536.0);
         }
@@ -45,11 +_,23 @@
     private static final double[] ASIN_TAB = new double[257];
     private static final double[] COS_TAB = new double[257];
 
+    // Canvas start
+    static {
+        CompactSineLUT.init();
+    }
+
     public static float sin(float value) {
+        if (Config.INSTANCE.useCompactSineLUT) {
+            return CompactSineLUT.sin(value);
+        }
         return SIN[(int)(value * 10430.378F) & 65535];
     }
 
     public static float cos(float value) {
+        if (Config.INSTANCE.useCompactSineLUT) {
+            return CompactSineLUT.cos(value);
+        }
+    // Canvas end
         return SIN[(int)(value * 10430.378F + 16384.0F) & 65535];
     }
 
