--- a/net/minecraft/util/RandomSource.java
+++ b/net/minecraft/util/RandomSource.java
@@ -12,7 +_,7 @@
     double GAUSSIAN_SPREAD_FACTOR = 2.297;
 
     static RandomSource create() {
-        return create(RandomSupport.generateUniqueSeed());
+        return createThreadSafe(); // Canvas - threadsafe
     }
 
     @Deprecated
@@ -21,7 +_,7 @@
     }
 
     static RandomSource create(long seed) {
-        return new LegacyRandomSource(seed);
+        return new ThreadSafeLegacyRandomSource(seed); // Canvas - threadsafe
     }
 
     static RandomSource createNewThreadLocalInstance() {
