From 2be0fbf099df19422462cdb5b961fc00062d256b Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java b/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
index f1cf029b98f1b927cf282a87bbdb0d59b30033ea..0141fc0df4ea025ed068a82dddcfff6a9da56a09 100644
--- a/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
+++ b/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
@@ -503,13 +503,20 @@ public class WalkNodeEvaluator extends NodeEvaluator {
         return blockPathType;
     }
 
-    protected static PathType getPathTypeFromState(final BlockGetter level, final BlockPos pos) {
+    // Canvas start - optimize path type cache
+    private static final PathType[] PATH_TYPES = PathType.values();
+    public static PathType getPathTypeFromState(final BlockGetter level, final BlockPos pos) {
         // Paper start - Do not load chunks during pathfinding
         BlockState blockState = level.getBlockStateIfLoaded(pos);
         if (blockState == null) {
             return PathType.BLOCKED;
         }
         // Paper end
+        return PATH_TYPES[blockState.pathType];
+    }
+
+    public static PathType canvas$getPathTypeFromBlockState(BlockState blockState) {
+    // Canvas end - optimize path type cache
         Block block = blockState.getBlock();
         if (blockState.isAir()) {
             return PathType.OPEN;
