--- a/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
+++ b/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
@@ -513,6 +_,8 @@
         return blockPathType;
     }
 
+    // Canvas start - optimize path type cache
+    private static final PathType[] PATH_TYPES = PathType.values();
     protected static PathType getPathTypeFromState(final BlockGetter level, final BlockPos pos) {
         // Paper start - Do not load chunks during pathfinding
         BlockState blockState = level.getBlockStateIfLoaded(pos);
@@ -520,6 +_,11 @@
             return PathType.BLOCKED;
         }
         // Paper end
+        return PATH_TYPES[blockState.pathType];
+    }
+
+    public static PathType canvas$getPathTypeFromBlockState(final BlockState blockState) {
+    // Canvas end - optimize path type cache
         Block block = blockState.getBlock();
         if (blockState.isAir()) {
             return PathType.OPEN;
