--- a/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
@@ -4,8 +_,12 @@
 import com.google.common.collect.Lists;
 import com.google.common.collect.Maps;
 import com.mojang.datafixers.util.Pair;
+import io.canvasmc.canvas.Config;
+import io.canvasmc.canvas.util.structure.PalettedStructureBlockInfoList;
+import io.canvasmc.canvas.util.structure.StructureTemplateOptimizer;
 import it.unimi.dsi.fastutil.objects.ObjectArrayList;
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.Comparator;
 import java.util.Iterator;
 import java.util.List;
@@ -71,8 +_,8 @@
     public static final String ENTITY_TAG_BLOCKPOS = "blockPos";
     public static final String ENTITY_TAG_NBT = "nbt";
     public static final String SIZE_TAG = "size";
-    public final List<StructureTemplate.Palette> palettes = Lists.newArrayList();
-    public final List<StructureTemplate.StructureEntityInfo> entityInfoList = Lists.newArrayList();
+    public final List<StructureTemplate.Palette> palettes = Collections.synchronizedList(Lists.newArrayList()); // Canvas - synchronize
+    public final List<StructureTemplate.StructureEntityInfo> entityInfoList = Collections.synchronizedList(Lists.newArrayList()); // Canvas - synchronize
     private Vec3i size = Vec3i.ZERO;
     private String author = "?";
     // CraftBukkit start - data containers
@@ -248,6 +_,12 @@
     public static BlockPos calculateRelativePosition(StructurePlaceSettings decorator, BlockPos pos) {
         return transform(pos, decorator.getMirror(), decorator.getRotation(), decorator.getRotationPivot());
     }
+    // Canvas start - StructureLayoutOptimizer
+
+    private List<StructureTemplate.StructureBlockInfo> structureLayoutOptimizer$shrinkStructureTemplateBlocksList(StructureTemplate.Palette palette, BlockPos offset, StructurePlaceSettings settings) {
+        return StructureTemplateOptimizer.getStructureBlockInfosInBounds(palette, offset, settings);
+    }
+    // Canvas end
 
     public boolean placeInWorld(ServerLevelAccessor serverLevel, BlockPos offset, BlockPos pos, StructurePlaceSettings settings, RandomSource random, int flags) {
         if (this.palettes.isEmpty()) {
@@ -266,7 +_,7 @@
                 }
             }
             // CraftBukkit end
-            List<StructureTemplate.StructureBlockInfo> list = settings.getRandomPalette(this.palettes, offset).blocks();
+            List<StructureTemplate.StructureBlockInfo> list = Config.INSTANCE.enableStructureLayoutOptimizer ? structureLayoutOptimizer$shrinkStructureTemplateBlocksList(settings.getRandomPalette(this.palettes, offset), offset, settings) : settings.getRandomPalette(this.palettes, offset).blocks(); // Canvas - StructureLayoutOptimizer
             if ((!list.isEmpty() || !settings.isIgnoreEntities() && !this.entityInfoList.isEmpty())
                 && this.size.getX() >= 1
                 && this.size.getY() >= 1
@@ -890,7 +_,7 @@
         private List<StructureTemplate.JigsawBlockInfo> cachedJigsaws;
 
         Palette(List<StructureTemplate.StructureBlockInfo> blocks) {
-            this.blocks = blocks;
+            this.blocks = Config.INSTANCE.enableStructureLayoutOptimizer ? new PalettedStructureBlockInfoList(blocks) : blocks; // Canvas - StructureLayoutOptimizer
         }
 
         public List<StructureTemplate.JigsawBlockInfo> jigsaws() {
