--- a/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
@@ -252,6 +_,12 @@
     public static BlockPos calculateRelativePosition(final StructurePlaceSettings settings, final BlockPos pos) {
         return transform(pos, settings.getMirror(), settings.getRotation(), settings.getRotationPivot());
     }
+    // Canvas start - optimize structure gen
+
+    private List<StructureTemplate.StructureBlockInfo> structureLayoutOptimizer$shrinkStructureTemplateBlocksList(final StructureTemplate.Palette palette, final BlockPos position, final StructurePlaceSettings settings) {
+        return io.canvasmc.canvas.util.structure.StructureTemplateOptimizer.getStructureBlockInfosInBounds(palette, position, settings);
+    }
+    // Canvas end - optimize structure gen
 
     public boolean placeInWorld(
         ServerLevelAccessor level, // Paper - make non-final
@@ -278,7 +_,7 @@
             }
         }
         // CraftBukkit end
-        List<StructureTemplate.StructureBlockInfo> blockInfoList = settings.getRandomPalette(this.palettes, position).blocks();
+        List<StructureTemplate.StructureBlockInfo> blockInfoList = io.canvasmc.canvas.GlobalConfiguration.getInstance().chunkSystem.structureOptimizations.enable ? structureLayoutOptimizer$shrinkStructureTemplateBlocksList(settings.getRandomPalette(this.palettes, position), position, settings) : settings.getRandomPalette(this.palettes, position).blocks(); // Canvas - optimize structure gen
         if ((!blockInfoList.isEmpty() || !settings.isIgnoreEntities() && !this.entityInfoList.isEmpty())
             && this.size.getX() >= 1
             && this.size.getY() >= 1
@@ -876,7 +_,7 @@
         private @Nullable List<StructureTemplate.JigsawBlockInfo> cachedJigsaws;
 
         private Palette(final List<StructureTemplate.StructureBlockInfo> blocks) {
-            this.blocks = blocks;
+            this.blocks = io.canvasmc.canvas.GlobalConfiguration.getInstance().chunkSystem.structureOptimizations.enable ? new io.canvasmc.canvas.util.structure.PalettedStructureBlockInfoList(blocks) : blocks; // Canvas - optimize structure gen
         }
 
         public List<StructureTemplate.JigsawBlockInfo> jigsaws() {
