--- a/net/minecraft/world/level/levelgen/structure/templatesystem/StructurePlaceSettings.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/StructurePlaceSettings.java
@@ -1,6 +_,7 @@
 package net.minecraft.world.level.levelgen.structure.templatesystem;
 
 import com.google.common.collect.Lists;
+import java.util.Collections;
 import java.util.List;
 import javax.annotation.Nullable;
 import net.minecraft.Util;
@@ -22,7 +_,7 @@
     @Nullable
     private RandomSource random;
     public int palette = -1; // CraftBukkit - Set initial value so we know if the palette has been set forcefully
-    private final List<StructureProcessor> processors = Lists.newArrayList();
+    private final List<StructureProcessor> processors = Collections.synchronizedList(Lists.newArrayList()); // Canvas - synchronize
     private boolean knownShape;
     private boolean finalizeEntities;
 
