From 53e1a57c72953de1bfcc94c3257c724e7fa1f88f Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/levelgen/structure/pools/SinglePoolElement.java b/net/minecraft/world/level/levelgen/structure/pools/SinglePoolElement.java
index 80319e8548bab23827667a91072cd751006632d5..15e3846fc7827c6a3d300432a8f65539f8511453 100644
--- a/net/minecraft/world/level/levelgen/structure/pools/SinglePoolElement.java
+++ b/net/minecraft/world/level/levelgen/structure/pools/SinglePoolElement.java
@@ -117,8 +117,15 @@ public class SinglePoolElement extends StructurePoolElement {
         final StructureTemplateManager structureTemplateManager, final BlockPos position, final Rotation rotation, final RandomSource random
     ) {
         List<StructureTemplate.JigsawBlockInfo> jigsaws = this.getTemplate(structureTemplateManager).getJigsaws(position, rotation);
+        // Canvas start - optimize structure gen
+        if (io.canvasmc.canvas.GlobalConfiguration.getInstance().chunkSystem.structureOptimizations.enable) {
+            structureLayoutOptimizer$fasterJigsawListShuffling1(jigsaws, random);
+            structureLayoutOptimizer$fasterJigsawListShuffling2(jigsaws);
+        } else {
         Util.shuffle(jigsaws, random);
         sortBySelectionPriority(jigsaws);
+        }
+        // Canvas end - optimize structure gen
         return jigsaws;
     }
 
@@ -191,6 +198,14 @@ public class SinglePoolElement extends StructurePoolElement {
     public String toString() {
         return "Single[" + this.template + "]";
     }
+    // Canvas start - optimize structure gen
+
+    private void structureLayoutOptimizer$fasterJigsawListShuffling1(List<StructureTemplate.JigsawBlockInfo> list, RandomSource randomSource) {
+        io.canvasmc.canvas.util.structure.GeneralUtils.shuffleAndPrioritize(list, randomSource);
+    }
+
+    private void structureLayoutOptimizer$fasterJigsawListShuffling2(List<StructureTemplate.JigsawBlockInfo> structureBlockInfos) {}
+    // Canvas end - optimize structure gen
 
     @VisibleForTesting
     public Identifier getTemplateLocation() {
