From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Mon, 13 Apr 2026 21:04:01 -0700
Subject: [PATCH] Aggressive Thread Checks


diff --git a/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java b/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java
index 4b91aca31f1b4a576f68146e8da5646848f758a3..1ea0e29fc82d691114b508fd4157915b4541a18a 100644
--- a/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/util/TickThread.java
@@ -309,4 +309,40 @@ public class TickThread extends Thread {
             return ((entity.hasNullCallback() || entity.isRemoved())) && isTickThreadFor((net.minecraft.server.level.ServerLevel)world, entity.chunkPosition());
         }
     }
+    // Canvas start - thread check more API
+
+    public static void canvas$tryEnsureThreadCheck(final Entity entity, String reason) {
+        final int mode = io.canvasmc.canvas.Config.INSTANCE.extraThreadChecksLevel;
+        if (mode == 0) {
+            return;
+        }
+
+        if (mode == 1) {
+            if (!isTickThreadFor(entity)) {
+                LOGGER.warn("Unsafe access detected for Spigot/Paper API, please contact the related plugin developers to resolve this");
+                LOGGER.warn(reason, new Throwable());
+            }
+            return;
+        }
+
+        ensureTickThread(entity, reason);
+    }
+
+    public static void canvas$tryEnsureThreadCheck(final Level world, final double blockX, final double blockZ, String reason) {
+        final int mode = io.canvasmc.canvas.Config.INSTANCE.extraThreadChecksLevel;
+        if (mode == 0) {
+            return;
+        }
+
+        if (mode == 1) {
+            if (!isTickThreadFor(world, blockX, blockZ)) {
+                LOGGER.warn("Unsafe access detected for Spigot/Paper API, please contact the related plugin developers to resolve this");
+                LOGGER.warn(reason, new Throwable());
+            }
+            return;
+        }
+
+        ensureTickThread(world, blockX, blockZ, reason);
+    }
+    // Canvas end - thread check more API
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index e6cd1c4a50b32b74f70a264e869b1ab4ac9b9eb0..8851cc357ec8fd2ae1d4742508dc9332080497dc 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -180,11 +180,13 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public Location getLocation() {
+        // TODO - Canvas - should we thread check this?
         return CraftLocation.toBukkit(this.entity.position(), this.getWorld(), this.entity.getBukkitYaw(), this.entity.getXRot());
     }
 
     @Override
     public Location getLocation(Location loc) {
+        // TODO - Canvas - should we thread check this?
         if (loc != null) {
             loc.setWorld(this.getWorld());
             loc.setX(this.entity.getX());
@@ -199,6 +201,7 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public Vector getVelocity() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot access entity velocity off region"); // Canvas - aggressive thread checks
         return CraftVector.toBukkit(this.entity.getDeltaMovement());
     }
 
@@ -206,6 +209,7 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
     public void setVelocity(Vector velocity) {
         Preconditions.checkArgument(velocity != null, "velocity");
         velocity.checkFinite();
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot set entity velocity off region"); // Canvas - aggressive thread checks
         // Paper start - Warn server owners when plugins try to set super high velocities
         if (!(this instanceof org.bukkit.entity.Projectile || this instanceof org.bukkit.entity.Minecart) && isUnsafeVelocity(velocity)) {
             CraftServer.excessiveVelEx = new Exception("Excessive velocity set detected: tried to set velocity of entity " + entity.getScoreboardName() + " id #" + getEntityId() + " to (" + velocity.getX() + "," + velocity.getY() + "," + velocity.getZ() + ").");
@@ -240,22 +244,26 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public double getHeight() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot access entity height off region"); // Canvas - aggressive thread checks
         return this.getHandle().getBbHeight();
     }
 
     @Override
     public double getWidth() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot access entity width off region"); // Canvas - aggressive thread checks
         return this.getHandle().getBbWidth();
     }
 
     @Override
     public BoundingBox getBoundingBox() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot access entity bounding box off region"); // Canvas - aggressive thread checks
         AABB bb = this.getHandle().getBoundingBox();
         return new BoundingBox(bb.minX, bb.minY, bb.minZ, bb.maxX, bb.maxY, bb.maxZ);
     }
 
     @Override
     public boolean isOnGround() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot access if entity is off the ground off the owning region"); // Canvas - aggressive thread checks
         if (this.entity instanceof AbstractArrow abstractArrow) {
             return abstractArrow.isInGround();
         }
@@ -264,16 +272,19 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public boolean isInWater() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot check if entity is in water off region"); // Canvas - aggressive thread checks
         return this.entity.isInWater();
     }
 
     @Override
     public World getWorld() {
+        // TODO - Canvas - do we thread check this?
         return this.entity.level().getWorld();
     }
 
     @Override
     public void setRotation(float yaw, float pitch) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot set rotation of entity off region"); // Canvas - aggressive thread checks
         NumberConversions.checkFinite(pitch, "pitch not finite");
         NumberConversions.checkFinite(yaw, "yaw not finite");
 
@@ -364,6 +375,7 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public void lookAt(double x, double y, double z, LookAnchor entityAnchor) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot make entity look at anchor off region"); // Canvas - aggressive thread checks
         this.getHandle().lookAt(toNmsAnchor(entityAnchor), new net.minecraft.world.phys.Vec3(x, y, z));
     }
 
@@ -384,7 +396,7 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
     @Override
     public List<org.bukkit.entity.Entity> getNearbyEntities(double x, double y, double z) {
         Preconditions.checkState(!this.entity.generation, "Cannot get nearby entities during world generation");
-        org.spigotmc.AsyncCatcher.catchOp("getNearbyEntities"); // Spigot
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get nearby entities off region"); // Canvas - aggressive thread checks
 
         List<Entity> entities = this.getHandle().level().getEntities(this.entity, this.entity.getBoundingBox().inflate(x, y, z), Predicates.alwaysTrue());
         List<org.bukkit.entity.Entity> result = new java.util.ArrayList<>(entities.size());
@@ -397,21 +409,25 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public int getEntityId() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get entity id off region"); // Canvas - aggressive thread checks
         return this.getHandle().getId();
     }
 
     @Override
     public int getFireTicks() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get fire ticks off region"); // Canvas - aggressive thread checks
         return this.getHandle().getRemainingFireTicks();
     }
 
     @Override
     public int getMaxFireTicks() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get max fire ticks off region"); // Canvas - aggressive thread checks
         return this.getHandle().getFireImmuneTicks();
     }
 
     @Override
     public void setFireTicks(int ticks) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot set fire ticks off region"); // Canvas - aggressive thread checks
         this.getHandle().setRemainingFireTicks(ticks);
     }
 
@@ -423,65 +439,77 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public void setVisualFire(final TriState fire) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot set visual fire off region"); // Canvas - aggressive thread checks
         Preconditions.checkArgument(fire != null, "TriState cannot be null");
         this.getHandle().visualFire = fire;
     }
 
     @Override
     public boolean isVisualFire() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot fetch is visual fire off region"); // Canvas - aggressive thread checks
         return getVisualFire().toBooleanOrElse(false);
     }
 
     @Override
     public TriState getVisualFire() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot fetch is visual fire off region"); // Canvas - aggressive thread checks
         return this.getHandle().visualFire;
     }
 
     @Override
     public int getFreezeTicks() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get freeze ticks off region"); // Canvas - aggressive thread checks
         return this.getHandle().getTicksFrozen();
     }
 
     @Override
     public int getMaxFreezeTicks() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get max freeze ticks off region"); // Canvas - aggressive thread checks
         return this.getHandle().getTicksRequiredToFreeze();
     }
 
     @Override
     public void setFreezeTicks(int ticks) {
         Preconditions.checkArgument(0 <= ticks, "Ticks (%s) cannot be less than 0", ticks);
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot set freeze ticks off region"); // Canvas - aggressive thread checks
 
         this.getHandle().setTicksFrozen(ticks);
     }
 
     @Override
     public boolean isFrozen() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot fetch if entity is frozen off region"); // Canvas - aggressive thread checks
         return this.getHandle().isFullyFrozen();
     }
 
     @Override
     public boolean isFreezeTickingLocked() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get if the entity has freezing ticks locked off region"); // Canvas - aggressive thread checks
         return this.entity.freezeLocked;
     }
 
     @Override
     public void lockFreezeTicks(boolean locked) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot lock or unlock freeze ticks off region"); // Canvas - aggressive thread checks
         this.entity.freezeLocked = locked;
     }
 
     @Override
     public void remove() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot remove entity off region"); // Canvas - aggressive thread checks
         this.entity.pluginRemoved = true;
         this.entity.discard(this.getHandle().generation ? null : EntityRemoveEvent.Cause.PLUGIN);
     }
 
     @Override
     public boolean isDead() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get if entity is dead off region"); // Canvas - aggressive thread checks
         return !this.entity.isAlive();
     }
 
     @Override
     public boolean isValid() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get if entity is valid off region"); // Canvas - aggressive thread checks
         return this.entity.isAlive() && this.entity.valid;
     }
 
@@ -492,22 +520,26 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public boolean isPersistent() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get if entity is persistent off region"); // Canvas - aggressive thread checks
         return this.entity.persist;
     }
 
     @Override
     public void setPersistent(boolean persistent) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot set if entity is persistent off region"); // Canvas - aggressive thread checks
         this.entity.persist = persistent;
     }
 
     @Override
     public org.bukkit.entity.Entity getPassenger() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get passenger off region"); // Canvas - aggressive thread checks
         return this.isEmpty() ? null : this.getHandle().getPassengers().getFirst().getBukkitEntity();
     }
 
     @Override
     public boolean setPassenger(org.bukkit.entity.Entity passenger) {
         Preconditions.checkArgument(!this.equals(passenger), "Entity cannot ride itself.");
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot set entity passengers off region"); // Canvas - aggressive thread checks
         if (passenger instanceof CraftEntity) {
             this.eject();
             return ((CraftEntity) passenger).getHandle().startRiding(this.getHandle());
@@ -518,6 +550,7 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public List<org.bukkit.entity.Entity> getPassengers() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get entity passengers off region"); // Canvas - aggressive thread checks
         return Lists.newArrayList(Lists.transform(this.getHandle().getPassengers(), (Function<Entity, org.bukkit.entity.Entity>) Entity::getBukkitEntity));
     }
 
@@ -525,6 +558,7 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
     public boolean addPassenger(org.bukkit.entity.Entity passenger) {
         Preconditions.checkArgument(passenger != null, "Entity passenger cannot be null");
         Preconditions.checkArgument(!this.equals(passenger), "Entity cannot ride itself.");
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot add passenger off region"); // Canvas - aggressive thread checks
 
         return ((CraftEntity) passenger).getHandle().startRiding(this.getHandle(), true, true);
     }
@@ -532,6 +566,7 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
     @Override
     public boolean removePassenger(org.bukkit.entity.Entity passenger) {
         Preconditions.checkArgument(passenger != null, "Entity passenger cannot be null");
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot remove passenger off region"); // Canvas - aggressive thread checks
 
         ((CraftEntity) passenger).getHandle().stopRiding();
         return true;
@@ -539,11 +574,13 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public boolean isEmpty() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot check if entity has passengers off region"); // Canvas - aggressive thread checks
         return !this.getHandle().isVehicle();
     }
 
     @Override
     public boolean eject() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot eject passengers off region"); // Canvas - aggressive thread checks
         if (this.isEmpty()) {
             return false;
         }
@@ -554,49 +591,58 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public ItemStack getPickItemStack() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get pick item off region"); // Canvas - aggressive thread checks
         net.minecraft.world.item.ItemStack stack = this.getHandle().getPickResult();
         return stack == null ? ItemStack.empty() : stack.asBukkitCopy();
     }
 
     @Override
     public float getFallDistance() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get fall distance off region"); // Canvas - aggressive thread checks
         return (float) this.getHandle().fallDistance;
     }
 
     @Override
     public void setFallDistance(float distance) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot set fall distance off region"); // Canvas - aggressive thread checks
         this.getHandle().fallDistance = distance;
     }
 
     @Override
     public void setLastDamageCause(EntityDamageEvent event) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot set last damage cause off region"); // Canvas - aggressive thread checks
         this.lastDamageEvent = event;
     }
 
     @Override
     public EntityDamageEvent getLastDamageCause() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get last damage cause off region"); // Canvas - aggressive thread checks
         return this.lastDamageEvent;
     }
 
     @Override
     public UUID getUniqueId() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get UUID off region"); // Canvas - aggressive thread checks
         return this.entity.getUUID();
     }
 
     @Override
     public int getTicksLived() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get ticks lived off region"); // Canvas - aggressive thread checks
         return this.getHandle().totalEntityAge;
     }
 
     @Override
     public void setTicksLived(int value) {
         Preconditions.checkArgument(value > 0, "Age value (%s) must be greater than 0", value);
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot set ticks lived off region"); // Canvas - aggressive thread checks
         this.getHandle().tickCount = value;
         this.getHandle().totalEntityAge = value;
     }
 
     @Override
     public final EntityType getType() {
+        // TODO - Canvas - do we thread check this?
         return this.entityType;
     }
 
@@ -605,52 +651,62 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         Preconditions.checkArgument(effect != null, "Entity effect cannot be null");
         Preconditions.checkState(!this.entity.generation, "Cannot play effect during world generation");
         Preconditions.checkArgument(effect.isApplicableTo(this), "Entity effect cannot apply to this entity");
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot play entity effect off region"); // Canvas - aggressive thread checks
 
         this.getHandle().level().broadcastEntityEvent(this.getHandle(), effect.getData());
     }
 
     @Override
     public Sound getSwimSound() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get swim sound off region"); // Canvas - aggressive thread checks
         return CraftSound.minecraftToBukkit(this.getHandle().getSwimSound());
     }
 
     @Override
     public Sound getSwimSplashSound() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get swim splash sound off region"); // Canvas - aggressive thread checks
         return CraftSound.minecraftToBukkit(this.getHandle().getSwimSplashSound());
     }
 
     @Override
     public Sound getSwimHighSpeedSplashSound() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get high speed splash sound off region"); // Canvas - aggressive thread checks
         return CraftSound.minecraftToBukkit(this.getHandle().getSwimHighSpeedSplashSound());
     }
 
     @Override
     public void setMetadata(String metadataKey, MetadataValue newMetadataValue) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot set metadata off region"); // Canvas - aggressive thread checks
         this.server.getEntityMetadata().setMetadata(this, metadataKey, newMetadataValue);
     }
 
     @Override
     public List<MetadataValue> getMetadata(String metadataKey) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get metadata off region"); // Canvas - aggressive thread checks
         return this.server.getEntityMetadata().getMetadata(this, metadataKey);
     }
 
     @Override
     public boolean hasMetadata(String metadataKey) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot check if entity has metadata key off region"); // Canvas - aggressive thread checks
         return this.server.getEntityMetadata().hasMetadata(this, metadataKey);
     }
 
     @Override
     public void removeMetadata(String metadataKey, Plugin owningPlugin) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot remove metadata off region"); // Canvas - aggressive thread checks
         this.server.getEntityMetadata().removeMetadata(this, metadataKey, owningPlugin);
     }
 
     @Override
     public boolean isInsideVehicle() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get if entity is a passenger off region"); // Canvas - aggressive thread checks
         return this.getHandle().isPassenger();
     }
 
     @Override
     public boolean leaveVehicle() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot force entity to leave vehicle off region"); // Canvas - aggressive thread checks
         if (!this.isInsideVehicle()) {
             return false;
         }
@@ -661,6 +717,7 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public org.bukkit.entity.Entity getVehicle() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get entity vehicle off region"); // Canvas - aggressive thread checks
         if (!this.isInsideVehicle()) {
             return null;
         }
@@ -670,12 +727,14 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public net.kyori.adventure.text.Component customName() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get custom name off region"); // Canvas - aggressive thread checks
         final Component name = this.getHandle().getCustomName();
         return name != null ? io.papermc.paper.adventure.PaperAdventure.asAdventure(name) : null;
     }
 
     @Override
     public void customName(final net.kyori.adventure.text.Component customName) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot set custom name off region"); // Canvas - aggressive thread checks
         this.getHandle().setCustomName(customName != null ? io.papermc.paper.adventure.PaperAdventure.asVanilla(customName) : null);
     }
 
@@ -686,6 +745,7 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public void setCustomName(String name) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot set custom name off region"); // Canvas - aggressive thread checks
         // sane limit for name length
         if (name != null && name.length() > 256) {
             name = name.substring(0, 256);
@@ -696,6 +756,7 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public String getCustomName() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get custom name off region"); // Canvas - aggressive thread checks
         Component name = this.getHandle().getCustomName();
 
         if (name == null) {
@@ -707,16 +768,19 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public void setCustomNameVisible(boolean flag) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot set if custom name should be visible for entity off region"); // Canvas - aggressive thread checks
         this.getHandle().setCustomNameVisible(flag);
     }
 
     @Override
     public boolean isCustomNameVisible() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get if custom name should be visible for entity off region"); // Canvas - aggressive thread checks
         return this.getHandle().isCustomNameVisible();
     }
 
     @Override
     public void setVisibleByDefault(boolean visible) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot set entity visible by default off region"); // Canvas - aggressive thread checks
         if (this.getHandle().visibleByDefault != visible) {
             if (visible) {
                 // Making visible by default, reset and show to all players
@@ -736,12 +800,14 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public boolean isVisibleByDefault() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get entity visible by default off region"); // Canvas - aggressive thread checks
         return this.getHandle().visibleByDefault;
     }
 
     @Override
     public Set<Player> getTrackedBy() {
         Preconditions.checkState(!this.entity.generation, "Cannot get tracking players during world generation");
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get players tracking entity off region"); // Canvas - aggressive thread checks
         ImmutableSet.Builder<Player> players = ImmutableSet.builder();
 
         ServerLevel world = ((CraftWorld) this.getWorld()).getHandle();
@@ -760,6 +826,7 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
     public boolean isTrackedBy(final Player player) {
         Preconditions.checkState(!this.entity.generation, "Cannot check tracking players during world generation");
         Preconditions.checkArgument(player != null, "Player cannot be null");
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot check if entity is tracked by player off region"); // Canvas - aggressive thread checks
 
         ServerLevel world = ((CraftWorld) this.getWorld()).getHandle();
         ChunkMap.TrackedEntity entityTracker = this.getHandle().moonrise$getTrackedEntity(); // Folia - region threading
@@ -790,16 +857,19 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public String getName() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get entity name off region"); // Canvas - aggressive thread checks
         return CraftChatMessage.fromComponent(this.getHandle().getName());
     }
 
     @Override
     public net.kyori.adventure.text.@org.jetbrains.annotations.NotNull Component name() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get entity name off region"); // Canvas - aggressive thread checks
         return io.papermc.paper.adventure.PaperAdventure.asAdventure(this.getHandle().getName());
     }
 
     @Override
     public net.kyori.adventure.text.@org.jetbrains.annotations.NotNull Component teamDisplayName() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get team display name name off region"); // Canvas - aggressive thread checks
         return io.papermc.paper.adventure.PaperAdventure.asAdventure(this.getHandle().getDisplayName());
     }
 
@@ -870,97 +940,116 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public void setGlowing(boolean flag) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot set glowing flag off region"); // Canvas - aggressive thread checks
         this.getHandle().setGlowingTag(flag);
     }
 
     @Override
     public boolean isGlowing() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get glowing flag off region"); // Canvas - aggressive thread checks
         return this.getHandle().isCurrentlyGlowing();
     }
 
     @Override
     public void setInvulnerable(boolean flag) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot set invulnerable flag off region"); // Canvas - aggressive thread checks
         this.getHandle().setInvulnerable(flag);
     }
 
     @Override
     public boolean isInvulnerable() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get invulnerable flag off region"); // Canvas - aggressive thread checks
         return this.getHandle().isInvulnerableToBase(this.getHandle().damageSources().generic());
     }
 
     @Override
     public boolean isSilent() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get if entity is silent off region"); // Canvas - aggressive thread checks
         return this.getHandle().isSilent();
     }
 
     @Override
     public void setSilent(boolean flag) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot set if entity is silent off region"); // Canvas - aggressive thread checks
         this.getHandle().setSilent(flag);
     }
 
     @Override
     public boolean hasGravity() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot check if entity has gravity off region"); // Canvas - aggressive thread checks
         return !this.getHandle().isNoGravity();
     }
 
     @Override
     public void setGravity(boolean gravity) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot set if entity has gravity off region"); // Canvas - aggressive thread checks
         this.getHandle().setNoGravity(!gravity);
     }
 
     @Override
     public int getPortalCooldown() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get portal cooldown off region"); // Canvas - aggressive thread checks
         return this.getHandle().getPortalCooldown();
     }
 
     @Override
     public void setPortalCooldown(int cooldown) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot set portal cooldown off region"); // Canvas - aggressive thread checks
         this.getHandle().setPortalCooldown(cooldown);
     }
 
     @Override
     public Set<String> getScoreboardTags() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get scoreboard tags off region"); // Canvas - aggressive thread checks
         return this.getHandle().getTags();
     }
 
     @Override
     public boolean addScoreboardTag(String tag) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot add scoreboard tags off region"); // Canvas - aggressive thread checks
         return this.getHandle().addTag(tag);
     }
 
     @Override
     public boolean removeScoreboardTag(String tag) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot remove scoreboard tags off region"); // Canvas - aggressive thread checks
         return this.getHandle().removeTag(tag);
     }
 
     @Override
     public PistonMoveReaction getPistonMoveReaction() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get the entity piston move reaction off region"); // Canvas - aggressive thread checks
         return PistonMoveReaction.getById(this.getHandle().getPistonPushReaction().ordinal());
     }
 
     @Override
     public BlockFace getFacing() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get facing blockface off region"); // Canvas - aggressive thread checks
         // Use this method over getDirection because it handles boats and minecarts.
         return CraftBlock.notchToBlockFace(this.getHandle().getMotionDirection());
     }
 
     @Override
     public CraftPersistentDataContainer getPersistentDataContainer() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get PDC off region"); // Canvas - aggressive thread checks
         return this.persistentDataContainer;
     }
 
     @Override
     public Pose getPose() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get entity pose off region"); // Canvas - aggressive thread checks
         return Pose.values()[this.getHandle().getPose().ordinal()];
     }
 
     @Override
     public void setSneaking(boolean sneak) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot set sneaking flag off region"); // Canvas - aggressive thread checks
         this.getHandle().setShiftKeyDown(sneak);
     }
 
     @Override
     public boolean isSneaking() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get sneaking flag off region"); // Canvas - aggressive thread checks
         return this.getHandle().isShiftKeyDown();
     }
 
@@ -971,6 +1060,7 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
     }
 
     public void setPose0(net.minecraft.world.entity.Pose pose, boolean fixed) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot set entity pose off region"); // Canvas - aggressive thread checks
         final Entity handle = this.getHandle();
         handle.fixedPose = false;
         handle.setPose(pose);
@@ -979,21 +1069,25 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public boolean hasFixedPose() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get if entity has a fixed pose off region"); // Canvas - aggressive thread checks
         return this.getHandle().fixedPose;
     }
 
     @Override
     public SpawnCategory getSpawnCategory() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get spawn category off region"); // Canvas - aggressive thread checks
         return CraftSpawnCategory.toBukkit(this.getHandle().getType().getCategory());
     }
 
     @Override
     public boolean isInWorld() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get if entity is in world off region"); // Canvas - aggressive thread checks
         return this.getHandle().inWorld;
     }
 
     @Override
     public String getAsString() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot serialize entity to string off region"); // Canvas - aggressive thread checks
         try (final ProblemReporter.ScopedCollector problemReporter = new ProblemReporter.ScopedCollector(
             () -> "Entity#toString", LOGGER
         )) {
@@ -1011,11 +1105,13 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public EntitySnapshot createSnapshot() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot create entity snapshot off region"); // Canvas - aggressive thread checks
         return CraftEntitySnapshot.create(this);
     }
 
     @Override
     public org.bukkit.entity.Entity copy() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot copy entity off region"); // Canvas - aggressive thread checks
         Entity copy = this.copy(this.getHandle().level());
         Preconditions.checkArgument(copy != null, "Error creating new entity.");
 
@@ -1024,6 +1120,7 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public org.bukkit.entity.Entity copy(Location location) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot copy entity off region"); // Canvas - aggressive thread checks
         Preconditions.checkArgument(location.getWorld() != null, "Location has no world");
 
         Entity copy = this.copy(((CraftWorld) location.getWorld()).getHandle());
@@ -1195,6 +1292,7 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public Location getOrigin() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get entity origin off region"); // Canvas - aggressive thread checks
         if (this.getHandle().origin == null) {
             return null;
         }
@@ -1209,36 +1307,43 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public boolean fromMobSpawner() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get if entity is from spawner off region"); // Canvas - aggressive thread checks
         return this.getHandle().spawnedViaMobSpawner;
     }
 
     @Override
     public org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason getEntitySpawnReason() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get entity spawn reason off region"); // Canvas - aggressive thread checks
         return this.getHandle().spawnReason;
     }
 
     @Override
     public boolean isUnderWater() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get if entity is underwater off region"); // Canvas - aggressive thread checks
         return this.getHandle().isUnderWater();
     }
 
     @Override
     public boolean isInRain() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get if entity is in rain off region"); // Canvas - aggressive thread checks
         return this.getHandle().isInRain();
     }
 
     @Override
     public boolean isInLava() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get if entity is in lava off region"); // Canvas - aggressive thread checks
         return this.getHandle().isInLava();
     }
 
     @Override
     public boolean isTicking() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get if entity is ticking off region"); // Canvas - aggressive thread checks
         return this.getHandle().isTicking();
     }
 
     @Override
     public Set<org.bukkit.entity.Player> getTrackedPlayers() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get tracked players off region"); // Canvas - aggressive thread checks
         ChunkMap.TrackedEntity tracker = this.entity.moonrise$getTrackedEntity(); // Folia - region threading
         if (tracker == null) {
             return java.util.Collections.emptySet();
@@ -1255,6 +1360,8 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
     public boolean spawnAt(Location location, CreatureSpawnEvent.SpawnReason reason) {
         Preconditions.checkNotNull(location, "location cannot be null");
         Preconditions.checkNotNull(reason, "reason cannot be null");
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot try spawn entity off region"); // Canvas - aggressive thread checks
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(((CraftWorld) location.getWorld()).getHandle(), location.getX(), location.getZ(), "Cannot spawn entity at position not owned by the current region"); // Canvas - aggressive thread checks
         this.entity.setLevel(((CraftWorld) location.getWorld()).getHandle());
         this.entity.setPos(location.getX(), location.getY(), location.getZ());
         this.entity.setRot(location.getYaw(), location.getPitch());
@@ -1267,6 +1374,7 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public boolean isInPowderedSnow() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get if entity is in powdered snow off region"); // Canvas - aggressive thread checks
         return this.getHandle().isInPowderSnow || this.getHandle().wasInPowderSnow; // depending on the location in the entity "tick" either could be needed.
     }
 
@@ -1297,27 +1405,32 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public boolean isInvisible() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get if entity is invisible off region"); // Canvas - aggressive thread checks
         return this.getHandle().isInvisible();
     }
 
     @Override
     public void setInvisible(boolean invisible) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot set invisible flag off region"); // Canvas - aggressive thread checks
         this.getHandle().persistentInvisibility = invisible;
         this.getHandle().setSharedFlag(Entity.FLAG_INVISIBLE, invisible);
     }
 
     @Override
     public void setNoPhysics(boolean noPhysics) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot set no physics flag off region"); // Canvas - aggressive thread checks
         this.getHandle().noPhysics = noPhysics;
     }
 
     @Override
     public boolean hasNoPhysics() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get no physics flag off region"); // Canvas - aggressive thread checks
         return this.getHandle().noPhysics;
     }
 
     @Override
     public boolean collidesAt(@org.jetbrains.annotations.NotNull Location location) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get if entity collides at location off region"); // Canvas - aggressive thread checks
         net.minecraft.world.phys.AABB box = this.getHandle().getBoundingBoxAt(location.getX(), location.getY(), location.getZ());
 
         return !this.getHandle().level().noCollision(this.getHandle(), box);
@@ -1325,6 +1438,7 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public boolean wouldCollideUsing(@org.jetbrains.annotations.NotNull BoundingBox boundingBox) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot check if entity would collide with a bounding box off the owning entity context"); // Canvas - aggressive thread checks
         net.minecraft.world.phys.AABB box = new AABB(boundingBox.getMinX(), boundingBox.getMinY(), boundingBox.getMinZ(), boundingBox.getMaxX(), boundingBox.getMaxY(), boundingBox.getMaxZ());
 
         return !this.getHandle().level().noCollision(this.getHandle(), box);
@@ -1332,12 +1446,14 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
 
     @Override
     public String getScoreboardEntryName() {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get entity scoreboard entry name off region"); // Canvas - aggressive thread checks
         return this.getHandle().getScoreboardName();
     }
 
     @Override
     public void broadcastHurtAnimation(java.util.Collection<Player> players) {
         //noinspection SuspiciousMethodCalls
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot broadcast hurt animation off region"); // Canvas - aggressive thread checks
         Preconditions.checkArgument(!players.contains(this), "Cannot broadcast hurt animation to self without a yaw");
         for (final org.bukkit.entity.Player player : players) {
             ((CraftPlayer) player).sendHurtAnimation(0, this);
@@ -1345,16 +1461,19 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
     }
     @Override
     public <T> @Nullable T getData(@NotNull final DataComponentType.Valued<T> type) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get data component value off region"); // Canvas - aggressive thread checks
         return this.entity.get(io.papermc.paper.datacomponent.PaperDataComponentType.bukkitToMinecraft(type));
     }
 
     @Override
     public <T> @Nullable T getDataOrDefault(@NotNull final DataComponentType.Valued<? extends T> type, @Nullable final T fallback) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot get data component value off region"); // Canvas - aggressive thread checks
         return this.entity.getOrDefault(io.papermc.paper.datacomponent.PaperDataComponentType.bukkitToMinecraft(type), fallback);
     }
 
     @Override
     public boolean hasData(final @NotNull DataComponentType type) {
+        ca.spottedleaf.moonrise.common.util.TickThread.canvas$tryEnsureThreadCheck(this.entity, "Cannot check if a data component type is present on entity off region"); // Canvas - aggressive thread checks
         return this.entity.get(io.papermc.paper.datacomponent.PaperDataComponentType.bukkitToMinecraft(type)) != null;
     }
 
