From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Mon, 20 Jan 2025 11:30:29 -0800
Subject: [PATCH] Allow disabling watchdog


diff --git a/src/main/java/org/spigotmc/WatchdogThread.java b/src/main/java/org/spigotmc/WatchdogThread.java
index 4d9afe27ee3c883835b96da060788ff45377adff..8d9dd4339e6fb8b5b97f48eef9a844254a783390 100644
--- a/src/main/java/org/spigotmc/WatchdogThread.java
+++ b/src/main/java/org/spigotmc/WatchdogThread.java
@@ -1,5 +1,6 @@
 package org.spigotmc;
 
+import io.canvasmc.canvas.Config;
 import io.papermc.paper.FeatureHooks;
 import io.papermc.paper.configuration.GlobalConfiguration;
 import java.lang.management.ManagementFactory;
@@ -72,6 +73,11 @@ public class WatchdogThread extends ca.spottedleaf.moonrise.common.util.TickThre
 
     @Override
     public void run() {
+        // Canvas start - allow disabling watchdog
+        if (!Config.INSTANCE.enableWatchdog) {
+            return;
+        }
+        // Canvas end
         while (!this.stopping) {
             // Paper start
             Logger logger = Bukkit.getServer().getLogger();
