--- a/ca/spottedleaf/concurrentutil/executor/thread/BalancedPrioritisedThreadPool.java
+++ b/ca/spottedleaf/concurrentutil/executor/thread/BalancedPrioritisedThreadPool.java
@@ -13,19 +_,19 @@
 import java.util.concurrent.atomic.AtomicLong;
 import java.util.function.Consumer;
 
-public final class BalancedPrioritisedThreadPool {
+public class BalancedPrioritisedThreadPool { // Canvas - non-final
 
     public static final long DEFAULT_GROUP_TIME_SLICE = (long)(15.0e6); // 15ms
 
     private static final Logger LOGGER = LoggerFactory.getLogger(BalancedPrioritisedThreadPool.class);
 
-    private final Consumer<Thread> threadModifier;
+    protected final Consumer<Thread> threadModifier; // Canvas - private -> protected
 
     private final COWArrayList<OrderedStreamGroup> groups = new COWArrayList<>(OrderedStreamGroup.class);
     private final COWArrayList<WorkerThread> threads = new COWArrayList<>(WorkerThread.class);
     private final COWArrayList<WorkerThread> aliveThreads = new COWArrayList<>(WorkerThread.class);
 
-    private final long groupTimeSliceNS;
+    protected final long groupTimeSliceNS; // Canvas - private -> protected
 
     private boolean shutdown;
 
