--- a/ca/spottedleaf/concurrentutil/executor/thread/PrioritisedThreadPool.java
+++ b/ca/spottedleaf/concurrentutil/executor/thread/PrioritisedThreadPool.java
@@ -304,7 +_,7 @@
         }
     }
 
-    private final class PrioritisedThread extends PrioritisedQueueExecutorThread {
+    private final class PrioritisedThread extends io.canvasmc.canvas.chunk.PrioritisedQueueExecutorThread { // Canvas - use optimized thread
 
         private final AtomicBoolean alertedHighPriority = new AtomicBoolean();
 
