From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/monster/Strider.java b/net/minecraft/world/entity/monster/Strider.java
index b451566378401e471a98db63df502b7c4cd20f42..9cba9f0b926aba52b88d3fdd92fe770b5f214f60 100644
--- a/net/minecraft/world/entity/monster/Strider.java
+++ b/net/minecraft/world/entity/monster/Strider.java
@@ -292,7 +292,7 @@ public class Strider extends Animal implements ItemSteerable {
             BlockState blockStateOnLegacy = this.getBlockStateOnLegacy();
             boolean flag = blockState.is(BlockTags.STRIDER_WARM_BLOCKS)
                 || blockStateOnLegacy.is(BlockTags.STRIDER_WARM_BLOCKS)
-                || this.getFluidHeight(FluidTags.LAVA) > 0.0;
+                || canvas$getLavaFluidHeight() > 0.0; // Canvas - optimize fluid height
             boolean flag1 = this.getVehicle() instanceof Strider strider && strider.isSuffocating();
             // CraftBukkit start
             boolean suffocating = !flag || flag1;
@@ -350,7 +350,7 @@ public class Strider extends Animal implements ItemSteerable {
 
     @Override
     protected boolean canAddPassenger(Entity passenger) {
-        return !this.isVehicle() && !this.isEyeInFluid(FluidTags.LAVA);
+        return !this.isVehicle() && !canvas$isEyeInLava(); // Canvas - optimize fluidOnEyes
     }
 
     @Override
