--- a/net/minecraft/world/entity/ai/behavior/GateBehavior.java
+++ b/net/minecraft/world/entity/ai/behavior/GateBehavior.java
@@ -85,7 +_,17 @@
             }
         }
         // Paper end - Perf: Remove streams from hot code
-        if (this.behaviors.stream().noneMatch(behavior -> behavior.getStatus() == Behavior.Status.RUNNING)) {
+        // Canvas start - remove streams
+        boolean hasRunningTask = false;
+        for (final BehaviorControl<? super E> behavior : this.behaviors) {
+            if (behavior.getStatus() == Behavior.Status.RUNNING) {
+                hasRunningTask = true;
+                break;
+            }
+        }
+
+        if (!hasRunningTask) {
+        // Canvas end - remove streams
             this.doStop(level, body, timestamp);
         }
     }
