--- a/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java
+++ b/net/minecraft/world/entity/ai/behavior/SetClosestHomeAsWalkTarget.java
@@ -58,11 +_,15 @@
                                 batchCache.put(key, lastUpdate.longValue() + CACHE_TIMEOUT);
                                 return true;
                             };
-                            Set<Pair<Holder<PoiType>, BlockPos>> pois = poiManager.findAllWithType(
-                                    p -> p.is(PoiTypes.HOME), cacheTest, body.blockPosition(), AcquirePoi.SCAN_RANGE, PoiManager.Occupancy.ANY
-                                )
-                                .collect(Collectors.toSet());
-                            Path path = AcquirePoi.findPathToPois(body, pois);
+                            // Canvas start - reroute to paper poi logic
+                            java.util.List<Pair<Holder<PoiType>, BlockPos>> poips = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
+                            ca.spottedleaf.moonrise.patches.poi_lookup.PoiAccess.findAnyPoiPositions(
+                                poiManager, p -> p.is(PoiTypes.HOME), cacheTest,
+                                body.blockPosition(), net.minecraft.world.entity.ai.behavior.AcquirePoi.SCAN_RANGE, // 48
+                                PoiManager.Occupancy.ANY, false, Integer.MAX_VALUE, poips
+                            );
+                            Path path = AcquirePoi.findPathToPois(body, new java.util.HashSet<>(poips));
+                            // Canvas end - reroute to paper poi logic
                             if (path != null && path.canReach()) {
                                 BlockPos targetPos = path.getTarget();
                                 Optional<Holder<PoiType>> type = poiManager.getType(targetPos);
