--- a/net/minecraft/world/entity/raid/Raid.java
+++ b/net/minecraft/world/entity/raid/Raid.java
@@ -555,7 +_,7 @@
             double x3 = playerLoc.x + 13.0 / distBtwn * (raidLoc.x - playerLoc.x);
             double z3 = playerLoc.z + 13.0 / distBtwn * (raidLoc.z - playerLoc.z);
             if (distBtwn <= 64.0 || playersInRaid.contains(player)) {
-                player.connection.send(new ClientboundSoundPacket(SoundEvents.RAID_HORN, SoundSource.NEUTRAL, x3, player.getY(), z3, 64.0F, 1.0F, seed));
+                player.connection.send(new ClientboundSoundPacket(SoundEvents.RAID_HORN, io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaFixes.mc245394 ? SoundSource.HOSTILE : SoundSource.NEUTRAL, x3, player.getY(), z3, 64.0F, 1.0F, seed)); // Canvas - MC-245394
             }
         }
     }
