--- a/net/minecraft/world/entity/raid/Raid.java
+++ b/net/minecraft/world/entity/raid/Raid.java
@@ -527,7 +_,7 @@
             double d1 = vec3.z + 13.0 / squareRoot * (vec31.z - vec3.z);
             if (squareRoot <= 64.0 || players.contains(serverPlayer)) {
                 serverPlayer.connection
-                    .send(new ClientboundSoundPacket(SoundEvents.RAID_HORN, SoundSource.NEUTRAL, d, serverPlayer.getY(), d1, 64.0F, 1.0F, randomLong));
+                    .send(new ClientboundSoundPacket(SoundEvents.RAID_HORN, io.canvasmc.canvas.Config.INSTANCE.fixes.mc245394 ? SoundSource.HOSTILE : SoundSource.NEUTRAL, d, serverPlayer.getY(), d1, 64.0F, 1.0F, randomLong)); // Canvas - MC-245394
             }
         }
     }
