From 53e1a57c72953de1bfcc94c3257c724e7fa1f88f Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/server/level/ChunkMap.java b/net/minecraft/server/level/ChunkMap.java
index 97c1f739253f986688178dbabc13bc114c6962bb..08823bc7a26a561a6cd3e4815e543eeec40236b2 100644
--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -244,7 +244,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
 
         final int index = entity.getType().getCategory().ordinal();
         final ca.spottedleaf.moonrise.common.list.ReferenceList<ServerPlayer> inRange =
-            this.level.moonrise$getNearbyPlayers().getPlayers(entity.chunkPosition(), ca.spottedleaf.moonrise.common.misc.NearbyPlayers.NearbyMapType.TICK_VIEW_DISTANCE);
+            this.level.moonrise$getNearbyPlayers().getPlayers(entity.chunkPosition(), ca.spottedleaf.moonrise.common.misc.NearbyPlayers.NearbyMapType.SPAWN_RANGE); // Canvas - fix mobs walking outside simulation distance with higher view distance causing mob caps to be ignored for per-player entities
         if (inRange == null) {
             return;
         }
@@ -650,7 +650,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
                 printFuture(holder.getEntityTickingChunkFuture()),
                 this.ticketStorage.getTicketDebugString(posKey, false),
                 this.anyPlayerCloseEnoughForSpawning(pos),
-                fullChunk.<Integer>map(c -> c.getBlockEntities().size()).orElse(0),
+                fullChunk.<Integer>map(LevelChunk::canvas$getBlockEntitiesCount).orElse(0), // Canvas - optimize block entity fetching
                 this.ticketStorage.getTicketDebugString(posKey, true),
                 this.distanceManager.getChunkLevel(posKey, true),
                 fullChunk.<Integer>map(levelChunk -> levelChunk.getBlockTicks().count()).orElse(0),
@@ -1331,7 +1331,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
                 double distanceSquared = deltaToPlayerX * deltaToPlayerX + deltaToPlayerZ * deltaToPlayerZ; // Paper
                 double rangeSquared = visibleRange * visibleRange;
                 // Paper start - Configurable entity tracking range by Y
-                boolean visibleToPlayer = distanceSquared <= rangeSquared;
+                boolean visibleToPlayer = distanceSquared <= rangeSquared && (!(this.entity instanceof net.minecraft.server.level.ServerPlayer) || !level.canvasConfig().visuals.dontTrackPlayersInEntityTracking); // Canvas - allow disabling player viewing eachother in worlds
                 if (visibleToPlayer && level.paperConfig().entities.trackingRangeY.enabled) {
                     double rangeY = level.paperConfig().entities.trackingRangeY.get(this.entity, -1);
                     if (rangeY != -1) {
@@ -1349,7 +1349,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
                 }
                 // Folia end - region threading
                 // CraftBukkit start - respect vanish API
-                if (visibleToPlayer && (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(player) || !player.getBukkitEntity().canSee(this.entity.getBukkitEntity()))) { // Paper - only consider hits // Folia - region threading
+                if (visibleToPlayer && (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(player) || !player.getBukkitEntity().canSeeChunkMapUpdatePlayer(this.entity.getBukkitEntity()))) { // Paper - only consider hits // Folia - region threading // Leaf - optimize canSee checks
                     visibleToPlayer = false;
                 }
                 // CraftBukkit end
