From 2be0fbf099df19422462cdb5b961fc00062d256b Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java b/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
index 630677ccc0276882169265e1f7aa63ed0fdd86a3..0721717790c143401b6011177a6a3c6098f0b61f 100644
--- a/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
+++ b/net/minecraft/world/entity/vehicle/minecart/OldMinecartBehavior.java
@@ -396,7 +396,7 @@ public class OldMinecartBehavior extends MinecartBehavior {
                 }
             }
         } else {
-            for (Entity entityx : this.level().getEntities(this.minecart, hitbox)) {
+            for (Entity entityx : this.level().getEntitiesOfClass(AbstractMinecart.class, hitbox, entity -> entity != this.minecart)) { // Canvas - optimize entity type predicates
                 if (!this.minecart.hasPassenger(entityx) && entityx.isPushable() && entityx instanceof AbstractMinecart) {
                     // CraftBukkit start
                     org.bukkit.event.vehicle.VehicleEntityCollisionEvent collisionEvent = new org.bukkit.event.vehicle.VehicleEntityCollisionEvent(
