From 56a92b2c479a7161c340134ba931fb3d6a8e6f79 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/chunk/ChunkGenerator.java b/net/minecraft/world/level/chunk/ChunkGenerator.java
index b7325170369299a74be35ecc66a6446c301605b8..5895eb456d58686c6319ac2ad3518d58b1eeb679 100644
--- a/net/minecraft/world/level/chunk/ChunkGenerator.java
+++ b/net/minecraft/world/level/chunk/ChunkGenerator.java
@@ -196,7 +196,7 @@ public abstract class ChunkGenerator {
             int chunkOriginX = SectionPos.blockToSectionCoord(pos.getX());
             int chunkOriginZ = SectionPos.blockToSectionCoord(pos.getZ());
 
-            for (int radius = 0; radius <= maxSearchRadius; radius++) {
+            for (int radius = (io.canvasmc.canvas.GlobalConfiguration.getInstance().chunkSystem.optimizeTreasureMapLocating ? maxSearchRadius : 0); radius <= maxSearchRadius; radius++) { // Canvas - optimize treasure map locating - just search the largest radius to avoid spamming iterations
                 boolean foundSomething = false;
 
                 for (Entry<StructurePlacement, Set<Holder<Structure>>> entry : randomSpreadEntries) {
@@ -507,6 +507,12 @@ public abstract class ChunkGenerator {
     ) {
         Map<Structure, LongSet> structures = structureManager.getAllStructuresAt(pos);
 
+    // Canvas start - optimize natural spawning
+        return getMobsAt(biome, structureManager, mobCategory, pos, structures);
+    }
+
+    public WeightedList<MobSpawnSettings.SpawnerData> getMobsAt(final Holder<Biome> biome, final StructureManager structureManager, final MobCategory mobCategory, final BlockPos pos, final Map<Structure, LongSet> structures) {
+    // Canvas end - optimize natural spawning
         for (Entry<Structure, LongSet> entry : structures.entrySet()) {
             Structure structure = entry.getKey();
             StructureSpawnOverride override = structure.spawnOverrides().get(mobCategory);
