From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/projectile/FireworkRocketEntity.java b/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
index ca9e604cb81cf3c3b0312dd02265388251872283..2ca8d79f2c4f89de9a9a5bf9cc6b7a85315857f7 100644
--- a/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
+++ b/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
@@ -147,7 +147,7 @@ public class FireworkRocketEntity extends Projectile implements ItemSupplier {
 
             if (this.attachedToEntity != null) {
                 Vec3 handAngle;
-                if (this.attachedToEntity.isFallFlying()) {
+                if (this.attachedToEntity.isFallFlying() && (!io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaFixes.mc119754 || !this.attachedToEntity.isSpectator())) { // Canvas - MC-119754
                     Vec3 lookAngle = this.attachedToEntity.getLookAngle();
                     double power = 1.5;
                     double powerAdd = 0.1;
