From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/projectile/FishingHook.java b/net/minecraft/world/entity/projectile/FishingHook.java
index 846dcec56a75bef9a226b3281b91c6788ef1924f..710c485160ec2cb112d417730d7b21893c7d3d06 100644
--- a/net/minecraft/world/entity/projectile/FishingHook.java
+++ b/net/minecraft/world/entity/projectile/FishingHook.java
@@ -507,6 +507,11 @@ public class FishingHook extends Projectile {
         if (!this.level().isClientSide() && owner != null && !this.shouldStopFishing(owner)) {
             int dmg = 0;
             if (this.hookedIn != null) {
+                // Canvas start - combat configs
+                if (!level().canvasConfig().combat.allowFishingRodsToPullEntities) {
+                    return 0; // act as if we canceled the pull
+                }
+                // Canvas end - combat configs
                 // CraftBukkit start
                 org.bukkit.event.player.PlayerFishEvent playerFishEvent = new org.bukkit.event.player.PlayerFishEvent((org.bukkit.entity.Player) owner.getBukkitEntity(), this.hookedIn.getBukkitEntity(), (org.bukkit.entity.FishHook) this.getBukkitEntity(), org.bukkit.craftbukkit.CraftEquipmentSlot.getHand(hand), org.bukkit.event.player.PlayerFishEvent.State.CAUGHT_ENTITY); // Paper - Add hand parameter to PlayerFishEvent
                 if (!playerFishEvent.callEvent()) {
@@ -600,6 +605,14 @@ public class FishingHook extends Projectile {
         if (owner != null) {
             Vec3 delta = new Vec3(owner.getX() - this.getX(), owner.getY() - this.getY(), owner.getZ() - this.getZ()).scale(0.1);
             entity.setDeltaMovement(entity.getDeltaMovement().add(delta));
+            // Canvas start - MC-100991
+            if (io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaFixes.mc100991 && entity instanceof net.minecraft.world.entity.LivingEntity livingEntity) {
+                livingEntity.getCombatTracker().recordDamage(
+                    level().damageSources().thrown(this, owner),
+                    livingEntity.getHealth()
+                );
+            }
+            // Canvas end - MC-100991
         }
     }
 
