From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/io/papermc/paper/threadedregions/RegionShutdownThread.java b/io/papermc/paper/threadedregions/RegionShutdownThread.java
index 11610b628bbf328fef39b4bfe1f8f522730a16e6..b266555faff7560294eb125b7c40f3170cc4709a 100644
--- a/io/papermc/paper/threadedregions/RegionShutdownThread.java
+++ b/io/papermc/paper/threadedregions/RegionShutdownThread.java
@@ -140,6 +140,7 @@ public class RegionShutdownThread extends ca.spottedleaf.moonrise.common.util.Ti
                         player.containerMenu.setCarried(ItemStack.EMPTY);
                         player.drop(carried, false);
                     }
+                    if (player.canvas$isTemporarilyBlocking) player.releaseUsingItem(); // Canvas - implement sword blocking
                 } catch (final Throwable thr) {
                     LOGGER.error("Failed to close player inventory for player: " + player, thr);
                 }
@@ -334,4 +335,4 @@ public class RegionShutdownThread extends ca.spottedleaf.moonrise.common.util.Ti
         MinecraftServer.getServer().stopPart2(); // stop part 2: close other resources (io thread, etc)
         // done, part 2 should call exit()
     }
-}
\ No newline at end of file
+}
