From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/io/papermc/paper/threadedregions/RegionizedServer.java b/io/papermc/paper/threadedregions/RegionizedServer.java
index aea2ffe9f02b6a1cf7a55f6b7c77c7e0431bbbf9..db46707c4d22a421b996f6d6cea1f554c4170c37 100644
--- a/io/papermc/paper/threadedregions/RegionizedServer.java
+++ b/io/papermc/paper/threadedregions/RegionizedServer.java
@@ -63,6 +63,14 @@ public final class RegionizedServer {
     }
 
     public void init() {
+        // Canvas start - default respawn dimension config
+        // validate respawn dimension config entry first
+        net.minecraft.resources.ResourceKey<net.minecraft.world.level.Level> respawnDimension = io.canvasmc.canvas.GlobalConfiguration.fetchRespawnDimensionKey();
+        ServerLevel level = MinecraftServer.getServer().getLevel(respawnDimension);
+        if (level == null) {
+            throw new IllegalStateException("Respawn dimension invalid, could not fetch dimension from value '" + respawnDimension + "'");
+        }
+        // Canvas end - default respawn dimension config
         // call init event _before_ scheduling anything
         new RegionizedServerInitEvent().callEvent();
 
@@ -563,4 +571,4 @@ public final class RegionizedServer {
     public static final record WorldLevelData(ServerLevel world, long nonRedstoneGameTime) {
 
     }
-}
\ No newline at end of file
+}
