From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Mon, 27 Apr 2026 22:46:36 -0700
Subject: [PATCH] Remove Dead Old Watchdog Code


diff --git a/net/minecraft/network/Connection.java b/net/minecraft/network/Connection.java
index c2bdede413d38ee5b211561f67c9897639c2e879..1e73ad5aad737946900227205c7639f6ba0d77b5 100644
--- a/net/minecraft/network/Connection.java
+++ b/net/minecraft/network/Connection.java
@@ -637,13 +637,7 @@ public class Connection extends SimpleChannelInboundHandler<Packet<?>> {
         if (this.packetListener instanceof TickablePacketListener tickable) {
             // Paper start - Buffer joins to world
             if (true) { // Folia - region threading
-            // Paper start - detailed watchdog information
-            net.minecraft.network.PacketProcessor.packetProcessing.push(this.packetListener);
-            try {
             tickable.tick();
-            } finally {
-                net.minecraft.network.PacketProcessor.packetProcessing.pop();
-            } // Paper end - detailed watchdog information
             } // Paper end - Buffer joins to world
         }
 
diff --git a/net/minecraft/network/PacketProcessor.java b/net/minecraft/network/PacketProcessor.java
index 0c84984fc72ae949c09300306783db23e0ed6cc0..c5c09e83583e8b413933e047d6256d6e9301b5d0 100644
--- a/net/minecraft/network/PacketProcessor.java
+++ b/net/minecraft/network/PacketProcessor.java
@@ -77,28 +77,8 @@ public class PacketProcessor implements AutoCloseable {
         this.closed = true;
     }
 
-    // Paper start - detailed watchdog information
-    public static final java.util.concurrent.ConcurrentLinkedDeque<PacketListener> packetProcessing = new java.util.concurrent.ConcurrentLinkedDeque<>();
-    static final java.util.concurrent.atomic.AtomicLong totalMainThreadPacketsProcessed = new java.util.concurrent.atomic.AtomicLong();
-
-    public static long getTotalProcessedPackets() {
-        return totalMainThreadPacketsProcessed.get();
-    }
-
-    public static java.util.List<PacketListener> getCurrentPacketProcessors() {
-        java.util.List<PacketListener> listeners = new java.util.ArrayList<>(4);
-        for (PacketListener listener : packetProcessing) {
-            listeners.add(listener);
-        }
-
-        return listeners;
-    }
-    // Paper end - detailed watchdog information
-
     private record ListenerAndPacket<T extends PacketListener>(T listener, Packet<T> packet) {
         public void handle() {
-            packetProcessing.push(this.listener); // Paper - detailed watchdog information
-            try { // Paper - detailed watchdog information
             if (this.listener.shouldHandleMessage(this.packet)) {
                 try {
                     this.packet.handle(this.listener);
@@ -112,12 +92,6 @@ public class PacketProcessor implements AutoCloseable {
             } else {
                 PacketProcessor.LOGGER.debug("Ignoring packet due to disconnection: {}", this.packet);
             }
-            // Paper start - detailed watchdog information
-            } finally {
-                totalMainThreadPacketsProcessed.getAndIncrement();
-                packetProcessing.pop();
-            }
-            // Paper end - detailed watchdog information
         }
     }
 }
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 5577c227684db30a34d7263043885cf2c5ca594b..29dd2844b599343e15a6bc7d3d45e7dbbeb82b50 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -1131,28 +1131,9 @@ public abstract class Entity
         return this.onGround;
     }
 
-    // Paper start - detailed watchdog information
-    public final Object posLock = new Object(); // Paper - log detailed entity tick information
-
-    @Nullable
-    private Vec3 moveVector;
-    private double moveStartX;
-    private double moveStartY;
-    private double moveStartZ;
-    // Paper end - detailed watchdog information
-
     public void move(final MoverType moverType, Vec3 delta) {
         final Vec3 originalMovement = delta; // Paper - Expose pre-collision velocity
-        // Paper start - detailed watchdog information
         ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread("Cannot move an entity off-main");
-        synchronized (this.posLock) {
-            this.moveStartX = this.getX();
-            this.moveStartY = this.getY();
-            this.moveStartZ = this.getZ();
-            this.moveVector = delta;
-        }
-        try {
-        // Paper end - detailed watchdog information
         if (this.noPhysics) {
             this.setPos(this.getX() + delta.x, this.getY() + delta.y, this.getZ() + delta.z);
             this.horizontalCollision = false;
@@ -1272,13 +1253,6 @@ public abstract class Entity
                 this.setDeltaMovement(this.getDeltaMovement().multiply(blockSpeedFactor, 1.0, blockSpeedFactor));
             }
         }
-        // Paper start - detailed watchdog information
-        } finally {
-            synchronized (this.posLock) { // Paper
-                this.moveVector = null;
-            } // Paper
-        }
-        // Paper end - detailed watchdog information
     }
 
     private void applyMovementEmissionAndPlaySound(
@@ -5662,9 +5636,7 @@ public abstract class Entity
 
     public void setDeltaMovement(final Vec3 deltaMovement) {
         if (deltaMovement.isFinite()) {
-            synchronized (this.posLock) { // Paper - detailed watchdog information
             this.deltaMovement = deltaMovement;
-            } // Paper - detailed watchdog information
         }
     }
 
@@ -5770,9 +5742,7 @@ public abstract class Entity
         // Paper end - Block invalid positions and bounding box
         boolean posChanged = this.position.x != x || this.position.y != y || this.position.z != z;
         if (posChanged) { // Folia - region threading
-            synchronized (this.posLock) { // Paper - detailed watchdog information
             this.position = new Vec3(x, y, z);
-            } // Paper - detailed watchdog information
             int fx = Mth.floor(x);
             int fy = Mth.floor(y);
             int fz = Mth.floor(z);
