From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Mon, 20 Apr 2026 12:05:53 -0700
Subject: [PATCH] PR Party 2


diff --git a/net/minecraft/commands/Commands.java b/net/minecraft/commands/Commands.java
index 50f1b26d3d848135a0d364a4a79d7414503a239d..e227bb12d843e2fc98906327fe192a0fc06f3c4b 100644
--- a/net/minecraft/commands/Commands.java
+++ b/net/minecraft/commands/Commands.java
@@ -203,7 +203,7 @@ public class Commands {
         //DataPackCommand.register(this.dispatcher, context); // Folia - region threading - TODO
         //DebugCommand.register(this.dispatcher); // Folia - region threading - TODO
         DefaultGameModeCommands.register(this.dispatcher);
-        //DialogCommand.register(this.dispatcher, context); // Folia - region threading - TODO
+        DialogCommand.register(this.dispatcher, context);
         DifficultyCommand.register(this.dispatcher);
         EffectCommands.register(this.dispatcher, context);
         EmoteCommands.register(this.dispatcher);
diff --git a/net/minecraft/server/commands/DialogCommand.java b/net/minecraft/server/commands/DialogCommand.java
index fcbe76e95a7de7c79afd0a116b6cbcf286ba85ff..393a3b573587b455327dc6f81e3099bd618e84de 100644
--- a/net/minecraft/server/commands/DialogCommand.java
+++ b/net/minecraft/server/commands/DialogCommand.java
@@ -46,7 +46,11 @@ public class DialogCommand {
 
     private static int showDialog(final CommandSourceStack sender, final Collection<ServerPlayer> targets, final Holder<Dialog> dialog) {
         for (ServerPlayer target : targets) {
-            target.openDialog(dialog);
+            // Folia start - region threading
+            target.getBukkitEntity().taskScheduler.scheduleOrExecute((ServerPlayer player) -> {
+                player.openDialog(dialog);
+            });
+            // Folia end - region threading
         }
 
         if (targets.size() == 1) {
@@ -60,7 +64,11 @@ public class DialogCommand {
 
     private static int clearDialog(final CommandSourceStack sender, final Collection<ServerPlayer> targets) {
         for (ServerPlayer target : targets) {
-            target.connection.send(ClientboundClearDialogPacket.INSTANCE);
+            // Folia start - region threading
+            target.getBukkitEntity().taskScheduler.scheduleOrExecute((ServerPlayer player) -> {
+                player.connection.send(ClientboundClearDialogPacket.INSTANCE);
+            });
+            // Folia end - region threading
         }
 
         if (targets.size() == 1) {
