From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Mon, 20 Apr 2026 12:07:26 -0700
Subject: [PATCH] PR Party 3


diff --git a/net/minecraft/commands/Commands.java b/net/minecraft/commands/Commands.java
index e227bb12d843e2fc98906327fe192a0fc06f3c4b..aa3e2b7aca2163c61de949519f332cd1dba5920a 100644
--- a/net/minecraft/commands/Commands.java
+++ b/net/minecraft/commands/Commands.java
@@ -243,7 +243,7 @@ public class Commands {
         SetBlockCommand.register(this.dispatcher, context);
         SetSpawnCommand.register(this.dispatcher);
         SetWorldSpawnCommand.register(this.dispatcher);
-        //SpectateCommand.register(this.dispatcher); // Folia - region threading - TODO later
+        SpectateCommand.register(this.dispatcher);
         //SpreadPlayersCommand.register(this.dispatcher); // Folia - region threading - TODO later
         StopSoundCommand.register(this.dispatcher);
         StopwatchCommand.register(this.dispatcher);
diff --git a/net/minecraft/server/commands/SpectateCommand.java b/net/minecraft/server/commands/SpectateCommand.java
index 38575e0ff4ae27beebf20a53ca04ed039acf9706..0f235515f9c39fbe6eef0d5e7fadced236a9bc63 100644
--- a/net/minecraft/server/commands/SpectateCommand.java
+++ b/net/minecraft/server/commands/SpectateCommand.java
@@ -37,7 +37,17 @@ public class SpectateCommand {
         );
     }
 
-    private static int spectate(final CommandSourceStack source, final @Nullable Entity target, final ServerPlayer player) throws CommandSyntaxException {
+    // Folia start - region threading
+    private static void sendMessage(CommandSourceStack src, CommandSyntaxException ex) {
+        src.sendFailure((Component)ex.getRawMessage());
+    }
+    // Folia end - region threading
+
+    private static int spectate(final CommandSourceStack source, final @Nullable Entity target, final ServerPlayer spectator) throws CommandSyntaxException { // Folia - region threading
+        // Folia start - region threading
+        spectator.getBukkitEntity().taskScheduler.scheduleOrExecute((ServerPlayer player) -> {
+            try {
+                // Folia end - region threading
         if (player == target) {
             throw ERROR_SELF.create();
         }
@@ -57,6 +67,13 @@ public class SpectateCommand {
             source.sendSuccess(() -> Component.translatable("commands.spectate.success.stopped"), false);
         }
 
-        return 1;
+        return; // Folia - region threading
+        // Folia start - region threading
+            } catch (CommandSyntaxException ex) {
+                sendMessage(source, ex);
+            }
+        });
+        return 0;
+        // Folia end - region threading
     }
 }
