--- a/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
+++ b/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
@@ -65,11 +_,17 @@
     }
 
     void updateConnectedPlayersWithinRange(ServerLevel level, BlockPos pos, VaultServerData serverData, VaultConfig config, double deactivationRange) {
-        Set<UUID> set = config.playerDetector()
-            .detect(level, config.entitySelector(), pos, deactivationRange, false)
-            .stream()
-            .filter(uuid -> !serverData.getRewardedPlayers().contains(uuid))
-            .collect(Collectors.toSet());
+        // Canvas start - remove streams
+        java.util.List<UUID> detectedPlayers = config.playerDetector().detect(level, config.entitySelector(), pos, deactivationRange, false);
+        Set<UUID> set = new ObjectLinkedOpenHashSet<>();
+
+        for (UUID uuid : detectedPlayers) {
+            if (!serverData.getRewardedPlayers().contains(uuid)) {
+                set.add(uuid);
+            }
+        }
+        // Canvas end - remove streams
+
         if (!this.connectedPlayers.equals(set)) {
             this.connectedPlayers = set;
             this.markDirty();
