--- a/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
+++ b/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
@@ -67,11 +_,14 @@
     void updateConnectedPlayersWithinRange(
         final ServerLevel serverLevel, final BlockPos pos, final VaultServerData serverData, final VaultConfig config, final double limit
     ) {
-        Set<UUID> currentConnectedPlayers = config.playerDetector()
-            .detect(serverLevel, config.entitySelector(), pos, limit, false)
-            .stream()
-            .filter(uuid -> !serverData.getRewardedPlayers().contains(uuid))
-            .collect(Collectors.toSet());
+        // Canvas start - remove streams
+        Set<UUID> currentConnectedPlayers = new ObjectLinkedOpenHashSet<>();
+        for (UUID uuid : config.playerDetector().detect(serverLevel, config.entitySelector(), pos, limit, false)) {
+            if (!serverData.getRewardedPlayers().contains(uuid)) {
+                currentConnectedPlayers.add(uuid);
+            }
+        }
+        // Canvas end - remove streams
         if (!this.connectedPlayers.equals(currentConnectedPlayers)) {
             this.connectedPlayers = currentConnectedPlayers;
             this.markDirty();
