From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/biome/TheEndBiomeSource.java b/net/minecraft/world/level/biome/TheEndBiomeSource.java
index c0c541e364e695e7b7eb0a05fbed8a2f1e14cd90..2899e14cedce5522e8214fd11fe0bb9b0cd7389e 100644
--- a/net/minecraft/world/level/biome/TheEndBiomeSource.java
+++ b/net/minecraft/world/level/biome/TheEndBiomeSource.java
@@ -56,9 +56,35 @@ public class TheEndBiomeSource extends BiomeSource {
     protected MapCodec<? extends BiomeSource> codec() {
         return CODEC;
     }
+    // Canvas start - end biome cache
+    private final ThreadLocal<it.unimi.dsi.fastutil.longs.Long2ObjectLinkedOpenHashMap<Holder<Biome>>> canvas$cache = ThreadLocal.withInitial(it.unimi.dsi.fastutil.longs.Long2ObjectLinkedOpenHashMap::new);
+    private final int canvas$cacheCapacity = io.canvasmc.canvas.GlobalConfiguration.getInstance().chunkSystem.endBiomeCacheSize;
 
     @Override
     public Holder<Biome> getNoiseBiome(final int quartX, final int quartY, final int quartZ, final Climate.Sampler sampler) {
+        if (!io.canvasmc.canvas.GlobalConfiguration.getInstance().chunkSystem.useEndBiomeCache) {
+            // biome cache not enabled, return raw vanilla
+            return getVanillaNoiseBiome(quartX, quartY, quartZ, sampler);
+        }
+        final long key = net.minecraft.world.level.ChunkPos.pack(quartX, quartZ);
+        final it.unimi.dsi.fastutil.longs.Long2ObjectLinkedOpenHashMap<Holder<Biome>> cacheThreadLocal = canvas$cache.get();
+        final Holder<Biome> biome = cacheThreadLocal.get(key);
+        if (biome != null) {
+            return biome;
+        } else {
+            final Holder<Biome> gennedBiome = getVanillaNoiseBiome(quartX, quartY, quartZ, sampler);
+            cacheThreadLocal.put(key, gennedBiome);
+            if (cacheThreadLocal.size() > canvas$cacheCapacity) {
+                for (int i = 0; i < canvas$cacheCapacity / 16; i ++) {
+                    cacheThreadLocal.removeFirst();
+                }
+            }
+            return gennedBiome;
+        }
+    }
+
+    public Holder<Biome> getVanillaNoiseBiome(final int quartX, final int quartY, final int quartZ, final Climate.Sampler sampler) {
+        // Canvas end - end biome cache
         int blockX = QuartPos.toBlock(quartX);
         int blockY = QuartPos.toBlock(quartY);
         int blockZ = QuartPos.toBlock(quartZ);
