From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Wed, 15 Jan 2025 01:06:07 -0800
Subject: [PATCH] Threaded Dimensions


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 249d6549543c059694a9372200b8581feb8dd8f8..e782e44b5773c704a240a9359ff8cebbce694f7c 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -67,7 +67,7 @@ import org.jetbrains.annotations.Nullable;
  * Represents the Bukkit core, for version and Server singleton handling
  */
 public final class Bukkit {
-    private static Server server;
+    public static Server server; // Canvas - private -> public
 
     /**
      * Static class cannot be initialized.
diff --git a/src/main/java/org/bukkit/event/server/ServerEvent.java b/src/main/java/org/bukkit/event/server/ServerEvent.java
index 05167fb34e4c42edc67af6e6700a2a3cc0f92769..c59bd09f2d97238f0c3647390ccff8a01e400f39 100644
--- a/src/main/java/org/bukkit/event/server/ServerEvent.java
+++ b/src/main/java/org/bukkit/event/server/ServerEvent.java
@@ -9,7 +9,7 @@ import org.bukkit.event.Event;
 public abstract class ServerEvent extends Event {
 
     public ServerEvent() {
-        super(!Bukkit.isPrimaryThread()); // Paper
+        super(Bukkit.server != null && !Bukkit.isPrimaryThread()); // Paper // Canvas
     }
 
     public ServerEvent(boolean isAsync) {
