From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
index e0a74852160193df7b7a4da98cf3d2c204c68ba9..d00313040e08c17c5533ee095e66b39840d9b515 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
@@ -383,8 +383,12 @@ public final class RegionizedPlayerChunkLoader {
             final int centerZ = PlayerChunkLoaderData.this.lastChunkZ;
 
             return Integer.compare(
-                Math.abs(c1x - centerX) + Math.abs(c1z - centerZ),
-                Math.abs(c2x - centerX) + Math.abs(c2z - centerZ)
+                // Canvas start - use Euclidean distance squared
+                io.canvasmc.canvas.Config.INSTANCE.chunks.useEuclideanDistanceSquared ?
+                    (c1x - centerX) * (c1x - centerX) + (c1z - centerZ) * (c1z - centerZ) : Math.abs(c1x - centerX) + Math.abs(c1z - centerZ),
+                io.canvasmc.canvas.Config.INSTANCE.chunks.useEuclideanDistanceSquared ?
+                    (c2x - centerX) * (c2x - centerX) + (c2z - centerZ) * (c2z - centerZ) : Math.abs(c2x - centerX) + Math.abs(c2z - centerZ)
+                // Canvas end - use Euclidean distance squared
             );
         };
         private final LongHeapPriorityQueue sendQueue = new LongHeapPriorityQueue(CLOSEST_MANHATTAN_DIST);
