From 56a92b2c479a7161c340134ba931fb3d6a8e6f79 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/item/ItemStack.java b/net/minecraft/world/item/ItemStack.java
index 0e0ab00b2836349cf15f1ac09ce9efaccbb073e9..5e5fafb906931ea5d220c0cb0c045c7ff69f32ce 100644
--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -837,11 +837,11 @@ public final class ItemStack implements DataComponentHolder, ItemInstance {
     }
 
     public static boolean isSameItem(final ItemStack a, final ItemStack b) {
-        return a.is(b.getItem());
+        return a == b || a.is(b.getItem()); // Canvas - optimize matching checks
     }
 
     public static boolean isSameItemSameComponents(final ItemStack a, final ItemStack b) {
-        return a.is(b.getItem()) && (a.isEmpty() && b.isEmpty() || Objects.equals(a.components, b.components));
+        return a == b || a.is(b.getItem()) && (a.isEmpty() && b.isEmpty() || Objects.equals(a.components, b.components)); // Canvas - optimize matching checks
     }
 
     public static boolean matchesIgnoringComponents(final ItemStack a, final ItemStack b, final Predicate<DataComponentType<?>> ignoredPredicate) {
