From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index 9d61d40c263c73b3ada563c4ade58d780fab7ab5..54562fbb6907dff0c4f0d02a0e13c4dc60d06c09 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -341,6 +341,7 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Leashab
             this.resetAmbientSoundTime();
             this.playAmbientSound();
         }
+        if (io.canvasmc.canvas.Config.INSTANCE.fixes.mc183990 && this.target != null && this.target.isDeadOrDying()) this.setTarget(null, org.bukkit.event.entity.EntityTargetEvent.TargetReason.TARGET_DIED); // Canvas - MC-183990
     }
 
     @Override
