From 60fbf10de1ffff26663053a967c13ae1e63a5a7a Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/vehicle/boat/AbstractBoat.java b/net/minecraft/world/entity/vehicle/boat/AbstractBoat.java
index de254df4f13dd1544481fc8a3762ced649930da8..0816b2aeae0c263eff78e084885d3c1ae23bf6dc 100644
--- a/net/minecraft/world/entity/vehicle/boat/AbstractBoat.java
+++ b/net/minecraft/world/entity/vehicle/boat/AbstractBoat.java
@@ -782,7 +782,7 @@ public abstract class AbstractBoat extends VehicleEntity implements Leashable {
 
     @Override
     protected boolean canAddPassenger(final Entity passenger) {
-        return this.getPassengers().size() < this.getMaxPassengers() && !this.isEyeInFluid(FluidTags.WATER);
+        return this.getPassengers().size() < this.getMaxPassengers() && !this.canvas$isEyeInWater(); // Canvas - optimize fluidOnEyes
     }
 
     protected int getMaxPassengers() {
