From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/commands/Commands.java b/net/minecraft/commands/Commands.java
index 434563cf7958aa4fec624293968d1236cb351c2e..2c4ecdd40ccc7b4b097bbcadbb25388d1eebc17a 100644
--- a/net/minecraft/commands/Commands.java
+++ b/net/minecraft/commands/Commands.java
@@ -261,6 +261,7 @@ public class Commands {
         WaypointCommand.register(this.dispatcher, context);
         WeatherCommand.register(this.dispatcher);
         WorldBorderCommand.register(this.dispatcher);
+        io.canvasmc.canvas.command.RootCommandTree.INSTANCE.build(this.dispatcher); // Canvas - Rebrand - register commands
         if (JvmProfiler.INSTANCE.isAvailable()) {
             JfrCommand.register(this.dispatcher);
         }
