From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/raid/Raid.java b/net/minecraft/world/entity/raid/Raid.java
index dcdd6f230ff153a7b58d13b963d32376b1905b72..a1f9bbe2850776a154fce4a77bb688077f26f78c 100644
--- a/net/minecraft/world/entity/raid/Raid.java
+++ b/net/minecraft/world/entity/raid/Raid.java
@@ -548,7 +548,7 @@ public class Raid {
             double d1 = vec3.z + 13.0 / squareRoot * (vec31.z - vec3.z);
             if (squareRoot <= 64.0 || players.contains(serverPlayer)) {
                 serverPlayer.connection
-                    .send(new ClientboundSoundPacket(SoundEvents.RAID_HORN, SoundSource.NEUTRAL, d, serverPlayer.getY(), d1, 64.0F, 1.0F, randomLong));
+                    .send(new ClientboundSoundPacket(SoundEvents.RAID_HORN, io.canvasmc.canvas.Config.INSTANCE.fixes.mc245394 ? SoundSource.HOSTILE : SoundSource.NEUTRAL, d, serverPlayer.getY(), d1, 64.0F, 1.0F, randomLong)); // Canvas - MC-245394
             }
         }
     }
