--- a/net/minecraft/world/entity/npc/villager/Villager.java
+++ b/net/minecraft/world/entity/npc/villager/Villager.java
@@ -863,7 +_,15 @@
 
     private int countFoodPointsInInventory() {
         SimpleContainer inventory = this.getInventory();
-        return FOOD_POINTS.entrySet().stream().mapToInt(entry -> inventory.countItem(entry.getKey()) * entry.getValue()).sum();
+        // Canvas start - remove streams
+        int sum = 0;
+        for (Map.Entry<Item, Integer> entry : FOOD_POINTS.entrySet()) {
+            int i = inventory.countItem(entry.getKey()) * entry.getValue();
+            sum += i;
+        }
+        return sum;
+        // return FOOD_POINTS.entrySet().stream().mapToInt(entry -> inventory.countItem(entry.getKey()) * entry.getValue()).sum();
+        // Canvas end - remove streams
     }
 
     public boolean hasFarmSeeds() {
