--- a/net/minecraft/world/entity/npc/villager/Villager.java
+++ b/net/minecraft/world/entity/npc/villager/Villager.java
@@ -150,6 +_,32 @@
     private long lastGossipDecayTime;
     private int lastRestockCheckDayTime;
+    // Canvas start - lobotomize villagers
+    private boolean lobotomized = false;
+
+    private boolean checkLobotomized() {
+        int interval = io.canvasmc.canvas.Config.INSTANCE.entities.villagers.lobotomizeCheckInterval;
+        if (this.tickCount % interval != 0) {
+            return this.lobotomized;
+        }
+        this.lobotomized = false;
+
+        if (this.isPassenger() || !this.onGround()) {
+            return false;
+        }
+
+        net.minecraft.world.phys.AABB aabb = this.getBoundingBox().inflate(0.0625D, 0.0625D, 0.0625D);
+        for (net.minecraft.world.phys.shapes.VoxelShape shape : this.level().getBlockCollisions(this, aabb)) {
+            if (!shape.isEmpty()) {
+                this.lobotomized = true;
+                break;
+            }
+        }
+
+        return this.lobotomized;
+    }
+    // Canvas end - lobotomize villagers
+
     public Villager(EntityType<? extends Villager> type, Level world) {
         this(type, world, VillagerType.PLAINS);
     }
@@ -293,10 +319,13 @@
     }
     protected void customServerAiStep(ServerLevel level, final boolean inactive) {
         // Paper end - EAR 2
-        if (!inactive) this.getBrain().tick(level, this); // Paper - EAR 2
+        // Canvas start - lobotomize villagers
+        if (!inactive && (!io.canvasmc.canvas.Config.INSTANCE.entities.villagers.lobotomize || !this.checkLobotomized())) {
+            this.getBrain().tick(level, this);
+        }
+        // Canvas end - lobotomize villagers
         if (this.assignProfessionWhenSpawned) {
             this.assignProfessionWhenSpawned = false;
@@ -851,7 +_,15 @@
 
     private int countFoodPointsInInventory() {
         SimpleContainer inventory = this.getInventory();
-        return FOOD_POINTS.entrySet().stream().mapToInt(entry -> inventory.countItem(entry.getKey()) * entry.getValue()).sum();
+        // Canvas start - remove streams
+        int sum = 0;
+        for (Map.Entry<Item, Integer> entry : FOOD_POINTS.entrySet()) {
+            int i = inventory.countItem(entry.getKey()) * entry.getValue();
+            sum += i;
+        }
+        return sum;
+        // return FOOD_POINTS.entrySet().stream().mapToInt(entry -> inventory.countItem(entry.getKey()) * entry.getValue()).sum();
+        // Canvas end - remove streams
     }
 
     public boolean hasFarmSeeds() {
