From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/storage/PrimaryLevelData.java b/net/minecraft/world/level/storage/PrimaryLevelData.java
index 8936b4b1ae9c82eed2a7d3191a29f01792438d14..396c0e244c37fd6939b6540644ec3f721134b8ef 100644
--- a/net/minecraft/world/level/storage/PrimaryLevelData.java
+++ b/net/minecraft/world/level/storage/PrimaryLevelData.java
@@ -72,6 +72,7 @@ public class PrimaryLevelData implements ServerLevelData, WorldData {
     private boolean wasModded;
     private final Set<String> removedFeatureFlags;
     private final TimerQueue<MinecraftServer> scheduledEvents;
+    public final io.canvasmc.canvas.world.PerWorldDistanceConfig canvas$distanceConfig; // Canvas - per world distance
 
     // CraftBukkit start - Add world and pdc
     public net.minecraft.core.Registry<net.minecraft.world.level.dimension.LevelStem> customDimensions;
@@ -115,6 +116,7 @@ public class PrimaryLevelData implements ServerLevelData, WorldData {
         WorldOptions worldOptions,
         PrimaryLevelData.SpecialWorldProperty specialWorldProperty,
         Lifecycle worldGenSettingsLifecycle
+        , io.canvasmc.canvas.world.PerWorldDistanceConfig canvas$distanceConfig // Canvas - per world distance
     ) {
         this.wasModded = wasModded;
         this.respawnData = respawnData;
@@ -142,6 +144,7 @@ public class PrimaryLevelData implements ServerLevelData, WorldData {
         this.worldOptions = worldOptions;
         this.specialWorldProperty = specialWorldProperty;
         this.worldGenSettingsLifecycle = worldGenSettingsLifecycle;
+        this.canvas$distanceConfig = canvas$distanceConfig; // Canvas - per world distance
     }
 
     public PrimaryLevelData(
@@ -174,6 +177,7 @@ public class PrimaryLevelData implements ServerLevelData, WorldData {
             worldOptions,
             specialWorldProperty,
             worldGenSettingsLifecycle
+            , io.canvasmc.canvas.world.PerWorldDistanceConfig.DEFAULT // Canvas - per world distance
         );
     }
 
@@ -218,6 +222,7 @@ public class PrimaryLevelData implements ServerLevelData, WorldData {
             worldOptions,
             specialWorldProperty,
             worldGenSettingsLifecycle
+            , tag.get("Canvas.Distance").read(io.canvasmc.canvas.world.PerWorldDistanceConfig.CODEC).resultOrPartial(LOGGER::error).orElse(io.canvasmc.canvas.world.PerWorldDistanceConfig.DEFAULT) // Canvas - per world distance
         );
         // Paper start
         data.respawnDimension = tag.get(PAPER_RESPAWN_DIMENSION)
@@ -293,6 +298,7 @@ public class PrimaryLevelData implements ServerLevelData, WorldData {
         tag.storeNullable("WanderingTraderId", UUIDUtil.CODEC, this.wanderingTraderId);
         tag.putString("Bukkit.Version", org.bukkit.Bukkit.getName() + "/" + org.bukkit.Bukkit.getVersion() + "/" + org.bukkit.Bukkit.getBukkitVersion()); // CraftBukkit
         this.world.getWorld().storeBukkitValues(tag); // CraftBukkit - add pdc
+        tag.store("Canvas.Distance", io.canvasmc.canvas.world.PerWorldDistanceConfig.CODEC, this.canvas$distanceConfig); // Canvas - per world distance
     }
 
     private static ListTag stringCollectionToTag(Set<String> stringCollection) {
