--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -851,6 +_,7 @@
      * @param save whether to save the chunks before unloading
      * @return true if successful, false otherwise
      */
+    @Deprecated(forRemoval = true) // Canvas - rewrite world load and unload
     public static boolean unloadWorld(@NotNull String name, boolean save) {
         return server.unloadWorld(name, save);
     }
@@ -865,9 +_,34 @@
      * @param save whether to save the chunks before unloading
      * @return true if successful, false otherwise
      */
+    @Deprecated(forRemoval = true) // Canvas - rewrite world load and unload
     public static boolean unloadWorld(@NotNull World world, boolean save) {
         return server.unloadWorld(world, save);
     }
+    // Canvas start - rewrite world load and unload
+
+    /**
+     * Unloads a world asynchronously with an optional callback result
+     *
+     * @param name Name of the world to unload
+     * @param save whether to save the chunks before unloading
+     * @param callback the result consumer for when the world unload is completed
+     */
+    public static void unloadWorldAsync(@NotNull String name, boolean save, final @NotNull Consumer<Boolean> callback) {
+        server.unloadWorldAsync(name, save, callback);
+    }
+
+    /**
+     * Unloads a world asynchronously with an optional callback result
+     *
+     * @param world the world to unload
+     * @param save whether to save the chunks before unloading
+     * @param callback the result consumer for when the world unload is completed
+     */
+    public static void unloadWorldAsync(@NotNull World world, boolean save, final @NotNull Consumer<Boolean> callback) {
+        server.unloadWorldAsync(world, save, callback);
+    }
+    // Canvas end - rewrite world load and unload
 
     /**
      * Gets the world with the given name.
@@ -3036,6 +_,11 @@
         return server.getRegionTPS(world, chunkX, chunkZ);
     }
     // Folia end - region TPS API
+    // Canvas start - region threading
+    <T> io.canvasmc.canvas.region.RegionTickData.IRegionizedData<T> createRegionizedData(java.util.function.BiFunction<io.canvasmc.canvas.region.RegionTickData, org.bukkit.World, T> supplier, io.canvasmc.canvas.region.RegionTickData.IRegionizedData.IRegionizedCallback<T> callback) {
+        return server.createRegionizedData(supplier, callback);
+    }
+    // Canvas end - region threading
 
     /**
      * @deprecated All methods on this class have been deprecated, see the individual methods for replacements.
