--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -838,6 +_,7 @@
      * @param save whether to save the chunks before unloading
      * @return true if successful, false otherwise
      */
+    @Deprecated(forRemoval = true) // Canvas - rewrite world load and unload
     public static boolean unloadWorld(@NotNull String name, boolean save) {
         return server.unloadWorld(name, save);
     }
@@ -852,9 +_,34 @@
      * @param save whether to save the chunks before unloading
      * @return true if successful, false otherwise
      */
+    @Deprecated(forRemoval = true) // Canvas - rewrite world load and unload
     public static boolean unloadWorld(@NotNull World world, boolean save) {
         return server.unloadWorld(world, save);
     }
+    // Canvas start - rewrite world load and unload
+
+    /**
+     * Unloads a world asynchronously with an optional callback result
+     *
+     * @param name Name of the world to unload
+     * @param save whether to save the chunks before unloading
+     * @param callback the result consumer for when the world unload is completed
+     */
+    public static void unloadWorldAsync(@NotNull String name, boolean save, final @NotNull Consumer<io.canvasmc.canvas.WorldUnloadResult> callback) {
+        server.unloadWorldAsync(name, save, callback);
+    }
+
+    /**
+     * Unloads a world asynchronously with an optional callback result
+     *
+     * @param world the world to unload
+     * @param save whether to save the chunks before unloading
+     * @param callback the result consumer for when the world unload is completed
+     */
+    public static void unloadWorldAsync(@NotNull World world, boolean save, final @NotNull Consumer<io.canvasmc.canvas.WorldUnloadResult> callback) {
+        server.unloadWorldAsync(world, save, callback);
+    }
+    // Canvas end - rewrite world load and unload
 
     /**
      * Gets the world with the given name.
@@ -2989,6 +_,29 @@
     public static boolean isGlobalTickThread() {
         return server.isGlobalTickThread();
     }
+    // Canvas start - expand region threading api
+
+    /**
+     * Returns whether the current thread is ticking a region and that the region being ticked
+     * owns the specified world and area in the world defined by the bounding box provided.
+     * @param world the world
+     * @param aabb the bounding box
+     */
+    public static boolean isOwnedByCurrentRegion(@NotNull World world, @NotNull org.bukkit.util.BoundingBox aabb) {
+        return server.isOwnedByCurrentRegion(world, aabb);
+    }
+
+    /**
+     * Returns whether the current thread is ticking a region and that the region being ticked
+     * owns the specified position that is adjusted as specified by the velocity provided, with a chunk buffer
+     * @param position the initial position
+     * @param velocity the velocity
+     * @param buffer the chunk buffer
+     */
+    public static boolean isOwnedByCurrentRegion(@NotNull Location position, @NotNull org.bukkit.util.Vector velocity, int buffer) {
+        return server.isOwnedByCurrentRegion(position, velocity, buffer);
+    }
+    // Canvas end - expand region threading api
     // Paper end - Folia region threading API
     // Folia start - region TPS API
     /**
@@ -3026,6 +_,39 @@
         return server.getRegionTPS(world, chunkX, chunkZ);
     }
     // Folia end - region TPS API
+    // Canvas start - region threading
+
+    /**
+     * Creates a new regionized data instance
+     * @param supplier the creator for tick region data and a world -> a new regionized data instance
+     * @param callback the split and merge callbacks for the region data
+     * @return a new regionized data instance
+     * @param <T> the generic type of your regionized data
+     */
+    public static <T> io.canvasmc.canvas.region.RegionTickData.IRegionizedData<T> createRegionizedData(java.util.function.BiFunction<io.canvasmc.canvas.region.RegionTickData, org.bukkit.World, T> supplier, io.canvasmc.canvas.region.RegionTickData.IRegionizedData.IRegionizedCallback<T> callback) {
+        return server.createRegionizedData(supplier, callback);
+    }
+
+    /**
+     * Gets or creates the current thread regionized data based off the {@link io.canvasmc.canvas.region.RegionTickData.IRegionizedData} provided
+     * @param regionizedData the regionized data
+     * @return the fetched or created regionized data
+     * @param <T> the generic type of your regionized data
+     * @throws java.lang.IllegalStateException if not on a tick thread runner, or if the current tick runner doesn't own a region
+     */
+    public static <T> T getLocalRegionizedData(io.canvasmc.canvas.region.RegionTickData.IRegionizedData<T> regionizedData) {
+        return server.getLocalRegionizedData(regionizedData);
+    }
+
+    /**
+     * Gets the region threading tick rate manager. This offers the same capabilities as {@link Bukkit#getServerTickManager()}
+     * but is compatible with Folia's region threading structure
+     * @return the server region threading tick manager
+     */
+    public static io.canvasmc.canvas.region.RegionThreadingTickManager getRegionThreadingTickRateManager() {
+        return server.getRegionThreadingTickRateManager();
+    }
+    // Canvas end - region threading
 
     /**
      * @deprecated All methods on this class have been deprecated, see the individual methods for replacements.
