--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -1,6 +_,7 @@
 package org.bukkit;
 
 import java.io.File;
+import io.canvasmc.canvas.LevelAccess;
 import io.papermc.paper.raytracing.PositionedRayTraceConfigurationBuilder;
 import org.bukkit.generator.ChunkGenerator;
 
@@ -4452,6 +_,15 @@
      */
     @NotNull
     public Collection<GeneratedStructure> getStructures(int x, int z, @NotNull Structure structure);
+    // Canvas start
+
+    /**
+     * Gets the {@link LevelAccess} for this world, allowing modification of its threaded context
+     * @return the level accessor corresponding to this world
+     */
+    @NotNull
+    public LevelAccess getLevelAccess();
+    // Canvas end
 
     /**
      * Represents various map environment types that a world may be
