--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -4473,4 +_,25 @@
             }
         }
     }
+    // Canvas start - region threading
+
+    /**
+     * Fetches the non-null regionizer for this world
+     * @return the regionizer
+     */
+    @NotNull
+    io.canvasmc.canvas.region.WorldRegionizer getRegionizer();
+
+    boolean isTickThreadFor(@NotNull Block block);
+    boolean isTickThreadFor(@NotNull Block block, int blockRadius);
+    boolean isTickThreadFor(@NotNull Chunk chunk);
+    boolean isTickThreadFor(Location position);
+    boolean isTickThreadFor(int chunkX, int chunkZ);
+    boolean isTickThreadFor(BoundingBox aabb);
+    boolean isTickThreadFor(double blockX, double blockZ);
+    boolean isTickThreadFor(Location position, Vector velocity, int buffer);
+    boolean isTickThreadFor(int fromChunkX, int fromChunkZ, int toChunkX, int toChunkZ);
+    boolean isTickThreadFor(int chunkX, int chunkZ, int radius);
+    boolean isTickThreadFor(Entity entity);
+    // Canvas end - region threading
 }
