--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -4691,4 +_,13 @@
             }
         }
     }
+    // Canvas start - region threading
+
+    /**
+     * Fetches the non-null regionizer for this world
+     * @return the regionizer
+     */
+    @NotNull
+    io.canvasmc.canvas.region.WorldRegionizer getRegionizer();
+    // Canvas end - region threading
 }
