From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/storage/LevelData.java b/net/minecraft/world/level/storage/LevelData.java
index 7f46f4529de72f2a0e3b38c4a8a401d3fee93829..85200e9ae711e8f5d9decab411e62f9360b440bd 100644
--- a/net/minecraft/world/level/storage/LevelData.java
+++ b/net/minecraft/world/level/storage/LevelData.java
@@ -31,7 +31,7 @@ public interface LevelData {
     }
 
     record RespawnData(GlobalPos globalPos, float yaw, float pitch) {
-        public static final LevelData.RespawnData DEFAULT = new LevelData.RespawnData(GlobalPos.of(Level.OVERWORLD, BlockPos.ZERO), 0.0F, 0.0F);
+        public static final LevelData.RespawnData DEFAULT = new LevelData.RespawnData(GlobalPos.of(io.canvasmc.canvas.GlobalConfiguration.fetchRespawnDimensionKey(), BlockPos.ZERO), 0.0F, 0.0F); // Canvas - default respawn dimension config
         public static final MapCodec<LevelData.RespawnData> MAP_CODEC = RecordCodecBuilder.mapCodec(
             i -> i.group(
                     GlobalPos.MAP_CODEC.forGetter(LevelData.RespawnData::globalPos),
