--- a/net/minecraft/world/entity/animal/allay/AllayAi.java
+++ b/net/minecraft/world/entity/animal/allay/AllayAi.java
@@ -114,6 +_,14 @@
         Optional<GlobalPos> memory = brain.getMemory(MemoryModuleType.LIKED_NOTEBLOCK_POSITION);
         if (memory.isPresent()) {
             GlobalPos globalPos = memory.get();
+            // Canvas start - region threading
+            if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(
+                net.minecraft.server.MinecraftServer.getServer().getLevel(globalPos.dimension()), globalPos.pos()
+            )) {
+                brain.eraseMemory(MemoryModuleType.LIKED_NOTEBLOCK_POSITION);
+                return Optional.empty();
+            }
+            // Canvas end - region threading
             if (shouldDepositItemsAtLikedNoteblock(entity, brain, globalPos)) {
                 return Optional.of(new BlockPosTracker(globalPos.pos().above()));
             }
