--- a/net/minecraft/world/level/material/FluidState.java
+++ b/net/minecraft/world/level/material/FluidState.java
@@ -166,6 +_,12 @@
     public Stream<TagKey<Fluid>> getTags() {
         return this.owner.builtInRegistryHolder().tags();
     }
+    // Canvas start - optimize fluidOnEyes - remove stream
+
+    public TagKey<Fluid>[] canvas$getTagsArray() {
+        return this.owner.builtInRegistryHolder().canvas$tagsArray;
+    }
+    // Canvas end - optimize fluidOnEyes - remove stream
 
     public void entityInside(Level level, BlockPos pos, Entity entity, InsideBlockEffectApplier effectApplier) {
         this.getType().entityInside(level, pos, entity, effectApplier);
