--- a/net/minecraft/world/level/material/LavaFluid.java
+++ b/net/minecraft/world/level/material/LavaFluid.java
@@ -2,6 +_,8 @@
 
 import java.util.Optional;
 import javax.annotation.Nullable;
+import io.canvasmc.canvas.Config;
+import io.canvasmc.canvas.util.TPSUtil;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.core.particles.ParticleOptions;
@@ -177,8 +_,19 @@
 
     @Override
     public int getTickDelay(LevelReader level) {
-        return level.dimensionType().ultraWarm() ? level.getWorldBorder().world.purpurConfig.lavaSpeedNether : level.getWorldBorder().world.purpurConfig.lavaSpeedNotNether; // Purpur - Make lava flow speed configurable
-    }
+    // Canvas start
+        int original = level.dimensionType().ultraWarm() ? level.getWorldBorder().world.purpurConfig.lavaSpeedNether : level.getWorldBorder().world.purpurConfig.lavaSpeedNotNether; // Purpur - Make lava flow speed configurable
+        if (level instanceof ServerLevel serverLevel) {
+            return lagCompensation(original, serverLevel);
+        }
+        return original;
+    }
+
+    private int lagCompensation(int original, ServerLevel level) {
+        if (!Config.INSTANCE.lagCompensation.enabled || !Config.INSTANCE.lagCompensation.fluidAcceleration) return original;
+        return TPSUtil.tt20(original, true, level);
+    }
+    // Canvas end
 
     @Override
     public int getSpreadDelay(Level level, BlockPos pos, FluidState currentState, FluidState newState) {
@@ -224,6 +_,7 @@
                     // CraftBukkit end
                 }
 
+                level.setBlock(pos, Blocks.STONE.defaultBlockState(), 3); // Canvas - ensure block is placed
                 this.fizz(level, pos);
                 return;
             }
