--- a/net/minecraft/world/level/material/WaterFluid.java
+++ b/net/minecraft/world/level/material/WaterFluid.java
@@ -2,6 +_,8 @@
 
 import java.util.Optional;
 import javax.annotation.Nullable;
+import io.canvasmc.canvas.Config;
+import io.canvasmc.canvas.util.TPSUtil;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.core.particles.ParticleOptions;
@@ -115,8 +_,18 @@
 
     @Override
     public int getTickDelay(LevelReader level) {
+    // Canvas start
+        if (level instanceof ServerLevel serverLevel) {
+            return lagCompensation(serverLevel);
+        }
         return 5;
     }
+
+    private int lagCompensation(ServerLevel level) {
+        if (!Config.INSTANCE.lagCompensation.enabled || !Config.INSTANCE.lagCompensation.fluidAcceleration) return 5;
+        return TPSUtil.tt20(5, true, level);
+    }
+    // Canvas end
 
     @Override
     public boolean canBeReplacedWith(FluidState fluidState, BlockGetter blockReader, BlockPos pos, Fluid fluid, Direction direction) {
