From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/levelgen/material/MaterialRuleList.java b/net/minecraft/world/level/levelgen/material/MaterialRuleList.java
index 2be3314197ffdaf3f25ea211c997db15a450915b..0c09d93747c66db2b84d0b3889aa39943a38d501 100644
--- a/net/minecraft/world/level/levelgen/material/MaterialRuleList.java
+++ b/net/minecraft/world/level/levelgen/material/MaterialRuleList.java
@@ -8,13 +8,16 @@ import org.jspecify.annotations.Nullable;
 public record MaterialRuleList(NoiseChunk.BlockStateFiller[] materialRuleList) implements NoiseChunk.BlockStateFiller {
     @Override
     public @Nullable BlockState calculate(final DensityFunction.FunctionContext context) {
-        for (NoiseChunk.BlockStateFiller rule : this.materialRuleList) {
-            BlockState state = rule.calculate(context);
-            if (state != null) {
-                return state;
-            }
-        }
+        // Canvas start - reduce worldgen allocations
+        // avoid iterator allocation
+        BlockState state = null;
+        int ruleCount = this.materialRuleList.length;
 
-        return null;
+        for (int i = 0; state == null && i < ruleCount; i++) {
+            NoiseChunk.BlockStateFiller rule = this.materialRuleList[i];
+            state = rule.calculate(context);
+        }
+        return state;
+        // Canvas end - reduce worldgen allocations
     }
 }
