From 53e1a57c72953de1bfcc94c3257c724e7fa1f88f Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/levelgen/Beardifier.java b/net/minecraft/world/level/levelgen/Beardifier.java
index c19a4c65f6ef9a953440046f27a6e7bbee65eb8d..71c1502ffa1e3970be79d5ac3fdd51762d09bc61 100644
--- a/net/minecraft/world/level/levelgen/Beardifier.java
+++ b/net/minecraft/world/level/levelgen/Beardifier.java
@@ -34,6 +34,15 @@ public class Beardifier implements DensityFunctions.BeardifierOrMarker {
     private final List<Beardifier.Rigid> pieces;
     private final List<JigsawJunction> junctions;
     private final @Nullable BoundingBox affectedBox;
+    // Canvas start - optimize beardifier
+    private Beardifier.Rigid[] c2me$pieceArray;
+    private JigsawJunction[] c2me$junctionArray;
+
+    private void c2me$initArrays() {
+        this.c2me$pieceArray = this.pieces.toArray(Beardifier.Rigid[]::new);
+        this.c2me$junctionArray = this.junctions.toArray(JigsawJunction[]::new);
+    }
+    // Canvas end - optimize beardifier
 
     public static Beardifier forStructuresInChunk(final StructureManager structureManager, final ChunkPos chunkPos) {
         List<StructureStart> structureStarts = structureManager.startsForStructure(chunkPos, s -> s.terrainAdaptation() != TerrainAdjustment.NONE);
@@ -109,6 +118,54 @@ public class Beardifier implements DensityFunctions.BeardifierOrMarker {
 
     @Override
     public double compute(final DensityFunction.FunctionContext context) {
+        // Canvas start - optimize beardifier
+        if (io.canvasmc.canvas.GlobalConfiguration.getInstance().chunkSystem.optimizeBeardifier) {
+            if (this.affectedBox == null) {
+                return 0.0;
+            }
+
+            int i = context.blockX();
+            int j = context.blockY();
+            int k = context.blockZ();
+
+            if (!this.affectedBox.isInside(i, j, k)) {
+                return 0.0;
+            }
+
+            if (this.c2me$pieceArray == null || this.c2me$junctionArray == null) {
+                this.c2me$initArrays();
+            }
+
+            double d = 0.0;
+
+
+            for (Rigid piece : this.c2me$pieceArray) {
+                BoundingBox blockBox = piece.box();
+                int l = piece.groundLevelDelta();
+                int m = Math.max(0, Math.max(blockBox.minX() - i, i - blockBox.maxX()));
+                int n = Math.max(0, Math.max(blockBox.minZ() - k, k - blockBox.maxZ()));
+                int o = blockBox.minY() + l;
+                int p = j - o;
+
+                d += switch (piece.terrainAdjustment()) {
+                    case NONE -> 0.0;
+                    case BURY -> getBuryContribution(m, (double)p / 2.0, n);
+                    case BEARD_THIN -> getBeardContribution(m, p, n, p) * 0.8;
+                    case BEARD_BOX -> getBeardContribution(m, Math.max(0, Math.max(o - j, j - blockBox.maxY())), n, p) * 0.8;
+                    case ENCAPSULATE -> getBuryContribution((double)m / 2.0, (double)Math.max(0, Math.max(blockBox.minY() - j, j - blockBox.maxY())) / 2.0, (double)n / 2.0) * 0.8;
+                };
+            }
+
+            for (JigsawJunction jigsawJunction : this.c2me$junctionArray) {
+                int r = i - jigsawJunction.getSourceX();
+                int l = j - jigsawJunction.getSourceGroundY();
+                int m = k - jigsawJunction.getSourceZ();
+                d += getBeardContribution(r, l, m, l) * 0.4;
+            }
+
+            return d;
+        }
+        // Canvas end - optimize beardifier
         if (this.affectedBox == null) {
             return 0.0;
         } else {
@@ -166,6 +223,16 @@ public class Beardifier implements DensityFunctions.BeardifierOrMarker {
     }
 
     private static double getBuryContribution(final double dx, final double dy, final double dz) {
+        // Canvas start - optimize beardifier
+        if (io.canvasmc.canvas.GlobalConfiguration.getInstance().chunkSystem.optimizeBeardifier) {
+            double d = Math.sqrt(dx * dx + dy * dy + dz * dz);
+            if (d > 6.0) {
+                return 0.0;
+            } else {
+                return 1.0 - d / 6.0;
+            }
+        }
+        // Canvas end - optimize beardifier
         double distance = Mth.length(dx, dy, dz);
         return Mth.clampedMap(distance, 0.0, 6.0, 1.0, 0.0);
     }
