From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/levelgen/NoiseChunk.java b/net/minecraft/world/level/levelgen/NoiseChunk.java
index b7ec70ef61a9d20cfc1d8150c630d052ea6e2aec..fbcc7b5b5f9c30eb7b3a14823cb8a6cdc7f0af29 100644
--- a/net/minecraft/world/level/levelgen/NoiseChunk.java
+++ b/net/minecraft/world/level/levelgen/NoiseChunk.java
@@ -377,7 +377,17 @@ public class NoiseChunk implements DensityFunction.FunctionContext, DensityFunct
     }
 
     protected DensityFunction wrap(final DensityFunction function) {
-        return this.wrapped.computeIfAbsent(function, this::wrapNew);
+        // Canvas start - reduce worldgen allocations
+        // avoid lambda allocation
+        DensityFunction func = this.wrapped.get(function);
+
+        if (func == null) {
+            func = this.wrapNew(function);
+            this.wrapped.put(function, func);
+        }
+
+        return func;
+        // Canvas end - reduce worldgen allocations
     }
 
     private DensityFunction wrapNew(final DensityFunction function) {
