From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/levelgen/SurfaceRules.java b/net/minecraft/world/level/levelgen/SurfaceRules.java
index 766452f7b5935d435eafc7089297a482bda9a980..ece430c948a79252e1d95ad947a6b7f7b6da4cec 100644
--- a/net/minecraft/world/level/levelgen/SurfaceRules.java
+++ b/net/minecraft/world/level/levelgen/SurfaceRules.java
@@ -592,8 +592,13 @@ public class SurfaceRules {
     private record SequenceRule(List<SurfaceRules.SurfaceRule> rules) implements SurfaceRules.SurfaceRule {
         @Override
         public @Nullable BlockState tryApply(final int blockX, final int blockY, final int blockZ) {
-            for (SurfaceRules.SurfaceRule rule : this.rules) {
-                BlockState state = rule.tryApply(blockX, blockY, blockZ);
+            // Canvas start - reduce worldgen allocations
+            // avoid iterator allocation
+            int size = this.rules.size();
+            //noinspection ForLoopReplaceableByForEach
+            for (int i = 0; i < size; i++) {
+                BlockState state = this.rules.get(i).tryApply(blockX, blockY, blockZ);
+            // Canvas end - reduce worldgen allocations
                 if (state != null) {
                     return state;
                 }
