From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/block/Blocks.java b/net/minecraft/world/level/block/Blocks.java
index 18bb49d16c9066973768fc8e9789573a06c1f23b..0b4e11ed6113f267a4a89948b371cd03dd1309af 100644
--- a/net/minecraft/world/level/block/Blocks.java
+++ b/net/minecraft/world/level/block/Blocks.java
@@ -7218,6 +7218,11 @@ public class Blocks {
     }
 
     private static Block register(final String id, final BlockBehaviour.Properties properties) {
+        // Canvas start - MC-223153
+        if (io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaFixes.mc223153 && id.equalsIgnoreCase("raw_copper_block")) {
+            return register(id, Block::new, properties.sound(SoundType.COPPER));
+        }
+        // Canvas end - MC-223153
         return register(id, Block::new, properties);
     }
 
