From 53e1a57c72953de1bfcc94c3257c724e7fa1f88f Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/block/CactusBlock.java b/net/minecraft/world/level/block/CactusBlock.java
index 1fbdb7a9ae8120cbe43cb185ad17c4769cf6e8b6..4fb031c33b2482ce732785a88d6598dc4c301f17 100644
--- a/net/minecraft/world/level/block/CactusBlock.java
+++ b/net/minecraft/world/level/block/CactusBlock.java
@@ -52,24 +52,45 @@ public class CactusBlock extends Block {
 
     @Override
     protected void randomTick(final BlockState state, final ServerLevel level, final BlockPos pos, final RandomSource random) {
-        BlockPos above = pos.above();
-        if (level.isEmptyBlock(above)) {
+        // Leaf start - decrease chunk/block lookups
+        final int x = pos.getX();
+        final int z = pos.getZ();
+        net.minecraft.world.level.chunk.LevelChunk chunk = level.getChunkIfLoaded(x >> 4, z >> 4);
+        if (chunk == null) return;
+
+        final int y = pos.getY();
+        if (chunk.getBlockState(x, y + 1, z).isAir()) {
+        // Leaf end - decrease chunk/block lookups
             int height = 1;
             int age = state.getValue(AGE);
 
-            while (level.getBlockState(pos.below(height)).is(this)) {
+            while (chunk.getBlockState(pos.below(height)).is(this)) { // Leaf - decrease chunk/block lookups
                 if (++height == level.paperConfig().maxGrowthHeight.cactus && age == 15) { // Paper - Configurable cactus/bamboo/reed growth height
                     return;
                 }
             }
 
-            if (age == 8 && this.canSurvive(this.defaultBlockState(), level, pos.above())) {
+            // Leaf start - decrease chunk/block lookups
+            BlockPos above = null;
+            if (age == 8 && this.canSurvive(this.defaultBlockState(), level, above = pos.above())) {
+                // Leaf end - decrease chunk/block lookups
                 double chanceToGrowFlower = height >= level.paperConfig().maxGrowthHeight.cactus ? 0.25 : 0.1; // Paper - Configurable cactus/bamboo/reed growth height
                 if (random.nextDouble() <= chanceToGrowFlower) {
                     org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, above, Blocks.CACTUS_FLOWER.defaultBlockState(), Block.UPDATE_ALL); // Paper - block grow event
                 }
             } else if (age == 15 && height < level.paperConfig().maxGrowthHeight.cactus) { // Paper - Configurable cactus/bamboo/reed growth height
-                if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, above, this.defaultBlockState(), Block.UPDATE_ALL)) return; // Paper - block grow event
+                // Leaf start - check if the cactus can even survive being placed
+                if (level.canvasConfig().cactusCheckSurvivalBeforeGrowth && !canSurvive(level, above = pos.above())) { // Leaf - Decrease chunk/block lookups
+                    level.levelEvent(LevelEvent.PARTICLES_DESTROY_BLOCK, above, Block.getId(state));
+                    // We're going to fake the block breaking to match vanilla standards.
+                    for (net.minecraft.world.item.ItemStack drop : Block.getDrops(state, level, pos, null)) { // Use base cactus since we don't place a block
+                        Block.popResource(level, above, drop);
+                    }
+                    level.setBlock(pos, state.setValue(CactusBlock.AGE, 0), Block.UPDATE_NONE);
+                    return;
+                }
+                // Leaf end - check if the cactus can even survive being placed
+                if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, above == null ? above = pos.above() : above, this.defaultBlockState(), Block.UPDATE_ALL)) return; // Paper - block grow event // Leaf - decrease chunk/block lookups
                 BlockState aboveBlock = state.setValue(AGE, 0);
                 level.setBlock(pos, aboveBlock, Block.UPDATE_NONE);
                 level.neighborChanged(aboveBlock, above, this, null, false);
@@ -111,6 +132,11 @@ public class CactusBlock extends Block {
 
     @Override
     protected boolean canSurvive(final BlockState state, final LevelReader level, final BlockPos pos) {
+    // Leaf start - check if the cactus can even survive being placed
+        return canSurvive(level, pos);
+    }
+    protected boolean canSurvive(final LevelReader level, final BlockPos pos) {
+    // Leaf end - check if the cactus can even survive being placed
         for (Direction direction : Direction.Plane.HORIZONTAL) {
             BlockState neighbor = level.getBlockState(pos.relative(direction));
             if (neighbor.isSolid() || level.getFluidState(pos.relative(direction)).is(FluidTags.LAVA)) {
