From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/block/ChestBlock.java b/net/minecraft/world/level/block/ChestBlock.java
index 599a11344c5dcfe1122ec10a89628490602572af..66608248dfa13aa1a37c168bc1be21017c49fcb5 100644
--- a/net/minecraft/world/level/block/ChestBlock.java
+++ b/net/minecraft/world/level/block/ChestBlock.java
@@ -92,6 +92,28 @@ public class ChestBlock extends AbstractChestBlock<ChestBlockEntity> implements
             return Optional.empty();
         }
     };
+    // Canvas start - optimize hoppers
+    private static final DoubleBlockCombiner.Combiner<ChestBlockEntity, Container> NULLABLE_DIRECT_CONTAINER = new DoubleBlockCombiner.Combiner<ChestBlockEntity, Container>() {
+        @Override
+        public Container acceptDouble(final ChestBlockEntity first, final ChestBlockEntity second) {
+            return new CompoundContainer(first, second);
+        }
+
+        @Override
+        public Container acceptSingle(final ChestBlockEntity single) {
+            return single;
+        }
+
+        @Override
+        public Container acceptNone() {
+            // the data flow issue is that all methods return non-null by default
+            // however when we use this we have nullability handling since we
+            // do expect the return value to be nullable
+            //noinspection DataFlowIssue
+            return null;
+        }
+    };
+    // Canvas end - optimize hoppers
     public static DoubleBlockCombiner.Combiner<ChestBlockEntity, Optional<MenuProvider>> MENU_PROVIDER_COMBINER = new DoubleBlockCombiner.Combiner<ChestBlockEntity, Optional<MenuProvider>>() {
         @Override
         public Optional<MenuProvider> acceptDouble(final ChestBlockEntity first, final ChestBlockEntity second) {
@@ -280,6 +302,16 @@ public class ChestBlock extends AbstractChestBlock<ChestBlockEntity> implements
         return this.blockEntityType.get();
     }
 
+    // Canvas start - optimize hoppers
+    public static @Nullable Container canvas$getContainerForHoppers(
+        final ChestBlock block, final BlockState state, final Level level, final BlockPos pos, final boolean ignoreBeingBlocked, final Container container
+    ) {
+        // when conducting tests, the code bellow the injected code took ~17.1ms out of the ~40.45ms total
+        // to tick the hoppers in our testing environment. ~20.14ms was from this method
+        return block.combine(state, level, pos, ignoreBeingBlocked).apply(NULLABLE_DIRECT_CONTAINER);
+    }
+
+    // Canvas end - optimize hoppers
     public static @Nullable Container getContainer(
         final ChestBlock block, final BlockState state, final Level level, final BlockPos pos, final boolean ignoreBeingBlocked
     ) {
@@ -348,7 +380,7 @@ public class ChestBlock extends AbstractChestBlock<ChestBlockEntity> implements
 
     public static boolean isBlockedChestByBlock(final BlockGetter level, final BlockPos pos) {
         BlockPos above = pos.above();
-        return level.getBlockState(above).isRedstoneConductor(level, above);
+        return (!(level instanceof net.minecraft.server.level.ServerLevel serverLevel) || !serverLevel.canvasConfig().blocks.chestsCanOpenWithFullBlockAbove) && level.getBlockState(above).isRedstoneConductor(level, above); // Canvas - block configs
     }
 
     private static boolean isCatSittingOnChest(final LevelAccessor level, final BlockPos pos) {
