From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/block/EnchantingTableBlock.java b/net/minecraft/world/level/block/EnchantingTableBlock.java
index d967111597076ecebd650c69637af6f3ad7b2e8b..425c63e2aa801727d2e66deaa42df8ba0b3984bc 100644
--- a/net/minecraft/world/level/block/EnchantingTableBlock.java
+++ b/net/minecraft/world/level/block/EnchantingTableBlock.java
@@ -45,8 +45,32 @@ public class EnchantingTableBlock extends BaseEntityBlock {
     }
 
     public static boolean isValidBookShelf(final Level level, final BlockPos pos, final BlockPos offset) {
-        return level.getBlockState(pos.offset(offset)).is(BlockTags.ENCHANTMENT_POWER_PROVIDER)
-            && level.getBlockState(pos.offset(offset.getX() / 2, offset.getY(), offset.getZ() / 2)).is(BlockTags.ENCHANTMENT_POWER_TRANSMITTER);
+        // Canvas start - block configs
+        final BlockState offsetState = level.getBlockState(pos.offset(offset));
+
+        // must be a power provider
+        if (!offsetState.is(BlockTags.ENCHANTMENT_POWER_PROVIDER)) {
+            if (level.canvasConfig().blocks.fullChiseledBookShelvesCountAsValidEnchantPowerSources
+                && offsetState.is(Blocks.CHISELED_BOOKSHELF)) {
+
+                // all slots have to be filled for it to be a valid bookshelf
+                for (final net.minecraft.world.level.block.state.properties.BooleanProperty slotOccupiedProperty : ChiseledBookShelfBlock.SLOT_OCCUPIED_PROPERTIES) {
+                    if (!offsetState.getValue(slotOccupiedProperty)) {
+                        return false;
+                    }
+                }
+
+            }
+            else return false;
+        }
+
+        // check if something is blocking the shelf
+        if (!level.getBlockState(pos.offset(offset.getX() / 2, offset.getY(), offset.getZ() / 2)).is(BlockTags.ENCHANTMENT_POWER_TRANSMITTER)) {
+            return false;
+        }
+
+        return true;
+        // Canvas end - block configs
     }
 
     @Override
