From 53e1a57c72953de1bfcc94c3257c724e7fa1f88f Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/block/FarmlandBlock.java b/net/minecraft/world/level/block/FarmlandBlock.java
index ef55faf83e39c8dfd6bbe33c4943ed88bc473715..2ca25e3b984806e2d07f02688512e6ef0c52ed76 100644
--- a/net/minecraft/world/level/block/FarmlandBlock.java
+++ b/net/minecraft/world/level/block/FarmlandBlock.java
@@ -40,7 +40,6 @@ public class FarmlandBlock extends Block {
 
     protected FarmlandBlock(final BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(MOISTURE, 0));
     }
 
     @Override
@@ -71,8 +70,17 @@ public class FarmlandBlock extends Block {
     public BlockState getStateForPlacement(final BlockPlaceContext context) {
         return !this.defaultBlockState().canSurvive(context.getLevel(), context.getClickedPos())
             ? Blocks.DIRT.defaultBlockState()
-            : super.getStateForPlacement(context);
+            : canvas$getModifiedPlacement(super.getStateForPlacement(context), context.getLevel()); // Canvas - farmland always moist config
     }
+    // Canvas start - farmland always moist config
+
+    private BlockState canvas$getModifiedPlacement(final BlockState state, final Level level) {
+        if (level.canvasConfig().farming.farmlandAlwaysMoist) {
+            return state.setValue(MOISTURE, MAX_MOISTURE);
+        }
+        return state;
+    }
+    // Canvas end - farmland always moist config
 
     @Override
     protected boolean useShapeForLightOcclusion(final BlockState state) {
@@ -96,7 +104,7 @@ public class FarmlandBlock extends Block {
         int moisture = state.getValue(MOISTURE);
         if (moisture > 0 && level.paperConfig().tickRates.wetFarmland != 1 && (level.paperConfig().tickRates.wetFarmland < 1 || (level.getRedstoneGameTime() + pos.hashCode()) % level.paperConfig().tickRates.wetFarmland != 0)) { return; } // Paper - Configurable random tick rates for blocks // Folia - region threading
         if (moisture == 0 && level.paperConfig().tickRates.dryFarmland != 1 && (level.paperConfig().tickRates.dryFarmland < 1 || (level.getRedstoneGameTime() + pos.hashCode()) % level.paperConfig().tickRates.dryFarmland != 0)) { return; } // Paper - Configurable random tick rates for blocks // Folia - region threading
-        if (!isNearWater(level, pos) && !level.isRainingAt(pos.above())) {
+        if (!level.canvasConfig().farming.farmlandAlwaysMoist && !isNearWater(level, pos) && !level.isRainingAt(pos.above())) { // Canvas - farmland always moist config - if farmland can be moist then we allow moisture changes, otherwise we assume near water
             if (moisture > 0) {
                 org.bukkit.craftbukkit.event.CraftEventFactory.handleMoistureChangeEvent(level, pos, state.setValue(MOISTURE, moisture - 1), Block.UPDATE_CLIENTS); // CraftBukkit
             } else if (!shouldMaintainFarmland(level, pos)) {
@@ -111,6 +119,7 @@ public class FarmlandBlock extends Block {
     public void fallOn(final Level level, final BlockState state, final BlockPos pos, final Entity entity, final double fallDistance) {
         super.fallOn(level, state, pos, entity, fallDistance); // CraftBukkit - moved here as game rules / events shouldn't affect fall damage.
         if (level instanceof ServerLevel serverLevel
+            && !level.canvasConfig().farming.disableFarmlandTrampling // Canvas - disable farmland trampling config
             && level.getRandom().nextFloat() < fallDistance - 0.5
             && entity instanceof LivingEntity
             && (entity instanceof Player || serverLevel.getGameRules().get(GameRules.MOB_GRIEFING))
