From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Fri, 29 Nov 2024 17:01:57 -0800
Subject: [PATCH] Chunk Access Optimizations


diff --git a/src/main/java/net/minecraft/world/level/LevelReader.java b/src/main/java/net/minecraft/world/level/LevelReader.java
index ade435de0af4ee3566fa4a490df53cddd2f6531c..0ccb6a64b39e8262d0705d91a10e183448ab21a2 100644
--- a/src/main/java/net/minecraft/world/level/LevelReader.java
+++ b/src/main/java/net/minecraft/world/level/LevelReader.java
@@ -1,5 +1,6 @@
 package net.minecraft.world.level;
 
+import java.util.concurrent.CompletableFuture;
 import java.util.stream.Stream;
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
@@ -10,6 +11,9 @@ import net.minecraft.core.Registry;
 import net.minecraft.core.RegistryAccess;
 import net.minecraft.core.SectionPos;
 import net.minecraft.resources.ResourceKey;
+import net.minecraft.server.level.ChunkHolder;
+import net.minecraft.server.level.ChunkResult;
+import net.minecraft.server.level.ServerLevel;
 import net.minecraft.tags.FluidTags;
 import net.minecraft.util.Mth;
 import net.minecraft.world.flag.FeatureFlagSet;
@@ -17,6 +21,7 @@ import net.minecraft.world.level.biome.Biome;
 import net.minecraft.world.level.biome.BiomeManager;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.chunk.ChunkAccess;
+import net.minecraft.world.level.chunk.LevelChunk;
 import net.minecraft.world.level.chunk.status.ChunkStatus;
 import net.minecraft.world.level.dimension.DimensionType;
 import net.minecraft.world.level.levelgen.Heightmap;
@@ -70,9 +75,26 @@ public interface LevelReader extends ca.spottedleaf.moonrise.patches.chunk_syste
 
     @Override
     default Holder<Biome> getNoiseBiome(int biomeX, int biomeY, int biomeZ) {
-        ChunkAccess chunkAccess = this.getChunk(QuartPos.toSection(biomeX), QuartPos.toSection(biomeZ), ChunkStatus.BIOMES, false);
+        ChunkAccess chunkAccess = this.canvas$fasterChunkAccess(this, QuartPos.toSection(biomeX), QuartPos.toSection(biomeZ), ChunkStatus.BIOMES, false); // Canvas
         return chunkAccess != null ? chunkAccess.getNoiseBiome(biomeX, biomeY, biomeZ) : this.getUncachedNoiseBiome(biomeX, biomeY, biomeZ);
     }
+    // Canvas start
+
+    private @Nullable ChunkAccess canvas$fasterChunkAccess(LevelReader instance, int x, int z, ChunkStatus chunkStatus, boolean create) {
+        if (!create && instance instanceof ServerLevel world) {
+            final ChunkHolder holder = (world.getChunkSource().chunkMap).getVisibleChunkIfPresent(ChunkPos.asLong(x, z));
+            if (holder != null) {
+                final CompletableFuture<ChunkResult<LevelChunk>> future = holder.getFullChunkFuture();
+                final ChunkResult<LevelChunk> either = future.getNow(null);
+                if (either != null) {
+                    final LevelChunk chunk = either.orElse(null);
+                    if (chunk != null) return chunk;
+                }
+            }
+        }
+        return instance.getChunk(x, z, chunkStatus, create);
+    }
+    // Canvas end
 
     Holder<Biome> getUncachedNoiseBiome(int biomeX, int biomeY, int biomeZ);
 
diff --git a/src/main/java/net/minecraft/world/level/biome/BiomeManager.java b/src/main/java/net/minecraft/world/level/biome/BiomeManager.java
index 90f8360f547ce709fd13ee34f8e67d8bfa94b498..c37b73290f52239abec7a74c332c7d77f4e8e095 100644
--- a/src/main/java/net/minecraft/world/level/biome/BiomeManager.java
+++ b/src/main/java/net/minecraft/world/level/biome/BiomeManager.java
@@ -29,39 +29,56 @@ public class BiomeManager {
     }
 
     public Holder<Biome> getBiome(BlockPos pos) {
-        int i = pos.getX() - 2;
-        int j = pos.getY() - 2;
-        int k = pos.getZ() - 2;
-        int l = i >> 2;
-        int m = j >> 2;
-        int n = k >> 2;
-        double d = (double)(i & 3) / 4.0;
-        double e = (double)(j & 3) / 4.0;
-        double f = (double)(k & 3) / 4.0;
-        int o = 0;
-        double g = Double.POSITIVE_INFINITY;
+        // Canvas start - optimize
+        final int var0 = pos.getX() - 2;
+        final int var1 = pos.getY() - 2;
+        final int var2 = pos.getZ() - 2;
+        final int var3 = var0 >> 2;
+        final int var4 = var1 >> 2;
+        final int var5 = var2 >> 2;
+        final double var6 = (double) (var0 & 3) / 4.0;
+        final double var7 = (double) (var1 & 3) / 4.0;
+        final double var8 = (double) (var2 & 3) / 4.0;
+        int var9 = 0;
+        double var10 = Double.POSITIVE_INFINITY;
 
-        for (int p = 0; p < 8; p++) {
-            boolean bl = (p & 4) == 0;
-            boolean bl2 = (p & 2) == 0;
-            boolean bl3 = (p & 1) == 0;
-            int q = bl ? l : l + 1;
-            int r = bl2 ? m : m + 1;
-            int s = bl3 ? n : n + 1;
-            double h = bl ? d : d - 1.0;
-            double t = bl2 ? e : e - 1.0;
-            double u = bl3 ? f : f - 1.0;
-            double v = getFiddledDistance(this.biomeZoomSeed, q, r, s, h, t, u);
-            if (g > v) {
-                o = p;
-                g = v;
+        for (int var11 = 0; var11 < 8; ++var11) {
+            boolean var12 = (var11 & 4) == 0;
+            boolean var13 = (var11 & 2) == 0;
+            boolean var14 = (var11 & 1) == 0;
+            long var15 = var12 ? var3 : var3 + 1;
+            long var16 = var13 ? var4 : var4 + 1;
+            long var17 = var14 ? var5 : var5 + 1;
+            double var18 = var12 ? var6 : var6 - 1.0;
+            double var19 = var13 ? var7 : var7 - 1.0;
+            double var20 = var14 ? var8 : var8 - 1.0;
+            long var21 = this.biomeZoomSeed * (this.biomeZoomSeed * 6364136223846793005L + 1442695040888963407L) + var15;
+            var21 = var21 * (var21 * 6364136223846793005L + 1442695040888963407L) + var16;
+            var21 = var21 * (var21 * 6364136223846793005L + 1442695040888963407L) + var17;
+            var21 = var21 * (var21 * 6364136223846793005L + 1442695040888963407L) + var15;
+            var21 = var21 * (var21 * 6364136223846793005L + 1442695040888963407L) + var16;
+            var21 = var21 * (var21 * 6364136223846793005L + 1442695040888963407L) + var17;
+            double var22 = (double)((var21 >> 24) & 1023) / 1024.0;
+            double var23 = (var22 - 0.5) * 0.9;
+            var21 = var21 * (var21 * 6364136223846793005L + 1442695040888963407L) + this.biomeZoomSeed;
+            double var24 = (double)((var21 >> 24) & 1023) / 1024.0;
+            double var25 = (var24 - 0.5) * 0.9;
+            var21 = var21 * (var21 * 6364136223846793005L + 1442695040888963407L) + this.biomeZoomSeed;
+            double var26 = (double)((var21 >> 24) & 1023) / 1024.0;
+            double var27 = (var26 - 0.5) * 0.9;
+            double var28 = Mth.square(var20 + var27) + Mth.square(var19 + var25) + Mth.square(var18 + var23);
+            if (var10 > var28) {
+                var9 = var11;
+                var10 = var28;
             }
         }
 
-        int w = (o & 4) == 0 ? l : l + 1;
-        int x = (o & 2) == 0 ? m : m + 1;
-        int y = (o & 1) == 0 ? n : n + 1;
-        return this.noiseBiomeSource.getNoiseBiome(w, x, y);
+        int resX = (var9 & 4) == 0 ? var3 : var3 + 1;
+        int resY = (var9 & 2) == 0 ? var4 : var4 + 1;
+        int resZ = (var9 & 1) == 0 ? var5 : var5 + 1;
+
+        return this.noiseBiomeSource.getNoiseBiome(resX, resY, resZ);
+        // Canvas end
     }
 
     public Holder<Biome> getNoiseBiomeAtPosition(double x, double y, double z) {
