From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Thu, 9 Jan 2025 18:52:25 -0800
Subject: [PATCH] Optimize InventoryChangeTrigger


diff --git a/src/main/java/io/canvasmc/canvas/Config.java b/src/main/java/io/canvasmc/canvas/Config.java
index 5a4f20b7dfaf65854e8eb814a08ad4e7645fbb00..f1712eb4ab626c869df37945c665082c1c99b4df 100644
--- a/src/main/java/io/canvasmc/canvas/Config.java
+++ b/src/main/java/io/canvasmc/canvas/Config.java
@@ -66,6 +66,8 @@ public class Config implements ConfigData {
 
 	@Comment("Enables a modified version of Pufferfish's async mob spawning patch")
 	public boolean enableAsyncSpawning = true;
+	@Comment("Delays the inventory change trigger to tick at an interval to avoid excessive usage of advancement updates and recipe updates")
+	public int skipTicksAdvancements = 3;
 
 	public static Config init() {
 		AutoConfig.register(Config.class, JanksonConfigSerializer::new);
diff --git a/src/main/java/net/minecraft/advancements/critereon/InventoryChangeTrigger.java b/src/main/java/net/minecraft/advancements/critereon/InventoryChangeTrigger.java
index ebbad97920df3d1645637e646a98e16cb04d361e..24a407a1175c756f964019143fcd4bc9e758c6e5 100644
--- a/src/main/java/net/minecraft/advancements/critereon/InventoryChangeTrigger.java
+++ b/src/main/java/net/minecraft/advancements/critereon/InventoryChangeTrigger.java
@@ -2,6 +2,7 @@ package net.minecraft.advancements.critereon;
 
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
+import io.canvasmc.canvas.Config;
 import it.unimi.dsi.fastutil.objects.ObjectArrayList;
 import java.util.List;
 import java.util.Map;
@@ -22,24 +23,45 @@ public class InventoryChangeTrigger extends SimpleCriterionTrigger<InventoryChan
         return InventoryChangeTrigger.TriggerInstance.CODEC;
     }
 
+    // Canvas start - optimize inventory change triggers via skip-ticks
+    private int ticksSkipped;
+
+    private boolean tryTick() {
+        int skipTicksAmount = Config.INSTANCE.skipTicksAdvancements;
+        if (skipTicksAmount <= 0)
+            return true;
+
+        this.ticksSkipped++;
+        if (this.ticksSkipped > skipTicksAmount) {
+            this.ticksSkipped = 0;
+            return true;
+        }
+
+        return false;
+    }
+
     public void trigger(ServerPlayer player, Inventory inventory, ItemStack stack) {
+        if (!this.tryTick())
+            return;
+
         int i = 0;
         int j = 0;
         int k = 0;
 
-        for (int l = 0; l < inventory.getContainerSize(); l++) {
-            ItemStack itemStack = inventory.getItem(l);
-            if (itemStack.isEmpty()) {
-                j++;
+        for (int l = 0; l < inventory.getContainerSize(); ++l) {
+            ItemStack itemstack = inventory.getItem(l);
+            if (itemstack.isEmpty()) {
+                ++j;
             } else {
-                k++;
-                if (itemStack.getCount() >= itemStack.getMaxStackSize()) {
-                    i++;
+                ++k;
+                if (itemstack.getCount() >= itemstack.getMaxStackSize()) {
+                    ++i;
                 }
             }
         }
 
         this.trigger(player, inventory, stack, i, j, k);
+    // Canvas end
     }
 
     private void trigger(ServerPlayer player, Inventory inventory, ItemStack stack, int full, int empty, int occupied) {
