From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Fri, 27 Dec 2024 00:12:01 -0800
Subject: [PATCH] Ensure Java 22 or Higher


diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
index b2b6b3edfecde01f4907d2ce28466bd53c9566dc..6259ed6815cccb07edd1131af59afeab3f33e3c1 100644
--- a/src/main/java/net/minecraft/server/Main.java
+++ b/src/main/java/net/minecraft/server/Main.java
@@ -92,6 +92,12 @@ public class Main {
     @SuppressForbidden(reason = "System.out needed before bootstrap") // CraftBukkit - decompile error
     @DontObfuscate
     public static void main(final OptionSet optionset) { // CraftBukkit - replaces main(String[] astring)
+        // Canvas start - ensure java 22 or higher
+        final int javaSpecVersion = Integer.parseInt(System.getProperty("java.specification.version"));
+        if (javaSpecVersion < 22) {
+            throw new RuntimeException("Canvas requires Java 22 or higher to run, but Java version " + javaSpecVersion + " was detected! Please update.");
+        }
+        // Canvas end
         io.papermc.paper.util.LogManagerShutdownThread.hook(); // Paper
         SharedConstants.tryDetectVersion();
         /* CraftBukkit start - Replace everything
