--- a/net/minecraft/server/network/PlayerChunkSender.java
+++ b/net/minecraft/server/network/PlayerChunkSender.java
@@ -2,6 +_,7 @@
 
 import com.google.common.collect.Comparators;
 import com.mojang.logging.LogUtils;
+import io.canvasmc.canvas.server.level.chunk.ChunkSendingExecutor;
 import it.unimi.dsi.fastutil.longs.LongOpenHashSet;
 import it.unimi.dsi.fastutil.longs.LongSet;
 import java.util.Comparator;
@@ -80,6 +_,7 @@
 
     // Paper start - Anti-Xray
     public static void sendChunk(ServerGamePacketListenerImpl packetListener, ServerLevel level, LevelChunk chunk) {
+        ChunkSendingExecutor.execute(() -> { // Canvas - async chunk sending
         final boolean shouldModify = level.chunkPacketBlockController.shouldModify(packetListener.player, chunk);
         packetListener.send(new ClientboundLevelChunkWithLightPacket(chunk, level.getLightEngine(), null, null, shouldModify));
         // Paper end - Anti-Xray
@@ -90,6 +_,7 @@
         // Paper end - PlayerChunkLoadEvent
         ChunkPos pos = chunk.getPos();
         DebugPackets.sendPoiPacketsForChunk(level, pos);
+        }); // Canvas - async chunk sending
     }
 
     private List<LevelChunk> collectChunksToSend(ChunkMap chunkMap, ChunkPos chunkPos) {
