From 30322d32a063dc3b6c41ba7f9d75c2e508ff7cf1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
index b83e3660ac6a91d1f8026bc8ac9b74ca40bea52a..217f4e4a095af1e1cf6d6b2f7817ea4194f520f5 100644
--- a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -322,10 +322,47 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
     // Canvas end - optimize flush calls with region threading
 
     public void send(final Packet<?> packet) {
+        // Canvas start - no chat reports
+        if (io.canvasmc.canvas.Config.INSTANCE.enableNoChatReports) {
+            Object self = this;
+            boolean cancel = false;
+
+            if (self instanceof ServerGamePacketListenerImpl listener
+                && packet instanceof net.minecraft.network.protocol.game.ClientboundPlayerChatPacket chat) {
+                packet = new net.minecraft.network.protocol.game.ClientboundSystemChatPacket(chat.chatType().decorate(
+                    chat.unsignedContent() != null ? chat.unsignedContent()
+                        : Component.literal(chat.body().content())
+                ), false);
+
+                cancel = true;
+                listener.send(packet);
+            }
+
+            if (cancel) {
+                return;
+            }
+        }
+        // Canvas end - no chat reports
         this.send(packet, null);
     }
 
     public void send(final Packet<?> packet, final @Nullable ChannelFutureListener listener) {
+        // Canvas start - no chat reports
+        if (io.canvasmc.canvas.Config.INSTANCE.enableNoChatReports) {
+            Object self = this;
+            boolean cancel = false;
+
+            if (self instanceof ServerGamePacketListenerImpl listenerImpl
+                && packet instanceof net.minecraft.network.protocol.game.ClientboundPlayerChatPacket chat && sendListener != null) {
+                cancel = true;
+                listenerImpl.send(chat);
+            }
+
+            if (cancel) {
+                return;
+            }
+        }
+        // Canvas end - no chat reports
         // CraftBukkit start
         if (packet == null || this.processedDisconnect) { // Spigot
             return;
