From 9a2c48187d2c733eeb18126deb4d18e2547c666f Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/server/network/ServerLoginPacketListenerImpl.java b/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
index ed905df09f24363044509e1e3d838fb25a69a002..4d70d6ddd504ddc93f086ff8a89582b48d1167a2 100644
--- a/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerLoginPacketListenerImpl.java
@@ -72,6 +72,7 @@ public class ServerLoginPacketListenerImpl implements ServerLoginPacketListener,
     public java.util.@Nullable UUID requestedUuid; // Paper
     private final io.papermc.paper.connection.PaperPlayerLoginConnection paperLoginConnection; // Paper - Config API
     private volatile boolean disconnecting = false; // Paper - Fix disconnect still ticking login
+    private volatile boolean canvas$switchedAlready = false; // Canvas - async protocol switch
 
     public ServerLoginPacketListenerImpl(final MinecraftServer minecraftserver, final Connection connection, final boolean transferred) {
         this.server = minecraftserver;
@@ -429,9 +430,29 @@ public class ServerLoginPacketListenerImpl implements ServerLoginPacketListener,
 
     @Override
     public void handleLoginAcknowledgement(final ServerboundLoginAcknowledgedPacket packet) {
+        // Canvas start - async protocol switch
+        if (io.canvasmc.canvas.Config.INSTANCE.networking.asyncProtocolSwitch && !canvas$switchedAlready) {
+            canvas$switchedAlready = true;
+            this.connection.canvas$setupOutboundProtocolAsync(ConfigurationProtocols.CLIENTBOUND)
+                .addListener(unused -> {
+                    try {
+                        // just run PacketUtils, this will schedule this to the global tick for us
+                        net.minecraft.network.protocol.PacketUtils.ensureRunningOnSameThread(packet, this, this.server.packetProcessor());
+                    } catch (net.minecraft.server.RunningOnDifferentThreadException | io.papermc.paper.util.ServerStopRejectedExecutionException ignored) {
+                    } catch (java.util.concurrent.RejectedExecutionException rejExc) {
+                        this.connection.disconnect(Component.translatable("multiplayer.disconnect.server_shutdown"));
+                    } catch (ClassCastException castClass) {
+                        LOGGER.error("Received {} that couldn't be processed", packet.getClass(), castClass);
+                        this.connection.disconnect(Component.translatable("multiplayer.disconnect.invalid_packet"));
+                    }
+                }
+            );
+            return;
+        }
         net.minecraft.network.protocol.PacketUtils.ensureRunningOnSameThread(packet, this, this.server.packetProcessor()); // CraftBukkit
         Validate.validState(this.state == ServerLoginPacketListenerImpl.State.PROTOCOL_SWITCHING, "Unexpected login acknowledgement packet");
-        this.connection.setupOutboundProtocol(ConfigurationProtocols.CLIENTBOUND);
+        if (!io.canvasmc.canvas.Config.INSTANCE.networking.asyncProtocolSwitch) this.connection.setupOutboundProtocol(ConfigurationProtocols.CLIENTBOUND);
+        // Canvas end - async protocol switch
         CommonListenerCookie cookie = CommonListenerCookie.createInitial(Objects.requireNonNull(this.authenticatedProfile), this.transferred);
         ServerConfigurationPacketListenerImpl configPacketListener = new ServerConfigurationPacketListenerImpl(this.server, this.connection, cookie);
         this.connection.setupInboundProtocol(ConfigurationProtocols.SERVERBOUND, configPacketListener);
