From 60fbf10de1ffff26663053a967c13ae1e63a5a7a Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java b/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
index 730274b901fc858461e98123937d2f3b85d33d23..3b52bdaf1c3cfcdec035fe76d64210b52350b183 100644
--- a/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
@@ -54,6 +54,7 @@ public class ServerConfigurationPacketListenerImpl extends ServerCommonPacketLis
     private @Nullable PrepareSpawnTask prepareSpawnTask;
     public io.papermc.paper.connection.PaperPlayerConfigurationConnection paperConnection; // Paper
     public net.minecraft.server.level.ServerPlayer switchToMain; // Folia - region threading - rewrite login process
+    private volatile boolean canvas$switchedAlready = false; // Canvas - async protocol switch
 
     public ServerConfigurationPacketListenerImpl(final MinecraftServer server, final Connection connection, final CommonListenerCookie cookie) {
         super(server, connection, cookie);
@@ -193,9 +194,32 @@ public class ServerConfigurationPacketListenerImpl extends ServerCommonPacketLis
 
     @Override
     public void handleConfigurationFinished(final ServerboundFinishConfigurationPacket packet) {
+        // Canvas start - async protocol switch
+        if (io.canvasmc.canvas.Config.INSTANCE.networking.asyncProtocolSwitch && !canvas$switchedAlready) {
+            canvas$switchedAlready = true;
+            this.finishCurrentTask(JoinWorldTask.TYPE);
+            this.connection.canvas$setupOutboundProtocolAsync(
+                GameProtocols.CLIENTBOUND_TEMPLATE.bind(
+                    RegistryFriendlyByteBuf.decorator(this.server.registryAccess())
+                )
+            ).addListener(unused -> {
+                try {
+                    // just run PacketUtils, this will schedule this to the global tick for us
+                    PacketUtils.ensureRunningOnSameThread(packet, this, this.server.packetProcessor());
+                } catch (net.minecraft.server.RunningOnDifferentThreadException | io.papermc.paper.util.ServerStopRejectedExecutionException ignored) {
+                } catch (java.util.concurrent.RejectedExecutionException rejExc) {
+                    this.connection.disconnect(Component.translatable("multiplayer.disconnect.server_shutdown"));
+                } catch (ClassCastException castClass) {
+                    LOGGER.error("Received {} that couldn't be processed", packet.getClass(), castClass);
+                    this.connection.disconnect(Component.translatable("multiplayer.disconnect.invalid_packet"));
+                }
+            });
+            return;
+        }
         PacketUtils.ensureRunningOnSameThread(packet, this, this.server.packetProcessor());
-        this.finishCurrentTask(JoinWorldTask.TYPE);
-        this.connection.setupOutboundProtocol(GameProtocols.CLIENTBOUND_TEMPLATE.bind(RegistryFriendlyByteBuf.decorator(this.server.registryAccess())));
+        if (!io.canvasmc.canvas.Config.INSTANCE.networking.asyncProtocolSwitch) this.finishCurrentTask(JoinWorldTask.TYPE);
+        if (!io.canvasmc.canvas.Config.INSTANCE.networking.asyncProtocolSwitch) this.connection.setupOutboundProtocol(GameProtocols.CLIENTBOUND_TEMPLATE.bind(RegistryFriendlyByteBuf.decorator(this.server.registryAccess())));
+        // Canvas end - async protocol switch
 
         try {
             PlayerList playerList = this.server.getPlayerList();
