--- a/net/minecraft/server/network/ServerConnectionListener.java
+++ b/net/minecraft/server/network/ServerConnectionListener.java
@@ -250,6 +_,13 @@
 
             while (iterator.hasNext()) {
                 Connection connection = iterator.next();
+                // Canvas start
+                if (connection.getPlayer() != null) {
+                    connection.getPlayer().serverLevel().notifyNewConnection(connection);
+                    iterator.remove();
+                    continue;
+                }
+                // Canvas end
                 if (!connection.isConnecting()) {
                     if (connection.isConnected()) {
                         try {
