--- a/net/minecraft/server/network/ServerTextFilter.java
+++ b/net/minecraft/server/network/ServerTextFilter.java
@@ -25,6 +_,7 @@
 import java.util.concurrent.ThreadFactory;
 import java.util.concurrent.atomic.AtomicInteger;
 import javax.annotation.Nullable;
+import io.canvasmc.canvas.Config;
 import net.minecraft.SharedConstants;
 import net.minecraft.Util;
 import net.minecraft.network.chat.FilterMask;
@@ -48,7 +_,7 @@
     final ExecutorService workerPool;
 
     protected static ExecutorService createWorkerPool(int size) {
-        return Executors.newFixedThreadPool(size, THREAD_FACTORY);
+        return Config.INSTANCE.virtualThreads.shouldReplaceTextFilterPool() ? Executors.newVirtualThreadPerTaskExecutor() : Executors.newFixedThreadPool(size, THREAD_FACTORY); // Canvas - virtual threads
     }
 
     protected ServerTextFilter(
