From a3c96b695f9e13d71875813ec2583d31d8569479 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 1acf16cb735a6bb54ed023206204214ec05ea048..1fe415fa19a291a1b61cab994f66707610a35edb 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -343,7 +343,7 @@ public class ServerGamePacketListenerImpl
         this.player = player;
         player.connection = this;
         player.getTextFilter().join();
-        this.signedMessageDecoder = SignedMessageChain.Decoder.unsigned(player.getUUID(), server::enforceSecureProfile);
+        this.signedMessageDecoder = SignedMessageChain.Decoder.unsigned(player.getUUID(), server::canvas$ncrEnforceSecureProfile); // Canvas - no chat reports
         this.chatMessageChain = new FutureChain(server.chatExecutor); // CraftBukkit - async chat
         this.tickEndEvent = new io.papermc.paper.event.packet.ClientTickEndEvent(player.getBukkitEntity()); // Paper - add client tick end event
         this.playerGameConnection =  new io.papermc.paper.connection.PaperPlayerGameConnection(this); // Paper
@@ -2469,7 +2469,7 @@ public class ServerGamePacketListenerImpl
         command = event.getMessage().substring(1);
         // CraftBukkit end
         ParseResults<CommandSourceStack> parsed = this.parseCommand(command);
-        if (this.server.enforceSecureProfile() && SignableCommand.hasSignableArguments(parsed)) {
+        if (this.server.canvas$ncrEnforceSecureProfile() && SignableCommand.hasSignableArguments(parsed)) { // Canvas - no chat reports
             LOGGER.error(
                 "Received unsigned command packet from {}, but the command requires signable arguments: {}", this.player.getGameProfile().name(), command
             );
@@ -3811,7 +3811,7 @@ public class ServerGamePacketListenerImpl
     private void resetPlayerChatState(final RemoteChatSession chatSession) {
         this.chatSession = chatSession;
         this.hasLoggedExpiry = false; // Paper - Prevent causing expired keys from impacting new joins
-        this.signedMessageDecoder = chatSession.createMessageDecoder(this.player.getUUID());
+        if (!io.canvasmc.canvas.Config.INSTANCE.enableNoChatReports) this.signedMessageDecoder = chatSession.createMessageDecoder(this.player.getUUID()); // Canvas - no chat reports
         this.chatMessageChain
             .append(
                 () -> {
