--- a/net/minecraft/server/network/ServerTextFilter.java
+++ b/net/minecraft/server/network/ServerTextFilter.java
@@ -48,7 +_,7 @@
     final ExecutorService workerPool;
 
     protected static ExecutorService createWorkerPool(int size) {
-        return Executors.newFixedThreadPool(size, THREAD_FACTORY);
+        return io.canvasmc.canvas.util.virtual.VirtualThreadUtils.createFixedExecutor(size, io.canvasmc.canvas.Config.INSTANCE.virtualThreads.serverTextFilter, THREAD_FACTORY, io.canvasmc.canvas.util.virtual.VirtualThreadUtils.buildFactory((virtual) -> virtual.name("Chat-Filter-Worker-" + WORKER_COUNT.getAndIncrement()))); // Canvas - virtual threads
     }
 
     protected ServerTextFilter(
