--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -8,6 +_,8 @@
 import com.mojang.brigadier.StringReader;
 import com.mojang.brigadier.suggestion.Suggestions;
 import com.mojang.logging.LogUtils;
+import io.canvasmc.canvas.Config;
+import io.canvasmc.canvas.region.ServerRegions;
 import it.unimi.dsi.fastutil.ints.Int2ObjectMaps;
 import it.unimi.dsi.fastutil.ints.Int2ObjectMap.Entry;
 import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
@@ -22,6 +_,7 @@
 import java.util.Set;
 import java.util.concurrent.CancellationException;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executors;
 import java.util.function.BiFunction;
 import java.util.function.Consumer;
 import java.util.stream.Collectors;
@@ -130,6 +_,7 @@
 import net.minecraft.resources.ResourceKey;
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.RunningOnDifferentThreadException;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.server.level.ServerPlayer;
 import net.minecraft.util.FutureChain;
@@ -185,6 +_,8 @@
 import net.minecraft.world.level.block.entity.StructureBlockEntity;
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.chunk.ChunkAccess;
+import net.minecraft.world.level.chunk.status.ChunkStatus;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.BlockHitResult;
 import net.minecraft.world.phys.Vec3;
@@ -294,8 +_,8 @@
     // CraftBukkit start - add fields
     private int lastTick = MinecraftServer.currentTick;
     private int allowedPlayerTicks = 1;
-    private int lastDropTick = MinecraftServer.currentTick;
-    private int lastBookTick  = MinecraftServer.currentTick;
+    private long lastDropTick = Util.getMillis() / 50L; // Canvas - Threaded Regions
+    private long lastBookTick  = Util.getMillis() / 50L; // Canvas - Threaded Regions
     private int dropCount = 0;
 
     private boolean hasMoved = false;
@@ -341,9 +_,30 @@
             }
         );
     // Purpur end - AFK API
+    long lastTickTime = Util.getNanos(); // Canvas - Threaded Regions
 
     @Override
     public void tick() {
+        // Canvas start - Threaded Regions
+        if (lastTickTime + (1_000_000 * 40) > Util.getNanos()) {
+            // cleanup, cannot double tick this.
+            // this is caused by a race condition where
+            // we accidentally create multiple references
+            // to this connection, making it tick multiple
+            // times, which we cannot have happening.
+            if (this.player.serverLevel().levelTickData.activeConnections.remove(this.connection)) {
+                // LOGGER.warn("Located duplicate connection reference for player '{}' in world", this.player);
+            }
+            this.player.serverLevel().regioniser.computeForAllRegionsUnsynchronised((region) -> {
+                if (region.getData().tickData.activeConnections.remove(this.connection)) {
+                    // LOGGER.warn("Located duplicate connection reference for player '{}' in region '{}'", this.player, region.getData());
+                }
+            });
+            ServerRegions.WorldTickData data = ServerRegions.getTickData(this.player.serverLevel());
+            data.activeConnections.add(this.connection);
+        }
+        else lastTickTime = Util.getNanos();
+        // Canvas end
         if (this.ackBlockChangesUpTo > -1) {
             this.send(new ClientboundBlockChangedAckPacket(this.ackBlockChangesUpTo));
             this.ackBlockChangesUpTo = -1;
@@ -392,7 +_,7 @@
             this.aboveGroundVehicleTickCount = 0;
         }
 
-        this.keepConnectionAlive();
+        // this.keepConnectionAlive(); // Canvas - handled on main
         this.chatSpamThrottler.tick();
         this.dropSpamThrottler.tick();
         this.tabSpamThrottler.tick(); // Paper - configurable tab spam limits
@@ -570,11 +_,17 @@
                     !serverLevel.areChunksLoadedForMove(this.player.getBoundingBox().expandTowards(new Vec3(toX, toY, toZ).subtract(this.player.position()))) ||
                         !serverLevel.areChunksLoadedForMove(rootVehicle.getBoundingBox().expandTowards(new Vec3(toX, toY, toZ).subtract(rootVehicle.position())))
                 )) {
+                    // Canvas start - schedule for load if we fail
+                    this.player.serverLevel().moonrise$getChunkTaskScheduler().scheduleChunkLoad(
+                        this.player.chunkPosition().x, this.player.chunkPosition().z, ChunkStatus.FULL, true, ca.spottedleaf.concurrentutil.util.Priority.BLOCKING,
+                        null
+                    );
+                    // Canvas end
                     this.connection.send(ClientboundMoveVehiclePacket.fromEntity(rootVehicle));
                     return;
                 }
                 // Paper end - Prevent moving into unloaded chunks
-                if (d7 - d6 > Math.max(100.0, Math.pow((double) (org.spigotmc.SpigotConfig.movedTooQuicklyMultiplier * (float) i * speed), 2)) && !this.isSingleplayerOwner()) {
+                if (!Config.INSTANCE.alwaysAllowWeirdMovement && (d7 - d6 > Math.max(100.0, Math.pow((double) (org.spigotmc.SpigotConfig.movedTooQuicklyMultiplier * (float) i * speed), 2)) && !this.isSingleplayerOwner())) { // Canvas
                     // CraftBukkit end
                     LOGGER.warn(
                         "{} (vehicle of {}) moved too quickly! {},{},{}", rootVehicle.getName().getString(), this.player.getName().getString(), d3, d4, d5
@@ -604,7 +_,7 @@
                 d5 = d2 - rootVehicle.getZ();
                 d7 = d3 * d3 + d4 * d4 + d5 * d5;
                 boolean flag2 = false;
-                if (d7 > org.spigotmc.SpigotConfig.movedWronglyThreshold) { // Spigot
+                if (!Config.INSTANCE.alwaysAllowWeirdMovement && (d7 > org.spigotmc.SpigotConfig.movedWronglyThreshold)) { // Spigot // Canvas
                     flag2 = true; // Paper - diff on change, this should be moved wrongly
                     LOGGER.warn("{} (vehicle of {}) moved wrongly! {}", rootVehicle.getName().getString(), this.player.getName().getString(), Math.sqrt(d7));
                 }
@@ -733,8 +_,14 @@
         PacketUtils.ensureRunningOnSameThread(packet, this, this.player.serverLevel());
         if (packet.getId() == this.awaitingTeleport) {
             if (this.awaitingPositionFromClient == null) {
+                // Canvas start - graceful teleport handling
+                if (Config.INSTANCE.gracefulTeleportHandling) {
+                    LOGGER.info("Was not expecting position data from client at this time, gracefully returning instead of disconnect.");
+                } else {
                 ServerGamePacketListenerImpl.LOGGER.warn("Disconnected on accept teleport packet. Was not expecting position data from client at this time"); // Purpur - Add more logger output for invalid movement kicks
                 this.disconnect(Component.translatable("multiplayer.disconnect.invalid_player_movement"), org.bukkit.event.player.PlayerKickEvent.Cause.INVALID_PLAYER_MOVEMENT); // Paper - kick event cause
+                }
+                // Canvas end
                 return;
             }
 
@@ -802,7 +_,12 @@
     }
 
     // Paper start - AsyncTabCompleteEvent
-    private static final java.util.concurrent.ExecutorService TAB_COMPLETE_EXECUTOR = java.util.concurrent.Executors.newFixedThreadPool(4,
+    // Canvas start
+    private static final java.util.concurrent.ExecutorService TAB_COMPLETE_EXECUTOR =
+        Config.INSTANCE.virtualThreads.shouldReplaceTabCompleteExecutor() ?
+            Executors.newVirtualThreadPerTaskExecutor() :
+            java.util.concurrent.Executors.newFixedThreadPool(4,
+    // Canvas end
         new com.google.common.util.concurrent.ThreadFactoryBuilder().setDaemon(true).setNameFormat("Async Tab Complete Thread - #%d").setUncaughtExceptionHandler(new net.minecraft.DefaultUncaughtExceptionHandlerWithName(MinecraftServer.LOGGER)).build());
     // Paper end - AsyncTabCompleteEvent
 
@@ -810,7 +_,7 @@
     public void handleCustomCommandSuggestions(ServerboundCommandSuggestionPacket packet) {
         // PacketUtils.ensureRunningOnSameThread(packet, this, this.player.serverLevel()); // Paper - AsyncTabCompleteEvent; run this async
         // CraftBukkit start
-        if (!this.tabSpamThrottler.isIncrementAndUnderThreshold() && !this.server.getPlayerList().isOp(this.player.getGameProfile()) && !this.server.isSingleplayerOwner(this.player.getGameProfile())) { // Paper - configurable tab spam limits
+        if (!this.tabSpamThrottler.isIncrementAndUnderThreshold() && !this.server.getPlayerList().isOp(this.player.getGameProfile()) && !this.server.isSingleplayerOwner(this.player.getGameProfile()) && !Config.INSTANCE.networking.disableDisconnectSpam) { // Paper - configurable tab spam limits // Canvas
             this.disconnectAsync(Component.translatable("disconnect.spam"), org.bukkit.event.player.PlayerKickEvent.Cause.SPAM); // Paper - Kick event cause // Paper - add proper async disconnect
             return;
         }
@@ -822,7 +_,7 @@
         // Paper end - Don't suggest if tab-complete is disabled
         // Paper start
         final int index;
-        if (packet.getCommand().length() > 64 && ((index = packet.getCommand().indexOf(' ')) == -1 || index >= 64)) {
+        if (packet.getCommand().length() > 64 && ((index = packet.getCommand().indexOf(' ')) == -1 || index >= 64) && !Config.INSTANCE.networking.disableDisconnectSpam) { // Canvas
             this.disconnectAsync(Component.translatable("disconnect.spam"), org.bukkit.event.player.PlayerKickEvent.Cause.SPAM); // Paper - add proper async disconnect
             return;
         }
@@ -882,6 +_,7 @@
         ParseResults<CommandSourceStack> parseResults = this.server.getCommands().getDispatcher().parse(stringReader, this.player.createCommandSourceStack());
         // Paper start - Handle non-recoverable exceptions
         if (!parseResults.getExceptions().isEmpty()
+            && !Config.INSTANCE.networking.disableDisconnectSpam // Canvas
             && parseResults.getExceptions().values().stream().anyMatch(e -> e instanceof io.papermc.paper.brigadier.TagParseCommandSyntaxException)) {
             this.disconnect(Component.translatable("disconnect.spam"), org.bukkit.event.player.PlayerKickEvent.Cause.SPAM);
             return;
@@ -1436,6 +_,12 @@
                                     // Paper start - Add fail move event
                                     io.papermc.paper.event.player.PlayerFailMoveEvent event = fireFailMove(io.papermc.paper.event.player.PlayerFailMoveEvent.FailReason.MOVED_INTO_UNLOADED_CHUNK,
                                         toX, toY, toZ, toYaw, toPitch, false);
+                                    // Canvas start - schedule for load if we fail
+                                    this.player.serverLevel().moonrise$getChunkTaskScheduler().scheduleChunkLoad(
+                                        this.player.chunkPosition().x, this.player.chunkPosition().z, ChunkStatus.FULL, true, ca.spottedleaf.concurrentutil.util.Priority.BLOCKING,
+                                        null
+                                    );
+                                    // Canvas end
                                     if (!event.isAllowed()) {
                                         this.internalTeleport(PositionMoveRotation.of(this.player), Collections.emptySet());
                                         return;
@@ -1447,6 +_,7 @@
                                 if (this.shouldCheckPlayerMovement(isFallFlying)) {
                                     float f2 = isFallFlying ? 300.0F : 100.0F;
                                     if (d7 - d6 > Math.max(f2, Mth.square(org.spigotmc.SpigotConfig.movedTooQuicklyMultiplier * (float) i * speed))) {
+                                        if (!Config.INSTANCE.alwaysAllowWeirdMovement && !(Config.INSTANCE.ignoreMovedTooQuicklyWhenLagging && player.serverLevel().tps5s.getAverage() < org.purpurmc.purpur.PurpurConfig.laggingThreshold)) { // Canvas
                                         // CraftBukkit end
                                         // Paper start - Add fail move event
                                         io.papermc.paper.event.player.PlayerFailMoveEvent event = fireFailMove(io.papermc.paper.event.player.PlayerFailMoveEvent.FailReason.MOVED_TOO_QUICKLY,
@@ -1458,6 +_,7 @@
                                             this.teleport(this.player.getX(), this.player.getY(), this.player.getZ(), this.player.getYRot(), this.player.getXRot());
                                             return;
                                         }
+                                        } // Canvas
                                         // Paper end - Add fail move event
                                     }
                                 }
@@ -1468,10 +_,11 @@
                             d4 = d1 - this.lastGoodY; // Paper - diff on change, used for checking large move vectors above
                             d5 = d2 - this.lastGoodZ; // Paper - diff on change, used for checking large move vectors above
                             boolean flag = d4 > 0.0;
+                            Location from = new Location(this.getCraftPlayer().getWorld(), this.lastPosX, this.lastPosY, this.lastPosZ, this.lastYaw, this.lastPitch); // Get the Players previous Event location. // Canvas
                             if (this.player.onGround() && !packet.isOnGround() && flag) {
                                 // Paper start - Add PlayerJumpEvent
                                 org.bukkit.entity.Player player = this.getCraftPlayer();
-                                Location from = new Location(player.getWorld(), this.lastPosX, this.lastPosY, this.lastPosZ, this.lastYaw, this.lastPitch); // Get the Players previous Event location.
+                                // Canvas - move up
                                 Location to = player.getLocation().clone(); // Start off the To location as the Players current location.
 
                                 // If the packet contains movement information then we update the To location with the correct XYZ.
@@ -1519,6 +_,7 @@
                             d7 = d3 * d3 + d4 * d4 + d5 * d5;
                             boolean movedWrongly = false; // Paper - Add fail move event; rename
                             if (!this.player.isChangingDimension()
+                                && !Config.INSTANCE.alwaysAllowWeirdMovement // Canvas
                                 && d7 > org.spigotmc.SpigotConfig.movedWronglyThreshold // Spigot
                                 && !this.player.isSleeping()
                                 && !this.player.gameMode.isCreative()
@@ -1569,7 +_,7 @@
                                     this.hasMoved = true;
                                 }
 
-                                Location from = new Location(player.getWorld(), this.lastPosX, this.lastPosY, this.lastPosZ, this.lastYaw, this.lastPitch); // Get the Players previous Event location.
+                                from = new Location(player.getWorld(), this.lastPosX, this.lastPosY, this.lastPosZ, this.lastYaw, this.lastPitch); // Get the Players previous Event location. // Canvas
                                 Location to = player.getLocation().clone(); // Start off the To location as the Players current location.
 
                                 // If the packet contains movement information then we update the To location with the correct XYZ.
@@ -1852,7 +_,16 @@
 
     @Override
     public void handlePlayerAction(ServerboundPlayerActionPacket packet) {
-        PacketUtils.ensureRunningOnSameThread(packet, this, this.player.serverLevel());
+        // Canvas start - rubberbanding fix
+        try {
+            PacketUtils.ensureRunningOnSameThread(packet, this, this.player.serverLevel());
+        } catch (RunningOnDifferentThreadException e) {
+            if (packet.getAction() == ServerboundPlayerActionPacket.Action.STOP_DESTROY_BLOCK) {
+                this.player.connection.send(new ClientboundBlockUpdatePacket(packet.getPos(), Blocks.AIR.defaultBlockState()));
+            }
+            throw e;
+        }
+        // Canvas end
         if (this.player.isImmobile()) return; // CraftBukkit
         if (this.player.hasClientLoaded()) {
             BlockPos pos = packet.getPos();
@@ -1890,9 +_,10 @@
                     if (!this.player.isSpectator()) {
                         // limit how quickly items can be dropped
                         // If the ticks aren't the same then the count starts from 0 and we update the lastDropTick.
-                        if (this.lastDropTick != MinecraftServer.currentTick) {
+                        ServerRegions.WorldTickData data = ServerRegions.getTickData(this.player.serverLevel()); // Canvas - Threaded Regions
+                        if (this.lastDropTick != data.peekTick()) { // Canvas - Threaded Regions
                             this.dropCount = 0;
-                            this.lastDropTick = MinecraftServer.currentTick;
+                            this.lastDropTick = data.peekTick(); // Canvas - Threaded Regions
                         } else {
                             // Else we increment the drop count and check the amount.
                             this.dropCount++;
@@ -2422,7 +_,7 @@
             this.player.resetLastActionTime();
             // CraftBukkit start
             if (sync) {
-                this.server.execute(handler);
+                this.player.getBukkitEntity().taskScheduler.schedule((_) -> handler.run(), null, 0L);
             } else {
                 handler.run();
             }
@@ -2600,6 +_,7 @@
         // this.chatSpamThrottler.increment();
         if (!this.chatSpamThrottler.isIncrementAndUnderThreshold()
             // CraftBukkit end
+            && !Config.INSTANCE.networking.disableDisconnectSpam // Canvas
             && !this.server.getPlayerList().isOp(this.player.getGameProfile())
             && !this.server.isSingleplayerOwner(this.player.getGameProfile())) {
             this.disconnectAsync(Component.translatable("disconnect.spam"), org.bukkit.event.player.PlayerKickEvent.Cause.SPAM); // Paper - kick event cause & add proper async disconnect
@@ -3378,7 +_,7 @@
     public void handlePlaceRecipe(ServerboundPlaceRecipePacket packet) {
         // Paper start - auto recipe limit
         if (!org.bukkit.Bukkit.isPrimaryThread()) {
-            if (!this.recipeSpamPackets.isIncrementAndUnderThreshold()) {
+            if (!this.recipeSpamPackets.isIncrementAndUnderThreshold() && !Config.INSTANCE.networking.disableDisconnectSpam) { // Canvas
                 this.disconnectAsync(net.minecraft.network.chat.Component.translatable("disconnect.spam"), org.bukkit.event.player.PlayerKickEvent.Cause.SPAM); // Paper - kick event cause // Paper - add proper async disconnect
                 return;
             }
