--- a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -314,10 +_,47 @@
     }
 
     public void send(Packet<?> packet) {
+        // Canvas start - no chat reports
+        if (io.canvasmc.canvas.Config.INSTANCE.enableNoChatReports) {
+            Object self = this;
+            boolean cancel = false;
+
+            if (self instanceof ServerGamePacketListenerImpl listener
+                && packet instanceof net.minecraft.network.protocol.game.ClientboundPlayerChatPacket chat) {
+                packet = new net.minecraft.network.protocol.game.ClientboundSystemChatPacket(chat.chatType().decorate(
+                    chat.unsignedContent() != null ? chat.unsignedContent()
+                        : Component.literal(chat.body().content())
+                ), false);
+
+                cancel = true;
+                listener.send(packet);
+            }
+
+            if (cancel) {
+                return;
+            }
+        }
+        // Canvas end - no chat reports
         this.send(packet, null);
     }
 
     public void send(Packet<?> packet, @Nullable ChannelFutureListener channelFutureListener) {
+        // Canvas start - no chat reports
+        if (io.canvasmc.canvas.Config.INSTANCE.enableNoChatReports) {
+            Object self = this;
+            boolean cancel = false;
+
+            if (self instanceof ServerGamePacketListenerImpl listenerImpl
+                && packet instanceof net.minecraft.network.protocol.game.ClientboundPlayerChatPacket chat && channelFutureListener != null) {
+                cancel = true;
+                listenerImpl.send(chat);
+            }
+
+            if (cancel) {
+                return;
+            }
+        }
+        // Canvas end - no chat reports
         // CraftBukkit start
         if (packet == null || this.processedDisconnect) { // Spigot
             return;
