From 56a92b2c479a7161c340134ba931fb3d6a8e6f79 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEgg.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEgg.java
index 21c4deda503679ebf91355f289ddeac5f82a1ee7..ece57cd4c2dafc1784ecd3c0af96b2651b40696c 100644
--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEgg.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEgg.java
@@ -61,6 +61,12 @@ public class ThrownEgg extends ThrowableItemProjectile {
     protected void onHitEntity(final EntityHitResult hitResult) {
         super.onHitEntity(hitResult);
         hitResult.getEntity().hurt(this.damageSources().thrown(this, this.getOwner()), 0.0F);
+        // Canvas start - egg can knockback player config
+        if (level().canvasConfig().combat.eggCanKnockbackPlayers && hitResult.getEntity() instanceof net.minecraft.server.level.ServerPlayer serverPlayer) {
+            serverPlayer.knockback(0.4F, this.getX() - hitResult.getEntity().getX(), this.getZ() - hitResult.getEntity().getZ(), damageSources().generic(), 0.0F);
+            serverPlayer.hurtMarked = true;
+        }
+        // Canvas end - egg can knockback player config
     }
 
     @Override
