From 53e1a57c72953de1bfcc94c3257c724e7fa1f88f Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/decoration/ArmorStand.java b/net/minecraft/world/entity/decoration/ArmorStand.java
index f413b0bc029c521cae05a6436fbc982656d8ee91..68f8b198b60c51a72966509df80f72ba8c34b40a 100644
--- a/net/minecraft/world/entity/decoration/ArmorStand.java
+++ b/net/minecraft/world/entity/decoration/ArmorStand.java
@@ -205,6 +205,7 @@ public class ArmorStand extends LivingEntity {
 
     @Override
     protected void pushEntities() {
+        if (!level().canvasConfig().entities.entityCollisionMode.allEntitiesCanBePushed()) return; // Canvas - only push players config
         if (!this.level().paperConfig().entities.armorStands.doCollisionEntityLookups) return; // Paper - Option to prevent armor stands from doing entity lookups
         for (Entity entity : this.level().getEntitiesOfClass(AbstractMinecart.class, this.getBoundingBox(), RIDABLE_MINECARTS)) { // Paper - optimise collisions
             if (this.distanceToSqr(entity) <= 0.2) {
@@ -335,7 +336,14 @@ public class ArmorStand extends LivingEntity {
             // CraftBukkit end
             // Paper start - avoid duplicate event call
             org.bukkit.event.entity.EntityDeathEvent event = this.brokenByAnything(level, source);
-            if (!event.isCancelled()) this.kill(level, source, false); // CraftBukkit
+            // Canvas start - MC-132878
+            if (!event.isCancelled()) {
+                if (io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaFixes.mc132878) {
+                    this.showBreakingParticles();
+                }
+                this.kill(level, source, false);
+            } // CraftBukkit
+            // Canvas end - MC-132878
             // Paper end
             return false;
         } else if (source.is(DamageTypeTags.IGNITES_ARMOR_STANDS)) {
@@ -439,7 +447,14 @@ public class ArmorStand extends LivingEntity {
         if (health <= 0.5F) {
             // Paper start - avoid duplicate event call
             org.bukkit.event.entity.EntityDeathEvent event = this.brokenByAnything(level, source);
-            if (!event.isCancelled()) this.kill(level, source, false); // CraftBukkit
+            // Canvas start - MC-132878
+            if (!event.isCancelled()) {
+                if (io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaFixes.mc132878) {
+                    this.showBreakingParticles();
+                }
+                this.kill(level, source, false);
+            } // CraftBukkit
+            // Canvas end - MC-132878
             // Paper end
         } else {
             this.setHealth(health);
