--- a/net/minecraft/world/level/levelgen/structure/pools/SinglePoolElement.java
+++ b/net/minecraft/world/level/levelgen/structure/pools/SinglePoolElement.java
@@ -12,6 +_,8 @@
 import java.util.List;
 import java.util.Optional;
 import java.util.function.Function;
+import io.canvasmc.canvas.Config;
+import io.canvasmc.canvas.util.structure.GeneralUtils;
 import net.minecraft.Util;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
@@ -119,8 +_,15 @@
         StructureTemplateManager structureTemplateManager, BlockPos pos, Rotation rotation, RandomSource random
     ) {
         List<StructureTemplate.JigsawBlockInfo> jigsaws = this.getTemplate(structureTemplateManager).getJigsaws(pos, rotation);
+        // Canvas start - StructureLayoutOptimizer
+        if (Config.INSTANCE.enableStructureLayoutOptimizer) {
+            structureLayoutOptimizer$fasterJigsawListShuffling1(jigsaws, random);
+            structureLayoutOptimizer$fasterJigsawListShuffling2(jigsaws);
+        } else {
         Util.shuffle(jigsaws, random);
         sortBySelectionPriority(jigsaws);
+        }
+        // Canvas end
         return jigsaws;
     }
 
@@ -191,4 +_,12 @@
     public String toString() {
         return "Single[" + this.template + "]";
     }
+    // Canvas start - StructureLayoutOptimizer
+
+    private void structureLayoutOptimizer$fasterJigsawListShuffling1(List<StructureTemplate.JigsawBlockInfo> list, RandomSource randomSource) {
+        GeneralUtils.shuffleAndPrioritize(list, randomSource);
+    }
+
+    private void structureLayoutOptimizer$fasterJigsawListShuffling2(List<StructureTemplate.JigsawBlockInfo> structureBlockInfos) {}
+    // Canvas end
 }
