--- a/net/minecraft/world/level/levelgen/structure/pools/SinglePoolElement.java
+++ b/net/minecraft/world/level/levelgen/structure/pools/SinglePoolElement.java
@@ -119,8 +_,15 @@
         StructureTemplateManager structureTemplateManager, BlockPos pos, Rotation rotation, RandomSource random
     ) {
         List<StructureTemplate.JigsawBlockInfo> jigsaws = this.getTemplate(structureTemplateManager).getJigsaws(pos, rotation);
+        // Canvas start - optimize structure gen
+        if (io.canvasmc.canvas.Config.INSTANCE.chunks.structures.optimizeStructureGen) {
+            structureLayoutOptimizer$fasterJigsawListShuffling1(jigsaws, random);
+            structureLayoutOptimizer$fasterJigsawListShuffling2(jigsaws);
+        } else {
         Util.shuffle(jigsaws, random);
         sortBySelectionPriority(jigsaws);
+        }
+        // Canvas end - optimize structure gen
         return jigsaws;
     }
 
@@ -191,6 +_,14 @@
     public String toString() {
         return "Single[" + this.template + "]";
     }
+    // Canvas start - optimize structure gen
+
+    private void structureLayoutOptimizer$fasterJigsawListShuffling1(List<StructureTemplate.JigsawBlockInfo> list, RandomSource randomSource) {
+        io.canvasmc.canvas.util.structure.GeneralUtils.shuffleAndPrioritize(list, randomSource);
+    }
+
+    private void structureLayoutOptimizer$fasterJigsawListShuffling2(List<StructureTemplate.JigsawBlockInfo> structureBlockInfos) {}
+    // Canvas end - optimize structure gen
 
     @VisibleForTesting
     public ResourceLocation getTemplateLocation() {
