--- a/net/minecraft/world/level/levelgen/structure/pools/StructureTemplatePool.java
+++ b/net/minecraft/world/level/levelgen/structure/pools/StructureTemplatePool.java
@@ -39,7 +_,7 @@
     public static final Codec<Holder<StructureTemplatePool>> CODEC = Util.make(
         RegistryFileCodec.create(Registries.TEMPLATE_POOL, DIRECT_CODEC), CODEC_REFERENCE::setValue
     );
-    private final List<Pair<StructurePoolElement, Integer>> rawTemplates;
+    public final List<Pair<StructurePoolElement, Integer>> rawTemplates; // Canvas - private -> public
     private final ObjectArrayList<StructurePoolElement> templates;
     private final Holder<StructureTemplatePool> fallback;
     private int maxSize = Integer.MIN_VALUE;
