--- a/net/minecraft/world/level/levelgen/structure/pools/StructurePoolElement.java
+++ b/net/minecraft/world/level/levelgen/structure/pools/StructurePoolElement.java
@@ -8,6 +_,7 @@
 import java.util.function.Function;
 import java.util.stream.Collectors;
 import javax.annotation.Nullable;
+import io.canvasmc.canvas.util.SynchronizedCodec;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
 import net.minecraft.core.Vec3i;
@@ -27,9 +_,9 @@
 import net.minecraft.world.level.levelgen.structure.templatesystem.StructureTemplateManager;
 
 public abstract class StructurePoolElement {
-    public static final Codec<StructurePoolElement> CODEC = BuiltInRegistries.STRUCTURE_POOL_ELEMENT
+    public static final Codec<StructurePoolElement> CODEC = new SynchronizedCodec<>(BuiltInRegistries.STRUCTURE_POOL_ELEMENT // Canvas - use synchronized codec
         .byNameCodec()
-        .dispatch("element_type", StructurePoolElement::getType, StructurePoolElementType::codec);
+        .dispatch("element_type", StructurePoolElement::getType, StructurePoolElementType::codec)); // Canvas - use synchronzied codec
     private static final Holder<StructureProcessorList> EMPTY = Holder.direct(new StructureProcessorList(List.of()));
     @Nullable
     private volatile StructureTemplatePool.Projection projection;
