From e64d3ec2ef48556e0a10e7686e1c057ce53f22aa Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/commands/Commands.java b/net/minecraft/commands/Commands.java
index ad764d57dd28f3c97faf9b2edd9bfee526da6a17..a121b8041632791e8ece54c3141f68a24b3744a4 100644
--- a/net/minecraft/commands/Commands.java
+++ b/net/minecraft/commands/Commands.java
@@ -196,7 +196,7 @@ public class Commands {
         AdvancementCommands.register(this.dispatcher);
         AttributeCommand.register(this.dispatcher, context);
         ExecuteCommand.register(this.dispatcher, context);
-        //BossBarCommands.register(this.dispatcher, context); // Folia - region threading - TODO
+        BossBarCommands.register(this.dispatcher, context);
         ClearInventoryCommands.register(this.dispatcher, context);
         //CloneCommands.register(this.dispatcher, context); // Folia - region threading - TODO
         DamageCommand.register(this.dispatcher, context);
@@ -223,7 +223,7 @@ public class Commands {
         KillCommand.register(this.dispatcher);
         ListPlayersCommand.register(this.dispatcher);
         LocateCommand.register(this.dispatcher, context);
-        //LootCommand.register(this.dispatcher, context); // Folia - region threading - TODO later
+        LootCommand.register(this.dispatcher, context);
         MsgCommand.register(this.dispatcher);
         SwingCommand.register(this.dispatcher);
         ParticleCommand.register(this.dispatcher, context);
@@ -245,11 +245,11 @@ public class Commands {
         SetSpawnCommand.register(this.dispatcher);
         SetWorldSpawnCommand.register(this.dispatcher);
         SpectateCommand.register(this.dispatcher);
-        //SpreadPlayersCommand.register(this.dispatcher); // Folia - region threading - TODO later
+        SpreadPlayersCommand.register(this.dispatcher);
         StopSoundCommand.register(this.dispatcher);
         StopwatchCommand.register(this.dispatcher);
         SummonCommand.register(this.dispatcher, context);
-        //TagCommand.register(this.dispatcher); // Folia - region threading - TODO later
+        TagCommand.register(this.dispatcher); // Canvas - TODO - do better
         //TeamCommand.register(this.dispatcher, context); // Folia - region threading - TODO later
         //TeamMsgCommand.register(this.dispatcher); // Folia - region threading - TODO later
         TeleportCommand.register(this.dispatcher);
@@ -262,6 +262,7 @@ public class Commands {
         WaypointCommand.register(this.dispatcher, context);
         WeatherCommand.register(this.dispatcher);
         WorldBorderCommand.register(this.dispatcher);
+        io.canvasmc.canvas.command.RootCommandTree.INSTANCE.build(this.dispatcher); // Canvas - Rebrand - register commands
         if (JvmProfiler.INSTANCE.isAvailable()) {
             JfrCommand.register(this.dispatcher);
         }
@@ -291,7 +292,7 @@ public class Commands {
             PardonCommand.register(this.dispatcher);
             PardonIpCommand.register(this.dispatcher);
             //PerfCommand.register(this.dispatcher); // Folia - region threading - TODO later
-            //SaveAllCommand.register(this.dispatcher); // Folia - region threading - TODO later
+            SaveAllCommand.register(this.dispatcher);
             SaveOffCommand.register(this.dispatcher);
             SaveOnCommand.register(this.dispatcher);
             SetPlayerIdleTimeoutCommand.register(this.dispatcher);
