From 5fef61e193d31b59e870ef514121ea8263f383b9 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas paperServer File Patches


diff --git a/src/main/java/io/papermc/paper/world/migration/WorldFolderMigration.java b/src/main/java/io/papermc/paper/world/migration/WorldFolderMigration.java
index 3e72dd125e5e7d577f0c1187d40b17cc6c2fea5a..c56a3e8e8adbb1bdeccf7dfcb1ad9cfe558dd976 100644
--- a/src/main/java/io/papermc/paper/world/migration/WorldFolderMigration.java
+++ b/src/main/java/io/papermc/paper/world/migration/WorldFolderMigration.java
@@ -54,6 +54,7 @@ public final class WorldFolderMigration {
             case VANILLA_MIGRATION -> VanillaWorldMigration.migrate(context);
             case NO_OP -> {}
         }
+        io.canvasmc.canvas.util.migration.CanvasFileMigration.initMigration(context); // Canvas - 26.1+ file migration
     }
 
     public static void migrateApiWorld(
