--- a/net/minecraft/world/entity/decoration/ArmorStand.java
+++ b/net/minecraft/world/entity/decoration/ArmorStand.java
@@ -205,6 +_,7 @@
 
     @Override
     protected void pushEntities() {
+        if (!io.canvasmc.canvas.Config.INSTANCE.entityCollisionMode.allEntitiesCanBePushed()) return; // Canvas - only push players config
         if (!this.level().paperConfig().entities.armorStands.doCollisionEntityLookups) return; // Paper - Option to prevent armor stands from doing entity lookups
         for (Entity entity : this.level().getEntitiesOfClass(AbstractMinecart.class, this.getBoundingBox(), RIDABLE_MINECARTS)) { // Paper - optimise collisions
             if (this.distanceToSqr(entity) <= 0.2) {
@@ -339,7 +_,14 @@
             // CraftBukkit end
             // Paper start - avoid duplicate event call
             org.bukkit.event.entity.EntityDeathEvent event = this.brokenByAnything(level, damageSource);
-            if (!event.isCancelled()) this.kill(level, damageSource, false); // CraftBukkit
+            // Canvas start - MC-132878
+            if (!event.isCancelled()) {
+                if (io.canvasmc.canvas.Config.INSTANCE.fixes.mc132878) {
+                    this.showBreakingParticles();
+                }
+                this.kill(level, damageSource, false);
+            } // CraftBukkit
+            // Canvas end - MC-132878
             // Paper end
             return false;
         } else if (damageSource.is(DamageTypeTags.IGNITES_ARMOR_STANDS)) {
@@ -446,7 +_,14 @@
         if (health <= 0.5F) {
             // Paper start - avoid duplicate event call
             org.bukkit.event.entity.EntityDeathEvent event = this.brokenByAnything(level, damageSource);
-            if (!event.isCancelled()) this.kill(level, damageSource, false); // CraftBukkit
+            // Canvas start - MC-132878
+            if (!event.isCancelled()) {
+                if (io.canvasmc.canvas.Config.INSTANCE.fixes.mc132878) {
+                    this.showBreakingParticles();
+                }
+                this.kill(level, damageSource, false);
+            } // CraftBukkit
+            // Canvas end - MC-132878
             // Paper end
         } else {
             this.setHealth(health);
