From 4d4ec3af021afd5dcd0fd24e659eea8297196156 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/server/network/config/PrepareSpawnTask.java b/net/minecraft/server/network/config/PrepareSpawnTask.java
index e2b97576d7d9a56fcc97146e532e1ff12e084d89..8c06fc75a08fa4be31534e855cda226aa4080261 100644
--- a/net/minecraft/server/network/config/PrepareSpawnTask.java
+++ b/net/minecraft/server/network/config/PrepareSpawnTask.java
@@ -80,7 +80,7 @@ public class PrepareSpawnTask implements ConfigurationTask {
                 if (bWorld != null) {
                     resourceKey = ((org.bukkit.craftbukkit.CraftWorld) bWorld).getHandle().dimension();
                 } else {
-                    resourceKey = net.minecraft.world.level.Level.OVERWORLD;
+                    resourceKey = io.canvasmc.canvas.Config.INSTANCE.fetchRespawnDimensionKey(); // Canvas - default respawn dimension config
                     invalidPlayerWorld[0] = true;
                 }
             }
