--- a/net/minecraft/server/network/config/PrepareSpawnTask.java
+++ b/net/minecraft/server/network/config/PrepareSpawnTask.java
@@ -78,7 +_,7 @@
                 if (bWorld != null) {
                     resourceKey = ((org.bukkit.craftbukkit.CraftWorld) bWorld).getHandle().dimension();
                 } else {
-                    resourceKey = net.minecraft.world.level.Level.OVERWORLD;
+                    resourceKey = io.canvasmc.canvas.Config.INSTANCE.fetchRespawnDimensionKey(); // Canvas - default respawn dimension config
                     invalidPlayerWorld[0] = true;
                 }
             }
