--- a/net/minecraft/world/level/biome/BiomeManager.java
+++ b/net/minecraft/world/level/biome/BiomeManager.java
@@ -14,6 +_,7 @@
     private static final int ZOOM_MASK = 3;
     private final BiomeManager.NoiseBiomeSource noiseBiomeSource;
     private final long biomeZoomSeed;
+    private static final double canvas$maxOffset = 0.4500000001D; // Canvas - optimize getBiome
 
     public BiomeManager(final BiomeManager.NoiseBiomeSource noiseBiomeSource, final long seed) {
         this.noiseBiomeSource = noiseBiomeSource;
@@ -28,30 +_,59 @@
         return new BiomeManager(biomeSource, this.biomeZoomSeed);
     }
 
-    public Holder<Biome> getBiome(final BlockPos pos) {
+    public net.minecraft.core.Holder<net.minecraft.world.level.biome.Biome> getBiome(final net.minecraft.core.BlockPos pos) {
+        // Canvas start - optimize getBiome
         int absX = pos.getX() - 2;
         int absY = pos.getY() - 2;
         int absZ = pos.getZ() - 2;
-        int parentX = absX >> 2;
+        int parentX = absX >> 2; // BlockPos to BiomePos
         int parentY = absY >> 2;
         int parentZ = absZ >> 2;
-        double fractX = (absX & 3) / 4.0;
-        double fractY = (absY & 3) / 4.0;
-        double fractZ = (absZ & 3) / 4.0;
+        double fractX = (absX & 3) / 4.0; // quartLocal divided by 4
+        double fractY = (absY & 3) / 4.0; // 0/4, 1/4, 2/4, 3/4
+        double fractZ = (absZ & 3) / 4.0; // [0, 0.25, 0.5, 0.75]
         int minI = 0;
         double minFiddledDistance = Double.POSITIVE_INFINITY;
-
         for (int i = 0; i < 8; i++) {
-            boolean xEven = (i & 4) == 0;
-            boolean yEven = (i & 2) == 0;
-            boolean zEven = (i & 1) == 0;
+            boolean xEven = (i & 4) == 0; // 1 1 1 1 0 0 0 0
+            boolean yEven = (i & 2) == 0; // 1 1 0 0 1 1 0 0
+            boolean zEven = (i & 1) == 0; // 1 0 1 0 1 0 1 0
+            double distanceX = xEven ? fractX : fractX - 1.0; //[-1.0,-0.75,-0.5,-0.25,0.0,0.25,0.5,0.75]
+            double distanceY = yEven ? fractY : fractY - 1.0;
+            double distanceZ = zEven ? fractZ : fractZ - 1.0;
+
+            // This code block is new
+            double maxDistanceY = 0.0D, maxDistanceZ = 0.0D;
+            if (i != 0) {
+                maxDistanceY = Mth.square(Math.max(distanceY + canvas$maxOffset, Math.abs(distanceY - canvas$maxOffset)));
+                maxDistanceZ = Mth.square(Math.max(distanceZ + canvas$maxOffset, Math.abs(distanceZ - canvas$maxOffset)));
+                double maxDistanceX = Mth.square(Math.max(distanceX + canvas$maxOffset, Math.abs(distanceX - canvas$maxOffset)));
+                if (minFiddledDistance < maxDistanceX + maxDistanceY + maxDistanceZ) continue;
+            }
+
             int cornerX = xEven ? parentX : parentX + 1;
             int cornerY = yEven ? parentY : parentY + 1;
             int cornerZ = zEven ? parentZ : parentZ + 1;
-            double distanceX = xEven ? fractX : fractX - 1.0;
-            double distanceY = yEven ? fractY : fractY - 1.0;
-            double distanceZ = zEven ? fractZ : fractZ - 1.0;
-            double next = getFiddledDistance(this.biomeZoomSeed, cornerX, cornerY, cornerZ, distanceX, distanceY, distanceZ);
+
+            //I transferred the code from method_38106 to here, so I could call continue halfway through
+            long seed = net.minecraft.util.LinearCongruentialGenerator.next(this.biomeZoomSeed, cornerX);
+            seed = net.minecraft.util.LinearCongruentialGenerator.next(seed, cornerY);
+            seed = net.minecraft.util.LinearCongruentialGenerator.next(seed, cornerZ);
+            seed = net.minecraft.util.LinearCongruentialGenerator.next(seed, cornerX);
+            seed = net.minecraft.util.LinearCongruentialGenerator.next(seed, cornerY);
+            seed = net.minecraft.util.LinearCongruentialGenerator.next(seed, cornerZ);
+            double offsetX = getFiddle(seed);
+            double sqrX = Mth.square(distanceX + offsetX);
+            if (i != 0 && minFiddledDistance < sqrX + maxDistanceY + maxDistanceZ) continue; //skip the rest of the loop
+            seed = net.minecraft.util.LinearCongruentialGenerator.next(seed, this.biomeZoomSeed);
+            double offsetY = getFiddle(seed);
+            double sqrY = Mth.square(distanceY + offsetY);
+            if (i != 0 && minFiddledDistance < sqrX + sqrY + maxDistanceZ) continue; // skip the rest of the loop
+            seed = net.minecraft.util.LinearCongruentialGenerator.next(seed, this.biomeZoomSeed);
+            double offsetZ = getFiddle(seed);
+            double next = sqrX + sqrY + Mth.square(distanceZ + offsetZ);
+            // Canvas end - optimize getBiome
+
             if (minFiddledDistance > next) {
                 minI = i;
                 minFiddledDistance = next;
@@ -98,9 +_,18 @@
         double fiddleZ = getFiddle(rval);
         return Mth.square(distanceZ + fiddleZ) + Mth.square(distanceY + fiddleY) + Mth.square(distanceX + fiddleX);
     }
+    // Canvas start - optimise getBiome
+
+    private static final double[] FIDDLE_TABLE = new double[1024];
+    static {
+        for (int i = 0; i < 1024; i++) {
+            FIDDLE_TABLE[i] = (i - 512) * (0.9 / 1024.0);
+        }
+    }
 
     private static double getFiddle(final long rval) {
-        return (double)(((rval >> 24) & (1024 - 1)) - (1024/2)) * (0.9 / 1024.0); // Paper - avoid floorMod, fp division, and fp subtraction
+        return FIDDLE_TABLE[(int)(rval >>> 24) & 1023];
+    // Canvas end - optimize getBiome
     }
 
     public interface NoiseBiomeSource {
