--- a/net/minecraft/world/level/biome/BiomeManager.java
+++ b/net/minecraft/world/level/biome/BiomeManager.java
@@ -14,6 +_,7 @@
     private static final int ZOOM_MASK = 3;
     private final BiomeManager.NoiseBiomeSource noiseBiomeSource;
     private final long biomeZoomSeed;
+    private static final double canvas$maxFiddle = 0.4500000001D; // Canvas - optimize getBiome
 
     public BiomeManager(final BiomeManager.NoiseBiomeSource noiseBiomeSource, final long seed) {
         this.noiseBiomeSource = noiseBiomeSource;
@@ -51,7 +_,33 @@
             double distanceX = xEven ? fractX : fractX - 1.0;
             double distanceY = yEven ? fractY : fractY - 1.0;
             double distanceZ = zEven ? fractZ : fractZ - 1.0;
-            double next = getFiddledDistance(this.biomeZoomSeed, cornerX, cornerY, cornerZ, distanceX, distanceY, distanceZ);
+            // Canvas start - optimize getBiome
+            // This code block is new
+            double minDistanceY = 0.0D, minDistanceZ = 0.0D;
+            if (i != 0) {
+                minDistanceY = Mth.square(Math.max(0.0, Math.abs(distanceY) - canvas$maxFiddle));
+                minDistanceZ = Mth.square(Math.max(0.0, Math.abs(distanceZ) - canvas$maxFiddle));
+                double minDistanceX = Mth.square(Math.max(0.0, Math.abs(distanceX) - canvas$maxFiddle));
+                if (minFiddledDistance <= minDistanceX + minDistanceY + minDistanceZ) continue;
+            }
+            // I transferred the code from getFiddledDistance to here, so I could call continue halfway through
+            long rval = LinearCongruentialGenerator.next(this.biomeZoomSeed, cornerX);
+            rval = LinearCongruentialGenerator.next(rval, cornerY);
+            rval = LinearCongruentialGenerator.next(rval, cornerZ);
+            rval = LinearCongruentialGenerator.next(rval, cornerX);
+            rval = LinearCongruentialGenerator.next(rval, cornerY);
+            rval = LinearCongruentialGenerator.next(rval, cornerZ);
+            double fiddleX = getFiddle(rval);
+            double sqrX = Mth.square(distanceX + fiddleX);
+            if (i != 0 && minFiddledDistance <= sqrX + minDistanceY + minDistanceZ) continue; //skip the rest of the loop
+            rval = LinearCongruentialGenerator.next(rval, this.biomeZoomSeed);
+            double fiddleY = getFiddle(rval);
+            double sqrY = Mth.square(distanceY + fiddleY);
+            if (i != 0 && minFiddledDistance <= sqrX + sqrY + minDistanceZ) continue; // skip the rest of the loop
+            rval = LinearCongruentialGenerator.next(rval, this.biomeZoomSeed);
+            double fiddleZ = getFiddle(rval);
+            double next = sqrX + sqrY + Mth.square(distanceZ + fiddleZ);
+            // Canvas end - optimize getBiome
             if (minFiddledDistance > next) {
                 minI = i;
                 minFiddledDistance = next;
@@ -99,9 +_,18 @@
         double fiddleZ = getFiddle(rval);
         return Mth.square(distanceZ + fiddleZ) + Mth.square(distanceY + fiddleY) + Mth.square(distanceX + fiddleX);
     }
+    // Canvas start - optimise getBiome
+
+    private static final double[] FIDDLE_TABLE = new double[1024];
+    static {
+        for (int i = 0; i < 1024; i++) {
+            FIDDLE_TABLE[i] = (i - 512) * (0.9 / 1024.0);
+        }
+    }
 
     private static double getFiddle(final long rval) {
-        return (double)(((rval >> 24) & (1024 - 1)) - (1024/2)) * (0.9 / 1024.0); // Paper - avoid floorMod, fp division, and fp subtraction
+        return FIDDLE_TABLE[(int)(rval >>> 24) & 1023];
+    // Canvas end - optimize getBiome
     }
 
     public interface NoiseBiomeSource {
