From 2be0fbf099df19422462cdb5b961fc00062d256b Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/biome/BiomeManager.java b/net/minecraft/world/level/biome/BiomeManager.java
index a0d7f6f8ecbc320267bdad27315fc2e2e1ff3702..bf87770ae4c5b8ee26fa7605d08402d508f1ffda 100644
--- a/net/minecraft/world/level/biome/BiomeManager.java
+++ b/net/minecraft/world/level/biome/BiomeManager.java
@@ -14,6 +14,7 @@ public class BiomeManager {
     private static final int ZOOM_MASK = 3;
     private final BiomeManager.NoiseBiomeSource noiseBiomeSource;
     private final long biomeZoomSeed;
+    private static final double canvas$maxOffset = 0.4500000001D; // Canvas - optimize getBiome
 
     public BiomeManager(final BiomeManager.NoiseBiomeSource noiseBiomeSource, final long seed) {
         this.noiseBiomeSource = noiseBiomeSource;
@@ -29,39 +30,68 @@ public class BiomeManager {
     }
 
     public Holder<Biome> getBiome(final BlockPos pos) {
-        int absX = pos.getX() - 2;
-        int absY = pos.getY() - 2;
-        int absZ = pos.getZ() - 2;
-        int parentX = absX >> 2;
-        int parentY = absY >> 2;
-        int parentZ = absZ >> 2;
-        double fractX = (absX & 3) / 4.0;
-        double fractY = (absY & 3) / 4.0;
-        double fractZ = (absZ & 3) / 4.0;
-        int minI = 0;
-        double minFiddledDistance = Double.POSITIVE_INFINITY;
-
-        for (int i = 0; i < 8; i++) {
-            boolean xEven = (i & 4) == 0;
-            boolean yEven = (i & 2) == 0;
-            boolean zEven = (i & 1) == 0;
-            int cornerX = xEven ? parentX : parentX + 1;
-            int cornerY = yEven ? parentY : parentY + 1;
-            int cornerZ = zEven ? parentZ : parentZ + 1;
-            double distanceX = xEven ? fractX : fractX - 1.0;
-            double distanceY = yEven ? fractY : fractY - 1.0;
-            double distanceZ = zEven ? fractZ : fractZ - 1.0;
-            double next = getFiddledDistance(this.biomeZoomSeed, cornerX, cornerY, cornerZ, distanceX, distanceY, distanceZ);
-            if (minFiddledDistance > next) {
-                minI = i;
-                minFiddledDistance = next;
+        // Canvas start - optimize getBiome
+        int xMinus2 = pos.getX() - 2;
+        int yMinus2 = pos.getY() - 2;
+        int zMinus2 = pos.getZ() - 2;
+        int x = xMinus2 >> 2; // BlockPos to BiomePos
+        int y = yMinus2 >> 2;
+        int z = zMinus2 >> 2;
+        double quartX = (double) (xMinus2 & 3) / 4.0D; // quartLocal divided by 4
+        double quartY = (double) (yMinus2 & 3) / 4.0D; // 0/4, 1/4, 2/4, 3/4
+        double quartZ = (double) (zMinus2 & 3) / 4.0D; // [0, 0.25, 0.5, 0.75]
+        int smallestX = 0;
+        double smallestDist = Double.POSITIVE_INFINITY;
+        for (int biomeX = 0; biomeX < 8; ++biomeX) {
+            boolean everyOtherQuad = (biomeX & 4) == 0; // 1 1 1 1 0 0 0 0
+            boolean everyOtherPair = (biomeX & 2) == 0; // 1 1 0 0 1 1 0 0
+            boolean everyOther = (biomeX & 1) == 0; // 1 0 1 0 1 0 1 0
+            double quartXX = everyOtherQuad ? quartX : quartX - 1.0D; //[-1.0,-0.75,-0.5,-0.25,0.0,0.25,0.5,0.75]
+            double quartYY = everyOtherPair ? quartY : quartY - 1.0D;
+            double quartZZ = everyOther ? quartZ : quartZ - 1.0D;
+
+            //This code block is new
+            double maxQuartYY = 0.0D, maxQuartZZ = 0.0D;
+            if (biomeX != 0) {
+                maxQuartYY = Mth.square(Math.max(quartYY + canvas$maxOffset, Math.abs(quartYY - canvas$maxOffset)));
+                maxQuartZZ = Mth.square(Math.max(quartZZ + canvas$maxOffset, Math.abs(quartZZ - canvas$maxOffset)));
+                double maxQuartXX = Mth.square(Math.max(quartXX + canvas$maxOffset, Math.abs(quartXX - canvas$maxOffset)));
+                if (smallestDist < maxQuartXX + maxQuartYY + maxQuartZZ) continue;
             }
-        }
 
-        int biomeX = (minI & 4) == 0 ? parentX : parentX + 1;
-        int biomeY = (minI & 2) == 0 ? parentY : parentY + 1;
-        int biomeZ = (minI & 1) == 0 ? parentZ : parentZ + 1;
-        return this.noiseBiomeSource.getNoiseBiome(biomeX, biomeY, biomeZ);
+            int xx = everyOtherQuad ? x : x + 1;
+            int yy = everyOtherPair ? y : y + 1;
+            int zz = everyOther ? z : z + 1;
+
+            //I transferred the code from method_38106 to here, so I could call continue halfway through
+            long seed = LinearCongruentialGenerator.next(this.biomeZoomSeed, xx);
+            seed = LinearCongruentialGenerator.next(seed, yy);
+            seed = LinearCongruentialGenerator.next(seed, zz);
+            seed = LinearCongruentialGenerator.next(seed, xx);
+            seed = LinearCongruentialGenerator.next(seed, yy);
+            seed = LinearCongruentialGenerator.next(seed, zz);
+            double offsetX = getFiddle(seed);
+            double sqrX = Mth.square(quartXX + offsetX);
+            if (biomeX != 0 && smallestDist < sqrX + maxQuartYY + maxQuartZZ) continue; //skip the rest of the loop
+            seed = LinearCongruentialGenerator.next(seed, this.biomeZoomSeed);
+            double offsetY = getFiddle(seed);
+            double sqrY = Mth.square(quartYY + offsetY);
+            if (biomeX != 0 && smallestDist < sqrX + sqrY + maxQuartZZ) continue; // skip the rest of the loop
+            seed = LinearCongruentialGenerator.next(seed, this.biomeZoomSeed);
+            double offsetZ = getFiddle(seed);
+            double biomeDist = sqrX + sqrY + Mth.square(quartZZ + offsetZ);
+
+            if (smallestDist > biomeDist) {
+                smallestX = biomeX;
+                smallestDist = biomeDist;
+            }
+        }
+        return this.noiseBiomeSource.getNoiseBiome(
+            (smallestX & 4) == 0 ? x : x + 1,
+            (smallestX & 2) == 0 ? y : y + 1,
+            (smallestX & 1) == 0 ? z : z + 1
+        );
+        // Canvas end - optimize getBiome
     }
 
     public Holder<Biome> getNoiseBiomeAtPosition(final double x, final double y, final double z) {
@@ -98,9 +128,18 @@ public class BiomeManager {
         double fiddleZ = getFiddle(rval);
         return Mth.square(distanceZ + fiddleZ) + Mth.square(distanceY + fiddleY) + Mth.square(distanceX + fiddleX);
     }
+    // Canvas start - optimise getBiome
+
+    private static final double[] FIDDLE_TABLE = new double[1024];
+    static {
+        for (int i = 0; i < 1024; i++) {
+            FIDDLE_TABLE[i] = (i - 512) * (0.9 / 1024.0);
+        }
+    }
 
     private static double getFiddle(final long rval) {
-        return (double)(((rval >> 24) & (1024 - 1)) - (1024/2)) * (0.9 / 1024.0); // Paper - avoid floorMod, fp division, and fp subtraction
+        return FIDDLE_TABLE[(int)(rval >>> 24) & 1023];
+    // Canvas end - optimize getBiome
     }
 
     public interface NoiseBiomeSource {
