From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/block/entity/BlockEntity.java b/net/minecraft/world/level/block/entity/BlockEntity.java
index d67b1ed5b11a6678a1aa4572b26e48a8cbec9b0a..b1878e868f3a4d716fd101177d00fde0185802db 100644
--- a/net/minecraft/world/level/block/entity/BlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BlockEntity.java
@@ -52,6 +52,7 @@ public abstract class BlockEntity implements DebugValueSource, TypedInstance<Blo
     protected boolean remove;
     private BlockState blockState;
     private DataComponentMap components = DataComponentMap.EMPTY;
+    public boolean canvas$isValid; // Canvas - optimize block entities
 
     // Folia start - region ticking
     public void updateTicks(final long fromTickOffset, final long fromRedstoneTimeOffset) {
@@ -65,6 +66,7 @@ public abstract class BlockEntity implements DebugValueSource, TypedInstance<Blo
         this.validateBlockState(blockState);
         this.blockState = blockState;
         this.persistentDataContainer = new org.bukkit.craftbukkit.persistence.CraftPersistentDataContainer(DATA_TYPE_REGISTRY); // Paper - always init
+        this.canvas$isValid = type.isValid(blockState); // Canvas - optimize block entities
     }
 
     private void validateBlockState(final BlockState blockState) {
@@ -224,6 +226,11 @@ public abstract class BlockEntity implements DebugValueSource, TypedInstance<Blo
             if (IGNORE_TILE_UPDATES.get().booleanValue()) return; // Paper - Perf: Optimize Hoppers // Folia - region threading
             setChanged(this.level, this.worldPosition, this.blockState);
         }
+        // Canvas start - block entity sleeping
+        if (this instanceof io.canvasmc.canvas.world.block.SleepingBlockEntity sleeping) {
+            sleeping.onChange();
+        }
+        // Canvas end - block entity sleeping
     }
 
     protected static void setChanged(final Level level, final BlockPos worldPosition, final BlockState blockState) {
@@ -299,6 +306,7 @@ public abstract class BlockEntity implements DebugValueSource, TypedInstance<Blo
     public void setBlockState(final BlockState blockState) {
         this.validateBlockState(blockState);
         this.blockState = blockState;
+        this.canvas$isValid = this.type.isValid(blockState); // Canvas - optimize block entities
     }
 
     protected void applyImplicitComponents(final DataComponentGetter components) {
@@ -376,7 +384,7 @@ public abstract class BlockEntity implements DebugValueSource, TypedInstance<Blo
 
     // CraftBukkit start - add method
     public org.bukkit.inventory.@Nullable InventoryHolder getOwner() {
-        return getOwner(true);
+        return getOwner(io.canvasmc.canvas.GlobalConfiguration.getInstance().tileEntitySnapshotCreation); // Canvas - don't use snapshots for acquiring blockstate
     }
 
     public org.bukkit.inventory.@Nullable InventoryHolder getOwner(boolean useSnapshot) {
