From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/block/entity/HopperBlockEntity.java b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
index 6ab58fc269f83f67e56213f9d3133c5e11754bc5..c6f33c8124fe26061503428e3d31bdba67cb6338 100644
--- a/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -95,6 +95,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
         if (!this.tryLoadLootTable(input)) {
             ContainerHelper.loadAllItems(input, this.items);
         }
+        canvas$recomputeFullState(this); // Canvas - optimize hoppers
 
         this.cooldownTime = input.getIntOr("TransferCooldown", -1);
     }
@@ -117,7 +118,11 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     @Override
     public ItemStack removeItem(final int slot, final int count) {
         this.unpackLootTable(null);
-        return ContainerHelper.removeItem(this.getItems(), slot, count);
+        // Canvas start - optimize hoppers
+        ItemStack retVal = ContainerHelper.removeItem(this.getItems(), slot, count);
+        canvas$recomputeFullState(this);
+        return retVal;
+        // Canvas end - optimize hoppers
     }
 
     @Override
@@ -125,13 +130,24 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
         this.unpackLootTable(null);
         this.getItems().set(slot, itemStack);
         itemStack.limitSize(this.getMaxStackSize(itemStack));
+        canvas$recomputeFullState(this); // Canvas - optimize hoppers
     }
 
     @Override
     public void setBlockState(final BlockState blockState) {
         super.setBlockState(blockState);
         this.facing = blockState.getValue(HopperBlock.FACING);
+        canvas$recomputeFullState(this); // Canvas - optimize hoppers
+        // Canvas - TODO - optimize hoppers - cache is locked and run tick skip system through here?
     }
+    // Canvas start - optimize hoppers
+
+    @Override
+    public void clearContent() {
+        super.clearContent();
+        canvas$recomputeFullState(this);
+    }
+    // Canvas end - optimize hoppers
 
     @Override
     protected Component getDefaultName() {
@@ -156,8 +172,17 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     private static final int HOPPER_EMPTY = 0;
     private static final int HOPPER_HAS_ITEMS = 1;
     private static final int HOPPER_IS_FULL = 2;
+    // Canvas start - optimize hoppers
+    private int fullState = -1;
+
+    @Override
+    public void setChanged() {
+        super.setChanged();
+        canvas$recomputeFullState(this);
+    }
 
-    private static int getFullState(final HopperBlockEntity hopper) {
+    private static void canvas$recomputeFullState(final HopperBlockEntity hopper) {
+    // Canvas end - optimize hoppers
         hopper.unpackLootTable(null);
 
         final List<ItemStack> hopperItems = hopper.items;
@@ -174,18 +199,20 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
 
             if (!full) {
                 // can't be full
-                return HOPPER_HAS_ITEMS;
+                hopper.fullState = HOPPER_HAS_ITEMS; // Canvas - optimize hoppers
+                return; // Canvas - optimize hoppers
             }
 
             empty = false;
 
             if (stack.getCount() != stack.getMaxStackSize()) {
                 // can't be full or empty
-                return HOPPER_HAS_ITEMS;
+                hopper.fullState = HOPPER_HAS_ITEMS; // Canvas - optimize hoppers
+                return; // Canvas - optimize hoppers
             }
         }
 
-        return empty ? HOPPER_EMPTY : (full ? HOPPER_IS_FULL : HOPPER_HAS_ITEMS);
+        hopper.fullState = empty ? HOPPER_EMPTY : (full ? HOPPER_IS_FULL : HOPPER_HAS_ITEMS); // Canvas - optimize hoppers
     }
     // Paper end - Perf: Optimize Hoppers
 
@@ -198,7 +225,8 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
 
         if (!entity.isOnCooldown() && state.getValue(HopperBlock.ENABLED)) {
             boolean changed = false;
-            final int fullState = getFullState(entity); // Paper - Perf: Optimize Hoppers
+            if (entity.fullState == -1) canvas$recomputeFullState(entity); // Canvas - optimize hoppers
+            final int fullState = entity.fullState; // Paper - Perf: Optimize Hoppers // Canvas - optimize hoppers
             if (fullState != HOPPER_EMPTY) { // Paper - Perf: Optimize Hoppers
                 changed = ejectItems(level, pos, entity);
             }
@@ -700,7 +728,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
         final boolean optimizeEntities
     ) {
         // Paper end - Perf: Optimize Hoppers
-        Container result = getBlockContainer(level, pos, state);
+        Container result = state.hasBlockEntity() || state.getBlock() instanceof WorldlyContainerHolder ? getBlockContainer(level, pos, state) : null; // Canvas - optimize hoppers - if the block has no block entity, why check it
         if (result == null && (!optimizeEntities || !level.paperConfig().hopper.ignoreOccludingBlocks || !state.getBukkitMaterial().isOccluding())) { // Paper - Perf: Optimize Hoppers
             result = getEntityContainer(level, x, y, z);
         }
@@ -717,7 +745,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
 
         if (state.hasBlockEntity() && level.getBlockEntity(pos) instanceof Container container) {
             if (container instanceof ChestBlockEntity && block instanceof ChestBlock) {
-                container = ChestBlock.getContainer((ChestBlock)block, state, level, pos, true);
+                container = ChestBlock.canvas$getContainerForHoppers((ChestBlock)block, state, level, pos, true, container); // Canvas - optimize hoppers
             }
 
             return container;
@@ -777,6 +805,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
     @Override
     protected void setItems(final NonNullList<ItemStack> items) {
         this.items = items;
+        canvas$recomputeFullState(this); // Canvas - optimize hoppers
     }
 
     public static void entityInside(final Level level, final BlockPos pos, final BlockState blockState, final Entity entity, final HopperBlockEntity hopper) {
