--- a/net/minecraft/world/level/gameevent/EuclideanGameEventListenerRegistry.java
+++ b/net/minecraft/world/level/gameevent/EuclideanGameEventListenerRegistry.java
@@ -13,9 +_,9 @@
 import net.minecraft.world.phys.Vec3;
 
 public class EuclideanGameEventListenerRegistry implements GameEventListenerRegistry {
-    private final List<GameEventListener> listeners = Lists.newArrayList();
-    private final Set<GameEventListener> listenersToRemove = Sets.newHashSet();
-    private final List<GameEventListener> listenersToAdd = Lists.newArrayList();
+    private final Set<GameEventListener> listeners = Sets.newConcurrentHashSet(); // Canvas
+    private final Set<GameEventListener> listenersToRemove = Sets.newConcurrentHashSet(); // Canvas
+    private final Set<GameEventListener> listenersToAdd = Sets.newConcurrentHashSet(); // Canvas
     private boolean processing;
     private final ServerLevel level;
     private final int sectionY;
