From 53e1a57c72953de1bfcc94c3257c724e7fa1f88f Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/network/codec/ByteBufCodecs.java b/net/minecraft/network/codec/ByteBufCodecs.java
index 36aa47326ecf0434378954dbec97f0bf8b7e83c3..9efc2f50bd54e7e4b2c5eca01487c296a00c9cb2 100644
--- a/net/minecraft/network/codec/ByteBufCodecs.java
+++ b/net/minecraft/network/codec/ByteBufCodecs.java
@@ -782,9 +782,35 @@ public interface ByteBufCodecs {
 
             @Override
             public void encode(final ByteBuf output, final JsonElement value) {
+                // Canvas start - no chat reports
+                if (io.canvasmc.canvas.GlobalConfiguration.getInstance().chat.disableChatReporting && this.canvas$isServerStatusElement(value)) {
+                    com.google.gson.JsonObject object = value.getAsJsonObject();
+                    object.addProperty("preventsChatReports", true);
+
+                    String string = GSON.toJson(object);
+                    Utf8String.write(output, string, 32767);
+                    return;
+                }
+                // Canvas end - no chat reports
                 String payload = GSON.toJson(value);
                 Utf8String.write(output, payload, maxStringLength);
             }
+            // Canvas start - no chat reports
+
+            private boolean canvas$isServerStatusElement(JsonElement element) {
+                if (!element.isJsonObject())
+                    return false;
+
+                com.google.gson.JsonObject object = element.getAsJsonObject();
+
+                if (!object.has("version") || !object.get("version").isJsonObject())
+                    return false;
+
+                com.google.gson.JsonObject version = object.get("version").getAsJsonObject();
+
+                return version.has("name") && version.has("protocol");
+            }
+            // Canvas end - no chat reports
         };
     }
 }
