From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/sensing/GolemSensor.java b/net/minecraft/world/entity/ai/sensing/GolemSensor.java
index 84d9e2a43adbabda8401e8ad8dd8d87f7dbeeea7..927fadde00c3fd85a5145cece524f35a1b908933 100644
--- a/net/minecraft/world/entity/ai/sensing/GolemSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/GolemSensor.java
@@ -34,7 +34,15 @@ public class GolemSensor extends Sensor<LivingEntity> {
     public static void checkForNearbyGolem(LivingEntity livingEntity) {
         Optional<List<LivingEntity>> memory = livingEntity.getBrain().getMemory(MemoryModuleType.NEAREST_LIVING_ENTITIES);
         if (!memory.isEmpty()) {
-            boolean flag = memory.get().stream().anyMatch(entity -> entity.getType().equals(EntityType.IRON_GOLEM));
+            // Canvas start - remove streams
+            boolean flag = false;
+            for (LivingEntity entity : memory.get()) {
+                if (entity.getType().equals(EntityType.IRON_GOLEM)) {
+                    flag = true;
+                    break;
+                }
+            }
+            // Canvas end - remove streams
             if (flag) {
                 golemDetected(livingEntity);
             }
