From 56a92b2c479a7161c340134ba931fb3d6a8e6f79 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/projectile/FishingHook.java b/net/minecraft/world/entity/projectile/FishingHook.java
index cb4a2efb90eef13d74be3f8a755a240625d5a659..7e7d88a9ce1cff9eecafa978a83b704a92e9f4fb 100644
--- a/net/minecraft/world/entity/projectile/FishingHook.java
+++ b/net/minecraft/world/entity/projectile/FishingHook.java
@@ -508,6 +508,11 @@ public class FishingHook extends Projectile {
         if (!this.level().isClientSide() && owner != null && !this.shouldStopFishing(owner)) {
             int dmg = 0;
             if (this.hookedIn != null) {
+                // Canvas start - combat configs
+                if (!level().canvasConfig().combat.allowFishingRodsToPullEntities) {
+                    return 0; // act as if we canceled the pull
+                }
+                // Canvas end - combat configs
                 // CraftBukkit start
                 org.bukkit.event.player.PlayerFishEvent playerFishEvent = new org.bukkit.event.player.PlayerFishEvent((org.bukkit.entity.Player) owner.getBukkitEntity(), this.hookedIn.getBukkitEntity(), (org.bukkit.entity.FishHook) this.getBukkitEntity(), org.bukkit.craftbukkit.CraftEquipmentSlot.getHand(hand), org.bukkit.event.player.PlayerFishEvent.State.CAUGHT_ENTITY); // Paper - Add hand parameter to PlayerFishEvent
                 if (!playerFishEvent.callEvent()) {
@@ -601,6 +606,14 @@ public class FishingHook extends Projectile {
         if (owner != null) {
             Vec3 delta = new Vec3(owner.getX() - this.getX(), owner.getY() - this.getY(), owner.getZ() - this.getZ()).scale(0.1);
             entity.setDeltaMovement(entity.getDeltaMovement().add(delta));
+            // Canvas start - MC-100991
+            if (io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaFixes.mc100991 && entity instanceof net.minecraft.world.entity.LivingEntity livingEntity) {
+                livingEntity.getCombatTracker().recordDamage(
+                    level().damageSources().thrown(this, owner),
+                    livingEntity.getHealth()
+                );
+            }
+            // Canvas end - MC-100991
         }
     }
 
