From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/projectile/ProjectileDeflection.java b/net/minecraft/world/entity/projectile/ProjectileDeflection.java
index d5ba900d25cd5c7a51b9050549af519fb35d4a89..08899ac9bbc79c8d6beea9907df9b426e1aaa3e1 100644
--- a/net/minecraft/world/entity/projectile/ProjectileDeflection.java
+++ b/net/minecraft/world/entity/projectile/ProjectileDeflection.java
@@ -17,7 +17,7 @@ public interface ProjectileDeflection {
     };
     ProjectileDeflection AIM_DEFLECT = (projectile, entity, random) -> {
         if (entity != null) {
-            Vec3 lookAngle = entity.getLookAngle();
+            Vec3 lookAngle = entity.canvas$getLookAngleThreadSafe(); // Canvas - region threading
             projectile.setDeltaMovement(lookAngle);
             projectile.needsSync = true;
         }
