--- a/src/main/java/org/bukkit/event/HandlerList.java
+++ b/src/main/java/org/bukkit/event/HandlerList.java
@@ -247,4 +_,12 @@
             return (ArrayList<HandlerList>) allLists.clone();
         }
     }
+    // Canvas start - allow getting the size
+
+    // if its null, it has no listeners
+    // if the length is <0, it has no listeners
+    public boolean hasRegisteredListeners() {
+        return this.handlers != null && this.handlers.length > 0;
+    }
+    // Canvas end
 }
