From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/sensing/MobSensor.java b/net/minecraft/world/entity/ai/sensing/MobSensor.java
index bda210b4809a5aade7ab4d0f26fdda4d5f53f619..cfb0ea1a89b074bcf9fc9b74eeae6e75e64bd707 100644
--- a/net/minecraft/world/entity/ai/sensing/MobSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/MobSensor.java
@@ -40,10 +40,15 @@ public class MobSensor<T extends LivingEntity> extends Sensor<T> {
     public void checkForMobsNearby(T sensingEntity) {
         Optional<List<LivingEntity>> memory = sensingEntity.getBrain().getMemory(MemoryModuleType.NEAREST_LIVING_ENTITIES);
         if (!memory.isEmpty()) {
-            boolean flag = memory.get().stream().anyMatch(livingEntity -> this.mobTest.test(sensingEntity, livingEntity));
-            if (flag) {
-                this.mobDetected(sensingEntity);
+            // Canvas start - remove streams
+            List<LivingEntity> entities = memory.get();
+            for (LivingEntity livingEntity : entities) {
+                if (this.mobTest.test(sensingEntity, livingEntity)) {
+                    this.mobDetected(sensingEntity);
+                    break;
+                }
             }
+            // Canvas end - remove streams
         }
     }
 
