From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/sensing/SecondaryPoiSensor.java b/net/minecraft/world/entity/ai/sensing/SecondaryPoiSensor.java
index 39a9e9a6f6dfddbf47a7f96150ef832efe61f89e..212333d88e4f68de62f11ae4db6c268783711e47 100644
--- a/net/minecraft/world/entity/ai/sensing/SecondaryPoiSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/SecondaryPoiSensor.java
@@ -22,6 +22,13 @@ public class SecondaryPoiSensor extends Sensor<Villager> {
 
     @Override
     protected void doTick(ServerLevel level, Villager entity) {
+        // Canvas start - skip secondary poi sensor if empty
+        final ImmutableSet<net.minecraft.world.level.block.Block> secondaryPoi = entity.getVillagerData().profession().value().secondaryPoi();
+        if (secondaryPoi.isEmpty()) {
+            entity.getBrain().eraseMemory(MemoryModuleType.SECONDARY_JOB_SITE);
+            return;
+        }
+        // Canvas end - skip secondary poi sensor if empty
         ResourceKey<Level> resourceKey = level.dimension();
         BlockPos blockPos = entity.blockPosition();
         List<GlobalPos> list = Lists.newArrayList();
@@ -31,7 +38,7 @@ public class SecondaryPoiSensor extends Sensor<Villager> {
             for (int i2 = -2; i2 <= 2; i2++) {
                 for (int i3 = -4; i3 <= 4; i3++) {
                     BlockPos blockPos1 = blockPos.offset(i1, i2, i3);
-                    if (entity.getVillagerData().profession().value().secondaryPoi().contains(level.getBlockState(blockPos1).getBlock())) {
+                    if (secondaryPoi.contains(level.getBlockState(blockPos1).getBlock())) { // Canvas - skip secondary poi sensor if empty
                         list.add(GlobalPos.of(resourceKey, blockPos1));
                     }
                 }
