From 56a92b2c479a7161c340134ba931fb3d6a8e6f79 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/block/entity/vault/VaultSharedData.java b/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
index 41154db05edbb777fbe08e1e79b4f30f12a3cc80..303213ad33d8133d3885141c96088d0173f47964 100644
--- a/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
+++ b/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
@@ -67,11 +67,14 @@ public class VaultSharedData {
     void updateConnectedPlayersWithinRange(
         final ServerLevel serverLevel, final BlockPos pos, final VaultServerData serverData, final VaultConfig config, final double limit
     ) {
-        Set<UUID> currentConnectedPlayers = config.playerDetector()
-            .detect(serverLevel, config.entitySelector(), pos, limit, false)
-            .stream()
-            .filter(uuid -> !serverData.getRewardedPlayers().contains(uuid))
-            .collect(Collectors.toSet());
+        // Canvas start - remove streams
+        Set<UUID> currentConnectedPlayers = new ObjectLinkedOpenHashSet<>();
+        for (UUID uuid : config.playerDetector().detect(serverLevel, config.entitySelector(), pos, limit, false)) {
+            if (!serverData.getRewardedPlayers().contains(uuid)) {
+                currentConnectedPlayers.add(uuid);
+            }
+        }
+        // Canvas end - remove streams
         if (!this.connectedPlayers.equals(currentConnectedPlayers)) {
             this.connectedPlayers = currentConnectedPlayers;
             this.markDirty();
