From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/block/entity/vault/VaultSharedData.java b/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
index 036087f8085f05b43703ceca3413e0b8ac0a412b..d500b884ce9ed8711a99bc0ee65526ec8aa28a4b 100644
--- a/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
+++ b/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
@@ -65,11 +65,17 @@ public class VaultSharedData {
     }
 
     void updateConnectedPlayersWithinRange(ServerLevel level, BlockPos pos, VaultServerData serverData, VaultConfig config, double deactivationRange) {
-        Set<UUID> set = config.playerDetector()
-            .detect(level, config.entitySelector(), pos, deactivationRange, false)
-            .stream()
-            .filter(uuid -> !serverData.getRewardedPlayers().contains(uuid))
-            .collect(Collectors.toSet());
+        // Canvas start - remove streams
+        java.util.List<UUID> detectedPlayers = config.playerDetector().detect(level, config.entitySelector(), pos, deactivationRange, false);
+        Set<UUID> set = new ObjectLinkedOpenHashSet<>();
+
+        for (UUID uuid : detectedPlayers) {
+            if (!serverData.getRewardedPlayers().contains(uuid)) {
+                set.add(uuid);
+            }
+        }
+        // Canvas end - remove streams
+
         if (!this.connectedPlayers.equals(set)) {
             this.connectedPlayers = set;
             this.markDirty();
