--- a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
@@ -28,6 +_,7 @@
 import java.util.Arrays;
 import java.util.Iterator;
 import java.util.List;
+import java.util.concurrent.locks.ReentrantLock;
 import java.util.function.Predicate;
 
 public final class ChunkEntitySlices {
@@ -203,8 +_,10 @@
         }
     }
 
-    private boolean preventStatusUpdates;
+    private final ReentrantLock statusLock = new ReentrantLock(); // Canvas - status locks
+    private volatile boolean preventStatusUpdates; // Canvas - volatile
     public boolean startPreventingStatusUpdates() {
+        this.statusLock.lock(); // Canvas - status locks
         final boolean ret = this.preventStatusUpdates;
         this.preventStatusUpdates = true;
         return ret;
@@ -216,6 +_,7 @@
 
     public void stopPreventingStatusUpdates(final boolean prev) {
         this.preventStatusUpdates = prev;
+        this.statusLock.unlock(); // Canvas - status locks
     }
 
     public void updateStatus(final FullChunkStatus status, final EntityLookup lookup) {
