From e34741f5a3aa1ffcca00d20d247b01a2d0c013ee Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
index 8d9ce3d301d5f7e4106587ae00adb8dd5b8b6f88..193c3ad4976a0239fd474f047a30d18c9aaa65b9 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
@@ -198,6 +198,12 @@ public final class ChunkEntitySlices {
         return this.entities.size() != 0;
     }
 
+    // Canvas start - getEntityArr
+    public Entity[] getEntityArr() {
+        return this.entities.getRawData();
+    }
+
+    // Canvas end - getEntityArr
     public List<Entity> getAllEntities() {
         final int len = this.entities.size();
         if (len == 0) {
