From 9a2c48187d2c733eeb18126deb4d18e2547c666f Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/sensing/GolemSensor.java b/net/minecraft/world/entity/ai/sensing/GolemSensor.java
index 730b372e46d75e43a553b49dad41cc2d2f7510ee..80bfd5f7eaef1cce6e93f0887cf3746e4d4fcd75 100644
--- a/net/minecraft/world/entity/ai/sensing/GolemSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/GolemSensor.java
@@ -33,7 +33,15 @@ public class GolemSensor extends Sensor<LivingEntity> {
     public static void checkForNearbyGolem(final LivingEntity body) {
         Optional<List<LivingEntity>> livingEntitiesMemory = body.getBrain().getMemory(MemoryModuleType.NEAREST_LIVING_ENTITIES);
         if (!livingEntitiesMemory.isEmpty()) {
-            boolean golemPresent = livingEntitiesMemory.get().stream().anyMatch(entity -> entity.is(EntityType.IRON_GOLEM));
+            // Canvas start - remove streams
+            boolean golemPresent = false;
+            for (final LivingEntity living : livingEntitiesMemory.get()) {
+                if (living.is(EntityType.IRON_GOLEM)) {
+                    golemPresent = true;
+                    break;
+                }
+            }
+            // Canvas end - remove streams
             if (golemPresent) {
                 golemDetected(body);
             }
