--- a/net/minecraft/world/entity/ai/sensing/MobSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/MobSensor.java
@@ -46,10 +_,14 @@
     public void checkForMobsNearby(final T body) {
         Optional<List<LivingEntity>> livingEntitiesMemory = body.getBrain().getMemory(MemoryModuleType.NEAREST_LIVING_ENTITIES);
         if (!livingEntitiesMemory.isEmpty()) {
-            boolean mobPresent = livingEntitiesMemory.get().stream().anyMatch(entity -> this.mobTest.test(body, entity));
-            if (mobPresent) {
-                this.mobDetected(body);
+            // Canvas start - remove streams
+            for (final LivingEntity living : livingEntitiesMemory.get()) {
+                if (this.mobTest.test(body, living)) {
+                    this.mobDetected(body);
+                    return;
+                }
             }
+            // Canvas end - remove streams
         }
     }
 
