--- a/net/minecraft/world/entity/ai/sensing/MobSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/MobSensor.java
@@ -40,10 +_,15 @@
     public void checkForMobsNearby(T sensingEntity) {
         Optional<List<LivingEntity>> memory = sensingEntity.getBrain().getMemory(MemoryModuleType.NEAREST_LIVING_ENTITIES);
         if (!memory.isEmpty()) {
-            boolean flag = memory.get().stream().anyMatch(livingEntity -> this.mobTest.test(sensingEntity, livingEntity));
-            if (flag) {
-                this.mobDetected(sensingEntity);
+            // Canvas start - remove streams
+            List<LivingEntity> entities = memory.get();
+            for (LivingEntity livingEntity : entities) {
+                if (this.mobTest.test(sensingEntity, livingEntity)) {
+                    this.mobDetected(sensingEntity);
+                    break;
+                }
             }
+            // Canvas end - remove streams
         }
     }
 
