From 31ebb67cda7f3f00de773448b77d99e750a63ace Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/sensing/MobSensor.java b/net/minecraft/world/entity/ai/sensing/MobSensor.java
index ac9c50d680814b5752f7c712633555d587633598..02cab0978f8d675aa2f057e79d83e07c62c1b8c4 100644
--- a/net/minecraft/world/entity/ai/sensing/MobSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/MobSensor.java
@@ -46,10 +46,14 @@ public class MobSensor<T extends LivingEntity> extends Sensor<T> {
     public void checkForMobsNearby(final T body) {
         Optional<List<LivingEntity>> livingEntitiesMemory = body.getBrain().getMemory(MemoryModuleType.NEAREST_LIVING_ENTITIES);
         if (!livingEntitiesMemory.isEmpty()) {
-            boolean mobPresent = livingEntitiesMemory.get().stream().anyMatch(entity -> this.mobTest.test(body, entity));
-            if (mobPresent) {
-                this.mobDetected(body);
+            // Canvas start - remove streams
+            for (final LivingEntity living : livingEntitiesMemory.get()) {
+                if (this.mobTest.test(body, living)) {
+                    this.mobDetected(body);
+                    return;
+                }
             }
+            // Canvas end - remove streams
         }
     }
 
