--- a/net/minecraft/world/entity/ai/sensing/NearestItemSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/NearestItemSensor.java
@@ -24,8 +_,8 @@
     @Override
     protected void doTick(ServerLevel level, Mob entity) {
         Brain<?> brain = entity.getBrain();
-        List<ItemEntity> entitiesOfClass = level.getEntitiesOfClass(ItemEntity.class, entity.getBoundingBox().inflate(32.0, 16.0, 32.0), itemEntity -> itemEntity.closerThan(entity, MAX_DISTANCE_TO_WANTED_ITEM) && entity.wantsToPickUp(level, itemEntity.getItem())); // Paper - Perf: Move predicate into getEntities
-        entitiesOfClass.sort(Comparator.comparingDouble(entity::distanceToSqr));
+        it.unimi.dsi.fastutil.objects.ObjectArrayList<ItemEntity> entitiesOfClass = (it.unimi.dsi.fastutil.objects.ObjectArrayList<ItemEntity>) level.getEntitiesOfClass(ItemEntity.class, entity.getBoundingBox().inflate(32.0, 16.0, 32.0), itemEntity -> itemEntity.closerThan(entity, MAX_DISTANCE_TO_WANTED_ITEM) && entity.wantsToPickUp(level, itemEntity.getItem())); // Paper - Perf: Move predicate into getEntities // Canvas - faster sorting
+        entitiesOfClass.unstableSort(Comparator.comparingDouble(entity::distanceToSqr)); // Canvas - faster sorting
         // Paper start - Perf: remove streams from hot code
         ItemEntity nearest = null;
         for (final ItemEntity itemEntity : entitiesOfClass) {
