From cedb5bb62375ffca015c927898897c43ca743e2f Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/sensing/NearestLivingEntitySensor.java b/net/minecraft/world/entity/ai/sensing/NearestLivingEntitySensor.java
index 3578ece216827fd5d8c1206c689fc608e97b50df..b516105ae8939739ee2b5c22322bd152b4fd5ec1 100644
--- a/net/minecraft/world/entity/ai/sensing/NearestLivingEntitySensor.java
+++ b/net/minecraft/world/entity/ai/sensing/NearestLivingEntitySensor.java
@@ -17,8 +17,10 @@ public class NearestLivingEntitySensor<T extends LivingEntity> extends Sensor<T>
     protected void doTick(final ServerLevel level, final T body) {
         double followRange = body.getAttributeValue(Attributes.FOLLOW_RANGE);
         AABB boundingBox = body.getBoundingBox().inflate(followRange, followRange, followRange);
-        List<LivingEntity> livingEntities = level.getEntitiesOfClass(LivingEntity.class, boundingBox, mob -> mob != body && mob.isAlive());
-        livingEntities.sort(Comparator.comparingDouble(body::distanceToSqr));
+        // Canvas start - faster sorting
+        it.unimi.dsi.fastutil.objects.ObjectArrayList<LivingEntity> livingEntities = (it.unimi.dsi.fastutil.objects.ObjectArrayList<LivingEntity>) level.getEntitiesOfClass(LivingEntity.class, boundingBox, mob -> mob != body && mob.isAlive());
+        livingEntities.unstableSort(Comparator.comparingDouble(body::distanceToSqr));
+        // Canvas end - faster sorting
         Brain<?> brain = body.getBrain();
         brain.setMemory(MemoryModuleType.NEAREST_LIVING_ENTITIES, livingEntities);
         brain.setMemory(MemoryModuleType.NEAREST_VISIBLE_LIVING_ENTITIES, new NearestVisibleLivingEntities(level, body, livingEntities));
