--- a/net/minecraft/world/entity/ai/sensing/NearestItemSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/NearestItemSensor.java
@@ -24,8 +_,8 @@
     @Override
     protected void doTick(final ServerLevel level, final Mob body) {
         Brain<?> brain = body.getBrain();
-        List<ItemEntity> items = level.getEntitiesOfClass(ItemEntity.class, body.getBoundingBox().inflate(32.0, 16.0, 32.0), item -> item.closerThan(body, MAX_DISTANCE_TO_WANTED_ITEM) && body.wantsToPickUp(level, item.getItem())); // Paper - Perf: Move predicate into getEntities
-        items.sort(Comparator.comparingDouble(body::distanceToSqr));
+        it.unimi.dsi.fastutil.objects.ObjectArrayList<ItemEntity> items = (it.unimi.dsi.fastutil.objects.ObjectArrayList<ItemEntity>) level.getEntitiesOfClass(ItemEntity.class, body.getBoundingBox().inflate(32.0, 16.0, 32.0), item -> item.closerThan(body, MAX_DISTANCE_TO_WANTED_ITEM) && body.wantsToPickUp(level, item.getItem())); // Paper - Perf: Move predicate into getEntities // Canvas - faster sorting
+        items.unstableSort(Comparator.comparingDouble(body::distanceToSqr)); // Canvas - faster sorting
         // Paper start - Perf: remove streams from hot code
         ItemEntity nearest = null;
         for (final ItemEntity item : items) {
