--- a/net/minecraft/world/entity/ai/sensing/NearestLivingEntitySensor.java
+++ b/net/minecraft/world/entity/ai/sensing/NearestLivingEntitySensor.java
@@ -17,8 +_,10 @@
     protected void doTick(final ServerLevel level, final T body) {
         double followRange = body.getAttributeValue(Attributes.FOLLOW_RANGE);
         AABB boundingBox = body.getBoundingBox().inflate(followRange, followRange, followRange);
-        List<LivingEntity> livingEntities = level.getEntitiesOfClass(LivingEntity.class, boundingBox, mob -> mob != body && mob.isAlive());
-        livingEntities.sort(Comparator.comparingDouble(body::distanceToSqr));
+        // Canvas start - faster sorting
+        it.unimi.dsi.fastutil.objects.ObjectArrayList<LivingEntity> livingEntities = (it.unimi.dsi.fastutil.objects.ObjectArrayList<LivingEntity>) level.getEntitiesOfClass(LivingEntity.class, boundingBox, mob -> mob != body && mob.isAlive());
+        livingEntities.unstableSort(Comparator.comparingDouble(body::distanceToSqr));
+        // Canvas end - faster sorting
         Brain<?> brain = body.getBrain();
         brain.setMemory(MemoryModuleType.NEAREST_LIVING_ENTITIES, livingEntities);
         brain.setMemory(MemoryModuleType.NEAREST_VISIBLE_LIVING_ENTITIES, new NearestVisibleLivingEntities(level, body, livingEntities));
