From e64d3ec2ef48556e0a10e7686e1c057ce53f22aa Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/sensing/NearestItemSensor.java b/net/minecraft/world/entity/ai/sensing/NearestItemSensor.java
index 8aeea55380462421cde43afb4b2fb0ce44b65195..05e1d03cebd6f4675abbb48c7903232a1a91d7d1 100644
--- a/net/minecraft/world/entity/ai/sensing/NearestItemSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/NearestItemSensor.java
@@ -24,8 +24,8 @@ public class NearestItemSensor extends Sensor<Mob> {
     @Override
     protected void doTick(final ServerLevel level, final Mob body) {
         Brain<?> brain = body.getBrain();
-        List<ItemEntity> items = level.getEntitiesOfClass(ItemEntity.class, body.getBoundingBox().inflate(32.0, 16.0, 32.0), item -> item.closerThan(body, MAX_DISTANCE_TO_WANTED_ITEM) && body.wantsToPickUp(level, item.getItem())); // Paper - Perf: Move predicate into getEntities
-        items.sort(Comparator.comparingDouble(body::distanceToSqr));
+        it.unimi.dsi.fastutil.objects.ObjectArrayList<ItemEntity> items = (it.unimi.dsi.fastutil.objects.ObjectArrayList<ItemEntity>) level.getEntitiesOfClass(ItemEntity.class, body.getBoundingBox().inflate(32.0, 16.0, 32.0), item -> item.closerThan(body, MAX_DISTANCE_TO_WANTED_ITEM) && body.wantsToPickUp(level, item.getItem())); // Paper - Perf: Move predicate into getEntities // Canvas - faster sorting
+        items.unstableSort(Comparator.comparingDouble(body::distanceToSqr)); // Canvas - faster sorting
         // Paper start - Perf: remove streams from hot code
         ItemEntity nearest = null;
         for (final ItemEntity item : items) {
