From 0ce24b7bccfa2d01cdd888671365a9ba0973fab1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/sensing/PlayerSensor.java b/net/minecraft/world/entity/ai/sensing/PlayerSensor.java
index 2fb0347495586a6d197066d46a8a1e831386d1bc..0e9349b2763f81f83fe1e09ea7e2ec86aa4810ce 100644
--- a/net/minecraft/world/entity/ai/sensing/PlayerSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/PlayerSensor.java
@@ -26,17 +26,33 @@ public class PlayerSensor extends Sensor<LivingEntity> {
 
     @Override
     protected void doTick(final ServerLevel level, final LivingEntity body) {
-        List<Player> players = level.getLocalPlayers() // Folia - region threading
-            .stream()
-            .filter(EntitySelector.NO_SPECTATORS)
-            .filter(player -> body.closerThan(player, this.getFollowDistance(body)))
-            .sorted(Comparator.comparingDouble(body::distanceToSqr))
-            .collect(Collectors.toList());
+        // Canvas start - remove streams
+        it.unimi.dsi.fastutil.objects.ObjectArrayList<Player> players = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
+        for (Player player : level.getLocalPlayers()) { // Folia - region threading
+            if (EntitySelector.NO_SPECTATORS.test(player) &&
+                body.closerThan(player, this.getFollowDistance(body))) {
+                players.add(player);
+            }
+        }
+        players.unstableSort(Comparator.comparingDouble(body::distanceToSqr));
+        // Canvas end - remove streams
         Brain<?> brain = body.getBrain();
         brain.setMemory(MemoryModuleType.NEAREST_PLAYERS, players);
-        List<Player> visiblePlayers = players.stream().filter(livingEntity -> isEntityTargetable(level, body, livingEntity)).collect(Collectors.toList());
-        brain.setMemory(MemoryModuleType.NEAREST_VISIBLE_PLAYER, visiblePlayers.isEmpty() ? null : visiblePlayers.get(0));
-        List<Player> visibleAttackablePlayers = visiblePlayers.stream().filter(livingEntity -> isEntityAttackable(level, body, livingEntity)).toList();
+        // Canvas start - remove streams
+        List<Player> visiblePlayers = new java.util.ArrayList<>();
+        for (Player player : players) {
+            if (isEntityTargetable(level, body, player)) {
+                visiblePlayers.add(player);
+            }
+        }
+        brain.setMemory(MemoryModuleType.NEAREST_VISIBLE_PLAYER, visiblePlayers.isEmpty() ? null : visiblePlayers.getFirst());
+        List<Player> visibleAttackablePlayers = new java.util.ArrayList<>();
+        for (Player livingEntity : visiblePlayers) {
+            if (isEntityAttackable(level, body, livingEntity)) {
+                visibleAttackablePlayers.add(livingEntity);
+            }
+        }
+        // Canvas end - remove streams
         brain.setMemory(MemoryModuleType.NEAREST_VISIBLE_ATTACKABLE_PLAYERS, visibleAttackablePlayers);
         brain.setMemory(MemoryModuleType.NEAREST_VISIBLE_ATTACKABLE_PLAYER, visibleAttackablePlayers.isEmpty() ? null : visibleAttackablePlayers.get(0));
     }
