--- a/net/minecraft/world/entity/ai/sensing/PlayerSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/PlayerSensor.java
@@ -22,7 +_,7 @@
 
     @Override
     protected void doTick(ServerLevel level, LivingEntity entity) {
-        List<Player> list = level.players()
+        List<Player> list = level.getLocalPlayers(entity.chunkPosition()) // Canvas - Threaded Regions
             .stream()
             .filter(EntitySelector.NO_SPECTATORS)
             .filter(serverPlayer -> entity.closerThan(serverPlayer, this.getFollowDistance(entity)))
