--- a/net/minecraft/world/entity/ai/sensing/SecondaryPoiSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/SecondaryPoiSensor.java
@@ -22,6 +_,13 @@
 
     @Override
     protected void doTick(ServerLevel level, Villager entity) {
+        // Canvas start - skip secondary poi sensor if empty
+        final ImmutableSet<net.minecraft.world.level.block.Block> secondaryPoi = entity.getVillagerData().profession().value().secondaryPoi();
+        if (secondaryPoi.isEmpty()) {
+            entity.getBrain().eraseMemory(MemoryModuleType.SECONDARY_JOB_SITE);
+            return;
+        }
+        // Canvas end - skip secondary poi sensor if empty
         ResourceKey<Level> resourceKey = level.dimension();
         BlockPos blockPos = entity.blockPosition();
         List<GlobalPos> list = Lists.newArrayList();
@@ -31,7 +_,7 @@
             for (int i2 = -2; i2 <= 2; i2++) {
                 for (int i3 = -4; i3 <= 4; i3++) {
                     BlockPos blockPos1 = blockPos.offset(i1, i2, i3);
-                    if (entity.getVillagerData().profession().value().secondaryPoi().contains(level.getBlockState(blockPos1).getBlock())) {
+                    if (secondaryPoi.contains(level.getBlockState(blockPos1).getBlock())) { // Canvas - skip secondary poi sensor if empty
                         list.add(GlobalPos.of(resourceKey, blockPos1));
                     }
                 }
