From cedb5bb62375ffca015c927898897c43ca743e2f Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/sensing/SecondaryPoiSensor.java b/net/minecraft/world/entity/ai/sensing/SecondaryPoiSensor.java
index ec12508c112b1e7748a5d197b73a0d540bed10fc..992b150bf33ebe1c5723da404896920ce295c38f 100644
--- a/net/minecraft/world/entity/ai/sensing/SecondaryPoiSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/SecondaryPoiSensor.java
@@ -22,6 +22,13 @@ public class SecondaryPoiSensor extends Sensor<Villager> {
 
     @Override
     protected void doTick(final ServerLevel level, final Villager body) {
+        // Canvas start - skip secondary poi sensor if empty
+        final ImmutableSet<net.minecraft.world.level.block.Block> secondaryPoi = body.getVillagerData().profession().value().secondaryPoi();
+        if (secondaryPoi.isEmpty()) {
+            body.getBrain().eraseMemory(MemoryModuleType.SECONDARY_JOB_SITE);
+            return;
+        }
+        // Canvas end - skip secondary poi sensor if empty
         ResourceKey<Level> dimensionType = level.dimension();
         BlockPos center = body.blockPosition();
         List<GlobalPos> jobSites = Lists.newArrayList();
@@ -31,7 +38,7 @@ public class SecondaryPoiSensor extends Sensor<Villager> {
             for (int y = -2; y <= 2; y++) {
                 for (int z = -4; z <= 4; z++) {
                     BlockPos testPos = center.offset(x, y, z);
-                    if (body.getVillagerData().profession().value().secondaryPoi().contains(level.getBlockState(testPos).getBlock())) {
+                    if (secondaryPoi.contains(level.getBlockState(testPos).getBlock())) { // Canvas - skip secondary poi sensor if empty
                         jobSites.add(GlobalPos.of(dimensionType, testPos));
                     }
                 }
