--- a/net/minecraft/world/entity/ai/sensing/SecondaryPoiSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/SecondaryPoiSensor.java
@@ -22,6 +_,13 @@
 
     @Override
     protected void doTick(final ServerLevel level, final Villager body) {
+        // Canvas start - skip secondary poi sensor if empty
+        final ImmutableSet<net.minecraft.world.level.block.Block> secondaryPoi = body.getVillagerData().profession().value().secondaryPoi();
+        if (secondaryPoi.isEmpty()) {
+            body.getBrain().eraseMemory(MemoryModuleType.SECONDARY_JOB_SITE);
+            return;
+        }
+        // Canvas end - skip secondary poi sensor if empty
         ResourceKey<Level> dimensionType = level.dimension();
         BlockPos center = body.blockPosition();
         List<GlobalPos> jobSites = Lists.newArrayList();
@@ -31,7 +_,7 @@
             for (int y = -2; y <= 2; y++) {
                 for (int z = -4; z <= 4; z++) {
                     BlockPos testPos = center.offset(x, y, z);
-                    if (body.getVillagerData().profession().value().secondaryPoi().contains(level.getBlockState(testPos).getBlock())) {
+                    if (secondaryPoi.contains(level.getBlockState(testPos).getBlock())) { // Canvas - skip secondary poi sensor if empty
                         jobSites.add(GlobalPos.of(dimensionType, testPos));
                     }
                 }
