--- a/net/minecraft/world/entity/ai/sensing/Sensor.java
+++ b/net/minecraft/world/entity/ai/sensing/Sensor.java
@@ -27,8 +_,8 @@
         .range(16.0)
         .ignoreLineOfSight()
         .ignoreInvisibilityTesting();
-    private final int scanRate;
-    private long timeToTick;
+    public final int scanRate; // Canvas - private -> public
+    public long timeToTick; // Canvas - private -> public
     private final String configKey; // Paper - configurable sensor tick rate and timings
 
     public Sensor(int scanRate) {
