--- a/net/minecraft/world/entity/ai/sensing/TemptingSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/TemptingSensor.java
@@ -36,7 +_,7 @@
     protected void doTick(ServerLevel level, PathfinderMob entity) {
         Brain<?> brain = entity.getBrain();
         TargetingConditions targetingConditions = TEMPT_TARGETING.copy().range((float)entity.getAttributeValue(Attributes.TEMPT_RANGE));
-        List<Player> list = level.players()
+        List<Player> list = level.getLocalPlayers(entity.chunkPosition()) // Canvas - Threaded Regions
             .stream()
             .filter(EntitySelector.NO_SPECTATORS)
             .filter(serverPlayer -> targetingConditions.test(level, entity, serverPlayer))
