From 0ce24b7bccfa2d01cdd888671365a9ba0973fab1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/sensing/TemptingSensor.java b/net/minecraft/world/entity/ai/sensing/TemptingSensor.java
index f5a1d35d41167d794409b4a585a7dc0f16ff8d96..5926579703127b0a5c2b91c7886877209add285d 100644
--- a/net/minecraft/world/entity/ai/sensing/TemptingSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/TemptingSensor.java
@@ -38,15 +38,20 @@ public class TemptingSensor extends Sensor<PathfinderMob> {
     protected void doTick(final ServerLevel level, final PathfinderMob body) {
         Brain<?> brain = body.getBrain();
         TargetingConditions targeting = TEMPT_TARGETING.copy().range((float)body.getAttributeValue(Attributes.TEMPT_RANGE));
-        List<Player> players = level.getLocalPlayers() // Folia - region threading
-            .stream()
-            .filter(EntitySelector.NO_SPECTATORS)
-            .filter(playerx -> targeting.test(level, body, playerx))
-            .filter(p -> this.playerHoldingTemptation(body, p))
-            .filter(playerx -> !body.hasPassenger(playerx))
-            .sorted(Comparator.comparingDouble(body::distanceToSqr))
-            .collect(Collectors.toList());
+        // Canvas start - remove streams
+        it.unimi.dsi.fastutil.objects.ObjectArrayList<net.minecraft.world.entity.player.Player> players = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
+        for (Player playerx : level.getLocalPlayers()) { // Folia - region threading
+            if (EntitySelector.NO_SPECTATORS.test(playerx) &&
+                targeting.test(level, body, playerx) &&
+                this.playerHoldingTemptation(body, playerx) &&
+                !body.hasPassenger(playerx)
+            ) {
+                players.add(playerx);
+            }
+        }
         if (!players.isEmpty()) {
+            players.unstableSort(Comparator.comparingDouble(body::distanceToSqr));
+        // Canvas end - remove streams
             Player player = players.get(0);
             // CraftBukkit start
             org.bukkit.event.entity.EntityTargetLivingEntityEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityTargetLivingEvent(
