From 4b22e4fb667ff3bf3c489ff7a94018cc27157519 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/sensing/VillagerBabiesSensor.java b/net/minecraft/world/entity/ai/sensing/VillagerBabiesSensor.java
index e3b2089e74d57ae8a15e9e5d29f43054ec6ec295..1727c54c89a0aa28483c1ea407ca0dd8f3e69fd2 100644
--- a/net/minecraft/world/entity/ai/sensing/VillagerBabiesSensor.java
+++ b/net/minecraft/world/entity/ai/sensing/VillagerBabiesSensor.java
@@ -22,7 +22,21 @@ public class VillagerBabiesSensor extends Sensor<LivingEntity> {
     }
 
     private List<LivingEntity> getNearestVillagerBabies(final LivingEntity myBody) {
-        return ImmutableList.copyOf(this.getVisibleEntities(myBody).findAll(this::isVillagerBaby));
+        // Leaf start - optimize baby villager sensor
+        NearestVisibleLivingEntities visibleEntities = this.getVisibleEntities(myBody);
+        ImmutableList.Builder<LivingEntity> babies = ImmutableList.builder();
+
+        // Inline and use single loop - copy from NearestVisibleLivingEntities#findAll and isVillagerBaby
+        for (LivingEntity target : visibleEntities.nearbyEntities) {
+            if (target.getType() == EntityType.VILLAGER
+                && target.isBaby()
+                && visibleEntities.lineOfSightTest.test(target)) {
+                babies.add(target);
+            }
+        }
+
+        return babies.build();
+        // Leaf end - optimize baby villager sensor
     }
 
     private boolean isVillagerBaby(final LivingEntity entity) {
