From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/material/FluidState.java b/net/minecraft/world/level/material/FluidState.java
index 3e5ada4a84bc1da06102dfacb0d3d45a4fc6db8a..512364bb91a44a925b615471b69399bc01ff51f9 100644
--- a/net/minecraft/world/level/material/FluidState.java
+++ b/net/minecraft/world/level/material/FluidState.java
@@ -164,6 +164,12 @@ public final class FluidState extends StateHolder<Fluid, FluidState> implements
     public Stream<TagKey<Fluid>> getTags() {
         return this.owner.builtInRegistryHolder().tags();
     }
+    // Canvas start - optimize fluidOnEyes - remove stream
+
+    public TagKey<Fluid>[] canvas$getTagsArray() {
+        return this.owner.builtInRegistryHolder().canvas$tagsArray;
+    }
+    // Canvas end - optimize fluidOnEyes - remove stream
 
     public void entityInside(Level level, BlockPos pos, Entity entity, InsideBlockEffectApplier effectApplier) {
         this.getType().entityInside(level, pos, entity, effectApplier);
