From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/network/protocol/game/ServerboundChatPacket.java b/net/minecraft/network/protocol/game/ServerboundChatPacket.java
index 11d18c1c13e25c01c5e8f2e3363932d1fce16236..921723f5c2475eea8b507c0ffeeafeaabf100066 100644
--- a/net/minecraft/network/protocol/game/ServerboundChatPacket.java
+++ b/net/minecraft/network/protocol/game/ServerboundChatPacket.java
@@ -11,6 +11,15 @@ import org.jspecify.annotations.Nullable;
 
 public record ServerboundChatPacket(String message, Instant timeStamp, long salt, @Nullable MessageSignature signature, LastSeenMessages.Update lastSeenMessages)
     implements Packet<ServerGamePacketListener> {
+    // Canvas start - no chat reports
+
+    @Override
+    @Nullable
+    public MessageSignature signature() {
+        return io.canvasmc.canvas.Config.INSTANCE.enableNoChatReports ? null : signature;
+    }
+
+    // Canvas end - no chat reports
     public static final StreamCodec<FriendlyByteBuf, ServerboundChatPacket> STREAM_CODEC = Packet.codec(
         ServerboundChatPacket::write, ServerboundChatPacket::new
     );
