From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sat, 7 Oct 2023 13:40:54 -0700
Subject: [PATCH] Mixin API Access


diff --git a/build.gradle.kts b/build.gradle.kts
index dc322b5d5458163c5b1de2a74d3d2edb1a4dac76..74055189fe69f9b5269e7f8c08d8f61fe6286dc4 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -49,6 +49,7 @@ dependencies {
     apiAndDocs("net.kyori:adventure-text-serializer-legacy")
     apiAndDocs("net.kyori:adventure-text-serializer-plain")
     apiAndDocs("net.kyori:adventure-text-logger-slf4j")
+    apiAndDocs("org.spongepowered:mixin:0.8.5") // Canvas - add mixin api
     api("org.apache.logging.log4j:log4j-api:$log4jVersion")
     api("org.slf4j:slf4j-api:$slf4jVersion")
 
