From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <159172167+PurpleWolfMC@users.noreply.github.com>
Date: Mon, 29 Apr 2024 18:54:25 +0000
Subject: [PATCH] Add Spongepowered Mixin API


diff --git a/build.gradle.kts b/build.gradle.kts
index 892e78b1d2d29dc54def03fcb6d85a93ad56d84c..11c6aafd050fd0fef55f56329bd3d4a34956c62a 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -49,6 +49,7 @@ dependencies {
     apiAndDocs("net.kyori:adventure-text-serializer-legacy")
     apiAndDocs("net.kyori:adventure-text-serializer-plain")
     apiAndDocs("net.kyori:adventure-text-logger-slf4j")
+    apiAndDocs("org.spongepowered:mixin:0.8.5") // Canvas - Add Spongepowered Mixin API
     api("org.apache.logging.log4j:log4j-api:$log4jVersion")
     api("org.slf4j:slf4j-api:$slf4jVersion")
 
