From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Tue, 5 Nov 2024 18:28:11 -0800
Subject: [PATCH] Server Properties API


diff --git a/src/main/java/io/github/dueris/canvas/properties/ServerProperties.java b/src/main/java/io/github/dueris/canvas/properties/ServerProperties.java
new file mode 100644
index 0000000000000000000000000000000000000000..a143f59c20a4bfa523e8a9dfdb8fc5511007bccb
--- /dev/null
+++ b/src/main/java/io/github/dueris/canvas/properties/ServerProperties.java
@@ -0,0 +1,14 @@
+package io.github.dueris.canvas.properties;
+
+import java.util.*;
+import java.io.*;
+
+public interface ServerProperties {
+    public ArrayList<String> getPropertyKeys();
+
+    public HashMap<String, Object> getPropertyMappings();
+
+    public File getFile();
+
+    public Properties getRawProperties();
+}
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index eb29794f6ca2efc9cde4dd1685822f9a3a73f3b9..4be483e33e54790a39cb07b72e5da93974030272 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -180,6 +180,15 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      * @return the port number of this server
      */
     public int getPort();
+    // Canvas start
+
+    /**
+     * Retrieves the server.properties file for reading
+     *
+     * @return ServerProperties for the server, aka the server.properties file
+     */
+    public io.github.dueris.canvas.properties.ServerProperties getServerProperties();
+    // Canvas end
 
     /**
      * Get the view distance from this server.
