From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Wed, 27 Nov 2024 00:39:00 -0800
Subject: [PATCH] Multithreaded Dimension Ticking


diff --git a/src/main/java/io/github/dueris/canvas/ThreadedContextStage.java b/src/main/java/io/github/dueris/canvas/ThreadedContextStage.java
new file mode 100644
index 0000000000000000000000000000000000000000..6d5f472b8d39a256d04690b0a52c0698ce93b7ed
--- /dev/null
+++ b/src/main/java/io/github/dueris/canvas/ThreadedContextStage.java
@@ -0,0 +1,5 @@
+package io.github.dueris.canvas;
+
+public enum ThreadedContextStage {
+    NOT_STARTED_YET, PREPARING, STARTED, DISABLING;
+}
diff --git a/src/main/java/io/github/dueris/canvas/ThreadedMinecraftServer.java b/src/main/java/io/github/dueris/canvas/ThreadedMinecraftServer.java
new file mode 100644
index 0000000000000000000000000000000000000000..1caaf8c77d5ff2a58f0ab1a193619faa2b9ef0b2
--- /dev/null
+++ b/src/main/java/io/github/dueris/canvas/ThreadedMinecraftServer.java
@@ -0,0 +1,12 @@
+package io.github.dueris.canvas;
+
+import org.bukkit.Server;
+import org.bukkit.World;
+import java.util.List;
+
+public interface ThreadedMinecraftServer {
+    List<World> getThreadedWorlds();
+    Server getBukkitServer();
+    boolean isBooted();
+    ThreadedContextStage getContextStage();
+}
diff --git a/src/main/java/io/github/dueris/canvas/event/ThreadedContextChangeEvent.java b/src/main/java/io/github/dueris/canvas/event/ThreadedContextChangeEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..04f67bff79c18c5629741b50a82662ee2829fa94
--- /dev/null
+++ b/src/main/java/io/github/dueris/canvas/event/ThreadedContextChangeEvent.java
@@ -0,0 +1,43 @@
+package io.github.dueris.canvas.event;
+
+import io.github.dueris.canvas.ThreadedContextStage;
+import io.github.dueris.canvas.ThreadedMinecraftServer;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.server.ServerEvent;
+import org.jetbrains.annotations.NotNull;
+
+public class ThreadedContextChangeEvent extends ServerEvent {
+    private static final HandlerList handlers = new HandlerList();
+    private final ThreadedContextStage oldStage;
+    private final ThreadedContextStage newStage;
+    private final ThreadedMinecraftServer threadedMinecraftServer;
+
+    public ThreadedContextChangeEvent(ThreadedContextStage oldStage, ThreadedContextStage newStage, ThreadedMinecraftServer threadedMinecraftServer) {
+        this.oldStage = oldStage;
+        this.newStage = newStage;
+        this.threadedMinecraftServer = threadedMinecraftServer;
+    }
+
+    @NotNull
+    @Override
+    public HandlerList getHandlers() {
+        return handlers;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return handlers;
+    }
+
+    public ThreadedContextStage getOldStage() {
+        return oldStage;
+    }
+
+    public ThreadedContextStage getNewStage() {
+        return newStage;
+    }
+
+    public ThreadedMinecraftServer getThreadedServer() {
+        return threadedMinecraftServer;
+    }
+}
diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index db453d04efb00baaeabb904a7bd1b99dd0a50735..28aac09b98d7536de927ebb12615ac371c9bf6fb 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -66,7 +66,7 @@ import org.jetbrains.annotations.Nullable;
  * Represents the Bukkit core, for version and Server singleton handling
  */
 public final class Bukkit {
-    private static Server server;
+    public static Server server; // Canvas - private -> public
 
     /**
      * Static class cannot be initialized.
diff --git a/src/main/java/org/bukkit/event/server/ServerEvent.java b/src/main/java/org/bukkit/event/server/ServerEvent.java
index 05167fb34e4c42edc67af6e6700a2a3cc0f92769..c59bd09f2d97238f0c3647390ccff8a01e400f39 100644
--- a/src/main/java/org/bukkit/event/server/ServerEvent.java
+++ b/src/main/java/org/bukkit/event/server/ServerEvent.java
@@ -9,7 +9,7 @@ import org.bukkit.event.Event;
 public abstract class ServerEvent extends Event {
 
     public ServerEvent() {
-        super(!Bukkit.isPrimaryThread()); // Paper
+        super(Bukkit.server != null && !Bukkit.isPrimaryThread()); // Paper // Canvas
     }
 
     public ServerEvent(boolean isAsync) {
