From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sat, 14 Oct 2023 14:16:38 -0700
Subject: [PATCH] Add Colored Logging


diff --git a/src/main/java/me/dueris/canvas/logger/CanvasLogger.java b/src/main/java/me/dueris/canvas/logger/CanvasLogger.java
index d01ae551c261fe8777eddab7c1c96fb0ee9b3190..e7c254815745b8a3e076b6d0a4fe1b4e5e9fcd54 100644
--- a/src/main/java/me/dueris/canvas/logger/CanvasLogger.java
+++ b/src/main/java/me/dueris/canvas/logger/CanvasLogger.java
@@ -6,6 +6,28 @@ public interface CanvasLogger {
         INFO, WARNING, ERROR
     }
 
+    public enum ConsoleColor {
+        WHITE("\u001B[37m"),
+        CYAN("\u001B[36m"),
+        RESET("\u001B[0m"),
+        PURPLE("\u001B[35m"),
+        BLUE("\u001B[34m"),
+        YELLOW("\u001B[33m"),
+        GREEN("\u001B[32m"),
+        RED("\u001B[31m"),
+        BLACK("\u001B[30m");
+
+        private final String stringValue;
+
+        ConsoleColor(String stringValue) {
+            this.stringValue = stringValue;
+        }
+
+        public String getStringValue() {
+            return stringValue;
+        }
+    }
+
     /**
      * Sends normal message to console
      * @param message
@@ -18,14 +40,14 @@ public interface CanvasLogger {
      * @param color
      * @param level
      */
-    public void logColored(String color, String level, String message);
+    public void logColored(ConsoleColor color, LogLevel level, String message);
 
     /**
      * Sends normal message to console with log level(INFO, SEVERE, WARNING) of your choice
      * @param message
      * @param level
      */
-    public void log(String level, String message);
+    public void log(LogLevel level, String message);
 
     /**
      * Sends error message to console
