From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Wed, 27 Nov 2024 00:39:00 -0800
Subject: [PATCH] Multithreaded Dimension Ticking


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index db453d04efb00baaeabb904a7bd1b99dd0a50735..28aac09b98d7536de927ebb12615ac371c9bf6fb 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -66,7 +66,7 @@ import org.jetbrains.annotations.Nullable;
  * Represents the Bukkit core, for version and Server singleton handling
  */
 public final class Bukkit {
-    private static Server server;
+    public static Server server; // Canvas - private -> public
 
     /**
      * Static class cannot be initialized.
diff --git a/src/main/java/org/bukkit/event/server/ServerEvent.java b/src/main/java/org/bukkit/event/server/ServerEvent.java
index 05167fb34e4c42edc67af6e6700a2a3cc0f92769..c59bd09f2d97238f0c3647390ccff8a01e400f39 100644
--- a/src/main/java/org/bukkit/event/server/ServerEvent.java
+++ b/src/main/java/org/bukkit/event/server/ServerEvent.java
@@ -9,7 +9,7 @@ import org.bukkit.event.Event;
 public abstract class ServerEvent extends Event {
 
     public ServerEvent() {
-        super(!Bukkit.isPrimaryThread()); // Paper
+        super(Bukkit.server != null && !Bukkit.isPrimaryThread()); // Paper // Canvas
     }
 
     public ServerEvent(boolean isAsync) {
