From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Mon, 29 Apr 2024 18:58:10 -0700
Subject: [PATCH] Dont Print Warnings to Console During Compile


diff --git a/build.gradle.kts b/build.gradle.kts
index b54b38b23eadeac5e6740c4dc04ee02f9d47eb0a..b730c49f05706c386ba0a947edaf25e46494c916 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -141,6 +141,13 @@ tasks.jar {
         )
     }
 }
+// Canvas start
+
+tasks.compileJava {
+    options.compilerArgs.add("-Xlint:-deprecation")
+    options.isWarnings = false
+}
+// Canvas end
 
 tasks.withType<Javadoc> {
     (options as StandardJavadocDocletOptions).addStringOption("-add-modules", "jdk.incubator.vector") // Purpur - our javadocs need this for pufferfish's SIMD patch
