From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Mon, 29 Apr 2024 18:58:10 -0700
Subject: [PATCH] Dont Print Warnings to Console During Compile


diff --git a/build.gradle.kts b/build.gradle.kts
index 11c6aafd050fd0fef55f56329bd3d4a34956c62a..46a2bcab38fda53d86ed16360a3d24c68fc20fd8 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -120,6 +120,13 @@ tasks.jar {
         )
     }
 }
+// Canvas start
+
+tasks.compileJava {
+    options.compilerArgs.add("-Xlint:-deprecation")
+    options.isWarnings = false
+}
+// Canvas end
 
 tasks.withType<Javadoc> {
     (options as StandardJavadocDocletOptions).addStringOption("-add-modules", "jdk.incubator.vector") // Purpur - our javadocs need this for pufferfish's SIMD patch
