From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Tue, 28 Nov 2023 00:46:28 +0000
Subject: [PATCH] Jar DependsOn SourcesJar


diff --git a/build.gradle.kts b/build.gradle.kts
index 46a2bcab38fda53d86ed16360a3d24c68fc20fd8..07566efcec0ccba8ca1c4133fac8f2494402959d 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -209,3 +209,4 @@ tasks.check {
     dependsOn(scanJarForOldGeneratedCode)
 }
 // Paper end
+tasks.getByName("jar").dependsOn("sourcesJar") // Canvas
