From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Tue, 28 Nov 2023 00:46:28 +0000
Subject: [PATCH] Jar DependsOn SourcesJar


diff --git a/build.gradle.kts b/build.gradle.kts
index 639ef49dcf4b2df9d2f807b50ea337cc50d2641a..4503be2b9f45cf04d2149a0352be2ac04af11f7b 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -241,3 +241,4 @@ tasks.check {
     dependsOn(scanJarForOldGeneratedCode)
 }
 // Paper end
+tasks.getByName("jar").dependsOn("sourcesJar") // Canvas
