From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sat, 21 Oct 2023 12:30:51 -0700
Subject: [PATCH] Add to iron golem api


diff --git a/src/main/java/org/bukkit/entity/IronGolem.java b/src/main/java/org/bukkit/entity/IronGolem.java
index 09fd716c8fc9ea34a1cbf87bcbe22df035422a51..8d02d982ea0b80caf891454dd7ccc857e26785df 100644
--- a/src/main/java/org/bukkit/entity/IronGolem.java
+++ b/src/main/java/org/bukkit/entity/IronGolem.java
@@ -1,5 +1,7 @@
 package org.bukkit.entity;
 
+import org.bukkit.Material;
+
 /**
  * An iron Golem that protects Villages.
  */
@@ -35,4 +37,34 @@ public interface IronGolem extends Golem {
      */
     void setSummoner(@org.jetbrains.annotations.Nullable java.util.UUID summoner);
     // Purpur end
+
+    // Canvas start
+    /**
+     * Returns the crackiness of the Iron Golem
+     * @return int value of Crackiness from the ordinal()
+     */
+    public int getCrackiness();
+
+    /**
+     * Sets the velocity that is applied to the entity when flung into the air by the Iron Golem
+     * Default is 0.4000000059604645D
+     * @param amt
+     */
+    public void setFlingVelocity(double amt);
+
+    /**
+     * Sets the velocity that is applied to the entity when flung into the air by the Iron Golem
+     * Default is 0.4000000059604645D
+     * @return velocity amplifier for the velocity applied to the entity upon being flung
+     */
+    public double getFlingVelocity();
+
+    public boolean canRepair();
+
+    public void setCanRepair(boolean can);
+
+    public void offerFlower(boolean lookingAtVillager);
+
+    public void setRepairItem(Material material);
+    // Canvas end
 }
