From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Tue, 28 Nov 2023 00:58:48 +0000
Subject: [PATCH] Publishing


diff --git a/.gitignore b/.gitignore
index 97e78e27ee0eea2c8b24886eeb19164d552323fe..78f6aa3384e944c6af857386d982ce27dfd89d2b 100644
--- a/.gitignore
+++ b/.gitignore
@@ -38,3 +38,4 @@
 # vs code
 /.vscode
 /.factorypath
+bin/
\ No newline at end of file
diff --git a/build.gradle.kts b/build.gradle.kts
index 07566efcec0ccba8ca1c4133fac8f2494402959d..53428de1f1e1524f0b711b471c43f654a00bcadd 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -1,7 +1,13 @@
+import com.github.jengelman.gradle.plugins.shadow.transformers.Log4j2PluginsCacheFileTransformer
+import com.github.jengelman.gradle.plugins.shadow.tasks.ShadowJar
+import io.papermc.paperweight.util.convertToPath
+import java.nio.file.Paths
+
 plugins {
     `java-library`
     `maven-publish`
     idea // Paper
+    id("io.github.goooler.shadow") // Canvas
 }
 
 java {
@@ -9,6 +15,19 @@ java {
     withJavadocJar()
 }
 
+// Canvas start
+tasks.named<ShadowJar>("shadowJar") {
+    mergeServiceFiles()
+
+    transform(Log4j2PluginsCacheFileTransformer())
+
+    from(tasks.getByName("jar"))
+
+    exclude("META-INF/versions/*/module-info.class")
+    exclude("module-info.class")
+}
+
+// Canvas end
 val annotationsVersion = "24.1.0"
 val bungeeCordChatVersion = "1.20-R0.2"
 val adventureVersion = "4.17.0-SNAPSHOT"
@@ -41,6 +60,7 @@ dependencies {
     api("com.googlecode.json-simple:json-simple:1.1.1") {
         isTransitive = false // includes junit
     }
+    implementation("org.spongepowered:mixin:0.8.5") // Canvas - Shadow Spongepowered Mixin API
     api("it.unimi.dsi:fastutil:8.5.6")
     apiAndDocs(platform("net.kyori:adventure-bom:$adventureVersion"))
     apiAndDocs("net.kyori:adventure-api")
@@ -49,7 +69,6 @@ dependencies {
     apiAndDocs("net.kyori:adventure-text-serializer-legacy")
     apiAndDocs("net.kyori:adventure-text-serializer-plain")
     apiAndDocs("net.kyori:adventure-text-logger-slf4j")
-    apiAndDocs("org.spongepowered:mixin:0.8.5") // Canvas - Add Spongepowered Mixin API
     api("org.apache.logging.log4j:log4j-api:$log4jVersion")
     api("org.slf4j:slf4j-api:$slf4jVersion")
 
@@ -95,9 +114,9 @@ sourceSets {
 // Paper end
 
 configure<PublishingExtension> {
-    publications.create<MavenPublication>("maven") {
-        from(components["java"])
-    }
+    // publications.create<MavenPublication>("maven") {
+    //     from(components["java"])
+    // }
 }
 
 val generateApiVersioningFile by tasks.registering {
@@ -208,5 +227,41 @@ val scanJarForOldGeneratedCode = tasks.register("scanJarForOldGeneratedCode", io
 tasks.check {
     dependsOn(scanJarForOldGeneratedCode)
 }
-// Paper end
-tasks.getByName("jar").dependsOn("sourcesJar") // Canvas
+// Paper end -/- Canvas start
+tasks.getByName("jar").dependsOn("sourcesJar")
+
+publishing {
+    publications {
+        create<MavenPublication>("mavenJava") {
+            pom {
+                name.set("Canvas-API")
+                description.set("Empowering the impossible in Minecraft")
+                url.set("https://github.me/dueris/Canvas")
+                licenses {
+                    license {
+                        name.set("GNU General Public License 3.0")
+                        url.set("https://www.gnu.org/licenses/gpl-3.0.en.html")
+                    }
+                }
+            }
+            groupId = "io.github.dueris"
+            artifactId = "canvas-api"
+            from(components["java"])
+        }
+    }
+    repositories {
+        maven {
+            name = "sonatype"
+            if (version.toString().endsWith("SNAPSHOT")) {
+                url = uri("https://s01.oss.sonatype.org/content/repositories/snapshots/")
+            } else {
+                url = uri("https://s01.oss.sonatype.org/service/local/staging/deploy/maven2/")
+            }
+            credentials {
+                username=System.getenv("OSSRH_USERNAME")
+                password=System.getenv("OSSRH_PASSWORD")
+            }
+        }
+    }
+}
+// Canvas end
diff --git a/gradle.properties b/gradle.properties
new file mode 100644
index 0000000000000000000000000000000000000000..bdae19149cb5e173a8538e3682a2efa6b19d6b08
--- /dev/null
+++ b/gradle.properties
@@ -0,0 +1,2 @@
+ossrhUsername=System.getenv("OSSRH_USERNAME")
+ossrhPassword=System.getenv("OSSRH_PASSWORD")
\ No newline at end of file
