From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Tue, 28 Nov 2023 00:46:28 +0000
Subject: [PATCH] Jar DependsOn SourcesJar


diff --git a/build.gradle.kts b/build.gradle.kts
index b730c49f05706c386ba0a947edaf25e46494c916..55bb1f34134a9e93e95694abdd169e4ee4cb0c4c 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -231,3 +231,4 @@ tasks.check {
     dependsOn(scanJarForOldGeneratedCode)
 }
 // Paper end
+tasks.getByName("jar").dependsOn("sourcesJar") // Canvas
