From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolf1909 <160778222+PurpleWolf1909@users.noreply.github.com>
Date: Tue, 30 Apr 2024 17:46:56 +0000
Subject: [PATCH] Expose Launch Args


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index a15d0ed710ff261f203d7e355e7d532f2e68abc9..a5c0b7834772a2d1aca5930814e57933e5bcd9e8 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -179,6 +179,15 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      * @return the port number of this server
      */
     public int getPort();
+    // Canvas start
+
+    /**
+     * Retrieves the array of startup arguments used to launch the server.
+     *
+     * @return The array of startup arguments passed to the server during its launch.
+     */
+    public String[] getStartupArgs();
+    // Canvas end
 
     /**
      * Get the view distance from this server.
