From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Tue, 28 Nov 2023 00:46:28 +0000
Subject: [PATCH] Jar DependsOn SourcesJar


diff --git a/build.gradle.kts b/build.gradle.kts
index cedefc5a346861a1474397dd845401df177ec48c..547165bc12eae068b86ef687b427a9562fd412a7 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -210,3 +210,4 @@ tasks.check {
     dependsOn(scanJarForOldGeneratedCode)
 }
 // Paper end
+tasks.getByName("jar").dependsOn("sourcesJar") // Canvas
