From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolf1909 <160778222+PurpleWolf1909@users.noreply.github.com>
Date: Tue, 30 Apr 2024 17:46:56 +0000
Subject: [PATCH] Expose Launch Args


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 1d594c256200d46882fafce5d2b33e60f0fa3a59..37e1df4bd144161d4b95709e0c2d0c8e32d0c2bc 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -180,6 +180,15 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      * @return the port number of this server
      */
     public int getPort();
+    // Canvas start
+
+    /**
+     * Retrieves the array of startup arguments used to launch the server.
+     *
+     * @return The array of startup arguments passed to the server during its launch.
+     */
+    public String[] getStartupArgs();
+    // Canvas end
 
     /**
      * Get the view distance from this server.
