From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Fri, 8 Dec 2023 03:43:26 +0000
Subject: [PATCH] RAM Details Exposure


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index f66fc28d1b49482f38a63cdb7fc8b53027d9707f..e51f1057c633ce0d62a958d7dca1511c51b2b5c2 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -125,6 +125,41 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull
     public CraftServerGui getServerGui();
 
+    /**
+     * Retrieves the total memory size.
+     *
+     * @return The total memory size in bytes, returned as a long.
+     */
+    public long getResourcesTotal();
+
+    /**
+     * Retrieves the amount of free memory.
+     *
+     * @return The amount of free memory in bytes, returned as a long.
+     */
+    public long getResourcesFree();
+
+    /**
+     * Retrieves the maximum memory size.
+     *
+     * @return The maximum memory size in bytes, returned as a long.
+     */
+    public long getResourcesMax();
+
+    /**
+     * Retrieves the amount of used memory (total - free).
+     *
+     * @return The amount of used memory in bytes, returned as a long.
+     */
+    public long getResourcesUsedMem();
+
+    /**
+     * Retrieves the used memory as a percentage of the maximum memory.
+     *
+     * @return The used memory percentage, where 0% represents no memory usage, and 100% represents full memory usage.
+     */
+    public int getResourcesUsedPercent();
+
     // Canvas end
 
     /**
