From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Tue, 24 Oct 2023 07:25:44 +0000
Subject: [PATCH] Added attribute listeners


diff --git a/src/main/java/org/bukkit/attribute/Attributable.java b/src/main/java/org/bukkit/attribute/Attributable.java
index 474ed1df364a5ca18661d0fbc29901760e39cb07..6fc55614705a1d2a9d18f56f41df325a6cfeb166 100644
--- a/src/main/java/org/bukkit/attribute/Attributable.java
+++ b/src/main/java/org/bukkit/attribute/Attributable.java
@@ -2,6 +2,9 @@ package org.bukkit.attribute;
 
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
+import org.bukkit.entity.*;
+import org.bukkit.attribute.*;
+import org.bukkit.plugin.*;
 
 /**
  * Represents an object which may contain attributes.
@@ -26,5 +29,16 @@ public interface Attributable {
      * @param attribute the generic attribute to register
      */
     void registerAttribute(@NotNull Attribute attribute);
-    // Paper end
-}
+    // Paper end -/- Canvas start
+    /**
+     * Registers an attribute listener based on the provided condition. When the condition is met, the attribute value is set to the
+     * trueValue; otherwise, it is set to the falseValue. The listener is associated with the specified attribute instance.
+     *
+     * @param condition    The condition to evaluate.
+     * @param falseValue   The value to set for the attribute when the condition is false.
+     * @param trueValue    The value to set for the attribute when the condition is true.
+     * @param attribute    The attribute instance to which the listener is associated.
+     */
+    void registerAttributeListener(boolean condition, int falseValue, int trueValue, Attribute attribute, Plugin plugin, int tickrate);
+    // Canvas end
+}
\ No newline at end of file
