From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Fri, 17 Nov 2023 22:28:04 +0000
Subject: [PATCH] More Block API


diff --git a/src/main/java/org/bukkit/DyeColor.java b/src/main/java/org/bukkit/DyeColor.java
index 47df858e095c4423c20e49e029d72f0f50d2c924..166f2cb522eddb53349a3f889b62d7f2817c074a 100644
--- a/src/main/java/org/bukkit/DyeColor.java
+++ b/src/main/java/org/bukkit/DyeColor.java
@@ -74,6 +74,47 @@ public enum DyeColor {
      * Represents black dye.
      */
     BLACK(0xF, 0x0, Color.fromRGB(0x1D1D21), Color.fromRGB(0x1E1B1B));
+    // Canvas start - add method to ensure its using the correct String for color because sometimes it didnt??
+
+    public String toDyeString(){
+        switch (this) {
+            case BLACK:
+                return "black";
+            case BLUE:
+                return "blue";
+            case BROWN:
+                return "brown";
+            case CYAN:
+                return "cyan";
+            case GRAY:
+                return "gray";
+            case GREEN:
+                return "green";
+            case LIGHT_BLUE:
+                return "light_blue";
+            case LIGHT_GRAY:
+                return "light_gray";
+            case LIME:
+                return "lime";
+            case MAGENTA:
+                return "magenta";
+            case ORANGE:
+                return "orange";
+            case PINK:
+                return "pink";
+            case PURPLE:
+                return "purple";
+            case RED:
+                return "red";
+            case WHITE:
+                return "white";
+            case YELLOW:
+                return "yellow";
+            default:
+                throw new IllegalArgumentException("Unknown DyeColor for " + this.toString());
+        }
+    }
+    // Canvas end
 
     private final byte woolData;
     private final byte dyeData;
diff --git a/src/main/java/org/bukkit/block/Campfire.java b/src/main/java/org/bukkit/block/Campfire.java
index 9c3952459ed216f727b3654b2ed536f17f320402..175a819eca9dfa9f204c911e3c59c046c9e466e0 100644
--- a/src/main/java/org/bukkit/block/Campfire.java
+++ b/src/main/java/org/bukkit/block/Campfire.java
@@ -104,5 +104,11 @@ public interface Campfire extends TileState {
      * @return {@code true} if the specified slot index cannot cook
      */
     boolean isCookingDisabled(int index);
-    // Paper end
+    // Paper end -/- Canvas start
+
+    /**
+     * Extinguishes the fire in the campfire
+     */
+    public void extinguish();
+    // Canvas end
 }
