From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Tue, 28 Nov 2023 00:46:28 +0000
Subject: [PATCH] Jar DependsOn SourcesJar


diff --git a/build.gradle.kts b/build.gradle.kts
index bcaebb026a0eccd448497665d8f0eee5ce841fe9..c2f045fbfa2d7189c1f08a2044c727231d304203 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -210,3 +210,4 @@ tasks.check {
     dependsOn(scanJarForOldGeneratedCode)
 }
 // Paper end
+tasks.getByName("jar").dependsOn("sourcesJar") // Canvas
