From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Tue, 28 Nov 2023 00:46:28 +0000
Subject: [PATCH] Jar DependsOn SourcesJar


diff --git a/build.gradle.kts b/build.gradle.kts
index 74055189fe69f9b5269e7f8c08d8f61fe6286dc4..594463f6844d743ba7312b8c103a3de739e11ef0 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -202,3 +202,4 @@ tasks.check {
     dependsOn(scanJarForOldGeneratedCode)
 }
 // Paper end
+tasks.getByName("jar").dependsOn("sourcesJar") // Canvas
