From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris Bauer <jedimastertoothless@hotmail.com>
Date: Tue, 24 Oct 2023 15:39:31 +0000
Subject: [PATCH] Make attribute listners use self instead of arg


diff --git a/src/main/java/org/bukkit/attribute/Attributable.java b/src/main/java/org/bukkit/attribute/Attributable.java
index 405f5255e69c43f8b1fa26653d9daa6bcc7a7cb4..6fc55614705a1d2a9d18f56f41df325a6cfeb166 100644
--- a/src/main/java/org/bukkit/attribute/Attributable.java
+++ b/src/main/java/org/bukkit/attribute/Attributable.java
@@ -39,6 +39,6 @@ public interface Attributable {
      * @param trueValue    The value to set for the attribute when the condition is true.
      * @param attribute    The attribute instance to which the listener is associated.
      */
-    void registerAttributeListener(boolean condition, int falseValue, int trueValue, Attribute attribute, Plugin plugin, int tickrate, Player player);
+    void registerAttributeListener(boolean condition, int falseValue, int trueValue, Attribute attribute, Plugin plugin, int tickrate);
     // Canvas end
 }
\ No newline at end of file
