From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Fri, 27 Oct 2023 07:18:33 +0000
Subject: [PATCH] Expose StartupArgs to Server


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index e51f1057c633ce0d62a958d7dca1511c51b2b5c2..a807356721d8ebb3b8c653c0e0ff99e31a64fd23 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -160,6 +160,12 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     public int getResourcesUsedPercent();
 
+    /**
+     * Retrieves the array of startup arguments used to launch the server.
+     *
+     * @return The array of startup arguments passed to the server during its launch.
+     */
+    public String[] getStartupArgs();
     // Canvas end
 
     /**
