From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Fri, 27 Oct 2023 18:00:03 +0000
Subject: [PATCH] Add ServerProperties API


diff --git a/src/main/java/me/dueris/canvas/properties/ServerProperties.java b/src/main/java/me/dueris/canvas/properties/ServerProperties.java
new file mode 100644
index 0000000000000000000000000000000000000000..7795874693df6f73f3d065ecbbe292f3aea90cd0
--- /dev/null
+++ b/src/main/java/me/dueris/canvas/properties/ServerProperties.java
@@ -0,0 +1,14 @@
+package me.dueris.canvas.properties;
+
+import java.util.*;
+import java.io.*;
+
+public interface ServerProperties {
+    public ArrayList<String> getPropertyKeys();
+
+    public HashMap<String, Object> getPropertyMappings();
+
+    public File getFile();
+
+    public Properties getRawProperties();
+}
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 1827468fab11a26b70849a0789aa903aa82cf0b1..de99c8b43be6e444eafea2c043d1954b25823eb9 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -165,6 +165,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      * @return The array of startup arguments passed to the server during its launch.
      */
     public String[] getStartupArgs();
+
+    /**
+     * Retrieves the server.properties file for reading
+     * 
+     * @return ServerProperties for the server, aka the server.properties file
+     */
+    public me.dueris.canvas.properties.ServerProperties getServerProperties();
     // Canvas end
 
     /**
