From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Tue, 24 Oct 2023 18:17:03 +0000
Subject: [PATCH] Added methods to Server to get RAM details


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 4c6435b3e7084aa00d912e46debc70585792e87b..985981277ee876b7ebd8ff1606eb79d515cbe6d4 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -124,8 +124,46 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull
     public CraftServerGui getServerGui();
 
+    /**
+     * @return ChatChannel manager
+     */
     @NotNull
     public ChatChannelContainer getChatChannelManager();
+    
+    /**
+     * Retrieves the total memory size.
+     *
+     * @return The total memory size in bytes, returned as a long.
+     */
+    public long getResourcesTotal();
+
+    /**
+     * Retrieves the amount of free memory.
+     *
+     * @return The amount of free memory in bytes, returned as a long.
+     */
+    public long getResourcesFree();
+
+    /**
+     * Retrieves the maximum memory size.
+     *
+     * @return The maximum memory size in bytes, returned as a long.
+     */
+    public long getResourcesMax();
+
+    /**
+     * Retrieves the amount of used memory (total - free).
+     *
+     * @return The amount of used memory in bytes, returned as a long.
+     */
+    public long getResourcesUsedMem();
+
+    /**
+     * Retrieves the used memory as a percentage of the maximum memory.
+     *
+     * @return The used memory percentage, where 0% represents no memory usage, and 100% represents full memory usage.
+     */
+    public int getResourcesUsedPercent();
 
     // Canvas end
 
