From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris Bauer <jedimastertoothless@hotmail.com>
Date: Fri, 27 Oct 2023 07:18:33 +0000
Subject: [PATCH] Add getStartupArgs to Server


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 985981277ee876b7ebd8ff1606eb79d515cbe6d4..1d5ce788678698eb2b81476eadeecb3372c7769c 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -165,6 +165,12 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     public int getResourcesUsedPercent();
 
+    /**
+     * Retrieves the array of startup arguments used to launch the server.
+     *
+     * @return The array of startup arguments passed to the server during its launch.
+     */
+    public String[] getStartupArgs();
     // Canvas end
 
     /**
