From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <122416109+Dueris@users.noreply.github.com>
Date: Sun, 14 Apr 2024 17:33:54 +0000
Subject: [PATCH] Enhance Projectile API


diff --git a/src/main/java/org/bukkit/entity/AbstractArrow.java b/src/main/java/org/bukkit/entity/AbstractArrow.java
index b36298679d6d52d09fe4bb8e52e19e18f6df742a..f4ef545bc11163f0e3259af90a8637d06327125b 100644
--- a/src/main/java/org/bukkit/entity/AbstractArrow.java
+++ b/src/main/java/org/bukkit/entity/AbstractArrow.java
@@ -238,4 +238,13 @@ public interface AbstractArrow extends Projectile {
      */
     void setHitSound(@NotNull org.bukkit.Sound sound);
     // Paper end - more projectile API
+    // Canvas start - enhance arrow api
+
+    /**
+     * Returns if the arrow is an instance of a Spectral arrow
+     *
+     * @return if arrow is a spectral arrow
+     */
+    public boolean isSpectral();
+    // Canvas end
 }
