From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <122416109+Dueris@users.noreply.github.com>
Date: Mon, 1 Apr 2024 16:54:11 +0000
Subject: [PATCH] Enhance Projectile API


diff --git a/src/main/java/org/bukkit/entity/AbstractArrow.java b/src/main/java/org/bukkit/entity/AbstractArrow.java
index 2d7521b71ec16e1b229bbe4c05d51673720df475..d54eedb60fe7095d282edbae1db755f4b8a1ed3a 100644
--- a/src/main/java/org/bukkit/entity/AbstractArrow.java
+++ b/src/main/java/org/bukkit/entity/AbstractArrow.java
@@ -213,4 +213,13 @@ public interface AbstractArrow extends Projectile {
      */
     void setHitSound(@NotNull org.bukkit.Sound sound);
     // Paper end
+    // Canvas start - enhance arrow api
+
+    /**
+     * Returns if the arrow is an instance of a Spectral arrow
+     *
+     * @return if arrow is a spectral arrow
+     */
+    public boolean isSpectral();
+    // Canvas end
 }
