From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sat, 28 Oct 2023 12:41:44 -0700
Subject: [PATCH] Change package from com.dueris to me.dueris


diff --git a/src/main/java/com/dueris/canvas/CraftServerGui.java b/src/main/java/me/dueris/canvas/CraftServerGui.java
similarity index 98%
rename from src/main/java/com/dueris/canvas/CraftServerGui.java
rename to src/main/java/me/dueris/canvas/CraftServerGui.java
index e2b37949653ee1bcee2beec3e450ca72bb10d33f..6796780e0d00e43a66a670517ecf84a5d606295e 100644
--- a/src/main/java/com/dueris/canvas/CraftServerGui.java
+++ b/src/main/java/me/dueris/canvas/CraftServerGui.java
@@ -1,4 +1,4 @@
-package com.dueris.canvas;
+package me.dueris.canvas;
 
 import javax.swing.*;
 import java.awt.*;
diff --git a/src/main/java/com/dueris/canvas/MojangServerGui.java b/src/main/java/me/dueris/canvas/MojangServerGui.java
similarity index 99%
rename from src/main/java/com/dueris/canvas/MojangServerGui.java
rename to src/main/java/me/dueris/canvas/MojangServerGui.java
index 9e3aabf1df8ce15b4f37e24b0b28225e30b654e0..223271f4acff947c13219b85c7da408a774eaf4c 100644
--- a/src/main/java/com/dueris/canvas/MojangServerGui.java
+++ b/src/main/java/me/dueris/canvas/MojangServerGui.java
@@ -1,4 +1,4 @@
-package com.dueris.canvas;
+package me.dueris.canvas;
 
 import org.bukkit.Server;
 import org.bukkit.scheduler.BukkitRunnable;
diff --git a/src/main/java/com/dueris/canvas/logger/CanvasLogger.java b/src/main/java/me/dueris/canvas/logger/CanvasLogger.java
similarity index 97%
rename from src/main/java/com/dueris/canvas/logger/CanvasLogger.java
rename to src/main/java/me/dueris/canvas/logger/CanvasLogger.java
index 560262e5c52067154044cfebb55fe91a3532a69d..e7c254815745b8a3e076b6d0a4fe1b4e5e9fcd54 100644
--- a/src/main/java/com/dueris/canvas/logger/CanvasLogger.java
+++ b/src/main/java/me/dueris/canvas/logger/CanvasLogger.java
@@ -1,4 +1,4 @@
-package com.dueris.canvas.logger;
+package me.dueris.canvas.logger;
 
 public interface CanvasLogger {
 
diff --git a/src/main/java/com/dueris/canvas/logger/Obfuscator.java b/src/main/java/me/dueris/canvas/logger/Obfuscator.java
similarity index 98%
rename from src/main/java/com/dueris/canvas/logger/Obfuscator.java
rename to src/main/java/me/dueris/canvas/logger/Obfuscator.java
index a8511d43667e7022e6b7ce5d2fb7f41f26764c6d..58f147a4be3012038820abf72a5aea0f0e775750 100644
--- a/src/main/java/com/dueris/canvas/logger/Obfuscator.java
+++ b/src/main/java/me/dueris/canvas/logger/Obfuscator.java
@@ -1,4 +1,4 @@
-package com.dueris.canvas.logger;
+package me.dueris.canvas.logger;
 
 import java.util.HashMap;
 import java.util.Map;
diff --git a/src/main/java/com/dueris/canvas/network/ChatChannelContainer.java b/src/main/java/me/dueris/canvas/network/ChatChannelContainer.java
similarity index 82%
rename from src/main/java/com/dueris/canvas/network/ChatChannelContainer.java
rename to src/main/java/me/dueris/canvas/network/ChatChannelContainer.java
index a11a624bd9f5c72b8e60f61a6900b742ce23be70..669151590c90398f40122b90907367e21e9d5e44 100644
--- a/src/main/java/com/dueris/canvas/network/ChatChannelContainer.java
+++ b/src/main/java/me/dueris/canvas/network/ChatChannelContainer.java
@@ -1,6 +1,6 @@
-package com.dueris.canvas.network;
+package me.dueris.canvas.network;
 
-import com.dueris.canvas.network.channels.ChatChannel;
+import me.dueris.canvas.network.channels.ChatChannel;
 import org.bukkit.NamespacedKey;
 import java.util.ArrayList;
 
diff --git a/src/main/java/com/dueris/canvas/network/channels/ChannelSettings.java b/src/main/java/me/dueris/canvas/network/channels/ChannelSettings.java
similarity index 95%
rename from src/main/java/com/dueris/canvas/network/channels/ChannelSettings.java
rename to src/main/java/me/dueris/canvas/network/channels/ChannelSettings.java
index fedc847f4db84308f9213463e90ef7c75c534f64..f1b49275bf8e8e8780217a94412adffa0ce01f68 100644
--- a/src/main/java/com/dueris/canvas/network/channels/ChannelSettings.java
+++ b/src/main/java/me/dueris/canvas/network/channels/ChannelSettings.java
@@ -1,4 +1,4 @@
-package com.dueris.canvas.network.channels;
+package me.dueris.canvas.network.channels;
 
 import org.bukkit.scheduler.BukkitRunnable;
 import org.bukkit.scheduler.BukkitTask;
diff --git a/src/main/java/com/dueris/canvas/network/channels/ChatChannel.java b/src/main/java/me/dueris/canvas/network/channels/ChatChannel.java
similarity index 94%
rename from src/main/java/com/dueris/canvas/network/channels/ChatChannel.java
rename to src/main/java/me/dueris/canvas/network/channels/ChatChannel.java
index 94681824ccb5b9ca80528fc5df3057a47fc9888c..c092a54f18dd725fbb6b137ed26fbf13979c2e81 100644
--- a/src/main/java/com/dueris/canvas/network/channels/ChatChannel.java
+++ b/src/main/java/me/dueris/canvas/network/channels/ChatChannel.java
@@ -1,4 +1,4 @@
-package com.dueris.canvas.network.channels;
+package me.dueris.canvas.network.channels;
 
 import org.bukkit.NamespacedKey;
 import org.bukkit.entity.Player;
diff --git a/src/main/java/com/dueris/canvas/placeholder b/src/main/java/me/dueris/canvas/placeholder
similarity index 100%
rename from src/main/java/com/dueris/canvas/placeholder
rename to src/main/java/me/dueris/canvas/placeholder
diff --git a/src/main/java/com/dueris/canvas/properties/ServerProperties.java b/src/main/java/me/dueris/canvas/properties/ServerProperties.java
similarity index 87%
rename from src/main/java/com/dueris/canvas/properties/ServerProperties.java
rename to src/main/java/me/dueris/canvas/properties/ServerProperties.java
index 7f4e03b4d04354c6be4c5cf1c6e30a37fcb43a15..7795874693df6f73f3d065ecbbe292f3aea90cd0 100644
--- a/src/main/java/com/dueris/canvas/properties/ServerProperties.java
+++ b/src/main/java/me/dueris/canvas/properties/ServerProperties.java
@@ -1,4 +1,4 @@
-package com.dueris.canvas.properties;
+package me.dueris.canvas.properties;
 
 import java.util.*;
 import java.io.*;
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 480bbc529645ad21304e143c460d1baa72bf0815..f21fea18803ceeb3ad012df56c3dfcd52aa161e1 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1,9 +1,9 @@
 package org.bukkit;
 
-import com.dueris.canvas.CraftServerGui;
-import com.dueris.canvas.logger.CanvasLogger;
-import com.dueris.canvas.logger.Obfuscator;
-import com.dueris.canvas.network.ChatChannelContainer;
+import me.dueris.canvas.CraftServerGui;
+import me.dueris.canvas.logger.CanvasLogger;
+import me.dueris.canvas.logger.Obfuscator;
+import me.dueris.canvas.network.ChatChannelContainer;
 import com.google.common.collect.ImmutableList;
 import java.awt.image.BufferedImage;
 import java.io.File;
@@ -18,6 +18,8 @@ import java.util.Set;
 import java.util.UUID;
 import java.util.function.Consumer;
 import java.util.logging.Logger;
+
+import me.dueris.canvas.properties.ServerProperties;
 import org.bukkit.Warning.WarningState;
 import org.bukkit.advancement.Advancement;
 import org.bukkit.block.data.BlockData;
@@ -178,7 +180,7 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      * 
      * @return ServerProperties for the server, aka the server.properties file
      */
-    public com.dueris.canvas.properties.ServerProperties getServerProperties();
+    public ServerProperties getServerProperties();
     // Canvas end
 
     /**
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index f64142b4b84a03b32c20647e3ae3a1b480a48109..0d36b13df0026afbe77d790ece75926ff4b581f1 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -7,7 +7,7 @@ import java.time.Instant;
 import java.util.Collection;
 import java.util.Date;
 import java.util.Map;
-import com.dueris.canvas.network.channels.ChatChannel;
+import me.dueris.canvas.network.channels.ChatChannel;
 import org.bukkit.BanEntry;
 import org.bukkit.Chunk;
 import org.bukkit.DyeColor;
@@ -44,7 +44,6 @@ import org.bukkit.inventory.ItemStack;
 import org.bukkit.map.MapView;
 import org.bukkit.plugin.Plugin;
 import org.bukkit.plugin.messaging.PluginMessageRecipient;
-import org.bukkit.profile.PlayerProfile;
 import org.bukkit.scoreboard.Scoreboard;
 import org.jetbrains.annotations.ApiStatus;
 import org.jetbrains.annotations.NotNull;
diff --git a/src/main/java/org/bukkit/event/player/AsyncPlayerChatEvent.java b/src/main/java/org/bukkit/event/player/AsyncPlayerChatEvent.java
index f4d3d9c1e470a24342bf1c932a68582a5c1e983e..96c36262578e4f596d0f56fd9d4da9f1b32d30cf 100644
--- a/src/main/java/org/bukkit/event/player/AsyncPlayerChatEvent.java
+++ b/src/main/java/org/bukkit/event/player/AsyncPlayerChatEvent.java
@@ -2,9 +2,6 @@ package org.bukkit.event.player;
 
 import java.util.IllegalFormatException;
 import java.util.Set;
-import com.dueris.canvas.network.ChatChannelContainer;
-import com.dueris.canvas.network.channels.ChatChannel;
-import org.bukkit.Bukkit;
 import org.bukkit.entity.Player;
 import org.bukkit.event.Cancellable;
 import org.bukkit.event.HandlerList;
