From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sat, 28 Oct 2023 12:41:44 -0700
Subject: [PATCH] Change package from com.dueris to me.dueris


diff --git a/src/main/java/com/dueris/canvas/CraftServerGui.java b/src/main/java/io/github/dueris/CraftServerGui.java
similarity index 98%
rename from src/main/java/com/dueris/canvas/CraftServerGui.java
rename to src/main/java/io/github/dueris/CraftServerGui.java
index e2b37949653ee1bcee2beec3e450ca72bb10d33f..9b16e050f4db56ee0a8399f0cae095b65cbcf958 100644
--- a/src/main/java/com/dueris/canvas/CraftServerGui.java
+++ b/src/main/java/io/github/dueris/CraftServerGui.java
@@ -1,4 +1,4 @@
-package com.dueris.canvas;
+package io.github.dueris;
 
 import javax.swing.*;
 import java.awt.*;
diff --git a/src/main/java/com/dueris/canvas/MojangServerGui.java b/src/main/java/io/github/dueris/MojangServerGui.java
similarity index 99%
rename from src/main/java/com/dueris/canvas/MojangServerGui.java
rename to src/main/java/io/github/dueris/MojangServerGui.java
index 9e3aabf1df8ce15b4f37e24b0b28225e30b654e0..09e298674dcb967afd32a318e3b89da90fb99665 100644
--- a/src/main/java/com/dueris/canvas/MojangServerGui.java
+++ b/src/main/java/io/github/dueris/MojangServerGui.java
@@ -1,4 +1,4 @@
-package com.dueris.canvas;
+package io.github.dueris;
 
 import org.bukkit.Server;
 import org.bukkit.scheduler.BukkitRunnable;
diff --git a/src/main/java/com/dueris/canvas/logger/CanvasLogger.java b/src/main/java/io/github/dueris/logger/CanvasLogger.java
similarity index 97%
rename from src/main/java/com/dueris/canvas/logger/CanvasLogger.java
rename to src/main/java/io/github/dueris/logger/CanvasLogger.java
index 560262e5c52067154044cfebb55fe91a3532a69d..e0aaa53eaa640baee026c88b5d9a0a2220baae6c 100644
--- a/src/main/java/com/dueris/canvas/logger/CanvasLogger.java
+++ b/src/main/java/io/github/dueris/logger/CanvasLogger.java
@@ -1,4 +1,4 @@
-package com.dueris.canvas.logger;
+package io.github.dueris.logger;
 
 public interface CanvasLogger {
 
diff --git a/src/main/java/com/dueris/canvas/logger/Obfuscator.java b/src/main/java/io/github/dueris/logger/Obfuscator.java
similarity index 98%
rename from src/main/java/com/dueris/canvas/logger/Obfuscator.java
rename to src/main/java/io/github/dueris/logger/Obfuscator.java
index a8511d43667e7022e6b7ce5d2fb7f41f26764c6d..e81987c0e367f9391e096fd12a93ef5e6c00cb02 100644
--- a/src/main/java/com/dueris/canvas/logger/Obfuscator.java
+++ b/src/main/java/io/github/dueris/logger/Obfuscator.java
@@ -1,4 +1,4 @@
-package com.dueris.canvas.logger;
+package io.github.dueris.logger;
 
 import java.util.HashMap;
 import java.util.Map;
diff --git a/src/main/java/com/dueris/canvas/network/ChatChannelContainer.java b/src/main/java/io/github/dueris/network/ChatChannelContainer.java
similarity index 82%
rename from src/main/java/com/dueris/canvas/network/ChatChannelContainer.java
rename to src/main/java/io/github/dueris/network/ChatChannelContainer.java
index a11a624bd9f5c72b8e60f61a6900b742ce23be70..a15cd692022b94ceef10538aa944dad096b0ac33 100644
--- a/src/main/java/com/dueris/canvas/network/ChatChannelContainer.java
+++ b/src/main/java/io/github/dueris/network/ChatChannelContainer.java
@@ -1,6 +1,6 @@
-package com.dueris.canvas.network;
+package io.github.dueris.network;
 
-import com.dueris.canvas.network.channels.ChatChannel;
+import io.github.dueris.network.channels.ChatChannel;
 import org.bukkit.NamespacedKey;
 import java.util.ArrayList;
 
diff --git a/src/main/java/com/dueris/canvas/network/channels/ChannelSettings.java b/src/main/java/io/github/dueris/network/channels/ChannelSettings.java
similarity index 95%
rename from src/main/java/com/dueris/canvas/network/channels/ChannelSettings.java
rename to src/main/java/io/github/dueris/network/channels/ChannelSettings.java
index fedc847f4db84308f9213463e90ef7c75c534f64..5beee5894b60b8064ec5dd94093ce86cda072698 100644
--- a/src/main/java/com/dueris/canvas/network/channels/ChannelSettings.java
+++ b/src/main/java/io/github/dueris/network/channels/ChannelSettings.java
@@ -1,4 +1,4 @@
-package com.dueris.canvas.network.channels;
+package io.github.dueris.network.channels;
 
 import org.bukkit.scheduler.BukkitRunnable;
 import org.bukkit.scheduler.BukkitTask;
diff --git a/src/main/java/com/dueris/canvas/network/channels/ChatChannel.java b/src/main/java/io/github/dueris/network/channels/ChatChannel.java
similarity index 94%
rename from src/main/java/com/dueris/canvas/network/channels/ChatChannel.java
rename to src/main/java/io/github/dueris/network/channels/ChatChannel.java
index 94681824ccb5b9ca80528fc5df3057a47fc9888c..fcc8f33f6e866c583da8d0365d156c1bc94854c8 100644
--- a/src/main/java/com/dueris/canvas/network/channels/ChatChannel.java
+++ b/src/main/java/io/github/dueris/network/channels/ChatChannel.java
@@ -1,4 +1,4 @@
-package com.dueris.canvas.network.channels;
+package io.github.dueris.network.channels;
 
 import org.bukkit.NamespacedKey;
 import org.bukkit.entity.Player;
diff --git a/src/main/java/com/dueris/canvas/placeholder b/src/main/java/io/github/dueris/placeholder
similarity index 100%
rename from src/main/java/com/dueris/canvas/placeholder
rename to src/main/java/io/github/dueris/placeholder
diff --git a/src/main/java/com/dueris/canvas/properties/ServerProperties.java b/src/main/java/io/github/dueris/properties/ServerProperties.java
similarity index 87%
rename from src/main/java/com/dueris/canvas/properties/ServerProperties.java
rename to src/main/java/io/github/dueris/properties/ServerProperties.java
index 7f4e03b4d04354c6be4c5cf1c6e30a37fcb43a15..a184066978b7b0755c07c79a3f5c73525180df3c 100644
--- a/src/main/java/com/dueris/canvas/properties/ServerProperties.java
+++ b/src/main/java/io/github/dueris/properties/ServerProperties.java
@@ -1,4 +1,4 @@
-package com.dueris.canvas.properties;
+package io.github.dueris.properties;
 
 import java.util.*;
 import java.io.*;
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 480bbc529645ad21304e143c460d1baa72bf0815..dd2f072f1e6c07d7049169b4f82bd89a7f42d05f 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1,9 +1,9 @@
 package org.bukkit;
 
-import com.dueris.canvas.CraftServerGui;
-import com.dueris.canvas.logger.CanvasLogger;
-import com.dueris.canvas.logger.Obfuscator;
-import com.dueris.canvas.network.ChatChannelContainer;
+import io.github.dueris.CraftServerGui;
+import io.github.dueris.logger.CanvasLogger;
+import io.github.dueris.logger.Obfuscator;
+import io.github.dueris.network.ChatChannelContainer;
 import com.google.common.collect.ImmutableList;
 import java.awt.image.BufferedImage;
 import java.io.File;
@@ -18,6 +18,8 @@ import java.util.Set;
 import java.util.UUID;
 import java.util.function.Consumer;
 import java.util.logging.Logger;
+
+import io.github.dueris.properties.ServerProperties;
 import org.bukkit.Warning.WarningState;
 import org.bukkit.advancement.Advancement;
 import org.bukkit.block.data.BlockData;
@@ -178,7 +180,7 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      * 
      * @return ServerProperties for the server, aka the server.properties file
      */
-    public com.dueris.canvas.properties.ServerProperties getServerProperties();
+    public ServerProperties getServerProperties();
     // Canvas end
 
     /**
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index f64142b4b84a03b32c20647e3ae3a1b480a48109..25d08e7e4afda3c45363eec347a0d7153aa5cbfb 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -7,7 +7,7 @@ import java.time.Instant;
 import java.util.Collection;
 import java.util.Date;
 import java.util.Map;
-import com.dueris.canvas.network.channels.ChatChannel;
+import io.github.dueris.network.channels.ChatChannel;
 import org.bukkit.BanEntry;
 import org.bukkit.Chunk;
 import org.bukkit.DyeColor;
@@ -44,7 +44,6 @@ import org.bukkit.inventory.ItemStack;
 import org.bukkit.map.MapView;
 import org.bukkit.plugin.Plugin;
 import org.bukkit.plugin.messaging.PluginMessageRecipient;
-import org.bukkit.profile.PlayerProfile;
 import org.bukkit.scoreboard.Scoreboard;
 import org.jetbrains.annotations.ApiStatus;
 import org.jetbrains.annotations.NotNull;
diff --git a/src/main/java/org/bukkit/event/player/AsyncPlayerChatEvent.java b/src/main/java/org/bukkit/event/player/AsyncPlayerChatEvent.java
index f4d3d9c1e470a24342bf1c932a68582a5c1e983e..96c36262578e4f596d0f56fd9d4da9f1b32d30cf 100644
--- a/src/main/java/org/bukkit/event/player/AsyncPlayerChatEvent.java
+++ b/src/main/java/org/bukkit/event/player/AsyncPlayerChatEvent.java
@@ -2,9 +2,6 @@ package org.bukkit.event.player;
 
 import java.util.IllegalFormatException;
 import java.util.Set;
-import com.dueris.canvas.network.ChatChannelContainer;
-import com.dueris.canvas.network.channels.ChatChannel;
-import org.bukkit.Bukkit;
 import org.bukkit.entity.Player;
 import org.bukkit.event.Cancellable;
 import org.bukkit.event.HandlerList;
