From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sat, 11 Nov 2023 14:48:53 -0800
Subject: [PATCH] Add PlayerUseRespawnAnchorEvent


diff --git a/src/main/java/me/dueris/canvas/event/player/PlayerUseRespawnAnchorEvent.java b/src/main/java/me/dueris/canvas/event/player/PlayerUseRespawnAnchorEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..566983916066c0f72c1c88f01f66c2aa26e35aae
--- /dev/null
+++ b/src/main/java/me/dueris/canvas/event/player/PlayerUseRespawnAnchorEvent.java
@@ -0,0 +1,50 @@
+package me.dueris.canvas.event.player;
+
+import org.bukkit.block.Block;
+import org.bukkit.block.data.type.RespawnAnchor;
+import org.bukkit.entity.Player;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.block.BlockEvent;
+import org.bukkit.event.player.PlayerEvent;
+import org.jetbrains.annotations.NotNull;
+
+public class PlayerUseRespawnAnchorEvent extends BlockEvent {
+    public static final HandlerList HANDLER_LIST = new HandlerList();
+    public Player player;
+    public Block anchor;
+    public boolean canceld;
+
+    public PlayerUseRespawnAnchorEvent(final @NotNull Block theBlock, final @NotNull Player player) {
+        super(theBlock);
+        this.anchor = theBlock;
+        this.player = player;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return HANDLER_LIST;
+    }
+
+    @NotNull
+    @Override
+    public HandlerList getHandlers() {
+        return HANDLER_LIST;
+    }
+
+    public Player getPlayer() {
+        return player;
+    }
+
+    @NotNull
+    public Block getRespawnAnchor() {
+        return anchor;
+    }
+
+    public void setCanceled(boolean canceled){
+        this.canceld = canceled;
+    }
+
+    public boolean getCanceled(){
+        return canceld;
+    }
+}
