From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Fri, 10 Nov 2023 20:32:54 -0800
Subject: [PATCH] Enhance Arrow API


diff --git a/src/main/java/org/bukkit/entity/AbstractArrow.java b/src/main/java/org/bukkit/entity/AbstractArrow.java
index d173010d51d05928c35bb4bf5fbc08ce221ec474..2744ba808b3d3b819f8948ed40ee67143a0b8163 100644
--- a/src/main/java/org/bukkit/entity/AbstractArrow.java
+++ b/src/main/java/org/bukkit/entity/AbstractArrow.java
@@ -226,5 +226,13 @@ public interface AbstractArrow extends Projectile {
         ALLOWED,
         CREATIVE_ONLY;
     }
-    // Paper end
+    // Paper end -/- Canvas start - enhance arrow api
+
+    /**
+     * Returns if the arrow is an instance of a Spectral arrow
+     *
+     * @return if arrow is a spectral arrow
+     */
+    public boolean isSpectral();
+    // Canvas end
 }
