From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <jedimastertoothless@hotmail.com>
Date: Mon, 30 Oct 2023 09:47:56 +0000
Subject: [PATCH] Remove old particle trail method for Fireball


diff --git a/src/main/java/org/bukkit/entity/Fireball.java b/src/main/java/org/bukkit/entity/Fireball.java
index ea12d01be3fb8b25ae2fb2d48c75b7624f07df4e..339034dc57aa9ac5b9febdd0ee1cd7cc55fd6fc8 100644
--- a/src/main/java/org/bukkit/entity/Fireball.java
+++ b/src/main/java/org/bukkit/entity/Fireball.java
@@ -24,9 +24,9 @@ public interface Fireball extends Projectile, Explosive {
     @NotNull
     public Vector getDirection();
 
-    public void setParticleTrailType(Particle particle);
+    public void setHasParticleTrail(boolean hasParticleTrail);
 
-    public Particle getParticleTrailType();
+    public boolean getHasParticleTrail();
 
     public void setShouldBurn(boolean should);
 
