From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Wed, 15 Nov 2023 07:30:28 +0000
Subject: [PATCH] Allow Jars and Datapacks to be Bundled in Plugins


diff --git a/src/main/java/io/papermc/paper/plugin/configuration/PluginMeta.java b/src/main/java/io/papermc/paper/plugin/configuration/PluginMeta.java
index ef393f1f93ca48264fc1b6e3a27787f6a9152e1b..cbb565352d32c7b8dde1055435e06177c00177d5 100644
--- a/src/main/java/io/papermc/paper/plugin/configuration/PluginMeta.java
+++ b/src/main/java/io/papermc/paper/plugin/configuration/PluginMeta.java
@@ -74,6 +74,18 @@ public interface PluginMeta {
      */
     @NotNull
     String getVersion();
+    // Canvas start
+
+    /**
+     * Provides the plugin pack name defined in the plugin.yml
+     * 
+     * If null, Canvas wont load any datapack within the plugin
+     * 
+     * @return
+     */
+    @Nullable
+    public String getPluginPack();
+    // Canvas end
 
     /**
      * Provides the prefix that should be used for the plugin logger.
diff --git a/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java b/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
index c0691a849831f99268fdcb7b0f471f80a1a2a70e..613c6254dcadacd2c0fcf9efc68498bf7dba0ed2 100644
--- a/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
+++ b/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
@@ -251,6 +251,7 @@ public final class PluginDescriptionFile implements io.papermc.paper.plugin.conf
     private List<String> authors = null;
     private List<String> contributors = null;
     private String website = null;
+    private String pluginPack = null;
     private String prefix = null;
     private PluginLoadOrder order = PluginLoadOrder.POSTWORLD;
     private List<Permission> permissions = null;
@@ -264,7 +265,30 @@ public final class PluginDescriptionFile implements io.papermc.paper.plugin.conf
      * Don't use this.
      */
     @org.jetbrains.annotations.ApiStatus.Internal
-    public PluginDescriptionFile(String rawName, String name, List<String> provides, String main, String classLoaderOf, List<String> depend, List<String> softDepend, List<String> loadBefore, String version, Map<String, Map<String, Object>> commands, String description, List<String> authors, List<String> contributors, String website, String prefix, PluginLoadOrder order, List<Permission> permissions, PermissionDefault defaultPerm, Set<PluginAwareness> awareness, String apiVersion, List<String> libraries) {
+    // Canvas start - expand for easier readability
+    public PluginDescriptionFile(String rawName,
+                                    String name,
+                                    List<String> provides,
+                                    String main,
+                                    String classLoaderOf,
+                                    List<String> depend,
+                                    List<String> softDepend,
+                                    List<String> loadBefore,
+                                    String version,
+                                    Map<String, Map<String, Object>> commands,
+                                    String description,
+                                    List<String> authors,
+                                    List<String> contributors,
+                                    String website,
+                                    String pluginPack, // Canvas - add pluginPack
+                                    String prefix,
+                                    PluginLoadOrder order,
+                                    List<Permission> permissions,
+                                    PermissionDefault defaultPerm,
+                                    Set<PluginAwareness> awareness,
+                                    String apiVersion,
+                                    List<String> libraries) {
+    // Canvas end
         this.rawName = rawName;
         this.name = name;
         this.provides = provides;
@@ -279,6 +303,7 @@ public final class PluginDescriptionFile implements io.papermc.paper.plugin.conf
         this.authors = authors;
         this.contributors = contributors;
         this.website = website;
+        this.pluginPack = pluginPack; // Canvas
         this.prefix = prefix;
         this.order = order;
         this.permissions = permissions;
@@ -589,6 +614,27 @@ public final class PluginDescriptionFile implements io.papermc.paper.plugin.conf
     public String getWebsite() {
         return website;
     }
+    // Canvas start
+
+    /**
+     * Gives the pack name of the plugin. This name is a unique identifier for
+     * plugins pack to be used for built-in datapack loading.
+     * <ul>
+     * <li>Must consist of all alphanumeric characters, underscores, hyphon,
+     *     and NO periods (a-z,A-Z,0-9, _-). Any other character will cause the
+     *     plugin.yml(and the canvas plugin loader) to fail loading.
+     * <li>Case sensitive.
+     * In the plugin.yml, this entry is named <code>plugin-pack</code>.
+     * <p>
+     * Example:<blockquote><pre>plugin-pack: MyPluginPack</pre></blockquote>
+     *
+     * @return the pack name of the plugin
+     */
+    @Nullable
+    public String getPluginPack() {
+        return this.pluginPack;
+    }
+    // Canvas end
 
     /**
      * Gives a list of other plugins that the plugin requires.
