From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Wed, 1 Nov 2023 21:41:23 +0000
Subject: [PATCH] More Animal APIs


diff --git a/src/main/java/org/bukkit/entity/Cow.java b/src/main/java/org/bukkit/entity/Cow.java
index cd4ed4de01c7376c1b35f31af0bba6ac4e26f947..3b526ba2dd6d0e716a498d1aee83509e0b0dd94f 100644
--- a/src/main/java/org/bukkit/entity/Cow.java
+++ b/src/main/java/org/bukkit/entity/Cow.java
@@ -3,4 +3,11 @@ package org.bukkit.entity;
 /**
  * Represents a Cow.
  */
-public interface Cow extends Animals {}
+public interface Cow extends Animals {
+    // Canvas start
+    /**
+     * Converts the Cow to a Mushroom Cow
+     */
+    public void convertToMushroom();
+    // Canvas end
+}
diff --git a/src/main/java/org/bukkit/entity/EntityType.java b/src/main/java/org/bukkit/entity/EntityType.java
index 6cf14afd2c1b8018c67c6cac302ae28bb12ab5a5..b983791793cb897b2ad8b0f2386535b49405d3e3 100644
--- a/src/main/java/org/bukkit/entity/EntityType.java
+++ b/src/main/java/org/bukkit/entity/EntityType.java
@@ -236,6 +236,7 @@ public enum EntityType implements Keyed, Translatable, net.kyori.adventure.trans
     WOLF("wolf", Wolf.class, 95),
     MUSHROOM_COW("mooshroom", MushroomCow.class, 96),
     SNOWMAN("snow_golem", Snowman.class, 97),
+    SNOW_GOLEM("snow_golem", Snowman.class, 97), // Canvas - EntityType conversion fixes
     OCELOT("ocelot", Ocelot.class, 98),
     IRON_GOLEM("iron_golem", IronGolem.class, 99),
     HORSE("horse", Horse.class, 100),
diff --git a/src/main/java/org/bukkit/entity/Mob.java b/src/main/java/org/bukkit/entity/Mob.java
index c06d28a06cfde8fdbb11fc59ab29d7d8cb9eacf4..a5e81937bc2b432893647bc643e578886ce69636 100644
--- a/src/main/java/org/bukkit/entity/Mob.java
+++ b/src/main/java/org/bukkit/entity/Mob.java
@@ -195,5 +195,14 @@ public interface Mob extends LivingEntity, Lootable {
      * @return the amount of experience the mob will possibly drop
      */
     public int getPossibleExperienceReward();
-    // Paper end
+    // Paper end -/- Canvas start
+
+    /**
+     * Allows you to convert any mob to another EntityType
+     * 
+     * @param entityType EntityType to convert the mob to
+     * @param keepEquipment Should keep equipment on transformation
+     */
+    public void convertTo(EntityType entityType, boolean keepEquipment);
+    // Canvas end
 }
diff --git a/src/main/java/org/bukkit/entity/MushroomCow.java b/src/main/java/org/bukkit/entity/MushroomCow.java
index a5ed2bb03684da44951ec8aecbb229fe2b355e63..4c69c6ff4ae0d9cc697ddb345299ea4df7796f6d 100644
--- a/src/main/java/org/bukkit/entity/MushroomCow.java
+++ b/src/main/java/org/bukkit/entity/MushroomCow.java
@@ -70,5 +70,10 @@ public interface MushroomCow extends Cow, io.papermc.paper.entity.Shearable { //
      *             or null if this cow does not give effects
      */
     void setStewEffect(@org.jetbrains.annotations.Nullable org.bukkit.potion.PotionEffectType type);
-    // Paper end
+    // Paper end -/- Canvas start
+    /**
+     * Converts the Mushroom Cow to a normal Cow
+     */
+    public void convertToNormal();
+    // Canvas end
 }
diff --git a/src/main/java/org/bukkit/entity/Panda.java b/src/main/java/org/bukkit/entity/Panda.java
index 57cf24cfd15a541f60aafc8507c189344aead0f7..aba1371c9084361c4ce8c92d66b666283249f7dc 100644
--- a/src/main/java/org/bukkit/entity/Panda.java
+++ b/src/main/java/org/bukkit/entity/Panda.java
@@ -214,4 +214,11 @@ public interface Panda extends Animals, Sittable {
             return recessive;
         }
     }
+    // Canvas start
+    public void roll();
+
+    public void sneeze();
+
+    public void lieOnBack();
+    // Canvas end
 }
diff --git a/src/main/java/org/bukkit/entity/Parrot.java b/src/main/java/org/bukkit/entity/Parrot.java
index c545a025c2ff5d3a0a4fff4e379277f38acf5c06..035c968963ce39c708680789c761bf8a6eb5d43c 100644
--- a/src/main/java/org/bukkit/entity/Parrot.java
+++ b/src/main/java/org/bukkit/entity/Parrot.java
@@ -54,4 +54,12 @@ public interface Parrot extends Tameable, Sittable {
          */
         GRAY;
     }
+    // Canvas start
+
+    public void landOnShoulder();
+
+    public void setPoisonousFood(org.bukkit.Material material);
+
+    public org.bukkit.Material getPoisonousFood();
+    // Canvas end
 }
diff --git a/src/main/java/org/bukkit/entity/Steerable.java b/src/main/java/org/bukkit/entity/Steerable.java
index da77a7188b92386fd07304b5e2bf385bd77f80d9..bc279ef86822ea834cdc804fda459222d655d5ff 100644
--- a/src/main/java/org/bukkit/entity/Steerable.java
+++ b/src/main/java/org/bukkit/entity/Steerable.java
@@ -68,4 +68,11 @@ public interface Steerable extends Animals {
      */
     @NotNull
     public Material getSteerMaterial();
+    // Canvas start
+
+    /**
+     * Sets the entity steer item
+     */
+    void setSteerMaterial(org.bukkit.Material material);
+    // Canvas end
 }
