From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Wed, 1 Nov 2023 21:41:23 +0000
Subject: [PATCH] More Mob APIs


diff --git a/src/main/java/org/bukkit/entity/Cow.java b/src/main/java/org/bukkit/entity/Cow.java
index cd4ed4de01c7376c1b35f31af0bba6ac4e26f947..3b526ba2dd6d0e716a498d1aee83509e0b0dd94f 100644
--- a/src/main/java/org/bukkit/entity/Cow.java
+++ b/src/main/java/org/bukkit/entity/Cow.java
@@ -3,4 +3,11 @@ package org.bukkit.entity;
 /**
  * Represents a Cow.
  */
-public interface Cow extends Animals {}
+public interface Cow extends Animals {
+    // Canvas start
+    /**
+     * Converts the Cow to a Mushroom Cow
+     */
+    public void convertToMushroom();
+    // Canvas end
+}
diff --git a/src/main/java/org/bukkit/entity/EntityType.java b/src/main/java/org/bukkit/entity/EntityType.java
index 6cf14afd2c1b8018c67c6cac302ae28bb12ab5a5..b983791793cb897b2ad8b0f2386535b49405d3e3 100644
--- a/src/main/java/org/bukkit/entity/EntityType.java
+++ b/src/main/java/org/bukkit/entity/EntityType.java
@@ -236,6 +236,7 @@ public enum EntityType implements Keyed, Translatable, net.kyori.adventure.trans
     WOLF("wolf", Wolf.class, 95),
     MUSHROOM_COW("mooshroom", MushroomCow.class, 96),
     SNOWMAN("snow_golem", Snowman.class, 97),
+    SNOW_GOLEM("snow_golem", Snowman.class, 97), // Canvas - EntityType conversion fixes
     OCELOT("ocelot", Ocelot.class, 98),
     IRON_GOLEM("iron_golem", IronGolem.class, 99),
     HORSE("horse", Horse.class, 100),
diff --git a/src/main/java/org/bukkit/entity/Mob.java b/src/main/java/org/bukkit/entity/Mob.java
index c06d28a06cfde8fdbb11fc59ab29d7d8cb9eacf4..a5e81937bc2b432893647bc643e578886ce69636 100644
--- a/src/main/java/org/bukkit/entity/Mob.java
+++ b/src/main/java/org/bukkit/entity/Mob.java
@@ -195,5 +195,14 @@ public interface Mob extends LivingEntity, Lootable {
      * @return the amount of experience the mob will possibly drop
      */
     public int getPossibleExperienceReward();
-    // Paper end
+    // Paper end -/- Canvas start
+
+    /**
+     * Allows you to convert any mob to another EntityType
+     * 
+     * @param entityType EntityType to convert the mob to
+     * @param keepEquipment Should keep equipment on transformation
+     */
+    public void convertTo(EntityType entityType, boolean keepEquipment);
+    // Canvas end
 }
diff --git a/src/main/java/org/bukkit/entity/MushroomCow.java b/src/main/java/org/bukkit/entity/MushroomCow.java
index a5ed2bb03684da44951ec8aecbb229fe2b355e63..4c69c6ff4ae0d9cc697ddb345299ea4df7796f6d 100644
--- a/src/main/java/org/bukkit/entity/MushroomCow.java
+++ b/src/main/java/org/bukkit/entity/MushroomCow.java
@@ -70,5 +70,10 @@ public interface MushroomCow extends Cow, io.papermc.paper.entity.Shearable { //
      *             or null if this cow does not give effects
      */
     void setStewEffect(@org.jetbrains.annotations.Nullable org.bukkit.potion.PotionEffectType type);
-    // Paper end
+    // Paper end -/- Canvas start
+    /**
+     * Converts the Mushroom Cow to a normal Cow
+     */
+    public void convertToNormal();
+    // Canvas end
 }
diff --git a/src/main/java/org/bukkit/entity/Panda.java b/src/main/java/org/bukkit/entity/Panda.java
index aa5686df134185334a74429576ff0709a604dbfd..e9d0877452ecb0880eba4728c675eedc3ea3fc0b 100644
--- a/src/main/java/org/bukkit/entity/Panda.java
+++ b/src/main/java/org/bukkit/entity/Panda.java
@@ -225,4 +225,11 @@ public interface Panda extends Animals, Sittable {
             return recessive;
         }
     }
+    // Canvas start
+    public void roll();
+
+    public void sneeze();
+
+    public void lieOnBack();
+    // Canvas end
 }
diff --git a/src/main/java/org/bukkit/entity/Parrot.java b/src/main/java/org/bukkit/entity/Parrot.java
index c545a025c2ff5d3a0a4fff4e379277f38acf5c06..035c968963ce39c708680789c761bf8a6eb5d43c 100644
--- a/src/main/java/org/bukkit/entity/Parrot.java
+++ b/src/main/java/org/bukkit/entity/Parrot.java
@@ -54,4 +54,12 @@ public interface Parrot extends Tameable, Sittable {
          */
         GRAY;
     }
+    // Canvas start
+
+    public void landOnShoulder();
+
+    public void setPoisonousFood(org.bukkit.Material material);
+
+    public org.bukkit.Material getPoisonousFood();
+    // Canvas end
 }
diff --git a/src/main/java/org/bukkit/entity/PigZombie.java b/src/main/java/org/bukkit/entity/PigZombie.java
index b2ec535bb1b0ce0c114ddd7638b90218b05cd835..102aac250adb83f2de61be93783c9088848f9f31 100644
--- a/src/main/java/org/bukkit/entity/PigZombie.java
+++ b/src/main/java/org/bukkit/entity/PigZombie.java
@@ -55,4 +55,8 @@ public interface PigZombie extends Zombie {
      */
     @Override
     public void setConversionTime(int time);
+    // Canvas start
+
+    public void convertToNormal();
+    // Canvas end
 }
diff --git a/src/main/java/org/bukkit/entity/Piglin.java b/src/main/java/org/bukkit/entity/Piglin.java
index eb0b7c18c1266748ff1e8e18e49b6c4f6e078b83..a3c8cf8b7d46c734c8138f279255a14b5e242659 100644
--- a/src/main/java/org/bukkit/entity/Piglin.java
+++ b/src/main/java/org/bukkit/entity/Piglin.java
@@ -131,6 +131,9 @@ public interface Piglin extends PiglinAbstract, InventoryHolder, com.destroystok
      * @return is dancing
      */
     boolean isDancing();
-    // Paper end
+    // Paper end -/- Canvas start
+
+    public void convertToZombified();
+    // Canvas end
 
 }
diff --git a/src/main/java/org/bukkit/entity/PufferFish.java b/src/main/java/org/bukkit/entity/PufferFish.java
index 2d26c19b53ba3887a171010a9a5b47fdc3d86995..c33fee71f9596624f30d8a0e11cfff25a15a7145 100644
--- a/src/main/java/org/bukkit/entity/PufferFish.java
+++ b/src/main/java/org/bukkit/entity/PufferFish.java
@@ -18,4 +18,16 @@ public interface PufferFish extends Fish {
      * @param state new puff state
      */
     void setPuffState(int state);
+    // Canvas start
+    
+    public void setScaleSizeSmall(float scale);
+
+    public void setScaleSizeMedium(float scale);
+
+    public void setScaleSizeLarge(float scale);
+
+    public void setCanPoison(boolean can);
+
+    public boolean getCanPoison();
+    // Canvas end
 }
diff --git a/src/main/java/org/bukkit/entity/Rabbit.java b/src/main/java/org/bukkit/entity/Rabbit.java
index 979062aa476e3bd75166458d8831894fba8778cd..55266ca0cc17c15fdac5da8661195d88b3d6f8d9 100644
--- a/src/main/java/org/bukkit/entity/Rabbit.java
+++ b/src/main/java/org/bukkit/entity/Rabbit.java
@@ -66,4 +66,8 @@ public interface Rabbit extends Animals {
          */
         THE_KILLER_BUNNY
     }
+    // Canvas start
+
+    public boolean wantsMoreFood();
+    // Canvas end
 }
diff --git a/src/main/java/org/bukkit/entity/Ravager.java b/src/main/java/org/bukkit/entity/Ravager.java
index 0eb7214472f3a43641a3526000af6beeefb7b36f..29c56a2502b9457be295c48d3e60600818bbc3d3 100644
--- a/src/main/java/org/bukkit/entity/Ravager.java
+++ b/src/main/java/org/bukkit/entity/Ravager.java
@@ -58,6 +58,9 @@ public interface Ravager extends Raider {
      * @param ticks ticks roaring or -1 if they should no longer be roaring
      */
     void setRoarTicks(int ticks);
+    // Canvas start
 
+    public void roar();
+    // Canvas end
 }
 // Paper end
diff --git a/src/main/java/org/bukkit/entity/Squid.java b/src/main/java/org/bukkit/entity/Squid.java
index 5127bd80dce2c03c525cafdf7dce96833ef160ba..92ce48cec3a54bec96a1fa13c220848722c52e69 100644
--- a/src/main/java/org/bukkit/entity/Squid.java
+++ b/src/main/java/org/bukkit/entity/Squid.java
@@ -9,5 +9,10 @@ public interface Squid extends WaterMob {
      * Sets the EntityType to a Glow Squid.
      */
     public void setGlowingSquid();
+
+    /**
+     * Squirts ink as if the Squid was ingured
+     */
+    public void squirtInk();
     // Canvas end
 }
diff --git a/src/main/java/org/bukkit/entity/Steerable.java b/src/main/java/org/bukkit/entity/Steerable.java
index da77a7188b92386fd07304b5e2bf385bd77f80d9..bc279ef86822ea834cdc804fda459222d655d5ff 100644
--- a/src/main/java/org/bukkit/entity/Steerable.java
+++ b/src/main/java/org/bukkit/entity/Steerable.java
@@ -68,4 +68,11 @@ public interface Steerable extends Animals {
      */
     @NotNull
     public Material getSteerMaterial();
+    // Canvas start
+
+    /**
+     * Sets the entity steer item
+     */
+    void setSteerMaterial(org.bukkit.Material material);
+    // Canvas end
 }
