From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Wed, 22 Nov 2023 09:00:22 +0000
Subject: [PATCH] Expose CreateEndPlatform API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index dbc92742e6433e8e3902b85b232e1c4d3d667480..c6bdd0e85d9388ccb1d1034463dd43cf34a7c8c8 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3667,6 +3667,21 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * @param pitch the pitch
      */
     void setRotation(float yaw, float pitch);
+    // Canvas start
+
+    /**
+     * Creates an end spawn platform at the current player position
+     * Doesnt call the PortalCreationEvnet
+     */
+    public void createEndPlatformAtCurrentPos();
+
+    /**
+     * Creates an end spawn platform at the current player position
+     * 
+     * @param callEvent decides weither to call the PortalCreationEvent
+     */
+    public void createEndPlatformAtCurrentPos(boolean callEvent);
+    // Canvas end
 
     /**
      * Causes the player to look towards the given position.
