From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Sat, 4 Nov 2023 16:44:22 +0000
Subject: [PATCH] Fix duplicated SNOWMAN bug in Registry


diff --git a/src/main/java/org/bukkit/entity/EntityType.java b/src/main/java/org/bukkit/entity/EntityType.java
index b983791793cb897b2ad8b0f2386535b49405d3e3..cd081dd809f41c9c9cff63d193a8a0d8f7a47893 100644
--- a/src/main/java/org/bukkit/entity/EntityType.java
+++ b/src/main/java/org/bukkit/entity/EntityType.java
@@ -236,7 +236,7 @@ public enum EntityType implements Keyed, Translatable, net.kyori.adventure.trans
     WOLF("wolf", Wolf.class, 95),
     MUSHROOM_COW("mooshroom", MushroomCow.class, 96),
     SNOWMAN("snow_golem", Snowman.class, 97),
-    SNOW_GOLEM("snow_golem", Snowman.class, 97), // Canvas - EntityType conversion fixes
+    // SNOW_GOLEM("snow_golem", Snowman.class, 97), // Canvas - EntityType conversion fixes
     OCELOT("ocelot", Ocelot.class, 98),
     IRON_GOLEM("iron_golem", IronGolem.class, 99),
     HORSE("horse", Horse.class, 100),
