From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Thu, 2 Nov 2023 17:23:11 +0000
Subject: [PATCH] FInished non-hostile entity apis


diff --git a/src/main/java/org/bukkit/entity/PufferFish.java b/src/main/java/org/bukkit/entity/PufferFish.java
index 2d26c19b53ba3887a171010a9a5b47fdc3d86995..c33fee71f9596624f30d8a0e11cfff25a15a7145 100644
--- a/src/main/java/org/bukkit/entity/PufferFish.java
+++ b/src/main/java/org/bukkit/entity/PufferFish.java
@@ -18,4 +18,16 @@ public interface PufferFish extends Fish {
      * @param state new puff state
      */
     void setPuffState(int state);
+    // Canvas start
+    
+    public void setScaleSizeSmall(float scale);
+
+    public void setScaleSizeMedium(float scale);
+
+    public void setScaleSizeLarge(float scale);
+
+    public void setCanPoison(boolean can);
+
+    public boolean getCanPoison();
+    // Canvas end
 }
diff --git a/src/main/java/org/bukkit/entity/Rabbit.java b/src/main/java/org/bukkit/entity/Rabbit.java
index 979062aa476e3bd75166458d8831894fba8778cd..55266ca0cc17c15fdac5da8661195d88b3d6f8d9 100644
--- a/src/main/java/org/bukkit/entity/Rabbit.java
+++ b/src/main/java/org/bukkit/entity/Rabbit.java
@@ -66,4 +66,8 @@ public interface Rabbit extends Animals {
          */
         THE_KILLER_BUNNY
     }
+    // Canvas start
+
+    public boolean wantsMoreFood();
+    // Canvas end
 }
diff --git a/src/main/java/org/bukkit/entity/Squid.java b/src/main/java/org/bukkit/entity/Squid.java
index 5127bd80dce2c03c525cafdf7dce96833ef160ba..92ce48cec3a54bec96a1fa13c220848722c52e69 100644
--- a/src/main/java/org/bukkit/entity/Squid.java
+++ b/src/main/java/org/bukkit/entity/Squid.java
@@ -9,5 +9,10 @@ public interface Squid extends WaterMob {
      * Sets the EntityType to a Glow Squid.
      */
     public void setGlowingSquid();
+
+    /**
+     * Squirts ink as if the Squid was ingured
+     */
+    public void squirtInk();
     // Canvas end
 }
diff --git a/src/main/java/org/bukkit/entity/PigZombie.java b/src/main/java/org/bukkit/entity/PigZombie.java
index b2ec535bb1b0ce0c114ddd7638b90218b05cd835..102aac250adb83f2de61be93783c9088848f9f31 100644
--- a/src/main/java/org/bukkit/entity/PigZombie.java
+++ b/src/main/java/org/bukkit/entity/PigZombie.java
@@ -55,4 +55,8 @@ public interface PigZombie extends Zombie {
      */
     @Override
     public void setConversionTime(int time);
+    // Canvas start
+
+    public void convertToNormal();
+    // Canvas end
 }
diff --git a/src/main/java/org/bukkit/entity/Piglin.java b/src/main/java/org/bukkit/entity/Piglin.java
index eb0b7c18c1266748ff1e8e18e49b6c4f6e078b83..a3c8cf8b7d46c734c8138f279255a14b5e242659 100644
--- a/src/main/java/org/bukkit/entity/Piglin.java
+++ b/src/main/java/org/bukkit/entity/Piglin.java
@@ -131,6 +131,9 @@ public interface Piglin extends PiglinAbstract, InventoryHolder, com.destroystok
      * @return is dancing
      */
     boolean isDancing();
-    // Paper end
+    // Paper end -/- Canvas start
+
+    public void convertToZombified();
+    // Canvas end
 
 }
diff --git a/src/main/java/org/bukkit/entity/Ravager.java b/src/main/java/org/bukkit/entity/Ravager.java
index 0eb7214472f3a43641a3526000af6beeefb7b36f..29c56a2502b9457be295c48d3e60600818bbc3d3 100644
--- a/src/main/java/org/bukkit/entity/Ravager.java
+++ b/src/main/java/org/bukkit/entity/Ravager.java
@@ -58,6 +58,9 @@ public interface Ravager extends Raider {
      * @param ticks ticks roaring or -1 if they should no longer be roaring
      */
     void setRoarTicks(int ticks);
+    // Canvas start
 
+    public void roar();
+    // Canvas end
 }
 // Paper end
