From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Fri, 17 Nov 2023 20:53:11 +0000
Subject: [PATCH] Add Entity Movement API


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index a36fc32dd83339b6939cb4046538ec6285042397..8787c9643c351bd947156febd3ffd2b0ee7160af 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -272,6 +272,18 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      * @return whether visual fire is enabled
      */
     boolean isVisualFire();
+    // Canvas start - movement api
+
+    public void step();
+
+    public void step(int steps);
+
+    public void setDeltaMovement(double x, double y, double z);
+
+    public void reapplyPosition();
+
+    public void turn(double cursorDeltaX, double cursorDeltaY);
+    // Canvas end
 
     /**
      * Returns the entity's current freeze ticks (amount of ticks the entity has
@@ -713,6 +725,17 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      * @param cooldown portal cooldown ticks
      */
     void setPortalCooldown(int cooldown);
+    // Canvas start - why not add "isInPortalCooldown"?
+
+    /**
+     * Retrieves if the entity is in the portal cooldown, or
+     * the period of time (int ticks) before the entity can use
+     * a portal.
+     * 
+     * @return if the entity is in "portal cooldown"
+     */
+    public boolean isInPortalCooldown();
+    // Canvas end
 
     /**
      * Returns a set of tags for this entity.
