From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Fri, 10 Nov 2023 20:39:46 -0800
Subject: [PATCH] Allow to configure orbs can merge to API


diff --git a/src/main/java/org/bukkit/entity/ExperienceOrb.java b/src/main/java/org/bukkit/entity/ExperienceOrb.java
index dec70bbfaf73a9d525b2c45682b804c684e1645b..53ae9f86eed0961d3276c9841241b9c5692a32a2 100644
--- a/src/main/java/org/bukkit/entity/ExperienceOrb.java
+++ b/src/main/java/org/bukkit/entity/ExperienceOrb.java
@@ -115,5 +115,22 @@ public interface ExperienceOrb extends Entity {
      */
     @NotNull
     SpawnReason getSpawnReason();
-    // Paper end
+    // Paper end -/- Canvas start
+
+    /**
+     * Sets if the orb is allowed to merge.
+     * @<code>false</code> If false, the orb will not be able to merge with other orbs
+     *
+     * @<code>true</code> If true, the orb will be able to merge with other orbs
+     *
+     * @param allowMerge
+     */
+    public void setAllowMerge(boolean allowMerge);
+
+    /**
+     * Returns if the orb is allowed to merge with other orbs
+     * @return if the orb can merge with other orbs
+     */
+    public boolean allowMerge();
+    // Canvas end
 }
