From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Fri, 17 Nov 2023 20:58:12 +0000
Subject: [PATCH] Add gamemode syncing api


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 4b0cd0494d83c3bba36aa8968aad195e0f1b86a8..cc725fa4cd547df6008aa79f6dfd85ad6ea11f1c 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -799,6 +799,41 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * Stop all sounds from playing.
      */
     public void stopAllSounds();
+    // Canvas start - more player api
+
+    /**
+     * This javadoc is for context of the CLIENT+SERVER gamemode methods
+     * The server is the dataholder of the world and can create how the
+     * data operates, thus meaning that the gamemodes can go out of sync
+     * from eachother, allowing for insane effects like having Spectator
+     * clipping while appearing to be in Creative mode.
+     * 
+     * Sets the CLIENT gamemode
+     * This controls what the Player sees
+     * 
+     * @param gamemode
+     */
+    public void setClientGamemode(GameMode gamemode);
+
+    /**
+     * This javadoc is for context of the CLIENT+SERVER gamemode methods
+     * The server is the dataholder of the world and can create how the
+     * data operates, thus meaning that the gamemodes can go out of sync
+     * from eachother, allowing for insane effects like having Spectator
+     * clipping while appearing to be in Creative mode.
+     * 
+     * Sets the SERVER gamemode
+     * This is the side that controls what the Player can do
+     * 
+     * @param gamemode
+     */
+    public void setServerGamemode(GameMode gamemode);
+
+    /**
+     * Resyncs the gamemode from the SERVER side gamemode
+     */
+    public void resyncGamemode();
+    // Canvas end
 
     /**
      * Plays an effect to just this player.
