From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Fri, 10 Nov 2023 20:15:28 -0800
Subject: [PATCH] ServerFinishStarting Event


diff --git a/src/main/java/io/github/dueris/event/ServerFinishStartingEvent.java b/src/main/java/io/github/dueris/event/ServerFinishStartingEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..aa14a59e98c04069489c426cf7891fd2a25d690b
--- /dev/null
+++ b/src/main/java/io/github/dueris/event/ServerFinishStartingEvent.java
@@ -0,0 +1,20 @@
+package io.github.dueris.event;
+
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+import org.jetbrains.annotations.NotNull;
+
+public class ServerFinishStartingEvent extends Event {
+    public static final HandlerList HANDLER_LIST = new HandlerList();
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return HANDLER_LIST;
+    }
+
+    @NotNull
+    @Override
+    public HandlerList getHandlers() {
+        return HANDLER_LIST;
+    }
+}
