From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sat, 25 Nov 2023 18:01:10 -0800
Subject: [PATCH] Add Modded Client API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 52569b7160eb2015f165c346a5f944aef685dca2..cbf9b910fb3f422bcfd38bad3c46d4249239fc8d 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -59,7 +59,40 @@ import org.jetbrains.annotations.Nullable;
  * Represents a player, connected or not
  */
 public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginMessageRecipient, net.kyori.adventure.identity.Identified, net.kyori.adventure.bossbar.BossBarViewer, com.destroystokyo.paper.network.NetworkClient { // Paper
+    // Canvas start - client modDetector
 
+    /**
+     * Gets if the client is running forge.
+     * WARNING: THIS CAN BE FAKED, THIS IS ACCURATE BUT OFTEN TIMES CAN
+     * BE WRONG DUE TO OTHER MODS INSTALLED
+     * @return if the client is on forge
+     */
+    public boolean isForgeClient();
+
+    /**
+     * Gets if the client is running fabric.
+     * WARNING: THIS CAN BE FAKED, THIS IS ACCURATE BUT OFTEN TIMES CAN
+     * BE WRONG DUE TO OTHER MODS INSTALLED
+     * @return if the client is on fabric
+     */
+    public boolean isFabricClient();
+
+    /**
+     * Upon the handshake of a client and server with a modded client,
+     * the client will send some data that the server reads and
+     * interperets as a CustomPacketPayload, that payload is decripted
+     * and can be used to tell if the client is running forge or fabric.
+     *
+     * This is the provider for that string. Once the data is read it is saved
+     * to here. If none is sent, it will return "No data found, or no data sent."
+     *
+     * If data is sent/found, it will return something like this:
+     * forge:tier_sortingforge:handshakeforge:login
+     *
+     * @return the string version of the modded client handshake data
+     */
+    public String getModdedClientHandshakeData();
+    // Canvas end
     // Paper start
     @Override
     default net.kyori.adventure.identity.@NotNull Identity identity() {
