From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <159172167+PurpleWolfMC@users.noreply.github.com>
Date: Wed, 7 Feb 2024 07:05:39 +0000
Subject: [PATCH] Allow setting respawn Dimension


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index eb1106259bad1a31141f8739f2d0bd088fdf3db4..6e62a6430911b1a188129c235c6e6415ffc1b0d9 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -608,7 +608,13 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     @Override
     @Deprecated
     public Location getBedSpawnLocation();
+    // Canvas start
+
+    public void setRespawnDimension(org.bukkit.World world);
 
+    @NotNull
+    public org.bukkit.World getRespawnDimension();
+    // Canvas end
     /**
      * Gets the Location where the player will spawn at, null if they
      * don't have a valid respawn point.
