From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Tue, 21 Nov 2023 19:14:53 +0000
Subject: [PATCH] Expose ReloadCount to Server


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index ef6e9a38fd96adfb0d175a7b7acdb5a9ff95ebfd..f2b0e20c3ebf5dffad2818c6cf66f90842bad42b 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -930,6 +930,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      * Reloads the server, refreshing settings and plugin information.
      */
     public void reload();
+    // Canvas start
+
+    /**
+     * Gets the amount of times the server has reloaded("/reload confirm")
+     */
+    public int getReloadCount();
+    // Canvas end
 
     /**
      * Reload only the Minecraft data for the server. This includes custom
