From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Thu, 16 Nov 2023 19:07:47 +0000
Subject: [PATCH] More Advancement API


diff --git a/src/main/java/org/bukkit/advancement/AdvancementDisplay.java b/src/main/java/org/bukkit/advancement/AdvancementDisplay.java
index aec6be7e121da3eb8a464b6934da29ab6b473885..898caa1e5b93d2c47f7aeec560e053e0badae91d 100644
--- a/src/main/java/org/bukkit/advancement/AdvancementDisplay.java
+++ b/src/main/java/org/bukkit/advancement/AdvancementDisplay.java
@@ -82,4 +82,65 @@ public interface AdvancementDisplay {
      */
     @NotNull
     AdvancementDisplayType getType();
+    // Canvas start - more advancement api
+
+    /**
+     * Sets the title of the advancement from the provided STRING
+     * 
+     * NOTE: IT WILL NOT MODIFY ALREADY SENT ADVANCEMENTS, AKA ONES
+     * BEING ACTIVELY DISPLAYED ON THE PLAYERS SCREEN
+     * 
+     * @param title
+     */
+    public void setTitle(String title);
+
+    /**
+     * Sets the description of the advancement from the provided STRING
+     * 
+     * NOTE: IT WILL NOT MODIFY ALREADY SENT ADVANCEMENTS, AKA ONES
+     * BEING ACTIVELY DISPLAYED ON THE PLAYERS SCREEN
+     * 
+     * @param description
+     */
+    public void setDescription(String description);
+
+    /**
+     * Sets if it should announce the "toast" for the player
+     * 
+     * NOTE: IT WILL NOT MODIFY ALREADY SENT ADVANCEMENTS, AKA ONES
+     * BEING ACTIVELY DISPLAYED ON THE PLAYERS SCREEN
+     * 
+     * @param should
+     */
+    public void setShouldShowToast(boolean should);
+
+    /**
+     * Sets if it should announce to chat that the player has
+     * completed the advancement
+     * 
+     * NOTE: IT WILL NOT MODIFY ALREADY SENT ADVANCEMENTS, AKA ONES
+     * BEING ACTIVELY DISPLAYED ON THE PLAYERS SCREEN
+     * 
+     * @param should
+     */
+    public void setShouldAnnounceChat(boolean should);
+
+    /**
+     * Sets if the advancement should be hidden or not from
+     * the advancement tree, along with its children
+     * 
+     * @param hidden
+     */
+    public void setHidden(boolean hidden);
+
+    /**
+     * Sets the icon of the advancement from the provided ItemStack
+     * 
+     * NOTE: IT WILL NOT MODIFY ALREADY SENT ADVANCEMENTS, AKA ONES
+     * BEING ACTIVELY DISPLAYED ON THE PLAYERS SCREEN
+     * 
+     * @param item
+     */
+    public void setIcon(org.bukkit.inventory.ItemStack item);
+    // Canvas end
 }
