From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <122416109+Dueris@users.noreply.github.com>
Date: Thu, 1 Feb 2024 03:52:11 +0000
Subject: [PATCH] Fix MixinPlugin ClassLoading


diff --git a/src/main/java/org/bukkit/plugin/java/JavaPlugin.java b/src/main/java/org/bukkit/plugin/java/JavaPlugin.java
index d5a3c3dce76c4ed0f1184ab5ba21db9c5f1c01ec..00796ff65a0b61bcb7fe24723c4eb89906c6d6c7 100644
--- a/src/main/java/org/bukkit/plugin/java/JavaPlugin.java
+++ b/src/main/java/org/bukkit/plugin/java/JavaPlugin.java
@@ -56,11 +56,16 @@ public abstract class JavaPlugin extends PluginBase {
 
     public JavaPlugin() {
         // Paper start
+        // Canvas start
+        if(!this.getClass().getClassLoader().getClass().getName().toString().equalsIgnoreCase("space.vectrix.ignite.launch.ember.EmberClassLoader"))
+        {
+        // Canvas end
         if (this.getClass().getClassLoader() instanceof io.papermc.paper.plugin.provider.classloader.ConfiguredPluginClassLoader configuredPluginClassLoader) {
             configuredPluginClassLoader.init(this);
         } else {
             throw new IllegalStateException("JavaPlugin requires to be created by a valid classloader.");
         }
+        } // Canvas
         // Paper end
     }
 
