From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Fri, 17 Nov 2023 17:05:26 +0000
Subject: [PATCH] Add setDecorationsDirty to MapRenderer


diff --git a/src/main/java/org/bukkit/map/MapRenderer.java b/src/main/java/org/bukkit/map/MapRenderer.java
index 22d8f31b1b8a5dbb5ab3275068642937c097abfe..d70b0cefc6699066aabe468913ead5f0743120fe 100644
--- a/src/main/java/org/bukkit/map/MapRenderer.java
+++ b/src/main/java/org/bukkit/map/MapRenderer.java
@@ -61,5 +61,11 @@ public abstract class MapRenderer {
      * @return True if explorer map
      */
     public abstract boolean isExplorerMap();
-    // Purpur - end
+    // Purpur end -/- Canvas start
+
+    /**
+     * Sets the decorations(or structures) on the map as "dirty"
+     */
+    public abstract void setDecorationsDirty();
+    // Canvas end
 }
