From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Tue, 21 Nov 2023 19:14:53 +0000
Subject: [PATCH] Expose ReloadCount to Server


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 6692d5f89ff5a35e5631a70f0b44e6ac1db3deb1..91375f6797f68d6cce0ec18bc711d0c511e63973 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -923,6 +923,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      * Reloads the server, refreshing settings and plugin information.
      */
     public void reload();
+    // Canvas start
+
+    /**
+     * Gets the amount of times the server has reloaded("/reload confirm")
+     */
+    public int getReloadCount();
+    // Canvas end
 
     /**
      * Reload only the Minecraft data for the server. This includes custom
