From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Tue, 21 Nov 2023 19:14:53 +0000
Subject: [PATCH] Expose ReloadCount to Server


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 9de241ca3159f115a6a0d6d5650352b84d11b738..9e2901b395500788692b69ce74fd634bf7557bc0 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -931,6 +931,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      * Reloads the server, refreshing settings and plugin information.
      */
     public void reload();
+    // Canvas start
+
+    /**
+     * Gets the amount of times the server has reloaded("/reload confirm")
+     */
+    public int getReloadCount();
+    // Canvas end
 
     /**
      * Reload only the Minecraft data for the server. This includes custom
