From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Fri, 17 May 2024 20:30:55 -0700
Subject: [PATCH] Better Multithreaded WorldTicking


diff --git a/src/main/java/org/bukkit/command/defaults/ReloadCommand.java b/src/main/java/org/bukkit/command/defaults/ReloadCommand.java
index bdfe68b386b5ca2878475e548d3c9a3808fce848..f466325f9185a5e4468f05129647c1b8e895509b 100644
--- a/src/main/java/org/bukkit/command/defaults/ReloadCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/ReloadCommand.java
@@ -10,6 +10,7 @@ import org.bukkit.command.CommandSender;
 import org.jetbrains.annotations.NotNull;
 
 public class ReloadCommand extends BukkitCommand {
+    public static boolean isReloading = false; // Canvas
     public ReloadCommand(@NotNull String name) {
         super(name);
         this.description = "Reloads the server configuration and plugins";
@@ -56,7 +57,9 @@ public class ReloadCommand extends BukkitCommand {
         Command.broadcastCommandMessage(sender, ChatColor.RED + "If you encounter any issues please use the /stop command to restart your server.");
         // Paper start - lifecycle events
         try {
+            isReloading = true; // Canvas
             Bukkit.reload();
+            isReloading = false; // Canvas
         } catch (final IllegalStateException ex) {
             if (ex.getMessage().equals(RELOADING_DISABLED_MESSAGE)) {
                 Command.broadcastCommandMessage(sender, ChatColor.RED + RELOADING_DISABLED_MESSAGE);
