From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Wed, 22 Nov 2023 01:46:15 +0000
Subject: [PATCH] Add BiomeSource API


diff --git a/src/main/java/me/dueris/canvas/biome/BiomeSource.java b/src/main/java/me/dueris/canvas/biome/BiomeSource.java
new file mode 100644
index 0000000000000000000000000000000000000000..0e83053547cf2541ba5f28ffb4f14c0093e28c5e
--- /dev/null
+++ b/src/main/java/me/dueris/canvas/biome/BiomeSource.java
@@ -0,0 +1,43 @@
+package me.dueris.canvas.biome;
+
+import org.bukkit.NamespacedKey;
+import org.bukkit.Particle;
+import org.bukkit.Sound;
+import org.bukkit.World;
+
+import java.util.Optional;
+
+public interface BiomeSource {
+
+    public NamespacedKey getKey();
+
+	public float getTemperature();
+
+    public org.bukkit.block.Biome getBukkitBiome();
+
+    public float getDownfall();
+
+    public boolean hasPrecipitation();
+
+    public boolean isFrozen();
+
+    public int getSkyColor();
+
+    public int getGrassColor(World world, double x, double y, double z);
+
+    public int getFoliageColor();
+
+    public float getBaseTemperature();
+
+    public int getWaterColor();
+
+    public int getWaterFogColor();
+
+    public float getCreatureProbability();
+
+    public Optional<Sound> getBackgroundMusic();
+
+    public Optional<Particle> getAmbientParticle();
+
+    public Optional<Sound> getAmbientLoop();
+}
\ No newline at end of file
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 0b09b24c34a68cfc9ce0edd62a82c772566ae2f7..89fd9bf37d2d5204728101300e612f35a005dac2 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -857,6 +857,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @Deprecated // Paper
     @NotNull
     public ItemStack createExplorerMap(@NotNull World world, @NotNull Location location, @NotNull StructureType structureType);
+    // Canvas start
+
+    /**
+     * Gets the biome source from an org.bukkit Biome
+     */
+    @NotNull
+    public me.dueris.canvas.biome.BiomeSource getBiomeSource(org.bukkit.NamespacedKey key);
+    // Canvas end
 
     /**
      * Create a new explorer map targeting the closest nearby structure of a
diff --git a/src/main/java/org/bukkit/block/Biome.java b/src/main/java/org/bukkit/block/Biome.java
index 5d8fa5b39a5d50cca48ba63af3a84b80f279b649..ab3c4fea460a99ef69137b95b12dc5b5713a2ee3 100644
--- a/src/main/java/org/bukkit/block/Biome.java
+++ b/src/main/java/org/bukkit/block/Biome.java
@@ -96,4 +96,10 @@ public enum Biome implements Keyed, net.kyori.adventure.translation.Translatable
         return "biome.minecraft." + this.key.getKey();
     }
     // Paper end
+    // Canvas start - more biome api
+
+    public me.dueris.canvas.biome.BiomeSource getBiomeMetaSource(){
+        return org.bukkit.Bukkit.getServer().getBiomeSource(this.getKey());
+    }
+    // Canvas end
 }
