From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sat, 18 Nov 2023 11:10:46 -0800
Subject: [PATCH] WardenSonicBoomEvent


diff --git a/src/main/java/io/github/dueris/event/entity/WardenSonicBoomEvent.java b/src/main/java/io/github/dueris/event/entity/WardenSonicBoomEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..312a85349caaa212ed2c6477e91e35553d64bb14
--- /dev/null
+++ b/src/main/java/io/github/dueris/event/entity/WardenSonicBoomEvent.java
@@ -0,0 +1,49 @@
+package io.github.dueris.event.entity;
+
+import org.bukkit.entity.Warden;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.entity.EntityEvent;
+import org.jetbrains.annotations.NotNull;
+
+public class WardenSonicBoomEvent extends EntityEvent {
+    public static final HandlerList HANDLER_LIST = new HandlerList();
+    private Warden warden;
+    private boolean canceled;
+    private float damage = 10.0F;
+
+    public WardenSonicBoomEvent(final @NotNull Warden what) {
+        super(what);
+        this.warden = what;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return HANDLER_LIST;
+    }
+
+    @NotNull
+    @Override
+    public HandlerList getHandlers() {
+        return HANDLER_LIST;
+    }
+
+    public Warden getWarden() {
+        return warden;
+    }
+
+    public boolean isCanceled(){
+        return canceled;
+    }
+
+    public void setCanceled(boolean canceled){
+        this.canceled = canceled;
+    }
+
+    public float getDamage(){
+        return damage;
+    }
+
+    public void setDamage(float damage){
+        this.damage = damage;
+    }
+}
