From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Wed, 22 Nov 2023 08:39:49 +0000
Subject: [PATCH] Add ModelPart API


diff --git a/src/main/java/me/dueris/canvas/entity/player/PlayerModelPart.java b/src/main/java/me/dueris/canvas/entity/player/PlayerModelPart.java
new file mode 100644
index 0000000000000000000000000000000000000000..a4e816378940fb2357533bd649316d368c0d5b37
--- /dev/null
+++ b/src/main/java/me/dueris/canvas/entity/player/PlayerModelPart.java
@@ -0,0 +1,21 @@
+package me.dueris.canvas.entity.player;
+
+public enum PlayerModelPart {
+    CAPE(0),
+    JACKET(1),
+    LEFT_SLEEVE(2),
+    RIGHT_SLEEVE(3),
+    LEFT_PANTS_LEG(4),
+    RIGHT_PANTS_LEG(5),
+    HAT(6);
+
+    int id;
+
+    private PlayerModelPart(int id) {
+        this.id = id;
+    }
+
+    public int getId(){
+        return this.id;
+    }
+}
\ No newline at end of file
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index cc725fa4cd547df6008aa79f6dfd85ad6ea11f1c..149cf28d60391fd817c0313e3ea517dd1f5e99a4 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -120,6 +120,13 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     public me.dueris.canvas.entity.Abilities getAbilities();
 
+    // ModelPartAPI start
+
+    /**
+     * Checks if the player has that modelpart visible or not
+     */
+    public boolean hasModelPartShown(me.dueris.canvas.entity.player.PlayerModelPart modelPart);
+    // ModelPartAPI end
     // Canvas end
 
     /**
