From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Wed, 22 Nov 2023 09:00:22 +0000
Subject: [PATCH] Expose CreateEndPlatform API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 9d436d2bb4eb69324adfba7332ca39a25ce5403c..6132a547de09e6cde76f858d4fe471f0b2e0c653 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3525,6 +3525,21 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * @param pitch the pitch
      */
     void setRotation(float yaw, float pitch);
+    // Canvas start
+
+    /**
+     * Creates an end spawn platform at the current player position
+     * Doesnt call the PortalCreationEvnet
+     */
+    public void createEndPlatformAtCurrentPos();
+
+    /**
+     * Creates an end spawn platform at the current player position
+     * 
+     * @param callEvent decides weither to call the PortalCreationEvent
+     */
+    public void createEndPlatformAtCurrentPos(boolean callEvent);
+    // Canvas end
 
     /**
      * Causes the player to look towards the given position.
