From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sun, 19 Nov 2023 12:25:47 -0800
Subject: [PATCH] Allow TIPPED_ARROW effect durations to be modified


diff --git a/src/main/java/org/bukkit/entity/Arrow.java b/src/main/java/org/bukkit/entity/Arrow.java
index 8e57082e8b19a2fd9625816c2ea4a90d41f78ca5..9e92e1c6ebdde02c468422b6fcdcf44a64d323a5 100644
--- a/src/main/java/org/bukkit/entity/Arrow.java
+++ b/src/main/java/org/bukkit/entity/Arrow.java
@@ -114,4 +114,28 @@ public interface Arrow extends AbstractArrow {
      * in a state with no Custom Effects and PotionType.UNCRAFTABLE
      */
     void clearCustomEffects();
+    // Canvas start
+
+    /**
+     * Sets the divisor that calculates TIPPED_ARROW effect durations
+     * DEFAULT = 8;
+     * Lower the number, longer the duration. Larger the number, shorter the duration.
+     *
+     * THE FINAL DURATION IS ROUNDED TO THE NEAREST HOLE NUMBER!
+     *
+     * @param divisor
+     */
+    public void setEffectDurationDivisor(float divisor);
+
+    /**
+     * Gets the divisor that calculates TIPPED_ARROW effect durations
+     * DEFAULT = 8;
+     * Lower the number, longer the duration. Larger the number, shorter the duration.
+     *
+     * THE FINAL DURATION IS ROUNDED TO THE NEAREST HOLE NUMBER!
+     *
+     * @return the divisor for TIPPED_ARROW effect durations
+     */
+    public float getEffectDurationDivisor();
+    // Canvas end
 }
