From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Wed, 22 Nov 2023 09:00:22 +0000
Subject: [PATCH] Expose CreateEndPlatform API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index c0b7e617163cbdf738d733dbddf6973255aded6c..64ffd5cc0f1fd59012107753f764c03d2b31bc53 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3486,6 +3486,21 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * @param pitch the pitch
      */
     void setRotation(float yaw, float pitch);
+    // Canvas start
+
+    /**
+     * Creates an end spawn platform at the current player position
+     * Doesnt call the PortalCreationEvnet
+     */
+    public void createEndPlatformAtCurrentPos();
+
+    /**
+     * Creates an end spawn platform at the current player position
+     * 
+     * @param callEvent decides weither to call the PortalCreationEvent
+     */
+    public void createEndPlatformAtCurrentPos(boolean callEvent);
+    // Canvas end
 
     /**
      * Causes the player to look towards the given position.
