From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Wed, 22 Nov 2023 09:00:22 +0000
Subject: [PATCH] Expose CreateEndPlatform API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 42c9be81309ac0af6d1679bb5ecd83f405bdf396..216c16436c5e19bdc12e77aad44cbcce5d1c9f95 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3435,6 +3435,21 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * @param pitch the pitch
      */
     void setRotation(float yaw, float pitch);
+    // Canvas start
+
+    /**
+     * Creates an end spawn platform at the current player position
+     * Doesnt call the PortalCreationEvnet
+     */
+    public void createEndPlatformAtCurrentPos();
+
+    /**
+     * Creates an end spawn platform at the current player position
+     * 
+     * @param callEvent decides weither to call the PortalCreationEvent
+     */
+    public void createEndPlatformAtCurrentPos(boolean callEvent);
+    // Canvas end
 
     /**
      * Causes the player to look towards the given position.
