From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Wed, 22 Nov 2023 09:00:22 +0000
Subject: [PATCH] Expose CreateEndPlatform API


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index fa3fc12e06f1434aac6b4412e5163698b7728611..b0db46189a0bceaec642e41fc45cdf5f41500e23 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -3365,6 +3365,21 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      * @param pitch the pitch
      */
     void setRotation(float yaw, float pitch);
+    // Canvas start
+
+    /**
+     * Creates an end spawn platform at the current player position
+     * Doesnt call the PortalCreationEvnet
+     */
+    public void createEndPlatformAtCurrentPos();
+
+    /**
+     * Creates an end spawn platform at the current player position
+     * 
+     * @param callEvent decides weither to call the PortalCreationEvent
+     */
+    public void createEndPlatformAtCurrentPos(boolean callEvent);
+    // Canvas end
 
     /**
      * Causes the player to look towards the given position.
