From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sun, 19 Nov 2023 16:39:22 -0800
Subject: [PATCH] Minor bossbar api addon


diff --git a/src/main/java/org/bukkit/boss/KeyedBossBar.java b/src/main/java/org/bukkit/boss/KeyedBossBar.java
index 6a1fe5f9160c758c22f53bf967d2181373c0a148..0cbb9e5f9753fca126ab145d9b6d7c55c8d13fce 100644
--- a/src/main/java/org/bukkit/boss/KeyedBossBar.java
+++ b/src/main/java/org/bukkit/boss/KeyedBossBar.java
@@ -6,4 +6,12 @@ import org.bukkit.Keyed;
  * Represents a custom {@link BossBar} that has a
  * {@link org.bukkit.NamespacedKey}
  */
-public interface KeyedBossBar extends BossBar, Keyed { }
+public interface KeyedBossBar extends BossBar, Keyed {
+    // Canvas start
+
+    /**
+     * Direct and safer access to the Bukkit.removeBossBar()
+     */
+    public void deleteBar();
+    // Canvas end
+}
