From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Mon, 20 Nov 2023 19:03:36 +0000
Subject: [PATCH] Add PlayerDisableShieldEvent


diff --git a/src/main/java/io/github/dueris/event/player/PlayerDisableShieldEvent.java b/src/main/java/io/github/dueris/event/player/PlayerDisableShieldEvent.java
new file mode 100644
index 0000000000000000000000000000000000000000..3fcdd0db77901363b2ae7eae40932d676e731f2c
--- /dev/null
+++ b/src/main/java/io/github/dueris/event/player/PlayerDisableShieldEvent.java
@@ -0,0 +1,52 @@
+package io.github.dueris.event.player;
+
+import org.bukkit.entity.Player;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.player.PlayerEvent;
+import org.jetbrains.annotations.NotNull;
+
+public class PlayerDisableShieldEvent extends org.bukkit.event.player.PlayerEvent{
+    public static final HandlerList HANDLER_LIST = new HandlerList();
+    public Player player;
+    public boolean canceled;
+    public int cooldown;
+    public org.bukkit.entity.LivingEntity attacker;
+
+    public PlayerDisableShieldEvent(final @NotNull Player player, int cooldown, org.bukkit.entity.LivingEntity attacker) {
+        super(player);
+        this.player = player;
+        this.cooldown = cooldown;
+        this.attacker = attacker;
+    }
+
+    @NotNull
+    public static HandlerList getHandlerList() {
+        return HANDLER_LIST;
+    }
+
+    @NotNull
+    @Override
+    public HandlerList getHandlers() {
+        return HANDLER_LIST;
+    }
+
+    public void setCanceled(boolean canceled){
+        this.canceled = canceled;
+    }
+
+    public boolean getCanceled(){
+        return canceled;
+    }
+
+    public int getCooldown(){
+        return this.cooldown;
+    }
+
+    public void setCooldown(int cooldown){
+        this.cooldown = cooldown;
+    }
+
+    public org.bukkit.entity.LivingEntity getAttacker(){
+        return this.attacker;
+    }
+}
