From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Wed, 29 Nov 2023 18:18:46 +0000
Subject: [PATCH] Add more World-Getter api


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index fc7807a706effe05ddae2462e11f25e5b395c1f9..f658c5dcc463bcdbbd64b14ec71f2522fda0c6ce 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -287,6 +287,26 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     @NotNull
     public String getWorldType();
+    // Canvas start
+
+    public java.util.ArrayList<World> getVanillaWorlds();
+
+    public java.util.ArrayList<World> getCustomWorlds();
+
+    public World getCustomWorld(int index);
+
+    public World getCustomWorld(org.bukkit.NamespacedKey key);
+
+    public World getOverworld();
+
+    public World getNether();
+
+    public World getEnd();
+
+    public World getVanillaWorld(int index);
+
+    public World getVanillaWorld(org.bukkit.NamespacedKey key);
+    // Canvas end
 
     /**
      * Get generate-structures setting.
