From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Tue, 21 Nov 2023 19:14:53 +0000
Subject: [PATCH] Add getReloadCount() to Server


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 2551e60e19cf9720892bc87075667e71f9b185d4..79e074ed85ec4c8b26af7d5f3db1b8eaf3113db1 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -956,6 +956,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      * Reloads the server, refreshing settings and plugin information.
      */
     public void reload();
+    // Canvas start
+
+    /**
+     * Gets the amount of times the server has reloaded("/reload confirm")
+     */
+    public int getReloadCount();
+    // Canvas end
 
     /**
      * Reload only the Minecraft data for the server. This includes custom
