From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Wed, 22 Nov 2023 01:46:15 +0000
Subject: [PATCH] Add BiomeSource API


diff --git a/src/main/java/io/github/dueris/biome/BiomeSource.java b/src/main/java/io/github/dueris/biome/BiomeSource.java
new file mode 100644
index 0000000000000000000000000000000000000000..dfa12006bcdd3371ccf43cc820b67ff262456678
--- /dev/null
+++ b/src/main/java/io/github/dueris/biome/BiomeSource.java
@@ -0,0 +1,43 @@
+package io.github.dueris.biome;
+
+import org.bukkit.NamespacedKey;
+import org.bukkit.Particle;
+import org.bukkit.Sound;
+import org.bukkit.World;
+
+import java.util.Optional;
+
+public interface BiomeSource {
+
+    public NamespacedKey getKey();
+
+	public float getTemperature();
+
+    public org.bukkit.block.Biome getBukkitBiome();
+
+    public float getDownfall();
+
+    public boolean hasPrecipitation();
+
+    public boolean isFrozen();
+
+    public int getSkyColor();
+
+    public int getGrassColor(World world, double x, double y, double z);
+
+    public int getFoliageColor();
+
+    public float getBaseTemperature();
+
+    public int getWaterColor();
+
+    public int getWaterFogColor();
+
+    public float getCreatureProbability();
+
+    public Optional<Sound> getBackgroundMusic();
+
+    public Optional<Particle> getAmbientParticle();
+
+    public Optional<Sound> getAmbientLoop();
+}
\ No newline at end of file
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index dec2e7110cd30f1fc09915983dc8a0f91db2106d..147dd88bc634efe223d0fe285d67676c2bb04e82 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -892,6 +892,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @Deprecated // Paper
     @NotNull
     public ItemStack createExplorerMap(@NotNull World world, @NotNull Location location, @NotNull StructureType structureType);
+    // Canvas start
+
+    /**
+     * Gets the biome source from an org.bukkit Biome
+     */
+    @NotNull
+    public io.github.dueris.biome.BiomeSource getBiomeSource(org.bukkit.NamespacedKey key);
+    // Canvas end
 
     /**
      * Create a new explorer map targeting the closest nearby structure of a
diff --git a/src/main/java/org/bukkit/block/Biome.java b/src/main/java/org/bukkit/block/Biome.java
index 5d8fa5b39a5d50cca48ba63af3a84b80f279b649..4e7ebf84c3f3f9cdaa2ccea835e3f0350dc3c0af 100644
--- a/src/main/java/org/bukkit/block/Biome.java
+++ b/src/main/java/org/bukkit/block/Biome.java
@@ -96,4 +96,10 @@ public enum Biome implements Keyed, net.kyori.adventure.translation.Translatable
         return "biome.minecraft." + this.key.getKey();
     }
     // Paper end
+    // Canvas start - more biome api
+
+    public io.github.dueris.biome.BiomeSource getBiomeMetaSource(){
+        return org.bukkit.Bukkit.getServer().getBiomeSource(this.getKey());
+    }
+    // Canvas end
 }
