From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Wed, 29 Nov 2023 18:18:46 +0000
Subject: [PATCH] Add more World-Getter api


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 89fd9bf37d2d5204728101300e612f35a005dac2..bcb13a0562cd5a03bd5f3b63aacbe5c62ad94c06 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -284,6 +284,26 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     @NotNull
     public String getWorldType();
+    // Canvas start
+
+    public java.util.ArrayList<World> getVanillaWorlds();
+
+    public java.util.ArrayList<World> getCustomWorlds();
+
+    public World getCustomWorld(int index);
+
+    public World getCustomWorld(org.bukkit.NamespacedKey key);
+
+    public World getOverworld();
+
+    public World getNether();
+
+    public World getEnd();
+
+    public World getVanillaWorld(int index);
+
+    public World getVanillaWorld(org.bukkit.NamespacedKey key);
+    // Canvas end
 
     /**
      * Get generate-structures setting.
