From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Wed, 22 Nov 2023 08:39:49 +0000
Subject: [PATCH] Add ModelPart API


diff --git a/src/main/java/io/github/dueris/entity/player/PlayerModelPart.java b/src/main/java/io/github/dueris/entity/player/PlayerModelPart.java
new file mode 100644
index 0000000000000000000000000000000000000000..1801fd7538570d2df2feecfaeb2980903e182806
--- /dev/null
+++ b/src/main/java/io/github/dueris/entity/player/PlayerModelPart.java
@@ -0,0 +1,21 @@
+package io.github.dueris.entity.player;
+
+public enum PlayerModelPart {
+    CAPE(0),
+    JACKET(1),
+    LEFT_SLEEVE(2),
+    RIGHT_SLEEVE(3),
+    LEFT_PANTS_LEG(4),
+    RIGHT_PANTS_LEG(5),
+    HAT(6);
+
+    int id;
+
+    private PlayerModelPart(int id) {
+        this.id = id;
+    }
+
+    public int getId(){
+        return this.id;
+    }
+}
\ No newline at end of file
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index ba7f8f43f18a9a4012293ea779a832c919d00dfd..0ce1162f08ae1740037d649250a43098506b0dd6 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -134,6 +134,13 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     public io.github.dueris.entity.Abilities getAbilities();
 
+    // ModelPartAPI start
+
+    /**
+     * Checks if the player has that modelpart visible or not
+     */
+    public boolean hasModelPartShown(io.github.dueris.entity.player.PlayerModelPart modelPart);
+    // ModelPartAPI end
     // Canvas end
 
     /**
