From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sat, 25 Nov 2023 12:36:56 -0800
Subject: [PATCH] Create DamageSource api


diff --git a/src/main/java/me/dueris/canvas/world/damage/DamageSource.java b/src/main/java/me/dueris/canvas/world/damage/DamageSource.java
new file mode 100644
index 0000000000000000000000000000000000000000..4daa16f8099d836c70371f8c9d4215fdb5eace73
--- /dev/null
+++ b/src/main/java/me/dueris/canvas/world/damage/DamageSource.java
@@ -0,0 +1,3 @@
+package me.dueris.canvas.world.damage;
+
+public interface DamageSource { }
diff --git a/src/main/java/me/dueris/canvas/world/damage/DamageSources.java b/src/main/java/me/dueris/canvas/world/damage/DamageSources.java
new file mode 100644
index 0000000000000000000000000000000000000000..f08632c31bebc63828dc456b95bfd46f95d020ee
--- /dev/null
+++ b/src/main/java/me/dueris/canvas/world/damage/DamageSources.java
@@ -0,0 +1,103 @@
+package me.dueris.canvas.world.damage;
+
+import org.bukkit.entity.AbstractArrow;
+import org.bukkit.entity.Entity;
+import org.bukkit.entity.Explosive;
+import org.bukkit.entity.Fireball;
+import org.bukkit.entity.Firework;
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.entity.Player;
+import org.bukkit.entity.WitherSkull;
+import org.bukkit.util.Vector;
+import javax.annotation.Nullable;
+
+public interface DamageSources {
+    public DamageSource sweetBerryBush();
+    
+    public DamageSource inFire();
+
+    public DamageSource lightningBolt();
+    
+    public DamageSource onFire();
+    
+    public DamageSource lava();
+    
+    public DamageSource hotFloor();
+    
+    public DamageSource inWall();
+    
+    public DamageSource cramming();
+    
+    public DamageSource drown();
+    
+    public DamageSource starve();
+    
+    public DamageSource cactus();
+    
+    public DamageSource fall();
+    
+    public DamageSource flyIntoWall();
+    
+    public DamageSource fellOutOfWorld();
+    
+    public DamageSource generic();
+    
+    public DamageSource magic();
+    
+    public DamageSource wither();
+    
+    public DamageSource dragonBreath();
+    
+    public DamageSource dryOut();
+    
+    public DamageSource freeze();
+    
+    public DamageSource stalagmite();
+    
+    public DamageSource fallingBlock(Entity attacker);
+    
+    public DamageSource anvil(Entity attacker);
+    
+    public DamageSource fallingStalactite(Entity attacker);
+    
+    public DamageSource sting(LivingEntity attacker);
+    
+    public DamageSource mobAttack(LivingEntity attacker);
+    
+    public DamageSource noAggroMobAttack(LivingEntity attacker);
+    
+    public DamageSource playerAttack(Player attacker);
+    
+    public DamageSource arrow(AbstractArrow source, @Nullable Entity attacker);
+    
+    public DamageSource trident(Entity source, @Nullable Entity attacker);
+    
+    public DamageSource mobProjectile(Entity source, @Nullable LivingEntity attacker);
+    
+    public DamageSource fireworks(Firework source, @Nullable Entity attacker);
+    
+    public DamageSource fireball(Fireball source, @Nullable Entity attacker);
+    
+    public DamageSource witherSkull(WitherSkull source, Entity attacker);
+    
+    public DamageSource thrown(Entity source, @Nullable Entity attacker);
+    
+    public DamageSource indirectMagic(Entity source, @Nullable Entity attacker);
+    
+    public DamageSource thorns(Entity attacker);
+    
+    public DamageSource explosion(@Nullable Entity source, @Nullable Entity attacker);
+    
+    public DamageSource sonicBoom(Entity attacker);
+
+    @Deprecated
+    @io.papermc.paper.annotation.DoNotUse
+    public DamageSource badRespawnPointExplosion(Vector position);
+    
+    public DamageSource badRespawnPointExplosion(Vector position, @Nullable org.bukkit.block.BlockState explodedBlockState);
+    
+    public DamageSource outOfBorder();
+    
+    public DamageSource genericKill();
+
+}
diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 23fadc701daded2d102dc87f13ed6d3f62cabe40..4382fd65c12c69a383b4b2b3b2631c8aafec8bac 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -311,6 +311,14 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
      */
     @NotNull
     public Block getHighestBlockAt(@NotNull Location location, @NotNull HeightMap heightMap);
+    // Canvas start
+
+    /**
+     * Gets the DamageSources for the world
+     * @return damagesources for the world
+     */
+    public me.dueris.canvas.world.damage.DamageSources damagesources();
+    // Canvas end
 
     /**
      * Gets the {@link Chunk} at the given coordinates
