From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <159172167+PurpleWolfMC@users.noreply.github.com>
Date: Wed, 7 Feb 2024 07:05:39 +0000
Subject: [PATCH] Allow setting respawn Dimension


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index fd3bf00bc3ff6fb4854579f0396b78262a6770b3..c10ac1c84742c3f1afe46c867f6ba9584317fde0 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -553,7 +553,13 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     @Override
     @Deprecated
     public Location getBedSpawnLocation();
+    // Canvas start
+
+    public void setRespawnDimension(org.bukkit.World world);
 
+    @NotNull
+    public org.bukkit.World getRespawnDimension();
+    // Canvas end
     /**
      * Gets the Location where the player will spawn at, null if they
      * don't have a valid respawn point.
