From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Sat, 6 Jan 2024 03:59:18 +0000
Subject: [PATCH] Allow setting the respawn dimension of the player


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index c44a3bb6d5a87cd7d3c82b14a656f3d15cacbf0e..febf3074b80d22dbed219ddf6b0b384ad159e334 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -548,6 +548,13 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     @Nullable
     @Override
     public Location getBedSpawnLocation();
+    // Canvas start
+
+    public void setRespawnDimension(org.bukkit.World world);
+
+    @NotNull
+    public org.bukkit.World getRespawnDimension();
+    // Canvas end
 
     /**
      * Sets the Location where the player will spawn at their bed.
