From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Thu, 30 Nov 2023 00:18:23 +0000
Subject: [PATCH] Remove ChatChannel API


diff --git a/src/main/java/me/dueris/canvas/network/ChatChannelContainer.java b/src/main/java/me/dueris/canvas/network/ChatChannelContainer.java
deleted file mode 100644
index 669151590c90398f40122b90907367e21e9d5e44..0000000000000000000000000000000000000000
--- a/src/main/java/me/dueris/canvas/network/ChatChannelContainer.java
+++ /dev/null
@@ -1,18 +0,0 @@
-package me.dueris.canvas.network;
-
-import me.dueris.canvas.network.channels.ChatChannel;
-import org.bukkit.NamespacedKey;
-import java.util.ArrayList;
-
-public interface ChatChannelContainer {
-
-    public ChatChannel createChannel(NamespacedKey namespacedKey, String title);
-
-    public void deleteChannel(ChatChannel chatChannel);
-
-    public ChatChannel getChatChannel(NamespacedKey namespacedKey);
-
-    public ArrayList<ChatChannel> getChannels();
-
-    public boolean containsChannel(ChatChannel chatChannel);
-}
diff --git a/src/main/java/me/dueris/canvas/network/channels/ChannelSettings.java b/src/main/java/me/dueris/canvas/network/channels/ChannelSettings.java
deleted file mode 100644
index f1b49275bf8e8e8780217a94412adffa0ce01f68..0000000000000000000000000000000000000000
--- a/src/main/java/me/dueris/canvas/network/channels/ChannelSettings.java
+++ /dev/null
@@ -1,43 +0,0 @@
-package me.dueris.canvas.network.channels;
-
-import org.bukkit.scheduler.BukkitRunnable;
-import org.bukkit.scheduler.BukkitTask;
-
-public interface ChannelSettings {
-
-    public void canViewOtherChannelLogs(boolean T);
-
-    /**
-     * Should publicly announce advancement messages or stay private inside the channel
-     * @return
-     */
-    public void sendAdvancements(boolean T);
-
-    /**
-     * Should publicly announce death messages or stay private inside the channel
-     * @return
-     */
-    public void sendDeathMessages(boolean T);
-
-    public void privateChannel(boolean T);
-
-    public void resetSettings();
-
-    public boolean isPrivateChannel();
-
-    /**
-     * Should publicly announce death messages or stay private inside the channel
-     * @return
-     */
-    public boolean canSendDeathMessages();
-
-    /**
-     * Should publicly announce advancement messages or stay private inside the channel
-     * @return
-     */
-    public boolean canSendAdvancements();
-
-    public boolean canViewOtherChannelLogs();
-
-    public boolean canReceiveWhisperFromOtherChannelMembers();
-}
diff --git a/src/main/java/me/dueris/canvas/network/channels/ChatChannel.java b/src/main/java/me/dueris/canvas/network/channels/ChatChannel.java
deleted file mode 100644
index c092a54f18dd725fbb6b137ed26fbf13979c2e81..0000000000000000000000000000000000000000
--- a/src/main/java/me/dueris/canvas/network/channels/ChatChannel.java
+++ /dev/null
@@ -1,34 +0,0 @@
-package me.dueris.canvas.network.channels;
-
-import org.bukkit.NamespacedKey;
-import org.bukkit.entity.Player;
-import java.util.ArrayList;
-
-public interface ChatChannel {
-
-    public ArrayList<Player> getPlayers();
-
-    public void addPlayer(Player player);
-
-    public void removePlayer(Player player);
-
-    public void changeNamespace(NamespacedKey namespacedKey);
-
-    public void sendMessage(String msg);
-
-    public void sendMessages(String[] msgs);
-
-    public void executeCommand(String cmd);
-
-    public void executeCommands(String[] cmds);
-
-    public boolean containsPlayer(Player player);
-
-    public NamespacedKey getChannelKey();
-
-    public String getChannelName();
-
-    public void changeName(String name);
-
-    public ChannelSettings getChannelSettings();
-}
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 65a0c4af8b3260d34ac228728a17438c60de9558..c9d8ea02dc9798574244edc70148e82ebeae85a8 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -3,7 +3,6 @@ package org.bukkit;
 import me.dueris.canvas.CraftServerGui;
 import me.dueris.canvas.logger.CanvasLogger;
 import me.dueris.canvas.logger.Obfuscator;
-import me.dueris.canvas.network.ChatChannelContainer;
 import com.google.common.collect.ImmutableList;
 import java.awt.image.BufferedImage;
 import java.io.File;
@@ -147,12 +146,6 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     @NotNull
     public CraftServerGui getServerGui();
-
-    /**
-     * @return ChatChannel manager
-     */
-    @NotNull
-    public ChatChannelContainer getChatChannelManager();
     
     /**
      * Retrieves the total memory size.
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index e5446b46f0aac6da885e62a0c97d1c370fba811b..7f7fb1ab11a4bde5ba0a5fbd619a186ea82a06ff 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -7,7 +7,6 @@ import java.time.Instant;
 import java.util.Collection;
 import java.util.Date;
 import java.util.Map;
-import me.dueris.canvas.network.channels.ChatChannel;
 import org.bukkit.BanEntry;
 import org.bukkit.Chunk;
 import org.bukkit.DyeColor;
@@ -130,22 +129,6 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
 
     // Canvas start
 
-    /**
-     * Retrives the chat channel the player is currently in
-     * 
-     * IF it returns null, the player is not inside a ChatChannel
-     * 
-     * @return the current chat channel the player is in
-     */
-    public ChatChannel getChatChannel();
-
-    /**
-     * Checks if the player is in a "chat channel"
-     * 
-     * @return the player is in a chat channel
-     */
-    public boolean isInChatChannel();
-
     /**
      * Retrives the chunks within the render distance of the player
      */
diff --git a/src/main/java/org/bukkit/event/player/AsyncPlayerChatEvent.java b/src/main/java/org/bukkit/event/player/AsyncPlayerChatEvent.java
index 96c36262578e4f596d0f56fd9d4da9f1b32d30cf..762671e6c203baea1090780690dd75296e849baa 100644
--- a/src/main/java/org/bukkit/event/player/AsyncPlayerChatEvent.java
+++ b/src/main/java/org/bukkit/event/player/AsyncPlayerChatEvent.java
@@ -44,17 +44,6 @@ public class AsyncPlayerChatEvent extends PlayerEvent implements Cancellable {
     public AsyncPlayerChatEvent(final boolean async, @NotNull final Player who, @NotNull final String message, @NotNull final Set<Player> players) {
         super(who, async);
         this.message = message;
-        // Canvas start
-        for (Player pl : players) {
-            if (pl.isInChatChannel()) {
-                if (pl.getChatChannel() != who.getChatChannel()) {
-                    if (!pl.getChatChannel().getChannelSettings().canViewOtherChannelLogs() || who.getChatChannel().getChannelSettings().isPrivateChannel()) {
-                        players.remove(pl);
-                    }
-                }
-            }
-        }
-        // Canvas end
 
         recipients = players;
     }
