From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sat, 7 Oct 2023 13:40:54 -0700
Subject: [PATCH] added mixin api access


diff --git a/build.gradle.kts b/build.gradle.kts
index 9ee4cc8afc99dd4da9f76bb7fb6740f5e715c1cc..252d637065e94cadc925082ec380121b61df7d38 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -44,6 +44,7 @@ dependencies {
     apiAndDocs("net.kyori:adventure-text-serializer-legacy")
     apiAndDocs("net.kyori:adventure-text-serializer-plain")
     apiAndDocs("net.kyori:adventure-text-logger-slf4j")
+    apiAndDocs("org.spongepowered:mixin:0.8.5") // Canvas - add mixin api
     api("org.apache.logging.log4j:log4j-api:$log4jVersion")
     api("org.slf4j:slf4j-api:$slf4jVersion")
     api("io.sentry:sentry:5.4.0") // Pufferfish
