From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sun, 15 Oct 2023 17:49:50 -0700
Subject: [PATCH] Create ChatChannel api


diff --git a/src/main/java/com/dueris/canvas/network/ChatChannelContainer.java b/src/main/java/com/dueris/canvas/network/ChatChannelContainer.java
new file mode 100644
index 0000000000000000000000000000000000000000..a11a624bd9f5c72b8e60f61a6900b742ce23be70
--- /dev/null
+++ b/src/main/java/com/dueris/canvas/network/ChatChannelContainer.java
@@ -0,0 +1,18 @@
+package com.dueris.canvas.network;
+
+import com.dueris.canvas.network.channels.ChatChannel;
+import org.bukkit.NamespacedKey;
+import java.util.ArrayList;
+
+public interface ChatChannelContainer {
+
+    public ChatChannel createChannel(NamespacedKey namespacedKey, String title);
+
+    public void deleteChannel(ChatChannel chatChannel);
+
+    public ChatChannel getChatChannel(NamespacedKey namespacedKey);
+
+    public ArrayList<ChatChannel> getChannels();
+
+    public boolean containsChannel(ChatChannel chatChannel);
+}
diff --git a/src/main/java/com/dueris/canvas/network/channels/ChannelSettings.java b/src/main/java/com/dueris/canvas/network/channels/ChannelSettings.java
new file mode 100644
index 0000000000000000000000000000000000000000..fedc847f4db84308f9213463e90ef7c75c534f64
--- /dev/null
+++ b/src/main/java/com/dueris/canvas/network/channels/ChannelSettings.java
@@ -0,0 +1,43 @@
+package com.dueris.canvas.network.channels;
+
+import org.bukkit.scheduler.BukkitRunnable;
+import org.bukkit.scheduler.BukkitTask;
+
+public interface ChannelSettings {
+
+    public void canViewOtherChannelLogs(boolean T);
+
+    /**
+     * Should publicly announce advancement messages or stay private inside the channel
+     * @return
+     */
+    public void sendAdvancements(boolean T);
+
+    /**
+     * Should publicly announce death messages or stay private inside the channel
+     * @return
+     */
+    public void sendDeathMessages(boolean T);
+
+    public void privateChannel(boolean T);
+
+    public void resetSettings();
+
+    public boolean isPrivateChannel();
+
+    /**
+     * Should publicly announce death messages or stay private inside the channel
+     * @return
+     */
+    public boolean canSendDeathMessages();
+
+    /**
+     * Should publicly announce advancement messages or stay private inside the channel
+     * @return
+     */
+    public boolean canSendAdvancements();
+
+    public boolean canViewOtherChannelLogs();
+
+    public boolean canReceiveWhisperFromOtherChannelMembers();
+}
diff --git a/src/main/java/com/dueris/canvas/network/channels/ChatChannel.java b/src/main/java/com/dueris/canvas/network/channels/ChatChannel.java
new file mode 100644
index 0000000000000000000000000000000000000000..94681824ccb5b9ca80528fc5df3057a47fc9888c
--- /dev/null
+++ b/src/main/java/com/dueris/canvas/network/channels/ChatChannel.java
@@ -0,0 +1,34 @@
+package com.dueris.canvas.network.channels;
+
+import org.bukkit.NamespacedKey;
+import org.bukkit.entity.Player;
+import java.util.ArrayList;
+
+public interface ChatChannel {
+
+    public ArrayList<Player> getPlayers();
+
+    public void addPlayer(Player player);
+
+    public void removePlayer(Player player);
+
+    public void changeNamespace(NamespacedKey namespacedKey);
+
+    public void sendMessage(String msg);
+
+    public void sendMessages(String[] msgs);
+
+    public void executeCommand(String cmd);
+
+    public void executeCommands(String[] cmds);
+
+    public boolean containsPlayer(Player player);
+
+    public NamespacedKey getChannelKey();
+
+    public String getChannelName();
+
+    public void changeName(String name);
+
+    public ChannelSettings getChannelSettings();
+}
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index cded9762f8bff025f49644c062f67ea5301e2038..47d9180ed8e7969c5d9eaa3264a15626a36b85e8 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -3,6 +3,7 @@ package org.bukkit;
 import com.dueris.canvas.CraftServerGui;
 import com.dueris.canvas.logger.CanvasLogger;
 import com.dueris.canvas.logger.Obfuscator;
+import com.dueris.canvas.network.ChatChannelContainer;
 import com.google.common.collect.ImmutableList;
 import java.awt.image.BufferedImage;
 import java.io.File;
@@ -123,6 +124,9 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull
     public CraftServerGui getServerGui();
 
+    @NotNull
+    public ChatChannelContainer getChatChannelManager();
+
     // Canvas end
 
     /**
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 0df709ed758401f04b3f565dc1f6492a55f5363c..330282040fc40eec72c3a63c0735a94b3de839ad 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -7,6 +7,7 @@ import java.time.Instant;
 import java.util.Collection;
 import java.util.Date;
 import java.util.Map;
+import com.dueris.canvas.network.channels.ChatChannel;
 import org.bukkit.BanEntry;
 import org.bukkit.DyeColor;
 import org.bukkit.Effect;
@@ -94,6 +95,14 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     @Override
     public String getName();
 
+    // Canvas start
+
+    public ChatChannel getChatChannel();
+
+    public boolean isInChatChannel();
+
+    // Canvas end
+
     /**
      * Gets the "friendly" name to display of this player. This may include
      * color.
diff --git a/src/main/java/org/bukkit/event/player/AsyncPlayerChatEvent.java b/src/main/java/org/bukkit/event/player/AsyncPlayerChatEvent.java
index 399afcd19fcb6acd24857ed6ab48cf0d105a01a3..f4d3d9c1e470a24342bf1c932a68582a5c1e983e 100644
--- a/src/main/java/org/bukkit/event/player/AsyncPlayerChatEvent.java
+++ b/src/main/java/org/bukkit/event/player/AsyncPlayerChatEvent.java
@@ -2,6 +2,9 @@ package org.bukkit.event.player;
 
 import java.util.IllegalFormatException;
 import java.util.Set;
+import com.dueris.canvas.network.ChatChannelContainer;
+import com.dueris.canvas.network.channels.ChatChannel;
+import org.bukkit.Bukkit;
 import org.bukkit.entity.Player;
 import org.bukkit.event.Cancellable;
 import org.bukkit.event.HandlerList;
@@ -32,7 +35,7 @@ public class AsyncPlayerChatEvent extends PlayerEvent implements Cancellable {
     private boolean cancel = false;
     private String message;
     private String format = "<%1$s> %2$s";
-    private final Set<Player> recipients;
+    private Set<Player> recipients; // Canvas
 
     /**
      * @param async This changes the event to a synchronous state.
@@ -44,6 +47,18 @@ public class AsyncPlayerChatEvent extends PlayerEvent implements Cancellable {
     public AsyncPlayerChatEvent(final boolean async, @NotNull final Player who, @NotNull final String message, @NotNull final Set<Player> players) {
         super(who, async);
         this.message = message;
+        // Canvas start
+        for (Player pl : players) {
+            if (pl.isInChatChannel()) {
+                if (pl.getChatChannel() != who.getChatChannel()) {
+                    if (!pl.getChatChannel().getChannelSettings().canViewOtherChannelLogs() || who.getChatChannel().getChannelSettings().isPrivateChannel()) {
+                        players.remove(pl);
+                    }
+                }
+            }
+        }
+        // Canvas end
+
         recipients = players;
     }
 
