From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris Bauer <jedimastertoothless@hotmail.com>
Date: Wed, 25 Oct 2023 17:54:43 +0000
Subject: [PATCH] Changed api library jar type


diff --git a/build.gradle.kts b/build.gradle.kts
index 252d637065e94cadc925082ec380121b61df7d38..c5f6a4caa17bfb0e1b5cfc1a4728e581ddf1afe4 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -179,3 +179,4 @@ tasks.check {
     dependsOn(scanJar)
 }
 // Paper end
+tasks.getByName("jar").dependsOn("sourcesJar") // Canvas
\ No newline at end of file
