From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris Bauer <jedimastertoothless@hotmail.com>
Date: Fri, 27 Oct 2023 07:18:33 +0000
Subject: [PATCH] Add getStartupArgs to Server


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index ed9443377cc98f50c818ec441351d18026938d80..aa31fb164c470426faaf229b9f023ed9f3feda78 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -165,6 +165,12 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     public int getResourcesUsedPercent();
 
+    /**
+     * Retrieves the array of startup arguments used to launch the server.
+     *
+     * @return The array of startup arguments passed to the server during its launch.
+     */
+    public String[] getStartupArgs();
     // Canvas end
 
     /**
