From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <jedimastertoothless@hotmail.com>
Date: Fri, 27 Oct 2023 18:00:03 +0000
Subject: [PATCH] Add ServerProperties api


diff --git a/src/main/java/com/dueris/canvas/properties/ServerProperties.java b/src/main/java/com/dueris/canvas/properties/ServerProperties.java
new file mode 100644
index 0000000000000000000000000000000000000000..7f4e03b4d04354c6be4c5cf1c6e30a37fcb43a15
--- /dev/null
+++ b/src/main/java/com/dueris/canvas/properties/ServerProperties.java
@@ -0,0 +1,14 @@
+package com.dueris.canvas.properties;
+
+import java.util.*;
+import java.io.*;
+
+public interface ServerProperties {
+    public ArrayList<String> getPropertyKeys();
+
+    public HashMap<String, Object> getPropertyMappings();
+
+    public File getFile();
+
+    public Properties getRawProperties();
+}
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index aa31fb164c470426faaf229b9f023ed9f3feda78..b33cf620536170701eaf9c3714883a48bb0cc732 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -171,6 +171,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      * @return The array of startup arguments passed to the server during its launch.
      */
     public String[] getStartupArgs();
+
+    /**
+     * Retrieves the server.properties file for reading
+     * 
+     * @return ServerProperties for the server, aka the server.properties file
+     */
+    public com.dueris.canvas.properties.ServerProperties getServerProperties();
     // Canvas end
 
     /**
