From 9e9970d708edecf8556f1b2db11dec8fc04b5815 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/network/PacketEncoder.java b/net/minecraft/network/PacketEncoder.java
index 03ea63536db3222c23fb870c286d68cb65d7773b..744cdb3f1235bccfd88e59db804b6be0c6c68066 100644
--- a/net/minecraft/network/PacketEncoder.java
+++ b/net/minecraft/network/PacketEncoder.java
@@ -60,7 +60,7 @@ public class PacketEncoder<T extends PacketListener> extends MessageToByteEncode
     // packet size is encoded into 3-byte varint
     private static final int MAX_FINAL_PACKET_SIZE = (1 << 21) - 1;
     // Vanilla Max size for the encoder (before compression)
-    private static final int MAX_PACKET_SIZE = 8388608;
+    private static int MAX_PACKET_SIZE = io.canvasmc.canvas.Config.INSTANCE.networking.maximumPacketBytes; // Canvas - configurable max packet bytes
 
     public static class PacketTooLargeException extends RuntimeException {
         private final Packet<?> packet;
