--- a/net/minecraft/network/PacketEncoder.java
+++ b/net/minecraft/network/PacketEncoder.java
@@ -55,7 +_,7 @@
     // packet size is encoded into 3-byte varint
     private static final int MAX_FINAL_PACKET_SIZE = (1 << 21) - 1;
     // Vanilla Max size for the encoder (before compression)
-    private static final int MAX_PACKET_SIZE = 8388608;
+    private static final int MAX_PACKET_SIZE = io.canvasmc.canvas.Config.INSTANCE.networking.maximumPacketBytes; // Canvas - configurable max packet bytes
 
     public static class PacketTooLargeException extends RuntimeException {
         private final Packet<?> packet;
