From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/item/MaceItem.java b/net/minecraft/world/item/MaceItem.java
index c9e6cf2643333a49cda78bbe2956b0f8a9b8a996..c4b1ea6484a86910c9806778c743b551a9e6f160 100644
--- a/net/minecraft/world/item/MaceItem.java
+++ b/net/minecraft/world/item/MaceItem.java
@@ -109,7 +109,7 @@ public class MaceItem extends Item {
             } else {
                 double d = 3.0;
                 double d1 = 8.0;
-                double d2 = livingEntity.fallDistance;
+                double d2 = io.canvasmc.canvas.Config.INSTANCE.combat.mace.ignoreFallDistance ? 1.0D : Math.min(livingEntity.fallDistance, io.canvasmc.canvas.Config.INSTANCE.combat.mace.fallDistanceLimit == -1 ? livingEntity.fallDistance : io.canvasmc.canvas.Config.INSTANCE.combat.mace.fallDistanceLimit); // Canvas - mace config
                 double d3;
                 if (d2 <= 3.0) {
                     d3 = 4.0 * d2;
