--- a/net/minecraft/advancements/critereon/InventoryChangeTrigger.java
+++ b/net/minecraft/advancements/critereon/InventoryChangeTrigger.java
@@ -23,6 +_,7 @@
     }
 
     public void trigger(ServerPlayer player, Inventory inventory, ItemStack stack) {
+        if (this.isDisabled) return; // Canvas
         int i = 0;
         int i1 = 0;
         int i2 = 0;
@@ -43,6 +_,7 @@
     }
 
     private void trigger(ServerPlayer player, Inventory inventory, ItemStack stack, int full, int empty, int occupied) {
+        if (this.isDisabled) return; // Canvas
         this.trigger(player, instance -> instance.matches(inventory, stack, full, empty, occupied));
     }
 
