From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <159172167+PurpleWolfMC@users.noreply.github.com>
Date: Mon, 29 Apr 2024 18:48:00 +0000
Subject: [PATCH] Dont Spam Warnings During Compile


diff --git a/build.gradle.kts b/build.gradle.kts
index b369f815a9d41d398b91f6c238436d32dae17408..21dc067af7bd5d9912fd5ac71eb211fdbbee1ba5 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -75,6 +75,12 @@ dependencies {
 paperweight {
     craftBukkitPackageVersion.set("v1_21_R1") // also needs to be updated in MappingEnvironment
 }
+// Canvas start - lets not spam warnings
+tasks.compileJava {
+    options.compilerArgs.add("-Xlint:-deprecation")
+    options.isWarnings = false
+}
+// Canvas end
 
 tasks.jar {
     archiveClassifier.set("dev")
