From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <159172167+PurpleWolfMC@users.noreply.github.com>
Date: Mon, 29 Apr 2024 18:41:18 +0000
Subject: [PATCH] Fix ReobfServer


diff --git a/src/main/java/io/papermc/paper/pluginremap/ReobfServer.java b/src/main/java/io/papermc/paper/pluginremap/ReobfServer.java
index aa5bf7ae042f3d43f7612d027ebef0e5c758ffc9..056f4c191b39028eeea7d2bf130b6344abf9ab58 100644
--- a/src/main/java/io/papermc/paper/pluginremap/ReobfServer.java
+++ b/src/main/java/io/papermc/paper/pluginremap/ReobfServer.java
@@ -83,9 +83,11 @@ final class ReobfServer {
     }
 
     private static Path serverJar() {
+        // Canvas start
         try {
-            return Path.of(ReobfServer.class.getProtectionDomain().getCodeSource().getLocation().toURI());
-        } catch (final URISyntaxException ex) {
+            return java.nio.file.Paths.get("./versions/1.20.6/canvas-1.20.6.jar");
+        } catch (final Exception ex) {
+        // Canvas end
             throw new RuntimeException(ex);
         }
     }
