From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sun, 8 Oct 2023 16:33:50 -0700
Subject: [PATCH] began work on mixin-launcher(dont use plz lol)


diff --git a/build.gradle.kts b/build.gradle.kts
index e52ee6bbe7b195686e6f548acdbdb8a310ac28b3..d64c6ef1b87e7db57866df8a2b095c8d674b4aa4 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -1,4 +1,9 @@
+import com.github.jengelman.gradle.plugins.shadow.tasks.ShadowJar
+import io.papermc.paperweight.tasks.RemapJar
 import io.papermc.paperweight.util.*
+import java.nio.file.Files
+import java.nio.file.Path
+import java.nio.file.StandardCopyOption
 
 plugins {
     java
@@ -19,6 +24,7 @@ dependencies {
         exclude("io.papermc.paper", "paper-api")
     }
     // Purpur end
+//    implementation(project(":canvas-launcher")) // Canvas
     // Paper start
     implementation("org.jline:jline-terminal-jansi:3.21.0")
     implementation("net.minecrell:terminalconsoleappender:1.3.0")
