From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Thu, 9 Nov 2023 21:21:31 +0000
Subject: [PATCH] Configurable max bees in hive


diff --git a/src/main/java/dev/etil/mirai/MiraiConfig.java b/src/main/java/dev/etil/mirai/MiraiConfig.java
index 54cab7abec16b0c7155fc494f317119f17804d49..71a7e3abaeca1b0f74c17543bd85f3257f701aae 100644
--- a/src/main/java/dev/etil/mirai/MiraiConfig.java
+++ b/src/main/java/dev/etil/mirai/MiraiConfig.java
@@ -131,4 +131,11 @@ public class MiraiConfig {
         return config.getStringList(key);
     }
 
+    public static int maxBees;
+    private static void maximumBees() {
+        maxBees = getInt("max-bees-in-hive", 3,
+                    "This value defines how many bees can fit in a hive.",
+                    "Allowing more bees in a hive can reduce the number of ticking hives.");
+    }
+
 }
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/world/level/block/entity/BeehiveBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/BeehiveBlockEntity.java
index f5bcfece7bba52234964da646370bc271043a631..9d204df9ebacb3337bff25d43570f2c935550496 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/BeehiveBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/BeehiveBlockEntity.java
@@ -43,7 +43,7 @@ public class BeehiveBlockEntity extends BlockEntity {
     private final List<BeehiveBlockEntity.BeeData> stored = Lists.newArrayList();
     @Nullable
     public BlockPos savedFlowerPos;
-    public int maxBees = org.purpurmc.purpur.PurpurConfig.beeInsideBeeHive; // CraftBukkit - allow setting max amount of bees a hive can hold // Purpur
+    public int maxBees = dev.etil.mirai.MiraiConfig.maxBees; // CraftBukkit - allow setting max amount of bees a hive can hold // Purpur // Mirai
 
     public BeehiveBlockEntity(BlockPos pos, BlockState state) {
         super(BlockEntityType.BEEHIVE, pos, state);
