From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris Bauer <jedimastertoothless@hotmail.com>
Date: Thu, 26 Oct 2023 20:54:10 +0000
Subject: [PATCH] Publishing


diff --git a/.gitignore b/.gitignore
index 3811c0d849a3eb028ed1a6b7a2d4747f7f570448..09d00408604e75c6ff418ce2d2a7ded744d3487f 100644
--- a/.gitignore
+++ b/.gitignore
@@ -46,3 +46,4 @@ dependency-reduced-pom.xml
 # vs code
 /.vscode
 /.factorypath
+bin/
diff --git a/build.gradle.kts b/build.gradle.kts
index 21dc067af7bd5d9912fd5ac71eb211fdbbee1ba5..f9f432b8de1b587371df7c25d8744cc5d6dbfebc 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -121,10 +121,10 @@ tasks.compileTestJava {
 }
 // Paper end
 
-publishing {
+/* publishing { // Canvas
     publications.create<MavenPublication>("maven") {
     }
-}
+} */ // Canvas
 
 // Paper start
 val scanJar = tasks.register("scanJarForBadCalls", io.papermc.paperweight.tasks.ScanJarForBadCalls::class) {
@@ -238,3 +238,43 @@ tasks.registerRunTask("runReobfPaperclip") {
     classpath(rootProject.tasks.named<io.papermc.paperweight.tasks.CreatePaperclipJar>("createReobfPaperclipJar").flatMap { it.outputZip })
     mainClass.set(null as String?)
 }
+// Canvas start
+publishing {
+    publications {
+        create<MavenPublication>("mavenJava") {
+            pom {
+                name.set("Canvas-Server")
+                description.set("Empowering the impossible in Minecraft")
+                url.set("https://github.com/Dueris/Canvas")
+                licenses {
+                    license {
+                        name.set("GNU General Public License 3.0")
+                        url.set("https://www.gnu.org/licenses/gpl-3.0.en.html")
+                    }
+                }
+            }
+            repositories {
+                mavenCentral()
+            }
+            groupId = "io.github.dueris"
+            artifactId = "canvas-server"
+            from(components["java"])
+        }
+    }
+    repositories {
+        maven {
+            name = "sonatype"
+            if (version.toString().endsWith("SNAPSHOT")) {
+                url = uri("https://s01.oss.sonatype.org/content/repositories/snapshots/")
+            } else {
+                url = uri("https://s01.oss.sonatype.org/service/local/staging/deploy/maven2/")
+            }
+            credentials {
+                username=System.getenv("OSSRH_USERNAME")
+                password=System.getenv("OSSRH_PASSWORD")
+            }
+        }
+    }
+}
+
+// Canvas end
diff --git a/gradle.properties b/gradle.properties
new file mode 100644
index 0000000000000000000000000000000000000000..bdae19149cb5e173a8538e3682a2efa6b19d6b08
--- /dev/null
+++ b/gradle.properties
@@ -0,0 +1,2 @@
+ossrhUsername=System.getenv("OSSRH_USERNAME")
+ossrhPassword=System.getenv("OSSRH_PASSWORD")
\ No newline at end of file
