From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <159172167+PurpleWolfMC@users.noreply.github.com>
Date: Thu, 25 Apr 2024 09:27:06 +0000
Subject: [PATCH] Expose Launch Args


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 52d639a4ed0ff34c27f1040aa473102c0374b45c..7b4ead249d7fcfafce914e49ad8e2bc84861d981 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -269,6 +269,7 @@ import javax.annotation.Nonnull; // Paper
 public final class CraftServer implements Server {
     private final String serverName = io.papermc.paper.ServerBuildInfo.buildInfo().brandName(); // Paper
     private final String serverVersion;
+    public static String[] startupArgs = null; // Canvas
     private final String bukkitVersion = Versioning.getBukkitVersion();
     private final Logger logger = Logger.getLogger("Minecraft");
     private final ServicesManager servicesManager = new SimpleServicesManager();
@@ -771,6 +772,13 @@ public final class CraftServer implements Server {
     public int getPort() {
         return this.getServer().getPort();
     }
+    // Canvas start
+ 
+    @Override
+    public String[] getStartupArgs(){
+        return CraftServer.startupArgs != null ? CraftServer.startupArgs : new String[0];
+    }
+    // Canvas end
 
     @Override
     public int getViewDistance() {
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 061352f377520ec42ab2719113e012b9ef655ec5..8da2e63744d0cbe970b2f4e36f341880f76b0864 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -390,6 +390,7 @@ public class Main {
             tryPreloadClass(io.papermc.paper.chunk.system.RegionizedPlayerChunkLoader.PlayerChunkLoaderData.class.getName());
             // Paper end
         }
+        CraftServer.startupArgs = args; // Canvas - Expose Launch Args
     }
 
     // Paper start
