From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Sun, 16 Jun 2024 18:13:25 -0700
Subject: [PATCH] Expose Launch Args


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 5578d0f2e0de5560c5049d73cf1a0f5a864f24f1..65d9c01536a1ef3e58bbbd03e6be8722bb3bffe5 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -269,6 +269,7 @@ import javax.annotation.Nonnull; // Paper
 public final class CraftServer implements Server {
     private final String serverName = io.papermc.paper.ServerBuildInfo.buildInfo().brandName(); // Paper
     private final String serverVersion;
+    public static String[] startupArgs = null; // Canvas
     private final String bukkitVersion = Versioning.getBukkitVersion();
     private final Logger logger = Logger.getLogger("Minecraft");
     private final ServicesManager servicesManager = new SimpleServicesManager();
@@ -784,6 +785,13 @@ public final class CraftServer implements Server {
     public int getPort() {
         return this.getServer().getPort();
     }
+    // Canvas start
+ 
+    @Override
+    public String[] getStartupArgs(){
+        return CraftServer.startupArgs != null ? CraftServer.startupArgs : new String[0];
+    }
+    // Canvas end
 
     @Override
     public int getViewDistance() {
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index b60aaaea106da01fb33caeb09f6376fbb9c15b5f..d9268fb923ee50205e8ef77176e52721a2f3032f 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -305,6 +305,7 @@ public class Main {
                 t.printStackTrace();
             }
         }
+        CraftServer.startupArgs = args; // Canvas - Expose Launch Args
     }
 
     private static List<String> asList(String... params) {
