From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Sun, 16 Jun 2024 18:13:25 -0700
Subject: [PATCH] Expose Launch Args


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 53efcb35a8ea6c5e1dd2328b97fa6261e2055564..3bf85c7725217ba6a84c4cc5d9e3c3a1ed8ee218 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -269,6 +269,7 @@ import javax.annotation.Nonnull; // Paper
 public final class CraftServer implements Server {
     private final String serverName = io.papermc.paper.ServerBuildInfo.buildInfo().brandName(); // Paper
     private final String serverVersion;
+    public static String[] startupArgs = null; // Canvas
     private final String bukkitVersion = Versioning.getBukkitVersion();
     private final Logger logger = Logger.getLogger("Minecraft");
     private final ServicesManager servicesManager = new SimpleServicesManager();
@@ -783,6 +784,13 @@ public final class CraftServer implements Server {
     public int getPort() {
         return this.getServer().getPort();
     }
+    // Canvas start
+ 
+    @Override
+    public String[] getStartupArgs(){
+        return CraftServer.startupArgs != null ? CraftServer.startupArgs : new String[0];
+    }
+    // Canvas end
 
     @Override
     public int getViewDistance() {
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 43380ed4ac82602b6f77743ed79eb5cc4778b862..29c102434bc7d7bf6ec4c1fae3dfbd2ef28407ca 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -305,6 +305,7 @@ public class Main {
                 t.printStackTrace();
             }
         }
+        CraftServer.startupArgs = args; // Canvas - Expose Launch Args
     }
 
     private static List<String> asList(String... params) {
