From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <159172167+PurpleWolfMC@users.noreply.github.com>
Date: Thu, 25 Apr 2024 09:27:06 +0000
Subject: [PATCH] Expose Launch Args


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index bd0e62668b1f4c623666a835a7eb2367254c779d..572d54e86fe764b09f7a664dd7e4bede55930843 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -272,6 +272,7 @@ import javax.annotation.Nonnull; // Paper
 public final class CraftServer implements Server {
     private final String serverName = "Purpur"; // Paper // Pufferfish // Purpur
     private final String serverVersion;
+    public static String[] startupArgs = null; // Canvas
     private final String bukkitVersion = Versioning.getBukkitVersion();
     private final Logger logger = Logger.getLogger("Minecraft");
     private final ServicesManager servicesManager = new SimpleServicesManager();
@@ -784,6 +785,13 @@ public final class CraftServer implements Server {
     public int getPort() {
         return this.getServer().getPort();
     }
+    // Canvas start
+ 
+    @Override
+    public String[] getStartupArgs(){
+        return CraftServer.startupArgs != null ? CraftServer.startupArgs : new String[0];
+    }
+    // Canvas end
 
     @Override
     public int getViewDistance() {
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 845e443be7eb59ea004995410d47bb0c321a0c48..3f80d7172a88b5611545702f8e61421710e4522f 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -389,6 +389,7 @@ public class Main {
             tryPreloadClass(io.papermc.paper.chunk.system.RegionizedPlayerChunkLoader.PlayerChunkLoaderData.class.getName());
             // Paper end
         }
+        CraftServer.startupArgs = args; // Canvas - Expose Launch Args
     }
 
     // Paper start
