From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Fri, 20 Oct 2023 19:03:42 -0700
Subject: [PATCH] Ensure IgniteEngine is Running


diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index e14843de0bd83bc99dbd57acfc6276e55bd2491e..90cb2f8008c87a13527af7e05a92ff10a853c300 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -11,6 +11,7 @@ import java.util.logging.Level;
 import java.util.logging.Logger;
 import me.dueris.canvas.logger.CanvasLogger;
 import me.dueris.canvas.logger.CraftCanvasLogger;
+import com.mojang.logging.LogUtils;
 import joptsimple.OptionParser;
 import joptsimple.OptionSet;
 import joptsimple.util.PathConverter;
@@ -28,7 +29,13 @@ public class Main {
     // Paper end
 
     public static void main(String[] args) {
-        // Paper start
+        // Canvas start - ensure loading canvas correctly
+        try {
+            Class.forName("space.vectrix.ignite.applaunch.IgniteBootstrap");
+        } catch (ClassNotFoundException e) {
+            Logger.getLogger("Canvas").info("Unable to locate CodeSource, Canvas may be unable to function in some areas");
+        }
+        // Canvas end -/- Paper start
         final String warnWhenLegacyFormattingDetected = String.join(".", "net", "kyori", "adventure", "text", "warnWhenLegacyFormattingDetected");
         if (false && System.getProperty(warnWhenLegacyFormattingDetected) == null) {
             System.setProperty(warnWhenLegacyFormattingDetected, String.valueOf(true));
