From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sun, 21 Jan 2024 12:41:40 -0800
Subject: [PATCH] Expose StartupArgs to Server


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 1d3b8b49ea420de1e5a3f53515b88160ad06dd75..c63b2802aaf19c3f6bd9cff49d84ef767c3ebf02 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -279,6 +279,7 @@ public final class CraftServer implements Server {
     // Canvas start
     public static CraftCanvasLogger.Obfuscator canvas_obfuscator = null;
     private final CraftCanvasLogger craftCanvasLogger = new CraftCanvasLogger();
+    public static String[] startupArgs = null;
     // Canvas end
     private final String bukkitVersion = Versioning.getBukkitVersion();
     private final Logger logger = Logger.getLogger("Minecraft");
@@ -695,26 +696,35 @@ public final class CraftServer implements Server {
         return this.console.getGui();
     }
 
+    @Override
     public long getResourcesTotal(){
         return this.getGuiHandle().getGuiStatsComponent().getRamDetails().getGraphData().getTotal();
     }
 
+    @Override
     public long getResourcesFree(){
         return this.getGuiHandle().getGuiStatsComponent().getRamDetails().getGraphData().getFree();
     }
 
+    @Override
     public long getResourcesMax(){
         return this.getGuiHandle().getGuiStatsComponent().getRamDetails().getGraphData().getMax();
     }
 
+    @Override
     public long getResourcesUsedMem(){
         return this.getGuiHandle().getGuiStatsComponent().getRamDetails().getGraphData().getUsedMem();
     }
 
+    @Override
     public int getResourcesUsedPercent(){
         return this.getGuiHandle().getGuiStatsComponent().getRamDetails().getGraphData().getUsedPercent();
     }
 
+    @Override
+    public String[] getStartupArgs(){
+        return CraftServer.startupArgs != null ? CraftServer.startupArgs : null;
+    }
     // Canvas end
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 92bccd1c8243cca0f1d7b15ffae8b4f83b31e60e..dd8294ca95baa7a1d6f85ce964e0e07827803fe7 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -410,6 +410,7 @@ public class Main {
         // Canvas start
 
         CraftServer.canvas_obfuscator = new CraftCanvasLogger.Obfuscator();
+        CraftServer.startupArgs = args;
 
         // Canvas end
     }
