From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sun, 21 Jan 2024 12:41:40 -0800
Subject: [PATCH] Expose StartupArgs to Server


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 95897fa67a1fd0724dcff8a0360bf982ca696531..9cf9d971333b10915450010daa92a85bcaad07ff 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -277,6 +277,7 @@ public final class CraftServer implements Server {
     // Canvas start
     public static CraftCanvasLogger.Obfuscator canvas_obfuscator = null;
     private final CraftCanvasLogger craftCanvasLogger = new CraftCanvasLogger();
+    public static String[] startupArgs = null;
     // Canvas end
     private final String bukkitVersion = Versioning.getBukkitVersion();
     private final Logger logger = Logger.getLogger("Minecraft");
@@ -693,26 +694,35 @@ public final class CraftServer implements Server {
         return this.console.getGui();
     }
 
+    @Override
     public long getResourcesTotal(){
         return this.getGuiHandle().getGuiStatsComponent().getRamDetails().getGraphData().getTotal();
     }
 
+    @Override
     public long getResourcesFree(){
         return this.getGuiHandle().getGuiStatsComponent().getRamDetails().getGraphData().getFree();
     }
 
+    @Override
     public long getResourcesMax(){
         return this.getGuiHandle().getGuiStatsComponent().getRamDetails().getGraphData().getMax();
     }
 
+    @Override
     public long getResourcesUsedMem(){
         return this.getGuiHandle().getGuiStatsComponent().getRamDetails().getGraphData().getUsedMem();
     }
 
+    @Override
     public int getResourcesUsedPercent(){
         return this.getGuiHandle().getGuiStatsComponent().getRamDetails().getGraphData().getUsedPercent();
     }
 
+    @Override
+    public String[] getStartupArgs(){
+        return CraftServer.startupArgs != null ? CraftServer.startupArgs : null};
+    }
     // Canvas end
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index b77ac5e784aba38c31f5fe0180549dcba737dcae..99d08754b5da26a69a6e477c519743e65c002412 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -409,6 +409,7 @@ public class Main {
         // Canvas start
 
         CraftServer.canvas_obfuscator = new CraftCanvasLogger.Obfuscator();
+        CraftServer.startupArgs = args;
 
         // Canvas end
     }
