From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Tue, 5 Nov 2024 19:09:43 -0800
Subject: [PATCH] PlayerUseRespawnAnchorEvent


diff --git a/src/main/java/net/minecraft/world/level/block/RespawnAnchorBlock.java b/src/main/java/net/minecraft/world/level/block/RespawnAnchorBlock.java
index 2c5e394156dbf76107adb4913a094dfd4a598dd7..57d86d5503f9ab8d163775440b8109d038d884aa 100644
--- a/src/main/java/net/minecraft/world/level/block/RespawnAnchorBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/RespawnAnchorBlock.java
@@ -64,17 +64,35 @@ public class RespawnAnchorBlock extends Block {
 
     @Override
     protected InteractionResult useItemOn(ItemStack stack, BlockState state, Level world, BlockPos pos, Player player, InteractionHand hand, BlockHitResult hit) {
+        // Canvas start - PlayerUseRespawnAnchorEvent
+        org.bukkit.entity.Player bukkitPlayer = player.getBukkitEntity() instanceof org.bukkit.entity.Player ? (org.bukkit.entity.Player) player.getBukkitEntity() : null;
+        org.bukkit.block.Block block = org.bukkit.craftbukkit.block.CraftBlock.at(world, pos);
+        io.github.dueris.canvas.event.player.PlayerUseRespawnAnchorEvent useRespawnAnchorEvent = new io.github.dueris.canvas.event.player.PlayerUseRespawnAnchorEvent(block, bukkitPlayer);
+        useRespawnAnchorEvent.callEvent();
+        if(!useRespawnAnchorEvent.getCanceled()){
+        // Canvas end
         if (RespawnAnchorBlock.isRespawnFuel(stack) && RespawnAnchorBlock.canBeCharged(state)) {
             RespawnAnchorBlock.charge(player, world, pos, state);
             stack.consume(1, player);
             return InteractionResult.SUCCESS;
         } else {
-            return (InteractionResult) (hand == InteractionHand.MAIN_HAND && RespawnAnchorBlock.isRespawnFuel(player.getItemInHand(InteractionHand.OFF_HAND)) && RespawnAnchorBlock.canBeCharged(state) ? InteractionResult.PASS : InteractionResult.TRY_WITH_EMPTY_HAND);
+             return (InteractionResult) (hand == InteractionHand.MAIN_HAND && RespawnAnchorBlock.isRespawnFuel(player.getItemInHand(InteractionHand.OFF_HAND)) && RespawnAnchorBlock.canBeCharged(state) ? InteractionResult.PASS : InteractionResult.TRY_WITH_EMPTY_HAND);
         }
+        // Canvas start
+        }
+        return InteractionResult.FAIL;
+        // Canvas end
     }
 
     @Override
     protected InteractionResult useWithoutItem(BlockState state, Level world, BlockPos pos, Player player, BlockHitResult hit) {
+        // Canvas start - PlayerUseRespawnAnchorEvent
+        org.bukkit.entity.Player bukkitPlayer = player.getBukkitEntity() instanceof org.bukkit.entity.Player ? (org.bukkit.entity.Player) player.getBukkitEntity() : null;
+        org.bukkit.block.Block block = org.bukkit.craftbukkit.block.CraftBlock.at(world, pos);
+        io.github.dueris.canvas.event.player.PlayerUseRespawnAnchorEvent useRespawnAnchorEvent = new io.github.dueris.canvas.event.player.PlayerUseRespawnAnchorEvent(block, bukkitPlayer);
+        useRespawnAnchorEvent.callEvent();
+        if(!useRespawnAnchorEvent.getCanceled()){
+        // Canvas end
         if ((Integer) state.getValue(RespawnAnchorBlock.CHARGE) == 0) {
             return InteractionResult.PASS;
         } else if (!RespawnAnchorBlock.canSetSpawn(world)) {
@@ -101,6 +119,8 @@ public class RespawnAnchorBlock extends Block {
 
             return InteractionResult.CONSUME;
         }
+        } // Canvas
+        return InteractionResult.FAIL; // Canvas
     }
 
     private static boolean isRespawnFuel(ItemStack stack) {
