From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Fri, 8 Dec 2023 06:02:10 +0000
Subject: [PATCH] Enhance ChunkAPI


diff --git a/src/main/java/me/dueris/canvas/CanvasConfig.java b/src/main/java/me/dueris/canvas/CanvasConfig.java
index 4d1df1549ed0d554db91dddc669944035e989251..2b2b831a6116a7a562c0bef82fe133a1b70f26fb 100644
--- a/src/main/java/me/dueris/canvas/CanvasConfig.java
+++ b/src/main/java/me/dueris/canvas/CanvasConfig.java
@@ -170,10 +170,15 @@ public class CanvasConfig {
 
     public static boolean disableMixin = false;
     public static boolean obfuscateIps = false;
+    public static boolean bypassStartupCheck = false;
+
     private static void mixin(){
         disableMixin = getBoolean("mixin.disableMixins", disableMixin);
     }
     private static void logger(){
         obfuscateIps = getBoolean("logger.obfuscateIps", obfuscateIps);
     }
+    private static void general(){
+        bypassStartupCheck = getBoolean("bypassStartupCheck", bypassStartupCheck);
+    }
 }
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 6425539317877cdfba2943e47e3efff002e6bda5..6c5291e75d00ba547d41d73c49d9da37e4731d65 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -209,11 +209,11 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         // Spigot end
         // Canvas start
         try {
-                CanvasConfig.init((java.io.File) options.valueOf("canvas-settings"));
-            } catch (Exception e) {
-                DedicatedServer.LOGGER.error("Unable to load canvas-server configuration", e);
-                return false;
-            }
+            CanvasConfig.init((java.io.File) options.valueOf("canvas-settings"));
+        } catch (Exception e) {
+            DedicatedServer.LOGGER.error("Unable to load canvas-server configuration", e);
+            return false;
+        }
         // Canvas end
         // Paper start
         io.papermc.paper.util.ObfHelper.INSTANCE.getClass(); // Paper - load mappings for stacktrace deobf and etc.
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
index b6716f2b0037d12fb9956d7d1d0fab9b86cf4433..24e34809fdb722af7da51d5fed93a60fd68890a5 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftChunk.java
@@ -5,10 +5,16 @@ import com.google.common.base.Predicates;
 import com.mojang.serialization.Codec;
 import java.util.Arrays;
 import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
 import java.util.Objects;
+import java.util.Random;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ThreadPoolExecutor;
 import java.util.concurrent.locks.LockSupport;
 import java.util.function.BooleanSupplier;
 import java.util.function.Predicate;
+import io.papermc.paper.math.Position;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
 import net.minecraft.core.Registry;
@@ -17,6 +23,8 @@ import net.minecraft.core.registries.Registries;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.nbt.NbtOps;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.util.Mth;
+import net.minecraft.util.RandomSource;
 import net.minecraft.util.thread.ProcessorMailbox;
 import net.minecraft.world.level.ChunkPos;
 import net.minecraft.world.level.LightLayer;
@@ -26,6 +34,7 @@ import net.minecraft.world.level.chunk.ChunkAccess;
 import net.minecraft.world.level.chunk.ChunkStatus;
 import net.minecraft.world.level.chunk.DataLayer;
 import net.minecraft.world.level.chunk.ImposterProtoChunk;
+import net.minecraft.world.level.chunk.LevelChunk;
 import net.minecraft.world.level.chunk.LevelChunkSection;
 import net.minecraft.world.level.chunk.PalettedContainer;
 import net.minecraft.world.level.chunk.PalettedContainerRO;
@@ -35,8 +44,11 @@ import net.minecraft.world.level.entity.PersistentEntitySectionManager;
 import net.minecraft.world.level.levelgen.Heightmap;
 import net.minecraft.world.level.levelgen.WorldgenRandom;
 import net.minecraft.world.level.lighting.LevelLightEngine;
+import org.bukkit.Bukkit;
 import org.bukkit.Chunk;
 import org.bukkit.ChunkSnapshot;
+import org.bukkit.Location;
+import org.bukkit.Material;
 import org.bukkit.World;
 import org.bukkit.block.Biome;
 import org.bukkit.block.Block;
@@ -45,9 +57,18 @@ import org.bukkit.block.data.BlockData;
 import org.bukkit.craftbukkit.block.CraftBiome;
 import org.bukkit.craftbukkit.block.CraftBlock;
 import org.bukkit.craftbukkit.block.data.CraftBlockData;
+import org.bukkit.craftbukkit.entity.CraftEntity;
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+import org.bukkit.craftbukkit.generator.CraftChunkData;
+import org.bukkit.craftbukkit.generator.CustomChunkGenerator;
+import org.bukkit.craftbukkit.util.RandomSourceWrapper;
 import org.bukkit.entity.Entity;
+import org.bukkit.entity.Player;
+import org.bukkit.generator.ChunkGenerator;
+import org.bukkit.generator.WorldInfo;
 import org.bukkit.persistence.PersistentDataContainer;
 import org.bukkit.plugin.Plugin;
+import org.jetbrains.annotations.Nullable;
 
 public class CraftChunk implements Chunk {
     private final ServerLevel worldServer;
@@ -349,6 +370,72 @@ public class CraftChunk implements Chunk {
         return LoadLevel.values()[chunk.getFullStatus().ordinal()];
     }
 
+    // Canvas start
+    @Override
+    public void clearAllEntities() {
+        for(Entity entity : this.getEntities()){
+            CraftEntity craftEntity = (CraftEntity) entity;
+            craftEntity.getHandle().remove(net.minecraft.world.entity.Entity.RemovalReason.DISCARDED);
+        }
+    }
+
+    @Override
+    public List<Entity> getEntitiesInChunk() {
+        return Arrays.stream(getEntities()).toList();
+    }
+
+    @Override
+    public void replaceBlocksInChunk(final Material oldMat, final Material newMat) {
+        for (Block block : this.getBlocksInChunk()) {
+            if (block.getType() == oldMat) {
+                block.setType(newMat);
+            }
+        }
+    }
+
+    @Override
+    public void reloadChunk() {
+        this.unload(true);
+        this.load();
+    }
+
+    @Override
+    public void addEntityToChunk(final Entity entity) {
+        entity.spawnAt(this.getCenter());
+    }
+
+    @Override
+    public void addEntityToChunk(final Entity entity, final Location location) {
+        entity.spawnAt(location);
+    }
+
+    @Override
+    public Location getCenter() {
+        return this.getBlock(0, 0, 0).getLocation();
+    }
+
+    @Override
+    public Block[] getBlocksInChunk() {
+        World world = this.getWorld();
+        int chunkX = this.getX();
+        int chunkZ = this.getZ();
+
+        Block[] blocks = new Block[16 * 256 * 16];
+        int index = 0;
+
+        for (int x = 0; x < 16; x++) {
+            for (int y = 0; y < 256; y++) {
+                for (int z = 0; z < 16; z++) {
+                    Location blockLocation = new Location(world, chunkX * 16 + x, y, chunkZ * 16 + z);
+                    Block block = blockLocation.getBlock();
+                    blocks[index++] = block;
+                }
+            }
+        }
+        return blocks;
+    }
+    // Canvas end
+
     @Override
     public boolean equals(Object o) {
         if (this == o) return true;
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 3dafeb6e303f56d6ba0c5e441d62c0a84498b660..d8a63992c5abed40004ef633615b210ace56817f 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -9,6 +9,7 @@ import java.util.Date;
 import java.util.List;
 import java.util.logging.Level;
 import java.util.logging.Logger;
+import me.dueris.canvas.CanvasConfig;
 import me.dueris.canvas.logger.CanvasLogger;
 import me.dueris.canvas.logger.CraftCanvasLogger;
 import com.mojang.logging.LogUtils;
@@ -30,10 +31,12 @@ public class Main {
 
     public static void main(String[] args) {
         // Canvas start - ensure loading canvas correctly
-        try {
-            Class.forName("space.vectrix.ignite.applaunch.IgniteBootstrap");
-        } catch (ClassNotFoundException e) {
-            Logger.getLogger("Canvas").info("Unable to locate CodeSource, Canvas may be unable to function in some areas");
+        if(!CanvasConfig.bypassStartupCheck){
+            try {
+                Class.forName("space.vectrix.ignite.applaunch.IgniteBootstrap");
+            } catch (ClassNotFoundException e) {
+                Logger.getLogger("Canvas").info("Unable to locate CodeSource, Canvas may be unable to function in some areas");
+            }
         }
         // Canvas end -/- Paper start
         final String warnWhenLegacyFormattingDetected = String.join(".", "net", "kyori", "adventure", "text", "warnWhenLegacyFormattingDetected");
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index aa7a5aa3a33d973dafe8683e9103b7bbf151d523..fcf228d3e6a7882addde9b1e4f4b0656d4869ed0 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -98,6 +98,7 @@ import net.minecraft.world.phys.Vec3;
 import org.bukkit.BanEntry;
 import org.bukkit.BanList;
 import org.bukkit.Bukkit;
+import org.bukkit.Chunk;
 import org.bukkit.DyeColor;
 import org.bukkit.Effect;
 import org.bukkit.GameMode;
@@ -2963,6 +2964,32 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         this.getHandle().adventure$displayName = displayName != null ? displayName : net.kyori.adventure.text.Component.text(this.getName());
         this.getHandle().displayName = null;
     }
+    // Canvas start
+    
+    @Override
+    public Chunk[] getChunksInRenderDistance() {
+        int viewDistance = this.getClientViewDistance();
+        int playerChunkX = this.getLocation().getChunk().getX();
+        int playerChunkZ = this.getLocation().getChunk().getZ();
+
+        int startChunkX = playerChunkX - viewDistance;
+        int endChunkX = playerChunkX + viewDistance;
+        int startChunkZ = playerChunkZ - viewDistance;
+        int endChunkZ = playerChunkZ + viewDistance;
+
+        Chunk[] chunksInPlayerView = new Chunk[(endChunkX - startChunkX + 1) * (endChunkZ - startChunkZ + 1)];
+        int index = 0;
+
+        for (int x = startChunkX; x <= endChunkX; x++) {
+            for (int z = startChunkZ; z <= endChunkZ; z++) {
+                chunksInPlayerView[index] = this.getWorld().getChunkAt(x, z);
+                index++;
+            }
+        }
+
+        return chunksInPlayerView;
+    }
+    // Canvas end
 
     @Override
     public void deleteMessage(net.kyori.adventure.chat.SignedMessage.Signature signature) {
diff --git a/src/main/java/org/bukkit/craftbukkit/generator/CustomChunkGenerator.java b/src/main/java/org/bukkit/craftbukkit/generator/CustomChunkGenerator.java
index c6e5d3b7ef3886d0ffa9302d1270c048eaaeb671..4e007965449665f7187a4aa799580de99713eca3 100644
--- a/src/main/java/org/bukkit/craftbukkit/generator/CustomChunkGenerator.java
+++ b/src/main/java/org/bukkit/craftbukkit/generator/CustomChunkGenerator.java
@@ -96,7 +96,7 @@ public class CustomChunkGenerator extends InternalChunkGenerator {
         return this.delegate;
     }
 
-    private static WorldgenRandom getSeededRandom() {
+    public static WorldgenRandom getSeededRandom() { // Canvas - private -> public
         return new WorldgenRandom(new LegacyRandomSource(0));
     }
 
