From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolf1909 <160778222+PurpleWolf1909@users.noreply.github.com>
Date: Tue, 30 Apr 2024 20:31:33 +0000
Subject: [PATCH] ServerFinishStartingEvent


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 4a8f056e4c86f10145de78f41eb889ce8da09062..2a7e00d6222ae8fb74fc8886962b572b6ffe22cb 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -21,6 +21,7 @@ import java.util.Objects;
 import java.util.Optional;
 import java.util.function.BooleanSupplier;
 import javax.annotation.Nullable;
+import me.dueris.canvas.event.ServerFinishStartingEvent;
 import net.minecraft.DefaultUncaughtExceptionHandler;
 import net.minecraft.DefaultUncaughtExceptionHandlerWithName;
 import net.minecraft.SharedConstants;
@@ -57,6 +58,7 @@ import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.GameType;
 import net.minecraft.world.level.block.entity.SkullBlockEntity;
 import net.minecraft.world.level.storage.LevelStorageSource;
+import org.bukkit.Bukkit;
 import org.slf4j.Logger;
 
 // CraftBukkit start
@@ -398,6 +400,10 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
 
             org.purpurmc.purpur.task.BossBarTask.startAll(); // Purpur
             if (org.purpurmc.purpur.PurpurConfig.beeCountPayload) org.purpurmc.purpur.task.BeehiveTask.instance().register(); // Purpur
+            // Canvas start - ServerFinishStartingEvent
+            ServerFinishStartingEvent finishStartingEvent = new ServerFinishStartingEvent();
+            Bukkit.getPluginManager().callEvent(finishStartingEvent);
+            // Canvas end
             return true;
         }
     }
