From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolf1909 <160778222+PurpleWolf1909@users.noreply.github.com>
Date: Tue, 30 Apr 2024 20:40:40 +0000
Subject: [PATCH] PlayerUseRespawnAnchorEvent


diff --git a/src/main/java/net/minecraft/world/level/block/RespawnAnchorBlock.java b/src/main/java/net/minecraft/world/level/block/RespawnAnchorBlock.java
index 00b6941951e1af9993f8f6da5425d31b8eaa85e4..e5a7d8e5e251c6c55ee1b03f53f3c9e1508ea28a 100644
--- a/src/main/java/net/minecraft/world/level/block/RespawnAnchorBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/RespawnAnchorBlock.java
@@ -65,6 +65,13 @@ public class RespawnAnchorBlock extends Block {
 
     @Override
     protected ItemInteractionResult useItemOn(ItemStack stack, BlockState state, Level world, BlockPos pos, Player player, InteractionHand hand, BlockHitResult hit) {
+        // Canvas start - PlayerUseRespawnAnchorEvent
+        org.bukkit.entity.Player bukkitPlayer = player.getBukkitEntity() instanceof org.bukkit.entity.Player ? (org.bukkit.entity.Player) player.getBukkitEntity() : null;
+        org.bukkit.block.Block block = org.bukkit.craftbukkit.block.CraftBlock.at(world, pos);
+        me.dueris.canvas.event.player.PlayerUseRespawnAnchorEvent useRespawnAnchorEvent = new me.dueris.canvas.event.player.PlayerUseRespawnAnchorEvent(block, bukkitPlayer);
+        useRespawnAnchorEvent.callEvent();
+        if(!useRespawnAnchorEvent.getCanceled()){
+        // Canvas end
         if (RespawnAnchorBlock.isRespawnFuel(stack) && RespawnAnchorBlock.canBeCharged(state)) {
             RespawnAnchorBlock.charge(player, world, pos, state);
             stack.consume(1, player);
@@ -72,10 +79,21 @@ public class RespawnAnchorBlock extends Block {
         } else {
             return hand == InteractionHand.MAIN_HAND && RespawnAnchorBlock.isRespawnFuel(player.getItemInHand(InteractionHand.OFF_HAND)) && RespawnAnchorBlock.canBeCharged(state) ? ItemInteractionResult.SKIP_DEFAULT_BLOCK_INTERACTION : ItemInteractionResult.PASS_TO_DEFAULT_BLOCK_INTERACTION;
         }
+        // Canvas start
+        }
+        return ItemInteractionResult.FAIL;
+        // Canvas end
     }
 
     @Override
     protected InteractionResult useWithoutItem(BlockState state, Level world, BlockPos pos, Player player, BlockHitResult hit) {
+        // Canvas start - PlayerUseRespawnAnchorEvent
+        org.bukkit.entity.Player bukkitPlayer = player.getBukkitEntity() instanceof org.bukkit.entity.Player ? (org.bukkit.entity.Player) player.getBukkitEntity() : null;
+        org.bukkit.block.Block block = org.bukkit.craftbukkit.block.CraftBlock.at(world, pos);
+        me.dueris.canvas.event.player.PlayerUseRespawnAnchorEvent useRespawnAnchorEvent = new me.dueris.canvas.event.player.PlayerUseRespawnAnchorEvent(block, bukkitPlayer);
+        useRespawnAnchorEvent.callEvent();
+        if(!useRespawnAnchorEvent.getCanceled()){
+        // Canvas end
         if ((Integer) state.getValue(RespawnAnchorBlock.CHARGE) == 0) {
             return InteractionResult.PASS;
         } else if (!RespawnAnchorBlock.canSetSpawn(world)) {
@@ -102,6 +120,8 @@ public class RespawnAnchorBlock extends Block {
 
             return InteractionResult.CONSUME;
         }
+        } // Canvas
+        return InteractionResult.FAIL; // Canvas
     }
 
     private static boolean isRespawnFuel(ItemStack stack) {
