From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Wed, 15 Nov 2023 21:03:11 +0000
Subject: [PATCH] Add player abilities api


diff --git a/src/main/java/io/github/dueris/canvas/entity/player/Abilities.java b/src/main/java/io/github/dueris/canvas/entity/player/Abilities.java
new file mode 100644
index 0000000000000000000000000000000000000000..e594dea70301e251cfff2467314be2ef1600c876
--- /dev/null
+++ b/src/main/java/io/github/dueris/canvas/entity/player/Abilities.java
@@ -0,0 +1,62 @@
+package io.github.dueris.canvas.entity.player;
+
+import org.bukkit.craftbukkit.entity.CraftPlayer;
+
+public class Abilities implements io.github.dueris.canvas.entity.Abilities {
+    private CraftPlayer player;
+
+    public Abilities(CraftPlayer player){
+        this.player = player;
+    }
+
+    @Override
+    public boolean invulnerable(){
+        return this.player.getHandle().getAbilities().invulnerable;
+    }
+    // NOTE FOR LATER: add ability to set some of these values, like "maybuild", "instabuild"
+
+    @Override
+    public boolean flying(){
+        return this.player.getHandle().getAbilities().flying;
+    }
+
+    @Override
+    public boolean maybuild(){
+        return this.player.getHandle().getAbilities().mayBuild;
+    }
+
+    @Override
+    public boolean mayFly(){
+        return this.player.getHandle().getAbilities().mayfly;
+    }
+
+    @Override
+    public boolean instabuild(){
+        return this.player.getHandle().getAbilities().instabuild;
+    }
+
+    @Override
+    public float getFlyingSpeed(){
+        return this.player.getHandle().getAbilities().flyingSpeed;
+    }
+
+    @Override
+    public void setFlyingSpeed(float speed){
+        this.player.getHandle().getAbilities().setFlyingSpeed(speed);
+    }
+
+    @Override
+    public float getWalkingSpeed(){
+        return this.player.getHandle().getAbilities().walkingSpeed;
+    }
+
+    @Override
+    public void setWalkingSpeed(float speed){
+        this.player.getHandle().getAbilities().setWalkingSpeed(speed);
+    }
+
+    @Override
+    public org.bukkit.entity.Player getPlayer(){
+        return this.player;
+    }
+}
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 5e8d87b25ed9ae5319b83582abb317ce36056ba0..21c6241cf43a962bf5bd412501b4dc6a58a37a9a 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -463,6 +463,13 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
 
          updatePlayerListHeaderFooter();
     }
+    // Canvas start
+
+    @Override
+    public io.github.dueris.canvas.entity.player.Abilities getAbilities(){
+        return new io.github.dueris.canvas.entity.player.Abilities(this);
+    }
+    // Canvas end
 
     @Override
     public void setPlayerListHeaderFooter(BaseComponent header, BaseComponent footer) {
