From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sun, 22 Oct 2023 18:15:49 -0700
Subject: [PATCH] Fix MojangGUI Logger not auto moving down on new line print


diff --git a/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java b/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
index 05b49a99f956d41ef1b344a0547df2d4fe091189..6a5d8e83e821fe22953f7212cc4a3e6227aef0ab 100644
--- a/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
+++ b/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
@@ -279,28 +279,30 @@ public class MinecraftServerGui extends CraftServerGui implements MojangServerGu
     }
 
     private static final java.util.regex.Pattern ANSI = java.util.regex.Pattern.compile("\\e\\[[\\d;]*[^\\d;]"); // CraftBukkit // Paper
-    public void print(org.purpurmc.purpur.gui.JColorTextPane textArea, JScrollPane scrollPane, String message) { // Purpur
+    public void print(org.purpurmc.purpur.gui.JColorTextPane textArea, JScrollPane scrollPane, String message) {
         Document document = textArea.getDocument();
         JScrollBar jscrollbar = scrollPane.getVerticalScrollBar();
-        boolean flag = false;
 
         if (scrollPane.getViewport().getView() == textArea) {
-            flag = (double) jscrollbar.getValue() + jscrollbar.getSize().getHeight() + (double) (MinecraftServerGui.FONT.getSize() * 4) > (double) jscrollbar.getMaximum(); // Canvas
+            boolean atBottom = (double) jscrollbar.getValue() + jscrollbar.getSize().getHeight() + (double) (MinecraftServerGui.FONT.getSize() * 4) > (double) jscrollbar.getMaximum();
+
+        /* // Purpur
+        try {
+            document.insertString(document.getLength(), MinecraftServerGui.ANSI.matcher(message).replaceAll(""), (AttributeSet) null); // CraftBukkit
+        } catch (BadLocationException badlocationexception) {
+            ;
         }
-            /* // Purpur
-            try {
-                document.insertString(document.getLength(), MinecraftServerGui.ANSI.matcher(message).replaceAll(""), (AttributeSet) null); // CraftBukkit
-            } catch (BadLocationException badlocationexception) {
-                ;
-            }
-            */ // Purpur
-        textArea.append(message+ "\n"); // Purpur // Canvas
+        */ // Purpur
+            textArea.append(message.replaceAll("\u001B\\[[;\\d]*m", "") + "\n"); // Purpur // Canvas
 
-        if (flag) {
-            jscrollbar.setValue(Integer.MAX_VALUE);
+            if (atBottom) {
+                textArea.setCaretPosition(textArea.getDocument().getLength());
+                jscrollbar.setValue(Integer.MAX_VALUE);
+            }
         }
     }
 
+
     private static int max = 1000; // Canvas
 
     // Purpur start
