From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Thu, 7 Dec 2023 17:13:32 +0000
Subject: [PATCH] Added attribute listeners


diff --git a/src/main/java/io/papermc/paper/attribute/UnmodifiableAttributeMap.java b/src/main/java/io/papermc/paper/attribute/UnmodifiableAttributeMap.java
index 68044b8439c302114240d0ae4da93ab3e0789cd2..2ddc4c7181461ab9eeea67950f668592f3f4baf5 100644
--- a/src/main/java/io/papermc/paper/attribute/UnmodifiableAttributeMap.java
+++ b/src/main/java/io/papermc/paper/attribute/UnmodifiableAttributeMap.java
@@ -7,6 +7,9 @@ import org.bukkit.attribute.AttributeInstance;
 import org.bukkit.craftbukkit.attribute.CraftAttribute;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
+import org.bukkit.entity.*;
+import org.bukkit.attribute.*;
+import org.bukkit.plugin.*;
 
 public class UnmodifiableAttributeMap implements Attributable {
 
@@ -29,4 +32,10 @@ public class UnmodifiableAttributeMap implements Attributable {
     public void registerAttribute(@NotNull Attribute attribute) {
         throw new UnsupportedOperationException("Cannot register new attributes here");
     }
-}
+    // Canvas start
+    @Override
+    public void registerAttributeListener(boolean condition, int falseValue, int trueValue, Attribute attribute, Plugin plugin, int tickrate){
+        throw new UnsupportedOperationException("Cannot register new listener here");
+    }
+    // Canvas end
+}
\ No newline at end of file
diff --git a/src/main/java/org/bukkit/craftbukkit/attribute/CraftAttributeMap.java b/src/main/java/org/bukkit/craftbukkit/attribute/CraftAttributeMap.java
index a3f989fd1775adbb84d197c51e22c1889a2f8e18..6fe4d77f4e13e25051d0fb4fe32c96b197ecf71d 100644
--- a/src/main/java/org/bukkit/craftbukkit/attribute/CraftAttributeMap.java
+++ b/src/main/java/org/bukkit/craftbukkit/attribute/CraftAttributeMap.java
@@ -5,6 +5,11 @@ import net.minecraft.world.entity.ai.attributes.AttributeMap;
 import org.bukkit.attribute.Attributable;
 import org.bukkit.attribute.Attribute;
 import org.bukkit.attribute.AttributeInstance;
+// Canvas start
+import org.bukkit.scheduler.BukkitRunnable;
+import org.bukkit.plugin.*;
+import org.bukkit.entity.*;
+// Canvas end
 
 public class CraftAttributeMap implements Attributable {
 
@@ -41,5 +46,29 @@ public class CraftAttributeMap implements Attributable {
         Preconditions.checkArgument(attribute != null, "attribute");
         handle.registerAttribute(CraftAttribute.bukkitToMinecraft(attribute));
     }
-    // Paper end
-}
+    // Paper end -/- Canvas start
+    @Override
+    public void registerAttributeListener(boolean condition, int falseValue, int trueValue, Attribute attribute, Plugin plugin, int tickrate){
+        new BukkitRunnable(){
+            @Override
+            public void run(){
+                // if(!getSelf().isDead()){ --- Canvas - method not exist because not instance of an entity.
+                    if(getSelf().getAttribute(attribute) != null){
+                        if(condition){
+                            getSelf().getAttribute(attribute).setBaseValue(trueValue);
+                        }else{
+                            getSelf().getAttribute(attribute).setBaseValue(falseValue);
+                        }
+                    }
+                // }else{
+                //     this.cancel(); --- Canvas - method not exist because not instance of an entity.
+                // }
+            }
+        }.runTaskTimer(plugin, 0, tickrate);
+    }
+
+    private CraftAttributeMap getSelf(){
+        return this;
+    }
+    // Canvas end
+}
\ No newline at end of file
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index d44278f62849a124e3329cdacd16b48b87e19aa4..3ab747cbcef41f77f54db81442dbb64680e40a7a 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -92,6 +92,7 @@ import org.bukkit.util.Vector;
 // Canvas start
 import org.bukkit.scheduler.BukkitRunnable;
 import org.bukkit.entity.*;
+import org.bukkit.attribute.*;
 import org.bukkit.plugin.*;
 // Canvas end
 
@@ -111,6 +112,31 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         return Math.min(Math.max(0, this.getHandle().getHealth()), this.getMaxHealth());
     }
 
+    // Canvas start
+    public void registerAttributeListener(boolean condition, int falseValue, int trueValue, Attribute attribute, Plugin plugin, int tickrate){
+        new BukkitRunnable(){
+            @Override
+            public void run(){
+                if(!getSelf().isDead()){
+                    if(getSelf().getAttribute(attribute) != null){
+                        if(condition){
+                            getSelf().getAttribute(attribute).setBaseValue(trueValue);
+                        }else{
+                            getSelf().getAttribute(attribute).setBaseValue(falseValue);
+                        }
+                    }
+                }else{
+                    this.cancel();
+                }
+            }
+        }.runTaskTimer(plugin, 0, tickrate);
+    }
+
+    private CraftLivingEntity getSelf(){
+        return this;
+    }
+    // Canvas end
+
     @Override
     public void setHealth(double health) {
         health = (float) health;
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
index a68669ba63a107f7f11100c7572ff2ad9a9ecab5..5fa6d582583da6c8551fae8727ba90aca73950a6 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
@@ -8,10 +8,13 @@ import org.bukkit.Sound;
 import org.bukkit.craftbukkit.CraftServer;
 import org.bukkit.craftbukkit.CraftSound;
 import org.bukkit.craftbukkit.util.CraftNamespacedKey;
-import org.bukkit.entity.LivingEntity;
-import org.bukkit.entity.Mob;
-import org.bukkit.entity.Player;
 import org.bukkit.loot.LootTable;
+// Canvas start
+import org.bukkit.scheduler.BukkitRunnable;
+import org.bukkit.entity.*;
+import org.bukkit.attribute.*;
+import org.bukkit.plugin.*;
+// Canvas end
 
 public abstract class CraftMob extends CraftLivingEntity implements Mob {
     public CraftMob(CraftServer server, net.minecraft.world.entity.Mob entity) {
@@ -33,6 +36,32 @@ public abstract class CraftMob extends CraftLivingEntity implements Mob {
         }
     }
 
+    // Canvas start
+    public void registerAttributeListener(boolean condition, int falseValue, int trueValue, Attribute attribute, Plugin plugin, int tickrate){
+        new BukkitRunnable(){
+            @Override
+            public void run(){
+                if(!getSelf().isDead()){
+                    if(getSelf().getAttribute(attribute) != null){
+                        if(condition){
+                            getSelf().getAttribute(attribute).setBaseValue(trueValue);
+                        }else{
+                            getSelf().getAttribute(attribute).setBaseValue(falseValue);
+                        }
+                    }
+                }else{
+                    this.cancel();
+                }
+            }
+        }.runTaskTimer(plugin, 0, tickrate);
+    }
+
+    private CraftMob getSelf(){
+        return this;
+    }
+
+    // Canvas end
+
     @Override
     public boolean canBeLeashed(Player player){
         return this.getHandle().canBeLeashed(((CraftPlayer) player).getHandle());
