From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Fri, 27 Dec 2024 00:12:01 -0800
Subject: [PATCH] Ensure Java 22 or Higher


diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
index 80402b8ba1a96c909677d310b253ad3cf7c0c0b6..54424db4969cfea2de6f9fe374fce21cbcb0f249 100644
--- a/src/main/java/net/minecraft/server/Main.java
+++ b/src/main/java/net/minecraft/server/Main.java
@@ -82,6 +82,12 @@ public class Main {
     @SuppressForbidden(reason = "System.out needed before bootstrap") // CraftBukkit - decompile error
     @DontObfuscate
     public static void main(final OptionSet optionset) { // CraftBukkit - replaces main(String[] astring)
+        // Canvas start - ensure java 22 or higher
+        final int javaSpecVersion = Integer.parseInt(System.getProperty("java.specification.version"));
+        if (javaSpecVersion < 22) {
+            throw new RuntimeException("Canvas requires Java 22 or higher to run, but Java version " + javaSpecVersion + " was detected! Please update.");
+        }
+        // Canvas end
         io.papermc.paper.util.LogManagerShutdownThread.hook(); // Paper
         SharedConstants.tryDetectVersion();
         /* CraftBukkit start - Replace everything
