From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Sun, 29 Oct 2023 17:54:28 -0700
Subject: [PATCH] Change some gradle configs for Sonatype - fix errors for
 missing depend


diff --git a/build.gradle.kts b/build.gradle.kts
index ec10df6c543581898686735493dff76769692d5e..75c93095727fcab9e3a2fb35eb230b011dc61854 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -45,6 +45,7 @@ dependencies {
     runtimeOnly("org.xerial:sqlite-jdbc:3.42.0.1")
     runtimeOnly("com.mysql:mysql-connector-j:8.2.0")
     runtimeOnly("com.lmax:disruptor:3.4.4") // Paper
+    // implementation("org.spongepowered:mixin:0.8.5") // Canvas - doesnt need to be here, already in api
     // Paper start - Use Velocity cipher
     implementation("com.velocitypowered:velocity-native:3.1.2-SNAPSHOT") {
         isTransitive = false
@@ -252,6 +253,9 @@ publishing {
                         url.set("https://www.gnu.org/licenses/gpl-3.0.en.html")
                     }
                 }
+                repositories{
+                    mavenCentral()
+                }
             }
             groupId = "io.github.dueris"
             artifactId = "canvas-server"
@@ -273,4 +277,4 @@ publishing {
         }
     }
 }
-// Canvas end
\ No newline at end of file
+// Canvas end
