From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Fri, 17 Nov 2023 17:05:25 +0000
Subject: [PATCH] Add setDecorationsDirty to MapRenderer


diff --git a/src/main/java/org/bukkit/craftbukkit/map/CraftMapRenderer.java b/src/main/java/org/bukkit/craftbukkit/map/CraftMapRenderer.java
index aef19cfbecb4ddfc8dc71c4f3b2a011364c12dc2..c3ea7cf4f1fdc2bea93938435e8b94a82af80252 100644
--- a/src/main/java/org/bukkit/craftbukkit/map/CraftMapRenderer.java
+++ b/src/main/java/org/bukkit/craftbukkit/map/CraftMapRenderer.java
@@ -52,5 +52,11 @@ public class CraftMapRenderer extends MapRenderer {
     public boolean isExplorerMap() {
         return this.worldMap.isExplorerMap;
     }
-    // Purpur - end
+    // Purpur end -/- Canvas start
+
+    @Override
+    public void setDecorationsDirty(){
+        this.worldMap.setDecorationsDirty();
+    }
+    // Canvas end
 }
