From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Fri, 8 Dec 2023 07:11:33 +0000
Subject: [PATCH] Remove old particle trail method for Fireball


diff --git a/src/main/java/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java b/src/main/java/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
index 20ce474b72894a8bb3fc21018b9d79bbdc8bb14b..ce26e74b3c941fa348ea2d85e3176a3fc37496ea 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
@@ -131,7 +131,7 @@ public abstract class AbstractHurtingProjectile extends Projectile {
             this.setDeltaMovement(vec3d.add(this.xPower, this.yPower, this.zPower).scale((double) f));
             ParticleOptions particleparam = this.getTrailParticle();
 
-            if (particleparam != null) {
+            if (particleparam != null && this.hasTrail) { // Canvas
                 this.level().addParticle(particleparam, d0, d1 + 0.5D, d2, 0.0D, 0.0D, 0.0D);
             }
 
@@ -144,8 +144,14 @@ public abstract class AbstractHurtingProjectile extends Projectile {
         return super.canHitEntity(entity) && !entity.noPhysics;
     }
 
-    protected boolean shouldBurn() {
-        return true;
+    // Canvas start
+    private boolean shouldBurn = true;
+    public boolean hasTrail = true;
+    private ParticleOptions trailType = ParticleTypes.SMOKE;
+    private float inertia = 0.95f;
+
+    public boolean shouldBurn() {
+        return shouldBurn;
     }
 
     @Nullable
@@ -153,14 +159,22 @@ public abstract class AbstractHurtingProjectile extends Projectile {
         return ParticleTypes.SMOKE;
     }
 
-    protected float getInertia() {
-        return 0.95F;
+    public float getInertia() {
+        return this.inertia;
     }
 
-    protected float getLiquidInertia() {
+    public float getLiquidInertia() {
         return 0.8F;
     }
 
+    public void setShouldBurn(boolean burn) {
+        this.shouldBurn = burn;
+    }
+
+    public void setInertia(float inertia) {
+        this.inertia = inertia;
+    }
+
     @Override
     public void addAdditionalSaveData(CompoundTag nbt) {
         super.addAdditionalSaveData(nbt);
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java
index 91259931f45b93483d1538b72fc2b66580d32fa9..f7ddec81e221a48b6ab2b8efcb2a2214181af2f6 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java
@@ -45,14 +45,13 @@ public class CraftFireball extends AbstractProjectile implements Fireball {
     }
 
     @Override
-    public void setParticleTrailType(final Particle particle) {
-        ParticleOptions particleType = CraftParticle.bukkitToMinecraft(particle);
-        this.getHandle().setTrailParticle(particleType);
+    public void setHasParticleTrail(boolean hasParticleTrail) {
+        this.getHandle().hasTrail = hasParticleTrail;
     }
 
     @Override
-    public Particle getParticleTrailType() {
-        return CraftParticle.minecraftToBukkit(this.getHandle().getTrailParticle().getType());
+    public boolean getHasParticleTrail() {
+        return this.getHandle().hasTrail;
     }
 
     @Override
