From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: dueris <jedimastertoothless@hotmail.com>
Date: Thu, 18 Jul 2024 08:15:56 -0700
Subject: [PATCH] Add illegal item check to ItemStacks


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
index 40fb5b5e00f6bc82e67d318b8b3d1e7606973f52..3f04a35e021ae81d44b38d713e225572af45bb48 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
@@ -147,6 +147,12 @@ public final class CraftItemStack extends ItemStack {
 
     public net.minecraft.world.item.ItemStack handle;
     private boolean isForInventoryDrop;
+    // Canvas start - why no getHandle method?
+
+    public net.minecraft.world.item.ItemStack getHandle(){
+        return this.handle;
+    }
+    // Canvas end
 
     /**
      * Mirror
@@ -398,6 +404,31 @@ public final class CraftItemStack extends ItemStack {
     static Material getType(net.minecraft.world.item.ItemStack item) {
         return item == null ? Material.AIR : CraftItemType.minecraftToBukkit(item.getItem());
     }
+    // Canvas start
+
+    public boolean isIllegalItem(){
+        return CraftItemStack.isIllegalItem(this);
+    }
+
+    public static boolean isIllegalItem(ItemStack itemStack){
+        for(Enchantment enchantment : itemStack.getEnchantments().keySet()){
+            if(enchantment.getMaxLevel() < itemStack.getEnchantments().get(enchantment)){
+                return true;
+            }
+        }
+
+        if(
+            itemStack.getDamage() < 0 ||
+            itemStack.getMaxItemUseDuration() < itemStack.getDamage() ||
+            itemStack.getItemMeta() == null ||
+            itemStack.getAmount() > 64
+        ){
+            return true;
+        } else {
+            return false;
+        }
+    }
+    // Canvas end
 
     @Override
     public boolean setItemMeta(ItemMeta itemMeta) {
