From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <jedimastertoothless@hotmail.com>
Date: Sat, 28 Oct 2023 04:51:09 +0000
Subject: [PATCH] Fix api gradle


diff --git a/build.gradle.kts b/build.gradle.kts
index d69977c52460be26469de5b35a74d3dd1d0c7cb0..ff603bebb2991c1370ecc9eef9a0043f1ac30901 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -19,7 +19,7 @@ val alsoShade: Configuration by configurations.creating
 
 dependencies {
     // Purpur start
-    implementation(project(":purpur-api"))
+    implementation(project(":canvas-api")) // Canvas
     implementation("io.papermc.paper:paper-mojangapi:${project.version}") {
         exclude("io.papermc.paper", "paper-api")
     }
