From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris Bauer <jedimastertoothless@hotmail.com>
Date: Fri, 27 Oct 2023 07:18:37 +0000
Subject: [PATCH] Expose StartupArgs to Server


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index a7dd74119aa579ad6c108febb094397bc26883aa..176a6cdc1837a3c6edced5f1e86c8f6f4e3f6b64 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -280,6 +280,7 @@ public final class CraftServer implements Server {
     public static CraftCanvasLogger.Obfuscator canvas_obfuscator = null;
     private final CraftCanvasLogger craftCanvasLogger = new CraftCanvasLogger();
     private final ChatChannelManager chatChannelManager = new ChatChannelManager();
+    public static String[] startupArgs = null;
     // Canvas end
     private final String bukkitVersion = Versioning.getBukkitVersion();
     private final Logger logger = Logger.getLogger("Minecraft");
@@ -677,7 +678,6 @@ public final class CraftServer implements Server {
     }
 
     // Canvas start
-
     @Override
     public @NotNull Obfuscator getObfuscator() {
         return this.canvas_obfuscator;
@@ -698,30 +698,40 @@ public final class CraftServer implements Server {
     //     return this.chatChannelManager;
     // }
 
+    @Override
     public net.minecraft.server.gui.MinecraftServerGui getGuiHandle(){
         return this.console.getGui();
     }
 
+    @Override
     public long getResourcesTotal(){
         return this.getGuiHandle().getGuiStatsComponent().getRamDetails().getGraphData().getTotal();
     }
 
+    @Override
     public long getResourcesFree(){
         return this.getGuiHandle().getGuiStatsComponent().getRamDetails().getGraphData().getFree();
     }
 
+    @Override
     public long getResourcesMax(){
         return this.getGuiHandle().getGuiStatsComponent().getRamDetails().getGraphData().getMax();
     }
 
+    @Override
     public long getResourcesUsedMem(){
         return this.getGuiHandle().getGuiStatsComponent().getRamDetails().getGraphData().getUsedMem();
     }
 
+    @Override
     public int getResourcesUsedPercent(){
         return this.getGuiHandle().getGuiStatsComponent().getRamDetails().getGraphData().getUsedPercent();
     }
 
+    @Override
+    public String[] getStartupArgs(){
+        return CraftServer.startupArgs != null ? CraftServer.startupArgs : null};
+    }
     // Canvas end
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index d8a63992c5abed40004ef633615b210ace56817f..e16a468821c764c44f6ce2c0d9824ff5396e4c3c 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -409,6 +409,7 @@ public class Main {
         // Canvas start
 
         CraftServer.canvas_obfuscator = new CraftCanvasLogger.Obfuscator();
+        CraftServer.startupArgs = args;
 
         // Canvas end
     }
