From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris Bauer <jedimastertoothless@hotmail.com>
Date: Fri, 27 Oct 2023 08:37:03 +0000
Subject: [PATCH] Add argument bypass for launcher check


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index de9d8c7f3ada1380f49f102ffd79fd802c3a01ce..cba4eb097bbf3354f70eb74eac797848f1564ae7 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -698,7 +698,7 @@ public final class CraftServer implements Server {
     //     return this.chatChannelManager;
     // }
 
-    @Override
+    @NotNull
     public net.minecraft.server.gui.MinecraftServerGui getGuiHandle(){
         return this.console.getGui();
     }
@@ -730,7 +730,7 @@ public final class CraftServer implements Server {
 
     @Override
     public String[] getStartupArgs(){
-        return CraftServer.startupArgs != null ? CraftServer.startupArgs : null};
+        return CraftServer.startupArgs != null ? CraftServer.startupArgs : null;
     }
     // Canvas end
 
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 3367482d0a687bfc3c307cd787319283ce2870e6..6b8663a29888a01d2b2bb19cc90ad9d73f540eaf 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -27,11 +27,22 @@ public class Main {
     static {
         System.setProperty("java.util.logging.manager", "io.papermc.paper.log.CustomLogManager");
     }
-    // Paper end
+    // Paper end -/- Canvas start
+    private static boolean checkBypassLauncher(String[] args){
+        Logger.getLogger("Canvas").info("Loading Canvas with args:" + Arrays.toString(args));
+        for (String element : args) {
+            if (element.equals("-bypass.launcher.check") || element.equals("-DPaper.isRunDev=true")) {
+                return true;
+            }
+        }
+        return false;
+    }
+    // Canvas end
 
     public static void main(String[] args) {
         // Canvas start - ensure loading canvas correctly
-        if(!CanvasConfig.bypassStartupCheck){
+        boolean checker = checkBypassLauncher(args);
+        if(!CanvasConfig.bypassStartupCheck || checker){
             try {
                 Class.forName("space.vectrix.ignite.applaunch.IgniteBootstrap");
             } catch (ClassNotFoundException e) {
@@ -201,6 +212,7 @@ public class Main {
                     .ofType(File.class)
                     .defaultsTo(new File("canvas.yml"))
                     .describedAs("Yml file");
+                acceptsAll(asList("bypass.launcher.check"), "Disables checking for correct Canvas Launcher loading");
                 // Canvas end
 
                 // Paper start
