From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris Bauer <jedimastertoothless@hotmail.com>
Date: Fri, 27 Oct 2023 08:37:03 +0000
Subject: [PATCH] Add argument bypass for launcher check


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 9b9ea932583aadd47eec22d176b4f99225bd30e8..e1ccd29184cee8e294719a02577330b2d51d4c9b 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -702,7 +702,7 @@ public final class CraftServer implements Server {
     //     return this.chatChannelManager;
     // }
 
-    @Override
+    @NotNull
     public net.minecraft.server.gui.MinecraftServerGui getGuiHandle(){
         return this.console.getGui();
     }
@@ -734,7 +734,7 @@ public final class CraftServer implements Server {
 
     @Override
     public String[] getStartupArgs(){
-        return CraftServer.startupArgs != null ? CraftServer.startupArgs : null};
+        return CraftServer.startupArgs != null ? CraftServer.startupArgs : null;
     }
     // Canvas end
 
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index d82658d2a416157dc36f9dc52fef056554be71a3..fd31a2f4f17e333e87eb4a06922b801bbd7cd372 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -27,11 +27,22 @@ public class Main {
     static {
         System.setProperty("java.util.logging.manager", "io.papermc.paper.log.CustomLogManager");
     }
-    // Paper end
+    // Paper end -/- Canvas start
+    private static boolean checkBypassLauncher(String[] args){
+        Logger.getLogger("Canvas").info("Loading Canvas with args:" + Arrays.toString(args));
+        for (String element : args) {
+            if (element.equals("-bypass.launcher.check") || element.equals("-DPaper.isRunDev=true")) {
+                return true;
+            }
+        }
+        return false;
+    }
+    // Canvas end
 
     public static void main(String[] args) {
         // Canvas start - ensure loading canvas correctly
-        if(!CanvasConfig.bypassStartupCheck){
+        boolean checker = checkBypassLauncher(args);
+        if(!CanvasConfig.bypassStartupCheck || checker){
             try {
                 Class.forName("space.vectrix.ignite.applaunch.IgniteBootstrap");
             } catch (ClassNotFoundException e) {
@@ -201,6 +212,7 @@ public class Main {
                     .ofType(File.class)
                     .defaultsTo(new File("canvas.yml"))
                     .describedAs("Yml file");
+                acceptsAll(asList("bypass.launcher.check"), "Disables checking for correct Canvas Launcher loading");
                 // Canvas end
 
                 // Paper start
