From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sun, 21 Jan 2024 12:44:32 -0800
Subject: [PATCH] Add LauncherRequirement bypass arg


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 9cf9d971333b10915450010daa92a85bcaad07ff..866f67a87fd76f809f27fb4654749675fde9ed77 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -721,7 +721,7 @@ public final class CraftServer implements Server {
 
     @Override
     public String[] getStartupArgs(){
-        return CraftServer.startupArgs != null ? CraftServer.startupArgs : null};
+        return CraftServer.startupArgs != null ? CraftServer.startupArgs : null;
     }
     // Canvas end
 
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 99d08754b5da26a69a6e477c519743e65c002412..a704c08b9952f7ead73af3b41e0ba4cfec3fcdf3 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -27,11 +27,22 @@ public class Main {
     static {
         System.setProperty("java.util.logging.manager", "io.papermc.paper.log.CustomLogManager");
     }
-    // Paper end
+    // Paper end -/- Canvas start
+    private static boolean checkBypassLauncher(String[] args){
+        Logger.getLogger("Canvas").info("Loading Canvas with args:" + Arrays.toString(args));
+        for (String element : args) {
+            if (element.equals("-bypass.launcher.check") || element.equals("-DPaper.isRunDev=true")) {
+                return true;
+            }
+        }
+        return false;
+    }
+    // Canvas end
 
     public static void main(String[] args) {
         // Canvas start - ensure loading canvas correctly
-        if(!CanvasConfig.bypassStartupCheck){
+        boolean checker = checkBypassLauncher(args);
+        if(!CanvasConfig.bypassStartupCheck || checker){
             try {
                 Class.forName("space.vectrix.ignite.applaunch.IgniteBootstrap");
             } catch (ClassNotFoundException e) {
@@ -207,6 +218,7 @@ public class Main {
                     .ofType(File.class)
                     .defaultsTo(new File("canvas.yml"))
                     .describedAs("Yml file");
+                acceptsAll(asList("bypass.launcher.check"), "Disables checking for correct Canvas Launcher loading");
                 // Canvas end
 
                 // Paper start
