From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Tue, 21 Nov 2023 02:27:58 +0000
Subject: [PATCH] Fix MC bugs and ExtringuishEvent


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index b11b0a1cba3027e11e5cc0490dd259ba9d5b1b64..72e4dce63ead0931fa6bdaae398c5e9d22230a91 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -1141,6 +1141,11 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
         this.awardStat(Stats.DEATHS);
         this.resetStat(Stats.CUSTOM.get(Stats.TIME_SINCE_DEATH));
         this.resetStat(Stats.CUSTOM.get(Stats.TIME_SINCE_REST));
+         // Canvas start
+        if (this.getRemainingFireTicks() > 0) {
+            new io.github.dueris.canvas.event.entity.EntityExtinguishEvent(this.getBukkitEntity(), io.github.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.DEATH).callEvent();
+        }
+        // Canvas end
         this.clearFire();
         this.setTicksFrozen(0);
         this.setSharedFlagOnFire(false);
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index ebf61dd7292667ddfeb608a97814fcaa6810ec1e..a174d77da3e9ddd1f350379c236fd34ec6572921 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -918,7 +918,13 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                     this.hurt(this.damageSources().onFire(), 1.0F);
                 }
 
-                this.setRemainingFireTicks(this.remainingFireTicks - 1);
+                // Canvas start
+                if (getRemainingFireTicks() - 1 == 0 && !new io.github.dueris.canvas.event.entity.EntityExtinguishEvent(this.getBukkitEntity(), io.github.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.TIME).callEvent()) {
+                    setRemainingFireTicks(20);
+                } else {
+                    this.setRemainingFireTicks(this.remainingFireTicks - 1);
+                }
+                // Canvas end
             }
 
             if (this.getTicksFrozen() > 0 && !freezeLocked) { // Paper - Freeze Tick Lock API
@@ -1310,7 +1316,11 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                 }
 
                 if (this.isOnFire() && (this.isInPowderSnow || this.isInWaterRainOrBubble())) {
-                    this.setRemainingFireTicks(-this.getFireImmuneTicks());
+                    // Canvas start
+                    if (this.callExtinguishEvent(iblockdata)){
+                        this.setRemainingFireTicks(-this.getFireImmuneTicks());
+                    }
+                    // Canvas end
                 }
 
                 //this.level().getProfiler().pop(); // Purpur
@@ -1367,7 +1377,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         }
     }
 
-    protected void playEntityOnFireExtinguishedSound() {
+    public void playEntityOnFireExtinguishedSound() { // Canvas - protected -> public
         this.playSound(SoundEvents.GENERIC_EXTINGUISH_FIRE, 0.7F, 1.6F + (this.random.nextFloat() - this.random.nextFloat()) * 0.4F);
     }
 
@@ -1961,12 +1971,41 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
 
             this.resetFallDistance();
             this.wasTouchingWater = true;
-            this.clearFire();
+            // Canvas start
+            if (this.getRemainingFireTicks() <= 0 || new io.github.dueris.canvas.event.entity.EntityExtinguishEvent(this.getBukkitEntity(), io.github.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.WATER).callEvent()) {
+                if (this.getRemainingFireTicks() > 0 && this instanceof net.minecraft.world.entity.projectile.AbstractArrow){
+                    this.playEntityOnFireExtinguishedSound();
+                }
+                this.clearFire();
+            }
+            // Canvas end
         } else {
             this.wasTouchingWater = false;
         }
 
     }
+    // Canvas start
+
+    public boolean callExtinguishEvent(@Nullable BlockState state) {
+        final io.github.dueris.canvas.event.entity.EntityExtinguishEvent.Cause cause;
+        final boolean onFire = getRemainingFireTicks() > 0;
+        if (!onFire) {
+            cause = null;
+        } else if (this.isInPowderSnow || (state != null && state.is(Blocks.POWDER_SNOW))) {
+            cause = io.github.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.POWDER_SNOW;
+        } else if (this.isInWater()) {
+            cause = io.github.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.WATER;
+        } else if (this.isInRain()) {
+            cause = io.github.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.RAIN;
+       } else if (this.isInBubbleColumn()) {
+            cause = io.github.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.BUBBLE_COLUMN;
+        } else {
+            cause = null;
+        }
+
+        return cause == null || new io.github.dueris.canvas.event.entity.EntityExtinguishEvent(this.getBukkitEntity(), cause).callEvent();
+    }
+    // Canvas end
 
     private void updateFluidOnEyes() {
         this.wasEyeInWater = this.isEyeInFluid(FluidTags.WATER);
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 38cceff26efd68df54f4fa1bcd64c2ad55adf49f..0232cc91146a7dedbafbc0fba5d450fbc430dbac 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -512,7 +512,11 @@ public abstract class LivingEntity extends Entity implements Attackable {
         }
 
         if (this.isAlive() && (this.isInWaterRainOrBubble() || this.isInPowderSnow)) {
-            this.extinguishFire();
+            // Canvas start
+            if (callExtinguishEvent(null)){
+                this.extinguishFire();
+            }
+            // Canvas end
         }
 
         if (this.hurtTime > 0) {
@@ -1870,6 +1874,11 @@ public abstract class LivingEntity extends Entity implements Attackable {
                         entity.killedEntity((ServerLevel) this.level(), this);
                     }
                     this.gameEvent(GameEvent.ENTITY_DIE);
+                    // Canvas start
+                    if (this.getRemainingFireTicks() > 0){
+                        new io.github.dueris.canvas.event.entity.EntityExtinguishEvent(this.getBukkitEntity(), io.github.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.DEATH).callEvent();
+                    }
+                    // Canvas end
                 } else {
                     this.dead = false;
                     this.setHealth((float) deathEvent.getReviveHealth());
diff --git a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
index 1dd61ee306231f77664dee7e4f2b617fe2e3659f..4ba9109921f1f995387136f2fe9bd0ea1f0b6746 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
@@ -227,7 +227,11 @@ public abstract class AbstractArrow extends Projectile {
         }
 
         if (this.isInWaterOrRain() || iblockdata.is(Blocks.POWDER_SNOW)) {
-            this.clearFire();
+            // Canvas start
+            if (this.callExtinguishEvent(iblockdata)){
+                this.clearFire();
+            }
+            // Canvas end
         }
 
         if (this.inGround && !flag) {
diff --git a/src/main/java/net/minecraft/world/entity/projectile/ThrownPotion.java b/src/main/java/net/minecraft/world/entity/projectile/ThrownPotion.java
index 86c4b593a97431efd062b8c9d86bf92269c00536..269d2e1515b0eb742a9e8a8bf92722bef771d6f8 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/ThrownPotion.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/ThrownPotion.java
@@ -167,7 +167,11 @@ public class ThrownPotion extends ThrowableItemProjectile implements ItemSupplie
                 ((CraftLivingEntity) affectedEntity).getHandle().hurt(this.damageSources().indirectMagic(this, this.getOwner()), 1.0F);
             }
             for (LivingEntity toExtinguish : event.getToExtinguish()) {
-                ((CraftLivingEntity) toExtinguish).getHandle().extinguishFire();
+                // Canvas start
+                if (new io.github.dueris.canvas.event.entity.EntityExtinguishEvent(toExtinguish, io.github.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.WATER_BOTTLE).callEvent()){
+                    ((CraftLivingEntity) toExtinguish).getHandle().extinguishFire();
+                }
+                // Canvas end
             }
             for (LivingEntity toRehydrate : event.getToRehydrate()) {
                 if (((CraftLivingEntity) toRehydrate).getHandle() instanceof Axolotl axolotl) {
diff --git a/src/main/java/net/minecraft/world/level/block/LayeredCauldronBlock.java b/src/main/java/net/minecraft/world/level/block/LayeredCauldronBlock.java
index a5c7c2d24498c66159316a4f92677625975ce5ca..80afa28542c0e694135f7f158de1e9fad3047480 100644
--- a/src/main/java/net/minecraft/world/level/block/LayeredCauldronBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/LayeredCauldronBlock.java
@@ -74,8 +74,11 @@ public class LayeredCauldronBlock extends AbstractCauldronBlock {
                     return;
                 }
             }
-            entity.clearFire();
-            // CraftBukkit end
+            // Canvas start
+            if (new io.github.dueris.canvas.event.entity.EntityExtinguishEvent(entity.getBukkitEntity(), io.github.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.CAULDRON).callEvent()){
+                entity.clearFire();
+            }
+            // CraftBukkit end -/- Canvas end
         }
 
     }
diff --git a/src/main/java/net/minecraft/world/level/block/PowderSnowBlock.java b/src/main/java/net/minecraft/world/level/block/PowderSnowBlock.java
index 5eac1a54398dfa5571b98fb6eefca9d2bf9b2793..2e7fb9ba169b4a3931a95fb57151073b062c3219 100644
--- a/src/main/java/net/minecraft/world/level/block/PowderSnowBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/PowderSnowBlock.java
@@ -83,7 +83,16 @@ public class PowderSnowBlock extends Block implements BucketPickup {
                 if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entity, pos, Blocks.AIR.defaultBlockState(), !((world.purpurConfig.powderSnowBypassMobGriefing || world.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) || entity instanceof Player))) {
                     return;
                 }
-                // CraftBukkit end
+                // CraftBukkit end -/- Canvas start
+                if (entity instanceof net.minecraft.world.entity.projectile.AbstractArrow) {
+                    if (!new io.github.dueris.canvas.event.entity.EntityExtinguishEvent(entity.getBukkitEntity(), io.github.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.POWDER_SNOW).callEvent()) {
+                        return;
+                    }
+
+                    entity.playEntityOnFireExtinguishedSound();
+                    entity.clearFire();
+                }
+                // Canvas end
                 world.destroyBlock(pos, false);
             }
 
