From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Tue, 21 Nov 2023 02:27:58 +0000
Subject: [PATCH] Fix MC bugs and ExtringuishEvent


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index eb3d7c76343935988229a252d09f093b4f482553..6839a0a53e92871691d8a519c83f1a139d97585c 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -1140,6 +1140,11 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
         this.awardStat(Stats.DEATHS);
         this.resetStat(Stats.CUSTOM.get(Stats.TIME_SINCE_DEATH));
         this.resetStat(Stats.CUSTOM.get(Stats.TIME_SINCE_REST));
+         // Canvas start
+        if (this.getRemainingFireTicks() > 0) {
+            new me.dueris.canvas.event.entity.EntityExtinguishEvent(this.getBukkitEntity(), me.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.DEATH).callEvent();
+        }
+        // Canvas end
         this.clearFire();
         this.setTicksFrozen(0);
         this.setSharedFlagOnFire(false);
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index fd208f6bb796d0dcb1742c87170507e992cd1fc8..d60825ce122def2b29dcef48b610f238288fd75d 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -897,7 +897,13 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                     this.hurt(this.damageSources().onFire(), 1.0F);
                 }
 
-                this.setRemainingFireTicks(this.remainingFireTicks - 1);
+                // Canvas start
+                if (getRemainingFireTicks() - 1 == 0 && !new me.dueris.canvas.event.entity.EntityExtinguishEvent(this.getBukkitEntity(), me.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.TIME).callEvent()) {
+                    setRemainingFireTicks(20);
+                } else {
+                    this.setRemainingFireTicks(this.remainingFireTicks - 1);
+                }
+                // Canvas end
             }
 
             if (this.getTicksFrozen() > 0 && !freezeLocked) { // Paper - Freeze Tick Lock API
@@ -1254,7 +1260,11 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                 }
 
                 if (this.isOnFire() && (this.isInPowderSnow || this.isInWaterRainOrBubble())) {
-                    this.setRemainingFireTicks(-this.getFireImmuneTicks());
+                    // Canvas start
+                    if (this.callExtinguishEvent(iblockdata)){
+                        this.setRemainingFireTicks(-this.getFireImmuneTicks());
+                    }
+                    // Canvas end
                 }
 
                 //this.level().getProfiler().pop(); // Purpur
@@ -1304,7 +1314,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         }
     }
 
-    protected void playEntityOnFireExtinguishedSound() {
+    public void playEntityOnFireExtinguishedSound() { // Canvas - protected -> public
         this.playSound(SoundEvents.GENERIC_EXTINGUISH_FIRE, 0.7F, 1.6F + (this.random.nextFloat() - this.random.nextFloat()) * 0.4F);
     }
 
@@ -1857,12 +1867,41 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
 
             this.resetFallDistance();
             this.wasTouchingWater = true;
-            this.clearFire();
+            // Canvas start
+            if (this.getRemainingFireTicks() <= 0 || new me.dueris.canvas.event.entity.EntityExtinguishEvent(this.getBukkitEntity(), me.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.WATER).callEvent()) {
+                if (this.getRemainingFireTicks() > 0 && this instanceof net.minecraft.world.entity.projectile.AbstractArrow){
+                    this.playEntityOnFireExtinguishedSound();
+                }
+                this.clearFire();
+            }
+            // Canvas end
         } else {
             this.wasTouchingWater = false;
         }
 
     }
+    // Canvas start
+
+    public boolean callExtinguishEvent(@Nullable BlockState state) {
+        final me.dueris.canvas.event.entity.EntityExtinguishEvent.Cause cause;
+        final boolean onFire = getRemainingFireTicks() > 0;
+        if (!onFire) {
+            cause = null;
+        } else if (this.isInPowderSnow || (state != null && state.is(Blocks.POWDER_SNOW))) {
+            cause = me.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.POWDER_SNOW;
+        } else if (this.isInWater()) {
+            cause = me.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.WATER;
+        } else if (this.isInRain()) {
+            cause = me.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.RAIN;
+       } else if (this.isInBubbleColumn()) {
+            cause = me.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.BUBBLE_COLUMN;
+        } else {
+            cause = null;
+        }
+
+        return cause == null || new me.dueris.canvas.event.entity.EntityExtinguishEvent(this.getBukkitEntity(), cause).callEvent();
+    }
+    // Canvas end
 
     private void updateFluidOnEyes() {
         this.wasEyeInWater = this.isEyeInFluid(FluidTags.WATER);
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 21865ff3c74dce69c6447a17d5ec897010831a93..480c0e2aea3b9002b2ea6bbb758acb0e6a624c72 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -512,7 +512,11 @@ public abstract class LivingEntity extends Entity implements Attackable {
         }
 
         if (this.isAlive() && (this.isInWaterRainOrBubble() || this.isInPowderSnow)) {
-            this.extinguishFire();
+            // Canvas start
+            if (callExtinguishEvent(null)){
+                this.extinguishFire();
+            }
+            // Canvas end
         }
 
         if (this.hurtTime > 0) {
@@ -1833,6 +1837,11 @@ public abstract class LivingEntity extends Entity implements Attackable {
                         entity.killedEntity((ServerLevel) this.level(), this);
                     }
                     this.gameEvent(GameEvent.ENTITY_DIE);
+                    // Canvas start
+                    if (this.getRemainingFireTicks() > 0){
+                        new me.dueris.canvas.event.entity.EntityExtinguishEvent(this.getBukkitEntity(), me.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.DEATH).callEvent();
+                    }
+                    // Canvas end
                 } else {
                     this.dead = false;
                     this.setHealth((float) deathEvent.getReviveHealth());
diff --git a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
index 0a7aa715674e5577e7f2327f6d0d253e5e6ec234..3d1210d0771a6f2820567b4b4e178a090cddaf25 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
@@ -227,7 +227,11 @@ public abstract class AbstractArrow extends Projectile {
         }
 
         if (this.isInWaterOrRain() || iblockdata.is(Blocks.POWDER_SNOW)) {
-            this.clearFire();
+            // Canvas start
+            if (this.callExtinguishEvent(iblockdata)){
+                this.clearFire();
+            }
+            // Canvas end
         }
 
         if (this.inGround && !flag) {
diff --git a/src/main/java/net/minecraft/world/entity/projectile/ThrownPotion.java b/src/main/java/net/minecraft/world/entity/projectile/ThrownPotion.java
index 86c4b593a97431efd062b8c9d86bf92269c00536..f6c058c768efc1d75ec8267ba9768e2c4c895fef 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/ThrownPotion.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/ThrownPotion.java
@@ -167,7 +167,11 @@ public class ThrownPotion extends ThrowableItemProjectile implements ItemSupplie
                 ((CraftLivingEntity) affectedEntity).getHandle().hurt(this.damageSources().indirectMagic(this, this.getOwner()), 1.0F);
             }
             for (LivingEntity toExtinguish : event.getToExtinguish()) {
-                ((CraftLivingEntity) toExtinguish).getHandle().extinguishFire();
+                // Canvas start
+                if (new me.dueris.canvas.event.entity.EntityExtinguishEvent(toExtinguish, me.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.WATER_BOTTLE).callEvent()){
+                    ((CraftLivingEntity) toExtinguish).getHandle().extinguishFire();
+                }
+                // Canvas end
             }
             for (LivingEntity toRehydrate : event.getToRehydrate()) {
                 if (((CraftLivingEntity) toRehydrate).getHandle() instanceof Axolotl axolotl) {
diff --git a/src/main/java/net/minecraft/world/level/block/LayeredCauldronBlock.java b/src/main/java/net/minecraft/world/level/block/LayeredCauldronBlock.java
index a5c7c2d24498c66159316a4f92677625975ce5ca..d55ff0e011e41310e9aaf27758bd408e839fab91 100644
--- a/src/main/java/net/minecraft/world/level/block/LayeredCauldronBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/LayeredCauldronBlock.java
@@ -74,8 +74,11 @@ public class LayeredCauldronBlock extends AbstractCauldronBlock {
                     return;
                 }
             }
-            entity.clearFire();
-            // CraftBukkit end
+            // Canvas start
+            if (new me.dueris.canvas.event.entity.EntityExtinguishEvent(entity.getBukkitEntity(), me.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.CAULDRON).callEvent()){
+                entity.clearFire();
+            }
+            // CraftBukkit end -/- Canvas end
         }
 
     }
diff --git a/src/main/java/net/minecraft/world/level/block/PowderSnowBlock.java b/src/main/java/net/minecraft/world/level/block/PowderSnowBlock.java
index 5eac1a54398dfa5571b98fb6eefca9d2bf9b2793..9c730b749abe6cf74905fc22b19b202191db3bcc 100644
--- a/src/main/java/net/minecraft/world/level/block/PowderSnowBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/PowderSnowBlock.java
@@ -83,7 +83,16 @@ public class PowderSnowBlock extends Block implements BucketPickup {
                 if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entity, pos, Blocks.AIR.defaultBlockState(), !((world.purpurConfig.powderSnowBypassMobGriefing || world.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) || entity instanceof Player))) {
                     return;
                 }
-                // CraftBukkit end
+                // CraftBukkit end -/- Canvas start
+                if (entity instanceof net.minecraft.world.entity.projectile.AbstractArrow) {
+                    if (!new me.dueris.canvas.event.entity.EntityExtinguishEvent(entity.getBukkitEntity(), me.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.POWDER_SNOW).callEvent()) {
+                        return;
+                    }
+
+                    entity.playEntityOnFireExtinguishedSound();
+                    entity.clearFire();
+                }
+                // Canvas end
                 world.destroyBlock(pos, false);
             }
 
