From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sun, 29 Oct 2023 19:22:22 -0700
Subject: [PATCH] Remove some Warden API mechanics from AI(no behavioral
 effect)


diff --git a/src/main/java/net/minecraft/world/entity/monster/warden/Warden.java b/src/main/java/net/minecraft/world/entity/monster/warden/Warden.java
index c111785b84d6bb70ca4685f2de649520a1eefcdc..4f579f5130a89714de55bc0e61a99f646e2dbcf9 100644
--- a/src/main/java/net/minecraft/world/entity/monster/warden/Warden.java
+++ b/src/main/java/net/minecraft/world/entity/monster/warden/Warden.java
@@ -78,7 +78,7 @@ import org.jetbrains.annotations.Contract;
 import org.joml.Vector3f;
 import org.slf4j.Logger;
 
-import static net.minecraft.world.entity.monster.warden.WardenAi.DIG_COOLDOWN_SETTER;
+// import static net.minecraft.world.entity.monster.warden.WardenAi.DIG_COOLDOWN_SETTER;
 
 public class Warden extends Monster implements VibrationSystem {
 
@@ -606,6 +606,7 @@ public class Warden extends Monster implements VibrationSystem {
     }
 
     // Canvas start
+    /*
     public void trySonicBoom(Warden warden, Brain<Warden> brain){
         brain.addActivityAndRemoveMemoryWhenStopped(Activity.FIGHT, 10, ImmutableList.of(DIG_COOLDOWN_SETTER, StopAttackingIfTargetInvalid.<Warden>create((entity) -> {
             return !warden.getAngerLevel().isAngry() || !warden.canTargetEntity(entity);
@@ -619,6 +620,7 @@ public class Warden extends Monster implements VibrationSystem {
             return entityx == entity;
         }).isPresent();
     }
+     */
     // Canvas end
 
     @Override
