From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Tue, 21 Nov 2023 19:14:51 +0000
Subject: [PATCH] Add getReloadCount() to Server


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 8b583b84bd3b6078ba84a3e386f38fd6ec525d1f..a5857d940bebad55760d0abb36a21b45fda29ebb 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1029,6 +1029,13 @@ public final class CraftServer implements Server {
 
         return false;
     }
+    // Canvas start
+
+    @Override
+    public int getReloadCount(){
+        return this.reloadCount;
+    }
+    // Canvas end
 
     @Override
     public void reload() {
