From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sun, 18 Feb 2024 16:39:20 -0800
Subject: [PATCH] ServerFinishStarting Event


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 3c5222ad07c5bd91347520f39dfbe420a38d7f28..ecb1644afb10aef022aa010b6aa382c91cfa8d96 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -21,6 +21,7 @@ import java.util.Objects;
 import java.util.Optional;
 import java.util.function.BooleanSupplier;
 import javax.annotation.Nullable;
+import me.dueris.canvas.event.ServerFinishStartingEvent;
 import net.minecraft.DefaultUncaughtExceptionHandler;
 import net.minecraft.DefaultUncaughtExceptionHandlerWithName;
 import net.minecraft.SharedConstants;
@@ -52,6 +53,7 @@ import net.minecraft.world.level.GameRules;
 import net.minecraft.world.level.GameType;
 import net.minecraft.world.level.block.entity.SkullBlockEntity;
 import net.minecraft.world.level.storage.LevelStorageSource;
+import org.bukkit.Bukkit;
 import org.slf4j.Logger;
 
 // CraftBukkit start
@@ -388,6 +390,10 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
             if (gg.pufferfish.pufferfish.PufferfishConfig.enableAsyncMobSpawning) mobSpawnExecutor.start(); // Pufferfish
             org.purpurmc.purpur.task.BossBarTask.startAll(); // Purpur
             if (org.purpurmc.purpur.PurpurConfig.beeCountPayload) org.purpurmc.purpur.task.BeehiveTask.instance().register(); // Purpur
+            // Canvas start - ServerFinishStartingEvent
+            ServerFinishStartingEvent finishStartingEvent = new ServerFinishStartingEvent();
+            Bukkit.getPluginManager().callEvent(finishStartingEvent);
+            // Canvas end
             return true;
         }
     }
