From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Wed, 19 Jun 2024 20:21:32 -0700
Subject: [PATCH] Optimize Experience Orbs


diff --git a/src/main/java/io/github/dueris/canvas/CanvasConfig.java b/src/main/java/io/github/dueris/canvas/CanvasConfig.java
index 8d14b98abcd06f67042bd5748482ecd8c2c9e19c..836931a6ad0fd4f6c6044871a441c73d661f24de 100644
--- a/src/main/java/io/github/dueris/canvas/CanvasConfig.java
+++ b/src/main/java/io/github/dueris/canvas/CanvasConfig.java
@@ -202,6 +202,10 @@ public class CanvasConfig {
     public static int spawnerMaxPlayerRange = 16;
     public static boolean spawnerEntitiesHaveAI = true;
 
+    public static boolean playerInstaAbsorbOrbs = false;
+    public static boolean useCanvasOrbOptimizations = true;
+    public static double orbMergeSearchInfaltionRadius = 0.5;
+
     private static void general(){
         disableFootsteps = getBoolean("disableFootsteps", disableFootsteps);
         checkLightForSnow = getBoolean("checkLightForSnow", checkLightForSnow);
@@ -236,6 +240,9 @@ public class CanvasConfig {
         fallingBlocksLoadChunks = getBoolean("optimizations.fallingBlocksLoadChunks", fallingBlocksLoadChunks);
         tickItemFrames = getBoolean("optimizations.tickItemFrames", tickItemFrames);
         disableRaidsFromSpawnerRaiders = getBoolean("optimizations.disableRaidsFromSpawnerRaiders", disableRaidsFromSpawnerRaiders);
+        playerInstaAbsorbOrbs = getBoolean("playerInstaAbsorbOrbs", playerInstaAbsorbOrbs);
+        useCanvasOrbOptimizations = getBoolean("useCanvasOrbOptimizations", useCanvasOrbOptimizations);
+        orbMergeSearchInfaltionRadius = getDouble("orbMergeSearchInfaltionRadius", orbMergeSearchInfaltionRadius);
     }
     
     private static void blocks(){
diff --git a/src/main/java/net/minecraft/world/entity/ExperienceOrb.java b/src/main/java/net/minecraft/world/entity/ExperienceOrb.java
index 32a0db7e8f974712bd8a05f16f3bd48ac66142d5..92f1724354af960f2766613f2dc38702d8b278f1 100644
--- a/src/main/java/net/minecraft/world/entity/ExperienceOrb.java
+++ b/src/main/java/net/minecraft/world/entity/ExperienceOrb.java
@@ -25,6 +25,7 @@ import net.minecraft.world.level.entity.EntityTypeTest;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 // CraftBukkit start
+import org.bukkit.craftbukkit.entity.CraftExperienceOrb;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.event.entity.EntityRemoveEvent;
 import org.bukkit.event.entity.EntityTargetLivingEntityEvent;
@@ -148,7 +149,21 @@ public class ExperienceOrb extends Entity {
         }
 
         if (this.tickCount % 20 == 1) {
-            this.scanForEntities();
+            // Canvas start
+            if (this.followingPlayer == null || this.followingPlayer.distanceToSqr((Entity) this) > 64.0D) {
+                this.followingPlayer = this.level().getNearestPlayer(this, 8.0D);
+            }
+            if(io.github.dueris.canvas.CanvasConfig.useCanvasOrbOptimizations){
+                List<ExperienceOrb> orbs = this.level().getEntities(EntityTypeTest.forClass(ExperienceOrb.class), this.getBoundingBox().inflate(io.github.dueris.canvas.CanvasConfig.orbMergeSearchInfaltionRadius), orb -> orb != this && !orb.isRemoved());
+                if(orbs.size() > 0){
+                    orbs.forEach(orb -> {
+                        this.count += orb.count;
+                        this.age = Math.min(this.age, orb.age);
+                        orb.remove(RemovalReason.DISCARDED);
+                    });
+                }
+            } else this.scanForEntities();
+            // Canvas end
         }
 
         if (this.followingPlayer != null && (this.followingPlayer.isSpectator() || this.followingPlayer.isDeadOrDying())) {
@@ -207,9 +222,10 @@ public class ExperienceOrb extends Entity {
     }
 
     private void scanForEntities() {
+        /* // Canvas
         if (this.followingPlayer == null || this.followingPlayer.distanceToSqr((Entity) this) > 64.0D) {
             this.followingPlayer = this.level().getNearestPlayer(this, 8.0D);
-        }
+        } */ // Canvas
 
         if (this.level() instanceof ServerLevel) {
             List<ExperienceOrb> list = this.level().getEntities(EntityTypeTest.forClass(ExperienceOrb.class), this.getBoundingBox().inflate(0.5D), this::canMerge);
@@ -332,8 +348,13 @@ public class ExperienceOrb extends Entity {
     @Override
     public void playerTouch(Player player) {
         if (player instanceof ServerPlayer entityplayer) {
-            if (player.takeXpDelay == 0 && new com.destroystokyo.paper.event.player.PlayerPickupExperienceEvent(entityplayer.getBukkitEntity(), (org.bukkit.entity.ExperienceOrb) this.getBukkitEntity()).callEvent()) { // Paper - PlayerPickupExperienceEvent
+            if ((io.github.dueris.canvas.CanvasConfig.playerInstaAbsorbOrbs || player.takeXpDelay == 0) && new com.destroystokyo.paper.event.player.PlayerPickupExperienceEvent(entityplayer.getBukkitEntity(), (org.bukkit.entity.ExperienceOrb) this.getBukkitEntity()).callEvent()) { // Paper - PlayerPickupExperienceEvent // Canvas
                 player.takeXpDelay = CraftEventFactory.callPlayerXpCooldownEvent(player, this.level().purpurConfig.playerExpPickupDelay, PlayerExpCooldownChangeEvent.ChangeReason.PICKUP_ORB).getNewCooldown(); // CraftBukkit - entityhuman.takeXpDelay = 2; // Purpur
+                // Canvas start
+                if(io.github.dueris.canvas.CanvasConfig.playerInstaAbsorbOrbs) {
+                    player.takeXpDelay = 0;
+                }
+                // Canvas end
                 player.take(this, 1);
                 int i = this.repairPlayerItems(entityplayer, this.value);
 
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 325f53a46203047a9340b36a1a4bf6873d62e8fd..25d9c03095fd1ad68de4e88e642079e315ebc455 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -662,7 +662,7 @@ public abstract class Player extends LivingEntity {
             while (iterator.hasNext()) {
                 Entity entity = (Entity) iterator.next();
 
-                if (entity.getType() == EntityType.EXPERIENCE_ORB && entity.level().purpurConfig.playerExpPickupDelay >= 0) { // Purpur
+                if ((entity.getType() == EntityType.EXPERIENCE_ORB && entity.level().purpurConfig.playerExpPickupDelay >= 0) && !io.github.dueris.canvas.CanvasConfig.playerInstaAbsorbOrbs) { // Purpur // Canvas
                     list1.add(entity);
                 } else if (!entity.isRemoved()) {
                     this.touch(entity);
