From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Tue, 21 Nov 2023 19:14:51 +0000
Subject: [PATCH] Add getReloadCount() to Server


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index da07095de49b48972b8d76d88783fd575dab1bb5..189983de751569807fbc7a4a14011cbc4d1f33b2 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1034,6 +1034,13 @@ public final class CraftServer implements Server {
 
         return false;
     }
+    // Canvas start
+
+    @Override
+    public int getReloadCount(){
+        return this.reloadCount;
+    }
+    // Canvas end
 
     @Override
     public void reload() {
