From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolf1909 <160778222+PurpleWolf1909@users.noreply.github.com>
Date: Tue, 30 Apr 2024 21:37:05 +0000
Subject: [PATCH] More ModelPart API


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index e3f6c1e132d668bd8473c8c3b6e10289c31a0e66..d47d2baf1859f515fa39c66aa1cf34b9559202c7 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -443,6 +443,16 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
         this.bukkitPickUpLoot = true;
         this.maxHealthCache = this.getMaxHealth();
     }
+    // Canvas start - ModelPart api
+
+    public me.dueris.canvas.entity.player.PlayerModelPart minecraftToCanvasModelPart(net.minecraft.world.entity.player.PlayerModelPart modelPart){
+        return me.dueris.canvas.entity.player.PlayerModelPart.valueOf(modelPart.toString());
+    }
+
+    public net.minecraft.world.entity.player.PlayerModelPart canvasToMinecraftModelPart(me.dueris.canvas.entity.player.PlayerModelPart modelPart){
+        return net.minecraft.world.entity.player.PlayerModelPart.valueOf(modelPart.toString());
+    }
+    // Canvas end
 
     // Use method to resend items in hands in case of client desync, because the item use got cancelled.
     // For example, when cancelling the leash event
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index f21e635f0b5c1603245b3462dc3c67184106a0a3..43f26fd4f8dd93bfef5f7d46630250dae1fec020 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -3287,6 +3287,13 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         }
         return (int) (duration.toMillis() / 50L);
     }
+    // Canvas start - PlayerModelPart api
+
+    @Override
+    public boolean hasModelPartShown(me.dueris.canvas.entity.player.PlayerModelPart modelPart){
+        return this.getHandle().isModelPartShown(this.getHandle().canvasToMinecraftModelPart(modelPart));
+    }
+    // Canvas end
 
     @Override
     public void clearTitle() {
