From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: satvrn <pastawho@gmail.com>
Date: Wed, 27 Sep 2023 19:38:55 -0700
Subject: [PATCH] Graphene (krypton): change Netty PooledByteBufAllocator arena
 size


diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
index 77eb8754cd0e6bb40146c5654041f1a3e77ff08a..71244e6d7e6cc4b848c7fa45f248651d797382ce 100644
--- a/src/main/java/net/minecraft/server/Main.java
+++ b/src/main/java/net/minecraft/server/Main.java
@@ -74,6 +74,21 @@ import org.bukkit.configuration.file.YamlConfiguration;
 
 public class Main {
 
+    // Graphene start - krypton: change Netty PooledByteBufAllocator arena size
+    static {
+        // By default, Netty allocates 16MiB arenas for the PooledByteBufAllocator. This is too much
+        // memory for Minecraft, which imposes a maximum packet size of 2MiB! We'll use 4MiB as a more
+        // sane default.
+        //
+        // Note: io.netty.allocator.pageSize << io.netty.allocator.maxOrder is the formula used to
+        // compute the chunk size. We lower maxOrder from its default of 11 to 9. (We also use a null
+        // check, so that the user is free to choose another setting if need be.)
+        if (System.getProperty("io.netty.allocator.maxOrder") == null) {
+            System.setProperty("io.netty.allocator.maxOrder", "9");
+        }
+    }
+    // Graphene end
+
     private static final Logger LOGGER = LogUtils.getLogger();
 
     public Main() {}
