From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sun, 29 Oct 2023 19:45:56 -0700
Subject: [PATCH] Fixed protected -> public patches


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 4aa6f35c80a96fbe68a22d504c534858718997a7..a3291e0945bbb629e3b1d9af50276256c1318178 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -2821,16 +2821,21 @@ public abstract class LivingEntity extends Entity implements Attackable {
         this.hasImpulse = true;
     }
 
-    protected void goDownInWater() {
+    public void goDownInWater() { // Canvas - protected -> public
         this.setDeltaMovement(this.getDeltaMovement().add(0.0D, -0.03999999910593033D, 0.0D));
     }
 
-    protected void jumpInLiquid(TagKey<Fluid> fluid) {
+    public void jumpInLiquid(TagKey<Fluid> fluid) { // Canvas - protected -> public
         this.setDeltaMovement(this.getDeltaMovement().add(0.0D, 0.03999999910593033D, 0.0D));
     }
 
-    protected float getWaterSlowDown() {
-        return 0.8F;
+    private float waterslowdown = 0.8f;
+    public float getWaterSlowDown() { // Canvas - protected -> public
+        return waterslowdown;
+    }
+
+    public void setWaterSlowdown(float slowdown){
+        this.waterslowdown = waterslowdown;
     }
 
     public boolean canStandOnFluid(FluidState state) {
diff --git a/src/main/java/net/minecraft/world/entity/animal/horse/SkeletonHorse.java b/src/main/java/net/minecraft/world/entity/animal/horse/SkeletonHorse.java
index c336cbb5c035974d44070d77e3bfe39d6b4f5e12..eba99d9b2e7b803d883078ef90249d8a19b5baa4 100644
--- a/src/main/java/net/minecraft/world/entity/animal/horse/SkeletonHorse.java
+++ b/src/main/java/net/minecraft/world/entity/animal/horse/SkeletonHorse.java
@@ -78,7 +78,7 @@ public class SkeletonHorse extends AbstractHorse {
     }
 
     @Override
-    protected void addBehaviourGoals() {
+    public void addBehaviourGoals() { // Canvas - protected -> public
         if (level().purpurConfig.skeletonHorseCanSwim) goalSelector.addGoal(0, new net.minecraft.world.entity.ai.goal.FloatGoal(this));
     }
 
diff --git a/src/main/java/net/minecraft/world/entity/animal/horse/ZombieHorse.java b/src/main/java/net/minecraft/world/entity/animal/horse/ZombieHorse.java
index 9e13a0c5ba2e31182c76df06380055a04c18a6a6..8919fba7ff88b3e0fa1070675f314867ab927364 100644
--- a/src/main/java/net/minecraft/world/entity/animal/horse/ZombieHorse.java
+++ b/src/main/java/net/minecraft/world/entity/animal/horse/ZombieHorse.java
@@ -105,7 +105,7 @@ public class ZombieHorse extends AbstractHorse {
     }
 
     @Override
-    protected void addBehaviourGoals() {
+    public void addBehaviourGoals() { // Canvas - protected -> public
         if (level().purpurConfig.zombieHorseCanSwim) goalSelector.addGoal(0, new net.minecraft.world.entity.ai.goal.FloatGoal(this)); // Purpur
     }
 
