From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolf1909 <160778222+PurpleWolf1909@users.noreply.github.com>
Date: Tue, 30 Apr 2024 21:37:05 +0000
Subject: [PATCH] More ModelPart API


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 6839a0a53e92871691d8a519c83f1a139d97585c..27e66f8866c1cd5327328f49f0589013b2d0bfdf 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -444,6 +444,16 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
         this.bukkitPickUpLoot = true;
         this.maxHealthCache = this.getMaxHealth();
     }
+    // Canvas start - ModelPart api
+
+    public me.dueris.canvas.entity.player.PlayerModelPart minecraftToCanvasModelPart(net.minecraft.world.entity.player.PlayerModelPart modelPart){
+        return me.dueris.canvas.entity.player.PlayerModelPart.valueOf(modelPart.toString());
+    }
+
+    public net.minecraft.world.entity.player.PlayerModelPart canvasToMinecraftModelPart(me.dueris.canvas.entity.player.PlayerModelPart modelPart){
+        return net.minecraft.world.entity.player.PlayerModelPart.valueOf(modelPart.toString());
+    }
+    // Canvas end
 
     // Use method to resend items in hands in case of client desync, because the item use got cancelled.
     // For example, when cancelling the leash event
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index ec2e7e0829f81b43427dfe611d5a4f469532d5b5..3d1e608da015ed7d3579f4e738ef25ab46c7fdd8 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -3285,6 +3285,13 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         }
         return (int) (duration.toMillis() / 50L);
     }
+    // Canvas start - PlayerModelPart api
+
+    @Override
+    public boolean hasModelPartShown(me.dueris.canvas.entity.player.PlayerModelPart modelPart){
+        return this.getHandle().isModelPartShown(this.getHandle().canvasToMinecraftModelPart(modelPart));
+    }
+    // Canvas end
 
     @Override
     public void clearTitle() {
