From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolf1909 <160778222+PurpleWolf1909@users.noreply.github.com>
Date: Tue, 30 Apr 2024 21:37:05 +0000
Subject: [PATCH] More ModelPart API


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index bd74c7dd0e2ffe8fef9e7738031d8ed9786e2059..1c9b24f627a49b74b8359738fac78fd98ceb5e2a 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -459,6 +459,16 @@ public class ServerPlayer extends Player {
         this.bukkitPickUpLoot = true;
         this.maxHealthCache = this.getMaxHealth();
     }
+    // Canvas start - ModelPart api
+
+    public me.dueris.canvas.entity.player.PlayerModelPart minecraftToCanvasModelPart(net.minecraft.world.entity.player.PlayerModelPart modelPart){
+        return me.dueris.canvas.entity.player.PlayerModelPart.valueOf(modelPart.toString());
+    }
+
+    public net.minecraft.world.entity.player.PlayerModelPart canvasToMinecraftModelPart(me.dueris.canvas.entity.player.PlayerModelPart modelPart){
+        return net.minecraft.world.entity.player.PlayerModelPart.valueOf(modelPart.toString());
+    }
+    // Canvas end
 
     // Use method to resend items in hands in case of client desync, because the item use got cancelled.
     // For example, when cancelling the leash event
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index dfef52a1a4ade765ccf39748e3a31decb64fb9b2..c86e521aba806426098f2bd564348ca56008b8d9 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -3269,6 +3269,13 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         }
         return (int) (duration.toMillis() / 50L);
     }
+    // Canvas start - PlayerModelPart api
+
+    @Override
+    public boolean hasModelPartShown(me.dueris.canvas.entity.player.PlayerModelPart modelPart){
+        return this.getHandle().isModelPartShown(this.getHandle().canvasToMinecraftModelPart(modelPart));
+    }
+    // Canvas end
 
     @Override
     public void clearTitle() {
