From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Fri, 20 Oct 2023 14:02:43 -0700
Subject: [PATCH] fixed Purpur upstream patches


diff --git a/build.gradle.kts b/build.gradle.kts
index 84251ed76069a5a002a16ad44a9ea9bfc7bac465..4d6ef998a790c54fceb6e2c427051d4ae22c9935 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -100,7 +100,7 @@ tasks.jar {
         attributes(
             "Main-Class" to "org.bukkit.craftbukkit.Main",
             "Implementation-Title" to "CraftBukkit",
-            "Implementation-Version" to "git-Purpur-$implementationVersion", // Pufferfish // Purpur
+            "Implementation-Version" to "git-Canvas-$implementationVersion", // Pufferfish // Purpur // Canvas
             "Implementation-Vendor" to date, // Paper
             "Specification-Title" to "Bukkit",
             "Specification-Version" to project.version,
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 5c6027a3ab7c52087dfebb1e21468a511ec09fa1..b94066ed7e3e7d7b26eec76fe968cfc722ada8a0 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -1,5 +1,6 @@
 package net.minecraft.server.dedicated;
 
+import com.dueris.canvas.CanvasConfig;
 import com.google.common.collect.Lists;
 import com.mojang.authlib.GameProfile;
 import com.mojang.datafixers.DataFixer;
@@ -16,6 +17,7 @@ import java.nio.file.Path;
 import java.util.Collections;
 import java.util.List;
 import java.util.Locale;
+import java.util.Objects;
 import java.util.Optional;
 import java.util.function.BooleanSupplier;
 import javax.annotation.Nullable;
@@ -205,6 +207,14 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         org.spigotmc.SpigotConfig.init((java.io.File) this.options.valueOf("spigot-settings"));
         org.spigotmc.SpigotConfig.registerCommands();
         // Spigot end
+        // Canvas start
+        try {
+                CanvasConfig.init((java.io.File) options.valueOf("canvas-settings"));
+            } catch (Exception e) {
+                DedicatedServer.LOGGER.error("Unable to load canvas-server configuration", e);
+                return false;
+            }
+        // Canvas end
         // Paper start
         io.papermc.paper.util.ObfHelper.INSTANCE.getClass(); // Paper - load mappings for stacktrace deobf and etc.
         paperConfigurations.initializeGlobalConfiguration(this.registryAccess());
@@ -586,11 +596,19 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
 
     public void showGui() {
         if (this.gui == null) {
-            this.gui = MinecraftServerGui.showFrameFor(this);
+            try {
+                this.gui = MinecraftServerGui.showFrameFor(this, javax.imageio.ImageIO.read(Objects.requireNonNull(MinecraftServerGui.class.getClassLoader().getResourceAsStream("logo.png"))));
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
         }
 
     }
 
+    public MinecraftServerGui getGui(){
+        return this.gui;
+    }
+
     @Override
     public boolean hasGui() {
         return this.gui != null;
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index efc002601050fede371a4bdd32474e3bca8e75c3..73e87f9fd8ab81f300ea156a65f98c720532252e 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1,5 +1,12 @@
 package org.bukkit.craftbukkit;
 
+import com.dueris.canvas.CanvasConfig;
+import com.dueris.canvas.CraftServerGui;
+import com.dueris.canvas.logger.CanvasLogger;
+import com.dueris.canvas.logger.CraftCanvasLogger;
+import com.dueris.canvas.logger.Obfuscator;
+import com.dueris.canvas.network.ChatChannelContainer;
+import com.dueris.canvas.network.ChatChannelManager;
 import com.google.common.base.Charsets;
 import com.google.common.base.Function;
 import com.google.common.base.Preconditions;
@@ -258,6 +265,7 @@ import org.bukkit.scoreboard.Criteria;
 import org.bukkit.structure.StructureManager;
 import org.bukkit.util.StringUtil;
 import org.bukkit.util.permissions.DefaultPermissions;
+import org.jetbrains.annotations.NotNull;
 import org.yaml.snakeyaml.LoaderOptions;
 import org.yaml.snakeyaml.Yaml;
 import org.yaml.snakeyaml.constructor.SafeConstructor;
@@ -269,8 +277,14 @@ import javax.annotation.Nullable; // Paper
 import javax.annotation.Nonnull; // Paper
 
 public final class CraftServer implements Server {
-    private final String serverName = "Purpur"; // Paper // Pufferfish // Purpur
+    private final String serverName = "Canvas"; // Paper // Purpur // Canvas
     private final String serverVersion;
+    // Canvas start
+    public static CraftCanvasLogger.Obfuscator canvas_obfuscator = null;
+    private final CraftCanvasLogger craftCanvasLogger = new CraftCanvasLogger();
+    private final ChatChannelManager chatChannelManager = new ChatChannelManager();
+
+    // Canvas end
     private final String bukkitVersion = Versioning.getBukkitVersion();
     private final Logger logger = Logger.getLogger("Minecraft");
     private final ServicesManager servicesManager = new SimpleServicesManager();
@@ -668,6 +682,26 @@ public final class CraftServer implements Server {
         return this.serverName;
     }
 
+    @Override
+    public @NotNull Obfuscator getObfuscator() {
+        return this.canvas_obfuscator;
+    }
+
+    @Override
+    public @NotNull CanvasLogger getCanvasLogger() {
+        return this.craftCanvasLogger;
+    }
+
+    @Override
+    public @NotNull CraftServerGui getServerGui() {
+        return this.console.getGui();
+    }
+
+    @Override
+    public @NotNull ChatChannelContainer getChatChannelManager() {
+        return this.chatChannelManager;
+    }
+
     @Override
     public String getVersion() {
         return this.serverVersion + " (MC: " + this.console.getServerVersion() + ")";
@@ -1056,6 +1090,7 @@ public final class CraftServer implements Server {
         org.spigotmc.SpigotConfig.init((File) this.console.options.valueOf("spigot-settings")); // Spigot
         this.console.paperConfigurations.reloadConfigs(this.console);
         org.purpurmc.purpur.PurpurConfig.init((File) console.options.valueOf("purpur-settings")); // Purpur
+        CanvasConfig.init((File) console.options.valueOf("canvas-settings")); // Canvas
         for (ServerLevel world : this.console.getAllLevels()) {
             // world.serverLevelData.setDifficulty(config.difficulty); // Paper - per level difficulty
             world.setSpawnSettings(world.serverLevelData.getDifficulty() != Difficulty.PEACEFUL && config.spawnMonsters, config.spawnAnimals); // Paper - per level difficulty (from MinecraftServer#setDifficulty(ServerLevel, Difficulty, boolean))
