From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolf1909 <160778222+PurpleWolf1909@users.noreply.github.com>
Date: Tue, 30 Apr 2024 21:51:15 +0000
Subject: [PATCH] Add Pufferfish SnakeYaml


diff --git a/build.gradle.kts b/build.gradle.kts
index ea8358f3d8270efa7081ec92547e9ecf6b8e07dc..1e4e90708e20b0eeb3361555fdc088b1a63be07d 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -16,6 +16,12 @@ dependencies {
     implementation(project(":canvas-api")) // Purpur // Canvas
     // Paper start
     implementation("org.jline:jline-terminal-jansi:3.21.0")
+    // Canvas - implement pufferfish snakeyaml
+    implementation("org.yaml:snakeyaml:1.32")
+    implementation ("com.github.carleslc.Simple-YAML:Simple-Yaml:1.8.4") {
+        exclude(group="org.yaml", module="snakeyaml")
+    }
+    // Canvas end
     implementation("net.minecrell:terminalconsoleappender:1.3.0")
     implementation("net.kyori:adventure-text-serializer-ansi:4.17.0") // Keep in sync with adventureVersion from Paper-API build file
     /*
