From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Fri, 24 Nov 2023 15:51:30 -0800
Subject: [PATCH] Expose ServerThread to CraftBukkit


diff --git a/src/main/java/org/bukkit/craftbukkit/bootstrap/Main.java b/src/main/java/org/bukkit/craftbukkit/bootstrap/Main.java
index 8a4f95049c63afb28bef6719c77b7a7092e75aae..047d7586ee1052c6c40bda4fb84ec58c21a79151 100644
--- a/src/main/java/org/bukkit/craftbukkit/bootstrap/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/bootstrap/Main.java
@@ -22,6 +22,7 @@ import java.util.ArrayList;
 import java.util.List;
 
 public class Main {
+    public static Thread runThread; // Canvas - expose runThread
 
     public static void main(String[] argv) {
         new Main().run(argv);
@@ -62,6 +63,9 @@ public class Main {
                     Thrower.INSTANCE.sneakyThrow(t);
                 }
             }, "ServerMain");
+            // Canvas start - expose runThread
+            Main.runThread = runThread;
+            // Canvas end
 
             runThread.setContextClassLoader(classLoader);
             runThread.start();
