From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Wed, 15 Nov 2023 21:54:36 +0000
Subject: [PATCH] Ensure Dir is Created


diff --git a/src/main/java/me/dueris/canvas/plugin/CanvasPluginLoader.java b/src/main/java/me/dueris/canvas/plugin/CanvasPluginLoader.java
index bb3ffcc364d4bb21a96a979949bea49ba61162d8..9dd42e352a98703d625d790fadcc08a687c4ba88 100644
--- a/src/main/java/me/dueris/canvas/plugin/CanvasPluginLoader.java
+++ b/src/main/java/me/dueris/canvas/plugin/CanvasPluginLoader.java
@@ -35,6 +35,7 @@ public class CanvasPluginLoader {
         pluginLoader.clearCache();
         // System.out.println(Paths.get("plugins"));
             File dir = Paths.get("plugins").toFile();
+            dir.mkdirs();
 
             for(File file : dir.listFiles()){
                 if(file.getName().endsWith(".jar")){
