From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Wed, 19 Jun 2024 20:13:02 -0700
Subject: [PATCH] Respawn Dimension Setter


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 27e66f8866c1cd5327328f49f0589013b2d0bfdf..c2474befa45d8f09ff742741bd869a00af7c154d 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -247,7 +247,7 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
     private Vec3 enteredLavaOnVehiclePosition;
     private SectionPos lastSectionPos;
     private ChunkTrackingView chunkTrackingView;
-    private ResourceKey<Level> respawnDimension;
+    public ResourceKey<Level> respawnDimension = Level.OVERWORLD; // Canvas - private -> public
     @Nullable
     private BlockPos respawnPosition;
     private boolean respawnForced;
@@ -314,6 +314,12 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
     public final boolean moonrise$isRealPlayer() {
         return this.isRealPlayer;
     }
+    // Canvas start
+
+    public void setRespawnDimension(ServerLevel level){
+        this.respawnDimension = level.dimension();
+    }
+    // Canvas end
 
     @Override
     public final void moonrise$setRealPlayer(final boolean real) {
@@ -352,7 +358,7 @@ public class ServerPlayer extends net.minecraft.world.entity.player.Player imple
         this.language =  null; // Paper - default to null
         this.lastSectionPos = SectionPos.of(0, 0, 0);
         this.chunkTrackingView = ChunkTrackingView.EMPTY;
-        this.respawnDimension = Level.OVERWORLD;
+        this.respawnDimension = getRespawnDimension();
         this.wardenSpawnTracker = new WardenSpawnTracker(0, 0, 0);
         this.lastKnownClientMovement = Vec3.ZERO;
         this.containerSynchronizer = new ContainerSynchronizer() {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index de282c99d40d35d0254ba752255296e94127b075..495dcc05623ce974812b7749f5550ddb91ed4f04 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1602,6 +1602,18 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         }
         return null;
     }
+    // Canvas start
+
+    @Override
+    public void setRespawnDimension(org.bukkit.World world){
+        this.getHandle().setRespawnDimension(((CraftWorld)world).getHandle());
+    }
+
+    @Override
+    public org.bukkit.World getRespawnDimension(){
+        return this.getHandle().level().getServer().getLevel(this.getHandle().respawnDimension).getWorld();
+    }
+    // Canvas end
 
     @Override
     public void setBedSpawnLocation(Location location) {
