From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sat, 21 Oct 2023 20:11:01 -0700
Subject: [PATCH] fixed protected values when they shouldnt


diff --git a/src/main/java/net/minecraft/world/entity/projectile/DragonFireball.java b/src/main/java/net/minecraft/world/entity/projectile/DragonFireball.java
index 6256ce68d6ecada66745fb09360cba2bf991360c..1d8aa652b52539ed64a5a0d8dc5dd399db4393b0 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/DragonFireball.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/DragonFireball.java
@@ -73,12 +73,12 @@ public class DragonFireball extends AbstractHurtingProjectile {
     }
 
     @Override
-    protected ParticleOptions getTrailParticle() {
+    public ParticleOptions getTrailParticle() { // Canvas
         return ParticleTypes.DRAGON_BREATH;
     }
 
     @Override
-    protected boolean shouldBurn() {
+    public boolean shouldBurn() { // Canvas
         return false;
     }
 }
diff --git a/src/main/java/net/minecraft/world/entity/projectile/WitherSkull.java b/src/main/java/net/minecraft/world/entity/projectile/WitherSkull.java
index b4687453256ead43cf5288994316c7bf946b86df..3f36646a03b9f9f37c09fc6e5adf1faf54a05df1 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/WitherSkull.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/WitherSkull.java
@@ -36,7 +36,7 @@ public class WitherSkull extends AbstractHurtingProjectile {
     }
 
     @Override
-    protected float getInertia() {
+    public float getInertia() {
         return this.isDangerous() ? 0.73F : super.getInertia();
     }
 
@@ -144,7 +144,7 @@ public class WitherSkull extends AbstractHurtingProjectile {
     }
 
     @Override
-    protected boolean shouldBurn() {
+    public boolean shouldBurn() {
         return false;
     }
 }
