From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Fri, 8 Dec 2023 07:11:33 +0000
Subject: [PATCH] Remove old particle trail method for Fireball


diff --git a/src/main/java/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java b/src/main/java/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
index c4ecc5faa4f61e7974e8c475762924a89615b377..862ee23ba2ab5df4cbe56356fa5ee1543d22c548 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
@@ -128,7 +128,7 @@ public abstract class AbstractHurtingProjectile extends Projectile {
             this.setDeltaMovement(vec3d.add(this.xPower, this.yPower, this.zPower).scale((double) f));
             ParticleOptions particleparam = this.getTrailParticle();
 
-            if (particleparam != null) {
+            if (particleparam != null && this.hasTrail) { // Canvas
                 this.level().addParticle(particleparam, d0, d1 + 0.5D, d2, 0.0D, 0.0D, 0.0D);
             }
 
@@ -141,8 +141,14 @@ public abstract class AbstractHurtingProjectile extends Projectile {
         return super.canHitEntity(entity) && !entity.noPhysics;
     }
 
-    protected boolean shouldBurn() {
-        return true;
+    // Canvas start
+    private boolean shouldBurn = true;
+    public boolean hasTrail = true;
+    private ParticleOptions trailType = ParticleTypes.SMOKE;
+    private float inertia = 0.95f;
+
+    public boolean shouldBurn() {
+        return shouldBurn;
     }
 
     @Nullable
@@ -150,14 +156,22 @@ public abstract class AbstractHurtingProjectile extends Projectile {
         return ParticleTypes.SMOKE;
     }
 
-    protected float getInertia() {
-        return 0.95F;
+    public float getInertia() {
+        return this.inertia;
     }
 
-    protected float getLiquidInertia() {
+    public float getLiquidInertia() {
         return 0.8F;
     }
 
+    public void setShouldBurn(boolean burn) {
+        this.shouldBurn = burn;
+    }
+
+    public void setInertia(float inertia) {
+        this.inertia = inertia;
+    }
+
     @Override
     public void addAdditionalSaveData(CompoundTag nbt) {
         super.addAdditionalSaveData(nbt);
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java
index c808b67959ee7f711cfb82f86bea90432354a895..9f18c48dd04717faa5e8ce5936d945894b709730 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java
@@ -45,14 +45,13 @@ public class CraftFireball extends AbstractProjectile implements Fireball {
     }
 
     @Override
-    public void setParticleTrailType(final Particle particle) {
-        ParticleOptions particleType = CraftParticle.bukkitToMinecraft(particle);
-        this.getHandle().setTrailParticle(particleType);
+    public void setHasParticleTrail(boolean hasParticleTrail) {
+        this.getHandle().hasTrail = hasParticleTrail;
     }
 
     @Override
-    public Particle getParticleTrailType() {
-        return CraftParticle.minecraftToBukkit(this.getHandle().getTrailParticle().getType());
+    public boolean getHasParticleTrail() {
+        return this.getHandle().hasTrail;
     }
 
     @Override
