From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Fri, 8 Dec 2023 05:00:21 +0000
Subject: [PATCH] Remove old particle trail method for Fireball


diff --git a/src/main/java/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java b/src/main/java/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
index 2724304fe8336f5adcd3f1cdad482b5adc08b4e0..b6b02617e862a398d1e6f6d78a543c7a66574919 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
@@ -129,7 +129,7 @@ public abstract class AbstractHurtingProjectile extends Projectile {
             this.setDeltaMovement(vec3d.add(this.xPower, this.yPower, this.zPower).scale((double) f));
             ParticleOptions particleparam = this.getTrailParticle();
 
-            if (particleparam != null) {
+            if (particleparam != null && this.hasTrail) { // Canvas
                 this.level().addParticle(particleparam, d0, d1 + 0.5D, d2, 0.0D, 0.0D, 0.0D);
             }
 
@@ -144,6 +144,7 @@ public abstract class AbstractHurtingProjectile extends Projectile {
 
     // Canvas start
     private boolean shouldBurn = true;
+    public boolean hasTrail = true;
     private ParticleOptions trailType = ParticleTypes.SMOKE;
     private float inertia = 0.95f;
 
@@ -164,10 +165,6 @@ public abstract class AbstractHurtingProjectile extends Projectile {
         this.shouldBurn = burn;
     }
 
-    public void setTrailParticle(ParticleOptions trailType) {
-        this.trailType = trailType;
-    }
-
     public void setInertia(float inertia) {
         this.inertia = inertia;
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java
index c808b67959ee7f711cfb82f86bea90432354a895..9f18c48dd04717faa5e8ce5936d945894b709730 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java
@@ -45,14 +45,13 @@ public class CraftFireball extends AbstractProjectile implements Fireball {
     }
 
     @Override
-    public void setParticleTrailType(final Particle particle) {
-        ParticleOptions particleType = CraftParticle.bukkitToMinecraft(particle);
-        this.getHandle().setTrailParticle(particleType);
+    public void setHasParticleTrail(boolean hasParticleTrail) {
+        this.getHandle().hasTrail = hasParticleTrail;
     }
 
     @Override
-    public Particle getParticleTrailType() {
-        return CraftParticle.minecraftToBukkit(this.getHandle().getTrailParticle().getType());
+    public boolean getHasParticleTrail() {
+        return this.getHandle().hasTrail;
     }
 
     @Override
