From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolf1909 <160778222+PurpleWolf1909@users.noreply.github.com>
Date: Tue, 30 Apr 2024 21:51:15 +0000
Subject: [PATCH] Add Pufferfish SnakeYaml


diff --git a/build.gradle.kts b/build.gradle.kts
index 8d48b8108f4fae1d3a64a33f57eabe05c69d92a1..e33410032c0cb06e266d79104bbc116820b55c07 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -15,6 +15,12 @@ dependencies {
     implementation(project(":canvas-api")) // Purpur // Canvas
     // Paper start
     implementation("org.jline:jline-terminal-jansi:3.21.0")
+    // Canvas - implement pufferfish snakeyaml
+    implementation("org.yaml:snakeyaml:1.32")
+    implementation ("com.github.carleslc.Simple-YAML:Simple-Yaml:1.8.4") {
+        exclude(group="org.yaml", module="snakeyaml")
+    }
+    // Canvas end
     implementation("net.minecrell:terminalconsoleappender:1.3.0")
     implementation("net.kyori:adventure-text-serializer-ansi:4.17.0") // Keep in sync with adventureVersion from Paper-API build file
     /*
