From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Fri, 17 Nov 2023 17:05:25 +0000
Subject: [PATCH] Add setDecorationsDirty to MapRenderer


diff --git a/src/main/java/org/bukkit/craftbukkit/map/CraftMapRenderer.java b/src/main/java/org/bukkit/craftbukkit/map/CraftMapRenderer.java
index b4b105c0190502328d5aeb680dd8e67c2875618f..9d0c65a8ae944ac896ee4b0582fdda07cc2170b4 100644
--- a/src/main/java/org/bukkit/craftbukkit/map/CraftMapRenderer.java
+++ b/src/main/java/org/bukkit/craftbukkit/map/CraftMapRenderer.java
@@ -51,5 +51,11 @@ public class CraftMapRenderer extends MapRenderer {
     public boolean isExplorerMap() {
         return this.worldMap.isExplorerMap;
     }
-    // Purpur - end
+    // Purpur end -/- Canvas start
+
+    @Override
+    public void setDecorationsDirty(){
+        this.worldMap.setDecorationsDirty();
+    }
+    // Canvas end
 }
