From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: dueris <jedimastertoothless@hotmail.com>
Date: Wed, 31 Jul 2024 09:41:56 -0700
Subject: [PATCH] Higher Startup Thread Priority


diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
index 228b7447de204573be94e46e888aebc6f4fda4e8..816e21790bcedb3f5f07c1c31b82a121fac23680 100644
--- a/src/main/java/net/minecraft/server/Main.java
+++ b/src/main/java/net/minecraft/server/Main.java
@@ -97,6 +97,8 @@ public class Main {
     public static void main(final OptionSet optionset) { // CraftBukkit - replaces main(String[] astring)
         io.papermc.paper.util.LogManagerShutdownThread.hook(); // Paper
         SharedConstants.tryDetectVersion();
+ 
+        if(Thread.currentThread().getPriority() < 6){ Thread.currentThread().setPriority(6); } // Canvas - improve startup times (smoothboot)
         /* CraftBukkit start - Replace everything
         OptionParser optionparser = new OptionParser();
         OptionSpec<Void> optionspec = optionparser.accepts("nogui");
