From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Fri, 24 May 2024 18:45:49 -0700
Subject: [PATCH] Optimization Configs


diff --git a/src/main/java/io/github/dueris/canvas/CanvasConfig.java b/src/main/java/io/github/dueris/canvas/CanvasConfig.java
index 54982d6550198d85659c94f84ba8fabdef482481..a7c6f58cc320f0e6030862614679dfb5b0e4e963 100644
--- a/src/main/java/io/github/dueris/canvas/CanvasConfig.java
+++ b/src/main/java/io/github/dueris/canvas/CanvasConfig.java
@@ -185,6 +185,29 @@ public class CanvasConfig {
     public static int mobWanderingPercentChance = 100;
     public static int mobWanderingReducedRateMultiplier = 6;
     public static int mobWanderingDelay = 20 * 5;
+    public static boolean tntLoadChunks = true;
+    public static boolean enderpearlsLoadChunks = true;
+    public static boolean fallingBlocksLoadChunks = true;
+    public static boolean tickItemFrames = true;
+    public static boolean onlyPlayersPushEntities = false;
+    public static boolean checkLightForSnow = false;
+    public static boolean lavaCatchesBlocksOnFire = true;
+    public static boolean disableFootsteps = false;
+    public static boolean disableRaidsFromSpawnerRaiders = false;
+    public static int spawnerMinSpawnDelay = 200;
+    public static int spawnerMaxSpawnDelay = 800;
+    public static int spawnerSpawnRange = 4;
+    public static int spawnerSpawnCount = 4;
+    public static boolean spawnerSpawnParticles = true;
+    public static int spawnerMaxNearbyEntities = 16;
+    public static int spawnerMaxPlayerRange = 16;
+    public static boolean spawnerEntitiesHaveAI = true;
+
+    private static void general(){
+        disableFootsteps = getBoolean("disableFootsteps", disableFootsteps);
+        checkLightForSnow = getBoolean("checkLightForSnow", checkLightForSnow);
+        onlyPlayersPushEntities = getBoolean("onlyPlayersPushEntities", onlyPlayersPushEntities);
+    }
 
     private static void optimizations(){
         asyncPathfinding = getBoolean("optimizations.async-pathfinding.enable", true);
@@ -209,11 +232,25 @@ public class CanvasConfig {
         mobWanderingDelay = getInt("optimizations.mobWanderingDelay", mobWanderingDelay);
         mobWanderingPercentChance = getInt("optimizations.mobWanderingPercentChance", mobWanderingPercentChance);
         mobWanderingReducedRateMultiplier = getInt("optimizations.mobWanderingReducedRateMultiplier", mobWanderingReducedRateMultiplier);
+        tntLoadChunks = getBoolean("optimizations.tntLoadChunks", tntLoadChunks);
+        enderpearlsLoadChunks = getBoolean("optimizations.enderpearlsLoadChunks", enderpearlsLoadChunks);
+        fallingBlocksLoadChunks = getBoolean("optimizations.fallingBlocksLoadChunks", fallingBlocksLoadChunks);
+        tickItemFrames = getBoolean("optimizations.tickItemFrames", tickItemFrames);
+        disableRaidsFromSpawnerRaiders = getBoolean("optimizations.disableRaidsFromSpawnerRaiders", disableRaidsFromSpawnerRaiders);
     }
     
     private static void blocks(){
         useCanvasBedSetDye = getBoolean("blocks.useCanvasBedSetDyeMethod", useCanvasBedSetDye);
         makeCampfireAlwaysLots = getBoolean("block.forceCampfireEmitLotsOfParticles", makeCampfireAlwaysLots);
         farmlandSearchRadius = getInt("block.farmlandSearchRadius", farmlandSearchRadius);
+        lavaCatchesBlocksOnFire = getBoolean("block.lavaCatchesBlocksOnFire", lavaCatchesBlocksOnFire);
+        spawnerMaxNearbyEntities = getInt("block.spawner.spawnerMaxNearbyEntities", spawnerMaxNearbyEntities);
+        spawnerMaxPlayerRange = getInt("block.spawner.spawnerMaxPlayerRange", spawnerMaxPlayerRange);
+        spawnerEntitiesHaveAI = getBoolean("block.spawner.spawnerEntitiesHaveAI", spawnerEntitiesHaveAI);
+        spawnerSpawnRange = getInt("block.spawner.spawnerSpawnRange", spawnerSpawnRange);
+        spawnerSpawnCount = getInt("block.spawner.spawnerSpawnCount", spawnerSpawnCount);
+        spawnerMaxSpawnDelay = getInt("block.spawner.spawnerMaxSpawnDelay", spawnerMaxSpawnDelay);
+        spawnerMinSpawnDelay = getInt("block.spawner.spawnerMinSpawnDelay", spawnerMinSpawnDelay);
+        spawnerSpawnParticles = getBoolean("block.spawner.spawnerSpawnParticles", spawnerSpawnParticles);
     }
 }
diff --git a/src/main/java/net/minecraft/server/level/ServerEntity.java b/src/main/java/net/minecraft/server/level/ServerEntity.java
index dfd09004063a8d8d93357633c8d23b6acff67b73..6ee78ccb8e63b1642d7d527c549977445c43e00a 100644
--- a/src/main/java/net/minecraft/server/level/ServerEntity.java
+++ b/src/main/java/net/minecraft/server/level/ServerEntity.java
@@ -124,7 +124,7 @@ public class ServerEntity {
 
         Entity entity = this.entity;
 
-        if (!this.trackedPlayers.isEmpty() && entity instanceof ItemFrame entityitemframe) { // Paper - Perf: Only tick item frames if players can see it
+        if (io.github.dueris.canvas.CanvasConfig.tickItemFrames && !this.trackedPlayers.isEmpty() && entity instanceof ItemFrame entityitemframe) { // Paper - Perf: Only tick item frames if players can see it // Canvas
             if (true || this.tickCount % 10 == 0) { // CraftBukkit - Moved below, should always enter this block
                 ItemStack itemstack = entityitemframe.getItem();
 
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 05c8498af6c8a86047d462864f42f9d728577b5e..c3d24a11d638edd049fedfd6a8b361ae17f27ef8 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -2611,12 +2611,19 @@ public class ServerLevel extends Level implements WorldGenLevel, ca.spottedleaf.
         }
 
         public void onTickingEnd(Entity entity) {
+            // Canvas start - More Configs -- Optimizations
+            if(!io.github.dueris.canvas.CanvasConfig.tntLoadChunks && !io.github.dueris.canvas.CanvasConfig.fallingBlocksLoadChunks && !io.github.dueris.canvas.CanvasConfig.enderpearlsLoadChunks || entity.isRemoved() ||
+                (!io.github.dueris.canvas.CanvasConfig.tntLoadChunks || entity.getType() != EntityType.TNT) &&
+                (!io.github.dueris.canvas.CanvasConfig.fallingBlocksLoadChunks || entity.getType() != EntityType.FALLING_BLOCK) &&
+                (!io.github.dueris.canvas.CanvasConfig.enderpearlsLoadChunks || entity.getType() != EntityType.ENDER_PEARL)) {
+            // Canvas end
             ServerLevel.this.entityTickList.remove(entity);
             // Paper start - Reset pearls when they stop being ticked
             if (paperConfig().fixes.disableUnloadedChunkEnderpearlExploit && entity instanceof net.minecraft.world.entity.projectile.ThrownEnderpearl pearl) {
                 pearl.cachedOwner = null;
                 pearl.ownerUUID = null;
             }
+            } // Canvas
             // Paper end - Reset pearls when they stop being ticked
         }
 
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index ece1faa0ebd66ab871fd54f2a604097ec67dadc5..4d70d2a64fed785e2b27a0fdb8415816a507a64a 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -3782,6 +3782,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
     protected void serverAiStep() {}
 
     protected void pushEntities() {
+        if(io.github.dueris.canvas.CanvasConfig.onlyPlayersPushEntities && !(this instanceof ServerPlayer)) return; // Canvas
         if (this.level().isClientSide()) {
             this.level().getEntities(EntityTypeTest.forClass(net.minecraft.world.entity.player.Player.class), this.getBoundingBox(), EntitySelector.pushableBy(this)).forEach(this::doPush);
         } else {
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 2605b605b37ec80c86165c64499c0034c603a43d..76d28def658a27dbf46537219718bd9df0021382 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -1830,6 +1830,7 @@ public abstract class Player extends LivingEntity {
 
     @Override
     protected void playStepSound(BlockPos pos, BlockState state) {
+        if(io.github.dueris.canvas.CanvasConfig.disableFootsteps) return; // Canvas - More Configs -- General
         if (this.isInWater()) {
             this.waterSwimSound();
             this.playMuffledStepSound(state);
diff --git a/src/main/java/net/minecraft/world/entity/raid/Raider.java b/src/main/java/net/minecraft/world/entity/raid/Raider.java
index 06487fc9ea416d8256e0c2cd1969d4e0283ffb05..348982a43b5491515d87b29382df37555da7d0e4 100644
--- a/src/main/java/net/minecraft/world/entity/raid/Raider.java
+++ b/src/main/java/net/minecraft/world/entity/raid/Raider.java
@@ -62,7 +62,9 @@ public abstract class Raider extends PatrollingMonster {
         super.registerGoals();
         this.goalSelector.addGoal(1, new Raider.ObtainRaidLeaderBannerGoal<>(this, this));
         this.goalSelector.addGoal(3, new PathfindToRaidGoal<>(this));
+        if(!spawnedViaMobSpawner || !io.github.dueris.canvas.CanvasConfig.disableRaidsFromSpawnerRaiders){ // Canvas - More Configs -- Optimizations
         this.goalSelector.addGoal(4, new Raider.RaiderMoveThroughVillageGoal(this, 1.0499999523162842D, 1));
+        } // Canvas
         this.goalSelector.addGoal(5, new Raider.RaiderCelebration(this));
     }
 
diff --git a/src/main/java/net/minecraft/world/level/BaseSpawner.java b/src/main/java/net/minecraft/world/level/BaseSpawner.java
index 967af8771ff8564c715d89f4b4b69b16c25add59..44ed64a1a43d75fd32f3df84a7100bd7ac3ae133 100644
--- a/src/main/java/net/minecraft/world/level/BaseSpawner.java
+++ b/src/main/java/net/minecraft/world/level/BaseSpawner.java
@@ -41,14 +41,16 @@ public abstract class BaseSpawner {
     public SpawnData nextSpawnData;
     private double spin;
     private double oSpin;
-    public int minSpawnDelay = 200;
-    public int maxSpawnDelay = 800;
-    public int spawnCount = 4;
+    // Canvas start - Configurable MobSpawners
+    public int minSpawnDelay = io.github.dueris.canvas.CanvasConfig.spawnerMinSpawnDelay == -1 ? 200 : io.github.dueris.canvas.CanvasConfig.spawnerMinSpawnDelay;
+    public int maxSpawnDelay = io.github.dueris.canvas.CanvasConfig.spawnerMaxSpawnDelay == -1 ? 800 : io.github.dueris.canvas.CanvasConfig.spawnerMaxSpawnDelay;
+    public int spawnCount = io.github.dueris.canvas.CanvasConfig.spawnerSpawnCount == -1 ? 4 : io.github.dueris.canvas.CanvasConfig.spawnerSpawnCount;
     @Nullable
     private Entity displayEntity;
-    public int maxNearbyEntities = 6;
-    public int requiredPlayerRange = 16;
-    public int spawnRange = 4;
+    public int maxNearbyEntities = io.github.dueris.canvas.CanvasConfig.spawnerMaxNearbyEntities == -1 ? 6 : io.github.dueris.canvas.CanvasConfig.spawnerMaxNearbyEntities;
+    public int requiredPlayerRange = io.github.dueris.canvas.CanvasConfig.spawnerMaxPlayerRange == -1 ? 16 : io.github.dueris.canvas.CanvasConfig.spawnerMaxPlayerRange;
+    // Canvas end
+    public int spawnRange = io.github.dueris.canvas.CanvasConfig.spawnerSpawnRange == -1 ? 4 : io.github.dueris.canvas.CanvasConfig.spawnerSpawnRange;
     private int tickDelay = 0; // Paper - Configurable mob spawner tick rate
 
     public BaseSpawner() {}
@@ -72,8 +74,10 @@ public abstract class BaseSpawner {
             double d1 = (double) pos.getY() + randomsource.nextDouble();
             double d2 = (double) pos.getZ() + randomsource.nextDouble();
 
+            if(io.github.dueris.canvas.CanvasConfig.spawnerSpawnParticles) { // Canvas - Configurable MobSpawners
             world.addParticle(ParticleTypes.SMOKE, d0, d1, d2, 0.0D, 0.0D, 0.0D);
             world.addParticle(ParticleTypes.FLAME, d0, d1, d2, 0.0D, 0.0D, 0.0D);
+            } // Canvas
             if (this.spawnDelay > 0) {
                 --this.spawnDelay;
             }
@@ -170,6 +174,7 @@ public abstract class BaseSpawner {
                         entity.moveTo(entity.getX(), entity.getY(), entity.getZ(), randomsource.nextFloat() * 360.0F, 0.0F);
                         if (entity instanceof Mob) {
                             Mob entityinsentient = (Mob) entity;
+                            if(!io.github.dueris.canvas.CanvasConfig.spawnerEntitiesHaveAI) entityinsentient.setNoAi(false); // Canvas - Configurable MobSpawners
 
                             if (mobspawnerdata.getCustomSpawnRules().isEmpty() && !entityinsentient.checkSpawnRules(world, MobSpawnType.SPAWNER) || !entityinsentient.checkSpawnObstruction(world)) {
                                 continue;
diff --git a/src/main/java/net/minecraft/world/level/biome/Biome.java b/src/main/java/net/minecraft/world/level/biome/Biome.java
index 90c165c890a2d998e3b0af9b4310e3995ede6f64..e8ee117f415769bc3946be0e7ccee6f7032c8ecd 100644
--- a/src/main/java/net/minecraft/world/level/biome/Biome.java
+++ b/src/main/java/net/minecraft/world/level/biome/Biome.java
@@ -157,7 +157,7 @@ public final class Biome {
         if (this.warmEnoughToRain(pos)) {
             return false;
         } else {
-            if (pos.getY() >= world.getMinBuildHeight() && pos.getY() < world.getMaxBuildHeight() && world.getBrightness(LightLayer.BLOCK, pos) < 10) {
+            if (pos.getY() >= world.getMinBuildHeight() && pos.getY() < world.getMaxBuildHeight() && (!io.github.dueris.canvas.CanvasConfig.checkLightForSnow || world.getBrightness(LightLayer.BLOCK, pos) < 10)) { // Canvas - More Configs -- General
                 BlockState blockState = world.getBlockState(pos);
                 if ((blockState.isAir() || blockState.is(Blocks.SNOW)) && Blocks.SNOW.defaultBlockState().canSurvive(world, pos)) {
                     return true;
diff --git a/src/main/java/net/minecraft/world/level/material/LavaFluid.java b/src/main/java/net/minecraft/world/level/material/LavaFluid.java
index 2d492d849ff73a738dfbcb16507feb89bf19a962..19ae57e2ee537654a6460913374f848ed48e0000 100644
--- a/src/main/java/net/minecraft/world/level/material/LavaFluid.java
+++ b/src/main/java/net/minecraft/world/level/material/LavaFluid.java
@@ -68,7 +68,7 @@ public abstract class LavaFluid extends FlowingFluid {
 
     @Override
     public void randomTick(Level world, BlockPos pos, FluidState state, RandomSource random) {
-        if (world.getGameRules().getBoolean(GameRules.RULE_DOFIRETICK)) {
+        if (world.getGameRules().getBoolean(GameRules.RULE_DOFIRETICK) && io.github.dueris.canvas.CanvasConfig.lavaCatchesBlocksOnFire) { // Canvas - More Configs -- Block
             int i = random.nextInt(3);
 
             if (i > 0) {
