From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <jedimastertoothless@hotmail.com>
Date: Mon, 30 Oct 2023 09:48:01 +0000
Subject: [PATCH] Remove old particle trail method for Fireball


diff --git a/src/main/java/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java b/src/main/java/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
index 326a58ecf85f70ddd1b223352811beb827ca9aad..a985ed12f0409904c90a1bf89265e2c62c212f46 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
@@ -114,7 +114,11 @@ public abstract class AbstractHurtingProjectile extends Projectile {
             }
 
             this.setDeltaMovement(vec3d.add(this.xPower, this.yPower, this.zPower).scale((double) f));
-            this.level().addParticle(this.getTrailParticle(), d0, d1 + 0.5D, d2, 0.0D, 0.0D, 0.0D);
+            // Canvas start
+            if(this.hasTrail){
+                this.level().addParticle(this.trailType, d0, d1 + 0.5D, d2, 0.0D, 0.0D, 0.0D);
+            }
+            // Canvas end
             this.setPos(d0, d1, d2);
         }
     }
@@ -126,6 +130,7 @@ public abstract class AbstractHurtingProjectile extends Projectile {
 
     // Canvas start
     private boolean shouldBurn = true;
+    public boolean hasTrail = true;
     private ParticleOptions trailType = ParticleTypes.SMOKE;
     private float inertia = 0.95f;
 
@@ -145,10 +150,6 @@ public abstract class AbstractHurtingProjectile extends Projectile {
         this.shouldBurn = burn;
     }
 
-    public void setTrailParticle(ParticleOptions trailType) {
-        this.trailType = trailType;
-    }
-
     public void setInertia(float inertia) {
         this.inertia = inertia;
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java
index c808b67959ee7f711cfb82f86bea90432354a895..9f18c48dd04717faa5e8ce5936d945894b709730 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java
@@ -45,14 +45,13 @@ public class CraftFireball extends AbstractProjectile implements Fireball {
     }
 
     @Override
-    public void setParticleTrailType(final Particle particle) {
-        ParticleOptions particleType = CraftParticle.bukkitToMinecraft(particle);
-        this.getHandle().setTrailParticle(particleType);
+    public void setHasParticleTrail(boolean hasParticleTrail) {
+        this.getHandle().hasTrail = hasParticleTrail;
     }
 
     @Override
-    public Particle getParticleTrailType() {
-        return CraftParticle.minecraftToBukkit(this.getHandle().getTrailParticle().getType());
+    public boolean getHasParticleTrail() {
+        return this.getHandle().hasTrail;
     }
 
     @Override
