From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Tue, 21 Nov 2023 01:33:57 +0000
Subject: [PATCH] Add SleepMessage API


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index 150d66079a41c9e11ef518936d1ccac00c8c1c31..0d218362259261470f31de0b82b8d4d5747a3838 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -1242,6 +1242,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
                 int i = this.getGameRules().getInt(GameRules.RULE_PLAYERS_SLEEPING_PERCENTAGE);
                 Component ichatmutablecomponent;
 
+                me.dueris.canvas.event.player.PlayerSleepMessageEvent.SleepMessageType messageType = me.dueris.canvas.event.player.PlayerSleepMessageEvent.SleepMessageType.UNKNOWN; // Canvas
                 if (this.sleepStatus.areEnoughSleeping(i)) {
                     // Purpur start
                     if (org.purpurmc.purpur.PurpurConfig.sleepSkippingNight.isBlank()) {
@@ -1250,6 +1251,7 @@ public class ServerLevel extends Level implements WorldGenLevel {
                     if (!org.purpurmc.purpur.PurpurConfig.sleepSkippingNight.equalsIgnoreCase("default")) {
                         ichatmutablecomponent = io.papermc.paper.adventure.PaperAdventure.asVanilla(net.kyori.adventure.text.minimessage.MiniMessage.miniMessage().deserialize(org.purpurmc.purpur.PurpurConfig.sleepSkippingNight));
                     } else
+                    messageType = me.dueris.canvas.event.player.PlayerSleepMessageEvent.SleepMessageType.SKIPPING_NIGHT; // Canvas
                     ichatmutablecomponent = Component.translatable("sleep.skipping_night");
                 } else {
                     if (org.purpurmc.purpur.PurpurConfig.sleepingPlayersPercent.isBlank()) {
@@ -1263,13 +1265,24 @@ public class ServerLevel extends Level implements WorldGenLevel {
                     // Purpur end
                     ichatmutablecomponent = Component.translatable("sleep.players_sleeping", this.sleepStatus.amountSleeping(), this.sleepStatus.sleepersNeeded(i));
                 }
+                // Canvas start
+                boolean shouldDisplayMessage =
+                    (messageType == me.dueris.canvas.event.player.PlayerSleepMessageEvent.SleepMessageType.SKIPPING_NIGHT)
+                    || (messageType == me.dueris.canvas.event.player.PlayerSleepMessageEvent.SleepMessageType.PLAYERS_SLEEPING_COUNT);
+                var adventureComponent = io.papermc.paper.adventure.PaperAdventure.asAdventure(ichatmutablecomponent);
+                // Canvas end
 
                 Iterator iterator = this.players.iterator();
 
                 while (iterator.hasNext()) {
                     ServerPlayer entityplayer = (ServerPlayer) iterator.next();
 
-                    entityplayer.displayClientMessage(ichatmutablecomponent, true);
+                    // Canvas start
+                    var event = new me.dueris.canvas.event.player.PlayerSleepMessageEvent(entityplayer.getBukkitEntity(), messageType, adventureComponent, !shouldDisplayMessage);
+                    if (event.callEvent()) {
+                        entityplayer.displayClientMessage(io.papermc.paper.adventure.PaperAdventure.asVanilla(event.getMessage()), true);
+                    }
+                    // Canvas end
                 }
 
             }
