From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Fri, 24 May 2024 18:45:49 -0700
Subject: [PATCH] Optimization Configs


diff --git a/src/main/java/me/dueris/canvas/CanvasConfig.java b/src/main/java/me/dueris/canvas/CanvasConfig.java
index 988a5ccbf31e8807b8d60444c16de44297521e51..bd501d0e3b0dcb5836952d3f79d044cf0ebff449 100644
--- a/src/main/java/me/dueris/canvas/CanvasConfig.java
+++ b/src/main/java/me/dueris/canvas/CanvasConfig.java
@@ -185,6 +185,29 @@ public class CanvasConfig {
     public static int mobWanderingPercentChance = 100;
     public static int mobWanderingReducedRateMultiplier = 6;
     public static int mobWanderingDelay = 20 * 5;
+    public static boolean tntLoadChunks = true;
+    public static boolean enderpearlsLoadChunks = true;
+    public static boolean fallingBlocksLoadChunks = true;
+    public static boolean tickItemFrames = true;
+    public static boolean onlyPlayersPushEntities = false;
+    public static boolean checkLightForSnow = false;
+    public static boolean lavaCatchesBlocksOnFire = true;
+    public static boolean disableFootsteps = false;
+    public static boolean disableRaidsFromSpawnerRaiders = false;
+    public static int spawnerMinSpawnDelay = 200;
+    public static int spawnerMaxSpawnDelay = 800;
+    public static int spawnerSpawnRange = 4;
+    public static int spawnerSpawnCount = 4;
+    public static boolean spawnerSpawnParticles = true;
+    public static int spawnerMaxNearbyEntities = 16;
+    public static int spawnerMaxPlayerRange = 16;
+    public static boolean spawnerEntitiesHaveAI = true;
+
+    private static void general(){
+        disableFootsteps = getBoolean("disableFootsteps", disableFootsteps);
+        checkLightForSnow = getBoolean("checkLightForSnow", checkLightForSnow);
+        onlyPlayersPushEntities = getBoolean("onlyPlayersPushEntities", onlyPlayersPushEntities);
+    }
 
     private static void optimizations(){
         asyncPathfinding = getBoolean("optimizations.async-pathfinding.enable", true);
@@ -209,11 +232,25 @@ public class CanvasConfig {
         mobWanderingDelay = getInt("optimizations.mobWanderingDelay", mobWanderingDelay);
         mobWanderingPercentChance = getInt("optimizations.mobWanderingPercentChance", mobWanderingPercentChance);
         mobWanderingReducedRateMultiplier = getInt("optimizations.mobWanderingReducedRateMultiplier", mobWanderingReducedRateMultiplier);
+        tntLoadChunks = getBoolean("optimizations.tntLoadChunks", tntLoadChunks);
+        enderpearlsLoadChunks = getBoolean("optimizations.enderpearlsLoadChunks", enderpearlsLoadChunks);
+        fallingBlocksLoadChunks = getBoolean("optimizations.fallingBlocksLoadChunks", fallingBlocksLoadChunks);
+        tickItemFrames = getBoolean("optimizations.tickItemFrames", tickItemFrames);
+        disableRaidsFromSpawnerRaiders = getBoolean("optimizations.disableRaidsFromSpawnerRaiders", disableRaidsFromSpawnerRaiders);
     }
     
     private static void blocks(){
         useCanvasBedSetDye = getBoolean("blocks.useCanvasBedSetDyeMethod", useCanvasBedSetDye);
         makeCampfireAlwaysLots = getBoolean("block.forceCampfireEmitLotsOfParticles", makeCampfireAlwaysLots);
         farmlandSearchRadius = getInt("block.farmlandSearchRadius", farmlandSearchRadius);
+        lavaCatchesBlocksOnFire = getBoolean("block.lavaCatchesBlocksOnFire", lavaCatchesBlocksOnFire);
+        spawnerMaxNearbyEntities = getInt("block.spawner.spawnerMaxNearbyEntities", spawnerMaxNearbyEntities);
+        spawnerMaxPlayerRange = getInt("block.spawner.spawnerMaxPlayerRange", spawnerMaxPlayerRange);
+        spawnerEntitiesHaveAI = getBoolean("block.spawner.spawnerEntitiesHaveAI", spawnerEntitiesHaveAI);
+        spawnerSpawnRange = getInt("block.spawner.spawnerSpawnRange", spawnerSpawnRange);
+        spawnerSpawnCount = getInt("block.spawner.spawnerSpawnCount", spawnerSpawnCount);
+        spawnerMaxSpawnDelay = getInt("block.spawner.spawnerMaxSpawnDelay", spawnerMaxSpawnDelay);
+        spawnerMinSpawnDelay = getInt("block.spawner.spawnerMinSpawnDelay", spawnerMinSpawnDelay);
+        spawnerSpawnParticles = getBoolean("block.spawner.spawnerSpawnParticles", spawnerSpawnParticles);
     }
 }
diff --git a/src/main/java/net/minecraft/server/level/ServerEntity.java b/src/main/java/net/minecraft/server/level/ServerEntity.java
index 3d4926e1d882bc41157f8402ae63cd96c4dc9c31..e81f89dc2357843de9b1cf8a54a1edaeb1a9a33a 100644
--- a/src/main/java/net/minecraft/server/level/ServerEntity.java
+++ b/src/main/java/net/minecraft/server/level/ServerEntity.java
@@ -124,7 +124,7 @@ public class ServerEntity {
 
         Entity entity = this.entity;
 
-        if (!this.trackedPlayers.isEmpty() && entity instanceof ItemFrame entityitemframe) { // Paper - Perf: Only tick item frames if players can see it
+        if (me.dueris.canvas.CanvasConfig.tickItemFrames && !this.trackedPlayers.isEmpty() && entity instanceof ItemFrame entityitemframe) { // Paper - Perf: Only tick item frames if players can see it // Canvas
             if (true || this.tickCount % 10 == 0) { // CraftBukkit - Moved below, should always enter this block
                 ItemStack itemstack = entityitemframe.getItem();
 
diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index de8c92bc3488d066eaf26c12bd4b36e0479ff924..4d1d6e349519411a16669918ce17c09731009f92 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -2570,12 +2570,19 @@ public class ServerLevel extends Level implements WorldGenLevel, ca.spottedleaf.
         }
 
         public void onTickingEnd(Entity entity) {
+            // Canvas start - More Configs -- Optimizations
+            if(!me.dueris.canvas.CanvasConfig.tntLoadChunks && !me.dueris.canvas.CanvasConfig.fallingBlocksLoadChunks && !me.dueris.canvas.CanvasConfig.enderpearlsLoadChunks || entity.isRemoved() ||
+                (!me.dueris.canvas.CanvasConfig.tntLoadChunks || entity.getType() != EntityType.TNT) &&
+                (!me.dueris.canvas.CanvasConfig.fallingBlocksLoadChunks || entity.getType() != EntityType.FALLING_BLOCK) &&
+                (!me.dueris.canvas.CanvasConfig.enderpearlsLoadChunks || entity.getType() != EntityType.ENDER_PEARL)) {
+            // Canvas end
             ServerLevel.this.entityTickList.remove(entity);
             // Paper start - Reset pearls when they stop being ticked
             if (paperConfig().fixes.disableUnloadedChunkEnderpearlExploit && entity instanceof net.minecraft.world.entity.projectile.ThrownEnderpearl pearl) {
                 pearl.cachedOwner = null;
                 pearl.ownerUUID = null;
             }
+            } // Canvas
             // Paper end - Reset pearls when they stop being ticked
         }
 
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index ee3c3d997f0f844c8e473326ef5596049beb3062..9c27ff39ac3c0e6d7671809a19d05965c08813f2 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -3756,6 +3756,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
     protected void serverAiStep() {}
 
     protected void pushEntities() {
+        if(me.dueris.canvas.CanvasConfig.onlyPlayersPushEntities && !(this instanceof ServerPlayer)) return; // Canvas
         if (this.level().isClientSide()) {
             this.level().getEntities(EntityTypeTest.forClass(net.minecraft.world.entity.player.Player.class), this.getBoundingBox(), EntitySelector.pushableBy(this)).forEach(this::doPush);
         } else {
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 2605b605b37ec80c86165c64499c0034c603a43d..26d111251c221c62ebf557540496e0a1c6ee1b20 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -1830,6 +1830,7 @@ public abstract class Player extends LivingEntity {
 
     @Override
     protected void playStepSound(BlockPos pos, BlockState state) {
+        if(me.dueris.canvas.CanvasConfig.disableFootsteps) return; // Canvas - More Configs -- General
         if (this.isInWater()) {
             this.waterSwimSound();
             this.playMuffledStepSound(state);
diff --git a/src/main/java/net/minecraft/world/entity/raid/Raider.java b/src/main/java/net/minecraft/world/entity/raid/Raider.java
index 06487fc9ea416d8256e0c2cd1969d4e0283ffb05..e73ff6af4be09963642e11a12715c61d7be9a136 100644
--- a/src/main/java/net/minecraft/world/entity/raid/Raider.java
+++ b/src/main/java/net/minecraft/world/entity/raid/Raider.java
@@ -62,7 +62,9 @@ public abstract class Raider extends PatrollingMonster {
         super.registerGoals();
         this.goalSelector.addGoal(1, new Raider.ObtainRaidLeaderBannerGoal<>(this, this));
         this.goalSelector.addGoal(3, new PathfindToRaidGoal<>(this));
+        if(!spawnedViaMobSpawner || !me.dueris.canvas.CanvasConfig.disableRaidsFromSpawnerRaiders){ // Canvas - More Configs -- Optimizations
         this.goalSelector.addGoal(4, new Raider.RaiderMoveThroughVillageGoal(this, 1.0499999523162842D, 1));
+        } // Canvas
         this.goalSelector.addGoal(5, new Raider.RaiderCelebration(this));
     }
 
diff --git a/src/main/java/net/minecraft/world/level/BaseSpawner.java b/src/main/java/net/minecraft/world/level/BaseSpawner.java
index 967af8771ff8564c715d89f4b4b69b16c25add59..93b02b5d96664ec65512f0a41bc62a65b7f88b2c 100644
--- a/src/main/java/net/minecraft/world/level/BaseSpawner.java
+++ b/src/main/java/net/minecraft/world/level/BaseSpawner.java
@@ -41,14 +41,16 @@ public abstract class BaseSpawner {
     public SpawnData nextSpawnData;
     private double spin;
     private double oSpin;
-    public int minSpawnDelay = 200;
-    public int maxSpawnDelay = 800;
-    public int spawnCount = 4;
+    // Canvas start - Configurable MobSpawners
+    public int minSpawnDelay = me.dueris.canvas.CanvasConfig.spawnerMinSpawnDelay == -1 ? 200 : me.dueris.canvas.CanvasConfig.spawnerMinSpawnDelay;
+    public int maxSpawnDelay = me.dueris.canvas.CanvasConfig.spawnerMaxSpawnDelay == -1 ? 800 : me.dueris.canvas.CanvasConfig.spawnerMaxSpawnDelay;
+    public int spawnCount = me.dueris.canvas.CanvasConfig.spawnerSpawnCount == -1 ? 4 : me.dueris.canvas.CanvasConfig.spawnerSpawnCount;
     @Nullable
     private Entity displayEntity;
-    public int maxNearbyEntities = 6;
-    public int requiredPlayerRange = 16;
-    public int spawnRange = 4;
+    public int maxNearbyEntities = me.dueris.canvas.CanvasConfig.spawnerMaxNearbyEntities == -1 ? 6 : me.dueris.canvas.CanvasConfig.spawnerMaxNearbyEntities;
+    public int requiredPlayerRange = me.dueris.canvas.CanvasConfig.spawnerMaxPlayerRange == -1 ? 16 : me.dueris.canvas.CanvasConfig.spawnerMaxPlayerRange;
+    // Canvas end
+    public int spawnRange = me.dueris.canvas.CanvasConfig.spawnerSpawnRange == -1 ? 4 : me.dueris.canvas.CanvasConfig.spawnerSpawnRange;
     private int tickDelay = 0; // Paper - Configurable mob spawner tick rate
 
     public BaseSpawner() {}
@@ -72,8 +74,10 @@ public abstract class BaseSpawner {
             double d1 = (double) pos.getY() + randomsource.nextDouble();
             double d2 = (double) pos.getZ() + randomsource.nextDouble();
 
+            if(me.dueris.canvas.CanvasConfig.spawnerSpawnParticles) { // Canvas - Configurable MobSpawners
             world.addParticle(ParticleTypes.SMOKE, d0, d1, d2, 0.0D, 0.0D, 0.0D);
             world.addParticle(ParticleTypes.FLAME, d0, d1, d2, 0.0D, 0.0D, 0.0D);
+            } // Canvas
             if (this.spawnDelay > 0) {
                 --this.spawnDelay;
             }
@@ -170,6 +174,7 @@ public abstract class BaseSpawner {
                         entity.moveTo(entity.getX(), entity.getY(), entity.getZ(), randomsource.nextFloat() * 360.0F, 0.0F);
                         if (entity instanceof Mob) {
                             Mob entityinsentient = (Mob) entity;
+                            if(!me.dueris.canvas.CanvasConfig.spawnerEntitiesHaveAI) entityinsentient.setNoAi(false); // Canvas - Configurable MobSpawners
 
                             if (mobspawnerdata.getCustomSpawnRules().isEmpty() && !entityinsentient.checkSpawnRules(world, MobSpawnType.SPAWNER) || !entityinsentient.checkSpawnObstruction(world)) {
                                 continue;
diff --git a/src/main/java/net/minecraft/world/level/biome/Biome.java b/src/main/java/net/minecraft/world/level/biome/Biome.java
index 15f82c9a1ce1fef2e951d1b3c7a65e64b82061ea..909c7ca17efad03618e4d1a62c53780c5f2dffa1 100644
--- a/src/main/java/net/minecraft/world/level/biome/Biome.java
+++ b/src/main/java/net/minecraft/world/level/biome/Biome.java
@@ -170,7 +170,7 @@ public final class Biome {
         if (this.warmEnoughToRain(pos)) {
             return false;
         } else {
-            if (pos.getY() >= world.getMinBuildHeight() && pos.getY() < world.getMaxBuildHeight() && world.getBrightness(LightLayer.BLOCK, pos) < 10) {
+            if (pos.getY() >= world.getMinBuildHeight() && pos.getY() < world.getMaxBuildHeight() && (!me.dueris.canvas.CanvasConfig.checkLightForSnow || world.getBrightness(LightLayer.BLOCK, pos) < 10)) { // Canvas - More Configs -- General
                 BlockState blockState = world.getBlockState(pos);
                 if ((blockState.isAir() || blockState.is(Blocks.SNOW)) && Blocks.SNOW.defaultBlockState().canSurvive(world, pos)) {
                     return true;
diff --git a/src/main/java/net/minecraft/world/level/material/LavaFluid.java b/src/main/java/net/minecraft/world/level/material/LavaFluid.java
index 2d492d849ff73a738dfbcb16507feb89bf19a962..f9b32e022444605906c32f608903b8ef22ed97ae 100644
--- a/src/main/java/net/minecraft/world/level/material/LavaFluid.java
+++ b/src/main/java/net/minecraft/world/level/material/LavaFluid.java
@@ -68,7 +68,7 @@ public abstract class LavaFluid extends FlowingFluid {
 
     @Override
     public void randomTick(Level world, BlockPos pos, FluidState state, RandomSource random) {
-        if (world.getGameRules().getBoolean(GameRules.RULE_DOFIRETICK)) {
+        if (world.getGameRules().getBoolean(GameRules.RULE_DOFIRETICK) && me.dueris.canvas.CanvasConfig.lavaCatchesBlocksOnFire) { // Canvas - More Configs -- Block
             int i = random.nextInt(3);
 
             if (i > 0) {
