From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <122416109+Dueris@users.noreply.github.com>
Date: Sun, 14 Apr 2024 19:46:42 +0000
Subject: [PATCH] WardenSonicBoom Event


diff --git a/src/main/java/net/minecraft/world/entity/ai/behavior/warden/SonicBoom.java b/src/main/java/net/minecraft/world/entity/ai/behavior/warden/SonicBoom.java
index fe9c5afadd9041ab32d9e03fe6dab2a2d00262f2..c21274b849240c27ac588118f358dd77de123a57 100644
--- a/src/main/java/net/minecraft/world/entity/ai/behavior/warden/SonicBoom.java
+++ b/src/main/java/net/minecraft/world/entity/ai/behavior/warden/SonicBoom.java
@@ -1,6 +1,7 @@
 package net.minecraft.world.entity.ai.behavior.warden;
 
 import com.google.common.collect.ImmutableMap;
+import me.dueris.canvas.event.entity.WardenSonicBoomEvent;
 import net.minecraft.core.particles.ParticleTypes;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.sounds.SoundEvents;
@@ -59,6 +60,10 @@ public class SonicBoom extends Behavior<Warden> {
 
     @Override
     protected void tick(ServerLevel world, Warden entity, long time) {
+        // Canvas start - WardenSonicBoomEvent
+        WardenSonicBoomEvent wardenSonicBoomEvent = new WardenSonicBoomEvent((org.bukkit.entity.Warden) entity.getBukkitEntity());
+        wardenSonicBoomEvent.callEvent();
+        if(!wardenSonicBoomEvent.isCanceled()){
         entity.getBrain().getMemory(MemoryModuleType.ATTACK_TARGET).ifPresent(target -> entity.getLookControl().setLookAt(target.position()));
         if (!entity.getBrain().hasMemoryValue(MemoryModuleType.SONIC_BOOM_SOUND_DELAY)
             && !entity.getBrain().hasMemoryValue(MemoryModuleType.SONIC_BOOM_SOUND_COOLDOWN)) {
@@ -84,6 +89,7 @@ public class SonicBoom extends Behavior<Warden> {
                     target.push(vec33.x() * e, vec33.y() * d, vec33.z() * e, entity); // Paper - Add EntityKnockbackByEntityEvent and EntityPushedByEntityAttackEvent
                 });
         }
+        } // Canvas end
     }
 
     @Override
