From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <159172167+PurpleWolfMC@users.noreply.github.com>
Date: Thu, 25 Apr 2024 08:41:58 +0000
Subject: [PATCH] Fix MixinPlugin Classloading


diff --git a/src/main/java/io/papermc/paper/plugin/entrypoint/strategy/modern/ModernPluginLoadingStrategy.java b/src/main/java/io/papermc/paper/plugin/entrypoint/strategy/modern/ModernPluginLoadingStrategy.java
index 9af388a8e56806ab44f8c3ef4f97086ce38ef3b4..6f3ed5d5685a0897ab83ad20bc680a09960f7d01 100644
--- a/src/main/java/io/papermc/paper/plugin/entrypoint/strategy/modern/ModernPluginLoadingStrategy.java
+++ b/src/main/java/io/papermc/paper/plugin/entrypoint/strategy/modern/ModernPluginLoadingStrategy.java
@@ -114,9 +114,18 @@ public class ModernPluginLoadingStrategy<T> implements ProviderLoadingStrategy<T
 
                 if (this.configuration.preloadProvider(retrievedProvider)) {
                     T instance = retrievedProvider.createInstance();
-                    if (this.configuration.load(retrievedProvider, instance)) {
-                        loadedPlugins.add(new ProviderPair<>(retrievedProvider, instance));
+                    if(instance != null){ // Canvas
+                        if (this.configuration.load(retrievedProvider, instance)) {
+                            loadedPlugins.add(new ProviderPair<>(retrievedProvider, instance));
+                        }
+                    // Canvas start
+                    }else{
+                        // MixinPlugin instance. We arent loading it the Paper way
+                        String name = this.getClass().getName();
+                        java.util.logging.Logger.getLogger("CanvasPluginLoader").log(java.util.logging.Level.WARNING, "MixinPlugin was found by Canvas by the name of [%name%].".replace("%name%", retrievedProvider.file().getName()));
+                        java.util.logging.Logger.getLogger("CanvasPluginLoader").log(java.util.logging.Level.WARNING, "Plugin startup for MixinPlugins are different than normal, see the PluginTemplate wiki for more info");
                     }
+                    // Canvas end
                 }
             } catch (Throwable ex) {
                 LOGGER.error("Could not load plugin '%s' in folder '%s'".formatted(retrievedProvider.getFileName(), retrievedProvider.getParentSource()), ex); // Paper
diff --git a/src/main/java/io/papermc/paper/plugin/manager/PaperPluginInstanceManager.java b/src/main/java/io/papermc/paper/plugin/manager/PaperPluginInstanceManager.java
index 3e82ea07ca4194844c5528446e2c4a46ff4acee5..ade868f42031458020bb8b131bfa851f186f6413 100644
--- a/src/main/java/io/papermc/paper/plugin/manager/PaperPluginInstanceManager.java
+++ b/src/main/java/io/papermc/paper/plugin/manager/PaperPluginInstanceManager.java
@@ -86,6 +86,7 @@ class PaperPluginInstanceManager {
     }
 
     public void loadPlugin(Plugin provided) {
+        if(provided.getClass().getClassLoader().getClass().getName().toString().equalsIgnoreCase("space.vectrix.ignite.launch.ember.EmberClassLoader")) return; // Canvas
         PluginMeta configuration = provided.getPluginMeta();
 
         this.plugins.add(provided);
@@ -156,6 +157,7 @@ class PaperPluginInstanceManager {
     // Plugins are disabled in order like this inorder to "rougly" prevent
     // their dependencies unloading first. But, eh.
     public void disablePlugins() {
+        me.dueris.canvas.server.MixinPluginHook.pluginDisableHook(org.bukkit.craftbukkit.CraftServer.startupArgs);
         Plugin[] plugins = this.getPlugins();
         for (int i = plugins.length - 1; i >= 0; i--) {
             this.disablePlugin(plugins[i]);
diff --git a/src/main/java/io/papermc/paper/plugin/provider/type/paper/PaperPluginParent.java b/src/main/java/io/papermc/paper/plugin/provider/type/paper/PaperPluginParent.java
index f2bc4d0b55d4c9877a442529e0b144656497dae6..1bf5baa5aadef605a9990bfc11ce94fdcfcc18b7 100644
--- a/src/main/java/io/papermc/paper/plugin/provider/type/paper/PaperPluginParent.java
+++ b/src/main/java/io/papermc/paper/plugin/provider/type/paper/PaperPluginParent.java
@@ -176,6 +176,11 @@ public class PaperPluginParent {
                 }
 
                 this.status = ProviderStatus.INITIALIZED;
+                // Canvas start
+                if(plugin.getClass().getClassLoader().getClass().getName().toString().equalsIgnoreCase("space.vectrix.ignite.launch.ember.EmberClassLoader")){
+                    return null;
+                }
+                // Canvas end
                 return plugin;
             } catch (Throwable throwable) {
                 this.status = ProviderStatus.ERRORED;
diff --git a/src/main/java/me/dueris/canvas/server/MixinPluginHook.java b/src/main/java/me/dueris/canvas/server/MixinPluginHook.java
new file mode 100644
index 0000000000000000000000000000000000000000..c3076835e63139efaebb63331694bbf42d7284b5
--- /dev/null
+++ b/src/main/java/me/dueris/canvas/server/MixinPluginHook.java
@@ -0,0 +1,20 @@
+package me.dueris.canvas.server;
+
+public class MixinPluginHook {
+
+    public static void preLoadHook(String[] args) {
+        // Preload for MixinPlugin instance registry
+    }
+
+    public static void preWorldLoadHook(String[] args) {
+        // PreWorldLoad for MixinPlugin instance registry
+    }
+
+    public static void postWorldLoadHook(String[] args) {
+        // PostWorldLoad for MixinPlugin instance registry
+    }
+
+    public static void pluginDisableHook(String[] args) {
+        // Plugin disable hook for MixinPlugin instance registry
+    }
+}
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 56e0ed7d9d6ebbbd05c2ba3b7124340cb0c93358..4043d0eea0ec578673e387af490d24f9811f8757 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -664,6 +664,13 @@ public final class CraftServer implements Server {
                 this.enablePlugin(plugin);
             }
         }
+        // Canvas start
+        if(type == PluginLoadOrder.POSTWORLD){
+            me.dueris.canvas.server.MixinPluginHook.postWorldLoadHook(org.bukkit.craftbukkit.CraftServer.startupArgs);
+        }else if(type == PluginLoadOrder.STARTUP){
+            me.dueris.canvas.server.MixinPluginHook.preWorldLoadHook(org.bukkit.craftbukkit.CraftServer.startupArgs);
+        }
+        // Canvas end
 
         if (type == PluginLoadOrder.POSTWORLD) {
             // Spigot start - Allow vanilla commands to be forced to be the main command
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 8da2e63744d0cbe970b2f4e36f341880f76b0864..4e50ce4083302f510381cd76c1bfd8802fead272 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -26,6 +26,7 @@ public class Main {
     // Paper end - Reset loggers after shutdown
 
     public static void main(String[] args) {
+        me.dueris.canvas.server.MixinPluginHook.preLoadHook(args); // Canvas - MixinPlugin injection point
         // Paper start
         final String warnWhenLegacyFormattingDetected = String.join(".", "net", "kyori", "adventure", "text", "warnWhenLegacyFormattingDetected");
         if (false && System.getProperty(warnWhenLegacyFormattingDetected) == null) {
