From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Wed, 19 Jun 2024 20:19:06 -0700
Subject: [PATCH] More Constants Improvements


diff --git a/src/main/java/me/dueris/canvas/Constants.java b/src/main/java/me/dueris/canvas/Constants.java
new file mode 100644
index 0000000000000000000000000000000000000000..1baff312ae42638d352ea5f4ab3e819e2df3158a
--- /dev/null
+++ b/src/main/java/me/dueris/canvas/Constants.java
@@ -0,0 +1,25 @@
+package me.dueris.canvas;
+
+import net.minecraft.world.entity.ai.behavior.DoNothing;
+import net.minecraft.world.item.ItemCooldowns;
+import net.minecraft.world.phys.Vec3;
+import java.io.File;
+
+public interface Constants {
+    Object[] EMPTY_object_arr = new Object[0];
+    int[] EMPTY_int_arr = new int[0];
+    int[] ZERO_int_arr = new int[]{0};
+    byte[] EMPTY_byte_arr = new byte[0];
+    String[] EMPTY_string_arr = new String[0];
+    long[] EMPTY_long_arr = new long[0];
+    org.bukkit.entity.Entity[] EMPTY_bukkit_entity_arr = new org.bukkit.entity.Entity[0];
+    net.minecraft.world.entity.Entity[] EMPTY_entity_arr = new net.minecraft.world.entity.Entity[0];
+    File SERVER_DIRECTORY = new File(".");
+    int[] ZERO = new int[0];
+    Vec3 EMPTY_VECTOR3 = new Vec3(0.0D, 0.0D, 1.0D);
+    ItemCooldowns ITEM_COOLDOWNS = new ItemCooldowns();
+    float[] WHITE_SHEEP_COLOR = new float[]{0.9019608F, 0.9019608F, 0.9019608F};
+    DoNothing DO_NOTHING = new DoNothing(30, 60);
+    DoNothing VILLAGERS_DO_NOTHING = new DoNothing(20, 40);
+    DoNothing PIGLIN_DO_NOTHING = new DoNothing(10, 20);
+}
diff --git a/src/main/java/me/dueris/canvas/SoundConstants.java b/src/main/java/me/dueris/canvas/SoundConstants.java
new file mode 100644
index 0000000000000000000000000000000000000000..581e5fd66a9a79f3a21b2c46edb5faadd43a1bec
--- /dev/null
+++ b/src/main/java/me/dueris/canvas/SoundConstants.java
@@ -0,0 +1,11 @@
+package me.dueris.canvas;
+
+import net.minecraft.sounds.SoundEvents;
+import net.minecraft.world.entity.LivingEntity;
+
+public interface SoundConstants {
+    LivingEntity.Fallsounds LIVING_ENTITY = new LivingEntity.Fallsounds(SoundEvents.GENERIC_SMALL_FALL, SoundEvents.GENERIC_BIG_FALL);
+    LivingEntity.Fallsounds PLAYER = new LivingEntity.Fallsounds(SoundEvents.PLAYER_SMALL_FALL, SoundEvents.PLAYER_BIG_FALL);
+    LivingEntity.Fallsounds MONSTER = new LivingEntity.Fallsounds(SoundEvents.HOSTILE_SMALL_FALL, SoundEvents.HOSTILE_BIG_FALL);
+    LivingEntity.Fallsounds ARMOR_STAND = new LivingEntity.Fallsounds(SoundEvents.ARMOR_STAND_FALL, SoundEvents.ARMOR_STAND_FALL);
+}
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 728417089e0c239dd0fc5257539a30b4973ee264..36faa9e22345250a19c3a08e1d33f2073ba51fad 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1630,7 +1630,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     }
 
     public Path getServerDirectory() {
-        return Path.of("");
+        return me.dueris.canvas.Constants.SERVER_DIRECTORY.toPath(); // Canvas
     }
 
     public void onServerCrash(CrashReport report) {}
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 9c27ff39ac3c0e6d7671809a19d05965c08813f2..892a7fa6071e0aca37b05cfb8f657b72ee108671 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -2104,7 +2104,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
     }
 
     public LivingEntity.Fallsounds getFallSounds() {
-        return new LivingEntity.Fallsounds(SoundEvents.GENERIC_SMALL_FALL, SoundEvents.GENERIC_BIG_FALL);
+        return me.dueris.canvas.SoundConstants.LIVING_ENTITY; // Canvas
     }
 
     protected SoundEvent getDrinkingSound(ItemStack stack) {
diff --git a/src/main/java/net/minecraft/world/entity/ai/behavior/VillagerGoalPackages.java b/src/main/java/net/minecraft/world/entity/ai/behavior/VillagerGoalPackages.java
index 3091d985ba9c55d404332576320718840538722e..abca41a75924080513acf6216feb3bb9b4f788de 100644
--- a/src/main/java/net/minecraft/world/entity/ai/behavior/VillagerGoalPackages.java
+++ b/src/main/java/net/minecraft/world/entity/ai/behavior/VillagerGoalPackages.java
@@ -102,7 +102,7 @@ public class VillagerGoalPackages {
                         Pair.of(VillageBoundRandomStroll.create(speed), 1),
                         Pair.of(SetWalkTargetFromLookTarget.create(speed, 2), 1),
                         Pair.of(new JumpOnBed(speed), 2),
-                        Pair.of(new DoNothing(20, 40), 2)
+                        Pair.of(me.dueris.canvas.Constants.VILLAGERS_DO_NOTHING, 2) // Canvas
                     )
                 )
             ),
@@ -123,7 +123,7 @@ public class VillagerGoalPackages {
                         Pair.of(SetClosestHomeAsWalkTarget.create(speed), 1),
                         Pair.of(InsideBrownianWalk.create(speed), 4),
                         Pair.of(GoToClosestVillage.create(speed, 4), 2),
-                        Pair.of(new DoNothing(20, 40), 2)
+                        Pair.of(me.dueris.canvas.Constants.VILLAGERS_DO_NOTHING, 2) // Canvas
                     )
                 )
             ),
@@ -172,7 +172,7 @@ public class VillagerGoalPackages {
                         Pair.of(VillageBoundRandomStroll.create(speed), 1),
                         Pair.of(SetWalkTargetFromLookTarget.create(speed, 2), 1),
                         Pair.of(new JumpOnBed(speed), 1),
-                        Pair.of(new DoNothing(30, 60), 1)
+                        Pair.of(me.dueris.canvas.Constants.DO_NOTHING, 1)
                     )
                 )
             ),
@@ -272,7 +272,7 @@ public class VillagerGoalPackages {
                     Pair.of(SetEntityLookTarget.create(MobCategory.UNDERGROUND_WATER_CREATURE, 8.0F), 1),
                     Pair.of(SetEntityLookTarget.create(MobCategory.WATER_AMBIENT, 8.0F), 1),
                     Pair.of(SetEntityLookTarget.create(MobCategory.MONSTER, 8.0F), 1),
-                    Pair.of(new DoNothing(30, 60), 2)
+                    Pair.of(me.dueris.canvas.Constants.DO_NOTHING, 2) // Canvas
                 )
             )
         );
@@ -285,7 +285,7 @@ public class VillagerGoalPackages {
                 ImmutableList.of(
                     Pair.of(SetEntityLookTarget.create(EntityType.VILLAGER, 8.0F), 2),
                     Pair.of(SetEntityLookTarget.create(EntityType.PLAYER, 8.0F), 2),
-                    Pair.of(new DoNothing(30, 60), 8)
+                    Pair.of(me.dueris.canvas.Constants.DO_NOTHING, 8) // Canvas
                 )
             )
         );
diff --git a/src/main/java/net/minecraft/world/entity/animal/Pig.java b/src/main/java/net/minecraft/world/entity/animal/Pig.java
index ce2a710fb258429a1aac3843b7afafeaefb11f49..6c65605562de29b54333330f8ab9aee8aa304782 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Pig.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Pig.java
@@ -334,7 +334,7 @@ public class Pig extends Animal implements ItemSteerable, Saddleable {
 
     @Override
     protected Vec3 getRiddenInput(Player controllingPlayer, Vec3 movementInput) {
-        return new Vec3(0.0D, 0.0D, 1.0D);
+        return me.dueris.canvas.Constants.EMPTY_VECTOR3; // Canvas
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java b/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java
index c0383771b08e1d89dfa4996b979248da1de8aca1..6226cc46a95f522d46d5dc08fd1930264b05195f 100644
--- a/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java
+++ b/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java
@@ -927,7 +927,7 @@ public class ArmorStand extends LivingEntity {
 
     @Override
     public LivingEntity.Fallsounds getFallSounds() {
-        return new LivingEntity.Fallsounds(SoundEvents.ARMOR_STAND_FALL, SoundEvents.ARMOR_STAND_FALL);
+        return me.dueris.canvas.SoundConstants.ARMOR_STAND; // Canvas
     }
 
     @Nullable
diff --git a/src/main/java/net/minecraft/world/entity/monster/Monster.java b/src/main/java/net/minecraft/world/entity/monster/Monster.java
index 6511b8c737a57fbc72122dd04e2b544987a49b92..153cab857bfcb70f197552a93ea418dd317f98a0 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Monster.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Monster.java
@@ -78,7 +78,7 @@ public abstract class Monster extends PathfinderMob implements Enemy {
 
     @Override
     public LivingEntity.Fallsounds getFallSounds() {
-        return new LivingEntity.Fallsounds(SoundEvents.HOSTILE_SMALL_FALL, SoundEvents.HOSTILE_BIG_FALL);
+        return me.dueris.canvas.SoundConstants.MONSTER; // Canvas
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/monster/Strider.java b/src/main/java/net/minecraft/world/entity/monster/Strider.java
index adf2bc17713622d540253fbb3d2ffbb81fa84c34..ebaf3141cdabf706d6a31966f490fbbc1a711053 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Strider.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Strider.java
@@ -322,7 +322,7 @@ public class Strider extends Animal implements ItemSteerable, Saddleable {
 
     @Override
     protected Vec3 getRiddenInput(Player controllingPlayer, Vec3 movementInput) {
-        return new Vec3(0.0D, 0.0D, 1.0D);
+        return me.dueris.canvas.Constants.EMPTY_VECTOR3; // Canvas
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/monster/Zoglin.java b/src/main/java/net/minecraft/world/entity/monster/Zoglin.java
index 0cf110be4ec7823996e89c1d1e014e091c1d3d8a..70a23df0cd74c13771555e1bd3d2b3d2df984428 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Zoglin.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Zoglin.java
@@ -142,7 +142,7 @@ public class Zoglin extends Monster implements Enemy, HoglinBase {
                 SetEntityLookTargetSometimes.create(8.0F, UniformInt.of(30, 60)),
                 new RunOne<>(
                     ImmutableList.of(
-                        Pair.of(RandomStroll.stroll(0.4F), 2), Pair.of(SetWalkTargetFromLookTarget.create(0.4F, 3), 2), Pair.of(new DoNothing(30, 60), 1)
+                        Pair.of(RandomStroll.stroll(0.4F), 2), Pair.of(SetWalkTargetFromLookTarget.create(0.4F, 3), 2), Pair.of(me.dueris.canvas.Constants.DO_NOTHING, 1)
                     )
                 )
             )
diff --git a/src/main/java/net/minecraft/world/entity/monster/piglin/PiglinAi.java b/src/main/java/net/minecraft/world/entity/monster/piglin/PiglinAi.java
index d079d5492b6ed709b1e0a7d84fb5f8f6896fc2fe..d8eea6e43bb8050e65585cc3918b5f53b167e694 100644
--- a/src/main/java/net/minecraft/world/entity/monster/piglin/PiglinAi.java
+++ b/src/main/java/net/minecraft/world/entity/monster/piglin/PiglinAi.java
@@ -159,7 +159,7 @@ public class PiglinAi {
     private static void initCelebrateActivity(Brain<Piglin> brain) {
         brain.addActivityAndRemoveMemoryWhenStopped(Activity.CELEBRATE, 10, ImmutableList.of(PiglinAi.avoidRepellent(), SetEntityLookTarget.create(PiglinAi::isPlayerHoldingLovedItem, 14.0F), StartAttacking.create(AbstractPiglin::isAdult, PiglinAi::findNearestValidAttackTarget), BehaviorBuilder.triggerIf((entitypiglin) -> {
             return !entitypiglin.isDancing();
-        }, GoToTargetLocation.create(MemoryModuleType.CELEBRATE_LOCATION, 2, 1.0F)), BehaviorBuilder.triggerIf(Piglin::isDancing, GoToTargetLocation.create(MemoryModuleType.CELEBRATE_LOCATION, 4, 0.6F)), new RunOne<>(ImmutableList.of(Pair.of(SetEntityLookTarget.create(EntityType.PIGLIN, 8.0F), 1), Pair.of(RandomStroll.stroll(0.6F, 2, 1), 1), Pair.of(new DoNothing(10, 20), 1)))), MemoryModuleType.CELEBRATE_LOCATION);
+        }, GoToTargetLocation.create(MemoryModuleType.CELEBRATE_LOCATION, 2, 1.0F)), BehaviorBuilder.triggerIf(Piglin::isDancing, GoToTargetLocation.create(MemoryModuleType.CELEBRATE_LOCATION, 4, 0.6F)), new RunOne<>(ImmutableList.of(Pair.of(SetEntityLookTarget.create(EntityType.PIGLIN, 8.0F), 1), Pair.of(RandomStroll.stroll(0.6F, 2, 1), 1), Pair.of(me.dueris.canvas.Constants.PIGLIN_DO_NOTHING, 1)))), MemoryModuleType.CELEBRATE_LOCATION); // Canvas
     }
 
     private static void initAdmireItemActivity(Brain<Piglin> brain) {
@@ -182,11 +182,11 @@ public class PiglinAi {
     }
 
     private static RunOne<LivingEntity> createIdleLookBehaviors() {
-        return new RunOne<>(ImmutableList.<Pair<? extends BehaviorControl<? super LivingEntity>, Integer>>builder().addAll(PiglinAi.createLookBehaviors()).add(Pair.of(new DoNothing(30, 60), 1)).build()); // CraftBukkit - decompile error
+        return new RunOne<>(ImmutableList.<Pair<? extends BehaviorControl<? super LivingEntity>, Integer>>builder().addAll(PiglinAi.createLookBehaviors()).add(Pair.of(me.dueris.canvas.Constants.DO_NOTHING, 1)).build()); // CraftBukkit - decompile error // Canvas
     }
 
     private static RunOne<Piglin> createIdleMovementBehaviors() {
-        return new RunOne<>(ImmutableList.of(Pair.of(RandomStroll.stroll(0.6F), 2), Pair.of(InteractWith.of(EntityType.PIGLIN, 8, MemoryModuleType.INTERACTION_TARGET, 0.6F, 2), 2), Pair.of(BehaviorBuilder.triggerIf(PiglinAi::doesntSeeAnyPlayerHoldingLovedItem, SetWalkTargetFromLookTarget.create(0.6F, 3)), 2), Pair.of(new DoNothing(30, 60), 1)));
+        return new RunOne<>(ImmutableList.of(Pair.of(RandomStroll.stroll(0.6F), 2), Pair.of(InteractWith.of(EntityType.PIGLIN, 8, MemoryModuleType.INTERACTION_TARGET, 0.6F, 2), 2), Pair.of(BehaviorBuilder.triggerIf(PiglinAi::doesntSeeAnyPlayerHoldingLovedItem, SetWalkTargetFromLookTarget.create(0.6F, 3)), 2), Pair.of(me.dueris.canvas.Constants.DO_NOTHING, 1))); // Canvas
     }
 
     private static BehaviorControl<PathfinderMob> avoidRepellent() {
diff --git a/src/main/java/net/minecraft/world/entity/monster/warden/WardenAi.java b/src/main/java/net/minecraft/world/entity/monster/warden/WardenAi.java
index d4e0ca72aaf7246cc641c8e3d38a0abe8dda18b9..29cba08e2e9b76bd8689419760396ca7f9b23af7 100644
--- a/src/main/java/net/minecraft/world/entity/monster/warden/WardenAi.java
+++ b/src/main/java/net/minecraft/world/entity/monster/warden/WardenAi.java
@@ -142,7 +142,7 @@ public class WardenAi {
                 TryToSniff.create(),
                 new RunOne<>(
                     ImmutableMap.of(MemoryModuleType.IS_SNIFFING, MemoryStatus.VALUE_ABSENT),
-                    ImmutableList.of(Pair.of(RandomStroll.stroll(0.5F), 2), Pair.of(new DoNothing(30, 60), 1))
+                    ImmutableList.of(Pair.of(RandomStroll.stroll(0.5F), 2), Pair.of(me.dueris.canvas.Constants.DO_NOTHING, 1)) // Canvas
                 )
             )
         );
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 26d111251c221c62ebf557540496e0a1c6ee1b20..d72b791691f6ad8156104f861db0dbc81034886c 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -408,7 +408,7 @@ public abstract class Player extends LivingEntity {
     }
 
     protected ItemCooldowns createItemCooldowns() {
-        return new ItemCooldowns();
+        return me.dueris.canvas.Constants.ITEM_COOLDOWNS; // Canvas
     }
 
     private void moveCloak() {
@@ -1854,7 +1854,7 @@ public abstract class Player extends LivingEntity {
 
     @Override
     public LivingEntity.Fallsounds getFallSounds() {
-        return new LivingEntity.Fallsounds(SoundEvents.PLAYER_SMALL_FALL, SoundEvents.PLAYER_BIG_FALL);
+        return me.dueris.canvas.SoundConstants.PLAYER;
     }
 
     @Override
