From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Wed, 19 Jun 2024 20:21:32 -0700
Subject: [PATCH] Optimize Experience Orbs


diff --git a/src/main/java/me/dueris/canvas/CanvasConfig.java b/src/main/java/me/dueris/canvas/CanvasConfig.java
index bd501d0e3b0dcb5836952d3f79d044cf0ebff449..9e84d1e11cbfd253adfc31e42cd1881c3115527f 100644
--- a/src/main/java/me/dueris/canvas/CanvasConfig.java
+++ b/src/main/java/me/dueris/canvas/CanvasConfig.java
@@ -203,6 +203,10 @@ public class CanvasConfig {
     public static int spawnerMaxPlayerRange = 16;
     public static boolean spawnerEntitiesHaveAI = true;
 
+    public static boolean playerInstaAbsorbOrbs = false;
+    public static boolean useCanvasOrbOptimizations = false;
+    public static double orbMergeSearchInfaltionRadius = 0.8;
+
     private static void general(){
         disableFootsteps = getBoolean("disableFootsteps", disableFootsteps);
         checkLightForSnow = getBoolean("checkLightForSnow", checkLightForSnow);
@@ -237,6 +241,9 @@ public class CanvasConfig {
         fallingBlocksLoadChunks = getBoolean("optimizations.fallingBlocksLoadChunks", fallingBlocksLoadChunks);
         tickItemFrames = getBoolean("optimizations.tickItemFrames", tickItemFrames);
         disableRaidsFromSpawnerRaiders = getBoolean("optimizations.disableRaidsFromSpawnerRaiders", disableRaidsFromSpawnerRaiders);
+        playerInstaAbsorbOrbs = getBoolean("playerInstaAbsorbOrbs", playerInstaAbsorbOrbs);
+        useCanvasOrbOptimizations = getBoolean("useCanvasOrbOptimizations", useCanvasOrbOptimizations);
+        orbMergeSearchInfaltionRadius = getDouble("orbMergeSearchInfaltionRadius", orbMergeSearchInfaltionRadius);
     }
     
     private static void blocks(){
diff --git a/src/main/java/net/minecraft/world/entity/ExperienceOrb.java b/src/main/java/net/minecraft/world/entity/ExperienceOrb.java
index 7d724ee7ecf4f50ea1bc4fb85c58988a4d52c9eb..4503c02a8f01cfbeffeaa6f902a23dd47955d85c 100644
--- a/src/main/java/net/minecraft/world/entity/ExperienceOrb.java
+++ b/src/main/java/net/minecraft/world/entity/ExperienceOrb.java
@@ -25,6 +25,7 @@ import net.minecraft.world.level.entity.EntityTypeTest;
 import net.minecraft.world.phys.AABB;
 import net.minecraft.world.phys.Vec3;
 // CraftBukkit start
+import org.bukkit.craftbukkit.entity.CraftExperienceOrb;
 import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.event.entity.EntityRemoveEvent;
 import org.bukkit.event.entity.EntityTargetLivingEntityEvent;
@@ -148,7 +149,21 @@ public class ExperienceOrb extends Entity {
         }
 
         if (this.tickCount % 20 == 1) {
-            this.scanForEntities();
+            // Canvas start
+            if (this.followingPlayer == null || this.followingPlayer.distanceToSqr((Entity) this) > 64.0D) {
+                this.followingPlayer = this.level().getNearestPlayer(this, 8.0D);
+            }
+            if(me.dueris.canvas.CanvasConfig.useCanvasOrbOptimizations){
+                List<ExperienceOrb> orbs = this.level().getEntities(EntityTypeTest.forClass(ExperienceOrb.class), this.getBoundingBox().inflate(me.dueris.canvas.CanvasConfig.orbMergeSearchInfaltionRadius), orb -> orb != this && !orb.isRemoved());
+                if(orbs.size() > 0){
+                    orbs.forEach(orb -> {
+                        this.count += orb.count;
+                        this.age = Math.min(this.age, orb.age);
+                        orb.remove(RemovalReason.DISCARDED);
+                    });
+                }
+            } else this.scanForEntities();
+            // Canvas end
         }
 
         if (this.followingPlayer != null && (this.followingPlayer.isSpectator() || this.followingPlayer.isDeadOrDying())) {
@@ -206,9 +221,10 @@ public class ExperienceOrb extends Entity {
     }
 
     private void scanForEntities() {
+        /* // Canvas
         if (this.followingPlayer == null || this.followingPlayer.distanceToSqr((Entity) this) > 64.0D) {
             this.followingPlayer = this.level().getNearestPlayer(this, 8.0D);
-        }
+        } */ // Canvas
 
         if (this.level() instanceof ServerLevel) {
             List<ExperienceOrb> list = this.level().getEntities(EntityTypeTest.forClass(ExperienceOrb.class), this.getBoundingBox().inflate(0.5D), this::canMerge);
@@ -335,8 +351,13 @@ public class ExperienceOrb extends Entity {
     @Override
     public void playerTouch(Player player) {
         if (player instanceof ServerPlayer entityplayer) {
-            if (player.takeXpDelay == 0 && new com.destroystokyo.paper.event.player.PlayerPickupExperienceEvent(entityplayer.getBukkitEntity(), (org.bukkit.entity.ExperienceOrb) this.getBukkitEntity()).callEvent()) { // Paper - PlayerPickupExperienceEvent
+            if ((me.dueris.canvas.CanvasConfig.playerInstaAbsorbOrbs || player.takeXpDelay == 0) && new com.destroystokyo.paper.event.player.PlayerPickupExperienceEvent(entityplayer.getBukkitEntity(), (org.bukkit.entity.ExperienceOrb) this.getBukkitEntity()).callEvent()) { // Paper - PlayerPickupExperienceEvent // Canvas
                 player.takeXpDelay = CraftEventFactory.callPlayerXpCooldownEvent(player, this.level().purpurConfig.playerExpPickupDelay, PlayerExpCooldownChangeEvent.ChangeReason.PICKUP_ORB).getNewCooldown(); // CraftBukkit - entityhuman.takeXpDelay = 2; // Purpur
+                // Canvas start
+                if(me.dueris.canvas.CanvasConfig.playerInstaAbsorbOrbs) {
+                    player.takeXpDelay = 0;
+                }
+                // Canvas end
                 player.take(this, 1);
                 int i = this.repairPlayerItems(entityplayer, this.value);
 
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index fe723c10cc09c49e5486f75965fe26f3d7fb72e3..d24f3ef757090e255572ea7c68302dd85b361a6e 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -636,7 +636,7 @@ public abstract class Player extends LivingEntity {
             while (iterator.hasNext()) {
                 Entity entity = (Entity) iterator.next();
 
-                if (entity.getType() == EntityType.EXPERIENCE_ORB && entity.level().purpurConfig.playerExpPickupDelay >= 0) { // Purpur
+                if ((entity.getType() == EntityType.EXPERIENCE_ORB && entity.level().purpurConfig.playerExpPickupDelay >= 0) && !me.dueris.canvas.CanvasConfig.playerInstaAbsorbOrbs) { // Purpur // Canvas
                     list1.add(entity);
                 } else if (!entity.isRemoved()) {
                     this.touch(entity);
