From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sat, 4 Nov 2023 13:32:28 -0700
Subject: [PATCH] Fix error causing logs in gui to stop


diff --git a/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java b/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
index 2288fafa9402254c2a84298e4d3bb98ca2be2e67..ba1b6f78b2e8cd2f8294a02de931844f6e930245 100644
--- a/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
+++ b/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
@@ -315,8 +315,12 @@ public class MinecraftServerGui extends CraftServerGui implements MojangServerGu
             textArea.append(message.replaceAll("\u001B\\[[;\\d]*m", "") + "\n"); // Purpur // Canvas
 
             if (atBottom) {
-                textArea.setCaretPosition(textArea.getDocument().getLength());
-                jscrollbar.setValue(Integer.MAX_VALUE);
+                try {
+                    textArea.setCaretPosition(textArea.getDocument().getLength());
+                    jscrollbar.setValue(Integer.MAX_VALUE);
+                } catch (Exception immaJustIgnoreThis){
+                    //ignore bc idk why this is erroring(also it does nothing)
+                }
             }
         }
     }
