From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sat, 4 Nov 2023 13:32:28 -0700
Subject: [PATCH] Fix error causing logs in gui to stop


diff --git a/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java b/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
index 7e3810d2609c7a89ac775de392ace4ed165ccc43..12df39e90acc4355b92326fa80d9c2533e040c0a 100644
--- a/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
+++ b/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
@@ -316,8 +316,12 @@ public class MinecraftServerGui extends CraftServerGui implements MojangServerGu
             textArea.append(message.replaceAll("\u001B\\[[;\\d]*m", "") + "\n"); // Purpur // Canvas
 
             if (atBottom) {
-                textArea.setCaretPosition(textArea.getDocument().getLength());
-                jscrollbar.setValue(Integer.MAX_VALUE);
+                try {
+                    textArea.setCaretPosition(textArea.getDocument().getLength());
+                    jscrollbar.setValue(Integer.MAX_VALUE);
+                } catch (Exception immaJustIgnoreThis){
+                    //ignore bc idk why this is erroring(also it does nothing)
+                }
             }
         }
     }
