From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <122416109+Dueris@users.noreply.github.com>
Date: Mon, 1 Apr 2024 20:53:00 +0000
Subject: [PATCH] ThreadRipper


diff --git a/src/main/java/me/dueris/canvas/CanvasConfig.java b/src/main/java/me/dueris/canvas/CanvasConfig.java
index 7dfd5fa4772cf21f25bfb18c28851b3ab38a7d97..47a97e19b4c8dbe28f1aba9c5a2724b3ab68287b 100644
--- a/src/main/java/me/dueris/canvas/CanvasConfig.java
+++ b/src/main/java/me/dueris/canvas/CanvasConfig.java
@@ -186,6 +186,7 @@ public class CanvasConfig {
     public static int mobWanderingPercentChance = 100;
     public static int mobWanderingReducedRateMultiplier = 6;
     public static int mobWanderingDelay = 20 * 5;
+    public static boolean threadripper = false;
 
     private static void optimizations(){
         maxTickerThreads = getInt("optimizations.maxTickerThreads", maxTickerThreads);
@@ -211,6 +212,7 @@ public class CanvasConfig {
         mobWanderingDelay = getInt("optimizations.mobWanderingDelay", mobWanderingDelay);
         mobWanderingPercentChance = getInt("optimizations.mobWanderingPercentChance", mobWanderingPercentChance);
         mobWanderingReducedRateMultiplier = getInt("optimizations.mobWanderingReducedRateMultiplier", mobWanderingReducedRateMultiplier);
+        threadripper = getBoolean("optimizations.threadripper", threadripper);
     }
     
     private static void blocks(){
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 6b8e985317df6fee56a4785583c9fcdfd0cf406f..a173e45ded71d57c9c817f0da4af18faaf68a3c5 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -229,6 +229,14 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
             DedicatedServer.LOGGER.error("Unable to load canvas-server configuration", e);
             return false;
         }
+        // Check threadripper
+
+        if(CanvasConfig.threadripper){
+            DedicatedServer.LOGGER.info("------------------------------------------------------------------------------------------------------------------");
+            DedicatedServer.LOGGER.info("You have enabled threadripper settings, this requires many threads/cores to truely have an impact on your server.");
+            DedicatedServer.LOGGER.info("There are a wide array of issues that may occur. Proceed with caution.");
+            DedicatedServer.LOGGER.info("------------------------------------------------------------------------------------------------------------------");
+        }
         // Canvas end
         // Paper start
         io.papermc.paper.util.ObfHelper.INSTANCE.getClass(); // Paper - load mappings for stacktrace deobf and etc.
