From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <122416109+Dueris@users.noreply.github.com>
Date: Mon, 18 Mar 2024 17:38:59 +0000
Subject: [PATCH] Enhance Projectile API


diff --git a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
index a9289c4179a78862361be87aaa83f49d6bf60714..2aea97fba2b37a58a67cc5e41eba9796fcda888d 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
@@ -717,7 +717,7 @@ public abstract class AbstractArrow extends Projectile {
 
     }
 
-    protected float getWaterInertia() {
+    public float getWaterInertia() { // Canvas - protected -> public
         return 0.6F;
     }
 
diff --git a/src/main/java/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java b/src/main/java/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
index ce26e74b3c941fa348ea2d85e3176a3fc37496ea..677c0f4e960e5d4da6dbbf13440532d84e234a94 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
@@ -147,6 +147,14 @@ public abstract class AbstractHurtingProjectile extends Projectile {
     // Canvas start
     private boolean shouldBurn = true;
     public boolean hasTrail = true;
+
+    public void setHasTrail(boolean hasTrail){
+        this.hasTrail = hasTrail;
+    }
+
+    public boolean hasTrail(){
+        return this.hasTrail;
+    }
     private ParticleOptions trailType = ParticleTypes.SMOKE;
     private float inertia = 0.95f;
 
diff --git a/src/main/java/net/minecraft/world/entity/projectile/DragonFireball.java b/src/main/java/net/minecraft/world/entity/projectile/DragonFireball.java
index 9d5bcfeb7bd59d9fe432d10eb8a463675cb347c5..d134a742fb5cb18a84aa62ce9fb0688b7488f377 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/DragonFireball.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/DragonFireball.java
@@ -29,6 +29,24 @@ public class DragonFireball extends AbstractHurtingProjectile {
     public DragonFireball(Level world, LivingEntity owner, double directionX, double directionY, double directionZ) {
         super(EntityType.DRAGON_FIREBALL, owner, directionX, directionY, directionZ, world);
     }
+    // Canvas start
+
+    public float radius = 3.0F;
+    public int duration = 600;
+    public int cloudAmplifier = 1;
+
+    public void setRadius(float radius){
+        this.radius = radius;
+    }
+
+    public void setDuration(int duration){
+        this.duration = duration;
+    }
+
+    public void setCloudAmplifier(int cloudAmplifier){
+        this.cloudAmplifier = cloudAmplifier;
+    }
+    // Canvas end
 
     @Override
     protected void onHit(HitResult hitResult) {
@@ -44,10 +62,10 @@ public class DragonFireball extends AbstractHurtingProjectile {
                 }
 
                 entityareaeffectcloud.setParticle(ParticleTypes.DRAGON_BREATH);
-                entityareaeffectcloud.setRadius(3.0F);
-                entityareaeffectcloud.setDuration(600);
+                entityareaeffectcloud.setRadius(this.radius); // Canvas
+                entityareaeffectcloud.setDuration(this.duration); // Canvas
                 entityareaeffectcloud.setRadiusPerTick((7.0F - entityareaeffectcloud.getRadius()) / (float) entityareaeffectcloud.getDuration());
-                entityareaeffectcloud.addEffect(new MobEffectInstance(MobEffects.HARM, 1, 1));
+                entityareaeffectcloud.addEffect(new MobEffectInstance(MobEffects.HARM, 1, this.cloudAmplifier)); // Canvas
                 if (!list.isEmpty()) {
                     Iterator iterator = list.iterator();
 
@@ -89,6 +107,6 @@ public class DragonFireball extends AbstractHurtingProjectile {
 
     @Override
     public boolean shouldBurn() { // Canvas
-        return false;
+        return super.shouldBurn();
     }
 }
diff --git a/src/main/java/net/minecraft/world/entity/projectile/EvokerFangs.java b/src/main/java/net/minecraft/world/entity/projectile/EvokerFangs.java
index e6f549f1fcd261f96f0e4fc4cbe26a04c389d191..83e74fc038469104b9890a05fa2e45d46afe2b6a 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/EvokerFangs.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/EvokerFangs.java
@@ -126,13 +126,20 @@ public class EvokerFangs extends Entity implements TraceableEntity {
         }
 
     }
+    // Canvas start
+
+    public float fangDamage = 6.0f;
+    public void setFangDamage(float damage){
+        this.fangDamage = damage;
+    }
+    // Canvas end
 
     private void dealDamageTo(LivingEntity target) {
         LivingEntity entityliving1 = this.getOwner();
 
         if (target.isAlive() && !target.isInvulnerable() && target != entityliving1) {
             if (entityliving1 == null) {
-                target.hurt(this.damageSources().magic().customEventDamager(this), 6.0F); // CraftBukkit // Paper - fix DamageSource API
+                target.hurt(this.damageSources().magic().customEventDamager(this), this.fangDamage); // CraftBukkit // Paper - fix DamageSource API // Canvas
             } else {
                 if (entityliving1.isAlliedTo((Entity) target)) {
                     return;
diff --git a/src/main/java/net/minecraft/world/entity/projectile/EyeOfEnder.java b/src/main/java/net/minecraft/world/entity/projectile/EyeOfEnder.java
index d06a02f1868b911721477c149f1b321979d8d178..df2bdeaa820ad59e922d2cd08427ce5b917ee37d 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/EyeOfEnder.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/EyeOfEnder.java
@@ -1,5 +1,6 @@
 package net.minecraft.world.entity.projectile;
 
+import me.dueris.canvas.event.item.EyeOfEnderTryBreakEvent;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.particles.ParticleTypes;
 import net.minecraft.nbt.CompoundTag;
@@ -161,11 +162,24 @@ public class EyeOfEnder extends Entity implements ItemSupplier {
             if (this.life > 80 && !this.level().isClientSide) {
                 this.playSound(SoundEvents.ENDER_EYE_DEATH, 1.0F, 1.0F);
                 this.discard(this.surviveAfterDeath ? EntityRemoveEvent.Cause.DROP : EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
+                // Canvas start
+                EyeOfEnderTryBreakEvent event = new EyeOfEnderTryBreakEvent(this.surviveAfterDeath, new org.bukkit.Location(this.level().getWorld(), this.getX(), this.getY(), this.getZ()));
+                event.callEvent();
+                if(!event.isCanceled()){
+                    if(event.surviveAfterDeath()){
+                        this.level().addFreshEntity(new ItemEntity(((org.bukkit.craftbukkit.CraftWorld) event.getLocation().getWorld()).getHandle(), event.getLocation().getX(), event.getLocation().getY(), event.getLocation().getZ(), this.getItem()));
+                    }else{
+                        this.level().levelEvent(2003, this.blockPosition(), 0);
+                    }
+                }
+                /*
                 if (this.surviveAfterDeath) {
                     this.level().addFreshEntity(new ItemEntity(this.level(), this.getX(), this.getY(), this.getZ(), this.getItem()));
                 } else {
                     this.level().levelEvent(2003, this.blockPosition(), 0);
                 }
+                 */
+                // Canvas end
             }
         } else {
             this.setPosRaw(d0, d1, d2);
diff --git a/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java b/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
index feccff3d240a5e1296c2c8a1f18dd24f012eb14c..6e1b32c1e5571c4be1c07e74b2d0d8d18a1015dd 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/FireworkRocketEntity.java
@@ -187,7 +187,7 @@ public class FireworkRocketEntity extends Projectile implements ItemSupplier {
 
     }
 
-    private void explode() {
+    public void explode() { // Canvas - private -> public
         this.level().broadcastEntityEvent(this, (byte) 17);
         this.gameEvent(GameEvent.EXPLODE, this.getOwner());
         this.dealExplosionDamage();
@@ -232,7 +232,7 @@ public class FireworkRocketEntity extends Projectile implements ItemSupplier {
 
     private void dealExplosionDamage() {
         float f = 0.0F;
-        ItemStack itemstack = (ItemStack) this.entityData.get(FireworkRocketEntity.DATA_ID_FIREWORKS_ITEM);
+        ItemStack itemstack = this.entityData.get(FireworkRocketEntity.DATA_ID_FIREWORKS_ITEM); // Canvas - remove redundant cast
         CompoundTag nbttagcompound = itemstack.isEmpty() ? null : itemstack.getTagElement("Fireworks");
         ListTag nbttaglist = nbttagcompound != null ? nbttagcompound.getList("Explosions", 10) : null;
 
@@ -282,7 +282,7 @@ public class FireworkRocketEntity extends Projectile implements ItemSupplier {
     }
 
     public boolean isShotAtAngle() {
-        return (Boolean) this.entityData.get(FireworkRocketEntity.DATA_SHOT_AT_ANGLE);
+        return this.entityData.get(FireworkRocketEntity.DATA_SHOT_AT_ANGLE); // Canvas - remove redundant cast
     }
 
     @Override
@@ -346,7 +346,7 @@ public class FireworkRocketEntity extends Projectile implements ItemSupplier {
 
     @Override
     public ItemStack getItem() {
-        ItemStack itemstack = (ItemStack) this.entityData.get(FireworkRocketEntity.DATA_ID_FIREWORKS_ITEM);
+        ItemStack itemstack = this.entityData.get(FireworkRocketEntity.DATA_ID_FIREWORKS_ITEM); // Canvas - remove redundant import
 
         return itemstack.isEmpty() ? new ItemStack(Items.FIREWORK_ROCKET) : itemstack;
     }
diff --git a/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java b/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
index 91497f5e6c07fcf1b05eca6846c51e1a15ed3bc0..89e62970fa154f01a598a3893b0182c1c285ce3e 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/FishingHook.java
@@ -56,7 +56,7 @@ public class FishingHook extends Projectile {
 
     private static final Logger LOGGER = LogUtils.getLogger();
     private final RandomSource syncronizedRandom;
-    private boolean biting;
+    public boolean biting; // Canvas
     public int outOfWaterTime;
     private static final int MAX_OUT_OF_WATER_TIME = 10;
     public static final EntityDataAccessor<Integer> DATA_HOOKED_ENTITY = SynchedEntityData.defineId(FishingHook.class, EntityDataSerializers.INT);
diff --git a/src/main/java/net/minecraft/world/entity/projectile/ShulkerBullet.java b/src/main/java/net/minecraft/world/entity/projectile/ShulkerBullet.java
index 1c50c870e59c35a39c83a0f136ee5d3c70268763..31ab56a41a046591182fbbddb682980f7e4f522a 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/ShulkerBullet.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/ShulkerBullet.java
@@ -323,12 +323,12 @@ public class ShulkerBullet extends Projectile {
         this.playSound(SoundEvents.SHULKER_BULLET_HIT, 1.0F, 1.0F);
     }
 
-    private void destroy() {
+    public void destroy() { // Canvas - private -> public
         // CraftBukkit start - add Bukkit remove cause
         this.destroy(null);
     }
 
-    private void destroy(EntityRemoveEvent.Cause cause) {
+    public void destroy(EntityRemoveEvent.Cause cause) { // Canvas - private -> public
         this.discard(cause);
         // CraftBukkit end
         this.level().gameEvent(GameEvent.ENTITY_DAMAGE, this.position(), GameEvent.Context.of((Entity) this));
diff --git a/src/main/java/net/minecraft/world/entity/projectile/ThrownTrident.java b/src/main/java/net/minecraft/world/entity/projectile/ThrownTrident.java
index 2bd77524313ae7b32f710e7d197e81a2ddd12965..b4e212a71c694e9979d334ecb307826b04e66ed3 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/ThrownTrident.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/ThrownTrident.java
@@ -182,9 +182,16 @@ public class ThrownTrident extends AbstractArrow {
         this.playSound(soundeffect, f1, 1.0F);
     }
 
+    // Canvas start
+    public boolean forceChanneling = false;
+    public void setForceChanneling(boolean forceChanneling){
+        this.forceChanneling = forceChanneling;
+    }
+
     public boolean isChanneling() {
-        return EnchantmentHelper.hasChanneling(this.getPickupItemStackOrigin());
+        return EnchantmentHelper.hasChanneling(this.getPickupItemStackOrigin()) || this.forceChanneling; // Canvas;
     }
+    // Canvas end
 
     @Override
     protected boolean tryPickup(Player player) {
@@ -226,10 +233,17 @@ public class ThrownTrident extends AbstractArrow {
         }
 
     }
+    // Canvas start
+
+    public float waterInertia = 0.99f;
+    public void setWaterInertia(float waterInertia){
+        this.waterInertia = waterInertia;
+    }
+    // Canvas end
 
     @Override
-    protected float getWaterInertia() {
-        return 0.99F;
+    public float getWaterInertia() { // Canvas - protected -> public
+        return this.waterInertia;
     }
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftDragonFireball.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftDragonFireball.java
index b884c6b0808f80222b1dbf13b855891fd80320af..07c349ef0ffec166c29c29a2a7901b427a7331ec 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftDragonFireball.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftDragonFireball.java
@@ -7,6 +7,42 @@ public class CraftDragonFireball extends CraftFireball implements DragonFireball
     public CraftDragonFireball(CraftServer server, net.minecraft.world.entity.projectile.DragonFireball entity) {
         super(server, entity);
     }
+    // Canvas start
+
+    public net.minecraft.world.entity.projectile.DragonFireball getHandle(){
+        return (net.minecraft.world.entity.projectile.DragonFireball) entity;
+    }
+
+    @Override
+    public void setEffectRadius(float radius){
+        this.getHandle().setRadius(radius);
+    }
+
+    @Override
+    public float getEffectRadius(){
+        return this.getHandle().radius;
+    }
+
+    @Override
+    public int getCloudAmplifier(){
+        return this.getHandle().cloudAmplifier;
+    }
+
+    @Override
+    public void setCloudAmplifier(int cloudAmplifier){
+        this.getHandle().setCloudAmplifier(cloudAmplifier);
+    }
+
+    @Override
+    public int getCloudDuration(){
+        return this.getHandle().duration;
+    }
+
+    @Override
+    public void setCloudDuration(int duration){
+        this.getHandle().setDuration(duration);
+    }
+    // Canvas end
 
     @Override
     public String toString() {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEvokerFangs.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEvokerFangs.java
index 19b368cc862cd7e3e1f0e89401a7d099e3eaefa3..823bbb70a7081a95104f60eacbe1a2f4727bf573 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEvokerFangs.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEvokerFangs.java
@@ -15,6 +15,18 @@ public class CraftEvokerFangs extends CraftEntity implements EvokerFangs {
     public net.minecraft.world.entity.projectile.EvokerFangs getHandle() {
         return (net.minecraft.world.entity.projectile.EvokerFangs) super.getHandle();
     }
+    // Canvas start
+
+    @Override
+    public float getFangDamage(){
+        return this.getHandle().fangDamage;
+    }
+
+    @Override
+    public void setFangDamage(float fangDamage){
+        this.getHandle().setFangDamage(fangDamage);
+    }
+    // Canvas end
 
     @Override
     public String toString() {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftFirework.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftFirework.java
index 1309fabd357eb9b8f90f882ec7ff3de9e60f187d..b28048dbed029a0867eb7534f48c07d93872f07d 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftFirework.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftFirework.java
@@ -41,6 +41,12 @@ public class CraftFirework extends CraftProjectile implements Firework {
     public FireworkRocketEntity getHandle() {
         return (FireworkRocketEntity) this.entity;
     }
+    // Canvas start
+    @Override
+    public void explode(){
+        this.getHandle().explode();
+    }
+    // Canvas end
 
     @Override
     public String toString() {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftFishHook.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftFishHook.java
index 116f23a3ee9fa409d7bc34c3769b94ed2bb07183..a5a752891aed817ea01d8e4a4ed802e0715d2a73 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftFishHook.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftFishHook.java
@@ -161,6 +161,13 @@ public class CraftFishHook extends CraftProjectile implements FishHook {
         Preconditions.checkArgument(chance >= 0 && chance <= 1, "The bite chance must be between 0 and 1");
         this.biteChance = chance;
     }
+    // Canvas start
+
+    @Override
+    public boolean isBiting(){
+        return this.getHandle().biting;
+    }
+    // Canvas end
 
     @Override
     public boolean isInOpenWater() {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftTrident.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftTrident.java
index 8a6af0db8e0aa0cffbf19584be747076c2c8ee44..7648f27632e0ea940f27b094374db7990ace300d 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftTrident.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftTrident.java
@@ -16,6 +16,28 @@ public class CraftTrident extends CraftArrow implements Trident {
     public ThrownTrident getHandle() {
         return (ThrownTrident) super.getHandle();
     }
+    // Canvas start
+
+    @Override
+    public boolean getChanneling(){
+        return this.getHandle().isChanneling();
+    }
+
+    @Override
+    public void forceChanneling(boolean force){
+        this.getHandle().setForceChanneling(force);
+    }
+
+    @Override
+    public float getWaterInertia(){
+        return this.getHandle().waterInertia;
+    }
+
+    @Override
+    public void setWaterInertia(float inertia){
+        this.getHandle().setWaterInertia(inertia);
+    }
+    // Canvas end
 
     @Override
     public ItemStack getItem() {
