From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Sun, 29 Oct 2023 17:54:28 -0700
Subject: [PATCH] Change some gradle configs for Sonatype - fix errors for
 missing depend


diff --git a/build.gradle.kts b/build.gradle.kts
index 77cd0fa49e7a300e9e5413ce585f3befb8fed0ed..b436ec7544f955e4e5d399f70e4294722af210b3 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -46,6 +46,7 @@ dependencies {
     runtimeOnly("org.xerial:sqlite-jdbc:3.42.0.1")
     runtimeOnly("com.mysql:mysql-connector-j:8.1.0")
     runtimeOnly("com.lmax:disruptor:3.4.4") // Paper
+    implementation("org.spongepowered:mixin:0.8.5") // Canvas - include mixin
     // Paper start - Use Velocity cipher
     implementation("com.velocitypowered:velocity-native:3.1.2-SNAPSHOT") {
         isTransitive = false
@@ -253,6 +254,9 @@ publishing {
                         url.set("https://www.gnu.org/licenses/gpl-3.0.en.html")
                     }
                 }
+                repositories{
+                    mavenCentral()
+                }
             }
             groupId = "io.github.dueris"
             artifactId = "canvas-server"
@@ -274,4 +278,4 @@ publishing {
         }
     }
 }
-// Canvas end
\ No newline at end of file
+// Canvas end
