From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Fri, 8 Dec 2023 05:17:14 +0000
Subject: [PATCH] Boat Wobble API


diff --git a/src/main/java/net/minecraft/world/entity/vehicle/Boat.java b/src/main/java/net/minecraft/world/entity/vehicle/Boat.java
index 03eb34b5e963072e4d8dff0893928ff166b8a9ce..ed99ef55b36feef2d08e6005de967277b7d6d407 100644
--- a/src/main/java/net/minecraft/world/entity/vehicle/Boat.java
+++ b/src/main/java/net/minecraft/world/entity/vehicle/Boat.java
@@ -98,7 +98,7 @@ public class Boat extends VehicleEntity implements VariantHolder<Boat.Type> {
     private Boat.Status oldStatus;
     private double lastYd;
     private boolean isAboveBubbleColumn;
-    private boolean bubbleColumnDirectionIsDown;
+    public boolean bubbleColumnDirectionIsDown; // Canvas - private -> public
     private float bubbleMultiplier;
     private float bubbleAngle;
     private float bubbleAngleO;
@@ -478,14 +478,25 @@ public class Boat extends VehicleEntity implements VariantHolder<Boat.Type> {
                     this.setBubbleTime(0);
                     Vec3 vec3d = this.getDeltaMovement();
 
+                    // Canvas start
+                    org.bukkit.util.Vector velocity = org.bukkit.craftbukkit.util.CraftVector.toBukkit(vec3d);
                     if (this.bubbleColumnDirectionIsDown) {
-                        this.setDeltaMovement(vec3d.add(0.0D, -0.7D, 0.0D));
-                        this.ejectPassengers();
+                        velocity.setY(velocity.getY() - 0.7D);
                     } else {
-                        this.setDeltaMovement(vec3d.x, this.hasPassenger((entity) -> {
+                        velocity.setY(this.hasPassenger((entity) -> {
                             return entity instanceof Player;
-                        }) ? 2.7D : 0.6D, vec3d.z);
+                        }) ? 2.7D : 0.6D);
                     }
+
+                    me.dueris.canvas.event.entity.BoatBobEvent event = new me.dueris.canvas.event.entity.BoatBobEvent((org.bukkit.entity.Boat) this.getBukkitEntity(), velocity);
+                    event.callEvent();
+                    if(event.isCanceled()) {
+                        if (event.getVelocity().getY() < 0.0D) {
+                            this.ejectPassengers();
+                        }
+                        this.getBukkitEntity().setVelocity(event.getVelocity());
+                    }
+                    // Canvas end
                 }
 
                 this.isAboveBubbleColumn = false;
@@ -924,11 +935,11 @@ public class Boat extends VehicleEntity implements VariantHolder<Boat.Type> {
         return (Boolean) this.entityData.get(paddle == 0 ? Boat.DATA_ID_PADDLE_LEFT : Boat.DATA_ID_PADDLE_RIGHT) && this.getControllingPassenger() != null;
     }
 
-    private void setBubbleTime(int wobbleTicks) {
+    public void setBubbleTime(int wobbleTicks) { // Canvas - private -> public
         this.entityData.set(Boat.DATA_ID_BUBBLE_TIME, wobbleTicks);
     }
 
-    private int getBubbleTime() {
+    public int getBubbleTime() { // Canvas - private -> public
         return (Integer) this.entityData.get(Boat.DATA_ID_BUBBLE_TIME);
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftBoat.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftBoat.java
index 5464e5064eb1935b01fd75b3b4d3413e34f110f1..9eb7d1329fe192f356b8115ac03cd326f3c11e0b 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftBoat.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftBoat.java
@@ -43,6 +43,29 @@ public class CraftBoat extends CraftVehicle implements Boat {
 
         this.getHandle().setVariant(CraftBoat.boatTypeToNms(type));
     }
+    // Canvas start
+
+    @Override
+    public int getWobbleTicks(){
+        return this.getHandle().getBubbleTime();
+    }
+
+    @Override
+    public void setWobbleTicks(int ticks){
+        Preconditions.checkArgument(0 < ticks, "Ticks cannot be less than 0");
+
+        this.getHandle().setBubbleTime(ticks);
+    }
+
+    @Override
+    public net.kyori.adventure.util.TriState willSinkAfterWobble() {
+        if (this.getWobbleTicks() == 0) {
+            return net.kyori.adventure.util.TriState.NOT_SET;
+        }
+
+        return net.kyori.adventure.util.TriState.byBoolean(this.getHandle().bubbleColumnDirectionIsDown);
+    }
+    // Canvas end
 
     @Override
     public double getMaxSpeed() {
