From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Fri, 10 Nov 2023 04:34:20 +0000
Subject: [PATCH] Fix Mirai patch errors


diff --git a/src/main/java/dev/etil/mirai/MiraiConfig.java b/src/main/java/dev/etil/mirai/MiraiConfig.java
index cf3fed88162efcc8cf4aa1d095066a2abce59887..43a453560658fa30bc58875e7b5db49126510c37 100644
--- a/src/main/java/dev/etil/mirai/MiraiConfig.java
+++ b/src/main/java/dev/etil/mirai/MiraiConfig.java
@@ -132,6 +132,12 @@ public class MiraiConfig {
         return config.getStringList(key);
     }
 
+    public static boolean fastSpeedCheck;
+    private static void speedCheck() {
+        fastSpeedCheck = getBoolean("fast-speed-check", true, 
+        "Whether or not server should listen to location triggers.");
+    }
+
     public static int maxBees;
     private static void maximumBees() {
         maxBees = getInt("max-bees-in-hive", 3,
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index cb4ff5d37013d764a2d5c13c878e323e5a9f360d..4a31243718a1a548a27303e96a81223c9ad451aa 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -1298,7 +1298,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
                 this.tryCheckInsideBlocks();
                 // NFT-Worlds start
                 float f;
-                if (dev.etil.mirai.MiraiConfig.fastSpeedCheck) {
+                if (dev.etil.mirai.MiraiConfig.fastSpeedCheck) { 
                     if (this.getDeltaMovement().x == 0 && this.getDeltaMovement().z == 0) {
                         f = 1;
                     } else {
