From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Fri, 24 May 2024 18:50:11 -0700
Subject: [PATCH] Optimize Entity Pathfinding


diff --git a/src/main/java/me/dueris/canvas/CanvasConfig.java b/src/main/java/me/dueris/canvas/CanvasConfig.java
index 2fe52fd3e619f74a77bae6d5f5f74d2cbca1e9f6..fff699215416a265a54aef79ca6ff401e2a9a3e1 100644
--- a/src/main/java/me/dueris/canvas/CanvasConfig.java
+++ b/src/main/java/me/dueris/canvas/CanvasConfig.java
@@ -210,11 +210,17 @@ public class CanvasConfig {
     public static boolean playerInstaAbsorbOrbs = false;
     public static boolean useCanvasOrbOptimizations = false;
     public static double orbMergeSearchInfaltionRadius = 0.8;
+    public static int maxEntitiesToReducePathfindingRates = 2;
 
     private static void general(){
         disableFootsteps = getBoolean("disableFootsteps", disableFootsteps);
         checkLightForSnow = getBoolean("checkLightForSnow", checkLightForSnow);
         onlyPlayersPushEntities = getBoolean("onlyPlayersPushEntities", onlyPlayersPushEntities);
+
+        playerInstaAbsorbOrbs = getBoolean("playerInstaAbsorbOrbs", playerInstaAbsorbOrbs);
+        useCanvasOrbOptimizations = getBoolean("useCanvasOrbOptimizations", useCanvasOrbOptimizations);
+        orbMergeSearchInfaltionRadius = getDouble("orbMergeSearchInfaltionRadius", orbMergeSearchInfaltionRadius);
+        maxEntitiesToReducePathfindingRates = getInt("maxEntitiesToReducePathfindingRates", maxEntitiesToReducePathfindingRates);
     }
 
     private static void optimizations(){
diff --git a/src/main/java/net/minecraft/world/entity/PathfinderMob.java b/src/main/java/net/minecraft/world/entity/PathfinderMob.java
index 645414720db0231a19c88f681608eb08200b8782..760ca63f199bed9d05d8ff1f7afd00a34fff77a0 100644
--- a/src/main/java/net/minecraft/world/entity/PathfinderMob.java
+++ b/src/main/java/net/minecraft/world/entity/PathfinderMob.java
@@ -118,4 +118,11 @@ public abstract class PathfinderMob extends Mob {
     }
 
     protected void onLeashDistance(float leashLength) {}
+    // Canvas start
+
+    public boolean canTickRandomStroll() {
+        java.util.List<Entity> entityList = this.level().getEntities(this, this.getBoundingBox(), entity -> !entity.isRemoved() && entity != this);
+		return entityList.isEmpty() || entityList.toArray().length < me.dueris.canvas.CanvasConfig.maxEntitiesToReducePathfindingRates;
+	}
+    // Canvas end
 }
diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/MoveToBlockGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/MoveToBlockGoal.java
index aee0147649d458b87d92496eda0c1723ebe570d2..aa46a1783323ac2c979c80e151be6ce410576355 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/MoveToBlockGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/MoveToBlockGoal.java
@@ -1,6 +1,7 @@
 package net.minecraft.world.entity.ai.goal;
 
 import java.util.EnumSet;
+import java.util.Random;
 import net.minecraft.core.BlockPos;
 import net.minecraft.world.entity.PathfinderMob;
 import net.minecraft.world.level.LevelReader;
@@ -43,6 +44,13 @@ public abstract class MoveToBlockGoal extends Goal {
 
     @Override
     public boolean canUse() {
+        // Canvas start
+        boolean chance = new Random().nextDouble(100) >= 10;
+        boolean canWander = this.mob.canTickRandomStroll();
+        if(!canWander && chance){
+            return false;
+        }
+        // Canvas end
         if (this.nextStartTick > 0) {
             this.nextStartTick--;
             return false;
diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/RandomStrollGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/RandomStrollGoal.java
index 128d7423b4bca1e121e26f5fc968a4f82462e227..9cf1217feace97aa28d6b297cb787fb6cd7ca720 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/RandomStrollGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/RandomStrollGoal.java
@@ -1,6 +1,7 @@
 package net.minecraft.world.entity.ai.goal;
 
 import java.util.EnumSet;
+import java.util.Random;
 import javax.annotation.Nullable;
 import net.minecraft.world.entity.PathfinderMob;
 import net.minecraft.world.entity.ai.util.DefaultRandomPos;
@@ -77,7 +78,9 @@ public class RandomStrollGoal extends Goal {
     @Override
     public boolean canUse() {
         // Canvas start - optimize entity goals
-        if(!this.canWander(this.mob)){
+        boolean chance = new Random().nextDouble(100) >= 10;
+        boolean canWander = this.mob.canTickRandomStroll();
+        if(!this.canWander(this.mob) || (!canWander && chance)){
             return false;
         }
         // Canvas end
