From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Sun, 2 Jun 2024 20:55:21 -0700
Subject: [PATCH] Allow to disable feature generation


diff --git a/src/main/java/me/dueris/canvas/CanvasConfig.java b/src/main/java/me/dueris/canvas/CanvasConfig.java
index 24f00e8ec31a7cf4ef554c778cebbe340970135a..9825b47fab89c05958ebfb93d830634fdd723ff2 100644
--- a/src/main/java/me/dueris/canvas/CanvasConfig.java
+++ b/src/main/java/me/dueris/canvas/CanvasConfig.java
@@ -211,6 +211,12 @@ public class CanvasConfig {
     public static double orbMergeSearchInfaltionRadius = 0.8;
     public static int maxEntitiesToReducePathfindingRates = 2;
 
+    public static boolean disableFeatures = false;
+
+    private static void chunkGen() {
+        disableFeatures = getBoolean("chunkgen.disablefeatures", disableFeatures);
+    }
+
     private static void general(){
         disableFootsteps = getBoolean("disableFootsteps", disableFootsteps);
         checkLightForSnow = getBoolean("checkLightForSnow", checkLightForSnow);
diff --git a/src/main/java/net/minecraft/world/level/levelgen/feature/ConfiguredFeature.java b/src/main/java/net/minecraft/world/level/levelgen/feature/ConfiguredFeature.java
index 25591124b6fadd3b976dfab79b06ae30278f8e6b..dd0b4e7100e7012afdc7c36472a1a15872fcb5d9 100644
--- a/src/main/java/net/minecraft/world/level/levelgen/feature/ConfiguredFeature.java
+++ b/src/main/java/net/minecraft/world/level/levelgen/feature/ConfiguredFeature.java
@@ -22,6 +22,7 @@ public record ConfiguredFeature<FC extends FeatureConfiguration, F extends Featu
     public static final Codec<HolderSet<ConfiguredFeature<?, ?>>> LIST_CODEC = RegistryCodecs.homogeneousList(Registries.CONFIGURED_FEATURE, DIRECT_CODEC);
 
     public boolean place(WorldGenLevel world, ChunkGenerator chunkGenerator, RandomSource random, BlockPos origin) {
+        if (me.dueris.canvas.CanvasConfig.disableFeatures) return false; // Canvas
         return this.feature.place(this.config, world, chunkGenerator, random, origin);
     }
 
