From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Tue, 30 Apr 2024 20:18:46 -0700
Subject: [PATCH] More Constants Improvements


diff --git a/src/main/java/me/dueris/canvas/Constants.java b/src/main/java/me/dueris/canvas/Constants.java
new file mode 100644
index 0000000000000000000000000000000000000000..1baff312ae42638d352ea5f4ab3e819e2df3158a
--- /dev/null
+++ b/src/main/java/me/dueris/canvas/Constants.java
@@ -0,0 +1,25 @@
+package me.dueris.canvas;
+
+import net.minecraft.world.entity.ai.behavior.DoNothing;
+import net.minecraft.world.item.ItemCooldowns;
+import net.minecraft.world.phys.Vec3;
+import java.io.File;
+
+public interface Constants {
+    Object[] EMPTY_object_arr = new Object[0];
+    int[] EMPTY_int_arr = new int[0];
+    int[] ZERO_int_arr = new int[]{0};
+    byte[] EMPTY_byte_arr = new byte[0];
+    String[] EMPTY_string_arr = new String[0];
+    long[] EMPTY_long_arr = new long[0];
+    org.bukkit.entity.Entity[] EMPTY_bukkit_entity_arr = new org.bukkit.entity.Entity[0];
+    net.minecraft.world.entity.Entity[] EMPTY_entity_arr = new net.minecraft.world.entity.Entity[0];
+    File SERVER_DIRECTORY = new File(".");
+    int[] ZERO = new int[0];
+    Vec3 EMPTY_VECTOR3 = new Vec3(0.0D, 0.0D, 1.0D);
+    ItemCooldowns ITEM_COOLDOWNS = new ItemCooldowns();
+    float[] WHITE_SHEEP_COLOR = new float[]{0.9019608F, 0.9019608F, 0.9019608F};
+    DoNothing DO_NOTHING = new DoNothing(30, 60);
+    DoNothing VILLAGERS_DO_NOTHING = new DoNothing(20, 40);
+    DoNothing PIGLIN_DO_NOTHING = new DoNothing(10, 20);
+}
diff --git a/src/main/java/me/dueris/canvas/SoundConstants.java b/src/main/java/me/dueris/canvas/SoundConstants.java
new file mode 100644
index 0000000000000000000000000000000000000000..581e5fd66a9a79f3a21b2c46edb5faadd43a1bec
--- /dev/null
+++ b/src/main/java/me/dueris/canvas/SoundConstants.java
@@ -0,0 +1,11 @@
+package me.dueris.canvas;
+
+import net.minecraft.sounds.SoundEvents;
+import net.minecraft.world.entity.LivingEntity;
+
+public interface SoundConstants {
+    LivingEntity.Fallsounds LIVING_ENTITY = new LivingEntity.Fallsounds(SoundEvents.GENERIC_SMALL_FALL, SoundEvents.GENERIC_BIG_FALL);
+    LivingEntity.Fallsounds PLAYER = new LivingEntity.Fallsounds(SoundEvents.PLAYER_SMALL_FALL, SoundEvents.PLAYER_BIG_FALL);
+    LivingEntity.Fallsounds MONSTER = new LivingEntity.Fallsounds(SoundEvents.HOSTILE_SMALL_FALL, SoundEvents.HOSTILE_BIG_FALL);
+    LivingEntity.Fallsounds ARMOR_STAND = new LivingEntity.Fallsounds(SoundEvents.ARMOR_STAND_FALL, SoundEvents.ARMOR_STAND_FALL);
+}
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 8b7d43ab868b0f26b88c4f33e8eca6c208423731..f28d18307fe059e465875566a132c80aeeef9864 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1631,7 +1631,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     }
 
     public File getServerDirectory() {
-        return new File(".");
+        return me.dueris.canvas.Constants.SERVER_DIRECTORY;
     }
 
     public void onServerCrash(CrashReport report) {}
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index be3e496910d574cb7f2a5685f0530983c14a14af..f2881800379c7ce2cea4124d3fc534a2d8293fc7 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -2071,7 +2071,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
     }
 
     public LivingEntity.Fallsounds getFallSounds() {
-        return new LivingEntity.Fallsounds(SoundEvents.GENERIC_SMALL_FALL, SoundEvents.GENERIC_BIG_FALL);
+        return me.dueris.canvas.SoundConstants.LIVING_ENTITY; // Canvas
     }
 
     protected SoundEvent getDrinkingSound(ItemStack stack) {
diff --git a/src/main/java/net/minecraft/world/entity/ai/behavior/VillagerGoalPackages.java b/src/main/java/net/minecraft/world/entity/ai/behavior/VillagerGoalPackages.java
index 3091d985ba9c55d404332576320718840538722e..abca41a75924080513acf6216feb3bb9b4f788de 100644
--- a/src/main/java/net/minecraft/world/entity/ai/behavior/VillagerGoalPackages.java
+++ b/src/main/java/net/minecraft/world/entity/ai/behavior/VillagerGoalPackages.java
@@ -102,7 +102,7 @@ public class VillagerGoalPackages {
                         Pair.of(VillageBoundRandomStroll.create(speed), 1),
                         Pair.of(SetWalkTargetFromLookTarget.create(speed, 2), 1),
                         Pair.of(new JumpOnBed(speed), 2),
-                        Pair.of(new DoNothing(20, 40), 2)
+                        Pair.of(me.dueris.canvas.Constants.VILLAGERS_DO_NOTHING, 2) // Canvas
                     )
                 )
             ),
@@ -123,7 +123,7 @@ public class VillagerGoalPackages {
                         Pair.of(SetClosestHomeAsWalkTarget.create(speed), 1),
                         Pair.of(InsideBrownianWalk.create(speed), 4),
                         Pair.of(GoToClosestVillage.create(speed, 4), 2),
-                        Pair.of(new DoNothing(20, 40), 2)
+                        Pair.of(me.dueris.canvas.Constants.VILLAGERS_DO_NOTHING, 2) // Canvas
                     )
                 )
             ),
@@ -172,7 +172,7 @@ public class VillagerGoalPackages {
                         Pair.of(VillageBoundRandomStroll.create(speed), 1),
                         Pair.of(SetWalkTargetFromLookTarget.create(speed, 2), 1),
                         Pair.of(new JumpOnBed(speed), 1),
-                        Pair.of(new DoNothing(30, 60), 1)
+                        Pair.of(me.dueris.canvas.Constants.DO_NOTHING, 1)
                     )
                 )
             ),
@@ -272,7 +272,7 @@ public class VillagerGoalPackages {
                     Pair.of(SetEntityLookTarget.create(MobCategory.UNDERGROUND_WATER_CREATURE, 8.0F), 1),
                     Pair.of(SetEntityLookTarget.create(MobCategory.WATER_AMBIENT, 8.0F), 1),
                     Pair.of(SetEntityLookTarget.create(MobCategory.MONSTER, 8.0F), 1),
-                    Pair.of(new DoNothing(30, 60), 2)
+                    Pair.of(me.dueris.canvas.Constants.DO_NOTHING, 2) // Canvas
                 )
             )
         );
@@ -285,7 +285,7 @@ public class VillagerGoalPackages {
                 ImmutableList.of(
                     Pair.of(SetEntityLookTarget.create(EntityType.VILLAGER, 8.0F), 2),
                     Pair.of(SetEntityLookTarget.create(EntityType.PLAYER, 8.0F), 2),
-                    Pair.of(new DoNothing(30, 60), 8)
+                    Pair.of(me.dueris.canvas.Constants.DO_NOTHING, 8) // Canvas
                 )
             )
         );
diff --git a/src/main/java/net/minecraft/world/entity/animal/Pig.java b/src/main/java/net/minecraft/world/entity/animal/Pig.java
index 748ab1a1f578a1ba4c46834607235d9d5856f185..ca164eee48bec1d313b17305eafcf037d7e9feb5 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Pig.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Pig.java
@@ -333,7 +333,7 @@ public class Pig extends Animal implements ItemSteerable, Saddleable {
 
     @Override
     protected Vec3 getRiddenInput(Player controllingPlayer, Vec3 movementInput) {
-        return new Vec3(0.0D, 0.0D, 1.0D);
+        return me.dueris.canvas.Constants.EMPTY_VECTOR3; // Canvas
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/animal/Sheep.java b/src/main/java/net/minecraft/world/entity/animal/Sheep.java
index 6b1244d3957e7f62c96ffd34692b8916337839fd..eac139bcc6a7880bcd95c8c125c77a1173d4a779 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Sheep.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Sheep.java
@@ -100,10 +100,10 @@ public class Sheep extends Animal implements Shearable {
 
     private static float[] createSheepColor(DyeColor color) {
         if (color == DyeColor.WHITE) {
-            return new float[]{0.9019608F, 0.9019608F, 0.9019608F};
+            return me.dueris.canvas.Constants.WHITE_SHEEP_COLOR; // Canvas
         } else {
             float[] afloat = color.getTextureDiffuseColors();
-            float f = 0.75F;
+            // float f = 0.75F; // Canvas
 
             return new float[]{afloat[0] * 0.75F, afloat[1] * 0.75F, afloat[2] * 0.75F};
         }
diff --git a/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java b/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java
index 92521cbedcf89a855f10a3401933acaf84bc3f98..69ee1358e88d10c700cd7c6c8d08e61213a09f76 100644
--- a/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java
+++ b/src/main/java/net/minecraft/world/entity/decoration/ArmorStand.java
@@ -920,7 +920,7 @@ public class ArmorStand extends LivingEntity {
 
     @Override
     public LivingEntity.Fallsounds getFallSounds() {
-        return new LivingEntity.Fallsounds(SoundEvents.ARMOR_STAND_FALL, SoundEvents.ARMOR_STAND_FALL);
+        return me.dueris.canvas.SoundConstants.ARMOR_STAND; // Canvas
     }
 
     @Nullable
diff --git a/src/main/java/net/minecraft/world/entity/monster/Monster.java b/src/main/java/net/minecraft/world/entity/monster/Monster.java
index e60e6b3e5ae5a468cfe649ed2222412f3bc8b268..c9b7115c4505f0eb61f820b931804365284e1756 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Monster.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Monster.java
@@ -79,7 +79,7 @@ public abstract class Monster extends PathfinderMob implements Enemy {
 
     @Override
     public LivingEntity.Fallsounds getFallSounds() {
-        return new LivingEntity.Fallsounds(SoundEvents.HOSTILE_SMALL_FALL, SoundEvents.HOSTILE_BIG_FALL);
+        return me.dueris.canvas.SoundConstants.MONSTER; // Canvas
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/monster/Strider.java b/src/main/java/net/minecraft/world/entity/monster/Strider.java
index edc2ca138088aef35754622153b63994910b8f98..3e5f8e3beb91d888217ecc400a494834b6494e64 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Strider.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Strider.java
@@ -321,7 +321,7 @@ public class Strider extends Animal implements ItemSteerable, Saddleable {
 
     @Override
     protected Vec3 getRiddenInput(Player controllingPlayer, Vec3 movementInput) {
-        return new Vec3(0.0D, 0.0D, 1.0D);
+        return me.dueris.canvas.Constants.EMPTY_VECTOR3; // Canvas
     }
 
     @Override
diff --git a/src/main/java/net/minecraft/world/entity/monster/Zoglin.java b/src/main/java/net/minecraft/world/entity/monster/Zoglin.java
index f9bb83f35a41e5b076ae438af613a97b8e60c0c3..e3fc01b9694946c9bdd363ef3353215b70262ca0 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Zoglin.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Zoglin.java
@@ -142,7 +142,7 @@ public class Zoglin extends Monster implements Enemy, HoglinBase {
                 SetEntityLookTargetSometimes.create(8.0F, UniformInt.of(30, 60)),
                 new RunOne<>(
                     ImmutableList.of(
-                        Pair.of(RandomStroll.stroll(0.4F), 2), Pair.of(SetWalkTargetFromLookTarget.create(0.4F, 3), 2), Pair.of(new DoNothing(30, 60), 1)
+                        Pair.of(RandomStroll.stroll(0.4F), 2), Pair.of(SetWalkTargetFromLookTarget.create(0.4F, 3), 2), Pair.of(me.dueris.canvas.Constants.DO_NOTHING, 1)
                     )
                 )
             )
diff --git a/src/main/java/net/minecraft/world/entity/monster/piglin/PiglinAi.java b/src/main/java/net/minecraft/world/entity/monster/piglin/PiglinAi.java
index b7d5c0b0e3741fcf04c4bac21a82fc41e2eeed5d..bb9cf1a5be306bfdb0b28911f3f51230fd8c64fe 100644
--- a/src/main/java/net/minecraft/world/entity/monster/piglin/PiglinAi.java
+++ b/src/main/java/net/minecraft/world/entity/monster/piglin/PiglinAi.java
@@ -159,7 +159,7 @@ public class PiglinAi {
     private static void initCelebrateActivity(Brain<Piglin> brain) {
         brain.addActivityAndRemoveMemoryWhenStopped(Activity.CELEBRATE, 10, ImmutableList.of(PiglinAi.avoidRepellent(), SetEntityLookTarget.create(PiglinAi::isPlayerHoldingLovedItem, 14.0F), StartAttacking.create(AbstractPiglin::isAdult, PiglinAi::findNearestValidAttackTarget), BehaviorBuilder.triggerIf((entitypiglin) -> {
             return !entitypiglin.isDancing();
-        }, GoToTargetLocation.create(MemoryModuleType.CELEBRATE_LOCATION, 2, 1.0F)), BehaviorBuilder.triggerIf(Piglin::isDancing, GoToTargetLocation.create(MemoryModuleType.CELEBRATE_LOCATION, 4, 0.6F)), new RunOne<>(ImmutableList.of(Pair.of(SetEntityLookTarget.create(EntityType.PIGLIN, 8.0F), 1), Pair.of(RandomStroll.stroll(0.6F, 2, 1), 1), Pair.of(new DoNothing(10, 20), 1)))), MemoryModuleType.CELEBRATE_LOCATION);
+        }, GoToTargetLocation.create(MemoryModuleType.CELEBRATE_LOCATION, 2, 1.0F)), BehaviorBuilder.triggerIf(Piglin::isDancing, GoToTargetLocation.create(MemoryModuleType.CELEBRATE_LOCATION, 4, 0.6F)), new RunOne<>(ImmutableList.of(Pair.of(SetEntityLookTarget.create(EntityType.PIGLIN, 8.0F), 1), Pair.of(RandomStroll.stroll(0.6F, 2, 1), 1), Pair.of(me.dueris.canvas.Constants.PIGLIN_DO_NOTHING, 1)))), MemoryModuleType.CELEBRATE_LOCATION); // Canvas
     }
 
     private static void initAdmireItemActivity(Brain<Piglin> brain) {
@@ -182,11 +182,11 @@ public class PiglinAi {
     }
 
     private static RunOne<LivingEntity> createIdleLookBehaviors() {
-        return new RunOne<>(ImmutableList.<Pair<? extends BehaviorControl<? super LivingEntity>, Integer>>builder().addAll(PiglinAi.createLookBehaviors()).add(Pair.of(new DoNothing(30, 60), 1)).build()); // CraftBukkit - decompile error
+        return new RunOne<>(ImmutableList.<Pair<? extends BehaviorControl<? super LivingEntity>, Integer>>builder().addAll(PiglinAi.createLookBehaviors()).add(Pair.of(me.dueris.canvas.Constants.DO_NOTHING, 1)).build()); // CraftBukkit - decompile error // Canvas
     }
 
     private static RunOne<Piglin> createIdleMovementBehaviors() {
-        return new RunOne<>(ImmutableList.of(Pair.of(RandomStroll.stroll(0.6F), 2), Pair.of(InteractWith.of(EntityType.PIGLIN, 8, MemoryModuleType.INTERACTION_TARGET, 0.6F, 2), 2), Pair.of(BehaviorBuilder.triggerIf(PiglinAi::doesntSeeAnyPlayerHoldingLovedItem, SetWalkTargetFromLookTarget.create(0.6F, 3)), 2), Pair.of(new DoNothing(30, 60), 1)));
+        return new RunOne<>(ImmutableList.of(Pair.of(RandomStroll.stroll(0.6F), 2), Pair.of(InteractWith.of(EntityType.PIGLIN, 8, MemoryModuleType.INTERACTION_TARGET, 0.6F, 2), 2), Pair.of(BehaviorBuilder.triggerIf(PiglinAi::doesntSeeAnyPlayerHoldingLovedItem, SetWalkTargetFromLookTarget.create(0.6F, 3)), 2), Pair.of(me.dueris.canvas.Constants.DO_NOTHING, 1))); // Canvas
     }
 
     private static BehaviorControl<PathfinderMob> avoidRepellent() {
diff --git a/src/main/java/net/minecraft/world/entity/monster/warden/WardenAi.java b/src/main/java/net/minecraft/world/entity/monster/warden/WardenAi.java
index d4e0ca72aaf7246cc641c8e3d38a0abe8dda18b9..29cba08e2e9b76bd8689419760396ca7f9b23af7 100644
--- a/src/main/java/net/minecraft/world/entity/monster/warden/WardenAi.java
+++ b/src/main/java/net/minecraft/world/entity/monster/warden/WardenAi.java
@@ -142,7 +142,7 @@ public class WardenAi {
                 TryToSniff.create(),
                 new RunOne<>(
                     ImmutableMap.of(MemoryModuleType.IS_SNIFFING, MemoryStatus.VALUE_ABSENT),
-                    ImmutableList.of(Pair.of(RandomStroll.stroll(0.5F), 2), Pair.of(new DoNothing(30, 60), 1))
+                    ImmutableList.of(Pair.of(RandomStroll.stroll(0.5F), 2), Pair.of(me.dueris.canvas.Constants.DO_NOTHING, 1)) // Canvas
                 )
             )
         );
diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 21e28cad221e343ec41e7ae193d7f45ae30ab4ab..4f123668284ae55e84fca65dbabe0165c8ffc844 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -402,7 +402,7 @@ public abstract class Player extends LivingEntity {
     }
 
     protected ItemCooldowns createItemCooldowns() {
-        return new ItemCooldowns();
+        return me.dueris.canvas.Constants.ITEM_COOLDOWNS; // Canvas
     }
 
     private void moveCloak() {
@@ -1840,7 +1840,7 @@ public abstract class Player extends LivingEntity {
 
     @Override
     public LivingEntity.Fallsounds getFallSounds() {
-        return new LivingEntity.Fallsounds(SoundEvents.PLAYER_SMALL_FALL, SoundEvents.PLAYER_BIG_FALL);
+        return me.dueris.canvas.SoundConstants.PLAYER;
     }
 
     @Override
