From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Tue, 24 Oct 2023 07:25:40 +0000
Subject: [PATCH] Added attribute listeners


diff --git a/src/main/java/io/papermc/paper/attribute/UnmodifiableAttributeMap.java b/src/main/java/io/papermc/paper/attribute/UnmodifiableAttributeMap.java
index 68044b8439c302114240d0ae4da93ab3e0789cd2..23aa17731d95dbcfe90c19ef169afb05dfaf7cc5 100644
--- a/src/main/java/io/papermc/paper/attribute/UnmodifiableAttributeMap.java
+++ b/src/main/java/io/papermc/paper/attribute/UnmodifiableAttributeMap.java
@@ -7,6 +7,9 @@ import org.bukkit.attribute.AttributeInstance;
 import org.bukkit.craftbukkit.attribute.CraftAttribute;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
+import org.bukkit.entity.*;
+import org.bukkit.attribute.*;
+import org.bukkit.plugin.*;
 
 public class UnmodifiableAttributeMap implements Attributable {
 
@@ -29,4 +32,10 @@ public class UnmodifiableAttributeMap implements Attributable {
     public void registerAttribute(@NotNull Attribute attribute) {
         throw new UnsupportedOperationException("Cannot register new attributes here");
     }
-}
+    // Canvas start
+    @Override
+    public void registerAttributeListener(boolean condition, int falseValue, int trueValue, Attribute attribute, Plugin plugin, int tickrate, Player player){
+        throw new UnsupportedOperationException("Cannot register new listener here");
+    }
+    // Canvas end
+}
\ No newline at end of file
diff --git a/src/main/java/org/bukkit/craftbukkit/attribute/CraftAttributeMap.java b/src/main/java/org/bukkit/craftbukkit/attribute/CraftAttributeMap.java
index a3f989fd1775adbb84d197c51e22c1889a2f8e18..36aed3ed152b9f46c5ee24848be46573f44b5e94 100644
--- a/src/main/java/org/bukkit/craftbukkit/attribute/CraftAttributeMap.java
+++ b/src/main/java/org/bukkit/craftbukkit/attribute/CraftAttributeMap.java
@@ -5,6 +5,11 @@ import net.minecraft.world.entity.ai.attributes.AttributeMap;
 import org.bukkit.attribute.Attributable;
 import org.bukkit.attribute.Attribute;
 import org.bukkit.attribute.AttributeInstance;
+// Canvas start
+import org.bukkit.scheduler.BukkitRunnable;
+import org.bukkit.plugin.*;
+import org.bukkit.entity.*;
+// Canvas end
 
 public class CraftAttributeMap implements Attributable {
 
@@ -41,5 +46,21 @@ public class CraftAttributeMap implements Attributable {
         Preconditions.checkArgument(attribute != null, "attribute");
         handle.registerAttribute(CraftAttribute.bukkitToMinecraft(attribute));
     }
-    // Paper end
-}
+    // Paper end -/- Canvas start
+    @Override
+    public void registerAttributeListener(boolean condition, int falseValue, int trueValue, Attribute attribute, Plugin plugin, int tickrate, Player player){
+        new BukkitRunnable(){
+            @Override
+            public void run(){
+                if(player.getAttribute(attribute) != null){
+                    if(condition){
+                        player.getAttribute(attribute).setBaseValue(trueValue);
+                    }else{
+                        player.getAttribute(attribute).setBaseValue(falseValue);
+                    }
+                }
+            }
+        }.runTaskTimer(plugin, 0, tickrate);
+    }
+    // Canvas end
+}
\ No newline at end of file
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index aa8efff39ca3b2c2ed69a19f1872a295b7103c6b..fac1725d919f5e25ccd89d0b1fe96c6802763c55 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -194,6 +194,10 @@ import org.bukkit.plugin.Plugin;
 import org.bukkit.util.BoundingBox;
 import org.bukkit.util.NumberConversions;
 import org.bukkit.util.Vector;
+import org.bukkit.scheduler.BukkitRunnable;
+import org.bukkit.entity.*;
+import org.bukkit.attribute.*;
+import org.bukkit.plugin.*;
 
 import net.md_5.bungee.api.chat.BaseComponent; // Spigot
 
@@ -228,6 +232,23 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         return getHandle().fireImmune();
     }
 
+    // Canvas start
+    public void registerAttributeListener(boolean condition, int falseValue, int trueValue, Attribute attribute, Plugin plugin, int tickrate, Player player){
+        new BukkitRunnable(){
+            @Override
+            public void run(){
+                if(player.getAttribute(attribute) != null){
+                    if(condition){
+                        player.getAttribute(attribute).setBaseValue(trueValue);
+                    }else{
+                        player.getAttribute(attribute).setBaseValue(falseValue);
+                    }
+                }
+            }
+        }.runTaskTimer(plugin, 0, tickrate);
+    }
+    // Canvas end
+
     @Override
     public void setImmuneToFire(Boolean fireImmune) {
         getHandle().immuneToFire = fireImmune;
