From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sun, 19 Nov 2023 16:39:21 -0800
Subject: [PATCH] Minor bossbar api addon


diff --git a/src/main/java/org/bukkit/craftbukkit/boss/CraftKeyedBossbar.java b/src/main/java/org/bukkit/craftbukkit/boss/CraftKeyedBossbar.java
index e380793e3436e5d311ae600fb7fa35f9a22794b9..a569a23122fb98f3ae4a52e44de03581af94488b 100644
--- a/src/main/java/org/bukkit/craftbukkit/boss/CraftKeyedBossbar.java
+++ b/src/main/java/org/bukkit/craftbukkit/boss/CraftKeyedBossbar.java
@@ -4,6 +4,7 @@ import net.minecraft.server.bossevents.CustomBossEvent;
 import org.bukkit.NamespacedKey;
 import org.bukkit.boss.KeyedBossBar;
 import org.bukkit.craftbukkit.util.CraftNamespacedKey;
+import org.bukkit.entity.Player; // Canvas
 
 public class CraftKeyedBossbar extends CraftBossBar implements KeyedBossBar {
 
@@ -20,4 +21,14 @@ public class CraftKeyedBossbar extends CraftBossBar implements KeyedBossBar {
     public CustomBossEvent getHandle() {
         return (CustomBossEvent) super.getHandle();
     }
+    // Canvas start - make a direct way for deleting bossbars
+
+    @Override
+    public void deleteBar(){
+        for(Player player : this.getPlayers()){
+            this.removePlayer(player);
+        }
+        org.bukkit.Bukkit.removeBossBar(this.getKey());
+    }
+    // Canvas end
 }
