From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Mon, 8 Jul 2024 12:40:13 -0700
Subject: [PATCH] Spectators can generate mobs


diff --git a/src/main/java/me/dueris/canvas/CanvasConfig.java b/src/main/java/me/dueris/canvas/CanvasConfig.java
index 9825b47fab89c05958ebfb93d830634fdd723ff2..6badaa66189d59e9888086a53f5a4af2af535c41 100644
--- a/src/main/java/me/dueris/canvas/CanvasConfig.java
+++ b/src/main/java/me/dueris/canvas/CanvasConfig.java
@@ -210,6 +210,7 @@ public class CanvasConfig {
     public static boolean useCanvasOrbOptimizations = false;
     public static double orbMergeSearchInfaltionRadius = 0.8;
     public static int maxEntitiesToReducePathfindingRates = 2;
+    public static boolean spectatorsSpawnMobs = false;
 
     public static boolean disableFeatures = false;
 
@@ -226,6 +227,7 @@ public class CanvasConfig {
         useCanvasOrbOptimizations = getBoolean("useCanvasOrbOptimizations", useCanvasOrbOptimizations);
         orbMergeSearchInfaltionRadius = getDouble("orbMergeSearchInfaltionRadius", orbMergeSearchInfaltionRadius);
         maxEntitiesToReducePathfindingRates = getInt("maxEntitiesToReducePathfindingRates", maxEntitiesToReducePathfindingRates);
+        spectatorsSpawnMobs = getBoolean("spectatorsCanSpawnNaturalMobs", spectatorsSpawnMobs);
     }
 
     private static void optimizations(){
diff --git a/src/main/java/net/minecraft/server/level/ChunkMap.java b/src/main/java/net/minecraft/server/level/ChunkMap.java
index be72b3e7a164e35ce6c7afbba6f9daf8f2d46b96..4b7c9c22c389cfbbc61c1445d478a63120618213 100644
--- a/src/main/java/net/minecraft/server/level/ChunkMap.java
+++ b/src/main/java/net/minecraft/server/level/ChunkMap.java
@@ -778,7 +778,7 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
     }
 
     private boolean playerIsCloseEnoughForSpawning(ServerPlayer entityplayer, ChunkPos chunkcoordintpair, double range) { // Spigot
-        if (entityplayer.isSpectator()) {
+        if (entityplayer.isSpectator() && !me.dueris.canvas.CanvasConfig.spectatorsSpawnMobs) { // Canvas
             return false;
         } else {
             double d0 = ChunkMap.euclideanDistanceSquared(chunkcoordintpair, entityplayer);
