From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Tue, 21 Nov 2023 02:27:58 +0000
Subject: [PATCH] Fix MC bugs and ExtringuishEvent


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 50916b224250526e2c8889cbb88d932292aa2ab3..33e9be7a89a9c2489cafd630775316b03ccc72c6 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -1073,6 +1073,11 @@ public class ServerPlayer extends Player {
         this.awardStat(Stats.DEATHS);
         this.resetStat(Stats.CUSTOM.get(Stats.TIME_SINCE_DEATH));
         this.resetStat(Stats.CUSTOM.get(Stats.TIME_SINCE_REST));
+         // Canvas start
+        if (this.getRemainingFireTicks() > 0) {
+            new me.dueris.canvas.event.entity.EntityExtinguishEvent(this.getBukkitEntity(), me.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.DEATH).callEvent();
+        }
+        // Canvas end
         this.clearFire();
         this.setTicksFrozen(0);
         this.setSharedFlagOnFire(false);
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 7ed6909c2eddc04f0943718163754aad722e250a..380adb0ebded098c0f52f6b2d1188e875a2cd8f1 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -886,7 +886,13 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
                     this.hurt(this.damageSources().onFire(), 1.0F);
                 }
 
-                this.setRemainingFireTicks(this.remainingFireTicks - 1);
+                // Canvas start
+                if (getRemainingFireTicks() - 1 == 0 && !new me.dueris.canvas.event.entity.EntityExtinguishEvent(this.getBukkitEntity(), me.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.TIME).callEvent()) {
+                    setRemainingFireTicks(20);
+                } else {
+                    this.setRemainingFireTicks(this.remainingFireTicks - 1);
+                }
+                // Canvas end
             }
 
             if (this.getTicksFrozen() > 0 && !freezeLocked) { // Paper - Freeze Tick Lock API
@@ -1315,7 +1321,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
                 }
 
                 if (this.isOnFire() && (this.isInPowderSnow || this.isInWaterRainOrBubble())) {
-                    this.setRemainingFireTicks(-this.getFireImmuneTicks());
+                    // Canvas start
+                    if (this.callExtinguishEvent(iblockdata)){
+                        this.setRemainingFireTicks(-this.getFireImmuneTicks());
+                    }
+                    // Canvas end
                 }
 
                 //this.level().getProfiler().pop(); // Purpur
@@ -1372,7 +1382,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
         }
     }
 
-    protected void playEntityOnFireExtinguishedSound() {
+    public void playEntityOnFireExtinguishedSound() { // Canvas - protected -> public
         this.playSound(SoundEvents.GENERIC_EXTINGUISH_FIRE, 0.7F, 1.6F + (this.random.nextFloat() - this.random.nextFloat()) * 0.4F);
     }
 
@@ -1908,12 +1918,41 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
 
             this.resetFallDistance();
             this.wasTouchingWater = true;
-            this.clearFire();
+            // Canvas start
+            if (this.getRemainingFireTicks() <= 0 || new me.dueris.canvas.event.entity.EntityExtinguishEvent(this.getBukkitEntity(), me.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.WATER).callEvent()) {
+                if (this.getRemainingFireTicks() > 0 && this instanceof net.minecraft.world.entity.projectile.AbstractArrow){
+                    this.playEntityOnFireExtinguishedSound();
+                }
+                this.clearFire();
+            }
+            // Canvas end
         } else {
             this.wasTouchingWater = false;
         }
 
     }
+    // Canvas start
+
+    public boolean callExtinguishEvent(@Nullable BlockState state) {
+        final me.dueris.canvas.event.entity.EntityExtinguishEvent.Cause cause;
+        final boolean onFire = getRemainingFireTicks() > 0;
+        if (!onFire) {
+            cause = null;
+        } else if (this.isInPowderSnow || (state != null && state.is(Blocks.POWDER_SNOW))) {
+            cause = me.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.POWDER_SNOW;
+        } else if (this.isInWater()) {
+            cause = me.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.WATER;
+        } else if (this.isInRain()) {
+            cause = me.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.RAIN;
+       } else if (this.isInBubbleColumn()) {
+            cause = me.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.BUBBLE_COLUMN;
+        } else {
+            cause = null;
+        }
+
+        return cause == null || new me.dueris.canvas.event.entity.EntityExtinguishEvent(this.getBukkitEntity(), cause).callEvent();
+    }
+    // Canvas end
 
     private void updateFluidOnEyes() {
         this.wasEyeInWater = this.isEyeInFluid(FluidTags.WATER);
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index be3fafb3b858070f529a9a75c149eefcc92228e7..c07d2f983574d2dec9943c03500056c87bf51032 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -473,7 +473,11 @@ public abstract class LivingEntity extends Entity implements Attackable {
         }
 
         if (this.isAlive() && (this.isInWaterRainOrBubble() || this.isInPowderSnow)) {
-            this.extinguishFire();
+            // Canvas start
+            if (callExtinguishEvent(null)){
+                this.extinguishFire();
+            }
+            // Canvas end
         }
 
         if (this.hurtTime > 0) {
@@ -1817,6 +1821,11 @@ public abstract class LivingEntity extends Entity implements Attackable {
                         entity.killedEntity((ServerLevel) this.level(), this);
                     }
                     this.gameEvent(GameEvent.ENTITY_DIE);
+                    // Canvas start
+                    if (this.getRemainingFireTicks() > 0){
+                        new me.dueris.canvas.event.entity.EntityExtinguishEvent(this.getBukkitEntity(), me.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.DEATH).callEvent();
+                    }
+                    // Canvas end
                 } else {
                     this.dead = false;
                     this.setHealth((float) deathEvent.getReviveHealth());
diff --git a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
index e027d5a982be5e3f56ca310eb066eb7fc05905fc..0eb6efc20ba3ff2cc3f8b3f07297e2f8e5d34b4d 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/AbstractArrow.java
@@ -196,7 +196,11 @@ public abstract class AbstractArrow extends Projectile {
         }
 
         if (this.isInWaterOrRain() || iblockdata.is(Blocks.POWDER_SNOW)) {
-            this.clearFire();
+            // Canvas start
+            if (this.callExtinguishEvent(iblockdata)){
+                this.clearFire();
+            }
+            // Canvas end
         }
 
         if (this.inGround && !flag) {
diff --git a/src/main/java/net/minecraft/world/entity/projectile/ThrownPotion.java b/src/main/java/net/minecraft/world/entity/projectile/ThrownPotion.java
index 114561c83c88265f9c0dc1ed66d8ecdc86a17ff8..a532f586b524ce13518e5f4ff20cb1239e446aac 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/ThrownPotion.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/ThrownPotion.java
@@ -167,7 +167,11 @@ public class ThrownPotion extends ThrowableItemProjectile implements ItemSupplie
                 ((CraftLivingEntity) affectedEntity).getHandle().hurt(this.damageSources().indirectMagic(this, this.getOwner()), 1.0F);
             }
             for (LivingEntity toExtinguish : event.getToExtinguish()) {
-                ((CraftLivingEntity) toExtinguish).getHandle().extinguishFire();
+                // Canvas start
+                if (new me.dueris.canvas.event.entity.EntityExtinguishEvent(toExtinguish, me.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.WATER_BOTTLE).callEvent()){
+                    ((CraftLivingEntity) toExtinguish).getHandle().extinguishFire();
+                }
+                // Canvas end
             }
             for (LivingEntity toRehydrate : event.getToRehydrate()) {
                 if (((CraftLivingEntity) toRehydrate).getHandle() instanceof Axolotl axolotl) {
diff --git a/src/main/java/net/minecraft/world/level/block/LayeredCauldronBlock.java b/src/main/java/net/minecraft/world/level/block/LayeredCauldronBlock.java
index a4857675772d4fe849ba85fc21a369decca42fc0..3a4b155041f61f16c9dcdf2fb69d318c502fc74d 100644
--- a/src/main/java/net/minecraft/world/level/block/LayeredCauldronBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/LayeredCauldronBlock.java
@@ -73,8 +73,11 @@ public class LayeredCauldronBlock extends AbstractCauldronBlock {
                     return;
                 }
             }
-            entity.clearFire();
-            // CraftBukkit end
+            // Canvas start
+            if (new me.dueris.canvas.event.entity.EntityExtinguishEvent(entity.getBukkitEntity(), me.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.CAULDRON).callEvent()){
+                entity.clearFire();
+            }
+            // CraftBukkit end -/- Canvas end
         }
 
     }
diff --git a/src/main/java/net/minecraft/world/level/block/PowderSnowBlock.java b/src/main/java/net/minecraft/world/level/block/PowderSnowBlock.java
index 2dbf71e421156093c8bc387941eae991f5e6c957..81fbbf119edebd39c30842b98a77a78710b26013 100644
--- a/src/main/java/net/minecraft/world/level/block/PowderSnowBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/PowderSnowBlock.java
@@ -83,7 +83,16 @@ public class PowderSnowBlock extends Block implements BucketPickup {
                 if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entity, pos, Blocks.AIR.defaultBlockState(), !((world.purpurConfig.powderSnowBypassMobGriefing || world.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) || entity instanceof Player))) {
                     return;
                 }
-                // CraftBukkit end
+                // CraftBukkit end -/- Canvas start
+                if (entity instanceof net.minecraft.world.entity.projectile.AbstractArrow) {
+                    if (!new me.dueris.canvas.event.entity.EntityExtinguishEvent(entity.getBukkitEntity(), me.dueris.canvas.event.entity.EntityExtinguishEvent.Cause.POWDER_SNOW).callEvent()) {
+                        return;
+                    }
+
+                    entity.playEntityOnFireExtinguishedSound();
+                    entity.clearFire();
+                }
+                // Canvas end
                 world.destroyBlock(pos, false);
             }
 
