From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Tue, 21 Nov 2023 19:14:51 +0000
Subject: [PATCH] Add getReloadCount() to Server


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 638bc1234d5856d890c19ba49a4baae18de6671a..c0fb5ca6542cc1e27edad7b1650e27e9731d903b 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1302,6 +1302,13 @@ public final class CraftServer implements Server {
 
         return false;
     }
+    // Canvas start
+
+    @Override
+    public int getReloadCount(){
+        return this.reloadCount;
+    }
+    // Canvas end
 
     @Override
     public void reload() {
@@ -1402,6 +1409,7 @@ public final class CraftServer implements Server {
         }
         io.papermc.paper.plugin.PluginInitializerManager.reload(this.console); // Paper
         me.dueris.canvas.plugin.CanvasPluginLoader.loader(); // Canvas - reload cache mappings
+        
         this.loadPlugins();
         this.enablePlugins(PluginLoadOrder.STARTUP);
         this.enablePlugins(PluginLoadOrder.POSTWORLD);
