From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Fri, 8 Dec 2023 05:05:48 +0000
Subject: [PATCH] Add PlayerUseRespawnAnchorEvent


diff --git a/src/main/java/net/minecraft/world/level/block/RespawnAnchorBlock.java b/src/main/java/net/minecraft/world/level/block/RespawnAnchorBlock.java
index ff315a10bcc605dc4e02a8de6b7d4f10db543a6c..780644ecb8fc3ddbade97062a1be2f91e8efd652 100644
--- a/src/main/java/net/minecraft/world/level/block/RespawnAnchorBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/RespawnAnchorBlock.java
@@ -8,6 +8,7 @@ import java.util.Objects;
 import java.util.Optional;
 import java.util.stream.Stream;
 import javax.annotation.Nullable;
+import me.dueris.canvas.event.player.PlayerUseRespawnAnchorEvent;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.core.Vec3i;
@@ -64,41 +65,49 @@ public class RespawnAnchorBlock extends Block {
     @Override
     public InteractionResult use(BlockState state, Level world, BlockPos pos, Player player, InteractionHand hand, BlockHitResult hit) {
         ItemStack itemstack = player.getItemInHand(hand);
+        // Canvas start - add PlayerUseRespawnAnchorEvent
+        org.bukkit.entity.Player bukkitPlayer = player.getBukkitEntity() instanceof org.bukkit.entity.Player ? (org.bukkit.entity.Player) player.getBukkitEntity() : null;
+        org.bukkit.block.Block block = org.bukkit.craftbukkit.block.CraftBlock.at(world, pos);
+        PlayerUseRespawnAnchorEvent useRespawnAnchorEvent = new PlayerUseRespawnAnchorEvent(block, bukkitPlayer);
+        useRespawnAnchorEvent.callEvent();
+        if(!useRespawnAnchorEvent.getCanceled()){
+            if (hand == InteractionHand.MAIN_HAND && !RespawnAnchorBlock.isRespawnFuel(itemstack) && RespawnAnchorBlock.isRespawnFuel(player.getItemInHand(InteractionHand.OFF_HAND))) {
+                return InteractionResult.PASS;
+            } else if (RespawnAnchorBlock.isRespawnFuel(itemstack) && RespawnAnchorBlock.canBeCharged(state)) {
+                RespawnAnchorBlock.charge(player, world, pos, state);
+                if (!player.getAbilities().instabuild) {
+                    itemstack.shrink(1);
+                }
 
-        if (hand == InteractionHand.MAIN_HAND && !RespawnAnchorBlock.isRespawnFuel(itemstack) && RespawnAnchorBlock.isRespawnFuel(player.getItemInHand(InteractionHand.OFF_HAND))) {
-            return InteractionResult.PASS;
-        } else if (RespawnAnchorBlock.isRespawnFuel(itemstack) && RespawnAnchorBlock.canBeCharged(state)) {
-            RespawnAnchorBlock.charge(player, world, pos, state);
-            if (!player.getAbilities().instabuild) {
-                itemstack.shrink(1);
-            }
-
-            return InteractionResult.sidedSuccess(world.isClientSide);
-        } else if ((Integer) state.getValue(RespawnAnchorBlock.CHARGE) == 0) {
-            return InteractionResult.PASS;
-        } else if (!RespawnAnchorBlock.canSetSpawn(world)) {
-            if (!world.isClientSide) {
-                this.explode(state, world, pos);
-            }
+                return InteractionResult.sidedSuccess(world.isClientSide);
+            } else if ((Integer) state.getValue(RespawnAnchorBlock.CHARGE) == 0) {
+                return InteractionResult.PASS;
+            } else if (!RespawnAnchorBlock.canSetSpawn(world)) {
+                if (!world.isClientSide) {
+                    this.explode(state, world, pos);
+                }
 
-            return InteractionResult.sidedSuccess(world.isClientSide);
-        } else {
-            if (!world.isClientSide) {
-                ServerPlayer entityplayer = (ServerPlayer) player;
-
-                if (entityplayer.getRespawnDimension() != world.dimension() || !pos.equals(entityplayer.getRespawnPosition())) {
-                    if (entityplayer.setRespawnPosition(world.dimension(), pos, 0.0F, false, true, com.destroystokyo.paper.event.player.PlayerSetSpawnEvent.Cause.RESPAWN_ANCHOR)) { // Paper - PlayerSetSpawnEvent
-                    world.playSound((Player) null, (double) pos.getX() + 0.5D, (double) pos.getY() + 0.5D, (double) pos.getZ() + 0.5D, SoundEvents.RESPAWN_ANCHOR_SET_SPAWN, SoundSource.BLOCKS, 1.0F, 1.0F);
-                    return InteractionResult.SUCCESS;
-                    // Paper start - handle failed set spawn
-                    } else {
-                        return InteractionResult.FAIL;
+                return InteractionResult.sidedSuccess(world.isClientSide);
+            } else {
+                if (!world.isClientSide) {
+                    ServerPlayer entityplayer = (ServerPlayer) player;
+
+                    if (entityplayer.getRespawnDimension() != world.dimension() || !pos.equals(entityplayer.getRespawnPosition())) {
+                        if (entityplayer.setRespawnPosition(world.dimension(), pos, 0.0F, false, true, com.destroystokyo.paper.event.player.PlayerSetSpawnEvent.Cause.RESPAWN_ANCHOR)) { // Paper - PlayerSetSpawnEvent
+                            world.playSound((Player) null, (double) pos.getX() + 0.5D, (double) pos.getY() + 0.5D, (double) pos.getZ() + 0.5D, SoundEvents.RESPAWN_ANCHOR_SET_SPAWN, SoundSource.BLOCKS, 1.0F, 1.0F);
+                            return InteractionResult.SUCCESS;
+                            // Paper start - handle failed set spawn
+                        } else {
+                            return InteractionResult.FAIL;
+                        }
+                        // Paper end
                     }
-                    // Paper end
                 }
-            }
 
-            return InteractionResult.CONSUME;
+                return InteractionResult.CONSUME;
+            }
+        }else{
+            return InteractionResult.PASS;
         }
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index f8756d74677180af6f0a5cbc09a5cf64ae0e98e4..f8b15eb6e8b42a1dfe544fa466c87522a56763a9 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -5,8 +5,7 @@ import me.dueris.canvas.CraftServerGui;
 import me.dueris.canvas.logger.CanvasLogger;
 import me.dueris.canvas.logger.CraftCanvasLogger;
 import me.dueris.canvas.logger.Obfuscator;
-import me.dueris.canvas.network.ChatChannelContainer;
-import me.dueris.canvas.network.ChatChannelManager;
+import me.dueris.canvas.properties.*;
 import com.google.common.base.Charsets;
 import com.google.common.base.Function;
 import com.google.common.base.Preconditions;
