From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Wed, 22 Nov 2023 09:00:21 +0000
Subject: [PATCH] Add createEndPlatformAPI


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index db14f96fd8a376070c9c6a853eeaeb4dec6695bf..7990834a859dfe34fe8d15e86f0a5550aa2e4581 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -1272,7 +1272,7 @@ public class ServerPlayer extends Player {
                 if (resourcekey == LevelStem.OVERWORLD && worldserver.getTypeKey() == LevelStem.NETHER) { // CraftBukkit
                     this.enteredNetherPosition = this.position();
                 } else if (worldserver.getTypeKey() == LevelStem.END && shapedetectorshape.portalEventInfo != null && shapedetectorshape.portalEventInfo.getCanCreatePortal()) { // CraftBukkit
-                    this.createEndPlatform(worldserver, BlockPos.containing(shapedetectorshape.pos));
+                    this.createEndPlatform(worldserver, BlockPos.containing(shapedetectorshape.pos), true); // Canvas
                 }
                 // CraftBukkit start
             } else {
@@ -1355,7 +1355,7 @@ public class ServerPlayer extends Player {
     }
     // CraftBukkit end
 
-    private void createEndPlatform(ServerLevel world, BlockPos centerPos) {
+    public void createEndPlatform(ServerLevel world, BlockPos centerPos, boolean callEvent) { // Canvas - private -> public
         BlockPos.MutableBlockPos blockposition_mutableblockposition = centerPos.mutable();
         org.bukkit.craftbukkit.util.BlockStateListPopulator blockList = new org.bukkit.craftbukkit.util.BlockStateListPopulator(world); // CraftBukkit
 
@@ -1368,13 +1368,17 @@ public class ServerPlayer extends Player {
                 }
             }
         }
-        // CraftBukkit start - call portal event
-        org.bukkit.event.world.PortalCreateEvent portalEvent = new org.bukkit.event.world.PortalCreateEvent((List<org.bukkit.block.BlockState>) (List) blockList.getList(), world.getWorld(), this.getBukkitEntity(), org.bukkit.event.world.PortalCreateEvent.CreateReason.END_PLATFORM);
-        world.getCraftServer().getPluginManager().callEvent(portalEvent);
-        if (!portalEvent.isCancelled()) {
-            blockList.updateList();
+        // Canvas start
+        if(callEvent){
+            // CraftBukkit start - call portal event
+            org.bukkit.event.world.PortalCreateEvent portalEvent = new org.bukkit.event.world.PortalCreateEvent((List<org.bukkit.block.BlockState>) (List) blockList.getList(), world.getWorld(), this.getBukkitEntity(), org.bukkit.event.world.PortalCreateEvent.CreateReason.END_PLATFORM);
+            world.getCraftServer().getPluginManager().callEvent(portalEvent);
+            if (!portalEvent.isCancelled()) {
+                blockList.updateList();
+            }
+            // CraftBukkit end
         }
-        // CraftBukkit end
+        // Canvas end        
 
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 139349d7ea857f608b73c3cce3430249ff8ec34b..8998d86c582efb9870a1bc8ab918e8ffadc2d265 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1337,6 +1337,24 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         this.lookAt(targetLocation, io.papermc.paper.entity.LookAnchor.EYES);
         // Paper end
     }
+    // Canvas start
+
+    @Override
+    public void createEndPlatformAtCurrentPos(){
+        this.createEndPlatformAtCurrentPos(false);
+    }
+
+    @Override
+    public void createEndPlatformAtCurrentPos(boolean callEvent){
+        ServerLevel level = ((CraftWorld) this.getWorld()).getHandle();
+        BlockPos pos = new BlockPos(
+            Math.toIntExact(carpetfixes.helpers.FastMath.round(this.getLocation().getX())),
+            Math.toIntExact(carpetfixes.helpers.FastMath.round(this.getLocation().getY())),
+            Math.toIntExact(carpetfixes.helpers.FastMath.round(this.getLocation().getZ()))
+        );
+        this.getHandle().createEndPlatform(level, pos, callEvent);
+    }
+    // Canvas end
 
     @Override
     public boolean teleport(Location location, PlayerTeleportEvent.TeleportCause cause) {
