From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Fri, 8 Dec 2023 05:29:33 +0000
Subject: [PATCH] Enhance Enchantment api


diff --git a/src/main/java/net/minecraft/world/item/enchantment/Enchantment.java b/src/main/java/net/minecraft/world/item/enchantment/Enchantment.java
index 7ff3741e189905637f96ed7ff99e1de0af0d43c0..acc0650087849362646dfab1eadd5c149f05f0d6 100644
--- a/src/main/java/net/minecraft/world/item/enchantment/Enchantment.java
+++ b/src/main/java/net/minecraft/world/item/enchantment/Enchantment.java
@@ -78,8 +78,16 @@ public abstract class Enchantment {
     }
 
     public final boolean isCompatibleWith(Enchantment other) {
-        return this.checkCompatibility(other) && other.checkCompatibility(this);
+        return canConflict ? this.checkCompatibility(other) && other.checkCompatibility(this) : true; // Canvas
     }
+    // Canvas start
+
+    public boolean canConflict = true;
+
+    public void setCanConflict(boolean canConflict){
+        this.canConflict = canConflict;
+    }
+    // Canvas end
 
     protected boolean checkCompatibility(Enchantment other) {
         return this != other;
diff --git a/src/main/java/org/bukkit/craftbukkit/enchantments/CraftEnchantment.java b/src/main/java/org/bukkit/craftbukkit/enchantments/CraftEnchantment.java
index f782e8dd9d3456f345859dffd69e5830d576ccd1..04e96e7ad847bda81343f3256ab604be9a75efeb 100644
--- a/src/main/java/org/bukkit/craftbukkit/enchantments/CraftEnchantment.java
+++ b/src/main/java/org/bukkit/craftbukkit/enchantments/CraftEnchantment.java
@@ -1,8 +1,10 @@
 package org.bukkit.craftbukkit.enchantments;
 
 import com.google.common.base.Preconditions;
+import me.dueris.canvas.world.damage.CraftDamageSource;
 import net.minecraft.core.registries.BuiltInRegistries;
 import net.minecraft.core.registries.Registries;
+import net.minecraft.world.damagesource.DamageSources;
 import net.minecraft.world.item.enchantment.BindingCurseEnchantment;
 import net.minecraft.world.item.enchantment.VanishingCurseEnchantment;
 import org.bukkit.NamespacedKey;
@@ -158,7 +160,18 @@ public class CraftEnchantment extends Enchantment {
         CraftEnchantment ench = (CraftEnchantment) other;
         return !this.handle.isCompatibleWith(ench.getHandle());
     }
-    // Paper start
+    // Canvas start
+
+    @Override
+    public void setCanConflict(boolean canConflict){
+        this.getHandle().setCanConflict(canConflict);
+    }
+
+    @Override
+    public boolean getCanConflict(){
+        return this.getHandle().canConflict;
+    }
+    // Canvas end -/- Paper start
     @Override
     public net.kyori.adventure.text.Component displayName(int level) {
         return io.papermc.paper.adventure.PaperAdventure.asAdventure(getHandle().getFullname(level));
