From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <159172167+PurpleWolfMC@users.noreply.github.com>
Date: Thu, 25 Apr 2024 08:43:43 +0000
Subject: [PATCH] Set ThreadPriority of startup thread


diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
index ce49590248d941e8245b8d7be4d1646d78950656..7f8ce4fc9d1bd90c5c0750b3a2f2bafc2fd4be0b 100644
--- a/src/main/java/net/minecraft/server/Main.java
+++ b/src/main/java/net/minecraft/server/Main.java
@@ -97,6 +97,8 @@ public class Main {
     public static void main(final OptionSet optionset) { // CraftBukkit - replaces main(String[] astring)
         // me.jellysquid.mods.hydrogen.common.jvm.ClassConstructors.init(); // JettPack // Canvas
         SharedConstants.tryDetectVersion();
+ 
+        if(Thread.currentThread().getPriority() < 6){ Thread.currentThread().setPriority(6); } // Canvas - improve startup times (smoothboot)
         /* CraftBukkit start - Replace everything
         OptionParser optionparser = new OptionParser();
         OptionSpec<Void> optionspec = optionparser.accepts("nogui");
