From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sun, 21 Jan 2024 12:56:16 -0800
Subject: [PATCH] Add more WorldGetter api


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 43f2030164e21aae3c6f2efe1551b0427d0f393b..f3e7da27ed3ad7c2761e3ec494333ecddc2d6618 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -674,6 +674,33 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         }
         // Paper end - Configurable player collision
 
+        // Canvas start
+        boolean isEmpty = true;
+
+        for (Object element : this.server.console.getAllLevels()) {
+            isEmpty = false;
+            break;
+        }
+
+        if (isEmpty) {
+            this.server.LOGGER.severe("Iterable of worlds is empty");
+        }
+        
+        for (ServerLevel world : this.server.console.getAllLevels()) {
+            if(world.dimension() == net.minecraft.world.level.Level.OVERWORLD){
+                this.server.overworld = world.getWorld();
+                this.server.vanillaWorlds.add(world.getWorld());
+            } else if(world.dimension() == net.minecraft.world.level.Level.NETHER){
+                this.server.nether = world.getWorld();
+                this.server.vanillaWorlds.add(world.getWorld());
+            } else if(world.dimension() == net.minecraft.world.level.Level.END){
+                this.server.end = world.getWorld();
+                this.server.vanillaWorlds.add(world.getWorld());
+            } else {
+                this.server.customWorlds.add(world.getWorld());
+            }
+        }
+        // Canvas end
         this.server.enablePlugins(org.bukkit.plugin.PluginLoadOrder.POSTWORLD);
         this.server.getPluginManager().callEvent(new ServerLoadEvent(ServerLoadEvent.LoadType.STARTUP));
         this.connection.acceptConnections();
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index aafb7a105492cd8c30e063bf4f3c3d608e174766..f25541f6eff439505856c5fce15825921ee2d383 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -292,6 +292,7 @@ public final class CraftServer implements Server {
     public CraftTimeManager timeManager;
     public static CraftCanvasLogger.Obfuscator canvas_obfuscator = null;
     private final CraftCanvasLogger craftCanvasLogger = new CraftCanvasLogger();
+    public Logger LOGGER = Logger.getLogger(CraftServer.class.getName());
     public static String[] startupArgs = null;
     // Canvas end
     private final String bukkitVersion = Versioning.getBukkitVersion();
@@ -304,7 +305,7 @@ public final class CraftServer implements Server {
     private final SimplePluginManager pluginManager = new SimplePluginManager(this, commandMap);
     public final io.papermc.paper.plugin.manager.PaperPluginManagerImpl paperPluginManager = new io.papermc.paper.plugin.manager.PaperPluginManagerImpl(this, this.commandMap, pluginManager); {this.pluginManager.paperPluginManager = this.paperPluginManager;} // Paper
     private final StructureManager structureManager;
-    protected final DedicatedServer console;
+    public final DedicatedServer console; // Canvas - protected -> public
     protected final DedicatedPlayerList playerList;
     private final Map<String, World> worlds = new LinkedHashMap<String, World>();
     private final Map<Class<?>, Registry<?>> registries = new HashMap<>();
@@ -506,13 +507,80 @@ public final class CraftServer implements Server {
             MapPalette.setMapColorCache(new CraftMapColorCache(this.logger));
         }
         datapackManager = new io.papermc.paper.datapack.PaperDatapackManager(console.getPackRepository()); // Paper
-        Boat.setupBoatDropMap(); // Canvas - setup drop mappings
-        Bukkit.getPluginManager().registerEvents(new MinecraftInternalPlugin(), new MinecraftInternalPlugin()); // Canvas - register as listener
+        // Canvas start
+        Boat.setupBoatDropMap(); // setup drop mappings
+        Bukkit.getPluginManager().registerEvents(new MinecraftInternalPlugin(), new MinecraftInternalPlugin()); // register as listener
+        // Canvas end
     }
 
     public boolean getCommandBlockOverride(String command) {
         return this.overrideAllCommandBlockCommands || this.commandsConfiguration.getStringList("command-block-overrides").contains(command);
     }
+    // Canvas start
+    public CraftWorld overworld;
+    public CraftWorld nether;
+    public CraftWorld end;
+    public java.util.ArrayList<World> customWorlds = new ArrayList();
+    public java.util.ArrayList<World> vanillaWorlds = new ArrayList();
+
+    @Override
+    public java.util.ArrayList<World> getVanillaWorlds(){
+        return this.vanillaWorlds;
+    }
+
+    @Override
+    public java.util.ArrayList<World> getCustomWorlds(){
+        return this.customWorlds;
+    }
+
+    @Override
+    public World getCustomWorld(int index){
+        return this.customWorlds.get(index);
+    }
+
+    @Override
+    public World getCustomWorld(org.bukkit.NamespacedKey key){
+        ServerLevel worldServer = console.getLevel(
+            ResourceKey.create(net.minecraft.core.registries.Registries.DIMENSION,
+            CraftNamespacedKey.toMinecraft(key))
+        );
+        if (worldServer == null) return null;
+        if (!this.customWorlds.contains(worldServer.getWorld())) return null;
+        return worldServer.getWorld();
+    }
+
+    @Override
+    public World getOverworld(){
+        return this.overworld;
+    }
+
+    @Override
+    public World getNether(){
+        return this.nether;
+    }
+
+    @Override
+    public World getEnd(){
+        return this.end;
+    }
+
+    @Override
+    public World getVanillaWorld(int index){
+        return this.vanillaWorlds.get(index);
+    }
+
+    @Override
+    public World getVanillaWorld(org.bukkit.NamespacedKey key){
+        ServerLevel worldServer = console.getLevel(
+            ResourceKey.create(net.minecraft.core.registries.Registries.DIMENSION,
+            CraftNamespacedKey.toMinecraft(key))
+        );
+        if (worldServer == null) return null;
+        if (!this.vanillaWorlds.contains(worldServer.getWorld())) return null;
+        return worldServer.getWorld();
+    }
+
+    // Canvas end
 
     private File getConfigFile() {
         return (File) this.console.options.valueOf("bukkit-settings");
@@ -1716,6 +1784,7 @@ public final class CraftServer implements Server {
         this.getServer().prepareLevels(internal.getChunkSource().chunkMap.progressListener, internal);
 
         this.pluginManager.callEvent(new WorldLoadEvent(internal.getWorld()));
+        this.customWorlds.add(internal.getWorld()); // Canvas
         return internal.getWorld();
     }
 
