From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Fri, 17 Nov 2023 20:58:10 +0000
Subject: [PATCH] Add gamemode syncing api


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index c9f01bef75d5ab8a94e95feee1b7dafab8e29363..c3ad2a5257cb71385fa76e06afa226f7f6de613d 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -281,6 +281,64 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
 
         this.perm.recalculatePermissions();
     }
+    // Canvas start - more player api
+
+    /**
+     * For those seeing this source, its documented what it means
+     * in the API interface, please see that for details on how
+     * this works and what this means
+     */
+    @Override
+    public void setClientGamemode(GameMode gamemode){
+        this.getHandle().connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.CHANGE_GAME_MODE, this.getGamemodeInt(gamemode)));
+    }
+
+    /**
+     * For those seeing this source, its documented what it means
+     * in the API interface, please see that for details on how
+     * this works and what this means
+     */
+    @Override
+    public void setServerGamemode(GameMode gamemode){
+        final GameMode lastGamemode = this.getGameMode();
+        this.setGameMode(gamemode);
+        this.getHandle().connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.CHANGE_GAME_MODE, this.getGamemodeInt(lastGamemode)));
+    }
+
+    private int getGamemodeInt(GameMode gamemode){
+        switch(gamemode){
+            case SURVIVAL -> {
+                return 0;
+            }
+
+            case CREATIVE -> {
+                return 1;
+            }
+
+            case SPECTATOR -> {
+                return 3;
+            }
+
+            case ADVENTURE -> {
+                return 2;
+            }
+
+            default -> {
+                return 0;
+            }
+        }
+    }
+
+    /**
+     * For those seeing this source, its documented what it means
+     * in the API interface, please see that for details on how
+     * this works and what this means
+     */
+    @Override
+    public void resyncGamemode(){
+        this.setGameMode(this.getGameMode());
+    }
+    // Canvas end
 
     @Override
     public boolean isOnline() {
