From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Fri, 8 Dec 2023 06:23:21 +0000
Subject: [PATCH] Expose convertTo to CraftBukkit


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
index ce8af08831e6fd4a2d05161a8049b5f49a7f9a74..24d777f6fd0599217e729cc61f57253962fba6b6 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftMob.java
@@ -83,6 +83,14 @@ public abstract class CraftMob extends CraftLivingEntity implements Mob {
     public boolean isAware() {
         return this.getHandle().aware;
     }
+    // Canvas start
+
+    @Override
+    public void convertTo(EntityType entityType, boolean keepEquipment){
+        net.minecraft.world.entity.EntityType ENT = org.bukkit.craftbukkit.entity.CraftEntityType.bukkitToMinecraft(entityType);
+        this.getHandle().convertTo(ENT, keepEquipment);
+    }
+    // Canvas end
 
     @Override
     public Sound getAmbientSound() {
