From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sun, 18 Feb 2024 16:22:31 -0800
Subject: [PATCH] Implement pufferfish snakeyaml


diff --git a/build.gradle.kts b/build.gradle.kts
index 17b90f6d52f8c9fdfcecd44e81f5a81de280afb3..67d63ab397e68b4b57c89623fbc3d61fbec69b5c 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -73,6 +73,13 @@ dependencies {
     testImplementation("org.mockito:mockito-core:5.5.0")
     testImplementation("org.ow2.asm:asm-tree:9.5")
     testImplementation("org.junit-pioneer:junit-pioneer:2.2.0") // Paper - CartesianTest
+    // Canvas - implement pufferfish snakeyaml
+
+    implementation("org.yaml:snakeyaml:1.32")
+    implementation ("com.github.carleslc.Simple-YAML:Simple-Yaml:1.8.4") {
+        exclude(group="org.yaml", module="snakeyaml")
+    }
+    // Canvas end
 }
 
 val craftbukkitPackageVersion = "1_20_R3" // Paper
