From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Fri, 17 Nov 2023 22:35:36 +0000
Subject: [PATCH] Ensure that plugins in cache are loaded on /reload


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 16847e6b4a71e222a4359e85aa51a3ebdc121c25..0ecb5747e0f6bed57c64d20fd88d0e2874767ee4 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1404,6 +1404,7 @@ public final class CraftServer implements Server {
             if (console.isDebugging()) io.papermc.paper.util.TraceUtil.dumpTraceForThread(worker.getThread(), "still running"); // Paper
         }
         io.papermc.paper.plugin.PluginInitializerManager.reload(this.console); // Paper
+        me.dueris.canvas.plugin.CanvasPluginLoader.loader(); // Canvas - reload cache mappings
         this.loadPlugins();
         this.enablePlugins(PluginLoadOrder.STARTUP);
         this.enablePlugins(PluginLoadOrder.POSTWORLD);
