From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sun, 18 Feb 2024 16:22:31 -0800
Subject: [PATCH] Implement pufferfish snakeyaml


diff --git a/build.gradle.kts b/build.gradle.kts
index 04fed13df340fb4a7aaa35ca30f32fa7ec12cf1d..702c83746217ef2b19a89c422476a9570aa50e13 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -72,6 +72,13 @@ dependencies {
     testImplementation("org.hamcrest:hamcrest:2.2")
     testImplementation("org.mockito:mockito-core:5.5.0")
     testImplementation("org.junit-pioneer:junit-pioneer:2.2.0") // Paper - CartesianTest
+    // Canvas - implement pufferfish snakeyaml
+
+    implementation("org.yaml:snakeyaml:1.32")
+    implementation ("com.github.carleslc.Simple-YAML:Simple-Yaml:1.8.4") {
+        exclude(group="org.yaml", module="snakeyaml")
+    }
+    // Canvas end
 }
 
 val craftbukkitPackageVersion = "1_20_R3" // Paper
