From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris Bauer <jedimastertoothless@hotmail.com>
Date: Fri, 27 Oct 2023 07:18:37 +0000
Subject: [PATCH] Add getStartupArgs to Server


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 56ae2eeb3967bd321878ddecbf47f6196127f988..aec6c1940ae3b2149ce741977a9652dd42993dd6 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -285,6 +285,7 @@ public final class CraftServer implements Server {
     public static CraftCanvasLogger.Obfuscator canvas_obfuscator = null;
     private final CraftCanvasLogger craftCanvasLogger = new CraftCanvasLogger();
     private final ChatChannelManager chatChannelManager = new ChatChannelManager();
+    public static String[] startupArgs = null;
     // Canvas end
     private final String bukkitVersion = Versioning.getBukkitVersion();
     private final Logger logger = Logger.getLogger("Minecraft");
@@ -686,7 +687,6 @@ public final class CraftServer implements Server {
     }
 
     // Canvas start
-
     @Override
     public @NotNull Obfuscator getObfuscator() {
         return this.canvas_obfuscator;
@@ -707,30 +707,40 @@ public final class CraftServer implements Server {
         return this.chatChannelManager;
     }
 
+    @Override
     public net.minecraft.server.gui.MinecraftServerGui getGuiHandle(){
         return this.console.getGui();
     }
 
+    @Override
     public long getResourcesTotal(){
         return this.getGuiHandle().getGuiStatsComponent().getRamDetails().getGraphData().getTotal();
     }
 
+    @Override
     public long getResourcesFree(){
         return this.getGuiHandle().getGuiStatsComponent().getRamDetails().getGraphData().getFree();
     }
 
+    @Override
     public long getResourcesMax(){
         return this.getGuiHandle().getGuiStatsComponent().getRamDetails().getGraphData().getMax();
     }
 
+    @Override
     public long getResourcesUsedMem(){
         return this.getGuiHandle().getGuiStatsComponent().getRamDetails().getGraphData().getUsedMem();
     }
 
+    @Override
     public int getResourcesUsedPercent(){
         return this.getGuiHandle().getGuiStatsComponent().getRamDetails().getGraphData().getUsedPercent();
     }
 
+    @Override
+    public String[] getStartupArgs(){
+        return CraftServer.startupArgs != null ? CraftServer.startupArgs : null};
+    }
     // Canvas end
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 0ee05e3cee7f59ad6197c414620ff3ec6a2f6dd9..609d34ae03dd32cf8ea574d323694f5dcf84506a 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -408,6 +408,7 @@ public class Main {
         // Canvas start
 
         CraftServer.canvas_obfuscator = new CraftCanvasLogger.Obfuscator();
+        CraftServer.startupArgs = args;
 
         // Canvas end
     }
