From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Fri, 8 Dec 2023 07:32:03 +0000
Subject: [PATCH] Implement pufferfish snakeyaml


diff --git a/build.gradle.kts b/build.gradle.kts
index 75c93095727fcab9e3a2fb35eb230b011dc61854..15ae70df0a942c6f2d9100232398abb4c8329996 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -71,6 +71,13 @@ dependencies {
     testImplementation("org.junit.jupiter:junit-jupiter:5.10.0")
     testImplementation("org.hamcrest:hamcrest:2.2")
     testImplementation("org.mockito:mockito-core:5.5.0")
+    // Canvas - implement pufferfish snakeyaml
+    
+    implementation("org.yaml:snakeyaml:1.32")
+    implementation ("com.github.carleslc.Simple-YAML:Simple-Yaml:1.8.4") {
+        exclude(group="org.yaml", module="snakeyaml")
+    }
+    // Canvas end
 }
 
 val craftbukkitPackageVersion = "1_20_R3" // Paper
