From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris Bauer <jedimastertoothless@hotmail.com>
Date: Fri, 27 Oct 2023 08:37:03 +0000
Subject: [PATCH] Add argument bypass for launcher check


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index aec6c1940ae3b2149ce741977a9652dd42993dd6..ab178c59b5154e95f4169325621c4d2151d28fb3 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -707,7 +707,7 @@ public final class CraftServer implements Server {
         return this.chatChannelManager;
     }
 
-    @Override
+    @NotNull
     public net.minecraft.server.gui.MinecraftServerGui getGuiHandle(){
         return this.console.getGui();
     }
@@ -739,7 +739,7 @@ public final class CraftServer implements Server {
 
     @Override
     public String[] getStartupArgs(){
-        return CraftServer.startupArgs != null ? CraftServer.startupArgs : null};
+        return CraftServer.startupArgs != null ? CraftServer.startupArgs : null;
     }
     // Canvas end
 
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 609d34ae03dd32cf8ea574d323694f5dcf84506a..ec342afe1ddf702f31534b964493878505c9089a 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -27,11 +27,22 @@ public class Main {
     static {
         System.setProperty("java.util.logging.manager", "io.papermc.paper.log.CustomLogManager");
     }
-    // Paper end
+    // Paper end -/- Canvas start
+    private static boolean checkBypassLauncher(String[] args){
+        Logger.getLogger("Canvas").info("Loading Canvas with args:" + Arrays.toString(args));
+        for (String element : args) {
+            if (element.equals("-bypass.launcher.check") || element.equals("-DPaper.isRunDev=true")) {
+                return true;
+            }
+        }
+        return false;
+    }
+    // Canvas end
 
     public static void main(String[] args) {
         // Canvas start - ensure loading canvas correctly
-        if(!CanvasConfig.bypassStartupCheck){
+        boolean checker = checkBypassLauncher(args);
+        if(!CanvasConfig.bypassStartupCheck || checker){
             try {
                 Class.forName("space.vectrix.ignite.applaunch.IgniteBootstrap");
             } catch (ClassNotFoundException e) {
@@ -206,6 +217,7 @@ public class Main {
                     .ofType(File.class)
                     .defaultsTo(new File("canvas.yml"))
                     .describedAs("Yml file");
+                acceptsAll(asList("bypass.launcher.check"), "Disables checking for correct Canvas Launcher loading");
                 // Canvas end
 
                 // Paper start
