From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Wed, 15 Nov 2023 21:54:36 +0000
Subject: [PATCH] Fix broken error on first server creation


diff --git a/src/main/java/me/dueris/canvas/plugin/CanvasPluginLoader.java b/src/main/java/me/dueris/canvas/plugin/CanvasPluginLoader.java
index 59b910cd5ad3d49f71005dd355ad3bd8efc6a36c..70ebb9fd6ecfd72ff35d334a96c3d17fe1209a99 100644
--- a/src/main/java/me/dueris/canvas/plugin/CanvasPluginLoader.java
+++ b/src/main/java/me/dueris/canvas/plugin/CanvasPluginLoader.java
@@ -35,6 +35,7 @@ public class CanvasPluginLoader {
         pluginLoader.clearCache();
         // System.out.println(Paths.get("plugins"));
             File dir = Paths.get("plugins").toFile();
+            dir.mkdirs();
 
             for(File file : dir.listFiles()){
                 if(file.getName().endsWith(".jar")){
