From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Mon, 18 Mar 2024 16:34:33 -0700
Subject: [PATCH] Set ThreadPriority of startup thread


diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
index cec369b7a6fd24c7003473c7eff1af8c5c330155..4338c7ed1a2cf49676ea1bc370d74ce113323731 100644
--- a/src/main/java/net/minecraft/server/Main.java
+++ b/src/main/java/net/minecraft/server/Main.java
@@ -102,6 +102,7 @@ public class Main {
         // me.jellysquid.mods.hydrogen.common.jvm.ClassConstructors.init(); // JettPack // Canvas
         SharedConstants.tryDetectVersion();
         Main.optionSet = optionset; // Canvas
+        if(Thread.currentThread().getPriority() < 6){ Thread.currentThread().setPriority(6); } // Canvas - improve startup times (smoothboot)
         /* CraftBukkit start - Replace everything
         OptionParser optionparser = new OptionParser();
         OptionSpec<Void> optionspec = optionparser.accepts("nogui");
