From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Fri, 10 Nov 2023 20:39:44 -0800
Subject: [PATCH] Allow to configure orbs can merge to API


diff --git a/src/main/java/net/minecraft/world/entity/ExperienceOrb.java b/src/main/java/net/minecraft/world/entity/ExperienceOrb.java
index 2bcbfe5516e30a150aa133cf8c1561b784443778..932f0e967a4f2baccd10998706f13b38c9cfacdc 100644
--- a/src/main/java/net/minecraft/world/entity/ExperienceOrb.java
+++ b/src/main/java/net/minecraft/world/entity/ExperienceOrb.java
@@ -207,7 +207,9 @@ public class ExperienceOrb extends Entity {
             while (iterator.hasNext()) {
                 ExperienceOrb entityexperienceorb = (ExperienceOrb) iterator.next();
 
-                this.merge(entityexperienceorb);
+                if(this.allowMerge){
+                    this.merge(entityexperienceorb);
+                }
             }
         }
 
@@ -250,20 +252,25 @@ public class ExperienceOrb extends Entity {
             return false;
         }
     }
+    // Canvas start
+    public boolean allowMerge = true;
 
     private boolean canMerge(ExperienceOrb other) {
-        return other != this && ExperienceOrb.canMerge(other, this.getId(), this.value);
+        return other != this && ExperienceOrb.canMerge(other, this.getId(), this.value) && this.allowMerge;
     }
 
     private static boolean canMerge(ExperienceOrb orb, int seed, int amount) {
-        return !orb.isRemoved() && (orb.getId() - seed) % 40 == 0 && orb.value == amount;
+        return !orb.isRemoved() && (orb.getId() - seed) % 40 == 0 && orb.value == amount && orb.allowMerge;
     }
 
     private void merge(ExperienceOrb other) {
-        this.count += other.count;
-        this.age = Math.min(this.age, other.age);
-        other.discard();
+        if(this.allowMerge && other.allowMerge) {
+            this.count += other.count;
+            this.age = Math.min(this.age, other.age);
+            other.discard();
+        }
     }
+    // Canvas end
 
     private void setUnderwaterMovement() {
         Vec3 vec3d = this.getDeltaMovement();
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftExperienceOrb.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftExperienceOrb.java
index 5a7d314ec0562e472f5dc45924a7b24841cff126..6696e68072a0ac1291a5cbb5fc904d113198c282 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftExperienceOrb.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftExperienceOrb.java
@@ -39,4 +39,16 @@ public class CraftExperienceOrb extends CraftEntity implements ExperienceOrb {
     public String toString() {
         return "CraftExperienceOrb";
     }
+    // Canvas start - allow configure can merge for xp orbs
+
+    @Override
+    public void setAllowMerge(final boolean allowMerge) {
+        this.getHandle().allowMerge = allowMerge;
+    }
+
+    @Override
+    public boolean allowMerge() {
+        return this.getHandle().allowMerge;
+    }
+    // Canvas end
 }
