From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <122416109+Dueris@users.noreply.github.com>
Date: Thu, 1 Feb 2024 03:51:59 +0000
Subject: [PATCH] Fix MixinPlugin ClassLoading


diff --git a/src/main/java/io/papermc/paper/plugin/entrypoint/strategy/modern/ModernPluginLoadingStrategy.java b/src/main/java/io/papermc/paper/plugin/entrypoint/strategy/modern/ModernPluginLoadingStrategy.java
index 9af388a8e56806ab44f8c3ef4f97086ce38ef3b4..8603391c6b96c65a8c54b7bfa8b4f14bed70cbf1 100644
--- a/src/main/java/io/papermc/paper/plugin/entrypoint/strategy/modern/ModernPluginLoadingStrategy.java
+++ b/src/main/java/io/papermc/paper/plugin/entrypoint/strategy/modern/ModernPluginLoadingStrategy.java
@@ -114,9 +114,18 @@ public class ModernPluginLoadingStrategy<T> implements ProviderLoadingStrategy<T
 
                 if (this.configuration.preloadProvider(retrievedProvider)) {
                     T instance = retrievedProvider.createInstance();
-                    if (this.configuration.load(retrievedProvider, instance)) {
-                        loadedPlugins.add(new ProviderPair<>(retrievedProvider, instance));
+                    if(instance != null){ // Canvas
+                        if (this.configuration.load(retrievedProvider, instance)) {
+                            loadedPlugins.add(new ProviderPair<>(retrievedProvider, instance));
+                        }
+                    // Canvas start
+                    }else{
+                        // MixinPlugin instance. We arent loading it the Paper way
+                        String name = this.getClass().getName();
+                        java.util.logging.Logger.getLogger("CanvasPluginLoader").log(java.util.logging.Level.WARNING, "MixinPlugin was found by Canvas by the name of [%name%].".replace("%name%", retrievedProvider.file().getName()));
+                        java.util.logging.Logger.getLogger("CanvasPluginLoader").log(java.util.logging.Level.WARNING, "Plugin startup for MixinPlugins are different than normal, see the PluginTemplate wiki for more info");
                     }
+                    // Canvas end
                 }
             } catch (Throwable ex) {
                 LOGGER.error("Could not load plugin '%s' in folder '%s'".formatted(retrievedProvider.getFileName(), retrievedProvider.getParentSource()), ex); // Paper
diff --git a/src/main/java/io/papermc/paper/plugin/manager/PaperPluginInstanceManager.java b/src/main/java/io/papermc/paper/plugin/manager/PaperPluginInstanceManager.java
index 04efba1517e2bde8ecced31ab2eb669b2e84a28c..cf672674941809732fa0e2b34871e80576547952 100644
--- a/src/main/java/io/papermc/paper/plugin/manager/PaperPluginInstanceManager.java
+++ b/src/main/java/io/papermc/paper/plugin/manager/PaperPluginInstanceManager.java
@@ -87,6 +87,7 @@ class PaperPluginInstanceManager {
     }
 
     public void loadPlugin(Plugin provided) {
+        if(provided.getClass().getClassLoader().getClass().getName().toString().equalsIgnoreCase("space.vectrix.ignite.launch.ember.EmberClassLoader")) return; // Canvas
         PluginMeta configuration = provided.getPluginMeta();
 
         this.plugins.add(provided);
@@ -144,6 +145,7 @@ class PaperPluginInstanceManager {
     // Plugins are disabled in order like this inorder to "rougly" prevent
     // their dependencies unloading first. But, eh.
     public void disablePlugins() {
+        me.dueris.canvas.server.MixinPluginHook.pluginDisableHook(org.bukkit.craftbukkit.CraftServer.startupArgs);
         Plugin[] plugins = this.getPlugins();
         for (int i = plugins.length - 1; i >= 0; i--) {
             this.disablePlugin(plugins[i]);
diff --git a/src/main/java/io/papermc/paper/plugin/provider/type/paper/PaperPluginParent.java b/src/main/java/io/papermc/paper/plugin/provider/type/paper/PaperPluginParent.java
index f2bc4d0b55d4c9877a442529e0b144656497dae6..c959d2a18150c1d4ec3d73eaedde7af2a0102e78 100644
--- a/src/main/java/io/papermc/paper/plugin/provider/type/paper/PaperPluginParent.java
+++ b/src/main/java/io/papermc/paper/plugin/provider/type/paper/PaperPluginParent.java
@@ -176,6 +176,11 @@ public class PaperPluginParent {
                 }
 
                 this.status = ProviderStatus.INITIALIZED;
+                // Canvas start
+                if(plugin.getClass().getClassLoader().getClass().getName().toString().equalsIgnoreCase("space.vectrix.ignite.launch.ember.EmberClassLoader")){
+                    return null;
+                }
+                // Canvas end
                 return plugin;
             } catch (Throwable throwable) {
                 this.status = ProviderStatus.ERRORED;
diff --git a/src/main/java/me/dueris/canvas/server/MixinPluginHook.java b/src/main/java/me/dueris/canvas/server/MixinPluginHook.java
new file mode 100644
index 0000000000000000000000000000000000000000..c3076835e63139efaebb63331694bbf42d7284b5
--- /dev/null
+++ b/src/main/java/me/dueris/canvas/server/MixinPluginHook.java
@@ -0,0 +1,20 @@
+package me.dueris.canvas.server;
+
+public class MixinPluginHook {
+
+    public static void preLoadHook(String[] args) {
+        // Preload for MixinPlugin instance registry
+    }
+
+    public static void preWorldLoadHook(String[] args) {
+        // PreWorldLoad for MixinPlugin instance registry
+    }
+
+    public static void postWorldLoadHook(String[] args) {
+        // PostWorldLoad for MixinPlugin instance registry
+    }
+
+    public static void pluginDisableHook(String[] args) {
+        // Plugin disable hook for MixinPlugin instance registry
+    }
+}
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index ca1db41b6380a5f52761d18eaf91f05e6d3169be..92a944ee9d5f31e70de39d420f7e52070902d6c0 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -661,6 +661,13 @@ public final class CraftServer implements Server {
                 this.enablePlugin(plugin);
             }
         }
+        // Canvas start
+        if(type == PluginLoadOrder.POSTWORLD){
+            me.dueris.canvas.server.MixinPluginHook.postWorldLoadHook(org.bukkit.craftbukkit.CraftServer.startupArgs);
+        }else if(type == PluginLoadOrder.STARTUP){
+            me.dueris.canvas.server.MixinPluginHook.preWorldLoadHook(org.bukkit.craftbukkit.CraftServer.startupArgs);
+        }
+        // Canvas end
 
         if (type == PluginLoadOrder.POSTWORLD) {
             // Spigot start - Allow vanilla commands to be forced to be the main command
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index d19133ab70b8dfcab07d4c09036ad043f66a8390..2cb178f19df0ca616a61aa2b3953dfdd0caf50a3 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -60,6 +60,7 @@ public class Main {
                 Logger.getLogger("Canvas").info("Unable to locate CodeSource, Canvas may be unable to function in some areas");
             }
         }
+        me.dueris.canvas.server.MixinPluginHook.preLoadHook(args); // MixinPlugin injection point
         // Canvas end -/- Paper start
         final String warnWhenLegacyFormattingDetected = String.join(".", "net", "kyori", "adventure", "text", "warnWhenLegacyFormattingDetected");
         if (false && System.getProperty(warnWhenLegacyFormattingDetected) == null) {
