From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Mon, 18 Mar 2024 16:35:34 -0700
Subject: [PATCH] Dont spawn Air ItemEntities


diff --git a/src/main/java/net/minecraft/server/level/ServerLevel.java b/src/main/java/net/minecraft/server/level/ServerLevel.java
index eea94fe0fd43b9485cf8b06d9a9d88ce4ba02261..90f8fdad3f32455ff753f7c7e855d42c087f7b2a 100644
--- a/src/main/java/net/minecraft/server/level/ServerLevel.java
+++ b/src/main/java/net/minecraft/server/level/ServerLevel.java
@@ -1817,6 +1817,15 @@ public class ServerLevel extends Level implements WorldGenLevel {
     // CraftBukkit start
     private boolean addEntity(Entity entity, CreatureSpawnEvent.SpawnReason spawnReason) {
         org.spigotmc.AsyncCatcher.catchOp("entity add"); // Spigot
+        // Canvas start - dont spawn item entities that are AIR items
+        boolean pass = true;
+        if(entity instanceof net.minecraft.world.entity.item.ItemEntity stack){
+            if (stack.getItem().isEmpty() || stack.getItem() == null || stack.getItem().is(net.minecraft.world.item.Items.AIR) || stack.getItem().asBukkitCopy().getType().equals(org.bukkit.Material.AIR)) {
+                pass = false;
+            }
+        }
+        if(!pass) return false;
+        // Canvas end
         entity.generation = false; // Paper - Don't fire sync event during generation; Reset flag if it was added during a ServerLevel generation process
         // Paper start - extra debug info
         if (entity.valid) {
