From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Mon, 20 Nov 2023 19:03:35 +0000
Subject: [PATCH] Add PlayerDisableShieldEvent


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 4cd021b04747fba7f75e2b85afbd6e24eefb0715..e5ec3322d1e35e28167323fca7e597a7571d32d6 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -1020,7 +1020,7 @@ public abstract class Player extends LivingEntity {
     protected void blockUsingShield(LivingEntity attacker) {
         super.blockUsingShield(attacker);
         if (attacker.canDisableShield()) {
-            this.disableShield(true);
+            this.disableShield(true, attacker);
         }
 
     }
@@ -1503,7 +1503,7 @@ public abstract class Player extends LivingEntity {
         this.attack(target);
     }
 
-    public void disableShield(boolean sprinting) {
+    public void disableShield(boolean sprinting, LivingEntity attacker) {
         float f = 0.25F + (float) EnchantmentHelper.getBlockEfficiency(this) * 0.05F;
 
         if (sprinting) {
@@ -1511,9 +1511,15 @@ public abstract class Player extends LivingEntity {
         }
 
         if (this.random.nextFloat() < f) {
-            this.getCooldowns().addCooldown(Items.SHIELD, 100);
-            this.stopUsingItem();
-            this.level().broadcastEntityEvent(this, (byte) 30);
+            // Canvas start - PlayerDisableShieldEvent
+            me.dueris.canvas.event.player.PlayerDisableShieldEvent event = new me.dueris.canvas.event.player.PlayerDisableShieldEvent((org.bukkit.entity.Player) this.getBukkitEntity(), 100, (org.bukkit.entity.LivingEntity) attacker.getBukkitEntity());
+            event.callEvent();
+            if(!event.getCanceled()){
+                this.getCooldowns().addCooldown(Items.SHIELD, event.getCooldown());
+                this.stopUsingItem();
+                this.level().broadcastEntityEvent(this, (byte) 30);
+            }
+            // Canvas end
         }
 
     }
