From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <122416109+Dueris@users.noreply.github.com>
Date: Fri, 2 Feb 2024 19:13:20 +0000
Subject: [PATCH] Support Ignite 1.0.0


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java
index 8518e7245b1da8d57566df8d292b63200e8a5b12..e93a6e510210aead22b3660a442f1fe1b890927d 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java
@@ -134,7 +134,7 @@ public class DedicatedServerProperties extends Settings<DedicatedServerPropertie
     }
 
     public File getFile(){
-        return new File(org.bukkit.craftbukkit.Main.jarLocation.getParent() + File.separator + "server.properties");
+        return java.nio.file.Paths.get("server.properties").toFile();
     }
 
     // Canvas end -/- CraftBukkit start
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 2cb178f19df0ca616a61aa2b3953dfdd0caf50a3..4e78897589ade5d83ca0d88622c34e9726221e36 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -29,37 +29,37 @@ public class Main {
         System.setProperty("java.util.logging.manager", "io.papermc.paper.log.CustomLogManager");
     }
     // Paper end -/- Canvas start
-    private static boolean checkBypassLauncher(String[] args){
-        Logger.getLogger("Canvas").info("Loading Canvas with args:" + Arrays.toString(args));
-        for (String element : args) {
-            if (element.equals("-bypass.launcher.check") || element.equals("-DPaper.isRunDev=true")) {
-                return true;
-            }
-        }
-        return false;
-    }
-
-    public static File jarLocation;
-    private static void findMainInstance(){
-        try {
-            jarLocation = new File(org.bukkit.craftbukkit.Main.class.getProtectionDomain().getCodeSource().getLocation().toURI());
-        } catch (URISyntaxException e) {
-            throw new RuntimeException(e);
-        }
-    }
+    // private static boolean checkBypassLauncher(String[] args){
+    //     Logger.getLogger("Canvas").info("Loading Canvas with args:" + Arrays.toString(args));
+    //     for (String element : args) {
+    //         if (element.equals("-bypass.launcher.check") || element.equals("-DPaper.isRunDev=true")) {
+    //             return true;
+    //         }
+    //     }
+    //     return false;
+    // }
+
+    // public static File jarLocation;
+    // private static void findMainInstance(){
+    //     try {
+    //         jarLocation = new File(org.bukkit.craftbukkit.Main.class.getProtectionDomain().getCodeSource().getLocation().toURI());
+    //     } catch (URISyntaxException e) {
+    //         throw new RuntimeException(e);
+    //     }
+    // }
     // Canvas end
 
     public static void main(String[] args) {
         // Canvas start - ensure loading canvas correctly
-        findMainInstance(); // Load the jarLocation instance
-        boolean checker = checkBypassLauncher(args);
-        if(!CanvasConfig.bypassStartupCheck || checker){
-            try {
-                Class.forName("space.vectrix.ignite.applaunch.IgniteBootstrap");
-            } catch (ClassNotFoundException e) {
-                Logger.getLogger("Canvas").info("Unable to locate CodeSource, Canvas may be unable to function in some areas");
-            }
-        }
+        // findMainInstance(); // Load the jarLocation instance
+        // boolean checker = checkBypassLauncher(args);
+        // if(!CanvasConfig.bypassStartupCheck || checker){
+        //     try {
+        //         Class.forName("space.vectrix.ignite.applaunch.IgniteBootstrap");
+        //     } catch (ClassNotFoundException e) {
+        //         Logger.getLogger("Canvas").info("Unable to locate CodeSource, Canvas may be unable to function in some areas");
+        //     }
+        // }
         me.dueris.canvas.server.MixinPluginHook.preLoadHook(args); // MixinPlugin injection point
         // Canvas end -/- Paper start
         final String warnWhenLegacyFormattingDetected = String.join(".", "net", "kyori", "adventure", "text", "warnWhenLegacyFormattingDetected");
