From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <122416109+Dueris@users.noreply.github.com>
Date: Mon, 1 Apr 2024 20:53:00 +0000
Subject: [PATCH] ThreadRipper


diff --git a/src/main/java/me/dueris/canvas/CanvasConfig.java b/src/main/java/me/dueris/canvas/CanvasConfig.java
index dec051ad7b11f2e131993a1f53f24846edc3a52e..f427aed8c26e021c29c7c5c2677a8a0249d68932 100644
--- a/src/main/java/me/dueris/canvas/CanvasConfig.java
+++ b/src/main/java/me/dueris/canvas/CanvasConfig.java
@@ -190,6 +190,7 @@ public class CanvasConfig {
     public static int mobWanderingPercentChance = 100;
     public static int mobWanderingReducedRateMultiplier = 6;
     public static int mobWanderingDelay = 20 * 5;
+    public static boolean threadripper = false;
 
     private static void mixin(){
         disableMixin = getBoolean("mixin.disableMixins", disableMixin);
@@ -227,6 +228,7 @@ public class CanvasConfig {
         mobWanderingDelay = getInt("optimizations.mobWanderingDelay", mobWanderingDelay);
         mobWanderingPercentChance = getInt("optimizations.mobWanderingPercentChance", mobWanderingPercentChance);
         mobWanderingReducedRateMultiplier = getInt("optimizations.mobWanderingReducedRateMultiplier", mobWanderingReducedRateMultiplier);
+        threadripper = getBoolean("optimizations.threadripper", threadripper);
     }
     private static void blocks(){
         useCanvasBedSetDye = getBoolean("blocks.useCanvasBedSetDyeMethod", useCanvasBedSetDye);
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 018a23162df163895d193848072e5cf72d18e01c..efdac022379c7a3360128cb24f39033777f978b1 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -216,6 +216,14 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
             DedicatedServer.LOGGER.error("Unable to load canvas-server configuration", e);
             return false;
         }
+        // Check threadripper
+
+        if(CanvasConfig.threadripper){
+            DedicatedServer.LOGGER.info("------------------------------------------------------------------------------------------------------------------");
+            DedicatedServer.LOGGER.info("You have enabled threadripper settings, this requires many threads/cores to truely have an impact on your server.");
+            DedicatedServer.LOGGER.info("There are a wide array of issues that may occur. Proceed with caution.");
+            DedicatedServer.LOGGER.info("------------------------------------------------------------------------------------------------------------------");
+        }
         // Canvas end
         // Paper start
         io.papermc.paper.util.ObfHelper.INSTANCE.getClass(); // Paper - load mappings for stacktrace deobf and etc.
