From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Tue, 21 Nov 2023 19:14:51 +0000
Subject: [PATCH] Add getReloadCount() to Server


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 0ecb5747e0f6bed57c64d20fd88d0e2874767ee4..6dfe8b111e26267d7e021224daec7913c56a9077 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1310,6 +1310,13 @@ public final class CraftServer implements Server {
 
         return false;
     }
+    // Canvas start
+
+    @Override
+    public int getReloadCount(){
+        return this.reloadCount;
+    }
+    // Canvas end
 
     @Override
     public void reload() {
@@ -1405,6 +1412,7 @@ public final class CraftServer implements Server {
         }
         io.papermc.paper.plugin.PluginInitializerManager.reload(this.console); // Paper
         me.dueris.canvas.plugin.CanvasPluginLoader.loader(); // Canvas - reload cache mappings
+        
         this.loadPlugins();
         this.enablePlugins(PluginLoadOrder.STARTUP);
         this.enablePlugins(PluginLoadOrder.POSTWORLD);
