From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: etil2jz <81570777+etil2jz@users.noreply.github.com>
Date: Sun, 8 May 2022 16:34:27 +0200
Subject: [PATCH] Fix brewing stands resetting their brewTime when being
 unloaded


diff --git a/src/main/java/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
index c57efcb9a79337ec791e4e8f6671612f0a82b441..111b9bfac0aafa0170c6e273662c4635963a0d34 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
@@ -301,6 +301,7 @@ public class BrewingStandBlockEntity extends BaseContainerBlockEntity implements
         ContainerHelper.loadAllItems(nbt, this.items);
         this.brewTime = nbt.getShort("BrewTime");
         this.fuel = nbt.getByte("Fuel");
+        this.ingredient = net.minecraft.core.registries.BuiltInRegistries.ITEM.get(new net.minecraft.resources.ResourceLocation(nbt.getString("Mirai.ingredient"))); // Mirai - fix brewing stands resetting their brewTime when being unloaded
     }
 
     @Override
@@ -309,6 +310,7 @@ public class BrewingStandBlockEntity extends BaseContainerBlockEntity implements
         nbt.putShort("BrewTime", (short) this.brewTime);
         ContainerHelper.saveAllItems(nbt, this.items);
         nbt.putByte("Fuel", (byte) this.fuel);
+        nbt.putString("Mirai.ingredient", net.minecraft.core.registries.BuiltInRegistries.ITEM.getKey(this.ingredient).toString()); // Mirai - fix brewing stands resetting their brewTime when being unloaded
     }
 
     @Override
