From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Wed, 22 Nov 2023 08:39:46 +0000
Subject: [PATCH] Add ModelPart API


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 49ab61d969a9f50543fefd488291f9eba79e2e79..f7f719f8a0a62bc7c264f783e3c8341146fcfb52 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -443,6 +443,16 @@ public class ServerPlayer extends Player {
         this.bukkitPickUpLoot = true;
         this.maxHealthCache = this.getMaxHealth();
     }
+    // Canvas start - ModelPart api
+
+    public me.dueris.canvas.entity.player.PlayerModelPart minecraftToCanvasModelPart(net.minecraft.world.entity.player.PlayerModelPart modelPart){
+        return me.dueris.canvas.entity.player.PlayerModelPart.valueOf(modelPart.toString());
+    }
+
+    public net.minecraft.world.entity.player.PlayerModelPart canvasToMinecraftModelPart(me.dueris.canvas.entity.player.PlayerModelPart modelPart){
+        return net.minecraft.world.entity.player.PlayerModelPart.valueOf(modelPart.toString());
+    }
+    // Canvas end
 
     // Yes, this doesn't match Vanilla, but it's the best we can do for now.
     // If this is an issue, PRs are welcome
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 8660fb6fb0c3d5297da4ce6daeb95753d72dc5da..bbb266a8a6d8c6f2947eb7983603e6f02f6e550d 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -3247,6 +3247,13 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         }
         return (int) (duration.toMillis() / 50L);
     }
+    // Canvas start - PlayerModelPart api
+
+    @Override
+    public boolean hasModelPartShown(me.dueris.canvas.entity.player.PlayerModelPart modelPart){
+        return this.getHandle().isModelPartShown(this.getHandle().canvasToMinecraftModelPart(modelPart));
+    }
+    // Canvas end
 
     @Override
     public void clearTitle() {
