From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sat, 6 Apr 2024 17:51:58 -0700
Subject: [PATCH] Configurable MobSpawners


diff --git a/src/main/java/me/dueris/canvas/CanvasConfig.java b/src/main/java/me/dueris/canvas/CanvasConfig.java
index d39c17d027030425df58b48e884074309280d50d..aba4ffe072bc191f2384da3a40f06b4f3bf6455e 100644
--- a/src/main/java/me/dueris/canvas/CanvasConfig.java
+++ b/src/main/java/me/dueris/canvas/CanvasConfig.java
@@ -199,6 +199,14 @@ public class CanvasConfig {
     public static boolean disableFootsteps = false;
     public static boolean disableRaidsFromSpawnerRaiders = false;
     public static int maxConnectionThreads = 5;
+    public static int spawnerMinSpawnDelay = 200;
+    public static int spawnerMaxSpawnDelay = 800;
+    public static int spawnerSpawnRange = 4;
+    public static int spawnerSpawnCount = 4;
+    public static boolean spawnerSpawnParticles = true;
+    public static int spawnerMaxNearbyEntities = 16;
+    public static int spawnerMaxPlayerRange = 16;
+    public static boolean spawnerEntitiesHaveAI = true;
 
     private static void logger(){
         obfuscateIps = getBoolean("logger.obfuscateIps", obfuscateIps);
@@ -247,5 +255,13 @@ public class CanvasConfig {
         makeCampfireAlwaysLots = getBoolean("block.forceCampfireEmitLotsOfParticles", makeCampfireAlwaysLots);
         farmlandSearchRadius = getInt("block.farmlandSearchRadius", farmlandSearchRadius);
         lavaCatchesBlocksOnFire = getBoolean("block.lavaCatchesBlocksOnFire", lavaCatchesBlocksOnFire);
+        spawnerMaxNearbyEntities = getInt("block.spawner.spawnerMaxNearbyEntities", spawnerMaxNearbyEntities);
+        spawnerMaxPlayerRange = getInt("block.spawner.spawnerMaxPlayerRange", spawnerMaxPlayerRange);
+        spawnerEntitiesHaveAI = getBoolean("block.spawner.spawnerEntitiesHaveAI", spawnerEntitiesHaveAI);
+        spawnerSpawnRange = getInt("block.spawner.spawnerSpawnRange", spawnerSpawnRange);
+        spawnerSpawnCount = getInt("block.spawner.spawnerSpawnCount", spawnerSpawnCount);
+        spawnerMaxSpawnDelay = getInt("block.spawner.spawnerMaxSpawnDelay", spawnerMaxSpawnDelay);
+        spawnerMinSpawnDelay = getInt("block.spawner.spawnerMinSpawnDelay", spawnerMinSpawnDelay);
+        spawnerSpawnParticles = getBoolean("block.spawner.spawnerSpawnParticles", spawnerSpawnParticles);
     }
 }
diff --git a/src/main/java/net/minecraft/world/level/BaseSpawner.java b/src/main/java/net/minecraft/world/level/BaseSpawner.java
index 40c199812ecf7b16fe5a17c18cb0d6d3ce258910..bd791b502a6a618a071c2b7e936f3cb84024690d 100644
--- a/src/main/java/net/minecraft/world/level/BaseSpawner.java
+++ b/src/main/java/net/minecraft/world/level/BaseSpawner.java
@@ -39,14 +39,16 @@ public abstract class BaseSpawner {
     public SpawnData nextSpawnData;
     private double spin;
     private double oSpin;
-    public int minSpawnDelay = 200;
-    public int maxSpawnDelay = 800;
-    public int spawnCount = 4;
+    // Canvas start - Configurable MobSpawners
+    public int minSpawnDelay = me.dueris.canvas.CanvasConfig.spawnerMinSpawnDelay == -1 ? 200 : me.dueris.canvas.CanvasConfig.spawnerMinSpawnDelay;
+    public int maxSpawnDelay = me.dueris.canvas.CanvasConfig.spawnerMaxSpawnDelay == -1 ? 800 : me.dueris.canvas.CanvasConfig.spawnerMaxSpawnDelay;
+    public int spawnCount = me.dueris.canvas.CanvasConfig.spawnerSpawnCount == -1 ? 4 : me.dueris.canvas.CanvasConfig.spawnerSpawnCount;
     @Nullable
     private Entity displayEntity;
-    public int maxNearbyEntities = 6;
-    public int requiredPlayerRange = 16;
-    public int spawnRange = 4;
+    public int maxNearbyEntities = me.dueris.canvas.CanvasConfig.spawnerMaxNearbyEntities == -1 ? 6 : me.dueris.canvas.CanvasConfig.spawnerMaxNearbyEntities;
+    public int requiredPlayerRange = me.dueris.canvas.CanvasConfig.spawnerMaxPlayerRange == -1 ? 16 : me.dueris.canvas.CanvasConfig.spawnerMaxPlayerRange;
+    // Canvas end
+    public int spawnRange = me.dueris.canvas.CanvasConfig.spawnerSpawnRange == -1 ? 4 : me.dueris.canvas.CanvasConfig.spawnerSpawnRange;
     private int tickDelay = 0; // Paper - Configurable mob spawner tick rate
 
     public BaseSpawner() {}
@@ -70,8 +72,10 @@ public abstract class BaseSpawner {
             double d1 = (double) pos.getY() + randomsource.nextDouble();
             double d2 = (double) pos.getZ() + randomsource.nextDouble();
 
+            if(me.dueris.canvas.CanvasConfig.spawnerSpawnParticles) { // Canvas - Configurable MobSpawners
             world.addParticle(ParticleTypes.SMOKE, d0, d1, d2, 0.0D, 0.0D, 0.0D);
             world.addParticle(ParticleTypes.FLAME, d0, d1, d2, 0.0D, 0.0D, 0.0D);
+            } // Canvas
             if (this.spawnDelay > 0) {
                 --this.spawnDelay;
             }
@@ -168,6 +172,7 @@ public abstract class BaseSpawner {
                         entity.moveTo(entity.getX(), entity.getY(), entity.getZ(), randomsource.nextFloat() * 360.0F, 0.0F);
                         if (entity instanceof Mob) {
                             Mob entityinsentient = (Mob) entity;
+                            if(!me.dueris.canvas.CanvasConfig.spawnerEntitiesHaveAI) entityinsentient.setNoAi(false); // Canvas - Configurable MobSpawners
 
                             if (mobspawnerdata.getCustomSpawnRules().isEmpty() && !entityinsentient.checkSpawnRules(world, MobSpawnType.SPAWNER) || !entityinsentient.checkSpawnObstruction(world)) {
                                 continue;
