From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Wed, 22 Nov 2023 09:00:21 +0000
Subject: [PATCH] Add createEndPlatformAPI


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index f7f719f8a0a62bc7c264f783e3c8341146fcfb52..9294164f1239318651a5e442c66661c5d1f57b90 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -1273,7 +1273,7 @@ public class ServerPlayer extends Player {
                 if (resourcekey == LevelStem.OVERWORLD && worldserver.getTypeKey() == LevelStem.NETHER) { // CraftBukkit
                     this.enteredNetherPosition = this.position();
                 } else if (worldserver.getTypeKey() == LevelStem.END && shapedetectorshape.portalEventInfo != null && shapedetectorshape.portalEventInfo.getCanCreatePortal()) { // CraftBukkit
-                    this.createEndPlatform(worldserver, BlockPos.containing(shapedetectorshape.pos));
+                    this.createEndPlatform(worldserver, BlockPos.containing(shapedetectorshape.pos), true); // Canvas
                 }
                 // CraftBukkit start
             } else {
@@ -1356,7 +1356,7 @@ public class ServerPlayer extends Player {
     }
     // CraftBukkit end
 
-    private void createEndPlatform(ServerLevel world, BlockPos centerPos) {
+    public void createEndPlatform(ServerLevel world, BlockPos centerPos, boolean callEvent) { // Canvas - private -> public
         BlockPos.MutableBlockPos blockposition_mutableblockposition = centerPos.mutable();
         org.bukkit.craftbukkit.util.BlockStateListPopulator blockList = new org.bukkit.craftbukkit.util.BlockStateListPopulator(world); // CraftBukkit
 
@@ -1369,13 +1369,17 @@ public class ServerPlayer extends Player {
                 }
             }
         }
-        // CraftBukkit start - call portal event
-        org.bukkit.event.world.PortalCreateEvent portalEvent = new org.bukkit.event.world.PortalCreateEvent((List<org.bukkit.block.BlockState>) (List) blockList.getList(), world.getWorld(), this.getBukkitEntity(), org.bukkit.event.world.PortalCreateEvent.CreateReason.END_PLATFORM);
-        world.getCraftServer().getPluginManager().callEvent(portalEvent);
-        if (!portalEvent.isCancelled()) {
-            blockList.updateList();
+        // Canvas start
+        if(callEvent){
+            // CraftBukkit start - call portal event
+            org.bukkit.event.world.PortalCreateEvent portalEvent = new org.bukkit.event.world.PortalCreateEvent((List<org.bukkit.block.BlockState>) (List) blockList.getList(), world.getWorld(), this.getBukkitEntity(), org.bukkit.event.world.PortalCreateEvent.CreateReason.END_PLATFORM);
+            world.getCraftServer().getPluginManager().callEvent(portalEvent);
+            if (!portalEvent.isCancelled()) {
+                blockList.updateList();
+            }
+            // CraftBukkit end
         }
-        // CraftBukkit end
+        // Canvas end        
 
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index bf403c9cd47a1ab6167bbc234b3b1efc41a8760f..f55e69b4477c876f6e16506aefd2b98dead6a8d7 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1337,6 +1337,24 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         this.lookAt(targetLocation, io.papermc.paper.entity.LookAnchor.EYES);
         // Paper end
     }
+    // Canvas start
+
+    @Override
+    public void createEndPlatformAtCurrentPos(){
+        this.createEndPlatformAtCurrentPos(false);
+    }
+
+    @Override
+    public void createEndPlatformAtCurrentPos(boolean callEvent){
+        ServerLevel level = ((CraftWorld) this.getWorld()).getHandle();
+        BlockPos pos = new BlockPos(
+            Math.toIntExact(carpetfixes.helpers.FastMath.round(this.getLocation().getX())),
+            Math.toIntExact(carpetfixes.helpers.FastMath.round(this.getLocation().getY())),
+            Math.toIntExact(carpetfixes.helpers.FastMath.round(this.getLocation().getZ()))
+        );
+        this.getHandle().createEndPlatform(level, pos, callEvent);
+    }
+    // Canvas end
 
     // Slice start
     public void teleportWithoutRespawn(Location location) {
