From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sun, 19 Nov 2023 12:25:46 -0800
Subject: [PATCH] Allow TIPPED_ARROW effect durations to be modified


diff --git a/src/main/java/net/minecraft/world/entity/projectile/Arrow.java b/src/main/java/net/minecraft/world/entity/projectile/Arrow.java
index c58bd47b045b8bfcbfef4444ba95f2e7d04dcc23..c4c3abc74767d07f4c5b2b65eb461f76db783375 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/Arrow.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/Arrow.java
@@ -199,6 +199,20 @@ public class Arrow extends AbstractArrow {
         }
 
     }
+    // Canvas start
+    protected float divisor = 8;
+
+    public float getDivisor(){
+        return divisor;
+    }
+
+    public void setDivisor(float divisor){
+        if(divisor == 0){
+            divisor = 1;
+        }
+        this.divisor = divisor;
+    }
+    // Canvas end
 
     @Override
     protected void doPostHurtEffects(LivingEntity target) {
@@ -211,7 +225,7 @@ public class Arrow extends AbstractArrow {
         while (iterator.hasNext()) {
             mobeffect = (MobEffectInstance) iterator.next();
             target.addEffect(new MobEffectInstance(mobeffect.getEffect(), Math.max(mobeffect.mapDuration((i) -> {
-                return i / 8;
+                return carpetfixes.helpers.FastMath.round(i / this.divisor); // Canvas
             }), 1), mobeffect.getAmplifier(), mobeffect.isAmbient(), mobeffect.isVisible()), entity, org.bukkit.event.entity.EntityPotionEffectEvent.Cause.ARROW); // CraftBukkit
         }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftTippedArrow.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftTippedArrow.java
index b2747e37600cadc385e2e9e1115ba1d4ec22f5da..13befbf7452c008d8b2363911f023ecebbaa2bbf 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftTippedArrow.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftTippedArrow.java
@@ -58,6 +58,18 @@ public class CraftTippedArrow extends CraftArrow implements Arrow {
         this.getHandle().effects.clear();
         this.getHandle().updateColor();
     }
+    // Canvas start
+
+    @Override
+    public void setEffectDurationDivisor(final float divisor) {
+        this.getHandle().setDivisor(divisor);
+    }
+
+    @Override
+    public float getEffectDurationDivisor() {
+        return this.getHandle().getDivisor();
+    }
+    // Canvas end
 
     @Override
     public List<PotionEffect> getCustomEffects() {
