From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sat, 28 Oct 2023 12:39:11 -0700
Subject: [PATCH] Fix ArrayIndexException for MojangGUI


diff --git a/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java b/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
index 6e8c1b91f4796e3b656549b5b05a85863ae8a90b..bf2e7de9b50bcdef872b6d05825bf64c9f15f4e6 100644
--- a/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
+++ b/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
@@ -297,8 +297,14 @@ public class MinecraftServerGui extends CraftServerGui implements MojangServerGu
         JScrollBar jscrollbar = scrollPane.getVerticalScrollBar();
 
         if (scrollPane.getViewport().getView() == textArea) {
-            boolean atBottom = (double) jscrollbar.getValue() + jscrollbar.getSize().getHeight() + (double) (MinecraftServerGui.FONT.getSize() * 4) > (double) jscrollbar.getMaximum();
-
+            // Canvas start
+            boolean atBottom = false;
+            try{
+                atBottom = (double) jscrollbar.getValue() + jscrollbar.getSize().getHeight() + (double) (MinecraftServerGui.FONT.getSize() * 4) > (double) jscrollbar.getMaximum();
+            } catch (ArrayIndexOutOfBoundsException ignored){
+                // ignore this
+            }
+            // Canvas end
         /* // Purpur
         try {
             document.insertString(document.getLength(), MinecraftServerGui.ANSI.matcher(message).replaceAll(""), (AttributeSet) null); // CraftBukkit
