From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sat, 28 Oct 2023 12:39:11 -0700
Subject: [PATCH] Fix ArrayIndexException for MojangGUI


diff --git a/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java b/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
index 6e8c1b91f4796e3b656549b5b05a85863ae8a90b..bf2e7de9b50bcdef872b6d05825bf64c9f15f4e6 100644
--- a/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
+++ b/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
@@ -297,8 +297,14 @@ public class MinecraftServerGui extends CraftServerGui implements MojangServerGu
         JScrollBar jscrollbar = scrollPane.getVerticalScrollBar();
 
         if (scrollPane.getViewport().getView() == textArea) {
-            boolean atBottom = (double) jscrollbar.getValue() + jscrollbar.getSize().getHeight() + (double) (MinecraftServerGui.FONT.getSize() * 4) > (double) jscrollbar.getMaximum();
-
+            // Canvas start
+            boolean atBottom = false;
+            try{
+                atBottom = (double) jscrollbar.getValue() + jscrollbar.getSize().getHeight() + (double) (MinecraftServerGui.FONT.getSize() * 4) > (double) jscrollbar.getMaximum();
+            } catch (ArrayIndexOutOfBoundsException ignored){
+                // ignore this
+            }
+            // Canvas end
         /* // Purpur
         try {
             document.insertString(document.getLength(), MinecraftServerGui.ANSI.matcher(message).replaceAll(""), (AttributeSet) null); // CraftBukkit

@@ -206,9 +206,11 @@ public class MinecraftServerGui extends CraftServerGui implements MojangServerGu
         jtextfield.getActionMap().put("up", new AbstractAction() {
             @Override
             public void actionPerformed(ActionEvent actionEvent) {
-                if (historyIndex < history.size() - 1) {
-                    historyIndex++;
-                    jtextfield.setText(history.get(history.size() - historyIndex - 1));
+                if(history.size() > 0){
+                    if (historyIndex < history.size() - 1) {
+                        historyIndex++;
+                        jtextfield.setText(history.get(history.size() - historyIndex - 1));
+                    }
                 }
             }
         });
@@ -216,7 +218,7 @@ public class MinecraftServerGui extends CraftServerGui implements MojangServerGu
         jtextfield.getActionMap().put("down", new AbstractAction() {
             @Override
             public void actionPerformed(ActionEvent actionEvent) {
-                if (historyIndex >= 0) {
+                if (historyIndex > 0 && history.size() > 0) {
                     historyIndex--;
                     if (historyIndex >= 0) {
                         jtextfield.setText(history.get(history.size() - historyIndex - 1));
@@ -240,17 +242,19 @@ public class MinecraftServerGui extends CraftServerGui implements MojangServerGu
             try {
                 java.util.List<String> allLines = Files.readAllLines(Path.of(path));
                 int starting = allLines.size();
-                for (String line : allLines) {
-                    print(jtextarea, jscrollpane, line);
-                }
-                while(true){
-                    if(starting < Files.readAllLines(Path.of(path)).size()){
-                        print(jtextarea, jscrollpane, Files.readAllLines(Path.of(path)).get(starting));
-                        starting++;
+                if(starting != 0){
+                    for (String line : allLines) {
+                        print(jtextarea, jscrollpane, line);
                     }
-                    if(this.isClosing.get()){
-                        print(jtextarea, jscrollpane, "Server shutting down, Mojang logger halted.");
-                        break;
+                    while(true){
+                        if(starting < Files.readAllLines(Path.of(path)).size()){
+                            print(jtextarea, jscrollpane, Files.readAllLines(Path.of(path)).get(starting));
+                            starting++;
+                        }
+                        if(this.isClosing.get()){
+                            print(jtextarea, jscrollpane, "Server shutting down, Mojang logger halted.");
+                            break;
+                        }
                     }
                 }
             } catch (IOException e) {
@@ -300,7 +304,9 @@ public class MinecraftServerGui extends CraftServerGui implements MojangServerGu
             // Canvas start
             boolean atBottom = false;
             try{
-                atBottom = (double) jscrollbar.getValue() + jscrollbar.getSize().getHeight() + (double) (MinecraftServerGui.FONT.getSize() * 4) > (double) jscrollbar.getMaximum();
+                if(jscrollbar.getMaximum() > 0){
+                    atBottom = (double) jscrollbar.getValue() + jscrollbar.getSize().getHeight() + (double) (MinecraftServerGui.FONT.getSize() * 4) > (double) jscrollbar.getMaximum();
+                }
             } catch (ArrayIndexOutOfBoundsException ignored){
                 // ignore this
             }
