From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Fri, 8 Dec 2023 07:02:49 +0000
Subject: [PATCH] Remove Old API


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 4442bfc7cd1d85964e70b22492ebd1dc6c0b0e19..cebfb17ded76353c14647fc754248c7160cac774 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -300,7 +300,7 @@ public final class CraftServer implements Server {
     public static CraftCanvasLogger.Obfuscator canvas_obfuscator = null;
     private final CraftCanvasLogger craftCanvasLogger = new CraftCanvasLogger();
     public Logger LOGGER = Logger.getLogger(CraftServer.class.getName());
-    private final ChatChannelManager chatChannelManager = new ChatChannelManager();
+    // private final ChatChannelManager chatChannelManager = new ChatChannelManager();
     public static String[] startupArgs = null;
     // Canvas end
     private final String bukkitVersion = Versioning.getBukkitVersion();
