From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sun, 12 Nov 2023 12:55:15 -0800
Subject: [PATCH] Change package from com.dueris to me.dueris


diff --git a/src/main/java/com/dueris/canvas/CanvasConfig.java b/src/main/java/io/github/dueris/CanvasConfig.java
similarity index 86%
rename from src/main/java/com/dueris/canvas/CanvasConfig.java
rename to src/main/java/io/github/dueris/CanvasConfig.java
index caae656d62922a73fd7ee7c228261fb67e2613d6..42230e9b8478705226d6cd5c5e13fb212c3ed908 100644
--- a/src/main/java/com/dueris/canvas/CanvasConfig.java
+++ b/src/main/java/io/github/dueris/CanvasConfig.java
@@ -1,42 +1,22 @@
-package com.dueris.canvas;
+package io.github.dueris;
 
 import com.google.common.base.Throwables;
 import com.google.common.collect.ImmutableMap;
-import com.mojang.datafixers.util.Pair;
-import net.kyori.adventure.bossbar.BossBar;
-import net.kyori.adventure.text.minimessage.MiniMessage;
-import net.minecraft.core.registries.BuiltInRegistries;
-import net.minecraft.resources.ResourceLocation;
 import net.minecraft.server.MinecraftServer;
-import net.minecraft.world.effect.MobEffect;
-import net.minecraft.world.effect.MobEffectInstance;
-import net.minecraft.world.entity.EntityDimensions;
-import net.minecraft.world.entity.EntityType;
-import net.minecraft.world.food.FoodProperties;
-import net.minecraft.world.food.Foods;
-import net.minecraft.world.item.enchantment.Enchantment;
-import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.Blocks;
 import org.bukkit.Bukkit;
 import org.bukkit.command.Command;
 import org.bukkit.configuration.ConfigurationSection;
 import org.bukkit.configuration.InvalidConfigurationException;
 import org.bukkit.configuration.file.YamlConfiguration;
-import org.purpurmc.purpur.command.PurpurCommand;
-import org.purpurmc.purpur.task.TPSBarTask;
 
 import java.io.File;
 import java.io.IOException;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.lang.reflect.Modifier;
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
-import java.util.Set;
 import java.util.logging.Level;
 
 @SuppressWarnings("unused")
diff --git a/src/main/java/com/dueris/canvas/logger/CraftCanvasLogger.java b/src/main/java/io/github/dueris/logger/CraftCanvasLogger.java
similarity index 98%
rename from src/main/java/com/dueris/canvas/logger/CraftCanvasLogger.java
rename to src/main/java/io/github/dueris/logger/CraftCanvasLogger.java
index 62d18b60b845a0802eade5b8e6fb6474a40fb725..c4f7cee005094b0b2fb5620bb17d0b6a7ecb6c35 100644
--- a/src/main/java/com/dueris/canvas/logger/CraftCanvasLogger.java
+++ b/src/main/java/io/github/dueris/logger/CraftCanvasLogger.java
@@ -1,4 +1,4 @@
-package com.dueris.canvas.logger;
+package io.github.dueris.logger;
 
 import org.bukkit.craftbukkit.CraftServer;
 import org.bukkit.craftbukkit.Main;
@@ -88,7 +88,7 @@ public class CraftCanvasLogger implements CanvasLogger {
 
     private LogLevel logLevel = LogLevel.INFO;
 
-    public static class Obfuscator implements com.dueris.canvas.logger.Obfuscator {
+    public static class Obfuscator implements io.github.dueris.logger.Obfuscator {
 
         String obfBuilder = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789!@#$%^&*()_+-={}[];`~ ";
 
diff --git a/src/main/java/com/dueris/canvas/network/ChatChannelManager.java b/src/main/java/io/github/dueris/network/ChatChannelManager.java
similarity index 82%
rename from src/main/java/com/dueris/canvas/network/ChatChannelManager.java
rename to src/main/java/io/github/dueris/network/ChatChannelManager.java
index 9b34071168294c7e2e81c8fb7ebe692635767437..3317c2f5166d4d7bf0aa6f55787a0195e28b6feb 100644
--- a/src/main/java/com/dueris/canvas/network/ChatChannelManager.java
+++ b/src/main/java/io/github/dueris/network/ChatChannelManager.java
@@ -1,11 +1,11 @@
-package com.dueris.canvas.network;
+package io.github.dueris.network;
 
-import com.dueris.canvas.network.channels.ChatChannel;
-import com.dueris.canvas.network.channels.CraftChatChannel;
+import io.github.dueris.network.channels.ChatChannel;
+import io.github.dueris.network.channels.CraftChatChannel;
 import org.bukkit.NamespacedKey;
 import java.util.ArrayList;
 
-public class ChatChannelManager implements ChatChannelContainer{
+public class ChatChannelManager implements ChatChannelContainer {
 
     public ArrayList<ChatChannel> craftChannels = new ArrayList<>();
 
diff --git a/src/main/java/com/dueris/canvas/network/channels/CraftChannelSettings.java b/src/main/java/io/github/dueris/network/channels/CraftChannelSettings.java
similarity index 93%
rename from src/main/java/com/dueris/canvas/network/channels/CraftChannelSettings.java
rename to src/main/java/io/github/dueris/network/channels/CraftChannelSettings.java
index bbea4aa68641121a60a594e8139cb2f533cd0cb9..ad9490d8a61dbf5b9737c03d92d62284fca2e37b 100644
--- a/src/main/java/com/dueris/canvas/network/channels/CraftChannelSettings.java
+++ b/src/main/java/io/github/dueris/network/channels/CraftChannelSettings.java
@@ -1,6 +1,6 @@
-package com.dueris.canvas.network.channels;
+package io.github.dueris.network.channels;
 
-public class CraftChannelSettings implements ChannelSettings{
+public class CraftChannelSettings implements ChannelSettings {
     CraftChatChannel channel;
     boolean viewOtherChannelLogs;
     boolean sendAdvancements;
diff --git a/src/main/java/com/dueris/canvas/network/channels/CraftChatChannel.java b/src/main/java/io/github/dueris/network/channels/CraftChatChannel.java
similarity index 97%
rename from src/main/java/com/dueris/canvas/network/channels/CraftChatChannel.java
rename to src/main/java/io/github/dueris/network/channels/CraftChatChannel.java
index abb7b9dc37d7093c8a729875f7fe52fbe0ea0355..03e8058cfcd84bef0e96b4fd1ddac8b7e9a0c3f8 100644
--- a/src/main/java/com/dueris/canvas/network/channels/CraftChatChannel.java
+++ b/src/main/java/io/github/dueris/network/channels/CraftChatChannel.java
@@ -1,4 +1,4 @@
-package com.dueris.canvas.network.channels;
+package io.github.dueris.network.channels;
 
 import org.bukkit.NamespacedKey;
 import org.bukkit.entity.Player;
diff --git a/src/main/java/com/dueris/canvas/plcaeholder b/src/main/java/io/github/dueris/plcaeholder
similarity index 100%
rename from src/main/java/com/dueris/canvas/plcaeholder
rename to src/main/java/io/github/dueris/plcaeholder
diff --git a/src/main/java/net/minecraft/network/Connection.java b/src/main/java/net/minecraft/network/Connection.java
index b458e79fb7d20ac89119bc3d7c6d87c5d8bc1675..91dcf1a0a701d83a24afc208e057f961c8ae9523 100644
--- a/src/main/java/net/minecraft/network/Connection.java
+++ b/src/main/java/net/minecraft/network/Connection.java
@@ -1,6 +1,6 @@
 package net.minecraft.network;
 
-import com.dueris.canvas.CanvasConfig;
+import io.github.dueris.CanvasConfig;
 import com.google.common.base.Suppliers;
 import com.google.common.collect.Queues;
 import com.google.common.util.concurrent.ThreadFactoryBuilder;
@@ -24,7 +24,6 @@ import io.netty.channel.epoll.EpollSocketChannel;
 import io.netty.channel.local.LocalChannel;
 import io.netty.channel.local.LocalServerChannel;
 import io.netty.channel.nio.NioEventLoopGroup;
-import io.netty.channel.socket.SocketChannel;
 import io.netty.channel.socket.nio.NioSocketChannel;
 import io.netty.handler.flow.FlowControlHandler;
 import io.netty.handler.timeout.ReadTimeoutHandler;
@@ -38,7 +37,6 @@ import java.util.concurrent.RejectedExecutionException;
 import java.util.function.Consumer;
 import java.util.function.Supplier;
 import javax.annotation.Nullable;
-import javax.crypto.Cipher;
 import net.minecraft.SharedConstants;
 import net.minecraft.Util;
 import net.minecraft.network.chat.Component;
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index e8c2602415639f17990471e65d62c3ec4ad49048..6300750ccfa6b08e59d29048880e23d3ffe1f315 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -1,21 +1,15 @@
 package net.minecraft.server.dedicated;
 
-import com.dueris.canvas.CanvasConfig;
-import com.google.common.collect.Lists;
+import io.github.dueris.CanvasConfig;
 import com.mojang.authlib.GameProfile;
 import com.mojang.datafixers.DataFixer;
 import com.mojang.logging.LogUtils;
-import java.io.BufferedReader;
 import java.io.BufferedWriter;
 import java.io.IOException;
-import java.io.InputStreamReader;
 import java.net.InetAddress;
 import java.net.Proxy;
-import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Path;
-import java.util.Collections;
-import java.util.List;
 import java.util.Locale;
 import java.util.Objects;
 import java.util.Optional;
@@ -60,7 +54,6 @@ import org.apache.logging.log4j.Level;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.io.IoBuilder;
 import org.bukkit.command.CommandSender;
-import co.aikar.timings.MinecraftTimings; // Paper
 import org.bukkit.event.server.ServerCommandEvent;
 import org.bukkit.craftbukkit.util.Waitable; // Paper
 import org.bukkit.event.server.RemoteServerCommandEvent;
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java
index b76571b367312351b799e440a37fdfa90adbb640..7038022786352d4b12757c6f7b21d6ed43f0a1fa 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java
@@ -19,6 +19,8 @@ import java.util.concurrent.TimeUnit;
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 import javax.annotation.Nullable;
+
+import io.github.dueris.properties.ServerProperties;
 import net.minecraft.core.Holder;
 import net.minecraft.core.HolderLookup;
 import net.minecraft.core.Registry;
@@ -51,7 +53,7 @@ import java.io.*;
 import joptsimple.OptionSet;
 // CraftBukkit end
 
-public class DedicatedServerProperties extends Settings<DedicatedServerProperties> implements com.dueris.canvas.properties.ServerProperties { // Canvas
+public class DedicatedServerProperties extends Settings<DedicatedServerProperties> implements ServerProperties { // Canvas
 
     static final Logger LOGGER = LogUtils.getLogger();
     private static final Pattern SHA1 = Pattern.compile("^[a-fA-F0-9]{40}$");
diff --git a/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java b/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
index bf2e7de9b50bcdef872b6d05825bf64c9f15f4e6..86551d1f7cd60a116e3e732634a4e08648759eef 100644
--- a/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
+++ b/src/main/java/net/minecraft/server/gui/MinecraftServerGui.java
@@ -1,7 +1,7 @@
 package net.minecraft.server.gui;
 
-import com.dueris.canvas.CraftServerGui;
-import com.dueris.canvas.MojangServerGui;
+import io.github.dueris.CraftServerGui;
+import io.github.dueris.MojangServerGui;
 import com.google.common.collect.Lists;
 import com.mojang.logging.LogUtils;
 import java.awt.*;
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 73a492714270253b2f8577d8eadefbeae4b51776..e9500bbaeb818cf1608cebf7acd81b3c1c5908b5 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1,12 +1,12 @@
 package org.bukkit.craftbukkit;
 
-import com.dueris.canvas.CanvasConfig;
-import com.dueris.canvas.CraftServerGui;
-import com.dueris.canvas.logger.CanvasLogger;
-import com.dueris.canvas.logger.CraftCanvasLogger;
-import com.dueris.canvas.logger.Obfuscator;
-import com.dueris.canvas.network.ChatChannelContainer;
-import com.dueris.canvas.network.ChatChannelManager;
+import io.github.dueris.CanvasConfig;
+import io.github.dueris.CraftServerGui;
+import io.github.dueris.logger.CanvasLogger;
+import io.github.dueris.logger.CraftCanvasLogger;
+import io.github.dueris.logger.Obfuscator;
+import io.github.dueris.network.ChatChannelContainer;
+import io.github.dueris.network.ChatChannelManager;
 import com.google.common.base.Charsets;
 import com.google.common.base.Function;
 import com.google.common.base.Preconditions;
@@ -256,7 +256,6 @@ import org.bukkit.plugin.PluginManager;
 import org.bukkit.plugin.ServicesManager;
 import org.bukkit.plugin.SimplePluginManager;
 import org.bukkit.plugin.SimpleServicesManager;
-import org.bukkit.plugin.java.JavaPluginLoader;
 import org.bukkit.plugin.messaging.Messenger;
 import org.bukkit.plugin.messaging.StandardMessenger;
 import org.bukkit.potion.Potion;
@@ -743,7 +742,7 @@ public final class CraftServer implements Server {
     }
 
     @Override
-    public com.dueris.canvas.properties.ServerProperties getServerProperties(){
+    public ServerProperties getServerProperties(){
         return getProperties();
     }
     // Canvas end
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index e9176efeb7b78a005cfafaa635c0b3da90a78d85..36aa7aa6beff9dfbadb98f87b3e0a9225e28a16b 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -9,15 +9,11 @@ import java.util.Date;
 import java.util.List;
 import java.util.logging.Level;
 import java.util.logging.Logger;
-import com.dueris.canvas.CanvasConfig;
-import com.dueris.canvas.logger.CanvasLogger;
-import com.dueris.canvas.logger.CraftCanvasLogger;
-import com.mojang.logging.LogUtils;
+import io.github.dueris.CanvasConfig;
+import io.github.dueris.logger.CraftCanvasLogger;
 import joptsimple.OptionParser;
 import joptsimple.OptionSet;
 import joptsimple.util.PathConverter;
-import org.bukkit.Bukkit;
-import javax.crypto.KeyGenerator;
 import java.net.*;
 
 public class Main {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 3fd266c4e9b4acb7fdc8c3152ca8d4b1af4bd4e1..8bab720b6db56bc811269f7df5a09efeee7dea56 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1,14 +1,12 @@
 package org.bukkit.craftbukkit.entity;
 
 import com.destroystokyo.paper.Title;
-import com.dueris.canvas.network.ChatChannelManager;
-import com.dueris.canvas.network.channels.ChatChannel;
+import io.github.dueris.network.channels.ChatChannel;
 import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableSet;
 import com.google.common.io.BaseEncoding;
 import com.mojang.authlib.GameProfile;
 import com.mojang.datafixers.util.Pair;
-import io.netty.buffer.Unpooled;
 import it.unimi.dsi.fastutil.shorts.ShortArraySet;
 import it.unimi.dsi.fastutil.shorts.ShortSet;
 import java.io.ByteArrayOutputStream;
@@ -38,12 +36,10 @@ import net.minecraft.advancements.AdvancementProgress;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
 import net.minecraft.core.SectionPos;
-import net.minecraft.core.registries.BuiltInRegistries;
 import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.FriendlyByteBuf;
 import net.minecraft.network.chat.Component;
 import net.minecraft.network.chat.PlayerChatMessage;
-import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.common.ClientboundCustomPayloadPacket;
 import net.minecraft.network.protocol.common.ClientboundResourcePackPacket;
 import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
@@ -56,7 +52,6 @@ import net.minecraft.network.protocol.game.ClientboundHurtAnimationPacket;
 import net.minecraft.network.protocol.game.ClientboundLevelEventPacket;
 import net.minecraft.network.protocol.game.ClientboundLevelParticlesPacket;
 import net.minecraft.network.protocol.game.ClientboundMapItemDataPacket;
-import net.minecraft.network.protocol.game.ClientboundPlayerChatPacket;
 import net.minecraft.network.protocol.game.ClientboundPlayerInfoRemovePacket;
 import net.minecraft.network.protocol.game.ClientboundPlayerInfoUpdatePacket;
 import net.minecraft.network.protocol.game.ClientboundSectionBlocksUpdatePacket;
@@ -142,7 +137,6 @@ import org.bukkit.craftbukkit.advancement.CraftAdvancement;
 import org.bukkit.craftbukkit.advancement.CraftAdvancementProgress;
 import org.bukkit.craftbukkit.block.CraftBlockEntityState;
 import org.bukkit.craftbukkit.block.CraftBlockState;
-import org.bukkit.craftbukkit.block.CraftBlockStates;
 import org.bukkit.craftbukkit.block.CraftSign;
 import org.bukkit.craftbukkit.block.data.CraftBlockData;
 import org.bukkit.craftbukkit.conversations.ConversationTracker;
@@ -150,7 +144,6 @@ import org.bukkit.craftbukkit.event.CraftEventFactory;
 import org.bukkit.craftbukkit.inventory.CraftItemStack;
 import org.bukkit.craftbukkit.map.CraftMapView;
 import org.bukkit.craftbukkit.map.RenderData;
-import org.bukkit.craftbukkit.profile.CraftPlayerProfile;
 import org.bukkit.craftbukkit.scoreboard.CraftScoreboard;
 import org.bukkit.craftbukkit.util.CraftChatMessage;
 import org.bukkit.craftbukkit.util.CraftLocation;
@@ -163,7 +156,6 @@ import org.bukkit.event.player.PlayerExpCooldownChangeEvent;
 import org.bukkit.event.player.PlayerHideEntityEvent;
 import org.bukkit.event.player.PlayerRegisterChannelEvent;
 import org.bukkit.event.player.PlayerShowEntityEvent;
-import org.bukkit.event.player.PlayerSpawnChangeEvent;
 import org.bukkit.event.player.PlayerTeleportEvent;
 import org.bukkit.event.player.PlayerUnregisterChannelEvent;
 import org.bukkit.inventory.EquipmentSlot;
@@ -174,7 +166,6 @@ import org.bukkit.map.MapView;
 import org.bukkit.metadata.MetadataValue;
 import org.bukkit.plugin.Plugin;
 import org.bukkit.plugin.messaging.StandardMessenger;
-import org.bukkit.profile.PlayerProfile;
 import org.bukkit.scoreboard.Scoreboard;
 import org.jetbrains.annotations.NotNull;
 
