From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Tue, 21 Nov 2023 01:36:07 +0000
Subject: [PATCH] MC-172801


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index c0f79535b009aee9b4518c735f500cecff36842f..9ab87240cd1fc0e73dbf89b15c971b44f67898ec 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -3104,7 +3104,13 @@ public abstract class LivingEntity extends Entity implements Attackable {
     }
 
     protected float getFlyingSpeed() {
-        return this.getControllingPassenger() instanceof net.minecraft.world.entity.player.Player ? this.getSpeed() * 0.1F : 0.02F;
+        // Canvas start - MC-172801
+        float flyingSpeed = 0.02F;
+        if(this.getAttributes().hasAttribute(Attributes.FLYING_SPEED)){
+            flyingSpeed = (float) (this.getAttribute(Attributes.FLYING_SPEED).getValue() * 0.049999999254942D);
+        }
+        return this.getControllingPassenger() instanceof net.minecraft.world.entity.player.Player ? this.getSpeed() * 0.1F : flyingSpeed;
+        // Canvas end
     }
 
     public float getSpeed() {
