From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: rafaelflromao <12960698+rafaelflromao@users.noreply.github.com>
Date: Mon, 12 Jun 2023 06:01:40 +0100
Subject: [PATCH] Sync event calls on async threads


diff --git a/src/main/java/dev/etil/mirai/MiraiConfig.java b/src/main/java/dev/etil/mirai/MiraiConfig.java
index e37c8f967b0fbaa15e35c53cd29f79e512dd2a53..cf3fed88162efcc8cf4aa1d095066a2abce59887 100644
--- a/src/main/java/dev/etil/mirai/MiraiConfig.java
+++ b/src/main/java/dev/etil/mirai/MiraiConfig.java
@@ -193,4 +193,17 @@ public class MiraiConfig {
             enableAsyncEntityTracker = temp;
         }
     }
+
+    public static boolean enableSyncEventCallsOnAsyncThreads;
+    public static boolean enableSyncEventCallsOnAsyncThreadsInitialized;
+
+    private static void syncEventCallsOnAsyncThreads() {
+        boolean temp = getBoolean("enable-sync-event-calls-on-async-threads", true,
+            "Whether or not sync event calls on async threads should be enabled. (If async entity tracker is enabled, this is enabled.)",
+            "You may encounter issues with plugins.");
+        if (!enableSyncEventCallsOnAsyncThreadsInitialized) {
+            enableSyncEventCallsOnAsyncThreadsInitialized = true;
+            enableSyncEventCallsOnAsyncThreads = temp;
+        }
+    }
 }
\ No newline at end of file
diff --git a/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java b/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
index d076b1a4429dcadb01bf2e1a5e09d500baef08d0..92d494894c3dd43fba888105711771c01f2f27a9 100644
--- a/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
+++ b/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
@@ -42,7 +42,7 @@ class PaperEventManager {
             throw new IllegalStateException(event.getEventName() + " may only be triggered asynchronously.");
         } else if (!event.isAsynchronous() && !this.server.isPrimaryThread() && !this.server.isStopping()) {
             // Mirai start
-            if(MiraiConfig.enableAsyncEntityTracker) {
+            if(MiraiConfig.enableAsyncEntityTracker || MiraiConfig.enableSyncEventCallsOnAsyncThreads) {
                 MinecraftServer.getServer().executeBlocking(event::callEvent);
                 return;
             } else {
