From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Fri, 10 Nov 2023 09:05:29 +0000
Subject: [PATCH] Sync event calls on async threads


diff --git a/src/main/java/dev/etil/mirai/MiraiConfig.java b/src/main/java/dev/etil/mirai/MiraiConfig.java
index 734ca538193d769cafeff8e97fc197a628544aa2..cf3fed88162efcc8cf4aa1d095066a2abce59887 100644
--- a/src/main/java/dev/etil/mirai/MiraiConfig.java
+++ b/src/main/java/dev/etil/mirai/MiraiConfig.java
@@ -182,4 +182,28 @@ public class MiraiConfig {
             Bukkit.getLogger().log(Level.INFO, "Using " + asyncPathfindingMaxThreads + " threads for Async Pathfinding");
     }
 
+    public static boolean enableAsyncEntityTracker;
+    public static boolean enableAsyncEntityTrackerInitialized;
+    private static void asyncEntityTracker() {
+        boolean temp = getBoolean("enable-async-entity-tracker", true,
+            "Whether or not async entity tracking should be enabled.",
+            "You may encounter issues with NPCs.");
+        if (!enableAsyncEntityTrackerInitialized) {
+            enableAsyncEntityTrackerInitialized = true;
+            enableAsyncEntityTracker = temp;
+        }
+    }
+
+    public static boolean enableSyncEventCallsOnAsyncThreads;
+    public static boolean enableSyncEventCallsOnAsyncThreadsInitialized;
+
+    private static void syncEventCallsOnAsyncThreads() {
+        boolean temp = getBoolean("enable-sync-event-calls-on-async-threads", true,
+            "Whether or not sync event calls on async threads should be enabled. (If async entity tracker is enabled, this is enabled.)",
+            "You may encounter issues with plugins.");
+        if (!enableSyncEventCallsOnAsyncThreadsInitialized) {
+            enableSyncEventCallsOnAsyncThreadsInitialized = true;
+            enableSyncEventCallsOnAsyncThreads = temp;
+        }
+    }
 }
\ No newline at end of file
diff --git a/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java b/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
index 7ce9ebba8ce304d1f3f21d4f15ee5f3560d7700b..ce3815cbc0cd240e2152ebf921d93c56fcde0bc3 100644
--- a/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
+++ b/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
@@ -39,7 +39,14 @@ class PaperEventManager {
         if (event.isAsynchronous() && this.server.isPrimaryThread()) {
             throw new IllegalStateException(event.getEventName() + " may only be triggered asynchronously.");
         } else if (!event.isAsynchronous() && !this.server.isPrimaryThread() && !this.server.isStopping()) {
-            throw new IllegalStateException(event.getEventName() + " may only be triggered synchronously.");
+            // Mirai start
+            if(MiraiConfig.enableAsyncEntityTracker || MiraiConfig.enableSyncEventCallsOnAsyncThreads) {
+                MinecraftServer.getServer().executeBlocking(event::callEvent);
+                return;
+            } else {
+                throw new IllegalStateException(event.getEventName() + " may only be triggered synchronously.");
+            }
+            // Mirai end
         }
 
         HandlerList handlers = event.getHandlers();
