From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sun, 5 Nov 2023 20:35:50 -0800
Subject: [PATCH] Fix AI issues with api


diff --git a/src/main/java/net/minecraft/world/entity/monster/Evoker.java b/src/main/java/net/minecraft/world/entity/monster/Evoker.java
index 5115832eece2798738dfbafc8547477fbc79ce62..b01f9fcee348caaa3260f86339ce6c19282f8f42 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Evoker.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Evoker.java
@@ -91,15 +91,17 @@ public class Evoker extends SpellcasterIllager {
     @Override
     protected void registerGoals() {
         super.registerGoals();
-        this.goalSelector.addGoal(0, new FloatGoal(this));
-        this.goalSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this)); // Purpur
-        this.goalSelector.addGoal(1, castingSpellGoal); // Canvas
-        this.goalSelector.addGoal(2, new AvoidEntityGoal<>(this, Player.class, 8.0F, 0.6D, 1.0D));
         // Canvas start
         summonSpellGoal = new EvokerSummonSpellGoal();
         castingSpellGoal = new EvokerCastingSpellGoal();
         attackSpellGoal = new EvokerAttackSpellGoal();
         wololoSpellGoal = new EvokerWololoSpellGoal();
+        // Canvas end
+        this.goalSelector.addGoal(0, new FloatGoal(this));
+        this.goalSelector.addGoal(0, new org.purpurmc.purpur.entity.ai.HasRider(this)); // Purpur
+        this.goalSelector.addGoal(1, castingSpellGoal); // Canvas
+        this.goalSelector.addGoal(2, new AvoidEntityGoal<>(this, Player.class, 8.0F, 0.6D, 1.0D));
+        // Canvas start
         this.goalSelector.addGoal(4, summonSpellGoal);
         this.goalSelector.addGoal(5, attackSpellGoal);
         this.goalSelector.addGoal(6, wololoSpellGoal);
