From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Thu, 9 Nov 2023 22:29:19 +0000
Subject: [PATCH] some entity micro opts


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 9ffc60ebba99f34b5273afd82bde1f0d5c5bc9f0..b922210f40b083551f60ad49a2802301341d9142 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -497,17 +497,34 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     public boolean canSaveToDisk() {
         return true;
     }
+    
+    // JettPack start - cache eye blockpos
+    private BlockPos cached_eye_blockpos;
+    private int cached_position_hashcode;
+    // JettPack end
 
     // Purpur start - copied from Mob
     public boolean isSunBurnTick() {
         if (this.level().isDay() && !this.level().isClientSide) {
-            float f = this.getLightLevelDependentMagicValue();
-            BlockPos blockposition = BlockPos.containing(this.getX(), this.getEyeY(), this.getZ());
+            // JettPack start - optimizations and cache eye blockpos
+            int positionHashCode = this.getPositionHashCode();
+            if (this.cached_position_hashcode != positionHashCode) {
+                this.cached_eye_blockpos = BlockPos.containing(this.getX(), this.getEyeY(), this.getZ());
+                this.cached_position_hashcode = positionHashCode;
+            }
+
+            float f = this.getLightLevelDependentMagicValue(cached_eye_blockpos); // Pass BlockPos to getBrightness
+
+            // Check brightness first
+            if (f <= 0.5F) return false;
+            if (this.random.nextFloat() * 30.0F >= (f - 0.4F) * 2.0F) return false;
+
             boolean flag = this.isInWaterRainOrBubble() || this.isInPowderSnow || this.wasInPowderSnow;
 
-            if (f > 0.5F && this.random.nextFloat() * 30.0F < (f - 0.4F) * 2.0F && !flag && this.level().canSeeSky(blockposition)) {
+            if (!flag && this.level().canSeeSky(this.cached_eye_blockpos)) { // JettPack - move brightness checks up
                 return true;
             }
+            // JettPack end
         }
 
         return false;
@@ -2030,12 +2047,18 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
         }
     }
 
+    // JettPack start - allow passing BlockPos to getLightLevelDependentMagicValue
     /** @deprecated */
     @Deprecated
     public float getLightLevelDependentMagicValue() {
-        return this.level().hasChunkAt(this.getBlockX(), this.getBlockZ()) ? this.level().getLightLevelDependentMagicValue(BlockPos.containing(this.getX(), this.getEyeY(), this.getZ())) : 0.0F;
+        return this.level().getLightLevelDependentMagicValue(BlockPos.containing(this.getX(), this.getEyeY(), this.getZ()));
     }
 
+    public float getLightLevelDependentMagicValue(BlockPos pos) {
++        return this.level().hasChunkAt(this.getBlockX(), this.getBlockZ()) ? this.level().getLightLevelDependentMagicValue(pos) : 0.0F;
++    }
++    // JettPack end
+
     public void absMoveTo(double x, double y, double z, float yaw, float pitch) {
         this.absMoveTo(x, y, z);
         this.setYRot(yaw % 360.0F);
@@ -4709,6 +4732,12 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     }
     // Paper end - block invalid positions
 
+    // JettPack start
+    public final int getPositionHashCode() {
+        return this.position.hashCode();
+    }
+    // JettPack end
+
     public final void setPosRaw(double x, double y, double z) {
         // Paper start
         this.setPosRaw(x, y, z, false);
