From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Thu, 30 Nov 2023 00:18:24 +0000
Subject: [PATCH] Remove chatchannel api


diff --git a/src/main/java/me/dueris/canvas/network/ChatChannelManager.java b/src/main/java/me/dueris/canvas/network/ChatChannelManager.java
deleted file mode 100644
index da6ffc327079cccab7db4b4a9f4d9c22fdcf9eff..0000000000000000000000000000000000000000
--- a/src/main/java/me/dueris/canvas/network/ChatChannelManager.java
+++ /dev/null
@@ -1,41 +0,0 @@
-package me.dueris.canvas.network;
-
-import me.dueris.canvas.network.channels.ChatChannel;
-import me.dueris.canvas.network.channels.CraftChatChannel;
-import org.bukkit.NamespacedKey;
-import java.util.ArrayList;
-
-public class ChatChannelManager implements ChatChannelContainer {
-
-    public ArrayList<ChatChannel> craftChannels = new ArrayList<>();
-
-    @Override
-    public CraftChatChannel createChannel(final NamespacedKey namespacedKey, String title) {
-        CraftChatChannel channel = new CraftChatChannel(namespacedKey, title);
-        craftChannels.add(channel);
-        return channel;
-    }
-
-    @Override
-    public void deleteChannel(ChatChannel chatChannel) {
-        craftChannels.remove(chatChannel);
-    }
-
-    @Override
-    public ChatChannel getChatChannel(final NamespacedKey namespacedKey) {
-        for(ChatChannel chatChannel : craftChannels){
-            if(chatChannel.getChannelKey() == namespacedKey) return chatChannel;
-        }
-        return null;
-    }
-
-    @Override
-    public ArrayList<ChatChannel> getChannels() {
-        return craftChannels;
-    }
-
-    @Override
-    public boolean containsChannel(final ChatChannel chatChannel) {
-        return craftChannels.contains(chatChannel);
-    }
-}
diff --git a/src/main/java/me/dueris/canvas/network/channels/CraftChannelSettings.java b/src/main/java/me/dueris/canvas/network/channels/CraftChannelSettings.java
deleted file mode 100644
index 6929758de4d8d8d356df6af49a312e840289d224..0000000000000000000000000000000000000000
--- a/src/main/java/me/dueris/canvas/network/channels/CraftChannelSettings.java
+++ /dev/null
@@ -1,68 +0,0 @@
-package me.dueris.canvas.network.channels;
-
-public class CraftChannelSettings implements ChannelSettings {
-    CraftChatChannel channel;
-    boolean viewOtherChannelLogs;
-    boolean sendAdvancements;
-    boolean sendDeathMessages;
-    boolean privateChannel;
-    boolean receiveWhispers;
-
-    public CraftChannelSettings(CraftChatChannel channel){
-        this.channel = channel;
-    }
-
-    @Override
-    public void canViewOtherChannelLogs(final boolean T) {
-        viewOtherChannelLogs = T;
-    }
-
-    @Override
-    public void sendAdvancements(final boolean T) {
-        sendAdvancements = T;
-    }
-
-    @Override
-    public void sendDeathMessages(final boolean T) {
-        sendDeathMessages = T;
-    }
-
-    @Override
-    public void privateChannel(final boolean T) {
-        privateChannel = T;
-    }
-
-    @Override
-    public void resetSettings() {
-        privateChannel = true;
-        sendDeathMessages = false;
-        sendAdvancements = false;
-        viewOtherChannelLogs = true;
-        receiveWhispers = true;
-    }
-
-    @Override
-    public boolean isPrivateChannel() {
-        return privateChannel;
-    }
-
-    @Override
-    public boolean canSendDeathMessages() {
-        return sendDeathMessages;
-    }
-
-    @Override
-    public boolean canSendAdvancements() {
-        return sendAdvancements;
-    }
-
-    @Override
-    public boolean canViewOtherChannelLogs() {
-        return viewOtherChannelLogs;
-    }
-
-    @Override
-    public boolean canReceiveWhisperFromOtherChannelMembers() {
-        return receiveWhispers;
-    }
-}
diff --git a/src/main/java/me/dueris/canvas/network/channels/CraftChatChannel.java b/src/main/java/me/dueris/canvas/network/channels/CraftChatChannel.java
deleted file mode 100644
index 7f1975601deeaeb88fb65d00240ad2840a33d354..0000000000000000000000000000000000000000
--- a/src/main/java/me/dueris/canvas/network/channels/CraftChatChannel.java
+++ /dev/null
@@ -1,89 +0,0 @@
-package me.dueris.canvas.network.channels;
-
-import org.bukkit.NamespacedKey;
-import org.bukkit.entity.Player;
-import java.util.ArrayList;
-
-public class CraftChatChannel implements ChatChannel {
-
-    private NamespacedKey namespacedKey;
-    private String title;
-    private ChannelSettings settings = null;
-
-    public CraftChatChannel(NamespacedKey namespacedKey, String title){
-        this.namespacedKey = namespacedKey;
-        this.title = title;
-        this.settings = new CraftChannelSettings(this);
-    }
-
-    public ArrayList<Player> players = new ArrayList<>();
-
-    @Override
-    public ArrayList<Player> getPlayers() {
-        return players;
-    }
-
-    @Override
-    public void addPlayer(final Player player) {
-        players.add(player);
-    }
-
-    @Override
-    public void removePlayer(final Player player) {
-        players.remove(player);
-    }
-
-    @Override
-    public void changeNamespace(final NamespacedKey namespacedKey) {
-        this.namespacedKey = namespacedKey;
-    }
-
-    @Override
-    public void sendMessage(final String msg) {
-        players.forEach((player) -> player.sendMessage(msg));
-    }
-
-    @Override
-    public void sendMessages(final String[] msgs) {
-        for(String msg : msgs){
-            players.forEach((player) -> player.sendMessage(msg));
-        }
-    }
-
-    @Override
-    public void executeCommand(final String cmd) {
-        players.forEach((player) -> player.sendMessage(cmd));
-    }
-
-    @Override
-    public void executeCommands(final String[] cmds) {
-        for(String cmd : cmds){
-            players.forEach((player) -> player.performCommand(cmd));
-        }
-    }
-
-    @Override
-    public boolean containsPlayer(final Player player) {
-        return players.contains(player);
-    }
-
-    @Override
-    public NamespacedKey getChannelKey() {
-        return this.namespacedKey;
-    }
-
-    @Override
-    public String getChannelName() {
-        return this.title;
-    }
-
-    @Override
-    public void changeName(final String name) {
-        this.title = name;
-    }
-
-    @Override
-    public ChannelSettings getChannelSettings() {
-        return this.settings;
-    }
-}
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 78cffcd9e32c92dc8199fe38c2602ea36f838302..690ef5f4b936968e7609afcdf6f22a9fa86f0a14 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -5,8 +5,6 @@ import me.dueris.canvas.CraftServerGui;
 import me.dueris.canvas.logger.CanvasLogger;
 import me.dueris.canvas.logger.CraftCanvasLogger;
 import me.dueris.canvas.logger.Obfuscator;
-import me.dueris.canvas.network.ChatChannelContainer;
-import me.dueris.canvas.network.ChatChannelManager;
 import me.dueris.canvas.properties.*;
 import com.google.common.base.Charsets;
 import com.google.common.base.Function;
@@ -300,7 +298,6 @@ public final class CraftServer implements Server {
     public static CraftCanvasLogger.Obfuscator canvas_obfuscator = null;
     private final CraftCanvasLogger craftCanvasLogger = new CraftCanvasLogger();
     public Logger LOGGER = Logger.getLogger(CraftServer.class.getName());
-    private final ChatChannelManager chatChannelManager = new ChatChannelManager();
     public static String[] startupArgs = null;
     // Canvas end
     private final String bukkitVersion = Versioning.getBukkitVersion();
@@ -980,11 +977,6 @@ public final class CraftServer implements Server {
         return this.console.getGui();
     }
 
-    @Override
-    public @NotNull ChatChannelContainer getChatChannelManager() {
-        return this.chatChannelManager;
-    }
-
     @NotNull
     public net.minecraft.server.gui.MinecraftServerGui getGuiHandle(){
         return this.console.getGui();
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index ec4311e3d991559e82ddab1cefc31eaa9c105051..cca9cb7ac6b1ca003bd8dbae97a512ea2a9e29de 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1,7 +1,6 @@
 package org.bukkit.craftbukkit.entity;
 
 import com.destroystokyo.paper.Title;
-import me.dueris.canvas.network.channels.ChatChannel;
 import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableSet;
 import com.google.common.io.BaseEncoding;
@@ -3029,26 +3028,6 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         this.getHandle().displayName = null;
     }
 
-    @Override
-    public ChatChannel getChatChannel() {
-        for(ChatChannel channel : this.getServer().getChatChannelManager().getChannels()){
-            if(channel.containsPlayer(this)) {
-                return channel;
-            }
-        }
-        return null;
-    }
-
-    @Override
-    public boolean isInChatChannel() {
-        for(ChatChannel channel : this.getServer().getChatChannelManager().getChannels()){
-            if(channel.containsPlayer(this)) {
-                return true;
-            }
-        }
-        return false;
-    }
-
     @Override
     public Chunk[] getChunksInRenderDistance() {
         int viewDistance = this.getClientViewDistance();
