From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Fri, 10 Nov 2023 20:32:53 -0800
Subject: [PATCH] Enhance Arrow API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftArrow.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftArrow.java
index 6d2fe30742f8b41d53dd2cbff120fcc042ea0e0c..230438530b56fb849adc486cc4560d0d23068802 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftArrow.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftArrow.java
@@ -2,6 +2,7 @@ package org.bukkit.craftbukkit.entity;
 
 import com.google.common.base.Preconditions;
 import net.minecraft.core.BlockPos;
+import net.minecraft.world.entity.projectile.SpectralArrow;
 import org.bukkit.block.Block;
 import org.bukkit.craftbukkit.CraftServer;
 import org.bukkit.entity.AbstractArrow;
@@ -122,7 +123,13 @@ public class CraftArrow extends AbstractProjectile implements AbstractArrow {
     public boolean hasNoPhysics() {
         return this.getHandle().isNoPhysics();
     }
-    // Paper end
+
+    // Canvas start
+    @Override
+    public boolean isSpectral() {
+        return this.getHandle() instanceof SpectralArrow;
+    }
+    // Paper end -/- Canvas end
 
     @Override
     public void setTicksLived(int value) {
