From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Wed, 15 Nov 2023 21:54:36 +0000
Subject: [PATCH] Fix broken error on first server creation


diff --git a/src/main/java/io/github/dueris/plugin/CanvasPluginLoader.java b/src/main/java/io/github/dueris/plugin/CanvasPluginLoader.java
index 7fd179a2e55e7aaa8f27c4d552e75c2df4a88079..10dbb9b5fd720ae5b6dc383c9f900e64ce6f94dd 100644
--- a/src/main/java/io/github/dueris/plugin/CanvasPluginLoader.java
+++ b/src/main/java/io/github/dueris/plugin/CanvasPluginLoader.java
@@ -35,6 +35,7 @@ public class CanvasPluginLoader {
         pluginLoader.clearCache();
         // System.out.println(Paths.get("plugins"));
             File dir = Paths.get("plugins").toFile();
+            dir.mkdirs();
 
             for(File file : dir.listFiles()){
                 if(file.getName().endsWith(".jar")){
