From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Thu, 16 Nov 2023 19:07:48 +0000
Subject: [PATCH] More Advancement API


diff --git a/src/main/java/net/minecraft/advancements/DisplayInfo.java b/src/main/java/net/minecraft/advancements/DisplayInfo.java
index 11b26d14b863075d776884544629cdd23f219298..d39597ad50680f6ba86662c74338cf1b3e4d9677 100644
--- a/src/main/java/net/minecraft/advancements/DisplayInfo.java
+++ b/src/main/java/net/minecraft/advancements/DisplayInfo.java
@@ -18,15 +18,17 @@ import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
 
 public class DisplayInfo {
-    private final Component title;
-    private final Component description;
-    private final ItemStack icon;
+    // Canvas start - private -> public && final -> !final
+    public Component title;
+    public Component description;
+    public ItemStack icon;
     @Nullable
-    private final ResourceLocation background;
-    private final FrameType frame;
-    private final boolean showToast;
-    private final boolean announceChat;
-    private final boolean hidden;
+    public ResourceLocation background;
+    public FrameType frame;
+    public boolean showToast;
+    public boolean announceChat;
+    public boolean hidden;
+    // Canvas end
     private float x;
     private float y;
     public final io.papermc.paper.advancement.AdvancementDisplay paper = new io.papermc.paper.advancement.PaperAdvancementDisplay(this); // Paper
@@ -50,6 +52,16 @@ public class DisplayInfo {
     public Component getTitle() {
         return this.title;
     }
+    // Canvas start
+
+    public void setTitle(Component title){
+        this.title = title;
+    }
+
+    public void setDescription(Component description){
+        this.description = description;
+    }
+    // Canvas end
 
     public Component getDescription() {
         return this.description;
@@ -58,6 +70,12 @@ public class DisplayInfo {
     public ItemStack getIcon() {
         return this.icon;
     }
+    // Canvas start
+
+    public void setIcon(ItemStack icon) {
+        this.icon = icon;
+    }
+    // Canvas end
 
     @Nullable
     public ResourceLocation getBackground() {
@@ -79,6 +97,16 @@ public class DisplayInfo {
     public boolean shouldShowToast() {
         return this.showToast;
     }
+    // Canvas start
+    
+    public void setShouldShowToast(boolean should){
+        this.showToast = should;
+    }
+
+    public void setShouldAnnounceChat(boolean should){
+        this.announceChat = should;
+    }
+    // Canvas end
 
     public boolean shouldAnnounceChat() {
         return this.announceChat;
@@ -87,6 +115,12 @@ public class DisplayInfo {
     public boolean isHidden() {
         return this.hidden;
     }
+    // Canvas start
+
+    public void setHidden(boolean hidden) {
+        this.hidden = hidden;
+    }
+    // Canvas end
 
     public static DisplayInfo fromJson(JsonObject obj) {
         Component component = Component.Serializer.fromJson(obj.get("title"));
diff --git a/src/main/java/org/bukkit/craftbukkit/advancement/CraftAdvancement.java b/src/main/java/org/bukkit/craftbukkit/advancement/CraftAdvancement.java
index bd640e0d8d796ee114ff787def7e07edbeffc0a5..ae06c2a2547af758b8a94e8cfbd89dd7a6ea591d 100644
--- a/src/main/java/org/bukkit/craftbukkit/advancement/CraftAdvancement.java
+++ b/src/main/java/org/bukkit/craftbukkit/advancement/CraftAdvancement.java
@@ -4,6 +4,7 @@ import java.util.Collection;
 import java.util.Collections;
 import net.minecraft.advancements.AdvancementHolder;
 import org.bukkit.NamespacedKey;
+import org.bukkit.advancement.Advancement;
 import org.bukkit.advancement.AdvancementDisplay;
 import org.bukkit.craftbukkit.util.CraftNamespacedKey;
 
diff --git a/src/main/java/org/bukkit/craftbukkit/advancement/CraftAdvancementDisplay.java b/src/main/java/org/bukkit/craftbukkit/advancement/CraftAdvancementDisplay.java
index e29d7c6e1cef10a76c8630855fada11cee583d30..189c62cba4e9c5096d9ac18fc53db215fdb7ceb4 100644
--- a/src/main/java/org/bukkit/craftbukkit/advancement/CraftAdvancementDisplay.java
+++ b/src/main/java/org/bukkit/craftbukkit/advancement/CraftAdvancementDisplay.java
@@ -23,6 +23,18 @@ public class CraftAdvancementDisplay implements org.bukkit.advancement.Advanceme
     public String getTitle() {
         return CraftChatMessage.fromComponent(this.handle.getTitle());
     }
+    // Canvas start
+
+    @Override
+    public void setTitle(String title){
+        this.getHandle().setTitle(CraftChatMessage.fromStringOrNull(title));
+    }
+
+    @Override
+    public void setDescription(String description){
+        this.getHandle().setTitle(CraftChatMessage.fromStringOrNull(description));
+    }
+    // Canvas end
 
     @Override
     public String getDescription() {
@@ -38,6 +50,18 @@ public class CraftAdvancementDisplay implements org.bukkit.advancement.Advanceme
     public boolean shouldShowToast() {
         return this.handle.shouldShowToast();
     }
+    // Canvas start
+
+    @Override
+    public void setShouldShowToast(boolean should){
+        this.getHandle().setShouldShowToast(should);
+    }
+
+    @Override
+    public void setShouldAnnounceChat(boolean should){
+        this.getHandle().setShouldAnnounceChat(should);
+    }
+    // Canvas end
 
     @Override
     public boolean shouldAnnounceChat() {
@@ -63,4 +87,16 @@ public class CraftAdvancementDisplay implements org.bukkit.advancement.Advanceme
     public AdvancementDisplayType getType() {
         return AdvancementDisplayType.values()[this.handle.getFrame().ordinal()];
     }
+    // Canvas start - more advancement apis
+
+    @Override
+    public void setHidden(boolean hidden) {
+        this.getHandle().setHidden(hidden);
+    }
+
+    @Override
+    public void setIcon(org.bukkit.inventory.ItemStack item){
+        this.getHandle().setIcon(CraftItemStack.unwrap(item));
+    }
+    // Canvas end
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/advancement/CraftAdvancementProgress.java b/src/main/java/org/bukkit/craftbukkit/advancement/CraftAdvancementProgress.java
index ec3b9cb901913b093c3eb0bda8dc0dbb738c165e..6bd0000ea3365f10876f8cd9462ac0d1195cd491 100644
--- a/src/main/java/org/bukkit/craftbukkit/advancement/CraftAdvancementProgress.java
+++ b/src/main/java/org/bukkit/craftbukkit/advancement/CraftAdvancementProgress.java
@@ -20,6 +20,12 @@ public class CraftAdvancementProgress implements AdvancementProgress {
         this.playerData = player;
         this.handle = handle;
     }
+    // Canvas start - no handle method?
+
+    public net.minecraft.advancements.AdvancementProgress getHandle(){
+        return this.handle;
+    }
+    // Canvas end
 
     @Override
     public Advancement getAdvancement() {
