From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Fri, 17 Nov 2023 20:53:18 +0000
Subject: [PATCH] Add entity movement api


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 3bd4a49d42b700f519067577891287dec17d16fc..40f8862e26c22c6a0e1a27342049af1e983e93d2 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -821,7 +821,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
         return this.dimensions.makeBoundingBox(this.position);
     }
 
-    protected void reapplyPosition() {
+    public void reapplyPosition() { // Canvas - protected -> public
         this.setPos(this.position.x, this.position.y, this.position.z);
     }
 
@@ -1658,7 +1658,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
         }
     }
 
-    protected float nextStep() {
+    public float nextStep() { // Canvas - protected -> public
         return (float) ((int) this.moveDist + 1);
     }
 
diff --git a/src/main/java/net/minecraft/world/entity/animal/Turtle.java b/src/main/java/net/minecraft/world/entity/animal/Turtle.java
index e1c012442ac80ccdef2c1f6744899629146ee750..73bf868c477f917f598c90515fd18fb6f070beb3 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Turtle.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Turtle.java
@@ -302,7 +302,7 @@ public class Turtle extends Animal {
     }
 
     @Override
-    protected float nextStep() {
+    public float nextStep() { // Canvas - protected -> public
         return this.moveDist + 0.15F;
     }
 
diff --git a/src/main/java/net/minecraft/world/entity/monster/Strider.java b/src/main/java/net/minecraft/world/entity/monster/Strider.java
index 1ea12e3fa05b3dbbef8e9daaeb114f2a433bf747..c0502deca7a929174ebfa0b2acf2aee1571ba4e6 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Strider.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Strider.java
@@ -352,7 +352,7 @@ public class Strider extends Animal implements ItemSteerable, Saddleable {
     }
 
     @Override
-    protected float nextStep() {
+    public float nextStep() { // Canvas - protected -> public
         return this.moveDist + 0.6F;
     }
 
diff --git a/src/main/java/net/minecraft/world/entity/monster/warden/Warden.java b/src/main/java/net/minecraft/world/entity/monster/warden/Warden.java
index f83d73c1d322374a345999ea65d23431e0779e85..92543f892d63976cd33f1c953f88b648c169a2e5 100644
--- a/src/main/java/net/minecraft/world/entity/monster/warden/Warden.java
+++ b/src/main/java/net/minecraft/world/entity/monster/warden/Warden.java
@@ -208,7 +208,7 @@ public class Warden extends Monster implements VibrationSystem {
     // Canvas end
 
     @Override
-    protected float nextStep() {
+    public float nextStep() { // Canvas - protected -> public
         return this.moveDist + 0.55F;
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index aa8efff39ca3b2c2ed69a19f1872a295b7103c6b..56d46f77e721eceef93730cc65aec7c8fe38be04 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -25,6 +25,7 @@ import net.minecraft.world.entity.Interaction;
 import net.minecraft.world.entity.LightningBolt;
 import net.minecraft.world.entity.LivingEntity;
 import net.minecraft.world.entity.Marker;
+import net.minecraft.world.entity.MoverType;
 import net.minecraft.world.entity.PathfinderMob;
 import net.minecraft.world.entity.TamableAnimal;
 import net.minecraft.world.entity.ambient.AmbientCreature;
@@ -161,6 +162,8 @@ import net.minecraft.world.entity.vehicle.MinecartHopper;
 import net.minecraft.world.entity.vehicle.MinecartSpawner;
 import net.minecraft.world.entity.vehicle.MinecartTNT;
 import net.minecraft.world.phys.AABB;
+import net.minecraft.world.phys.Vec3;
+
 import org.bukkit.EntityEffect;
 import org.bukkit.Location;
 import org.bukkit.Server;
@@ -495,6 +498,39 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
     public Vector getVelocity() {
         return CraftVector.toBukkit(this.entity.getDeltaMovement());
     }
+    // Canvas start - movement api
+
+    @Override
+    public void step(){
+        this.step(1);
+    }
+
+    @Override
+    public void step(int steps){
+        for(int i = 0; i < steps; i++){
+            double x = this.getLocation().getDirection().getX();
+            double y = this.getLocation().getDirection().getY();
+            double z = this.getLocation().getDirection().getZ();
+
+            this.getHandle().move(MoverType.SELF, new Vec3(x, y, z));
+        }
+    }
+
+    @Override
+    public void setDeltaMovement(double x, double y, double z){
+        this.getHandle().setDeltaMovement(x, y, z);
+    }
+
+    @Override
+    public void reapplyPosition(){
+        this.getHandle().reapplyPosition();
+    }
+
+    @Override
+    public void turn(double cursorDeltaX, double cursorDeltaY){
+        this.getHandle().turn(cursorDeltaX, cursorDeltaY);
+    }
+    // Canvas end
 
     @Override
     public void setVelocity(Vector velocity) {
@@ -1226,6 +1262,13 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
     public void setPortalCooldown(int cooldown) {
         this.getHandle().portalCooldown = cooldown;
     }
+    // Canvas - why not have "isInPortalCooldown?"
+
+    @Override
+    public boolean isInPortalCooldown(){
+        return this.getHandle().portalCooldown > 0;
+    }
+    // Canvas end
 
     @Override
     public Set<String> getScoreboardTags() {
