From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Fri, 17 Nov 2023 22:28:02 +0000
Subject: [PATCH] More block apis


diff --git a/src/main/java/me/dueris/canvas/CanvasConfig.java b/src/main/java/me/dueris/canvas/CanvasConfig.java
index 1236d20c1be82e0c5cf505a72a01391ca2aa2f63..3a4b237b9acce62556639e14b4aa673e3d88e6b5 100644
--- a/src/main/java/me/dueris/canvas/CanvasConfig.java
+++ b/src/main/java/me/dueris/canvas/CanvasConfig.java
@@ -152,6 +152,9 @@ public class CanvasConfig {
     public static boolean obfuscateIps = false;
     public static boolean bypassStartupCheck = false;
     public static boolean wantChunky = true;
+    public static boolean useCanvasBedSetDye = true;
+    public static boolean makeCampfireAlwaysLots = false;
+    public static int farmlandSearchRadius = 4;
 
     private static void mixin(){
         disableMixin = getBoolean("mixin.disableMixins", disableMixin);
@@ -165,4 +168,9 @@ public class CanvasConfig {
     private static void plugins(){
         wantChunky = getBoolean("plugins.use_provided_chunky", wantChunky);
     }
+    private static void blocks(){
+        useCanvasBedSetDye = getBoolean("blocks.useCanvasBedSetDyeMethod", useCanvasBedSetDye);
+        makeCampfireAlwaysLots = getBoolean("block.forceCampfireEmitLotsOfParticles", makeCampfireAlwaysLots);
+        farmlandSearchRadius = getInt("block.farmlandSearchRadius", farmlandSearchRadius);
+    }
 }
diff --git a/src/main/java/net/minecraft/world/level/block/FarmBlock.java b/src/main/java/net/minecraft/world/level/block/FarmBlock.java
index d566f67f8f6f1748023430de4f191881b79e44a1..db0c40ed8d76c82322773d8431588182dd9a6ee8 100644
--- a/src/main/java/net/minecraft/world/level/block/FarmBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/FarmBlock.java
@@ -157,15 +157,18 @@ public class FarmBlock extends Block {
         return world.getBlockState(pos.above()).is(BlockTags.MAINTAINS_FARMLAND);
     }
 
+    // Canvas start
+    private static int searchRadius = me.dueris.canvas.CanvasConfig.farmlandSearchRadius;
+    
     private static boolean isNearWater(LevelReader world, BlockPos pos) {
         // Paper start - remove abstract block iteration
         int xOff = pos.getX();
         int yOff = pos.getY();
         int zOff = pos.getZ();
 
-        for (int dz = -4; dz <= 4; ++dz) {
+        for (int dz = -searchRadius; dz <= searchRadius; ++dz) {
             int z = dz + zOff;
-            for (int dx = -4; dx <= 4; ++dx) {
+            for (int dx = -searchRadius; dx <= searchRadius; ++dx) {
                 int x = xOff + dx;
                 for (int dy = 0; dy <= 1; ++dy) {
                     int y = dy + yOff;
@@ -180,6 +183,7 @@ public class FarmBlock extends Block {
 
         return ((ServerLevel) world).purpurConfig.farmlandGetsMoistFromBelow && world.getFluidState(pos.relative(Direction.DOWN)).is(FluidTags.WATER); // Purpur;
     }
+    // Canvas end
 
     @Override
     protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
diff --git a/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
index d80855b22dc10dbf697578d5f78664ed7b6ac572..87859e4d938002e87c357d2f83745b7d868411e8 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
@@ -129,7 +129,7 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
 
         if (randomsource.nextFloat() < 0.11F) {
             for (i = 0; i < randomsource.nextInt(2) + 2; ++i) {
-                CampfireBlock.makeParticles(world, pos, (Boolean) state.getValue(CampfireBlock.SIGNAL_FIRE), false);
+                CampfireBlock.makeParticles(world, pos, (Boolean) state.getValue(CampfireBlock.SIGNAL_FIRE), getShouldBeLots(false)); // Canvas
             }
         }
 
@@ -150,6 +150,13 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
         }
 
     }
+    // Canvas start
+    private static boolean forceLots = me.dueris.canvas.CanvasConfig.makeCampfireAlwaysLots;
+
+    public static boolean getShouldBeLots(boolean original){
+        return forceLots ? true : original;
+    }
+    // Canvas end
 
     public NonNullList<ItemStack> getItems() {
         return this.items;
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBed.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBed.java
index 6595ca064fa410d6249c928c0009ea167d9e6d6e..dccd7ad325c98a81770dd573895cbc7a7b1e5d09 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBed.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBed.java
@@ -1,7 +1,11 @@
 package org.bukkit.craftbukkit.block;
 
 import net.minecraft.world.level.block.entity.BedBlockEntity;
+
+import java.nio.channels.UnsupportedAddressTypeException;
+
 import org.bukkit.DyeColor;
+import org.bukkit.Material;
 import org.bukkit.World;
 import org.bukkit.block.Bed;
 
@@ -57,7 +61,15 @@ public class CraftBed extends CraftBlockEntityState<BedBlockEntity> implements B
 
     @Override
     public void setColor(DyeColor color) {
-        throw new UnsupportedOperationException("Must set block type to appropriate bed colour");
+        // throw new UnsupportedOperationException("Must set block type to appropriate bed colour"); - Canvas - why tho lol
+        if(me.dueris.canvas.CanvasConfig.useCanvasBedSetDye){
+            this.getBlock().setType(Material.valueOf(
+                color.toDyeString().toUpperCase() + "_BED"
+            ));
+        }else{
+            // Then we can use Paper/Bukkits throw exception
+            throw new UnsupportedOperationException("Must set block type to appropriate bed colour");
+        }
     }
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftCampfire.java b/src/main/java/org/bukkit/craftbukkit/block/CraftCampfire.java
index cc641dc70dc12b902bd9cad93fffded9032f1e82..5e3df69406a391563761de4cd76e7091222567d6 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftCampfire.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftCampfire.java
@@ -20,6 +20,13 @@ public class CraftCampfire extends CraftBlockEntityState<CampfireBlockEntity> im
     public int getSize() {
         return this.getSnapshot().getItems().size();
     }
+    // Canvas start
+
+    @Override
+    public void extinguish(){
+        this.getSnapshot().dowse();
+    }
+    // Canvas end
 
     @Override
     public ItemStack getItem(int index) {
