From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sat, 18 Nov 2023 12:17:13 -0800
Subject: [PATCH] Remove Redundant Cast


diff --git a/src/main/java/net/minecraft/advancements/AdvancementTree.java b/src/main/java/net/minecraft/advancements/AdvancementTree.java
index 8aa7439d9f3c071f88c84c6c75b7a65e5e5cfa82..acb059dc92438ac02c192365f22d4921c8c86ca2 100644
--- a/src/main/java/net/minecraft/advancements/AdvancementTree.java
+++ b/src/main/java/net/minecraft/advancements/AdvancementTree.java
@@ -85,7 +85,7 @@ public class AdvancementTree {
         Map map = this.nodes;
 
         Objects.requireNonNull(this.nodes);
-        AdvancementNode advancementnode = (AdvancementNode) optional.map(map::get).orElse((Object) null);
+        AdvancementNode advancementnode = (AdvancementNode) optional.map(map::get).orElse(null);
 
         if (advancementnode == null && optional.isPresent()) {
             return false;
