From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Tue, 21 Nov 2023 01:36:07 +0000
Subject: [PATCH] MC-172801


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 52fb92fad221d63b0868c392ea2718113f6d59a8..e436886edd4a35caad5f3a739e3b42e12bfcf557 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -3096,7 +3096,13 @@ public abstract class LivingEntity extends Entity implements Attackable {
     }
 
     protected float getFlyingSpeed() {
-        return this.getControllingPassenger() instanceof net.minecraft.world.entity.player.Player ? this.getSpeed() * 0.1F : 0.02F;
+        // Canvas start - MC-172801
+        float flyingSpeed = 0.02F;
+        if(this.getAttributes().hasAttribute(Attributes.FLYING_SPEED)){
+            flyingSpeed = (float) (this.getAttribute(Attributes.FLYING_SPEED).getValue() * 0.049999999254942D);
+        }
+        return this.getControllingPassenger() instanceof net.minecraft.world.entity.player.Player ? this.getSpeed() * 0.1F : flyingSpeed;
+        // Canvas end
     }
 
     public float getSpeed() {
