From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris MC <purplewolf.mc1@gmail.com>
Date: Tue, 21 Nov 2023 19:14:51 +0000
Subject: [PATCH] Add getReloadCount() to Server


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index e3cf2f617e00b6f391d45462b6d1dd780637b374..7919d2ec1ff59a7a6fc17f88d7e064b794c57abc 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1315,6 +1315,13 @@ public final class CraftServer implements Server {
 
         return false;
     }
+    // Canvas start
+
+    @Override
+    public int getReloadCount(){
+        return this.reloadCount;
+    }
+    // Canvas end
 
     @Override
     public void reload() {
@@ -1410,6 +1417,7 @@ public final class CraftServer implements Server {
         }
         io.papermc.paper.plugin.PluginInitializerManager.reload(this.console); // Paper
         me.dueris.canvas.plugin.CanvasPluginLoader.loader(); // Canvas - reload cache mappings
+        
         this.loadPlugins();
         this.enablePlugins(PluginLoadOrder.STARTUP);
         this.enablePlugins(PluginLoadOrder.POSTWORLD);
