From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sun, 15 Oct 2023 15:32:59 -0700
Subject: [PATCH] remove old configuration options


diff --git a/src/main/java/com/dueris/canvas/CanvasConfig.java b/src/main/java/com/dueris/canvas/CanvasConfig.java
index 5b3381df08ca70f6a00f6b923c20c31e2f73c669..6ab31ffe2d10c9a3391fe9e10911baa8b9d408cf 100644
--- a/src/main/java/com/dueris/canvas/CanvasConfig.java
+++ b/src/main/java/com/dueris/canvas/CanvasConfig.java
@@ -170,15 +170,10 @@ public class CanvasConfig {
 
     public static boolean disableMixin = false;
     public static boolean obfuscateIps = false;
-    public static boolean darkModeGui = false;
-    public static boolean worldMapGui = false;
     private static void mixin(){
         disableMixin = getBoolean("mixin.disableMixins", disableMixin);
     }
     private static void logger(){
         obfuscateIps = getBoolean("logger.obfuscateIps", obfuscateIps);
     }
-    private static void gui(){
-        darkModeGui = getBoolean("gui.darkModeGui", darkModeGui);
-    }
 }
