From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sun, 15 Oct 2023 15:40:11 -0700
Subject: [PATCH] Remove old server.properties option


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java
index 148e2b4768eb6fc0569a2c7fdedbcd69a7766451..cda4544ae96a4fcb5c6c4483df67a59f1b53fd27 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServerProperties.java
@@ -66,9 +66,6 @@ public class DedicatedServerProperties extends Settings<DedicatedServerPropertie
     public final String motd = this.get("motd", "A Minecraft Server");
     public final boolean forceGameMode = this.get("force-gamemode", false);
     public final boolean enforceWhitelist = this.get("enforce-whitelist", false);
-    // Canvas start
-    public boolean allowMixin = this.get("allow-mixin", true);
-    // Canvas end
     public final Difficulty difficulty;
     public final GameType gamemode;
     public final String levelName;
@@ -119,9 +116,6 @@ public class DedicatedServerProperties extends Settings<DedicatedServerPropertie
     public DedicatedServerProperties(Properties properties, OptionSet optionset) {
         super(properties, optionset);
         // CraftBukkit end
-        // Canvas start
-        this.allowMixin = this.get("allow-mixin", true);
-        // Canvas end
         this.difficulty = (Difficulty) this.get("difficulty", dispatchNumberOrString(Difficulty::byId, Difficulty::byName), Difficulty::getKey, Difficulty.EASY);
         this.gamemode = (GameType) this.get("gamemode", dispatchNumberOrString(GameType::byId, GameType::byName), GameType::getName, GameType.SURVIVAL);
         this.levelName = this.get("level-name", "world");
