From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Fri, 20 Oct 2023 19:03:42 -0700
Subject: [PATCH] Ensure IgniteEngine is running for startup


diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 432d086518f0c7a5ab83b755236648f020d3e5f7..5e561bec1830602c3a5133d1c4ae676701548f41 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -11,6 +11,7 @@ import java.util.logging.Level;
 import java.util.logging.Logger;
 import com.dueris.canvas.logger.CanvasLogger;
 import com.dueris.canvas.logger.CraftCanvasLogger;
+import com.mojang.logging.LogUtils;
 import joptsimple.OptionParser;
 import joptsimple.OptionSet;
 import joptsimple.util.PathConverter;
@@ -28,7 +29,13 @@ public class Main {
     // Paper end
 
     public static void main(String[] args) {
-        // Paper start
+        // Canvas start - ensure loading canvas correctly
+        try {
+            Class.forName("space.vectrix.ignite.applaunch.IgniteBootstrap");
+        } catch (ClassNotFoundException e) {
+            throw new RuntimeException("Unable to start Canvas, please use the Canvas Launcher");
+        }
+        // Canvas end -/- Paper start
         final String warnWhenLegacyFormattingDetected = String.join(".", "net", "kyori", "adventure", "text", "warnWhenLegacyFormattingDetected");
         if (false && System.getProperty(warnWhenLegacyFormattingDetected) == null) {
             System.setProperty(warnWhenLegacyFormattingDetected, String.valueOf(true));
