From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PurpleWolfMC <jedimastertoothless@hotmail.com>
Date: Sat, 21 Oct 2023 12:30:50 -0700
Subject: [PATCH] Add to iron golem api


diff --git a/src/main/java/net/minecraft/world/entity/animal/IronGolem.java b/src/main/java/net/minecraft/world/entity/animal/IronGolem.java
index 64aba511e615983988cdb6a0fd45b7d9d4f2f16d..edd90334b464b96dc4a20784e90fc6bfe734a736 100644
--- a/src/main/java/net/minecraft/world/entity/animal/IronGolem.java
+++ b/src/main/java/net/minecraft/world/entity/animal/IronGolem.java
@@ -4,6 +4,7 @@ import com.google.common.collect.ImmutableList;
 import java.util.Comparator;
 import java.util.Iterator;
 import java.util.List;
+import java.util.Map;
 import java.util.UUID;
 import java.util.stream.Stream;
 import javax.annotation.Nullable;
@@ -41,6 +42,7 @@ import net.minecraft.world.entity.ai.goal.target.ResetUniversalAngerTargetGoal;
 import net.minecraft.world.entity.monster.Creeper;
 import net.minecraft.world.entity.monster.Enemy;
 import net.minecraft.world.entity.player.Player;
+import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.Items;
 import net.minecraft.world.level.Level;
@@ -49,6 +51,8 @@ import net.minecraft.world.level.NaturalSpawner;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.material.Fluids;
 import net.minecraft.world.phys.Vec3;
+import org.bukkit.Material;
+import org.bukkit.craftbukkit.util.CraftMagicNumbers;
 
 public class IronGolem extends AbstractGolem implements NeutralMob {
 
@@ -230,6 +234,7 @@ public class IronGolem extends AbstractGolem implements NeutralMob {
     @Override
     public boolean doHurtTarget(Entity target) {
         this.attackAnimationTick = 10;
+        Vec3 velocity = translateIntToVec3(getFlingVelocity()); // Canvas
         this.level().broadcastEntityEvent(this, (byte) 4);
         float f = this.getAttackDamage();
         float f1 = (int) f > 0 ? f / 2.0F + (float) this.random.nextInt((int) f) : f;
@@ -249,7 +254,8 @@ public class IronGolem extends AbstractGolem implements NeutralMob {
             double d1 = d0;
             double d2 = Math.max(0.0D, 1.0D - d1);
 
-            target.setDeltaMovement(target.getDeltaMovement().add(0.0D, 0.4000000059604645D * d2, 0.0D));
+            target.setDeltaMovement(target.getDeltaMovement().add(velocity)); // Canvas
+            this.setRepairItem(Material.IRON_INGOT);
             this.doEnchantDamageEffects(this, target);
         }
 
@@ -257,6 +263,22 @@ public class IronGolem extends AbstractGolem implements NeutralMob {
         return flag;
     }
 
+    // Canvas start
+    private static double yVelocity = 0.4000000059604645D;
+
+    public void setFlingVelocity(double newYVelocity) {
+        yVelocity = newYVelocity;
+    }
+
+    public double getFlingVelocity(){
+        return yVelocity;
+    }
+
+    private Vec3 translateIntToVec3(double amount) {
+        return new Vec3(0.0D, amount * yVelocity, 0.0D);
+    }
+    // Canvas end
+
     @Override
     public boolean hurt(DamageSource source, float amount) {
         IronGolem.Crackiness entityirongolem_cracklevel = this.getCrackiness();
@@ -317,7 +339,7 @@ public class IronGolem extends AbstractGolem implements NeutralMob {
     protected InteractionResult mobInteract(Player player, InteractionHand hand) {
         ItemStack itemstack = player.getItemInHand(hand);
 
-        if (!itemstack.is(Items.IRON_INGOT)) {
+        if (!itemstack.is(repairitem) || !canRepair) { // Canvas
             return tryRide(player, hand); // Purpur
         } else {
             float f = this.getHealth();
@@ -339,6 +361,20 @@ public class IronGolem extends AbstractGolem implements NeutralMob {
         }
     }
 
+    // Canvas start
+    public boolean canRepair = true;
+
+    public void setCanRepair(boolean can){
+        canRepair = can;
+    }
+
+    public Item repairitem = Items.IRON_INGOT;
+
+    public void setRepairItem(Material material){
+        repairitem = CraftMagicNumbers.getItem(material);
+    }
+    // Canvas end
+
     @Override
     protected void playStepSound(BlockPos pos, BlockState state) {
         this.playSound(SoundEvents.IRON_GOLEM_STEP, 1.0F, 1.0F);
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftIronGolem.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftIronGolem.java
index 208018981a2a5666c455eb34614b03f617354165..ec5668b56797735f3eeddc36a5afbe8ea33d1680 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftIronGolem.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftIronGolem.java
@@ -1,5 +1,6 @@
 package org.bukkit.craftbukkit.entity;
 
+import org.bukkit.Material;
 import org.bukkit.craftbukkit.CraftServer;
 import org.bukkit.entity.IronGolem;
 
@@ -40,4 +41,41 @@ public class CraftIronGolem extends CraftGolem implements IronGolem {
         getHandle().setSummoner(summoner);
     }
     // Purpur end
+
+    // Canvas start
+    @Override
+    public int getCrackiness(){
+        return this.getHandle().getCrackiness().ordinal();
+    }
+
+    @Override
+    public void setFlingVelocity(double amt){
+        this.getHandle().setFlingVelocity(amt);
+    }
+
+    @Override
+    public double getFlingVelocity(){
+        return this.getHandle().getFlingVelocity();
+    }
+
+    @Override
+    public boolean canRepair() {
+        return this.getHandle().canRepair;
+    }
+
+    @Override
+    public void setCanRepair(boolean can) {
+        this.getHandle().setCanRepair(can);
+    }
+
+    @Override
+    public void offerFlower(final boolean lookingAtVillager) {
+        this.getHandle().offerFlower(lookingAtVillager);
+    }
+
+    @Override
+    public void setRepairItem(final Material material) {
+        this.getHandle().setRepairItem(material);
+    }
+    // Canvas end
 }
