From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris Bauer <jedimastertoothless@hotmail.com>
Date: Thu, 26 Oct 2023 20:54:10 +0000
Subject: [PATCH] Changed to use Sonatype repository


diff --git a/.gitignore b/.gitignore
index 37c3a00659ce21623be07317f4f6a45bf990d799..dffb55df77046746d70ecd6bd95c5df6b9765e2e 100644
--- a/.gitignore
+++ b/.gitignore
@@ -45,3 +45,4 @@ dependency-reduced-pom.xml
 # vs code
 /.vscode
 /.factorypath
+bin/
diff --git a/build.gradle.kts b/build.gradle.kts
index c2dab95be770fffc678409b595c9b3847154628a..0a5eed3e2679655cd5bba01329f72dc190387c91 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -127,9 +127,9 @@ tasks.compileTestJava {
 // Paper end
 
 publishing {
-    publications.create<MavenPublication>("maven") {
-        artifact(tasks.shadowJar)
-    }
+    // publications.create<MavenPublication>("maven") {
+    //     artifact(tasks.shadowJar)
+    // }
 }
 
 relocation {
@@ -244,3 +244,40 @@ tasks.registerRunTask("runDev") {
     classpath(runtimeClasspathForRunDev)
     jvmArgs("-DPaper.isRunDev=true")
 }
+
+// Canvas start
+publishing {
+    publications {
+        create<MavenPublication>("mavenJava") {
+            pom {
+                name.set("Canvas-Server")
+                description.set("Empowering the impossible in Minecraft")
+                url.set("https://github.com/Dueris/Canvas")
+                licenses {
+                    license {
+                        name.set("GNU General Public License 3.0")
+                        url.set("https://www.gnu.org/licenses/gpl-3.0.en.html")
+                    }
+                }
+            }
+            groupId = "io.github.dueris"
+            artifactId = "canvas-server"
+            from(components["java"])
+        }
+    }
+    repositories {
+        maven {
+            name = "sonatype"
+            if (version.toString().endsWith("SNAPSHOT")) {
+                url = uri("https://s01.oss.sonatype.org/content/repositories/snapshots/")
+            } else {
+                url = uri("https://s01.oss.sonatype.org/service/local/staging/deploy/maven2/")
+            }
+            credentials {
+                username=System.getenv("OSSRH_USERNAME")
+                password=System.getenv("OSSRH_PASSWORD")
+            }
+        }
+    }
+}
+// Canvas end
\ No newline at end of file
diff --git a/gradle.properties b/gradle.properties
new file mode 100644
index 0000000000000000000000000000000000000000..bdae19149cb5e173a8538e3682a2efa6b19d6b08
--- /dev/null
+++ b/gradle.properties
@@ -0,0 +1,2 @@
+ossrhUsername=System.getenv("OSSRH_USERNAME")
+ossrhPassword=System.getenv("OSSRH_PASSWORD")
\ No newline at end of file
