From 53e1a57c72953de1bfcc94c3257c724e7fa1f88f Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/navigation/WallClimberNavigation.java b/net/minecraft/world/entity/ai/navigation/WallClimberNavigation.java
index da73c38857960aa2977164dfbe421cbf02673d25..0150997a7338390c1e4c500d588662a81f19c7be 100644
--- a/net/minecraft/world/entity/ai/navigation/WallClimberNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/WallClimberNavigation.java
@@ -44,11 +44,11 @@ public class WallClimberNavigation extends GroundPathNavigation {
             super.tick();
         } else {
             if (this.pathToPosition != null) {
-                if (!this.pathToPosition.closerToCenterThan(this.mob.position(), this.mob.getBbWidth())
+                if (!this.pathToPosition.closerToCenterThan(this.mob.position(), io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaFixes.mc94054 ? Math.max(this.mob.getBbWidth(), 1.0F) : this.mob.getBbWidth()) // Canvas - MC-94054
                     && (
                         !(this.mob.getY() > this.pathToPosition.getY())
                             || !BlockPos.containing(this.pathToPosition.getX(), this.mob.getY(), this.pathToPosition.getZ())
-                                .closerToCenterThan(this.mob.position(), this.mob.getBbWidth())
+                                .closerToCenterThan(this.mob.position(), io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaFixes.mc94054 ? Math.max(this.mob.getBbWidth(), 1.0F) : this.mob.getBbWidth()) // Canvas - MC-94054
                     )) {
                     this.mob
                         .getMoveControl()
