From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/navigation/WallClimberNavigation.java b/net/minecraft/world/entity/ai/navigation/WallClimberNavigation.java
index 2b5ba8f01b57b5e54bba4a8624f17e1b550d818e..c8ee59ce4c7e84ec88f9171ca736fa31a34680c2 100644
--- a/net/minecraft/world/entity/ai/navigation/WallClimberNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/WallClimberNavigation.java
@@ -44,11 +44,11 @@ public class WallClimberNavigation extends GroundPathNavigation {
             super.tick();
         } else {
             if (this.pathToPosition != null) {
-                if (!this.pathToPosition.closerToCenterThan(this.mob.position(), this.mob.getBbWidth())
+                if (!this.pathToPosition.closerToCenterThan(this.mob.position(), io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaFixes.mc94054 ? Math.max(this.mob.getBbWidth(), 1.0F) : this.mob.getBbWidth()) // Canvas - MC-94054
                     && (
                         !(this.mob.getY() > this.pathToPosition.getY())
                             || !BlockPos.containing(this.pathToPosition.getX(), this.mob.getY(), this.pathToPosition.getZ())
-                                .closerToCenterThan(this.mob.position(), this.mob.getBbWidth())
+                                .closerToCenterThan(this.mob.position(), io.canvasmc.canvas.GlobalConfiguration.getInstance().vanillaFixes.mc94054 ? Math.max(this.mob.getBbWidth(), 1.0F) : this.mob.getBbWidth()) // Canvas - MC-94054
                     )) {
                     this.mob
                         .getMoveControl()
