--- a/net/minecraft/Util.java
+++ b/net/minecraft/Util.java
@@ -98,7 +_,7 @@
     public static final TracingExecutor DIMENSION_DATA_IO_POOL = makeExtraIoExecutor("Dimension-Data-IO-Worker-"); // Paper - Separate dimension data IO pool
     private static final TracingExecutor DOWNLOAD_POOL = makeIoExecutor("Download-", true);
     // Paper start - don't submit BLOCKING PROFILE LOOKUPS to the world gen thread
-    public static final ExecutorService PROFILE_EXECUTOR = Executors.newFixedThreadPool(2, new java.util.concurrent.ThreadFactory() {
+    public static final ExecutorService PROFILE_EXECUTOR = io.canvasmc.canvas.util.virtual.VirtualThreadUtils.createFixedExecutor(2, io.canvasmc.canvas.Config.INSTANCE.virtualThreads.profileLookupExecutor, new java.util.concurrent.ThreadFactory() { // Canvas - virtual threads
 
         private final AtomicInteger count = new AtomicInteger();
 
