--- a/net/minecraft/Util.java
+++ b/net/minecraft/Util.java
@@ -16,6 +_,7 @@
 import com.mojang.logging.LogUtils;
 import com.mojang.serialization.DataResult;
 import com.mojang.serialization.Dynamic;
+import io.canvasmc.canvas.Config;
 import it.unimi.dsi.fastutil.ints.IntArrayList;
 import it.unimi.dsi.fastutil.objects.Object2IntMap;
 import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
@@ -97,7 +_,7 @@
     public static final TracingExecutor DIMENSION_DATA_IO_POOL = makeExtraIoExecutor("Dimension-Data-IO-Worker-"); // Paper - Separate dimension data IO pool
     private static final TracingExecutor DOWNLOAD_POOL = makeIoExecutor("Download-", true);
     // Paper start - don't submit BLOCKING PROFILE LOOKUPS to the world gen thread
-    public static final ExecutorService PROFILE_EXECUTOR = Executors.newFixedThreadPool(2, new java.util.concurrent.ThreadFactory() {
+    public static final ExecutorService PROFILE_EXECUTOR = Config.INSTANCE.virtualThreads.shouldReplaceProfileLookupPool() ? Executors.newVirtualThreadPerTaskExecutor() : Executors.newFixedThreadPool(2, new java.util.concurrent.ThreadFactory() { // Canvas - virtual threads
 
         private final AtomicInteger count = new AtomicInteger();
 
