--- a/net/minecraft/world/entity/animal/dolphin/Dolphin.java
+++ b/net/minecraft/world/entity/animal/dolphin/Dolphin.java
@@ -496,7 +_,7 @@
 
         @Override
         public boolean canContinueToUse() {
-            return this.player != null && this.player.isSwimming() && this.dolphin.distanceToSqr(this.player) < 256.0;
+            return this.player != null && ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.player) && this.player.isSwimming() && this.dolphin.distanceToSqr(this.player) < 256.0; // Canvas - ensure dolphin can safely access player
         }
 
         @Override
