From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Wed, 17 Dec 2025 17:30:02 -0800
Subject: [PATCH] Rewrite Moonrise Executor


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java
index 2abdaf3da3e4216d32113a0fae09ce2fa4d59aee..b72a7a6950cba2d0b6f4473ab4ff7df81d8b7c58 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java
@@ -190,9 +190,7 @@ public final class ChunkTaskScheduler {
                 // FULL is executed on main.
         );
 
-        for (final ChunkStatus status : parallelCapableStatus) {
-            ((ChunkSystemChunkStatus)status).moonrise$setParallelCapable(true);
-        }
+        // Canvas - rewrite moonrise executor - TODO - check this patch
     }
 
     private static final int[] ACCESS_RADIUS_TABLE_LOAD = new int[ChunkStatus.getStatusList().size()];
diff --git a/com/mojang/datafixers/types/Type.java b/com/mojang/datafixers/types/Type.java
index b05a8f5f1da64d8b821a1423e865bdb4bc764694..e15d4bfb3c96df88316a890265159a1b7dc34f1c 100644
--- a/com/mojang/datafixers/types/Type.java
+++ b/com/mojang/datafixers/types/Type.java
@@ -193,7 +193,7 @@ public abstract class Type<A> implements App<Type.Mu, A> {
             PENDING_REWRITE_CACHE.remove(key);
             return result;
         }
-        return (Optional<RewriteResult<A, ?>>) pending.join();
+        return (Optional<RewriteResult<A, ?>>) io.canvasmc.canvas.util.CFUtil.join(pending); // Canvas - rewrite moonrise executor
     }
 
     public <FT, FR> Type<?> getSetType(final OpticFinder<FT> optic, final Type<FR> newType) {
diff --git a/com/mojang/math/Transformation.java b/com/mojang/math/Transformation.java
index 9b1a6dfbe5028cffa78d297b9621e39e9182dec5..109047724f86a65a60e539cfb1bcf747072363c1 100644
--- a/com/mojang/math/Transformation.java
+++ b/com/mojang/math/Transformation.java
@@ -50,6 +50,7 @@ public final class Transformation {
         } else {
             this.matrix = matrix;
         }
+        this.ensureDecomposed(); // Canvas - rewrite moonrise executor
     }
 
     public Transformation(@Nullable Vector3fc translation, @Nullable Quaternionfc leftRotation, @Nullable Vector3fc scale, @Nullable Quaternionfc rightRotation) {
@@ -59,6 +60,7 @@ public final class Transformation {
         this.scale = (Vector3fc)(scale != null ? scale : new Vector3f(1.0F, 1.0F, 1.0F));
         this.rightRotation = (Quaternionfc)(rightRotation != null ? rightRotation : new Quaternionf());
         this.decomposed = true;
+        this.ensureDecomposed(); // Canvas - rewrite moonrise executor
     }
 
     public static Transformation identity() {
diff --git a/net/minecraft/world/level/levelgen/feature/stateproviders/RandomizedIntStateProvider.java b/net/minecraft/world/level/levelgen/feature/stateproviders/RandomizedIntStateProvider.java
index 1b1de30fbbddc76b14b4b93fbb235ee05741f187..e6466281cbf295c04016f1262f14840eeddac9f3 100644
--- a/net/minecraft/world/level/levelgen/feature/stateproviders/RandomizedIntStateProvider.java
+++ b/net/minecraft/world/level/levelgen/feature/stateproviders/RandomizedIntStateProvider.java
@@ -54,17 +54,20 @@ public class RandomizedIntStateProvider extends BlockStateProvider {
 
     @Override
     public BlockState getState(RandomSource random, BlockPos pos) {
-        BlockState state = this.source.getState(random, pos);
-        if (this.property == null || !state.hasProperty(this.property)) {
-            IntegerProperty integerProperty = findProperty(state, this.propertyName);
-            if (integerProperty == null) {
-                return state;
+        // Canvas start - rewrite moonrise executor
+        BlockState blockState = this.source.getState(random, pos);
+        IntegerProperty propertyLocal = this.property; // used as cache only
+        if (propertyLocal == null || !blockState.hasProperty(propertyLocal)) {
+            IntegerProperty intProperty = findProperty(blockState, this.propertyName);
+            if (intProperty == null) {
+                return blockState;
             }
 
-            this.property = integerProperty;
+            this.property = propertyLocal = intProperty;
         }
 
-        return state.setValue(this.property, this.values.sample(random));
+        return blockState.trySetValue(propertyLocal, Integer.valueOf(this.values.sample(random)));
+        // Canvas end - rewrite moonrise executor
     }
 
     private static @Nullable IntegerProperty findProperty(BlockState state, String propertyName) {
diff --git a/net/minecraft/world/level/levelgen/structure/ScatteredFeaturePiece.java b/net/minecraft/world/level/levelgen/structure/ScatteredFeaturePiece.java
index e2036a80eff3dc1a9ec625880d4aab6ef71d84fa..2046eab8e85d6f48ab9d9c7c9643fffc0c06b1ec 100644
--- a/net/minecraft/world/level/levelgen/structure/ScatteredFeaturePiece.java
+++ b/net/minecraft/world/level/levelgen/structure/ScatteredFeaturePiece.java
@@ -12,7 +12,7 @@ public abstract class ScatteredFeaturePiece extends StructurePiece {
     protected final int width;
     protected final int height;
     protected final int depth;
-    protected int heightPosition = -1;
+    protected volatile int heightPosition = -1; // Canvas - rewrite moonrise executor
 
     protected ScatteredFeaturePiece(StructurePieceType type, int x, int y, int z, int width, int height, int depth, Direction orientation) {
         super(type, 0, StructurePiece.makeBoundingBox(x, y, z, orientation, width, height, depth));
diff --git a/net/minecraft/world/level/levelgen/structure/structures/DesertPyramidPiece.java b/net/minecraft/world/level/levelgen/structure/structures/DesertPyramidPiece.java
index 0d2451a9ade43650dbbcbab69ce0f6e8f69b5aee..0a3baf3036ffc10eb964a35494a0b15d2eeb571f 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/DesertPyramidPiece.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/DesertPyramidPiece.java
@@ -23,29 +23,44 @@ import net.minecraft.world.level.storage.loot.BuiltInLootTables;
 public class DesertPyramidPiece extends ScatteredFeaturePiece {
     public static final int WIDTH = 21;
     public static final int DEPTH = 21;
-    private final boolean[] hasPlacedChest = new boolean[4];
-    private final List<BlockPos> potentialSuspiciousSandWorldPositions = new ArrayList<>();
+    private final java.util.concurrent.atomic.AtomicReferenceArray<Boolean> hasPlacedChestAtomic = new java.util.concurrent.atomic.AtomicReferenceArray<>(new Boolean[4]); // Canvas - rewrite moonrise executor
+    private final java.util.Set<BlockPos> potentialSuspiciousSandWorldPositions = com.google.common.collect.Sets.newConcurrentHashSet(); // Canvas - rewrite moonrise executor
     private BlockPos randomCollapsedRoofPos = BlockPos.ZERO;
+    // Canvas start - rewrite moonrise executor
+    private void onInit() {
+        for(int i = 0; i < this.hasPlacedChestAtomic.length(); ++i) {
+            if (this.hasPlacedChestAtomic.get(i) == null) {
+                this.hasPlacedChestAtomic.set(i, false);
+            }
+        }
+    }
+    // Canvas end - rewrite moonrise executor
 
     public DesertPyramidPiece(RandomSource random, int x, int z) {
         super(StructurePieceType.DESERT_PYRAMID_PIECE, x, 64, z, 21, 15, 21, getRandomHorizontalDirection(random));
+        onInit(); // Canvas - rewrite moonrise executor
     }
 
     public DesertPyramidPiece(CompoundTag tag) {
         super(StructurePieceType.DESERT_PYRAMID_PIECE, tag);
-        this.hasPlacedChest[0] = tag.getBooleanOr("hasPlacedChest0", false);
-        this.hasPlacedChest[1] = tag.getBooleanOr("hasPlacedChest1", false);
-        this.hasPlacedChest[2] = tag.getBooleanOr("hasPlacedChest2", false);
-        this.hasPlacedChest[3] = tag.getBooleanOr("hasPlacedChest3", false);
+        // Canvas start - rewrite moonrise executor
+        this.hasPlacedChestAtomic.set(0, tag.getBooleanOr("hasPlacedChest0", false));
+        this.hasPlacedChestAtomic.set(1, tag.getBooleanOr("hasPlacedChest1", false));
+        this.hasPlacedChestAtomic.set(2, tag.getBooleanOr("hasPlacedChest2", false));
+        this.hasPlacedChestAtomic.set(3, tag.getBooleanOr("hasPlacedChest3", false));
+        onInit();
+        // Canvas end - rewrite moonrise executor
     }
 
     @Override
     protected void addAdditionalSaveData(StructurePieceSerializationContext context, CompoundTag tag) {
         super.addAdditionalSaveData(context, tag);
-        tag.putBoolean("hasPlacedChest0", this.hasPlacedChest[0]);
-        tag.putBoolean("hasPlacedChest1", this.hasPlacedChest[1]);
-        tag.putBoolean("hasPlacedChest2", this.hasPlacedChest[2]);
-        tag.putBoolean("hasPlacedChest3", this.hasPlacedChest[3]);
+        // Canvas start - rewrite moonrise executor
+        tag.putBoolean("hasPlacedChest0", this.hasPlacedChestAtomic.get(0));
+        tag.putBoolean("hasPlacedChest1", this.hasPlacedChestAtomic.get(1));
+        tag.putBoolean("hasPlacedChest2", this.hasPlacedChestAtomic.get(2));
+        tag.putBoolean("hasPlacedChest3", this.hasPlacedChestAtomic.get(3));
+        // Canvas end - rewrite moonrise executor
     }
 
     @Override
@@ -287,12 +302,12 @@ public class DesertPyramidPiece extends ScatteredFeaturePiece {
             this.placeBlock(level, Blocks.CUT_SANDSTONE.defaultBlockState(), 10, -11, 13, box);
 
             for (Direction direction : Direction.Plane.HORIZONTAL) {
-                if (!this.hasPlacedChest[direction.get2DDataValue()]) {
+                if (!this.hasPlacedChestAtomic.get(direction.get2DDataValue())) { // Canvas - rewrite moonrise executor
                     int i4 = direction.getStepX() * 2;
                     int i5 = direction.getStepZ() * 2;
-                    this.hasPlacedChest[direction.get2DDataValue()] = this.createChest(
+                    this.hasPlacedChestAtomic.set(direction.get2DDataValue(), this.createChest( // Canvas - rewrite moonrise executor
                         level, box, random, 10 + i4, -11, 10 + i5, BuiltInLootTables.DESERT_PYRAMID
-                    );
+                    )); // Canvas - rewrite moonrise executor
                 }
             }
 
@@ -419,7 +434,7 @@ public class DesertPyramidPiece extends ScatteredFeaturePiece {
         this.randomCollapsedRoofPos = new BlockPos(this.getWorldX(i1, randomInt), this.getWorldY(y), this.getWorldZ(i1, randomInt));
     }
 
-    public List<BlockPos> getPotentialSuspiciousSandWorldPositions() {
+    public java.util.Set<BlockPos> getPotentialSuspiciousSandWorldPositions() { // Canvas - rewrite moonrise executor
         return this.potentialSuspiciousSandWorldPositions;
     }
 
diff --git a/net/minecraft/world/level/levelgen/structure/structures/JungleTemplePiece.java b/net/minecraft/world/level/levelgen/structure/structures/JungleTemplePiece.java
index d28f1a25e2137955402e41679d1a4220a0136579..c346e9026a093938d6959d3f024b4f3667b767ba 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/JungleTemplePiece.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/JungleTemplePiece.java
@@ -30,10 +30,12 @@ import net.minecraft.world.level.storage.loot.BuiltInLootTables;
 public class JungleTemplePiece extends ScatteredFeaturePiece {
     public static final int WIDTH = 12;
     public static final int DEPTH = 15;
-    private boolean placedMainChest;
-    private boolean placedHiddenChest;
-    private boolean placedTrap1;
-    private boolean placedTrap2;
+    // Canvas start - rewrite moonrise executor
+    private final java.util.concurrent.atomic.AtomicBoolean placedMainChest = new java.util.concurrent.atomic.AtomicBoolean(false);
+    private final java.util.concurrent.atomic.AtomicBoolean placedHiddenChest = new java.util.concurrent.atomic.AtomicBoolean(false);
+    private final java.util.concurrent.atomic.AtomicBoolean placedTrap1 = new java.util.concurrent.atomic.AtomicBoolean(false);
+    private final java.util.concurrent.atomic.AtomicBoolean placedTrap2 = new java.util.concurrent.atomic.AtomicBoolean(false);
+    // Canvas end - rewrite moonrise executor
     private static final JungleTemplePiece.MossStoneSelector STONE_SELECTOR = new JungleTemplePiece.MossStoneSelector();
 
     public JungleTemplePiece(RandomSource random, int x, int z) {
@@ -42,19 +44,23 @@ public class JungleTemplePiece extends ScatteredFeaturePiece {
 
     public JungleTemplePiece(CompoundTag tag) {
         super(StructurePieceType.JUNGLE_PYRAMID_PIECE, tag);
-        this.placedMainChest = tag.getBooleanOr("placedMainChest", false);
-        this.placedHiddenChest = tag.getBooleanOr("placedHiddenChest", false);
-        this.placedTrap1 = tag.getBooleanOr("placedTrap1", false);
-        this.placedTrap2 = tag.getBooleanOr("placedTrap2", false);
+        // Canvas start - rewrite moonrise executor
+        this.placedMainChest.set(tag.getBooleanOr("placedMainChest", false));
+        this.placedHiddenChest.set(tag.getBooleanOr("placedHiddenChest", false));
+        this.placedTrap1.set(tag.getBooleanOr("placedTrap1", false));
+        this.placedTrap2.set(tag.getBooleanOr("placedTrap2", false));
+        // Canvas end - rewrite moonrise executor
     }
 
     @Override
     protected void addAdditionalSaveData(StructurePieceSerializationContext context, CompoundTag tag) {
         super.addAdditionalSaveData(context, tag);
-        tag.putBoolean("placedMainChest", this.placedMainChest);
-        tag.putBoolean("placedHiddenChest", this.placedHiddenChest);
-        tag.putBoolean("placedTrap1", this.placedTrap1);
-        tag.putBoolean("placedTrap2", this.placedTrap2);
+        // Canvas start - rewrite moonrise executor
+        tag.putBoolean("placedMainChest", this.placedMainChest.get());
+        tag.putBoolean("placedHiddenChest", this.placedHiddenChest.get());
+        tag.putBoolean("placedTrap1", this.placedTrap1.get());
+        tag.putBoolean("placedTrap2", this.placedTrap2.get());
+        // Canvas end - rewrite moonrise executor
     }
 
     @Override
@@ -236,8 +242,10 @@ public class JungleTemplePiece extends ScatteredFeaturePiece {
                 box
             );
             this.placeBlock(level, Blocks.MOSSY_COBBLESTONE.defaultBlockState(), 3, -3, 1, box);
-            if (!this.placedTrap1) {
-                this.placedTrap1 = this.createDispenser(level, box, random, 3, -2, 1, Direction.NORTH, BuiltInLootTables.JUNGLE_TEMPLE_DISPENSER);
+            // Canvas start - rewrite moonrise executor
+            if (!this.placedTrap1.get()) {
+                this.placedTrap1.set(this.createDispenser(level, box, random, 3, -2, 1, Direction.NORTH, BuiltInLootTables.JUNGLE_TEMPLE_DISPENSER));
+            // Canvas end - rewrite moonrise executor
             }
 
             this.placeBlock(level, Blocks.VINE.defaultBlockState().setValue(VineBlock.SOUTH, true), 3, -2, 2, box);
@@ -328,14 +336,18 @@ public class JungleTemplePiece extends ScatteredFeaturePiece {
             );
             this.placeBlock(level, Blocks.MOSSY_COBBLESTONE.defaultBlockState(), 9, -3, 4, box);
             this.placeBlock(level, blockState4, 9, -2, 4, box);
-            if (!this.placedTrap2) {
-                this.placedTrap2 = this.createDispenser(level, box, random, 9, -2, 3, Direction.WEST, BuiltInLootTables.JUNGLE_TEMPLE_DISPENSER);
+            // Canvas start - rewrite moonrise executor
+            if (!this.placedTrap2.get()) {
+                this.placedTrap2.set(this.createDispenser(level, box, random, 9, -2, 3, Direction.WEST, BuiltInLootTables.JUNGLE_TEMPLE_DISPENSER));
+            // Canvas end - rewrite moonrise executor
             }
 
             this.placeBlock(level, Blocks.VINE.defaultBlockState().setValue(VineBlock.EAST, true), 8, -1, 3, box);
             this.placeBlock(level, Blocks.VINE.defaultBlockState().setValue(VineBlock.EAST, true), 8, -2, 3, box);
-            if (!this.placedMainChest) {
-                this.placedMainChest = this.createChest(level, box, random, 8, -3, 3, BuiltInLootTables.JUNGLE_TEMPLE);
+            // Canvas start - rewrite moonrise executor
+            if (!this.placedMainChest.get()) {
+                this.placedMainChest.set(this.createChest(level, box, random, 8, -3, 3, BuiltInLootTables.JUNGLE_TEMPLE));
+            // Canvas end - rewrite moonrise executor
             }
 
             this.placeBlock(level, Blocks.MOSSY_COBBLESTONE.defaultBlockState(), 9, -3, 2, box);
@@ -378,8 +390,10 @@ public class JungleTemplePiece extends ScatteredFeaturePiece {
             this.placeBlock(level, Blocks.STICKY_PISTON.defaultBlockState().setValue(PistonBaseBlock.FACING, Direction.WEST), 10, -2, 8, box);
             this.placeBlock(level, Blocks.STICKY_PISTON.defaultBlockState().setValue(PistonBaseBlock.FACING, Direction.WEST), 10, -1, 8, box);
             this.placeBlock(level, Blocks.REPEATER.defaultBlockState().setValue(RepeaterBlock.FACING, Direction.NORTH), 10, -2, 10, box);
-            if (!this.placedHiddenChest) {
-                this.placedHiddenChest = this.createChest(level, box, random, 9, -3, 10, BuiltInLootTables.JUNGLE_TEMPLE);
+            // Canvas start - rewrite moonrise executor
+            if (!this.placedHiddenChest.get()) {
+                this.placedHiddenChest.set(this.createChest(level, box, random, 9, -3, 10, BuiltInLootTables.JUNGLE_TEMPLE));
+            // Canvas end - rewrite moonrise executor
             }
         }
     }
diff --git a/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java b/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
index 008a11d83a9efd5886eb0ace3fb91524ac1d4f34..dc4a6ff2297d4edb6b109681a35b8077249c2261 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
@@ -91,7 +91,7 @@ public class MineshaftPieces {
     public static class MineShaftCorridor extends MineshaftPieces.MineShaftPiece {
         private final boolean hasRails;
         private final boolean spiderCorridor;
-        private boolean hasPlacedSpider;
+        private volatile boolean hasPlacedSpider; // Canvas - rewrite moonrise executor
         private final int numSections;
 
         public MineShaftCorridor(CompoundTag tag) {
@@ -948,7 +948,7 @@ public class MineshaftPieces {
     }
 
     public static class MineShaftRoom extends MineshaftPieces.MineShaftPiece {
-        private final List<BoundingBox> childEntranceBoxes = Lists.newLinkedList();
+        private final List<BoundingBox> childEntranceBoxes = java.util.Collections.synchronizedList(Lists.newLinkedList()); // Canvas - rewrite moonrise executor
 
         public MineShaftRoom(int genDepth, RandomSource random, int x, int z, MineshaftStructure.Type type) {
             super(
diff --git a/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java b/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java
index 205fbd89b0c44b3038f198b13b54bbd016e7da6d..e95bc2aa642bf2f44b325e794b364f5fcb1a8eef 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java
@@ -647,7 +647,7 @@ public class NetherFortressPieces {
         private static final int WIDTH = 5;
         private static final int HEIGHT = 7;
         private static final int DEPTH = 5;
-        private boolean isNeedingChest;
+        private volatile boolean isNeedingChest; // Canvas - rewrite moonrise executor
 
         public CastleSmallCorridorLeftTurnPiece(int genDepth, RandomSource random, BoundingBox box, Direction orientation) {
             super(StructurePieceType.NETHER_FORTRESS_CASTLE_SMALL_CORRIDOR_LEFT_TURN, genDepth, box);
@@ -777,7 +777,7 @@ public class NetherFortressPieces {
         private static final int WIDTH = 5;
         private static final int HEIGHT = 7;
         private static final int DEPTH = 5;
-        private boolean isNeedingChest;
+        private volatile boolean isNeedingChest; // Canvas - rewrite moonrise executor
 
         public CastleSmallCorridorRightTurnPiece(int genDepth, RandomSource random, BoundingBox box, Direction orientation) {
             super(StructurePieceType.NETHER_FORTRESS_CASTLE_SMALL_CORRIDOR_RIGHT_TURN, genDepth, box);
@@ -1014,7 +1014,7 @@ public class NetherFortressPieces {
         private static final int WIDTH = 7;
         private static final int HEIGHT = 8;
         private static final int DEPTH = 9;
-        private boolean hasPlacedSpawner;
+        private volatile boolean hasPlacedSpawner; // Canvas - rewrite moonrise executor
 
         public MonsterThrone(int genDepth, BoundingBox box, Direction orientation) {
             super(StructurePieceType.NETHER_FORTRESS_MONSTER_THRONE, genDepth, box);
@@ -1102,7 +1102,7 @@ public class NetherFortressPieces {
         }
     }
 
-    abstract static class NetherBridgePiece extends StructurePiece {
+    public abstract static class NetherBridgePiece extends StructurePiece { // Canvas - TODO - at
         protected NetherBridgePiece(StructurePieceType type, int genDepth, BoundingBox boundingBox) {
             super(type, genDepth, boundingBox);
         }
@@ -1120,7 +1120,7 @@ public class NetherFortressPieces {
             int i = 0;
 
             for (NetherFortressPieces.PieceWeight pieceWeight : weights) {
-                if (pieceWeight.maxPlaceCount > 0 && pieceWeight.placeCount < pieceWeight.maxPlaceCount) {
+                if (pieceWeight.maxPlaceCount > 0 && pieceWeight.placeCount.get() < pieceWeight.maxPlaceCount) { // Canvas - rewrite moonrise executor
                     flag = true;
                 }
 
@@ -1160,7 +1160,7 @@ public class NetherFortressPieces {
                             pieceWeight, pieces, random, x, y, z, orientation, genDepth
                         );
                         if (netherBridgePiece != null) {
-                            pieceWeight.placeCount++;
+                            pieceWeight.placeCount.set(pieceWeight.placeCount.get() + 1); // Canvas - rewrite moonrise executor
                             startPiece.previousPiece = pieceWeight;
                             if (!pieceWeight.isValid()) {
                                 weights.remove(pieceWeight);
@@ -1392,7 +1392,7 @@ public class NetherFortressPieces {
     static class PieceWeight {
         public final Class<? extends NetherFortressPieces.NetherBridgePiece> pieceClass;
         public final int weight;
-        public int placeCount;
+        public final ThreadLocal<Integer> placeCount = ThreadLocal.withInitial(() -> 0); // Canvas - rewrite moonrise executor
         public final int maxPlaceCount;
         public final boolean allowInRow;
 
@@ -1408,11 +1408,11 @@ public class NetherFortressPieces {
         }
 
         public boolean doPlace(int genDepth) {
-            return this.maxPlaceCount == 0 || this.placeCount < this.maxPlaceCount;
+            return this.maxPlaceCount == 0 || this.placeCount.get() < this.maxPlaceCount; // Canvas - rewrite moonrise executor
         }
 
         public boolean isValid() {
-            return this.maxPlaceCount == 0 || this.placeCount < this.maxPlaceCount;
+            return this.maxPlaceCount == 0 || this.placeCount.get() < this.maxPlaceCount; // Canvas - rewrite moonrise executor
         }
     }
 
@@ -1554,21 +1554,21 @@ public class NetherFortressPieces {
     }
 
     public static class StartPiece extends NetherFortressPieces.BridgeCrossing {
-        NetherFortressPieces.@Nullable PieceWeight previousPiece;
-        final List<NetherFortressPieces.PieceWeight> availableBridgePieces = new ArrayList<>();
-        final List<NetherFortressPieces.PieceWeight> availableCastlePieces = new ArrayList<>();
+        volatile NetherFortressPieces.@Nullable PieceWeight previousPiece; // Canvas - rewrite moonrise executor
+        final List<NetherFortressPieces.PieceWeight> availableBridgePieces = java.util.Collections.synchronizedList(new ArrayList<>()); // Canvas - rewrite moonrise executor
+        final List<NetherFortressPieces.PieceWeight> availableCastlePieces = java.util.Collections.synchronizedList(new ArrayList<>()); // Canvas - rewrite moonrise executor
         public final List<StructurePiece> pendingChildren = Lists.newArrayList();
 
         public StartPiece(RandomSource random, int x, int z) {
             super(x, z, getRandomHorizontalDirection(random));
 
             for (NetherFortressPieces.PieceWeight pieceWeight : NetherFortressPieces.BRIDGE_PIECE_WEIGHTS) {
-                pieceWeight.placeCount = 0;
+                pieceWeight.placeCount.remove(); // Canvas - rewrite moonrise executor
                 this.availableBridgePieces.add(pieceWeight);
             }
 
             for (NetherFortressPieces.PieceWeight pieceWeight : NetherFortressPieces.CASTLE_PIECE_WEIGHTS) {
-                pieceWeight.placeCount = 0;
+                pieceWeight.placeCount.remove(); // Canvas - rewrite moonrise executor
                 this.availableCastlePieces.add(pieceWeight);
             }
         }
diff --git a/net/minecraft/world/level/levelgen/structure/structures/OceanMonumentPieces.java b/net/minecraft/world/level/levelgen/structure/structures/OceanMonumentPieces.java
index e24b8556dce62bc38d4e22beb15276a5f8475ad7..1aa465bb35106201f909258a75a098b71da0bf1f 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/OceanMonumentPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/OceanMonumentPieces.java
@@ -160,8 +160,8 @@ public class OceanMonumentPieces {
         private static final int DEPTH = 58;
         public static final int BIOME_RANGE_CHECK = 29;
         private static final int TOP_POSITION = 61;
-        private OceanMonumentPieces.RoomDefinition sourceRoom;
-        private OceanMonumentPieces.RoomDefinition coreRoom;
+        private volatile OceanMonumentPieces.RoomDefinition sourceRoom; // Canvas - rewrite moonrise executor
+        private volatile OceanMonumentPieces.RoomDefinition coreRoom; // Canvas - rewrite moonrise executor
         private final List<OceanMonumentPieces.OceanMonumentPiece> childPieces = Lists.newArrayList();
 
         public MonumentBuilding(RandomSource random, int x, int z, Direction orientation) {
@@ -1897,9 +1897,9 @@ public class OceanMonumentPieces {
         final int index;
         final OceanMonumentPieces.RoomDefinition[] connections = new OceanMonumentPieces.RoomDefinition[6];
         final boolean[] hasOpening = new boolean[6];
-        boolean claimed;
-        boolean isSource;
-        private int scanIndex;
+        volatile boolean claimed; // Canvas - rewrite moonrise executor
+        volatile boolean isSource; // Canvas - rewrite moonrise executor
+        private volatile int scanIndex; // Canvas - rewrite moonrise executor
 
         public RoomDefinition(int index) {
             this.index = index;
diff --git a/net/minecraft/world/level/levelgen/structure/structures/StrongholdPieces.java b/net/minecraft/world/level/levelgen/structure/structures/StrongholdPieces.java
index dbb02512abecd64a30b9be1c4f1251b403343d84..73563d4322bb352fc717ab082fe00e11850de7a5 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/StrongholdPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/StrongholdPieces.java
@@ -66,32 +66,32 @@ public class StrongholdPieces {
             }
         }
     };
-    private static List<StrongholdPieces.PieceWeight> currentPieces;
-    static @Nullable Class<? extends StrongholdPieces.StrongholdPiece> imposedPiece;
-    private static int totalWeight;
+    private static final ThreadLocal<List<PieceWeight>> currentPieces = new ThreadLocal<List<PieceWeight>>(); // Canvas - rewrite moonrise executor
+    static final ThreadLocal<Class<? extends StrongholdPiece>> imposedPiece = new ThreadLocal<Class<? extends StrongholdPiece>>(); // Canvas - rewrite moonrise executor
+    private static final ThreadLocal<Integer> totalWeight = ThreadLocal.withInitial(() -> 0); // Canvas - rewrite moonrise executor
     static final StrongholdPieces.SmoothStoneSelector SMOOTH_STONE_SELECTOR = new StrongholdPieces.SmoothStoneSelector();
 
     public static void resetPieces() {
-        currentPieces = Lists.newArrayList();
+        currentPieces.set(Lists.newArrayList()); // Canvas - rewrite moonrise executor
 
         for (StrongholdPieces.PieceWeight pieceWeight : STRONGHOLD_PIECE_WEIGHTS) {
-            pieceWeight.placeCount = 0;
-            currentPieces.add(pieceWeight);
+            pieceWeight.placeCount.set(0); // Canvas - rewrite moonrise executor
+            currentPieces.get().add(pieceWeight); // Canvas - rewrite moonrise executor
         }
 
-        imposedPiece = null;
+        imposedPiece.remove(); // Canvas - rewrite moonrise executor
     }
 
     private static boolean updatePieceWeight() {
         boolean flag = false;
-        totalWeight = 0;
+        totalWeight.set(0); // Canvas - rewrite moonrise executor
 
-        for (StrongholdPieces.PieceWeight pieceWeight : currentPieces) {
-            if (pieceWeight.maxPlaceCount > 0 && pieceWeight.placeCount < pieceWeight.maxPlaceCount) {
+        for (StrongholdPieces.PieceWeight pieceWeight : currentPieces.get()) { // Canvas - rewrite moonrise executor
+            if (pieceWeight.maxPlaceCount > 0 && pieceWeight.placeCount.get() < pieceWeight.maxPlaceCount) { // Canvas - rewrite moonrise executor
                 flag = true;
             }
 
-            totalWeight = totalWeight + pieceWeight.weight;
+            totalWeight.set(totalWeight.get() + pieceWeight.weight); // Canvas - rewrite moonrise executor
         }
 
         return flag;
@@ -141,9 +141,9 @@ public class StrongholdPieces {
         if (!updatePieceWeight()) {
             return null;
         } else {
-            if (imposedPiece != null) {
-                StrongholdPieces.StrongholdPiece strongholdPiece = findAndCreatePieceFactory(imposedPiece, pieces, random, x, y, z, direction, genDepth);
-                imposedPiece = null;
+            if (imposedPiece.get() != null) { // Canvas - rewrite moonrise executor
+                StrongholdPieces.StrongholdPiece strongholdPiece = findAndCreatePieceFactory(imposedPiece.get(), pieces, random, x, y, z, direction, genDepth); // Canvas - rewrite moonrise executor
+                imposedPiece.remove(); // Canvas - rewrite moonrise executor
                 if (strongholdPiece != null) {
                     return strongholdPiece;
                 }
@@ -153,9 +153,9 @@ public class StrongholdPieces {
 
             while (i < 5) {
                 i++;
-                int randomInt = random.nextInt(totalWeight);
+                int randomInt = random.nextInt(totalWeight.get()); // Canvas - rewrite moonrise executor
 
-                for (StrongholdPieces.PieceWeight pieceWeight : currentPieces) {
+                for (StrongholdPieces.PieceWeight pieceWeight : currentPieces.get()) { // Canvas - rewrite moonrise executor
                     randomInt -= pieceWeight.weight;
                     if (randomInt < 0) {
                         if (!pieceWeight.doPlace(genDepth) || pieceWeight == piece.previousPiece) {
@@ -166,10 +166,10 @@ public class StrongholdPieces {
                             pieceWeight.pieceClass, pieces, random, x, y, z, direction, genDepth
                         );
                         if (strongholdPiece1 != null) {
-                            pieceWeight.placeCount++;
+                            pieceWeight.placeCount.set(pieceWeight.placeCount.get() + 1); // Canvas - rewrite moonrise executor
                             piece.previousPiece = pieceWeight;
                             if (!pieceWeight.isValid()) {
-                                currentPieces.remove(pieceWeight);
+                                currentPieces.get().remove(pieceWeight); // Canvas - rewrite moonrise executor
                             }
 
                             return strongholdPiece1;
@@ -205,7 +205,7 @@ public class StrongholdPieces {
         private static final int WIDTH = 5;
         private static final int HEIGHT = 5;
         private static final int DEPTH = 7;
-        private boolean hasPlacedChest;
+        private volatile boolean hasPlacedChest; // Canvas - rewrite moonrise executor
 
         public ChestCorridor(int genDepth, RandomSource random, BoundingBox box, Direction orientation) {
             super(StructurePieceType.STRONGHOLD_CHEST_CORRIDOR, genDepth, box);
@@ -691,7 +691,7 @@ public class StrongholdPieces {
     static class PieceWeight {
         public final Class<? extends StrongholdPieces.StrongholdPiece> pieceClass;
         public final int weight;
-        public int placeCount;
+        public final ThreadLocal<Integer> placeCount = ThreadLocal.withInitial(() -> 0); // Canvas - rewrite moonrise executor
         public final int maxPlaceCount;
 
         public PieceWeight(Class<? extends StrongholdPieces.StrongholdPiece> pieceClass, int weight, int maxPlaceCount) {
@@ -701,11 +701,11 @@ public class StrongholdPieces {
         }
 
         public boolean doPlace(int genDepth) {
-            return this.maxPlaceCount == 0 || this.placeCount < this.maxPlaceCount;
+            return this.maxPlaceCount == 0 || this.placeCount.get() < this.maxPlaceCount; // Canvas - rewrite moonrise executor
         }
 
         public boolean isValid() {
-            return this.maxPlaceCount == 0 || this.placeCount < this.maxPlaceCount;
+            return this.maxPlaceCount == 0 || this.placeCount.get() < this.maxPlaceCount; // Canvas - rewrite moonrise executor
         }
     }
 
@@ -713,7 +713,7 @@ public class StrongholdPieces {
         protected static final int WIDTH = 11;
         protected static final int HEIGHT = 8;
         protected static final int DEPTH = 16;
-        private boolean hasPlacedSpawner;
+        private volatile boolean hasPlacedSpawner; // Canvas - rewrite moonrise executor
 
         public PortalRoom(int genDepth, BoundingBox box, Direction orientation) {
             super(StructurePieceType.STRONGHOLD_PORTAL_ROOM, genDepth, box);
@@ -1175,7 +1175,7 @@ public class StrongholdPieces {
         @Override
         public void addChildren(StructurePiece piece, StructurePieceAccessor pieces, RandomSource random) {
             if (this.isSource) {
-                StrongholdPieces.imposedPiece = StrongholdPieces.FiveCrossing.class;
+                StrongholdPieces.imposedPiece.set(StrongholdPieces.FiveCrossing.class); // Canvas - rewrite moonrise executor
             }
 
             this.generateSmallDoorChildForward((StrongholdPieces.StartPiece)piece, pieces, random, 1, 1);
@@ -1224,9 +1224,9 @@ public class StrongholdPieces {
     }
 
     public static class StartPiece extends StrongholdPieces.StairsDown {
-        public StrongholdPieces.@Nullable PieceWeight previousPiece;
-        public StrongholdPieces.@Nullable PortalRoom portalRoomPiece;
-        public final List<StructurePiece> pendingChildren = Lists.newArrayList();
+        public volatile StrongholdPieces.@Nullable PieceWeight previousPiece; // Canvas - rewrite moonrise executor
+        public volatile StrongholdPieces.@Nullable PortalRoom portalRoomPiece; // Canvas - rewrite moonrise executor
+        public final List<StructurePiece> pendingChildren = java.util.Collections.synchronizedList(Lists.newArrayList()); // Canvas - rewrite moonrise executor
 
         public StartPiece(RandomSource random, int x, int z) {
             super(StructurePieceType.STRONGHOLD_START, 0, x, z, getRandomHorizontalDirection(random));
@@ -1377,7 +1377,7 @@ public class StrongholdPieces {
         }
     }
 
-    abstract static class StrongholdPiece extends StructurePiece {
+    public abstract static class StrongholdPiece extends StructurePiece { // Canvas - TODO - at
         protected StrongholdPieces.StrongholdPiece.SmallDoorType entryDoor = StrongholdPieces.StrongholdPiece.SmallDoorType.OPENING;
 
         protected StrongholdPiece(StructurePieceType type, int genDepth, BoundingBox boundingBox) {
diff --git a/net/minecraft/world/level/levelgen/structure/structures/SwampHutPiece.java b/net/minecraft/world/level/levelgen/structure/structures/SwampHutPiece.java
index 7e03e99cc177f59d606e9ac21e95790968602308..e9858957769b54041890bfe148c03a4df2b0acbe 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/SwampHutPiece.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/SwampHutPiece.java
@@ -23,8 +23,8 @@ import net.minecraft.world.level.levelgen.structure.pieces.StructurePieceSeriali
 import net.minecraft.world.level.levelgen.structure.pieces.StructurePieceType;
 
 public class SwampHutPiece extends ScatteredFeaturePiece {
-    private boolean spawnedWitch;
-    private boolean spawnedCat;
+    private volatile boolean spawnedWitch; // Canvas - rewrite moonrise executor
+    private volatile boolean spawnedCat; // Canvas - rewrite moonrise executor
 
     public SwampHutPiece(RandomSource random, int x, int z) {
         super(StructurePieceType.SWAMPLAND_HUT, x, 64, z, 7, 7, 9, getRandomHorizontalDirection(random));
diff --git a/net/minecraft/world/level/levelgen/structure/structures/WoodlandMansionPieces.java b/net/minecraft/world/level/levelgen/structure/structures/WoodlandMansionPieces.java
index c9029e15218063789fa2ac280e6fd1398ecf504a..7333628f4686338a3abddc56c489e9e94339c8da 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/WoodlandMansionPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/WoodlandMansionPieces.java
@@ -127,7 +127,7 @@ public class WoodlandMansionPieces {
             int i = 11;
             this.entranceX = 7;
             this.entranceY = 4;
-            this.baseGrid = new WoodlandMansionPieces.SimpleGrid(DEFAULT_SIZE, DEFAULT_SIZE, BLOCKED);
+            this.baseGrid = new io.canvasmc.canvas.util.chunk.ConcurrentFlagMatrix(DEFAULT_SIZE, DEFAULT_SIZE, BLOCKED); // Canvas - rewrite moonrise executor
             this.baseGrid.set(this.entranceX, this.entranceY, this.entranceX + 1, this.entranceY + 1, START_ROOM);
             this.baseGrid.set(this.entranceX - 1, this.entranceY, this.entranceX - 1, this.entranceY + 1, ROOM);
             this.baseGrid.set(this.entranceX + 2, this.entranceY - 2, this.entranceX + 3, this.entranceY + 3, BLOCKED);
@@ -146,14 +146,14 @@ public class WoodlandMansionPieces {
             }
 
             this.floorRooms = new WoodlandMansionPieces.SimpleGrid[3];
-            this.floorRooms[0] = new WoodlandMansionPieces.SimpleGrid(DEFAULT_SIZE, DEFAULT_SIZE, BLOCKED);
-            this.floorRooms[1] = new WoodlandMansionPieces.SimpleGrid(DEFAULT_SIZE, DEFAULT_SIZE, BLOCKED);
-            this.floorRooms[2] = new WoodlandMansionPieces.SimpleGrid(DEFAULT_SIZE, DEFAULT_SIZE, BLOCKED);
+            this.floorRooms[0] = new io.canvasmc.canvas.util.chunk.ConcurrentFlagMatrix(DEFAULT_SIZE, DEFAULT_SIZE, BLOCKED); // Canvas - rewrite moonrise executor
+            this.floorRooms[1] = new io.canvasmc.canvas.util.chunk.ConcurrentFlagMatrix(DEFAULT_SIZE, DEFAULT_SIZE, BLOCKED); // Canvas - rewrite moonrise executor
+            this.floorRooms[2] = new io.canvasmc.canvas.util.chunk.ConcurrentFlagMatrix(DEFAULT_SIZE, DEFAULT_SIZE, BLOCKED); // Canvas - rewrite moonrise executor
             this.identifyRooms(this.baseGrid, this.floorRooms[0]);
             this.identifyRooms(this.baseGrid, this.floorRooms[1]);
             this.floorRooms[0].set(this.entranceX + 1, this.entranceY, this.entranceX + 1, this.entranceY + 1, 8388608);
             this.floorRooms[1].set(this.entranceX + 1, this.entranceY, this.entranceX + 1, this.entranceY + 1, 8388608);
-            this.thirdFloorGrid = new WoodlandMansionPieces.SimpleGrid(this.baseGrid.width, this.baseGrid.height, BLOCKED);
+            this.thirdFloorGrid = new io.canvasmc.canvas.util.chunk.ConcurrentFlagMatrix(this.baseGrid.width, this.baseGrid.height, BLOCKED); // Canvas - rewrite moonrise executor
             this.setupThirdFloor();
             this.identifyRooms(this.thirdFloorGrid, this.floorRooms[2]);
         }
@@ -1147,9 +1147,9 @@ public class WoodlandMansionPieces {
     }
 
     static class PlacementData {
-        public Rotation rotation;
-        public BlockPos position;
-        public String wallType;
+        public volatile Rotation rotation; // Canvas - rewrite moonrise executor
+        public volatile BlockPos position; // Canvas - rewrite moonrise executor
+        public volatile String wallType; // Canvas - rewrite moonrise executor
     }
 
     static class SecondFloorRoomCollection extends WoodlandMansionPieces.FloorRoomCollection {
@@ -1189,7 +1189,7 @@ public class WoodlandMansionPieces {
         }
     }
 
-    static class SimpleGrid {
+    public static class SimpleGrid { // Canvas - TODO - at
         private final int[][] grid;
         final int width;
         final int height;
diff --git a/net/minecraft/world/level/levelgen/structure/templatesystem/StructurePlaceSettings.java b/net/minecraft/world/level/levelgen/structure/templatesystem/StructurePlaceSettings.java
index 088540c7acb585eda11c4b556909c5b66cd11160..887c4f9a98c8c7725598e5a187c187ddad275fec 100644
--- a/net/minecraft/world/level/levelgen/structure/templatesystem/StructurePlaceSettings.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/StructurePlaceSettings.java
@@ -20,7 +20,7 @@ public class StructurePlaceSettings {
     private LiquidSettings liquidSettings = LiquidSettings.APPLY_WATERLOGGING;
     private @Nullable RandomSource random;
     public int palette = -1; // CraftBukkit - Set initial value so we know if the palette has been set forcefully
-    private final List<StructureProcessor> processors = Lists.newArrayList();
+    private final List<StructureProcessor> processors = java.util.Collections.synchronizedList(Lists.newArrayList()); // Canvas - rewrite moonrise executor
     private boolean knownShape;
     private boolean finalizeEntities;
 
diff --git a/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java b/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
index dd804657ed9935155aaaacfd5a1738df39d016e8..059fa5d5ac4d2de20e3665623ae0c59f24742be6 100644
--- a/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
@@ -76,8 +76,8 @@ public class StructureTemplate {
     public static final String ENTITY_TAG_BLOCKPOS = "blockPos";
     public static final String ENTITY_TAG_NBT = "nbt";
     public static final String SIZE_TAG = "size";
-    public final List<StructureTemplate.Palette> palettes = Lists.newArrayList();
-    public final List<StructureTemplate.StructureEntityInfo> entityInfoList = Lists.newArrayList();
+    public final List<StructureTemplate.Palette> palettes = java.util.Collections.synchronizedList(Lists.newArrayList()); // Canvas - rewrite moonrise executor
+    public final List<StructureTemplate.StructureEntityInfo> entityInfoList = java.util.Collections.synchronizedList(Lists.newArrayList()); // Canvas - rewrite moonrise executor
     private Vec3i size = Vec3i.ZERO;
     private String author = "?";
     // CraftBukkit start - data containers
