--- a/ca/spottedleaf/moonrise/patches/collisions/CollisionUtil.java
+++ b/ca/spottedleaf/moonrise/patches/collisions/CollisionUtil.java
@@ -2090,7 +_,7 @@
         aabb = aabb.inflate(-COLLISION_EPSILON, -COLLISION_EPSILON, -COLLISION_EPSILON);
         final List<Entity> entities;
         if (entity != null && ((ChunkSystemEntity)entity).moonrise$isHardColliding()) {
-            entities = world.getEntities(entity, aabb, predicate);
+            entities = ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(world, aabb) ? world.getEntities(entity, aabb, predicate) : List.of(); // Canvas - fix entities on edge of region with large bounding boxes crashing the server
         } else {
             entities = ((ChunkSystemEntityGetter)world).moonrise$getHardCollidingEntities(entity, aabb, predicate);
         }
