From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/npc/villager/Villager.java b/net/minecraft/world/entity/npc/villager/Villager.java
index 31e5dc8d33552d86abeed57d354c3223339c8cfb..d986ebb161a45941247d9e1570442b25e8880f1b 100644
--- a/net/minecraft/world/entity/npc/villager/Villager.java
+++ b/net/minecraft/world/entity/npc/villager/Villager.java
@@ -841,7 +841,15 @@ public class Villager extends AbstractVillager implements VillagerDataHolder, Re
 
     private int countFoodPointsInInventory() {
         SimpleContainer inventory = this.getInventory();
-        return FOOD_POINTS.entrySet().stream().mapToInt(entry -> inventory.countItem(entry.getKey()) * entry.getValue()).sum();
+        // Canvas start - remove streams
+        int sum = 0;
+        for (Map.Entry<Item, Integer> entry : FOOD_POINTS.entrySet()) {
+            int i = inventory.countItem(entry.getKey()) * entry.getValue();
+            sum += i;
+        }
+        return sum;
+        // return FOOD_POINTS.entrySet().stream().mapToInt(entry -> inventory.countItem(entry.getKey()) * entry.getValue()).sum();
+        // Canvas end - remove streams
     }
 
     public boolean hasFarmSeeds() {
