From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/io/papermc/paper/threadedregions/RegionizedServer.java b/io/papermc/paper/threadedregions/RegionizedServer.java
index b95f76adf2d37badc4d2dcacf340e569054bea5f..ffc558755c5fb21c190030ddfffe223283aa584a 100644
--- a/io/papermc/paper/threadedregions/RegionizedServer.java
+++ b/io/papermc/paper/threadedregions/RegionizedServer.java
@@ -63,6 +63,14 @@ public final class RegionizedServer {
     }
 
     public void init() {
+        // Canvas start - default respawn dimension config
+        // validate respawn dimension config entry first
+        net.minecraft.resources.ResourceKey<net.minecraft.world.level.Level> respawnDimension = io.canvasmc.canvas.Config.INSTANCE.fetchRespawnDimensionKey();
+        ServerLevel level = MinecraftServer.getServer().getLevel(respawnDimension);
+        if (level == null) {
+            throw new IllegalStateException("Respawn dimension invalid, could not fetch dimension from value '" + respawnDimension + "'");
+        }
+        // Canvas end - default respawn dimension config
         // call init event _before_ scheduling anything
         new RegionizedServerInitEvent().callEvent();
 
@@ -554,4 +562,4 @@ public final class RegionizedServer {
     public static final record WorldLevelData(ServerLevel world, long nonRedstoneGameTime, long dayTime) {
 
     }
-}
\ No newline at end of file
+}
