--- a/net/minecraft/world/level/levelgen/feature/DripstoneClusterFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/DripstoneClusterFeature.java
@@ -3,6 +_,7 @@
 import com.mojang.serialization.Codec;
 import java.util.Optional;
 import java.util.OptionalInt;
+import io.canvasmc.canvas.util.ThreadLocalRandomSource;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.tags.BlockTags;
@@ -28,7 +_,7 @@
         WorldGenLevel worldGenLevel = context.level();
         BlockPos blockPos = context.origin();
         DripstoneClusterConfiguration dripstoneClusterConfiguration = context.config();
-        RandomSource randomSource = context.random();
+        RandomSource randomSource = ThreadLocalRandomSource.INSTANCE; // context.random(); // Canvas - threadlocal ran
         if (!DripstoneUtils.isEmptyOrWater(worldGenLevel, blockPos)) {
             return false;
         } else {
