--- a/net/minecraft/world/level/levelgen/feature/OreFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/OreFeature.java
@@ -71,6 +_,6 @@
     ) {
         int placed = 0;
-        BitSet tested = new BitSet(sizeXZ * sizeY * sizeXZ);
+        BitSet tested = com.ishland.c2me.opts.allocs.CachedOrNewBitsGetter.getCachedOrNewBitSet(sizeXZ * sizeY * sizeXZ); // Leaf - C2ME - Reduce Allocations
         BlockPos.MutableBlockPos orePos = new BlockPos.MutableBlockPos();
         int size = config.size;
         double[] data = new double[size * 4];
