--- a/net/minecraft/world/level/levelgen/feature/TreeFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/TreeFeature.java
@@ -198,6 +_,7 @@
                 if (box.isInside(blockPos1)) {
                     if (i2 != 0) {
                         BlockState blockState = level.getBlockState(blockPos1);
+                        if (!blockState.is(BlockTags.LEAVES)) continue; // Canvas
                         setBlockKnownShape(level, blockPos1, blockState.setValue(BlockStateProperties.DISTANCE, Integer.valueOf(i2)));
                     }
 
