From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/block/entity/vault/VaultSharedData.java b/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
index 877dbe309089df9ffb27ebbe3af527feb75d97a6..c6a7edfe36c3488d633640dee4843121e1bf2332 100644
--- a/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
+++ b/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
@@ -67,11 +67,14 @@ public class VaultSharedData {
     void updateConnectedPlayersWithinRange(
         final ServerLevel serverLevel, final BlockPos pos, final VaultServerData serverData, final VaultConfig config, final double limit
     ) {
-        Set<UUID> currentConnectedPlayers = config.playerDetector()
-            .detect(serverLevel, config.entitySelector(), pos, limit, false)
-            .stream()
-            .filter(uuid -> !serverData.getRewardedPlayers().contains(uuid))
-            .collect(Collectors.toSet());
+        // Canvas start - remove streams
+        Set<UUID> currentConnectedPlayers = new ObjectLinkedOpenHashSet<>();
+        for (UUID uuid : config.playerDetector().detect(serverLevel, config.entitySelector(), pos, limit, false)) {
+            if (!serverData.getRewardedPlayers().contains(uuid)) {
+                currentConnectedPlayers.add(uuid);
+            }
+        }
+        // Canvas end - remove streams
         if (!this.connectedPlayers.equals(currentConnectedPlayers)) {
             this.connectedPlayers = currentConnectedPlayers;
             this.markDirty();
