--- a/com/mojang/math/OctahedralGroup.java
+++ b/com/mojang/math/OctahedralGroup.java
@@ -112,6 +_,7 @@
         this.transformation = new Matrix3f().scaling(invertX ? -1.0F : 1.0F, invertY ? -1.0F : 1.0F, invertZ ? -1.0F : 1.0F);
         this.transformation.mul(permutation.transformation());
         this.initializeRotationDirections(); // Paper - Avoid Lazy Initialization for Enum Fields
+        this.rotate(Direction.UP); // Canvas
     }
 
     private BooleanList packInversions() {
