--- a/net/minecraft/world/level/chunk/storage/SerializableChunkData.java
+++ b/net/minecraft/world/level/chunk/storage/SerializableChunkData.java
@@ -574,7 +_,7 @@
                 }
             }
 
-            ChunkAccess.PackedTicks ticksForSerialization = chunk.getTicksForSerialization(level.getGameTime());
+            ChunkAccess.PackedTicks ticksForSerialization = chunk.getTicksForSerialization(level.getRedstoneGameTime()); // Canvas - Threaded Regions
             ShortList[] lists = Arrays.stream(chunk.getPostProcessing())
                 .map(list3 -> list3 != null ? new ShortArrayList(list3) : null)
                 .toArray(ShortList[]::new);
