--- a/src/main/java/ca/spottedleaf/moonrise/common/misc/Delayed26WayDistancePropagator3D.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/misc/Delayed26WayDistancePropagator3D.java
@@ -15,7 +_,7 @@
 
     // Generally updates to positions are made close to other updates, so we link to decrease cache misses when
     // propagating updates
-    protected final LongLinkedOpenHashSet updatedSources = new LongLinkedOpenHashSet();
+    protected final it.unimi.dsi.fastutil.longs.LongSet updatedSources = it.unimi.dsi.fastutil.longs.LongSets.synchronize(new LongLinkedOpenHashSet()); // Canvas - synchronize
 
     @FunctionalInterface
     public static interface LevelChangeCallback {
@@ -94,29 +_,29 @@
 
     protected final void addToIncreaseWorkQueue(final long coordinate, final byte level) {
         final Delayed8WayDistancePropagator2D.WorkQueue queue = this.levelIncreaseWorkQueues[level];
-        queue.queuedCoordinates.enqueue(coordinate);
-        queue.queuedLevels.enqueue(level);
+        queue.queuedCoordinates.add(coordinate); // Canvas
+        queue.queuedLevels.add(level); // Canvas
 
         this.levelIncreaseWorkQueueBitset |= (1L << level);
     }
 
     protected final void addToIncreaseWorkQueue(final long coordinate, final byte index, final byte level) {
         final Delayed8WayDistancePropagator2D.WorkQueue queue = this.levelIncreaseWorkQueues[index];
-        queue.queuedCoordinates.enqueue(coordinate);
-        queue.queuedLevels.enqueue(level);
+        queue.queuedCoordinates.add(coordinate); // Canvas
+        queue.queuedLevels.add(level); // Canvas
 
         this.levelIncreaseWorkQueueBitset |= (1L << index);
     }
 
     protected final void addToRemoveWorkQueue(final long coordinate, final byte level) {
         final Delayed8WayDistancePropagator2D.WorkQueue queue = this.levelRemoveWorkQueues[level];
-        queue.queuedCoordinates.enqueue(coordinate);
-        queue.queuedLevels.enqueue(level);
+        queue.queuedCoordinates.add(coordinate); // Canvas
+        queue.queuedLevels.add(level); // Canvas
 
         this.levelRemoveWorkQueueBitset |= (1L << level);
     }
 
-    public boolean propagateUpdates() {
+    public synchronized boolean propagateUpdates() { // Canvas - synchronize
         if (this.updatedSources.isEmpty()) {
             return false;
         }
@@ -157,15 +_,15 @@
         return ret;
     }
 
-    protected void propagateIncreases() {
+    protected synchronized void propagateIncreases() { // Canvas - synchronize
         for (int queueIndex = 63 ^ Long.numberOfLeadingZeros(this.levelIncreaseWorkQueueBitset);
              this.levelIncreaseWorkQueueBitset != 0L;
              this.levelIncreaseWorkQueueBitset ^= (1L << queueIndex), queueIndex = 63 ^ Long.numberOfLeadingZeros(this.levelIncreaseWorkQueueBitset)) {
 
             final Delayed8WayDistancePropagator2D.WorkQueue queue = this.levelIncreaseWorkQueues[queueIndex];
             while (!queue.queuedLevels.isEmpty()) {
-                final long coordinate = queue.queuedCoordinates.removeFirstLong();
-                byte level = queue.queuedLevels.removeFirstByte();
+                final long coordinate = queue.queuedCoordinates.removeFirst(); // Canvas
+                byte level = queue.queuedLevels.removeFirst(); // Canvas
 
                 final boolean neighbourCheck = level < 0;
 
@@ -226,15 +_,15 @@
         }
     }
 
-    protected void propagateDecreases() {
+    protected synchronized void propagateDecreases() { // Canvas - synchronize
         for (int queueIndex = 63 ^ Long.numberOfLeadingZeros(this.levelRemoveWorkQueueBitset);
              this.levelRemoveWorkQueueBitset != 0L;
              this.levelRemoveWorkQueueBitset ^= (1L << queueIndex), queueIndex = 63 ^ Long.numberOfLeadingZeros(this.levelRemoveWorkQueueBitset)) {
 
             final Delayed8WayDistancePropagator2D.WorkQueue queue = this.levelRemoveWorkQueues[queueIndex];
             while (!queue.queuedLevels.isEmpty()) {
-                final long coordinate = queue.queuedCoordinates.removeFirstLong();
-                final byte level = queue.queuedLevels.removeFirstByte();
+                final long coordinate = queue.queuedCoordinates.removeFirst(); // Canvas
+                final byte level = queue.queuedLevels.removeFirst(); // Canvas
 
                 final byte currentLevel = this.levels.removeIfGreaterOrEqual(coordinate, level);
                 if (currentLevel == 0) {
