--- a/src/main/java/ca/spottedleaf/moonrise/common/misc/Delayed8WayDistancePropagator2D.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/misc/Delayed8WayDistancePropagator2D.java
@@ -561,7 +_,7 @@
         }
 
         // copied from superclass
-        private int find(final long k) {
+        private synchronized int find(final long k) { // Canvas - synchronize
             if (k == 0L) {
                 return this.containsNullKey ? this.n : -(this.n + 1);
             } else {
@@ -585,7 +_,7 @@
         }
 
         // copied from superclass
-        private void insert(final int pos, final long k, final byte v) {
+        private synchronized void insert(final int pos, final long k, final byte v) { // Canvas - synchronize
             if (pos == this.n) {
                 this.containsNullKey = true;
             }
@@ -598,7 +_,7 @@
         }
 
         // copied from superclass
-        public byte putIfGreater(final long key, final byte value) {
+        public synchronized byte putIfGreater(final long key, final byte value) { // Canvas - synchronize
             final int pos = this.find(key);
             if (pos < 0) {
                 if (this.defRetValue < value) {
@@ -616,7 +_,7 @@
         }
 
         // copied from superclass
-        private void removeEntry(final int pos) {
+        private synchronized void removeEntry(final int pos) { // Canvas - synchronize
             --this.size;
             this.shiftKeys(pos);
             if (this.n > this.minN && this.size < this.maxFill / 4 && this.n > 16) {
@@ -625,7 +_,7 @@
         }
 
         // copied from superclass
-        private void removeNullEntry() {
+        private synchronized void removeNullEntry() { // Canvas - synchronize
             this.containsNullKey = false;
             --this.size;
             if (this.n > this.minN && this.size < this.maxFill / 4 && this.n > 16) {
@@ -634,7 +_,7 @@
         }
 
         // copied from superclass
-        public byte removeIfGreaterOrEqual(final long key, final byte value) {
+        public synchronized byte removeIfGreaterOrEqual(final long key, final byte value) { // Canvas - synchronize
             if (key == 0L) {
                 if (!this.containsNullKey) {
                     return this.defRetValue;
