--- a/src/main/java/ca/spottedleaf/moonrise/common/misc/PositionCountingAreaMap.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/misc/PositionCountingAreaMap.java
@@ -3,12 +_,14 @@
 import ca.spottedleaf.concurrentutil.util.IntPairUtil;
 import it.unimi.dsi.fastutil.longs.Long2IntOpenHashMap;
 import it.unimi.dsi.fastutil.longs.LongSet;
+import it.unimi.dsi.fastutil.objects.Reference2ReferenceMap;
+import it.unimi.dsi.fastutil.objects.Reference2ReferenceMaps;
 import it.unimi.dsi.fastutil.objects.Reference2ReferenceOpenHashMap;
 import it.unimi.dsi.fastutil.objects.ReferenceSet;
 
 public final class PositionCountingAreaMap<T> {
 
-    private final Reference2ReferenceOpenHashMap<T, PositionCounter> counters = new Reference2ReferenceOpenHashMap<>();
+    private final Reference2ReferenceMap<T, PositionCounter> counters = Reference2ReferenceMaps.synchronize(new Reference2ReferenceOpenHashMap<>()); // Canvas - synchronize
     private final Long2IntOpenHashMap positions = new Long2IntOpenHashMap();
 
     public ReferenceSet<T> getObjects() {
