--- a/src/main/java/ca/spottedleaf/moonrise/common/misc/PositionCountingAreaMap.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/misc/PositionCountingAreaMap.java
@@ -3,32 +_,43 @@
 import ca.spottedleaf.concurrentutil.util.IntPairUtil;
 import it.unimi.dsi.fastutil.longs.Long2IntOpenHashMap;
 import it.unimi.dsi.fastutil.longs.LongSet;
+import it.unimi.dsi.fastutil.objects.Reference2ReferenceMap;
+import it.unimi.dsi.fastutil.objects.Reference2ReferenceMaps;
 import it.unimi.dsi.fastutil.objects.Reference2ReferenceOpenHashMap;
 import it.unimi.dsi.fastutil.objects.ReferenceSet;
 
 public final class PositionCountingAreaMap<T> {
 
-    private final Reference2ReferenceOpenHashMap<T, PositionCounter> counters = new Reference2ReferenceOpenHashMap<>();
+    private final Reference2ReferenceMap<T, PositionCounter> counters = Reference2ReferenceMaps.synchronize(new Reference2ReferenceOpenHashMap<>()); // Canvas - synchronize
     private final Long2IntOpenHashMap positions = new Long2IntOpenHashMap();
+    private final Object lock = new Object(); // Canvas - lock
 
     public ReferenceSet<T> getObjects() {
         return this.counters.keySet();
     }
 
     public LongSet getPositions() {
+        synchronized (lock) { // Canvas - lock
         return this.positions.keySet();
+        } // Canvas - lock
     }
 
     public int getTotalPositions() {
+        synchronized (lock) { // Canvas - lock
         return this.positions.size();
+        } // Canvas - lock
     }
 
     public boolean hasObjectsNear(final int toX, final int toZ) {
+        synchronized (lock) { // Canvas - lock
         return this.positions.containsKey(IntPairUtil.key(toX, toZ));
+        } // Canvas - lock
     }
 
     public int getObjectsNear(final int toX, final int toZ) {
+        synchronized (lock) { // Canvas - lock
         return this.positions.get(IntPairUtil.key(toX, toZ));
+        } // Canvas - lock
     }
 
     public boolean add(final T parameter, final int toX, final int toZ, final int distance) {
@@ -85,15 +_,19 @@
 
         @Override
         protected void addCallback(final T parameter, final int toX, final int toZ) {
+            synchronized (lock) { // Canvas - lock
             PositionCountingAreaMap.this.positions.addTo(IntPairUtil.key(toX, toZ), 1);
+            } // Canvas - lock
         }
 
         @Override
         protected void removeCallback(final T parameter, final int toX, final int toZ) {
+            synchronized (lock) { // Canvas - lock
             final long key = IntPairUtil.key(toX, toZ);
             if (PositionCountingAreaMap.this.positions.addTo(key, -1) == 1) {
                 PositionCountingAreaMap.this.positions.remove(key);
             }
+            } // Canvas - lock
         }
     }
 }
