From f76ca63fb77ad4d368205b19bd95b232db44ca70 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas paperServer File Patches


diff --git a/src/main/java/ca/spottedleaf/moonrise/common/misc/SingleUserAreaMap.java b/src/main/java/ca/spottedleaf/moonrise/common/misc/SingleUserAreaMap.java
index 7e85bb58b8a5de17fdab9ad3c574f2f80952e1ca..031ed0ed56cea85eb178b20da80fcdf60ae7f66a 100644
--- a/src/main/java/ca/spottedleaf/moonrise/common/misc/SingleUserAreaMap.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/misc/SingleUserAreaMap.java
@@ -86,6 +86,7 @@ public abstract class SingleUserAreaMap<T> {
         if (fromX == NOT_SET) {
             return false;
         }
+        if (fromX == toX && fromZ == toZ && oldViewDistance == newViewDistance) return true; // Leaf - optimize player movement
 
         this.lastChunkX = toX;
         this.lastChunkZ = toZ;
