--- a/src/main/java/ca/spottedleaf/moonrise/common/misc/SingleUserAreaMap.java
+++ b/src/main/java/ca/spottedleaf/moonrise/common/misc/SingleUserAreaMap.java
@@ -86,6 +_,7 @@
         if (fromX == NOT_SET) {
             return false;
         }
+        if (fromX == toX && fromZ == toZ && oldViewDistance == newViewDistance) return true; // Leaf - optimize player movement
 
         this.lastChunkX = toX;
         this.lastChunkZ = toZ;
