From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/server/bossevents/CustomBossEvents.java b/net/minecraft/server/bossevents/CustomBossEvents.java
index 3a506aade575bfcc788b7ad3e33ad76f198bfd11..bc277c9c91d64b6396a85aa5dcec369a7733d87e 100644
--- a/net/minecraft/server/bossevents/CustomBossEvents.java
+++ b/net/minecraft/server/bossevents/CustomBossEvents.java
@@ -18,7 +18,7 @@ import org.slf4j.Logger;
 public class CustomBossEvents {
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final Codec<Map<Identifier, CustomBossEvent.Packed>> EVENTS_CODEC = Codec.unboundedMap(Identifier.CODEC, CustomBossEvent.Packed.CODEC);
-    private final Map<Identifier, CustomBossEvent> events = Maps.newHashMap();
+    private final Map<Identifier, CustomBossEvent> events = Maps.newConcurrentMap(); // Canvas - fix commands with region threading - TODO - do better
 
     public @Nullable CustomBossEvent get(Identifier id) {
         return this.events.get(id);
