From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/server/bossevents/CustomBossEvent.java b/net/minecraft/server/bossevents/CustomBossEvent.java
index 152f08f77c4d8f665380aa02b9241628857e2683..11841a6c0a83895e40278798f6cdf71d84ab3b13 100644
--- a/net/minecraft/server/bossevents/CustomBossEvent.java
+++ b/net/minecraft/server/bossevents/CustomBossEvent.java
@@ -20,7 +20,7 @@ import net.minecraft.world.BossEvent;
 public class CustomBossEvent extends ServerBossEvent {
     private static final int DEFAULT_MAX = 100;
     private final Identifier id;
-    private final Set<UUID> players = Sets.newHashSet();
+    private final Set<UUID> players = Sets.newConcurrentHashSet(); // Canvas - fix commands with region threading - TODO - do better
     private int value;
     private int max = 100;
     // CraftBukkit start
