From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Mon, 28 Jul 2025 02:20:45 +0200
Subject: [PATCH] Rebrand


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index 652ff54e7c50412503725d628bfe72ed03059790..e394a3643ad582cd823ac61b8ce56b8b069c3ab4 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -18,6 +18,16 @@ public interface ServerBuildInfo {
      * The brand id for Paper.
      */
     Key BRAND_PAPER_ID = Key.key("papermc", "paper");
+    // Canvas start - Rebrand
+    /**
+     * The brand id for Folia.
+     */
+    Key BRAND_FOLIA_ID = Key.key("papermc", "folia");
+    /**
+     * The brand id for Canvas.
+     */
+    Key BRAND_CANVAS_ID = Key.key("canvasmc", "canvas");
+    // Canvas end - Rebrand
 
     /**
      * Gets the {@code ServerBuildInfo}.
