From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Sat, 26 Jul 2025 15:50:20 -0700
Subject: [PATCH] Purpur Ender Chest 6 Rows Config


diff --git a/src/main/java/org/bukkit/event/inventory/InventoryType.java b/src/main/java/org/bukkit/event/inventory/InventoryType.java
index 54ff294e64abd692c0f64ba2db7341c9c71436d0..d32550766889103436eedd395267501f11844b92 100644
--- a/src/main/java/org/bukkit/event/inventory/InventoryType.java
+++ b/src/main/java/org/bukkit/event/inventory/InventoryType.java
@@ -166,7 +166,12 @@ public enum InventoryType {
     SMITHING_NEW(4, "Upgrade Gear", MenuType.SMITHING),
     ;
 
-    private final int size;
+    private int size; // Canvas - mutable
+    // Canvas start - add setter
+    public void setDefaultSize(int size) {
+        this.size = size;
+    }
+    // Canvas end - add setter
     private final String title;
     private final MenuType menuType;
     private final boolean isCreatable;
