From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Thu, 23 Feb 2023 20:12:48 -0800
Subject: [PATCH] Region Threading


diff --git a/src/main/java/co/aikar/timings/Timings.java b/src/main/java/co/aikar/timings/Timings.java
index 95b7cdf0677ef71e6885fa78aa5c75bb500f5f53..89017af09ce32e7a66014fc3aeb50155921252a5 100644
--- a/src/main/java/co/aikar/timings/Timings.java
+++ b/src/main/java/co/aikar/timings/Timings.java
@@ -145,6 +145,7 @@ public final class Timings {
      * @param enabled Should timings be reported
      */
     public static void setTimingsEnabled(boolean enabled) {
+        enabled = false; // Folia - region threading - disable timings
         if (enabled && !warnedAboutDeprecationOnEnable) {
             Bukkit.getLogger().severe(PlainTextComponentSerializer.plainText().serialize(deprecationMessage()));
             warnedAboutDeprecationOnEnable = true;
diff --git a/src/main/java/io/papermc/paper/plugin/configuration/PluginMeta.java b/src/main/java/io/papermc/paper/plugin/configuration/PluginMeta.java
index a399f87e925a72aac91b3066317c3efb61c0421a..0100108fd87268d225a60e51eaf87071b359ef6f 100644
--- a/src/main/java/io/papermc/paper/plugin/configuration/PluginMeta.java
+++ b/src/main/java/io/papermc/paper/plugin/configuration/PluginMeta.java
@@ -189,4 +189,12 @@ public interface PluginMeta extends Namespaced {
     default String namespace() {
         return this.getName().toLowerCase(Locale.ROOT);
     }
+    
+    // Folia start - block plugins not marked as supported
+    /**
+     * Returns whether the plugin has been marked to be compatible with regionised threading as provided
+     * by Folia
+     */
+    public boolean isFoliaSupported();
+    // Folia end - block plugins not marked as supported
 }
diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index d083e3e5f00aac3e5c48a248cae665f3517a4b4a..7a6e1697cfe7d0a687ba7b38720a8f3decfa3f14 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -838,6 +838,7 @@ public final class Bukkit {
      * @param save whether to save the chunks before unloading
      * @return true if successful, false otherwise
      */
+    @Deprecated(forRemoval = true) // Canvas - region threading
     public static boolean unloadWorld(@NotNull String name, boolean save) {
         return server.unloadWorld(name, save);
     }
@@ -852,6 +853,7 @@ public final class Bukkit {
      * @param save whether to save the chunks before unloading
      * @return true if successful, false otherwise
      */
+    @Deprecated(forRemoval = true) // Canvas - region threading
     public static boolean unloadWorld(@NotNull World world, boolean save) {
         return server.unloadWorld(world, save);
     }
@@ -2994,6 +2996,42 @@ public final class Bukkit {
         return server.isGlobalTickThread();
     }
     // Paper end - Folia region threading API
+    // Folia start - region TPS API
+    /**
+     * Gets the TPS from the region which owns the specified location, or {@code null} if no region owns
+     * the specified location.
+     *
+     * @param location The location for which to get the TPS
+     * @return TPS (5s, 15s, 1m, 5m, 15m), or null if the region doesn't exist
+     */
+    public static double @Nullable [] getRegionTPS(@NotNull Location location) {
+        return server.getRegionTPS(location);
+    }
+
+    /**
+     * Gets the TPS from the region which owns the specified chunk, or {@code null} if no region owns
+     * the specified location.
+     *
+     * @param chunk - The specified chunk
+     * @return TPS (5s, 15s, 1m, 5m, 15m), or null if the region doesn't exist
+     */
+    public static double @Nullable [] getRegionTPS(@NotNull Chunk chunk) {
+        return server.getRegionTPS(chunk);
+    }
+
+    /**
+     * Gets the TPS from the region which owns the specified chunk, or {@code null} if no region owns
+     * the specified location.
+     *
+     * @param world - World containing the chunk
+     * @param chunkX - X-coordinate of the chunk
+     * @param chunkZ - Z-coordinate of the chunk
+     * @return TPS (5s, 15s, 1m, 5m, 15m), or null if the region doesn't exist
+     */
+    public static double @Nullable [] getRegionTPS(@NotNull World world, int chunkX, int chunkZ) {
+        return server.getRegionTPS(world, chunkX, chunkZ);
+    }
+    // Folia end - region TPS API
 
     /**
      * @deprecated All methods on this class have been deprecated, see the individual methods for replacements.
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 62341f72e8457c5547994bee1f3fda9faade6ce0..9c0163f53ded34517db707216b8b24829c2bde3c 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -761,6 +761,7 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      * @return true if successful, false otherwise
      * @throws IllegalStateException when {@link #isTickingWorlds() isTickingWorlds} is true
      */
+    @Deprecated(forRemoval = true) // Canvas - region threading
     public boolean unloadWorld(@NotNull String name, boolean save);
 
     /**
@@ -774,7 +775,28 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      * @return true if successful, false otherwise
      * @throws IllegalStateException when {@link #isTickingWorlds() isTickingWorlds} is true
      */
+    @Deprecated(forRemoval = true) // Canvas - region threading
     public boolean unloadWorld(@NotNull World world, boolean save);
+    // Canvas start - region threading
+
+    /**
+     * Unloads a world asynchronously with an optional callback result
+     *
+     * @param name Name of the world to unload
+     * @param save whether to save the chunks before unloading
+     * @param callback the result consumer for when the world unload is completed
+     */
+    void unloadWorldAsync(@NotNull String name, boolean save, final @NotNull Consumer<io.canvasmc.canvas.WorldUnloadResult> callback);
+
+    /**
+     * Unloads a world asynchronously with an optional callback result
+     *
+     * @param world the world to unload
+     * @param save whether to save the chunks before unloading
+     * @param callback the result consumer for when the world unload is completed
+     */
+    void unloadWorldAsync(@NotNull World world, boolean save, final @NotNull Consumer<io.canvasmc.canvas.WorldUnloadResult> callback);
+    // Canvas end - region threading
 
     /**
      * Gets the world that players respawn in.
@@ -2753,6 +2775,25 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      * @param entity Specified entity.
      */
     boolean isOwnedByCurrentRegion(@NotNull Entity entity);
+    // Canvas start - expand region threading api
+
+    /**
+     * Returns whether the current thread is ticking a region and that the region being ticked
+     * owns the specified world and area in the world defined by the bounding box provided.
+     * @param world the world
+     * @param aabb the bounding box
+     */
+    boolean isOwnedByCurrentRegion(@NotNull World world, @NotNull org.bukkit.util.BoundingBox aabb);
+
+    /**
+     * Returns whether the current thread is ticking a region and that the region being ticked
+     * owns the specified position that is adjusted as specified by the velocity provided, with a chunk buffer
+     * @param position the initial position
+     * @param velocity the velocity
+     * @param buffer the chunk buffer
+     */
+    boolean isOwnedByCurrentRegion(@NotNull Location position, @NotNull org.bukkit.util.Vector velocity, int buffer);
+    // Canvas end - expand region threading api
 
     /**
      * Returns whether the current thread is ticking the global region.
@@ -2776,4 +2817,92 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     void allowPausing(@NotNull org.bukkit.plugin.Plugin plugin, boolean value);
     // Paper end - API to check if the server is sleeping
+    // Folia start - region TPS API
+    /**
+     * Gets the TPS from the region which owns the specified location, or {@code null} if no region owns
+     * the specified location.
+     *
+     * @param location The location for which to get the TPS
+     * @return TPS (5s, 15s, 1m, 5m, 15m), or null if the region doesn't exist
+     */
+    double @Nullable [] getRegionTPS(final @NotNull Location location);
+
+    /**
+     * Gets the TPS from the region which owns the specified chunk, or {@code null} if no region owns
+     * the specified location.
+     *
+     * @param chunk - The specified chunk
+     * @return TPS (5s, 15s, 1m, 5m, 15m), or null if the region doesn't exist
+     */
+    double @Nullable [] getRegionTPS(final @NotNull Chunk chunk);
+
+    /**
+     * Gets the TPS from the region which owns the specified chunk, or {@code null} if no region owns
+     * the specified location.
+     *
+     * @param world - World containing the chunk
+     * @param chunkX - X-coordinate of the chunk
+     * @param chunkZ - Z-coordinate of the chunk
+     * @return TPS (5s, 15s, 1m, 5m, 15m), or null if the region doesn't exist
+     */
+    double @Nullable [] getRegionTPS(final @NotNull World world, final int chunkX, final int chunkZ);
+    // Folia end - region TPS API
+    // Canvas start - expand region TPS API (pulled from Folia#391)
+
+    /**
+     * Gets the average tick time from the region which owns the specified location, or {@code null} if no region owns
+     * the specified location.
+     *
+     * @param location The location for which to get the TPS
+     * @return Average tick time (5s, 15s, 1m, 5m, 15m), or null if the region doesn't exist
+     */
+    double @Nullable [] getRegionAverageTickTimes(final @NotNull Location location);
+
+    /**
+     * Gets the average tick time from the region which owns the specified chunk, or {@code null} if no region owns
+     * the specified location.
+     *
+     * @param chunk - The specified chunk
+     * @return Average tick time (5s, 15s, 1m, 5m, 15m), or null if the region doesn't exist
+     */
+    double @Nullable [] getRegionAverageTickTimes(final @NotNull Chunk chunk);
+
+    /**
+     * Gets the average tick time from the region which owns the specified chunk, or {@code null} if no region owns
+     * the specified location.
+     *
+     * @param world - World containing the chunk
+     * @param chunkX - X-coordinate of the chunk
+     * @param chunkZ - Z-coordinate of the chunk
+     * @return Average tick time (5s, 15s, 1m, 5m, 15m), or null if the region doesn't exist
+     */
+    double @Nullable [] getRegionAverageTickTimes(final @NotNull World world, final int chunkX, final int chunkZ);
+    // Canvas end - expand region TPS API (pulled from Folia#391)
+    // Canvas start - region threading
+
+    /**
+     * Creates a new regionized data instance
+     * @param supplier the creator for tick region data and a world -> a new regionized data instance
+     * @param callback the split and merge callbacks for the region data
+     * @return a new regionized data instance
+     * @param <T> the generic type of your regionized data
+     */
+    <T> io.canvasmc.canvas.region.RegionTickData.IRegionizedData<@NotNull T> createRegionizedData(final @NotNull java.util.function.BiFunction<io.canvasmc.canvas.region.RegionTickData, org.bukkit.World, T> supplier, final io.canvasmc.canvas.region.RegionTickData.IRegionizedData.IRegionizedCallback<@NotNull T> callback);
+
+    /**
+     * Gets or creates the current thread regionized data based off the {@link io.canvasmc.canvas.region.RegionTickData.IRegionizedData} provided
+     * @param regionizedData the regionized data
+     * @return the fetched or created regionized data
+     * @param <T> the generic type of your regionized data
+     * @throws java.lang.IllegalStateException if not on a tick thread runner, or if the current tick runner doesn't own a region
+     */
+    <T> T getLocalRegionizedData(final @NotNull io.canvasmc.canvas.region.RegionTickData.IRegionizedData<T> regionizedData);
+
+    /**
+     * Gets the region threading tick rate manager. This offers the same capabilities as {@link Bukkit#getServerTickManager()}
+     * but is compatible with Folia's region threading structure
+     * @return the server region threading tick manager
+     */
+    io.canvasmc.canvas.region.RegionThreadingTickManager getRegionThreadingTickRateManager();
+    // Canvas end - region threading
 }
diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index c66cd17fd037b73d8ff04dd2ed328ca25e8b0f1a..d663a8e71dc6514116e6440b6ea4ece7e08d07eb 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -4708,4 +4708,13 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
             }
         }
     }
+    // Canvas start - region threading
+
+    /**
+     * Fetches the non-null regionizer for this world
+     * @return the regionizer
+     */
+    @NotNull
+    io.canvasmc.canvas.region.WorldRegionizer getRegionizer();
+    // Canvas end - region threading
 }
diff --git a/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java b/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
index c14ce1e7d7a5e30cd29b406a412e47104b529511..4a5ac5193bee2107ddc65f37a39046f92e875350 100644
--- a/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
+++ b/src/main/java/org/bukkit/plugin/PluginDescriptionFile.java
@@ -267,6 +267,19 @@ public final class PluginDescriptionFile implements io.papermc.paper.plugin.conf
         return this.paperPluginLoader;
     }
     // Paper end - plugin loader api
+    // Folia start - block plugins not marked as supported
+    private String foliaSupported;
+    private static final String FOLIA_SUPPORTED_KEY = "folia-supported";
+
+    /**
+     * Returns whether the plugin has been marked to be compatible with regionised threading as provided
+     * by Folia
+     */
+    @Override
+    public boolean isFoliaSupported() {
+        return this.foliaSupported != null && this.foliaSupported.equalsIgnoreCase("true");
+    }
+    // Folia end - block plugins not marked as supported
     // Paper start - oh my goddddd
     /**
      * @hidden
@@ -1267,6 +1280,11 @@ public final class PluginDescriptionFile implements io.papermc.paper.plugin.conf
         if (map.get("prefix") != null) {
             prefix = map.get("prefix").toString();
         }
+        // Folia start - block plugins not marked as supported
+        if (map.get(FOLIA_SUPPORTED_KEY) != null) {
+            foliaSupported = map.get(FOLIA_SUPPORTED_KEY).toString();
+        }
+        // Folia end - block plugins not marked as supported
     }
 
     @NotNull
@@ -1343,6 +1361,11 @@ public final class PluginDescriptionFile implements io.papermc.paper.plugin.conf
         if (prefix != null) {
             map.put("prefix", prefix);
         }
+        // Folia start - block plugins not marked as supported
+        if (foliaSupported != null) {
+            map.put(FOLIA_SUPPORTED_KEY, foliaSupported);
+        }
+        // Folia end - block plugins not marked as supported
 
         return map;
     }
diff --git a/src/main/java/org/bukkit/plugin/SimplePluginManager.java b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
index d1a13d8893ffbbd2a3f90ec8f89bd584ef96032e..457e75bbecb70ac3da4c21382f97d68de5a00547 100644
--- a/src/main/java/org/bukkit/plugin/SimplePluginManager.java
+++ b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
@@ -556,9 +556,9 @@ public final class SimplePluginManager implements PluginManager {
             }
 
             try {
-                server.getScheduler().cancelTasks(plugin);
+                server.getAsyncScheduler().cancelTasks(plugin); // Folia - new schedulers
             } catch (Throwable ex) {
-                handlePluginException("Error occurred (in the plugin loader) while cancelling tasks for "
+                handlePluginException("Error occurred (in the plugin loader) while cancelling async tasks for " // Folia - new schedulers
                         + plugin.getDescription().getFullName() + " (Is it up to date?)", ex, plugin); // Paper
             }
 
diff --git a/src/main/java/org/bukkit/scheduler/BukkitScheduler.java b/src/main/java/org/bukkit/scheduler/BukkitScheduler.java
index b65d3aea348bda0f21ed1a5b17bbc05d0df8ca29..09639e03301726fcab650e6c952444f522b553e5 100644
--- a/src/main/java/org/bukkit/scheduler/BukkitScheduler.java
+++ b/src/main/java/org/bukkit/scheduler/BukkitScheduler.java
@@ -7,6 +7,15 @@ import java.util.function.Consumer;
 import org.bukkit.plugin.Plugin;
 import org.jetbrains.annotations.NotNull;
 
+// Folia start - add new schedulers
+/**
+ * @deprecated Use one of {@link io.papermc.paper.threadedregions.scheduler.RegionScheduler},
+ *                        {@link io.papermc.paper.threadedregions.scheduler.AsyncScheduler},
+ *                        {@link io.papermc.paper.threadedregions.scheduler.EntityScheduler},
+ *                        or {@link io.papermc.paper.threadedregions.scheduler.GlobalRegionScheduler}
+ */
+// Folia end - add new schedulers
+@Deprecated
 public interface BukkitScheduler {
 
     /**
