From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Sat, 26 Jul 2025 15:50:20 -0700
Subject: [PATCH] Purpur Ender Chest 6 Rows Config


diff --git a/src/main/java/org/bukkit/event/inventory/InventoryType.java b/src/main/java/org/bukkit/event/inventory/InventoryType.java
index c80ac017b7f720e10d739bb6554271249933930b..fbed18a09138d11dfb3c7501cc9212e837a2a136 100644
--- a/src/main/java/org/bukkit/event/inventory/InventoryType.java
+++ b/src/main/java/org/bukkit/event/inventory/InventoryType.java
@@ -171,6 +171,11 @@ public enum InventoryType {
     ;
 
     int size;
+    // Canvas start - add setter
+    public void setDefaultSize(int size) {
+        this.size = size;
+    }
+    // Canvas end - add setter
     private final String title;
     private final MenuType menuType;
     private final boolean isCreatable;
