From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Sat, 26 Jul 2025 15:50:20 -0700
Subject: [PATCH] Purpur Ender Chest 6 Rows Config


diff --git a/src/main/java/org/bukkit/event/inventory/InventoryType.java b/src/main/java/org/bukkit/event/inventory/InventoryType.java
index 450d5a5866554401adab0e0df02566b80ca4f765..39ca4aad8b7cbd239df8afbc38a6c4dec1c95549 100644
--- a/src/main/java/org/bukkit/event/inventory/InventoryType.java
+++ b/src/main/java/org/bukkit/event/inventory/InventoryType.java
@@ -171,6 +171,11 @@ public enum InventoryType {
     ;
 
     private int size;
+    // Canvas start - add setter
+    public void setDefaultSize(int size) {
+        this.size = size;
+    }
+    // Canvas end - add setter
     private final String title;
     private final MenuType menuType;
     private final boolean isCreatable;
