From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/chunk/ChunkGenerator.java b/net/minecraft/world/level/chunk/ChunkGenerator.java
index d57b3ed1c53d2818caf420004296297f006f36d7..80da91af3b3c7bdfa67a33a91941147105c9c7f2 100644
--- a/net/minecraft/world/level/chunk/ChunkGenerator.java
+++ b/net/minecraft/world/level/chunk/ChunkGenerator.java
@@ -491,6 +491,12 @@ public abstract class ChunkGenerator {
     public WeightedList<MobSpawnSettings.SpawnerData> getMobsAt(Holder<Biome> biome, StructureManager structureManager, MobCategory category, BlockPos pos) {
         Map<Structure, LongSet> allStructuresAt = structureManager.getAllStructuresAt(pos);
 
+    // Canvas start - optimize natural spawning
+        return getMobsAt(biome, structureManager, category, pos, allStructuresAt);
+    }
+
+    public WeightedList<MobSpawnSettings.SpawnerData> getMobsAt(Holder<Biome> biome, StructureManager structureManager, MobCategory category, BlockPos pos, Map<Structure, LongSet> allStructuresAt) {
+    // Canvas end - optimize natural spawning
         for (Entry<Structure, LongSet> entry : allStructuresAt.entrySet()) {
             Structure structure = entry.getKey();
             StructureSpawnOverride structureSpawnOverride = structure.spawnOverrides().get(category);
