From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/chunk/LevelChunkSection.java b/net/minecraft/world/level/chunk/LevelChunkSection.java
index 655fb8dcedfd70a54e3afdb7859f55d2979a7e37..88d0ce9ddfd6772c9d6927245a709b4a97d40ade 100644
--- a/net/minecraft/world/level/chunk/LevelChunkSection.java
+++ b/net/minecraft/world/level/chunk/LevelChunkSection.java
@@ -19,6 +19,9 @@ public class LevelChunkSection implements ca.spottedleaf.moonrise.patches.block_
     public short tickingFluidCount;
     public final PalettedContainer<BlockState> states;
     private PalettedContainer<Holder<Biome>> biomes; // CraftBukkit - read/write
+    public short canvas$waterCount; // Canvas - optimize fluid pushing
+    public short canvas$bubbleCount; // Canvas - optimize fluid pushing
+    public short canvas$lavaCount; // Canvas - optimize fluid pushing
 
     // Paper start - block counting
     private static final it.unimi.dsi.fastutil.shorts.ShortArrayList FULL_LIST = new it.unimi.dsi.fastutil.shorts.ShortArrayList(16*16*16);
@@ -48,6 +51,9 @@ public class LevelChunkSection implements ca.spottedleaf.moonrise.patches.block_
         this.nonEmptyBlockCount = section.nonEmptyBlockCount;
         this.tickingBlockCount = section.tickingBlockCount;
         this.tickingFluidCount = section.tickingFluidCount;
+        this.canvas$waterCount = section.canvas$waterCount; // Canvas - optimize fluid pushing
+        this.canvas$bubbleCount = section.canvas$bubbleCount; // Canvas - optimize fluid pushing
+        this.canvas$lavaCount = section.canvas$lavaCount; // Canvas - optimize fluid pushing
         this.states = section.states.copy();
         this.biomes = section.biomes.copy();
     }
@@ -155,6 +161,26 @@ public class LevelChunkSection implements ca.spottedleaf.moonrise.patches.block_
                 this.tickingBlockCount++;
             }
         }
+        // Canvas start - optimize fluid pushing
+        if (blockState.canvas$isWater) {
+            this.canvas$waterCount--;
+        }
+        if (blockState.canvas$isBubble) {
+            this.canvas$bubbleCount--;
+        }
+        if (blockState.canvas$isLava) {
+            this.canvas$lavaCount--;
+        }
+        if (state.canvas$isWater) {
+            this.canvas$waterCount++;
+        }
+        if (state.canvas$isBubble) {
+            this.canvas$bubbleCount++;
+        }
+        if (state.canvas$isLava) {
+            this.canvas$lavaCount++;
+        }
+        // Canvas end - optimize fluid pushing
 
         if (!!fluidState1.isRandomlyTicking()) { // Paper - block counting
             this.tickingFluidCount++;
@@ -185,6 +211,8 @@ public class LevelChunkSection implements ca.spottedleaf.moonrise.patches.block_
         // Paper start - block counting
         // reset, then recalculate
         this.nonEmptyBlockCount = (short)0;
+        this.canvas$waterCount = (short)0; // Canvas - optimize fluid pushing
+        this.canvas$lavaCount = (short)0; // Canvas - optimize fluid pushing
         this.tickingBlockCount = (short)0;
         this.tickingFluidCount = (short)0;
         this.specialCollidingBlocks = (short)0;
@@ -242,6 +270,17 @@ public class LevelChunkSection implements ca.spottedleaf.moonrise.patches.block_
                     if (fluid.isRandomlyTicking()) {
                         this.tickingFluidCount += (short)paletteCount;
                     }
+                    // Canvas start - optimize fluid pushing
+                    if (state.canvas$isWater) {
+                        this.canvas$waterCount += (short)paletteCount;
+                    }
+                    if (state.canvas$isBubble) {
+                        this.canvas$bubbleCount += (short)paletteCount;
+                    }
+                    if (state.canvas$isLava) {
+                        this.canvas$lavaCount += (short)paletteCount;
+                    }
+                    // Canvas end - optimize fluid pushing
                 }
             }
         }
