--- a/net/minecraft/server/dedicated/DedicatedServerProperties.java
+++ b/net/minecraft/server/dedicated/DedicatedServerProperties.java
@@ -80,8 +_,8 @@
     public final long maxTickTime = this.get("max-tick-time", TimeUnit.MINUTES.toMillis(1L));
     public final int maxChainedNeighborUpdates = this.get("max-chained-neighbor-updates", 1000000);
     public final int rateLimitPacketsPerSecond = this.get("rate-limit", 0);
-    public final int viewDistance = this.get("view-distance", 10);
-    public final int simulationDistance = this.get("simulation-distance", 10);
+    public int viewDistance = this.get("view-distance", 10); // Canvas - mutable
+    public int simulationDistance = this.get("simulation-distance", 10); // Canvas - mutable
     public final int maxPlayers = this.get("max-players", 20);
     public final int networkCompressionThreshold = this.get("network-compression-threshold", 256);
     public final boolean broadcastRconToOps = this.get("broadcast-rcon-to-ops", true);
