--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -281,6 +_,20 @@
         // Paper start - initialize global and world-defaults configuration
         this.paperConfigurations.initializeGlobalConfiguration(this.registryAccess());
         this.paperConfigurations.initializeWorldDefaultsConfiguration(this.registryAccess());
+        // Canvas start - pre start server config hook
+        // we have this here as a per-file patch since we might add more here
+        io.canvasmc.canvas.GlobalConfiguration.PurpurContainers containers = io.canvasmc.canvas.GlobalConfiguration.getInstance().purpurContainers;
+        org.bukkit.event.inventory.InventoryType.BARREL.setDefaultSize(switch (containers.barrelRows) {
+            case 6 -> 54;
+            case 5 -> 45;
+            case 4 -> 36;
+            case 2 -> 18;
+            case 1 -> 9;
+            default -> 27;
+        });
+        org.bukkit.event.inventory.InventoryType.ENDER_CHEST.setDefaultSize(containers.enderChestSixRows ? 54 : 27);
+        net.minecraft.world.level.block.state.BlockBehaviour.BlockStateBase.canvas$initPathTypeCache();
+        // Canvas end - pre start server config hook
         // Paper end - initialize global and world-defaults configuration
         this.server.spark.enableEarlyIfRequested(); // Paper - spark
         // Paper start - fix converting txt to json file; convert old users earlier after PlayerList creation but before file load/save
