From c8924af9ea84aa98154140092c21d52c00c82d9c Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index 5ba8c945284d8ff42b6faf0bc4e38fa2ec341948..4198946c01e777da34da8414d968ffaa75d04e4b 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -281,6 +281,43 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         // Paper start - initialize global and world-defaults configuration
         this.paperConfigurations.initializeGlobalConfiguration(this.registryAccess());
         this.paperConfigurations.initializeWorldDefaultsConfiguration(this.registryAccess());
+        // Canvas start - pre start server config hook
+        // we have this here as a per-file patch since we might add more here
+        io.canvasmc.canvas.Config.Containers containers = io.canvasmc.canvas.Config.INSTANCE.containers;
+        org.bukkit.event.inventory.InventoryType.BARREL.setDefaultSize(switch (containers.barrelRows) {
+            case 6 -> 54;
+            case 5 -> 45;
+            case 4 -> 36;
+            case 2 -> 18;
+            case 1 -> 9;
+            default -> 27;
+        });
+        org.bukkit.event.inventory.InventoryType.ENDER_CHEST.setDefaultSize(containers.enderChestSixRows ? 54 : 27);
+        if (io.canvasmc.canvas.Config.INSTANCE.combat.enableOldEnchantedGoldenAppleCrafting) {
+            org.bukkit.inventory.ShapedRecipe recipe =
+                new org.bukkit.inventory.ShapedRecipe(
+                    org.bukkit.NamespacedKey.minecraft("enchanted_golden_apple"),
+                    org.bukkit.inventory.ItemStack.of(org.bukkit.Material.ENCHANTED_GOLDEN_APPLE)
+                )
+                    .shape(new String[]{
+                        "111",
+                        "101",
+                        "111"
+                    })
+                    .setIngredient('1', org.bukkit.Material.GOLD_BLOCK)
+                    .setIngredient('0', org.bukkit.Material.APPLE);
+            recipe.setCategory(org.bukkit.inventory.recipe.CraftingBookCategory.EQUIPMENT);
+            this.server.addRecipe(recipe);
+        }
+        net.minecraft.world.level.block.state.BlockBehaviour.BlockStateBase.canvas$initPathTypeCache();
+        if (io.canvasmc.canvas.Config.INSTANCE.enableNoChatReports && !properties.enforceSecureProfile) {
+            LOGGER.warn("====================================================================================");
+            LOGGER.warn("You have 'enforce-secure-profile` disabled in your server.properties file, while enabling Canvas' no chat reports option.");
+            LOGGER.warn("This can actually cause the toast to display on the client that this server doesn't process chat reports.");
+            LOGGER.warn("To resolve this, please change 'enforce-secure-profile` in your server.properties back to true. Canvas' NCR option disables reporting on its own");
+            LOGGER.warn("====================================================================================");
+        }
+        // Canvas end - pre start server config hook
         // Paper end - initialize global and world-defaults configuration
         this.server.spark.enableEarlyIfRequested(); // Paper - spark
         // Paper start - fix converting txt to json file; convert old users earlier after PlayerList creation but before file load/save
