--- a/net/minecraft/world/entity/ai/behavior/MoveToTargetSink.java
+++ b/net/minecraft/world/entity/ai/behavior/MoveToTargetSink.java
@@ -119,7 +_,7 @@
 
     private boolean tryComputePath(Mob mob, WalkTarget target, long time) {
         BlockPos blockPos = target.getTarget().currentBlockPosition();
-        this.path = mob.getNavigation().createPath(blockPos, 0);
+        this.path = ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(mob.level(), blockPos) ? mob.getNavigation().createPath(blockPos, 0) : null; // Canvas - region threading
         this.speedModifier = target.getSpeedModifier();
         Brain<?> brain = mob.getBrain();
         if (this.reachedTarget(mob, target)) {
