From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/data/advancements/packs/VanillaNetherAdvancements.java b/net/minecraft/data/advancements/packs/VanillaNetherAdvancements.java
index 968c5fba05fc14e44ea710c072bcce3a56750d5c..416b8ff65985041d4e8435243dba2dc9dd36d918 100644
--- a/net/minecraft/data/advancements/packs/VanillaNetherAdvancements.java
+++ b/net/minecraft/data/advancements/packs/VanillaNetherAdvancements.java
@@ -151,7 +151,7 @@ public class VanillaNetherAdvancements implements AdvancementSubProvider {
             .addCriterion(
                 "killed_ghast",
                 KilledTrigger.TriggerInstance.playerKilledEntity(
-                    EntityPredicate.Builder.entity().of(holderGetter, EntityType.GHAST).located(LocationPredicate.Builder.inDimension(Level.OVERWORLD))
+                    EntityPredicate.Builder.entity().of(holderGetter, EntityType.GHAST).located(LocationPredicate.Builder.inDimension(io.canvasmc.canvas.Config.INSTANCE.fetchRespawnDimensionKey())) // Canvas - default respawn dimension config - this is where the nether portal ties to, so we should change this to match that
                 )
             )
             .save(writer, "nether/uneasy_alliance");
@@ -444,7 +444,7 @@ public class VanillaNetherAdvancements implements AdvancementSubProvider {
                 "ride_entity_distance",
                 DistanceTrigger.TriggerInstance.rideEntityInLava(
                     EntityPredicate.Builder.entity()
-                        .located(LocationPredicate.Builder.inDimension(Level.OVERWORLD))
+                        .located(LocationPredicate.Builder.inDimension(io.canvasmc.canvas.Config.INSTANCE.fetchRespawnDimensionKey())) // Canvas - default respawn dimension config - this is where the nether portal ties to, so we should change this to match that
                         .vehicle(EntityPredicate.Builder.entity().of(holderGetter, EntityType.STRIDER)),
                     DistancePredicate.horizontal(MinMaxBounds.Doubles.atLeast(50.0))
                 )
