--- a/net/minecraft/world/level/levelgen/feature/stateproviders/RandomizedIntStateProvider.java
+++ b/net/minecraft/world/level/levelgen/feature/stateproviders/RandomizedIntStateProvider.java
@@ -55,17 +_,21 @@
 
     @Override
     public BlockState getState(RandomSource random, BlockPos pos) {
-        BlockState state = this.source.getState(random, pos);
-        if (this.property == null || !state.hasProperty(this.property)) {
-            IntegerProperty integerProperty = findProperty(state, this.propertyName);
-            if (integerProperty == null) {
-                return state;
+        // Canvas start
+        BlockState blockState = this.source.getState(random, pos);
+        IntegerProperty propertyLocal = this.property;
+        if (propertyLocal == null || !blockState.hasProperty(propertyLocal)) {
+            IntegerProperty intProperty = findProperty(blockState, this.propertyName);
+            if (intProperty == null) {
+                return blockState;
             }
 
-            this.property = integerProperty;
+            propertyLocal = intProperty;
+            this.property = intProperty;
         }
 
-        return state.setValue(this.property, Integer.valueOf(this.values.sample(random)));
+        return (BlockState)blockState.setValue(propertyLocal, this.values.sample(random));
+        // Canvas end
     }
 
     @Nullable
