--- a/src/main/java/io/papermc/paper/PaperBootstrap.java
+++ b/src/main/java/io/papermc/paper/PaperBootstrap.java
@@ -1,8 +_,11 @@
 package io.papermc.paper;
 
+import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.util.List;
 import joptsimple.OptionSet;
 import net.minecraft.SharedConstants;
+import net.minecraft.server.Eula;
 import net.minecraft.server.Main;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -15,6 +_,21 @@
 
     public static void boot(final OptionSet options) {
         SharedConstants.tryDetectVersion();
+        // Canvas start - move EULA verification sooner
+        Path path2 = Paths.get("eula.txt");
+        Eula eula = new Eula(path2);
+        boolean eulaAgreed = Boolean.getBoolean("com.mojang.eula.agree");
+        if (eulaAgreed) {
+            LOGGER.error("You have used the Spigot command line EULA agreement flag.");
+            LOGGER.error("By using this setting you are indicating your agreement to Mojang's EULA (https://aka.ms/MinecraftEULA).");
+            LOGGER.error("If you do not agree to the above EULA please stop your server and remove this flag immediately.");
+        }
+        if (!eula.hasAgreedToEULA() && !eulaAgreed) {
+            LOGGER.info("You need to agree to the EULA in order to run the server. Go to eula.txt for more info.");
+            return;
+        }
+        System.out.println("Loading libraries, please wait..."); // Restore CraftBukkit log
+        // Canvas end
 
         getStartupVersionMessages().forEach(LOGGER::info);
 
