--- a/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
@@ -23,15 +_,17 @@
     Optional<String> gitBranch,
     Optional<String> gitCommit
 ) implements ServerBuildInfo {
+    public static boolean IS_EXPERIMENTAL = false; // Canvas - experimental toggle
     private static final String ATTRIBUTE_BRAND_ID = "Brand-Id";
     private static final String ATTRIBUTE_BRAND_NAME = "Brand-Name";
     private static final String ATTRIBUTE_BUILD_TIME = "Build-Time";
     private static final String ATTRIBUTE_BUILD_NUMBER = "Build-Number";
     private static final String ATTRIBUTE_GIT_BRANCH = "Git-Branch";
     private static final String ATTRIBUTE_GIT_COMMIT = "Git-Commit";
+    private static final String ATTRIBUTE_EXPERIMENTAL = "Experimental"; // Canvas - experimental toggle
 
     private static final String BRAND_PAPER_NAME = "Paper";
-    private static final String BRAND_PURPUR_NAME = "Purpur"; // Purpur - Rebrand
+    private static final String BRAND_PURPUR_NAME = "Canvas"; // Purpur - Rebrand // Canvas - Rebrand
 
     private static final String BUILD_DEV = "DEV";
 
@@ -58,6 +_,7 @@
             getManifestAttribute(manifest, ATTRIBUTE_GIT_BRANCH),
             getManifestAttribute(manifest, ATTRIBUTE_GIT_COMMIT)
         );
+        IS_EXPERIMENTAL = Boolean.parseBoolean(getManifestAttribute(manifest, ATTRIBUTE_EXPERIMENTAL).orElse("false"));
     }
 
     @Override
