--- a/net/minecraft/world/level/levelgen/structure/structures/JungleTemplePiece.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/JungleTemplePiece.java
@@ -26,14 +_,17 @@
 import net.minecraft.world.level.levelgen.structure.pieces.StructurePieceSerializationContext;
 import net.minecraft.world.level.levelgen.structure.pieces.StructurePieceType;
 import net.minecraft.world.level.storage.loot.BuiltInLootTables;
+import java.util.concurrent.atomic.AtomicBoolean;
 
 public class JungleTemplePiece extends ScatteredFeaturePiece {
     public static final int WIDTH = 12;
     public static final int DEPTH = 15;
-    private boolean placedMainChest;
-    private boolean placedHiddenChest;
-    private boolean placedTrap1;
-    private boolean placedTrap2;
+    // Canvas start
+    private final AtomicBoolean placedMainChest = new AtomicBoolean(false);
+    private final AtomicBoolean placedHiddenChest = new AtomicBoolean(false);
+    private final AtomicBoolean placedTrap1 = new AtomicBoolean(false);
+    private final AtomicBoolean placedTrap2 = new AtomicBoolean(false);
+    // Canvas end
     private static final JungleTemplePiece.MossStoneSelector STONE_SELECTOR = new JungleTemplePiece.MossStoneSelector();
 
     public JungleTemplePiece(RandomSource random, int x, int z) {
@@ -42,19 +_,23 @@
 
     public JungleTemplePiece(CompoundTag tag) {
         super(StructurePieceType.JUNGLE_PYRAMID_PIECE, tag);
-        this.placedMainChest = tag.getBoolean("placedMainChest");
-        this.placedHiddenChest = tag.getBoolean("placedHiddenChest");
-        this.placedTrap1 = tag.getBoolean("placedTrap1");
-        this.placedTrap2 = tag.getBoolean("placedTrap2");
+        // Canvas start
+        this.placedMainChest.set(tag.getBoolean("placedMainChest"));
+        this.placedHiddenChest.set(tag.getBoolean("placedHiddenChest"));
+        this.placedTrap1.set(tag.getBoolean("placedTrap1"));
+        this.placedTrap2.set(tag.getBoolean("placedTrap2"));
+        // Canvas end
     }
 
     @Override
     protected void addAdditionalSaveData(StructurePieceSerializationContext context, CompoundTag tag) {
         super.addAdditionalSaveData(context, tag);
-        tag.putBoolean("placedMainChest", this.placedMainChest);
-        tag.putBoolean("placedHiddenChest", this.placedHiddenChest);
-        tag.putBoolean("placedTrap1", this.placedTrap1);
-        tag.putBoolean("placedTrap2", this.placedTrap2);
+        // Canvas start
+        tag.putBoolean("placedMainChest", this.placedMainChest.get());
+        tag.putBoolean("placedHiddenChest", this.placedHiddenChest.get());
+        tag.putBoolean("placedTrap1", this.placedTrap1.get());
+        tag.putBoolean("placedTrap2", this.placedTrap2.get());
+        // Canvas end
     }
 
     @Override
@@ -242,8 +_,10 @@
                 box
             );
             this.placeBlock(level, Blocks.MOSSY_COBBLESTONE.defaultBlockState(), 3, -3, 1, box);
-            if (!this.placedTrap1) {
-                this.placedTrap1 = this.createDispenser(level, box, random, 3, -2, 1, Direction.NORTH, BuiltInLootTables.JUNGLE_TEMPLE_DISPENSER);
+            // Canvas start
+            if (!this.placedTrap1.get()) {
+                this.placedTrap1.set(this.createDispenser(level, box, random, 3, -2, 1, Direction.NORTH, BuiltInLootTables.JUNGLE_TEMPLE_DISPENSER));
+            // Canvas end
             }
 
             this.placeBlock(level, Blocks.VINE.defaultBlockState().setValue(VineBlock.SOUTH, Boolean.valueOf(true)), 3, -2, 2, box);
@@ -340,14 +_,18 @@
             );
             this.placeBlock(level, Blocks.MOSSY_COBBLESTONE.defaultBlockState(), 9, -3, 4, box);
             this.placeBlock(level, blockState4, 9, -2, 4, box);
-            if (!this.placedTrap2) {
-                this.placedTrap2 = this.createDispenser(level, box, random, 9, -2, 3, Direction.WEST, BuiltInLootTables.JUNGLE_TEMPLE_DISPENSER);
+            // Canvas start
+            if (!this.placedTrap2.get()) {
+                this.placedTrap2.set(this.createDispenser(level, box, random, 9, -2, 3, Direction.WEST, BuiltInLootTables.JUNGLE_TEMPLE_DISPENSER));
+            // Canvas end
             }
 
             this.placeBlock(level, Blocks.VINE.defaultBlockState().setValue(VineBlock.EAST, Boolean.valueOf(true)), 8, -1, 3, box);
             this.placeBlock(level, Blocks.VINE.defaultBlockState().setValue(VineBlock.EAST, Boolean.valueOf(true)), 8, -2, 3, box);
-            if (!this.placedMainChest) {
-                this.placedMainChest = this.createChest(level, box, random, 8, -3, 3, BuiltInLootTables.JUNGLE_TEMPLE);
+            // Canvas start
+            if (!this.placedMainChest.get()) {
+                this.placedMainChest.set(this.createChest(level, box, random, 8, -3, 3, BuiltInLootTables.JUNGLE_TEMPLE));
+            // Canvas end
             }
 
             this.placeBlock(level, Blocks.MOSSY_COBBLESTONE.defaultBlockState(), 9, -3, 2, box);
@@ -390,8 +_,10 @@
             this.placeBlock(level, Blocks.STICKY_PISTON.defaultBlockState().setValue(PistonBaseBlock.FACING, Direction.WEST), 10, -2, 8, box);
             this.placeBlock(level, Blocks.STICKY_PISTON.defaultBlockState().setValue(PistonBaseBlock.FACING, Direction.WEST), 10, -1, 8, box);
             this.placeBlock(level, Blocks.REPEATER.defaultBlockState().setValue(RepeaterBlock.FACING, Direction.NORTH), 10, -2, 10, box);
-            if (!this.placedHiddenChest) {
-                this.placedHiddenChest = this.createChest(level, box, random, 9, -3, 10, BuiltInLootTables.JUNGLE_TEMPLE);
+            // Canvas start
+            if (!this.placedHiddenChest.get()) {
+                this.placedHiddenChest.set(this.createChest(level, box, random, 9, -3, 10, BuiltInLootTables.JUNGLE_TEMPLE));
+            // Canvas end
             }
         }
     }
