--- a/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
@@ -2,6 +_,7 @@
 
 import com.google.common.collect.Lists;
 import com.mojang.logging.LogUtils;
+import java.util.Collections;
 import java.util.List;
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
@@ -95,7 +_,7 @@
     public static class MineShaftCorridor extends MineshaftPieces.MineShaftPiece {
         private final boolean hasRails;
         private final boolean spiderCorridor;
-        private boolean hasPlacedSpider;
+        private volatile boolean hasPlacedSpider; // Canvas - volatile
         private final int numSections;
 
         public MineShaftCorridor(CompoundTag tag) {
@@ -954,7 +_,7 @@
     }
 
     public static class MineShaftRoom extends MineshaftPieces.MineShaftPiece {
-        private final List<BoundingBox> childEntranceBoxes = Lists.newLinkedList();
+        private final List<BoundingBox> childEntranceBoxes = Collections.synchronizedList(Lists.newLinkedList()); // Canvas
 
         public MineShaftRoom(int genDepth, RandomSource random, int x, int z, MineshaftStructure.Type type) {
             super(
