--- a/net/minecraft/world/level/levelgen/structure/structures/OceanMonumentPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/OceanMonumentPieces.java
@@ -160,8 +_,8 @@
         private static final int DEPTH = 58;
         public static final int BIOME_RANGE_CHECK = 29;
         private static final int TOP_POSITION = 61;
-        private OceanMonumentPieces.RoomDefinition sourceRoom;
-        private OceanMonumentPieces.RoomDefinition coreRoom;
+        private volatile OceanMonumentPieces.RoomDefinition sourceRoom; // Canvas - volatile
+        private volatile OceanMonumentPieces.RoomDefinition coreRoom; // Canvas - volatile
         private final List<OceanMonumentPieces.OceanMonumentPiece> childPieces = Lists.newArrayList();
 
         public MonumentBuilding(RandomSource random, int x, int z, Direction orientation) {
@@ -1897,9 +_,9 @@
         final int index;
         final OceanMonumentPieces.RoomDefinition[] connections = new OceanMonumentPieces.RoomDefinition[6];
         final boolean[] hasOpening = new boolean[6];
-        boolean claimed;
-        boolean isSource;
-        private int scanIndex;
+        volatile boolean claimed; // Canvas - volatile
+        volatile boolean isSource; // Canvas - volatile
+        private volatile int scanIndex; // Canvas - volatile
 
         public RoomDefinition(int index) {
             this.index = index;
