--- a/net/minecraft/world/level/levelgen/structure/structures/SwampHutPiece.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/SwampHutPiece.java
@@ -23,8 +_,8 @@
 import net.minecraft.world.level.levelgen.structure.pieces.StructurePieceType;
 
 public class SwampHutPiece extends ScatteredFeaturePiece {
-    private boolean spawnedWitch;
-    private boolean spawnedCat;
+    private volatile boolean spawnedWitch; // Canvas - volatile
+    private volatile boolean spawnedCat; // Canvas - volatile
 
     public SwampHutPiece(RandomSource random, int x, int z) {
         super(StructurePieceType.SWAMPLAND_HUT, x, 64, z, 7, 7, 9, getRandomHorizontalDirection(random));
