--- a/net/minecraft/core/BlockPos.java
+++ b/net/minecraft/core/BlockPos.java
@@ -28,6 +_,7 @@
 import net.minecraft.world.phys.Vec3;
 import org.apache.commons.lang3.Validate;
 import org.apache.commons.lang3.tuple.Pair;
+import org.jetbrains.annotations.NotNull;
 import org.slf4j.Logger;
 
 @Immutable
@@ -223,7 +_,7 @@
     }
 
     @Override
-    public BlockPos relative(Direction direction) {
+    public synchronized @NotNull BlockPos relative(Direction direction) { // Canvas
         // Paper start - Perf: Optimize BlockPosition
         switch(direction) {
             case UP:
