--- a/net/minecraft/core/Holder.java
+++ b/net/minecraft/core/Holder.java
@@ -116,6 +_,7 @@
         private final HolderOwner<T> owner;
         @Nullable
         private Set<TagKey<T>> tags;
+        public TagKey<T>[] canvas$tagsArray = new TagKey[0]; // Canvas - optimize fluidOnEyes - remove stream
         private final Holder.Reference.Type type;
         @Nullable
         private ResourceKey<T> key;
@@ -231,6 +_,7 @@
 
         void bindTags(Collection<TagKey<T>> tags) {
             this.tags = it.unimi.dsi.fastutil.objects.ReferenceSets.unmodifiable(new it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<>(tags)); // Paper - use reference set because TagKey are interned
+            this.canvas$tagsArray = this.tags.toArray(new TagKey[0]); // Canvas - optimize fluidOnEyes - remove stream
         }
 
         @Override
