--- a/net/minecraft/core/Holder.java
+++ b/net/minecraft/core/Holder.java
@@ -115,6 +_,7 @@
     public static class Reference<T> implements Holder<T> {
         private final HolderOwner<T> owner;
         private @Nullable Set<TagKey<T>> tags;
+        public TagKey<T>[] canvas$tagsArray = new TagKey[0]; // Canvas - optimize fluidOnEyes - remove stream
         private final Holder.Reference.Type type;
         private @Nullable ResourceKey<T> key;
         private @Nullable T value;
@@ -228,6 +_,7 @@
 
         void bindTags(Collection<TagKey<T>> tags) {
             this.tags = it.unimi.dsi.fastutil.objects.ReferenceSets.unmodifiable(new it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<>(tags)); // Paper - use reference set because TagKey are interned
+            this.canvas$tagsArray = this.tags.toArray(new TagKey[0]); // Canvas - optimize fluidOnEyes - remove stream
         }
 
         @Override
