From a97bf5e0abe5bf59c656f1cf543b54003c29b383 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/core/Holder.java b/net/minecraft/core/Holder.java
index 3ca8678b6dddae039cf35963dfe1d578284cf398..a67ed42adda8cfc35d8b83ff98d14350519b9992 100644
--- a/net/minecraft/core/Holder.java
+++ b/net/minecraft/core/Holder.java
@@ -131,6 +131,7 @@ public interface Holder<T> {
         private final HolderOwner<T> owner;
         private @Nullable Set<TagKey<T>> tags;
         private @Nullable DataComponentMap components;
+        public TagKey<T>[] canvas$tagsArray = new TagKey[0]; // Canvas - optimize fluidOnEyes - remove stream
         private final Holder.Reference.Type type;
         private @Nullable ResourceKey<T> key;
         private @Nullable T value;
@@ -249,6 +250,7 @@ public interface Holder<T> {
 
         void bindTags(final Collection<TagKey<T>> tags) {
             this.tags = it.unimi.dsi.fastutil.objects.ReferenceSets.unmodifiable(new it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<>(tags)); // Paper - use reference set because TagKey are interned
+            this.canvas$tagsArray = this.tags.toArray(new TagKey[0]); // Canvas - optimize fluidOnEyes - remove stream
         }
 
         public void bindComponents(final DataComponentMap components) {
