--- a/net/minecraft/core/NonNullList.java
+++ b/net/minecraft/core/NonNullList.java
@@ -14,11 +_,11 @@
     private final E defaultValue;
 
     public static <E> NonNullList<E> create() {
-        return new NonNullList<>(Lists.newArrayList(), null);
+        return new NonNullList<>(new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(), null); // Canvas - optimized collection
     }
 
     public static <E> NonNullList<E> createWithCapacity(int initialCapacity) {
-        return new NonNullList<>(Lists.newArrayListWithCapacity(initialCapacity), null);
+        return new NonNullList<>(new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(initialCapacity), null); // Canvas - optimized collection
     }
 
     /**
@@ -28,12 +_,12 @@
         Validate.notNull(defaultValue);
         Object[] objects = new Object[size];
         Arrays.fill(objects, defaultValue);
-        return new NonNullList<>(Arrays.asList((E[])objects), defaultValue);
+        return new NonNullList<>(new it.unimi.dsi.fastutil.objects.ObjectArrayList<>((E[])objects), defaultValue); // Canvas - optimized collection
     }
 
     @SafeVarargs
     public static <E> NonNullList<E> of(E defaultValue, E... elements) {
-        return new NonNullList<>(Arrays.asList(elements), defaultValue);
+        return new NonNullList<>(new it.unimi.dsi.fastutil.objects.ObjectArrayList<>(elements), defaultValue); // Canvas - optimized collection
     }
 
     protected NonNullList(List<E> list, @Nullable E defaultValue) {
