From 30322d32a063dc3b6c41ba7f9d75c2e508ff7cf1 Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 06:37:42 -0700
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/ai/memory/MemorySlot.java b/net/minecraft/world/entity/ai/memory/MemorySlot.java
index 88a89b4c72cc99dc89d3f3cc928b2dfda4125759..2e636e5894f757b6351298b05e8cf6211fcaa15f 100644
--- a/net/minecraft/world/entity/ai/memory/MemorySlot.java
+++ b/net/minecraft/world/entity/ai/memory/MemorySlot.java
@@ -30,6 +30,11 @@ public class MemorySlot<T> {
     public void set(final T value, final long timeToLive) {
         this.value = value;
         this.timeToLive = timeToLive;
+        // Canvas start - optimize brain
+        if (value == null && timeToLive != Long.MAX_VALUE) {
+            this.timeToLive = Long.MAX_VALUE;
+        }
+        // Canvas end - optimize brain
     }
 
     public void set(final T value) {
