--- a/net/minecraft/world/entity/ai/memory/MemorySlot.java
+++ b/net/minecraft/world/entity/ai/memory/MemorySlot.java
@@ -30,6 +_,11 @@
     public void set(final T value, final long timeToLive) {
         this.value = value;
         this.timeToLive = timeToLive;
+        // Canvas start - optimize brain
+        if (value == null && timeToLive != Long.MAX_VALUE) {
+            this.timeToLive = Long.MAX_VALUE;
+        }
+        // Canvas end - optimize brain
     }
 
     public void set(final T value) {
