From 56fbeba69dc02a23c55fcd9bf06e3cae1da93bce Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java b/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
index 17293eeda23265ca7b0b607576737fbd1c5d52eb..23f476509d554e557f8c4389cb2722817100ab2d 100644
--- a/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
+++ b/net/minecraft/world/level/levelgen/structure/templatesystem/StructureTemplate.java
@@ -251,6 +251,12 @@ public class StructureTemplate {
     public static BlockPos calculateRelativePosition(final StructurePlaceSettings settings, final BlockPos pos) {
         return transform(pos, settings.getMirror(), settings.getRotation(), settings.getRotationPivot());
     }
+    // Canvas start - optimize structure gen
+
+    private List<StructureTemplate.StructureBlockInfo> structureLayoutOptimizer$shrinkStructureTemplateBlocksList(final StructureTemplate.Palette palette, final BlockPos position, final StructurePlaceSettings settings) {
+        return io.canvasmc.canvas.util.structure.StructureTemplateOptimizer.getStructureBlockInfosInBounds(palette, position, settings);
+    }
+    // Canvas end - optimize structure gen
 
     public boolean placeInWorld(
         ServerLevelAccessor level, // Paper - make non-final
@@ -277,7 +283,7 @@ public class StructureTemplate {
             }
         }
         // CraftBukkit end
-        List<StructureTemplate.StructureBlockInfo> blockInfoList = settings.getRandomPalette(this.palettes, position).blocks();
+        List<StructureTemplate.StructureBlockInfo> blockInfoList = io.canvasmc.canvas.GlobalConfiguration.getInstance().chunkSystem.structureOptimizations.enable ? structureLayoutOptimizer$shrinkStructureTemplateBlocksList(settings.getRandomPalette(this.palettes, position), position, settings) : settings.getRandomPalette(this.palettes, position).blocks(); // Canvas - optimize structure gen
         if ((!blockInfoList.isEmpty() || !settings.isIgnoreEntities() && !this.entityInfoList.isEmpty())
             && this.size.getX() >= 1
             && this.size.getY() >= 1
@@ -882,7 +888,7 @@ public class StructureTemplate {
         private @Nullable List<StructureTemplate.JigsawBlockInfo> cachedJigsaws;
 
         private Palette(final List<StructureTemplate.StructureBlockInfo> blocks) {
-            this.blocks = blocks;
+            this.blocks = io.canvasmc.canvas.GlobalConfiguration.getInstance().chunkSystem.structureOptimizations.enable ? new io.canvasmc.canvas.util.structure.PalettedStructureBlockInfoList(blocks) : blocks; // Canvas - optimize structure gen
         }
 
         public List<StructureTemplate.JigsawBlockInfo> jigsaws() {
