From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/server/Main.java b/net/minecraft/server/Main.java
index 9216e85e74bb0121c72665f155b07cf9a37f430a..5dbf208ddc4dd2373a00b4674583a8d20682d8ae 100644
--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -67,6 +67,20 @@ public class Main {
     public static void main(final OptionSet optionSet) { // CraftBukkit - replaces main(String[] args)
         io.papermc.paper.log.LogManagerShutdownThread.hook(); // Paper - Improved watchdog support
         SharedConstants.tryDetectVersion();
+        // Canvas start - validate EULA early
+        Path path2 = Paths.get("eula.txt");
+        Eula eula = new Eula(path2);
+        boolean eulaAgreed = Boolean.getBoolean("com.mojang.eula.agree");
+        if (eulaAgreed) {
+            LOGGER.error("You have used the Spigot command line EULA agreement flag.");
+            LOGGER.error("By using this setting you are indicating your agreement to Mojang's EULA (https://aka.ms/MinecraftEULA).");
+            LOGGER.error("If you do not agree to the above EULA please stop your server and remove this flag immediately.");
+        }
+        if (!eula.hasAgreedToEULA() && !eulaAgreed) {
+            LOGGER.info("You need to agree to the EULA in order to run the server. Go to eula.txt for more info.");
+            return;
+        }
+        // Canvas end - validate EULA early
         /* CraftBukkit start - Replace everything
         OptionParser optionParser = new OptionParser();
         OptionSpec<Void> optionSpec = optionParser.accepts("nogui");
@@ -113,8 +127,7 @@ public class Main {
             DedicatedServerSettings dedicatedServerSettings = new DedicatedServerSettings(optionSet); // CraftBukkit - CLI argument support
             dedicatedServerSettings.forceSave();
             RegionFileVersion.configure(dedicatedServerSettings.getProperties().regionFileComression);
-            Path path2 = Paths.get("eula.txt");
-            Eula eula = new Eula(path2);
+            // Canvas - validate EULA early - move up
             // Paper start - load config files early for access below if needed
             org.bukkit.configuration.file.YamlConfiguration bukkitConfiguration = io.papermc.paper.configuration.PaperConfigurations.loadLegacyConfigFile((File) optionSet.valueOf("bukkit-settings"));
             org.bukkit.configuration.file.YamlConfiguration spigotConfiguration = io.papermc.paper.configuration.PaperConfigurations.loadLegacyConfigFile((File) optionSet.valueOf("spigot-settings"));
@@ -137,18 +150,7 @@ public class Main {
                 return;
             }
 
-            // Spigot start
-            boolean eulaAgreed = Boolean.getBoolean("com.mojang.eula.agree");
-            if (eulaAgreed) {
-                LOGGER.error("You have used the Spigot command line EULA agreement flag.");
-                LOGGER.error("By using this setting you are indicating your agreement to Mojang's EULA (https://aka.ms/MinecraftEULA).");
-                LOGGER.error("If you do not agree to the above EULA please stop your server and remove this flag immediately.");
-            }
-            if (!eula.hasAgreedToEULA() && !eulaAgreed) {
-                // Spigot end
-                LOGGER.info("You need to agree to the EULA in order to run the server. Go to eula.txt for more info.");
-                return;
-            }
+            // Canvas - validate EULA early - move up
 
             // Paper start - Detect headless JRE
             String awtException = io.papermc.paper.util.ServerEnvironment.awtDependencyCheck();
