From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Toffikk <slodkitofik@gmail.com>
Date: Thu, 9 Apr 2026 22:05:45 +0200
Subject: [PATCH] Paper Changes


diff --git a/net/minecraft/nbt/NbtAccounter.java b/net/minecraft/nbt/NbtAccounter.java
index 6f516a6e7d193e53b407ad54a2cb027bee2258df..98185e8382603d08ab74e6da73d266dde6f39203 100644
--- a/net/minecraft/nbt/NbtAccounter.java
+++ b/net/minecraft/nbt/NbtAccounter.java
@@ -48,6 +48,16 @@ public class NbtAccounter {
         }
     }
 
+    // Paper start - Track codec depth
+    public void pushDepth(final int depth) {
+        if (this.depth + depth >= this.maxDepth) {
+            throw new NbtAccounterException("Tried to read NBT tag with too high complexity, depth > " + this.maxDepth);
+        } else {
+            this.depth += depth;
+        }
+    }
+    // Paper end - Track codec depth
+
     public void pushDepth() {
         if (this.depth >= this.maxDepth) {
             throw new NbtAccounterException("Tried to read NBT tag with too high complexity, depth > " + this.maxDepth);
diff --git a/net/minecraft/network/RegistryFriendlyByteBuf.java b/net/minecraft/network/RegistryFriendlyByteBuf.java
index e1651db791e55a6621157116359aa9aca6919475..cf8f037dc0926e7cadf41f4341e0476aa3457fc9 100644
--- a/net/minecraft/network/RegistryFriendlyByteBuf.java
+++ b/net/minecraft/network/RegistryFriendlyByteBuf.java
@@ -12,6 +12,15 @@ public class RegistryFriendlyByteBuf extends FriendlyByteBuf {
         this.registryAccess = registryAccess;
     }
 
+    // Paper start - Track codec depth
+    public RegistryFriendlyByteBuf(final ByteBuf source, final RegistryFriendlyByteBuf parent) {
+        super(source);
+        this.registryAccess = parent.registryAccess;
+        this.trackCodecDepth = parent.trackCodecDepth;
+        this.codecDepth = parent.codecDepth;
+    }
+    // Paper end - Track codec depth
+
     public RegistryAccess registryAccess() {
         return this.registryAccess;
     }
diff --git a/net/minecraft/network/codec/ByteBufCodecs.java b/net/minecraft/network/codec/ByteBufCodecs.java
index 43270171bf22ed87c015593289380ea05eebfb57..0b7c57cc8cd34e2552f6431576d6fb3ef9125147 100644
--- a/net/minecraft/network/codec/ByteBufCodecs.java
+++ b/net/minecraft/network/codec/ByteBufCodecs.java
@@ -334,7 +334,13 @@ public interface ByteBufCodecs {
         return new StreamCodec<ByteBuf, Tag>() {
             @Override
             public Tag decode(ByteBuf buffer) {
-                Tag nbt = FriendlyByteBuf.readNbt(buffer, accounter.get());
+                // Paper start - Track codec depth
+                final NbtAccounter acc = accounter.get();
+                if (buffer instanceof FriendlyByteBuf friendlyByteBuf && friendlyByteBuf.trackCodecDepth) {
+                    acc.pushDepth(friendlyByteBuf.codecDepth * 3);
+                }
+                Tag nbt = FriendlyByteBuf.readNbt(buffer, acc);
+                // Paper end - Track codec depth
                 if (nbt == null) {
                     throw new DecoderException("Expected non-null compound tag");
                 } else {
@@ -634,7 +640,7 @@ public interface ByteBufCodecs {
     }
 
     static <V> StreamCodec.CodecOperation<RegistryFriendlyByteBuf, V, V> registryFriendlyLengthPrefixed(int length) {
-        return lengthPrefixed(length, (registryFriendlyByteBuf, byteBuf) -> new RegistryFriendlyByteBuf(byteBuf, registryFriendlyByteBuf.registryAccess()));
+        return lengthPrefixed(length, (registryFriendlyByteBuf, byteBuf) -> new RegistryFriendlyByteBuf(byteBuf, registryFriendlyByteBuf)); // Paper - Track codec depth
     }
 
     static <T> StreamCodec<ByteBuf, T> idMapper(final IntFunction<T> idLookup, final ToIntFunction<T> idGetter) {
diff --git a/net/minecraft/world/effect/MobEffectInstance.java b/net/minecraft/world/effect/MobEffectInstance.java
index 5888dcb3ce9b988a6ba1ff4f810f41522bc25161..3bcfe2faca53e85a20bdbf847b25aef9c61e24f6 100644
--- a/net/minecraft/world/effect/MobEffectInstance.java
+++ b/net/minecraft/world/effect/MobEffectInstance.java
@@ -253,13 +253,16 @@ public class MobEffectInstance implements Comparable<MobEffectInstance> {
     }
 
     private boolean downgradeToHiddenEffect() {
-        if (this.duration == 0 && this.hiddenEffect != null) {
+        // Paper start - Fix MC-259832
+        if (this.duration != 0 || this.hiddenEffect == null) return false;
+
+        do {
             this.setDetailsFrom(this.hiddenEffect);
             this.hiddenEffect = this.hiddenEffect.hiddenEffect;
-            return true;
-        } else {
-            return false;
-        }
+        } while (this.duration == 0 && this.hiddenEffect != null);
+
+        return true;
+        // Paper end - Fix MC-259832
     }
 
     public void onEffectStarted(LivingEntity entity) {
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 624d9b86f775dbf80f36bc1c1d995af683262a3c..394768573f1b3475d1cd0db715c05704cef1ad3c 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -1263,6 +1263,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
 
                 float blockSpeedFactor = this.getBlockSpeedFactor();
                 this.setDeltaMovement(this.getDeltaMovement().multiply(blockSpeedFactor, 1.0, blockSpeedFactor));
+                if (this.vehicle == null && type != MoverType.PLAYER) this.repositionAllPassengers(false); // Paper
                 profilerFiller.pop();
             }
         }
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index f9d1fddefdf3e8ca1b77e4640cc7de0c0d8856b7..c4351e6140ef7ed75f1387758d454d2263444a71 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -2474,16 +2474,17 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
                 return false;
             }
 
-            if (damageSource.getEntity() instanceof net.minecraft.world.entity.player.Player) {
+            if (damageSource.getEntity() instanceof net.minecraft.world.entity.player.Player attackingPlayer) {
                 // Paper start - PlayerAttackEntityCooldownResetEvent
                 //((net.minecraft.world.entity.player.Player) damagesource.getEntity()).resetOnlyAttackStrengthTicker(); // Moved from Player in order to make the cooldown reset get called after the damage event is fired
-                if (damageSource.getEntity() instanceof ServerPlayer) {
-                    ServerPlayer player = (ServerPlayer) damageSource.getEntity();
-                    if (new com.destroystokyo.paper.event.player.PlayerAttackEntityCooldownResetEvent(player.getBukkitEntity(), this.getBukkitEntity(), player.getAttackStrengthScale(0F)).callEvent()) {
-                        player.resetOnlyAttackStrengthTicker();
+                if (!attackingPlayer.isUsingItem() || !attackingPlayer.getUseItem().has(DataComponents.KINETIC_WEAPON)) { // Paper - Preserve vanilla spear charge cooldown
+                    if (attackingPlayer instanceof ServerPlayer player) {
+                        if (new com.destroystokyo.paper.event.player.PlayerAttackEntityCooldownResetEvent(player.getBukkitEntity(), this.getBukkitEntity(), player.getAttackStrengthScale(0F)).callEvent()) {
+                            player.resetOnlyAttackStrengthTicker();
+                        }
+                    } else {
+                        attackingPlayer.resetOnlyAttackStrengthTicker();
                     }
-                } else {
-                    ((net.minecraft.world.entity.player.Player) damageSource.getEntity()).resetOnlyAttackStrengthTicker();
                 }
                 // Paper end - PlayerAttackEntityCooldownResetEvent
             }
diff --git a/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java b/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
index 1123ec0038552e0b40774f4a433ff325695ea071..8af57f1625b2eb67950815047466624f8f65082b 100644
--- a/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
+++ b/net/minecraft/world/entity/vehicle/minecart/AbstractMinecart.java
@@ -390,6 +390,7 @@ public abstract class AbstractMinecart extends VehicleEntity {
 
     protected void moveAlongTrack(ServerLevel level) {
         this.behavior.moveAlongTrack(level);
+        if (this.getVehicle() == null) this.repositionAllPassengers(false); // Paper
     }
 
     protected void comeOffTrack(ServerLevel level) {
diff --git a/net/minecraft/world/item/ItemStack.java b/net/minecraft/world/item/ItemStack.java
index 8660c37560442153787e854a0e95bdd81c1bc152..165d7bb826943a2bf050442e6f154e312edb8436 100644
--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -227,7 +227,7 @@ public final class ItemStack implements DataComponentHolder {
             public ItemStack decode(RegistryFriendlyByteBuf buffer) {
                 ItemStack itemStack = codec.decode(buffer);
                 if (!itemStack.isEmpty()) {
-                    RegistryOps<Unit> registryOps = buffer.registryAccess().createSerializationContext(NullOps.INSTANCE);
+                    RegistryOps<io.papermc.paper.util.CountingOps.Value> registryOps = buffer.registryAccess().createSerializationContext(io.papermc.paper.util.CountingOps.INSTANCE); // Paper - Track codec depth
                     ItemStack.CODEC.encodeStart(registryOps, itemStack).getOrThrow(DecoderException::new);
                 }
 
diff --git a/net/minecraft/world/item/component/CustomData.java b/net/minecraft/world/item/component/CustomData.java
index 418608ab95401f9c6266223ec7f0a9ec7d03d1e8..3dc4b262c81c393f2082b3c938e860fecf0578e0 100644
--- a/net/minecraft/world/item/component/CustomData.java
+++ b/net/minecraft/world/item/component/CustomData.java
@@ -13,19 +13,18 @@ import net.minecraft.world.item.ItemStack;
 
 public final class CustomData {
     public static final CustomData EMPTY = new CustomData(new CompoundTag());
-    public static final Codec<CompoundTag> COMPOUND_TAG_CODEC = Codec.withAlternative(CompoundTag.CODEC, TagParser.FLATTENED_CODEC);
     // Paper start - Item serialization as json
     public static ThreadLocal<Boolean> SERIALIZE_CUSTOM_AS_SNBT = ThreadLocal.withInitial(() -> false);
-    public static final Codec<CustomData> CODEC = Codec.either(CompoundTag.CODEC, TagParser.FLATTENED_CODEC)
+    public static final Codec<CompoundTag> COMPOUND_TAG_CODEC = Codec.either(CompoundTag.CODEC, TagParser.FLATTENED_CODEC)
         .xmap(com.mojang.datafixers.util.Either::unwrap, data -> { // Both will be used for deserialization, but we decide which one to use for serialization
             if (!SERIALIZE_CUSTOM_AS_SNBT.get()) {
                 return com.mojang.datafixers.util.Either.left(data); // First codec
             } else {
                 return com.mojang.datafixers.util.Either.right(data); // Second codec
             }
-        })
-        .xmap(CustomData::new, customData -> customData.tag);
+        });
     // Paper end - Item serialization as json
+    public static final Codec<CustomData> CODEC = COMPOUND_TAG_CODEC.xmap(CustomData::new, data -> data.tag);
     @Deprecated
     public static final StreamCodec<ByteBuf, CustomData> STREAM_CODEC = ByteBufCodecs.COMPOUND_TAG.map(CustomData::new, data -> data.tag);
     private final CompoundTag tag;
diff --git a/net/minecraft/world/item/equipment/Equippable.java b/net/minecraft/world/item/equipment/Equippable.java
index 73bb882252fec07b93fdbf13ea61ec846a17d475..fe24aff29f04027449abd68b9d0630c2b106877f 100644
--- a/net/minecraft/world/item/equipment/Equippable.java
+++ b/net/minecraft/world/item/equipment/Equippable.java
@@ -129,6 +129,19 @@ public record Equippable(
             ItemStack itemBySlot = player.getItemBySlot(this.slot);
             if ((!EnchantmentHelper.has(itemBySlot, EnchantmentEffectComponents.PREVENT_ARMOR_CHANGE) || player.isCreative())
                 && !ItemStack.isSameItemSameComponents(stack, itemBySlot)) {
+                // Paper start
+                final io.papermc.paper.event.player.PlayerSwapWithEquipmentSlotEvent event = new io.papermc.paper.event.player.PlayerSwapWithEquipmentSlotEvent(
+                    (org.bukkit.entity.Player) player.getBukkitEntity(),
+                    org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(stack),
+                    org.bukkit.craftbukkit.CraftEquipmentSlot.getSlot(this.slot),
+                    org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(itemBySlot)
+                );
+                event.callEvent();
+                if (event.isCancelled()) {
+                    ((org.bukkit.entity.Player) player.getBukkitEntity()).updateInventory();
+                    return InteractionResult.FAIL;
+                }
+                // Paper end
                 if (!player.level().isClientSide()) {
                     player.awardStat(Stats.ITEM_USED.get(stack.getItem()));
                 }
diff --git a/net/minecraft/world/level/block/ComparatorBlock.java b/net/minecraft/world/level/block/ComparatorBlock.java
index 578e7c8dc601f53cf2a068685c6901bfaf47e763..7e87d56cd42f27c115b575b9acf2ce9ac0953af8 100644
--- a/net/minecraft/world/level/block/ComparatorBlock.java
+++ b/net/minecraft/world/level/block/ComparatorBlock.java
@@ -136,7 +136,7 @@ public class ComparatorBlock extends DiodeBlock implements EntityBlock {
             float f = state.getValue(MODE) == ComparatorMode.SUBTRACT ? 0.55F : 0.5F;
             level.playSound(player, pos, SoundEvents.COMPARATOR_CLICK, SoundSource.BLOCKS, 0.3F, f);
             level.setBlock(pos, state, Block.UPDATE_CLIENTS);
-            this.refreshOutputState(level, pos, state);
+            if (level.getBlockState(pos).is(this)) { this.refreshOutputState(level, pos, state); } // Paper
             return InteractionResult.SUCCESS;
         }
     }
diff --git a/net/minecraft/world/level/block/DiodeBlock.java b/net/minecraft/world/level/block/DiodeBlock.java
index 02ffb5569e2405d86b3a4a695dd17c9372169ff7..a750e2d7f6c5ac7561c8b06870e022b07875c587 100644
--- a/net/minecraft/world/level/block/DiodeBlock.java
+++ b/net/minecraft/world/level/block/DiodeBlock.java
@@ -92,6 +92,7 @@ public abstract class DiodeBlock extends HorizontalDirectionalBlock {
 
     @Override
     protected void neighborChanged(BlockState state, Level level, BlockPos pos, Block neighborBlock, @Nullable Orientation orientation, boolean movedByPiston) {
+        if (!level.getBlockState(pos).is(this)) return; // Paper
         if (state.canSurvive(level, pos)) {
             this.checkTickOnNeighbor(level, pos, state);
         } else {
diff --git a/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java b/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
index c8bfa1eeb847d6d9972d2f6625cc1f60063713d4..1215ac44fd30dc872f9d319be34e5e08ba880ea2 100644
--- a/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
@@ -490,7 +490,7 @@ public abstract class AbstractFurnaceBlockEntity extends BaseContainerBlockEntit
             event = new org.bukkit.event.inventory.FurnaceExtractEvent(
                 serverPlayer.getBukkitEntity(),
                 org.bukkit.craftbukkit.block.CraftBlock.at(level, blockPos),
-                org.bukkit.craftbukkit.inventory.CraftItemType.minecraftToBukkit(itemStack.getItem()),
+                itemStack.asBukkitCopy(),
                 amount,
                 floor
             );
