From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Thu, 9 Apr 2026 14:49:29 -0700
Subject: [PATCH] Fix Riding Stats


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 44ec79d6c7f24c019c3416b791358d1858b503ec..a3a8d17d6ebb73b5ceed498e06fd56ee33c614f9 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -4125,6 +4125,11 @@ public abstract class Entity
         }
 
         public void adjustRiders(boolean teleport) {
+        // Canvas start - region threading - fix ride stats
+            this.adjustRiders(teleport, false);
+        }
+        public void adjustRiders(boolean teleport, boolean trackVehicleMoveStats) {
+        // Canvas end - region threading - fix ride stats
             java.util.ArrayDeque<EntityTreeNode> queue = new java.util.ArrayDeque<>();
             queue.add(this);
 
@@ -4137,14 +4142,25 @@ public abstract class Entity
 
                 for (EntityTreeNode passenger : passengers) {
                     queue.add(passenger);
+                    // Canvas start - region threading - fix ride stats
+                    // this is required to fix the riding statistics for players
+                    // this is because of how Folia does their fix for repositioning riders
+                    // that makes it so it ignores checking riding statistics
+                    double x = passenger.root.getX();
+                    double y = passenger.root.getY();
+                    double z = passenger.root.getZ();
                     curr.root.positionRider(passenger.root, teleport ? Entity::snapTo : Entity::setPos);
+                    if (trackVehicleMoveStats && passenger.root instanceof ServerPlayer player) {
+                        player.checkRidingStatistics(player.getX() - x, player.getY() - y, player.getZ() - z);
+                    }
+                    // Canvas end - region threading - fix ride stats
                 }
             }
         }
     }
 
     public void repositionAllPassengers(boolean teleport) {
-        this.makePassengerTree().adjustRiders(teleport);
+        this.makePassengerTree().adjustRiders(teleport, true); // Canvas - region threading - fix ride stats
     }
 
     protected EntityTreeNode makePassengerTree() {
