From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Tue, 28 Apr 2026 01:23:10 -0700
Subject: [PATCH] Per world Canvas configs


diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 874abf77ec74693dffa3c2a97387ecaaa4d7f79c..e31275091ebb11b6c1764d75e2c36507cadf6229 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -159,6 +159,17 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         return this.paperConfig;
     }
     // Paper end - add paper world config
+    // Canvas start - world config
+    private io.canvasmc.canvas.WorldConfig canvasConfig;
+
+    public io.canvasmc.canvas.WorldConfig canvasConfig() {
+        return this.canvasConfig;
+    }
+
+    public void reloadCanvasConfig() {
+        this.canvasConfig = io.canvasmc.canvas.WorldConfig.buildForWorld((ServerLevel) this, dimension);
+    }
+    // Canvas end - world config
 
     public static @Nullable BlockPos lastPhysicsProblem; // Spigot
     //public final Map<ServerExplosion.CacheKey, Float> explosionDensityCache = new java.util.HashMap<>(); // Paper - Optimize explosions // Folia - region threading
@@ -887,6 +898,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         this.damageSources = new DamageSources(registryAccess);
         this.entityLookup = new ca.spottedleaf.moonrise.patches.chunk_system.level.entity.dfl.DefaultEntityLookup(this); // Paper - rewrite chunk system
         this.chunkPacketBlockController = this.paperConfig().anticheat.antiXray.enabled ? new io.papermc.paper.antixray.ChunkPacketBlockControllerAntiXray(this, executor) : io.papermc.paper.antixray.ChunkPacketBlockController.NO_OPERATION_INSTANCE; // Paper - Anti-Xray
+        reloadCanvasConfig(); // Canvas - world config
     }
 
     // Paper start - Cancel hit for vanished players
