From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Tue, 28 Apr 2026 01:23:10 -0700
Subject: [PATCH] Per world Canvas configs


diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 874abf77ec74693dffa3c2a97387ecaaa4d7f79c..8c4e4e244a4cfdd35a47061052b62c5aea2c49aa 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -159,6 +159,12 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         return this.paperConfig;
     }
     // Paper end - add paper world config
+    // Canvas start - world config
+    private final io.canvasmc.canvas.WorldConfig canvasConfig;
+    private io.canvasmc.canvas.WorldConfig canvasConfig() {
+        return this.canvasConfig;
+    }
+    // Canvas end - world config
 
     public static @Nullable BlockPos lastPhysicsProblem; // Spigot
     //public final Map<ServerExplosion.CacheKey, Float> explosionDensityCache = new java.util.HashMap<>(); // Paper - Optimize explosions // Folia - region threading
@@ -865,6 +871,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         final org.bukkit.NamespacedKey worldKey = CraftNamespacedKey.fromMinecraft(dimension.identifier()); // Paper
         this.spigotConfig = new org.spigotmc.SpigotWorldConfig(bukkitName, worldKey); // Spigot
         this.paperConfig = paperWorldConfigCreator.apply(this.spigotConfig); // Paper - create paper world config
+        this.canvasConfig = io.canvasmc.canvas.WorldConfig.buildForWorld((ServerLevel) this, dimension); // Canvas - world config
         this.generator = generator;
         this.world = new CraftWorld((ServerLevel) this, worldKey, biomeProvider, environment);
 
