From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Fri, 19 Jun 2026 21:35:59 -0700
Subject: [PATCH] Canvas RegionizedWorldData


diff --git a/io/papermc/paper/threadedregions/RegionizedWorldData.java b/io/papermc/paper/threadedregions/RegionizedWorldData.java
index 00d135681846f106a15578bd86907e6941fcf977..e979793795a358e3fe0f1671d312e9730cfc1872 100644
--- a/io/papermc/paper/threadedregions/RegionizedWorldData.java
+++ b/io/papermc/paper/threadedregions/RegionizedWorldData.java
@@ -323,6 +323,20 @@ public final class RegionizedWorldData {
     };
 
     public final ServerLevel world;
+    // Canvas start - canvas regionized data
+
+    // this is because I don't want to deal with an insane amount of diff constantly
+    // because of canvas adding a ton of regionized data
+
+    private io.canvasmc.canvas.threadedregions.CanvasRegionizedWorldData canvasWorldData;
+
+    public io.canvasmc.canvas.threadedregions.CanvasRegionizedWorldData getCanvasWorldData() {
+        if (this.canvasWorldData == null) {
+            this.canvasWorldData = io.canvasmc.canvas.threadedregions.CanvasRegionizedWorldData.getCurrentWorldData();
+        }
+        return canvasWorldData;
+    }
+    // Canvas end - canvas regionized data
 
     private RegionizedServer.WorldLevelData tickData;
 
