From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Fri, 19 Jun 2026 21:35:59 -0700
Subject: [PATCH] Canvas RegionizedWorldData


diff --git a/io/canvasmc/canvas/threadedregions/CanvasRegionizedWorldData.java b/io/canvasmc/canvas/threadedregions/CanvasRegionizedWorldData.java
new file mode 100644
index 0000000000000000000000000000000000000000..c14a3c84a3fca89f3e386c15b322042bc97aa025
--- /dev/null
+++ b/io/canvasmc/canvas/threadedregions/CanvasRegionizedWorldData.java
@@ -0,0 +1,101 @@
+package io.canvasmc.canvas.threadedregions;
+
+import io.canvasmc.canvas.GlobalConfiguration;
+import io.canvasmc.canvas.region.RegionTickData;
+import io.canvasmc.canvas.world.RegionResourceBar;
+import io.canvasmc.canvas.world.RegionizedRamBar;
+import io.canvasmc.canvas.world.RegionizedTpsBar;
+import io.canvasmc.canvas.world.chunk.NatureSpawnChunkMap;
+import io.papermc.paper.threadedregions.RegionizedWorldData;
+import io.papermc.paper.threadedregions.ThreadedRegionizer;
+import io.papermc.paper.threadedregions.TickRegionScheduler;
+import io.papermc.paper.threadedregions.TickRegions;
+import it.unimi.dsi.fastutil.longs.Long2ReferenceOpenHashMap;
+import it.unimi.dsi.fastutil.longs.LongOpenHashSet;
+import it.unimi.dsi.fastutil.longs.LongSet;
+import it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.util.RandomSource;
+import net.minecraft.world.level.material.FlowingFluid;
+import org.bukkit.craftbukkit.CraftWorld;
+import org.jspecify.annotations.NonNull;
+import org.jspecify.annotations.Nullable;
+
+public class CanvasRegionizedWorldData {
+    private static final RegionTickData.IRegionizedData<CanvasRegionizedWorldData> CANVAS_REGION_DATA = TickRegions.TickRegionData.createRegionizedDataApiHandle(
+        (regionTickData, world) -> new CanvasRegionizedWorldData((TickRegions.TickRegionData) regionTickData, ((CraftWorld) world).getHandle()),
+        new RegionTickData.IRegionizedData.IRegionizedCallback<>() {
+            @Override
+            public void merge(
+                final @NonNull CanvasRegionizedWorldData from,
+                final @NonNull CanvasRegionizedWorldData into,
+                final long fromTickOffset
+            ) {
+                // no-op currently
+            }
+
+            @Override
+            public void split(
+                final @NonNull CanvasRegionizedWorldData from,
+                final int chunkToRegionShift,
+                final @NonNull Long2ReferenceOpenHashMap<CanvasRegionizedWorldData> regionToData,
+                final @NonNull ReferenceOpenHashSet<CanvasRegionizedWorldData> dataSet
+            ) {
+                // no-op currently
+            }
+        }
+    );
+
+    public static @Nullable CanvasRegionizedWorldData getCurrentWorldData() {
+        final ThreadedRegionizer.ThreadedRegion<TickRegions.TickRegionData, TickRegions.TickRegionSectionData>
+            region = TickRegionScheduler.getCurrentRegion();
+        if (region == null) {
+            return null;
+        }
+        return region.getData().getOrCreateFromIRegionizedData(CANVAS_REGION_DATA);
+    }
+
+    private final TickRegions.TickRegionData regionData;
+    private final ServerLevel level;
+    private final RegionizedWorldData worldData;
+
+    // this is transient, from Leaf optimized flowing fluids
+    public LongSet slopeDistanceCacheVisited = new LongOpenHashSet(512);
+    public FlowingFluid.SlopeDistanceNodeDeque slopeDistanceCacheQueue = new FlowingFluid.SlopeDistanceNodeDeque();
+
+    // natural spawn chunk optimizations from Leaf
+    public final NatureSpawnChunkMap natureSpawnChunkMap = new NatureSpawnChunkMap();
+
+    public final RandomSource simpleUnsafeLocalRandom = GlobalConfiguration.createFastRandom();
+
+    // tpsbar and rambar
+    public final RegionResourceBar tpsbar;
+    public final RegionResourceBar rambar;
+
+    public long projectilesLoadedTick = 0;
+    public long projectilesLoadedThisTick = 0;
+
+    private CanvasRegionizedWorldData(final TickRegions.TickRegionData regionData, final ServerLevel level) {
+        this.regionData = regionData;
+        this.level = level;
+
+        final RegionizedWorldData worldData = TickRegionScheduler.getCurrentRegionizedWorldData();
+
+        this.tpsbar = new RegionizedTpsBar();
+        this.rambar = new RegionizedRamBar();
+
+        this.worldData = worldData;
+    }
+
+    public ServerLevel getLevel() {
+        return level;
+    }
+
+    public RegionizedWorldData getFoliaWorldData() {
+        return worldData;
+    }
+
+    public TickRegions.TickRegionData getRegionData() {
+        return regionData;
+    }
+}
diff --git a/io/papermc/paper/threadedregions/RegionizedWorldData.java b/io/papermc/paper/threadedregions/RegionizedWorldData.java
index 02224c5b8466dfaf985eb9f77b2d2b504ea9286b..6810b850bdbf4a936488d3cdc48b830948e758ba 100644
--- a/io/papermc/paper/threadedregions/RegionizedWorldData.java
+++ b/io/papermc/paper/threadedregions/RegionizedWorldData.java
@@ -321,6 +321,20 @@ public final class RegionizedWorldData {
     };
 
     public final ServerLevel world;
+    // Canvas start - canvas regionized data
+
+    // this is because I don't want to deal with an insane amount of diff constantly
+    // because of canvas adding a ton of regionized data
+
+    private io.canvasmc.canvas.threadedregions.CanvasRegionizedWorldData canvasWorldData;
+
+    public io.canvasmc.canvas.threadedregions.CanvasRegionizedWorldData getCanvasWorldData() {
+        if (this.canvasWorldData == null) {
+            this.canvasWorldData = io.canvasmc.canvas.threadedregions.CanvasRegionizedWorldData.getCurrentWorldData();
+        }
+        return canvasWorldData;
+    }
+    // Canvas end - canvas regionized data
 
     private RegionizedServer.WorldLevelData tickData;
 
