From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Sat, 20 Dec 2025 12:33:05 -0800
Subject: [PATCH] Replace Moonrise Executor


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java
index 5068de75f6f38c0459b77eb7354ce9db2d251d2c..b6b98c125ed4c66c33c5000d03a1708a28fb47c0 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java
@@ -108,11 +108,13 @@ public final class ChunkTaskScheduler {
 
     public final ServerLevel world;
     public final AreaDependentQueue radiusAwareScheduler;
-    public final BalancedPrioritisedThreadPool.OrderedStreamGroup.Queue parallelGenExecutor;
-    public final BalancedPrioritisedThreadPool.OrderedStreamGroup.Queue loadExecutor;
-    public final BalancedPrioritisedThreadPool.OrderedStreamGroup.Queue ioExecutor;
-    public final BalancedPrioritisedThreadPool.OrderedStreamGroup.Queue compressionExecutor;
-    public final BalancedPrioritisedThreadPool.OrderedStreamGroup.Queue saveExecutor;
+    // Canvas start - replace moonrise executor
+    public final io.canvasmc.canvas.world.chunk.TheChunkSystem.OrderedStreamGroup.Queue parallelGenExecutor;
+    public final io.canvasmc.canvas.world.chunk.TheChunkSystem.OrderedStreamGroup.Queue loadExecutor;
+    public final io.canvasmc.canvas.world.chunk.TheChunkSystem.OrderedStreamGroup.Queue compressionExecutor;
+    public final io.canvasmc.canvas.world.chunk.TheChunkSystem.OrderedStreamGroup.Queue saveExecutor;
+    public final io.canvasmc.canvas.world.chunk.TheChunkSystem.OrderedStreamGroup.Queue ioExecutor;
+    // Canvas end - replace moonrise executor
 
     // Folia - regionised ticking
 
diff --git a/ca/spottedleaf/moonrise/patches/starlight/light/StarLightInterface.java b/ca/spottedleaf/moonrise/patches/starlight/light/StarLightInterface.java
index 7faf226b753a03e03b0dec9bbfcc94a6c3f5bc0f..74737c3e938df5cc509c6f34ea76715b6e553801 100644
--- a/ca/spottedleaf/moonrise/patches/starlight/light/StarLightInterface.java
+++ b/ca/spottedleaf/moonrise/patches/starlight/light/StarLightInterface.java
@@ -891,10 +891,7 @@ public final class StarLightInterface {
             public ServerChunkTasks(final long chunkCoordinate, final StarLightInterface lightEngine,
                                     final ServerLightQueue queue, final Priority priority) {
                 super(chunkCoordinate, lightEngine, queue);
-                this.task = ((ChunkSystemServerLevel)(ServerLevel)lightEngine.getWorld()).moonrise$getChunkTaskScheduler().radiusAwareScheduler.createTask(
-                        CoordinateUtils.getChunkX(chunkCoordinate), CoordinateUtils.getChunkZ(chunkCoordinate),
-                        ((ChunkSystemChunkStatus)ChunkStatus.LIGHT).moonrise$getWriteRadius(), this, priority
-                );
+                this.task = ((ChunkSystemServerLevel) lightEngine.getWorld()).moonrise$getChunkTaskScheduler().parallelGenExecutor.createTask(this, priority, CoordinateUtils.getChunkX(chunkCoordinate), CoordinateUtils.getChunkZ(chunkCoordinate)); // Canvas - replace moonrise executor
             }
 
             public boolean markTicketAdded() {
