From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Mon, 20 Apr 2026 11:48:28 -0700
Subject: [PATCH] Replace Moonrise Executor


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java
index 693a43a0639a66c200f37a576077db4626cc2d39..e83c6129e2e637324a7021787c44a21c292b00b4 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java
@@ -108,11 +108,13 @@ public final class ChunkTaskScheduler {
 
     public final ServerLevel world;
     public final AreaDependentQueue radiusAwareScheduler;
-    public final BalancedPrioritisedThreadPool.OrderedStreamGroup.Queue parallelGenExecutor;
-    public final BalancedPrioritisedThreadPool.OrderedStreamGroup.Queue loadExecutor;
-    public final BalancedPrioritisedThreadPool.OrderedStreamGroup.Queue ioExecutor;
-    public final BalancedPrioritisedThreadPool.OrderedStreamGroup.Queue compressionExecutor;
-    public final BalancedPrioritisedThreadPool.OrderedStreamGroup.Queue saveExecutor;
+    // Canvas start - replace moonrise executor
+    public final io.canvasmc.canvas.world.chunk.BalancedChunkSystem.OrderedStreamGroup.Queue parallelGenExecutor;
+    public final io.canvasmc.canvas.world.chunk.BalancedChunkSystem.OrderedStreamGroup.Queue loadExecutor;
+    public final io.canvasmc.canvas.world.chunk.BalancedChunkSystem.OrderedStreamGroup.Queue compressionExecutor;
+    public final io.canvasmc.canvas.world.chunk.BalancedChunkSystem.OrderedStreamGroup.Queue saveExecutor;
+    public final io.canvasmc.canvas.world.chunk.BalancedChunkSystem.OrderedStreamGroup.Queue ioExecutor;
+    // Canvas end - replace moonrise executor
 
     // Folia - region threading
 
