From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Mon, 20 Apr 2026 12:05:53 -0700
Subject: [PATCH] PR Party 2


diff --git a/net/minecraft/commands/Commands.java b/net/minecraft/commands/Commands.java
index a3c60500d3aa85b9b824f5a02261d6aaa4dae30b..589671edefaf2bf6cbe43fec25f4032a9bebaa13 100644
--- a/net/minecraft/commands/Commands.java
+++ b/net/minecraft/commands/Commands.java
@@ -204,7 +204,7 @@ public class Commands {
         //DataPackCommand.register(this.dispatcher, context); // Folia - region threading - TODO
         //DebugCommand.register(this.dispatcher); // Folia - region threading - TODO
         DefaultGameModeCommands.register(this.dispatcher);
-        //DialogCommand.register(this.dispatcher, context); // Folia - region threading - TODO
+        DialogCommand.register(this.dispatcher, context);
         DifficultyCommand.register(this.dispatcher);
         EffectCommands.register(this.dispatcher, context);
         EmoteCommands.register(this.dispatcher);
diff --git a/net/minecraft/server/commands/DialogCommand.java b/net/minecraft/server/commands/DialogCommand.java
index fcbe76e95a7de7c79afd0a116b6cbcf286ba85ff..6c544e5d79a7d076f9dd5212891d6d22b433f9b5 100644
--- a/net/minecraft/server/commands/DialogCommand.java
+++ b/net/minecraft/server/commands/DialogCommand.java
@@ -46,7 +46,11 @@ public class DialogCommand {
 
     private static int showDialog(final CommandSourceStack sender, final Collection<ServerPlayer> targets, final Holder<Dialog> dialog) {
         for (ServerPlayer target : targets) {
-            target.openDialog(dialog);
+            // Folia start - region threading
+            target.getBukkitEntity().taskScheduler.scheduleOrExecute((ServerPlayer serverPlayer) -> {
+                serverPlayer.openDialog(dialog);
+            });
+            // Folia end - region threading
         }
 
         if (targets.size() == 1) {
@@ -60,7 +64,11 @@ public class DialogCommand {
 
     private static int clearDialog(final CommandSourceStack sender, final Collection<ServerPlayer> targets) {
         for (ServerPlayer target : targets) {
-            target.connection.send(ClientboundClearDialogPacket.INSTANCE);
+            // Folia start - region threading
+            target.getBukkitEntity().taskScheduler.scheduleOrExecute((ServerPlayer serverPlayer) -> {
+                serverPlayer.connection.send(ClientboundClearDialogPacket.INSTANCE);
+            });
+            // Folia end - region threading
         }
 
         if (targets.size() == 1) {
