From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Mon, 20 Apr 2026 11:48:28 -0700
Subject: [PATCH] Replace Moonrise Executor


diff --git a/ca/spottedleaf/concurrentutil/executor/thread/BalancedPrioritisedThreadPool.java b/ca/spottedleaf/concurrentutil/executor/thread/BalancedPrioritisedThreadPool.java
index 67d083bfd4afe187d2c997b49bbce36bba94c790..0c3d7a9b4a65a0550755b40d8c5fa821bf5ef49d 100644
--- a/ca/spottedleaf/concurrentutil/executor/thread/BalancedPrioritisedThreadPool.java
+++ b/ca/spottedleaf/concurrentutil/executor/thread/BalancedPrioritisedThreadPool.java
@@ -309,7 +309,7 @@ public class BalancedPrioritisedThreadPool {
         }
     }
 
-    public final class OrderedStreamGroup {
+    public final class OrderedStreamGroup implements io.canvasmc.canvas.world.chunk.StreamGroup { // Canvas - replace moonrise executor
 
         private final AtomicLong subOrderGenerator;
         private final boolean streamCompare;
@@ -387,7 +387,7 @@ public class BalancedPrioritisedThreadPool {
             }
         }
 
-        public final class Queue implements PrioritisedExecutor {
+        public final class Queue implements PrioritisedExecutor, io.canvasmc.canvas.world.chunk.ChunkSystemExecutor { // Canvas - replace moonrise executor
 
             private final PrioritisedTaskQueue wrapped;
             private volatile boolean halt;
diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java
index 5068de75f6f38c0459b77eb7354ce9db2d251d2c..bf8f07de2e0bc1e0f43814b5b3888781bc4718a3 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkTaskScheduler.java
@@ -108,11 +108,13 @@ public final class ChunkTaskScheduler {
 
     public final ServerLevel world;
     public final AreaDependentQueue radiusAwareScheduler;
-    public final BalancedPrioritisedThreadPool.OrderedStreamGroup.Queue parallelGenExecutor;
-    public final BalancedPrioritisedThreadPool.OrderedStreamGroup.Queue loadExecutor;
-    public final BalancedPrioritisedThreadPool.OrderedStreamGroup.Queue ioExecutor;
-    public final BalancedPrioritisedThreadPool.OrderedStreamGroup.Queue compressionExecutor;
-    public final BalancedPrioritisedThreadPool.OrderedStreamGroup.Queue saveExecutor;
+    // Canvas start - replace moonrise executor
+    public final io.canvasmc.canvas.world.chunk.ChunkSystemExecutor parallelGenExecutor;
+    public final io.canvasmc.canvas.world.chunk.ChunkSystemExecutor loadExecutor;
+    public final io.canvasmc.canvas.world.chunk.ChunkSystemExecutor compressionExecutor;
+    public final io.canvasmc.canvas.world.chunk.ChunkSystemExecutor saveExecutor;
+    public final io.canvasmc.canvas.world.chunk.ChunkSystemExecutor ioExecutor;
+    // Canvas end - replace moonrise executor
 
     // Folia - regionised ticking
 
