From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Sat, 17 Jan 2026 21:01:05 -0800
Subject: [PATCH] PR Party 3


diff --git a/net/minecraft/commands/Commands.java b/net/minecraft/commands/Commands.java
index 668ce850c9197b0f52859d74a8d3a8a2d3a390e3..d86d5a8e9abc5f941d89fed410db323bb57b7dab 100644
--- a/net/minecraft/commands/Commands.java
+++ b/net/minecraft/commands/Commands.java
@@ -241,7 +241,7 @@ public class Commands {
         SetBlockCommand.register(this.dispatcher, context);
         SetSpawnCommand.register(this.dispatcher);
         SetWorldSpawnCommand.register(this.dispatcher);
-        //SpectateCommand.register(this.dispatcher); // Folia - region threading - TODO later
+        SpectateCommand.register(this.dispatcher);
         //SpreadPlayersCommand.register(this.dispatcher); // Folia - region threading - TODO later
         StopSoundCommand.register(this.dispatcher);
         StopwatchCommand.register(this.dispatcher);
diff --git a/net/minecraft/server/commands/SpectateCommand.java b/net/minecraft/server/commands/SpectateCommand.java
index bf3b0bb46d623afd538cf4bc675eb801aca40e38..3edbebcc3a4b8f235cea3b5cfc353df501fe4907 100644
--- a/net/minecraft/server/commands/SpectateCommand.java
+++ b/net/minecraft/server/commands/SpectateCommand.java
@@ -43,7 +43,17 @@ public class SpectateCommand {
         );
     }
 
-    private static int spectate(CommandSourceStack source, @Nullable Entity target, ServerPlayer player) throws CommandSyntaxException {
+    // Folia start - region threading
+    private static void sendMessage(CommandSourceStack src, CommandSyntaxException ex) {
+        src.sendFailure((Component)ex.getRawMessage());
+    }
+    // Folia end - region threading
+
+    private static int spectate(CommandSourceStack source, @Nullable Entity target, ServerPlayer spectator) throws CommandSyntaxException { // Folia - region threading
+        // Folia start - region threading
+        spectator.getBukkitEntity().taskScheduler.schedule((ServerPlayer player) -> {
+            try {
+                // Folia end - region threading
         if (player == target) {
             throw ERROR_SELF.create();
         } else if (!player.isSpectator()) {
@@ -58,7 +68,14 @@ public class SpectateCommand {
                 source.sendSuccess(() -> Component.translatable("commands.spectate.success.stopped"), false);
             }
 
-            return 1;
+            return; // Folia - region threading
         }
+        // Folia start - region threading
+            } catch (CommandSyntaxException ex) {
+                sendMessage(source, ex);
+            }
+        }, null, 1L);
+        return 0;
+        // Folia end - region threading
     }
 }
