From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Mon, 20 Apr 2026 12:08:18 -0700
Subject: [PATCH] PR Party 4


diff --git a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
index 8dc8c6cfc01be3787142a117ef2b98f5294a8cef..4f48a29023714e34ea2cff799bbeb9c4f0e4d741 100644
--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
@@ -262,7 +262,7 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
             int previousChunkZ = SectionPos.blockToSectionCoord(this.position().z());
             Entity owner = this.owner != null ? findOwnerIncludingDeadPlayer(serverLevel, this.owner.getUUID()) : null;
             if (owner instanceof ServerPlayer serverPlayer
-                && !owner.isAlive()
+                && serverPlayer.getHealth() <= 0.0f // Folia - region threading
                 && !serverPlayer.wonGame
                 && serverPlayer.level().getGameRules().get(GameRules.ENDER_PEARLS_VANISH_ON_DEATH)) {
                 this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
