From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Sat, 17 Jan 2026 18:43:05 -0800
Subject: [PATCH] PR Party 4


diff --git a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
index 7d8c528a76e6a95d27368bc459d654c6ef66dd8f..a9c8f1b1c8bb32fbbc2048ac0f8a435922fa483e 100644
--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
@@ -257,7 +257,7 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
             int sectionPosZ = SectionPos.blockToSectionCoord(this.position().z());
             Entity entity = this.owner != null ? findOwnerIncludingDeadPlayer(serverLevel, this.owner.getUUID()) : null;
             if (entity instanceof ServerPlayer serverPlayer
-                && !entity.isAlive()
+                && serverPlayer.getHealth() <= 0.0f // Folia - region threading
                 && !serverPlayer.wonGame
                 && serverPlayer.level().getGameRules().get(GameRules.ENDER_PEARLS_VANISH_ON_DEATH)) {
                 this.discard(org.bukkit.event.entity.EntityRemoveEvent.Cause.DESPAWN); // CraftBukkit - add Bukkit remove cause
