From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Mon, 20 Apr 2026 13:27:03 -0700
Subject: [PATCH] Fixup Region Threading


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
index 961a9c3a31a5cabc250932df823b9e3f1e6e9662..59fe9191f09d5eb1001c5074cf925b1820763e18 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
@@ -256,7 +256,12 @@ public final class ChunkHolderManager {
         // Folia start - region threading
         this.close(save, halt, true, true, true);
     }
+    // Canvas start - region threading
     public void close(final boolean save, final boolean halt, final boolean first, final boolean last, final boolean checkRegions) {
+        this.close(save, halt, first, last, checkRegions, true);
+    }
+    public void close(final boolean save, final boolean halt, final boolean first, final boolean last, final boolean checkRegions, final boolean flush) {
+    // Canvas end - region threading
         // Folia end - region threading
         TickThread.ensureTickThread("Closing world off-main");
         if (first && halt) { // Folia - region threading
@@ -269,7 +274,7 @@ public final class ChunkHolderManager {
         }
 
         if (save) {
-            this.saveAllChunksRegionised(true, true, true, false, first, last, checkRegions); // Folia - region threading
+            this.saveAllChunksRegionised(flush, true, true, false, first, last, checkRegions); // Folia - region threading // Canvas - region threading
         }
 
         if (last) { // Folia - region threading
diff --git a/io/papermc/paper/threadedregions/RegionShutdownThread.java b/io/papermc/paper/threadedregions/RegionShutdownThread.java
index f2208bcd5ccf3bd5a428e8ea3bf3ff26b3c9483b..0da94ecfe36b11ca37945dcb8952fb9df87521a9 100644
--- a/io/papermc/paper/threadedregions/RegionShutdownThread.java
+++ b/io/papermc/paper/threadedregions/RegionShutdownThread.java
@@ -148,9 +148,48 @@ public final class RegionShutdownThread extends ca.spottedleaf.moonrise.common.u
             this.shuttingDown = null;
         }
     }
+    // Canvas start - region threading
+
+    private static volatile boolean IS_IN_SHUTDOWN = false;
+
+    public static boolean canvas$isShutdown() {
+        return IS_IN_SHUTDOWN;
+    }
+
+    private void savePlayers() {
+        net.minecraft.server.players.PlayerList playerList = MinecraftServer.getServer().getPlayerList();
+        LOGGER.info("Removing players");
+        // waiting for 100ms is a stupid fix
+        List<java.util.concurrent.CompletableFuture<Void>> futures = new ArrayList<>();
+        // prepare futures
+        for (final ServerPlayer serverPlayer : playerList.players) {
+            final net.minecraft.network.Connection connection = serverPlayer.connection.connection;
+            if (connection.channel == null || !connection.channel.isOpen()) continue;
+            java.util.concurrent.CompletableFuture<Void> future = new java.util.concurrent.CompletableFuture<>();
+            connection.channel.closeFuture().addListener(f -> future.complete(null));
+            futures.add(future);
+        }
+
+        // now, remove players and wait
+        playerList.removeAll(MinecraftServer.getServer().isRestarting);
+
+        // wait for player removal
+        if (!futures.isEmpty()) {
+            if (io.canvasmc.canvas.util.Util.waitFor(java.util.concurrent.CompletableFuture.allOf(futures.toArray(new java.util.concurrent.CompletableFuture[0])),
+                java.util.concurrent.TimeUnit.SECONDS, 20)) {
+                LOGGER.info("Players removed");
+            } else {
+                LOGGER.warn("Players did not all get removed within 20 seconds, proceeding with shutdown anyways");
+            }
+        }
+
+        MinecraftServer.getServer().getConnection().stop();
+    }
+    // Canvas end - region threading
 
     @Override
     public final void run() {
+        IS_IN_SHUTDOWN = true; // Canvas - region threading
         // await scheduler termination
         LOGGER.info("Awaiting scheduler termination for 60s...");
         if (TickRegions.getScheduler().halt(true, TimeUnit.SECONDS.toNanos(60L))) {
@@ -161,6 +200,11 @@ public final class RegionShutdownThread extends ca.spottedleaf.moonrise.common.u
         }
 
         MinecraftServer.getServer().stopServer(); // stop part 1: most logic, kicking players, plugins, etc
+        // Canvas start - region threading
+        // this.playerList.removeAll(this.isRestarting); // Paper
+        // try { Thread.sleep(100); } catch (InterruptedException ex) {}
+        this.savePlayers();
+        // Canvas end - region threading
         // halt all chunk systems first so that any in-progress chunk generation stops
         LOGGER.info("Halting chunk systems...");
         for (final ServerLevel world : MinecraftServer.getServer().getAllLevels()) {
diff --git a/io/papermc/paper/threadedregions/RegionizedServer.java b/io/papermc/paper/threadedregions/RegionizedServer.java
index c4f6053762d0280e8299b38526455955cc42b97a..158ee7f333bc6726aaae6ed24b8d07433ad34212 100644
--- a/io/papermc/paper/threadedregions/RegionizedServer.java
+++ b/io/papermc/paper/threadedregions/RegionizedServer.java
@@ -32,7 +32,7 @@ import java.util.function.BooleanSupplier;
 
 public final class RegionizedServer {
 
-    private static final Logger LOGGER = LogUtils.getLogger();
+    private static final Logger LOGGER = LogUtils.getClassLogger(); // Canvas - region threading
     private static final RegionizedServer INSTANCE = new RegionizedServer();
 
     public final RegionizedTaskQueue taskQueue = new RegionizedTaskQueue();
@@ -67,7 +67,7 @@ public final class RegionizedServer {
         new RegionizedServerInitEvent().callEvent();
 
         // now we can schedule
-        this.tickHandle.setInitialStart(System.nanoTime() + TickRegionScheduler.TIME_BETWEEN_TICKS);
+        this.tickHandle.setInitialStart(System.nanoTime() + TickRegionScheduler.getTimeBetweenTicks()); // Canvas - region threading
         TickRegions.getScheduler().scheduleRegion(this.tickHandle);
         TickRegions.start();
     }
@@ -481,6 +481,7 @@ public final class RegionizedServer {
     }
 
     private void advanceWeatherCycle(final ServerLevel world) {
+        if (!getGlobalTickData().getTickManager().doesRunGameElements()) return; // Canvas - region threading
         world.advanceWeatherCycle();
     }
 
@@ -497,6 +498,7 @@ public final class RegionizedServer {
     }
 
     private void tickTime(final ServerLevel world, final long tickCount) {
+        if (!getGlobalTickData().getTickManager().doesRunGameElements()) return; // Canvas - region threading
         if (world.tickTime) {
             world.serverLevelData.setGameTime(world.serverLevelData.getGameTime() + tickCount);
         }
diff --git a/io/papermc/paper/threadedregions/RegionizedWorldData.java b/io/papermc/paper/threadedregions/RegionizedWorldData.java
index bc946ce8f24d7654eefd77a99ac9d037f3d0fb7e..f8ed575c12cdf0608d3af27d306a70bf4391222c 100644
--- a/io/papermc/paper/threadedregions/RegionizedWorldData.java
+++ b/io/papermc/paper/threadedregions/RegionizedWorldData.java
@@ -65,7 +65,7 @@ import java.util.function.Predicate;
 
 public final class RegionizedWorldData {
 
-    private static final Logger LOGGER = LogUtils.getLogger();
+    private static final Logger LOGGER = LogUtils.getClassLogger(); // Canvas - region threading
 
     private static final Entity[] EMPTY_ENTITY_ARRAY = new Entity[0];
 
@@ -477,7 +477,7 @@ public final class RegionizedWorldData {
         this.hasEntityMoveEvent = io.papermc.paper.event.entity.EntityMoveEvent.getHandlerList().getRegisteredListeners().length > 0; // Paper - Add EntityMoveEvent
         this.skipHopperEvents = this.world.paperConfig().hopper.disableMoveEvent || org.bukkit.event.inventory.InventoryMoveItemEvent.getHandlerList().getRegisteredListeners().length == 0; // Paper - Perf: Optimize Hoppers
         // always subtract from server init so that the tick starts at zero, allowing us to cast to int without much worry
-        this.lagCompensationTick = (System.nanoTime() - MinecraftServer.SERVER_INIT) / TickRegionScheduler.TIME_BETWEEN_TICKS;
+        this.lagCompensationTick = (System.nanoTime() - MinecraftServer.SERVER_INIT) / TickRegionScheduler.getTimeBetweenTicks(); // Canvas - region threading
     }
 
     public NearbyPlayers getNearbyPlayers() {
diff --git a/io/papermc/paper/threadedregions/TickRegionScheduler.java b/io/papermc/paper/threadedregions/TickRegionScheduler.java
index 4c560afc4f25e911c88c6347857ab710b1052d8b..b5e551694f013bde5a6eee7511769ca79705cdb6 100644
--- a/io/papermc/paper/threadedregions/TickRegionScheduler.java
+++ b/io/papermc/paper/threadedregions/TickRegionScheduler.java
@@ -28,7 +28,7 @@ import java.util.function.BooleanSupplier;
 
 public final class TickRegionScheduler {
 
-    private static final Logger LOGGER = LogUtils.getLogger();
+    public static final Logger LOGGER = LogUtils.getClassLogger(); // Canvas - region threading
     private static final ThreadMXBean THREAD_MX_BEAN = ManagementFactory.getThreadMXBean();
     private static final boolean MEASURE_CPU_TIME;
     static {
@@ -40,8 +40,41 @@ public final class TickRegionScheduler {
         }
     }
 
-    public static final int TICK_RATE = 20;
-    public static final long TIME_BETWEEN_TICKS = 1_000_000_000L / TICK_RATE; // ns
+    // Canvas start - region threading
+    private static final float DEFAULT_TICK_RATE = io.canvasmc.canvas.Config.INSTANCE.scheduler.defaultTickRate;
+    private static final long DEFAULT_TICK_TIME_NANOS = (long) (1_000_000_000 / DEFAULT_TICK_RATE);
+
+    private static float CURRENT_TICK_RATE = DEFAULT_TICK_RATE;
+    private static long CURRENT_TIME_BETWEEN_TICKS = DEFAULT_TICK_TIME_NANOS;
+
+    // Note: used for setting new tick time nanos, not really for anything else, utility method
+    //    this is fine for decimal rates, as we round it so if anything, we are like 1 nanosecond off... which isn't bad
+    private static long tickTimeNanosFromRate(float rate) {
+        return Math.round(1_000_000_000 / rate);
+    }
+
+    public static void setTickRate(final float newRate) {
+        CURRENT_TICK_RATE = newRate;
+        CURRENT_TIME_BETWEEN_TICKS = tickTimeNanosFromRate(newRate);
+        io.canvasmc.canvas.tick.ScheduledHandleTickState.sendStateToAllPlayers();
+    }
+
+    public static float getTickRate() {
+        return CURRENT_TICK_RATE;
+    }
+
+    public static long getTimeBetweenTicks() {
+        return CURRENT_TIME_BETWEEN_TICKS;
+    }
+
+    public static float getDefaultTickRate() {
+        return DEFAULT_TICK_RATE;
+    }
+
+    public static long getDefaultTickTimeNanos() {
+        return DEFAULT_TICK_TIME_NANOS;
+    }
+    // Canvas end - region threading
     // Folia start - watchdog
     public static final FoliaWatchdogThread WATCHDOG_THREAD = new FoliaWatchdogThread();
     static {
@@ -53,6 +86,7 @@ public final class TickRegionScheduler {
 
     public static enum SchedulerType {
         EDF,
+        AFFINITY, // Canvas - add AFFINITY scheduler
         WORK_STEALING;
     }
 
@@ -68,32 +102,16 @@ public final class TickRegionScheduler {
             public Thread newThread(final Runnable run) {
                 final Thread ret = new TickThreadRunner(this.threadGroup, run, "Folia Region Scheduler Thread #" + this.idGenerator.getAndIncrement());
                 ret.setUncaughtExceptionHandler(TickRegionScheduler.this::uncaughtException);
+                ret.setPriority(8); // Canvas - bump priority to match main thread in Paper
                 return ret;
             }
         };
 
-        switch (schedulerType) {
-            case EDF: {
-                this.scheduler = new EDFSchedulerThreadPool(initialThreads, threadFactory);
-                break;
-            }
-            case WORK_STEALING: {
-                this.scheduler = new StealingScheduledThreadPool(
-                        threadFactory, MoonriseConstants.NUMA_ENABLE ? OSNuma.getNativeInstance() : OSNuma.NoOp.INSTANCE
-                );
-                ((StealingScheduledThreadPool)this.scheduler).setFlags(StealingScheduledThreadPool.FLAG_SCHEDULE_EVENLY);
-                break;
-            }
-            default: {
-                throw new IllegalStateException("Unknown scheduler type: " + schedulerType);
-            }
-        }
+        this.scheduler = io.canvasmc.canvas.tick.SchedulerUtil.decideScheduler(schedulerType,  initialThreads, threadFactory); // Canvas - region threading
     }
 
     public void start() {
-        if (this.scheduler instanceof EDFSchedulerThreadPool edfSchedulerThreadPool) {
-            edfSchedulerThreadPool.start();
-        }
+        io.canvasmc.canvas.tick.SchedulerUtil.startScheduler(); // Canvas - region threading
     }
 
     public void setThreads(final int threads) {
@@ -172,6 +190,16 @@ public final class TickRegionScheduler {
         }
         return tickThreadRunner.currentTickingRegion;
     }
+    // Canvas start - region threading
+
+    /**
+     * Returns if the current thread is a tick thread runner
+     */
+    public static boolean isTickRunner() {
+        final Thread currThread = Thread.currentThread();
+        return currThread instanceof TickThreadRunner;
+    }
+    // Canvas end - region threading
 
     /**
      * Returns the current ticking region's world regionised data, or {@code null} if there is no ticking region.
@@ -241,6 +269,16 @@ public final class TickRegionScheduler {
     private void uncaughtException(final Thread thread, final Throwable thr) {
         LOGGER.error("Uncaught exception in tick thread \"" + thread.getName() + "\"", thr);
 
+        // Canvas start - region threading
+        net.minecraft.CrashReport crashReport = net.minecraft.CrashReport.forThrowable(thr, "Region Crash");
+        net.minecraft.server.MinecraftServer.getServer().fillSystemReport(crashReport.getSystemReport());
+        java.nio.file.Path path = net.minecraft.server.MinecraftServer.getServer().getServerDirectory().resolve("crash-reports").resolve("crash-" + net.minecraft.util.Util.getFilenameFormattedDateTime() + "-server.txt");
+        if (crashReport.saveToFile(path, net.minecraft.ReportType.CRASH)) {
+            LOGGER.error("This crash report has been saved to: {}", path.toAbsolutePath());
+        } else {
+            LOGGER.error("We were unable to save this crash report to disk.");
+        }
+        // Canvas end - region threading
         // prevent further ticks from occurring
         // we CANNOT sync, because WE ARE ON A SCHEDULER THREAD
         this.scheduler.halt();
@@ -251,6 +289,19 @@ public final class TickRegionScheduler {
     private void regionFailed(final RegionScheduleHandle handle, final boolean executingTasks, final Throwable thr) {
         // when a region fails, we need to shut down the server gracefully
 
+        // Canvas start - region threading
+        net.minecraft.CrashReport crashReport = net.minecraft.CrashReport.forThrowable(thr, "Region Crash");
+        net.minecraft.server.MinecraftServer.getServer().fillSystemReport(crashReport.getSystemReport());
+        net.minecraft.CrashReportCategory crashReportCategory = crashReport.addCategory("Region Context");
+        crashReportCategory.setDetail("Executing Tasks", () -> String.valueOf(executingTasks));
+        crashReportCategory.setDetail("RegionScheduleHandle", handle.region == null ? "Not available" : handle.region.region);
+        java.nio.file.Path path = net.minecraft.server.MinecraftServer.getServer().getServerDirectory().resolve("crash-reports").resolve("crash-" + net.minecraft.util.Util.getFilenameFormattedDateTime() + "-server.txt");
+        if (crashReport.saveToFile(path, net.minecraft.ReportType.CRASH)) {
+            LOGGER.error("This crash report has been saved to: {}", path.toAbsolutePath());
+        } else {
+            LOGGER.error("We were unable to save this crash report to disk.");
+        }
+        // Canvas end - region threading
         // prevent further ticks from occurring
         // we CANNOT sync, because WE ARE ON A SCHEDULER THREAD
         this.scheduler.halt();
@@ -297,6 +348,13 @@ public final class TickRegionScheduler {
         protected final Schedule tickSchedule;
 
         private TickRegionScheduler scheduler;
+        // Canvas start - region threading
+        private final io.canvasmc.canvas.tick.ScheduledHandleTickState tickManager = new io.canvasmc.canvas.tick.ScheduledHandleTickState(this);
+
+        public io.canvasmc.canvas.tick.ScheduledHandleTickState getTickManager() {
+            return tickManager;
+        }
+        // Canvas end - region threading
 
         public RegionScheduleHandle(final TickRegions.TickRegionData region, final long firstStart) {
             this.currentTick = 0L;
@@ -309,7 +367,7 @@ public final class TickRegionScheduler {
             this.region = region;
 
             this.setScheduledStart(firstStart);
-            this.tickSchedule = new Schedule(firstStart == TimeUtil.DEADLINE_NOT_SET ? firstStart : firstStart - TIME_BETWEEN_TICKS);
+            this.tickSchedule = new Schedule(firstStart == TimeUtil.DEADLINE_NOT_SET ? firstStart : firstStart - getTimeBetweenTicks()); // Canvas - region threading
         }
 
         /**
@@ -318,7 +376,7 @@ public final class TickRegionScheduler {
          */
         protected final void updateScheduledStart(final long to) {
             this.setScheduledStart(to);
-            this.tickSchedule.setLastPeriod(to == TimeUtil.DEADLINE_NOT_SET ? to : to - TIME_BETWEEN_TICKS);
+            this.tickSchedule.setLastPeriod(to == TimeUtil.DEADLINE_NOT_SET ? to : to - getTimeBetweenTicks()); // Canvas - region threading
         }
 
         public final void markNonSchedulable() {
@@ -409,9 +467,13 @@ public final class TickRegionScheduler {
 
             final long cpuStart = MEASURE_CPU_TIME ? THREAD_MX_BEAN.getCurrentThreadCpuTime() : 0L;
             final long tickStart = System.nanoTime();
+            // Canvas start - region threading
+            io.canvasmc.canvas.tick.ScheduledHandleTickState tickState = getTickManager();
+            long deadlineFuture = tickState.tickStart(tickStart);
+            // Canvas end - region threading
 
             // use max(), don't assume that tickStart >= scheduledStart
-            final long tickCount = Math.max(1L, this.tickSchedule.getPeriodsAhead(TIME_BETWEEN_TICKS, tickStart));
+            final long tickCount = Math.max(1L, this.tickSchedule.getPeriodsAhead(deadlineFuture, tickStart)); // Canvas - region threading
 
             if (!this.tryMarkTicking()) {
                 if (!this.cancelled.get()) {
@@ -435,7 +497,7 @@ public final class TickRegionScheduler {
             this.lastTickStart = tickStart;
 
             final long scheduledStart = this.getScheduledStart();
-            final long scheduledEnd = scheduledStart + TIME_BETWEEN_TICKS;
+            final long scheduledEnd = scheduledStart + deadlineFuture; // Canvas - region threading
 
             final long intermediateTaskTime;
             final long intermediateTaskTimeCPU;
@@ -472,8 +534,10 @@ public final class TickRegionScheduler {
                 // in order to ensure all regions get their chance at scheduling, we have to ensure that regions
                 // that exceed the max tick time are not always prioritised over everything else. Thus, we use the greatest
                 // of the current time and "ideal" next tick start.
-                this.tickSchedule.advanceBy(tickCount, TIME_BETWEEN_TICKS);
-                this.setScheduledStart(TimeUtil.getGreatestTime(tickEnd, this.tickSchedule.getDeadline(TIME_BETWEEN_TICKS)));
+                // Canvas start - region threading
+                this.tickSchedule.advanceBy(tickCount, deadlineFuture);
+                this.setScheduledStart(TimeUtil.getGreatestTime(tickEnd, this.tickSchedule.getDeadline(deadlineFuture)));
+                // Canvas end - region threading
 
                 final TickTime time = new TickTime(
                     lastTickStart, scheduledStart, tickStart, cpuStart, tickEnd, cpuEnd,
@@ -482,6 +546,7 @@ public final class TickRegionScheduler {
                 );
 
                 this.addTickTime(time);
+                tickState.tickEnd(tickEnd); // Canvas - region threading
                 TickRegionScheduler.setTickTask(null);
                 if (this.region != null) {
                     TickRegionScheduler.setTickingRegion(null);
@@ -525,31 +590,31 @@ public final class TickRegionScheduler {
 
         public final TickData.TickReportData getTickReport5s(final long currTime) {
             synchronized (this) {
-                return this.tickTimes5s.generateTickReport(this.adjustCurrentTickData(currTime), currTime, TIME_BETWEEN_TICKS);
+                return this.tickTimes5s.generateTickReport(this.adjustCurrentTickData(currTime), currTime, getTimeBetweenTicks()); // Canvas - region threading
             }
         }
 
         public final TickData.TickReportData getTickReport15s(final long currTime) {
             synchronized (this) {
-                return this.tickTimes15s.generateTickReport(this.adjustCurrentTickData(currTime), currTime, TIME_BETWEEN_TICKS);
+                return this.tickTimes15s.generateTickReport(this.adjustCurrentTickData(currTime), currTime, getTimeBetweenTicks()); // Canvas - region threading
             }
         }
 
         public final TickData.TickReportData getTickReport1m(final long currTime) {
             synchronized (this) {
-                return this.tickTimes1m.generateTickReport(this.adjustCurrentTickData(currTime), currTime, TIME_BETWEEN_TICKS);
+                return this.tickTimes1m.generateTickReport(this.adjustCurrentTickData(currTime), currTime, getTimeBetweenTicks()); // Canvas - region threading
             }
         }
 
         public final TickData.TickReportData getTickReport5m(final long currTime) {
             synchronized (this) {
-                return this.tickTimes5m.generateTickReport(this.adjustCurrentTickData(currTime), currTime, TIME_BETWEEN_TICKS);
+                return this.tickTimes5m.generateTickReport(this.adjustCurrentTickData(currTime), currTime, getTimeBetweenTicks()); // Canvas - region threading
             }
         }
 
         public final TickData.TickReportData getTickReport15m(final long currTime) {
             synchronized (this) {
-                return this.tickTimes15m.generateTickReport(this.adjustCurrentTickData(currTime), currTime, TIME_BETWEEN_TICKS);
+                return this.tickTimes15m.generateTickReport(this.adjustCurrentTickData(currTime), currTime, getTimeBetweenTicks()); // Canvas - region threading
             }
         }
 
diff --git a/io/papermc/paper/threadedregions/TickRegions.java b/io/papermc/paper/threadedregions/TickRegions.java
index 50cf0f76be33a671c8143abd778a4a18dd2a8400..6958f760f9c0cb27f23b35c971400810c87de2a0 100644
--- a/io/papermc/paper/threadedregions/TickRegions.java
+++ b/io/papermc/paper/threadedregions/TickRegions.java
@@ -26,7 +26,7 @@ import java.util.function.BooleanSupplier;
 
 public final class TickRegions implements ThreadedRegionizer.RegionCallbacks<TickRegions.TickRegionData, TickRegions.TickRegionSectionData> {
 
-    private static final Logger LOGGER = LogUtils.getLogger();
+    private static final Logger LOGGER = LogUtils.getClassLogger(); // Canvas - region threading
     private static int regionShift = 31;
 
     public static int getRegionChunkShift() {
@@ -103,6 +103,7 @@ public final class TickRegions implements ThreadedRegionizer.RegionCallbacks<Tic
     @Override
     public void onRegionDestroy(final ThreadedRegionizer.ThreadedRegion<TickRegionData, TickRegionSectionData> region) {
         // nothing for now
+        io.canvasmc.canvas.tick.SchedulerUtil.getHandle().onRegionDestroy(region); // Canvas - region profiler
     }
 
     @Override
@@ -118,6 +119,8 @@ public final class TickRegions implements ThreadedRegionizer.RegionCallbacks<Tic
         final TickRegionData data = region.getData();
 
         scheduler.descheduleRegion(data.tickHandle);
+        // we call inactive here, since unpinning needs to occur before copying, or else this will cause issues
+        io.canvasmc.canvas.tick.SchedulerUtil.getHandle().onRegionInactive(region); // Canvas - region profiler
         // old handle cannot be scheduled anymore, copy to a new handle
         data.tickHandle = data.tickHandle.copy();
     }
@@ -301,6 +304,7 @@ public final class TickRegions implements ThreadedRegionizer.RegionCallbacks<Tic
                 }
             }
 
+            io.canvasmc.canvas.tick.SchedulerUtil.getHandle().onRegionSplit(this, into, world); // Canvas - region profiler
             // generic regionised data
             for (final Iterator<Reference2ReferenceMap.Entry<RegionizedData<?>, Object>> dataIterator = this.regionizedData.reference2ReferenceEntrySet().fastIterator();
                  dataIterator.hasNext();) {
@@ -388,6 +392,7 @@ public final class TickRegions implements ThreadedRegionizer.RegionCallbacks<Tic
 
             // task queue
             this.taskQueueData.mergeInto(data.taskQueueData);
+            io.canvasmc.canvas.tick.SchedulerUtil.getHandle().onRegionMerge(this, data, world); // Canvas - region profiler
         }
     }
 
@@ -406,6 +411,7 @@ public final class TickRegions implements ThreadedRegionizer.RegionCallbacks<Tic
             ret.currentTick = this.currentTick;
             ret.lastTickStart = this.lastTickStart;
             ret.tickSchedule.setLastPeriod(this.tickSchedule.getLastPeriod());
+            io.canvasmc.canvas.tick.SchedulerUtil.getHandle().tryTransferPinningState(this, ret); // Canvas - region profiler
 
             return ret;
         }
@@ -436,7 +442,7 @@ public final class TickRegions implements ThreadedRegionizer.RegionCallbacks<Tic
 
         private void checkInitialSchedule() {
             if (this.getScheduledStart() == TimeUtil.DEADLINE_NOT_SET) {
-                this.updateScheduledStart(System.nanoTime() + TickRegionScheduler.TIME_BETWEEN_TICKS);
+                this.updateScheduledStart(System.nanoTime() + TickRegionScheduler.getTimeBetweenTicks()); // Canvas - region threading
             }
         }
 
diff --git a/io/papermc/paper/threadedregions/commands/CommandServerHealth.java b/io/papermc/paper/threadedregions/commands/CommandServerHealth.java
index 7826c50b7c7c59e148905337ddc0dc8085f3f7c1..9100fa2575707d9fdacf7e731fd4d89f64bd12e3 100644
--- a/io/papermc/paper/threadedregions/commands/CommandServerHealth.java
+++ b/io/papermc/paper/threadedregions/commands/CommandServerHealth.java
@@ -55,14 +55,14 @@ public final class CommandServerHealth extends Command {
     }
 
     private static Component formatRegionInfo(final String prefix, final double util, final double mspt, final double tps,
-                                              final boolean newline) {
+                                              final boolean newline, final io.papermc.paper.threadedregions.TickRegionScheduler.RegionScheduleHandle regionHandle) { // Canvas - region threading
         return Component.text()
             .append(Component.text(prefix, PRIMARY, TextDecoration.BOLD))
-            .append(Component.text(ONE_DECIMAL_PLACES.get().format(util * 100.0), CommandUtil.getUtilisationColourRegion(util)))
+            .append(Component.text(ONE_DECIMAL_PLACES.get().format(util * 100.0), regionHandle.getTickManager().isSprinting() ? CommandUtil.SPRINTING_COLOR : CommandUtil.getUtilisationColourRegion(util))) // Canvas - region threading
             .append(Component.text("% util at ", PRIMARY))
-            .append(Component.text(TWO_DECIMAL_PLACES.get().format(mspt), CommandUtil.getColourForMSPT(mspt)))
+            .append(Component.text(TWO_DECIMAL_PLACES.get().format(mspt), regionHandle.getTickManager().isSprinting() ? CommandUtil.SPRINTING_COLOR : CommandUtil.getColourForMSPT(mspt))) // Canvas - region threading
             .append(Component.text(" MSPT at ", PRIMARY))
-            .append(Component.text(TWO_DECIMAL_PLACES.get().format(tps), CommandUtil.getColourForTPS(tps)))
+            .append(Component.text(TWO_DECIMAL_PLACES.get().format(tps), regionHandle.getTickManager().isSprinting() ? CommandUtil.SPRINTING_COLOR : CommandUtil.getColourForTPS(tps))) // Canvas - region threading
             .append(Component.text(" TPS" + (newline ? "\n" : ""), PRIMARY))
             .build();
     }
@@ -113,10 +113,10 @@ public final class CommandServerHealth extends Command {
             .append(Component.text(":\n", PRIMARY))
 
             .append(
-                formatRegionInfo("15s: ", util15s, mspt15s, tps15s, true)
+                formatRegionInfo("15s: ", util15s, mspt15s, tps15s, true, region.getData().getRegionSchedulingHandle()) // Canvas - region threading
             )
             .append(
-                formatRegionInfo("1m: ", util1m, mspt1m, tps1m, true)
+                formatRegionInfo("1m: ", util1m, mspt1m, tps1m, true, region.getData().getRegionSchedulingHandle()) // Canvas - region threading
             )
             .append(
                 formatRegionStats(region.getData().getRegionStats(), false)
@@ -250,11 +250,11 @@ public final class CommandServerHealth extends Command {
                 .append(Component.text(":\n", PRIMARY))
 
                 .append(Component.text("    ", PRIMARY))
-                .append(Component.text(ONE_DECIMAL_PLACES.get().format(util * 100.0), CommandUtil.getUtilisationColourRegion(util)))
+                .append(Component.text(ONE_DECIMAL_PLACES.get().format(util * 100.0), region.getData().getRegionSchedulingHandle().getTickManager().isSprinting() ? CommandUtil.SPRINTING_COLOR : CommandUtil.getUtilisationColourRegion(util))) // Canvas - region threading
                 .append(Component.text("% util at ", PRIMARY))
-                .append(Component.text(TWO_DECIMAL_PLACES.get().format(mspt), CommandUtil.getColourForMSPT(mspt)))
+                .append(Component.text(TWO_DECIMAL_PLACES.get().format(mspt), region.getData().getRegionSchedulingHandle().getTickManager().isSprinting() ? CommandUtil.SPRINTING_COLOR : CommandUtil.getColourForMSPT(mspt))) // Canvas - region threading
                 .append(Component.text(" MSPT at ", PRIMARY))
-                .append(Component.text(TWO_DECIMAL_PLACES.get().format(tps), CommandUtil.getColourForTPS(tps)))
+                .append(Component.text(TWO_DECIMAL_PLACES.get().format(tps), region.getData().getRegionSchedulingHandle().getTickManager().isSprinting() ? CommandUtil.SPRINTING_COLOR : CommandUtil.getColourForTPS(tps))) // Canvas - region threading
                 .append(Component.text(" TPS\n", PRIMARY))
 
                 .append(Component.text("    ", PRIMARY))
diff --git a/io/papermc/paper/threadedregions/commands/CommandUtil.java b/io/papermc/paper/threadedregions/commands/CommandUtil.java
index 8d4f092ec65d285508be81cbfadd6e747e1d067e..2db474ced22d5867b0f6315185af31d3b0879f92 100644
--- a/io/papermc/paper/threadedregions/commands/CommandUtil.java
+++ b/io/papermc/paper/threadedregions/commands/CommandUtil.java
@@ -9,6 +9,16 @@ import java.util.List;
 import java.util.function.Function;
 
 public final class CommandUtil {
+    // Canvas start - region threading
+    public static final TextColor SPRINTING_COLOR = net.kyori.adventure.text.format.NamedTextColor.LIGHT_PURPLE;
+    public static final double diff2MultTps = 0.1D;
+    public static final double diff5MultTps = 0.25D;
+    public static final double diff10MultTps = 0.5D;
+    public static final double diff15MultMspt = 0.3D;
+    public static final double diff25MultMspt = 0.5D;
+    public static final double diff35MultMspt = 0.7D;
+    public static final double diff40MultMspt = 0.8D;
+    // Canvas end - region threading
 
     public static List<String> getSortedList(final Iterable<String> iterable) {
         final List<String> ret = new ArrayList<>();
@@ -63,39 +73,45 @@ public final class CommandUtil {
     }
 
     public static TextColor getColourForTPS(final double tps) {
-        final double difference = Math.min(Math.abs(20.0 - tps), 20.0);
+        // Canvas start - region threading
+        final float rate = io.papermc.paper.threadedregions.TickRegionScheduler.getTickRate();
+        final double difference = Math.min(Math.abs(rate - tps), rate);
         final double coordinate;
-        if (difference <= 2.0) {
+        if (difference <= (rate * diff2MultTps)) {
             // >= 18 tps
-            coordinate = 70.0 + ((140.0 - 70.0)/(0.0 - 2.0)) * (difference - 2.0);
-        } else if (difference <= 5.0) {
+            coordinate = 70.0 + ((140.0 - 70.0)/(0.0 - (rate * diff2MultTps))) * (difference - (rate * diff2MultTps));
+        } else if (difference <= (rate * diff5MultTps)) {
             // >= 15 tps
-            coordinate = 30.0 + ((70.0 - 30.0)/(2.0 - 5.0)) * (difference - 5.0);
-        } else if (difference <= 10.0) {
+            coordinate = 30.0 + ((70.0 - 30.0)/((rate * diff2MultTps) - (rate * diff5MultTps))) * (difference - (rate * diff5MultTps));
+        } else if (difference <= (rate * diff10MultTps)) {
             // >= 10 tps
-            coordinate = 10.0 + ((30.0 - 10.0)/(5.0 - 10.0)) * (difference - 10.0);
+            coordinate = 10.0 + ((30.0 - 10.0)/((rate * diff5MultTps) - (rate * diff10MultTps))) * (difference - (rate * diff10MultTps));
         } else {
             // >= 0.0 tps
-            coordinate = 0.0 + ((10.0 - 0.0)/(10.0 - 20.0)) * (difference - 20.0);
+            coordinate = 0.0 + ((10.0 - 0.0)/((rate * diff10MultTps) - rate)) * (difference - rate);
         }
+        // Canvas end - region threading
 
         return TextColor.color(HSVLike.hsvLike((float)(coordinate / 360.0), 85.0f / 100.0f, 80.0f / 100.0f));
     }
 
     public static TextColor getColourForMSPT(final double mspt) {
-        final double clamped = Math.min(Math.abs(mspt), 50.0);
+        // Canvas start - region threading
+        final float time = 1000 / io.papermc.paper.threadedregions.TickRegionScheduler.getTickRate();
+        final double clamped = Math.min(Math.abs(mspt), time);
         final double coordinate;
-        if (clamped <= 15.0) {
-            coordinate = 130.0 + ((140.0 - 130.0)/(0.0 - 15.0)) * (clamped - 15.0);
-        } else if (clamped <= 25.0) {
-            coordinate = 90.0 + ((130.0 - 90.0)/(15.0 - 25.0)) * (clamped - 25.0);
-        } else if (clamped <= 35.0) {
-            coordinate = 30.0 + ((90.0 - 30.0)/(25.0 - 35.0)) * (clamped - 35.0);
-        } else if (clamped <= 40.0) {
-            coordinate = 15.0 + ((30.0 - 15.0)/(35.0 - 40.0)) * (clamped - 40.0);
+        if (clamped <= (time * diff15MultMspt)) {
+            coordinate = 130.0 + ((140.0 - 130.0)/(0.0 - (time * diff15MultMspt))) * (clamped - (time * diff15MultMspt));
+        } else if (clamped <= (time * diff25MultMspt)) {
+            coordinate = 90.0 + ((130.0 - 90.0)/((time * diff15MultMspt) - (time * diff25MultMspt))) * (clamped - (time * diff25MultMspt));
+        } else if (clamped <= (time * diff35MultMspt)) {
+            coordinate = 30.0 + ((90.0 - 30.0)/((time * diff25MultMspt) - (time * diff35MultMspt))) * (clamped - (time * diff35MultMspt));
+        } else if (clamped <= (time * diff40MultMspt)) {
+            coordinate = 15.0 + ((30.0 - 15.0)/((time * diff35MultMspt) - (time * diff40MultMspt))) * (clamped - (time * diff40MultMspt));
         } else {
-            coordinate = 0.0 + ((15.0 - 0.0)/(40.0 - 50.0)) * (clamped - 50.0);
+            coordinate = 0.0 + ((15.0 - 0.0)/((time * diff40MultMspt) - time)) * (clamped - time);
         }
+        // Canvas end - region threading
 
         return TextColor.color(HSVLike.hsvLike((float)(coordinate / 360.0), 85.0f / 100.0f, 80.0f / 100.0f));
     }
@@ -103,7 +119,7 @@ public final class CommandUtil {
     public static TextColor getUtilisationColourRegion(final double util) {
         // TODO anything better?
         // assume 20TPS
-        return getColourForMSPT(util * 50.0);
+        return getColourForMSPT(util * (io.papermc.paper.threadedregions.TickRegionScheduler.getTimeBetweenTicks() / 1_000_000F)); // Canvas - region threading
     }
 
     public static ServerPlayer getPlayer(final String name) {
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 11ac702624315605358b46c9d9ef3299a39dfe7b..f171c3b9cceffbfb382d214da4214443a708a0ed 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -287,7 +287,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         thread.setUncaughtExceptionHandler((t, e) -> LOGGER.error("Uncaught exception in server thread", e));
         thread.setPriority(Thread.NORM_PRIORITY + 2); // Paper - Perf: Boost priority
         if (Runtime.getRuntime().availableProcessors() > 4) {
-            thread.setPriority(8);
+            thread.setPriority(10); // Canvas - bump priority of boot thread
         }
 
         S server = (S)factory.apply(thread);
@@ -1097,15 +1097,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             this.server.waitForAsyncTasksShutdown(); // Paper - Wait for Async Tasks during shutdown
         }
         // CraftBukkit end
-        this.getConnection().stop();
-        this.isSaving = true;
-        if (this.playerList != null) {
-            //LOGGER.info("Saving players"); // Folia - move to shutdown thread logic
-            //this.playerList.saveAll(); // Folia - move to shutdown thread logic
-            this.playerList.removeAll(this.isRestarting); // Paper
-            try { Thread.sleep(100); } catch (InterruptedException ex) {} // CraftBukkit - SPIGOT-625 - give server at least a chance to send packets
-        }
-
+        this.isSaving = true; // Canvas - region threading
         // Folia start - region threading
         if (true) {
             // the rest till part 2 is handled by the region shutdown thread
@@ -1339,11 +1331,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
                 io.papermc.paper.threadedregions.RegionizedServer.getInstance().init(); // Folia - region threading - only after loading worlds
                 final long actualDoneTimeMs = System.currentTimeMillis() - org.bukkit.craftbukkit.Main.BOOT_TIME.toEpochMilli(); // Paper - Improve startup message
                 LOGGER.info("Done ({})! For help, type \"help\"", String.format(java.util.Locale.ROOT, "%.3fs", actualDoneTimeMs / 1000.00D)); // Paper - Improve startup message
-                for (;;) {
-                    try {
-                        Thread.sleep(Long.MAX_VALUE);
-                    } catch (final InterruptedException ex) {}
-                }
+                return; // Canvas - region threading
             }
             // Folia end - region threading
             // Spigot start
@@ -1442,7 +1430,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
             }
 
             this.onServerCrash(report);
-        } finally {
+        } if (false) { // Canvas - region threading
             try {
                 this.stopped = true;
                 this.stopServer();
@@ -1691,7 +1679,31 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         // Folia end - region threading
         //this.tickCount++; // Folia - region threading
         //this.tickRateManager.tick(); // Folia - region threading
-        this.tickChildren(haveTime, region); // Folia - region threading
+        // Canvas start - optimize flush calls with region threading
+        try {
+            List<ServerPlayer> flushing = new java.util.ArrayList<>(region.world.getCurrentWorldData().getLocalPlayers());
+            for (final net.minecraft.server.level.ServerPlayer localPlayer : flushing) {
+                // by suspending flushing, this will suspend flushing on the owning thread of the player
+                // however, if a diff thread sends a packet, it automatically flushes
+                localPlayer.connection.suspendFlushing();
+            }
+            this.tickChildren(haveTime, region);
+            for (final net.minecraft.server.level.ServerPlayer localPlayer : flushing) {
+                // if this thread doesn't own the player for some reason, or they were marked to
+                // continue flushing, we don't even try and touch it
+                if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(localPlayer)
+                    || !localPlayer.connection.suspendFlushingOnServerThread) {
+                    continue;
+                }
+                // note, this also flushes the channel
+                localPlayer.connection.resumeFlushing();
+            }
+        } catch (Throwable thrown) {
+            CrashReport crashReport = CrashReport.forThrowable(thrown, "Exception ticking region");
+            region.world.fillReportDetails(crashReport);
+            throw new ReportedException(crashReport);
+        }
+        // Canvas end - optimize flush calls with region threading
         if (false && nano - this.lastServerStatus >= STATUS_EXPIRE_TIME_NANOS) { // Folia - region threading
             this.lastServerStatus = nano;
             this.status = this.buildServerStatus();
@@ -1875,7 +1887,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         regionizedWorldData.tickConnections(); // Folia - region threading
         //this.playerList.tick(); // Folia - region threading
         this.debugSubscribers.tick();
-        if (this.tickRateManager.runsNormally()) {
+        if (false) { // Canvas - region threading
             GameTestTicker.SINGLETON.tick();
         }
 
diff --git a/net/minecraft/server/ServerTickRateManager.java b/net/minecraft/server/ServerTickRateManager.java
index 3687f52e5c2f765a1e4fd3e82a767903cce99d00..f0a7532bfba1a7710515c944d1636918e705af28 100644
--- a/net/minecraft/server/ServerTickRateManager.java
+++ b/net/minecraft/server/ServerTickRateManager.java
@@ -94,12 +94,7 @@ public class ServerTickRateManager extends TickRateManager {
         );
         this.scheduledCurrentSprintTicks = 0L;
         this.sprintTimeSpend = 0L;
-        // Paper start - silence feedback when API requests sprint
-        if (!this.silent) this.server
-            .createCommandSourceStack()
-            .sendSuccess(() -> Component.translatable("commands.tick.sprint.report", ticksPerSecond, millisecondsPerTick), true);
-        this.silent = false;
-        // Paper end - silence feedback when API requests sprint
+        // Canvas - region threading
         this.remainingSprintTicks = 0L;
         this.setFrozen(this.previousIsFrozen);
         this.server.onTickRateChanged();
diff --git a/net/minecraft/server/level/ServerChunkCache.java b/net/minecraft/server/level/ServerChunkCache.java
index 62b832cd135163a28b402386747ffd7df0f5fa55..ea909ea1143c5fb62afcd302d4ffbd156d457f76 100644
--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -478,7 +478,7 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
 
     @Override
     public void tick(final BooleanSupplier haveTime, final boolean tickChunks) {
-        if (this.level.tickRateManager().runsNormally() || !tickChunks || this.level.spigotConfig.unloadFrozenChunks) { // Spigot
+        if (this.level.getCurrentWorldData().regionData.tickHandle.getTickManager().doesRunGameElements() || !tickChunks || this.level.spigotConfig.unloadFrozenChunks) { // Spigot // Canvas - region threading
             this.ticketStorage.purgeStaleTickets(this.chunkMap);
         }
 
@@ -499,7 +499,7 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
         long timeDiff = 1L; // Folia - region threading
         //this.lastInhabitedUpdate = time; // Folia - region threading
         if (!this.level.isDebug()) {
-            if (this.level.tickRateManager().runsNormally()) {
+            if (regionizedWorldData.regionData.tickHandle.getTickManager().doesRunGameElements()) { // Canvas - region threading
                 this.tickChunks(timeDiff);
             }
 
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 400d0589e6d15ff8725065010e2cdfe081c216f3..1cc43387d4c61c368909e983d18157cbce48e1e1 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -279,6 +279,57 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         return true;
     }
 
+    // Canvas start - region threading
+    public final void canvas$loadOrRunAtChunksAsync(final BlockPos pos, final int radiusBlocks,
+                                                    final ca.spottedleaf.concurrentutil.util.Priority priority,
+                                                    final Runnable callback) {
+        if (ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(
+            this, pos, radiusBlocks
+        )) {
+            // already on the running region, run callback
+            callback.run();
+        } else if (moonrise$areChunksLoaded(
+            (pos.getX() - radiusBlocks) >> 4,
+            (pos.getX() + radiusBlocks) >> 4,
+            (pos.getZ() - radiusBlocks) >> 4,
+            (pos.getZ() + radiusBlocks) >> 4
+        )) {
+            // chunks are loaded, schedule callback
+            io.papermc.paper.threadedregions.RegionizedServer.getInstance().taskQueue.queueTickTaskQueue(
+                this, pos.getX() >> 4, pos.getZ() >> 4, callback, priority
+            );
+        } else {
+            // chunks are not loaded, schedule load and callback
+            moonrise$loadChunksAsync(
+                pos, radiusBlocks, priority, (unused) -> callback.run()
+            );
+        }
+    }
+
+    public final void canvas$loadOrRunAtChunksAsync(final int minChunkX, final int maxChunkX, final int minChunkZ, final int maxChunkZ,
+                                                    final ca.spottedleaf.concurrentutil.util.Priority priority,
+                                                    final Runnable callback) {
+        if (ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(
+            this, minChunkX, minChunkZ, maxChunkX, maxChunkZ
+        )) {
+            // already on the running region, run callback
+            callback.run();
+        } else if (moonrise$areChunksLoaded(
+            minChunkX, minChunkZ, maxChunkX, maxChunkZ
+        )) {
+            // chunks are loaded, schedule callback
+            io.papermc.paper.threadedregions.RegionizedServer.getInstance().taskQueue.queueTickTaskQueue(
+                this, minChunkX, minChunkZ, callback, priority
+            );
+        } else {
+            // chunks are not loaded, schedule load and callback
+            moonrise$loadChunksAsync(
+                minChunkX, minChunkZ, maxChunkX, maxChunkZ, priority, (unused) -> callback.run()
+            );
+        }
+    }
+
+    // Canvas end - region threading
     public final void loadChunksForMoveAsync(AABB box, ca.spottedleaf.concurrentutil.util.Priority priority,
                                              java.util.function.Consumer<List<net.minecraft.world.level.chunk.ChunkAccess>> onLoad) {
         // Paper - rewrite chunk system
@@ -791,8 +842,10 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     public void tick(BooleanSupplier haveTime, final io.papermc.paper.threadedregions.TickRegions.TickRegionData region) { // Folia - regionised ticking
         final io.papermc.paper.threadedregions.RegionizedWorldData regionizedWorldData = this.getCurrentWorldData(); // Folia - regionised ticking
         regionizedWorldData.setHandlingTick(true); // Folia - regionised ticking
-        TickRateManager tickRateManager = this.tickRateManager();
-        boolean runs = tickRateManager.runsNormally();
+        // Canvas start - region threading
+        io.canvasmc.canvas.tick.ScheduledHandleTickState handleTickState = regionizedWorldData.regionData.tickHandle.getTickManager();
+        boolean runs = handleTickState.doesRunGameElements();
+        // Canvas end - region threading
         if (runs) {
             //this.getWorldBorder().tick(); // Folia - regionised ticking
             // Folia start - regionised ticking
@@ -888,7 +941,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
                 .forEachTickingEntity( // Folia - regionised ticking
                     entity -> {
                         if (!entity.isRemoved()) {
-                            if (!tickRateManager.isEntityFrozen(entity)) {
+                            if (!handleTickState.isEntityFrozen(entity)) { // Canvas - region threading
                                 entity.checkDespawn();
                                 if (entity.isRemoved()) return; // Folia - region threading - if we despawned, DON'T TICK IT!
                                 if (true) { // Paper - rewrite chunk system
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index b26c816242da5a206cfb3ea996fab6dffd2d7cc2..f163a5f9f106a40d79ea32ed7477b63d0b6715a8 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1843,7 +1843,18 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     public void preChangeDimension() {
         super.preChangeDimension();
         this.stopUsingItem();
+        // if not suspending, we resume flushing. we check this on pre dimension change
+        // since this is called for portaling *and* cross-region teleports
+        if (!this.connection.suspendFlushingOnServerThread) this.connection.resumeFlushing(); // Canvas - optimize flush calls with region threading
     }
+    // Canvas start - region threading
+
+    @Override
+    public void postChangeDimension() {
+        this.level().getCurrentWorldData().regionData.getRegionSchedulingHandle().getTickManager().sendStateToPlayer(this); // Canvas - region threading
+        super.postChangeDimension();
+    }
+    // Canvas end - region threading
 
     @Override
     protected void placeSingleSync(ServerLevel originWorld, ServerLevel destination, EntityTreeNode treeNode, long teleportFlags) {
@@ -2290,7 +2301,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
 
     @Override
     protected void pushEntities() {
-        if (this.level().tickRateManager().runsNormally()) {
+        if (true) { // Canvas - region threading
             super.pushEntities();
         }
     }
diff --git a/net/minecraft/server/level/TicketType.java b/net/minecraft/server/level/TicketType.java
index af80a4c81211048d078388aa1063ddcb99686f24..37bae6fc56db1d511894652798878856a5c6e3be 100644
--- a/net/minecraft/server/level/TicketType.java
+++ b/net/minecraft/server/level/TicketType.java
@@ -75,6 +75,7 @@ public final class TicketType<T> implements ca.spottedleaf.moonrise.patches.chun
     public static final TicketType<Long> TELEPORT_HOLD_TICKET = ca.spottedleaf.moonrise.patches.chunk_system.ticket.ChunkSystemTicketType.create("folia:teleport_hold_ticket", Long::compareTo);
     public static final TicketType REGION_SCHEDULER_API_HOLD = register("folia:region_scheduler_api_hold", 0L, FLAG_LOADING | FLAG_SIMULATION);
     // Folia end - region threading
+    public static final TicketType REGION_PROFILING_HOLD = register("canvas:region_profiler_hold", NO_TIMEOUT, FLAG_LOADING | FLAG_SIMULATION | FLAG_KEEP_DIMENSION_ACTIVE); // Canvas - region threading
 
     private static TicketType register(final String name, final long timeout, @TicketType.Flags final int flags) {
         return Registry.register(BuiltInRegistries.TICKET_TYPE, name, new TicketType(timeout, flags));
diff --git a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
index f9ba96cb2039135e690f3044ff37deac4f9949e3..b83e3660ac6a91d1f8026bc8ac9b74ca40bea52a 100644
--- a/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerCommonPacketListenerImpl.java
@@ -309,10 +309,17 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
         this.suspendFlushingOnServerThread = true;
     }
 
+    // Canvas start - optimize flush calls with region threading
+    private int lastSentNonFlushPacketCount = 0;
+    private int sentNonFlushPacketCount = 0;
     public void resumeFlushing() {
         this.suspendFlushingOnServerThread = false;
-        this.connection.flushChannel();
+        if (lastSentNonFlushPacketCount != sentNonFlushPacketCount) {
+            connection.flushChannel();
+            lastSentNonFlushPacketCount = sentNonFlushPacketCount;
+        }
     }
+    // Canvas end - optimize flush calls with region threading
 
     public void send(final Packet<?> packet) {
         this.send(packet, null);
@@ -330,10 +337,13 @@ public abstract class ServerCommonPacketListenerImpl implements ServerCommonPack
             this.close();
         }
 
-        boolean flush = !this.suspendFlushingOnServerThread || !this.server.isSameThread();
-
+        // Canvas start - optimize flush calls with region threading
+        boolean flush = !this.suspendFlushingOnServerThread ||
+            !ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(connection.getPlayer());
+        if (!flush) sentNonFlushPacketCount++;
         try {
             this.connection.send(packet, listener, flush);
+        // Canvas end - optimize flush calls with region threading
         } catch (Throwable var7) {
             CrashReport report = CrashReport.forThrowable(var7, "Sending packet");
             CrashReportCategory category = report.addCategory("Packet being sent");
diff --git a/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java b/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
index 8f760aa6a38bdc7003b622a044cd42288383b8fb..b6a7038445b7d856b2718d124860b4d626ccace9 100644
--- a/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
@@ -288,13 +288,13 @@ public class ServerConfigurationPacketListenerImpl extends ServerCommonPacketLis
     // Paper start
     @Override
     public void disconnectAsync(final net.minecraft.network.DisconnectionDetails disconnectionInfo) {
-        if (this.cserver.isPrimaryThread()) {
+        if (io.papermc.paper.threadedregions.RegionizedServer.isGlobalTickThread()) { // Canvas - region threading
             this.disconnect(disconnectionInfo);
             return;
         }
 
         this.connection.setReadOnly();
-        this.server.scheduleOnMain(() -> {
+        io.papermc.paper.threadedregions.RegionizedServer.getInstance().addTask(() -> { // Canvas - region threading
             this.disconnect(disconnectionInfo); // Currently you cannot cancel disconnect during the config stage
         });
     }
diff --git a/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index a67a9ac716c1517640275dcb1ccd19084a7740f5..919af5b2971b5168d5c8288678c3048c6abd1de5 100644
--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -1552,7 +1552,7 @@ public class ServerGamePacketListenerImpl
                                 }
                             } else {
                                 boolean isFallFlying = this.player.isFallFlying();
-                                if (level.tickRateManager().runsNormally()) {
+                                if (true) { // Canvas - region threading
                                     this.receivedMovePacketCount++;
                                     int deltaPackets = this.receivedMovePacketCount - this.knownMovePacketCount;
                                     // CraftBukkit start - handle custom speeds and skipped ticks
@@ -2414,7 +2414,7 @@ public class ServerGamePacketListenerImpl
     public void handleChat(final ServerboundChatPacket packet) {
         // CraftBukkit start - async chat
         // SPIGOT-3638
-        if (this.server.isStopped()) {
+        if (io.papermc.paper.threadedregions.RegionShutdownThread.canvas$isShutdown()) { // Canvas - region threading
             return;
         }
         // CraftBukkit end
diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index e39d8689303eac729f76c671516384b6c97d56c3..53d79f58a1794e8670ae5c11f130ebb2749dd7ac 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -1058,7 +1058,7 @@ public abstract class PlayerList {
         // Paper start - Incremental chunk and player saving
             int numSaved = 0;
             final long now = System.nanoTime(); // Folia - region threading
-            final long timeInterval = (long)interval * io.papermc.paper.threadedregions.TickRegionScheduler.TIME_BETWEEN_TICKS; // Folia - region threading
+            final long timeInterval = (long)interval * io.papermc.paper.threadedregions.TickRegionScheduler.getTimeBetweenTicks(); // Folia - region threading // Canvas - region threading
             for (final ServerPlayer player : this.players) {
             // Folia start - region threading
             if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(player)) {
@@ -1115,7 +1115,7 @@ public abstract class PlayerList {
         }
 
         player.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.LEVEL_CHUNKS_LOAD_START, 0.0F));
-        this.server.tickRateManager().updateJoiningPlayer(player);
+        level.getCurrentWorldData().regionData.getRegionSchedulingHandle().getTickManager().sendStateToPlayer(player); // Canvas - region threading
     }
 
     public void sendAllPlayerInfo(final ServerPlayer player) {
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index e6f93db780975dbde98ff1ebcfd4e9030796764c..b89389857b62b1a3e590fc1e575a832b4fa446a6 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -331,7 +331,7 @@ public abstract class Entity
     protected UUID uuid = Mth.createInsecureUUID(this.random);
     protected String stringUUID = this.uuid.toString();
     private boolean hasGlowingTag;
-    private final Set<String> tags = new io.papermc.paper.util.SizeLimitedSet<>(new it.unimi.dsi.fastutil.objects.ObjectOpenHashSet<>(), MAX_ENTITY_TAG_COUNT); // Paper - fully limit tag size - replace set impl
+    private final Set<String> tags = new io.papermc.paper.util.SizeLimitedSet<>(new java.util.concurrent.CopyOnWriteArraySet<>(), MAX_ENTITY_TAG_COUNT); // Paper - fully limit tag size - replace set impl // Canvas - region threading
     private final double[] pistonDeltas = new double[]{0.0, 0.0, 0.0};
     private long pistonDeltasGameTime = Long.MIN_VALUE; // Folia - region threading
     protected EntityDimensions dimensions;
@@ -396,6 +396,18 @@ public abstract class Entity
         return this.dimensions.makeBoundingBox(x, y, z);
     }
     // Paper end
+    // Canvas start - region threading
+
+    public void canvas$scheduleToOrRun(Runnable task) {
+        if (ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this)) {
+            task.run();
+            return;
+        }
+        this.getBukkitEntity().taskScheduler.schedule(
+            (entity) -> task.run(), null, 1L
+        );
+    }
+    // Canvas end - region threading
     // Paper start - rewrite chunk system
     private final boolean isHardColliding = this.moonrise$isHardCollidingUncached();
     private net.minecraft.server.level.FullChunkStatus chunkStatus;
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index d38969ed70445275eccc2fe2e554b0eef1d8b391..20b043cea20570fd32bc9bcfa09c09275a3df068 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -4530,21 +4530,12 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
                 this.setPos(xo, yo, zo);
 
                 if (ok) {
-                    if (!(this instanceof ServerPlayer)) {
-                        org.bukkit.event.entity.EntityTeleportEvent teleport = new org.bukkit.event.entity.EntityTeleportEvent(this.getBukkitEntity(), new Location(this.level().getWorld(), xo, yo, zo), new Location(this.level().getWorld(), xx, y, zz));
-                        this.level().getCraftServer().getPluginManager().callEvent(teleport);
-                        if (!teleport.isCancelled() && teleport.getTo() != null) { // Paper
-                            Location to = teleport.getTo();
-                            this.teleportTo(to.getX(), to.getY(), to.getZ());
-                        } else {
-                            return Optional.empty();
-                        }
-                    } else {
-                        // player teleport event is called in the underlining code
-                        if (!((ServerPlayer) this).connection.teleport(xx, y, zz, this.getYRot(), this.getXRot(), cause)) {
-                            return Optional.empty();
-                        }
-                    }
+                    // Canvas start - region threading
+                    boolean teleportResult = this.teleportAsync(
+                        (ServerLevel) this.level(), new Vec3(xx, y, zz), this.getYRot(), this.getXRot(), this.getDeltaMovement(), cause, Entity.TELEPORT_FLAG_LOAD_CHUNK | Entity.TELEPORT_FLAG_UNMOUNT, null
+                    );
+                    if (!teleportResult) return Optional.empty();
+                    // Canvas end - region threading
                 }
                 // CraftBukkit end
             }
diff --git a/net/minecraft/world/entity/Mob.java b/net/minecraft/world/entity/Mob.java
index 244765a11e2cea836295f591573f170fb3e763b9..edccf6b99f2565b3fef6cea80af24a6f6b15b8a2 100644
--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -290,7 +290,7 @@ public abstract class Mob extends LivingEntity implements Targeting, EquipmentUs
 
     public boolean setTarget(@Nullable LivingEntity target, EntityTargetEvent.@Nullable TargetReason reason) {
         LivingEntity currentTarget = this.getTargetUnchecked();
-        if (Objects.equals(currentTarget, target)) {
+        if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(target) || Objects.equals(currentTarget, target)) { // Canvas - fix Folia#386
             return false;
         }
         LivingEntity originalTarget = target;
diff --git a/net/minecraft/world/entity/ai/behavior/BabyFollowAdult.java b/net/minecraft/world/entity/ai/behavior/BabyFollowAdult.java
index 801ba5da1fa5fb8b2fc6e76053cf150c040f455a..f32971c003aee48dca2aea6d212bef6081a70d63 100644
--- a/net/minecraft/world/entity/ai/behavior/BabyFollowAdult.java
+++ b/net/minecraft/world/entity/ai/behavior/BabyFollowAdult.java
@@ -27,6 +27,12 @@ public class BabyFollowAdult {
                             return false;
                         } else {
                             LivingEntity adult = i.get(nearestAdult);
+                            // Canvas start - region threading
+                            if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(adult)) {
+                                nearestAdult.erase();
+                                return true;
+                            }
+                            // Canvas end - region threading
                             if (body.closerThan(adult, followRange.maxInclusive() + 1) && !body.closerThan(adult, followRange.minInclusive())) {
                                 // CraftBukkit start
                                 org.bukkit.event.entity.EntityTargetLivingEntityEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callEntityTargetLivingEvent(body, adult, org.bukkit.event.entity.EntityTargetEvent.TargetReason.FOLLOW_LEADER);
@@ -37,6 +43,12 @@ public class BabyFollowAdult {
                                     walkTarget.erase();
                                     return true;
                                 }
+                                // Canvas start - region threading
+                                if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(((org.bukkit.craftbukkit.entity.CraftEntity)event.getTarget()).getHandleRaw())) {
+                                    nearestAdult.erase();
+                                    return true;
+                                }
+                                // Canvas end - region threading
                                 adult = ((org.bukkit.craftbukkit.entity.CraftLivingEntity) event.getTarget()).getHandle();
                                 // CraftBukkit end
                                 WalkTarget target = new WalkTarget(
diff --git a/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java b/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
index 7172721045c7b40f49c714260c31c72073ea9cc4..e369afaeb17d5fe142c0d211ab257170619ca85c 100644
--- a/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
+++ b/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
@@ -81,6 +81,7 @@ public class BehaviorUtils {
     public static void setWalkAndLookTargetMemories(
         final LivingEntity walker, final PositionTracker target, final float speedModifier, final int closeEnoughDistance
     ) {
+        if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(entity.level(), positionTracker.currentBlockPosition())) return; // Canvas - region threading
         WalkTarget walkTarget = new WalkTarget(target, speedModifier, closeEnoughDistance);
         walker.getBrain().setMemory(MemoryModuleType.LOOK_TARGET, target);
         walker.getBrain().setMemory(MemoryModuleType.WALK_TARGET, walkTarget);
diff --git a/net/minecraft/world/entity/ai/behavior/JumpOnBed.java b/net/minecraft/world/entity/ai/behavior/JumpOnBed.java
index 9cf3fe563508c12679dc7ed1ec9f6864128357e1..9119bdf774d628874a39d3648c82fff8dc1143a3 100644
--- a/net/minecraft/world/entity/ai/behavior/JumpOnBed.java
+++ b/net/minecraft/world/entity/ai/behavior/JumpOnBed.java
@@ -101,6 +101,11 @@ public class JumpOnBed extends Behavior<Mob> {
     }
 
     private boolean isBed(final ServerLevel level, final BlockPos bodyPos) {
+        // Canvas start - region threading
+        if (!level.hasChunkAt(bodyPos)) {
+            return false;
+        }
+        // Canvas end - region threading
         return level.getBlockState(bodyPos).is(BlockTags.BEDS);
     }
 
diff --git a/net/minecraft/world/entity/ai/behavior/MoveToTargetSink.java b/net/minecraft/world/entity/ai/behavior/MoveToTargetSink.java
index 194813475271364831cdd1bb5ae244967341f9da..f2456235a329307a17df99d6f7eba3b061071d1f 100644
--- a/net/minecraft/world/entity/ai/behavior/MoveToTargetSink.java
+++ b/net/minecraft/world/entity/ai/behavior/MoveToTargetSink.java
@@ -117,7 +117,7 @@ public class MoveToTargetSink extends Behavior<Mob> {
 
     private boolean tryComputePath(final Mob body, final WalkTarget walkTarget, final long timestamp) {
         BlockPos targetPos = walkTarget.getTarget().currentBlockPosition();
-        this.path = body.getNavigation().createPath(targetPos, 0);
+        this.path = ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(body.level(), targetPos) ? body.getNavigation().createPath(targetPos, 0) : null; // Canvas - region threading
         this.speedModifier = walkTarget.getSpeedModifier();
         Brain<?> brain = body.getBrain();
         if (this.reachedTarget(body, walkTarget)) {
diff --git a/net/minecraft/world/entity/ai/behavior/SleepInBed.java b/net/minecraft/world/entity/ai/behavior/SleepInBed.java
index 112572f5a6813abd3cf3bd26e2b087b7edb48f41..4df6c95aa3cf521e87626975e1088dec7ed93098 100644
--- a/net/minecraft/world/entity/ai/behavior/SleepInBed.java
+++ b/net/minecraft/world/entity/ai/behavior/SleepInBed.java
@@ -55,6 +55,7 @@ public class SleepInBed extends Behavior<LivingEntity> {
                     }
                 }
 
+                if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(owner.level(), globalPos.pos())) return false; // Canvas - region threading
                 BlockState blockState = level.getBlockStateIfLoaded(target.pos()); // Paper - Prevent sync chunk loads when villagers try to find beds
                 if (blockState == null) { return false; } // Paper - Prevent sync chunk loads when villagers try to find beds
                 return target.pos().closerToCenterThan(body.position(), 2.0) && blockState.is(BlockTags.BEDS) && !blockState.getValue(BedBlock.OCCUPIED);
diff --git a/net/minecraft/world/entity/animal/allay/AllayAi.java b/net/minecraft/world/entity/animal/allay/AllayAi.java
index 05436572581d369fe392566ec9c0afa8d8f03c6e..1e18b5ed68ea2ef180d232b272b82fac2cb84e5e 100644
--- a/net/minecraft/world/entity/animal/allay/AllayAi.java
+++ b/net/minecraft/world/entity/animal/allay/AllayAi.java
@@ -112,6 +112,14 @@ public class AllayAi {
         Optional<GlobalPos> likedNoteblockPos = brain.getMemory(MemoryModuleType.LIKED_NOTEBLOCK_POSITION);
         if (likedNoteblockPos.isPresent()) {
             GlobalPos position = likedNoteblockPos.get();
+            // Canvas start - region threading
+            if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(
+                net.minecraft.server.MinecraftServer.getServer().getLevel(globalPos.dimension()), globalPos.pos()
+            )) {
+                brain.eraseMemory(MemoryModuleType.LIKED_NOTEBLOCK_POSITION);
+                return Optional.empty();
+            }
+            // Canvas end - region threading
             if (shouldDepositItemsAtLikedNoteblock(allay, brain, position)) {
                 return Optional.of(new BlockPosTracker(position.pos().above()));
             }
diff --git a/net/minecraft/world/entity/animal/axolotl/AxolotlAi.java b/net/minecraft/world/entity/animal/axolotl/AxolotlAi.java
index 05de2a22fea3d8573cdd00e8b102e9732dd4c5dc..5766a3db624a019579345140b72f0d9a08e91524 100644
--- a/net/minecraft/world/entity/animal/axolotl/AxolotlAi.java
+++ b/net/minecraft/world/entity/animal/axolotl/AxolotlAi.java
@@ -134,7 +134,13 @@ public class AxolotlAi {
         Optional<PositionTracker> lookTarget = body.getBrain().getMemory(MemoryModuleType.LOOK_TARGET);
         if (lookTarget.isPresent()) {
             BlockPos pos = lookTarget.get().currentBlockPosition();
-            return level.isWaterAt(pos) == body.isInWater();
+            // Canvas start - region threading
+            net.minecraft.world.level.material.FluidState fluidIfLoaded = level.getFluidIfLoaded(pos);
+            if (fluidIfLoaded == null) {
+                return false;
+            }
+            return fluidIfLoaded.is(net.minecraft.tags.FluidTags.WATER) == body.isInWater();
+            // Canvas end - region threading
         } else {
             return false;
         }
diff --git a/net/minecraft/world/entity/animal/sniffer/Sniffer.java b/net/minecraft/world/entity/animal/sniffer/Sniffer.java
index b84a7c923d52ee1265ce79ae994697d3c348734a..3d981e1065f4d2f5197de23f57773ab190d67640 100644
--- a/net/minecraft/world/entity/animal/sniffer/Sniffer.java
+++ b/net/minecraft/world/entity/animal/sniffer/Sniffer.java
@@ -276,7 +276,16 @@ public class Sniffer extends Animal {
 
     private boolean canDig(final BlockPos position) {
         return this.level().getBlockState(position).is(BlockTags.SNIFFER_DIGGABLE_BLOCK)
-            && this.getExploredPositions().noneMatch(explored -> GlobalPos.of(this.level().dimension(), position).equals(explored))
+            // Canvas start - region threading
+            && this.getExploredPositions().noneMatch(explored -> {
+                if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(
+                    net.minecraft.server.MinecraftServer.getServer().getLevel(explored.dimension()), explored.pos()
+                )) {
+                    return false;
+                }
+                return GlobalPos.of(this.level().dimension(), position).equals(explored);
+            })
+            // Canvas end - region threading
             && Optional.ofNullable(this.getNavigation().createPath(position, 1)).map(Path::canReach).orElse(false);
     }
 
diff --git a/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java b/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
index 3ffb48f19c8b58a31d8641f398a705b578fada0d..4c069d692254c7ed4b1b5d434c161f7fe999d7ae 100644
--- a/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
+++ b/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
@@ -291,20 +291,36 @@ public class EnderDragon extends Mob implements Enemy {
                 this.tickPart(this.wing1, cc1 * 4.5F, 2.0, ss1 * 4.5F);
                 this.tickPart(this.wing2, cc1 * -4.5F, 2.0, ss1 * -4.5F);
                 if (this.level() instanceof ServerLevel serverLevelx && this.hurtTime == 0) {
-                    this.knockBack(
+                    // Canvas start - region threading
+                    final AABB wing1AABB = this.wing1.getBoundingBox()
+                        .inflate(4.0, 2.0, 4.0)
+                        .move(0.0, -2.0, 0.0);
+                    final net.minecraft.world.phys.AABB wing2AABB = this.wing2.getBoundingBox()
+                        .inflate(4.0, 2.0, 4.0)
+                        .move(0.0, -2.0, 0.0);
+                    final net.minecraft.world.phys.AABB neckAABB = this.neck.getBoundingBox().inflate(1.0);
+                    final net.minecraft.world.phys.AABB headAABB = this.head.getBoundingBox().inflate(1.0);
+                    if (ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(serverLevelx, wing1AABB)) this.knockBack(
+                    // Canvas end - region threading
                         serverLevelx,
                         serverLevelx.getEntities(
-                            this, this.wing1.getBoundingBox().inflate(4.0, 2.0, 4.0).move(0.0, -2.0, 0.0), EntitySelector.NO_CREATIVE_OR_SPECTATOR
+                            this, wing1AABB, EntitySelector.NO_CREATIVE_OR_SPECTATOR // Canvas - region threading
                         )
                     );
-                    this.knockBack(
+                    if (ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(serverLevelx, wing2AABB)) this.knockBack( // Canvas - region threading
                         serverLevelx,
                         serverLevelx.getEntities(
-                            this, this.wing2.getBoundingBox().inflate(4.0, 2.0, 4.0).move(0.0, -2.0, 0.0), EntitySelector.NO_CREATIVE_OR_SPECTATOR
+                            this, wing2AABB, EntitySelector.NO_CREATIVE_OR_SPECTATOR // Canvas - region threading
                         )
                     );
-                    this.hurt(serverLevelx, serverLevelx.getEntities(this, this.head.getBoundingBox().inflate(1.0), EntitySelector.NO_CREATIVE_OR_SPECTATOR));
-                    this.hurt(serverLevelx, serverLevelx.getEntities(this, this.neck.getBoundingBox().inflate(1.0), EntitySelector.NO_CREATIVE_OR_SPECTATOR));
+                    // Canvas start - region threading
+                    if (ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(serverLevelx, headAABB)) {
+                        this.hurt(serverLevelx, serverLevelx.getEntities(this, headAABB, EntitySelector.NO_CREATIVE_OR_SPECTATOR));
+                    }
+                    if (ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(serverLevelx, neckAABB)) {
+                        this.hurt(serverLevelx, serverLevelx.getEntities(this, neckAABB, EntitySelector.NO_CREATIVE_OR_SPECTATOR));
+                    }
+                    // Canvas end - region threading
                 }
 
                 float ss2 = Mth.sin(this.getYRot() * Mth.DEG_TO_RAD - this.yRotA * 0.01F);
diff --git a/net/minecraft/world/entity/projectile/FishingHook.java b/net/minecraft/world/entity/projectile/FishingHook.java
index 0c8a48988c859669cf7e38816aaf32782c245244..9116d87463f4068c03e733de1f73fb464219ebed 100644
--- a/net/minecraft/world/entity/projectile/FishingHook.java
+++ b/net/minecraft/world/entity/projectile/FishingHook.java
@@ -207,7 +207,7 @@ public class FishingHook extends Projectile {
             } else {
                 if (this.currentState == FishingHook.FishHookState.HOOKED_IN_ENTITY) {
                     if (this.hookedIn != null) {
-                        if (!this.hookedIn.isRemoved() && this.hookedIn.canInteractWithLevel() && this.hookedIn.level().dimension() == this.level().dimension()
+                        if (!this.hookedIn.isRemoved() && this.hookedIn.canInteractWithLevel() && this.hookedIn.level().dimension() == this.level().dimension() && ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.hookedIn) // Canvas - region threading
                             )
                          {
                             this.setPos(this.hookedIn.getX(), this.hookedIn.getY(0.8), this.hookedIn.getZ());
diff --git a/net/minecraft/world/item/Item.java b/net/minecraft/world/item/Item.java
index a592cdacb46f95ced9f995beb6b5e1e9c5151778..bfec5537e71bd245ff79c7a7ff49030583bc8e71 100644
--- a/net/minecraft/world/item/Item.java
+++ b/net/minecraft/world/item/Item.java
@@ -730,7 +730,7 @@ public class Item implements ItemLike, FeatureElement {
 
                 @Override
                 public float tickRate() {
-                    return level.tickRateManager().tickrate();
+                    return io.papermc.paper.threadedregions.TickRegionScheduler.getTickRate();
                 }
 
                 @Override
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 91fb2c1594de9625bc8ea7e34137c2ebeb79c23c..e8931817133aa2e88c9f12d84ca8a686bb4d5264 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1523,7 +1523,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
         List<TickingBlockEntity> blockEntityTickers = regionizedWorldData.getBlockEntityTickers(); // Folia - regionised ticking
 
         // Paper - Fix MC-117075 use removeAll - remove iterator in favour of indexed for loop, ensuring compile error if something uses iter incorrectly
-        boolean tickBlockEntities = this.tickRateManager().runsNormally();
+        boolean tickBlockEntities = regionizedWorldData.regionData.tickHandle.getTickManager().doesRunGameElements(); // Canvas - region threading
 
         int tickedEntities = 0; // Paper - rewrite chunk system
         // Paper start - Fix MC-117075 use removeAll
diff --git a/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java b/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
index 832e04b97ffae60127537c0a8eb448af6602da1e..83b40c182cce1c7a77e8ba3c66c2273a49c10881 100644
--- a/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
@@ -216,7 +216,7 @@ public class TheEndGatewayBlockEntity extends TheEndPortalBlockEntity {
                 teleport, net.minecraft.world.entity.Entity.TELEPORT_FLAG_LOAD_CHUNK | net.minecraft.world.entity.Entity.TELEPORT_FLAG_TELEPORT_PASSENGERS,
                 (net.minecraft.world.entity.Entity teleportedEntity) -> {
                     // adjust to the final exit position
-                    Vec3 adjusted = Vec3.atCenterOf(TheEndGatewayBlockEntity.findExitPosition(portalWorld, teleportPos));
+                    Vec3 adjusted = Vec3.atCenterOf(TheEndGatewayBlockEntity.findExitPosition(portalWorld, teleportedEntity.blockPosition())); // Canvas - region threading
                     // teleportTo will adjust rider positions
                     teleportedEntity.teleportTo(adjusted.x, adjusted.y, adjusted.z);
 
