From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <jedimastertoothless@hotmail.com>
Date: Sat, 17 Jan 2026 18:41:55 -0800
Subject: [PATCH] PR Party 5


diff --git a/net/minecraft/world/entity/TamableAnimal.java b/net/minecraft/world/entity/TamableAnimal.java
index fbb1e3912fbf7777997aecae60861abe16ddb420..5baff40285ee8f99729fbae09e0ea46a879e2704 100644
--- a/net/minecraft/world/entity/TamableAnimal.java
+++ b/net/minecraft/world/entity/TamableAnimal.java
@@ -279,6 +279,16 @@ public abstract class TamableAnimal extends Animal implements OwnableEntity {
     }
 
     private boolean maybeTeleportTo(int x, int y, int z) {
+        // Folia start - region threading
+        // the tamable entity can, when trying to see if it can teleport to the
+        // block position, try and load chunks that may or may not be outside the
+        // region boundaries. if it tries to load chunks outside the region,
+        // the server crashes. so we do a quick check here first to ensure
+        // that the chunk *is* loaded first, and if it is, then we attempt teleport
+        if (this.level().getChunkIfLoaded(x >> 4, z >> 4) == null) {
+            return false; // chunk not loaded, don't teleport
+        }
+        // Folia end - region threading
         if (!this.canTeleportTo(new BlockPos(x, y, z))) {
             return false;
         } else {
