From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Mon, 20 Apr 2026 13:51:43 -0700
Subject: [PATCH] Fix world load and unload API


diff --git a/io/papermc/paper/threadedregions/RegionizedServer.java b/io/papermc/paper/threadedregions/RegionizedServer.java
index 3a2a9cc67a32983bbabdee62664c9c29dee1fce3..3eed775c55d9e72f7725b4b0ece5653f03f6bab8 100644
--- a/io/papermc/paper/threadedregions/RegionizedServer.java
+++ b/io/papermc/paper/threadedregions/RegionizedServer.java
@@ -436,6 +436,37 @@ public final class RegionizedServer {
     // A global tick only updates things like weather / worldborder, basically anything in the world that is
     // NOT tied to a specific region, but rather shared amongst all of them.
     private void globalTick(final ServerLevel world, final long tickCount) {
+        // Canvas start - rewrite world load and unload
+        if (world.canvas$unloadTicket.isPresent()) {
+            // we are unloading, return.
+            AtomicBoolean isSingularRegionAlive = new AtomicBoolean(false);
+            world.regioniser.computeForAllRegions((region) -> {
+                if (!region.getData().getRegionSchedulingHandle().isMarkedAsNonSchedulable()) {
+                    // region still active, possibly still saving or hasn't gotten to that yet
+                    isSingularRegionAlive.set(true);
+                }
+            });
+            if (!isSingularRegionAlive.get()) {
+                io.canvasmc.canvas.util.ticket.UnloadTicket ticket = world.canvas$unloadTicket.getOrThrow();
+                LOGGER.info("Finalizing unload of world '{}'", world.getWorld().getName());
+                final boolean save = ticket.save();
+                if (save) {
+                    org.bukkit.Bukkit.getPluginManager().callEvent(new org.bukkit.event.world.WorldSaveEvent(world.getWorld()));
+                    world.saveLevelData(true);
+                }
+                world.chunkSource.getDataStorage().close();
+                // 'save' SHOULD be false, we run this on regions, and that MUST be regionized, or we crash :)
+                world.moonrise$getChunkTaskScheduler().chunkHolderManager.close(false, true, true, true, false);
+                io.papermc.paper.FeatureHooks.closeEntityManager(world, save);
+                this.worlds.remove(world);
+                MinecraftServer.getServer().removeLevel(world);
+                MinecraftServer.getServer().server.worlds.remove(world.getWorld().getName().toLowerCase(java.util.Locale.ROOT));
+                LOGGER.info("Completed unload of world '{}'", world.getWorld().getName());
+                ticket.callback().accept(io.canvasmc.canvas.WorldUnloadResult.SUCCESS);
+            }
+            return;
+        }
+        // Canvas end - rewrite world load and unload
         // needs
         // worldborder tick
         // advancing the weather cycle
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index 60ea8b8e5066e3e9c037498d2588973a653ef607..64231ca4becdd778eee8ed7c00c811feeb2b568c 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1686,6 +1686,69 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
                 scheduler.executeTick();
             }
         }
+        // Canvas start - rewrite world load and unload
+        ServerLevel world = region.world;
+        if (world.canvas$unloadTicket.isPresent()) {
+            // finish teleports
+            final ChunkPos center = region.region.getCenterChunk();
+            LOGGER.info("Beginning unload of region around chunk {} in unloading world '{}'", center, world.getWorld().getName());
+            final List<ServerLevel.PendingTeleport> pendingTeleports = world.removeAllRegionTeleports();
+            if (!pendingTeleports.isEmpty()) {
+                LOGGER.info("Completing {} pending teleports in region around chunk {} in world '{}'", pendingTeleports.size(), center, world.getWorld().getName());
+                for (final ServerLevel.PendingTeleport pendingTeleport : pendingTeleports) {
+                    LOGGER.info("Completing teleportation to target position {}", pendingTeleport.to());
+
+                    // first, add entities to entity chunk so that they will be saved
+                    for (final net.minecraft.world.entity.Entity.EntityTreeNode node : pendingTeleport.rootVehicle().getFullTree()) {
+                        // assume that world and position are set to destination here
+                        node.root.setLevel(world); // in case the pending teleport is from a portal before it finds the exact destination
+                        world.moonrise$getEntityLookup().addEntityForShutdownTeleportComplete(node.root);
+                    }
+
+                    // then, rebuild the passenger tree so that when saving only the root vehicle will be written - and if
+                    // there are any player passengers, that the later player saving will save the tree
+                    pendingTeleport.rootVehicle().restore();
+
+                    // now we are finished
+                    LOGGER.info("Completed teleportation to target position {}", pendingTeleport.to());
+                }
+            }
+
+            // essentially we need to cover the ass of dumb plugins or race conditions
+            // where a player is teleported and then the world is unloaded, allowing a small
+            // window for a race condition where the player can be thrown into an unloading
+            // world. to resolve this issue we teleport the player back to where they came from
+            // Note: this covers the following:
+            //  - Portaling
+            //  - Teleporting (via command, plugin, etc.)
+            //  - Plugin
+            for (ServerPlayer localPlayer : world.getCurrentWorldData().getLocalPlayers()) {
+                io.canvasmc.canvas.util.ServerLocation lastTeleportPos = localPlayer.canvas$lastTeleportOrigin;
+                // this *really* shouldn't ever happen... ever
+                if (lastTeleportPos == null) {
+                    LOGGER.warn("Player does not contain a last origin teleport position, skipping unload return");
+                    continue;
+                }
+                LOGGER.info("Returning player to position {}/{} due to world unload", lastTeleportPos.world(), lastTeleportPos.pos());
+                localPlayer.teleportAsync(
+                    lastTeleportPos.world(),
+                    lastTeleportPos.pos(),
+                    lastTeleportPos.yaw(), lastTeleportPos.pitch(),
+                    null,
+                    org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.UNKNOWN,
+                    net.minecraft.world.entity.Entity.TELEPORT_FLAG_LOAD_CHUNK,
+                    null
+                );
+            }
+
+            LOGGER.info("Saving chunks in region around chunk {} in world '{}'", center, world.getWorld().getName());
+            world.moonrise$getChunkTaskScheduler().chunkHolderManager.close(world.canvas$unloadTicket.getOrThrow().save(), true, false, false, false);
+
+            LOGGER.info("Descheduling region around chunk {} in world '{}'", center, world.getWorld().getName());
+            region.getRegionSchedulingHandle().markNonSchedulable();
+            return;
+        }
+        // Canvas end - rewrite world load and unload
         // Folia end - region threading
         //this.tickCount++; // Folia - region threading
         //this.tickRateManager.tick(); // Folia - region threading
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 68c1ff11c9458b586888253d287d016c45a3e1ef..ca96610ba3da579e90b783a33fd03f3af475ee18 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -765,6 +765,38 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         );
     }
     public final io.papermc.paper.threadedregions.RegionizedTaskQueue.WorldRegionTaskData taskQueueRegionData = new io.papermc.paper.threadedregions.RegionizedTaskQueue.WorldRegionTaskData(this);
+    // Canvas start - region threading
+    public final io.canvasmc.canvas.util.ticket.TicketHolder<io.canvasmc.canvas.util.ticket.UnloadTicket>
+        canvas$unloadTicket = new io.canvasmc.canvas.util.ticket.TicketHolder<>();
+    // this lock should be used whenever we need to check or modify the unload state of the world
+    private final java.util.concurrent.locks.ReentrantLock canvas$worldStageLock = new java.util.concurrent.locks.ReentrantLock();
+    public final java.util.concurrent.CopyOnWriteArrayList<String> canvas$joiningPlayers = new java.util.concurrent.CopyOnWriteArrayList<>();
+
+    public void canvas$executeUnderWorldStageLock(Runnable runnable) {
+        canvas$obtainWorldStageLock();
+        try {
+            runnable.run();
+        } finally {
+            canvas$releaseWorldStageLock();
+        }
+    }
+
+    public void canvas$obtainWorldStageLock() {
+        canvas$worldStageLock.lock();
+    }
+
+    public void canvas$releaseWorldStageLock() {
+        canvas$worldStageLock.unlock();
+    }
+
+    public void canvas$incrementJoiningPlayers(String key) {
+        canvas$executeUnderWorldStageLock(() -> canvas$joiningPlayers.add(key));
+    }
+
+    public void canvas$decrementJoiningPlayers(String key) {
+        canvas$executeUnderWorldStageLock(() -> canvas$joiningPlayers.remove(key));
+    }
+    // Canvas end - region threading
 
     public static final record PendingTeleport(Entity.EntityTreeNode rootVehicle, Vec3 to) {}
     private final it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<PendingTeleport> pendingTeleports = new it.unimi.dsi.fastutil.objects.ReferenceOpenHashSet<>();
diff --git a/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java b/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
index b6a7038445b7d856b2718d124860b4d626ccace9..c0aa26f228a0db304b9d210eda8bae17c7c18ce2 100644
--- a/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerConfigurationPacketListenerImpl.java
@@ -90,6 +90,13 @@ public class ServerConfigurationPacketListenerImpl extends ServerCommonPacketLis
         if (this.prepareSpawnTask != null) {
             this.prepareSpawnTask.close();
             this.prepareSpawnTask = null;
+            // Canvas start - region threading
+            // ensure that if we disconnect for some reason, we clear us from the join list
+            // this covers if exceptions are thrown during join
+            for (final net.minecraft.server.level.ServerLevel level : net.minecraft.server.MinecraftServer.getServer().getAllLevels()) {
+                level.canvas$decrementJoiningPlayers(gameProfile.name());
+            }
+            // Canvas end - region threading
         }
 
         super.onDisconnect(details);
diff --git a/net/minecraft/server/network/config/PrepareSpawnTask.java b/net/minecraft/server/network/config/PrepareSpawnTask.java
index 475751ccb4dfcd1f2022b9883ce5cb7a30841f88..93ac90ccfbd5815b471e909dbca139582f4a4089 100644
--- a/net/minecraft/server/network/config/PrepareSpawnTask.java
+++ b/net/minecraft/server/network/config/PrepareSpawnTask.java
@@ -96,15 +96,31 @@ public class PrepareSpawnTask implements ConfigurationTask {
                 spawnDataLevel = this.server.overworld();
             }
             ServerLevel spawnLevel;
+            boolean obtainedWorldStageLock = false; // Canvas - region threading
             if (resourceKey == null) {
                 spawnLevel = spawnDataLevel;
             } else {
                 spawnLevel = this.server.getLevel(resourceKey);
-                if (spawnLevel == null) {
+                // Canvas start - region threading
+                if (spawnLevel != null) {
+                    spawnLevel.canvas$obtainWorldStageLock();
+                    obtainedWorldStageLock = true;
+                }
+                // if it is null, the dimension doesn't exist
+                if (spawnLevel == null || spawnLevel.canvas$unloadTicket.isPresent()) {
+                    if (spawnLevel != null) {
+                        // if we reach here, we are unloading
+                        spawnLevel.canvas$releaseWorldStageLock();
+                        obtainedWorldStageLock = false;
+                    }
+                    // the overworld cannot be unloaded because of the bukkit api rule
+                // Canvas end - region threading
                     LOGGER.warn("Unknown respawn dimension {}, defaulting to overworld", resourceKey);
                     spawnLevel = spawnDataLevel;
                 }
             }
+            spawnLevel.canvas$incrementJoiningPlayers(nameAndId.name()); // Canvas - region threading
+            if (obtainedWorldStageLock) spawnLevel.canvas$releaseWorldStageLock(); // Canvas - region threading
             // Folia start - region threading
             CompletableFuture<Vec3> spawnPosition = new java.util.concurrent.CompletableFuture<>();
             if (loadedPosition.position().isPresent()) {
diff --git a/net/minecraft/server/players/PlayerList.java b/net/minecraft/server/players/PlayerList.java
index 1475868b15c6ef98436f2d913ca1cfd09534c33b..cfdb4a8471a80dbe4f9d860fbc910e1c830ab4f3 100644
--- a/net/minecraft/server/players/PlayerList.java
+++ b/net/minecraft/server/players/PlayerList.java
@@ -299,6 +299,7 @@ public abstract class PlayerList {
         player.suppressTrackerForLogin = true;
         this.sendLevelInfo(player, level);
         level.addNewPlayer(player);
+        level.canvas$decrementJoiningPlayers(gameProfile.name()); // Canvas - region threading - we are added to player list by now, meaning it will be caught by unload preconditions
         this.server.getCustomBossEvents().onPlayerConnect(player); // see commented out section below serverLevel.addPlayerJoin(player);
         // Paper end - Fire PlayerJoinEvent when Player is actually ready
         player.initInventoryMenu();
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index cff9c19709e669f80a1ceeaa1414047747cfb373..699d4aad8a84c63943aae50200290df4cf7d90ae 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -397,6 +397,7 @@ public abstract class Entity
     }
     // Paper end
     // Canvas start - region threading
+    public io.canvasmc.canvas.util.@Nullable ServerLocation canvas$lastTeleportOrigin = null;
 
     // Canvas end - region threading
     // Paper start - rewrite chunk system
@@ -4357,6 +4358,12 @@ public abstract class Entity
             return false;
         }
         this.getBukkitEntity(); // force bukkit entity to be created before TPing
+
+        if (destination.canvas$unloadTicket.isPresent()) {
+            // destination world unload propagated, do not teleport
+            return false;
+        }
+
         if ((teleportFlags & TELEPORT_FLAG_UNMOUNT) == 0L) {
             for (Entity entity : this.getIndirectPassengers()) {
                 if (!entity.canTeleportAsync()) {
@@ -4420,6 +4427,7 @@ public abstract class Entity
         EntityTreeNode passengerTree = this.detachPassengers();
         List<EntityTreeNode> fullPassengerTree = passengerTree.getFullTree();
         ServerLevel originWorld = (ServerLevel)this.level;
+        this.canvas$lastTeleportOrigin = new io.canvasmc.canvas.util.ServerLocation(originWorld, currPos, this.getYRot(), this.getXRot()); // Canvas - region threading
 
         for (EntityTreeNode node : fullPassengerTree) {
             node.root.preChangeDimension();
@@ -4685,6 +4693,7 @@ public abstract class Entity
             }
         }
 
+        if (to != null && to.canvas$unloadTicket.isPresent()) return false; // Canvas - region threading
         return true;
     }
 
@@ -4704,6 +4713,7 @@ public abstract class Entity
             ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkX(initialPosition),
             ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkZ(initialPosition)
         );
+        this.canvas$lastTeleportOrigin = new io.canvasmc.canvas.util.ServerLocation(originWorld, initialPosition, this.getYRot(), this.getXRot()); // Canvas - region threading
 
         // first, remove entity/passengers from world
         EntityTreeNode passengerTree = this.detachPassengers();
