From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Mon, 20 Apr 2026 14:15:44 -0700
Subject: [PATCH] Fix respawn anchor charges not being processed


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 534aea2ab6f58c74acd813fae626b65d59efd175..29b2faee0d59dedf6d0159f84c8462fc269a24e2 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1637,7 +1637,11 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
             Runnable consumeAnchorCharge = null; // Paper - Fix SPIGOT-5989 (don't use charge until after respawn event)
             Optional<Vec3> standUpPosition = RespawnAnchorBlock.findStandUpPosition(EntityType.PLAYER, level, pos);
             if (!forced && consumeSpawnBlock && standUpPosition.isPresent()) {
-                consumeAnchorCharge = () -> level.setBlock(pos, blockState.setValue(RespawnAnchorBlock.CHARGE, blockState.getValue(RespawnAnchorBlock.CHARGE) - 1), Block.UPDATE_ALL); // Paper - Fix SPIGOT-5989 (don't use charge until after respawn event)
+                // Canvas start - region threading
+                consumeAnchorCharge = () -> level.canvas$loadOrRunAtChunksAsync(pos, 1, ca.spottedleaf.concurrentutil.util.Priority.NORMAL, () -> {
+                    level.setBlock(pos, blockState.setValue(net.minecraft.world.level.block.RespawnAnchorBlock.CHARGE, blockState.getValue(net.minecraft.world.level.block.RespawnAnchorBlock.CHARGE) - 1), net.minecraft.world.level.block.Block.UPDATE_ALL); // Paper - Fix SPIGOT-5989 (don't use charge until after respawn event)
+                });
+                // Canvas end - region threading
             }
             final Runnable finalConsumeAnchorCharge = consumeAnchorCharge; // Paper - Fix SPIGOT-5989
 
@@ -1727,6 +1731,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         ca.spottedleaf.concurrentutil.completable.CallbackCompletable<org.bukkit.Location> spawnPosComplete =
             new ca.spottedleaf.concurrentutil.completable.CallbackCompletable<>();
         boolean[] usedRespawnAnchor = new boolean[1];
+        RespawnPosAngle[] respawnPosAngle = new RespawnPosAngle[1]; // Canvas - region threading
 
         // set up post spawn location logic
         spawnPosComplete.addWaiter((spawnLoc, throwable) -> {
@@ -1755,6 +1760,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                                 1.0F, 1.0F, ServerPlayer.this.level().getRandom().nextLong()
                             )
                         );
+                        if (respawnPosAngle[0] != null && respawnPosAngle[0].consumeAnchorCharge != null) respawnPosAngle[0].consumeAnchorCharge.run(); // Canvas - region threading
                     }
                     ServerPlayer.this.connection.restartClientLoadTimerAfterRespawn();
                     // now the respawn logic is complete
@@ -1797,6 +1803,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                     boolean isRespawnAnchor = respawnWorld.getBlockState(respawnConfig.respawnData.pos()).is(net.minecraft.world.level.block.Blocks.RESPAWN_ANCHOR);
                     boolean isBed = respawnWorld.getBlockState(respawnConfig.respawnData.pos()).is(net.minecraft.tags.BlockTags.BEDS);
                     usedRespawnAnchor[0] = !alive && isRespawnAnchor;
+                    respawnPosAngle[0] = spawnPos; // Canvas - region threading
 
                     // finished now, pass the location on
                     spawnPosComplete.complete(
