From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Thu, 9 Apr 2026 11:32:22 -0700
Subject: [PATCH] Fix respawn anchor charges not being processed


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 34b23f7be46ec74ee7321f0111cf58c1fd1300c3..437d63fdee3dfd317eab5ad9238e592716f0fa66 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1608,7 +1608,11 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
             Optional<Vec3> optional = RespawnAnchorBlock.findStandUpPosition(EntityType.PLAYER, level, blockPos);
             Runnable consumeAnchorCharge = null; // Paper - Fix SPIGOT-5989 (don't use charge until after respawn event)
             if (!flag && useCharge && optional.isPresent()) {
-                consumeAnchorCharge = () -> level.setBlock(blockPos, blockState.setValue(RespawnAnchorBlock.CHARGE, blockState.getValue(RespawnAnchorBlock.CHARGE) - 1), Block.UPDATE_ALL); // Paper - Fix SPIGOT-5989 (don't use charge until after respawn event)
+                // Canvas start - region threading
+                consumeAnchorCharge = () -> level.canvas$loadOrRunAtChunksAsync(blockPos, 1, ca.spottedleaf.concurrentutil.util.Priority.NORMAL, () -> {
+                    level.setBlock(blockPos, blockState.setValue(net.minecraft.world.level.block.RespawnAnchorBlock.CHARGE, blockState.getValue(net.minecraft.world.level.block.RespawnAnchorBlock.CHARGE) - 1), net.minecraft.world.level.block.Block.UPDATE_ALL); // Paper - Fix SPIGOT-5989 (don't use charge until after respawn event)
+                });
+                // Canvas end - region threading
             }
             final Runnable finalConsumeAnchorCharge = consumeAnchorCharge; // Paper - Fix SPIGOT-5989
 
@@ -1726,6 +1730,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                                 1.0F, 1.0F, ServerPlayer.this.level().getRandom().nextLong()
                             )
                         );
+                        if (respawnPosAngle[0] != null && respawnPosAngle[0].consumeAnchorCharge != null) respawnPosAngle[0].consumeAnchorCharge.run(); // Canvas - region threading
                     }
                     ServerPlayer.this.connection.restartClientLoadTimerAfterRespawn();
                     // now the respawn logic is complete
