From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Thu, 9 Apr 2026 11:32:22 -0700
Subject: [PATCH] Fix respawn anchor charges not being processed


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 07a3b57dc63095c8a5f818725b58cd551d6d2a77..187eae2106b8c91c4b165be5ab46df42f18daa8c 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1611,7 +1611,11 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
             Optional<Vec3> optional = RespawnAnchorBlock.findStandUpPosition(EntityType.PLAYER, level, blockPos);
             Runnable consumeAnchorCharge = null; // Paper - Fix SPIGOT-5989 (don't use charge until after respawn event)
             if (!flag && useCharge && optional.isPresent()) {
-                consumeAnchorCharge = () -> level.setBlock(blockPos, blockState.setValue(RespawnAnchorBlock.CHARGE, blockState.getValue(RespawnAnchorBlock.CHARGE) - 1), Block.UPDATE_ALL); // Paper - Fix SPIGOT-5989 (don't use charge until after respawn event)
+                // Canvas start - region threading
+                consumeAnchorCharge = () -> level.canvas$loadOrRunAtChunksAsync(blockPos, 1, ca.spottedleaf.concurrentutil.util.Priority.NORMAL, () -> {
+                    level.setBlock(blockPos, blockState.setValue(net.minecraft.world.level.block.RespawnAnchorBlock.CHARGE, blockState.getValue(net.minecraft.world.level.block.RespawnAnchorBlock.CHARGE) - 1), net.minecraft.world.level.block.Block.UPDATE_ALL); // Paper - Fix SPIGOT-5989 (don't use charge until after respawn event)
+                });
+                // Canvas end - region threading
             }
             final Runnable finalConsumeAnchorCharge = consumeAnchorCharge; // Paper - Fix SPIGOT-5989
 
@@ -1701,6 +1705,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
         ca.spottedleaf.concurrentutil.completable.CallbackCompletable<org.bukkit.Location> spawnPosComplete =
             new ca.spottedleaf.concurrentutil.completable.CallbackCompletable<>();
         boolean[] usedRespawnAnchor = new boolean[1];
+        RespawnPosAngle[] respawnPosAngle = new RespawnPosAngle[1]; // Canvas - region threading
 
         // set up post spawn location logic
         spawnPosComplete.addWaiter((spawnLoc, throwable) -> {
@@ -1729,6 +1734,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                                 1.0F, 1.0F, ServerPlayer.this.level().getRandom().nextLong()
                             )
                         );
+                        if (respawnPosAngle[0] != null && respawnPosAngle[0].consumeAnchorCharge != null) respawnPosAngle[0].consumeAnchorCharge.run(); // Canvas - region threading
                     }
                     ServerPlayer.this.connection.restartClientLoadTimerAfterRespawn();
                     // now the respawn logic is complete
@@ -1771,6 +1777,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                     boolean isRespawnAnchor = respawnWorld.getBlockState(respawnConfig.respawnData.pos()).is(net.minecraft.world.level.block.Blocks.RESPAWN_ANCHOR);
                     boolean isBed = respawnWorld.getBlockState(respawnConfig.respawnData.pos()).is(net.minecraft.tags.BlockTags.BEDS);
                     usedRespawnAnchor[0] = !alive && isRespawnAnchor;
+                    respawnPosAngle[0] = spawnPos; // Canvas - region threading
 
                     // finished now, pass the location on
                     spawnPosComplete.complete(
