From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Mon, 20 Apr 2026 14:40:42 -0700
Subject: [PATCH] Fix Riding Stats


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 2703a713368db46a534e1b24886d083500c49c88..81b3e577943e3322900d723bc2b6deaaab94deaf 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -4085,6 +4085,11 @@ public abstract class Entity
         }
 
         public void adjustRiders(boolean teleport) {
+        // Canvas start - region threading - fix ride stats
+            this.adjustRiders(teleport, false);
+        }
+        public void adjustRiders(boolean teleport, boolean trackVehicleMoveStats) {
+        // Canvas end - region threading - fix ride stats
             java.util.ArrayDeque<EntityTreeNode> queue = new java.util.ArrayDeque<>();
             queue.add(this);
 
@@ -4097,14 +4102,25 @@ public abstract class Entity
 
                 for (EntityTreeNode passenger : passengers) {
                     queue.add(passenger);
+                    // Canvas start - region threading - fix ride stats
+                    // this is required to fix the riding statistics for players
+                    // this is because of how Folia does their fix for repositioning riders
+                    // that makes it so it ignores checking riding statistics
+                    double x = passenger.root.getX();
+                    double y = passenger.root.getY();
+                    double z = passenger.root.getZ();
                     curr.root.positionRider(passenger.root, teleport ? Entity::snapTo : Entity::setPos);
+                    if (trackVehicleMoveStats && passenger.root instanceof ServerPlayer player) {
+                        player.checkRidingStatistics(player.getX() - x, player.getY() - y, player.getZ() - z);
+                    }
+                    // Canvas end - region threading - fix ride stats
                 }
             }
         }
     }
 
     public void repositionAllPassengers(boolean teleport) {
-        this.makePassengerTree().adjustRiders(teleport);
+        this.makePassengerTree().adjustRiders(teleport, true); // Canvas - region threading - fix ride stats
     }
 
     protected EntityTreeNode makePassengerTree() {
