From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Mon, 20 Apr 2026 14:40:57 -0700
Subject: [PATCH] Implement Teleport Portal and Respawn Events


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 7d92db1240eb7273d3c532b02eefb07c66c8fff6..9fe9e73ce327ac135f796f3af7c375e49e158313 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1754,14 +1754,29 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
             new ca.spottedleaf.concurrentutil.completable.CallbackCompletable<>();
         boolean[] usedRespawnAnchor = new boolean[1];
         RespawnPosAngle[] respawnPosAngle = new RespawnPosAngle[1]; // Canvas - region threading
+        boolean[] isBedSpawn = new boolean[]{false}; // Canvas - region threading
+        boolean[] missingRespawnBlock = new boolean[]{false}; // Canvas - region threading
 
         // set up post spawn location logic
         spawnPosComplete.addWaiter((spawnLoc, throwable) -> {
             // update pos and velocity
-            ServerPlayer.this.setPosRaw(spawnLoc.getX(), spawnLoc.getY(), spawnLoc.getZ());
-            ServerPlayer.this.setYRot(spawnLoc.getYaw());
-            ServerPlayer.this.setYHeadRot(spawnLoc.getYaw());
-            ServerPlayer.this.setXRot(spawnLoc.getPitch());
+            // Canvas start - region threading
+            io.canvasmc.canvas.event.PlayerRespawnAsyncEvent respawnEvent = new io.canvasmc.canvas.event.PlayerRespawnAsyncEvent(
+                ServerPlayer.this.getBukkitEntity(),
+                spawnLoc,
+                isBedSpawn[0],
+                usedRespawnAnchor[0],
+                missingRespawnBlock[0],
+                reason
+            );
+            respawnEvent.callEvent();
+            java.util.concurrent.atomic.AtomicReference<org.bukkit.Location> spawnLocAtomic = new java.util.concurrent.atomic.AtomicReference<>();
+            Runnable finalize = () -> {
+            ServerPlayer.this.setPosRaw(spawnLocAtomic.get().getX(), spawnLocAtomic.get().getY(), spawnLocAtomic.get().getZ());
+            ServerPlayer.this.setYRot(spawnLocAtomic.get().getYaw());
+            ServerPlayer.this.setYHeadRot(spawnLocAtomic.get().getYaw());
+            ServerPlayer.this.setXRot(spawnLocAtomic.get().getPitch());
+            // Canvas end - region threading
             ServerPlayer.this.setDeltaMovement(Vec3.ZERO);
             // placeInAsync will update the world
 
@@ -1769,7 +1784,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                 origin,
                 // use the load chunk flag just in case the spawn loc isn't loaded, and to ensure the chunks
                 // stay loaded for a bit with the teleport ticket
-                ((org.bukkit.craftbukkit.CraftWorld)spawnLoc.getWorld()).getHandle(),
+                ((org.bukkit.craftbukkit.CraftWorld)spawnLocAtomic.get().getWorld()).getHandle(), // Canvas - region threading
                 TELEPORT_FLAG_LOAD_CHUNK | TELEPORT_FLAGS_PLAYER_RESPAWN,
                 passengerTree, // note: we expect this to just be the player, no passengers
                 (entity) -> {
@@ -1791,8 +1806,35 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                     if (respawnComplete != null) {
                         respawnComplete.accept(ServerPlayer.this);
                     }
+                    // Canvas start - region threading
+                    io.canvasmc.canvas.event.PlayerPostRespawnAsyncEvent postEvent = new io.canvasmc.canvas.event.PlayerPostRespawnAsyncEvent(
+                        ServerPlayer.this.getBukkitEntity(),
+                        spawnLoc,
+                        isBedSpawn[0],
+                        usedRespawnAnchor[0],
+                        missingRespawnBlock[0],
+                        reason
+                    );
+                    postEvent.callEvent();
+                    // Canvas end - region threading
                 }
             );
+            // Canvas start - region threading
+            };
+            spawnLocAtomic.set(respawnEvent.getRespawnLocation());
+            if (spawnLoc != respawnEvent.getRespawnLocation()) {
+                // event changed the location, load async and then call finalizer
+                // use all the events location for this, they may have changed the world
+                ServerLevel respawnAt = ((org.bukkit.craftbukkit.CraftWorld) respawnEvent.getRespawnLocation().getWorld()).getHandle();
+                respawnAt.canvas$loadOrRunAtChunksAsync(
+                    new BlockPos(respawnEvent.getRespawnLocation().getBlockX(), respawnEvent.getRespawnLocation().getBlockY(), respawnEvent.getRespawnLocation().getBlockZ()),
+                    16, ca.spottedleaf.concurrentutil.util.Priority.HIGHER, finalize
+                );
+            } else {
+                // event didn't modify location, run finalizer
+                finalize.run();
+            }
+            // Canvas end - region threading
         });
 
         // find and modify respawn block state
@@ -1810,6 +1852,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                         respawnWorld, respawnConfig, !alive
                     ).orElse(null);
                     if (spawnPos == null) {
+                        missingRespawnBlock[0] = true; // Canvas - region threading
                         // no spawn
                         ServerPlayer.this.connection.send(
                             new ClientboundGameEventPacket(ClientboundGameEventPacket.NO_RESPAWN_BLOCK_AVAILABLE, 0.0F)
@@ -1823,7 +1866,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                     }
 
                     boolean isRespawnAnchor = respawnWorld.getBlockState(respawnConfig.respawnData.pos()).is(net.minecraft.world.level.block.Blocks.RESPAWN_ANCHOR);
-                    boolean isBed = respawnWorld.getBlockState(respawnConfig.respawnData.pos()).is(net.minecraft.tags.BlockTags.BEDS);
+                    isBedSpawn[0] = respawnWorld.getBlockState(respawnConfig.respawnData.pos()).is(net.minecraft.tags.BlockTags.BEDS); // Canvas - region threading
                     usedRespawnAnchor[0] = !alive && isRespawnAnchor;
                     respawnPosAngle[0] = spawnPos; // Canvas - region threading
 
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index d4190d4309c6e35cbad700d9bf87e5c5403f9321..514093dc52472a5f3c6024c75d2e651041b11b72 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -4371,29 +4371,28 @@ public abstract class Entity
         );
     }
 
-    public final boolean teleportAsync(ServerLevel destination, Vec3 pos, Float yaw, Float pitch, Vec3 velocity,
-                                       org.bukkit.event.player.PlayerTeleportEvent.TeleportCause cause, long teleportFlags,
-                                       java.util.function.Consumer<Entity> teleportComplete) {
+    // Canvas start - region threading
+    private void canvas$ensureCanTeleportAsync(ServerLevel destination, Vec3 pos, long teleportFlags) throws TeleportValidationException {
         ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this, "Cannot teleport entity async");
 
         if (!ServerLevel.isInSpawnableBounds(new BlockPos(ca.spottedleaf.moonrise.common.util.CoordinateUtils.getBlockX(pos), ca.spottedleaf.moonrise.common.util.CoordinateUtils.getBlockY(pos), ca.spottedleaf.moonrise.common.util.CoordinateUtils.getBlockZ(pos)))) {
-            return false;
+            throw new TeleportValidationException("Destination pos not in spawnable bounds");
         }
 
         if (!this.canTeleportAsync()) {
-            return false;
+            throw new TeleportValidationException("Entity not able to teleport async. Is the entity removed or in a bed?");
         }
         this.getBukkitEntity(); // force bukkit entity to be created before TPing
 
         if (destination.canvas$unloadTicket.isPresent()) {
             // destination world unload propagated, do not teleport
-            return false;
+            throw new TeleportValidationException("Destination world is marked for unload");
         }
 
         if ((teleportFlags & TELEPORT_FLAG_UNMOUNT) == 0L) {
             for (Entity entity : this.getIndirectPassengers()) {
                 if (!entity.canTeleportAsync()) {
-                    return false;
+                    throw new TeleportValidationException("Indirect passenger not able to teleport async. Is the passenger removed or dead?");
                 }
                 entity.getBukkitEntity(); // force bukkit entity to be created before TPing
             }
@@ -4402,21 +4401,87 @@ public abstract class Entity
         }
 
         if ((teleportFlags & TELEPORT_FLAG_TELEPORT_PASSENGERS) != 0L) {
-            if (this.isPassenger()) {
-                return false;
-            }
+            if (isPassenger()) throw new TeleportValidationException("Entity is passenger");
         } else {
-            if (this.isVehicle() || this.isPassenger()) {
+            if (isVehicle() || isPassenger()) throw new TeleportValidationException("Entity is passenger or vehicle");
+        }
+    }
+
+    public static class TeleportValidationException extends IllegalStateException {
+        private final String reason;
+
+        public TeleportValidationException() {
+            this("Unknown reason");
+        }
+
+        public TeleportValidationException(String reason) {
+            this.reason = reason;
+        }
+
+        @Override
+        public String getMessage() {
+            return "Teleport validation failed. Did an event change the entity state?: " + reason;
+        }
+    }
+
+    public final boolean teleportAsync(ServerLevel destination, Vec3 pos, @Nullable Float yaw, @Nullable Float pitch, Vec3 velocity,
+                                       org.bukkit.event.player.PlayerTeleportEvent.TeleportCause cause, long teleportFlags,
+                                       java.util.function.Consumer<Entity> teleportComplete) {
+        try {
+            // we silence the catch to replicate what it is in Folia
+            canvas$ensureCanTeleportAsync(destination, pos, teleportFlags);
+        } catch (TeleportValidationException ignored) {
+            return false;
+        } // else | we passed first go
+        // TODO - all teleport events things actually go HERE
+        // when we call events, they need to allow destination modification but also need to account
+        // for the fact that the teleport passed BEFORE, and now we need to pass AGAIN to be considered valid still
+        // we check this because of this comment in Folia#68:
+        //     "The comment where I wrote that events may be placed there is actually wrong.
+        //     It also needs to check the preconditions for teleporting again, as plugins may
+        //     modify them. Ideally, given that this is "new" API, if any of those checks fail
+        //     (i.e wrong tick thread, entity removed, or any other check it made before) then
+        //     the code needs to throw an exception." - SpottedLeaf
+        // fire teleport events
+        Vec3 currPos = this.position();
+        ServerLevel originWorld = (ServerLevel) this.level();
+        final boolean isSameRegion = destination == this.level()
+            && ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(destination, ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkX(pos), ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkZ(pos), 8)
+            && (((teleportFlags & TELEPORT_FLAG_LOAD_CHUNK) == 0L) || destination.areChunksLoadedForMove(this.getBoundingBoxAt(pos.x, pos.y, pos.z)));
+        final io.canvasmc.canvas.event.EntityTeleportAsyncEvent.TeleportType teleportType =
+            isSameRegion ? io.canvasmc.canvas.event.EntityTeleportAsyncEvent.TeleportType.SAME_REGION :
+                (destination == this.level() ? io.canvasmc.canvas.event.EntityTeleportAsyncEvent.TeleportType.CROSS_REGION :
+                 io.canvasmc.canvas.event.EntityTeleportAsyncEvent.TeleportType.CROSS_WORLD);
+        final boolean shouldCallEvents =
+            // this is purely so we don't do unneeded event calls and we don't run checks twice for no reason
+            // TODO - any other events should be added here too, that way we can properly verify validation of the event POST
+            io.canvasmc.canvas.event.EntityTeleportAsyncEvent.getHandlerList().getRegisteredListeners().length > 0;
+        if (shouldCallEvents) {
+            // we have listeners, call
+            org.bukkit.Location bukkitFrom = new org.bukkit.Location(originWorld.getWorld(), currPos.x, currPos.y, currPos.z, this.getYRot(), this.getXRot());
+            org.bukkit.Location bukkitTo = new org.bukkit.Location(destination.getWorld(), pos.x, pos.y, pos.z, yaw == null ? this.getYRot() : yaw, pitch == null ? this.getXRot() : pitch);
+
+            io.canvasmc.canvas.event.EntityTeleportAsyncEvent teleportAsyncEvent =
+                new io.canvasmc.canvas.event.EntityTeleportAsyncEvent(this.getBukkitEntity(), bukkitFrom, bukkitTo, cause, teleportType);
+            if (!teleportAsyncEvent.callEvent()) {
+                // event was cancelled
                 return false;
             }
-        }
 
-        // TODO any events that can modify go HERE
+            // set the new teleport location
+            org.bukkit.Location resultLocation = teleportAsyncEvent.getTo();
+            pos = new Vec3(teleportAsyncEvent.getTo().x(), teleportAsyncEvent.getTo().y(), teleportAsyncEvent.getTo().z());
+            destination = resultLocation.getWorld() == null ? destination : ((org.bukkit.craftbukkit.CraftWorld) resultLocation.getWorld()).getHandle();
+            yaw = resultLocation.getYaw();
+            pitch = resultLocation.getPitch();
+
+            // if plugin did something funky, kill
+            canvas$ensureCanTeleportAsync(destination, pos, teleportFlags);
+        }
+    // Canvas end - region threading
 
         // check for same region
-        if (destination == this.level()
-                && ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(destination, ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkX(pos), ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkZ(pos), 8)
-                && (((teleportFlags & TELEPORT_FLAG_LOAD_CHUNK) == 0L) || destination.areChunksLoadedForMove(this.getBoundingBoxAt(pos.x, pos.y, pos.z)))) {
+        if (isSameRegion) { // Canvas - region threading
             boolean hasPassengers = !this.passengers.isEmpty();
             EntityTreeNode passengerTree = this.detachPassengers();
 
@@ -4449,12 +4514,21 @@ public abstract class Entity
             if (teleportComplete != null) {
                 teleportComplete.accept(this);
             }
+            // Canvas start - region threading
+            if (io.canvasmc.canvas.event.EntityPostTeleportAsyncEvent.getHandlerList().getRegisteredListeners().length > 0) {
+                org.bukkit.Location bukkitFrom = new org.bukkit.Location(originWorld.getWorld(), currPos.x, currPos.y, currPos.z, this.getYRot(), this.getXRot());
+                org.bukkit.Location bukkitTo = new org.bukkit.Location(destination.getWorld(), pos.x, pos.y, pos.z, yaw == null ? this.getYRot() : yaw, pitch == null ? this.getXRot() : pitch);
+                new io.canvasmc.canvas.event.EntityPostTeleportAsyncEvent(
+                    this.getBukkitEntity(), bukkitFrom, bukkitTo, cause, teleportType
+                ).callEvent();
+            }
+            // Canvas end - region threading
             return true;
         }
 
         EntityTreeNode passengerTree = this.detachPassengers();
         List<EntityTreeNode> fullPassengerTree = passengerTree.getFullTree();
-        ServerLevel originWorld = (ServerLevel)this.level;
+        // Canvas - region threading
         this.canvas$lastTeleportOrigin = new io.canvasmc.canvas.util.ServerLocation(originWorld, currPos, this.getYRot(), this.getXRot()); // Canvas - region threading
 
         for (EntityTreeNode node : fullPassengerTree) {
@@ -4465,7 +4539,22 @@ public abstract class Entity
             node.root = node.root.transformForAsyncTeleport(destination, pos, yaw, pitch, velocity);
         }
 
-        passengerTree.root.placeInAsync(originWorld, destination, teleportFlags, passengerTree, teleportComplete);
+        // Canvas start - region threading
+        final ServerLevel finalLevelDest = destination;
+        final Vec3 finalPos = pos;
+        final Float finalYaw = yaw;
+        final Float finalPitch = pitch;
+        passengerTree.root.placeInAsync(originWorld, destination, teleportFlags, passengerTree, (entity) -> {
+            if (teleportComplete != null) teleportComplete.accept(entity);
+            if (io.canvasmc.canvas.event.EntityPostTeleportAsyncEvent.getHandlerList().getRegisteredListeners().length > 0) {
+                org.bukkit.Location bukkitFrom = new org.bukkit.Location(originWorld.getWorld(), currPos.x, currPos.y, currPos.z, this.getYRot(), this.getXRot());
+                org.bukkit.Location bukkitTo = new org.bukkit.Location(finalLevelDest.getWorld(), finalPos.x, finalPos.y, finalPos.z, finalYaw == null ? 0 : finalYaw, finalPitch == null ? 0 : finalPitch);
+                new io.canvasmc.canvas.event.EntityPostTeleportAsyncEvent(
+                    this.getBukkitEntity(), bukkitFrom, bukkitTo, cause, teleportType
+                ).callEvent();
+            }
+        });
+        // Canvas end - region threading
 
         return true;
     }
@@ -4695,35 +4784,46 @@ public abstract class Entity
         }
     }
 
+    // Canvas start - region threading
+    // required for ServerPlayer end portal async logic
     public boolean canPortalAsync(ServerLevel to, boolean considerPassengers) {
-        return this.canPortalAsync(to, considerPassengers, false);
+        try {
+            canvas$ensureCanPortalAsync(to, considerPassengers);
+            return true;
+        } catch (TeleportValidationException e) {
+            return false;
+        }
     }
 
-    protected boolean canPortalAsync(ServerLevel to, boolean considerPassengers, boolean skipPassengerCheck) {
+    public void canvas$ensureCanPortalAsync(ServerLevel to, boolean considerPassengers) throws TeleportValidationException {
+        this.canvas$ensureCanPortalAsync(to, considerPassengers, false);
+    }
+
+    protected void canvas$ensureCanPortalAsync(ServerLevel to, boolean considerPassengers, boolean skipPassengerCheck) throws TeleportValidationException {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this, "Cannot portal entity async");
         if (considerPassengers) {
             if (!skipPassengerCheck && this.isPassenger()) {
-                return false;
+                throw new TeleportValidationException("Entity is passenger");
             }
         } else {
             if (this.isVehicle() || (!skipPassengerCheck && this.isPassenger())) {
-                return false;
+                throw new TeleportValidationException("Entity is passenger or vehicle");
             }
         }
         this.getBukkitEntity(); // force bukkit entity to be created before TPing
         if (!this.canTeleportAsync()) {
-            return false;
+            throw new TeleportValidationException("Entity not able to teleport async. Is the entity removed or in a bed?");
         }
         if (considerPassengers) {
             for (Entity entity : this.passengers) {
-                if (!entity.canPortalAsync(to, true, true)) {
-                    return false;
-                }
+                entity.canvas$ensureCanPortalAsync(to, true, true);
             }
         }
-
-        if (to != null && to.canvas$unloadTicket.isPresent()) return false; // Canvas - region threading
-        return true;
+        if (to != null && to.canvas$unloadTicket.isPresent()) {
+            throw new TeleportValidationException("Destination world is marked for unload");
+        }
     }
+    // Canvas end - region threading
 
     protected void prePortalLogic(ServerLevel origin, ServerLevel destination, PortalType type) {
 
@@ -4731,10 +4831,41 @@ public abstract class Entity
 
     protected boolean portalToAsync(ServerLevel destination, BlockPos portalPos, boolean takePassengers,
                                     PortalType type, java.util.function.Consumer<Entity> teleportComplete) {
-        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this, "Cannot portal entity async");
-        if (!this.canPortalAsync(destination, takePassengers)) {
+        // Canvas start - region threading
+        try {
+            canvas$ensureCanPortalAsync(destination, takePassengers);
+        } catch (TeleportValidationException ignored) {
             return false;
+        } // else | we passed first go
+
+        // TODO - all portal events go here
+        // fire portal events
+        ServerLevel originWorld = (ServerLevel) this.level();
+
+        boolean shouldCallEvents =
+            // this is purely so we don't do unneeded event calls and we don't run checks twice for no reason
+            // TODO - any other events should be added here too, that way we can properly verify validation of the event POST
+            io.canvasmc.canvas.event.EntityPortalAsyncEvent.getHandlerList().getRegisteredListeners().length > 0;
+        if (shouldCallEvents) {
+            io.canvasmc.canvas.event.EntityPortalAsyncEvent portalAsyncEvent = new io.canvasmc.canvas.event.EntityPortalAsyncEvent(
+                this.getBukkitEntity(), originWorld.getWorld(), destination.getWorld(),
+                switch (type) {
+                    case END -> org.bukkit.PortalType.ENDER;
+                    case NETHER -> org.bukkit.PortalType.NETHER;
+                }
+            );
+
+            if (!portalAsyncEvent.callEvent()) {
+                return false;
+            }
+
+            destination = ((org.bukkit.craftbukkit.CraftWorld) portalAsyncEvent.getTo()).getHandle();
+
+            // if a plugin did a stupid, kill
+            canvas$ensureCanPortalAsync(destination, takePassengers);
         }
+        final ServerLevel finalDestination = destination;
+        // Canvas end - region threading
 
         Vec3 initialPosition = this.position();
         ChunkPos initialPositionChunk = new ChunkPos(
@@ -4746,7 +4877,7 @@ public abstract class Entity
         // first, remove entity/passengers from world
         EntityTreeNode passengerTree = this.detachPassengers();
         List<EntityTreeNode> fullPassengerTree = passengerTree.getFullTree();
-        ServerLevel originWorld = (ServerLevel)this.level;
+        // Canvas - region threading
 
         for (EntityTreeNode node : fullPassengerTree) {
             node.root.preChangeDimension();
@@ -4793,7 +4924,7 @@ public abstract class Entity
 
             // place
             passengerTree.root.placeInAsync(
-                originWorld, destination, Entity.TELEPORT_FLAG_LOAD_CHUNK | (takePassengers ? Entity.TELEPORT_FLAG_TELEPORT_PASSENGERS : 0L),
+                originWorld, finalDestination, Entity.TELEPORT_FLAG_LOAD_CHUNK | (takePassengers ? Entity.TELEPORT_FLAG_TELEPORT_PASSENGERS : 0L), // Canvas - region threading
                 passengerTree,
                 (Entity teleported) -> {
                     if (info.postTeleportTransition() != null) {
@@ -4803,6 +4934,17 @@ public abstract class Entity
                     if (teleportComplete != null) {
                         teleportComplete.accept(teleported);
                     }
+                    // Canvas start - region threading
+                    if (io.canvasmc.canvas.event.EntityPostPortalAsyncEvent.getHandlerList().getRegisteredListeners().length > 0) {
+                        new io.canvasmc.canvas.event.EntityPostPortalAsyncEvent(
+                            this.getBukkitEntity(), originWorld.getWorld(), finalDestination.getWorld(),
+                            switch (type) {
+                                case END -> org.bukkit.PortalType.ENDER;
+                                case NETHER -> org.bukkit.PortalType.NETHER;
+                            }
+                        ).callEvent();
+                    }
+                    // Canvas end - region threading
                 }
             );
         });
