From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Mon, 20 Apr 2026 14:41:25 -0700
Subject: [PATCH] Fix End Credits


diff --git a/io/papermc/paper/threadedregions/RegionizedServer.java b/io/papermc/paper/threadedregions/RegionizedServer.java
index 6771355d1bd4b759232a75b656e7d56745719147..ee70bdd32927fbbce5cb6975df31df41f3fa8edb 100644
--- a/io/papermc/paper/threadedregions/RegionizedServer.java
+++ b/io/papermc/paper/threadedregions/RegionizedServer.java
@@ -404,9 +404,14 @@ public final class RegionizedServer {
             }
 
             if (isNotOwnedByGlobalRegion(conn)) {
-                // we actually require that the owning regions remove the connection for us, as it is possible
-                // that ownership is transferred back to us
-                continue;
+                // Canvas start - region threading
+                if (!(conn.getPacketListener() instanceof ServerGamePacketListenerImpl gamePacketListener
+                    && gamePacketListener.player.canvas$isDisplayingEndCredits)) {
+                    // we actually require that the owning regions remove the connection for us,
+                    // as it is possible that ownership is transferred back to us
+                    continue;
+                }
+                // Canvas end - region threading
             }
 
             if (!conn.isConnected()) {
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 678b452eff0c312edffd0d8f52eb70f8fcea2cc9..a015526498702351c089773ee662926bb5fdb44d 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -271,6 +271,8 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     // Canvas start - region threading
     public final it.unimi.dsi.fastutil.objects.Object2ObjectMap<net.minecraft.world.waypoints.WaypointTransmitter, net.minecraft.world.waypoints.WaypointTransmitter.Connection>
         canvas$activeWaypoints = new it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<>();
+    public boolean canvas$isDisplayingEndCredits = false;
+    private @Nullable Runnable canvas$exitEndCreditsCallback = null;
     // Canvas end - region threading
     private final Set<ThrownEnderpearl> enderPearls = java.util.concurrent.ConcurrentHashMap.newKeySet(); // Canvas - region threading
     private long timeEntitySatOnShoulder;
@@ -1705,8 +1707,17 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
             // not in the end credits anymore
             return;
         }
-        this.wonGame = false;
+        // Canvas start - region threading
+        if (!this.canvas$isDisplayingEndCredits) this.wonGame = false;
 
+        if (this.canvas$exitEndCreditsCallback != null) {
+            this.wonGame = false;
+            this.canvas$isDisplayingEndCredits = false;
+            this.canvas$exitEndCreditsCallback.run();
+            this.canvas$exitEndCreditsCallback = null;
+            return;
+        }
+        // Canvas end - region threading
         this.respawn((player) -> {
             CriteriaTriggers.CHANGED_DIMENSION.trigger(player, Level.END, Level.OVERWORLD);
         }, org.bukkit.event.player.PlayerRespawnEvent.RespawnReason.END_PORTAL, true);
@@ -1837,6 +1848,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
             // Canvas end - region threading
         });
 
+        Runnable finalizer = () -> { // Canvas - region threading
         // find and modify respawn block state
         if (respawnWorld == null || respawnConfig == null) {
             // default to regular spawn
@@ -1878,6 +1890,16 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                 }
             );
         }
+        // Canvas start - region threading
+        };
+        if (this.canvas$isDisplayingEndCredits) {
+            // we need to assign to callback, we are displaying the credits
+            this.canvas$exitEndCreditsCallback = finalizer;
+        } else {
+            // not displaying credits, run finalizer
+            finalizer.run();
+        }
+        // Canvas end - region threading
     }
 
     @Override
@@ -2038,12 +2060,22 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                 return false;
             }
             this.wonGame = true;
-            // TODO is there a better solution to this that DOESN'T skip the credits?
-            this.seenCredits = true;
-            if (!this.seenCredits) {
+            // Canvas start - region threading
+            // we need to remove the player *first*, and then display
+            // the end credits, or else we get the player stuck in the
+            // void, and they can't escape...
+            this.canvas$isDisplayingEndCredits = !this.seenCredits && !this.level().paperConfig().misc.disableEndCredits; // ensure we respect paper config
+            io.canvasmc.canvas.event.PlayerViewEndCreditsEvent viewEndCreditsEvent = new io.canvasmc.canvas.event.PlayerViewEndCreditsEvent(
+                getBukkitEntity(), this.canvas$isDisplayingEndCredits
+            );
+            viewEndCreditsEvent.callEvent();
+            this.canvas$isDisplayingEndCredits = viewEndCreditsEvent.willShowCredits();
+            this.exitEndCredits(); // note: this calls 'unRide()'
+            if (this.canvas$isDisplayingEndCredits) {
                 this.connection.send(new ClientboundGameEventPacket(ClientboundGameEventPacket.WIN_GAME, 0.0F));
+                this.seenCredits = true; // mark seen the end credits
             }
-            this.exitEndCredits();
+            // Canvas end - region threading
             return true;
         } else {
             return super.endPortalLogicAsync(portalPos);
