From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dueris <duerismc@gmail.com>
Date: Mon, 4 May 2026 18:45:25 -0700
Subject: [PATCH] Region Tick Guards


diff --git a/net/minecraft/server/level/ChunkMap.java b/net/minecraft/server/level/ChunkMap.java
index 6fcf0ee02caf94164c1d9a6e0fe2f813a8cc73a7..97c1f739253f986688178dbabc13bc114c6962bb 100644
--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -956,7 +956,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
     }
 
     public void addEntity(final Entity entity) {
-        org.spigotmc.AsyncCatcher.catchOp("entity track"); // Spigot
+        io.canvasmc.canvas.util.TickGuard.guard(entity, "Cannot add entity to tracking async"); // Canvas - region guards
         // Paper start - ignore and warn about illegal addEntity calls instead of crashing server
         if (!entity.valid || entity.level() != this.level || entity.moonrise$getTrackedEntity() != null) { // Folia - region threading
             LOGGER.error("Illegal ChunkMap::addEntity for world " + io.papermc.paper.util.MCUtil.getLevelName(this.level)
@@ -1004,7 +1004,7 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
     }
 
     protected void removeEntity(final Entity entity) {
-        org.spigotmc.AsyncCatcher.catchOp("entity untrack"); // Spigot
+        io.canvasmc.canvas.util.TickGuard.guard(entity, "Cannot remove entity from tracking async"); // Canvas - region guards
         if (entity instanceof ServerPlayer player) {
             this.updatePlayerStatus(player, false);
 
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index ae4c386eba2407bedb25861870bb3f55019321bc..c4cb32e8dfe18b41570fdd4c84a85d338bc8b68b 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1773,7 +1773,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
 
     // CraftBukkit start
     private boolean addEntity(final Entity entity, final org.bukkit.event.entity.CreatureSpawnEvent.@Nullable SpawnReason spawnReason) {
-        org.spigotmc.AsyncCatcher.catchOp("entity add"); // Spigot
+        io.canvasmc.canvas.util.TickGuard.guard(entity, "Cannot add entity async"); // Canvas - region guards
         entity.generation = false; // Paper - Don't fire sync event during generation; Reset flag if it was added during a ServerLevel generation process
         // Paper start - extra debug info
         if (entity.valid) {
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 0b936e53ea6ce50ccd9f8fe23b875d35a814a1ad..8d9ebacb10f2c1510a50d0eced89eb0c642f7710 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -891,6 +891,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     }
 
     public void initMenu(final AbstractContainerMenu container) {
+        io.canvasmc.canvas.util.TickGuard.guard(this, "Cannot init menu async"); // Canvas - region guards
         container.addSlotListener(this.containerListener);
         container.setSynchronizer(this.containerSynchronizer);
     }
@@ -2445,6 +2446,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
 
     @Override
     public OptionalInt openMenu(final @Nullable MenuProvider provider) {
+        io.canvasmc.canvas.util.TickGuard.guard(this, "Cannot open menu async"); // Canvas - region guards
         if (provider == null) {
             return OptionalInt.empty();
         } else {
@@ -2583,6 +2585,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     }
     @Override
     public void closeContainer(org.bukkit.event.inventory.InventoryCloseEvent.Reason reason) {
+        io.canvasmc.canvas.util.TickGuard.guard(this, "Cannot close container async"); // Canvas - region guards
         org.bukkit.craftbukkit.event.CraftEventFactory.handleInventoryCloseEvent(this, reason); // CraftBukkit
         // Paper end - Inventory close reason
         this.connection.send(new ClientboundContainerClosePacket(this.containerMenu.containerId));
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index ba8f83d1f74fb17d15f732a5282b89e3452dc270..27bbe51f490d4248fda5347ba21b1ecdaca8265f 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -1831,6 +1831,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     }
 
     public void die(final DamageSource source) {
+        io.canvasmc.canvas.util.TickGuard.guard(this, "Cannot die async"); // Canvas - region guards
         if (!this.isRemoved() && !this.dead) {
             Entity sourceEntity = source.getEntity();
             LivingEntity killer = this.getKillCredit();
diff --git a/net/minecraft/world/level/block/state/BlockBehaviour.java b/net/minecraft/world/level/block/state/BlockBehaviour.java
index 42918460ae0fffdde1e7fab5baec6a3662b10be7..6c17f84e69a2a60d8a7170a9e16c27f356288ee9 100644
--- a/net/minecraft/world/level/block/state/BlockBehaviour.java
+++ b/net/minecraft/world/level/block/state/BlockBehaviour.java
@@ -170,7 +170,7 @@ public abstract class BlockBehaviour implements FeatureElement {
     }
 
     protected void onPlace(final BlockState state, final Level level, final BlockPos pos, final BlockState oldState, final boolean movedByPiston) {
-        org.spigotmc.AsyncCatcher.catchOp("block onPlace"); // Spigot
+        io.canvasmc.canvas.util.TickGuard.guard(pos, level, "Cannot place block async"); // Canvas - region guards
     }
 
     protected void affectNeighborsAfterRemoval(final BlockState state, final ServerLevel level, final BlockPos pos, final boolean movedByPiston) {
