From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/npc/villager/Villager.java b/net/minecraft/world/entity/npc/villager/Villager.java
index d5360e2a0ad491dea6d30accdfbe55d797a2ddcb..2c5be1b48cddfb70322eb6d00ffdb6b38a1077e3 100644
--- a/net/minecraft/world/entity/npc/villager/Villager.java
+++ b/net/minecraft/world/entity/npc/villager/Villager.java
@@ -851,7 +851,15 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
 
     private int countFoodPointsInInventory() {
         SimpleContainer inventory = this.getInventory();
-        return FOOD_POINTS.entrySet().stream().mapToInt(entry -> inventory.countItem(entry.getKey()) * entry.getValue()).sum();
+        // Canvas start - remove streams
+        int sum = 0;
+        for (Map.Entry<Item, Integer> entry : FOOD_POINTS.entrySet()) {
+            int i = inventory.countItem(entry.getKey()) * entry.getValue();
+            sum += i;
+        }
+        return sum;
+        // return FOOD_POINTS.entrySet().stream().mapToInt(entry -> inventory.countItem(entry.getKey()) * entry.getValue()).sum();
+        // Canvas end - remove streams
     }
 
     public boolean hasFarmSeeds() {
