--- a/net/minecraft/world/level/levelgen/synth/BlendedNoise.java
+++ b/net/minecraft/world/level/levelgen/synth/BlendedNoise.java
@@ -154,6 +_,39 @@
         return this.maxValue;
     }
 
+    // Leaf start - C2ME - DensityFunction compiler
+    @Override
+    public boolean equals(final Object object) {
+        if (this == object) return true;
+        if (object == null || this.getClass() != object.getClass()) return false;
+        BlendedNoise that = (BlendedNoise)object;
+        return Double.compare(this.xzMultiplier, that.xzMultiplier) == 0
+            && Double.compare(this.yMultiplier, that.yMultiplier) == 0
+            && Double.compare(this.xzFactor, that.xzFactor) == 0
+            && Double.compare(this.yFactor, that.yFactor) == 0
+            && Double.compare(this.smearScaleMultiplier, that.smearScaleMultiplier) == 0
+            && java.util.Objects.equals(this.minLimitNoise, that.minLimitNoise)
+            && java.util.Objects.equals(this.maxLimitNoise, that.maxLimitNoise)
+            && java.util.Objects.equals(this.mainNoise, that.mainNoise);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = 1;
+
+        result = 31 * result + java.util.Objects.hashCode(this.minLimitNoise);
+        result = 31 * result + java.util.Objects.hashCode(this.maxLimitNoise);
+        result = 31 * result + java.util.Objects.hashCode(this.mainNoise);
+        result = 31 * result + Double.hashCode(this.xzMultiplier);
+        result = 31 * result + Double.hashCode(this.yMultiplier);
+        result = 31 * result + Double.hashCode(this.xzFactor);
+        result = 31 * result + Double.hashCode(this.yFactor);
+        result = 31 * result + Double.hashCode(this.smearScaleMultiplier);
+
+        return result;
+    }
+    // Leaf end - C2ME - DensityFunction compiler
+
     @VisibleForTesting
     public void parityConfigString(final StringBuilder sb) {
         sb.append("BlendedNoise{minLimitNoise=");
