--- a/net/minecraft/world/level/levelgen/synth/ImprovedNoise.java
+++ b/net/minecraft/world/level/levelgen/synth/ImprovedNoise.java
@@ -6,7 +_,7 @@
 
 public final class ImprovedNoise {
     private static final float SHIFT_UP_EPSILON = 1.0E-7F;
-    private final byte[] p;
+    public final byte[] p; // Canvas start - private -> public
     public final double xo;
     public final double yo;
     public final double zo;
