--- a/net/minecraft/world/level/levelgen/synth/NormalNoise.java
+++ b/net/minecraft/world/level/levelgen/synth/NormalNoise.java
@@ -83,6 +_,29 @@
         return this.parameters;
     }
 
+    // Leaf start - C2ME - DensityFunction compiler
+    @Override
+    public boolean equals(final Object object) {
+        if (this == object) return true;
+        if (object == null || this.getClass() != object.getClass()) return false;
+        NormalNoise that = (NormalNoise)object;
+        return Double.compare(this.valueFactor, that.valueFactor) == 0
+            && java.util.Objects.equals(this.first, that.first)
+            && java.util.Objects.equals(this.second, that.second);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = 1;
+
+        result = 31 * result + Double.hashCode(this.valueFactor);
+        result = 31 * result + java.util.Objects.hashCode(this.first);
+        result = 31 * result + java.util.Objects.hashCode(this.second);
+
+        return result;
+    }
+    // Leaf end - C2ME - DensityFunction compiler
+
     @VisibleForTesting
     public void parityConfigString(final StringBuilder sb) {
         sb.append("NormalNoise {");
