--- a/net/minecraft/world/level/levelgen/synth/NormalNoise.java
+++ b/net/minecraft/world/level/levelgen/synth/NormalNoise.java
@@ -3,9 +_,13 @@
 import com.google.common.annotations.VisibleForTesting;
 import com.mojang.serialization.Codec;
 import com.mojang.serialization.codecs.RecordCodecBuilder;
+import io.canvasmc.canvas.Config;
+import io.canvasmc.canvas.util.BindingsTemplate;
 import it.unimi.dsi.fastutil.doubles.DoubleArrayList;
 import it.unimi.dsi.fastutil.doubles.DoubleList;
 import it.unimi.dsi.fastutil.doubles.DoubleListIterator;
+import java.lang.foreign.Arena;
+import java.lang.foreign.MemorySegment;
 import java.util.List;
 import net.minecraft.Util;
 import net.minecraft.core.Holder;
@@ -21,6 +_,15 @@
     private final PerlinNoise second;
     private final double maxValue;
     private final NormalNoise.NoiseParameters parameters;
+    // Canvas start
+    private final Arena c2me$arena = Arena.ofAuto();
+    private MemorySegment c2me$samplerData = null;
+    private long c2me$samplerDataPtr;
+
+    public long c2me$getPointer() {
+        return this.c2me$samplerDataPtr;
+    }
+    // Canvas end
 
     @Deprecated
     public static NormalNoise createLegacyNetherBiome(RandomSource random, NormalNoise.NoiseParameters parameters) {
@@ -62,6 +_,12 @@
 
         this.valueFactor = 0.16666666666666666 / expectedDeviation(i2 - i1);
         this.maxValue = (this.first.maxValue() + this.second.maxValue()) * this.valueFactor;
+        // Canvas start
+        if (Config.INSTANCE.chunks.nativeAccelerationEnabled) {
+            this.c2me$samplerData = BindingsTemplate.double_octave_sampler_data$create(this.c2me$arena, this.first, this.second, this.valueFactor);
+            this.c2me$samplerDataPtr = this.c2me$samplerData.address();
+        }
+        // Canvas end
     }
 
     public double maxValue() {
