--- a/net/minecraft/world/level/levelgen/synth/PerlinNoise.java
+++ b/net/minecraft/world/level/levelgen/synth/PerlinNoise.java
@@ -20,12 +_,14 @@
 
 public class PerlinNoise {
     private static final int ROUND_OFF = 33554432;
-    private final ImprovedNoise[] noiseLevels;
-    private final int firstOctave;
-    private final DoubleList amplitudes;
-    private final double lowestFreqValueFactor;
-    private final double lowestFreqInputFactor;
-    private final double maxValue;
+    // Canvas start - AT
+    public final ImprovedNoise[] noiseLevels;
+    public final int firstOctave;
+    public final DoubleList amplitudes;
+    public final double lowestFreqValueFactor;
+    public final double lowestFreqInputFactor;
+    public final double maxValue;
+    // Canvas end
 
     @Deprecated
     public static PerlinNoise createLegacyForBlendedNoise(RandomSource random, IntStream octaves) {
