--- a/net/minecraft/world/level/levelgen/synth/SimplexNoise.java
+++ b/net/minecraft/world/level/levelgen/synth/SimplexNoise.java
@@ -191,4 +_,29 @@
         double n3 = this.getCornerNoise3D(gi3, x3, y3, z3, 0.6);
         return 32.0 * (n0 + n1 + n2 + n3);
     }
+
+    // Leaf start - C2ME - DensityFunction compiler
+    @Override
+    public boolean equals(final Object object) {
+        if (this == object) return true;
+        if (object == null || this.getClass() != object.getClass()) return false;
+        SimplexNoise that = (SimplexNoise) object;
+        return Double.compare(this.xo, that.xo) == 0
+            && Double.compare(this.yo, that.yo) == 0
+            && Double.compare(this.zo, that.zo) == 0
+            && java.util.Arrays.equals(this.p, that.p);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = 1;
+
+        result = 31 * result + java.util.Arrays.hashCode(this.p);
+        result = 31 * result + Double.hashCode(this.xo);
+        result = 31 * result + Double.hashCode(this.yo);
+        result = 31 * result + Double.hashCode(this.zo);
+
+        return result;
+    }
+    // Leaf end - C2ME - DensityFunction compiler
 }
