--- a/net/minecraft/world/level/levelgen/synth/SimplexNoise.java
+++ b/net/minecraft/world/level/levelgen/synth/SimplexNoise.java
@@ -25,7 +_,7 @@
     private static final double SQRT_3 = Math.sqrt(3.0);
     private static final double F2 = 0.5 * (SQRT_3 - 1.0);
     private static final double G2 = (3.0 - SQRT_3) / 6.0;
-    private final int[] p = new int[512];
+    public final int[] p = new int[512]; // Canvas - private -> public
     public final double xo;
     public final double yo;
     public final double zo;
