From fb2fa76bcfb03ba2831bd257c68fc51821e86ead Mon Sep 17 00:00:00 2001
From: File <noreply+automated@papermc.io>
Date: Sun, 20 Apr 1997 15:37:42 +0200
Subject: [PATCH] canvas File Patches


diff --git a/net/minecraft/world/entity/animal/fish/AbstractFish.java b/net/minecraft/world/entity/animal/fish/AbstractFish.java
index 7b4215058325812e8dc785277d1ece0fd5dd6ea3..9e991c4b3ae705f3979477b775c536f2c4b4a020 100644
--- a/net/minecraft/world/entity/animal/fish/AbstractFish.java
+++ b/net/minecraft/world/entity/animal/fish/AbstractFish.java
@@ -170,7 +170,7 @@ public abstract class AbstractFish extends WaterAnimal implements Bucketable {
 
         @Override
         public void tick() {
-            if (this.fish.isEyeInFluid(FluidTags.WATER)) {
+            if (this.fish.canvas$isEyeInWater()) { // Canvas - optimize fluidOnEyes
                 this.fish.setDeltaMovement(this.fish.getDeltaMovement().add(0.0, 0.005, 0.0));
             }
 
