From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 1229ed65 AH <1229ed65b2ccd3@crankymonkey.info>
Date: Fri, 26 Jan 2024 17:54:53 +0000
Subject: [PATCH] LevelResource api


diff --git a/src/main/java/io/github/dueris/canvas/util/LevelResource.java b/src/main/java/io/github/dueris/canvas/util/LevelResource.java
new file mode 100644
index 0000000000000000000000000000000000000000..db68e8fefa3bf6d38787373f6eafc40794cc00b5
--- /dev/null
+++ b/src/main/java/io/github/dueris/canvas/util/LevelResource.java
@@ -0,0 +1,39 @@
+package io.github.dueris.canvas.util;
+
+import java.nio.file.*;
+import java.io.*;
+
+public interface LevelResource {
+    
+    public Path datapackDir();
+    
+    public Path advancementsDir();
+    
+    public Path playerStatsDir();
+    
+    public Path playerdataDir();
+    
+    public Path legacyPlayerdataDir();
+    
+    public File leveldataFile();
+    
+    public File leveldataOldFile();
+    
+    public File iconFile();
+    
+    public File lockFile();
+    
+    public Path generatedDir();
+    
+    public File mapResourceFile();
+    
+    public Path mainWorldRoot();
+    
+    public Path serverRoot();
+    
+    public File serverPropertiesFile();
+    
+    public Path logsDir();
+    
+    public Path cacheDir();
+}
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index fa5019254efc2611b80fe647c15cfd48c59c774a..4e2a36f5daf170e8e54be89490a5ce45b1e7d9c4 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -198,6 +198,11 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
      */
     public ServerProperties getServerProperties();
 
+    /**
+     * Gets the LevelResource Constants from Canvas
+     */
+    public io.github.dueris.canvas.util.LevelResource getLevelResources();
+
     public TimeManager getTimeManager();
     // Canvas end
 
